diff --git a/piwik.install b/piwik.install
index a5ccb08..59d2d1e 100644
--- a/piwik.install
+++ b/piwik.install
@@ -68,15 +68,6 @@ function piwik_requirements($phase) {
 
   switch ($phase) {
     case 'runtime' :
-      // Module cannot validate piwik URL without external HTTP requests.
-      if (variable_get('drupal_http_request_fails', TRUE) && !system_check_http_request()) {
-        $requirements['http requests'] = array(
-          'title' => $t('HTTP request status'),
-          'value' => $t('Fails'),
-          'severity' => REQUIREMENT_ERROR,
-          'description' => $t('Your system or network configuration does not allow Drupal to access web pages, resulting in reduced functionality. This could be due to your webserver configuration or PHP settings, and should be resolved in order to download information about available updates, fetch aggregator feeds, sign in via OpenID, or use other network-dependent services.'),
-        );
-      }
 
       // Raise warning if Piwik user account has not been set yet.
       if (!preg_match('/^\d{1,}$/', variable_get('piwik_site_id', ''))) {
