// @lrnwebcomponents/haxcms-elements/lib/ui-components/site/site-search.js
define(["exports", "require", "../../../../../lit/index.js", "../../../../simple-fields/lib/simple-fields-field.js", "../../../../simple-icon/simple-icon.js", "../../../../simple-icon/lib/simple-icons.js", "../../../../iframe-loader/lib/loading-indicator.js"], function (_exports, _require, _index, _simpleFieldsField, _simpleIcon, _simpleIcons, _loadingIndicator) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SiteSearch = void 0;
  _require = _interopRequireWildcard(_require);

  var _templateObject_63cd1fd0fac911ebb072cd67fd432833, _templateObject2_63cd1fd0fac911ebb072cd67fd432833, _templateObject3_63cd1fd0fac911ebb072cd67fd432833, _templateObject4_63cd1fd0fac911ebb072cd67fd432833, _templateObject5_63cd1fd0fac911ebb072cd67fd432833;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `site-search`
   * `Searching HAXcms content using the auto-generated lunr search configuration`
   *
   * @demo demo/index.html
   */
  var SiteSearch = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(SiteSearch, _LitElement);

    var _super = _createSuper(SiteSearch);

    function SiteSearch() {
      var _this;

      babelHelpers.classCallCheck(this, SiteSearch);
      _this = _super.call(this);
      _this.whileLoading = false;
      _this.hideInput = false;
      _this.search = "";
      _this.showPath = false;
      _this.showDate = false;
      _this.__results = [];
      setTimeout(function () {
        new Promise(function (res, rej) {
          return _require.default(["../../../../lunr-search/lunr-search.js"], res, rej);
        });
        new Promise(function (res, rej) {
          return _require.default(["../../../../simple-datetime/simple-datetime.js"], res, rej);
        });
      }, 0);
      return _this;
    } // render function


    babelHelpers.createClass(SiteSearch, [{
      key: "render",
      value: function render() {
        var _this2 = this;

        return (0, _index.html)(_templateObject_63cd1fd0fac911ebb072cd67fd432833 || (_templateObject_63cd1fd0fac911ebb072cd67fd432833 = babelHelpers.taggedTemplateLiteral(["\n      <simple-fields-field\n        ?hidden=\"", "\"\n        id=\"search\"\n        always-float-label\n        label=\"Search\"\n        placeholder=\"Type at least 3 letters to start search..\"\n        type=\"text\"\n        @value-changed=\"", "\"\n      >\n        <simple-icon icon=\"search\" slot=\"prefix\"></simple-icon>\n      </simple-fields-field>\n      ", "\n      <lunr-search id=\"lunr\"></lunr-search>\n      <loading-indicator\n        full\n        ?loading=\"", "\"\n      ></loading-indicator>\n      ", "\n    "])), this.hideInput, this._searchValueChanged, this.search.length > 0 ? (0, _index.html)(_templateObject2_63cd1fd0fac911ebb072cd67fd432833 || (_templateObject2_63cd1fd0fac911ebb072cd67fd432833 = babelHelpers.taggedTemplateLiteral(["\n            <h1 class=\"results-found-text\">\n              Found ", " results.\n            </h1>\n          "])), this.__results.length) : (0, _index.html)(_templateObject3_63cd1fd0fac911ebb072cd67fd432833 || (_templateObject3_63cd1fd0fac911ebb072cd67fd432833 = babelHelpers.taggedTemplateLiteral([""]))), this.whileLoading, this.__results.map(function (item) {
          return (0, _index.html)(_templateObject4_63cd1fd0fac911ebb072cd67fd432833 || (_templateObject4_63cd1fd0fac911ebb072cd67fd432833 = babelHelpers.taggedTemplateLiteral(["\n          <a\n            class=\"result\"\n            .href=\"", "\"\n            @click=\"", "\"\n          >\n            <div class=\"title\">\n              ", "<span\n                ?hidden=\"", "\"\n                class=\"link-text\"\n                aria-hidden=\"true\"\n                >(", ")</span\n              >\n            </div>\n            <div class=\"date\" ?hidden=\"", "\">\n              <simple-datetime format=\"M jS\" .timestamp=\"", "\" unix\n                >", "</simple-datetime\n              >\n            </div>\n            <p>", "..</p>\n          </a>\n        "])), item.location, _this2.selectionMade, item.title, !_this2.showPath, item.location, !_this2.showDate, item.created, item.created, item.description);
        }));
      }
    }, {
      key: "selectionMade",
      value: function selectionMade(e) {
        this.dispatchEvent(new CustomEvent("search-item-selected", {
          detail: {
            value: e.detail
          }
        })); // hide modal if it's there

        window.dispatchEvent(new CustomEvent("simple-modal-hide", {
          bubbles: true,
          cancelable: true,
          detail: {}
        }));
      }
    }, {
      key: "_searchValueChanged",
      value: function _searchValueChanged(e) {
        this.search = e.detail.value;
      }
    }, {
      key: "__resultsChanged",
      value: function __resultsChanged(e) {
        var _this3 = this;

        if (e.detail.value) {
          setTimeout(function () {
            _this3.whileLoading = false;
          }, 100);
          this.__results = babelHelpers.toConsumableArray(e.detail.value);
        } else {
          this.__results = [];
        }
      }
      /**
       * Mix in an opened status
       */

    }, {
      key: "firstUpdated",
      value:
      /**
       * LitElement life cycle - ready callback
       */
      function firstUpdated(changedProperties) {
        this.shadowRoot.querySelector("#lunr").addEventListener("results-changed", this.__resultsChanged.bind(this));
      }
      /**
       * LitElement life cycle - properties changed callback
       */

    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this4 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "search" && _this4[propName]) {
            _this4._searchChanged(_this4[propName], oldValue);

            _this4.shadowRoot.querySelector("#lunr").search = _this4[propName];
          }

          if (propName == "dataSource" && _this4[propName]) {
            _this4.shadowRoot.querySelector("#lunr").dataSource = _this4[propName];
          }
        });
      }
      /**
       * Notice search term changed and let's fire up some results
       */

    }, {
      key: "_searchChanged",
      value: function _searchChanged(term, oldTerm) {
        if (term.length >= 3) {
          this.whileLoading = true; // only load up the lunr source data once they have 3 or more characters

          if (babelHelpers.typeof(this.dataSource) === (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
            this.dataSource = "lunrSearchIndex.json";
          }
        }
      }
    }], [{
      key: "styles",
      get:
      /**
       * LitElement constructable styles enhancement
       */
      function get() {
        return [(0, _index.css)(_templateObject5_63cd1fd0fac911ebb072cd67fd432833 || (_templateObject5_63cd1fd0fac911ebb072cd67fd432833 = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          font-size: 16px;\n          color: var(--site-search-color, #383f45);\n        }\n        [hidden] {\n          display: none;\n        }\n        .result {\n          display: block;\n          background-color: var(--site-search-result-background-color, #eeeeee);\n          color: var(--site-search-result-color, #222222);\n          margin-bottom: 1.5rem;\n          padding-bottom: 0.5rem;\n        }\n        .result:hover,\n        .result:focus {\n          background-color: var(\n            --site-search-result-background-color-hover,\n            #dddddd\n          );\n          color: var(--site-search-link-color-hover, #000000);\n          text-decoration: none;\n          outline: 2px solid grey;\n          outline-offset: 4px;\n        }\n        .result .title {\n          font-size: 28px;\n          margin: 0 0 8px 0;\n          line-height: 1;\n        }\n        .result {\n          color: var(--site-search-link-color, #444444);\n          text-decoration: none;\n        }\n        simple-datetime {\n          color: var(--site-search-link-color, #444444);\n        }\n        .result .link-text {\n          font-size: 12px;\n          color: var(--site-search-link-text-color, #999999);\n          font-style: italic;\n          padding-left: 8px;\n        }\n        .results-found-text {\n          margin-bottom: 1.5rem;\n          padding-bottom: 0.5rem;\n        }\n        #search {\n          flex-grow: 2;\n          margin-right: 4px;\n          --simple-fields-accent-color: var(--site-search-text-color, #000);\n          color: var(--site-search-placeholder-color, #222);\n        }\n      "])))];
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       */

    }, {
      key: "tag",
      get: function get() {
        return "site-search";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          dataSource: {
            type: String,
            attribute: "data-source"
          },
          whileLoading: {
            type: Boolean
          },
          showDate: {
            type: Boolean,
            attribute: "show-date"
          },
          showPath: {
            type: Boolean,
            attribute: "show-path"
          },
          hideInput: {
            type: Boolean,
            attribute: "hide-input"
          },
          search: {
            type: String
          },
          __results: {
            type: Array
          }
        };
      }
    }]);
    return SiteSearch;
  }(_index.LitElement);

  _exports.SiteSearch = SiteSearch;
  window.customElements.define(SiteSearch.tag, SiteSearch);
});
