// @lrnwebcomponents/haxcms-elements/lib/ui-components/site/site-title.js
define(["exports", "require", "../../../../../lit-element/lit-element.js", "../../core/haxcms-site-store.js", "../../../../../mobx/lib/mobx.module.js", "../../../../../@polymer/polymer/lib/elements/custom-style.js"], function (_exports, _require, _litElement, _haxcmsSiteStore, _mobxModule, _customStyle) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SiteTitle = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject4_575f2ac0019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral([" <h1>", "</h1> "]);

    _templateObject4_575f2ac0019d11eb9027c17497ee62cd = function _templateObject4_575f2ac0019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _templateObject3_575f2ac0019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral([""]);

    _templateObject3_575f2ac0019d11eb9027c17497ee62cd = function _templateObject3_575f2ac0019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _templateObject2_575f2ac0019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <custom-style>\n        <style>\n          a {\n            @apply --site-title-link;\n          }\n          a:hover,\n          a:focus,\n          a:active {\n            @apply --site-title-link-hover;\n          }\n          a h1 {\n            text-rendering: optimizelegibility;\n            @apply --site-title-heading;\n          }\n          iron-icon {\n            @apply --site-title-icon;\n          }\n        </style>\n      </custom-style>\n      <a\n        href=\"", "\"\n        title=\"", "\"\n        ?disabled=\"", "\"\n      >\n        <iron-icon\n          ?hidden=\"", "\"\n          icon=\"", "\"\n        ></iron-icon>\n        ", "\n      </a>\n    "]);

    _templateObject2_575f2ac0019d11eb9027c17497ee62cd = function _templateObject2_575f2ac0019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _templateObject_575f2ac0019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          text-rendering: optimizelegibility;\n          position: relative;\n        }\n      "]);

    _templateObject_575f2ac0019d11eb9027c17497ee62cd = function _templateObject_575f2ac0019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `site-title`
   * `Title of the site`
   *
  
   * @demo demo/index.html
   */
  var SiteTitle = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(SiteTitle, _LitElement);

    var _super = _createSuper(SiteTitle);

    babelHelpers.createClass(SiteTitle, null, [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject_575f2ac0019d11eb9027c17497ee62cd())];
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       */

    }, {
      key: "tag",
      get: function get() {
        return "site-title";
      }
    }]);

    function SiteTitle() {
      var _this;

      babelHelpers.classCallCheck(this, SiteTitle);
      _this = _super.call(this);
      _this.__disposer = [];
      _this.label = "Home";
      _this.notitle = false;
      new Promise(function (res, rej) {
        return _require.default(["../../../../../@polymer/iron-icon/iron-icon.js"], res, rej);
      });
      (0, _mobxModule.autorun)(function (reaction) {
        _this.siteTitle = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.siteTitle);

        _this.__disposer.push(reaction);
      });
      (0, _mobxModule.autorun)(function (reaction) {
        _this.homeLink = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.homeLink);

        _this.__disposer.push(reaction);
      });
      return _this;
    }
    /**
     * LitElement
     */


    babelHelpers.createClass(SiteTitle, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject2_575f2ac0019d11eb9027c17497ee62cd(), this.homeLink, this.label, this.disabled, this.icon ? false : true, this.icon, this.notitle ? (0, _litElement.html)(_templateObject3_575f2ac0019d11eb9027c17497ee62cd()) : (0, _litElement.html)(_templateObject4_575f2ac0019d11eb9027c17497ee62cd(), this.siteTitle));
      }
      /**
       * Props
       */

    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        for (var i in this.__disposer) {
          this.__disposer[i].dispose();
        }

        babelHelpers.get(babelHelpers.getPrototypeOf(SiteTitle.prototype), "disconnectedCallback", this).call(this);
      }
    }], [{
      key: "properties",
      get: function get() {
        return {
          disabled: {
            type: Boolean,
            reflect: true
          },

          /**
           * Site title
           */
          siteTitle: {
            type: String,
            attribute: "site-title"
          },

          /**
           * HREF to the home page
           */
          homeLink: {
            type: String,
            attribute: "home-link"
          },

          /**
           * Label
           */
          label: {
            type: String
          },

          /**
           * optional icon
           */
          icon: {
            type: String
          },

          /**
           * If the title should be displayed or not
           */
          notitle: {
            type: Boolean,
            reflect: true
          }
        };
      }
    }]);
    return SiteTitle;
  }(_litElement.LitElement);

  _exports.SiteTitle = SiteTitle;
  window.customElements.define(SiteTitle.tag, SiteTitle);
});
