// @lrnwebcomponents/lrndesign-paperstack/lrndesign-paperstack.js
define(["exports", "../../@polymer/polymer/polymer-element.js", "../simple-colors/lib/simple-colors-polymer.js", "../a11y-behaviors/a11y-behaviors.js", "../simple-icon/simple-icon.js", "../simple-icon/lib/simple-icons.js", "../hax-iconset/lib/simple-hax-iconset.js"], function (_exports, _polymerElement, _simpleColorsPolymer, _a11yBehaviors, _simpleIcon, _simpleIcons, _simpleHaxIconset) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrndesignPaperstack = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject_75c0ddd029dc11eb8473978bc22b9cf7() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style include=\"simple-colors-shared-styles-polymer\">\n        :host {\n          display: block;\n          font-size: 14px;\n          --lrndesign-paperstack-bg: var(--simple-colors-default-theme-grey-1);\n          --lrndesign-paperstack-faded-bg: var(\n            --simple-colors-default-theme-grey-2\n          );\n          --lrndesign-paperstack-border: var(\n            --simple-colors-default-theme-grey-4\n          );\n          --lrndesign-paperstack-shadow: var(\n            --simple-colors-default-theme-grey-12\n          );\n          --lrndesign-paperstack-accent: var(\n            --simple-colors-default-theme-accent-8\n          );\n          --lrndesign-paperstack-text: var(\n            --simple-colors-default-theme-grey-12\n          );\n          --lrndesign-paperstack-heading-font: cursive;\n        }\n        :host([dark]) {\n          --lrndesign-paperstack-bg: var(\n            --simple-colors-default-theme-accent-4\n          );\n          --lrndesign-paperstack-faded-bg: var(\n            --simple-colors-default-theme-accent-3\n          );\n          --lrndesign-paperstack-border: var(\n            --simple-colors-default-theme-accent-1\n          );\n          --lrndesign-paperstack-shadow: var(\n            --simple-colors-default-theme-grey-1\n          );\n          --lrndesign-paperstack-accent: var(\n            --simple-colors-default-theme-grey-12\n          );\n          --lrndesign-paperstack-text: var(\n            --simple-colors-default-theme-grey-12\n          );\n        }\n        .stack {\n          width: 100%;\n          position: relative;\n          display: flex;\n          flex-direction: column;\n          align-items: stretch;\n        }\n        .paper {\n          min-height: 160px;\n          width: 80%;\n          padding: 12px 24px;\n          position: absolute;\n          flex: 1 1 auto;\n          left: 0;\n          top: 0;\n          box-shadow: 0 0 1px var(--lrndesign-paperstack-shadow);\n          border: 1px solid var(--lrndesign-paperstack-border);\n        }\n        .paper:first-of-type {\n          left: -0.32px;\n          top: 0.32px;\n          transform: rotate(-2.5deg);\n          background-color: var(--lrndesign-paperstack-faded-bg);\n        }\n        .paper:nth-of-type(2) {\n          right: -0.32px;\n          top: 1.6px;\n          transform: rotate(1.4deg);\n          background-color: var(--lrndesign-paperstack-faded-bg);\n        }\n        .front {\n          flex: 0 0 100%;\n          position: relative;\n          margin-bottom: 48px;\n          background-color: var(--lrndesign-paperstack-bg);\n          color: var(--lrndesign-paperstack-text);\n        }\n\n        simple-icon {\n          --simple-icon-width: 40px;\n          --simple-icon-height: 40px;\n          margin: 4px;\n          color: var(--lrndesign-paperstack-bg);\n        }\n        .heading {\n          display: flex;\n          align-items: center;\n        }\n        .icon-container {\n          width: 48px;\n          height: 48px;\n          border-radius: 50%;\n          background-color: var(--lrndesign-paperstack-accent);\n          margin-right: 8px;\n          display: flex;\n          align-items: center;\n        }\n        .title {\n          font-size: 16px;\n          color: var(--lrndesign-paperstack-accent);\n          font-family: var(--lrndesign-paperstack-heading-font);\n        }\n      </style>\n      <div class=\"stack\">\n        <div class=\"paper\"></div>\n        <div class=\"paper\"></div>\n        <div class=\"front paper\">\n          <div class=\"heading\">\n            <div class=\"icon-container circle\">\n              <simple-icon icon=\"[[icon]]\"></simple-icon>\n            </div>\n            <span class=\"title\">[[title]]</span>\n          </div>\n          <p><slot></slot></p>\n        </div>\n      </div>\n    "]);

    _templateObject_75c0ddd029dc11eb8473978bc22b9cf7 = function _templateObject_75c0ddd029dc11eb8473978bc22b9cf7() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `lrndesign-paperstack`
   * @element lrndesign-paperstack
   * `Stack of papers visually`
   *
   * @demo demo/index.html
   */
  var LrndesignPaperstack = /*#__PURE__*/function (_A11yBehaviors) {
    babelHelpers.inherits(LrndesignPaperstack, _A11yBehaviors);

    var _super = _createSuper(LrndesignPaperstack);

    function LrndesignPaperstack() {
      babelHelpers.classCallCheck(this, LrndesignPaperstack);
      return _super.apply(this, arguments);
    }

    babelHelpers.createClass(LrndesignPaperstack, null, [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_75c0ddd029dc11eb8473978bc22b9cf7());
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrndesign-paperstack";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(LrndesignPaperstack), "properties", this)), {}, {
          /**
           * Title
           */
          title: {
            type: String,
            value: ""
          },

          /**
           * icon
           */
          icon: {
            type: String,
            value: "lrn:assignment"
          }
        });
      }
    }, {
      key: "haxProperties",
      get: function get() {
        return {
          canScale: true,
          canPosition: true,
          canEditSource: false,
          gizmo: {
            title: "Paper stack",
            description: "A stack of papers",
            icon: "icons:content-copy",
            color: "grey",
            groups: ["Content", "Layout"],
            handles: [],
            meta: {
              author: "ELMS:LN"
            }
          },
          settings: {
            configure: [{
              property: "title",
              title: "Title",
              description: "Title of the cards",
              inputMethod: "boolean"
            }, {
              property: "accentColor",
              title: "Accent Color",
              description: "Select the accent color use",
              inputMethod: "colorpicker",
              icon: "editor:format-color-fill"
            }, {
              property: "dark",
              title: "Dark",
              description: "Use dark theme",
              inputMethod: "boolean",
              icon: "invert-colors"
            }, {
              property: "icon",
              title: "Icon",
              description: "Icon for the card",
              inputMethod: "iconpicker"
            }, {
              slot: "",
              title: "Contents",
              description: "card contents",
              inputMethod: "code-editor"
            }],
            advanced: []
          }
        };
      }
    }]);
    return LrndesignPaperstack;
  }((0, _a11yBehaviors.A11yBehaviors)(_simpleColorsPolymer.SimpleColorsPolymer));

  _exports.LrndesignPaperstack = LrndesignPaperstack;
  window.customElements.define(LrndesignPaperstack.tag, LrndesignPaperstack);
});
