// @lrnwebcomponents/json-outline-schema/lib/jos-render.js
define(["exports", "../../../lit-element/lit-element.js", "../json-outline-schema.js", "../../utils/utils.js", "../../dynamic-import-registry/dynamic-import-registry.js"], function (_exports, _litElement, _jsonOutlineSchema, _utils, _dynamicImportRegistry) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.JosRender = void 0;

  function _templateObject2_a48b24a0d10711ea97eb69ec2e44a486() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        .children {\n          display: block;\n          margin: 16px auto;\n        }\n        .children ::slotted(*) {\n          display: inline-flex;\n          margin: var(--jos-render-margin, 8px);\n          padding: var(--jos-render-padding, 8px);\n        }\n      "]);

    _templateObject2_a48b24a0d10711ea97eb69ec2e44a486 = function _templateObject2_a48b24a0d10711ea97eb69ec2e44a486() {
      return data;
    };

    return data;
  }

  function _templateObject_a48b24a0d10711ea97eb69ec2e44a486() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"children\">\n        <slot></slot>\n      </div>\n    "]);

    _templateObject_a48b24a0d10711ea97eb69ec2e44a486 = function _templateObject_a48b24a0d10711ea97eb69ec2e44a486() {
      return data;
    };

    return data;
  }

  var JosRender =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(JosRender, _LitElement);
    babelHelpers.createClass(JosRender, null, [{
      key: "tag",
      get: function get() {
        return "jos-render";
      }
    }]);

    function JosRender() {
      var _this;

      babelHelpers.classCallCheck(this, JosRender);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(JosRender).call(this));
      _this.registry = window.DynamicImportRegistry.requestAvailability();

      if (window.WCAutoloadBasePath) {
        _this.registry.basePath = window.WCAutoloadBasePath;
      }

      _this.items = [];
      return _this;
    }

    babelHelpers.createClass(JosRender, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_a48b24a0d10711ea97eb69ec2e44a486());
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(
        /*#__PURE__*/
        function () {
          var _ref = babelHelpers.asyncToGenerator(
          /*#__PURE__*/
          regeneratorRuntime.mark(function _callee(oldValue, propName) {
            var site;
            return regeneratorRuntime.wrap(function _callee$(_context) {
              while (1) {
                switch (_context.prev = _context.next) {
                  case 0:
                    if (!(propName == "source")) {
                      _context.next = 6;
                      break;
                    }

                    site = window.JSONOutlineSchema.requestAvailability(); // load source

                    _context.next = 4;
                    return site.load(_this2[propName]);

                  case 4:
                    if (!_context.sent) {
                      _context.next = 6;
                      break;
                    }

                    _this2.items = babelHelpers.toConsumableArray(site.items);

                  case 6:
                    if (propName == "map" && _this2.map.path && _this2.map.tag) {
                      // register
                      _this2.registry.register({
                        tag: _this2.map.tag,
                        path: _this2.map.path
                      }); // load the definition


                      _this2.registry.loadDefinition(_this2.map.tag);

                      if (_this2.items.length > 0) {
                        _this2.renderItems(_this2.items);
                      }
                    }

                    if (propName == "items") {
                      _this2.renderItems(_this2.items);
                    }

                  case 8:
                  case "end":
                    return _context.stop();
                }
              }
            }, _callee);
          }));

          return function (_x, _x2) {
            return _ref.apply(this, arguments);
          };
        }());
      }
      /**
       * Render any item list passed in and handle it via our map
       * Separate function to support dynamic property remapping
       */

    }, {
      key: "renderItems",
      value: function renderItems(items) {
        var _this3 = this;

        // ensure we have a map to render
        if (this.map && this.map.properties) {
          // wipe slot
          (0, _utils.wipeSlot)(this);
          var values = (0, _utils.valueMapTransform)(items, this.map.properties);
          values.forEach(function (item) {
            var n = document.createElement(_this3.map.tag);
            Object.assign(n, item);

            _this3.appendChild(n);
          });
        }
      }
    }], [{
      key: "styles",
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_a48b24a0d10711ea97eb69ec2e44a486())];
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          source: {
            type: String
          },
          map: {
            type: Object
          },
          items: {
            type: Array
          }
        };
      }
    }]);
    return JosRender;
  }(_litElement.LitElement);

  _exports.JosRender = JosRender;
  window.customElements.define(JosRender.tag, JosRender);
});
