// @lrnwebcomponents/a11y-carousel/lib/a11y-carousel-buttons.js
define(["exports", "../../../lit-element/lit-element.js", "../../radio-behaviors/radio-behaviors.js"], function (_exports, _litElement, _radioBehaviors) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.a11yCarouselButtons = void 0;

  function _templateObject2_309fd570bff711ea911389a091fca225() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: inline;\n          background-color: transparent;\n          border: 1px solid black;\n          padding: 2px 5px;\n          margin: 15px 2px;\n          text-align: center;\n        }\n        :host([hidden]) {\n          display: none !important;\n        }\n        :host([disabled]) {\n          opacity: 0.5;\n        }\n      "]);

    _templateObject2_309fd570bff711ea911389a091fca225 = function _templateObject2_309fd570bff711ea911389a091fca225() {
      return data;
    };

    return data;
  }

  function _templateObject_309fd570bff711ea911389a091fca225() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <slot></slot>\n    "]);

    _templateObject_309fd570bff711ea911389a091fca225 = function _templateObject_309fd570bff711ea911389a091fca225() {
      return data;
    };

    return data;
  }

  /**
   * `a11y-carousel-buttons`
   * button that controls a carousel
   * @demo demo/index.html
   * @element a11y-carousel-buttons
   */
  var a11yCarouselButtons =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(a11yCarouselButtons, _LitElement);
    babelHelpers.createClass(a11yCarouselButtons, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_309fd570bff711ea911389a091fca225());
      } // properties available to the custom element for data binding

    }], [{
      key: "styles",
      //styles function
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_309fd570bff711ea911389a091fca225())];
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          active: {
            type: Boolean,
            attribute: "active",
            reflect: true
          },
          buttonType: {
            type: String,
            attribute: "button-type",
            reflect: true
          },
          controls: {
            type: String,
            attribute: "controls",
            reflect: true
          },
          disabled: {
            type: Boolean,
            attribute: "disabled",
            reflect: true
          },
          hidden: {
            type: Boolean,
            attribute: "hidden",
            reflect: true
          }
        };
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */

    }, {
      key: "tag",
      get: function get() {
        return "a11y-carousel-buttons";
      }
    }]);

    function a11yCarouselButtons() {
      var _this;

      babelHelpers.classCallCheck(this, a11yCarouselButtons);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(a11yCarouselButtons).call(this));
      _this.tabindex = 0;

      _this.setAttribute("role", "button");

      _this.setAttribute("tabindex", 0);

      _this.addEventListener("click", _this._handleClick);

      _this.addEventListener("keydown", _this._handleClick);

      _this.active = false;
      _this.hidden = false;
      return _this;
    }

    babelHelpers.createClass(a11yCarouselButtons, [{
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        this.removeEventListener("click", this._handleClick);
        this.removeEventListener("keydown", this._handleClick);
        if (babelHelpers.get(babelHelpers.getPrototypeOf(a11yCarouselButtons.prototype), "disconnectedCallback", this)) babelHelpers.get(babelHelpers.getPrototypeOf(a11yCarouselButtons.prototype), "disconnectedCallback", this).call(this);
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        if (babelHelpers.get(babelHelpers.getPrototypeOf(a11yCarouselButtons.prototype), "updated", this)) babelHelpers.get(babelHelpers.getPrototypeOf(a11yCarouselButtons.prototype), "updated", this).call(this, changedProperties);
        changedProperties.forEach(function (oldValue, propName) {
          if (["controls", "active"].includes(propName)) _this2.disabled = !_this2.controls || _this2.active;
        });
      }
      /**
       * Fires when clicked.
       *
       * @event select-carousel-item
       */

    }, {
      key: "_handleClick",
      value: function _handleClick(e) {
        if (e.type !== "keydown" || e.key === " " || e.key === "Enter" || e.key === "Spacebar") {
          if (!this.disabled) {
            event.preventDefault();
            this.dispatchEvent(new CustomEvent("select-carousel-item", {
              bubbles: true,
              cancelable: true,
              composed: true,
              detail: this
            }));
          }
        }
      }
    }]);
    return a11yCarouselButtons;
  }(_litElement.LitElement);

  _exports.a11yCarouselButtons = a11yCarouselButtons;
  window.customElements.define(a11yCarouselButtons.tag, a11yCarouselButtons);
});
