// @lrnwebcomponents/lrnapp-fab-menu/lib/lrnapp-fab-speed-dial-action.js
define(["exports","../../../@polymer/polymer/polymer-element.js","../../../@polymer/paper-fab/paper-fab.js","../../../@polymer/iron-flex-layout/iron-flex-layout-classes.js","../../materializecss-styles/lib/colors.js"],function(_exports,_polymerElement,_paperFab,_ironFlexLayoutClasses,_colors){"use strict";Object.defineProperty(_exports,"__esModule",{value:!0});_exports.LrnappFabSpeedDialAction=void 0;/*
An action within a material design [Floating Action Button with Speed Dial](https://www.google.com/design/spec/components/buttons-floating-action-button.html#buttons-floating-action-button-transitions)

### Styling

Style                                                   | Description
------------------------------------------------------- | ------------
--lrnapp-fab-speed-dial-action-background                | The background color of the Floating Action Button
--lrnapp-fab-speed-dial-action-keyboard-focus-background | The background color of the Floating Action Button when focused

### Example

```html
<lrnapp-fab-speed-dial-action icon="icons:content-copy">Copy</lrnapp-fab-speed-dial-action>
```

* @demo demo/index.html
*/class LrnappFabSpeedDialAction extends _polymerElement.PolymerElement{static get tag(){return"lrnapp-fab-speed-dial-action"}static get template(){return _polymerElement.html`
      <style
        include="iron-flex iron-flex-alignment materializecss-styles-colors"
      >
        :host {
          @apply --layout-horizontal;
          @apply --layout-center;
          @apply --layout-end-justified;
          margin-top: 15px;
          margin-right: 8px;
          min-width: 270px;
        }
        .label {
          color: black;
          background: white;
          padding: 0 16px;
          border-radius: 4px;
          margin-right: 24px;
        }
        .fab {
          --lrnapp-fab-background: var(
            --lrnapp-fab-speed-dial-action-background
          );
          --lrnapp-fab-keyboard-focus-background: var(
            --lrnapp-fab-speed-dial-action-keyboard-focus-background
          );
        }
        .label,
        .fab {
          display: flex;
        }
      </style>
      <div class="flex">
        <span class="label"><slot></slot></span>
      </div>
      <paper-fab class$="fab [[color]]" icon="[[icon]]" mini></paper-fab>
    `}static get properties(){return{/**
       * Icon that is shown next to the content
       */icon:{type:String},/**
       * Color class work to apply
       */color:{type:String,value:"blue"}}}}_exports.LrnappFabSpeedDialAction=LrnappFabSpeedDialAction;window.customElements.define(LrnappFabSpeedDialAction.tag,LrnappFabSpeedDialAction)});
