// @lrnwebcomponents/lrnapp-fab-menu/lib/lrnapp-fab-speed-dial-action.js
define(["exports", "../../../lit-element/lit-element.js", "../../../@polymer/iron-flex-layout/iron-flex-layout-classes.js", "../../simple-icon/simple-icon.js", "../../simple-icon/lib/simple-icons.js", "../../simple-icon/lib/simple-icon-button.js", "../../materializecss-styles/lib/colors.js"], function (_exports, _litElement, _ironFlexLayoutClasses, _simpleIcon, _simpleIcons, _simpleIconButton, _colors) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappFabSpeedDialAction = void 0;

  /*
  An action within a material design [Floating Action Button with Speed Dial](https://www.google.com/design/spec/components/buttons-floating-action-button.html#buttons-floating-action-button-transitions)
  
  ### Styling
  
  Style                                                   | Description
  ------------------------------------------------------- | ------------
  --lrnapp-fab-speed-dial-action-background                | The background color of the Floating Action Button
  --lrnapp-fab-speed-dial-action-keyboard-focus-background | The background color of the Floating Action Button when focused
  
  ### Example
  
  ```html
  <lrnapp-fab-speed-dial-action icon="icons:content-copy">Copy</lrnapp-fab-speed-dial-action>
  ```
  
  * @demo demo/index.html
  */
  class LrnappFabSpeedDialAction extends _litElement.LitElement {
    /**
     * LitElement constructable styles enhancement
     */
    static get styles() {
      return [_colors.materialCssStyles, (0, _litElement.css)`
        :host {
          align-items: center;
          justify-content: flex-end;
          display: flex;
          margin-top: 15px;
          margin-right: 8px;
          min-width: 270px;
        }
        .label {
          color: black;
          background: white;
          padding: 0 16px;
          border-radius: 4px;
          margin-right: 24px;
        }
        .flex {
          display: inline-flex;
        }
        .label,
        .fab {
          display: flex;
        }
      `];
    }

    static get tag() {
      return "lrnapp-fab-speed-dial-action";
    }
    /**
     * HTMLElement
     */


    constructor() {
      super();
      this.color = "blue";
    }
    /**
     * LitElement render
     */


    render() {
      return (0, _litElement.html)`
      <div class="flex">
        <span class="label"><slot></slot></span>
      </div>
      <simple-icon-button
        class="fab ${this.color}"
        accent-color="${this.color}"
        icon="${this.icon}"
        mini
      ></simple-icon-button>
    `;
    }

    static get properties() {
      return {
        /**
         * Icon that is shown next to the content
         */
        icon: {
          type: String
        },

        /**
         * Color class work to apply
         */
        color: {
          type: String
        }
      };
    }

  }

  _exports.LrnappFabSpeedDialAction = LrnappFabSpeedDialAction;
  window.customElements.define(LrnappFabSpeedDialAction.tag, LrnappFabSpeedDialAction);
});
