// @lrnwebcomponents/simple-pages/simple-pages.js
define(["exports", "meta", "require", "../../lit-element/lit-element.js"], function (_exports, meta, _require, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SimplePages = void 0;
  meta = babelHelpers.interopRequireWildcard(meta);
  _require = babelHelpers.interopRequireWildcard(_require);

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject2_6b11c330545d11eba21e89082e94d7aa() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n\n        :host([hidden]) {\n          display: none;\n        }\n\n        iron-pages:not(:defined) {\n          display: none;\n        }\n      "]);

    _templateObject2_6b11c330545d11eba21e89082e94d7aa = function _templateObject2_6b11c330545d11eba21e89082e94d7aa() {
      return data;
    };

    return data;
  }

  function _templateObject_6b11c330545d11eba21e89082e94d7aa() {
    var data = babelHelpers.taggedTemplateLiteral([" <iron-pages\n      selected=\"", "\"\n      selected-attribute=\"", "\"\n      @selected-changed=\"", "\"\n    >\n      <slot></slot>\n    </iron-pages>"]);

    _templateObject_6b11c330545d11eba21e89082e94d7aa = function _templateObject_6b11c330545d11eba21e89082e94d7aa() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `simple-pages`
   * `toggle between pages`
   *
   * @microcopy - language worth noting:
   *  -
   *
   * @demo demo/index.html
   * @element simple-pages
   */
  var SimplePages = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(SimplePages, _LitElement);

    var _super = _createSuper(SimplePages);

    babelHelpers.createClass(SimplePages, [{
      key: "render",
      // render function
      value: function render() {
        return (0, _litElement.html)(_templateObject_6b11c330545d11eba21e89082e94d7aa(), this.selected, this.selectedAttribute, this._selectedChanged);
      } // properties available to the custom element for data binding

    }], [{
      key: "styles",
      //styles function
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_6b11c330545d11eba21e89082e94d7aa())];
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(SimplePages), "properties", this)), {}, {
          selected: {
            type: Number
          },
          selectedAttribute: {
            type: String,
            attribute: "selected-attribute"
          }
        });
      }
      /**
       * Convention we use
       */

    }, {
      key: "tag",
      get: function get() {
        return "simple-pages";
      }
      /**
       * HTMLElement
       */

    }]);

    function SimplePages() {
      var _this;

      babelHelpers.classCallCheck(this, SimplePages);
      _this = _super.call(this);
      _this.selected = 0;
      setTimeout(function () {
        new Promise(function (res, rej) {
          return _require.default(["../../@polymer/iron-pages/iron-pages.js"], res, rej);
        });
      }, 0);
      return _this;
    }
    /**
     * Selected changed
     */


    babelHelpers.createClass(SimplePages, [{
      key: "_selectedChanged",
      value: function _selectedChanged(e) {
        if (this.children && this.children[e.detail.value] && this.children[e.detail.value].tagName && this.children[e.detail.value].getAttribute("data-dimport")) {
          var el = this.children[e.detail.value];

          if (!window.customElements.get(el.tagName.toLowerCase())) {
            var basePath = this.pathFromUrl(decodeURIComponent(meta.url));
            new Promise(function (res, rej) {
              return _require.default(["".concat(basePath, "../../").concat(el.getAttribute("data-dimport"))], res, rej);
            }).then(function (response) {
              setTimeout(function () {
                window.dispatchEvent(new Event("resize"));
              }, 0);
            });
          }
        }
      } // simple path from a url modifier

    }, {
      key: "pathFromUrl",
      value: function pathFromUrl(url) {
        return url.substring(0, url.lastIndexOf("/") + 1);
      }
      /**
       * LitElement ready
       */

    }, {
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {}
      /**
       * LitElement life cycle - property changed
       */

    }, {
      key: "updated",
      value: function updated(changedProperties) {
        changedProperties.forEach(function (oldValue, propName) {
          /* notify example
          // notify
          if (propName == 'format') {
            this.dispatchEvent(
              new CustomEvent(`${propName}-changed`, {
                detail: {
                  value: this[propName],
                }
              })
            );
          }
          */

          /* observer example
          if (propName == 'activeNode') {
            this._activeNodeChanged(this[propName], oldValue);
          }
          */

          /* computed example
          if (['id', 'selected'].includes(propName)) {
            this.__selectedChanged(this.selected, this.id);
          }
          */
        });
      }
    }]);
    return SimplePages;
  }(_litElement.LitElement);

  _exports.SimplePages = SimplePages;
  customElements.define(SimplePages.tag, SimplePages);
});
