// @lrnwebcomponents/mtz-marked-editor/lib/mtz-marked-control-generic-wrap.js
define(["exports", "../../../@polymer/polymer/polymer-element.js", "../../../@polymer/iron-a11y-keys/iron-a11y-keys.js", "../../simple-icon/simple-icon.js", "../../simple-icon/lib/simple-icons.js", "../../simple-icon/lib/simple-icon-button.js", "./mtz-marked-control-wrapper-behavior.js", "./mtz-marked-control-behavior.js"], function (_exports, _polymerElement, _ironA11yKeys, _simpleIcon, _simpleIcons, _simpleIconButton, _mtzMarkedControlWrapperBehavior, _mtzMarkedControlBehavior) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.MtzMarkedControlGenericWrap = void 0;

  class MtzMarkedControlGenericWrap extends (0, _mtzMarkedControlBehavior.mtzMarkedControlBehavior)((0, _mtzMarkedControlWrapperBehavior.mtzMarkedControlWrapperBehaviorImpl)(_polymerElement.PolymerElement)) {
    static get template() {
      return (0, _polymerElement.html)`
      <style>
        :host {
          display: inline-block;
        }
      </style>

      <simple-icon-button
        icon="[[icon]]"
        noink="[[noink]]"
        on-click="_handleCommand"
        alt="[[title]]"
      ></simple-icon-button>

      <iron-a11y-keys
        keys="[[keys]]"
        on-keys-pressed="_handleCommand"
        target="[[__editor]]"
      ></iron-a11y-keys>
    `;
    }

    static get tag() {
      return "mtz-marked-control-generic-wrap";
    }

    static get properties() {
      return { ...super.properties,
        title: String,
        icon: String,
        keys: String,
        noink: Boolean // Pass-through

      };
    }

  }

  _exports.MtzMarkedControlGenericWrap = MtzMarkedControlGenericWrap;
  window.customElements.define(MtzMarkedControlGenericWrap.tag, MtzMarkedControlGenericWrap);
});
