// @lrnwebcomponents/mtz-marked-editor/lib/mtz-marked-control-behavior.js
define(["exports"], function (_exports) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.mtzMarkedControlBehavior = void 0;

  const mtzMarkedControlBehavior = function (SuperClass) {
    return class extends SuperClass {
      static get properties() {
        let props = {
          /* Reference to the editor. Set programatically */
          __editor: Object
        };

        if (super.properties) {
          props = Object.assign(props, super.properties);
        }

        return props;
      }
      /**
       * Registers the control with the editor to obtain the editor instance
       * @fires register-control
       */


      connectedCallback() {
        super.connectedCallback();
        this.dispatchEvent(new CustomEvent("register-control", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: true
        }));
      }

    };
  };

  _exports.mtzMarkedControlBehavior = mtzMarkedControlBehavior;
});
