// @lrnwebcomponents/mtz-marked-editor/lib/mtz-marked-control-generic-line.js
define(["exports", "../../../@polymer/polymer/polymer-element.js", "../../../@polymer/iron-a11y-keys/iron-a11y-keys.js", "../../../@polymer/paper-icon-button/paper-icon-button.js", "./mtz-marked-control-line-behavior.js", "./mtz-marked-control-behavior.js"], function (_exports, _polymerElement, _ironA11yKeys, _paperIconButton, _mtzMarkedControlLineBehavior, _mtzMarkedControlBehavior) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.MtzMarkedControlGenericLine = void 0;

  class MtzMarkedControlGenericLine extends (0, _mtzMarkedControlBehavior.mtzMarkedControlBehavior)((0, _mtzMarkedControlLineBehavior.mtzMarkedControlLineBehaviorImpl)(_polymerElement.PolymerElement)) {
    static get template() {
      return (0, _polymerElement.html)`
      <style>
        :host {
          display: inline-block;
        }
      </style>

      <paper-icon-button
        icon="[[icon]]"
        noink="[[noink]]"
        on-click="_handleCommand"
        alt="[[title]]"
      ></paper-icon-button>

      <iron-a11y-keys
        keys="[[keys]]"
        on-keys-pressed="_handleCommand"
        target="[[__editor]]"
      ></iron-a11y-keys>
    `;
    }

    static get tag() {
      return "mtz-marked-control-generic-line";
    }

    static get properties() {
      return { ...super.properties,
        title: String,
        icon: String,
        keys: String,
        noink: Boolean // Pass-through

      };
    }

  }

  _exports.MtzMarkedControlGenericLine = MtzMarkedControlGenericLine;
  window.customElements.define(MtzMarkedControlGenericLine.tag, MtzMarkedControlGenericLine);
});
