// @lrnwebcomponents/relative-heading/lib/relative-heading-state-manager.js
define(["exports", "require", "../../../lit/index.js"], function (_exports, _require, _index) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.RelativeHeadingStateManager = void 0;
  _require = _interopRequireWildcard(_require);

  var _templateObject_6b9544b0942811ecb1937969798b231a, _templateObject2_6b9544b0942811ecb1937969798b231a;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
  // register globally so we can make sure there is only one
  window.RelativeHeadingStateManager = window.RelativeHeadingStateManager || {}; // request if this exists. This helps invoke the element existing in the dom
  // as well as that there is only one of them. That way we can ensure everything
  // is rendered through the same modal

  window.RelativeHeadingStateManager.requestAvailability = function () {
    if (!window.RelativeHeadingStateManager.instance) {
      window.RelativeHeadingStateManager.instance = document.createElement("relative-heading-state-manager");
      document.body.appendChild(window.RelativeHeadingStateManager.instance);
    }

    return window.RelativeHeadingStateManager.instance;
  };
  /**
   * `relative-heading-state-manager`
   * ` A utility that determines headings relative to their parents.`
   * @element relative-heading-state-manager
   *
   * @microcopy - language worth noting:
   *  -
   *
  
   * @polymer
   */


  var RelativeHeadingStateManager = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(RelativeHeadingStateManager, _LitElement);

    var _super = _createSuper(RelativeHeadingStateManager);

    /**
     * Makes sure there is a utility ready and listening for elements.
     */
    function RelativeHeadingStateManager() {
      var _this;

      babelHelpers.classCallCheck(this, RelativeHeadingStateManager);
      _this = _super.call(this);
      _this.closeIcon = "close";
      _this.closeLabel = "Close";
      _this.copyMessage = "Copied to Clipboard";
      _this.headings = [];
      _this.copyHeading = {};
      _this.usesCopyLink = false; // sets the instance to the current instance

      if (!window.RelativeHeadingStateManager.instance) window.RelativeHeadingStateManager.instance = babelHelpers.assertThisInitialized(_this);
      return _this;
    }

    babelHelpers.createClass(RelativeHeadingStateManager, [{
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_6b9544b0942811ecb1937969798b231a || (_templateObject_6b9544b0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <simple-toast\n        id=\"relative-heading-toast\"\n        duration=\"5000\"\n        ?hidden=\"", "\"\n        ?disabled=\"", "\"\n        text=\"", ": ", "\"\n      >\n        <simple-icon-button-lite\n          icon=\"", "\"\n          label=\"", "\"\n          title=\"", "\"\n          @click=\"", "\"\n        ></simple-icon-button-lite>\n      </simple-toast>\n    "])), !this.usesCopyLink, !this.usesCopyLink, this.copyHeading.copyMessage || this.copyMessage, this.copyUrl, this.copyHeading.closeIcon || this.closeIcon, this.copyHeading.closeLabel || this.closeLabel, this.copyHeading.closeLabel || this.closeLabel, this.closeCopyLink);
      }
      /**
       * gets URL to be copied
       * @readonly
       * @returns {string}
       */

    }, {
      key: "copyUrl",
      get: function get() {
        return "".concat(window.location.href.replace(window.location.hash, ""), "#").concat(this.copyHeading && this.copyHeading.id ? this.copyHeading.id : "");
      }
      /**
       * imports toast if needed and not already loaded
       */

    }, {
      key: "useCopyLink",
      value: function useCopyLink() {
        if (!this.usesCopyLink) {
          this.usesCopyLink = true;
          new Promise(function (res, rej) {
            return _require.default(["../../simple-icon/lib/simple-icons.js"], res, rej);
          });
          new Promise(function (res, rej) {
            return _require.default(["../../simple-icon/lib/simple-icon-button-lite.js"], res, rej);
          });
          new Promise(function (res, rej) {
            return _require.default(["../../simple-toast/simple-toast.js"], res, rej);
          });
        }
      }
      /**
       * handles copying the share link
       * @param {object} active heading
       */

    }, {
      key: "copyLink",
      value: function copyLink(heading) {
        this.copyHeading = heading;
        var el = document.createElement("textarea");
        el.value = this.copyUrl;
        document.body.appendChild(el);
        el.select();
        document.execCommand("copy");
        document.body.removeChild(el);
        if (this.shadowRoot.querySelector("#relative-heading-toast") && this.shadowRoot.querySelector("#relative-heading-toast").open) this.shadowRoot.querySelector("#relative-heading-toast").open();
      }
      /**
       * handles closing link toast
       */

    }, {
      key: "closeCopyLink",
      value: function closeCopyLink() {
        if (this.shadowRoot.querySelector("#relative-heading-toast") && this.shadowRoot.querySelector("#relative-heading-toast").close) this.shadowRoot.querySelector("#relative-heading-toast").close();
      }
      /**
       * adds heading to manager data
       * @param {object} heading to be added
       */

    }, {
      key: "addHeading",
      value: function addHeading(heading) {
        if (heading) {
          this.addSubhead(heading);
          this.setHeading(heading.id, heading);
          this.updateLevel(heading);
        }
      }
      /**
       * adds heading from manager data
       * @param {object} heading to be removed
       */

    }, {
      key: "removeHeading",
      value: function removeHeading(heading) {
        var _this2 = this;

        if (heading) {
          if (heading.id && this.headings[heading.id]) {
            this.headings[heading.id].heading = null;
            this.headings[heading.id].subheads.forEach(function (subhead) {
              return _this2.updateLevel(subhead);
            });
          }

          this.removeSubhead(heading.parent, heading);
        }
      }
      /**
       * updates heading id in manager data
       * @param {object} heading to be updated
       * @param {string} old heading id
       */

    }, {
      key: "updateId",
      value: function updateId(heading) {
        var old = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;

        if (heading) {
          if (old) this.setHeading(old, null);
          this.setHeading(heading.id, heading);
        }
      }
      /**
       * updates heading parent id in manager data
       * @param {object} heading to be updated
       * @param {string} old heading parent
       */

    }, {
      key: "updateParent",
      value: function updateParent(heading) {
        var old = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;

        if (heading) {
          if (old) this.removeSubhead(old, heading);
          this.addSubhead(heading);
        }
      }
      /**
       * updates heading level based on default level
       * @param {object} heading to be updated
       */

    }, {
      key: "updateDefaultLevel",
      value: function updateDefaultLevel(heading) {
        var old = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
        if (heading) this.updateLevel(heading);
      }
      /**
       * adds heading to subhead data
       * @param {object} heading to be added
       */

    }, {
      key: "addSubhead",
      value: function addSubhead(heading) {
        if (heading) {
          if (heading.parent) {
            if (!this.headings[heading.parent]) this.headings[heading.parent] = {
              heading: null,
              subheads: []
            };
            if (!this.headings[heading.parent].subheads.includes(heading)) this.headings[heading.parent].subheads.push(heading);
          }

          this.updateLevel(heading);
        }
      }
      /**
       * removes heading from subhead data
       * @param {string} id to be updated
       * @param {object} heading to be removed
       */

    }, {
      key: "removeSubhead",
      value: function removeSubhead(id, heading) {
        if (id && this.headings[id] && this.headings[id].subheads) {
          this.headings[id].subheads = this.headings[id].subheads.filter(function (subhead) {
            return subhead != heading;
          });
        }
      }
      /**
       * sets manager's heading data
       * @param {string} id to be updated
       * @param {object} heading to be set
       */

    }, {
      key: "setHeading",
      value: function setHeading(id, heading) {
        if (id) {
          if (!this.headings[id]) this.headings[id] = {
            heading: null,
            subheads: []
          };
          this.headings[id].heading = heading;
        }
      }
      /**
       * updates heading level
       * @param {object} heading to be updated
       */

    }, {
      key: "updateLevel",
      value: function updateLevel(heading) {
        var _this3 = this;

        if (heading) {
          var parent = heading.parent && this.headings[heading.parent].heading ? parseInt(this.headings[heading.parent].heading.__level) : null,
              plevel = parent ? Math.min(6, Math.max(parent + 1, 1)) : null,
              level = plevel ? plevel : heading.defaultLevel;

          heading._setLevel(level);

          if (this.headings[heading.id]) this.headings[heading.id].subheads.forEach(function (subhead) {
            return _this3.updateLevel(subhead);
          });
        }
      }
    }], [{
      key: "tag",
      get:
      /**
       * Store the tag name to make it easier to obtain directly.
       */
      function get() {
        return "relative-heading-state-manager";
      } // properties available to the custom element for data binding

    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(RelativeHeadingStateManager), "properties", this)), {}, {
          /**
           * icon for copy link's toast's close button
           */
          closeIcon: {
            type: String
          },

          /**
           * label for copy link's toast's close button
           */
          closeLabel: {
            type: String
          },

          /**
           * active heading for copying link
           */
          copyHeading: {
            type: Object
          },

          /**
           * message for copy link's toast
           */
          copyMessage: {
            type: String
          },

          /**
           * Stores an array of all the players on the page.
           */
          headings: {
            type: Object
          },

          /**
           * indicates is toast is already imported for copy link feature
           */
          usesCopyLink: {
            type: Boolean
          }
        });
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _index.css)(_templateObject2_6b9544b0942811ecb1937969798b231a || (_templateObject2_6b9544b0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        simple-toast:not(:defined) {\n          display: none;\n        }\n      "])))];
      }
    }]);
    return RelativeHeadingStateManager;
  }(_index.LitElement);

  _exports.RelativeHeadingStateManager = RelativeHeadingStateManager;
  window.customElements.define(RelativeHeadingStateManager.tag, RelativeHeadingStateManager);
});
