// @lrnwebcomponents/b-r/lib/star-rating.js
import{html as t,css as s,LitElement as e}from"../../../lit/index.js";import"../../simple-icon/lib/simple-icon-button.js";export class StarRating extends e{static get styles(){return s`
      :host {
        display: block;
        padding: 25px;
        color: var(--star-rating-text-color, #000);
      }
      #star-rating {
        display: inline-flex;
      }
      .star {
        background-color: black;
      }
      #rating {
        padding-left: 20px; //responsive sizes weren't working well
      }
      h2 {
        margin: 0; //keeps them both on the same line
      }
      :host(:not([interactive])) simple-icon-button::part(button) {
        cursor: unset;
      }
    `}static get properties(){return{score:{type:Number},possible:{type:Number},interactive:{type:Boolean,reflect:!0},numStars:{type:Number,attribute:"num-stars"},_calPercent:{type:Number}}}constructor(){super(),this._calPercent=0,this.numStars=5,this.score=0,this.possible=100,this.interactive=!1}render(){return t`
      <div class="star-rating">
        <div id="stars">
          ${this.renderStar(this.numStars,this.interactive)}
        </div>
        <div id="rating">
          <h2>${this._calPercent} (${this.score}/${this.possible})</h2>
        </div>
      </div>
    `}renderStar(s,e){if(0===s)return t``;let r=0;const i=[];for(;r<s;)r+1<s*this._calPercent?i.push(t`<simple-icon-button
            icon="star"
            dark
            contrast="1"
            class="star"
            ?disabled="${!e}"
          ></simple-icon-button>`):s*this._calPercent-(r+1)<-.01&&s*this._calPercent-(r+1)>=-.5?i.push(t`<simple-icon-button
            icon="star-half"
            dark
            contrast="1"
            class="star"
            ?disabled="${!e}"
          ></simple-icon-button>`):i.push(t`<simple-icon-button
            icon="star-border"
            dark
            contrast="1"
            class="star"
            ?disabled="${!e}"
          ></simple-icon-button>`),r++;return i}updated(t){t.forEach(((t,s)=>{["score","possible"].includes(s)&&0!==this.possible&&(this._calPercent=Math.round(this.score/this.possible*100)/100)}))}}customElements.define("star-rating",StarRating);
