// @lrnwebcomponents/simple-popover/simple-popover.js
define(["exports","../../lit-element/lit-element.js","../absolute-position-behavior/absolute-position-behavior.js"],(function(_exports,_litElement,_absolutePositionBehavior){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.SimplePopover=void 0;
/**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class SimplePopover extends _absolutePositionBehavior.AbsolutePositionBehavior{static get styles(){return[_litElement.css`
        :host([hidden]) {
          display: none;
        }
        :host > div {
          display: flex;
          flex-direction: column-reverse;
          justify-content: stretch;
          z-index: 1;
        }
        :host([position="left"]) > div {
          justify-content: flex-start;
          flex-direction: row;
        }
        :host([position="right"]) > div {
          justify-content: flex-end;
          flex-direction: row-reverse;
        }
        :host([position="top"]) > div {
          flex-direction: column;
        }
        :host > div > * {
          width: 100%;
        }
        :host([position="left"]) > div > *,
        :host([position="right"]) > div > * {
          width: unset;
        }
        #content {
          margin: 0;
          padding: var(--simple-popover-padding, 10px);
          color: var(--simple-popover-color, #222);
          background-color: var(--simple-popover-background-color, white);
          border: 1px solid var(--simple-popover-border-color, #bbb);
          min-height: 20px;
          border-radius: var(--simple-popover-border-radius, 3px);
          box-shadow: var(
            --simple-popover-box-shadow,
            rgba(60, 64, 67, 0.3) 0px 4px 8px 3px
          );
        }
        #pointer-outer {
          margin: -1px;
        }
        #pointer {
          width: 20px;
          height: 20px;
          position: relative;
          overflow: hidden;
          flex: 0 0 20px;
        }
        #pointer:after {
          content: "";
          position: absolute;
          width: 10px;
          height: 10px;
          background-color: var(--simple-popover-background-color, white);
          border: 1px solid var(--simple-popover-border-color, #bbb);
          transform: rotate(45deg);
          top: 15px;
          left: 5px;
        }
        :host([position="top"]) #pointer:after {
          top: -5px;
          left: 5px;
        }
        :host([position="right"]) #pointer:after {
          top: 5px;
          left: 15px;
        }
        :host([position="left"]) #pointer:after {
          top: 5px;
          left: -5px;
        }
      `]}render(){return _litElement.html`
      <div>
        <div id="content" role="alertdialog">
          <slot></slot>
        </div>
        <div id="pointer-outer">
          <div id="pointer" .style=${this._getMargins(this.__positions)}></div>
        </div>
      </div>
    `}static get haxProperties(){return{canScale:!0,canPosition:!0,canEditSource:!1,gizmo:{title:"Simple popover",description:"A popover alertdialog that is positioned next to a target element",icon:"icons:android",color:"green",groups:["Popover"],handles:[{type:"todo:read-the-docs-for-usage"}],meta:{author:"nikkimk",owner:"The Pennsylvania State University"}},settings:{quick:[],configure:[{property:"title",description:"",inputMethod:"textfield",required:!1,icon:"icons:android"}],advanced:[]}}}static get properties(){return{...super.properties}}constructor(){super(),this.offset=-10,this.fitToVisibleBounds=!0}static get tag(){return"simple-popover"}_getMargins(positions){if(positions&&positions.target){let self=this.getBoundingClientRect(),h="bottom"===this.position||"top"===this.position,max=h?self.width:self.height,sStart=h?self.left:self.top,center=(h?positions.target.left:positions.target.top)+(h?positions.target.width/2:positions.target.height/2)-10,margin=Math.min(max-20,Math.max(0,center-sStart));return h?`margin: 0 0 0 ${margin}px;`:`margin: ${margin}px 0 0 0;`}return""}}_exports.SimplePopover=SimplePopover,window.customElements.define(SimplePopover.tag,SimplePopover)}));
