// @lrnwebcomponents/sorting-question/sorting-question.js
import{LitElement as t,html as e,css as r}from"../../lit/index.js";import{I18NMixin as i}from"../i18n-manager/lib/I18NMixin.js";import"../simple-icon/lib/simple-icon-button-lite.js";import"../simple-icon/lib/simple-icons.js";import"./lib/sorting-option.js";export class SortingQuestion extends(i(t)){static get tag(){return"sorting-question"}constructor(){super(),this.celebrate=!1,this.shame=!1,this.numberOfOptions=this.children.length,this.numberCorrrect=0,this.correctOrder=[],this.mute=!1,this.noBackground=!1,this.question="Sort the following in order!",this.t={numCorrectLeft:"You Have",numCorrectRight:"Correct.",submit:"Submit",reset:"Reset"},this.registerLocalization({context:this,localesPath:new URL("../locales/",import.meta.url).href,locales:["he","ja","es"]}),this.getCorrectOrder(),setTimeout(this.randomizeOptions(),100)}getCorrectOrder(){let t=this;this.childNodes.forEach((function(e){"SORTING-OPTION"==e.tagName&&t.correctOrder.push(e)}))}randomizeOptions(){let t=[];for(var e=0;e<this.numberOfOptions;e++)t.push(e);for(e=0;e<this.numberOfOptions;e++){var r=Math.floor(Math.random()*(this.numberOfOptions-0)+0);void 0===r&&(r=0);let e=this.children[t[r]];var i=Math.floor(Math.random()*(t.length-0)+0);void 0===i&&(i=0),this.insertBefore(e,this.children[t[i]])}}checkOrder(){for(var t=0,e=0;e<this.numberOfOptions;e++)this.children[e].isEqualNode(this.correctOrder[e])?(t+=1,this.children[e].removeAttribute("incorrect"),this.children[e].setAttribute("correct",!0),this.children[e].shadowRoot.querySelector("#correct-icon").style.display="flex",this.children[e].shadowRoot.querySelector("#incorrect-icon").style.display="none"):(this.children[e].removeAttribute("correct"),this.children[e].setAttribute("incorrect",!0),this.children[e].shadowRoot.querySelector("#incorrect-icon").style.display="flex",this.children[e].shadowRoot.querySelector("#correct-icon").style.display="none");this.numberCorrrect=t,this.celebrate&&this.celebration(this.numberCorrrect,this.numberOfOptions),this.shame&&this.celebration(this.numberCorrrect,this.numberOfOptions)}celebration(t,e){t==e&&import("../multiple-choice/lib/confetti-container.js").then((t=>{setTimeout((()=>{this.shadowRoot.querySelector("#confetti").setAttribute("popped","")}),0)}))}reset(){this.childNodes.forEach((function(t){"SORTING-OPTION"==t.tagName&&(t.shadowRoot.querySelector("#incorrect-icon").style.display="none",t.shadowRoot.querySelector("#correct-icon").style.display="none",t.removeAttribute("incorrect"),t.removeAttribute("correct"))})),this.numberCorrrect=0,this.randomizeOptions()}static get properties(){return{...super.properties,question:{type:String,reflect:!0},correctOrder:{type:Array},numberOfOptions:{type:Number},numberCorrrect:{type:Number},celebrate:{type:Boolean},shame:{type:Boolean},mute:{type:Boolean},disabled:{type:Boolean},noBackground:{type:Boolean,attribute:"no-background"}}}updated(t){t.forEach(((t,e)=>{if("need"===e&&"joy"===this[e]&&this.classList.add("joyful"),this.disabled){var r=this.shadowRoot.querySelector(".reset-button"),i=this.shadowRoot.querySelector(".submit-button");r.setAttribute("disabled",!0),i.setAttribute("disabled",!0),this.childNodes.forEach((function(t){"SORTING-OPTION"==t.tagName&&(t.setAttribute("disabled",!0),t.setAttribute("draggable",!1),t.removeAttribute("correct"),t.removeAttribute("incorrect"),t.shadowRoot.querySelector("#incorrect-icon").style.display="none",t.shadowRoot.querySelector("#correct-icon").style.display="none",t.style.opacity="0.5")}))}else{r=this.shadowRoot.querySelector(".reset-button"),i=this.shadowRoot.querySelector(".submit-button");r.removeAttribute("disabled"),i.removeAttribute("disabled"),this.childNodes.forEach((function(t){"SORTING-OPTION"==t.tagName&&(t.removeAttribute("disabled"),t.setAttribute("draggable",!0),t.style.opacity="1")}))}"question"==e&&this.requestUpdate()}))}firstUpdated(t){if(super.firstUpdated&&(super.firstUpdated(t),this.disabled)){var e=this.shadowRoot.querySelector(".reset-button"),r=this.shadowRoot.querySelector(".submit-button");e.setAttribute("disabled",!0),r.setAttribute("disabled",!0),this.childNodes.forEach((function(t){"SORTING-OPTION"==t.tagName&&(t.setAttribute("disabled",!0),t.setAttribute("draggable",!1),t.style.opacity="0.5")}))}}static get styles(){return r`
      :host {
        background-color: var(--sorting-question-background-color, white);

        background-position: center;
        border: 2px solid black;
        padding: 15px 10px;
        display: flex;
        flex-direction: column;
        border-radius: 10px;
        box-shadow: 2px 1px 2px -1px black;
        margin-top: 10px;
        margin-bottom: 10px;
        font-weight: bold;
        height: inherit;
        width: inherit;
      }

      .sorting-question-header {
        text-align: center;
        font-size: 20px;
        font-family: revert;
        margin-bottom: 10px;
      }

      .options {
        margin-bottom: 10px;
      }

      .sorting-controls {
        display: flex;
        justify-content: space-between;
        padding-right: 5%;
        font-size: 20px;
        font-family: revert;
      }

      .submit-button {
        border-radius: 5px;
        border-width: 2px;
        cursor: pointer;
        background-color: inherit;
        box-shadow: 1px 1px 1px 0px;
        font-weight: bold;
      }

      .submit-button:hover {
        opacity: 0.8;
      }

      .submit-button:active {
        cursor: default;
      }

      .button-container {
        display: flex;
        justify-content: center;
        align-items: center;
      }

      .reset-button {
        margin-right: 5px;
        border-radius: 5px;
        padding: 0 4px;
        background-color: inherit;
        box-shadow: 1px 1px 1px 0px;
        cursor: pointer;
      }

      .reset-button:disabled {
        opacity: 0.5;
      }

      .reset-button simple-icon-button {
        height: inherit;
        width: inherit;
        padding: 1px;
      }
    `}render(){return e`
      <div class="sorting-question-header">${this.question}</div>
      <confetti-container id="confetti">
        <div class="options"><slot></slot></div>
      </confetti-container>
      <div class="sorting-controls">
        <span
          >${this.t.numCorrectLeft}
          ${this.numberCorrrect}/${this.numberOfOptions}
          ${this.t.numCorrectRight}</span
        >
        <div class="button-container">
          <simple-icon-button-lite
            @click="${this.reset}"
            class="reset-button"
            icon="refresh"
            >${this.t.reset}</simple-icon-button-lite
          >
          <button
            @click="${this.checkOrder}"
            class="submit-button"
            icon="submit"
          >
            ${this.t.submit}
          </button>
        </div>
      </div>
    `}static get haxProperties(){return new URL(`./lib/${this.tag}.haxProperties.json`,import.meta.url).href}}window.customElements.define(SortingQuestion.tag,SortingQuestion);
