// @vaadin/router/src/triggers/popstate.js
define(["exports", "../utils.js"], function (_exports, _utils) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.default = void 0;
  // PopStateEvent constructor shim
  var isIE = /Trident/.test(navigator.userAgent);
  /* istanbul ignore next: coverage is calculated in Chrome, this code is for IE */

  if (isIE && !(0, _utils.isFunction)(window.PopStateEvent)) {
    window.PopStateEvent = function (inType, params) {
      params = params || {};
      var e = document.createEvent('Event');
      e.initEvent(inType, Boolean(params.bubbles), Boolean(params.cancelable));
      e.state = params.state || null;
      return e;
    };

    window.PopStateEvent.prototype = window.Event.prototype;
  }

  function vaadinRouterGlobalPopstateHandler(event) {
    if (event.state === 'vaadin-router-ignore') {
      return;
    }

    var _window$location = window.location,
        pathname = _window$location.pathname,
        search = _window$location.search,
        hash = _window$location.hash;
    (0, _utils.fireRouterEvent)('go', {
      pathname: pathname,
      search: search,
      hash: hash
    });
  }
  /**
   * A navigation trigger for Vaadin Router that translates popstate events into
   * Vaadin Router navigation events.
   *
   * @memberOf Router.NavigationTrigger
   * @type {NavigationTrigger}
   */


  var POPSTATE = {
    activate: function activate() {
      window.addEventListener('popstate', vaadinRouterGlobalPopstateHandler);
    },
    inactivate: function inactivate() {
      window.removeEventListener('popstate', vaadinRouterGlobalPopstateHandler);
    }
  };
  var _default = POPSTATE;
  _exports.default = _default;
});
