// @lrnwebcomponents/html-block/html-block.js
define(["exports"], function (_exports) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HtmlBlock = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  /**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `html-block`
   * @element html-block
   * `A basic HTML block that provides HAXschema wiring`
   *
   * @microcopy - language worth noting:
   *  -
   *
  
   * @demo demo/index.html
   */
  var HtmlBlock =
  /*#__PURE__*/
  function (_HTMLElement) {
    babelHelpers.inherits(HtmlBlock, _HTMLElement);
    babelHelpers.createClass(HtmlBlock, [{
      key: "html",
      // render function
      get: function get() {
        return "\n<style>\n\n        </style>\n<slot></slot>";
      } // haxProperty definition

    }], [{
      key: "haxProperties",
      get: function get() {
        return {
          canScale: true,
          canPosition: true,
          canEditSource: false,
          gizmo: {
            title: "Html block",
            description: "A basic HTML block that provides HAXschema wiring",
            icon: "icons:warning",
            color: "red",
            groups: ["Block"],
            handles: [{
              type: "html",
              content: "slot"
            }],
            meta: {
              author: "btopro",
              owner: "The Pennsylvania State University"
            }
          },
          settings: {
            quick: [],
            configure: [{
              slot: "",
              title: "HTML",
              description: "HTML code you want to present in content",
              inputMethod: "code-editor"
            }],
            advanced: []
          }
        };
      } // properties available to the custom element for data binding

    }, {
      key: "properties",
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(HtmlBlock), "properties", this));
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */

    }, {
      key: "tag",
      get: function get() {
        return "html-block";
      }
      /**
       * life cycle
       */

    }]);

    function HtmlBlock() {
      var _this;

      var delayRender = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
      babelHelpers.classCallCheck(this, HtmlBlock);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(HtmlBlock).call(this)); // set tag for later use

      _this.tag = HtmlBlock.tag; // map our imported properties json to real props on the element
      // @notice static getter of properties is built via tooling
      // to edit modify src/HtmlBlock-properties.json

      return _this;
    }
    /**
     * life cycle, element is afixed to the DOM
     */


    babelHelpers.createClass(HtmlBlock, [{
      key: "connectedCallback",
      value: function connectedCallback() {
        // default we block all script unless the user says to do so
        // @todo ensure HAX actually respects this down the road, right now it sanitizes it
        this.allowscript = false;
        this.__ignoreChange = false;
        this.style.display = "block"; // ensure we keep applying sanitization as needed while monitoring the tree

        this.__observer = new MutationObserver(this.render.bind(this));

        this.__observer.observe(this, {
          attributes: true,
          characterData: true,
          childList: true,
          subtree: true
        });
      }
    }, {
      key: "render",
      value: function render() {
        if (!this.__ignoreChange) {
          if (this.allowscript == null || !this.allowscript || babelHelpers.typeof(this.allowscript) === (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
            this.__sanitizeHTML();
          }
        } else {
          this.__ignoreChange = false;
        }
      }
    }, {
      key: "attributeChangedCallback",
      // disconnectedCallback() {}
      value: function attributeChangedCallback(attr, oldValue, newValue) {
        if (attr === "allowscript") {
          if (newValue == null || !newValue || babelHelpers.typeof(newValue) === (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
            // we should sanitize innerHTML but create a holding pen for the rawHTML first
            this.__sanitizeHTML();
          } else {
            // see if we had anything in the holding pen
            if (this.__rawHTML) {
              this.__ignoreChange = true;
              this.innerHTML = this.__rawHTML;
            }
          }
        }
      }
    }, {
      key: "__sanitizeHTML",
      value: function __sanitizeHTML() {
        if (!this.__pen) {
          this.__pen = document.createElement("div");
        }

        this.__pen.innerHTML = this.innerHTML;
        this.__rawHTML = this.__pen.cloneNode(true).innerHTML; // clear it up

        if (typeof this.innerHTML === "function") {
          this.innerHTML = this.innerHTML.replace(/<script[\s\S]*?>/gi, "&lt;script&gt;");
          this.innerHTML = this.innerHTML.replace(/<\/script>/gi, "&lt;/script&gt;");
        }
      }
    }, {
      key: "allowscript",
      get: function get() {
        return this.getAttribute("allowscript");
      },
      set: function set(value) {
        if (value) {
          this.setAttribute("allowscript", "allowscript");
        } else {
          this.removeAttribute("allowscript");
        }
      }
    }], [{
      key: "observedAttributes",
      get: function get() {
        return ["allowscript"];
      }
    }]);
    return HtmlBlock;
  }(babelHelpers.wrapNativeSuper(HTMLElement));

  _exports.HtmlBlock = HtmlBlock;
  window.customElements.define(HtmlBlock.tag, HtmlBlock);
});
