// @lrnwebcomponents/bootstrap-theme/lib/BootstrapFooter.js
define(["exports", "meta", "../../../lit/index.js", "../../simple-icon/lib/simple-icon-lite.js", "../../simple-icon/lib/simple-icons.js", "../../haxcms-elements/lib/core/haxcms-site-store.js", "../../../mobx/dist/mobx.esm.js"], function (_exports, meta, _index, _simpleIconLite, _simpleIcons, _haxcmsSiteStore, _mobxEsm) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.BootstrapFooter = void 0;
  meta = _interopRequireWildcard(meta);

  var _templateObject_58d6fd50942811ecb1937969798b231a, _templateObject2_58d6fd50942811ecb1937969798b231a;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var BootstrapFooter = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(BootstrapFooter, _LitElement);

    var _super = _createSuper(BootstrapFooter);

    function BootstrapFooter() {
      var _this;

      babelHelpers.classCallCheck(this, BootstrapFooter);
      _this = _super.call(this);

      var basePath = _this.getBasePath(decodeURIComponent(meta.url));

      _this._bootstrapPath = basePath + "bootstrap/dist/css/bootstrap.min.css";
      _this._forwardItem = {};
      _this._backwardItem = {};
      _this._activeItemIndex = -1;
      _this._routerManifest = {};
      _this.__disposer = _this.__disposer ? _this.__disposer : [];
      (0, _mobxEsm.autorun)(function (reaction) {
        var storeActiveItemIndex = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.activeManifestIndex);
        var storeRouterManifest = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.routerManifest);

        if (_this._activeItem !== storeActiveItemIndex || _this._routerManifest !== storeRouterManifest) {
          _this._activeItem = storeActiveItemIndex;
          _this._routerManifest = storeRouterManifest;

          if (storeRouterManifest.items[storeActiveItemIndex - 1]) {
            _this._backwardItem = storeRouterManifest.items[storeActiveItemIndex - 1];
          }

          if (storeRouterManifest.items[storeActiveItemIndex + 1]) {
            _this._forwardItem = storeRouterManifest.items[storeActiveItemIndex + 1];
          }
        }

        _this.__disposer.push(reaction);
      });
      return _this;
    }

    babelHelpers.createClass(BootstrapFooter, [{
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_58d6fd50942811ecb1937969798b231a || (_templateObject_58d6fd50942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <link rel=\"stylesheet\" href=\"", "\" />\n      <div class=\"container\">\n        <a class=\"backward\" href=\"", "\">\n          <button class=\"btn btn-outline-primary\">\n            <simple-icon-lite icon=\"av:fast-rewind\"></simple-icon-lite>\n            ", "\n          </button>\n        </a>\n        <a class=\"forward\" href=\"", "\">\n          <button class=\"btn btn-outline-primary forward\">\n            ", "\n            <simple-icon-lite icon=\"av:fast-forward\"></simple-icon-lite>\n          </button>\n        </a>\n      </div>\n    "])), this._bootstrapPath, this._backwardItem.slug, this._backwardItem.title, this._forwardItem.slug, this._forwardItem.title);
      }
    }, {
      key: "getBasePath",
      value: function getBasePath(url) {
        return url.substring(0, url.lastIndexOf("/@lrnwebcomponents/") + 1);
      }
    }], [{
      key: "tag",
      get: function get() {
        return "bootstrap-footer";
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _index.css)(_templateObject2_58d6fd50942811ecb1937969798b231a || (_templateObject2_58d6fd50942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          margin-bottom: 10px;\n          --simple-icon-height: 18px;\n          --simple-icon-width: 18px;\n          --simple-icon-color: #007bff;\n        }\n        .btn-outline-primary {\n          color: #007bff;\n          border-color: #007bff;\n        }\n\n        .btn-outline-primary:active,\n        .btn-outline-primary:focus,\n        .btn-outline-primary:hover {\n          --simple-icon-color: white;\n          color: #fff;\n          background-color: #007bff;\n          border-color: #007bff;\n        }\n\n        .container {\n          display: flex;\n          background-color: #e9ecef;\n          border-radius: 5px;\n          padding: 10px;\n          -moz-box-shadow: inset 0 2px 4px 0 rgb(0 0 0 / 10%);\n          -webkit-box-shadow: inset 0 2px 4px 0 rgb(0 0 0 / 10%);\n          box-shadow: inset 0 2px 4px 0 rgb(0 0 0 / 10%);\n        }\n\n        .forward {\n          margin-left: auto;\n        }\n\n        .backward {\n          margin-right: auto;\n        }\n\n        simple-icon-lite {\n          margin-bottom: 3px;\n        }\n\n        a {\n          text-decoration: none;\n          color: #007bff;\n        }\n        /* dark theme */\n        :host([color-theme=\"1\"]) {\n          --simple-icon-color: var(--bootstrap-theme-dark-color);\n        }\n\n        :host([color-theme=\"1\"]) .btn {\n          color: var(--bootstrap-theme-dark-color);\n          background-color: transparent;\n          background-image: none;\n          border-color: #6c757d;\n        }\n\n        :host([color-theme=\"1\"]) .btn:hover {\n          color: var(--bootstrap-theme-dark-color);\n          background-color: #6c757d;\n          border-color: #6c757d;\n        }\n\n        :host([color-theme=\"1\"]) .container {\n          background-color: var(\n            --bootstrap-theme-dark-secondary-background-color\n          );\n          -moz-box-shadow: inset 0 2px 4px 0\n            rgb(var(--bootstrap-theme-light-secondary-background-color), 0.7);\n          -webkit-box-shadow: inset 0 2px 4px 0\n            rgb(var(--bootstrap-theme-light-secondary-background-color), 0.7);\n          box-shadow: inset 0 2px 4px 0\n            rgb(var(--bootstrap-theme-light-secondary-background-color), 0.7);\n        }\n        /* palenight theme */\n        :host([color-theme=\"2\"]) {\n          --simple-icon-color: var(--bootstrap-theme-palenight-color);\n        }\n\n        :host([color-theme=\"2\"]) .btn {\n          color: var(--bootstrap-theme-palenight-color);\n          background-color: transparent;\n          background-image: none;\n          border-color: #6c757d;\n        }\n\n        :host([color-theme=\"2\"]) .btn:hover {\n          color: var(--bootstrap-theme-palenight-color);\n          background-color: #6c757d;\n          border-color: #6c757d;\n        }\n\n        :host([color-theme=\"2\"]) .container {\n          background-color: var(\n            --bootstrap-theme-palenight-secondary-background-color\n          );\n          -moz-box-shadow: inset 0 2px 4px 0\n            rgb(var(--bootstrap-theme-light-secondary-background-color), 0.7);\n          -webkit-box-shadow: inset 0 2px 4px 0\n            rgb(var(--bootstrap-theme-light-secondary-background-color), 0.7);\n          box-shadow: inset 0 2px 4px 0\n            rgb(var(--bootstrap-theme-light-secondary-background-color), 0.7);\n        }\n      "])))];
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          _backwardItem: {
            type: Object
          },
          _forwardItem: {
            type: Object
          },
          colorTheme: {
            type: String,
            attribute: "color-theme"
          }
        };
      }
    }]);
    return BootstrapFooter;
  }(_index.LitElement);

  _exports.BootstrapFooter = BootstrapFooter;
  customElements.define(BootstrapFooter.tag, BootstrapFooter);
});
