// @lrnwebcomponents/bootstrap-theme/lib/BootstrapBreadcrumb.js
define(["exports", "meta", "../../../lit/index.js", "../../simple-icon/lib/simple-icon-lite.js", "../../simple-icon/lib/simple-icons.js", "../../haxcms-elements/lib/core/haxcms-site-store.js", "../../../mobx/dist/mobx.esm.js"], function (_exports, meta, _index, _simpleIconLite, _simpleIcons, _haxcmsSiteStore, _mobxEsm) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.BootstrapBreadcrumb = void 0;
  meta = _interopRequireWildcard(meta);

  var _templateObject_58d3a1f0942811ecb1937969798b231a, _templateObject2_58d3a1f0942811ecb1937969798b231a, _templateObject3_58d3a1f0942811ecb1937969798b231a;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `bootstrap-breadcrumb`
   * `Breadcrumb element for bootstrap theme`
   * @demo demo/index.html
   * @element bootstrap-breadcrumb
   */
  var BootstrapBreadcrumb = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(BootstrapBreadcrumb, _LitElement);

    var _super = _createSuper(BootstrapBreadcrumb);

    function BootstrapBreadcrumb() {
      var _this;

      babelHelpers.classCallCheck(this, BootstrapBreadcrumb);
      _this = _super.call(this);
      _this.items = [];
      _this._activeItem = {};
      _this.homeItem = {};

      var basePath = _this.getBasePath(decodeURIComponent(meta.url));

      _this._bootstrapPath = basePath + "bootstrap/dist/css/bootstrap.min.css";
      _this.__disposer = _this.__disposer ? _this.__disposer : [];
      (0, _mobxEsm.autorun)(function (reaction) {
        var manifestHomeItem = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.manifest.items[0]);
        var storeActiveItem = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.activeItem); // check if home item has changed, if it has set new home item

        if (_this.homeItem !== manifestHomeItem) {
          _this.homeItem = manifestHomeItem;
        } // check if we have a new active item
        // if so we clear our items array, set a new activeItem, push it to the items array
        // then check for a parent, if a parent is present call recursive function that keeps adding subsequent parents


        if (storeActiveItem && _this._activeItem !== storeActiveItem) {
          _this.items = [];
          _this._activeItem = storeActiveItem;

          _this.items.push(storeActiveItem);

          if (storeActiveItem.parent) {
            _this.addParentToItems(storeActiveItem);
          }
        }

        _this.__disposer.push(reaction);
      });
      return _this;
    }

    babelHelpers.createClass(BootstrapBreadcrumb, [{
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_58d3a1f0942811ecb1937969798b231a || (_templateObject_58d3a1f0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <link rel=\"stylesheet\" href=\"", "\" />\n      <div class=\"container p-0 mb-3\">\n        <nav aria-label=\"breadcrumb\">\n          <ol class=\"breadcrumb m-auto\">\n            <li\n              id=\"home\"\n              class=\"breadcrumb-item\"\n              data-bs-toggle=\"tooltip\"\n              data-bs-placement=\"top\"\n              title=\"", "\"\n            >\n              <a href=\"", "\">\n                <simple-icon-lite accent-color=\"blue\" icon=\"home\">\n                </simple-icon-lite>\n              </a>\n            </li>\n            ", "\n          </ol>\n        </nav>\n      </div>\n    "])), this._bootstrapPath, this.homeItem.title, this.homeItem.slug, this.items.map(function (item) {
          return (0, _index.html)(_templateObject2_58d3a1f0942811ecb1937969798b231a || (_templateObject2_58d3a1f0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                  <li\n                    class=\"breadcrumb-item\"\n                    data-bs-toggle=\"tooltip\"\n                    data-bs-placement=\"top\"\n                    title=\"", "\"\n                  >\n                    <a href=\"", "\"> ", " </a>\n                  </li>\n                "])), item.title, item.slug, item.title);
        }));
      }
    }, {
      key: "getBasePath",
      value: function getBasePath(url) {
        return url.substring(0, url.lastIndexOf("/@lrnwebcomponents/") + 1);
      } // gets parents item by the items id

    }, {
      key: "getParentById",
      value: function getParentById(parentId) {
        var elementPos = _haxcmsSiteStore.store.manifest.items.map(function (item) {
          return (0, _mobxEsm.toJS)(item.id);
        }).indexOf(parentId);

        var parentFound = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.manifest.items[elementPos]);
        return parentFound;
      } // recursive function that keeps adding parents to items array

    }, {
      key: "addParentToItems",
      value: function addParentToItems(item) {
        var parentItem = this.getParentById(item.parent);
        this.items.unshift(parentItem);

        if (parentItem.parent) {
          this.addParentToItems(parentItem);
        }
      }
    }, {
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {}
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        changedProperties.forEach(function (oldValue, propName) {});
      }
    }], [{
      key: "styles",
      get: function get() {
        return [(0, _index.css)(_templateObject3_58d3a1f0942811ecb1937969798b231a || (_templateObject3_58d3a1f0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          --bootstrap-dark-theme-secondary-background-color: #343a40;\n          --simple-icon-height: 18px;\n          --simple-icon-width: 18px;\n          --simple-icon-color: #007bff;\n        }\n        a {\n          color: #007bff;\n        }\n\n        .container {\n          background-color: #e9ecef;\n          border-radius: 5px;\n        }\n\n        .breadcrumb {\n          -moz-box-shadow: inset 0 2px 4px 0 rgb(0 0 0 / 10%);\n          -webkit-box-shadow: inset 0 2px 4px 0 rgb(0 0 0 / 10%);\n          box-shadow: inset 0 2px 4px 0 rgb(0 0 0 / 10%);\n        }\n\n        simple-icon-lite {\n          margin-bottom: 5px;\n        }\n\n        a:hover simple-icon-lite {\n          text-decoration: underline;\n        }\n\n        /* dark mode */\n        :host([color-theme=\"1\"]) {\n          --simple-icon-color: #999;\n        }\n\n        :host([color-theme=\"1\"]) simple-icon-lite:hover {\n          --simple-icon-color: #fff;\n        }\n\n        :host([color-theme=\"1\"]) .breadcrumb {\n          background-color: var(\n            --bootstrap-dark-theme-secondary-background-color\n          );\n          -moz-box-shadow: inset 0 2px 4px 0\n            rgb(var(--bootstrap-theme-light-secondary-background-color), 0.7);\n          -webkit-box-shadow: inset 0 2px 4px 0\n            rgb(var(--bootstrap-theme-light-secondary-background-color), 0.7);\n          box-shadow: inset 0 2px 4px 0\n            rgb(var(--bootstrap-theme-light-secondary-background-color), 0.7);\n        }\n\n        :host([color-theme=\"1\"]) .container {\n          background-color: var(\n            --bootstrap-dark-theme-secondary-background-color\n          );\n        }\n\n        :host([color-theme=\"1\"]) a {\n          color: #999;\n        }\n\n        :host([color-theme=\"1\"]) a:hover {\n          color: #fff;\n        }\n\n        /* palenight theme */\n        :host([color-theme=\"2\"]) {\n          --simple-icon-color: var(--bootstrap-theme-palenight-color);\n        }\n\n        :host([color-theme=\"2\"]) simple-icon-lite:hover {\n          --simple-icon-color: var(--bootstrap-theme-palenight-color);\n        }\n\n        :host([color-theme=\"2\"]) .breadcrumb {\n          background-color: var(\n            --bootstrap-theme-palenight-secondary-background-color\n          );\n          -moz-box-shadow: inset 0 2px 4px 0\n            rgb(var(--bootstrap-theme-light-secondary-background-color), 0.7);\n          -webkit-box-shadow: inset 0 2px 4px 0\n            rgb(var(--bootstrap-theme-light-secondary-background-color), 0.7);\n          box-shadow: inset 0 2px 4px 0\n            rgb(var(--bootstrap-theme-light-secondary-background-color), 0.7);\n        }\n\n        :host([color-theme=\"2\"]) .container {\n          background-color: var(\n            --bootstrap-theme-palenight-secondary-background-color\n          );\n        }\n\n        :host([color-theme=\"2\"]) a {\n          color: var(--bootstrap-theme-palenight-color);\n        }\n\n        :host([color-theme=\"2\"]) a:hover {\n          color: var(--bootstrap-theme-palenight-color);\n        }\n      "])))];
      }
    }, {
      key: "tag",
      get: function get() {
        return "bootstrap-breadcrumb";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          items: {
            type: Array
          },
          homeItem: {
            type: Object
          },
          colorTheme: {
            type: String,
            reflect: true,
            attribute: "color-theme"
          }
        };
      }
    }]);
    return BootstrapBreadcrumb;
  }(_index.LitElement);

  _exports.BootstrapBreadcrumb = BootstrapBreadcrumb;
  customElements.define(BootstrapBreadcrumb.tag, BootstrapBreadcrumb);
});
