// @lrnwebcomponents/bootstrap-theme/lib/BootstrapSearch.js
define(["exports", "meta", "../../../lit-element/lit-element.js", "../../../mobx/dist/mobx.esm.js", "../../haxcms-elements/lib/core/haxcms-site-store.js"], function (_exports, meta, _litElement, _mobxEsm, _haxcmsSiteStore) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.BootstrapSearch = void 0;
  meta = babelHelpers.interopRequireWildcard(meta);

  var _templateObject_790e7dc0b81c11eb80aa4b4d3d30a499, _templateObject2_790e7dc0b81c11eb80aa4b4d3d30a499;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var BootstrapSearch = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(BootstrapSearch, _LitElement);

    var _super = _createSuper(BootstrapSearch);

    function BootstrapSearch() {
      var _this;

      babelHelpers.classCallCheck(this, BootstrapSearch);
      _this = _super.call(this);

      var basePath = _this.getBasePath(decodeURIComponent(meta.url));

      _this._bootstrapPath = basePath + "bootstrap/dist/css/bootstrap.min.css";
      _this.searchText = "";
      return _this;
    }

    babelHelpers.createClass(BootstrapSearch, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_790e7dc0b81c11eb80aa4b4d3d30a499 || (_templateObject_790e7dc0b81c11eb80aa4b4d3d30a499 = babelHelpers.taggedTemplateLiteral(["\n      <link rel=\"stylesheet\" href=\"", "\" />\n      <form class=\"d-flex\">\n        <input\n          class=\"form-control me-2\"\n          type=\"search\"\n          placeholder=\"Type to search\"\n          aria-label=\"Search\"\n          @input=", "\n        />\n      </form>\n    "])), this._bootstrapPath, this.inputChanged);
      }
    }, {
      key: "inputChanged",
      value: function inputChanged(evt) {
        this.searchText = evt.target.value;
        this.dispatchEvent(new CustomEvent("search-changed", {
          bubbles: true,
          composed: true,
          detail: this
        }));
      }
    }, {
      key: "getBasePath",
      value: function getBasePath(url) {
        return url.substring(0, url.lastIndexOf("/@lrnwebcomponents/") + 1);
      }
    }], [{
      key: "tag",
      get: function get() {
        return "bootstrap-search";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          colorTheme: {
            type: String,
            reflect: true,
            attribute: "color-theme"
          },
          searchText: {
            type: String
          }
        };
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_790e7dc0b81c11eb80aa4b4d3d30a499 || (_templateObject2_790e7dc0b81c11eb80aa4b4d3d30a499 = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          float: right;\n        }\n        input {\n          max-width: var(--bootstrap-search-max-width, 300px);\n          width: var(--bootstrap-search-width, 300px);\n        }\n      "])))];
      }
    }]);
    return BootstrapSearch;
  }(_litElement.LitElement);

  _exports.BootstrapSearch = BootstrapSearch;
  customElements.define(BootstrapSearch.tag, BootstrapSearch);
});
