// @vaadin/component-base/src/disabled-mixin.js
define(["exports", "../../../@polymer/polymer/lib/utils/mixin.js"], function (_exports, _mixin) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.DisabledMixin = void 0;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * A mixin to provide disabled property for field components.
   *
   * @polymerMixin
   */
  var DisabledMixin = (0, _mixin.dedupingMixin)(function (superclass) {
    return /*#__PURE__*/function (_superclass) {
      babelHelpers.inherits(DisabledMixinClass, _superclass);

      var _super = _createSuper(DisabledMixinClass);

      function DisabledMixinClass() {
        babelHelpers.classCallCheck(this, DisabledMixinClass);
        return _super.apply(this, arguments);
      }

      babelHelpers.createClass(DisabledMixinClass, [{
        key: "_disabledChanged",
        value:
        /**
         * @param {boolean} disabled
         * @protected
         */
        function _disabledChanged(disabled) {
          this._setAriaDisabled(disabled);
        }
        /**
         * @param {boolean} disabled
         * @protected
         */

      }, {
        key: "_setAriaDisabled",
        value: function _setAriaDisabled(disabled) {
          if (disabled) {
            this.setAttribute('aria-disabled', 'true');
          } else {
            this.removeAttribute('aria-disabled');
          }
        }
        /**
         * Overrides the default element `click` method in order to prevent
         * firing the `click` event when the element is disabled.
         * @protected
         * @override
         */

      }, {
        key: "click",
        value: function click() {
          if (!this.disabled) {
            babelHelpers.get(babelHelpers.getPrototypeOf(DisabledMixinClass.prototype), "click", this).call(this);
          }
        }
      }], [{
        key: "properties",
        get: function get() {
          return {
            /**
             * If true, the user cannot interact with this element.
             */
            disabled: {
              type: Boolean,
              value: false,
              observer: '_disabledChanged',
              reflectToAttribute: true
            }
          };
        }
      }]);
      return DisabledMixinClass;
    }(superclass);
  });
  _exports.DisabledMixin = DisabledMixin;
});
