// @lrnwebcomponents/lrn-assignment/lib/lrn-assignment-button.js
define(["exports", "../../../lit-element/lit-element.js", "../../simple-icon/simple-icon.js", "../../simple-icon/lib/simple-icons.js", "../../simple-icon/lib/simple-icon-button.js", "../../hax-iconset/lib/simple-hax-iconset.js"], function (_exports, _litElement, _simpleIcon, _simpleIcons, _simpleIconButton, _simpleHaxIconset) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnAssignmentButton = void 0;

  function _templateObject4_7e35054029dc11eb8473978bc22b9cf7() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        a {\n          color: inherit;\n          text-decoration: inherit;\n        }\n        button {\n          background: #b0e6f9;\n        }\n        button[complete] {\n          background: #e7ffe7;\n        }\n        simple-icon {\n          margin-left: 8px;\n          opacity: 0.8;\n        }\n      "]);

    _templateObject4_7e35054029dc11eb8473978bc22b9cf7 = function _templateObject4_7e35054029dc11eb8473978bc22b9cf7() {
      return data;
    };

    return data;
  }

  function _templateObject3_7e35054029dc11eb8473978bc22b9cf7() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <button raised complete>\n                ", " <simple-icon icon=\"lrn-icons:done\"></simple-icon>\n              </button>\n            "]);

    _templateObject3_7e35054029dc11eb8473978bc22b9cf7 = function _templateObject3_7e35054029dc11eb8473978bc22b9cf7() {
      return data;
    };

    return data;
  }

  function _templateObject2_7e35054029dc11eb8473978bc22b9cf7() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <button raised open>\n                ", " <simple-icon icon=\"lrn-icons:input\"></simple-icon>\n              </button>\n            "]);

    _templateObject2_7e35054029dc11eb8473978bc22b9cf7 = function _templateObject2_7e35054029dc11eb8473978bc22b9cf7() {
      return data;
    };

    return data;
  }

  function _templateObject_7e35054029dc11eb8473978bc22b9cf7() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <a href=\"", "\">\n        ", "\n        ", "\n      </a>\n    "]);

    _templateObject_7e35054029dc11eb8473978bc22b9cf7 = function _templateObject_7e35054029dc11eb8473978bc22b9cf7() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  var LrnAssignmentButton = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(LrnAssignmentButton, _LitElement);

    var _super = _createSuper(LrnAssignmentButton);

    babelHelpers.createClass(LrnAssignmentButton, [{
      key: "render",

      /**
       * LitElement render
       */
      value: function render() {
        return (0, _litElement.html)(_templateObject_7e35054029dc11eb8473978bc22b9cf7(), this.url, this.open ? (0, _litElement.html)(_templateObject2_7e35054029dc11eb8473978bc22b9cf7(), this.title) : "", this.complete ? (0, _litElement.html)(_templateObject3_7e35054029dc11eb8473978bc22b9cf7(), this.title) : "");
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject4_7e35054029dc11eb8473978bc22b9cf7())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrn-assignment-button";
      }
    }]);

    function LrnAssignmentButton() {
      var _this;

      babelHelpers.classCallCheck(this, LrnAssignmentButton);
      _this = _super.call(this);
      _this.open = false;
      _this.complete = false;
      return _this;
    }

    babelHelpers.createClass(LrnAssignmentButton, null, [{
      key: "properties",
      get: function get() {
        return {
          title: {
            type: String
          },
          url: {
            type: String
          },
          open: {
            type: Boolean
          },
          complete: {
            type: Boolean
          }
        };
      }
    }]);
    return LrnAssignmentButton;
  }(_litElement.LitElement);

  _exports.LrnAssignmentButton = LrnAssignmentButton;
  window.customElements.define(LrnAssignmentButton.tag, LrnAssignmentButton);
});
