// @lrnwebcomponents/haxcms-elements/lib/ui-components/site/site-search.js
define(["exports","require","../../../../../lit-element/lit-element.js"],(function(e,t,r){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.SiteSearch=void 0,t=babelHelpers.interopRequireWildcard(t);
/**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class SiteSearch extends r.LitElement{static get styles(){return[r.css`
        :host {
          display: block;
          font-size: 16px;
          color: var(--site-search-color, #383f45);
          height: 55vh;
          width: 60vh;
        }
        .result {
          display: block;
          background-color: #eeeeee;
          color: #222222;
          border: 1px solid black;
          padding: 16px;
        }
        .result:hover,
        .result:focus {
          background-color: #dddddd;
          color: black;
          color: var(--site-search-link-color-hover, #000000);
          text-decoration: underline;
        }
        .result .title {
          font-size: 28px;
          margin: 0 0 8px 0;
          line-height: 1;
        }
        .result {
          color: var(--site-search-link-color, #444444);
          text-decoration: none;
        }
        .result .link-text {
          font-size: 12px;
          color: var(--site-search-link-text-color, #999999);
          font-style: italic;
          padding-left: 8px;
        }
        #search {
          flex-grow: 2;
          margin-right: 4px;
          --paper-input-container-input-color: var(
            --site-search-text-color,
            #000
          );
          --paper-input-container-shared-input-style_-_color: var(
            --site-search-text-color,
            #000
          );
          --paper-input-container-focus-color: var(
            --site-search-line-color,
            #000
          );
          --paper-input-container-color: var(
            --site-search-placeholder-color,
            #222
          );
          color: var(--site-search-placeholder-color, #222);
        }
      `]}static get tag(){return"site-search"}constructor(){super(),this.search="",this.__results=[],setTimeout(()=>{new Promise((e,r)=>t.default(["../../../../../@polymer/paper-input/paper-input.js"],e,r)),new Promise((e,r)=>t.default(["../../../../../@polymer/iron-icons/iron-icons.js"],e,r)),new Promise((e,r)=>t.default(["../../../../../@polymer/iron-icon/iron-icon.js"],e,r)),new Promise((e,r)=>t.default(["../../../../lunr-search/lunr-search.js"],e,r)),new Promise((e,r)=>t.default(["../../../../simple-datetime/simple-datetime.js"],e,r))},0)}render(){return r.html`
      <paper-input
        id="search"
        always-float-label
        label="Search"
        @value-changed="${this._searchValueChanged}"
      >
        <iron-icon icon="search" slot="prefix"></iron-icon>
      </paper-input>
      ${this.search.length>0?r.html`
            Found ${this.__results.length} results.
          `:r.html``}
      <lunr-search id="lunr"></lunr-search>

      ${this.__results.map(e=>r.html`
          <a class="result" .href="${e.location}">
            <div class="title">
              ${e.title}<span class="link-text" aria-hidden="true"
                >(${e.location})</span
              >
            </div>
            <div class="date">
              <simple-datetime format="M jS" .timestamp="${e.created}" unix
                >${e.created}</simple-datetime
              >
            </div>
            <p>${e.description}..</p>
          </a>
        `)}
    `}_searchValueChanged(e){this.search=e.detail.value}__resultsChanged(e){e.detail.value?this.__results=[...e.detail.value]:this.__results=[]}static get properties(){return{dataSource:{type:String,attribute:"data-source"},search:{type:String},__results:{type:Array}}}firstUpdated(e){this.shadowRoot.querySelector("#lunr").addEventListener("results-changed",this.__resultsChanged.bind(this))}updated(e){e.forEach((e,t)=>{"search"==t&&(this._searchChanged(this[t],e),this.shadowRoot.querySelector("#lunr").search=this[t]),"dataSource"==t&&this[t]&&(this.shadowRoot.querySelector("#lunr").dataSource=this[t])})}_searchChanged(e,t){e.length>=3&&void 0===this.dataSource&&(this.dataSource="lunrSearchIndex.json")}}e.SiteSearch=SiteSearch,window.customElements.define(SiteSearch.tag,SiteSearch)}));
