// @lrnwebcomponents/lrnsys-render-html/lrnsys-render-html.js
define(["exports", "../../lit/index.js"], function (_exports, _index) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnsysRenderHtml = void 0;

  var _templateObject_87d8c290a15311ecad533bc0497e45a3, _templateObject2_87d8c290a15311ecad533bc0497e45a3;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
  `lrnsys-render-html`
  A legacy element which just directly renders HTML.
  WARNING: DO NOT USE THIS UNLESS YOU KNOW WHAT YOU ARE DOING!
  
  This element is meant to render html from a source that has already been sanitized.
  Polymer will, by design, not render html for security reasons. This element gets around
  that. Do not render raw user input through this element! This would allow XSS attacks for
  your users.
  
  * @demo demo/index.html
  */
  var LrnsysRenderHtml = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(LrnsysRenderHtml, _LitElement);

    var _super = _createSuper(LrnsysRenderHtml);

    function LrnsysRenderHtml() {
      babelHelpers.classCallCheck(this, LrnsysRenderHtml);
      return _super.apply(this, arguments);
    }

    babelHelpers.createClass(LrnsysRenderHtml, [{
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_87d8c290a15311ecad533bc0497e45a3 || (_templateObject_87d8c290a15311ecad533bc0497e45a3 = babelHelpers.taggedTemplateLiteral([" <div id=\"container\"></div> "])));
      }
    }, {
      key: "firstUpdated",
      value: function firstUpdated() {
        this.shadowRoot.querySelector("#container").innerHTML = this.html;
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "html" && _this.shadowRoot) {
            _this.shadowRoot.querySelector("#container").innerHTML = _this[propName];
          }
        });
      }
    }], [{
      key: "styles",
      get:
      /**
       * LitElement constructable styles enhancement
       */
      function get() {
        return [(0, _index.css)(_templateObject2_87d8c290a15311ecad533bc0497e45a3 || (_templateObject2_87d8c290a15311ecad533bc0497e45a3 = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n      "])))];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrnsys-render-html";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * String to render as HTML directly
           */
          html: {
            type: String
          }
        };
      }
    }]);
    return LrnsysRenderHtml;
  }(_index.LitElement);

  _exports.LrnsysRenderHtml = LrnsysRenderHtml;
  window.customElements.define(LrnsysRenderHtml.tag, LrnsysRenderHtml);
});
