diff --git a/h5p_tincan_bridge.module b/h5p_tincan_bridge.module
index 287a9b7..d8d9035 100644
--- a/h5p_tincan_bridge.module
+++ b/h5p_tincan_bridge.module
@@ -46,11 +46,8 @@ function h5p_tincan_bridge_view() {
 
     h5p_tincan_bridge_tincanapi_add_og_context(node_load($nid), $data);
 
-    $hook = 'tincanapi_ajax_data_alter';
-    foreach (module_implements($hook) as $module) {
-      $function = $module . '_' . $hook;
-      $function('h5p_tincan_bridge', $data, $_POST);
-    }
+    // allow other projects to jump in here
+    drupal_alter('tincanapi_ajax_data', $data, $_POST);
 
     tincanapi_send("statements", "POST", $data);
   }
