// @haxtheweb/sorting-question/sorting-question.js
import{html as t,css as e,nothing as s}from"../../lit/index.js";import{QuestionElement as i}from"../multiple-choice/lib/QuestionElement.js";import"../simple-icon/lib/simple-icons.js";import"../simple-toolbar/lib/simple-toolbar-button.js";import"../simple-toast/simple-toast.js";import"../grid-plate/grid-plate.js";import"./lib/sorting-option.js";export class SortingQuestion extends i{static get tag(){return"sorting-question"}getOptions(t=""){if(this.shadowRoot)return t?this.shadowRoot.querySelectorAll(`sorting-option[${t}]`):this.shadowRoot.querySelectorAll("sorting-option")}constructor(){super(),this.randomize=!0,this.numberCorrect=0,this.hasHint=this.querySelector('[slot="hint"]'),this.hasContent=this.querySelector('[slot="content"]'),this.hasFeedbackCorrect=this.querySelector('[slot="feedbackCorrect"]'),this.hasFeedbackIncorrect=this.querySelector('[slot="feedbackIncorrect"]'),this.hasEvidence=this.querySelector('[slot="evidence"]'),this.correctText="That is correct, Great job!",this.correctIcon="icons:thumb-up",this.incorrectIcon="icons:thumb-down",this.quizName="default",this.question="Put the following in order",this.t={numCorrectLeft:"You have",numCorrectRight:"correct",checkAnswer:"Check answer",tryAgain:"Try again"},this.checkLabel="Check answer",this.resetLabel="Try again",this.registerLocalization({context:this,localesPath:new URL("../locales/sorting-question.es.json",import.meta.url).href+"/../",locales:["he","ja","es"]})}checkAnswer(){this.showAnswer=!0}resetAnswer(){if(this.getOptions("incorrect")){const t=this.getOptions("incorrect")[0];t&&setTimeout((()=>{t.shadowRoot.querySelector("#downArrow").shadowRoot.querySelector("button").focus()}),0)}if(this.showAnswer=!1,globalThis.dispatchEvent(new CustomEvent("simple-toast-hide",{bubbles:!0,composed:!0,cancelable:!1,detail:!1})),this.getOptions().forEach((t=>{t.disabled=!1,t.incorrect=null,t.correct=null})),this.numberCorrect===this.answers.length){this.displayedAnswers=[],this.answers.forEach((t=>{t.userGuess=""}));const t=JSON.parse(JSON.stringify(this.answers));this.answers=[...t]}this.numberCorrect=0}static get properties(){return{...super.properties,numberCorrect:{type:Number}}}updated(t){super.updated&&super.updated(t),t.forEach(((t,e)=>{if("showAnswer"===e)if(this.showAnswer){var s=0;let t=this.getOptions();for(var i=0;i<this.answers.length;i++)t[i].innerText===this.answers[i].label&&i===this.answers[i].order?(s+=1,t[i].correct=!0,t[i].incorrect=null):(t[i].correct=null,t[i].incorrect=!0),t[i].disabled=!0;this.numberCorrect=s;let e=document.createElement("simple-icon-lite"),r={},o="simple-toast-show";globalThis.HAXCMSToast&&(o="haxcms-toast-show",e.style.setProperty("--simple-icon-height","40px"),e.style.setProperty("--simple-icon-width","40px"),e.style.height="150px",e.style.marginLeft="8px"),this.shadowRoot.querySelector("#feedback").focus();let n=this.numberCorrect===this.answers.length;n?(this.playSound("success"),this.__toastColor="green",this.__toastIcon=this.correctIcon,this.__toastText=this.correctText,this.makeItRain(),r.hat="party"):(this.playSound("error"),this.__toastColor="red",this.__toastIcon=this.incorrectIcon,this.__toastText=`${this.t.numCorrectLeft} ${this.numberCorrect} of ${this.answers.length} ${this.t.numCorrectRight}`,r.fire=!0),e.icon=this.__toastIcon,e.style.marginLeft="16px",e.accentColor=this.__toastColor,e.dark=!0,globalThis.dispatchEvent(new CustomEvent(o,{bubbles:!0,composed:!0,cancelable:!0,detail:{text:this.__toastText,accentColor:this.__toastColor,duration:3e3,slot:e,...r}}));let a={activityDisplay:"answered",objectName:this.quizName,resultSuccess:n};this.dispatchEvent(new CustomEvent("user-engagement",{bubbles:!0,composed:!0,cancelable:!1,detail:a}))}else this.getOptions().forEach((t=>{t.disabled=!1}));"disabled"===e&&this.shadowRoot&&(this.disabled?this.getOptions().forEach((t=>{t.disabled=!0,t.correct=null,t.incorrect=null})):this.getOptions().forEach((t=>{t.disabled=!1})))}))}static get styles(){return[super.styles,e`
        :host {
          display: block;
          min-width: 160px;
          padding: var(--ddd-spacing-8);
          border: var(--ddd-border-md);
          border-radius: var(--ddd-radius-xs);
          transition: all 0.3s ease-in-out;
        }
        :host(:focus),
        :host(:focus-within),
        :host(:hover) {
          border-color: var(--simple-colors-default-theme-accent-12);
        }

        simple-toolbar-button {
          font-size: var(--ddd-font-size-xs);
          font-family: var(--ddd-font-navigation);
          transition: all 0.3s ease-in-out;
          border: none;
          border-radius: var(--ddd-radius-xs);
        }
        simple-toolbar-button {
          background-color: light-dark(
            var(--ddd-theme-default-link),
            var(--ddd-theme-default-linkLight)
          );
          color: light-dark(white, black);
        }
        simple-toolbar-button[disabled] {
          background-color: light-dark(
            var(--ddd-theme-default-limestoneLight),
            var(--ddd-theme-default-slateGray)
          );
          color: light-dark(black, white);
          opacity: 0.5;
        }
        :host simple-toolbar-button:hover::part(button),
        :host simple-toolbar-button:focus::part(button),
        :host simple-toolbar-button:focus-within::part(button),
        :host simple-toolbar-button:active::part(button) {
          cursor: pointer;
          box-shadow: var(--ddd-boxShadow-sm);
          border-color: black;
        }
        simple-toolbar-button::part(button) {
          border: var(--ddd-border-sm);
          border-radius: var(--ddd-radius-xs);
          padding: var(--ddd-spacing-2);
        }
        simple-toolbar-button::part(label) {
          font-size: var(--ddd-font-size-s);
          font-family: var(--ddd-font-navigation);
          padding: 0;
          margin: 0;
        }
        #check {
          margin-right: var(--ddd-spacing-4);
        }
        .options {
          margin-bottom: var(--ddd-spacing-6);
          border-radius: var(--ddd-radius-xs);
          border: var(--ddd-border-xs);
        }

        #buttons {
          display: flex;
          justify-content: center;
          align-items: center;
        }
        h3 {
          padding: 0;
          margin: 0 0 var(--ddd-spacing-8) 0;
          font-family: var(--ddd-font-navigation);
        }
        .feedback {
          margin: var(--ddd-spacing-3) 0;
          font-size: var(--ddd-font-size-m);
          font-weight: bold;
          text-align: center;
        }
        sorting-option img {
          border: var(--ddd-border-sm);
          border-radius: var(--ddd-radius-sm);
          box-shadow: var(--ddd-boxShadow-sm);
          margin-right: var(--ddd-spacing-4);
        }
      `]}renderInteraction(){return t`
      <fieldset class="options">
        ${this.displayedAnswers.map((e=>t`<sorting-option ?disabled="${this.disabled}"
              >${e.image?t`<img src="${e.image}" alt="${e.alt}" />`:s}${e.label}</sorting-option
            >`))}
      </fieldset>
    `}render(){return t`
      <meta property="oer:assessing" content="${this.relatedResource}" />
      <confetti-container id="confetti">
        <grid-plate layout="1-1">
          <div slot="col-1">
            <h3 property="oer:name">${this.question}</h3>
            ${this.renderInteraction()}
            ${this.hideButtons?"":this.renderButtons()}
          </div>
          <div slot="col-2">
            <details
              tabindex="${this.disabled?"-1":""}"
              ?disabled="${this.disabled}"
              ?open="${!this.hasContent}"
              id="directions"
            >
              <summary>Directions</summary>
              <div>${this.renderDirections()}</div>
            </details>
            ${this.hasContent?t` <details
                  tabindex="${this.disabled?"-1":""}"
                  ?disabled="${this.disabled}"
                  ?open="${!this.showAnswer}"
                  id="related"
                >
                  <summary>Related content</summary>
                  <div>
                    <slot name="content"></slot>
                  </div>
                </details>`:""}
            <details
              tabindex="${!this.showAnswer||this.disabled?"-1":""}"
              ?disabled="${!this.showAnswer||this.disabled}"
              ?open="${this.showAnswer}"
            >
              <summary id="feedback">Feedback</summary>
              <div>${this.renderFeedback()}</div>
            </details>
          </div>
        </grid-plate>
      </confetti-container>
    `}inactiveCase(){return!0}renderButtons(){return t`
      <div id="buttons">
        <simple-toolbar-button
          id="check"
          ?disabled="${this.disabled||!this.inactiveCase()||this.showAnswer}"
          @click="${this.checkAnswer}"
          label="${this.checkLabel}"
        >
        </simple-toolbar-button>
        <simple-toolbar-button
          id="reset"
          ?disabled="${this.disabled||!this.inactiveCase()||this.inactiveCase()&&!this.showAnswer}"
          @click="${this.resetAnswer}"
          label="${this.resetLabel}"
        >
        </simple-toolbar-button>
      </div>
    `}renderDirections(){return t`<p>
      Place the items in the correct order either by clicking the up and down
      arrows or drag and drop. When you believe you have them in the correct
      order, test your answer by selecting
      <strong>${this.t.checkAnswer}</strong>. You will get feedback just below
      here indicating correctness of your answer.
    </p>`}renderFeedback(){return t`
      ${this.showAnswer&&this.numberCorrect!==this.answers.length?t` <p class="feedback">
              ${this.t.numCorrectLeft}
              ${this.numberCorrect}/${this.answers.length}
              ${this.t.numCorrectRight}
            </p>
            ${this.hasFeedbackIncorrect?t`<slot name="feedbackIncorrect"></slot>`:""}`:""}
      ${this.showAnswer&&this.numberCorrect===this.answers.length?t` <p class="feedback">${this.correctText}</p>
            ${this.hasFeedbackCorrect?t`<slot name="feedbackCorrect"></slot>`:""}`:""}
      ${this.hasHint&&this.showAnswer&&this.numberCorrect!==this.answers.length?t`
            <h4>Need a hint?</h4>
            <div>
              <slot name="hint"></slot>
            </div>
          `:""}
      ${this.hasEvidence&&this.showAnswer&&this.numberCorrect===this.answers.length?t`
            <h4>Evidence</h4>
            <div>
              <slot name="evidence"></slot>
            </div>
          `:""}
      <simple-toolbar-button
        ?disabled="${this.disabled||!this.showAnswer}"
        @click="${this.resetAnswer}"
        label="${this.t.tryAgain}"
      >
      </simple-toolbar-button>
    `}static get haxProperties(){return new URL(`./lib/${this.tag}.haxProperties.json`,import.meta.url).href}}globalThis.customElements.define(SortingQuestion.tag,SortingQuestion);
