// @lrnwebcomponents/haxcms-elements/lib/ui-components/active-item/site-active-title.js
define(["exports", "../../../../../@polymer/polymer/polymer-element.js", "../../core/haxcms-site-store.js", "../../../../../mobx/lib/mobx.module.js"], function (_exports, _polymerElement, _haxcmsSiteStore, _mobxModule) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SiteActiveTitle = void 0;

  /**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `site-active-title`
   * `Title of the active page in the site`
   *
   * @customElement
   * @polymer
   * @demo demo/index.html
   */
  class SiteActiveTitle extends _polymerElement.PolymerElement {
    /**
     * Store the tag name to make it easier to obtain directly.
     * @notice function name must be here for tooling to operate correctly
     */
    static get tag() {
      return "site-active-title";
    } // render function


    static get template() {
      return _polymerElement.html`
      <style>
        :host {
          display: block;
          transition: 0.2s opacity linear;
          opacity: 1;
        }
        :host([edit-mode]) {
          opacity: 0.2;
        }
        h2 {
          text-rendering: optimizelegibility;
          font-family: sans-serif;
          color: var(--site-active-title-color, #383f45);
          @apply --site-active-title-heading;
        }
      </style>
      <h2>[[__title]]</h2>
    `;
    }
    /**
     * Props
     */


    static get properties() {
      return {
        /**
         * Turn off the title fallback to avoid empty titles
         */
        noFallback: {
          type: Boolean,
          value: false
        },

        /**
         * How we should obtain the parent who's children we should show
         * options are active, parent, or ancestor
         */
        dynamicMethodology: {
          type: String,
          value: "active"
        },
        __title: {
          type: String,
          computed: "_makeTitle(dynamicMethodology, activeTitle, parentTitle, ancestorTitle)"
        },
        editMode: {
          type: Boolean,
          reflectToAttribute: true
        }
      };
    }

    _makeTitle(dynamicMethodology, activeTitle, parentTitle, ancestorTitle) {
      switch (dynamicMethodology) {
        case "above":
          if (parentTitle === "" && !this.noFallback) {
            return activeTitle;
          }

          return parentTitle;
          break;

        case "ancestor":
          if (ancestorTitle === "" && !this.noFallback) {
            return activeTitle;
          }

          return ancestorTitle;
          break;

        default:
          return activeTitle;
          break;
      }
    }

    connectedCallback() {
      super.connectedCallback();
      this.__disposer = [];
      (0, _mobxModule.autorun)(reaction => {
        this.editMode = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.editMode);

        this.__disposer.push(reaction);
      });
      (0, _mobxModule.autorun)(reaction => {
        this.activeTitle = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.activeTitle);

        this.__disposer.push(reaction);
      });
      (0, _mobxModule.autorun)(reaction => {
        this.ancestorTitle = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.ancestorTitle);

        this.__disposer.push(reaction);
      });
      (0, _mobxModule.autorun)(reaction => {
        this.parentTitle = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.parentTitle);

        this.__disposer.push(reaction);
      });
    }

    disconnectedCallback() {
      // clean up state
      for (var i in this.__disposer) {
        this.__disposer[i].dispose();
      }

      super.disconnectedCallback();
    }

  }

  _exports.SiteActiveTitle = SiteActiveTitle;
  window.customElements.define(SiteActiveTitle.tag, SiteActiveTitle);
});
