// @lrnwebcomponents/haxcms-elements/lib/ui-components/active-item/site-active-fields.js
define(["exports", "../../../../../lit-element/lit-element.js", "../../core/haxcms-site-store.js", "../../../../../mobx/lib/mobx.module.js"], function (_exports, _litElement, _haxcmsSiteStore, _mobxModule) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SiteActiveFields = void 0;

  /**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `site-active-fields`
   * `Title of the active page in the site`
   *
  
   * @demo demo/index.html
   */
  class SiteActiveFields extends _litElement.LitElement {
    /**
     * Store the tag name to make it easier to obtain directly.
     */
    static get tag() {
      return "site-active-fields";
    }
    /**
     * LitElement properties decorator
     */


    static get properties() {
      return {
        fields: {
          type: Object
        }
      };
    }
    /**
     * LitElement life cycle callback
     */


    updated(changedProperties) {
      changedProperties.forEach((oldValue, propName) => {
        if (propName == "fields") {
          this.dispatchEvent(new CustomEvent("fields-changed", {
            detail: this[propName]
          }));
        }
      });
    }
    /**
     * VanillaJS life cycle callback
     */


    connectedCallback() {
      super.connectedCallback();
      this.__disposer = (0, _mobxModule.autorun)(() => {
        this.fields = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.activeItemFields);
      });
    }
    /**
     * VanillaJS life cycle callback
     */


    disconnectedCallback() {
      this.__disposer();

      super.disconnectedCallback();
    }

  }

  _exports.SiteActiveFields = SiteActiveFields;
  window.customElements.define(SiteActiveFields.tag, SiteActiveFields);
});
