// @haxtheweb/page-break/page-break.js
/**
 * Copyright 2021 The Pennsylvania State University
 * @license Apache-2.0, see License.md for full text.
 */
import{html as e,css as t,LitElement as i}from"../../lit/index.js";import{SchemaBehaviors as a}from"../schema-behaviors/schema-behaviors.js";import{IntersectionObserverMixin as o}from"../intersection-element/lib/IntersectionObserverMixin.js";import{I18NMixin as s}from"../i18n-manager/lib/I18NMixin.js";import"../simple-icon/lib/simple-icon-button-lite.js";import"../simple-icon/lib/simple-icon-lite.js";import"../simple-icon/lib/simple-icons.js";import"../haxcms-elements/lib/core/micros/haxcms-button-add.js";import{pageBreakManager as l}from"./lib/page-break-manager.js";import{DDDExtra as n}from"../d-d-d/lib/DDDStyles.js";import{store as r}from"../haxcms-elements/lib/core/haxcms-site-store.js";import{autorun as d,toJS as h}from"../../mobx/dist/mobx.esm.js";export class PageBreak extends(o(s(a(i)))){static get tag(){return"page-break"}constructor(){super(),this.relatedItems=null,this.icon=null,this.accentColor=null,this.entityType="page",this.status="",this.author=null,this.linkUrl=null,this.linkTarget="_self",this.t={newPage:"New page",pageBreak:"Page break",selectToEditPageDetails:"Select to edit Page details",noParent:"No parent",toggleLock:"Toggle lock",togglePublished:"Toggle published",linkMessage:"Users will be redirected to:",linkOpensInNewWindow:"Opens in new window",linkOpensInSameWindow:"Opens in same window",pageActions:"Page Actions",editPage:"Edit page",modifyPageTitle:"Modify page title",modifyPageIcon:"Modify page icon",editMedia:"Edit Media",editTags:"Edit tags",lock:"Lock",unlock:"Unlock",publish:"Publish",unpublish:"Unpublish",savePage:"Save page",saveAndEdit:"Save & Edit",cancel:"Cancel"},this.registerLocalization({context:this,localesPath:new URL("./locales/page-break.es.json",import.meta.url).href+"/../"}),this.description=null,this.hideInMenu=!1,this.noderefs=[],this.developerTheme=null,this.tags=null,this.title=this.t.newPage,this.pageType=null,this.slug="",this.published=!1,this.image=null,this.target=null,this.locked=!1,this.order=null,this.depth=0,this.itemId=null,this._haxState=!1,this.IORemoveOnVisible=!1,this.IODelay=250,this.remoteHeadingobserver=new MutationObserver((()=>{this.title!=this.target.innerText&&(this.__moUpdate=!0,this.title=this.target.innerText)})),this.iconType="editor:format-page-break",this.breakType="node",this.isLoggedIn=!1,this._editingUILoaded=!1,this.__disposer=[],d((e=>{this.isLoggedIn=h(r.isLoggedIn),this.__disposer.push(e)}))}static get properties(){return{...super.properties||{},iconType:{type:String},noderefs:{type:Array,attribute:!1},relatedItems:{type:String,attribute:"related-items"},icon:{type:String},accentColor:{type:String,attribute:"accent-color"},entityType:{type:String,attribute:"entity-type"},description:{type:String},order:{type:Number},hideInMenu:{type:Boolean,reflect:!0,attribute:"hide-in-menu"},tags:{type:String,reflect:!0},developerTheme:{type:String,attribute:"developer-theme"},title:{type:String,reflect:!0},slug:{type:String},image:{type:String},parent:{type:String,reflect:!0},published:{type:Boolean,reflect:!0},locked:{type:Boolean,reflect:!0},depth:{type:Number,reflect:!0},itemId:{type:String,attribute:"item-id",reflect:!0},breakType:{type:String,attribute:"break-type"},status:{type:String},pageType:{type:String,attribute:"page-type"},author:{type:String},linkUrl:{type:String,attribute:"link-url"},linkTarget:{type:String,attribute:"link-target"},_haxState:{type:Boolean},platformHidden:{type:Boolean,attribute:"platform-hidden",reflect:!0},isLoggedIn:{type:Boolean,reflect:!0,attribute:"is-logged-in"}}}async _ensureEditingUI(){this._editingUILoaded||(await Promise.all([import("../simple-toolbar/lib/simple-toolbar-button.js"),import("../simple-fields/lib/simple-context-menu.js")]),this._editingUILoaded=!0,this.requestUpdate())}connectedCallback(){super.connectedCallback(),"node"===this.breakType&&(this.nextElementSibling&&this.nextElementSibling.tagName&&["H1","H2","H3","H4","H5","H6"].includes(this.nextElementSibling.tagName)?(this.title=this.nextElementSibling.innerText,this.target=this.nextElementSibling,this.setupTargetData(this.target)):setTimeout((()=>{if(null===this.target)if(this.nextElementSibling&&this.nextElementSibling.tagName&&["H1","H2","H3","H4","H5","H6"].includes(this.nextElementSibling.tagName))this.title=this.nextElementSibling.innerText,this.target=this.nextElementSibling,this.setupTargetData(this.target);else{let e=0===this.depth?"h2":`h${this.depth+2}`,t=globalThis.document.createElement(e);t.setAttribute("data-original-level","H2"),t.innerText=this.title,this.parentNode.insertBefore(t,this.nextElementSibling),setTimeout((()=>{this.setupTargetData(this.nextElementSibling)}),100)}}),0)),globalThis.dispatchEvent(new CustomEvent("page-break-registration",{composed:!0,bubbles:!0,cancelable:!0,detail:{value:this,action:"add"}})),globalThis.dispatchEvent(new CustomEvent("page-break-change",{composed:!0,bubbles:!0,cancelable:!0,detail:{value:this}}))}disconnectedCallback(){for(var e in globalThis.dispatchEvent(new CustomEvent("page-break-registration",{detail:{value:this,action:"remove"}})),globalThis.dispatchEvent(new CustomEvent("page-break-change",{composed:!0,bubbles:!0,cancelable:!0,detail:{value:this}})),this.remoteHeadingobserver.disconnect(),this.__disposer)this.__disposer[e].dispose();super.disconnectedCallback()}setupTargetData(e){this.target&&this.remoteHeadingobserver.disconnect(),this.target=e,this.target&&this.target instanceof Node?(this._haxSibling=this,this.remoteHeadingobserver.observe(this.target,{characterData:!0,childList:!0,subtree:!0})):console.warn("page-break: setupTargetData called with invalid target",this.target)}getCurrentUser(){return globalThis.HAXCMS&&globalThis.HAXCMS.requestAvailability().store.userData&&globalThis.HAXCMS.requestAvailability().store.userData.userName?globalThis.HAXCMS.requestAvailability().store.userData.userName:globalThis.store&&globalThis.store.user&&globalThis.store.user.name?globalThis.store.user.name:null}updated(e){super.updated&&super.updated(e),e.forEach(((e,t)=>{if("isLoggedIn"===t&&this.isLoggedIn&&!this._editingUILoaded&&this._ensureEditingUI(),["title","description","tags","published","locked"].includes(t)&&void 0!==e){const e=this.getCurrentUser();e&&e!==this.author&&(this.author=e)}if("noderefs"===t){let e=[];for(let t=0;t<this.noderefs.length;t++)e.push(this.noderefs[t].node);this.relatedItems=e.join(",")}"schemaResourceID"===t&&null==this.itemId&&void 0!==e&&(this.itemId=this.schemaResourceID.replace("#","item-")),this.locked&&"locked"===t?l.elementsBetween(this).forEach((e=>{e.setAttribute("data-hax-lock","data-hax-lock")})):!this.locked&&"locked"===t&&e&&l.elementsBetween(this).forEach((e=>{e.removeAttribute("data-hax-lock")})),this._ceMenu&&["locked","parent","published"].includes(t)&&this._updateHAXCEMenu(),["title","parent","slug"].includes(t)&&globalThis.dispatchEvent(new CustomEvent("page-break-change",{composed:!0,bubbles:!0,cancelable:!0,detail:{value:this}})),"node"===this.breakType&&this.target&&("title"===t&&this[t]&&(this.__moUpdate?this.__moUpdate=!1:this.title!=this.target.innerText&&(this.target.innerText=this.title)),!this._haxState&&"depth"===t&&this.depth>=0&&l.elementsBetween(this,"page-break","h1,h2,h3,h4,h5,h6").forEach((e=>{let t=(e.getAttribute("data-original-level")?new Number(e.getAttribute("data-original-level").replace("H","")):new Number(e.tagName.replace("H","")))+this.depth;t=t>6?6:t;const i=globalThis.document.createElement(`h${t}`);i.setAttribute("data-original-level",e.tagName);for(var a=0,o=e.attributes.length;a<o;++a)i.setAttribute(e.attributes.item(a).nodeName,e.attributes.item(a).nodeValue);i.innerHTML=e.innerHTML,this.setupTargetData(i),e.parentNode.replaceChild(i,e)})),"_haxState"===t&&void 0!==e&&(this._haxState?l.elementsBetween(this,"page-break","h1,h2,h3,h4,h5,h6").forEach((e=>{if(e.getAttribute("data-original-level")){let a=new Number(e.getAttribute("data-original-level").replace("H",""));const o=globalThis.document.createElement(`h${a}`);for(var t=0,i=e.attributes.length;t<i;++t)o.setAttribute(e.attributes.item(t).nodeName,e.attributes.item(t).nodeValue);o.innerHTML=e.innerHTML,e.parentNode.replaceChild(o,e),this.setupTargetData(o)}})):l.elementsBetween(this,"page-break","h1,h2,h3,h4,h5,h6").forEach((e=>{let t=(e.getAttribute("data-original-level")?new Number(e.getAttribute("data-original-level").replace("H","")):new Number(e.tagName.replace("H","")))+this.depth;t=t>6?6:t;const i=globalThis.document.createElement(`h${t}`);i.setAttribute("data-original-level",e.tagName);for(var a=0,o=e.attributes.length;a<o;++a)i.setAttribute(e.attributes.item(a).nodeName,e.attributes.item(a).nodeValue);i.innerHTML=e.innerHTML,e.parentNode.replaceChild(i,e),this.setupTargetData(i)})))),"breakType"===t&&("node"===this[t]?this.iconType="editor:format-page-break":this.iconType="hax:page-details")}))}static get styles(){return[n,t`
        :host {
          display: block;
          position: relative;
          height: 0;
        }
        /* Platform configuration can force page-break to always be hidden */
        :host([platform-hidden]) {
          display: none !important;
          opacity: 0 !important;
          pointer-events: none;
        }
        :host([data-hax-ray]) {
          display: block;
          margin: var(--ddd-spacing-1) 0 var(--ddd-spacing-20) 0;
          padding: var(--ddd-spacing-4);
          border: 2px solid var(--ddd-theme-default-limestoneGray);
          border-radius: var(--ddd-radius-xs);
          background-color: var(--ddd-theme-default-limestoneMaxLight);
          position: relative;
          opacity: 0.9;
          transition:
            opacity 0.2s ease-in-out,
            border-color 0.2s ease-in-out,
            background-color 0.2s ease-in-out;
        }
        /* Increase bottom margin when link URL is present to prevent clipping */
        :host([data-hax-ray][link-url]) {
          margin-bottom: var(--ddd-spacing-24);
        }
        :host([data-hax-ray]:hover) {
          opacity: 1;
          background-color: var(--ddd-theme-default-white);
        }
        :host([data-hax-active]) {
          opacity: 1;
          border-color: var(--ddd-theme-default-skyBlue);
          background-color: var(--ddd-theme-default-white);
          box-shadow: var(--ddd-boxShadow-sm);
        }
        :host([data-hax-ray]) .text {
          display: block;
        }
        .link-info {
          display: none;
          background-color: light-dark(
            var(--ddd-theme-default-limestoneLight),
            var(--ddd-theme-default-slateGray)
          );
          border: var(--ddd-border-xs);
          border-color: light-dark(
            var(--ddd-theme-default-limestoneGray),
            var(--ddd-theme-default-coalyGray)
          );
          border-radius: var(--ddd-radius-xs);
          padding: var(--ddd-spacing-3);
          margin-top: var(--ddd-spacing-2);
          font-size: var(--ddd-font-size-xs);
          color: light-dark(
            var(--ddd-theme-default-coalyGray),
            var(--ddd-theme-default-limestoneLight)
          );
        }
        :host([data-hax-ray]) .link-info {
          display: block;
        }
        .link-url {
          font-family: monospace;
          background-color: light-dark(
            var(--ddd-theme-default-white),
            var(--ddd-theme-default-coalyGray)
          );
          padding: var(--ddd-spacing-1) var(--ddd-spacing-2);
          border-radius: var(--ddd-radius-xs);
          border: var(--ddd-border-xs) solid
            light-dark(
              var(--ddd-theme-default-limestoneGray),
              var(--ddd-theme-default-slateGray)
            );
          display: inline-block;
          margin: var(--ddd-spacing-1) 0;
          word-break: break-all;
          text-decoration: none;
          color: light-dark(
            var(--ddd-theme-default-coalyGray),
            var(--ddd-theme-default-limestoneLight)
          );
          transition: all 0.2s ease;
        }
        .link-url:hover {
          background-color: light-dark(
            var(--ddd-theme-default-limestoneMaxLight),
            var(--ddd-theme-default-slateMaxLight)
          );
          color: var(--ddd-theme-default-skyBlue);
          border-color: var(--ddd-theme-default-skyBlue);
        }
        .text {
          display: none;
          font-weight: var(--ddd-font-weight-medium);
          color: var(--ddd-theme-default-coalyGray);
          background-color: var(--ddd-theme-default-skyBlue);
          font-size: var(--ddd-font-size-4xs);
          padding: var(--ddd-spacing-1) var(--ddd-spacing-4);
          position: absolute;
          top: 0;
          left: 0;
          z-index: 10;
          color: var(--ddd-theme-default-white);
          height: 24px;
          line-height: 24px;
        }
        simple-toolbar-button.menu-button,
        simple-toolbar-button.save-button,
        simple-toolbar-button.save-edit-button,
        simple-toolbar-button.cancel-button {
          position: absolute;
          top: 0;
          right: 0;
          --simple-toolbar-button-height: 20px;
          --simple-toolbar-button-width: 20px;
          padding: var(--ddd-spacing-1);
          border: var(--ddd-border-xs);
          box-shadow: var(--ddd-boxShadow-sm);
          --simple-toolbar-button-hover-border-color: transparent;
        }
        simple-toolbar-button.menu-button {
          top: -8px;
          border: var(--ddd-border-xs);
          border-radius: var(--ddd-radius-circle);
          box-shadow: var(--ddd-boxShadow-sm);
          --simple-toolbar-button-height: 12px;
          --simple-toolbar-button-width: 12px;
          color: var(--ddd-theme-default-white);
          background-color: var(--ddd-theme-default-skyBlue);
        }
        simple-toolbar-button.save-edit-button {
          right: 32px;
        }

        simple-toolbar-button.save-button {
          right: 64px;
        }

        simple-toolbar-button.save-button,
        simple-toolbar-button.save-edit-button,
        simple-toolbar-button.cancel-button {
          --simple-toolbar-button-height: 20px;
          --simple-toolbar-button-width: 20px;
          background-color: var(--ddd-theme-default-skyBlue);
          color: var(--ddd-theme-default-white);
          border-color: var(--ddd-theme-default-limestoneGray);
          border-width: 1px;
        }

        simple-toolbar-button.cancel-button {
          background-color: var(--ddd-theme-default-discoveryCoral);
        }

        simple-toolbar-button.menu-item {
          --simple-toolbar-button-justify: flex-start;
          cursor: pointer;
          --simple-icon-height: 16px;
          --simple-icon-width: 16px;
        }
        simple-toolbar-button.menu-item simple-icon-lite {
          --simple-icon-height: 16px;
          --simple-icon-width: 16px;
        }
        simple-toolbar-button.menu-item-delete {
          --simple-toolbar-button-justify: flex-start;
          --simple-toolbar-button-hover-border-color: transparent;
          cursor: pointer;
          --simple-icon-height: 16px;
          --simple-icon-width: 16px;
          border-top: var(--ddd-border-sm) solid var(--ddd-theme-default-limestoneGray);
          margin-top: var(--ddd-spacing-2);
          padding-top: var(--ddd-spacing-2);
        }
        simple-toolbar-button.menu-item-delete:focus,
        simple-toolbar-button.menu-item-delete:hover {
          color: black;
          background-color: var(--ddd-theme-default-discoveryCoral);
        }
      `]}firstUpdated(e){if(super.firstUpdated&&super.firstUpdated(e),this.relatedItems){const e=this.relatedItems.split(",");for(let t=0;t<e.length;t++)this.noderefs.push({node:e[t]})}null!=this.itemId&&(this.schemaResourceID=this.itemId)}render(){return e`
      <a .href="${this.slug}" .name="#${this.itemId}" class="sr-only">${this.title}</a>
      <div class="text">
        <simple-icon-lite icon="${this.iconType}"></simple-icon-lite>${this.t.selectToEditPageDetails}
      </div>
      ${this.linkUrl?e`<div class="link-info">
            <div>
              <strong>${this.t.linkMessage}</strong>
            </div>
            <a
              class="link-url"
              href="${this.linkUrl}"
              target="_blank"
              rel="noopener noreferrer"
            >
              ${this.linkUrl}
            </a>
            <div>
              ${"_blank"===this.linkTarget?this.t.linkOpensInNewWindow:this.t.linkOpensInSameWindow}
            </div>
          </div>`:""}
      ${this.isLoggedIn&&!this._haxState?e`
            <simple-toolbar-button
              class="menu-button"
              icon="icons:create"
              label="${this.t.pageActions}"
              @click="${this._toggleMenu}"
            ></simple-toolbar-button>
            <simple-context-menu id="menu" title="${this.t.pageActions}">
              <simple-toolbar-button
                class="menu-item"
                icon="hax:page-edit"
                label="${this.t.editPage}"
                show-text-label
                @click="${this._editPage}"
                ?disabled="${this.locked}"
                autofocus
              ></simple-toolbar-button>
              <simple-toolbar-button
                class="menu-item"
                icon="editor:title"
                label="${this.t.modifyPageTitle}"
                show-text-label
                @click="${this._editTitle}"
                ?disabled="${this.locked}"
              ></simple-toolbar-button>
              <simple-toolbar-button
                class="menu-item"
                icon="hax:hax2022"
                label="${this.t.modifyPageIcon}"
                show-text-label
                @click="${this._editIcon}"
                ?disabled="${this.locked}"
              ></simple-toolbar-button>
              <simple-toolbar-button
                class="menu-item"
                icon="image:photo-library"
                label="${this.t.editMedia}"
                show-text-label
                @click="${this._editMedia}"
                ?disabled="${this.locked}"
              ></simple-toolbar-button>
              <simple-toolbar-button
                class="menu-item"
                icon="icons:label"
                label="${this.t.editTags}"
                show-text-label
                @click="${this._editTags}"
                ?disabled="${this.locked}"
              ></simple-toolbar-button>
              <simple-toolbar-button
                class="menu-item"
                icon="${this.published?"icons:visibility-off":"icons:visibility"}"
                label="${this.published?this.t.unpublish:this.t.publish}"
                show-text-label
                @click="${this._togglePublished}"
                ?disabled="${this.locked}"
              ></simple-toolbar-button>
              <simple-toolbar-button
                class="menu-item"
                icon="${this.locked?"icons:lock":"icons:lock-open"}"
                label="${this.locked?this.t.unlock:this.t.lock}"
                show-text-label
                @click="${this._toggleLocked}"
              ></simple-toolbar-button>
              <simple-toolbar-button
                class="menu-item-delete"
                icon="icons:delete"
                label="${this.t.delete||"Delete"}"
                show-text-label
                @click="${this._deletePage}"
                ?disabled="${this.locked}"
              ></simple-toolbar-button>
            </simple-context-menu>
          `:""}
      ${this.isLoggedIn&&this._haxState?e`
            <simple-toolbar-button
              class="save-button"
              icon="icons:save"
              label="${this.t.savePage}"
              @click="${this._savePage}"
            ></simple-toolbar-button>
            <simple-toolbar-button
              class="save-edit-button"
              icon="hax:page-edit"
              label="${this.t.saveAndEdit}"
              @click="${this._saveAndEdit}"
            ></simple-toolbar-button>
            <simple-toolbar-button
              class="cancel-button"
              icon="icons:cancel"
              label="${this.t.cancel}"
              @click="${this._cancelEdit}"
            ></simple-toolbar-button>
          `:""}
    `}_toggleMenu(e){e.stopPropagation();const t=this.shadowRoot.querySelector("#menu");t&&t.toggle(e.target)}_editPage(e){if(this.locked){const e=this.shadowRoot.querySelector("#menu");return e&&e.close(),r.toast("This page is locked. Unlock it first to edit.",3e3,{hat:"error"}),void r.playSound("error")}const t=this.shadowRoot.querySelector("#menu");t&&t.close(),r.cmsSiteEditor.haxCmsSiteEditorUIElement._editButtonTap()}_saveAndEdit(e){r.cmsSiteEditor.haxCmsSiteEditorUIElement._saveAndEditButtonTap()}_cancelEdit(e){r.cmsSiteEditor.haxCmsSiteEditorUIElement._cancelButtonTap(e)}async _editTitle(e){if(this.locked){const e=this.shadowRoot.querySelector("#menu");return e&&e.close(),r.toast("This page is locked. Unlock it first to edit.",3e3,{hat:"error"}),void r.playSound("error")}const t=this.shadowRoot.querySelector("#menu");t&&t.close();const i=h(r.activeItem);if(!i||!i.id)return void console.warn("page-break _editTitle: No active item found");if(!globalThis.SuperDaemonManager)return void console.error("page-break _editTitle: SuperDaemonManager not available");const a=globalThis.SuperDaemonManager.requestAvailability();r.playSound("click"),a.waveWand(["","/",{},"edit-title","Edit title"])}async _editIcon(e){if(this.locked){const e=this.shadowRoot.querySelector("#menu");return e&&e.close(),r.toast("This page is locked. Unlock it first to edit.",3e3,{hat:"error"}),void r.playSound("error")}const t=this.shadowRoot.querySelector("#menu");t&&t.close();const{SimpleIconsetStore:i}=await import("../simple-icon/lib/simple-iconset.js"),a=h(r.activeItem);if(!a||!a.id)return;const o=globalThis.SuperDaemonManager.requestAvailability();r.playSound("click");const s=i&&i.iconlist?[...i.iconlist].sort():[];o.defineOption({title:"Edit Icon",icon:"icons:image",priority:-1e4,tags:["edit","icon"],eventName:"super-daemon-run-program",path:"CMS/edit/icon",value:{name:"edit-icon",machineName:"edit-icon",placeholder:"Type to search icons by name",program:async e=>{const t=e?e.toLowerCase():"",i=[];return(t?s.filter((e=>e.toLowerCase().includes(t))):s.slice(0,50)).forEach((e=>{const t=e.replace(/^.*:/,"").replace(/-/g," ").replace(/\b\w/g,(e=>e.toUpperCase()));i.push({title:`${t} (${e})`,icon:e,tags:["icon"],value:{target:globalThis,method:"dispatchEvent",args:[new CustomEvent("haxcms-save-node-details",{bubbles:!0,composed:!0,cancelable:!0,detail:{id:a.id,operation:"setIcon",icon:e}})]},eventName:"super-daemon-element-method",path:`CMS/edit/icon/${e}`})})),0===i.length?[{title:t?`No icons found for "${t}"`:"No icons available",icon:"icons:search",tags:["empty"],value:{disabled:!0},eventName:"disabled",path:"No results"}]:(!t&&s.length>50&&i.push({title:`Showing 50 of ${s.length} icons - type to search`,icon:"icons:info",tags:["hint"],value:{disabled:!0},eventName:"disabled",path:"Hint"}),i)}}}),o.waveWand(["","/",{},"edit-icon","Edit Icon"])}async _editMedia(e){if(this.locked){const e=this.shadowRoot.querySelector("#menu");return e&&e.close(),r.toast("This page is locked. Unlock it first to edit.",3e3,{hat:"error"}),void r.playSound("error")}const t=this.shadowRoot.querySelector("#menu");t&&t.close();const i=globalThis.SuperDaemonManager.requestAvailability();r.playSound("click"),i.waveWand(["","/",{},"hax-agent","Agent"])}async _editTags(e){if(this.locked){const e=this.shadowRoot.querySelector("#menu");return e&&e.close(),r.toast("This page is locked. Unlock it first to edit.",3e3,{hat:"error"}),void r.playSound("error")}const t=this.shadowRoot.querySelector("#menu");t&&t.close();const i=h(r.activeItem);if(!i||!i.id)return;const a=globalThis.SuperDaemonManager.requestAvailability();a.close(),r.playSound("click");const o=i.metadata&&i.metadata.tags||"";this._originalTags=o,a.waveWand([o,"/",{},"edit-tags","Edit tags"]);const handleEscape=e=>{"Escape"===e.key&&(this.tags=this._originalTags,globalThis.removeEventListener("keydown",handleEscape))};globalThis.addEventListener("keydown",handleEscape,{once:!0})}async _toggleLocked(e){const t=this.shadowRoot.querySelector("#menu");t&&t.close();const i=h(r.activeItem);i&&i.id&&(r.playSound("click"),globalThis.dispatchEvent(new CustomEvent("haxcms-save-node-details",{bubbles:!0,composed:!0,cancelable:!0,detail:{id:i.id,operation:"setLocked",locked:!this.locked}})))}async _togglePublished(e){if(this.locked){const e=this.shadowRoot.querySelector("#menu");return e&&e.close(),r.toast("This page is locked. Unlock it first to change publish status.",3e3,{hat:"error"}),void r.playSound("error")}const t=this.shadowRoot.querySelector("#menu");t&&t.close();const i=h(r.activeItem);i&&i.id&&(r.playSound("click"),globalThis.dispatchEvent(new CustomEvent("haxcms-save-node-details",{bubbles:!0,composed:!0,cancelable:!0,detail:{id:i.id,operation:"setPublished",published:!this.published}})))}async _savePage(e){r.cmsSiteEditor.haxCmsSiteEditorUIElement._editButtonTap()}_addPage(e){const t=this.shadowRoot.querySelector("#menu");t&&t.close()}_deletePage(e){if(this.locked){const e=this.shadowRoot.querySelector("#menu");return e&&e.close(),r.toast("This page is locked. Unlock it first to delete.",3e3,{hat:"error"}),void r.playSound("error")}const t=this.shadowRoot.querySelector("#menu");t&&t.close(),r.cmsSiteEditor.haxCmsSiteEditorUIElement._deleteButtonTap()}static get haxProperties(){return new URL("./lib/page-break.haxProperties.json",import.meta.url).href}haxHooks(){return{editModeChanged:"haxeditModeChanged",inlineContextMenu:"haxinlineContextMenu",activeElementChanged:"haxactiveElementChanged",setupActiveElementForm:"haxsetupActiveElementForm",preProcessInsertContent:"haxpreProcessInsertContent",trayDragNDropToNode:"haxtrayDragNDropToNode"}}async haxpreProcessInsertContent(e,t){let i=t;if(i&&i.parentNode){for(;"HAX-BODY"!==i.parentNode.tagName;)i=i.parentNode;const t=await l.associatedPageBreak(i);t&&(e.properties.parent=t.parent,e.properties.order=t.order+1,e.properties.published=t.published,e.properties.locked=t.locked)}return e}async haxtrayDragNDropToNode(e){let t=e;for(;"HAX-BODY"!==t.parentNode.tagName;)t=t.parentNode;const i=await l.associatedPageBreak(t);i&&(e.parent=i.parent,e.order=i.order+1,e.published=i.published,e.locked=i.locked)}haxsetupActiveElementForm(e){if(globalThis.HAXCMS){const i=globalThis.HAXCMS.requestAvailability().store.getManifestItems(!0);var t=[{text:`-- ${this.t.noParent} --`,value:null}];i.forEach((e=>{if(e.id!=this.itemId){let a=e,o="- ";for(;a&&null!=a.parent;)a=i.find((e=>e.id==a.parent)),a&&(o="--"+o);t.push({text:o+e.title,value:e.id})}})),e.settings.advanced.forEach(((i,a)=>{"parent"===i.property&&(e.settings.advanced[a].inputMethod="select",e.settings.advanced[a].itemsList=t),"noderefs"===i.property&&(e.settings.advanced[a].properties[0].inputMethod="select",e.settings.advanced[a].properties[0].itemsList=t)})),e.settings.developer.forEach(((t,i)=>{if("developerTheme"===t.property&&globalThis.appSettings&&globalThis.appSettings.themes){var a=[{text:"",value:null}];Object.keys(globalThis.appSettings.themes).map((e=>{a.push({text:globalThis.appSettings.themes[e].name,value:globalThis.appSettings.themes[e].element})})),e.settings.developer[i].inputMethod="select",e.settings.developer[i].itemsList=a}})),e.settings.developer.forEach(((t,i)=>{if("author"===t.property){const t=this.getCurrentUser();t&&!this.author&&(this.author=t,e.settings.developer[i].value=t)}}))}}haxactiveElementChanged(e,t){!t&&this._ceMenu&&("site"===this.breakType?(this._ceMenu.disableOps=!1,this._ceMenu.canMoveElement=!0,this._ceMenu.insertAbove=!0):this._ceMenu.disableDuplicate=!1)}haxeditModeChanged(e){this._haxState=e}haxinlineContextMenu(e){this._ceMenu=e,this._updateHAXCEMenu(),"site"===this.breakType?(this._ceMenu.disableOps=!0,this._ceMenu.canMoveElement=!1,this._ceMenu.insertAbove=!1):this._ceMenu.disableDuplicate=!0}_updateHAXCEMenu(){this._ceMenu.ceButtons=[{icon:this.locked?"icons:lock":"icons:lock-open",callback:"haxClickInlineLock",label:this.t.toggleLock},{icon:this.published?"lrn:view":"lrn:view-off",callback:"haxClickInlinePublished",label:this.t.togglePublished}]}haxClickLockInPage(e){this.locked=!this.locked,globalThis.dispatchEvent(new CustomEvent("hax-refresh-tray-form",{}))}haxClickInlineLock(e){return this.locked=!this.locked,!0}haxClickInlinePublished(e){return this.published=!this.published,!0}}globalThis.customElements.define(PageBreak.tag,PageBreak);
