// @lrnwebcomponents/topic-heading/topic-heading.js
define(["exports", "require", "../../lit-element/lit-element.js"], function (_exports, _require, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.TopicHeading = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject4_e1673ef01bb211eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n\n        :host([hidden]) {\n          display: none;\n        }\n\n        relative-heading {\n          color: var(--topic-heading-text-color);\n          display: inline-flex;\n        }\n        iron-icon {\n          color: var(--topic-heading-icon-color);\n          display: inline-flex;\n          font-size: 16px;\n          height: 32px;\n          width: 32px;\n          padding: 16px;\n          line-height: 16px;\n          vertical-align: bottom;\n        }\n      "]);

    _templateObject4_e1673ef01bb211eabd7cc30119585ddb = function _templateObject4_e1673ef01bb211eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject3_e1673ef01bb211eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <relative-heading text=\"", "\"\n              >", "</relative-heading\n            >\n          "]);

    _templateObject3_e1673ef01bb211eabd7cc30119585ddb = function _templateObject3_e1673ef01bb211eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject2_e1673ef01bb211eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <iron-icon icon=\"", "\"></iron-icon>\n          "]);

    _templateObject2_e1673ef01bb211eabd7cc30119585ddb = function _templateObject2_e1673ef01bb211eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject_e1673ef01bb211eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      ", "\n      ", "\n    "]);

    _templateObject_e1673ef01bb211eabd7cc30119585ddb = function _templateObject_e1673ef01bb211eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  /**
   * `topic-heading`
   * `Semantic and visual meaning behind a heading break`
   * @demo demo/index.html
   * @customElement topic-heading
   */
  var TopicHeading =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(TopicHeading, _LitElement);
    babelHelpers.createClass(TopicHeading, [{
      key: "render",
      // render function
      value: function render() {
        return (0, _litElement.html)(_templateObject_e1673ef01bb211eabd7cc30119585ddb(), this.icon ? (0, _litElement.html)(_templateObject2_e1673ef01bb211eabd7cc30119585ddb(), this.icon) : "", this.title ? (0, _litElement.html)(_templateObject3_e1673ef01bb211eabd7cc30119585ddb(), this.title, this.title) : "");
      } // haxProperty definition

    }], [{
      key: "styles",
      //styles function
      get: function get() {
        return [(0, _litElement.css)(_templateObject4_e1673ef01bb211eabd7cc30119585ddb())];
      }
    }, {
      key: "haxProperties",
      get: function get() {
        return {
          canScale: true,
          canPosition: true,
          canEditSource: false,
          gizmo: {
            title: "Topic heading",
            description: "Semantic and visual meaning behind a heading break",
            icon: "icons:android",
            color: "green",
            groups: ["Heading"],
            handles: [{
              type: "todo:read-the-docs-for-usage"
            }],
            meta: {
              author: "btopro",
              owner: "The Pennsylvania State University"
            }
          },
          settings: {
            quick: [],
            configure: [{
              property: "icon",
              description: "",
              inputMethod: "iconpicker",
              required: false,
              icon: "icons:code"
            }, {
              property: "title",
              description: "",
              inputMethod: "textfield",
              required: false,
              icon: "icons:heading"
            }],
            advanced: []
          }
        };
      } // properties available to the custom element for data binding

    }, {
      key: "properties",
      get: function get() {
        return babelHelpers.objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(TopicHeading), "properties", this), {
          icon: {
            type: String
          },
          title: {
            type: String
          }
        });
      }
    }]);

    function TopicHeading() {
      var _this;

      babelHelpers.classCallCheck(this, TopicHeading);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(TopicHeading).call(this));
      _this.icon = "";
      _this.title = "";
      new Promise(function (res, rej) {
        return _require.default(["../relative-heading/relative-heading.js"], res, rej);
      });
      return _this;
    }

    babelHelpers.createClass(TopicHeading, [{
      key: "firstUpdated",
      value: function firstUpdated() {
        if (this.icon) {
          new Promise(function (res, rej) {
            return _require.default(["../../@polymer/iron-icons/iron-icons.js"], res, rej);
          });
          new Promise(function (res, rej) {
            return _require.default(["../../@polymer/iron-icon/iron-icon.js"], res, rej);
          });
        }
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */

    }], [{
      key: "tag",
      get: function get() {
        return "topic-heading";
      }
    }]);
    return TopicHeading;
  }(_litElement.LitElement);

  _exports.TopicHeading = TopicHeading;
  window.customElements.define(TopicHeading.tag, TopicHeading);
});
