// @lrnwebcomponents/product-card/product-card.js
define(["exports", "../../lit-element/lit-element.js", "../a11y-collapse/lib/a11y-collapse-group.js", "../a11y-collapse/a11y-collapse.js", "../simple-colors/simple-colors.js", "../accent-card/accent-card.js"], function (_exports, _litElement, _a11yCollapseGroup, _a11yCollapse, _simpleColors, _accentCard) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ProductCard = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject3_30f79c40c5f311ea81c635a4a0b7726b() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: flex;\n        }\n      "]);

    _templateObject3_30f79c40c5f311ea81c635a4a0b7726b = function _templateObject3_30f79c40c5f311ea81c635a4a0b7726b() {
      return data;
    };

    return data;
  }

  function _templateObject2_30f79c40c5f311ea81c635a4a0b7726b() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                <iron-icon icon=\"", "\"></iron-icon>\n              "]);

    _templateObject2_30f79c40c5f311ea81c635a4a0b7726b = function _templateObject2_30f79c40c5f311ea81c635a4a0b7726b() {
      return data;
    };

    return data;
  }

  function _templateObject_30f79c40c5f311ea81c635a4a0b7726b() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: block;\n          opacity: 1;\n          visibility: visible;\n        }\n        :host([hidden]) {\n          display: none;\n        }\n        :host([disabled]) accent-card {\n          cursor: not-allowed;\n          opacity: 0.5;\n        }\n        :host([disabled]) accent-card:hover {\n          cursor: not-allowed;\n          opacity: 0.8;\n        }\n        accent-card {\n          margin: 0;\n          --accent-card-footer-padding-left: 0;\n          --accent-card-footer-padding-right: 0;\n        }\n        accent-card > div[slot=\"heading\"] {\n          height: 40px;\n        }\n        div[slot=\"subheading\"] {\n          height: 30px;\n        }\n        a11y-collapse-group {\n          margin: 0;\n          --a11y-collapse-border-color: #ddd;\n        }\n        a11y-collapse > div[slot=\"heading\"] {\n          display: flex;\n          flex-direction: row;\n          align-items: center;\n          min-height: 40px;\n        }\n      </style>\n      <accent-card\n        accent-color=\"", "\"\n        accent-heading\n        ?flat=\"", "\"\n      >\n        <div slot=\"heading\">\n          ", "\n          ", "\n          <slot name=\"card-header\"></slot>\n        </div>\n        <div slot=\"subheading\">\n          ", "\n          <slot name=\"card-subheader\"></slot>\n        </div>\n        <div slot=\"content\"><slot></slot></div>\n        <div slot=\"footer\">\n          <a11y-collapse-group>\n            <a11y-collapse ?disabled=\"", "\" accordion>\n              <div slot=\"heading\">\n                <slot name=\"details-collapse-header\"></slot>\n              </div>\n              <div slot=\"content\">\n                <slot name=\"details-collapse-content\"></slot>\n              </div>\n            </a11y-collapse>\n            <a11y-collapse\n              ?disabled=\"", "\"\n              accordion\n              @expand=\"", "\"\n              @collapse=\"", "\"\n            >\n              <div slot=\"heading\">\n                <slot name=\"demo-collapse-header\"></slot>\n              </div>\n              <div slot=\"content\">\n                <slot name=\"demo-collapse-content\"></slot>\n              </div>\n            </a11y-collapse>\n          </a11y-collapse-group>\n        </div>\n      </accent-card>\n    "]);

    _templateObject_30f79c40c5f311ea81c635a4a0b7726b = function _templateObject_30f79c40c5f311ea81c635a4a0b7726b() {
      return data;
    };

    return data;
  }

  /**
   * `product-card`
   * `basic display of a product with advanced data driven helpers`
   * @demo demo/index.html
   * @demo demo/hax.html HAX Demo
   * @element product-card
   */
  var ProductCard =
  /*#__PURE__*/
  function (_SimpleColors) {
    babelHelpers.inherits(ProductCard, _SimpleColors);
    babelHelpers.createClass(ProductCard, [{
      key: "render",
      // render function
      value: function render() {
        return (0, _litElement.html)(_templateObject_30f79c40c5f311ea81c635a4a0b7726b(), !this.disabled ? this.accentColor : "grey", this.disabled, this.icon ? (0, _litElement.html)(_templateObject2_30f79c40c5f311ea81c635a4a0b7726b(), this.icon) : "", this.heading, this.subheading, this.disabled, this.disabled || !this.hasDemo, this.__demoCollapseStatusChange, this.__demoCollapseStatusChange);
      }
    }, {
      key: "__demoCollapseStatusChange",
      value: function __demoCollapseStatusChange(e) {
        this.dispatchEvent(new CustomEvent("product-card-demo-show", {
          detail: e.detail
        }));
      } // properties available to the custom element for data binding

    }], [{
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(ProductCard), "styles", this)), [(0, _litElement.css)(_templateObject3_30f79c40c5f311ea81c635a4a0b7726b())]);
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(ProductCard), "properties", this), {
          disabled: {
            type: Boolean,
            reflect: true
          },
          heading: {
            type: String
          },
          subheading: {
            type: String
          },
          icon: {
            type: String
          },
          hasDemo: {
            type: Boolean,
            attribute: "has-demo"
          }
        });
      }
      /**
       * Convention we use
       */

    }, {
      key: "tag",
      get: function get() {
        return "product-card";
      }
      /**
       * HTMLElement
       */

    }]);

    function ProductCard() {
      var _this;

      babelHelpers.classCallCheck(this, ProductCard);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(ProductCard).call(this));
      _this.disabled = false;
      return _this;
    }

    return ProductCard;
  }(_simpleColors.SimpleColors);

  _exports.ProductCard = ProductCard;
  customElements.define(ProductCard.tag, ProductCard);
});
