// @polymer/polymer/lib/utils/settings.js
define(["exports", "./boot.js", "./resolve-url.js"], function (_exports, _boot, _resolveUrl) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.setCancelSyntheticClickEvents = _exports.cancelSyntheticClickEvents = _exports.setSyncInitialRender = _exports.syncInitialRender = _exports.setLegacyOptimizations = _exports.legacyOptimizations = _exports.setAllowTemplateFromDomModule = _exports.allowTemplateFromDomModule = _exports.setStrictTemplatePolicy = _exports.strictTemplatePolicy = _exports.setPassiveTouchGestures = _exports.passiveTouchGestures = _exports.setSanitizeDOMValue = _exports.sanitizeDOMValue = _exports.setRootPath = _exports.rootPath = _exports.useNativeCustomElements = _exports.useNativeCSSProperties = _exports.useShadow = void 0;

  /**
  @license
  Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
  This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
  The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
  The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
  Code distributed by Google as part of the polymer project is also
  subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
  */
  const useShadow = !window.ShadyDOM;
  _exports.useShadow = useShadow;
  const useNativeCSSProperties = Boolean(!window.ShadyCSS || window.ShadyCSS.nativeCss);
  _exports.useNativeCSSProperties = useNativeCSSProperties;
  const useNativeCustomElements = !window.customElements.polyfillWrapFlushCallback;
  /**
   * Globally settable property that is automatically assigned to
   * `ElementMixin` instances, useful for binding in templates to
   * make URL's relative to an application's root.  Defaults to the main
   * document URL, but can be overridden by users.  It may be useful to set
   * `rootPath` to provide a stable application mount path when
   * using client side routing.
   */

  _exports.useNativeCustomElements = useNativeCustomElements;
  let rootPath = (0, _resolveUrl.pathFromUrl)(document.baseURI || window.location.href);
  /**
   * Sets the global rootPath property used by `ElementMixin` and
   * available via `rootPath`.
   *
   * @param {string} path The new root path
   * @return {void}
   */

  _exports.rootPath = rootPath;

  const setRootPath = function (path) {
    _exports.rootPath = rootPath = path;
  };
  /**
   * A global callback used to sanitize any value before inserting it into the DOM.
   * The callback signature is:
   *
   *  function sanitizeDOMValue(value, name, type, node) { ... }
   *
   * Where:
   *
   * `value` is the value to sanitize.
   * `name` is the name of an attribute or property (for example, href).
   * `type` indicates where the value is being inserted: one of property, attribute, or text.
   * `node` is the node where the value is being inserted.
   *
   * @type {(function(*,string,string,Node):*)|undefined}
   */


  _exports.setRootPath = setRootPath;
  let sanitizeDOMValue = window.Polymer && window.Polymer.sanitizeDOMValue || undefined;
  /**
   * Sets the global sanitizeDOMValue available via this module's exported
   * `sanitizeDOMValue` variable.
   *
   * @param {(function(*,string,string,Node):*)|undefined} newSanitizeDOMValue the global sanitizeDOMValue callback
   * @return {void}
   */

  _exports.sanitizeDOMValue = sanitizeDOMValue;

  const setSanitizeDOMValue = function (newSanitizeDOMValue) {
    _exports.sanitizeDOMValue = sanitizeDOMValue = newSanitizeDOMValue;
  };
  /**
   * Globally settable property to make Polymer Gestures use passive TouchEvent listeners when recognizing gestures.
   * When set to `true`, gestures made from touch will not be able to prevent scrolling, allowing for smoother
   * scrolling performance.
   * Defaults to `false` for backwards compatibility.
   */


  _exports.setSanitizeDOMValue = setSanitizeDOMValue;
  let passiveTouchGestures = false;
  /**
   * Sets `passiveTouchGestures` globally for all elements using Polymer Gestures.
   *
   * @param {boolean} usePassive enable or disable passive touch gestures globally
   * @return {void}
   */

  _exports.passiveTouchGestures = passiveTouchGestures;

  const setPassiveTouchGestures = function (usePassive) {
    _exports.passiveTouchGestures = passiveTouchGestures = usePassive;
  };
  /**
   * Setting to ensure Polymer template evaluation only occurs based on tempates
   * defined in trusted script.  When true, `<dom-module>` re-registration is
   * disallowed, `<dom-bind>` is disabled, and `<dom-if>`/`<dom-repeat>`
   * templates will only evaluate in the context of a trusted element template.
   */


  _exports.setPassiveTouchGestures = setPassiveTouchGestures;
  let strictTemplatePolicy = false;
  /**
   * Sets `strictTemplatePolicy` globally for all elements
   *
   * @param {boolean} useStrictPolicy enable or disable strict template policy
   *   globally
   * @return {void}
   */

  _exports.strictTemplatePolicy = strictTemplatePolicy;

  const setStrictTemplatePolicy = function (useStrictPolicy) {
    _exports.strictTemplatePolicy = strictTemplatePolicy = useStrictPolicy;
  };
  /**
   * Setting to enable dom-module lookup from Polymer.Element.  By default,
   * templates must be defined in script using the `static get template()`
   * getter and the `html` tag function.  To enable legacy loading of templates
   * via dom-module, set this flag to true.
   */


  _exports.setStrictTemplatePolicy = setStrictTemplatePolicy;
  let allowTemplateFromDomModule = false;
  /**
   * Sets `lookupTemplateFromDomModule` globally for all elements
   *
   * @param {boolean} allowDomModule enable or disable template lookup
   *   globally
   * @return {void}
   */

  _exports.allowTemplateFromDomModule = allowTemplateFromDomModule;

  const setAllowTemplateFromDomModule = function (allowDomModule) {
    _exports.allowTemplateFromDomModule = allowTemplateFromDomModule = allowDomModule;
  };
  /**
   * Setting to skip processing style includes and re-writing urls in css styles.
   * Normally "included" styles are pulled into the element and all urls in styles
   * are re-written to be relative to the containing script url.
   * If no includes or relative urls are used in styles, these steps can be
   * skipped as an optimization.
   */


  _exports.setAllowTemplateFromDomModule = setAllowTemplateFromDomModule;
  let legacyOptimizations = false;
  /**
   * Sets `legacyOptimizations` globally for all elements to enable optimizations
   * when only legacy based elements are used.
   *
   * @param {boolean} useLegacyOptimizations enable or disable legacy optimizations
   * includes and url rewriting
   * @return {void}
   */

  _exports.legacyOptimizations = legacyOptimizations;

  const setLegacyOptimizations = function (useLegacyOptimizations) {
    _exports.legacyOptimizations = legacyOptimizations = useLegacyOptimizations;
  };
  /**
   * Setting to perform initial rendering synchronously when running under ShadyDOM.
   * This matches the behavior of Polymer 1.
   */


  _exports.setLegacyOptimizations = setLegacyOptimizations;
  let syncInitialRender = false;
  /**
   * Sets `syncInitialRender` globally for all elements to enable synchronous
   * initial rendering.
   *
   * @param {boolean} useSyncInitialRender enable or disable synchronous initial
   * rendering globally.
   * @return {void}
   */

  _exports.syncInitialRender = syncInitialRender;

  const setSyncInitialRender = function (useSyncInitialRender) {
    _exports.syncInitialRender = syncInitialRender = useSyncInitialRender;
  };
  /**
   * Setting to cancel synthetic click events fired by older mobile browsers. Modern browsers
   * no longer fire synthetic click events, and the cancellation behavior can interfere
   * when programmatically clicking on elements.
   */


  _exports.setSyncInitialRender = setSyncInitialRender;
  let cancelSyntheticClickEvents = true;
  /**
   * Sets `setCancelSyntheticEvents` globally for all elements to cancel synthetic click events.
   *
   * @param {boolean} useCancelSyntheticClickEvents enable or disable cancelling synthetic
   * events
   * @return {void}
   */

  _exports.cancelSyntheticClickEvents = cancelSyntheticClickEvents;

  const setCancelSyntheticClickEvents = function (useCancelSyntheticClickEvents) {
    _exports.cancelSyntheticClickEvents = cancelSyntheticClickEvents = useCancelSyntheticClickEvents;
  };

  _exports.setCancelSyntheticClickEvents = setCancelSyntheticClickEvents;
});
