// @haxtheweb/haxcms-elements/lib/core/utils/HAXCMSOperationButtons.js
import{css as t,html as e}from"../../../../../lit/index.js";import{store as i}from"../haxcms-site-store.js";export const HAXCMSOperationButtons=function(i){return class extends i{constructor(){super(),this.t||(this.t={}),this.t&&(this.t={...this.t,editPage:"Edit page"})}static get styles(){let e=[];return super.styles&&(e=super.styles),[e,t`
          .operation-buttons-wrapper {
            display: flex;
            justify-content: flex-end;
            height: 0px;
            margin: 0;
            padding: 0;
          }
          .operation-buttons-wrapper .btn {
            margin: 0 16px 0 0;
            transition: all 0.3s ease-in-out;
          }
          .operation-buttons-wrapper .btn:active {
            opacity: 1;
          }
        `]}HAXCMSRenderOperationButtons(){return e`
        ${this.isLoggedIn?e`
              <div class="operation-buttons-wrapper">
                <simple-icon-button-lite
                  class="btn ${this.editMode?"edit-mode-active":""}"
                  icon="hax:page-edit"
                  ?disabled="${this.editMode}"
                  @click="${this._editButtonTap}"
                  >${this.t.editPage}</simple-icon-button-lite
                >
              </div>
            `:""}
      `}_editButtonTap(t){this.editMode=!this.editMode}}};
