// @lrnwebcomponents/moment-element/lib/moment/src/lib/duration/get.js
define(["exports", "../units/aliases.js", "../utils/abs-floor.js"], function (_exports, _aliases, _absFloor) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.days = void 0;
  _exports.get = get;
  _exports.seconds = _exports.months = _exports.minutes = _exports.milliseconds = _exports.hours = void 0;
  _exports.weeks = weeks;
  _exports.years = void 0;
  _absFloor = babelHelpers.interopRequireDefault(_absFloor);

  function get(units) {
    units = (0, _aliases.normalizeUnits)(units);
    return this.isValid() ? this[units + "s"]() : NaN;
  }

  function makeGetter(name) {
    return function () {
      return this.isValid() ? this._data[name] : NaN;
    };
  }

  var milliseconds = makeGetter("milliseconds");
  _exports.milliseconds = milliseconds;
  var seconds = makeGetter("seconds");
  _exports.seconds = seconds;
  var minutes = makeGetter("minutes");
  _exports.minutes = minutes;
  var hours = makeGetter("hours");
  _exports.hours = hours;
  var days = makeGetter("days");
  _exports.days = days;
  var months = makeGetter("months");
  _exports.months = months;
  var years = makeGetter("years");
  _exports.years = years;

  function weeks() {
    return (0, _absFloor.default)(this.days() / 7);
  }
});
