// @lrnwebcomponents/paper-fab-transitions/lib/paper-square-grow-animation.js
define(["exports", "../../../@polymer/polymer/polymer-element.js", "../../../@polymer/neon-animation/neon-animation.js", "../../../@polymer/neon-animation/neon-animation-behavior.js", "../../../@polymer/polymer/lib/legacy/class.js"], function (_exports, _polymerElement, _neonAnimation, _neonAnimationBehavior, _class) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.PaperSquareGrowAnimation = void 0;

  /**
  `<paper-square-grow-animation>` increases the element's width and height from an initial
  predefined pixel square size to its final size.
  
  Configuration:
  ```
  {
    name: 'expand-animation',
    node: <node>,
    startSize: <start-size-in-px>,
    timing: <animation-timing>,
  }
  ```
  @hero hero.svg
  * @demo demo/index.html
  */
  var PaperSquareGrowAnimation =
  /*#__PURE__*/
  function (_mixinBehaviors) {
    babelHelpers.inherits(PaperSquareGrowAnimation, _mixinBehaviors);

    function PaperSquareGrowAnimation() {
      babelHelpers.classCallCheck(this, PaperSquareGrowAnimation);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(PaperSquareGrowAnimation).apply(this, arguments));
    }

    babelHelpers.createClass(PaperSquareGrowAnimation, [{
      key: "configure",
      value: function configure(config) {
        var node = config.node;
        var startSize = config.startSize;
        var height = node.getBoundingClientRect().height;
        var width = node.getBoundingClientRect().width;
        this._effect = new KeyframeEffect(node, [{
          height: startSize + "px",
          width: startSize + "px"
        }, {
          height: height + "px",
          width: width + "px"
        }], this.timingFromConfig(config));
        return this._effect;
      }
    }], [{
      key: "tag",
      get: function get() {
        return "paper-square-grow-animation";
      }
    }]);
    return PaperSquareGrowAnimation;
  }((0, _class.mixinBehaviors)([_neonAnimationBehavior.NeonAnimationBehavior], _polymerElement.PolymerElement));

  _exports.PaperSquareGrowAnimation = PaperSquareGrowAnimation;
  window.customElements.define(PaperSquareGrowAnimation.tag, PaperSquareGrowAnimation);
});
