// @lrnwebcomponents/paper-fab-transitions/lib/paper-square-grow-animation.js
define(["exports", "../../../@polymer/polymer/polymer-element.js", "../../../@polymer/neon-animation/neon-animation.js", "../../../@polymer/neon-animation/neon-animation-behavior.js", "../../../@polymer/polymer/lib/legacy/class.js"], function (_exports, _polymerElement, _neonAnimation, _neonAnimationBehavior, _class) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.PaperSquareGrowAnimation = void 0;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
  `<paper-square-grow-animation>` increases the element's width and height from an initial
  predefined pixel square size to its final size.
  
  Configuration:
  ```
  {
    name: 'expand-animation',
    node: <node>,
    startSize: <start-size-in-px>,
    timing: <animation-timing>,
  }
  ```
  @hero hero.svg
  * @demo demo/index.html
  */
  var PaperSquareGrowAnimation = /*#__PURE__*/function (_mixinBehaviors) {
    babelHelpers.inherits(PaperSquareGrowAnimation, _mixinBehaviors);

    var _super = _createSuper(PaperSquareGrowAnimation);

    function PaperSquareGrowAnimation() {
      babelHelpers.classCallCheck(this, PaperSquareGrowAnimation);
      return _super.apply(this, arguments);
    }

    babelHelpers.createClass(PaperSquareGrowAnimation, [{
      key: "configure",
      value: function configure(config) {
        var node = config.node;
        var startSize = config.startSize;
        var height = node.getBoundingClientRect().height;
        var width = node.getBoundingClientRect().width;
        this._effect = new KeyframeEffect(node, [{
          height: startSize + "px",
          width: startSize + "px"
        }, {
          height: height + "px",
          width: width + "px"
        }], this.timingFromConfig(config));
        return this._effect;
      }
    }], [{
      key: "tag",
      get: function get() {
        return "paper-square-grow-animation";
      }
    }]);
    return PaperSquareGrowAnimation;
  }((0, _class.mixinBehaviors)([_neonAnimationBehavior.NeonAnimationBehavior], _polymerElement.PolymerElement));

  _exports.PaperSquareGrowAnimation = PaperSquareGrowAnimation;
  window.customElements.define(PaperSquareGrowAnimation.tag, PaperSquareGrowAnimation);
});
