// @lrnwebcomponents/lrndesign-comment/lrndesign-comment.js
define(["exports", "../../@polymer/polymer/polymer-element.js", "../lrndesign-avatar/lrndesign-avatar.js", "../../time-elements/dist/time-elements.js"], function (_exports, _polymerElement, _lrndesignAvatar, _timeElements) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrndesignComment = void 0;

  /**
   * `lrndesign-comment`
   * @element lrndesign-comment
   * A LRN element
   *
   * @demo demo/index.html
   */
  class LrndesignComment extends _polymerElement.PolymerElement {
    static get template() {
      return (0, _polymerElement.html)`
      <style>
        :host {
          display: block;
        }
        .comment-left {
          float: left;
          display: inline-block;
        }
        .comment-right {
          display: inline-block;
        }
      </style>
      <div class="comment-container">
        <div class="comment-left"><lrndesign-avatar></lrndesign-avatar></div>
        <div class="comment-right">
          <div class="row-1">
            <span>{{name}}</span>
            <relative-time datetime\$="{{date}}"> </relative-time>
          </div>
          <div class="row-2"><slot></slot></div>
          <div class="row-3">{{links}}</div>
        </div>
      </div>
    `;
    }

    static get tag() {
      return "lrndesign-comment";
    }

    static get properties() {
      return {
        avatar: {
          type: Object,
          reflectToAttribute: true,
          notify: true
        },
        name: {
          type: String,
          reflectToAttribute: true,
          notify: true
        },
        date: {
          type: String,
          value: "2014-04-01T00:00:00.000Z",
          reflectToAttribute: true,
          notify: true
        },
        links: {
          type: Object,
          notify: true
        }
      };
    }

  }

  _exports.LrndesignComment = LrndesignComment;
  window.customElements.define(LrndesignComment.tag, LrndesignComment);
});
