// @lrnwebcomponents/simple-datetime/simple-datetime.js
define(["exports", "../../lit-element/lit-element.js", "./lib/date.format.js"], function (_exports, _litElement, _dateFormat) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SimpleDatetime = void 0;

  function _templateObject2_ed7cb5c0544211ebad6c77197a564a47() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          font-size: 14px;\n          color: #b3b3b1;\n          line-height: 30px;\n        }\n      "]);

    _templateObject2_ed7cb5c0544211ebad6c77197a564a47 = function _templateObject2_ed7cb5c0544211ebad6c77197a564a47() {
      return data;
    };

    return data;
  }

  function _templateObject_ed7cb5c0544211ebad6c77197a564a47() {
    var data = babelHelpers.taggedTemplateLiteral([" <time datetime=\"", "\">", "</time> "]);

    _templateObject_ed7cb5c0544211ebad6c77197a564a47 = function _templateObject_ed7cb5c0544211ebad6c77197a564a47() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `simple-datetime`
   * @element simple-datetime
   * A simple datetime element that takes in unix timestamp and outputs a date.
   * @demo demo/index.html
   * @microcopy - the mental model for this element
   * - passing in a timestamp from unix and having it be php based date formatting to render is super helpful
   */
  var SimpleDatetime = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(SimpleDatetime, _LitElement);

    var _super = _createSuper(SimpleDatetime);

    /**
     * HTMLElement life cycle
     */
    function SimpleDatetime() {
      var _this;

      babelHelpers.classCallCheck(this, SimpleDatetime);
      _this = _super.call(this);
      _this.format = "M jS, Y";
      _this.unix = false;
      return _this;
    }
    /**
     * LitElement constructable styles enhancement
     */


    babelHelpers.createClass(SimpleDatetime, [{
      key: "render",

      /**
       * LitElement life cycle - render callback
       */
      value: function render() {
        return (0, _litElement.html)(_templateObject_ed7cb5c0544211ebad6c77197a564a47(), this.date, this.date);
      }
    }, {
      key: "updated",

      /**
       * LitElement life cycle - property changed callback
       */
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (["timestamp", "format", "unix"].includes(propName)) {
            _this2.date = _this2.formatDate(_this2.timestamp, _this2.format, _this2.unix);
          }
        });
      }
      /**
       * Figure out the date
       */

    }, {
      key: "formatDate",
      value: function formatDate(timestamp, format, unix) {
        // unix timestamp is seconds, JS is milliseconds
        if (unix) {
          timestamp = timestamp * 1000;
        }

        return new Date(timestamp).format(format);
      }
    }], [{
      key: "styles",
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_ed7cb5c0544211ebad6c77197a564a47())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "simple-datetime";
      }
      /**
       * LitElement life cycle -  properties definitions
       */

    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Javascript timestamp
           */
          timestamp: {
            type: Number
          },

          /**
           * Format to output, see https://github.com/jacwright/date.format#supported-identifiers
           */
          format: {
            type: String
          },

          /**
           * Date, generated from timestamp + format
           */
          date: {
            type: String
          },

          /**
           * Support for UNIX timestamp conversion on the fly
           */
          unix: {
            type: Boolean
          }
        };
      }
    }]);
    return SimpleDatetime;
  }(_litElement.LitElement);

  _exports.SimpleDatetime = SimpleDatetime;
  window.customElements.define(SimpleDatetime.tag, SimpleDatetime);
});
