// @lrnwebcomponents/post-card/lib/PostCardPostmark.js
define(["exports", "meta", "../../../lit/index.js"], function (_exports, meta, _index) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.PostCardPostmark = void 0;
  meta = _interopRequireWildcard(meta);

  var _templateObject_6ab78ee0942811ecb1937969798b231a, _templateObject2_6ab78ee0942811ecb1937969798b231a;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var PostCardPostmark = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(PostCardPostmark, _LitElement);

    var _super = _createSuper(PostCardPostmark);

    function PostCardPostmark() {
      var _this;

      babelHelpers.classCallCheck(this, PostCardPostmark);
      _this = _super.call(this);
      _this.image = new URL("./assets/postcard-postmark.svg", meta.url).href;
      _this.alt = "";
      _this.locations = "";
      return _this;
    }

    babelHelpers.createClass(PostCardPostmark, [{
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_6ab78ee0942811ecb1937969798b231a || (_templateObject_6ab78ee0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <div>\n        <img loading=\"lazy\" src=\"", "\" alt=\"", "\" />\n        <p>", "</p>\n      </div>\n    "])), this.image, this.alt, this.locations);
      }
    }], [{
      key: "tag",
      get: function get() {
        return "post-card-postmark";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          image: {
            type: String
          },
          alt: {
            type: String
          },
          locations: {
            type: String
          }
        };
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _index.css)(_templateObject2_6ab78ee0942811ecb1937969798b231a || (_templateObject2_6ab78ee0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          --post-card-img-width: 250px;\n          font-family: \"Bebas Neue\", sans-serif;\n        }\n        div {\n          width: var(--post-card-img-width);\n          display: flex;\n          flex-direction: column;\n          justify-content: center;\n          align-items: center;\n          opacity: 0.8;\n        }\n        img {\n          width: var(--post-card-img-width);\n          height: calc(var(--post-card-img-width) * (1 / 3));\n          filter: invert(62%) sepia(0%) saturate(329%) hue-rotate(162deg)\n            brightness(98%) contrast(95%); /* created using: https://codepen.io/sosuke/pen/Pjoqqp */\n        }\n        p {\n          width: var(--post-card-img-width);\n          text-transform: uppercase;\n          text-align: center;\n          color: black;\n          margin: 0px;\n          font-size: 16px;\n          letter-spacing: 2px;\n          text-align: center;\n          overflow: hidden;\n        }\n      "])))];
      }
    }]);
    return PostCardPostmark;
  }(_index.LitElement);

  _exports.PostCardPostmark = PostCardPostmark;
  customElements.define(PostCardPostmark.tag, PostCardPostmark);
});
