// @lrnwebcomponents/post-card/lib/PostCardPhoto.js
define(["exports", "meta", "../../../lit/index.js", "../../simple-colors/simple-colors.js"], function (_exports, meta, _index, _simpleColors) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.PostCardPhoto = void 0;
  meta = _interopRequireWildcard(meta);

  var _templateObject_6ab4a8b0942811ecb1937969798b231a, _templateObject2_6ab4a8b0942811ecb1937969798b231a;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var tape = new URL("./assets/postcard-tape.png", meta.url).href;
  var PostCardPhotoShadow = new URL("./assets/postcard-photo-shadow.png", meta.url).href;

  var PostCardPhoto = /*#__PURE__*/function (_SimpleColors) {
    babelHelpers.inherits(PostCardPhoto, _SimpleColors);

    var _super = _createSuper(PostCardPhoto);

    function PostCardPhoto() {
      var _this;

      babelHelpers.classCallCheck(this, PostCardPhoto);
      _this = _super.call(this);
      _this.accentColor = "grey";
      _this.image = new URL(_this.image, meta.url).href;
      _this.alt = "";
      return _this;
    }

    babelHelpers.createClass(PostCardPhoto, [{
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_6ab4a8b0942811ecb1937969798b231a || (_templateObject_6ab4a8b0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <div>\n        <img\n          loading=\"lazy\"\n          src=\"", "\"\n          alt=\"\"\n          class=\"cardShadow\"\n        />\n        <img\n          loading=\"lazy\"\n          src=\"", "\"\n          alt=\"", "\"\n          class=\"cardImage\"\n        />\n        <img loading=\"lazy\" src=\"", "\" alt=\"\" class=\"cardTape\" />\n      </div>\n    "])), PostCardPhotoShadow, this.image, this.alt, tape);
      }
    }], [{
      key: "tag",
      get: function get() {
        return "post-card-photo";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(PostCardPhoto), "properties", this)), {}, {
          image: {
            type: String
          },
          alt: {
            type: String,
            reflect: true
          }
        });
      }
    }, {
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(PostCardPhoto), "styles", this)), [(0, _index.css)(_templateObject2_6ab4a8b0942811ecb1937969798b231a || (_templateObject2_6ab4a8b0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          --post-card-img-width: 340px;\n        }\n\n        div {\n          display: inline-grid;\n          grid-template-columns: 1;\n          grid-template-rows: 1;\n          align-items: center;\n        }\n\n        img {\n          grid-column: 1;\n          grid-row: 1;\n          justify-self: center;\n        }\n\n        .cardShadow {\n          width: calc(var(--post-card-img-width) * 1.05);\n          height: calc(var(--post-card-img-width) * 0.78);\n          z-index: 2;\n          opacity: 0.5;\n          transform: translate(1%, 1.5%) rotate(0.5deg);\n        }\n\n        .cardImage {\n          width: var(--post-card-img-width);\n          height: calc(var(--post-card-img-width) * 0.7);\n          z-index: 2;\n          transform: rotate(-3deg);\n          border-radius: 5px 5px 0px 5px;\n        }\n\n        .cardTape {\n          width: auto;\n          height: calc(var(--post-card-img-width) * 0.8);\n          z-index: 3;\n        }\n      "])))]);
      }
    }]);
    return PostCardPhoto;
  }(_simpleColors.SimpleColors);

  _exports.PostCardPhoto = PostCardPhoto;
  customElements.define(PostCardPhoto.tag, PostCardPhoto);
});
