// @lrnwebcomponents/rich-text-editor/lib/toolbars/rich-text-editor-toolbar-full.js
define(["exports", "../../../../lit-element/lit-element.js", "./rich-text-editor-toolbar.js", "./rich-text-editor-breadcrumbs.js"], function (_exports, _litElement, _richTextEditorToolbar, _richTextEditorBreadcrumbs) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.RichTextEditorToolbarFull = void 0;

  /**
   * Copyright 2019 Penn State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `rich-text-editor-toolbar-full`
   * `a full toolbar with breadcrumbs for the rich text editor`
   *
   * @element rich-text-editor-toolbar-full
   * @demo ./demo/index.html demo
   * @demo ./demo/full.html toolbar with breadcrumb
   */
  class RichTextEditorToolbarFull extends (0, _richTextEditorToolbar.RichTextEditorToolbarBehaviors)(_litElement.LitElement) {
    /**
     * Store the tag name to make it easier to obtain directly.
     */
    static get tag() {
      return "rich-text-editor-toolbar-full";
    }

    static get styles() {
      return [...super.baseStyles, ...super.stickyStyles];
    } // render function for template


    render() {
      return super.render();
    } // properties available to the custom element for data binding


    static get properties() {
      return { ...super.properties,

        /**
         * The label for the breadcrums area.
         */
        breadcrumbsLabel: {
          name: "breadcrumbsLabel",
          type: String,
          attribute: "breadcrumbs-label"
        }
      };
    }

    constructor() {
      super();
      this.breadcrumbsLabel = "Expand selection: ";
      this.breadcrumbs = document.createElement("rich-text-editor-breadcrumbs");

      this.breadcrumbs.onselectnode = e => this._selectNode(e.detail);

      document.body.appendChild(this.breadcrumbs);
    }

  }

  _exports.RichTextEditorToolbarFull = RichTextEditorToolbarFull;
  window.customElements.define(RichTextEditorToolbarFull.tag, RichTextEditorToolbarFull);
});
