// @lrnwebcomponents/lunr-search/lunr-search.js
define(["exports", "meta", "../../lit-element/lit-element.js", "../es-global-bridge/es-global-bridge.js", "../../@polymer/iron-ajax/iron-ajax.js"], function (_exports, meta, _litElement, _esGlobalBridge, _ironAjax) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LunrSearch = void 0;
  meta = babelHelpers.interopRequireWildcard(meta);

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject2_870400a0e0cf11eab599756c9ebe7c2f() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n\n        :host([hidden]) {\n          display: none;\n        }\n      "]);

    _templateObject2_870400a0e0cf11eab599756c9ebe7c2f = function _templateObject2_870400a0e0cf11eab599756c9ebe7c2f() {
      return data;
    };

    return data;
  }

  function _templateObject_870400a0e0cf11eab599756c9ebe7c2f() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <iron-ajax\n        ?auto=\"", "\"\n        url=\"", "\"\n        method=\"", "\"\n        handle-as=\"json\"\n        @response=\"", "\"\n      ></iron-ajax>\n    "]);

    _templateObject_870400a0e0cf11eab599756c9ebe7c2f = function _templateObject_870400a0e0cf11eab599756c9ebe7c2f() {
      return data;
    };

    return data;
  }

  /**
   * `lunr-search`
   * `LunrJS search element`
   * @demo demo/index.html
   * @element lunr-search
   */
  var LunrSearch =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LunrSearch, _LitElement);
    babelHelpers.createClass(LunrSearch, [{
      key: "render",
      // render function
      value: function render() {
        return (0, _litElement.html)(_templateObject_870400a0e0cf11eab599756c9ebe7c2f(), this.__auto, this.dataSource, this.method, this._dataResponse);
      } // properties available to the custom element for data binding

    }], [{
      key: "styles",
      //styles function
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_870400a0e0cf11eab599756c9ebe7c2f())];
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(LunrSearch), "properties", this), {
          dataSource: {
            type: String,
            attribute: "data-source"
          },
          __auto: {
            type: Boolean
          },
          data: {
            type: Array
          },
          method: {
            type: String
          },
          search: {
            type: String
          },
          results: {
            type: Array
          },
          noStopWords: {
            type: Boolean,
            attribute: "no-stop-words"
          },
          fields: {
            type: Array
          },
          indexNoStopWords: {
            type: Object
          },
          index: {
            type: Object
          },
          __lunrLoaded: {
            type: Boolean
          },
          limit: {
            type: Number
          },
          minScore: {
            type: Number
          },
          log: {
            type: Boolean
          }
        });
      }
    }]);

    function LunrSearch() {
      var _this;

      babelHelpers.classCallCheck(this, LunrSearch);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LunrSearch).call(this));
      _this.method = "GET";
      _this.noStopWords = false;
      _this.dataSource = null;
      _this.fields = [];
      _this.limit = 500;
      _this.__auto = false;
      _this.minScore = 0;
      _this.log = false;

      var basePath = _this.pathFromUrl(meta.url);

      var location = "".concat(basePath, "../../lunr/lunr.js");
      window.addEventListener("es-bridge-lunr-loaded", _this._lunrLoaded.bind(babelHelpers.assertThisInitialized(_this)));
      window.ESGlobalBridge.requestAvailability();
      window.ESGlobalBridge.instance.load("lunr", location);

      if (window.ESGlobalBridge.imports && window.ESGlobalBridge.imports["lunr"]) {
        _this.__lunrLoaded = true;
      }

      return _this;
    }

    babelHelpers.createClass(LunrSearch, [{
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          var notifiedProps = ["data", "search", "results", "noStopWords"];

          if (notifiedProps.includes(propName)) {
            // notify
            var eventName = "".concat(propName.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, "$1-$2").toLowerCase(), "-changed");

            _this2.dispatchEvent(new CustomEvent(eventName, {
              detail: {
                value: _this2[propName]
              }
            }));
          } // only request data when we actually have a data source


          if (propName == "dataSource" && _this2[propName]) {
            _this2.__auto = true;
          }

          if (["data", "search", "index", "minScore", "limit"].includes(propName)) {
            _this2.results = _this2.searched(_this2.data, _this2.search, _this2.index, _this2.minScore, _this2.limit);
          }

          if (["data", "fields", "noStopWords", "__lunrLoaded"].includes(propName)) {
            _this2.index = _this2._createIndex(_this2.data, _this2.fields, _this2.noStopWords, _this2.__lunrLoaded);
          }
        });
      } // simple path from a url modifier

    }, {
      key: "pathFromUrl",
      value: function pathFromUrl(url) {
        return url.substring(0, url.lastIndexOf("/") + 1);
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        window.removeEventListener("es-bridge-lunr-loaded", this._lunrLoaded.bind(this));
        babelHelpers.get(babelHelpers.getPrototypeOf(LunrSearch.prototype), "disconnectedCallback", this).call(this);
      }
    }, {
      key: "_lunrLoaded",
      value: function _lunrLoaded(e) {
        // callback when loaded
        this.__lunrLoaded = true;
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */

    }, {
      key: "_dataResponse",
      value: function _dataResponse(e) {
        // must get a real response
        if (e && e.detail && e.detail.response) {
          try {
            this.data = babelHelpers.toConsumableArray(e.detail.response);
          } catch (e) {
            console.warn(e);
          }
        }
      }
      /**
        Filters your input data
        
        @param {Array} data Array of Objects with common properties.
        @param {String} search The search term that filters results.
        @param {Object} index The lunr Index..
        @param {Number} minScore The minimum score of your results.
        @param {Number} limit The maximum number of results you'd like your results.
        
        @return {Array} The filtered data.
       */

    }, {
      key: "searched",
      value: function searched(data, search, index, minScore, limit) {
        if (data && index && search) {
          var results = [];

          if ("" + search !== "") {
            var searched = index.search(search);

            for (var i = 0; i < searched.length; i++) {
              if (i === limit || searched[i].score < minScore) {
                break;
              } // match on the id within the array of options


              var tmpItem = data.find(function (j) {
                return j.id == searched[i].ref;
              });
              results.push(tmpItem);
            }
          }

          if (results.length === 0 && !this.noStopWords && "" + search !== "") {
            if (!this.indexNoStopWords) {
              this.indexNoStopWords = this._createIndex(data, this.fields, true, this.__lunrLoaded);
            }

            searched = this.indexNoStopWords.search(search);
            var results = [];

            for (var i = 0; i < searched.length; i++) {
              if (i === limit || searched[i].score < minScore) {
                break;
              }

              var _tmpItem = data.find(function (j) {
                return j.id == searched[i].ref;
              });

              results.push(_tmpItem);
            }
          }

          return results;
        }
      }
    }, {
      key: "_createIndex",
      value: function _createIndex(data, fields, noStopWords, ready) {
        if (ready) {
          var root = this;

          if (Array.isArray(data) && data.length > 0) {
            if (Array.isArray(fields) && fields.length > 0) {
              return lunr(function () {
                for (var i = 0; i < fields.length; i++) {
                  if (fields[i].charAt(0) === fields[i].charAt(0).toUpperCase()) {
                    this.field(fields[i], {
                      boost: 10
                    });
                  } else {
                    this.field(fields[i]);
                  }
                }

                for (var i = 0; i < data.length; i++) {
                  var toIndex = {
                    id: i
                  };

                  for (var f = 0; f < fields.length; f++) {
                    if (data[i].hasOwnProperty(fields[f]) && data[i][fields[f]] !== null && typeof data[i][fields[f]].toString == "function" && (data[i][fields[f]].toString().split(" ").length > 2 || data[i][fields[f]].toString().length < 30)) {
                      //indicate that they might be words in it
                      toIndex[fields[f]] = data[i][fields[f]].toString();
                    } else {
                      toIndex[fields[f]] = "";
                    }
                  }

                  this.add(toIndex);
                }

                if (noStopWords) {
                  this.pipeline.remove(lunr.stopWordFilter);
                }
              });
            } else {
              // find fields
              // TODO only word best fields.
              var fields = [];
              var ddup = {};
              return lunr(function () {
                for (var indexOfData = 0; indexOfData < data.length; indexOfData++) {
                  for (var prop in data[indexOfData]) {
                    if (prop.charAt(0) !== "_" && !ddup.hasOwnProperty(prop) && (prop.toString().split(" ").length > 2 || prop.toString().length < 30)) {
                      fields.push(prop);

                      if (prop.charAt(0) === prop.charAt(0).toUpperCase()) {
                        this.field(prop, {
                          boost: 10
                        });
                      } else {
                        this.field(prop);
                      }

                      ddup[prop] = 1;
                    }
                  }
                }

                if (fields.length > 0) {
                  root.fields = fields;
                }

                for (var i = 0; i < data.length; i++) {
                  var toIndex = {
                    id: i
                  };

                  for (var f = 0; f < fields.length; f++) {
                    if (data[i].hasOwnProperty(fields[f]) && data[i][fields[f]] !== null && typeof data[i][fields[f]].toString == "function" && (data[i][fields[f]].toString().split(" ").length > 2 || data[i][fields[f]].toString().length < 30)) {
                      //indicate that they might be words in it
                      toIndex[fields[f]] = data[i][fields[f]].toString();
                    } else {
                      toIndex[fields[f]] = "";
                    }
                  }

                  this.add(toIndex);
                }

                if (noStopWords) {
                  this.pipeline.remove(lunr.stopWordFilter);
                }
              });
            }
          }
        }
      }
    }], [{
      key: "tag",
      get: function get() {
        return "lunr-search";
      }
    }]);
    return LunrSearch;
  }(_litElement.LitElement);

  _exports.LunrSearch = LunrSearch;
  window.customElements.define(LunrSearch.tag, LunrSearch);
});
