// @lrnwebcomponents/haxschema-builder/lib/hax-schema-form.js
define(["exports", "require", "../../../@polymer/polymer/polymer-element.js", "../../hax-body-behaviors/lib/HAXFields.js", "../../simple-icon/simple-icon.js", "../../simple-icon/lib/simple-icons.js", "../../simple-icon/lib/simple-icon-button.js"], function (_exports, _require, _polymerElement, _HAXFields, _simpleIcon, _simpleIcons, _simpleIconButton) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HaxSchemaForm = void 0;
  _require = _interopRequireWildcard(_require);

  var _templateObject_6508a1f0942811ecb1937969798b231a;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `hax-schema-form`
   * @element hax-schema-form
   * `An element that can generate a form from HAXschema`
   * @demo demo/index.html
   * @microcopy - the mental model for this element
   *  - source - a json object from some place loaded in remotely which will then be in json-schema format. This will then be parsed into a form which can be used to manipulate the element.
   */
  var HaxSchemaForm = /*#__PURE__*/function (_PolymerElement) {
    babelHelpers.inherits(HaxSchemaForm, _PolymerElement);

    var _super = _createSuper(HaxSchemaForm);

    function HaxSchemaForm() {
      var _this;

      babelHelpers.classCallCheck(this, HaxSchemaForm);
      _this = _super.call(this);
      new Promise(function (res, rej) {
        return _require.default(["../../../@polymer/paper-tabs/paper-tabs.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../../@polymer/paper-tabs/paper-tab.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../simple-fields/simple-fields.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../simple-fields/lib/simple-fields-form.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../code-editor/code-editor.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../simple-fields/lib/simple-fields-field.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../simple-picker/simple-picker.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../simple-icon-picker/simple-icon-picker.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../simple-colors/lib/simple-colors-picker.js"], res, rej);
      });
      return _this;
    }

    babelHelpers.createClass(HaxSchemaForm, [{
      key: "_computedFormKey",
      value:
      /**
       * Compute form key to use.
       */
      function _computedFormKey(advanced) {
        if (advanced) {
          return "advanced";
        } else {
          return "configure";
        }
      }
      /**
       * Form key changed, rebuild schema for the form.
       */

    }, {
      key: "_formKeyChanged",
      value: function _formKeyChanged(newValue, oldValue) {
        if (newValue) {
          if (newValue === "advanced") {
            this.set("schema", this.advancedSchema);
          } else {
            this.set("schema", this.configureSchema);
          }

          this.notifyPath("schema.*");
        }
      }
      /**
       * Value in the form has changed, reflect to the preview.
       */

    }, {
      key: "_valueChanged",
      value: function _valueChanged(newValue) {
        if (newValue && this.schema) {
          for (var i in newValue) {
            this.schema[i].value = newValue[i];
          }
        }
      }
      /**
       * Editor mode changed handler
       */

    }, {
      key: "_editorModeChanged",
      value: function _editorModeChanged(mode) {
        if (mode) {
          // if it's the advanced setting then toggle the advancedForm setting
          if (mode === "advanced") {
            this.advancedForm = true;
          } else {
            this.advancedForm = false;
          }
        }
      }
      /**
       * add a field to the form in question in refresh
       */

    }, {
      key: "addField",
      value: function addField(key, field) {
        var type = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : "configure";

        if (type === "configure") {
          this.configureSchema.properties[key] = field;
          this.set("schema", this.configureSchema);
        } else {
          this.set("schema", this.advancedSchema);
        }

        this.notifyPath("schema.*");
      }
      /**
       * remove a field from one of the forms in question and refresh the display
       */

    }, {
      key: "removeField",
      value: function removeField(key) {
        var type = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "configure";

        if (type === "configure") {
          delete this.configureSchema.properties[key];
          this.set("schema", this.configureSchema);
        } else {
          delete this.advancedSchema.properties[key];
          this.set("schema", this.advancedSchema);
        }

        this.notifyPath("schema.*");
      }
    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_6508a1f0942811ecb1937969798b231a || (_templateObject_6508a1f0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: block;\n          background-color: #ffffff;\n          overflow: hidden;\n        }\n        div.card.form-wrapper {\n          margin: 0;\n          padding: 0 16px 80px 16px;\n          width: 100%;\n          min-height: 160px;\n          background-color: transparent;\n          overflow: auto;\n          height: 100%;\n          box-shadow: 0 5px 5px rgba(0, 0, 0, 0.7);\n        }\n\n        simple-fields {\n          width: 50%;\n        }\n\n        #modetabs {\n          height: 64px;\n          padding: 0px;\n          margin: 16px 0 0 0;\n          box-sizing: content-box;\n          color: var(--hax-color-text, #000000);\n          text-align: center;\n          background-color: transparent;\n          border-bottom: 1px solid var(--hax-color-border-outline);\n          display: block;\n          justify-content: space-evenly;\n          --paper-tabs-selection-bar-color: var(\n            --hax-color-accent1,\n            --simple-colors-default-theme-light-blue-7\n          );\n          --paper-tabs: {\n            background: transparent;\n          }\n        }\n\n        #modetabs paper-tab {\n          display: inline-flex;\n          height: 100%;\n          --paper-tab-ink: var(\n            --hax-color-accent1,\n            --simple-colors-default-theme-light-blue-7\n          );\n          --paper-tab: {\n            font-size: 16px;\n          }\n        }\n        #modetabs paper-tab button {\n          min-width: unset;\n          width: 100%;\n          background-color: var(--hax-color-menu-heading-bg, #eeeeee);\n          color: var(--hax-color-menu-heading-color, black);\n        }\n        simple-fields {\n          color: var(--hax-text-color);\n          --simple-colors-picker-preview-size: 20px;\n        }\n      </style>\n      <paper-tabs\n        id=\"modetabs\"\n        selected=\"{{modeTab}}\"\n        attr-for-selected=\"data-mode\"\n      >\n        <paper-tab id=\"configurebutton\" data-mode=\"configure\"\n          ><button>Configure</button></paper-tab\n        >\n        <paper-tab id=\"advancedbutton\" data-mode=\"advanced\"\n          ><button>Advanced</button></paper-tab\n        >\n      </paper-tabs>\n      <div class=\"card form-wrapper\">\n        <simple-fields\n          id=\"form\"\n          schema=\"[[schema]]\"\n          .schematizer=\"", "\"\n          .elementizer=\"", "\"\n          value=\"{{value}}\"\n        ></simple-fields>\n      </div>\n    "])), _HAXFields.HaxSchematizer, _HAXFields.HaxElementizer);
      }
    }, {
      key: "tag",
      get: function get() {
        return "hax-schema-form";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Returned value from the form input.
           */
          initialValue: {
            type: Object,
            notify: true,
            value: {},
            observer: "_valueChanged"
          },
          value: {
            type: Object,
            notify: true,
            value: {}
          },

          /**
           * State of mode tabs.
           */
          modeTab: {
            type: String,
            observer: "_editorModeChanged"
          },

          /**
           * If this is the advancedForm or not. Default to not but slider allows
           * switching mode for the form to be presented.
           */
          advancedForm: {
            type: Boolean,
            value: false
          },

          /**
           * If we should show source view or not.
           */
          canEditSource: {
            type: Boolean
          },

          /**
           * Form key from hax to target.
           */
          formKey: {
            type: String,
            computed: "_computedFormKey(advancedForm)",
            observer: "_formKeyChanged"
          },

          /**
           * JSON Schema.
           */
          schema: {
            type: Object,
            notify: true
          },

          /**
           * JSON Schema.
           */
          configureSchema: {
            type: Object,
            value: {
              schema: {}
            }
          },

          /**
           * JSON Schema.
           */
          advancedSchema: {
            type: Object,
            value: {
              schema: {}
            }
          }
        };
      }
    }]);
    return HaxSchemaForm;
  }(_polymerElement.PolymerElement);

  _exports.HaxSchemaForm = HaxSchemaForm;
  window.customElements.define(HaxSchemaForm.tag, HaxSchemaForm);
});
