// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-kanban/lrnapp-studio-assignment-display.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@polymer/marked-element/marked-element.js"], function (_exports, _litElement, _markedElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioAssignmentDisplay = void 0;

  class LrnappStudioAssignmentDisplay extends _litElement.LitElement {
    /**
     * LitElement constructable styles enhancement
     */
    static get styles() {
      return [_litElement.css`
        :host {
          display: block;
          padding: 16px;
        }
        #assignment-body {
          font-size: 21px;
        }
      `];
    }

    render() {
      return _litElement.html`
      <h1>${this.assignment.attributes.title}</h1>
      <marked-element
        id="assignment-body"
        markdown="${this.assignment.attributes.body}"
      ></marked-element>
    `;
    }

    static get tag() {
      return "lrnapp-studio-assignment-display";
    }

    static get properties() {
      return {
        elmslnCourse: {
          type: String,
          attribute: "elmsln-course"
        },
        elmslnSection: {
          type: String,
          attribute: "elmsln-section"
        },
        basePath: {
          type: String,
          attribute: "base-path"
        },
        csrfToken: {
          type: String,
          attribute: "csrf-token"
        },
        endPoint: {
          type: String,
          attribute: "end-point"
        },
        assignment: {
          type: Object
        }
      };
    }

  }

  _exports.LrnappStudioAssignmentDisplay = LrnappStudioAssignmentDisplay;
  window.customElements.define(LrnappStudioAssignmentDisplay.tag, LrnappStudioAssignmentDisplay);
});
