// @lrnwebcomponents/haxcms-elements/lib/ui-components/navigation/site-menu-content.js
define(["exports", "../../../../page-contents-menu/page-contents-menu.js", "../../core/haxcms-site-store.js", "../../../../../mobx/lib/mobx.module.js", "../../../../../lit-element/lit-element.js"], function (_exports, _pageContentsMenu, _haxcmsSiteStore, _mobxModule, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SiteMenuContent = void 0;

  class SiteMenuContent extends _pageContentsMenu.PageContentsMenu {
    static get tag() {
      return "site-menu-content";
    }

    constructor() {
      super();
      this.hideIfEmpty = true;
      this.__disposer = this.__disposer ? this.__disposer : [];
      (0, _mobxModule.autorun)(reaction => {
        this.contentContainer = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.themeElement);
        setTimeout(() => {
          this.updateMenu();
        }, 10);

        this.__disposer.push(reaction);
      });
      (0, _mobxModule.autorun)(reaction => {
        let content = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.activeItemContent);
        setTimeout(() => {
          this.updateMenu();
        }, 10);

        this.__disposer.push(reaction);
      });
    }

    static get styles() {
      return [...super.styles, (0, _litElement.css)`
        :host {
          --page-contents-menu-link-hover: var(
            --haxcms-color,
            var(--simple-colors-default-theme-purple-7)
          );
        }
        :host([hide-if-empty][is-empty]) {
          display: none !important;
        }
      `];
    }

  }

  _exports.SiteMenuContent = SiteMenuContent;
  customElements.define(SiteMenuContent.tag, SiteMenuContent);
});
