// @lrnwebcomponents/haxcms-elements/lib/ui-components/navigation/site-top-menu.js
define(["exports","require","../../../../../@polymer/polymer/polymer-element.js","../../core/haxcms-site-store.js","../../../../../mobx/lib/mobx.module.js","../query/site-query.js","../../../../../@polymer/polymer/lib/elements/dom-repeat.js"],(function(e,t,i,o,s,a,n){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.SiteTopMenu=void 0,t=babelHelpers.interopRequireWildcard(t);
/**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class SiteTopMenu extends i.PolymerElement{static get tag(){return"site-top-menu"}constructor(){super(),this.__disposer=[],new Promise((e,i)=>t.default(["../../../../../@polymer/paper-icon-button/paper-icon-button.js"],e,i)),new Promise((e,i)=>t.default(["../../../../../@polymer/paper-button/paper-button.js"],e,i))}static get template(){return i.html`
      <style>
        :host {
          display: block;
          --site-top-menu-bg: var(--haxcms-color, #ffffff);
          --site-top-menu-indicator-arrow: 6px;
          transition: 0.2s opacity linear;
          opacity: 1;
        }
        :host([edit-mode]) {
          opacity: 0.2;
          pointer-events: none;
        }
        :host([sticky]) {
          position: fixed;
          top: 0;
          left: 0;
          right: 0;
          z-index: 1000;
          @apply --site-top-menu-sticky;
        }
        .wrapper {
          display: flex;
          justify-content: space-evenly;
          background-color: var(--site-top-menu-bg);
          @apply --site-top-menu-wrapper;
        }
        :host .wrapper ::slotted(div.spacing) {
          display: inline-flex;
          @apply --site-top-menu-spacing;
        }
        .spacing {
          display: inline-flex;
          @apply --site-top-menu-spacing;
        }
        .link {
          color: var(--site-top-menu-link-color, #444444);
          @apply --site-top-menu-link;
        }
        paper-button {
          text-transform: unset;
          min-width: unset;
          @apply --site-top-menu-button;
        }
        .active {
          color: var(--site-top-menu-link-active-color, #000000);
          @apply --site-top-menu-link-active;
        }
        #indicator {
          transition: 0.4s ease-in-out left;
          transition-delay: 0.2s;
          position: relative;
          width: 0;
          height: 0;
          visibility: hidden;
        }
        :host([indicator="line"]) #indicator {
          border-bottom: 2px solid var(--site-top-menu-indicator-color, #000000);
          @apply --site-top-menu-indicator;
        }
        :host([indicator="arrow"]) #indicator {
          border-left: var(--site-top-menu-indicator-arrow) solid transparent;
          border-right: var(--site-top-menu-indicator-arrow) solid transparent;
          border-bottom: var(--site-top-menu-indicator-arrow) solid
            var(--site-top-menu-indicator-color, #000000);
          @apply --site-top-menu-indicator;
        }
        #indicator.activated {
          visibility: visible;
          position: absolute;
          @apply --site-top-menu-indicator-activated;
        }
        :host([notitle]) .spacing .link-title {
          display: none;
        }
        .spacing .link-index {
          display: none;
        }
        :host([showindex]) .spacing .link-index {
          display: inline-flex;
        }
        .mobiletitle,
        paper-icon-button {
          display: none;
        }
        @media screen and (max-width: 640px) {
          .wrapper .spacing {
            display: none;
          }
          .wrapper .mobiletitle,
          .wrapper paper-icon-button {
            display: inline-block;
          }
          .wrapper {
            display: block;
          }
        }

        @media screen and (max-width: 640px) {
          #indicator {
            display: none !important;
          }
          .wrapper.responsive {
            position: relative;
          }
          .wrapper.responsive .spacing {
            float: none;
            display: block;
            text-align: left;
          }
        }
      </style>
      <site-query
        result="{{__items}}"
        sort="[[sort]]"
        conditions="[[conditions]]"
      ></site-query>
      <div id="wrapper" class="wrapper">
        <paper-icon-button
          icon="menu"
          id="menu"
          title="Open navigation"
        ></paper-icon-button>
        <span class="mobiletitle">[[mobileTitle]]</span>
        <slot name="prefix"></slot>
        <dom-repeat items="[[__items]]" mutable-data>
          <template>
            <div class="spacing">
              <a
                data-id$="[[item.id]]"
                class="link"
                tabindex="-1"
                title$="Go to [[item.title]]"
                href$="[[item.location]]"
              >
                <paper-button id$="item-[[item.id]]" noink="[[noink]]">
                  <span class="link-index">[[humanIndex(index)]]</span>
                  <span class="link-title">[[item.title]]</span>
                </paper-button>
              </a>
            </div>
          </template>
        </dom-repeat>
        <slot name="suffix"></slot>
      </div>
      <div id="indicator"></div>
    `}static get properties(){return{manifest:{type:Object},activeId:{type:String,observer:"_activeIdChanged"},sticky:{type:Boolean,reflectToAttribute:!0,value:!1},indicator:{type:String,reflectToAttribute:!0,value:"line"},noink:{type:Boolean,reflectToAttribute:!0,value:!1},notitle:{type:Boolean,reflectToAttribute:!0,value:!1},showindex:{type:Boolean,reflectToAttribute:!0,value:!1},arrowSize:{type:Number,value:6},sort:{type:Object,value:{order:"ASC"}},conditions:{type:Object,value:{parent:null}},mobileTitle:{type:String,value:"Navigation"},editMode:{type:Boolean,reflectToAttribute:!0}}}humanIndex(e){return e+1}toggleOpen(){var e=this.shadowRoot.querySelector("#wrapper");e.classList.contains("responsive")?e.classList.remove("responsive"):e.classList.add("responsive")}_activeIdChanged(e){if("none"!=this.indicator)if(e){this.shadowRoot.querySelector("#indicator").classList.add("activated");let t=null;if(this.shadowRoot.querySelector('[data-id="'+e+'"]'))t=this.shadowRoot.querySelector('[data-id="'+e+'"]');else{let i=this.manifest.items.find(t=>t.id==e);for(;null===t&&i&&null!=i.parent;)i=this.manifest.items.find(e=>e.id==i.parent),i&&this.shadowRoot.querySelector('[data-id="'+i.id+'"]')&&(t=this.shadowRoot.querySelector('[data-id="'+i.id+'"]'))}this._prevEl&&this._prevEl.classList.remove("active"),t&&(t.classList.add("active"),this._prevEl=t,"arrow"==this.indicator?(this.shadowRoot.querySelector("#indicator").style.left=t.offsetLeft+t.offsetWidth/2-this.arrowSize+"px",this.shadowRoot.querySelector("#indicator").style.top=t.offsetTop+t.offsetHeight-this.arrowSize+"px"):(this.shadowRoot.querySelector("#indicator").style.left=t.offsetLeft+"px",this.shadowRoot.querySelector("#indicator").style.top=t.offsetTop+t.offsetHeight+"px",this.shadowRoot.querySelector("#indicator").style.width=t.offsetWidth+"px"))}else this.shadowRoot.querySelector("#indicator").classList.remove("activated")}connectedCallback(){super.connectedCallback(),this.shadowRoot.querySelector("#menu").addEventListener("click",this.toggleOpen.bind(this)),(0,s.autorun)(e=>{this.manifest=(0,s.toJS)(o.store.manifest),this.__disposer.push(e)}),(0,s.autorun)(e=>{this.editMode=(0,s.toJS)(o.store.editMode),this.__disposer.push(e)}),setTimeout(()=>{(0,s.autorun)(e=>{this.activeId=(0,s.toJS)(o.store.activeId),this.__disposer.push(e)})},50),window.addEventListener("resize",()=>{this._activeIdChanged(this.activeId)},!0)}disconnectedCallback(){for(var e in this.__disposer)this.__disposer[e].dispose();window.removeEventListener("resize",()=>{this._activeIdChanged(this.activeId)},!0),super.disconnectedCallback()}}e.SiteTopMenu=SiteTopMenu,window.customElements.define(SiteTopMenu.tag,SiteTopMenu)}));
