// @lrnwebcomponents/haxcms-elements/lib/ui-components/navigation/site-breadcrumb.js
define(["exports","require","../../../../../lit-element/lit-element.js","../../core/haxcms-site-store.js","../../../../../mobx/lib/mobx.module.js"],(function(e,t,i,r,o){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.SiteBreadcrumb=void 0,t=babelHelpers.interopRequireWildcard(t);
/**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class SiteBreadcrumb extends i.LitElement{static get styles(){return[i.css`
        :host {
          display: block;
          font-size: 16px;
          color: var(--site-breadcrumb-color, #383f45);
        }
        #space {
          height: 24px;
          line-height: 24px;
        }
        a {
          height: 24px;
          color: var(--site-breadcrumb-color, #383f45);
          display: inline-flex;
          line-height: 24px;
          padding: 0 8px 0 0;
          text-decoration: var(--site-breadcrumb-text-decoration, underline);
        }
        paper-button {
          margin: 0;
          padding: 0;
          height: 24px;
          min-width: unset;
          display: inline-flex;
          text-transform: unset;
        }
        paper-button:hover,
        paper-button:focus,
        paper-button:active {
          background-color: var(--site-breadcrumb-hover-bg, transparent);
          color: var(--site-breadcrumb-hover-color, #222222);
        }
        span {
          margin: 0;
          padding: 0 8px 0 0;
          height: 24px;
          display: inline-flex;
        }
        iron-icon {
          display: inline-flex;
          height: 24px;
          width: 24px;
          padding: 0 8px 0 0;
          color: var(--site-breadcrumb-color, #383f45);
        }
      `]}static get tag(){return"site-breadcrumb"}constructor(){super(),this.__disposer=[],new Promise((e,i)=>t.default(["../../../../../@polymer/iron-icon/iron-icon.js"],e,i)),new Promise((e,i)=>t.default(["../../../../../@polymer/paper-button/paper-button.js"],e,i)),(0,o.autorun)(e=>{this.manifest=(0,o.toJS)(r.store.routerManifest),this.__disposer.push(e)}),(0,o.autorun)(e=>{this.editMode=(0,o.toJS)(r.store.editMode),this.__disposer.push(e)}),(0,o.autorun)(e=>{this._activeItemChanged((0,o.toJS)(r.store.activeItem)),this.__disposer.push(e)})}render(){return i.html`
      <div
        id="space"
        itemscope
        itemtype="http://data-vocabulary.org/Breadcrumb"
      ></div>
    `}_activeItemChanged(e){const t=e;if(t&&this.shadowRoot.querySelector("#space")){for(;null!==this.shadowRoot.querySelector("#space").firstChild;)this.shadowRoot.querySelector("#space").removeChild(this.shadowRoot.querySelector("#space").firstChild);var i=[{title:t.title,location:null}];let e=t;for(;e&&null!=e.parent;)e=this.manifest.items.find(t=>t.id==e.parent),e&&i.unshift({title:e.title,location:e.location});for(var r in i){let e=document.createElement("iron-icon");if(e.icon="icons:chevron-right",null!=i[r].location){let t=document.createElement("paper-button");if(t.innerText=i[r].title,t.noink=!0,this.editMode)t.setAttribute("disabled","disabled"),this.shadowRoot.querySelector("#space").appendChild(t);else{let e=document.createElement("a");e.setAttribute("href",i[r].location),e.setAttribute("tabindex","-1"),e.setAttribute("itemprop","url"),e.appendChild(t),this.shadowRoot.querySelector("#space").appendChild(e)}this.shadowRoot.querySelector("#space").appendChild(e)}else{let e=document.createElement("span");e.innerText=i[r].title,this.shadowRoot.querySelector("#space").appendChild(e)}}}}disconnectedCallback(){for(var e in this.__disposer)this.__disposer[e].dispose();super.disconnectedCallback()}}e.SiteBreadcrumb=SiteBreadcrumb,window.customElements.define(SiteBreadcrumb.tag,SiteBreadcrumb)}));
