// @haxtheweb/outline-designer/outline-designer.js
/**
 * Copyright 2022 The Pennsylvania State University
 * @license Apache-2.0, see License.md for full text.
 */
import{LitElement as e,css as t,html as i}from"../../lit/index.js";import{I18NMixin as o}from"../i18n-manager/lib/I18NMixin.js";import{unsafeHTML as a}from"../../lit/directives/unsafe-html.js";import{HAXStore as s}from"../hax-body/lib/hax-store.js";import{store as n}from"../haxcms-elements/lib/core/haxcms-site-store.js";import{autorun as d,toJS as r}from"../../mobx/dist/mobx.esm.js";import{JSONOutlineSchemaItem as l}from"../json-outline-schema/lib/json-outline-schema-item.js";import{JsonOutlineSchema as c}from"../json-outline-schema/json-outline-schema.js";import"../simple-popover/simple-popover.js";import"../simple-icon/lib/simple-icon-lite.js";import"../hax-iconset/lib/simple-hax-iconset.js";import"../simple-icon/lib/simple-icon-button-lite.js";import"../simple-fields/lib/simple-fields-field.js";import"../simple-fields/lib/simple-context-menu.js";import"../simple-toolbar/lib/simple-toolbar-button.js";import{encapScript as h,haxElementToNode as p}from"../utils/utils.js";export class OutlineDesigner extends(o(e)){static get styles(){return[t`
        :host {
          display: block;
          font-family: var(--ddd-font-navigation);
          --outline-designer-zoom: 1;
        }
        :host(.zoom-out-1) {
          --outline-designer-zoom: 0.85;
        }
        :host(.zoom-out-2) {
          --outline-designer-zoom: 0.7;
        }
        :host(.zoom-out-3) {
          --outline-designer-zoom: 0.55;
        }
        .item {
          transform: scale(var(--outline-designer-zoom));
          transform-origin: left center;
          margin-bottom: calc(
            var(--ddd-spacing-1) * var(--outline-designer-zoom)
          );
        }
        simple-icon-button[hidden] {
          visibility: hidden !important;
          opacity: 0;
          pointer-events: none;
          padding: 0;
          margin: 0;
          height: 0;
          border: 0;
        }
        .controls {
          position: sticky;
          top: -32px;
          background-color: light-dark(
            var(--ddd-accent-6),
            var(--ddd-primary-4)
          );
          z-index: 1;
          padding: var(--ddd-spacing-4) 0 var(--ddd-spacing-2) 0;
          border-bottom: var(--ddd-border-xs);
          border-color: light-dark(
            var(--ddd-theme-default-limestoneLight),
            var(--ddd-primary-5)
          );
        }
        .controls simple-toolbar-button.control {
          --simple-toolbar-button-border-width: 1px;
          --simple-toolbar-border-color: light-dark(
            var(--ddd-primary-4),
            var(--ddd-accent-6)
          );
          --simple-toolbar-border-radius: var(--ddd-radius-xs);
          --simple-toolbar-button-padding: var(--ddd-spacing-1);
          background-color: light-dark(
            var(--ddd-accent-6),
            var(--ddd-primary-3)
          );
          color: light-dark(var(--ddd-primary-4), var(--ddd-accent-6));
        }
        .controls simple-toolbar-button.control:hover {
          background-color: light-dark(
            var(--ddd-theme-default-limestoneMaxLight),
            var(--ddd-primary-5)
          );
          --simple-toolbar-border-color: light-dark(
            var(--ddd-primary-1),
            var(--ddd-accent-5)
          );
        }
        simple-popover {
          --simple-popover-max-height: 300px;
        }
        simple-popover *:not(.close-btn) {
          max-width: 40vw;
        }
        simple-popover::part(simple-popover-content) {
          overflow: auto;
        }
        .close-btn {
          z-index: 1000;
          background-color: light-dark(
            var(--ddd-accent-6),
            var(--ddd-primary-4)
          );
          color: light-dark(var(--ddd-primary-4), var(--ddd-accent-6));
          border: var(--ddd-border-xs);
          border-color: light-dark(var(--ddd-primary-5), var(--ddd-accent-5));
          border-radius: 50%;
          position: absolute;
          top: 24px;
          right: 0;
        }
        .container {
          text-align: left;
        }
        ul {
          list-style: none;
          padding: 0;
          margin: 0;
        }
        ul li {
          margin: 0;
          padding: 0;
        }
        .operation {
          display: inline-flex;
          --simple-icon-width: 24px;
          --simple-icon-height: 24px;
          margin: 0 4px;
        }
        .content-adding-operations .operation {
          display: inline-flex;
          --simple-icon-width: 24px;
          --simple-icon-height: 24px;
          margin: 0 var(--ddd-spacing-1);
          border: var(--ddd-border-xs);
          border-color: light-dark(var(--ddd-primary-4), var(--ddd-accent-6));
          border-radius: var(--ddd-radius-xs);
          padding: var(--ddd-spacing-1);
          background-color: light-dark(
            var(--ddd-accent-6),
            var(--ddd-primary-3)
          );
          color: light-dark(var(--ddd-primary-4), var(--ddd-accent-6));
        }
        .content-adding-operations .operation:hover {
          background-color: light-dark(
            var(--ddd-theme-default-limestoneMaxLight),
            var(--ddd-primary-5)
          );
          border-color: light-dark(var(--ddd-primary-1), var(--ddd-accent-5));
        }
        .lock {
          margin-right: 16px !important;
          visibility: var(--outline-designer-lock-visibility);
        }
        .del {
          margin-left: 32px !important;
        }
        .goto {
          margin-left: 32px !important;
        }
        .add {
          margin-left: 16px !important;
        }
        li .operations {
          justify-content: space-evenly;
          display: flex;
        }
        .content-operation {
          display: inline-flex;
          opacity: 0;
          visibility: hidden;
          --simple-icon-width: 24px;
          --simple-icon-height: 24px;
          margin: 0 4px;
        }
        .content-operations {
          justify-content: space-evenly;
          display: flex;
        }
        li.content-child:hover .content-operation,
        li.content-child:focus-within .content-operation {
          visibility: visible;
          opacity: 1;
        }
        li[class*="collapsed-by-"] {
          opacity: 0;
          height: 0px !important;
          visibility: hidden;
          padding: 0px !important;
          margin: 0px !important;
          width: 0px !important;
          padding: 0px !important;
          margin: 0px !important;
          border: 0px !important;
          pointer-events: none;
          z-index: -1;
        }
        /* content not rendered if hidden but in case we change that */
        li[data-contents-collapsed] {
          opacity: 0;
          height: 0px !important;
          visibility: hidden;
          padding: 0px !important;
          margin: 0px !important;
          width: 0px !important;
          padding: 0px !important;
          margin: 0px !important;
          border: 0px !important;
          pointer-events: none;
          z-index: -1;
        }
        li simple-icon-button-lite:hover {
          background-color: light-dark(
            var(--ddd-theme-default-limestoneMaxLight),
            var(--ddd-primary-5)
          );
        }
        .active-preview-item {
          outline: var(--ddd-border-xs);
          outline-color: light-dark(var(--ddd-primary-5), var(--ddd-accent-5));
          outline-offset: -1px;
        }
        .label,
        .label-edit {
          display: none;
        }
        span[disabled].label {
          pointer-events: none;
          opacity: 0.6;
        }
        .shown {
          display: inline-block;
        }
        .outline-designer-hovered {
          outline: var(--ddd-border-sm);
          outline-color: light-dark(var(--ddd-primary-4), var(--ddd-accent-6));
          outline-offset: -1px;
          background-color: light-dark(
            var(--ddd-theme-default-limestoneMaxLight),
            var(--ddd-primary-5)
          );
        }
        .make-child-btn {
          transition: 0.3s all ease-in-out;
          visibility: hidden;
          opacity: 0;
        }
        .outline-designer-hovered .make-child-btn {
          visibility: visible;
          opacity: 0.6;
        }
        .outline-designer-hovered .make-child-btn:hover {
          opacity: 1;
        }
        .modified .label::after {
          content: "*";
          color: light-dark(var(--ddd-primary-22), var(--ddd-primary-12));
          font-size: 20px;
          line-height: 20px;
        }
        .new {
          --simple-icon-width: 16px;
          --simple-icon-height: 16px;
          background-color: light-dark(
            var(--ddd-primary-4),
            var(--ddd-accent-6)
          );
          color: light-dark(var(--ddd-accent-6), var(--ddd-primary-4));
          display: block;
          margin: -14px 0 0 4px;
        }
        :host([stop-animation]) .item {
          transition: none !important;
        }
        .item {
          display: -webkit-box;
          border: var(--ddd-border-xs);
          border-color: light-dark(var(--ddd-primary-5), var(--ddd-accent-5));
          margin: 0;
          padding: var(--ddd-spacing-1);
          cursor: pointer;
          opacity: 1;
          visibility: visible;
          height: 42px;
          transition:
            0.3s padding ease-in-out,
            0.3s border ease-in-out,
            0.3s margin ease-in-out;
          overflow: hidden;
          align-items: center;
          justify-content: left;
          display: flex;
        }
        .collapse-btn {
          visibility: hidden;
        }
        .item[data-has-children] .collapse-btn {
          visibility: visible;
        }
        .item[data-about-to-delete] {
          background-color: light-dark(
            var(--ddd-theme-default-errorLight),
            var(--ddd-primary-22)
          );
          opacity: 0.8;
          border-color: light-dark(
            var(--ddd-primary-22),
            var(--ddd-primary-12)
          );
        }
        .item[data-about-to-delete][hidden] {
          visibility: hidden !important;
          opacity: 0 !important;
          padding: 0;
          margin: 0;
          height: 0;
          border: 0;
        }
        .item:hover,
        .item:focus,
        .item:focus-within {
          background-color: light-dark(
            var(--ddd-theme-default-limestoneMaxLight),
            var(--ddd-primary-5)
          );
          outline: var(--ddd-border-sm);
          outline-color: light-dark(var(--ddd-primary-1), var(--ddd-accent-5));
          outline-offset: -2px;
          box-shadow: var(--ddd-boxShadow-sm);
        }
        .item.active-page {
          background-color: light-dark(
            var(--ddd-accent-5),
            var(--ddd-primary-6)
          );
          outline: 2px solid
            light-dark(var(--ddd-primary-2), var(--ddd-accent-4));
          outline-offset: -2px;
          box-shadow: var(--ddd-boxShadow-md);
        }
        .item.selected-page {
          background-color: light-dark(
            var(--ddd-accent-4),
            var(--ddd-primary-7)
          );
          outline: 2px solid
            light-dark(var(--ddd-primary-3), var(--ddd-accent-3));
          outline-offset: -2px;
        }
        .item[data-dragging="true"] {
          box-shadow: var(--ddd-boxShadow-lg);
          opacity: 0.8;
          transform: scale(1.02);
          z-index: 1000;
        }
        .item[data-dragging="true"] {
          box-shadow: var(--ddd-boxShadow-lg);
          opacity: 0.8;
          transform: scale(1.02);
          z-index: 1000;
        }
        /* Menu button styling */
        .actions-menu {
          margin-left: var(--ddd-spacing-2);
        }
        .actions-menu::part(button) {
          padding: var(--ddd-spacing-1);
          border-radius: var(--ddd-radius-xs);
          border: var(--ddd-border-xs);
          background: light-dark(var(--ddd-accent-6), var(--ddd-primary-3));
          color: light-dark(var(--ddd-primary-4), var(--ddd-accent-6));
        }
        .actions-menu::part(button):hover {
          background: light-dark(
            var(--ddd-theme-default-limestoneMaxLight),
            var(--ddd-primary-5)
          );
        }
        .actions-menu simple-toolbar-button {
          text-align: left;
          justify-content: flex-start;
          padding-left: var(--ddd-spacing-2);
          padding-right: var(--ddd-spacing-4);
          --simple-toolbar-button-justify: flex-start;
          --simple-icon-height: 16px;
          --simple-icon-width: 16px;
        }
        .actions-menu simple-toolbar-button.delete-button {
          border-top: var(--ddd-border-sm) solid
            var(--ddd-theme-default-limestoneGray);
          margin-top: var(--ddd-spacing-2);
          padding-top: var(--ddd-spacing-2);
        }
        .actions-menu simple-toolbar-button.delete-button:hover {
          color: var(--ddd-theme-default-error);
          background-color: var(--ddd-theme-default-errorLight);
        }

        simple-icon-button-lite {
          color: light-dark(var(--ddd-primary-4), var(--ddd-accent-6));
        }

        simple-context-menu {
          text-align: left;
        }
        simple-context-menu::part(content) {
          text-align: left;
          align-items: flex-start;
        }
        .item:hover .label,
        .item:focus .label,
        .item:focus-within .label {
          color: light-dark(var(--ddd-primary-4), var(--ddd-accent-6));
        }
        ul {
          list-style: none;
        }
        .item .label-edit,
        .item .label {
          cursor: text;
          font-size: 14px;
          font-weight: bold;
          min-width: 200px;
          margin-right: 8px;
          max-width: 40%;
          line-height: 1.2;
          padding: 0 4px;
          color: light-dark(
            var(--ddd-theme-default-originalBlack),
            var(--ddd-theme-default-white)
          );
        }

        .content-child {
          margin-left: 46px;
          padding: 8px;
          height: 24px;
          border-top: none;
          border-bottom: none;
        }
        .content-heading,
        .content-non-heading {
          margin-left: 32px;
        }
        .indent-0 {
          padding-left: 0;
          position: relative;
        }
        .indent-1 {
          padding-left: calc(16px * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-2 {
          padding-left: calc(16px * 2 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-3 {
          padding-left: calc(16px * 3 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-4 {
          padding-left: calc(16px * 4 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-5 {
          padding-left: calc(16px * 5 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-6 {
          padding-left: calc(16px * 6 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-7 {
          padding-left: calc(16px * 7 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-8 {
          padding-left: calc(16px * 8 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-9 {
          padding-left: calc(16px * 9 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-10 {
          padding-left: calc(16px * 10 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-11 {
          padding-left: calc(16px * 11 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-12 {
          padding-left: calc(16px * 12 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-13 {
          padding-left: calc(16px * 13 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-14 {
          padding-left: calc(16px * 14 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-15 {
          padding-left: calc(16px * 15 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-16 {
          padding-left: calc(16px * 16 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-17 {
          padding-left: calc(16px * 17 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-18 {
          padding-left: calc(16px * 18 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-19 {
          padding-left: calc(16px * 19 * var(--outline-designer-zoom));
          position: relative;
        }
        .indent-20 {
          padding-left: calc(16px * 20 * var(--outline-designer-zoom));
          position: relative;
        }
        .sr-only {
          position: absolute;
          left: -10000px;
          top: auto;
          width: 1px;
          height: 1px;
          overflow: hidden;
        }
        /* Drop zone indicators for drag and drop - kanban style */
        .drop-indicator {
          height: calc(42px * var(--outline-designer-zoom, 1));
          background: light-dark(
            rgba(33, 150, 243, 0.08),
            rgba(33, 150, 243, 0.12)
          );
          border: 2px dashed var(--ddd-accent-4);
          border-radius: var(--ddd-radius-sm);
          margin: var(--ddd-spacing-1) 0;
          position: relative;
          z-index: 100;
          opacity: 0;
          transform: scaleY(0);
          transform-origin: top;
          transition: all 0.15s ease-out;
        }
        .drop-indicator.show {
          opacity: 1;
          transform: scaleY(1);
        }
        .drop-indicator.indent-visual {
          border-left: 4px solid var(--ddd-accent-3);
          background: light-dark(
            rgba(33, 150, 243, 0.15),
            rgba(33, 150, 243, 0.2)
          );
        }
        .drop-target-child {
          outline: 2px dashed var(--ddd-accent-4) !important;
          outline-offset: 2px;
          background: light-dark(
            rgba(135, 206, 235, 0.15),
            rgba(135, 206, 235, 0.1)
          ) !important;
        }
        /* Drag preview styles */
        .drag-preview {
          position: absolute;
          top: -1000px;
          left: -1000px;
          padding: 8px 12px;
          background: light-dark(var(--ddd-accent-6), var(--ddd-primary-3));
          border: 2px solid
            light-dark(var(--ddd-primary-4), var(--ddd-accent-6));
          border-radius: var(--ddd-radius-sm);
          font-weight: var(--ddd-font-weight-bold);
          box-shadow: var(--ddd-boxShadow-lg);
          z-index: 10000;
          pointer-events: none;
          display: flex;
          align-items: center;
          gap: var(--ddd-spacing-2);
        }
        /* Stacked effect for items with children */
        .drag-preview.has-children {
          position: relative;
        }
        .drag-preview.has-children::before,
        .drag-preview.has-children::after {
          content: "";
          position: absolute;
          left: 3px;
          right: -3px;
          height: 100%;
          background: light-dark(var(--ddd-accent-6), var(--ddd-primary-3));
          border: 2px solid
            light-dark(var(--ddd-primary-4), var(--ddd-accent-6));
          border-radius: var(--ddd-radius-sm);
          z-index: -1;
        }
        .drag-preview.has-children::before {
          top: -3px;
          opacity: 0.7;
        }
        .drag-preview.has-children::after {
          top: -6px;
          opacity: 0.4;
        }
        .drag-preview-children {
          opacity: 0.7;
          font-size: 12px;
          font-weight: var(--ddd-font-weight-regular);
        }
        .drag-handle-active {
          outline: 2px solid
            light-dark(var(--ddd-accent-4), var(--ddd-accent-3));
          outline-offset: 2px;
          background-color: light-dark(
            var(--ddd-accent-5),
            var(--ddd-primary-6)
          );
        }
        /* Push-aside animation for drop target */
        .item.drop-above-target {
          margin-top: 48px;
          transition: margin-top 0.2s ease-out;
        }
        .item.drop-below-target {
          margin-bottom: 48px;
          transition: margin-bottom 0.2s ease-out;
        }
      `]}constructor(){super(),this._blurBlock=!1,this.fidelity="medium",this.haxGizmos=[],this.hideDelete=!1,this.activeItemForActions=null,this.storeTools=!1,this.hideContentOps=!1,this.items=[],this.appReady=!1,this.eventData={},this.activeId=null,this.activePreview=null,this.activePreviewIndex=-1,this.liveRegionText="",this._dropZone=null,this._dragPreviewElement=null,this.activePage=null,this.selectedPages=[],this.zoomLevel=1,this._dragHandleActive=null,this.t={selectTarget:"Select target",importContentUnderThisPage:"Import content under this page",importThisContent:"Import this content",thisPage:"this page",newPage:"New page",copyOf:"Copy of",pageActions:"Page Actions",editTitle:"Edit title",lock:"Lock",unlock:"Unlock",moveUp:"Move up",moveDown:"Move down",makeChild:"Make child",moveToParentLevel:"Move to parent level",addPage:"Add page",duplicate:"Duplicate",delete:"Delete",restore:"Restore",goToPage:"Go to page"},this.registerLocalization({context:this,basePath:import.meta.url}),d((()=>{this.activeId=r(n.activeId)})),d((()=>{this.appReady=r(n.appReady)})),d((()=>{this.haxGizmos=r(s.gizmoList).filter((e=>!!(e&&e.meta&&e.meta.outlineDesigner)))})),this.addEventListener("click",this.resetPopOver.bind(this))}resetPopOver(){this.activePreview&&(this.shadowRoot.querySelector("simple-popover").setAttribute("hidden","hidden"),this.activePreview=null,this.activePreviewIndex=-1)}getSiteItems(){var e=[{text:this.t.selectTarget,value:null}];if(this.appReady&&this.items.length>0){const t=n.getManifestItems(!0);t.forEach((i=>{let o=i,a="- ";for(;o&&null!=o.parent;)o=t.find((e=>e.id==o.parent)),o&&(a="--"+a);e.push({text:a+i.title,value:i.id})}))}return e}render(){return i` <div class="controls">
        ${this.storeTools?i`
              <label for="targetselector">${this.t.importThisContent}</label>
              <simple-fields-field
                id="targetselector"
                type="select"
                value="children"
                .itemsList="${[{text:"as children of",value:"children"},{text:"Above",value:"above"},{text:"Below",value:"below"}]}"
              ></simple-fields-field>
              ${this.t.thisPage}:
              <simple-fields-field
                id="itemselector"
                type="select"
                value="${this.activeId}"
                .itemsList="${this.getSiteItems()}"
              ></simple-fields-field>
              <label for="itemselector"
                >${this.t.importContentUnderThisPage}</label
              >
            `:""}
        <simple-toolbar-button
          class="control"
          icon="add"
          @click="${this.addItemToTop}"
          label="Add page"
        ></simple-toolbar-button>
        <simple-toolbar-button
          icon="hardware:keyboard-arrow-right"
          @click="${this.collapseAll}"
          class="control"
          label="Collapse all"
        ></simple-toolbar-button>
        <simple-toolbar-button
          icon="hardware:keyboard-arrow-down"
          @click="${this.expandAll}"
          class="control"
          label="Expand all"
        ></simple-toolbar-button>
        ${this.selectedPages.length>0?i`<simple-toolbar-button
              icon="delete"
              @click="${this.deleteSelected}"
              class="control"
              label="Delete Selected (${this.selectedPages.length})"
            ></simple-toolbar-button>`:""}
        ${this.hasDeletedItems()?i`<simple-toolbar-button
              icon="delete"
              @click="${this.toggleDelete}"
              class="control"
              label="${this.hideDelete?"Show Deleted":"Hide Deleted"}"
            ></simple-toolbar-button>`:""}
        <simple-toolbar-button
          icon="zoom-in"
          @click="${this.zoomIn}"
          class="control"
          ?disabled="${this.zoomLevel>=1}"
          label="Zoom In"
        ></simple-toolbar-button>
        <simple-toolbar-button
          icon="zoom-out"
          @click="${this.zoomOut}"
          class="control"
          ?disabled="${this.zoomLevel<=-3}"
          label="Zoom Out"
        ></simple-toolbar-button>
      </div>
      <ul
        id="list"
        role="tree"
        aria-label="Outline structure of pages and content"
      >
        ${this.items.map(((e,t)=>""===this.getItemParentsCollapsed(e)?this.renderItem(e,t):""))}
      </ul>
      <div aria-live="polite" aria-atomic="true" class="sr-only">
        ${this.liveRegionText}
      </div>
      <simple-popover auto for="list" hidden>
        <simple-icon-button-lite
          @click="${this.resetPopOver}"
          title="Close"
          icon="cancel"
          class="close-btn"
        ></simple-icon-button-lite>
        ${this.renderActiveContentItem(this.activePreview,this.activePreviewIndex)}
      </simple-popover>`}hasDeletedItems(){return!!this.items.find((e=>1==e.delete))}toggleDelete(e){this.hideDelete=!this.hideDelete}renderActiveContentItem(e,t){if(e&&-1!=t){let o=this.items.find((t=>t.id===e.getAttribute("data-content-parent-id")));if(o.contents){let e=globalThis.document.createElement("div");e.innerHTML=o.contents;for(let o=0;o<e.childNodes.length;o++){let s=e.childNodes[o];if(o===t)return i`${a(h(s.outerHTML))}`}}}}setActiveItemForActions(e){this.activeItemForActions=e.target.closest("[data-item-id]").getAttribute("data-item-id")}handleItemClick(e){const t=e.target.closest("[data-item-id]");if(!t)return;const i=t.getAttribute("data-item-id");e.ctrlKey||e.metaKey?this.togglePageSelection(i):e.shiftKey&&this.activePage?this.selectPageRange(this.activePage,i):e.target.closest("simple-toolbar-button")||e.target.closest("simple-icon-button-lite")||e.target.closest(".label-edit")||(this.activePage=i,this.requestUpdate())}togglePageSelection(e){const t=this.selectedPages.indexOf(e);t>-1?this.selectedPages.splice(t,1):this.selectedPages.push(e),this.activePage=e,this.requestUpdate()}selectPageRange(e,t){const i=this.items.findIndex((t=>t.id===e)),o=this.items.findIndex((e=>e.id===t));if(-1===i||-1===o)return;const a=Math.min(i,o),s=Math.max(i,o);this.selectedPages=[];for(let e=a;e<=s;e++)""===this.getItemParentsCollapsed(this.items[e])&&this.selectedPages.push(this.items[e].id);this.requestUpdate()}deleteSelected(){this.selectedPages.forEach((e=>{const t=this.items.findIndex((t=>t.id===e));-1===t||this.isLocked(t)||(this.items[t].delete=!0,this.hasChildren(this.items[t].id)&&this.recurseAction(this.items[t].id,"delete",!0))})),this.selectedPages=[],this.__syncUIAndDataModel()}zoomIn(){this.zoomLevel<1&&(this.zoomLevel++,this.updateZoomClass())}zoomOut(){this.zoomLevel>-3&&(this.zoomLevel--,this.updateZoomClass())}updateZoomClass(){this.classList.remove("zoom-out-1","zoom-out-2","zoom-out-3"),0===this.zoomLevel?this.classList.add("zoom-out-1"):-1===this.zoomLevel?this.classList.add("zoom-out-2"):this.zoomLevel<=-2&&this.classList.add("zoom-out-3"),this.requestUpdate()}_toggleActionsMenu(e,t){e.stopPropagation();const i=e.target,o=i.closest("li").querySelector(".actions-menu");o&&o.toggle(i)}_handleMenuAction(e,t,i){e.stopPropagation();const o=e.target.closest("li"),a=o.querySelector(".actions-menu");if(a&&a.close(),"edit-title"===i&&!1!==t){const e=o.querySelector(".label.shown");e&&!this.isLocked(t)&&this.editTitle({target:e})}else i&&!1!==t&&this.itemOp(t,i)}renderItem(e,t){return i`
      <li
        role="treeitem"
        tabindex="${0===t?"0":"-1"}"
        aria-label="${e.title}${e.modified?" (modified)":""}${e.delete?" (marked for deletion)":""}"
        aria-expanded="${this.hasChildren(e.id)?this.isCollapsed(e.id)?"false":"true":"undefined"}"
        aria-level="${e.indent+1}"
        aria-setsize="${this.items.filter((t=>t.parent===e.parent)).length}"
        aria-posinset="${this.items.filter((t=>t.parent===e.parent)).indexOf(e)+1}"
        @keydown="${this.handleTreeItemKeydown}"
        @dragenter="${this._dragEnter}"
        @dragleave="${this._dragLeave}"
        @mouseenter="${this.setActiveItemForActions}"
        @click="${this.handleItemClick}"
        class="item indent-${e.indent<20?e.indent:20} ${e.modified?"modified":""} ${this.getItemParentsCollapsed(e)} ${this.activePage===e.id?"active-page":""} ${this.selectedPages.includes(e.id)?"selected-page":""}"
        data-item-id="${e.id}"
        @focusin="${this.setActiveItemForActions}"
        data-parents="${this.getItemParents(e)}"
        ?data-has-children="${this.hasChildren(e.id)}"
        ?data-about-to-delete="${e.delete}"
        ?hidden="${this.hideDelete&&e.delete}"
      >
        <simple-toolbar-button
          class="actions-menu-button"
          icon="icons:more-vert"
          label="${this.t.pageActions}"
          @click="${e=>this._toggleActionsMenu(e,t)}"
        ></simple-toolbar-button>
        <simple-icon-button-lite
          ?disabled="${this.isLocked(t)}"
          class="collapse-btn"
          icon="${this.isCollapsed(e.id)?"hardware:keyboard-arrow-right":"hardware:keyboard-arrow-down"}"
          @click="${this.collapseExpand}"
        ></simple-icon-button-lite>
        <simple-icon-button-lite
          ?disabled="${this.isLocked(t)}"
          @dragstart="${this._dragStart}"
          @dragend="${this._dragEnd}"
          @drag="${this._onDrag}"
          @keydown="${this._handleDragHandleKeydown}"
          draggable="${!this.isLocked(t)}"
          icon="icons:reorder"
          class="drag-handle ${this._dragHandleActive===e.id?"drag-handle-active":""}"
          title="Drag to reorder or press Enter to activate keyboard controls"
          data-drag-handle-id="${e.id}"
        ></simple-icon-button-lite>
        <simple-icon-button-lite
          ?disabled="${this.isLocked(t)}"
          ?hidden="${this.hideContentOps||""===e.contents||!e.contents}"
          icon="editor:insert-drive-file"
          @click="${this.toggleContent}"
          title="Content structure"
        >
          ${e.new?i`<simple-icon
                ?disabled="${this.isLocked(t)}"
                icon="av:fiber-new"
                title="${this.t.newPage}"
                class="new"
                accent-color="green"
                dark
                contrast="1"
              ></simple-icon>`:""}
        </simple-icon-button-lite>
        <span
          class="label shown"
          ?disabled="${this.isLocked(t)}"
          @dblclick="${this.editTitle}"
          @keydown="${this.handleLabelKeydown}"
          tabindex="0"
          >${e.title}</span
        >
        <span
          class="label-edit"
          @blur="${this.blurTitle}"
          @keypress="${this.monitorTitle}"
          @keydown="${this.monitorEsc}"
        ></span>
        <simple-context-menu
          class="actions-menu"
          data-item-index="${t}"
          title="${this.t.pageActions}"
        >
          <simple-toolbar-button
            value="up"
            ?disabled="${this.isLocked(t)}"
            icon="icons:arrow-upward"
            label="${this.t.moveUp}"
            show-text-label
            @click="${e=>this._handleMenuAction(e,t,"up")}"
          ></simple-toolbar-button>
          <simple-toolbar-button
            value="down"
            ?disabled="${this.isLocked(t)}"
            icon="icons:arrow-downward"
            label="${this.t.moveDown}"
            show-text-label
            @click="${e=>this._handleMenuAction(e,t,"down")}"
          ></simple-toolbar-button>
          <simple-toolbar-button
            value="in"
            ?disabled="${this.isLocked(t)}"
            icon="hax:outline-designer-indent"
            label="Indent"
            show-text-label
            @click="${e=>this._handleMenuAction(e,t,"in")}"
          ></simple-toolbar-button>
          <simple-toolbar-button
            value="out"
            ?disabled="${this.isLocked(t)}"
            icon="hax:outline-designer-outdent"
            label="Outdent"
            show-text-label
            @click="${e=>this._handleMenuAction(e,t,"out")}"
          ></simple-toolbar-button>
          <simple-toolbar-button
            value="edit-title"
            ?disabled="${this.isLocked(t)}"
            icon="editor:mode-edit"
            label="${this.t.editTitle}"
            show-text-label
            @click="${e=>this._handleMenuAction(e,t,"edit-title")}"
            autofocus
          ></simple-toolbar-button>
          <simple-toolbar-button
            value="add"
            ?disabled="${this.isLocked(t)}"
            icon="hax:add-page"
            label="${this.t.addPage}"
            show-text-label
            @click="${e=>this._handleMenuAction(e,t,"add")}"
          ></simple-toolbar-button>
          <simple-toolbar-button
            value="duplicate"
            ?disabled="${this.isLocked(t)}"
            icon="content-copy"
            label="${this.t.duplicate}"
            show-text-label
            @click="${e=>this._handleMenuAction(e,t,"duplicate")}"
          ></simple-toolbar-button>
          <simple-toolbar-button
            value="goto"
            icon="open-in-browser"
            label="${this.t.goToPage}"
            show-text-label
            @click="${e=>this._handleMenuAction(e,t,"goto")}"
          ></simple-toolbar-button>
          <simple-toolbar-button
            value="lock"
            icon="${this.isLocked(t)?"icons:lock":"icons:lock-open"}"
            label="${this.isLocked(t)?this.t.unlock:this.t.lock}"
            show-text-label
            @click="${e=>this._handleMenuAction(e,t,"lock")}"
          ></simple-toolbar-button>
          <simple-toolbar-button
            class="delete-button"
            value="delete"
            ?disabled="${this.isLocked(t)}"
            icon="icons:delete"
            label="Delete"
            show-text-label
            @click="${e=>this._handleMenuAction(e,t,"delete")}"
          ></simple-toolbar-button>
        </simple-context-menu>
      </li>
      ${!this.hideContentOps&&e.showContent?this.renderItemContents(e):""}
    `}hasContents(e){return!(!e.contents||""==e.contents)}renderItemContents(e){let t=[this.itemContentsOperations(e)];if(e.contents){let i=globalThis.document.createElement("div");i.innerHTML=e.contents;let o=1,a=0;i.childNodes.forEach(((i,s)=>{a=0,["h1","h2","h3","h4","h5","h6"].includes(i.tagName.toLowerCase())&&(o=parseInt(i.tagName.toLowerCase().replace("h","")),a=-1),t.push(this.renderNodeAsItem(i,s,e,parseInt(e.indent)+o+a))}))}return t}itemContentsOperations(e){return i` <li
      class="item content-adding-operations indent-${e.indent<20?e.indent:20}"
      data-item-for-content-id="${e.id}"
      ?data-contents-collapsed="${!e.showContent}"
      ?data-about-to-delete="${e.delete}"
      ?hidden="${this.hideDelete&&e.delete}"
    >
      ${this.haxGizmos.map((e=>i`
          <simple-icon-button-lite
            class="operation"
            icon="${e.icon}"
            value="${e.tag}"
            @click="${this.prependNodeToContent}"
            >Add ${e.title}</simple-icon-button-lite
          >
        `))}
    </li>`}prependNodeToContent(e){let t=e.target.closest("[data-item-for-content-id]").getAttribute("data-item-for-content-id");this.items.map(((i,o)=>{if(i.id===t&&e.target.value){let t,a=s.haxSchemaFromTag(e.target.value);t=a.gizmo&&a.gizmo.tag&&a.demoSchema&&a.demoSchema[0]?p(a.demoSchema[0]):globalThis.document.createElement(tag),this.items[o].contents=t.outerHTML+i.contents,this.resetPopOver(),this.__syncUIAndDataModel()}}))}renderNodeAsItem(e,t,o,a){let n=e.tagName.toLowerCase(),d="hax:bricks",r=n,l="non-heading",c=s.haxSchemaFromTag(n);if(c&&c.gizmo)switch(d=c.gizmo.icon,r=c.gizmo.title,n){case"h1":case"h2":case"h3":case"h4":case"h5":case"h6":r=e.innerText,l="heading"}return i` <li
      class="item content-child content-${l} indent-${a<20?a:20}"
      data-node-index="${t}"
      data-content-parent-id="${o.id}"
      ?data-contents-collapsed="${!o.showContent}"
      ?data-about-to-delete="${o.delete}"
      ?hidden="${this.hideDelete&&o.delete}"
    >
      <div class="btn-contrast">
        <simple-icon-button-lite
          icon="${d}"
          title="Click to preview"
          @click="${this.setActivePreview}"
        ></simple-icon-button-lite>
      </div>
      ${"heading"===l?i`
            <span
              class="label shown"
              ?disabled="${o.metadata.locked}"
              @dblclick="${this.editTitle}"
              >${r}</span
            >
            <span
              class="label-edit"
              @keypress="${this.monitorHeading}"
              @keydown="${this.monitorEsc}"
            ></span>
          `:i`<span class="label shown">${r}</span>`}
      <div class="content-operations">
        <simple-icon-button-lite
          class="content-operation"
          icon="hax:outline-designer-outdent"
          @click="${e=>this.modifyContentAction(e,o,"in")}"
          title="Increase heading"
          ?disabled="${"h1"===n||o.metadata.locked}"
          ?hidden="${"heading"!==l}"
        ></simple-icon-button-lite>
        <simple-icon-button-lite
          icon="hax:keyboard-arrow-up"
          @click="${e=>this.modifyContentAction(e,o,"up")}"
          title="Move up"
          ?disabled="${o.metadata.locked}"
          class="content-operation"
        ></simple-icon-button-lite>
        <simple-icon-button-lite
          icon="hax:keyboard-arrow-down"
          @click="${e=>this.modifyContentAction(e,o,"down")}"
          title="Move down"
          ?disabled="${o.metadata.locked}"
          class="content-operation"
        ></simple-icon-button-lite>
        <simple-icon-button-lite
          class="content-operation"
          icon="hax:outline-designer-indent"
          @click="${e=>this.modifyContentAction(e,o,"out")}"
          title="Decrease Heading"
          ?disabled="${"h6"===n||o.metadata.locked}"
          ?hidden="${"heading"!==l}"
        ></simple-icon-button-lite>
        <simple-icon-button-lite
          class="content-operation"
          icon="editor:format-page-break"
          @click="${e=>this.pageBreakHere(e,o)}"
          title="Promote to page"
          ?disabled="${o.metadata.locked}"
          ?hidden="${"heading"!==l}"
        ></simple-icon-button-lite>
        <simple-icon-button-lite
          icon="delete"
          @click="${e=>this.modifyContentAction(e,o,"delete")}"
          class="content-operation del"
          title="Delete"
          ?disabled="${o.metadata.locked}"
          accent-color="red"
        ></simple-icon-button-lite>
      </div>
    </li>`}setActivePreview(e){let t=e.target.closest("[data-content-parent-id]"),i=parseInt(t.getAttribute("data-node-index"));this.activePreview=t,this.shadowRoot.querySelector("simple-popover").removeAttribute("hidden"),this.shadowRoot.querySelector("simple-popover").target=t,this.activePreviewIndex=i,e.preventDefault(),e.stopPropagation(),e.stopImmediatePropagation()}modifyContentAction(e,t,i){if(!t.metadata.locked){e.preventDefault(),e.stopPropagation(),e.stopImmediatePropagation();let t=e.target.closest("[data-content-parent-id]"),o=t.getAttribute("data-content-parent-id"),a=parseInt(t.getAttribute("data-node-index")),s=this.items.find((e=>e.id===o));if(s.contents){let e=globalThis.document.createElement("div");e.innerHTML=s.contents;let t="";switch(i){case"up":0!==a&&e.childNodes[a].previousElementSibling.insertAdjacentElement("beforebegin",e.childNodes[a]);break;case"down":a!==e.childNodes.length-1&&e.childNodes[a].nextElementSibling.insertAdjacentElement("afterend",e.childNodes[a])}for(let o=0;o<e.childNodes.length;o++){let s=e.childNodes[o];if(o<a)t+=s.outerHTML;else if(o===a)switch(i){case"delete":this.setAttribute("stop-animation","true");break;case"up":case"down":t+=s.outerHTML;break;case"in":case"out":let e,o=parseInt(s.tagName.toLowerCase().replace("h",""));"in"===i&&o>1?(e=globalThis.document.createElement("h"+(o-1)),e.innerText=s.innerText):"out"===i&&o<6?(e=globalThis.document.createElement(`h${o+1}`),e.innerText=s.innerText):e=s,t+=e.outerHTML}else t+=s.outerHTML}s.contents=t,this.resetPopOver(),this.__syncUIAndDataModel()}}}pageBreakHere(e,t){if(!t.metadata.locked){e.preventDefault(),e.stopPropagation(),e.stopImmediatePropagation();let t,i=e.target.closest("[data-content-parent-id]"),o=i.getAttribute("data-content-parent-id"),a=parseInt(i.getAttribute("data-node-index")),s=this.items.find((e=>e.id===o));if(this.items.map(((e,i)=>e.id===o?t=i:null)),s.contents){let e=globalThis.document.createElement("div");e.innerHTML=s.contents;let i="",o="",n=this.t.newPage;for(let t=0;t<e.childNodes.length;t++){let s=e.childNodes[t];t<a?i+=s.outerHTML:t===a?""!=s.innerText&&(n=s.innerText):o+=s.outerHTML}s.contents=i;let d=new l;d.title=n,d.order=s.order+1,d.parent=s.parent,d.indent=s.indent,d.metadata.locked=!1,d.new=!0,d.contents=o,this.items[t].modified=!0,this.items.splice(t+1,0,d),this.resetPopOver(),this.__syncUIAndDataModel()}}}__syncUIAndDataModel(){this._recurseUpdateIndent(),this._schemaOrderUpdate(),setTimeout((()=>{this.requestUpdate(),setTimeout((()=>{this.removeAttribute("stop-animation")}),300)}),0)}collapseAll(){this.items.map(((e,t)=>{this.hasChildren(e.id)&&!e.collapsed&&(this.items[t].collapsed=!0)})),setTimeout((()=>{this.requestUpdate()}),0)}expandAll(){this.items.map(((e,t)=>{this.hasChildren(e.id)&&e.collapsed&&(this.items[t].collapsed=!1)})),setTimeout((()=>{this.requestUpdate()}),0)}getItemParents(e){let t=this.items.find((t=>t.id==e.parent)),i="";for(;t;)i+=t.id+" ",t=this.items.find((e=>e.id==t.parent));return i}getItemParentsCollapsed(e){let t=this.items.find((t=>t.id==e.parent)),i="";for(;t;)t.collapsed&&t.id&&(i+=`collapsed-by-${t.id} `),t=this.items.find((e=>e.id==t.parent));return i}isCollapsed(e){return!!this.items.find((t=>t.id==e)).collapsed}hasChildren(e){return!!this.items.find((t=>t.parent==e))}collapseExpand(e){let t=e.target.closest("[data-item-id]").getAttribute("data-item-id");this.items.map(((e,i)=>{e.id!==t||this.isLocked(i)||(this.items[i].collapsed?this.items[i].collapsed=!1:this.items[i].collapsed=!0,this.requestUpdate())}))}toggleContent(e){let t=e.target.closest("[data-item-id]");if(t&&!e.target.disabled){let e=t.getAttribute("data-item-id");this.items.map(((t,i)=>{t.id===e&&(this.items[i].showContent?this.items[i].showContent=!1:this.items[i].showContent=!0)})),this.requestUpdate()}}editTitle(e){e.target.classList.remove("shown");let t=e.target.nextElementSibling;if(t.setAttribute("contenteditable","true"),t.classList.add("shown"),t.innerText=e.target.innerText,t.focus(),this.shadowRoot.getSelection){var i=globalThis.document.createRange();i.selectNodeContents(t),this.shadowRoot.getSelection().removeAllRanges(),this.shadowRoot.getSelection().addRange(i)}else try{globalThis.document.execCommand("selectAll",!1,null)}catch(e){console.warn(e)}}monitorTitle(e){if("Enter"===e.key){e.target.classList.remove("shown"),e.target.previousElementSibling.classList.add("shown"),e.target.removeAttribute("contenteditable");let t=e.target.closest("[data-item-id]").getAttribute("data-item-id");for(let i=0;i<this.items.length;i++)this.items[i].id===t&&""!=e.target.innerText&&(this.items[i].new||(this.items[i].modified=!0),this.items[i].title=e.target.innerText);this.requestUpdate()}}monitorHeading(e){if("Enter"===e.key){e.target.classList.remove("shown"),e.target.previousElementSibling.classList.add("shown"),e.target.removeAttribute("contenteditable");let t=e.target.closest("[data-content-parent-id]"),i=t.getAttribute("data-content-parent-id"),o=parseInt(t.getAttribute("data-node-index")),a=this.items.find((e=>e.id===i));if(a.contents){let t=globalThis.document.createElement("div");t.innerHTML=a.contents;let i="";for(let a=0;a<t.childNodes.length;a++){let s=t.childNodes[a];a<o?i+=s.outerHTML:a===o?(s.innerText=e.target.innerText,i+=s.outerHTML):i+=s.outerHTML}a.contents=i,this.resetPopOver(),this.requestUpdate()}}}monitorEsc(e){"Escape"===e.key?(e.target.classList.remove("shown"),e.target.removeAttribute("contenteditable"),e.target.previousElementSibling.classList.add("shown"),e.target.innerText=e.target.previousElementSibling.innerText):"Enter"===e.key&&(this._blurBlock=!0)}blurTitle(e){this._blurBlock||(e.target.classList.remove("shown"),e.target.removeAttribute("contenteditable"),e.target.previousElementSibling.classList.add("shown"),e.target.innerText=e.target.previousElementSibling.innerText),setTimeout((()=>{this._blurBlock=!1}),0)}_handleDragHandleKeydown(e){const t=e.target.getAttribute("data-drag-handle-id"),i=this.items.findIndex((e=>e.id===t));if(-1===i||this.isLocked(i))return;if("Enter"===e.key)return e.preventDefault(),e.stopPropagation(),this._dragHandleActive===t?(this._dragHandleActive=null,this.announceAction("Keyboard controls deactivated")):(this._dragHandleActive=t,this.announceAction("Keyboard controls activated. Use arrow keys to move, Enter or Escape to deactivate")),void this.requestUpdate();if("Escape"===e.key&&this._dragHandleActive===t)return e.preventDefault(),e.stopPropagation(),this._dragHandleActive=null,this.announceAction("Keyboard controls deactivated"),void this.requestUpdate();if(this._dragHandleActive!==t)return;let o=null;switch(e.key){case"ArrowUp":e.preventDefault(),e.stopPropagation(),o="up",this.announceAction("Moving item up");break;case"ArrowDown":e.preventDefault(),e.stopPropagation(),o="down",this.announceAction("Moving item down");break;case"ArrowLeft":e.preventDefault(),e.stopPropagation(),o="in",this.announceAction("Indenting item");break;case"ArrowRight":e.preventDefault(),e.stopPropagation(),o="out",this.announceAction("Outdenting item")}o&&this.itemOp(i,o)}handleLabelKeydown(e){"Enter"!==e.key||e.target.hasAttribute("disabled")||(e.preventDefault(),e.stopPropagation(),this.editTitle(e),this.announceAction("Title editing activated"))}handleTreeItemKeydown(e){if("li"!==e.target.tagName.toLowerCase()||!e.target.hasAttribute("role"))return;const t=e.target.getAttribute("data-item-id"),i=this.items.findIndex((e=>e.id===t));switch(e.key){case"ArrowDown":e.preventDefault(),this.focusNextItem(i),this.announceNavigation("Moved to next item");break;case"ArrowUp":e.preventDefault(),this.focusPreviousItem(i),this.announceNavigation("Moved to previous item");break;case"ArrowRight":e.preventDefault(),this.hasChildren(t)&&this.isCollapsed(t)?(this.collapseExpand(e),this.announceStateChange("Expanded")):(this.focusFirstChild(i),this.announceNavigation("Moved to first child"));break;case"ArrowLeft":e.preventDefault(),this.hasChildren(t)&&!this.isCollapsed(t)?(this.collapseExpand(e),this.announceStateChange("Collapsed")):(this.focusParent(i),this.announceNavigation("Moved to parent"));break;case"Enter":case" ":if(this.activeItemForActions!==t&&this.hasChildren(t)){e.preventDefault(),this.collapseExpand(e);const i=this.isCollapsed(t)?"Collapsed":"Expanded";this.announceStateChange(i)}break;case"Home":e.preventDefault(),this.focusFirstItem(),this.announceNavigation("Moved to first item");break;case"End":e.preventDefault(),this.focusLastItem(),this.announceNavigation("Moved to last item");break;case"Delete":case"Backspace":this.activeItemForActions!==t||this.isLocked(i)||(e.preventDefault(),this.itemOp(i,"delete"),this.announceAction("Item marked for deletion"));break;case"d":e.ctrlKey&&this.activeItemForActions===t&&!this.isLocked(i)&&(e.preventDefault(),this.itemOp(i,"duplicate"),this.announceAction("Item duplicated"))}}focusNextItem(e){const t=this.items.filter(((e,t)=>""===this.getItemParentsCollapsed(e)&&!this.hideDelete||!e.delete)),i=t.findIndex((t=>t.id===this.items[e].id)),o=Math.min(i+1,t.length-1);this.focusItem(t[o].id)}focusPreviousItem(e){const t=this.items.filter(((e,t)=>""===this.getItemParentsCollapsed(e)&&!this.hideDelete||!e.delete)),i=t.findIndex((t=>t.id===this.items[e].id)),o=Math.max(i-1,0);this.focusItem(t[o].id)}focusFirstChild(e){const t=this.items[e].id,i=this.items.find((e=>e.parent===t));i&&""===this.getItemParentsCollapsed(i)&&this.focusItem(i.id)}focusParent(e){const t=this.items[e].parent;t&&this.focusItem(t)}focusFirstItem(){const e=this.items.find((e=>!(""!==this.getItemParentsCollapsed(e)||this.hideDelete&&e.delete)));e&&this.focusItem(e.id)}focusLastItem(){const e=this.items.filter((e=>!(""!==this.getItemParentsCollapsed(e)||this.hideDelete&&e.delete)));e.length>0&&this.focusItem(e[e.length-1].id)}focusItem(e){this.shadowRoot.querySelectorAll('[role="treeitem"]').forEach((e=>{e.setAttribute("tabindex","-1")}));const t=this.shadowRoot.querySelector(`[data-item-id="${e}"]`);t&&(t.setAttribute("tabindex","0"),t.focus())}announceNavigation(e){this.liveRegionText=e,this.requestUpdate(),setTimeout((()=>{this.liveRegionText="",this.requestUpdate()}),1e3)}announceStateChange(e){this.liveRegionText=e,this.requestUpdate(),setTimeout((()=>{this.liveRegionText="",this.requestUpdate()}),1e3)}announceAction(e){this.liveRegionText=e,this.requestUpdate(),setTimeout((()=>{this.liveRegionText="",this.requestUpdate()}),1e3)}scrollIntoViewIfNeeded(e,t=2e3){setTimeout((()=>{const t=this.shadowRoot.querySelector(`[data-item-id="${e}"]`);if(t){const e=t.getBoundingClientRect(),i=globalThis.innerHeight,o=globalThis.innerWidth;(e.bottom<0||e.top>i||e.right<0||e.left>o)&&t.scrollIntoView({behavior:"smooth",block:"center",inline:"nearest"})}}),t)}_createDragPreview(e){const t=e.getAttribute("data-item-id"),i=this.items.find((e=>e.id===t));if(!i)return null;const o=globalThis.document.createElement("div");o.className="drag-preview";const a=this._countAllDescendants(t);a>0&&o.classList.add("has-children");const s=globalThis.document.createElement("simple-icon-lite");s.setAttribute("icon","icons:reorder");const n=globalThis.document.createElement("span");if(n.textContent=i.title.length>40?i.title.substring(0,40)+"...":i.title,o.appendChild(s),o.appendChild(n),a>0){const e=globalThis.document.createElement("span");e.className="drag-preview-children",e.textContent=`(+${a} page${1!==a?"s":""})`,o.appendChild(e)}return o}_countAllDescendants(e){const t=this.items.filter((t=>t.parent===e));let i=t.length;return t.forEach((e=>{i+=this._countAllDescendants(e.id)})),i}_onDrag(e){if(this._targetDrag&&e.clientX&&e.clientY){const t=globalThis.document.elementFromPoint(e.clientX,e.clientY);if(t){const i=t.closest("[data-item-id]");i&&i!==this._lastDragTarget&&(this._lastDragTarget=i,this._dragEnter(e))}}}_mouseDownDrag(e){let t=e.target.closest("[data-item-id]").getAttribute("data-item-id");this.items.map(((e,i)=>{e.id===t&&this.hasChildren(e.id)&&(this.items[i].collapsed=!0)})),setTimeout((()=>{this.requestUpdate()}),0)}_dragEnter(e){const t=e.target.closest("[data-item-id]");if(!t||!this._targetDrag)return;if(t===this._targetDrag)return;e.preventDefault();const i=t.getBoundingClientRect(),o=e.clientY-i.top,a=e.clientX-i.left,s=i.height,n=(i.width,a>40);let d="child";if(o<.3*s&&!n?d="above":o>.7*s&&!n?d="below":n&&(d="child"),this._targetDrop!==t||this._dropZone!==d){if(this.shadowRoot.querySelectorAll(".drop-indicator").forEach((e=>e.remove())),this.shadowRoot.querySelectorAll(".drop-target-child").forEach((e=>{e.classList.remove("drop-target-child")})),this.shadowRoot.querySelectorAll(".outline-designer-hovered").forEach((e=>{e.classList.remove("outline-designer-hovered")})),this.shadowRoot.querySelectorAll(".drop-above-target").forEach((e=>{e.classList.remove("drop-above-target")})),this.shadowRoot.querySelectorAll(".drop-below-target").forEach((e=>{e.classList.remove("drop-below-target")})),"above"===d){const e=globalThis.document.createElement("div");e.className="drop-indicator drop-above",t.insertAdjacentElement("beforebegin",e),setTimeout((()=>e.classList.add("show")),10),t.classList.add("drop-above-target")}else if("below"===d){const e=globalThis.document.createElement("div");e.className="drop-indicator drop-below",t.insertAdjacentElement("afterend",e),setTimeout((()=>e.classList.add("show")),10),t.classList.add("drop-below-target")}else{t.classList.add("drop-target-child");const e=globalThis.document.createElement("div");e.className="drop-indicator indent-visual",e.style.marginLeft="40px",t.insertAdjacentElement("afterend",e),setTimeout((()=>e.classList.add("show")),10)}this._targetDrop=t,this._dropZone=d;const e=this.items.find((e=>e.id===t.getAttribute("data-item-id")));if(e){this.items.find((e=>e.id===this._targetDrag.getAttribute("data-item-id")));let t="";t="above"===d?`Drop above ${e.title}`:"below"===d?`Drop below ${e.title}`:`Drop as child of ${e.title}`,this.announceAction(t)}}}_dragLeave(e){}_dragEnd(e){if(this._targetDrag&&this._targetDrag.removeAttribute("data-dragging"),this._dragPreviewElement&&(this._dragPreviewElement.remove(),this._dragPreviewElement=null),this.shadowRoot.querySelectorAll(".drop-indicator").forEach((e=>e.remove())),this.shadowRoot.querySelectorAll(".drop-target-child").forEach((e=>{e.classList.remove("drop-target-child")})),this.shadowRoot.querySelectorAll(".outline-designer-hovered").forEach((e=>{e.classList.remove("outline-designer-hovered")})),this.shadowRoot.querySelectorAll(".drop-above-target").forEach((e=>{e.classList.remove("drop-above-target")})),this.shadowRoot.querySelectorAll(".drop-below-target").forEach((e=>{e.classList.remove("drop-below-target")})),this._targetDrag&&this._targetDrop&&this._dropZone){let e=null,t=null;for(let i=0;i<this.items.length;i++){let o=this.items[i];o.id===this._targetDrop.getAttribute("data-item-id")&&(e=i),o.id===this._targetDrag.getAttribute("data-item-id")&&(t=i)}if(null!==t&&null!==e){switch(this.items[t].new||(this.items[t].modified=!0),this._dropZone){case"above":this.items[t].order=this.items[e].order-.5,this.items[t].parent=this.items[e].parent,this.items[t].indent=this.items[e].indent;break;case"below":this.items[t].order=this.items[e].order+.5,this.items[t].parent=this.items[e].parent,this.items[t].indent=this.items[e].indent;break;case"child":this.items[t].parent=this.items[e].id,this.items[t].order=0,this.items[t].indent=this.items[e].indent+1}this.hasChildren(this.items[t].id)&&(this.items[t].collapsed=!1),this.scrollIntoViewIfNeeded(this.items[t].id)}this.setAttribute("stop-animation","true"),this.__syncUIAndDataModel()}this._targetDrag=null,this._targetDrop=null,this._dropZone=null}_dragStart(e){if(null==e.target.getAttribute("disabled")){let t=e.target.closest("[data-item-id]");this._targetDrop=null,this._targetDrag=t,t.setAttribute("data-dragging","true");const i=this._createDragPreview(t);i&&(globalThis.document.body.appendChild(i),this._dragPreviewElement=i,e.dataTransfer&&(e.dataTransfer.effectAllowed="move",e.dataTransfer.dropEffect="move",e.dataTransfer.setDragImage(i,20,20))),this._mouseDownDrag(e),e.stopPropagation(),e.stopImmediatePropagation()}}static get properties(){return{haxGizmos:{type:Array},hideDelete:{type:Boolean},activeItemForActions:{type:String},storeTools:{type:Boolean},eventData:{type:Object},items:{type:Array},appReady:{type:Boolean},activePreview:{type:Object},activePreviewIndex:{type:Number},hideContentOps:{type:Boolean,reflect:!0,attribute:"hide-content-ops"},fidelity:{type:String},liveRegionText:{type:String},activePage:{type:String},selectedPages:{type:Array},zoomLevel:{type:Number},_dragHandleActive:{type:String}}}updated(e){super.updated(e),e.forEach(((e,t)=>{"fidelity"===t&&this[t]&&this[t],"activePreview"===t&&e&&e.classList.remove("active-preview-item"),"activePreview"===t&&this[t]&&this[t].classList.add("active-preview-item")}))}firstUpdated(e){super.firstUpdated&&super.firstUpdated(e),this.items&&setTimeout((()=>{this.__syncUIAndDataModel()}),0)}static get tag(){return"outline-designer"}addItemToTop(){this.setAttribute("stop-animation","true"),this.addNewItem("top"),this.__syncUIAndDataModel()}async getData(){let e=this.eventData;if(e.items=[...this.items],this.storeTools){const t=this.shadowRoot.querySelector("#itemselector").value,i=this.shadowRoot.querySelector("#targetselector").value;let o=0;await e.items.map((async(a,s)=>{if(t&&null==a.parent){o++;let a=await n.findItemAsObject(t);switch(i){case"below":e.items[s].parent=a.parent,e.items[s].order=parseInt(a.order)+o;break;case"above":e.items[s].parent=a.parent,e.items[s].order=parseInt(a.order)-o;break;case"children":e.items[s].parent=t}}}))}return e}isLocked(e){return!!(!1!==e&&this.items[e]&&this.items[e].metadata&&this.items[e].metadata.locked)}addNewItem(e,t=!1,i=[]){let o=1;"top"===e&&(e=0,o=-1);let a={indent:0,parent:null,order:0};this.items&&this.items.length>0&&this.items[e]&&(a=this.items[e]);let s=new l;if(s.order=a.order+o,s.parent=a.parent,s.indent=a.indent,s.metadata.locked=!1,s.new=!0,t?(s.title=`${this.t.copyOf} ${a.title}`,s.contents=a.contents,s.duplicate=a.id):s.contents="<p></p>",i.push(s),this.items&&this.items.length>0&&this.items[e]&&this.hasChildren(this.items[e].id)&&t){let t={};t[this.items[e].id]=s.id,i=this.recurseCopyChildren(this.items[e].id,t,i)}this.items&&this.items.length>0?i.forEach(((t,i)=>this.items.splice(e+i+1,0,t))):i.forEach((e=>this.items.push(e)))}recurseCopyChildren(e,t,i){let o=this.items.filter((t=>t.parent==e));for(let e=0;e<o.length;e++){const a=o[e];let s=new l;s.order=a.order,s.parent=t[a.parent],s.indent=a.indent,s.metadata.locked=!1,s.new=!0,s.title=`${this.t.copyOf} ${a.title}`,s.contents=a.contents,s.collapsed=a.collapsed,s.duplicate=o[e].id,t[o[e].id]=s.id,i.push(s),this.hasChildren(o[e].id)&&this.recurseCopyChildren(o[e].id,t,i)}return i}recurseAction(e,t,i=!0){let o=this.items.filter((t=>t.parent==e));for(let e=0;e<o.length;e++)this.items.map(((a,s)=>{if(a.id===o[e].id)switch(t){case"delete":this.items[s].delete=i;break;case"lock":this.items[s].metadata.locked=i}})),this.hasChildren(o[e].id)&&this.recurseAction(o[e].id,t,i);return!0}itemOp(e,t){if(!1!==e&&this.items[e]&&t){if(this.items[e].metadata.locked)"lock"===t&&(this.items[e].metadata.locked=!1,this.hasChildren(this.items[e].id)&&this.recurseAction(this.items[e].id,t,this.items[e].metadata.locked));else switch(t){case"goto":this.dispatchEvent(new CustomEvent("simple-modal-hide",{bubbles:!0,cancelable:!0,detail:{}}));let i=this.items[e].slug||this.items[e].location;globalThis.location.href=i;case"lock":this.items[e].metadata.locked=!0,this.hasChildren(this.items[e].id)&&this.recurseAction(this.items[e].id,t,this.items[e].metadata.locked);break;case"delete":this.items[e].delete?this.items[e].delete=!1:this.items[e].delete=!0,this.hasChildren(this.items[e].id)&&this.recurseAction(this.items[e].id,t,this.items[e].delete);break;case"add":this.setAttribute("stop-animation","true"),this.addNewItem(e);break;case"duplicate":this.setAttribute("stop-animation","true"),this.addNewItem(e,!0);break;case"in":if(0!==e&&this.items[e].parent!=this.items[e-1].id){let t=this.items[e-1];this.items[e].parent=t.id,this.items[e].order=0,this.items[e].indent=parseInt(t.indent)+1,this.items[e].new||(this.items[e].modified=!0)}break;case"out":if(null!==this.items[e].parent){let t=this.items.find((t=>this.items[e].parent===t.id));this.items[e].parent=t.parent,this.items[e].order=parseInt(t.order)+1,this.items[e].indent=parseInt(t.indent),this.items[e].new||(this.items[e].modified=!0)}break;case"up":case"down":this.setAttribute("stop-animation","true");let o=this.items[e],a=[];this.items.map((e=>{e.parent===o.parent&&a.push(e)})),a.sort(((e,t)=>e.order<t.order?-1:e.order>t.order?1:0));let s=null;if(a.map(((e,i)=>{"up"===t&&i>0&&e.id===o.id?s=a[i-1]:"down"===t&&i<a.length-1&&e.id===o.id&&(s=a[i+1])})),s){const e=parseInt(s.order+".0"),t=parseInt(o.order+".0");this.items.map(((i,a)=>{i.id===s.id?(this.items[a].order=t,this.items[a].new||(this.items[a].modified=!0)):i.id===o.id&&(this.items[a].order=e,this.items[a].new||(this.items[a].modified=!0))})),this.scrollIntoViewIfNeeded(o.id)}}this.__syncUIAndDataModel()}}_recurseUpdateIndent(e={id:null},t=0){this.items.map(((i,o)=>{i.parent==e.id&&(this.items[o].indent=t,this._recurseUpdateIndent(this.items[o],t+1))}))}_schemaOrderUpdate(){let e=new c,t=e.itemsToNodes(this.items),i=e.nodesToItems(t),o=[];for(var a in i){let e=this.items.find((e=>e.id===i[a].id));e&&(delete e.children,o.push(e))}this.items=o}}globalThis.customElements.define(OutlineDesigner.tag,OutlineDesigner);
