// @lrnwebcomponents/product-glance/product-glance.js
define(["exports", "meta", "require", "../../lit/index.js", "../simple-colors/simple-colors.js"], function (_exports, meta, _require, _index, _simpleColors) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ProductGlance = void 0;
  meta = _interopRequireWildcard(meta);
  _require = _interopRequireWildcard(_require);

  var _templateObject_2111a550942811ecb1937969798b231a, _templateObject2_2111a550942811ecb1937969798b231a, _templateObject3_2111a550942811ecb1937969798b231a, _templateObject4_2111a550942811ecb1937969798b231a;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var ProductGlance = /*#__PURE__*/function (_SimpleColors) {
    babelHelpers.inherits(ProductGlance, _SimpleColors);

    var _super = _createSuper(ProductGlance);

    function ProductGlance() {
      babelHelpers.classCallCheck(this, ProductGlance);
      return _super.call(this);
    }

    babelHelpers.createClass(ProductGlance, [{
      key: "updated",
      value: function updated() {
        if (this.icon) {
          new Promise(function (res, rej) {
            return _require.default(["../simple-icon/simple-icon.js"], res, rej);
          });
        }
      }
    }, {
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_2111a550942811ecb1937969798b231a || (_templateObject_2111a550942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"wrapper\">\n        <div class=\"icon-wrapper\">\n          ", "\n        </div>\n        <div></div>\n        <div class=\"text-wrapper\">\n          <div class=\"title-text\"><slot name=\"title\">", "</slot></div>\n          <div class=\"subtitle-text\">\n            <slot name=\"subtitle\">", "</slot>\n          </div>\n        </div>\n      </div>\n    "])), this.icon ? (0, _index.html)(_templateObject2_2111a550942811ecb1937969798b231a || (_templateObject2_2111a550942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<simple-icon\n                class=\"icon\"\n                icon=\"", "\"\n                accent-color=\"", "\"\n                .contrast=\"", "\"\n                ?dark=\"", "\"\n              ></simple-icon>"])), this.icon, this.accentColor, this.contrast, this.dark) : (0, _index.html)(_templateObject3_2111a550942811ecb1937969798b231a || (_templateObject3_2111a550942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral([""]))), this.title, this.subtitle);
      }
    }], [{
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(ProductGlance), "styles", this)), [(0, _index.css)(_templateObject4_2111a550942811ecb1937969798b231a || (_templateObject4_2111a550942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: inline-block;\n          --product-glance-icon-width: 20px;\n          --product-glance-icon-height: var(--product-glance-icon-width, 20px);\n        }\n        .wrapper {\n          display: grid;\n          grid-template-columns: 3.5em 0.5em 21em;\n          margin: var(--product-glance-internal-margin, 0.25em);\n          background-color: var(--simple-colors-default-theme-grey-1, #eeeeee);\n          padding: var(--product-glance-internal-padding, 0.5em);\n        }\n        .icon-wrapper {\n          padding: var(--product-glance-internal-padding, 0.5em);\n          display: block;\n          margin: 0 auto;\n        }\n\n        .icon {\n          margin: 0;\n          width: var(--product-glance-icon-width);\n          height: var(--product-glance-icon-height);\n          --simple-icon-height: var(--product-glance-icon-height);\n          --simple-icon-width: var(--product-glance-icon-width);\n          border: 2px solid var(--simple-colors-default-theme-grey-4, #eeeeee);\n          border-radius: 50%;\n          padding: var(--product-glance-internal-padding, 0.5em);\n          display: block;\n        }\n        .title-text {\n          margin-top: var(--product-glance-title-margin-top, 8px);\n        }\n        .title-text ::slotted(p),\n        .subtitle-text ::slotted(p) {\n          margin: 0;\n        }\n        .title-text,\n        .title-text ::slotted(*) {\n          color: var(\n            --product-glance-text-color,\n            var(--simple-colors-default-theme-grey-12, #222222)\n          );\n          font-family: \"OpenSans-Bold\", \"OpenSans\", \"Arial\", sans-serif;\n          font-size: 1.25em;\n          font-weight: bold;\n        }\n\n        .subtitle-text,\n        .subtitle-text ::slotted(*) {\n          color: var(\n            --product-glance-subtext-color,\n            var(--simple-colors-default-theme-grey-8, #555555)\n          );\n          font-family: \"OpenSans-Bold\", \"OpenSans\", \"Arial\", sans-serif;\n          font-size: 0.9em;\n          line-height: 1em;\n        }\n      "])))]);
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(ProductGlance), "properties", this)), {}, {
          title: {
            type: String
          },
          subtitle: {
            type: String
          },
          icon: {
            type: String
          }
        });
      }
    }, {
      key: "tag",
      get: function get() {
        return "product-glance";
      }
      /**
       * haxProperties integration via file reference
       */

    }, {
      key: "haxProperties",
      get: function get() {
        return new URL("./lib/".concat(this.tag, ".haxProperties.json"), meta.url).href;
      }
    }]);
    return ProductGlance;
  }(_simpleColors.SimpleColors);

  _exports.ProductGlance = ProductGlance;
  customElements.define(ProductGlance.tag, ProductGlance);
});
