// @lrnwebcomponents/oer-schema/oer-schema.js
define(["exports", "../../@polymer/polymer/polymer-element.js", "../schema-behaviors/schema-behaviors.js", "./lib/oerschema.js", "../hax-iconset/hax-iconset.js"], function (_exports, _polymerElement, _schemaBehaviors, _oerschema, _haxIconset) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.OerSchemaElement = void 0;

  function _templateObject_13494830e14a11e9bce923988a998e3c() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: inline-block;\n        }\n      </style>\n      <span property$=\"oer:[[oerProperty]]\"> <slot></slot> [[text]] </span>\n    "], ["\n      <style>\n        :host {\n          display: inline-block;\n        }\n      </style>\n      <span property\\$=\"oer:[[oerProperty]]\"> <slot></slot> [[text]] </span>\n    "]);

    _templateObject_13494830e14a11e9bce923988a998e3c = function _templateObject_13494830e14a11e9bce923988a998e3c() {
      return data;
    };

    return data;
  }

  /**
   * `oer-schema`
   * `A LRN element to wrap an oer schema prefix onto materials.`
   * @demo demo/index.html
   */
  var OerSchemaElement =
  /*#__PURE__*/
  function (_SchemaBehaviors) {
    babelHelpers.inherits(OerSchemaElement, _SchemaBehaviors);

    function OerSchemaElement() {
      babelHelpers.classCallCheck(this, OerSchemaElement);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(OerSchemaElement).apply(this, arguments));
    }

    babelHelpers.createClass(OerSchemaElement, [{
      key: "_generateforComponentLink",
      value: function _generateforComponentLink(source) {
        // remove existing if this is moving around
        if (this._OERLink) {
          document.head.removeChild(this._OERLink);
        }

        var link = document.createElement("link");
        link.setAttribute("property", "oer:forComponent");
        link.setAttribute("content", this.relatedResource);
        document.head.appendChild(link);
        return link;
      }
    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_13494830e14a11e9bce923988a998e3c());
      }
    }, {
      key: "tag",
      get: function get() {
        return "oer-schema";
      }
    }, {
      key: "properties",
      get: function get() {
        var props = {
          /**
           * Text to wire into the middle of the element.
           * This is easier to manage then slotted data though
           * this supports both methods.
           */
          text: {
            type: String,
            value: ""
          },

          /**
           * Property value for this oer resource
           */
          oerProperty: {
            type: String,
            value: "name"
          },

          /**
           * Property value for this oer resource
           */
          typeof: {
            type: String,
            value: "Resource"
          },

          /**
           * Related Resource ID
           */
          relatedResource: {
            type: String
          },

          /**
           * License link object for semantic meaning
           */
          _OERLink: {
            type: Object,
            computed: "_generateforComponentLink(relatedResource)"
          }
        };

        if (babelHelpers.get(babelHelpers.getPrototypeOf(OerSchemaElement), "properties", this)) {
          props = Object.assign(props, babelHelpers.get(babelHelpers.getPrototypeOf(OerSchemaElement), "properties", this));
        }

        return props;
      }
    }, {
      key: "haxProperties",
      get: function get() {
        var oerSchema = new _oerschema.OERSchema();
        return {
          canScale: false,
          canPosition: false,
          canEditSource: false,
          gizmo: {
            title: "Schema",
            description: "Schematized element area",
            icon: "hax:oerschema",
            color: "blue",
            groups: ["Instructional"],
            handles: [{
              type: "inline",
              text: "text"
            }],
            meta: {
              author: "LRNWebComponents"
            }
          },
          settings: {
            quick: [{
              slot: "",
              title: "Text",
              inputMethod: "textfield",
              icon: "editor:title"
            }],
            configure: [{
              slot: "",
              title: "Text",
              inputMethod: "textfield",
              icon: "editor:title"
            }, {
              property: "typeof",
              title: "Schema typeof",
              inputMethod: "select",
              allowNull: true,
              options: oerSchema.types
            }, {
              property: "oerProperty",
              title: "Schema property",
              description: "The OER Schema property this represents",
              inputMethod: "select",
              allowNull: true,
              options: {
                name: "name",
                additionalType: "additionalType",
                description: "description",
                image: "image",
                mainEntityOfPage: "mainEntityOfPage",
                sameAs: "sameAs",
                uri: "uri"
              }
            }, {
              property: "relatedResource",
              title: "Related resource",
              description: "A reference to the related Schema resource",
              inputMethod: "textfield",
              icon: "editor:title"
            }],
            advanced: []
          },
          saveOptions: {
            unsetAttributes: ["_oerlink"]
          }
        };
      }
    }]);
    return OerSchemaElement;
  }((0, _schemaBehaviors.SchemaBehaviors)(_polymerElement.PolymerElement));

  _exports.OerSchemaElement = OerSchemaElement;
  window.customElements.define(OerSchemaElement.tag, OerSchemaElement);
});
