// @lrnwebcomponents/elmsln-loading/elmsln-loading.js
define(["exports","require","../../lit-element/lit-element.js","../simple-colors/simple-colors.js"],(function(i,e,n,t){"use strict";Object.defineProperty(i,"__esModule",{value:!0}),i.ElmslnLoading=void 0,e=babelHelpers.interopRequireWildcard(e);
/**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class ElmslnLoading extends t.SimpleColors{static get styles(){return[...super.styles,n.css`
        @-moz-keyframes spin {
          100% {
            -moz-transform: rotate(60deg);
            filter: saturate(10) invert(0.9);
          }
        }
        @-webkit-keyframes spin {
          100% {
            -webkit-transform: rotate(60deg);
            filter: saturate(10) invert(0.9);
          }
        }
        @keyframes spin {
          100% {
            -webkit-transform: rotate(60deg);
            transform: rotate(60deg);
          }
        }
        :host iron-icon {
          color: var(--simple-colors-default-theme-accent-6);
          display: block;
          -webkit-animation: spin 1.25s ease-out infinite;
          -moz-animation: spin 1.25s ease-out infinite;
          animation: spin 1.25s ease-out infinite;
        }
        :host([size="tiny"]) iron-icon {
          width: 16px;
          height: 16px;
          -webkit-animation: spin 0.75s ease-out infinite;
          -moz-animation: spin 0.75s ease-out infinite;
          animation: spin 0.75s ease-out infinite;
        }
        :host([size="small"]) iron-icon {
          width: 32px;
          height: 32px;
          -webkit-animation: spin 1s ease-out infinite;
          -moz-animation: spin 1s ease-out infinite;
          animation: spin 1s ease-out infinite;
        }
        :host([size="medium"]) iron-icon {
          width: 64px;
          height: 64px;
          -webkit-animation: spin 1.25s ease-out infinite;
          -moz-animation: spin 1.25s ease-out infinite;
          animation: spin 1.25s ease-out infinite;
        }
        :host([size="large"]) iron-icon {
          width: 80px;
          height: 80px;
          -webkit-animation: spin 1.25s ease-out infinite;
          -moz-animation: spin 1.25s ease-out infinite;
          animation: spin 1.25s ease-out infinite;
        }
        :host([size="epic"]) iron-icon {
          width: 400px;
          height: 400px;
          -webkit-animation: spin 2s ease-out infinite;
          -moz-animation: spin 2s ease-out infinite;
          animation: spin 2s ease-out infinite;
        }
      `]}static get tag(){return"elmsln-loading"}constructor(){super(),this.size="medium",new Promise((i,n)=>e.default(["../lrn-icons/lrn-icons.js"],i,n)),new Promise((i,n)=>e.default(["../../@polymer/iron-icon/iron-icon.js"],i,n))}updated(i){i.forEach((i,e)=>{"color"==e&&this._getAccentColor(this[e])})}static get properties(){return{...super.properties,color:{type:String},size:{type:String,reflect:!0}}}render(){return n.html`
      <iron-icon icon="lrn:network"></iron-icon>
    `}_getAccentColor(i){i=i.replace("-text",""),this.accentColor&&"grey"!==this.accentColor||!this.colors[i]||(this.accentColor=i)}}i.ElmslnLoading=ElmslnLoading,window.customElements.define(ElmslnLoading.tag,ElmslnLoading)}));
