// @material/mwc-base/form-element.js
define(["exports", "./base-element.js"], function (_exports, _baseElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  Object.defineProperty(_exports, "addHasRemoveClass", {
    enumerable: true,
    get: function () {
      return _baseElement.addHasRemoveClass;
    }
  });
  Object.defineProperty(_exports, "BaseElement", {
    enumerable: true,
    get: function () {
      return _baseElement.BaseElement;
    }
  });
  _exports.FormElement = void 0;

  /**
  @license
  Copyright 2018 Google Inc. All Rights Reserved.
  
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at
  
      http://www.apache.org/licenses/LICENSE-2.0
  
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  */

  /** @soyCompatible */
  class FormElement extends _baseElement.BaseElement {
    createRenderRoot() {
      return this.attachShadow({
        mode: 'open',
        delegatesFocus: true
      });
    }

    click() {
      if (this.formElement) {
        this.formElement.focus();
        this.formElement.click();
      }
    }

    setAriaLabel(label) {
      if (this.formElement) {
        this.formElement.setAttribute('aria-label', label);
      }
    }

    firstUpdated() {
      super.firstUpdated();
      this.mdcRoot.addEventListener('change', e => {
        this.dispatchEvent(new Event('change', e));
      });
    }

  }

  _exports.FormElement = FormElement;
});
