// @lrnwebcomponents/mark-the-words/mark-the-words.js
/**
 * Copyright 2021 The Pennsylvania State University
 * @license Apache-2.0, see License.md for full text.
 * @demo demo/index.html
 */
import{LitElement as t,html as e,css as r}from"../../lit/index.js";import"../simple-toolbar/lib/simple-toolbar-button.js";import{I18NMixin as s}from"../i18n-manager/lib/I18NMixin.js";export class MarkTheWords extends(s(t)){static get tag(){return"mark-the-words"}constructor(){super(),this.answers=null,this.correctAnswers=[],this.wordList=this.innerText.trim().split(/\s+/g),this.question="Mark the words that are correct",this.isEnabled=!0,this.numberCorrect=0,this.numberGuessed=0,this._haxstate=!1,this.demoMode=!1,this.t={check:"Check",tryAgain:"Try Again",editMode:"Edit Mode",tryMode:"Try Mode"}}static get properties(){return{...super.properties,wordList:{type:Array},demoMode:{type:Boolean},answers:{type:String},correctAnswers:{type:Array},missedAnswers:{type:Array},question:{type:String},isEnabled:{type:Boolean},_haxstate:{type:Boolean},numberCorrect:{type:Number},numberGuessed:{type:Number}}}updated(t){super.updated(t),t.forEach(((t,e)=>{if("wordList"===e&&this.rebuildContents(this[e]),"answers"===e&&this[e]){this.correctAnswers=this[e].split(",");for(var r=0;r<this.correctAnswers.length;r++)this.correctAnswers[r]=this.correctAnswers[r].trim().toUpperCase()}"numberCorrect"===e&&this.numberCorrect===this.numberGuessed&&0!==this.numberCorrect&&0!==this.numberGuessed&&import("../multiple-choice/lib/confetti-container.js").then((t=>{setTimeout((()=>{this.shadowRoot.querySelector("#confetti").setAttribute("popped","")}),0)}))}))}rebuildContents(t){this.shadowRoot.querySelector(".text").innerHTML="",t.forEach((t=>{let e=document.createElement("button");e.innerText=t,e.addEventListener("click",this.selectWord.bind(this)),this.shadowRoot.querySelector(".text").appendChild(e)}))}static get styles(){return[r`
        :host {
          display: grid;
          padding: 8px;
          margin: 0px;
          border: 2px solid black;
        }
        simple-toolbar-button {
          font-size: 24px;
        }
        button {
          display: inline-flex;
          font-size: var(--x-ample-font-size, 20px);
          padding: 4px 8px;
          margin: 0px 2px;
          border-radius: none;
          transition: outline 0.2s ease-in-out;
        }
        button:active,
        button:hover,
        button:focus {
          outline: 3px dashed blue;
          cursor: pointer;
        }
        button[data-selected] {
          outline: 3px solid orange;
        }
        button[data-selected]:hover,
        button[data-selected]:focus {
          outline: 3px solid blue;
        }
        button[data-status="correct"] {
          outline: 3px solid green;
        }
        button[data-status="correct"]::after {
          content: "+1";
          color: green;
          font-size: 12px;
          border-radius: 50%;
          border: 1px solid green;
          padding: 3px;
          margin-left: 4px;
          line-height: 16px;
          height: 16px;
          width: 16px;
        }
        button[data-status="incorrect"] {
          outline: 3px solid purple;
        }
        button[data-status="incorrect"]::after {
          content: "-1";
          color: purple;
          font-size: 12px;
          border-radius: 50%;
          border: 1px solid purple;
          padding: 3px;
          margin-left: 4px;
          line-height: 16px;
          height: 16px;
          width: 16px;
        }
        button[check-mode="active"] {
          pointer-events: none;
        }
        .buttons {
          grid-column: 1;
          grid-row: 3;
          margin: 8px;
          display: flex;
        }
        .correct {
          grid-column: 1;
          grid-row: 3;
          margin-top: 20px;
        }
        .question {
          margin-bottom: 0px;
          margin-top: 0px;
          grid-column: 1;
          grid-row: 1;
          font-size: 32px;
        }
        .text {
          grid-column: 1;
          grid-row: 2;
          margin: 16px;
          border: 2px solid black;
          padding: 16px;
        }
        .scoreExp {
          margin-top: 1px;
          margin-bottom: 1px;
          text-decoration: underline;
        }

        .score {
          margin: 0 16px;
          font-size: 22px;
          font-weight: bold;
          line-height: 30px;
          display: inline-block;
          vertical-align: middle;
        }
      `]}selectWord(t){t.target.getAttribute("data-selected")?t.target.removeAttribute("data-selected"):t.target.setAttribute("data-selected","data-selected")}checkAnswer(t){if(this.isEnabled){this.isEnabled=!1;for(var e=0;e<this.wordList.length;e++)this.correctAnswers.includes(this.wordList[e].replace(/[&#^,+()$~%.":*?<>{}]/g,"").toUpperCase())&&this.numberCorrect++;const t=this.shadowRoot.querySelectorAll(".text button[data-selected]");for(e=0;e<t.length;e++){const r=t[e];this.correctAnswers.includes(r.innerText.replace(/[&#^,+()$~%.":*?<>{}]/g,"").toUpperCase())?(r.setAttribute("data-status","correct"),this.numberGuessed++):(r.setAttribute("data-status","incorrect"),this.numberGuessed>0&&this.numberGuessed--)}const r=this.shadowRoot.querySelectorAll(".text button");for(e=0;e<r.length;e++){r[e].setAttribute("check-mode","active")}}else{this.isEnabled=!0,this.numberGuessed=0,this.numberCorrect=0;const t=this.shadowRoot.querySelectorAll(".text button[data-selected]");for(e=0;e<t.length;e++){const r=t[e];r.getAttribute("data-status")&&(r.removeAttribute("data-status"),r.removeAttribute("data-selected"))}const r=this.shadowRoot.querySelectorAll(".text button");for(e=0;e<r.length;e++){r[e].setAttribute("check-mode","inactive")}}}render(){return e`
      <confetti-container id="confetti">
        <p class="question">${this.question}</p>
        ${this._haxstate&&this.demoMode?e`<slot></slot>`:e`<div class="text"></div>`}
        <div class="buttons">
          <simple-toolbar-button
            @click="${this.checkAnswer}"
            icon="${this.isEnabled?"check":"refresh"}"
            label="${this.isEnabled?this.t.check:this.t.tryAgain}"
            show-text-label
            icon-position="left"
            align-vertical="center"
            align-horizontal="center"
          ></simple-toolbar-button>
          <span class="score">
            ${this.isEnabled?"":e`${this.numberGuessed}/${this.numberCorrect}
                ${Math.round(this.numberGuessed/this.numberCorrect*100*10)/10}%`}
          </span>
          ${this._haxstate?e`<simple-toolbar-button
                @click="${this.toggleDemo}"
                icon="${this.demoMode?"av:play-circle-filled":"image:edit"}"
                label="${this.demoMode?this.t.tryMode:this.t.editMode}"
                show-text-label
                icon-position="left"
                align-vertical="center"
                align-horizontal="center"
              ></simple-toolbar-button>`:""}
        </div>
      </confetti-container>
    `}toggleDemo(){this.demoMode=!this.demoMode,this.demoMode||(this.wordList=this.innerText.trim().split(/\s+/g))}haxHooks(){return{editModeChanged:"haxeditModeChanged",activeElementChanged:"haxactiveElementChanged"}}haxactiveElementChanged(t,e){e&&(this._haxstate=e)}haxeditModeChanged(t){this._haxstate=t}static get haxProperties(){return new URL(`./lib/${this.tag}.haxProperties.json`,import.meta.url).href}}customElements.define(MarkTheWords.tag,MarkTheWords);
