// @lrnwebcomponents/a11y-tabs/lib/a11y-tab.js
define(["exports","../../../lit-element/lit-element.js","../../../@polymer/iron-icons/iron-icons.js"],(function(e,t,a){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.A11yTab=void 0;
/**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class A11yTab extends t.LitElement{static get styles(){return[t.css`
        :host {
          display: block;
          height: var(--a11y-tabs-tab-height, --a11y-tabs-height);
          overflow: var(--a11y-tabs-tab-overflow, --a11y-tabs-overflow);
        }
        /*:host([disabled]) {
          display:none;
        }*/
        :host([hidden]) {
          display: none;
        }
        .sr-only {
          position: absolute;
          left: -99999px;
          height: 0;
          overflow: hidden;
        }
        #content-inner {
          max-width: 100%;
          overflow: auto;
        }
        ::slotted(*[slot="label"]) {
          display: none;
        }
      `]}render(){return t.html`
      <span class="sr-only">Tab ${this.xOfY}</span>
      <slot name="label"></slot>
      <div id="content-inner"><slot></slot></div>
      <span class="sr-only">
        End of tab ${this.xOfY}. Back to <a href="#${this.id}">tabs</a>.
      </span>
    `}static get properties(){return{flag:{type:String,reflect:!0},flagIcon:{type:String,attribute:"flag-icon"},disabled:{type:Boolean,reflect:!0,attribute:"disabled"},hidden:{type:Boolean,reflect:!0},icon:{type:String},id:{type:String,reflect:!0,attribute:"id"},label:{type:String},order:{type:Number},total:{type:Number}}}constructor(){super(),this.hidden=!1,this.order=1,this.total=1,this.addEventListener("a11y-tab-flag",e=>this.handleFlag(e))}get observer(){return new MutationObserver(e=>this._labelChanged())}get xOfY(){return`${this.order} of ${this.total}`}connectedCallback(){super.connectedCallback(),this._labelChanged(),this.observer.observe(this,{childList:!0})}disconnectedCallback(){this.observer.disconnect(),this.removeEventListener("a11y-tab-flag",e=>this.handleFlag(e)),super.disconnectedCallback()}updated(e){e.forEach((e,t)=>{["id","order"].includes(t)&&!this.id&&(this.id=`tab-${this.order}`),["label","order"].includes(t)&&!this.label&&this._labelChanged(),"flag"===t&&this._fireTabChanged(),"flagIcon"===t&&this._fireTabChanged(),"icon"===t&&this._fireTabChanged(),"id"===t&&this._fireTabChanged(),"label"===t&&this._fireTabChanged()})}static get tag(){return"a11y-tab"}_handleFlag(e){this.flag=e.detail.flag,this.flagIcon=e.detail.flagIcon}_labelChanged(){this.label&&""!==this.label.trim()||(this.label=this.querySelector('*[slot="label"]')?this.querySelector('*[slot="label"]').innerHTML:`Tab ${this.order}`)}_fireTabChanged(){this.dispatchEvent(new CustomEvent("a11y-tab-changed",{bubbles:!0,cancelable:!0,composed:!1,detail:this}))}}e.A11yTab=A11yTab,window.customElements.define(A11yTab.tag,A11yTab)}));
