// @lrnwebcomponents/a11y-tabs/lib/a11y-tab.js
define(["exports","../../../lit-element/lit-element.js","../../../@polymer/iron-icons/iron-icons.js"],function(_exports,_litElement,_ironIcons){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.A11yTab=void 0;class A11yTab extends _litElement.LitElement{static get styles(){return[_litElement.css`
        :host {
          display: block;
          height: var(--a11y-tabs-tab-height, --a11y-tabs-height);
          overflow: var(--a11y-tabs-tab-overflow, --a11y-tabs-overflow);
        }
        :host([hidden]) {
          display: none;
        }
        .sr-only {
          position: absolute;
          left: -99999px;
          height: 0;
          overflow: hidden;
        }
      `]}render(){return _litElement.html`
      <span class="sr-only">Tab ${this.__xOfY}</span>
      <slot></slot>
      <span class="sr-only"
        >End of tab ${this.__xOfY}. Back to
        <a href="${this.__toTop}">tabs</a>.</span
      >
    `}static get properties(){return{flag:{type:String,reflect:!0},flagIcon:{type:String,attribute:"flag-icon"},hidden:{type:Boolean,reflect:!0},icon:{type:String},id:{type:String},label:{type:String},__toTop:{type:String},__xOfY:{type:String}}}constructor(){super(),this.flag=null,this.flagIcon=null,this.hidden=!1,this.icon=null,this.id=null,this.label=null,this.__xOfY=null,this.__toTop=null,this.addEventListener("a11y-tab-flag",e=>this.handleFlag(e))}disconnectedCallback(){this.removeEventListener("a11y-tab-flag",e=>this.handleFlag(e)),super.disconnectedCallback()}updated(changedProperties){changedProperties.forEach((oldValue,propName)=>{"flag"===propName&&this._tabChanged(),"flagIcon"===propName&&this._tabChanged(),"icon"===propName&&this._tabChanged(),"id"===propName&&this._tabChanged(),"label"===propName&&this._tabChanged()})}static get tag(){return"a11y-tab"}_handleFlag(e){this.flag=e.detail.flag,this.flagIcon=e.detail.flagIcon}_tabChanged(){this.dispatchEvent(new CustomEvent("a11y-tab-changed",{bubbles:!0,cancelable:!0,composed:!1,detail:this}))}}_exports.A11yTab=A11yTab,window.customElements.define(A11yTab.tag,A11yTab)});
