// @lrnwebcomponents/a11y-tabs/lib/a11y-tab.js
define(["exports","../../../@polymer/polymer/polymer-element.js","../../../@polymer/iron-icons/iron-icons.js","../../../@polymer/iron-a11y-keys/iron-a11y-keys.js"],function(_exports,_polymerElement,_ironIcons,_ironA11yKeys){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.A11yTab=void 0;class A11yTab extends _polymerElement.PolymerElement{static get template(){return _polymerElement.html`
      <style>
        :host {
          display: block;
          @apply --a11y-tab-content;
        }
        :host([flag]) {
          @apply --a11y-tab-flagged-content;
        }
        :host([hidden]) {
          display: none;
        }
        :host .sr-only {
          position: absolute;
          left: -99999px;
          height: 0;
          overflow: hidden;
        }
      </style>
      <span class="sr-only">Tab [[__xOfY]]</span>
      <slot></slot>
      <span class="sr-only"
        >End of tab [[__xOfY]]. Back to <a href$="[[__toTop]]">tabs</a>.</span
      >
    `}static get properties(){return{id:{name:"id",type:"String",value:null,observer:"_tabChanged"},flag:{name:"flag",type:"String",value:null,observer:"_tabChanged",reflectToAttribute:!0},flagIcon:{name:"flagIcon",type:"String",value:null,observer:"_tabChanged"},icon:{name:"id",type:"String",value:null,observer:"_tabChanged"},label:{name:"label",type:"String",value:null,observer:"_tabChanged"},hidden:{name:"hidden",type:"Boolean",value:!1,reflectToAttribute:!0},__toTop:{name:"__toTop",type:"String",value:null},__xOfY:{name:"__xOfY",type:"String",value:null}}}static get tag(){return"a11y-tab"}connectedCallback(){super.connectedCallback(),this.__target=this,this.addEventListener("a11y-tab-flag",e=>{this.flag=e.detail.flag,this.flagIcon=e.detail.flagIcon})}_tabChanged(e){this.dispatchEvent(new CustomEvent("a11y-tab-changed",{bubbles:!0,cancelable:!0,composed:!1,detail:this}))}}_exports.A11yTab=A11yTab,window.customElements.define(A11yTab.tag,A11yTab)});
