// @lrnwebcomponents/a11y-carousel/a11y-carousel.js
define(["exports", "../../lit-element/lit-element.js", "../radio-behaviors/radio-behaviors.js", "./lib/a11y-carousel-button.js"], function (_exports, _litElement, _radioBehaviors, _a11yCarouselButton) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.a11yCarousel = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject2_0c82f1c0c07b11ea85f79731808de288() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          margin: 15px 0;\n        }\n        ::slotted(figure) {\n          margin: 0 auto;\n        }\n        #inner {\n          position: relative;\n        }\n        a11y-carousel-button {\n          border: 1px solid red;\n          position: absolute;\n          margin: 0;\n          padding: 0;\n          top: 0;\n          bottom: 0;\n          display: flex;\n          flex-direction: column;\n          align-items: stretch;\n          justify-content: center;\n          width: 50%;\n        }\n        a11y-carousel-button[button-type=prev] {\n          left: 0;\n          right: 50%;\n        }\n        a11y-carousel-button[button-type=next]{\n          right: 0;\n          left: 50%;\n        }\n        :host([hidden]),\n        ::slotted(figure:not([active])) {\n          display: none !important;\n        }\n        ::part(img){\n          border: 4px solid green;\n        }\n        ::part(fig) img{\n          border: 4px solid green;\n        }\n      "]);

    _templateObject2_0c82f1c0c07b11ea85f79731808de288 = function _templateObject2_0c82f1c0c07b11ea85f79731808de288() {
      return data;
    };

    return data;
  }

  function _templateObject_0c82f1c0c07b11ea85f79731808de288() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <slot name=\"top-buttons\"></slot>\n      <div id=\"inner\">\n        <slot name=\"img\"></slot>\n        <slot></slot>\n        <a11y-carousel-button button-type=\"prev\" controls=\"", "\">\n          <slot name=\"prev\"></slot>\n        </a11y-carousel-button>\n        <a11y-carousel-button button-type=\"next\" controls=\"", "\">\n          <slot name=\"next\"></slot>\n        </a11y-carousel-button>\n      </div>\n      <slot name=\"bottom-buttons\"></slot>\n    "]);

    _templateObject_0c82f1c0c07b11ea85f79731808de288 = function _templateObject_0c82f1c0c07b11ea85f79731808de288() {
      return data;
    };

    return data;
  }

  /**
   * `a11y-carousel`
   * Layers images over each other with a slider interface to compare them
   * @demo demo/index.html
   * @element a11y-carousel
   */
  var a11yCarousel =
  /*#__PURE__*/
  function (_RadioBehaviors) {
    babelHelpers.inherits(a11yCarousel, _RadioBehaviors);
    babelHelpers.createClass(a11yCarousel, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_0c82f1c0c07b11ea85f79731808de288(), this.prev, this.next);
      } // properties available to the custom element for data binding

    }], [{
      key: "styles",
      //styles function
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_0c82f1c0c07b11ea85f79731808de288())];
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(a11yCarousel), "properties", this), {
          figures: {
            type: Array
          }
        });
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */

    }, {
      key: "tag",
      get: function get() {
        return "a11y-carousel";
      }
    }]);

    function a11yCarousel() {
      babelHelpers.classCallCheck(this, a11yCarousel);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(a11yCarousel).call(this));
    }
    /**
     * overrides query selector for slotted children
     * @readonly
     */


    babelHelpers.createClass(a11yCarousel, [{
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        if (babelHelpers.get(babelHelpers.getPrototypeOf(a11yCarousel.prototype), "firstUpdated", this)) babelHelpers.get(babelHelpers.getPrototypeOf(a11yCarousel.prototype), "firstUpdated", this).call(this, changedProperties);

        this._handleSelectionChange();
      }
      /**
       * shows or hides items based on selection
       */

    }, {
      key: "_handleSelectionChange",
      value: function _handleSelectionChange() {
        var _this = this;

        babelHelpers.get(babelHelpers.getPrototypeOf(a11yCarousel.prototype), "_handleSelectionChange", this).call(this);

        this._updateItemData();

        var image = this.querySelector("figure#".concat(this.selection, "[active] > img, figure > img")),
            buttons = this.querySelectorAll("a11y-carousel-button"),
            first = this.itemData && this.itemData[0] ? this.itemData[0].id : undefined,
            last = this.itemData && this.itemData[this.itemData.length - 1] ? this.itemData[this.itemData.length - 1].id : undefined,
            prev = this.itemData[this.selectedIndex - 1] ? this.itemData[this.selectedIndex - 1].id : first,
            next = this.itemData[this.selectedIndex + 1] ? this.itemData[this.selectedIndex + 1].id : last;
        this.style.setProperty("--a11y-carousel-bg-image", "url(".concat(image.src, ")"));
        Object.keys(buttons || {}).forEach(function (key) {
          var button = buttons[key];
          if (button.buttonType === "first") button.controls = first;
          if (button.buttonType === "prev") button.controls = prev;
          if (button.buttonType === "next") button.controls = next;
          if (button.buttonType === "last") button.controls = last;
          button.active = button.controls === _this.selection;
        });
      }
    }, {
      key: "_query",
      get: function get() {
        return "figure";
      }
      /**
       *
       * overrides attribute to apply to selected item
       * @readonly
       */

    }, {
      key: "_selected",
      get: function get() {
        return "active";
      }
      /**
       * overrides name of event that selects item
       * @readonly
       */

    }, {
      key: "_selectEvent",
      get: function get() {
        return "select-carousel-item";
      }
    }, {
      key: "first",
      get: function get() {
        return this.itemData && this.itemData[0] ? this.itemData[0].id : undefined;
      }
    }, {
      key: "prev",
      get: function get() {
        return this.itemData && this.itemData[this.selectedIndex - 1] ? this.itemData[this.selectedIndex - 1].id : this.first;
      }
    }, {
      key: "last",
      get: function get() {
        return this.itemData && this.itemData[this.itemData.length - 1] ? this.itemData[this.itemData.length - 1].id : undefined;
      }
    }, {
      key: "next",
      get: function get() {
        return this.itemData && this.itemData[this.selectedIndex + 1] ? this.itemData[this.selectedIndex + 1].id : this.last;
      }
    }]);
    return a11yCarousel;
  }((0, _radioBehaviors.RadioBehaviors)(_litElement.LitElement));

  _exports.a11yCarousel = a11yCarousel;
  window.customElements.define(a11yCarousel.tag, a11yCarousel);
});
