// @lrnwebcomponents/a11y-carousel/a11y-carousel.js
define(["exports", "../../lit-element/lit-element.js", "../radio-behaviors/radio-behaviors.js", "../../@polymer/iron-icon/iron-icon.js", "../../@polymer/iron-icons/iron-icons.js", "./lib/a11y-carousel-button.js"], function (_exports, _litElement, _radioBehaviors, _ironIcon, _ironIcons, _a11yCarouselButton) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.a11yCarousel = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject5_a51ebf50c61511ea9828e39058a15749() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          margin: 15px 0;\n          --a11y-carousel-background-image: none;\n        }\n        ::slotted(figure) {\n          margin: 0 auto;\n          display: flex;\n          flex-direction: column;\n          justify-content: center;\n          align-items: center;\n        }\n        :host([hidden]),\n        ::slotted(figure:not([active])) {\n          display: none !important;\n        }\n        .sr-only {\n          position: absolute;\n          left: -99999px;\n          width: 0;\n          overflow: hidden;\n        }\n        @media screen {\n          #inner {\n            margin-bottom: 10px;\n            position: relative;\n            z-index: 2;\n          }\n          .buttons {\n            display: flex;\n            align-items: stretch;\n            justify-content: center;\n            z-index: 200;\n          }\n          a11y-carousel-button.prevnext {\n            position: absolute;\n            margin: 0;\n            padding: 0;\n            top: 0;\n            bottom: 0;\n            display: flex;\n            flex-direction: column;\n            align-items: stretch;\n            justify-content: center;\n            border: 0px solid transparent;\n            width: 50%;\n          }\n          a11y-carousel-button.prevnext iron-icon {\n            opacity: 0;\n            width: 30%;\n            height: 30%;\n            transition: 0.5s all ease-in-out;\n          }\n          a11y-carousel-button.prevnext:focus,\n          a11y-carousel-button.prevnext:hover {\n            outline: none;\n          }\n          a11y-carousel-button.prevnext:focus iron-icon,\n          a11y-carousel-button.prevnext:hover iron-icon {\n            opacity: 1;\n            color: var(--a11y-carousel-button-background-color, black);\n          }\n          a11y-carousel-button[button-type=\"prev\"] {\n            left: 0;\n            right: 50%;\n          }\n          a11y-carousel-button[button-type=\"next\"] {\n            right: 0;\n            left: 50%;\n            align-items: flex-end;\n          }\n          a11y-carousel-button.button {\n            flex: 0 0 var(--a11y-carousel-button-width, 40px);\n            height: var(--a11y-carousel-button-height, 10px);\n            margin: var(--a11y-carousel-button-margin, 5px);\n            padding: var(--a11y-carousel-button-padding, 0px);\n            background-color: var(\n              --a11y-carousel-button-background-color,\n              black\n            );\n            border: var(--a11y-carousel-button-border, 1px solid black);\n            border-radius: var(\n              --a11y-carousel-button-border-radius,\n              1px solid black\n            );\n          }\n        }\n      "]);

    _templateObject5_a51ebf50c61511ea9828e39058a15749 = function _templateObject5_a51ebf50c61511ea9828e39058a15749() {
      return data;
    };

    return data;
  }

  function _templateObject4_a51ebf50c61511ea9828e39058a15749() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <a11y-carousel-button class=\"button\" controls=\"", "\">\n                    <span class=\"sr-only\">", "</span>\n                  </a11y-carousel-button>\n                "]);

    _templateObject4_a51ebf50c61511ea9828e39058a15749 = function _templateObject4_a51ebf50c61511ea9828e39058a15749() {
      return data;
    };

    return data;
  }

  function _templateObject3_a51ebf50c61511ea9828e39058a15749() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <div class=\"buttons\">\n              ", "\n            </div>\n          "]);

    _templateObject3_a51ebf50c61511ea9828e39058a15749 = function _templateObject3_a51ebf50c61511ea9828e39058a15749() {
      return data;
    };

    return data;
  }

  function _templateObject2_a51ebf50c61511ea9828e39058a15749() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <a11y-carousel-button\n                class=\"prevnext\"\n                button-type=\"prev\"\n                controls=\"", "\"\n                title=\"", "\"\n              >\n                <span class=\"sr-only\">", "</span>\n                <iron-icon icon=\"icons:chevron-left\"></iron-icon>\n              </a11y-carousel-button>\n              <a11y-carousel-button\n                class=\"prevnext\"\n                button-type=\"next\"\n                controls=\"", "\"\n                title=\"", "\"\n              >\n                <span class=\"sr-only\">", "</span>\n                <iron-icon icon=\"icons:chevron-right\"></iron-icon>\n              </a11y-carousel-button>\n            "]);

    _templateObject2_a51ebf50c61511ea9828e39058a15749 = function _templateObject2_a51ebf50c61511ea9828e39058a15749() {
      return data;
    };

    return data;
  }

  function _templateObject_a51ebf50c61511ea9828e39058a15749() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div><slot name=\"above\"></slot></div>\n      <div id=\"inner\">\n        <slot name=\"img\"></slot>\n        <slot></slot>\n        ", "\n      </div>\n      ", "\n      <div><slot name=\"below\"></slot></div>\n    "]);

    _templateObject_a51ebf50c61511ea9828e39058a15749 = function _templateObject_a51ebf50c61511ea9828e39058a15749() {
      return data;
    };

    return data;
  }

  /**
   * `a11y-carousel`
   * Layers images over each other with a slider interface to compare them
   * @demo demo/index.html
   * @element a11y-carousel
   */
  var a11yCarousel =
  /*#__PURE__*/
  function (_RadioBehaviors) {
    babelHelpers.inherits(a11yCarousel, _RadioBehaviors);
    babelHelpers.createClass(a11yCarousel, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_a51ebf50c61511ea9828e39058a15749(), this.noPrevNext ? "" : (0, _litElement.html)(_templateObject2_a51ebf50c61511ea9828e39058a15749(), this.prev, this.prevLabel, this.nextLabel, this.next, this.nextLabel, this.nextLabel), this.noButtons ? "" : (0, _litElement.html)(_templateObject3_a51ebf50c61511ea9828e39058a15749(), (this.itemData || []).map(function (item) {
          return (0, _litElement.html)(_templateObject4_a51ebf50c61511ea9828e39058a15749(), item.id, item.index + 1);
        })));
      } // properties available to the custom element for data binding

    }], [{
      key: "styles",
      //styles function
      get: function get() {
        return [(0, _litElement.css)(_templateObject5_a51ebf50c61511ea9828e39058a15749())];
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(a11yCarousel), "properties", this), {
          nextLabel: {
            type: String
          },
          noPrevNext: {
            type: Boolean,
            attribute: "no-prev-next"
          },
          noButtons: {
            type: Boolean,
            attribute: "no-buttons"
          },
          prevLabel: {
            type: String
          }
        });
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */

    }, {
      key: "tag",
      get: function get() {
        return "a11y-carousel";
      }
    }]);

    function a11yCarousel() {
      var _this;

      babelHelpers.classCallCheck(this, a11yCarousel);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(a11yCarousel).call(this));
      _this.noButtons = false;
      _this.noPrevNext = false;
      _this.prevLabel = "previous";
      _this.nextLabel = "next";
      return _this;
    }
    /**
     * overrides query selector for slotted children
     * @readonly
     */


    babelHelpers.createClass(a11yCarousel, [{
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        if (babelHelpers.get(babelHelpers.getPrototypeOf(a11yCarousel.prototype), "firstUpdated", this)) babelHelpers.get(babelHelpers.getPrototypeOf(a11yCarousel.prototype), "firstUpdated", this).call(this, changedProperties);

        this._handleSelectionChange();
      }
      /**
       * gets image css for selected image
       *
       * @returns
       * @memberof a11yCarousel
       */

    }, {
      key: "_getImage",
      value: function _getImage() {
        var img = this.querySelector("figure#".concat(this.selection, " > img")) || this.querySelector("figure > img");
        return img ? "url(\"".concat(img.src, "\")") : undefined;
      }
      /**
       * shows or hides items based on selection
       */

    }, {
      key: "_handleSelectionChange",
      value: function _handleSelectionChange() {
        var _this2 = this;

        babelHelpers.get(babelHelpers.getPrototypeOf(a11yCarousel.prototype), "_handleSelectionChange", this).call(this);

        this._updateItemData();

        var buttons = this.querySelectorAll("a11y-carousel-button");
        this.style.setProperty("--a11y-carousel-background-image", this._getImage());
        Object.keys(buttons || {}).forEach(function (key) {
          var button = buttons[key];
          if (button.buttonType === "first") button.controls = _this2.first;
          if (button.buttonType === "prev") button.controls = _this2.prev;
          if (button.buttonType === "next") button.controls = _this2.next;
          if (button.buttonType === "last") button.controls = _this2.last;
          button.active = button.controls === _this2.selection;
        });
      }
    }, {
      key: "__query",
      get: function get() {
        return "figure";
      }
      /**
       *
       * overrides attribute to apply to selected item
       * @readonly
       */

    }, {
      key: "__selected",
      get: function get() {
        return "active";
      }
      /**
       * overrides name of event that selects item
       * @readonly
       */

    }, {
      key: "__selectEvent",
      get: function get() {
        return "select-carousel-item";
      }
      /**
       * gets id for first figure
       *
       * @readonly
       * @memberof a11yCarousel
       */

    }, {
      key: "first",
      get: function get() {
        return this.itemData && this.itemData[0] ? this.itemData[0].id : undefined;
      }
      /**
       * gets id for previous figure
       *
       * @readonly
       * @memberof a11yCarousel
       */

    }, {
      key: "prev",
      get: function get() {
        return this.itemData && this.itemData[this.selectedIndex - 1] ? this.itemData[this.selectedIndex - 1].id : this.first;
      }
      /**
       * gets id for last figure
       *
       * @readonly
       * @memberof a11yCarousel
       */

    }, {
      key: "last",
      get: function get() {
        return this.itemData && this.itemData[this.itemData.length - 1] ? this.itemData[this.itemData.length - 1].id : undefined;
      }
      /**
       * gets id for next figure
       *
       * @readonly
       * @memberof a11yCarousel
       */

    }, {
      key: "next",
      get: function get() {
        return this.itemData && this.itemData[this.selectedIndex + 1] ? this.itemData[this.selectedIndex + 1].id : this.last;
      }
    }]);
    return a11yCarousel;
  }((0, _radioBehaviors.RadioBehaviors)(_litElement.LitElement));

  _exports.a11yCarousel = a11yCarousel;
  window.customElements.define(a11yCarousel.tag, a11yCarousel);
});
