// @lrnwebcomponents/cms-hax/cms-hax.js
define(["exports", "require", "../../lit-element/lit-element.js", "../../@polymer/iron-ajax/iron-ajax.js", "../h-a-x/h-a-x.js"], function (_exports, _require, _litElement, _ironAjax, _hAX) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.CmsHax = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject2_96a621601bb211eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          font-size: 16px;\n          box-sizing: content-box;\n        }\n      "]);

    _templateObject2_96a621601bb211eabd7cc30119585ddb = function _templateObject2_96a621601bb211eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject_96a621601bb211eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <iron-ajax\n        id=\"pageupdateajax\"\n        url=\"", "\"\n        method=\"", "\"\n        body=\"", "\"\n        content-type=\"application/json\"\n        handle-as=\"json\"\n        @response=\"", "\"\n      ></iron-ajax>\n      <h-a-x app-store=\"", "\"></h-a-x>\n    "]);

    _templateObject_96a621601bb211eabd7cc30119585ddb = function _templateObject_96a621601bb211eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  /**
   * `cms-hax`
   * @customElement cms-hax
   * @demo ../../demo/index.html
   */
  var CmsHax =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(CmsHax, _LitElement);
    babelHelpers.createClass(CmsHax, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_96a621601bb211eabd7cc30119585ddb(), this.endPoint, this.method, this.updatePageData, this._handleUpdateResponse, this.appStoreConnection);
      }
    }, {
      key: "_activeHaxBodyUpdated",

      /**
       * Ensure we've imported our content on initial setup
       */
      value: function _activeHaxBodyUpdated(newValue, oldValue) {
        // ensure we import our content once we get an initial registration of active body
        if (newValue != null && !this.__imported) {
          this.__imported = true; // see what's inside of this, in a template tag

          var children = this.querySelector("template"); // convert this template content into the real thing
          // this helps with correctly preserving everything on the way down

          if (children != null) {
            newValue.importContent(children.innerHTML);
          }
        }
      }
      /**
       * Calculate if we have anywhere to redirect to.
       */

    }, {
      key: "_computeRedirectOnSave",
      value: function _computeRedirectOnSave(redirectLocation) {
        if (babelHelpers.typeof(redirectLocation) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          return true;
        }

        return false;
      }
      /**
       * LitElement without shadowRoot
       */

    }, {
      key: "createRenderRoot",
      value: function createRenderRoot() {
        return this;
      }
      /**
       * Set certain data bound values to the store once it's ready
       */

    }, {
      key: "_noticeTagChanges",
      value: function _noticeTagChanges(openDefault, allowedTags, hideExportButton, hidePanelOps, hidePreferencesButton, align) {
        if (window.HaxStore.ready) {
          // double check because this can cause issues
          if (allowedTags) {
            window.HaxStore.instance.validTagList = allowedTags;
          }

          window.HaxStore.instance.haxPanel.hideExportButton = hideExportButton;
          window.HaxStore.instance.haxPanel.hidePanelOps = hidePanelOps;
          window.HaxStore.instance.haxPanel.hidePreferencesButton = hidePreferencesButton;
          window.HaxStore.instance.haxPanel.align = align;

          if (openDefault) {
            window.HaxStore.write("editMode", openDefault, this);
          }
        }
      }
      /**
       * LitElement ready
       */

    }, {
      key: "firstUpdated",
      value: function firstUpdated() {
        this.__applyMO();
      }
      /**
       * Set certain data bound values to the store once it's ready
       */

    }, {
      key: "_storeReady",
      value: function _storeReady(e) {
        var _this2 = this;

        // delay as there can be some timing issues with attributes in CMSs
        setTimeout(function () {
          // trigger the update of different parts of the global state
          _this2._noticeTagChanges(_this2.openDefault, _this2.allowedTags, _this2.hideExportButton, _this2.hidePanelOps, _this2.hidePreferencesButton, _this2.align);

          _this2.__applyMO();
        }, 250);
      }
      /**
       * Created life cycle
       */

    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_96a621601bb211eabd7cc30119585ddb())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "cms-hax";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Default the panel to open
           */
          openDefault: {
            type: Boolean,
            reflect: true,
            attribute: "open-default"
          },

          /**
           * Hide the export button, showing it is good for developers
           * or those doing QA testing of new elements.
           */
          hideExportButton: {
            type: Boolean,
            attribute: "hide-export-button"
          },

          /**
           * Hide the panel operations (save and cancel),
           */
          hidePanelOps: {
            type: Boolean,
            attribute: "hide-panel-ops"
          },

          /**
           * Hide preferences button
           */
          hidePreferencesButton: {
            type: Boolean,
            attribute: "hide-preferences-button"
          },

          /**
           * Direction to align the hax edit panel
           */
          align: {
            type: String
          },

          /**
           * allowed Tags, usually as dictated by the input filtering
           * layer of the backend system that HAX is riding on.
           * While not fullproof, this at least will enforce front-end
           * filtering to match what actually is going to be allowed
           * to be saved in the first place.
           */
          allowedTags: {
            type: Array
          },

          /**
           * Location to save content to.
           */
          endPoint: {
            type: String,
            attribute: "end-point"
          },

          /**
           * Method to save content.
           */
          method: {
            type: String
          },

          /**
           * Page data, body of text as a string.
           */
          updatePageData: {
            type: String,
            attribute: "udpate-page-data"
          },

          /**
           * Connection object for talking to an app store.
           */
          appStoreConnection: {
            type: String,
            attribute: "app-store-connection"
          },

          /**
           * State of the panel
           */
          editMode: {
            type: Boolean,
            reflect: true,
            attribute: "edit-mode"
          },

          /**
           * syncBody
           */
          syncBody: {
            type: Boolean,
            attribute: "sync-body"
          },

          /**
           * Only available if syncBody is true; this allows data binding to the value being worked on in hax-body tag
           */
          bodyValue: {
            type: String,
            attribute: "body-value"
          },

          /**
           * Flag to hide the toast.
           */
          hideMessage: {
            type: Boolean,
            attribute: "hide-message"
          },

          /**
           * Optional URL to redirect to once we save.
           */
          redirectLocation: {
            type: String,
            attribute: "redirect-location"
          },

          /**
           * Option to redirect once we save successfully
           */
          redirectOnSave: {
            type: Boolean,
            attribute: "redirect-on-save"
          },

          /**
           * Reference to activeBody.
           */
          activeHaxBody: {
            type: Object
          },
          __imported: {
            type: Boolean
          }
        };
      }
    }]);

    function CmsHax() {
      var _this;

      babelHelpers.classCallCheck(this, CmsHax);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(CmsHax).call(this));
      _this.__lock = false;
      _this.openDefault = false;
      _this.hideExportButton = true;
      _this.hidePanelOps = false;
      _this.hidePreferencesButton = false;
      _this.align = "right";
      _this.method = "PUT";
      _this.syncBody = false;
      _this.bodyValue = "";
      _this.hideMessage = false;
      _this.__imported = false;
      window.SimpleToast.requestAvailability();
      new Promise(function (res, rej) {
        return _require.default(["./lib/cms-token.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["./lib/cms-block.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["./lib/cms-views.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["./lib/cms-entity.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../simple-toast/simple-toast.js"], res, rej);
      });
      return _this;
    }

    babelHelpers.createClass(CmsHax, [{
      key: "updated",
      value: function updated(changedProperties) {
        var _this3 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "redirectLocation") {
            _this3.redirectOnSave = _this3._computeRedirectOnSave(_this3[propName]);
          }

          if (propName == "activeHaxBody") {
            _this3._activeHaxBodyUpdated(_this3[propName]);
          }

          if (["openDefault", "allowedTags", "hideExportButton", "hidePanelOps", "hidePreferencesButton", "align"].includes(propName)) {
            _this3._noticeTagChanges(_this3.openDefault, _this3.allowedTags, _this3.hideExportButton, _this3.hidePanelOps, _this3.hidePreferencesButton, _this3.align);
          }
        });
      }
      /**
       * detached life cycle
       */

    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        if (this._observer) {
          this._observer.disconnect();

          this._observer = null;
        }

        window.removeEventListener("hax-store-property-updated", this._haxStorePropertyUpdated.bind(this));
        window.removeEventListener("hax-store-ready", this._storeReady.bind(this));
        window.removeEventListener("hax-save", this._saveFired.bind(this));
        babelHelpers.get(babelHelpers.getPrototypeOf(CmsHax.prototype), "disconnectedCallback", this).call(this);
      }
      /**
       * Attached to the DOM; now we can fire event to the store that
       * we exist and are the thing being edited.
       */

    }, {
      key: "connectedCallback",
      value: function connectedCallback() {
        var _this4 = this;

        babelHelpers.get(babelHelpers.getPrototypeOf(CmsHax.prototype), "connectedCallback", this).call(this);
        setTimeout(function () {
          window.addEventListener("hax-store-property-updated", _this4._haxStorePropertyUpdated.bind(_this4));
          window.addEventListener("hax-store-ready", _this4._storeReady.bind(_this4));
          window.addEventListener("hax-save", _this4._saveFired.bind(_this4));
        }, 0);

        this.__applyMO();
      }
    }, {
      key: "__applyMO",
      value: function __applyMO() {
        var _this5 = this;

        // notice ANY change to body and bubble up, only when we are attached though
        if (!this._observer && this.syncBody && window.HaxStore && window.HaxStore.instance && window.HaxStore.instance.activeHaxBody) {
          this._observer = new MutationObserver(function (mutations) {
            if (!_this5.__lock) {
              _this5.__lock = true;

              _this5.dispatchEvent(new CustomEvent("hax-body-content-changed", {
                bubbles: true,
                cancelable: true,
                composed: true,
                detail: window.HaxStore.instance.activeHaxBody.haxToContent()
              }));

              setTimeout(function () {
                _this5.__lock = false;
              }, 100);
            }
          });

          this._observer.observe(window.HaxStore.instance.activeHaxBody, {
            childList: true,
            subtree: true
          });
        }
      }
      /**
       * Store updated, sync.
       */

    }, {
      key: "_haxStorePropertyUpdated",
      value: function _haxStorePropertyUpdated(e) {
        if (e.detail && babelHelpers.typeof(e.detail.value) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined)) && e.detail.property) {
          if (babelHelpers.typeof(e.detail.value) === "object") {
            this[e.detail.property] = babelHelpers.objectSpread({}, e.detail.value);
          } else {
            this[e.detail.property] = e.detail.value;
          }
        }
      }
      /**
       * _saveFired
       */

    }, {
      key: "_saveFired",
      value: function _saveFired(e) {
        // generate sanitized content
        this.updatePageData = window.HaxStore.instance.activeHaxBody.haxToContent(); // send the request

        this.querySelector("#pageupdateajax").generateRequest();
      }
      /**
       * _handleUpdateResponse
       */

    }, {
      key: "_handleUpdateResponse",
      value: function _handleUpdateResponse(e) {
        var _this6 = this;

        if (!this.hideMessage) {
          var evt = new CustomEvent("simple-toast-show", {
            bubbles: true,
            cancelable: true,
            composed: true,
            detail: {
              text: "Saved!",
              duration: 3000
            }
          });
          this.dispatchEvent(evt); // support auto redirecting on save if that's been requested
          // in the integration point

          if (this.redirectOnSave) {
            setTimeout(function () {
              // trigger redirect
              window.location = _this6.redirectLocation;
            }, 500);
          }
        }
      }
    }]);
    return CmsHax;
  }(_litElement.LitElement);

  _exports.CmsHax = CmsHax;
  window.customElements.define(CmsHax.tag, CmsHax);
});
