// @lrnwebcomponents/elmsln-apps/lib/elmsmedia-dashboard/elmsmedia-dashboard-toolbar-filter.js
define(["exports", "../../../../@polymer/polymer/polymer-element.js", "../../../../@polymer/paper-icon-button/paper-icon-button.js", "../../../../@polymer/paper-tooltip/paper-tooltip.js", "./paper-chip.js"], function (_exports, _polymerElement, _paperIconButton, _paperTooltip, _paperChip) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmsmediaDashboardToolbarFilter = void 0;

  class ElmsmediaDashboardToolbarFilter extends _polymerElement.PolymerElement {
    static get template() {
      return _polymerElement.html`
      <style>
        :host {
          display: flex;
          font-size: 0.8em;
        }
        paper-icon-button {
          --paper-icon-button: {
            height: 30px;
            width: 30px;
            margin-right: -12px;
          }
        }
      </style>
      <paper-chip>
        [[title]]
        <paper-icon-button
          id="clear"
          icon="clear"
          title="Remove filter"
          on-click="remove"
        ></paper-icon-button>
      </paper-chip>
      <paper-tooltip
        for="clear"
        position="top"
        animation-delay="100"
        margin-top="5"
        aria-hidden="true"
        >Remove Filter</paper-tooltip
      >
    `;
    }

    static get tag() {
      return "elmsmedia-dashboard-toolbar-filter";
    }

    static get properties() {
      return {
        path: {
          type: String,
          value: ""
        },
        propValue: {
          type: String,
          value: ""
        },
        title: {
          type: String,
          value: ""
        }
      };
    }

    remove(e) {
      this.dispatchEvent(new CustomEvent("remove-filter", {
        bubbles: true,
        cancelable: true,
        composed: true,
        detail: {
          path: this.path,
          propValue: this.propValue
        }
      }));
    }

  }

  _exports.ElmsmediaDashboardToolbarFilter = ElmsmediaDashboardToolbarFilter;
  window.customElements.define(ElmsmediaDashboardToolbarFilter.tag, ElmsmediaDashboardToolbarFilter);
});
