// @lrnwebcomponents/haxcms-elements/lib/ui-components/blocks/site-children-block.js
define(["exports","require","../../../../../@polymer/polymer/polymer-element.js","../../core/haxcms-site-store.js","../../../../../mobx/lib/mobx.module.js","../../../../../@polymer/polymer/lib/utils/render-status.js","../query/site-query-menu-slice.js","../../../../../@polymer/polymer/lib/elements/dom-repeat.js"],(function(e,t,i,n,l,r,a,s){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.SiteChildrenBlock=void 0,t=babelHelpers.interopRequireWildcard(t);
/**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class SiteChildrenBlock extends i.PolymerElement{static get tag(){return"site-children-block"}constructor(){super(),this.__disposer=[],new Promise((e,i)=>t.default(["../../../../../@polymer/paper-button/paper-button.js"],e,i))}static get template(){return i.html`
      <style>
        :host {
          display: block;
          --site-children-block-indent: 16px;
          --site-children-block-link-active-bg: rgba(255, 255, 255, 0.1);
          transition: 0.2s opacity linear;
          opacity: 1;
        }
        :host([edit-mode]) {
          opacity: 0.2;
          pointer-events: none;
        }
        .wrapper {
          @apply --site-children-block-wrapper;
        }
        .spacing {
          @apply --site-children-block-spacing;
        }
        .link {
          display: block;
          color: var(--site-children-block-link-color, #444444);
          @apply --site-children-block-link;
        }
        paper-button {
          text-transform: unset;
          min-width: unset;
          width: 100%;
          margin: 0;
          border-radius: 0;
          justify-content: flex-start;
          @apply --site-children-block-button;
        }
        paper-button:hover,
        paper-button:focus,
        paper-button:active {
          @apply --site-children-block-button-active;
        }
        .active {
          color: var(--site-children-block-link-active-color, #000000);
          background-color: var(--site-children-block-link-active-bg);
          @apply --site-children-block-link-active;
        }
        .spacing .indent {
          display: inline-flex;
        }
        .indent-1 {
          margin-left: calc(var(--site-children-block-indent) * 1);
          @apply --site-children-block-indent-1;
        }
        .indent-2 {
          margin-left: calc(var(--site-children-block-indent) * 2);
          @apply --site-children-block-indent-1;
        }
        .indent-3 {
          margin-left: calc(var(--site-children-block-indent) * 3);
          @apply --site-children-block-indent-1;
        }
        .indent-4 {
          margin-left: calc(var(--site-children-block-indent) * 4);
          @apply --site-children-block-indent-1;
        }
        .indent-5 {
          margin-left: calc(var(--site-children-block-indent) * 5);
          @apply --site-children-block-indent-1;
        }
        .indent-6 {
          margin-left: calc(var(--site-children-block-indent) * 6);
          @apply --site-children-block-indent-1;
        }
        .indent-7 {
          margin-left: calc(var(--site-children-block-indent) * 7);
          @apply --site-children-block-indent-1;
        }
        .indent-8 {
          margin-left: calc(var(--site-children-block-indent) * 8);
          @apply --site-children-block-indent-1;
        }
        .indent-9 {
          margin-left: calc(var(--site-children-block-indent) * 9);
          @apply --site-children-block-indent-1;
        }
        .indent-10 {
          margin-left: calc(var(--site-children-block-indent) * 10);
          @apply --site-children-block-indent-1;
        }
      </style>
      <div class="wrapper">
        <site-query-menu-slice
          result="{{__items}}"
          dynamic-methodology="[[dynamicMethodology]]"
          start="[[start]]"
          end="[[end]]"
          parent="[[parent]]"
          fixed-id="[[fixedId]]"
        ></site-query-menu-slice>
        <dom-repeat items="[[__items]]">
          <template>
            <div class="spacing">
              <a
                data-id$="[[item.id]]"
                class="link"
                tabindex="-1"
                href$="[[item.location]]"
              >
                <paper-button noink="[[noink]]">
                  <div class$="indent indent-[[item.indent]]"></div>
                  [[item.title]]
                </paper-button>
              </a>
            </div>
          </template>
        </dom-repeat>
      </div>
    `}static get properties(){return{dynamicMethodology:{type:String,value:"direct"},start:{type:Number,value:1},end:{type:Number,value:1e3},parent:{type:String},fixedId:{type:Boolean,value:!1},noink:{type:Boolean,value:!1},__items:{type:Array,notify:!0},activeId:{type:String,observer:"_activeIdChanged"},editMode:{type:Boolean,reflectToAttribute:!0}}}_activeIdChanged(e){if(e){let t=null;if(this.shadowRoot.querySelector('[data-id="'+e+'"]'))t=this.shadowRoot.querySelector('[data-id="'+e+'"]');else{let i=this.manifest.items.find(t=>t.id==e);for(;null===t&&i&&null!=i.parent;)i=this.manifest.items.find(e=>e.id==i.parent),i&&this.shadowRoot.querySelector('[data-id="'+i.id+'"]')&&(t=this.shadowRoot.querySelector('[data-id="'+i.id+'"]'))}this._prevEl&&this._prevEl.classList.remove("active"),t&&(t.classList.add("active"),this._prevEl=t)}else this._prevEl&&this._prevEl.classList.remove("active")}connectedCallback(){super.connectedCallback(),(0,l.autorun)(e=>{this.editMode=(0,l.toJS)(n.store.editMode),this.__disposer.push(e)}),(0,l.autorun)(e=>{this.manifest=(0,l.toJS)(n.store.manifest),this.__disposer.push(e)}),(0,r.afterNextRender)(this,(function(){setTimeout(()=>{(0,l.autorun)(e=>{this.activeId=(0,l.toJS)(n.store.activeId),this.__disposer.push(e)})},250)}))}disconnectedCallback(){for(var e in this.__disposer)this.__disposer[e].dispose();super.disconnectedCallback()}}e.SiteChildrenBlock=SiteChildrenBlock,window.customElements.define(SiteChildrenBlock.tag,SiteChildrenBlock)}));
