// @lrnwebcomponents/elmsln-apps/lib/elmsmedia-dashboard/elmsmedia-dashboard.js
define(["exports", "../../../../@polymer/polymer/polymer-element.js", "../../../../@polymer/app-route/app-route.js", "../../../../@polymer/app-route/app-location.js", "../../../../@polymer/paper-dialog/paper-dialog.js", "../../../../@polymer/paper-button/paper-button.js", "../../../hax-body/lib/hax-app.js", "./elmsmedia-dashboard-filters.js", "./elmsmedia-dashboard-toolbar-filters.js", "./elmsmedia-dashboard-toolbar-button.js"], function (_exports, _polymerElement, _appRoute, _appLocation, _paperDialog, _paperButton, _haxApp, _elmsmediaDashboardFilters, _elmsmediaDashboardToolbarFilters, _elmsmediaDashboardToolbarButton) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmsmediaDashboard = void 0;

  function _templateObject_54788100e3b511e9b0d5a5c01151ac03() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: block;\n        }\n        paper-dialog {\n          padding: 1em;\n        }\n        #toolbar {\n          display: flex;\n          justify-content: flex-end;\n          align-items: center;\n        }\n      </style>\n\n      <app-location route=\"{{route}}\"></app-location>\n      <app-route\n        route=\"{{route}}\"\n        pattern=\"/:page\"\n        data=\"{{data}}\"\n        tail=\"{{tail}}\"\n        query-params=\"{{queryParams}}\"\n      ></app-route>\n\n      <div id=\"toolbar\">\n        <elmsmedia-dashboard-toolbar-filters\n          filters=\"[[queryParams]]\"\n        ></elmsmedia-dashboard-toolbar-filters>\n        <elmsmedia-dashboard-toolbar-button\n          icon=\"filter-list\"\n          title=\"Filter\"\n          on-click=\"toggleFilters\"\n        ></elmsmedia-dashboard-toolbar-button>\n        <elmsmedia-dashboard-toolbar-button\n          icon=\"search\"\n          title=\"Search\"\n          on-click=\"toggleSearch\"\n        ></elmsmedia-dashboard-toolbar-button>\n      </div>\n\n      <paper-dialog id=\"filterDialog\">\n        <h3>Filter Media</h3>\n        <elmsmedia-dashboard-filters\n          form=\"{{queryParams}}\"\n          on-filter-changed=\"_filterChanged\"\n        ></elmsmedia-dashboard-filters>\n        <div class=\"buttons\">\n          <paper-button dialog-dismiss=\"\">Dismiss dialog</paper-button>\n        </div>\n      </paper-dialog>\n\n      <hax-app\n        id=\"haxSource\"\n        auto=\"\"\n        query-param=\"title\"\n        request-end-point=\"[[requestEndPoint]]\"\n        request-params=\"{{queryParams}}\"\n        data='{\n      \"root\": \"list\",\n      \"gizmoType\": \"video\",\n      \"url\": \"http://media.elmsln.local/entity_iframe/node/\",\n      \"id\": \"id\",\n      \"title\": \"attributes.title\",\n      \"description\": \"attributes.body\",\n      \"image\": \"display.image\",\n      \"customGizmoType\": \"type\"}'\n      ></hax-app>\n    "]);

    _templateObject_54788100e3b511e9b0d5a5c01151ac03 = function _templateObject_54788100e3b511e9b0d5a5c01151ac03() {
      return data;
    };

    return data;
  }

  var ElmsmediaDashboard =
  /*#__PURE__*/
  function (_PolymerElement) {
    babelHelpers.inherits(ElmsmediaDashboard, _PolymerElement);

    function ElmsmediaDashboard() {
      babelHelpers.classCallCheck(this, ElmsmediaDashboard);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(ElmsmediaDashboard).apply(this, arguments));
    }

    babelHelpers.createClass(ElmsmediaDashboard, [{
      key: "_computeRequestEndPoint",
      value: function _computeRequestEndPoint(endPoint, csrfToken) {
        return "".concat(endPoint, "/api/data?token=").concat(csrfToken);
      }
      /**
       * Listen for filter changes and reset the page count
       * @todo unsetting this object does not always work
       */

    }, {
      key: "_filterChanged",
      value: function _filterChanged(e) {
        var newParams = Object.assign({}, e.detail);
        newParams = this._cleanParams(newParams);
        this.set("queryParams", {});
        this.set("queryParams", newParams);
      }
      /**
       * Helper function to prepare parameters object for the url
       * @todo this is horribly written
       */

    }, {
      key: "_cleanParams",
      value: function _cleanParams(params) {
        var newParams = {};

        for (x in params) {
          var prop = x;
          var value = params[x];

          if (value === "" || value === null || value === undefined) {// do nothing
          } else {
            // add the clean property to the new params object
            newParams[prop] = value;
          }
        }

        return newParams;
      }
    }, {
      key: "toggleSearch",
      value: function toggleSearch(e) {
        var searchEnabled = this.$.haxSource.search;
        this.$.haxSource.search = !searchEnabled;
      }
    }, {
      key: "toggleFilters",
      value: function toggleFilters(e) {
        this.$.filterDialog.toggle();
      }
    }, {
      key: "removeFilter",
      value: function removeFilter(path, propValue) {
        var currentParams = Object.assign({}, this.queryParams); // get the current value of the filter

        var currentValue = _.get(currentParams, path); // remove the specified filter from the current value
        // by converting to an array and filtering it


        var newValue = currentValue.split(",").filter(function (v) {
          return v !== propValue;
        }).join(",");

        var newParams = _.set(currentParams, path, newValue);

        var newCleanParams = this._cleanParams(newParams);

        this.set("queryParams", {});
        this.set("queryParams", newCleanParams);
      }
    }, {
      key: "ready",
      value: function ready() {
        var _this = this;

        babelHelpers.get(babelHelpers.getPrototypeOf(ElmsmediaDashboard.prototype), "ready", this).call(this);
        this.addEventListener("remove-filter", function (e) {
          _this.removeFilter(e.detail.path, e.detail.propValue);
        });
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        var _this2 = this;

        babelHelpers.get(babelHelpers.getPrototypeOf(ElmsmediaDashboard.prototype), "disconnectedCallback", this).call(this);
        this.removeEventListener("remove-filter", function (e) {
          _this2.removeFilter(e.detail.path, e.detail.propValue);
        });
      }
    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_54788100e3b511e9b0d5a5c01151ac03());
      }
    }, {
      key: "tag",
      get: function get() {
        return "elmsmedia-dashboard";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          requestEndPoint: {
            type: String,
            value: ""
          },
          queryParams: {
            type: Object,
            value: {}
          }
        };
      }
    }]);
    return ElmsmediaDashboard;
  }(_polymerElement.PolymerElement);

  _exports.ElmsmediaDashboard = ElmsmediaDashboard;
  window.customElements.define(ElmsmediaDashboard.tag, ElmsmediaDashboard);
});
