// @lrnwebcomponents/elmsln-apps/lib/elmsmedia-dashboard/elmsmedia-dashboard-toolbar-filters.js
define(["exports", "../../../../@polymer/polymer/polymer-element.js", "../../../../@polymer/polymer/lib/elements/dom-repeat.js", "./elmsmedia-dashboard-toolbar-filter.js"], function (_exports, _polymerElement, _domRepeat, _elmsmediaDashboardToolbarFilter) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmsmediaDashboardToolbarFilters = void 0;

  function _templateObject_5ba15aa0dbdc11e9bcafdb98156fc1d6() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: flex;\n          align-items: center;\n        }\n        elmsmedia-dashboard-toolbar-filter {\n          margin-right: 4.8px;\n        }\n      </style>\n\n      <template is=\"dom-repeat\" items=\"[[_filtersArray]]\" as=\"item\">\n        <elmsmedia-dashboard-toolbar-filter\n          path=\"[[item.path]]\"\n          prop-value=\"[[item.propValue]]\"\n          title=\"[[item.title]]\"\n        ></elmsmedia-dashboard-toolbar-filter>\n      </template>\n    "]);

    _templateObject_5ba15aa0dbdc11e9bcafdb98156fc1d6 = function _templateObject_5ba15aa0dbdc11e9bcafdb98156fc1d6() {
      return data;
    };

    return data;
  }

  var ElmsmediaDashboardToolbarFilters =
  /*#__PURE__*/
  function (_PolymerElement) {
    babelHelpers.inherits(ElmsmediaDashboardToolbarFilters, _PolymerElement);

    function ElmsmediaDashboardToolbarFilters() {
      babelHelpers.classCallCheck(this, ElmsmediaDashboardToolbarFilters);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(ElmsmediaDashboardToolbarFilters).apply(this, arguments));
    }

    babelHelpers.createClass(ElmsmediaDashboardToolbarFilters, [{
      key: "_filtersArrayCompute",
      value: function _filtersArrayCompute(filters) {
        var filtersArray = [];

        var _loop = function _loop() {
          var prop = f; // account for mulitple filter values

          var values = filters[f].split(","); // ignore the page property

          if (prop !== "page") {
            values.forEach(function (value) {
              filtersArray.push({
                path: prop,
                propValue: value,
                title: "".concat(prop, ": ").concat(value)
              });
            });
          }
        };

        for (f in filters) {
          _loop();
        }

        return filtersArray;
      }
    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_5ba15aa0dbdc11e9bcafdb98156fc1d6());
      }
    }, {
      key: "tag",
      get: function get() {
        return "elmsmedia-dashboard-toolbar-filters";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          filters: {
            type: Object,
            value: {}
          },
          _filtersArray: {
            type: Array,
            computed: "_filtersArrayCompute(filters, filters.*)"
          }
        };
      }
    }]);
    return ElmsmediaDashboardToolbarFilters;
  }(_polymerElement.PolymerElement);

  _exports.ElmsmediaDashboardToolbarFilters = ElmsmediaDashboardToolbarFilters;
  window.customElements.define(ElmsmediaDashboardToolbarFilters.tag, ElmsmediaDashboardToolbarFilters);
});
