// @lrnwebcomponents/elmsln-apps/lib/elmsmedia-dashboard/elmsmedia-dashboard-toolbar-filter.js
define(["exports", "../../../../@polymer/polymer/polymer-element.js", "../../../../@polymer/paper-icon-button/paper-icon-button.js", "../../../../@polymer/paper-tooltip/paper-tooltip.js", "./paper-chip.js"], function (_exports, _polymerElement, _paperIconButton, _paperTooltip, _paperChip) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmsmediaDashboardToolbarFilter = void 0;

  function _templateObject_54386ac0e3b511e9b0d5a5c01151ac03() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: flex;\n          font-size: 0.8em;\n        }\n        paper-icon-button {\n          --paper-icon-button: {\n            height: 30px;\n            width: 30px;\n            margin-right: -12px;\n          }\n        }\n      </style>\n      <paper-chip>\n        [[title]]\n        <paper-icon-button\n          id=\"clear\"\n          icon=\"clear\"\n          title=\"Remove filter\"\n          on-click=\"remove\"\n        ></paper-icon-button>\n      </paper-chip>\n      <paper-tooltip\n        for=\"clear\"\n        position=\"top\"\n        animation-delay=\"100\"\n        margin-top=\"5\"\n        aria-hidden=\"true\"\n        >Remove Filter</paper-tooltip\n      >\n    "]);

    _templateObject_54386ac0e3b511e9b0d5a5c01151ac03 = function _templateObject_54386ac0e3b511e9b0d5a5c01151ac03() {
      return data;
    };

    return data;
  }

  var ElmsmediaDashboardToolbarFilter =
  /*#__PURE__*/
  function (_PolymerElement) {
    babelHelpers.inherits(ElmsmediaDashboardToolbarFilter, _PolymerElement);

    function ElmsmediaDashboardToolbarFilter() {
      babelHelpers.classCallCheck(this, ElmsmediaDashboardToolbarFilter);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(ElmsmediaDashboardToolbarFilter).apply(this, arguments));
    }

    babelHelpers.createClass(ElmsmediaDashboardToolbarFilter, [{
      key: "remove",
      value: function remove(e) {
        this.dispatchEvent(new CustomEvent("remove-filter", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: {
            path: this.path,
            propValue: this.propValue
          }
        }));
      }
    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_54386ac0e3b511e9b0d5a5c01151ac03());
      }
    }, {
      key: "tag",
      get: function get() {
        return "elmsmedia-dashboard-toolbar-filter";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          path: {
            type: String,
            value: ""
          },
          propValue: {
            type: String,
            value: ""
          },
          title: {
            type: String,
            value: ""
          }
        };
      }
    }]);
    return ElmsmediaDashboardToolbarFilter;
  }(_polymerElement.PolymerElement);

  _exports.ElmsmediaDashboardToolbarFilter = ElmsmediaDashboardToolbarFilter;
  window.customElements.define(ElmsmediaDashboardToolbarFilter.tag, ElmsmediaDashboardToolbarFilter);
});
