// @lrnwebcomponents/elmsln-apps/lib/elmsmedia-dashboard/elmsmedia-dashboard-filters.js
define(["exports", "../../../../@polymer/polymer/polymer-element.js", "../../../../@polymer/paper-icon-button/paper-icon-button.js", "../../../../@polymer/paper-dropdown-menu/paper-dropdown-menu.js", "../../../../@polymer/paper-listbox/paper-listbox.js", "../../../../@polymer/paper-item/paper-item.js"], function (_exports, _polymerElement, _paperIconButton, _paperDropdownMenu, _paperListbox, _paperItem) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmsmediaDashboardFilters = void 0;

  function _templateObject_69bbb6d0e14a11e9bce923988a998e3c() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: block;\n        }\n        paper-dropdown-menu {\n          display: block;\n          width: 100%;\n        }\n      </style>\n\n      <paper-dropdown-menu label=\"Order\">\n        <paper-listbox\n          slot=\"dropdown-content\"\n          attr-for-selected=\"name\"\n          selected=\"{{form.order}}\"\n        >\n          <paper-item name=\"ASC\">Ascending</paper-item>\n          <paper-item name=\"DESC\">Descending</paper-item>\n        </paper-listbox>\n      </paper-dropdown-menu>\n\n      <paper-dropdown-menu label=\"Media Type\">\n        <paper-listbox\n          slot=\"dropdown-content\"\n          attr-for-selected=\"name\"\n          selected=\"{{form.media_type}}\"\n        >\n          <paper-item name=\"elmsmedia_image\">Image</paper-item>\n          <paper-item name=\"h5p\">H5P</paper-item>\n          <paper-item name=\"video\">Video</paper-item>\n          <paper-item name=\"external_video\">External Video</paper-item>\n          <paper-item name=\"audio\">audio</paper-item>\n        </paper-listbox>\n      </paper-dropdown-menu>\n    "]);

    _templateObject_69bbb6d0e14a11e9bce923988a998e3c = function _templateObject_69bbb6d0e14a11e9bce923988a998e3c() {
      return data;
    };

    return data;
  }

  var ElmsmediaDashboardFilters =
  /*#__PURE__*/
  function (_PolymerElement) {
    babelHelpers.inherits(ElmsmediaDashboardFilters, _PolymerElement);

    function ElmsmediaDashboardFilters() {
      babelHelpers.classCallCheck(this, ElmsmediaDashboardFilters);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(ElmsmediaDashboardFilters).apply(this, arguments));
    }

    babelHelpers.createClass(ElmsmediaDashboardFilters, [{
      key: "_formChanged",
      value: function _formChanged(form) {
        var path = form.path.replace("form.", "");
        var propValue = form.value;
        this.dispatchEvent(new CustomEvent("add-filter", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: {
            path: path,
            propValue: propValue
          }
        }));
      }
    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_69bbb6d0e14a11e9bce923988a998e3c());
      }
    }, {
      key: "tag",
      get: function get() {
        return "elmsmedia-dashboard-filters";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          form: {
            type: Object,
            value: {}
          }
        };
      }
    }, {
      key: "observers",
      get: function get() {
        return ["_formChanged(form.*)"];
      }
    }]);
    return ElmsmediaDashboardFilters;
  }(_polymerElement.PolymerElement);

  _exports.ElmsmediaDashboardFilters = ElmsmediaDashboardFilters;
  window.customElements.define(ElmsmediaDashboardFilters.tag, ElmsmediaDashboardFilters);
});
