// @lrnwebcomponents/h5p-element/lib/h5p-wrapped-element.js
define(["exports", "../../haxcms-elements/lib/core/haxcms-site-store.js", "../../../@polymer/polymer/lib/utils/flattened-nodes-observer.js", "../../../lit/index.js", "../../../mobx/dist/mobx.esm.js", "../../iframe-loader/iframe-loader.js"], function (_exports, _haxcmsSiteStore, _flattenedNodesObserver, _index, _mobxEsm, _iframeLoader) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.H5pWrappedElement = void 0;

  var _templateObject_6384e820942811ecb1937969798b231a, _templateObject2_6384e820942811ecb1937969798b231a, _templateObject3_6384e820942811ecb1937969798b231a, _templateObject4_6384e820942811ecb1937969798b231a;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  // wrapped implementation of h5p which is an iframe on the page
  // if we use this method then we can try and see if HAXcms is around
  // this won't hurt anything if it's not, but will opt into edit mode if it is
  var H5pWrappedElement = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(H5pWrappedElement, _LitElement);

    var _super = _createSuper(H5pWrappedElement);

    function H5pWrappedElement() {
      var _this;

      babelHelpers.classCallCheck(this, H5pWrappedElement);
      _this = _super.call(this);
      _this.src = "";
      _this.__disposer = (0, _mobxEsm.autorun)(function () {
        _this.__editMode = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.editMode);
      });
      return _this;
    }

    babelHelpers.createClass(H5pWrappedElement, [{
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        babelHelpers.get(babelHelpers.getPrototypeOf(H5pWrappedElement.prototype), "disconnectedCallback", this).call(this);

        this.__disposer();
      }
    }, {
      key: "firstUpdated",
      value: function firstUpdated() {
        var _this2 = this;

        this._observer = new _flattenedNodesObserver.FlattenedNodesObserver(this, function (info) {
          info.addedNodes.forEach(function (item) {
            var iframe = item.nodeName === "IFRAME" ? item : _this2.querySelector("iframe");

            if (iframe) {
              if (typeof iframe.src !== "undefined") {
                // specific to elms / drupal paths from media system
                _this2.__editLink = iframe.src.replace("entity_iframe/", "") + "/edit";
              }
            }
          });
        });
      }
    }, {
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_6384e820942811ecb1937969798b231a || (_templateObject_6384e820942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <div part=\"container\" class=\"", "\">\n        ", "\n      </div>\n    "])), this.__editMode ? "editing" : "editing", this.__editMode ? (0, _index.html)(_templateObject2_6384e820942811ecb1937969798b231a || (_templateObject2_6384e820942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n              <slot></slot>\n              <div part=\"edit-screen\">\n                <div part=\"source-link\">\n                  <a part=\"anchor\" href=\"", "\" target=\"_blank\"\n                    >Edit H5P Source</a\n                  >\n                </div>\n              </div>\n            "])), this.__editLink) : (0, _index.html)(_templateObject3_6384e820942811ecb1937969798b231a || (_templateObject3_6384e820942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n              <iframe-loader><slot></slot></iframe-loader>\n            "]))));
      }
    }], [{
      key: "properties",
      get: function get() {
        return {
          __editLink: {
            type: String,
            attribute: false
          },
          __editMode: {
            type: Boolean,
            attribute: false
          }
        };
      }
    }, {
      key: "styles",
      get: function get() {
        return (0, _index.css)(_templateObject4_6384e820942811ecb1937969798b231a || (_templateObject4_6384e820942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      :host {\n        display: block;\n      }\n      :host [part=\"container\"] {\n        display: block;\n        position: relative;\n      }\n      .editing[part=\"container\"] {\n        min-height: 100px;\n      }\n      :host [part=\"edit-screen\"] {\n        position: absolute;\n        left: 0;\n        top: 0;\n        width: 100%;\n        height: 100%;\n        display: flex;\n        align-items: flex-end;\n        justify-content: flex-end;\n      }\n      :host [part=\"source-link\"] a {\n        background: black;\n        color: white;\n        padding: 1em;\n      }\n      :host [part=\"anchor\"] {\n        display: block;\n      }\n    "])));
      }
    }]);
    return H5pWrappedElement;
  }(_index.LitElement);

  _exports.H5pWrappedElement = H5pWrappedElement;
  customElements.define("h5p-wrapped-element", H5pWrappedElement);
});
