// @lrnwebcomponents/lrndesign-gallerycard/lrndesign-gallerycard.js
define(["exports", "require", "../../@polymer/polymer/polymer-element.js", "../../@polymer/polymer/lib/utils/render-status.js"], function (_exports, _require, _polymerElement, _renderStatus) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrndesignGallerycard = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject_31ce5c00b3b911e9b58d79562a23326c() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: inline-flex;\n        }\n        :host([size=\"micro\"]) {\n          transform: scale(0.5);\n        }\n        :host([size=\"small\"]) {\n          transform: scale(0.8);\n        }\n\n        paper-card {\n          border-radius: 4px;\n          margin: 0;\n          width: 100%;\n        }\n\n        .card-actions {\n          background-color: #f5f5f5;\n          border-radius: 0 0 4px 4px;\n          padding: 0 8px;\n        }\n        .card-actions .card-action-details {\n          display: inline-block;\n          vertical-align: middle;\n          vertical-align: -webkit-baseline-middle;\n          width: 80%;\n        }\n        #avatar {\n          display: inline-block;\n          vertical-align: text-top;\n          transform: scale(0.8);\n        }\n\n        .card-control-height {\n          height: 240px;\n        }\n\n        [elevation=\"0\"] {\n          border: solid 1px #eeeeee;\n        }\n\n        .text-right {\n          text-align: right;\n        }\n\n        .text-left {\n          text-align: left;\n        }\n\n        .title {\n          color: #222;\n          font-size: 12.8px;\n          font-weight: 600;\n          line-height: 20px;\n          padding: 0 12px;\n          overflow: hidden;\n          text-overflow: ellipsis;\n          margin-top: 8px;\n        }\n\n        .comments {\n          font-size: 12px;\n          float: right;\n        }\n\n        .divider {\n          height: 1px;\n          width: 100%;\n          background: #efefef;\n        }\n\n        .project-icon {\n          --iron-icon-height: 100%;\n          --iron-icon-width: 100%;\n          overflow: hidden;\n          color: grey;\n        }\n        .project-icon:hover,\n        .project-icon:focus {\n          color: black;\n        }\n\n        .center {\n          margin: auto;\n          width: 80%;\n          padding: 16px;\n        }\n\n        .link {\n          font-size: 16px;\n          line-height: 16px;\n        }\n\n        .submission-info {\n          width: 100%;\n        }\n        .submission-preview {\n          height: 160px;\n        }\n\n        .card-content {\n          padding: 0;\n          margin: 0;\n          overflow: hidden;\n        }\n\n        .inline {\n          display: inline;\n        }\n      </style>\n      <paper-card elevation=\"[[elevation]]\">\n        <div class=\"card-content card-control-height card-control-center\">\n          <div class=\"submission-preview\">\n            <iron-icon\n              class=\"project-icon\"\n              icon=\"[[icon]]\"\n              hidden$=\"[[!icon]]\"\n            ></iron-icon>\n            <iron-image\n              style=\"width:100%; height:100%; background-color: lightgray;\"\n              sizing=\"cover\"\n              preload=\"\"\n              fade=\"\"\n              src=\"[[image]]\"\n              hidden$=\"[[!image]]\"\n            ></iron-image>\n          </div>\n          <div class=\"submission-info\">\n            <div class=\"divider\"></div>\n            <div class=\"title\">[[title]]</div>\n          </div>\n        </div>\n        <div class=\"card-actions\">\n          <lrndesign-avatar\n            id=\"avatar\"\n            label=\"[[author.name]]\"\n            src=\"[[author.avatar]]\"\n          ></lrndesign-avatar>\n          <div class=\"card-action-details\">\n            <span class=\"text-left author\">[[author.display_name]]</span>\n            <span class=\"comments text-right\">Comments: [[comments]]</span>\n          </div>\n        </div>\n      </paper-card>\n    "], ["\n      <style>\n        :host {\n          display: inline-flex;\n        }\n        :host([size=\"micro\"]) {\n          transform: scale(0.5);\n        }\n        :host([size=\"small\"]) {\n          transform: scale(0.8);\n        }\n\n        paper-card {\n          border-radius: 4px;\n          margin: 0;\n          width: 100%;\n        }\n\n        .card-actions {\n          background-color: #f5f5f5;\n          border-radius: 0 0 4px 4px;\n          padding: 0 8px;\n        }\n        .card-actions .card-action-details {\n          display: inline-block;\n          vertical-align: middle;\n          vertical-align: -webkit-baseline-middle;\n          width: 80%;\n        }\n        #avatar {\n          display: inline-block;\n          vertical-align: text-top;\n          transform: scale(0.8);\n        }\n\n        .card-control-height {\n          height: 240px;\n        }\n\n        [elevation=\"0\"] {\n          border: solid 1px #eeeeee;\n        }\n\n        .text-right {\n          text-align: right;\n        }\n\n        .text-left {\n          text-align: left;\n        }\n\n        .title {\n          color: #222;\n          font-size: 12.8px;\n          font-weight: 600;\n          line-height: 20px;\n          padding: 0 12px;\n          overflow: hidden;\n          text-overflow: ellipsis;\n          margin-top: 8px;\n        }\n\n        .comments {\n          font-size: 12px;\n          float: right;\n        }\n\n        .divider {\n          height: 1px;\n          width: 100%;\n          background: #efefef;\n        }\n\n        .project-icon {\n          --iron-icon-height: 100%;\n          --iron-icon-width: 100%;\n          overflow: hidden;\n          color: grey;\n        }\n        .project-icon:hover,\n        .project-icon:focus {\n          color: black;\n        }\n\n        .center {\n          margin: auto;\n          width: 80%;\n          padding: 16px;\n        }\n\n        .link {\n          font-size: 16px;\n          line-height: 16px;\n        }\n\n        .submission-info {\n          width: 100%;\n        }\n        .submission-preview {\n          height: 160px;\n        }\n\n        .card-content {\n          padding: 0;\n          margin: 0;\n          overflow: hidden;\n        }\n\n        .inline {\n          display: inline;\n        }\n      </style>\n      <paper-card elevation=\"[[elevation]]\">\n        <div class=\"card-content card-control-height card-control-center\">\n          <div class=\"submission-preview\">\n            <iron-icon\n              class=\"project-icon\"\n              icon=\"[[icon]]\"\n              hidden\\$=\"[[!icon]]\"\n            ></iron-icon>\n            <iron-image\n              style=\"width:100%; height:100%; background-color: lightgray;\"\n              sizing=\"cover\"\n              preload=\"\"\n              fade=\"\"\n              src=\"[[image]]\"\n              hidden\\$=\"[[!image]]\"\n            ></iron-image>\n          </div>\n          <div class=\"submission-info\">\n            <div class=\"divider\"></div>\n            <div class=\"title\">[[title]]</div>\n          </div>\n        </div>\n        <div class=\"card-actions\">\n          <lrndesign-avatar\n            id=\"avatar\"\n            label=\"[[author.name]]\"\n            src=\"[[author.avatar]]\"\n          ></lrndesign-avatar>\n          <div class=\"card-action-details\">\n            <span class=\"text-left author\">[[author.display_name]]</span>\n            <span class=\"comments text-right\">Comments: [[comments]]</span>\n          </div>\n        </div>\n      </paper-card>\n    "]);

    _templateObject_31ce5c00b3b911e9b58d79562a23326c = function _templateObject_31ce5c00b3b911e9b58d79562a23326c() {
      return data;
    };

    return data;
  }

  /**
  `lrndesign-gallerycard`
  A LRN element for presenting a gallery of items as cards
  that can pop up to display more info
  * @demo demo/index.html
  */
  var LrndesignGallerycard =
  /*#__PURE__*/
  function (_PolymerElement) {
    babelHelpers.inherits(LrndesignGallerycard, _PolymerElement);

    function LrndesignGallerycard() {
      var _this;

      babelHelpers.classCallCheck(this, LrndesignGallerycard);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrndesignGallerycard).call(this));
      new Promise(function (res, rej) {
        return _require.default(["../lrndesign-avatar/lrndesign-avatar.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/iron-image/iron-image.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/paper-card/paper-card.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/iron-icon/iron-icon.js"], res, rej);
      });
      return _this;
    }

    babelHelpers.createClass(LrndesignGallerycard, [{
      key: "connectedCallback",
      value: function connectedCallback() {
        babelHelpers.get(babelHelpers.getPrototypeOf(LrndesignGallerycard.prototype), "connectedCallback", this).call(this);
        (0, _renderStatus.afterNextRender)(this, function () {
          this.addEventListener("mouseenter", this._mouseEnter.bind(this));
          this.addEventListener("mouseleave", this._mouseLeave.bind(this));
        });
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        this.removeEventListener("mouseenter", this._mouseEnter.bind(this));
        this.removeEventListener("mouseleave", this._mouseLeave.bind(this));
        babelHelpers.get(babelHelpers.getPrototypeOf(LrndesignGallerycard.prototype), "disconnectedCallback", this).call(this);
      }
    }, {
      key: "_mouseEnter",
      value: function _mouseEnter(e) {
        this.elevation += 2;
      }
    }, {
      key: "_mouseLeave",
      value: function _mouseLeave(e) {
        this.elevation -= 2;
      }
    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_31ce5c00b3b911e9b58d79562a23326c());
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrndesign-gallerycard";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          size: {
            type: String,
            notify: true,
            reflectToAttribute: true
          },

          /**
           * Cover image src.
           */
          image: {
            type: String,
            notify: true,
            reflectToAttribute: true
          },

          /**
           * Icon to use if image isn't there.
           */
          icon: {
            type: String,
            notify: true,
            reflectToAttribute: true
          },

          /**
           * Title of the gallery item
           */
          title: {
            type: String,
            value: "Project",
            notify: true
          },

          /**
           * Gallery creator
           */
          author: {
            type: Object,
            value: {
              name: "author",
              display_name: "Author"
            },
            notify: true
          },

          /**
           * Visual elevation of the item off the UI via paper element height
           */
          elevation: {
            type: Number,
            value: 1,
            reflectToAttribute: true,
            notify: true
          },

          /**
           * Number of comments this has
           */
          comments: {
            type: Number,
            value: 0,
            reflectToAttribute: true,
            notify: true
          }
        };
      }
    }]);
    return LrndesignGallerycard;
  }(_polymerElement.PolymerElement);

  _exports.LrndesignGallerycard = LrndesignGallerycard;
  window.customElements.define(LrndesignGallerycard.tag, LrndesignGallerycard);
});
