// @lrnwebcomponents/lrndesign-gallerycard/lrndesign-gallerycard.js
define(["exports", "require", "../../lit/index.js"], function (_exports, _require, _index) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrndesignGallerycard = void 0;
  _require = _interopRequireWildcard(_require);

  var _templateObject_1e52fa80942811ecb1937969798b231a, _templateObject2_1e52fa80942811ecb1937969798b231a, _templateObject3_1e52fa80942811ecb1937969798b231a, _templateObject4_1e52fa80942811ecb1937969798b231a;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `lrndesign-gallerycard`
   * @demo demo/index.html
   */
  var LrndesignGallerycard = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(LrndesignGallerycard, _LitElement);

    var _super = _createSuper(LrndesignGallerycard);

    function LrndesignGallerycard() {
      var _this;

      babelHelpers.classCallCheck(this, LrndesignGallerycard);
      _this = _super.call(this);
      _this.author = {
        name: "author",
        display_name: "Author"
      };
      _this.title = "Project";
      _this.elevation = 1;
      _this.comments = 0;
      setTimeout(function () {
        new Promise(function (res, rej) {
          return _require.default(["../lrndesign-avatar/lrndesign-avatar.js"], res, rej);
        });

        _this.addEventListener("mouseenter", _this._mouseEnter.bind(babelHelpers.assertThisInitialized(_this)));

        _this.addEventListener("mouseleave", _this._mouseLeave.bind(babelHelpers.assertThisInitialized(_this)));
      }, 0);
      return _this;
    }

    babelHelpers.createClass(LrndesignGallerycard, [{
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (["size", "image", "icon", "title", "author", "elevation", "comments"].includes(propName)) {
            _this2.dispatchEvent(new CustomEvent("".concat(propName, "-changed"), {
              value: _this2[propName]
            }));
          }

          if (propName == "icon" && _this2[propName] != "") {
            new Promise(function (res, rej) {
              return _require.default(["../simple-icon/simple-icon.js"], res, rej);
            });
            new Promise(function (res, rej) {
              return _require.default(["../simple-icon/lib/simple-icons.js"], res, rej);
            });
          }
        });
      }
    }, {
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_1e52fa80942811ecb1937969798b231a || (_templateObject_1e52fa80942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"card\">\n        <div class=\"card-content card-control-height card-control-center\">\n          <div class=\"submission-preview\">\n            ", "\n            ", "\n          </div>\n          <div class=\"submission-info\">\n            <div class=\"divider\"></div>\n            <div class=\"title\">", "</div>\n          </div>\n        </div>\n        <div class=\"card-actions\">\n          <lrndesign-avatar\n            id=\"avatar\"\n            label=\"", "\"\n            src=\"", "\"\n          ></lrndesign-avatar>\n          <div class=\"card-action-details\">\n            <div class=\"author\">", "</div>\n            <span class=\"comments text-right\">Comments: ", "</span>\n          </div>\n        </div>\n      </div>\n    "])), this.icon ? (0, _index.html)(_templateObject2_1e52fa80942811ecb1937969798b231a || (_templateObject2_1e52fa80942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                  <simple-icon\n                    class=\"project-icon\"\n                    icon=\"", "\"\n                  ></simple-icon>\n                "])), this.icon) : "", this.image ? (0, _index.html)(_templateObject3_1e52fa80942811ecb1937969798b231a || (_templateObject3_1e52fa80942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                  <img\n                    style=\"width:100%; height:100%; background-color: lightgray;\"\n                    loading=\"lazy\"\n                    src=\"", "\"\n                  />\n                "])), this.image) : "", this.title, this.author.name, this.author.avatar, this.author.display_name, this.comments);
      }
    }, {
      key: "_mouseEnter",
      value: function _mouseEnter(e) {
        this.elevation += 2;
      }
    }, {
      key: "_mouseLeave",
      value: function _mouseLeave(e) {
        this.elevation -= 2;
      }
    }], [{
      key: "styles",
      get: function get() {
        return [(0, _index.css)(_templateObject4_1e52fa80942811ecb1937969798b231a || (_templateObject4_1e52fa80942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: inline-flex;\n        }\n        :host([size=\"micro\"]) {\n          transform: scale(0.5);\n        }\n        :host([size=\"small\"]) {\n          transform: scale(0.8);\n        }\n\n        div.card {\n          border-radius: 4px;\n          margin: 0;\n          width: 100%;\n          box-shadow: 0 5px 5px rgba(0, 0, 0, 0.7);\n        }\n\n        .card-actions {\n          background-color: #f5f5f5;\n          border-radius: 0 0 4px 4px;\n          padding: 0 8px;\n        }\n        .card-actions .card-action-details {\n          display: inline-block;\n          vertical-align: middle;\n          vertical-align: -webkit-baseline-middle;\n          width: 80%;\n        }\n        #avatar {\n          display: inline-block;\n          vertical-align: text-top;\n          transform: scale(0.8);\n        }\n\n        .card-control-height {\n          height: 216px;\n        }\n\n        :host([elevation=\"0\"]) {\n          border: solid 1px #eeeeee;\n        }\n\n        :host([elevation=\"1\"]) {\n          box-shadow: 0 2px 2px rgba(0, 0, 0, 0.7);\n        }\n        :host([elevation=\"2\"]) {\n          box-shadow: 0 5px 5px rgba(0, 0, 0, 0.7);\n        }\n\n        .text-right {\n          text-align: right;\n        }\n\n        .text-left {\n          text-align: left;\n        }\n        .author {\n          text-align: left;\n          font-size: 12px;\n          line-height: 12px;\n          color: black;\n          display: block;\n          padding: 8px;\n        }\n\n        .title {\n          color: #222;\n          font-size: 12.8px;\n          font-weight: 600;\n          line-height: 20px;\n          padding: 0 12px;\n          overflow: hidden;\n          text-overflow: ellipsis;\n          margin-top: 8px;\n        }\n\n        .comments {\n          font-size: 12px;\n          float: right;\n          color: black;\n          line-height: 12px;\n        }\n\n        .divider {\n          height: 1px;\n          width: 100%;\n          background: #efefef;\n        }\n\n        .project-icon {\n          --simple-icon-height: 100%;\n          --simple-icon-width: 100%;\n          overflow: hidden;\n          color: grey;\n        }\n        .project-icon:hover,\n        .project-icon:focus {\n          color: black;\n        }\n\n        .center {\n          margin: auto;\n          width: 80%;\n          padding: 16px;\n        }\n\n        .link {\n          font-size: 16px;\n          line-height: 16px;\n        }\n\n        .submission-info {\n          width: 100%;\n        }\n        .submission-preview {\n          height: 160px;\n        }\n\n        .card-content {\n          padding: 0;\n          margin: 0;\n          overflow: hidden;\n        }\n\n        .inline {\n          display: inline;\n        }\n      "])))];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrndesign-gallerycard";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          size: {
            type: String,
            reflect: true
          },

          /**
           * Cover image src.
           */
          image: {
            type: String,
            reflect: true
          },

          /**
           * Icon to use if image isn't there.
           */
          icon: {
            type: String,
            reflect: true
          },

          /**
           * Title of the gallery item
           */
          title: {
            type: String
          },

          /**
           * Gallery creator
           */
          author: {
            type: Object
          },

          /**
           * Visual elevation of the item off the UI via paper element height
           */
          elevation: {
            type: Number,
            reflect: true
          },

          /**
           * Number of comments this has
           */
          comments: {
            type: Number,
            reflect: true
          }
        };
      }
    }]);
    return LrndesignGallerycard;
  }(_index.LitElement);

  _exports.LrndesignGallerycard = LrndesignGallerycard;
  window.customElements.define(LrndesignGallerycard.tag, LrndesignGallerycard);
});
