// @lrnwebcomponents/lrndesign-gallerycard/lrndesign-gallerycard.js
define(["exports", "require", "../../lit-element/lit-element.js"], function (_exports, _require, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrndesignGallerycard = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject4_9e34a130d27411eaa0a3df3cfaf5ad22() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: inline-flex;\n        }\n        :host([size=\"micro\"]) {\n          transform: scale(0.5);\n        }\n        :host([size=\"small\"]) {\n          transform: scale(0.8);\n        }\n\n        paper-card {\n          border-radius: 4px;\n          margin: 0;\n          width: 100%;\n        }\n\n        .card-actions {\n          background-color: #f5f5f5;\n          border-radius: 0 0 4px 4px;\n          padding: 0 8px;\n        }\n        .card-actions .card-action-details {\n          display: inline-block;\n          vertical-align: middle;\n          vertical-align: -webkit-baseline-middle;\n          width: 80%;\n        }\n        #avatar {\n          display: inline-block;\n          vertical-align: text-top;\n          transform: scale(0.8);\n        }\n\n        .card-control-height {\n          height: 240px;\n        }\n\n        [elevation=\"0\"] {\n          border: solid 1px #eeeeee;\n        }\n\n        .text-right {\n          text-align: right;\n        }\n\n        .text-left {\n          text-align: left;\n        }\n\n        .title {\n          color: #222;\n          font-size: 12.8px;\n          font-weight: 600;\n          line-height: 20px;\n          padding: 0 12px;\n          overflow: hidden;\n          text-overflow: ellipsis;\n          margin-top: 8px;\n        }\n\n        .comments {\n          font-size: 12px;\n          float: right;\n        }\n\n        .divider {\n          height: 1px;\n          width: 100%;\n          background: #efefef;\n        }\n\n        .project-icon {\n          --iron-icon-height: 100%;\n          --iron-icon-width: 100%;\n          overflow: hidden;\n          color: grey;\n        }\n        .project-icon:hover,\n        .project-icon:focus {\n          color: black;\n        }\n\n        .center {\n          margin: auto;\n          width: 80%;\n          padding: 16px;\n        }\n\n        .link {\n          font-size: 16px;\n          line-height: 16px;\n        }\n\n        .submission-info {\n          width: 100%;\n        }\n        .submission-preview {\n          height: 160px;\n        }\n\n        .card-content {\n          padding: 0;\n          margin: 0;\n          overflow: hidden;\n        }\n\n        .inline {\n          display: inline;\n        }\n      "]);

    _templateObject4_9e34a130d27411eaa0a3df3cfaf5ad22 = function _templateObject4_9e34a130d27411eaa0a3df3cfaf5ad22() {
      return data;
    };

    return data;
  }

  function _templateObject3_9e34a130d27411eaa0a3df3cfaf5ad22() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <iron-image\n                    style=\"width:100%; height:100%; background-color: lightgray;\"\n                    sizing=\"cover\"\n                    preload=\"\"\n                    fade=\"\"\n                    src=\"", "\"\n                  ></iron-image>\n                "]);

    _templateObject3_9e34a130d27411eaa0a3df3cfaf5ad22 = function _templateObject3_9e34a130d27411eaa0a3df3cfaf5ad22() {
      return data;
    };

    return data;
  }

  function _templateObject2_9e34a130d27411eaa0a3df3cfaf5ad22() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <iron-icon\n                    class=\"project-icon\"\n                    icon=\"", "\"\n                  ></iron-icon>\n                "]);

    _templateObject2_9e34a130d27411eaa0a3df3cfaf5ad22 = function _templateObject2_9e34a130d27411eaa0a3df3cfaf5ad22() {
      return data;
    };

    return data;
  }

  function _templateObject_9e34a130d27411eaa0a3df3cfaf5ad22() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <paper-card elevation=\"", "\">\n        <div class=\"card-content card-control-height card-control-center\">\n          <div class=\"submission-preview\">\n            ", "\n            ", "\n          </div>\n          <div class=\"submission-info\">\n            <div class=\"divider\"></div>\n            <div class=\"title\">", "</div>\n          </div>\n        </div>\n        <div class=\"card-actions\">\n          <lrndesign-avatar\n            id=\"avatar\"\n            label=\"", "\"\n            src=\"", "\"\n          ></lrndesign-avatar>\n          <div class=\"card-action-details\">\n            <span class=\"text-left author\">", "</span>\n            <span class=\"comments text-right\">Comments: ", "</span>\n          </div>\n        </div>\n      </paper-card>\n    "]);

    _templateObject_9e34a130d27411eaa0a3df3cfaf5ad22 = function _templateObject_9e34a130d27411eaa0a3df3cfaf5ad22() {
      return data;
    };

    return data;
  }

  /**
   * `lrndesign-gallerycard`
   * @demo demo/index.html
   */
  var LrndesignGallerycard =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrndesignGallerycard, _LitElement);

    function LrndesignGallerycard() {
      var _this;

      babelHelpers.classCallCheck(this, LrndesignGallerycard);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrndesignGallerycard).call(this));
      _this.author = {
        name: "author",
        display_name: "Author"
      };
      _this.title = "Project";
      _this.elevation = 1;
      _this.comments = 0;
      setTimeout(function () {
        new Promise(function (res, rej) {
          return _require.default(["../lrndesign-avatar/lrndesign-avatar.js"], res, rej);
        });
        new Promise(function (res, rej) {
          return _require.default(["../../@polymer/paper-card/paper-card.js"], res, rej);
        });

        _this.addEventListener("mouseenter", _this._mouseEnter.bind(babelHelpers.assertThisInitialized(_this)));

        _this.addEventListener("mouseleave", _this._mouseLeave.bind(babelHelpers.assertThisInitialized(_this)));
      }, 0);
      return _this;
    }

    babelHelpers.createClass(LrndesignGallerycard, [{
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (["size", "image", "icon", "title", "author", "elevation", "comments"].includes(propName)) {
            _this2.dispatchEvent(new CustomEvent("".concat(propName, "-changed"), {
              value: _this2[propName]
            }));
          }

          if (propName == "image" && _this2[propName] != "") {
            new Promise(function (res, rej) {
              return _require.default(["../../@polymer/iron-image/iron-image.js"], res, rej);
            });
          }

          if (propName == "icon" && _this2[propName] != "") {
            new Promise(function (res, rej) {
              return _require.default(["../../@polymer/iron-icon/iron-icon.js"], res, rej);
            });
            new Promise(function (res, rej) {
              return _require.default(["../../@polymer/iron-icons/iron-icons.js"], res, rej);
            });
          }
        });
      }
    }, {
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_9e34a130d27411eaa0a3df3cfaf5ad22(), this.elevation, this.icon ? (0, _litElement.html)(_templateObject2_9e34a130d27411eaa0a3df3cfaf5ad22(), this.icon) : "", this.image ? (0, _litElement.html)(_templateObject3_9e34a130d27411eaa0a3df3cfaf5ad22(), this.image) : "", this.title, this.author.name, this.author.avatar, this.author.display_name, this.comments);
      }
    }, {
      key: "_mouseEnter",
      value: function _mouseEnter(e) {
        this.elevation += 2;
      }
    }, {
      key: "_mouseLeave",
      value: function _mouseLeave(e) {
        this.elevation -= 2;
      }
    }], [{
      key: "styles",
      get: function get() {
        return [(0, _litElement.css)(_templateObject4_9e34a130d27411eaa0a3df3cfaf5ad22())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrndesign-gallerycard";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          size: {
            type: String,
            reflect: true
          },

          /**
           * Cover image src.
           */
          image: {
            type: String,
            reflect: true
          },

          /**
           * Icon to use if image isn't there.
           */
          icon: {
            type: String,
            reflect: true
          },

          /**
           * Title of the gallery item
           */
          title: {
            type: String
          },

          /**
           * Gallery creator
           */
          author: {
            type: Object
          },

          /**
           * Visual elevation of the item off the UI via paper element height
           */
          elevation: {
            type: Number,
            reflect: true
          },

          /**
           * Number of comments this has
           */
          comments: {
            type: Number,
            reflect: true
          }
        };
      }
    }]);
    return LrndesignGallerycard;
  }(_litElement.LitElement);

  _exports.LrndesignGallerycard = LrndesignGallerycard;
  window.customElements.define(LrndesignGallerycard.tag, LrndesignGallerycard);
});
