// @haxtheweb/app-hax/lib/v2/app-hax-search-results.js
import{SimpleColors as e}from"../../../simple-colors/simple-colors.js";import{html as t,css as r}from"../../../../lit/index.js";import{autorun as a,toJS as s}from"../../../../mobx/dist/mobx.esm.js";import{varGet as d}from"../../../utils/utils.js";import{store as o}from"./AppHaxStore.js";import"./app-hax-site-bar.js";import"./app-hax-site-details.js";import"../../../simple-tooltip/simple-tooltip.js";export class AppHaxSearchResults extends e{static get tag(){return"app-hax-search-results"}constructor(){super(),this.searchItems=[],this.displayItems=[],this.searchTerm="",this.dark=!1,this.isPointerDown=!1,this.startX=0,this.scrollLeftVal=0,this.isDragging=!1,this.currentIndex=1,this.totalItems=0,this.sortOption="az",a((()=>{this.searchTerm=s(o.searchTerm)})),a((()=>{this.dark=s(o.darkMode)})),a((()=>{const e=s(o.manifest);e&&e.items&&(this.searchItems=e.items,this.displayItems=[...this.searchItems],o.loadThemesData())}))}static get properties(){return{...super.properties,searchTerm:{type:String,reflect:!0},searchItems:{type:Array},displayItems:{type:Array},siteUrl:{type:String,attribute:"site-url"},isPointerDown:{type:Boolean},startX:{type:Number},scrollLeftVal:{type:Number},isDragging:{type:Boolean},currentIndex:{type:Number},totalItems:{type:Number},sortOption:{type:String,attribute:"sort-option"}}}static get styles(){return[super.styles,r`
        :host {
          display: block;
          width: 100%;
          overflow: visible;
        }

        .carousel-container {
          display: flex;
          align-items: center;
          justify-content: flex-start;
          gap: var(--ddd-spacing-2, 8px);
          width: 65vw;
          max-height: 280px;
          position: relative;
          border-radius: var(--ddd-radius-md, 8px);
          border: var(--ddd-border-xs, 1px solid)
            var(--ddd-theme-default-limestoneLight, #e4e5e7);
          box-shadow: var(--ddd-boxShadow-sm);
          padding: var(--ddd-spacing-2, 8px);
          transition: box-shadow 0.2s ease;
          overflow: visible;
        }
        :host([dark]) .carousel-container,
        body.dark-mode .carousel-container {
          border-color: var(--ddd-theme-default-slateGray, #666);
          color: var(--ddd-theme-default-white, white);
        }
        .carousel-container:hover {
          box-shadow: var(--ddd-boxShadow-md);
        }
        .pager-container {
          display: flex;
          flex-direction: column;
          justify-content: center;
          align-items: center;
          gap: var(--ddd-spacing-2, 8px);
          margin-right: var(--ddd-spacing-4, 16px);
          padding: var(--ddd-spacing-1, 4px);
          flex-shrink: 0;
          scrollbar-width: none;
          -ms-overflow-style: none;
        }

        .pager-container::-webkit-scrollbar {
          display: none;
        }
        .pager-dot {
          width: var(--ddd-spacing-3, 12px);
          height: var(--ddd-spacing-3, 12px);
          border-radius: var(--ddd-radius-circle, 50%);
          background: var(--ddd-theme-default-limestoneGray, #a2aaad);
          border: var(--ddd-border-xs, 1px solid) transparent;
          cursor: pointer;
          transition: all 0.2s ease;
          outline: none;
        }
        .pager-dot:hover,
        .pager-dot:focus {
          transform: scale(1.2);
          outline: var(--ddd-border-xs, 1px solid)
            var(--ddd-theme-default-keystoneYellow, #ffd100);
        }
        .pager-dot.active {
          background: var(--ddd-theme-default-nittanyNavy, #001e44);
          border-color: var(--ddd-theme-default-keystoneYellow, #ffd100);
          transform: scale(1.1);
        }
        :host([dark]) .pager-dot,
        body.dark-mode .pager-dot {
          background: var(--ddd-theme-default-slateGray, #666);
        }
        :host([dark]) .pager-dot.active,
        body.dark-mode .pager-dot.active {
          background: var(--ddd-theme-default-keystoneYellow, #ffd100);
          border-color: var(--ddd-theme-default-white, white);
        }
        .scroll-left,
        .scroll-right {
          background: var(--ddd-theme-default-nittanyNavy, #001e44);
          color: var(--ddd-theme-default-white, white);
          border: var(--ddd-border-sm, 2px solid)
            var(--ddd-theme-default-keystoneYellow, #ffd100);
          border-radius: var(--ddd-radius-sm, 4px);
          padding: var(--ddd-spacing-4, 16px);
          cursor: pointer;
          height: 240px;
          min-width: var(--ddd-spacing-12, 56px);
          display: flex;
          align-items: center;
          justify-content: center;
          font-size: var(--ddd-font-size-s, 20px);
          transition: all 0.2s ease;
          flex-shrink: 0;
          z-index: 10;
          position: relative;
        }
        .scroll-right {
          margin-right: 0;
        }

        :host([dark]) .scroll-left,
        :host([dark]) .scroll-right,
        body.dark-mode .scroll-left,
        body.dark-mode .scroll-right {
          background: var(--ddd-theme-default-keystoneYellow, #ffd100);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          border-color: var(--ddd-theme-default-white, white);
        }

        .scroll-left:hover:not(:disabled),
        .scroll-right:hover:not(:disabled) {
          background: var(--ddd-theme-default-keystoneYellow, #ffd100);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          transform: translateY(-2px);
          box-shadow: var(--ddd-boxShadow-md);
        }

        .scroll-left:disabled,
        .scroll-right:disabled {
          opacity: 0.3;
          cursor: not-allowed;
          background: var(--ddd-theme-default-limestoneGray, #a2aaad);
          color: var(--ddd-theme-default-slateGray, #666);
          border-color: var(--ddd-theme-default-limestoneGray, #a2aaad);
          transform: none;
          box-shadow: none;
        }

        :host([dark]) .scroll-left:hover:not(:disabled),
        :host([dark]) .scroll-right:hover:not(:disabled),
        body.dark-mode .scroll-left:hover:not(:disabled),
        body.dark-mode .scroll-right:hover:not(:disabled) {
          background: var(--ddd-theme-default-nittanyNavy, #001e44);
          color: var(--ddd-theme-default-white, white);
        }

        :host([dark]) .scroll-left:disabled,
        :host([dark]) .scroll-right:disabled,
        body.dark-mode .scroll-left:disabled,
        body.dark-mode .scroll-right:disabled {
          opacity: 0.3;
          cursor: not-allowed;
          background: var(--ddd-theme-default-coalyGray, #444);
          color: var(--ddd-theme-default-slateGray, #666);
          border-color: var(--ddd-theme-default-coalyGray, #444);
        }

        #results {
          display: flex;
          flex-wrap: nowrap;
          overflow-x: auto;
          overflow-y: hidden;
          scroll-snap-type: x mandatory;
          gap: var(--ddd-spacing-6, 24px);
          padding: var(--ddd-spacing-1, 4px) var(--ddd-spacing-2, 8px);
          flex: 1;
          min-width: 0;
          cursor: grab;
          user-select: none;
          /* Keep scrollbar visible for multiple interaction methods */
          scrollbar-width: thin;
          scrollbar-color: var(--ddd-theme-default-slateGray, #666)
            var(--ddd-theme-default-limestoneLight, #e4e5e7);
        }

        :host([dark]) #results,
        body.dark-mode #results {
          scrollbar-color: var(--ddd-theme-default-limestoneGray, #a2aaad)
            var(--ddd-theme-default-coalyGray, #444);
        }

        #results.dragging {
          cursor: grabbing;
          scroll-behavior: auto;
        }

        #results::-webkit-scrollbar {
          height: 8px;
        }

        #results::-webkit-scrollbar-track {
          background: var(--ddd-theme-default-limestoneLight, #e4e5e7);
          border-radius: var(--ddd-radius-xs, 2px);
        }

        #results::-webkit-scrollbar-thumb {
          background: var(--ddd-theme-default-slateGray, #666);
          border-radius: var(--ddd-radius-xs, 2px);
          transition: background 0.2s ease;
        }

        #results::-webkit-scrollbar-thumb:hover {
          background: var(--ddd-theme-default-nittanyNavy, #001e44);
        }

        :host([dark]) #results::-webkit-scrollbar-track,
        body.dark-mode #results::-webkit-scrollbar-track {
          background: var(--ddd-theme-default-coalyGray, #444);
        }

        :host([dark]) #results::-webkit-scrollbar-thumb,
        body.dark-mode #results::-webkit-scrollbar-thumb {
          background: var(--ddd-theme-default-limestoneGray, #a2aaad);
        }

        :host([dark]) #results::-webkit-scrollbar-thumb:hover,
        body.dark-mode #results::-webkit-scrollbar-thumb:hover {
          background: var(--ddd-theme-default-keystoneYellow, #ffd100);
        }

        li {
          flex: 0 0 auto;
          scroll-snap-align: center;
          width: 220px;
          min-width: 220px;
          height: 260px;
          display: flex;
          flex-direction: column;
          align-items: center;
          justify-content: center;
        }

        app-hax-site-bar {
          margin: 0 var(--ddd-spacing-3, 12px);
          width: 100%;
        }
        .description {
          max-height: 64px;
          overflow: hidden;
          max-width: 80%;
          text-overflow: ellipsis;
          word-break: break-word;
          font-family: var(--ddd-font-primary, sans-serif);
          font-size: var(--ddd-font-size-xs, 14px);
          line-height: var(--ddd-lh-150, 1.5);
        }

        @media (max-width: 1200px) {
          :host {
            min-width: calc(
              2 * 264px + var(--ddd-spacing-6, 24px) + 2 *
                var(--ddd-spacing-12, 56px)
            );
          }
        }

        @media (max-width: 800px) {
          :host {
            min-width: calc(264px + 2 * var(--ddd-spacing-12, 56px));
          }

          app-hax-site-bar {
            --main-banner-width: 60vw;
          }

          .description {
            max-height: 24px;
            font-size: var(--ddd-font-size-3xs, 8px);
            font-family: var(--ddd-font-primary, sans-serif);
          }

          .scroll-left,
          .scroll-right {
            min-width: var(--ddd-spacing-10, 40px);
            padding: var(--ddd-spacing-2, 8px);
            font-size: var(--ddd-font-size-s, 16px);
          }
        }
        @media (max-width: 640px) {
          app-hax-site-bar a {
            font-size: var(--ddd-font-size-xs, 14px);
          }
          app-hax-site-bar {
            --main-banner-width: 70vw;
          }

          li {
            width: 240px;
            min-width: 240px;
          }

          #results::after {
            min-width: 240px;
          }
          /* Mobile: Show only 1 item, hide arrows */
          .scroll-left,
          .scroll-right {
            display: none;
          }

          .carousel-container {
            padding: var(--ddd-spacing-2, 8px) var(--ddd-spacing-4, 16px);
            justify-content: center;
          }

          #results {
            /* Single item takes full width on mobile */
            gap: 0;
            scroll-snap-type: x mandatory;
            overflow-x: auto;
            scrollbar-width: none;
            -ms-overflow-style: none;
          }
        }
        span[slot="band"] {
          height: var(--ddd-spacing-12, 48px);
          overflow: hidden;
          text-overflow: ellipsis;
          margin-bottom: var(--ddd-spacing-2, 8px);
        }
        :host([dark]) #noResult,
        body.dark-mode #noResult {
          color: var(--ddd-theme-default-limestoneGray, #f5f5f5);
        }

        #noResult {
          font-family: var(--ddd-font-primary, sans-serif);
          font-size: var(--ddd-font-size-s, 16px);
          color: var(--ddd-theme-default-coalyGray, #444);
          text-align: center;
          padding: var(--ddd-spacing-6, 24px);
          margin: var(--ddd-spacing-4, 16px);
        }

        /* Screen reader only text */
        .sr-only {
          position: absolute;
          width: 1px;
          height: 1px;
          padding: 0;
          margin: -1px;
          overflow: hidden;
          clip: rect(0, 0, 0, 0);
          white-space: nowrap;
          border: 0;
        }
      `]}getVisibleDotRange(){if(this.totalItems<=10)return{start:1,end:this.totalItems};const e=Math.floor(5);let t=Math.max(1,this.currentIndex-e),r=Math.min(this.totalItems,t+10-1);return r===this.totalItems&&(t=Math.max(1,r-10+1)),{start:t,end:r}}getSortedItems(){const e=Array.isArray(this.displayItems)?[...this.displayItems]:[],t=this.sortOption||"az",getTitle=e=>(e.title||"").toString().toLowerCase(),getDate=(e,t)=>{const r=Number(d(e,`metadata.site.${t}`,d(e,"metadata.site.created",0)));return Number.isNaN(r)?0:r},getThemeName=e=>{const t=d(e,"metadata.theme.element","");return t&&o.themesData&&o.themesData[t]&&o.themesData[t].name?o.themesData[t].name.toString().toLowerCase():t.toString().toLowerCase()};return e.sort(((e,r)=>{switch(t){case"za":return getTitle(r).localeCompare(getTitle(e));case"newest":return getDate(r,"updated")-getDate(e,"updated");case"oldest":return getDate(e,"updated")-getDate(r,"updated");case"theme":return getThemeName(e).localeCompare(getThemeName(r));default:return getTitle(e).localeCompare(getTitle(r))}})),e}render(){const e=this.getSortedItems();this.totalItems=e.length;const r=this.getVisibleDotRange();return t`
      <div
        class="carousel-container"
        role="region"
        aria-label="Site results carousel"
        aria-live="polite"
      >
        <simple-tooltip for="scroll-left-btn" position="top"
          >Previous</simple-tooltip
        >
        <button
          id="scroll-left-btn"
          class="scroll-left"
          @click="${this.scrollLeft}"
          ?disabled="${this.currentIndex<=1||this.totalItems<=1}"
          aria-label="Previous sites"
          aria-describedby="scroll-left-desc"
        >
          ◀
          <span id="scroll-left-desc" class="sr-only"
            >View previous sites in the carousel</span
          >
        </button>
        <ul
          id="results"
          class="${this.isDragging?"dragging":""}"
          @mousedown="${this.handlePointerStart}"
          @touchstart="${this.handlePointerStart}"
          @mousemove="${this.handlePointerMove}"
          @touchmove="${this.handlePointerMove}"
          @mouseup="${this.handlePointerEnd}"
          @touchend="${this.handlePointerEnd}"
          @mouseleave="${this.handlePointerEnd}"
          @scroll="${this.handleScroll}"
>\n          ${e.length>0?e.map((e=>t` <li>
                    <app-hax-site-bar
                      ?dark="${this.dark}"
                      site-id="${e.id}"
                      .siteUrl="${e.slug}"
                      .slug="${e.slug}"
                      .image="${this.getThemeImage(e)}"
                      accent-color="${d(e,"metadata.theme.variables.cssVariable","").replace("--simple-colors-default-theme-","").replace("-7","")}"
                    >
                      <a
                        slot="heading"
                        href="${e.slug}"
                        target="_blank"
                        rel="noopener noreferrer"
                        title="${e.title}"
                      >
                        ${e.title}
                      </a>
                      <span slot="subHeading">${e.author}</span>
                      <app-hax-site-details
                        slot="band"
                        .details="${this.getItemDetails(e)}"
                        site-id="${e.id}"
                      >
                        <div class="description" slot="pre">
                          ${e.description}
                        </div>
                      </app-hax-site-details>
                    </app-hax-site-bar>
                  </li>`)):t`<div id="noResult">
                No
                results${""!==this.searchTerm?t`<strong>"${this.searchTerm}"</strong>`:", Create a new site!"}
              </div>`}
        </ul>
        <simple-tooltip for="scroll-right-btn" position="top"
          >Next</simple-tooltip
        >
        <button
          id="scroll-right-btn"
          class="scroll-right"
          @click="${this.scrollRight}"
          ?disabled="${this.currentIndex>=this.totalItems||this.totalItems<=1}"
          aria-label="Next sites"
          aria-describedby="scroll-right-desc"
        >
          ▶
          <span id="scroll-right-desc" class="sr-only"
            >View next sites in the carousel</span
          >
        </button>
        ${this.totalItems>1?t`
              <div class="pager-container">
                ${Array.from({length:r.end-r.start+1},((e,a)=>{const s=r.start+a;return t`
                      <button
                        class="pager-dot ${this.currentIndex===s?"active":""}"
                        @click="${()=>this.goToPage(s)}"
                        aria-label="Go to page ${s} of ${this.totalItems}"
                        aria-current="${this.currentIndex===s?"page":"false"}"
                        tabindex="0"
                      ></button>
                    `}))}
              </div>
            `:""}
      </div>
    `}scrollLeft(){if(this.currentIndex<=1||this.totalItems<=1)return;this.shadowRoot.querySelector("#results").scrollBy({left:-288,behavior:"smooth"})}scrollRight(){if(this.currentIndex>=this.totalItems||this.totalItems<=1)return;this.shadowRoot.querySelector("#results").scrollBy({left:288,behavior:"smooth"})}handlePointerStart(e){const t=this.shadowRoot.querySelector("#results");this.isPointerDown=!0,this.isDragging=!1,this.startX=("mousedown"===e.type?e.clientX:e.touches[0].clientX)-t.offsetLeft,this.scrollLeftVal=t.scrollLeft,e.preventDefault()}handlePointerMove(e){if(!this.isPointerDown)return;const t=this.shadowRoot.querySelector("#results"),r=2*(("mousemove"===e.type?e.clientX:e.touches[0].clientX)-t.offsetLeft-this.startX);Math.abs(r)>5&&(this.isDragging=!0,t.style.scrollBehavior="auto"),this.isDragging&&(t.scrollLeft=this.scrollLeftVal-r,e.preventDefault())}handlePointerEnd(e){this.isPointerDown=!1;const t=this.shadowRoot.querySelector("#results");t&&(t.style.scrollBehavior="smooth"),setTimeout((()=>{this.isDragging=!1}),100)}handleScroll(e){if(this.totalItems<=1)return;const t=e.target,r=t.scrollLeft,a=t.scrollWidth;if(r+t.clientWidth>=a-50)this.currentIndex=this.totalItems;else{const e=r/288;this.currentIndex=Math.min(Math.max(1,Math.floor(e)+1),this.totalItems)}}goToPage(e){const t=288*(e-1);this.shadowRoot.querySelector("#results").scrollTo({left:t,behavior:"smooth"}),this.currentIndex=e}getThemeImage(e){const t=d(e,"metadata.theme.element","");if(t&&o.themesData&&o.themesData[t]){let e=o.themesData[t].thumbnail||"";if(e&&e.startsWith("@haxtheweb/")){e=new URL("../../../../"+e,import.meta.url).href}return e}return""}getItemDetails(e){return{created:d(e,"metadata.site.created",new Date/1e3),updated:d(e,"metadata.site.updated",new Date/1e3),pages:d(e,"metadata.pageCount",0),url:e.slug}}openedChanged(e){o.appEl.playSound("click"),e.detail.value?this.shadowRoot.querySelector("app-hax-site-details").removeAttribute("tabindex"):this.shadowRoot.querySelector("app-hax-site-details").setAttribute("tabindex","-1")}}customElements.define(AppHaxSearchResults.tag,AppHaxSearchResults);
