// @haxtheweb/app-hax/lib/v2/app-hax-scroll-button.js
import{LitElement as e,html as t,css as o}from"../../../../lit/index.js";import"../../../simple-tooltip/simple-tooltip.js";export class AppHaxScrollButton extends e{static get tag(){return"app-hax-scroll-button"}constructor(){super(),this.label="",this.targetId="",this.isDarkMode=document.body.classList.contains("dark-mode"),this.addEventListener("keydown",this._handleKeydown.bind(this))}static get properties(){return{label:{type:String},targetId:{type:String},isDarkMode:{type:Boolean,reflect:!0}}}scrollToTarget(){if(this.targetId){let e=null,t=this.closest("app-hax")||document.querySelector("app-hax");if(t){let o=t.shadowRoot?t.shadowRoot.querySelector("app-hax-use-case-filter"):t.querySelector("app-hax-use-case-filter");o&&(e=o.shadowRoot?o.shadowRoot.getElementById(this.targetId):o.querySelector(`#${this.targetId}`))}e||(e=document.getElementById(this.targetId)),e?(e.scrollIntoView({behavior:"smooth",block:"start"}),this._announceNavigation()):console.warn(`Element with id '${this.targetId}' not found inside app-hax-use-case-filter.`)}}connectedCallback(){super.connectedCallback(),this._darkModeObserver=new MutationObserver((()=>{this.isDarkMode=document.body.classList.contains("dark-mode")})),this._darkModeObserver.observe(document.body,{attributes:!0,attributeFilter:["class"]})}disconnectedCallback(){super.disconnectedCallback(),this._darkModeObserver&&this._darkModeObserver.disconnect()}static get styles(){return[o`
        :host {
          display: flex;
          background-image: url("/elements/app-hax/lib/assets/images/scroll-button-LM.png");
          background-size: cover;
          background-position: center;
          width: 300px;
          height: 50px;
          color: var(--app-hax-accent-color, var(--accent-color));
          align-items: center;
          justify-content: center;
          text-align: center;
          margin: 8px;
        }
        :host([isDarkMode]) {
          background-image: url("/elements/app-hax/lib/assets/images/scroll-button-DM.png");
        }
        div {
          display: flex;
          align-items: center;
          cursor: pointer;
        }

        div:hover,
        div:focus {
          outline: 2px solid var(--ddd-theme-default-keystoneYellow, #ffd100);
          outline-offset: 2px;
        }

        /* Screen reader only text */
        .sr-only {
          position: absolute;
          width: 1px;
          height: 1px;
          padding: 0;
          margin: -1px;
          overflow: hidden;
          clip: rect(0, 0, 0, 0);
          white-space: nowrap;
          border: 0;
        }

        /* Live region for announcements */
        .live-region {
          position: absolute;
          width: 1px;
          height: 1px;
          margin: -1px;
          overflow: hidden;
          clip: rect(0, 0, 0, 0);
          white-space: nowrap;
        }
      `]}render(){return t`
      <div
        @click="${this.scrollToTarget}"
        tabindex="0"
        role="button"
        aria-label="${this.label} - Navigate to section"
        aria-describedby="scroll-desc"
      >
        <h5 aria-hidden="true">${this.label}</h5>
        <div id="scroll-desc" class="sr-only">
          Click to scroll to the ${this.label} section
        </div>
      </div>
      <div
        class="live-region"
        aria-live="polite"
        aria-atomic="true"
        id="scroll-announcement"
      ></div>
    `}_handleKeydown(e){"Enter"!==e.key&&" "!==e.key||(e.preventDefault(),this.scrollToTarget())}_announceNavigation(){const e=this.shadowRoot.querySelector("#scroll-announcement");e&&(e.textContent=`Navigated to ${this.label} section`,setTimeout((()=>{e.textContent=""}),1e3))}}customElements.define(AppHaxScrollButton.tag,AppHaxScrollButton);
