// @haxtheweb/app-hax/lib/v2/app-hax-simple-hat-progress.js
import{LitElement as t,html as e,css as s}from"../../../../lit/index.js";import{DDDSuper as r}from"../../../d-d-d/d-d-d.js";export class AppHaxSimpleHatProgress extends(r(t)){static get tag(){return"app-hax-simple-hat-progress"}constructor(){super(),this.progress=0,this.max=100}static get properties(){return{progress:{type:Number},max:{type:Number}}}static get styles(){return[super.styles,s`
        :host {
          display: block;
          width: 160px;
          height: 160px;
          position: relative;
        }

        .hat-container {
          position: relative;
          width: 100%;
          height: 100%;
        }

        .hat-image {
          width: 100%;
          height: 100%;
          pointer-events: none;
          z-index: 2;
          position: relative;
        }

        .progress-bar {
          position: absolute;
          bottom: 38px;
          left: 8px;
          right: 8px;
          height: 16px;
          background: var(--ddd-theme-default-limestoneGray, #f5f5f5);
          border-radius: var(--ddd-radius-sm, 4px);
          overflow: hidden;
          z-index: 1;
        }

        .progress-fill {
          height: 100%;
          background: linear-gradient(
            90deg,
            var(--ddd-theme-default-original87Pink, #e4007c) 0%,
            var(--ddd-theme-default-keystoneYellow, #ffd100) 50%,
            var(--ddd-theme-default-futureLime, #99cc33) 100%
          );
          border-radius: var(--ddd-radius-sm, 4px);
          transition: width 0.5s ease;
          width: 0%;
        }

        .progress-text {
          position: absolute;
          top: 32%;
          left: 52%;
          transform: translate(-50%, -50%);
          color: var(--ddd-theme-default-white, white);
          font-family: var(--ddd-font-primary, sans-serif);
          font-size: var(--ddd-font-size-s, 16px);
          font-weight: var(--ddd-font-weight-bold, 700);
          text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.5);
          z-index: 3;
        }
      `]}updated(t){if(t.has("progress")){const t=this.shadowRoot.querySelector(".progress-fill");if(t){const e=Math.min(100,Math.max(0,this.progress/this.max*100));t.style.width=`${e}%`}}}render(){const t=Math.min(100,Math.max(0,this.progress/this.max*100));return e`
      <div class="hat-container">
        <img
          class="hat-image"
          src="${new URL("../assets/images/HatBlank.svg",import.meta.url).href}"
          alt="Progress Hat"
        />
        <div class="progress-bar">
          <div class="progress-fill" style="width: ${t}%"></div>
        </div>
        <div class="progress-text">${Math.round(t)}%</div>
      </div>
    `}}customElements.define(AppHaxSimpleHatProgress.tag,AppHaxSimpleHatProgress);
