// @haxtheweb/app-hax/lib/v2/app-hax-search-bar.js
import{LitElement as e,html as t,css as i}from"../../../../lit/index.js";import"../../../simple-tooltip/simple-tooltip.js";import{store as o}from"./AppHaxStore.js";export class AppHaxSearchBar extends e{static get tag(){return"app-hax-search-bar"}constructor(){super(),this.searchTerm="",this.disabled=!1,this.showSearch=!1}static get properties(){return{searchTerm:{type:String},showSearch:{type:Boolean,reflect:!0,attribute:"show-search"},disabled:{type:Boolean,reflect:!0}}}updated(e){e.forEach(((e,t)=>{"searchItems"===t?this.displayItems=[...this.searchItems]:"searchTerm"===t?o.searchTerm=this.searchTerm:"showSearch"===t&&void 0!==e&&!1===this[t]&&(this.searchTerm="")}))}static get styles(){return[i`
        :host {
          overflow: hidden;
          position: relative;
          display: inline-block;
        }

        [role="search"] {
          position: relative;
          display: inline-block;
        }

        .search-input-container {
          position: relative;
          display: inline-block;
        }
        input {
          opacity: 1;
          width: 750px;
          transition: all ease-in-out 0.3s;
          padding: 4px;
          padding-left: 35px;
          font-family: sans-serif;
          font-size: 16px;
          margin: 2px 0 0 16px;
        }
        @media (max-width: 780px) {
          :host([show-search]) input {
            width: 250px;
            max-width: 20vw;
          }
        }
        @media (max-width: 600px) {
          :host([show-search]) input {
            width: 200px;
            max-width: 20vw;
          }
        }

        simple-toolbar-button[disabled] {
          background-color: #cccccc;
          pointer-events: none;
          cursor: help;
        }
        simple-toolbar-button {
          min-width: 48px;
          margin: 0;
          --simple-toolbar-border-color: #dddddddd;
          height: 48px;
          --simple-toolbar-button-disabled-opacity: 0.3;
          --simple-toolbar-button-padding: 3px 6px;
          --simple-toolbar-border-radius: 0;
        }
        simple-toolbar-button:hover,
        simple-toolbar-button:active,
        simple-toolbar-button:focus {
          background-color: var(--hax-ui-background-color-accent);
          color: var(--hax-ui-color);
        }
        simple-toolbar-button:hover,
        simple-toolbar-button:active,
        simple-toolbar-button:focus {
          --simple-toolbar-border-color: var(--hax-ui-color-accent);
        }
        .search-icon {
          position: absolute;
          left: 24px;
          top: 50%;
          transform: translateY(-50%);
          font-size: 16px;
          align-self: center;
        }
      `]}testKeydown(e){"Escape"!==e.key&&"Enter"!==e.key||this.toggleSearch()}search(){this.searchTerm=this.shadowRoot.querySelector("#searchField").value}render(){return t`
      <div role="search" aria-label="Site search">
        <div class="search-input-container">
          <simple-icon
            class="search-icon"
            icon="icons:search"
            aria-hidden="true"
            id="search-icon"
          ></simple-icon>
          <input
            id="searchField"
            @click="${this.toggleSearch}"
            @input="${this.search}"
            @keydown="${this.testKeydown}"
            type="text"
            placeholder="Search Existing Sites"
            aria-label="Search existing sites"
            aria-describedby="search-icon"
            role="searchbox"
          />
        </div>
        <slot></slot>
      </div>
    `}toggleSearch(){this.disabled||(this.shadowRoot.querySelector("#searchField").value="")}}customElements.define(AppHaxSearchBar.tag,AppHaxSearchBar);
