// @haxtheweb/app-hax/lib/v2/app-hax-site-login.js
import{LitElement as e,html as t,css as i}from"../../../../lit/index.js";import"../../../simple-icon/lib/simple-icons.js";import"../../../simple-icon/lib/simple-icon-lite.js";import{DDDSuper as a}from"../../../d-d-d/d-d-d.js";import"../../../rpg-character/rpg-character.js";import{store as s}from"./AppHaxStore.js";export class AppHaxSiteLogin extends(a(e)){static get tag(){return"app-hax-site-login"}constructor(){super(),this.windowControllers=new AbortController,this.username="",this.password="",this.errorMSG="Enter User name",this.hidePassword=!0,this.hasPass=!1}static get properties(){return{...super.properties,username:{type:String},password:{type:String},errorMSG:{type:String},hidePassword:{type:Boolean},hasPass:{type:Boolean}}}firstUpdated(){super.firstUpdated(),setTimeout((()=>{this.shadowRoot.querySelector("input").focus()}),0)}static get styles(){return[super.styles,i`
        :host {
          display: flex;
          flex-direction: column;
          justify-content: center;
          align-items: center;
          padding: var(--ddd-spacing-6, 24px);
          text-align: center;
          font-family: var(--ddd-font-primary, sans-serif);
          background: var(--ddd-theme-default-white, white);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
        }

        rpg-character {
          display: block;
          margin: 0 0 var(--ddd-spacing-4, 16px) 0;
          width: 120px;
          height: 120px;
        }

        #errorText {
          color: var(--ddd-theme-default-original87Pink, #e4007c);
          font-size: var(--ddd-font-size-xs, 14px);
          margin: var(--ddd-spacing-2, 8px) 0;
          min-height: var(--ddd-spacing-5, 20px);
          font-weight: var(--ddd-font-weight-medium, 500);
        }

        #inputcontainer {
          display: flex;
          flex-direction: column;
          justify-content: center;
          align-items: center;
          gap: var(--ddd-spacing-4, 16px);
        }

        .form-group {
          width: 100%;
          position: relative;
        }

        input {
          width: 100%;
          padding: var(--ddd-spacing-3, 12px);
          border: var(--ddd-border-sm, 2px solid)
            var(--ddd-theme-default-slateGray, #666);
          border-radius: var(--ddd-radius-sm, 4px);
          font-size: var(--ddd-font-size-s, 16px);
          font-family: var(--ddd-font-primary, sans-serif);
          box-sizing: border-box;
          transition: border-color 0.2s ease;
          background-color: var(--ddd-theme-default-limestoneMaxLight, #f5f5f5);
          color: var(--ddd-theme-default-coalyGray, #444);
        }

        input:focus {
          outline: none;
          border-color: var(--ddd-theme-default-nittanyNavy, #001e44);
        }

        input::placeholder {
          color: var(--ddd-theme-default-slateGray, #666);
          text-transform: capitalize;
        }

        button {
          padding: var(--ddd-spacing-3, 12px) var(--ddd-spacing-4, 16px);
          border-radius: var(--ddd-radius-sm, 4px);
          font-size: var(--ddd-font-size-s, 16px);
          font-weight: var(--ddd-font-weight-medium, 500);
          font-family: var(--ddd-font-primary, sans-serif);
          cursor: pointer;
          transition: all 0.2s ease;
          border: none;
          display: flex;
          align-items: center;
          justify-content: center;
          gap: var(--ddd-spacing-2, 8px);
          width: 100%;
          min-height: var(--ddd-spacing-10, 40px);
          background: var(--ddd-theme-default-nittanyNavy, #001e44);
          color: var(--ddd-theme-default-white, white);
        }

        button:disabled {
          opacity: 0.5;
          cursor: not-allowed;
        }

        button:hover:not(:disabled) {
          background: var(--ddd-theme-default-keystoneYellow, #ffd100);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          transform: translateY(-1px);
          box-shadow: var(--ddd-boxShadow-md);
        }

        .notyou {
          padding: var(--ddd-spacing-2, 8px);
          font-size: var(--ddd-font-size-s, 16px);
          color: var(--ddd-theme-default-coalyGray, #444);
        }

        .notyou a {
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          text-decoration: underline;
          cursor: pointer;
          font-weight: var(--ddd-font-weight-medium, 500);
        }

        .notyou a:hover {
          color: var(--ddd-theme-default-keystoneYellow, #ffd100);
        }

        .visibility-icon {
          position: absolute;
          right: var(--ddd-spacing-3, 12px);
          top: 50%;
          transform: translateY(-50%);
          background: transparent;
          border: none;
          color: var(--ddd-theme-default-slateGray, #666);
          cursor: pointer;
          padding: var(--ddd-spacing-1, 4px);
          border-radius: var(--ddd-radius-xs, 2px);
          transition: color 0.2s ease;
          --simple-icon-width: var(--ddd-icon-xs, 16px);
          --simple-icon-height: var(--ddd-icon-xs, 16px);
        }

        .visibility-icon:hover {
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
        }

        .external {
          text-align: center;
          width: 100%;
          margin-top: var(--ddd-spacing-4, 16px);
        }

        @media (max-width: 600px) {
          :host {
            padding: var(--ddd-spacing-4, 16px);
          }

          rpg-character {
            width: 80px;
            height: 80px;
          }

          button {
            font-size: var(--ddd-font-size-xs, 14px);
            padding: var(--ddd-spacing-2, 8px) var(--ddd-spacing-3, 12px);
          }
        }
      `]}checkUsername(){const e=this.shadowRoot.querySelector("#username").value;this.hidePassword=!1,this.errorMSG="",this.username=e,s.appEl.playSound("click2"),setTimeout((()=>{this.shadowRoot.querySelector("input").focus()}),0)}async checkPassword(){s.appEl.playSound("click2");const e=this.shadowRoot.querySelector("#password").value;globalThis.dispatchEvent(new CustomEvent("jwt-login-login",{composed:!0,bubbles:!0,cancelable:!1,detail:{username:this.username,password:e}}))}reset(){this.errorMSG="",this.username="",this.hasPass=!1,this.hidePassword=!0}nameChange(){this.username=this.shadowRoot.querySelector("#username").value}connectedCallback(){super.connectedCallback(),globalThis.addEventListener("jwt-logged-in",this._jwtLoggedIn.bind(this),{signal:this.windowControllers.signal}),globalThis.addEventListener("jwt-login-login-failed",this._jwtLoginFailed.bind(this),{signal:this.windowControllers.signal})}disconnectedCallback(){this.windowControllers.abort(),super.disconnectedCallback()}_jwtLoginFailed(e){this.hidePassword=!0,this.errorMSG="Invalid Username or Password",s.appEl.playSound("error")}_jwtLoggedIn(e){e.detail&&(s.user={name:this.username},s.appEl.playSound("success"),this.dispatchEvent(new CustomEvent("simple-modal-hide",{bubbles:!0,composed:!0,cancelable:!0,detail:{}})),s.toast(`Welcome ${this.username}! Let's go!`,5e3,{hat:"construction"}),s.appEl.reset())}passChange(e){const t=this.shadowRoot.querySelector("#password").value;this.hasPass=!!t}toggleViewPass(e){const t=this.shadowRoot.querySelector("#password"),i="password"===t.getAttribute("type")?"text":"password";t.setAttribute("type",i),e.target.icon="text"===i?"lrn:visible":"lrn:view-off"}render(){return t`
      <rpg-character seed="${this.username}"></rpg-character>
      <p id="errorText">${this.errorMSG}</p>
      <div id="inputcontainer">
        ${this.hidePassword?t` <div class="form-group">
                <input
                  id="username"
                  type="text"
                  placeholder="Enter your username"
                  aria-label="Username"
                  @input="${this.nameChange}"
                />
              </div>
              <button
                ?disabled="${!this.username}"
                @click=${this.checkUsername}
              >
                <simple-icon-lite icon="icons:arrow-forward"></simple-icon-lite>
                Next
              </button>`:t` <div class="notyou">
                Welcome back, ${this.username}!
                <a @click=${this.reset}>Not you?</a>
              </div>
              <div class="form-group">
                <input
                  id="password"
                  type="password"
                  placeholder="Enter your password"
                  aria-label="Password"
                  @input="${this.passChange}"
                />
                <simple-icon-button-lite
                  icon="lrn:view-off"
                  tabindex="-1"
                  title="Toggle password display"
                  @click="${this.toggleViewPass}"
                  class="visibility-icon"
                ></simple-icon-button-lite>
              </div>
              <button ?disabled="${!this.hasPass}" @click=${this.checkPassword}>
                <simple-icon-lite icon="icons:login"></simple-icon-lite>
                Login
              </button>`}
        <div class="external">
          <slot name="externalproviders"></slot>
        </div>
      </div>
    `}}customElements.define(AppHaxSiteLogin.tag,AppHaxSiteLogin);
