// @haxtheweb/app-hax/lib/v2/app-hax-wired-toggle.js
import{autorun as e,toJS as t}from"../../../../mobx/dist/mobx.esm.js";import{html as d,css as r}from"../../../../lit/index.js";import{store as a}from"./AppHaxStore.js";import{WiredDarkmodeToggle as i}from"../../../haxcms-elements/lib/core/ui/wired-darkmode-toggle/wired-darkmode-toggle.js";import{SimpleTourFinder as o}from"../../../simple-popover/lib/SimpleTourFinder.js";export class AppHAXWiredToggle extends(o(i)){constructor(){super(),this.tourName="hax",this.darkModeMediaQuery=globalThis.matchMedia("(prefers-color-scheme: dark)"),this._updateToggleState=()=>{this.checked=t(a.darkMode),this.darkModeMediaQuery.matches?this.disabled=!0:this.disabled=!1},e(this._updateToggleState),this.darkModeMediaQuery.addEventListener("change",this._updateToggleState)}static get tag(){return"app-hax-wired-toggle"}disconnectedCallback(){super.disconnectedCallback(),this.darkModeMediaQuery&&this._updateToggleState&&this.darkModeMediaQuery.removeEventListener("change",this._updateToggleState)}updated(e){super.updated&&super.updated(e),e.forEach(((e,t)=>{"checked"===t&&void 0!==e&&(a.darkMode=this[t])}))}static get styles(){return[super.styles,r`
        /* Screen reader only text */
        .sr-only {
          position: absolute;
          width: 1px;
          height: 1px;
          padding: 0;
          margin: -1px;
          overflow: hidden;
          clip: rect(0, 0, 0, 0);
          white-space: nowrap;
          border: 0;
        }
      `]}render(){return d`
      <div
        data-simple-tour-stop
        data-stop-title="data-label"
        data-label="${this.label}"
        aria-describedby="dark-mode-desc"
      >
        ${super.render()}
        <div id="dark-mode-desc" class="sr-only">
          Toggle between light and dark mode themes
        </div>
        <div data-stop-content style="display:none;">
          You can toggle your user interface between "light" and "dark" for your
          viewing enjoyment.
        </div>
      </div>
    `}}customElements.define(AppHAXWiredToggle.tag,AppHAXWiredToggle);
