// @haxtheweb/glossy-portfolio-theme/lib/glossy-portfolio-footer.js
/**
 * Copyright 2025 NazmanRosman
 * @license Apache-2.0, see LICENSE for full text.
 */
import{LitElement as e,html as t,css as i}from"../../../lit/index.js";import{DDDSuper as s}from"../../d-d-d/d-d-d.js";import{I18NMixin as r}from"../../i18n-manager/lib/I18NMixin.js";import{store as n}from"../../haxcms-elements/lib/core/haxcms-site-store.js";import{autorun as a,toJS as o}from"../../../mobx/dist/mobx.esm.js";import{licenseList as l}from"../../license-element/license-element.js";export class GlossyPortfolioFooter extends(s(r(e))){static get tag(){return"glossy-portfolio-footer"}constructor(){super(),this.title="Title",this.homeLink="",this.__disposer=this.__disposer||[],a((e=>{this.manifest=o(n.manifest);let t=new l;this.manifest.license&&t[this.manifest.license]&&(this.licenseName=t[this.manifest.license].name,this.licenseLink=t[this.manifest.license].link,this.licenseImage=t[this.manifest.license].image),this.__disposer.push(e)})),a((e=>{this.lastUpdated=new Date(1e3*n.manifest.metadata.site.updated).toDateString(),this.copyrightYear=new Date(1e3*n.manifest.metadata.site.created).getFullYear(),this.__disposer.push(e)}))}static get properties(){return{...super.properties,title:{type:String},thumbnail:{type:String},link:{type:String},topItems:{type:Array}}}static get styles(){return[super.styles,i`
      *{
        box-sizing: border-box;
      }
      .background{
        background-color: #161616;
        display: flex;
        flex-direction: column;
        justify-content: center;
        align-items: center;
      }
      .wrapper {
        height: var(--footer-height, 76px); /* Default height if not set */
        padding: var(--page-padding);
        padding-top: 20px;
        padding-bottom: 20px;
        display: grid;
        grid-template-columns: 1fr auto 1fr; /* Three columns: left, center, right */
        align-items: center;
        width: 100%;
        max-width: var(--max-width);

        
      }
      p {
        margin: 0;
        font-family: Inter;
        color: #ffffff97;
        font-size: 0.8rem;
        text-align: center; /* Center-align text inside the grid cell */
        font-weight: 300;
      }
      .item {
        justify-self: start; /* Align left-side text to the start */
        text-align: left; /* Ensure text is left-aligned */

      }
      .center {
        justify-self: center; /* Center-align the copyright text */
      }
      .license {
        justify-self: end; /* Align the license image to the end */
      }


    `]}render(){return t`
<div class="background">
  <div class="wrapper">
      <p class="item"> Last updated: ${this.lastUpdated} </p>
      <p class="center">© ${this.copyrightYear}   </p>
    <div class="license item"> 
      <a href="${this.licenseLink}" target="_blank">
        <img src="${this.licenseImage}" alt="${this.licenseName}" />
      </a>
    </div>
  </div>
</div>


  `}static get haxProperties(){return new URL(`./lib/${this.tag}.haxProperties.json`,import.meta.url).href}}globalThis.customElements.define(GlossyPortfolioFooter.tag,GlossyPortfolioFooter);
