// @haxtheweb/glossy-portfolio-theme/lib/glossy-portfolio-breadcrumb.js
/**
 * Copyright 2025 NazmanRosman
 * @license Apache-2.0, see LICENSE for full text.
 */
import{LitElement as e,html as t,css as i}from"../../../lit/index.js";import{DDDSuper as r}from"../../d-d-d/d-d-d.js";import{I18NMixin as s}from"../../i18n-manager/lib/I18NMixin.js";import{SiteBreadcrumb as o}from"../../haxcms-elements/lib/ui-components/navigation/site-breadcrumb.js";import{store as l}from"../../haxcms-elements/lib/core/haxcms-site-store.js";import{autorun as a,toJS as m}from"../../../mobx/dist/mobx.esm.js";export class GlossyPortfolioBreadcrumb extends(r(s(e))){static get tag(){return"glossy-portfolio-breadcrumb"}static get styles(){return[super.styles,i`
        :host {
          display: block;
          box-sizing: border-box;
        }

        :host{
          --breadcrumb-font-size: 0.9rem;
        }

        ol.breadcrumb{
          list-style: none;
          margin: 0;
          padding: 0;
          display: flex;
          flex-direction: row;
          flex-wrap: wrap;
          font-family: inherit;
          font-size: var(--breadcrumb-font-size);
          gap: 8px;
          list-style: ">";
          list-style-position: inside; /* Place bullets inside the container */

        }


        ol.breadcrumb li {
          font-size: var(--breadcrumb-font-size);
          font-weight: 300;
        }
        ol.breadcrumb li a {
          margin-left: 4px;
          font-size: var(--breadcrumb-font-size);
          color: #ffffffa6;
          font-weight: 400;
        }
        ol.breadcrumb li a:hover {
          color: var(--link-color);

        }
        ol.breadcrumb li::marker {
          font-size: var(--breadcrumb-font-size);
          color: #ffffffa6;
        }
        ol.breadcrumb li:last-child a{
          color: var(--text-color);
        }
        ol.breadcrumb li:last-child a:hover{
          text-decoration: none;
          cursor: default
        }
        ol.breadcrumb li:first-child::marker {
          content: "";
          
        }
        ol.breadcrumb li:first-child a {
          margin-left: 0;
      
        }


      `]}constructor(){super(),this.__disposer=[],this.items=[],this.includeHome=!1,this.t=this.t||{},this.t={...this.t,home:"Home"}}render(){return t` ${this.items.length>0?t`
          <ol
            class="breadcrumb"
            itemscope
            itemtype="https://schema.org/BreadcrumbList"
          >
            ${this.items.map((e=>t`<li
                  itemprop="itemListElement"
                  itemscope
                  itemtype="https://schema.org/ListItem"
                >
                  <a itemprop="item" href="${e.slug}">${e.title}</a>
                </li>`))}
          </ol>
        `:""}`}static get properties(){return{...super.properties,items:{type:Array},editMode:{type:Boolean,reflect:!0,attribute:"edit-mode"},includeHome:{type:Boolean,reflect:!0,attribute:"include-home"}}}firstUpdated(e){super.firstUpdated&&super.firstUpdated(e),a((e=>{this.editMode=m(l.editMode),this.__disposer.push(e)})),a((e=>{this._activeItemChanged(m(l.activeItem)),this.__disposer.push(e)}))}_activeItemChanged(e){const t=e;if(t&&this.shadowRoot){var i=[{title:t.title,slug:t.slug}];let e=t,r=m(l.routerManifest);for(;e&&null!=e.parent;)e=r.items.find((t=>t.id==e.parent)),e&&i.unshift({title:e.title,slug:e.slug});this.includeHome&&i.unshift({title:this.t.home,slug:l.homeLink}),this.includeHome||1!==i.length?this.includeHome&&2===i.length&&i[0].slug===i[1].slug?this.items=[]:this.items=[...i]:this.items=[]}}disconnectedCallback(){for(var e in this.__disposer)this.__disposer[e].dispose();super.disconnectedCallback()}static get haxProperties(){return new URL(`./lib/${this.tag}.haxProperties.json`,import.meta.url).href}}globalThis.customElements.define(GlossyPortfolioBreadcrumb.tag,GlossyPortfolioBreadcrumb);
