// lit-html/development/private-ssr-support.js
define(["exports", "./lit-html.js"], function (_exports, _litHtml) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports._Σ = void 0;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * END USERS SHOULD NOT RELY ON THIS OBJECT.
   *
   * We currently do not make a mangled rollup build of the lit-ssr code. In order
   * to keep a number of (otherwise private) top-level exports mangled in the
   * client side code, we export a _Σ object containing those members (or
   * helper methods for accessing private fields of those members), and then
   * re-export them for use in lit-ssr. This keeps lit-ssr agnostic to whether the
   * client-side code is being used in `dev` mode or `prod` mode.
   * @private
   */
  var _Σ = {
    boundAttributeSuffix: _litHtml._Σ._boundAttributeSuffix,
    marker: _litHtml._Σ._marker,
    markerMatch: _litHtml._Σ._markerMatch,
    HTML_RESULT: _litHtml._Σ._HTML_RESULT,
    getTemplateHtml: _litHtml._Σ._getTemplateHtml,
    overrideDirectiveResolve: function overrideDirectiveResolve(directiveClass, resolveOverrideFn) {
      return /*#__PURE__*/function (_directiveClass) {
        babelHelpers.inherits(_class, _directiveClass);

        var _super = _createSuper(_class);

        function _class() {
          babelHelpers.classCallCheck(this, _class);
          return _super.apply(this, arguments);
        }

        babelHelpers.createClass(_class, [{
          key: "_$resolve",
          value: function _$resolve(_part, values) {
            return resolveOverrideFn(this, values);
          }
        }]);
        return _class;
      }(directiveClass);
    },
    getAttributePartCommittedValue: function getAttributePartCommittedValue(part, value, index) {
      // Use the part setter to resolve directives/concatenate multiple parts
      // into a final value (captured by passing in a commitValue override)
      var committedValue = _litHtml.noChange; // Note that _commitValue need not be in `stableProperties` because this
      // method is only run on `AttributePart`s created by lit-ssr using the same
      // version of the library as this file

      part._commitValue = function (value) {
        return committedValue = value;
      };

      part._$setValue(value, part, index);

      return committedValue;
    },
    resolveDirective: _litHtml._Σ._resolveDirective,
    AttributePart: _litHtml._Σ._AttributePart,
    PropertyPart: _litHtml._Σ._PropertyPart,
    BooleanAttributePart: _litHtml._Σ._BooleanAttributePart,
    EventPart: _litHtml._Σ._EventPart,
    ElementPart: _litHtml._Σ._ElementPart
  };
  _exports._Σ = _Σ;
});
