// @haxtheweb/super-daemon/lib/super-daemon-ui.js
import{html as e,css as t,nothing as s}from"../../../lit/index.js";import{SimpleFilterMixin as i}from"../../simple-filter/simple-filter.js";import{I18NMixin as o}from"../../i18n-manager/lib/I18NMixin.js";import{SimpleColors as r}from"../../simple-colors/simple-colors.js";import{UserScaffoldInstance as a}from"../../user-scaffold/user-scaffold.js";import{autorun as n,toJS as l}from"../../../mobx/dist/mobx.esm.js";import"../../../@lit-labs/virtualizer/lit-virtualizer.js";import"./super-daemon-row.js";import"./super-daemon-search.js";export class SuperDaemonUI extends(i(o(r))){constructor(){super(),this.focused=!1,this.like="",this.voiceSearch=!1,this.iconAccent="purple",this.multiMatch=!0,this.t=this.t||{},this.t={...this.t,noResultsForThisTerm:this._defaultTextEmpty,voiceSearch:"Voice search",results:"Results",loadingResults:"Loading results",insertBlocks:"Insert blocks",findMedia:"Find media",submitIdeas:"Submit your ideas",dropFilesHere:"Drop files here",typeWhatYouWant:"Type what you want to do",opensMemoryPalace:"opens Merlin",clickToDoAnything:"Click to do anything!"},this.registerLocalization({context:this,namespace:"super-daemon",basePath:import.meta.url+"/../../"}),this.opened=!1,this.items=[],this.mini=!1,this.wand=!1,this.loading=!0,this.listeningForInput=!1,this.programSearch="",this.commandContext="*",this.programName=null,this.programPlaceholder=null,this.shadowRootOptions={...r.shadowRootOptions,delegatesFocus:!0},this.activeDrag=!1,this.activeType=null,this.where="index",this.icon="hardware:keyboard-return",this._selectedIndex=-1,this._activeDescendant="",n((()=>{const e=l(a.action),t=l(a.data),s=globalThis.SuperDaemonManager.requestAvailability();a.active&&a.memory.isLoggedIn&&null===s.programName&&"drag"===e.type?(this.activeDrag=!0,this.activeType=t.value||t.architype):a.active&&a.memory.isLoggedIn&&null===s.programName&&"dragleave"===e.type&&(this.activeDrag=!1,this.activeType=null)}))}static get tag(){return"super-daemon-ui"}static get properties(){return{...super.properties,icon:{type:String},iconAccent:{type:String,attribute:"icon-accent"},voiceSearch:{type:Boolean,reflect:!0,attribute:"voice-search"},listeningForInput:{type:Boolean,reflect:!0,attribute:"listening-for-input"},activeDrag:{type:Boolean},activeType:{type:String},mini:{type:Boolean,reflect:!0},wand:{type:Boolean,reflect:!0},loading:{type:Boolean,reflect:!0},programSearch:{type:String,attribute:"program-search"},programName:{type:String,attribute:"program-name"},programPlaceholder:{type:String,attribute:"program-placeholder"},commandContext:{type:String,attribute:"command-context"},opened:{type:Boolean,reflect:!0},focused:{type:Boolean,reflect:!0},_selectedIndex:{type:Number,state:!0},_activeDescendant:{type:String,state:!0}}}static get styles(){let e=[];return super.styles&&(e=super.styles),[e,t`
        :host {
          display: block;
        }
        super-daemon-search {
          display: flex;
          margin: var(--ddd-spacing-4);
        }
        :host([wand]) super-daemon-search {
          margin: calc(-1 * var(--ddd-spacing-6)) 0 0 0;
          height: var(--ddd-spacing-12);
        }
        .voice {
          --simple-icon-height: 50px;
          --simple-icon-width: 100px;
          --simple-icon-button-border-radius: 0;
          color: var(--simple-colors-default-theme-grey-10, grey);
          transition: color 0.6s ease-in-out;
        }
        :host([mini]) .voice {
          --simple-icon-height: var(--ddd-icon-xxs);
          --simple-icon-width: var(--ddd-icon-xxs);
        }
        .voice:hover,
        .voice:focus {
          color: var(--simple-colors-default-theme-purple-6, purple);
        }
        .voice.listening {
          color: var(--simple-colors-default-theme-purple-4, purple);
        }

        .search .user-context-icon {
          display: inline-flex;
          --simple-icon-height: 50px;
          --simple-icon-width: 30px;
        }
        :host([mini]) .search .user-context-icon {
          --simple-icon-height: var(--ddd-icon-xxs);
          --simple-icon-width: var(--ddd-icon-xxs);
          margin-top: 0;
        }
        .loading {
          font-size: 12px;
          font-style: italic;
          margin: var(--ddd-spacing-4);
        }
        .results-stats {
          font-size: 12px;
          color: var(--simple-colors-default-theme-grey-10, black);
          padding: var(--ddd-spacing-2);
          float: right;
        }
        :host([focused][wand]) .results {
          display: block;
        }
        .results {
          width: 100%;
          padding: var(--ddd-spacing-4) 0;
        }
        :host([mini]) .results lit-virtualizer {
          max-height: unset;
          height: unset;
        }
        .results lit-virtualizer {
          max-height: 50vh;
          min-height: 264px !important;
          width: 100%;
          display: block;
          height: 50vh;
          border: var(--ddd-border-sm) solid
            var(--simple-colors-default-theme-grey-10, black);
        }
        .results super-daemon-row {
          scroll-snap-align: start;
          scroll-snap-stop: always;
          width: -webkit-fill-available;
        }
        .no-results {
          font-size: var(--ddd-font-size-3xs);
          font-weight: bold;
          word-break: break-all;
          overflow: hidden;
          line-height: var(--ddd-spacing-8);
          margin: var(--ddd-spacing-8);
          border: var(--ddd-border-xs) solid transparent;
          box-shadow: none;
          outline: 0;
        }
        .slotted {
          display: block;
          font-size: 12px;
          line-height: 18px;
        }
        .slotted ::slotted(a) {
          color: var(--simple-colors-default-theme-grey-8, blue);
          font-weight: bold;
          text-decoration: underline;
          cursor: pointer;
        }
        :host([mini]) .no-results {
          margin: var(--ddd-spacing-4);
        }

        @media screen and (max-width: 800px) {
          .voice {
            --simple-icon-height: 30px;
            --simple-icon-width: 30px;
          }
          super-daemon-search {
            margin: var(--ddd-spacing-2);
          }
          .results-stats {
            display: none;
          }
          .results {
            padding: 0px;
          }
          super-daemon-row {
            --super-daemon-row-icon: 30px;
            margin: var(--ddd-spacing-1);
          }

          super-daemon-row::part(label-wrap) {
            min-width: 70%;
          }
          super-daemon-row::part(button) {
            padding: var(--ddd-spacing-1);
          }
          super-daemon-row::part(action) {
            max-width: unset;
          }
          super-daemon-row::part(tags) {
            display: none;
          }
          super-daemon-row::part(path) {
            font-size: 10px;
          }
        }
        :host([mini]) {
          color: var(--simple-colors-default-theme-grey-12, black);
          background-color: var(--simple-colors-default-theme-grey-1, white);
        }
        :host([mini]) super-daemon-row {
          --super-daemon-row-icon: var(--ddd-icon-xxs);
          border-radius: 0;
          margin: var(--ddd-spacing-1);
        }
        :host([mini]) .results-stats {
          display: none;
        }
        .mini-results-counter {
          font-size: 11px;
          color: var(
            --ddd-theme-default-keystoneYellow,
            var(--simple-colors-default-theme-grey-8, #666)
          );
          padding: var(--ddd-spacing-1);
          text-align: center;
          border-top: 1px solid
            var(
              --ddd-theme-default-limestoneGray,
              var(--simple-colors-default-theme-grey-4, #ddd)
            );
          background: var(
            --ddd-theme-default-coalyGray,
            var(--simple-colors-default-theme-grey-2, #f8f8f8)
          );
          display: none;
        }
        :host([mini]) .mini-results-counter {
          display: block;
        }
        :host([dark][mini]) .mini-results-counter {
          color: var(
            --ddd-theme-default-coalyGray,
            var(--simple-colors-default-theme-grey-4, #ccc)
          );
          border-top-color: var(
            --ddd-theme-default-coalyGray,
            var(--simple-colors-default-theme-grey-8, #444)
          );
          background: var(
            --ddd-theme-default-slateMaxLight,
            var(--simple-colors-default-theme-grey-10, #222)
          );
        }
        :host([mini]) .results {
          padding: 0;
        }
      `]}updated(e){super.updated&&super.updated(e),e.forEach(((e,t)=>{if("filtered"==t&&void 0!==e){this.filtered.length>0&&(this.loading=!1),this._announceResults(),this._updateActiveDescendant(-1);const e=globalThis.SuperDaemonManager.requestAvailability();(e.santaMode||this.listeningForInput)&&(clearTimeout(this._selectTimeout),this._selectTimeout=setTimeout((()=>{(1===this.filtered.length||this.filtered&&this.filtered[0]&&this.filtered[0].title.toLocaleLowerCase()==e.value.toLocaleLowerCase())&&(this.shadowRoot.querySelector("super-daemon-row").selected(),e.listeningForInput=!0)}),600))}"opened"==t&&this.shadowRoot&&this.opened&&(this.activeType=null,this.activeDrag=!1,this.focusInput(),this.mini&&!this.wand&&this.shadowRoot.querySelector(".results").scrollTo(0,0)),"commandContext"==t&&this.dispatchEvent(new CustomEvent("super-daemon-command-context-changed",{detail:{value:this[t]}})),"like"==t&&this.dispatchEvent(new CustomEvent("like-changed",{detail:{value:this[t]}}))}))}focusInput(){this.shadowRoot.querySelector("super-daemon-search").focusInput()}selectInput(){this.shadowRoot.querySelector("super-daemon-search").selectInput()}setupProgram(e=""){this.programSearch=e,this.focusInput(),this.selectInput(),this.shadowRoot.querySelector(".results").scrollTo(0,0)}_updateActiveDescendant(e){this._selectedIndex=e,this._activeDescendant=e>=0?`option-${e}`:"",this.requestUpdate(),requestAnimationFrame((()=>{this.shadowRoot.querySelectorAll("super-daemon-row").forEach(((t,s)=>{const i=s===e;t.active=i})),e>=0&&this.shadowRoot.querySelector("lit-virtualizer")&&this.shadowRoot.querySelector("lit-virtualizer").scrollToIndex(e,"center")}))}_announceResults(){const e=this.filtered.length,t=`${e} ${1===e?"result":"results"} available`,s=globalThis.document.createElement("div");s.setAttribute("aria-live","polite"),s.setAttribute("aria-atomic","true"),s.style.position="absolute",s.style.left="-10000px",s.style.width="1px",s.style.height="1px",s.style.overflow="hidden",s.textContent=t,globalThis.document.body.appendChild(s),setTimeout((()=>{s.parentNode&&globalThis.document.body.removeChild(s)}),1e3)}itemSelected(e){this.like="",this.programSearch=""}_resultsKeydown(e){if(this.filtered.length>0&&this.shadowRoot.querySelector("super-daemon-row[active]"))switch(e.key){case"ArrowUp":case"ArrowLeft":this.shadowRoot.querySelector("super-daemon-row[active]")===this.shadowRoot.querySelector("super-daemon-row")?(this.shadowRoot.querySelector("lit-virtualizer").scrollToIndex(this.filtered.length-1,"center"),requestAnimationFrame((()=>{this.shadowRoot.querySelector("super-daemon-row:last-of-type").focus()}))):this.shadowRoot.querySelector("super-daemon-row[active]").previousElementSibling.focus();break;case"ArrowDown":case"ArrowRight":this.shadowRoot.querySelector("super-daemon-row[active]")===this.shadowRoot.querySelector("super-daemon-row:last-of-type")?(this.shadowRoot.querySelector("lit-virtualizer").scrollToIndex(0,"center"),requestAnimationFrame((()=>{this.shadowRoot.querySelector("super-daemon-row").focus()}))):this.shadowRoot.querySelector("super-daemon-row[active]").nextElementSibling.focus()}}focusedChanged(e){this.focused=e.detail.value}inputfilterChanged(e){const t=e.target&&"string"==typeof e.target.value?e.target.value:"";this.programName?this.programSearch=t:this.like=t,this.dispatchEvent(new CustomEvent("value-changed",{bubbles:!0,composed:!0,detail:{value:t}}))}listeningForInputChanged(e){e.detail.value&&this.shadowRoot.querySelector(".results").scrollTo(0,0)}commandContextChanged(e){this.commandContext=e.detail.value}_inputKeydown(e){if(this.filtered.length>0)switch(e.key){case"Enter":e.preventDefault(),this._selectedIndex>=0?this.shadowRoot.querySelectorAll("super-daemon-row")[this._selectedIndex].selected():this.shadowRoot.querySelector("super-daemon-row").selected();break;case"ArrowUp":e.preventDefault();const t=this._selectedIndex<=0?this.filtered.length-1:this._selectedIndex-1;this._updateActiveDescendant(t);break;case"ArrowDown":e.preventDefault();const s=this._selectedIndex>=this.filtered.length-1?0:this._selectedIndex+1;this._updateActiveDescendant(s);break;case"Escape":e.preventDefault(),this._updateActiveDescendant(-1),this.dispatchEvent(new CustomEvent("super-daemon-close",{bubbles:!0,composed:!0,cancelable:!0}))}else"Enter"===e.key&&this.programName&&""!==this.programSearch.trim()&&(this.dispatchEvent(new CustomEvent("super-daemon-program-enter",{bubbles:!0,cancelable:!0,composed:!0,detail:{programName:this.programName,input:this.programSearch.trim()}})),e.preventDefault());switch(e.key){case"!":case"/":case"\\":case">":case"<":"\\"===e.key&&""==this.like||"!"===e.key&&""==this.like?(this.commandContext="/",e.preventDefault()):"<"===e.key&&""==this.like?(this.commandContext=">",e.preventDefault()):""==this.like&&(this.commandContext=e.key,e.preventDefault());break;case"Backspace":""==this.programSearch&&this.programName?(this.dispatchEvent(new CustomEvent("super-daemon-run-program",{bubbles:!0,cancelable:!0,composed:!0,detail:!1})),e.preventDefault()):!this.programName&&""==this.like&&this.commandContext&&(this.commandContext="*",e.preventDefault())}}dropEvent(e){e.preventDefault(),this.activeDrag=!1,this.activeType=null;globalThis.SuperDaemonManager.requestAvailability().waveWand(["","/",e,"hax-agent","Agent"],this.shadowRoot.querySelector("#merlin"),"coin2")}dragenterEvent(e){e.preventDefault(),this.shadowRoot.querySelector("super-daemon-search").dragover=!0}dragoverEvent(e){e.preventDefault(),this.shadowRoot.querySelector("super-daemon-search").dragover=!0}dragleaveEvent(e){e.preventDefault(),this.shadowRoot.querySelector("super-daemon-search").dragover=!1}render(){return e`
      <super-daemon-search
        role="combobox"
        aria-expanded="${this.filtered.length>0?"true":"false"}"
        aria-haspopup="listbox"
        aria-autocomplete="list"
        aria-controls="results-listbox"
        aria-activedescendant="${this._activeDescendant||""}"
        @keydown="${this._inputKeydown}"
        @focused-changed="${this.focusedChanged}"
        @value-changed="${this.inputfilterChanged}"
        @command-context-changed="${this.commandContextChanged}"
        @listening-for-input-changed="${this.listeningForInputChanged}"
        @drop="${this.dropEvent}"
        @dragenter="${this.dragenterEvent}"
        @dragleave="${this.dragleaveEvent}"
        @dragover="${this.dragoverEvent}"
        icon="${this.icon}"
        icon-accent="${this.iconAccent}"
        value="${this.like}"
        ?voice-search="${this.voiceSearch}"
        ?mini="${this.mini}"
        ?wand="${this.wand}"
        ?loading="${this.loading}"
        program-search="${this.programSearch}"
        program-placeholder="${this.programPlaceholder||""}"
        ?listening-for-input="${this.listeningForInput}"
        command-context="${this.commandContext}"
        droppable-type="${this.activeType}"
        ?droppable="${this.activeDrag}"
      >
      </super-daemon-search>
      <div
        class="results"
        @keydown="${this._resultsKeydown}"
        @super-daemon-row-selected="${this.itemSelected}"
      >
        ${this.loading?e`<div class="loading">
              ${this.t.loadingResults||"Loading results"}..
            </div>`:e`
              ${this.filtered.length&&0!==this.filtered.length?e`
                    <lit-virtualizer
                      role="listbox"
                      id="results-listbox"
                      aria-label="${this.t.results||"Results"}"
                      scroller
                      .items=${this.filtered}
                      .renderItem=${(t,i)=>t?e`<super-daemon-row
                              role="option"
                              id="option-${i}"
                              tabindex="-1"
                              aria-selected="${this._selectedIndex===i}"
                              data-row-index="${i}"
                              .value="${t.value||{}}"
                              icon="${t.icon}"
                              image="${t.image}"
                              ?dark="${this.dark}"
                              text-character="${t.textCharacter}"
                              title="${t.title}"
                              .tags="${t.tags||[]}"
                              event-name="${t.eventName}"
                              path="${t.path}"
                              ?more="${t.more&&(!this.mini||this.wand)}"
                              ?mini="${this.mini}"
                              >${t.more?t.more:s}</super-daemon-row
                            >`:s}
                    ></lit-virtualizer>
                  `:e`<div class="no-results">
                    ${this.t.noResultsForThisTerm||"No results for this term"}
                    <div class="slotted"><slot></slot></div>
                  </div> `}
            `}
        <div class="results-stats">
          ${this.filtered.length} / ${this.items.length}
          ${this.t.results||"Results"}
        </div>
        ${this.mini&&this.filtered.length>1?e`<div class="mini-results-counter">
              ${this.filtered.length} results
            </div>`:s}
      </div>
      <div id="bottom"></div>
    `}}globalThis.customElements.define(SuperDaemonUI.tag,SuperDaemonUI);
