// @lrnwebcomponents/lrnsys-button/lrnsys-button.js
define(["exports","require","../../lit-element/lit-element.js","../materializecss-styles/lib/colors.js","../../@polymer/paper-button/paper-button.js","../../@polymer/iron-icon/iron-icon.js"],(function(_exports,_require,_litElement,_colors,_paperButton,_ironIcon){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.LrnsysButton=void 0,_require=babelHelpers.interopRequireWildcard(_require);
/**
   * Copyright 2018 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class LrnsysButton extends _litElement.LitElement{static get styles(){return[_colors.materialCssStyles,_litElement.css`
        :host {
          display: block;
          --lrnsys-button-height: 48px;
        }
        :host([disabled]) {
          pointer-events: none;
        }
        a {
          text-decoration: none;
          display: block;
          color: var(--lrnsys-button-link-color, #000000);
          display: flex;
        }
        paper-button {
          padding: 0;
          margin: 0;
          min-width: 0.16px;
          height: inherit;
          -webkit-justify-content: flex-start;
          justify-content: flex-start;
          align-items: center;
          width: 100%;
          text-transform: unset;
          border-radius: unset;
          display: flex;
        }
        .no-padding {
          padding: 0;
        }
        paper-button iron-icon {
          height: var(--lrnsys-button-height);
          margin: 0 4px;
        }
        paper-button iron-icon:first-child {
          margin: 0 4px 0 0;
        }
        paper-button iron-icon:last-child {
          margin: 0 0 0 4px;
        }
        paper-button div.inner {
          height: var(--lrnsys-button-height);
          line-height: var(--lrnsys-button-height);
          display: flex;
          padding: 0 16px;
        }
        paper-button span.label {
          height: var(--lrnsys-button-height);
          line-height: var(--lrnsys-button-height);
        }
        .no-margin {
          margin: 0 !important;
        }
        .no-right-padding {
          padding-right: 0 !important;
        }
        .no-left-padding {
          padding-left: 0 !important;
        }
      `]}constructor(){super(),this.href="#",this.label="",this.target="_self",this.icon="",this.alt="",this.focusState=!1,this.disabled=!1,setTimeout(()=>{this.addEventListener("mousedown",this.tapEventOn.bind(this)),this.addEventListener("mouseover",this.tapEventOn.bind(this)),this.addEventListener("focusin",this.tapEventOn.bind(this)),this.addEventListener("focusout",this.tapEventOff.bind(this)),this.addEventListener("mouseout",this.tapEventOff.bind(this)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/iron-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["../simple-tooltip/simple-tooltip.js"],res,rej))},0)}render(){return _litElement.html`
      <a
        tabindex="-1"
        id="lrnsys-button-link"
        href="${this.href}"
        target="${this.target}"
        ?disabled="${this.disabled}"
      >
        <paper-button
          id="button"
          title="${this.alt}"
          ?raised="${this.raised}"
          class="${this.buttonClass} ${this.color} ${this.textColor}"
          ?disabled="${this.disabled}"
          @focus-changed="${this.focusToggle}"
        >
          <div class="inner ${this.innerClass}">
            <slot name="prefix"></slot>
            <iron-icon
              icon="${this.icon}"
              id="icon"
              class="${this.iconClass}"
              ?hidden="${!this.icon}"
            ></iron-icon>
            <span class="label" ?hidden="${!this.label}"> ${this.label} </span>
            <slot></slot>
          </div>
        </paper-button>
      </a>
      <simple-tooltip
        for="lrnsys-button-link"
        animation-delay="0"
        ?hidden="${!this.alt}"
        >${this.alt}</simple-tooltip
      >
    `}static get tag(){return"lrnsys-button"}static get properties(){return{href:{type:String,reflect:!0},raised:{type:Boolean,reflect:!0},label:{type:String},target:{type:String},icon:{type:String},hoverClass:{type:String,attribute:"hover-class"},buttonClass:{type:String,attribute:"button-class"},iconClass:{type:String,attribute:"icon-class"},innerClass:{type:String,attribute:"inner-class"},color:{type:String,reflect:!0},textColor:{type:String,attribute:"text-color"},alt:{type:String},disabled:{type:Boolean},focusState:{type:Boolean,attribute:"focus-state"}}}tapEventOn(e){void 0===this.hoverClass||this.disabled||this.hoverClass.split(" ").forEach((item,index)=>{""!=item&&(this.shadowRoot.querySelector("#button").classList.add(item),-1!=item.indexOf("-")&&this.shadowRoot.querySelector("#icon").classList.add(item))})}tapEventOff(e){void 0===this.hoverClass||this.disabled||this.hoverClass.split(" ").forEach((item,index)=>{""!=item&&(this.shadowRoot.querySelector("#button").classList.remove(item),-1!=item.indexOf("-")&&this.shadowRoot.querySelector("#icon").classList.remove(item))})}focusToggle(e){void 0===this.hoverClass||this.disabled||this.hoverClass.split(" ").forEach((item,index)=>{""!=item&&(this.focusState?(this.shadowRoot.querySelector("#button").classList.remove(item),-1!=item.indexOf("-")&&this.shadowRoot.querySelector("#icon").classList.remove(item)):(this.shadowRoot.querySelector("#button").classList.add(item),-1!=item.indexOf("-")&&this.shadowRoot.querySelector("#icon").classList.add(item)))});this.focusState=!this.focusState}}_exports.LrnsysButton=LrnsysButton,window.customElements.define(LrnsysButton.tag,LrnsysButton)}));
