// @lrnwebcomponents/q-r/q-r.js
define(["exports", "require", "../../lit/index.js"], function (_exports, _require, _index) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.QR = void 0;
  _require = _interopRequireWildcard(_require);

  var _templateObject_3491acf0012611ec9676957c409166bc, _templateObject2_3491acf0012611ec9676957c409166bc;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `q-r`
   * `Polymer wrapper for a qr code.`
   *
   * @demo demo/index.html
   * @element q-r
   */
  var QR = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(QR, _LitElement);

    var _super = _createSuper(QR);

    function QR() {
      var _this;

      babelHelpers.classCallCheck(this, QR);
      _this = _super.call(this);
      _this.modulesize = 4;
      _this.margin = 2;
      _this.format = "png";
      new Promise(function (res, rej) {
        return _require.default(["./lib/qr-code.js"], res, rej);
      });
      return _this;
    }

    babelHelpers.createClass(QR, [{
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_3491acf0012611ec9676957c409166bc || (_templateObject_3491acf0012611ec9676957c409166bc = babelHelpers.taggedTemplateLiteral(["\n      <qr-code\n        id=\"qr\"\n        data=\"", "\"\n        modulesize=\"", "\"\n        margin=\"", "\"\n        format=\"", "\"\n      ></qr-code>\n      <a href=\"", "\" id=\"link\">", "</a>\n    "])), this.data, this.modulesize, this.margin, this.format, this.data, this.title);
      }
    }], [{
      key: "styles",
      get:
      /**
       * LitElement constructable styles enhancement
       */
      function get() {
        return [(0, _index.css)(_templateObject2_3491acf0012611ec9676957c409166bc || (_templateObject2_3491acf0012611ec9676957c409166bc = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        #link {\n          visibility: hidden;\n          opacity: 0;\n        }\n      "])))];
      }
    }, {
      key: "tag",
      get: function get() {
        return "q-r";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Data to code via QR code
           */
          data: {
            type: String
          },

          /**
           * Alternate title for the data
           */
          title: {
            type: String
          },

          /**
           * module size of the square
           */
          modulesize: {
            type: Number
          },

          /**
           * Margin on the square
           */
          margin: {
            type: Number
          },

          /**
           * format to output
           */
          format: {
            type: String
          }
        };
      }
      /**
       * Attached to the DOM, now fire.
       */

    }, {
      key: "haxProperties",
      get: function get() {
        // Establish hax property binding
        return {
          canScale: true,
          canPosition: true,
          canEditSource: true,
          gizmo: {
            title: "QR Code",
            description: "A code to scan from a smartphone.",
            icon: "hax:qr-code",
            color: "grey",
            groups: ["QR"],
            handles: [{
              type: "video",
              source: "data",
              title: "title"
            }, {
              type: "image",
              source: "data",
              title: "title"
            }, {
              type: "link",
              source: "data",
              title: "title"
            }],
            meta: {
              author: "ELMS:LN"
            }
          },
          settings: {
            configure: [{
              property: "data",
              title: "QR data",
              description: "Source of the data for the QR code.",
              inputMethod: "haxupload",
              icon: "hardware:developer-board"
            }, {
              property: "title",
              title: "Alternate title",
              description: "An alternate title to go to the source of the QR code.",
              inputMethod: "alt",
              icon: "editor:title"
            }, {
              property: "modulesize",
              title: "Size",
              description: "Size of the QR code",
              inputMethod: "number",
              icon: "image:photo-size-select-small"
            }, {
              property: "margin",
              title: "Margin",
              description: "Wrapper to the code.",
              inputMethod: "number",
              icon: "icons:settings-overscan"
            }, {
              property: "format",
              title: "Output format",
              description: "Format to put it out.",
              inputMethod: "select",
              options: {
                png: "PNG",
                html: "HTML",
                svg: "SVG"
              },
              icon: "editor:bubble-chart"
            }],
            advanced: []
          }
        };
      }
    }]);
    return QR;
  }(_index.LitElement);

  _exports.QR = QR;
  window.customElements.define(QR.tag, QR);
});
