// @lrnwebcomponents/haxcms-elements/lib/core/site-list/site-login.js
define(["exports","require","../../../../../lit-element/lit-element.js","../../../../../@polymer/paper-styles/shadow.js","../../../../../@polymer/paper-styles/typography.js","../../../../../@polymer/paper-styles/color.js","../../../../../@polymer/polymer/lib/elements/custom-style.js"],(function(e,t,r,i,s,o,n){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.SiteLogin=void 0,t=babelHelpers.interopRequireWildcard(t);class SiteLogin extends r.LitElement{static get styles(){return[r.css`
        #loginform {
          width: 450px;
          height: 450px;
          background: var(--login-form-background-color, white);
          box-shadow: 0 12px 16px 1px rgba(0, 0, 0, 0.14),
            0 4px 22px 3px rgba(0, 0, 0, 0.12),
            0 6px 7px -4px rgba(0, 0, 0, 0.4);
        }
        #loginformcontent {
          padding: 48px;
        }
        #loginformcontent > * {
          margin-top: 8px;
          margin-bottom: 8px;
        }
        #loginbtn {
          margin-top: 24px;
          float: right;
          background-color: var(
            --login-btn-background-color,
            var(--paper-indigo-500)
          );
          color: var(--login-btn-text-color, white);
        }
        #loginbtn[disabled] {
          background-color: var(
            --login-btn-disabled-background-color,
            var(--paper-indigo-100)
          );
        }
        h1 {
          margin: 0;
        }
        h2 {
          margin: 0;
        }
        paper-progress {
          width: 100%;
        }
        #errormsg {
          margin-top: 16px;
          color: var(--login-error-label-color, var(--error-color));
        }
      `]}constructor(){super(),this.loading=!1,this.userInputLabel="Username",this.userInputErrMsg="Username required",this.passwordInputLabel="Password",this.passwordInputErrMsg="Password required",this.loginBtnText="Login",new Promise((e,r)=>t.default(["../../../../../@polymer/paper-button/paper-button.js"],e,r)),new Promise((e,r)=>t.default(["../../../../../@polymer/paper-input/paper-input.js"],e,r)),new Promise((e,r)=>t.default(["../../../../../@polymer/paper-progress/paper-progress.js"],e,r))}render(){return r.html`
      <custom-style>
        <style>
          #loginbtn {
            --paper-button-raised-keyboard-focus: {
              background-color: var(
                --login-btn-raised-background-color,
                var(--paper-pink-a200)
              ) !important;
              color: var(--login-btn-text-color, white) !important;
            }
            @apply --login-btn;
          }

          h1 {
            @apply --paper-font-display1;
            @apply --login-title;
          }

          h2 {
            @apply --paper-font-title;
            @apply --login-subtitle;
          }

          #errormsg {
            @apply --paper-font-menu;
          }
        </style>
      </custom-style>
      <div id="loginform">
        <paper-progress
          ?disabled="${!this.loading}"
          indeterminate
        ></paper-progress>
        <div id="loginformcontent">
          <h1>${this.title}</h1>
          <h2>${this.subtitle}</h2>
          <div id="errormsg">${this.errorMsg}</div>
          <paper-input
            id="userinput"
            .value="${this.username}"
            @value-changed="${this.usernameChanged}"
            ?disabled="${this.loading}"
            type="text"
            label="${this.userInputLabel}"
            required
            error-message="${this.userInputErrMsg}"
          ></paper-input>
          <paper-input
            id="passinput"
            .value="${this.password}"
            @value-changed="${this.passwordChanged}"
            ?disabled="${this.loading}"
            type="password"
            label="${this.passwordInputLabel}"
            required
            error-message="${this.passwordInputErrMsg}"
          ></paper-input>
          <paper-button
            @click="${this._login}"
            ?disabled="${this.loading}"
            id="loginbtn"
            raised
            class="indigo"
            >${this.loginBtnText}</paper-button
          >
          <slot name="links"></slot>
        </div>
      </div>
    `}static get tag(){return"site-login"}usernameChanged(e){this.username=e.detail.value}passwordChanged(e){this.password=e.detail.value}static get properties(){return{title:String,subtitle:String,errorMsg:String,username:{type:String},password:{type:String},loading:{type:Boolean},userInputLabel:{type:String,attribute:"user-input-label"},userInputErrMsg:{type:String,attribute:"user-input-err-msg"},passwordInputLabel:{type:String,attribute:"password-input-label"},passwordInputErrMsg:{type:String,attribute:"password-input-err-msg"},loginBtnText:{type:String,attribute:"login-btn-text"}}}updated(e){e.forEach((e,t)=>{["username","password"].includes(t)&&this.dispatchEvent(new CustomEvent(`${t}-changed`,{detail:{value:this[t]}}))})}firstUpdated(e){this.shadowRoot.querySelector("#loginform").addEventListener("keypress",this._keyPress.bind(this))}disconnectedCallback(){this.shadowRoot.querySelector("#loginform").removeEventListener("keypress",this._keyPress.bind(this)),super.disconnectedCallback()}_keyPress(e){if(13==e.keyCode)return this._login(),!1}_login(){this.shadowRoot.querySelector("#userinput").validate()&&this.shadowRoot.querySelector("#passinput").validate()&&this.dispatchEvent(new CustomEvent("login-btn-click",{bubbles:!0,composed:!0}))}}e.SiteLogin=SiteLogin,window.customElements.define(SiteLogin.tag,SiteLogin)}));
