diff --git a/adminimal_admin_menu.module b/adminimal_admin_menu.module
index f48f388..90e2e0b 100644
--- a/adminimal_admin_menu.module
+++ b/adminimal_admin_menu.module
@@ -31,8 +31,8 @@ function adminimal_admin_menu_page_build(&$page) {
     return;
   }
   $path = drupal_get_path('module', 'adminimal_admin_menu');
-  $load_slicknav = variable_get('adminimal_admin_menu_slicknav', TRUE );
-  $load_jQuery = variable_get('adminimal_admin_menu_jquery', TRUE );
+  $load_slicknav = variable_get('adminimal_admin_menu_slicknav', FALSE );
+  $load_jQuery = variable_get('adminimal_admin_menu_jquery', FALSE );
 
   // Attach the CSS and JavaScript assets.
   drupal_add_css($path . '/adminimal_admin_menu.css');
@@ -153,7 +153,7 @@ function adminimal_admin_menu_admin_menu_output_alter(&$content) {
   }
 
   // Create the responsive menu.
-  if (variable_get('adminimal_admin_menu_slicknav', 'TRUE')) {
+  if (variable_get('adminimal_admin_menu_slicknav', FALSE)) {
     // Prepare the responsive menu and Join the Icon menu with the administration menu.
     $responsivemenu = array_merge($content['icon'], $content['menu']);
 
diff --git a/adminimal_menu_settings.inc b/adminimal_menu_settings.inc
index e1147f4..7cdde9d 100644
--- a/adminimal_menu_settings.inc
+++ b/adminimal_menu_settings.inc
@@ -42,7 +42,7 @@ function adminimal_admin_menu_settings($form, &$form_state) {
 
 	$form['advanced_settings']['adminimal_admin_menu_slicknav'] = array(
 	  '#type' => 'checkbox',
-	  '#default_value' => variable_get('adminimal_admin_menu_slicknav', 'TRUE'),
+	  '#default_value' => variable_get('adminimal_admin_menu_slicknav', FALSE),
 	  '#title' => t('Enable Responsive Menu.'),
 	  '#description' => t('<strong>Default value => Checked</strong>. 
 	  Enable this option if you want to have responsive menu and mobile device support. 
@@ -51,7 +51,7 @@ function adminimal_admin_menu_settings($form, &$form_state) {
 
 	$form['advanced_settings']['adminimal_admin_menu_jquery'] = array(
 	  '#type' => 'checkbox',
-	  '#default_value' => variable_get('adminimal_admin_menu_jquery', 'TRUE'),
+	  '#default_value' => variable_get('adminimal_admin_menu_jquery', FALSE),
 	  '#title' => t('Load the requred jQuery 1.7 library automagically.'),
 	  '#description' => t('<strong>Default value => Checked</strong>. This will load the newer jQuery version 1.7 using 
 	  the no-conflict method so it wont interfere with any existing jQuery or other java-script libraries. 
@@ -63,7 +63,7 @@ function adminimal_admin_menu_settings($form, &$form_state) {
        ':input[name="adminimal_admin_menu_slicknav"]' => array('checked' => TRUE),
       ),
       'unchecked' => array(
-       variable_get('adminimal_admin_menu_jquery', 'TRUE') => FALSE,
+       variable_get('adminimal_admin_menu_jquery', FALSE) => FALSE,
       ),
      ),
 	);
