// @haxtheweb/screen-recorder/screen-recorder.js
/**
 * Copyright 2025 haxtheweb
 * @license Apache-2.0, see LICENSE for full text.
 */
import{LitElement as e,html as t,css as i}from"../../lit/index.js";import{DDDSuper as o}from"../d-d-d/d-d-d.js";import{I18NMixin as r}from"../i18n-manager/lib/I18NMixin.js";import"../simple-icon/lib/simple-icon-button-lite.js";import"../simple-icon/lib/simple-icons.js";export class ScreenRecorder extends(o(r(e))){static get tag(){return"screen-recorder"}constructor(){super(),this.recording=!1,this.videoSrc="",this.downloadUrl="",this.completeBlob=null,this.recorder=null,this.chunks=[],this.stream=null,this.audioStream=null,this.includeSystemAudio=!0,this.includeMicrophoneAudio=!0,this.t=this.t||{},this.t={...this.t,startRecording:"Start Recording",stopRecording:"Stop Recording",downloadVideo:"Download Video",systemAudio:"Include System Audio",microphoneAudio:"Include Microphone"},this.registerLocalization({context:this,localesPath:new URL("./locales/screen-recorder.ar.json",import.meta.url).href+"/../"})}static get properties(){return{...super.properties,recording:{type:Boolean},videoSrc:{type:String},downloadUrl:{type:String},includeSystemAudio:{type:Boolean},includeMicrophoneAudio:{type:Boolean}}}static get styles(){return[super.styles,i`
      :host {
        display: block;
        color: light-dark(var(--ddd-theme-default-coalyGray), var(--ddd-theme-default-white));
        font-family: var(--ddd-font-navigation);
      }
      .video-container {
        text-align: center;
      }
      .video-container video {
        max-width: 100%;
        height: auto;
        border: var(--ddd-border-sm);
        border-radius: var(--ddd-radius-sm);
      }
      .controls {
        text-align: center;
        margin: var(--ddd-spacing-2) 0;
      }
      .controls simple-icon-button-lite {
        background-color: var(--ddd-theme-default-error, #d32f2f);
        color: white;
        border: none;
        padding: var(--ddd-spacing-1, 4px) var(--ddd-spacing-2, 8px);
        margin: var(--ddd-spacing-1, 4px);
        border-radius: var(--ddd-radius-sm, 4px);
        font-family: var(--ddd-font-navigation, sans-serif);
        font-size: var(--ddd-font-size-4xs, 10px);
        cursor: pointer;
        transition: background-color 0.2s ease;
        min-width: 60px;
        text-align: center;
        display: inline-flex;
        align-items: center;
        justify-content: center;
        --simple-icon-width: 12px;
        --simple-icon-height: 12px;
      }
      .controls button:hover {
        background-color: var(--ddd-theme-default-original87Pink);
      }
      .controls button:disabled {
        opacity: 0.6;
        cursor: not-allowed;
      }
      .download-link {
        text-decoration: none;
      }
      .download-link simple-icon-button-lite {
        background-color: var(--ddd-theme-default-keystoneYellow, #ffd100) !important;
        color: var(--ddd-theme-default-coalyGray, #444) !important;
      }
      .download-link simple-icon-button-lite:hover {
        background-color: var(--ddd-theme-default-original87Pink, #ff6b9d) !important;
      }
      .hidden {
        display: none !important;
      }
      .audio-options {
        text-align: center;
      }
      .audio-options h4 {
        margin: 0 0 var(--ddd-spacing-1) 0;
        font-size: var(--ddd-font-size-3xs);
        color: light-dark(var(--ddd-theme-default-coalyGray), var(--ddd-theme-default-white));
        font-family: var(--ddd-font-navigation);
      }
      .checkbox-container {
        display: inline-block;
        margin: var(--ddd-spacing-1);
      }
      .checkbox-container label {
        display: flex;
        align-items: center;
        cursor: pointer;
        font-size: var(--ddd-font-size-4xs);
        color: light-dark(var(--ddd-theme-default-coalyGray), var(--ddd-theme-default-white));
        font-family: var(--ddd-font-navigation);
      }
      .checkbox-container input[type="checkbox"] {
        margin-right: var(--ddd-spacing-1);
        cursor: pointer;
      }
    `]}render(){return t`
      <div class="wrapper">
        <div class="video-container">
          <video 
            controls 
            class="${this.videoSrc?"":"hidden"}"
            .src="${this.videoSrc}"
          ></video>
        </div>
        
        <div class="audio-options ${this.recording?"hidden":""}">
          <h4>Audio Options</h4>
          <div class="checkbox-container">
            <label>
              <input 
                type="checkbox" 
                .checked="${this.includeSystemAudio}"
                @change="${this._toggleSystemAudio}"
              />
              ${this.t.systemAudio}
            </label>
          </div>
          <div class="checkbox-container">
            <label>
              <input 
                type="checkbox" 
                .checked="${this.includeMicrophoneAudio}"
                @change="${this._toggleMicrophoneAudio}"
              />
              ${this.t.microphoneAudio}
            </label>
          </div>
        </div>
        
        <div class="controls">
          <simple-icon-button-lite
            icon="av:fiber-smart-record"
            class="${this.recording?"hidden":""}"
            @click="${this._startRecording}"
          >
            ${this.t.startRecording}
          </simple-icon-button-lite>
          
          <simple-icon-button-lite
            icon="av:stop"
            class="${this.recording?"":"hidden"}"
            @click="${this._stopRecording}"
          >
            ${this.t.stopRecording}
          </simple-icon-button-lite>
          
          <a 
            class="download-link ${this.downloadUrl?"":"hidden"}"
            href="${this.downloadUrl}"
            download="screen-recording-${Date.now()}.webm"
          >
            <simple-icon-button-lite icon="icons:file-download">
              ${this.t.downloadVideo}
            </simple-icon-button-lite>
          </a>
        </div>
        
        <slot></slot>
      </div>
    `}_toggleSystemAudio(e){this.includeSystemAudio=e.target.checked}_toggleMicrophoneAudio(e){this.includeMicrophoneAudio=e.target.checked}async _startRecording(){try{this.stream=await navigator.mediaDevices.getDisplayMedia({video:{mediaSource:"screen"},audio:this.includeSystemAudio});let e=this.stream;if(this.includeMicrophoneAudio)try{this.audioStream=await navigator.mediaDevices.getUserMedia({audio:!0});const t=new AudioContext,i=t.createMediaStreamDestination();if(this.includeSystemAudio&&this.stream.getAudioTracks().length>0){t.createMediaStreamSource(this.stream).connect(i)}t.createMediaStreamSource(this.audioStream).connect(i),e=new MediaStream([...this.stream.getVideoTracks(),...i.stream.getAudioTracks()])}catch(e){console.warn("Could not access microphone:",e)}this.recorder=new MediaRecorder(e),this.chunks=[],this.recorder.ondataavailable=e=>{this.chunks.push(e.data)},this.recorder.onstop=()=>{this._onRecordingStop()},this.recorder.start(),this.recording=!0}catch(e){console.error("Error starting screen recording:",e),alert("Error starting screen recording: "+e.message)}}_stopRecording(){this.recorder&&"recording"===this.recorder.state&&this.recorder.stop(),this.stream&&this.stream.getTracks().forEach((e=>{e.stop()})),this.audioStream&&(this.audioStream.getTracks().forEach((e=>{e.stop()})),this.audioStream=null),this.recording=!1}_onRecordingStop(){this.completeBlob=new Blob(this.chunks,{type:this.chunks[0].type}),this.videoSrc=URL.createObjectURL(this.completeBlob),this.downloadUrl=this.videoSrc,this.dispatchEvent(new CustomEvent("screen-recorder-blob",{bubbles:!0,composed:!0,cancelable:!0,detail:{blob:this.completeBlob}})),this.chunks=[]}disconnectedCallback(){super.disconnectedCallback(),this.videoSrc&&this.videoSrc.startsWith("blob:")&&URL.revokeObjectURL(this.videoSrc),this.stream&&this.stream.getTracks().forEach((e=>e.stop())),this.audioStream&&this.audioStream.getTracks().forEach((e=>e.stop()))}static get haxProperties(){return new URL(`./lib/${this.tag}.haxProperties.json`,import.meta.url).href}}globalThis.customElements.define(ScreenRecorder.tag,ScreenRecorder);
