// @lrnwebcomponents/lrn-markdown-editor/lrn-markdown-editor.js
define(["exports","../../@polymer/polymer/polymer-element.js","../../@polymer/paper-tabs/paper-tabs.js","../../@polymer/marked-element/marked-element.js","../../@polymer/iron-pages/iron-pages.js","./lib/lrn-markdown-editor-editor.js"],function(_exports,_polymerElement,_paperTabs,_markedElement,_ironPages,_lrnMarkdownEditorEditor){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.LrnMarkdownEditor=void 0;class LrnMarkdownEditor extends _polymerElement.PolymerElement{static get template(){return _polymerElement.html`
      <style>
        :host {
          display: block;
        }

        #split-pane {
          display: flex;
        }

        .split-pane > * {
          flex: 1 1 auto;
          min-height: 160px;
        }

        .preview-pane {
          background: lightblue;
        }

        paper-card {
          padding: 16px;
          width: calc(100% - 32px);
        }

        paper-tabs {
          background: #f5f5f5;
          border-style: solid;
          border-color: #dcdcdc;
          border-width: 1px;
          min-width: 500px;
        }

        marked-element.lrn-markdown-editor {
          width: 100%;
          word-wrap: break-word;
        }

        .container-flex {
          display: flex;
          flex-wrap: nowrap;
        }

        .split-pane .container-flex > * {
          width: 50%;
        }

        .split-pane marked-element {
          width: calc(100% - 32px);
          min-width: 150px;
          margin: 0 16px;
          padding: 0 16px;
          background: #fff;
          border-left: solid #dcdcdc 1px;
        }
      </style>

      <div class="mtz-toolbar">
        <paper-tabs selected="{{selected}}">
          <paper-tab>Write</paper-tab>
          <paper-tab>Preview</paper-tab>
          <paper-tab>Split View</paper-tab>
        </paper-tabs>
      </div>

      <iron-pages selected="{{selected}}">
        <section>
          <paper-card>
            <lrn-markdown-editor-editor
              content="{{content}}"
            ></lrn-markdown-editor-editor>
          </paper-card>
        </section>

        <section>
          <paper-card>
            <marked-element markdown="{{content}}"></marked-element>
          </paper-card>
        </section>

        <section class="split-pane">
          <paper-card>
            <div class="container-flex">
              <lrn-markdown-editor-editor
                content="{{content}}"
              ></lrn-markdown-editor-editor>
              <marked-element
                class="preview-pane"
                markdown="{{content}}"
              ></marked-element>
            </div>
          </paper-card>
        </section>
      </iron-pages>
    `}static get tag(){return"lrn-markdown-editor"}static get properties(){return{content:{type:String,notify:!0},selected:{type:String,value:"0",reflectToAttribute:!0},layout:{type:String,value:0},cookies:{type:Boolean,value:!0},elReady:{type:Boolean,value:!1}}}static get observers(){return["_selectedChanged(selected)"]}_selectedChanged(selected){var cookieName=this._getCookieName();2===selected?this._createCookie(cookieName,"true","30"):2!==selected&&!0===this.elReady&&this._eraseCookie(cookieName)}_createCookie(name,value,days){var expires="";if(days){var date=new Date;date.setTime(date.getTime()+24*days*60*60*1e3),expires="; expires="+date.toUTCString()}document.cookie=name+"="+value+expires+"; path=/"}_readCookie(name){for(var nameEQ=name+"=",ca=document.cookie.split(";"),i=0;i<ca.length;i++){for(var c=ca[i];" "==c.charAt(0);)c=c.substring(1,c.length);if(0==c.indexOf(nameEQ))return c.substring(nameEQ.length,c.length)}return null}_eraseCookie(name){this._createCookie(name,"",-1)}_getCookieName(){return"lrnmarkdowneditorsplitview"}ready(){super.ready();this.elReady=!0;var cookieName=this._getCookieName(),cookie=this._readCookie(cookieName);cookie&&"true"===cookie&&(this.selected=2)}}_exports.LrnMarkdownEditor=LrnMarkdownEditor,window.customElements.define(LrnMarkdownEditor.tag,LrnMarkdownEditor)});
