// @lrnwebcomponents/hax-body/lib/hax-app-search-inputs.js
define(["exports", "require", "../../../@polymer/polymer/polymer-element.js", "../../eco-json-schema-form/lib/eco-json-schema-object.js", "../../simple-colors/simple-colors.js", "./hax-shared-styles.js"], function (_exports, _require, _polymerElement, _ecoJsonSchemaObject, _simpleColors, _haxSharedStyles) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HaxAppSearchInputs = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject_4b6c2110b3b911e9b58d79562a23326c() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style include=\"hax-shared-styles\">\n        :host {\n          display: block;\n        }\n        eco-json-schema-object {\n          color: var(--hax-color-text);\n        }\n        .search-label {\n          font-size: 24px;\n          color: var(--hax-color-text);\n          font-weight: bold;\n          margin: 0;\n          padding: 0;\n        }\n      </style>\n      <div class=\"search-label\">Search [[label]]</div>\n      <eco-json-schema-object\n        id=\"form\"\n        schema=\"[[schema]]\"\n        value=\"{{values}}\"\n      ></eco-json-schema-object>\n    "]);

    _templateObject_4b6c2110b3b911e9b58d79562a23326c = function _templateObject_4b6c2110b3b911e9b58d79562a23326c() {
      return data;
    };

    return data;
  }

  /**
  `hax-app-search-inputs`
   An element that brokers the visual display of a listing of material from an end point. The goal is to normalize data from some location which is media centric. This expects to get at least enough data in order to form a grid of items which are selectable. It's also generically implemented so that anything can be hooked up as a potential source for input (example: youtube API or custom in-house solution). The goal is to return enough info via fired event so that hax-manager can tell hax-body that the user selected a tag, properties, slot combination so that hax-body can turn the selection into a custom element / element injected into the hax-body slot.
  
  * @demo demo/index.html
  
  @microcopy - the mental model for this element
   - hax-app - a source of input we're querying for media / content
   - hax-app-search - element controlling the experience of searching an app
   - hax-body - the text are ultimately we are trying to insert this item into
  */
  var HaxAppSearchInputs =
  /*#__PURE__*/
  function (_PolymerElement) {
    babelHelpers.inherits(HaxAppSearchInputs, _PolymerElement);

    function HaxAppSearchInputs() {
      var _this;

      babelHelpers.classCallCheck(this, HaxAppSearchInputs);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(HaxAppSearchInputs).call(this));
      new Promise(function (res, rej) {
        return _require.default(["../../../@polymer/paper-input/paper-input.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../../@polymer/paper-item/paper-item.js"], res, rej);
      });
      return _this;
    }

    babelHelpers.createClass(HaxAppSearchInputs, [{
      key: "_valueChanged",

      /**
       * Search input was added.
       */
      value: function _valueChanged(change) {
        this.dispatchEvent(new CustomEvent("hax-app-search-values-changed", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: change.base
        }));
      }
    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_4b6c2110b3b911e9b58d79562a23326c());
      }
    }, {
      key: "tag",
      get: function get() {
        return "hax-app-search-inputs";
      }
    }, {
      key: "observers",
      get: function get() {
        return ["_valueChanged(values.*)"];
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Title.
           */
          label: {
            type: String,
            value: "app"
          },

          /**
           * Search input values mapped to schema inputs.
           */
          values: {
            type: Object
          },

          /**
           * Schema used to generate the input types.
           */
          schema: {
            type: Object
          }
        };
      }
    }]);
    return HaxAppSearchInputs;
  }(_polymerElement.PolymerElement);

  _exports.HaxAppSearchInputs = HaxAppSearchInputs;
  window.customElements.define(HaxAppSearchInputs.tag, HaxAppSearchInputs);
});
