// @lrnwebcomponents/hax-body/lib/hax-toolbar.js
define(["exports", "require", "../../../lit-element/lit-element.js"], function (_exports, _require, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HaxToolbar = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject2_67e850c02d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <hax-context-item\n        ?hidden=\"", "\"\n        mini\n        light\n        icon=\"close\"\n        label=\"Hide menu\"\n        event-name=\"close-menu\"\n        class=\"close-cap\"\n        direction=\"left\"\n      ></hax-context-item>\n      <div class=\"wrapper\">\n        <hax-context-item-menu\n          ?hidden=\"", "\"\n          @selected-value-changed=\"", "\"\n          id=\"justify\"\n          icon=\"", "\"\n          label=\"Alignment\"\n        >\n          <hax-context-item\n            menu\n            icon=\"editor:format-align-left\"\n            event-name=\"hax-align-left\"\n            >Left</hax-context-item\n          >\n          <hax-context-item\n            menu\n            icon=\"editor:format-align-center\"\n            event-name=\"hax-align-center\"\n            >Center</hax-context-item\n          >\n        </hax-context-item-menu>\n        <paper-slider\n          ?hidden=\"", "\"\n          id=\"slider\"\n          pin\n          min=\"", "\"\n          step=\"", "\"\n          max=\"", "\"\n          value=\"", "\"\n          @immediate-value-changed=\"", "\"\n          @value-changed=\"", "\"\n        ></paper-slider>\n        <paper-tooltip\n          ?hidden=\"", "\"\n          for=\"slider\"\n          position=\"top\"\n          offset=\"10\"\n        >\n          Resize\n        </paper-tooltip>\n        <slot name=\"primary\"></slot>\n        <hax-context-item\n          ?hidden=\"", "\"\n          icon=\"hax:bricks\"\n          label=\"Change type\"\n          event-name=\"grid-plate-convert\"\n        ></hax-context-item>\n        <hax-context-item\n          ?hidden=\"", "\"\n          icon=\"delete\"\n          label=\"Remove\"\n          event-name=\"grid-plate-delete\"\n        ></hax-context-item>\n        <hax-context-item-menu\n          ?hidden=\"", "\"\n          icon=\"more-vert\"\n          label=\"More operations\"\n          id=\"moremenu\"\n          event-name=\"grid-plate-op\"\n          reset-on-select\n        >\n          <paper-item value=\"\" hidden></paper-item>\n          <slot name=\"more\"></slot>\n          <hax-context-item-textop\n            menu\n            icon=\"icons:content-copy\"\n            event-name=\"grid-plate-duplicate\"\n            >Duplicate</hax-context-item-textop\n          >\n        </hax-context-item-menu>\n      </div>\n    "]);

    _templateObject2_67e850c02d8511ea80ccf124101da415 = function _templateObject2_67e850c02d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject_67e850c02d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: flex;\n          justify-content: flex-start;\n          visibility: visible;\n          transition: 0.3s all ease;\n          box-sizing: border-box;\n          height: 36px;\n          pointer-events: all;\n        }\n        :host *[hidden] {\n          display: none;\n        }\n        .wrapper {\n          display: flex;\n          border: 1px solid var(--hax-color-border-outline, black);\n          border-bottom: none;\n          color: #222222;\n          background-color: #ffffff;\n          height: 38px;\n          align-items: center;\n        }\n        :host .wrapper ::slotted(*) {\n          pointer-events: all;\n        }\n        .close-cap {\n          margin: 10px 10px 0 8px;\n          display: flex;\n          padding: 0;\n        }\n        :host * ::slotted(hax-context-item:hover),\n        :host * ::slotted(hax-context-item-textop:hover),\n        :host * ::slotted(hax-context-item-menu:hover),\n        hax-context-item:hover,\n        hax-context-item-textop:hover,\n        hax-context-item-menu:hover {\n          z-index: 2;\n        }\n        #moremenu hax-context-item,\n        #moremenu ::slotted(paper-item),\n        paper-item {\n          height: 32px;\n          padding: 0;\n          min-width: 100px;\n        }\n\n        #moremenu hax-context-item:hover,\n        :host #moremenu ::slotted(paper-item:hover),\n        paper-item:hover {\n          background-color: var(--hax-color-menu-heading-bg);\n          cursor: pointer;\n        }\n        paper-slider {\n          background-color: var(--hax-color-bg-accent);\n          color: #ffffff;\n          font-weight: bold;\n          height: 36px;\n          min-width: 100px;\n          --paper-slider-font-color: white;\n          --paper-slider-active-color: var(--hax-color-accent1);\n          --paper-slider-knob-color: var(--hax-color-accent1);\n          --paper-slider-pin-start-color: var(--hax-color-accent1);\n          --paper-slider-pin-color: var(--hax-color-accent1);\n        }\n        .convert-button {\n          border-top: 1px solid var(--hax-color-bg-accent);\n        }\n      "]);

    _templateObject_67e850c02d8511ea80ccf124101da415 = function _templateObject_67e850c02d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  var HaxToolbar =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(HaxToolbar, _LitElement);
    babelHelpers.createClass(HaxToolbar, null, [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject_67e850c02d8511ea80ccf124101da415())];
      }
    }]);

    function HaxToolbar() {
      var _this;

      babelHelpers.classCallCheck(this, HaxToolbar);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(HaxToolbar).call(this));
      _this.hideTransform = false;
      _this.selected = false;
      _this.haxProperties = {};
      _this.hideMore = false;
      _this.size = 100;
      _this.justifyIcon = "editor:format-align-left";
      _this.inline = false;
      _this.justifyValue = "";
      new Promise(function (res, rej) {
        return _require.default(["../../../@polymer/paper-slider/paper-slider.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../../@polymer/paper-tooltip/paper-tooltip.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../../@polymer/paper-item/paper-item.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../../@polymer/iron-icons/iron-icons.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../../@polymer/iron-icons/editor-icons.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["./hax-toolbar-item.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["./hax-toolbar-menu.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["./hax-context-item.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["./hax-context-item-menu.js"], res, rej);
      });
      setTimeout(function () {
        _this.addEventListener("hax-context-item-selected", _this._haxContextOperation.bind(babelHelpers.assertThisInitialized(_this)));
      }, 0);
      return _this;
    }

    babelHelpers.createClass(HaxToolbar, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject2_67e850c02d8511ea80ccf124101da415(), this.inline, !this.haxProperties.canPosition, this.justifyValueChanged, this.justifyIcon, !this.haxProperties.canScale, this.haxProperties.canScale && this.haxProperties.canScale.min ? this.haxProperties.canScale.min : 12.5, this.haxProperties.canScale && this.haxProperties.canScale.step ? this.haxProperties.canScale.step : 12.5, this.haxProperties.canScale && this.haxProperties.canScale.max ? this.haxProperties.canScale.max : 100, this.size, this.sizeChanged, this.sizeChanged, this.inline, this.hideTransform, this.inline, this.hideMode);
      }
    }, {
      key: "sizeChanged",
      value: function sizeChanged(e) {
        this.size = e.detail.value;
      }
    }, {
      key: "justifyValueChanged",
      value: function justifyValueChanged(e) {
        this.justifyValue = e.detail;
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "haxProperties") {
            _this2._haxPropertiesChanged(_this2[propName], oldValue);
          }

          if (propName == "size") {
            // notify
            _this2.dispatchEvent(new CustomEvent("size-changed", {
              detail: _this2[propName]
            }));
          }

          if (propName == "justifyValue") {
            // notify
            _this2.dispatchEvent(new CustomEvent("justify-value-changed", {
              detail: _this2[propName]
            }));
          }
        });
      }
      /**
       * If hax properties changes, let's see what the initial state
       * of the buttons should be.
       */

    }, {
      key: "_haxPropertiesChanged",
      value: function _haxPropertiesChanged(newValue, oldValue) {
        // value doesn't matter, just look at what's active
        if (babelHelpers.typeof(window.HaxStore.instance.activeNode) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          if (window.HaxStore.instance.isTextElement(window.HaxStore.instance.activeNode) || window.HaxStore.instance.activeNode.tagName == "HR") {
            this.hideTransform = true;
          } else {
            this.hideTransform = false;
          }

          if (window.HaxStore.instance.activeNode.style.width != "") {
            this.size = window.HaxStore.instance.activeNode.style.width.replace("%", "");
          } else {
            this.size = 100;
          }

          if (window.HaxStore.instance.activeNode.style.margin == "0px auto" && window.HaxStore.instance.activeNode.style.display == "block") {
            this.justifyValue = "hax-align-center";
            this.justifyIcon = "editor:format-align-center";
          } else {
            this.justifyValue = "hax-align-left";
            this.justifyIcon = "editor:format-align-left";
          }
        }
      }
      /**
       * Respond to simple modifications.
       */

    }, {
      key: "_haxContextOperation",
      value: function _haxContextOperation(e) {
        var _this3 = this;

        var detail = e.detail; // support a simple insert event to bubble up or everything else

        switch (detail.eventName) {
          case "hax-align-left":
          case "hax-align-center":
            this.justifyIcon = detail.target.icon;
            break;

          case "close-menu":
            setTimeout(function () {
              _this3.shadowRoot.querySelector("#moremenu").shadowRoot.querySelector("#menu").hideMenu();

              _this3.shadowRoot.querySelector("#justify").shadowRoot.querySelector("#menu").hideMenu();
            }, 200);
            break;
        }
      }
    }], [{
      key: "tag",
      get: function get() {
        return "hax-toolbar";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Hide the transform button as its not supported
           */
          hideTransform: {
            type: Boolean,
            attribute: "hide-transform"
          },

          /**
           * See what's selected
           */
          selected: {
            type: Boolean,
            reflect: true
          },

          /**
           * Selected value to match ce direction currently.
           */
          haxProperties: {
            type: Object,
            attribute: "hax-properties"
          },

          /**
           * Hide the more menu.
           */
          hideMore: {
            type: Boolean,
            attribute: "hide-more"
          },

          /**
           * size of the slider if it exists.
           */
          size: {
            type: Number
          },

          /**
           * Justify icon to reflect state.
           */
          justifyIcon: {
            type: String,
            attribute: "justify-icon"
          },

          /**
           * This is an inline context menu
           */
          inline: {
            type: Boolean,
            reflect: true
          },

          /**
           * Selected value to match ce direction currently.
           */
          justifyValue: {
            type: String,
            attribute: "justify-value"
          }
        };
      }
    }]);
    return HaxToolbar;
  }(_litElement.LitElement);

  _exports.HaxToolbar = HaxToolbar;
  window.customElements.define(HaxToolbar.tag, HaxToolbar);
});
