// @lrnwebcomponents/hax-body/lib/hax-blox-browser-item.js
define(["exports", "require", "../../../@polymer/polymer/polymer-element.js", "../../../@polymer/polymer/lib/utils/render-status.js", "./hax-shared-styles.js"], function (_exports, _require, _polymerElement, _renderStatus, _haxSharedStyles) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HaxBloxBrowserItem = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject_d19b63a0aa5311e9b999174ecd579fc6() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style include=\"hax-shared-styles\">\n        :host {\n          display: inline-flex;\n        }\n        :host([elevation=\"1\"]) {\n          -webkit-transform: scale(1, 1);\n          transform: scale(1, 1);\n        }\n        :host([elevation=\"2\"]) {\n          -webkit-transform: scale(1.4, 1.4);\n          transform: scale(1.4, 1.4);\n        }\n        paper-card {\n          margin: 4px 0;\n          border-radius: 10px;\n        }\n        paper-button {\n          color: var(--hax-color-text);\n          background-color: #ffffff;\n          border: 2px solid var(--hax-color-border-outline);\n          text-transform: none;\n          margin: 0;\n          height: 80px !important;\n          width: 200px !important;\n          display: flex;\n          border-radius: 10px;\n          min-width: unset;\n        }\n        paper-button .item-title {\n          font-size: 14px;\n          display: inline-flex;\n        }\n        .flip-icon {\n          transform: rotateY(180deg);\n        }\n        iron-icon {\n          width: 40px;\n          height: 40px;\n          display: inline-block;\n          color: var(--hax-color-text);\n        }\n        @media screen and (max-width: 550px) {\n          paper-button .item-title {\n            font-size: 12px;\n          }\n        }\n      </style>\n      <paper-card id=\"card\" elevation=\"[[elevation]]\">\n        <paper-button\n          id=\"button\"\n          on-click=\"_fireEvent\"\n          data-voicecommand$=\"select [[title]]\"\n        >\n          <div class=\"button-inner\">\n            <iron-icon icon=\"[[icon]]\"></iron-icon>\n            <div class=\"item-title\">[[title]]</div>\n          </div>\n        </paper-button>\n      </paper-card>\n    "], ["\n      <style include=\"hax-shared-styles\">\n        :host {\n          display: inline-flex;\n        }\n        :host([elevation=\"1\"]) {\n          -webkit-transform: scale(1, 1);\n          transform: scale(1, 1);\n        }\n        :host([elevation=\"2\"]) {\n          -webkit-transform: scale(1.4, 1.4);\n          transform: scale(1.4, 1.4);\n        }\n        paper-card {\n          margin: 4px 0;\n          border-radius: 10px;\n        }\n        paper-button {\n          color: var(--hax-color-text);\n          background-color: #ffffff;\n          border: 2px solid var(--hax-color-border-outline);\n          text-transform: none;\n          margin: 0;\n          height: 80px !important;\n          width: 200px !important;\n          display: flex;\n          border-radius: 10px;\n          min-width: unset;\n        }\n        paper-button .item-title {\n          font-size: 14px;\n          display: inline-flex;\n        }\n        .flip-icon {\n          transform: rotateY(180deg);\n        }\n        iron-icon {\n          width: 40px;\n          height: 40px;\n          display: inline-block;\n          color: var(--hax-color-text);\n        }\n        @media screen and (max-width: 550px) {\n          paper-button .item-title {\n            font-size: 12px;\n          }\n        }\n      </style>\n      <paper-card id=\"card\" elevation=\"[[elevation]]\">\n        <paper-button\n          id=\"button\"\n          on-click=\"_fireEvent\"\n          data-voicecommand\\$=\"select [[title]]\"\n        >\n          <div class=\"button-inner\">\n            <iron-icon icon=\"[[icon]]\"></iron-icon>\n            <div class=\"item-title\">[[title]]</div>\n          </div>\n        </paper-button>\n      </paper-card>\n    "]);

    _templateObject_d19b63a0aa5311e9b999174ecd579fc6 = function _templateObject_d19b63a0aa5311e9b999174ecd579fc6() {
      return data;
    };

    return data;
  }

  /**
   * `hax-blox-browser-item`
   * `A button on the hax-gizmo-browser app display`
   */
  var HaxBloxBrowserItem =
  /*#__PURE__*/
  function (_PolymerElement) {
    babelHelpers.inherits(HaxBloxBrowserItem, _PolymerElement);

    function HaxBloxBrowserItem() {
      var _this;

      babelHelpers.classCallCheck(this, HaxBloxBrowserItem);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(HaxBloxBrowserItem).call(this));
      new Promise(function (res, rej) {
        return _require.default(["../../../@polymer/paper-button/paper-button.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../../@polymer/paper-card/paper-card.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../../@polymer/iron-icon/iron-icon.js"], res, rej);
      });

      _this.addEventListener("mousedown", _this.tapEventOn.bind(babelHelpers.assertThisInitialized(_this)));

      _this.addEventListener("mouseover", _this.tapEventOn.bind(babelHelpers.assertThisInitialized(_this)));

      _this.addEventListener("mouseout", _this.tapEventOff.bind(babelHelpers.assertThisInitialized(_this)));

      _this.addEventListener("focusin", _this.tapEventOn.bind(babelHelpers.assertThisInitialized(_this)));

      _this.addEventListener("focusout", _this.tapEventOff.bind(babelHelpers.assertThisInitialized(_this)));

      return _this;
    }

    babelHelpers.createClass(HaxBloxBrowserItem, [{
      key: "tapEventOn",

      /**
       * special handling for taps on the thing
       */
      value: function tapEventOn(e) {
        this.elevation = 2;
      }
      /**
       * Hover off stop showing the deeper shadow.
       */

    }, {
      key: "tapEventOff",
      value: function tapEventOff(e) {
        this.elevation = 1;
      }
      /**
       * Fire an event that includes the eventName of what was just pressed.
       */

    }, {
      key: "_fireEvent",
      value: function _fireEvent(e) {
        var content = "";

        for (var i = 0; i < this.blox.length; i++) {
          var node = window.HaxStore.haxElementToNode(this.blox[i].tag, this.blox[i].content, this.blox[i].properties);
          content += window.HaxStore.nodeToContent(node);
        } // generate a hax element


        var blox = {
          tag: "grid-plate",
          properties: {
            layout: this.layout
          },
          content: content
        };
        this.dispatchEvent(new CustomEvent("hax-insert-content", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: blox
        }));
        window.HaxStore.instance.haxBloxPicker.close();
      }
    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_d19b63a0aa5311e9b999174ecd579fc6());
      }
    }, {
      key: "tag",
      get: function get() {
        return "hax-blox-browser-item";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Title
           */
          title: {
            type: String
          },

          /**
           * Index position in the original list of imports
           */
          bloxReference: {
            type: Object
          },

          /**
           * icon for the button, optional.
           */
          icon: {
            type: String
          },

          /**
           * Author related to this gizmo
           */
          author: {
            type: String
          },

          /**
           * Description for this.
           */
          description: {
            type: String
          },

          /**
           * Examples, optional.
           */
          examples: {
            type: Array
          },

          /**
           * Status, whether disabled, enabled, or other future states.
           */
          status: {
            type: Array
          },

          /**
           * Layout string to use
           */
          layout: {
            type: String
          },

          /**
           * Tag
           */
          tag: {
            type: String
          },

          /**
           * Elevation off the UI
           */
          elevation: {
            type: Number,
            value: 1,
            reflectToAttribute: true
          }
        };
      }
    }]);
    return HaxBloxBrowserItem;
  }(_polymerElement.PolymerElement);

  _exports.HaxBloxBrowserItem = HaxBloxBrowserItem;
  window.customElements.define(HaxBloxBrowserItem.tag, HaxBloxBrowserItem);
});
