// @lrnwebcomponents/simple-colors/lib/simple-colors-polymer.js
define(["exports","../../../@polymer/polymer/polymer-element.js","../../simple-colors-shared-styles/simple-colors-shared-styles.js","./simple-colors-shared-styles-polymer.js"],function(_exports,_polymerElement,_simpleColorsSharedStyles,_simpleColorsSharedStylesPolymer){"use strict";Object.defineProperty(_exports,"__esModule",{value:!0});_exports.SimpleColorsPolymer=void 0;function _templateObject_2ab560e07a0011eab074ed896f61b288(){var data=babelHelpers.taggedTemplateLiteral(["\n      <style include=\"simple-colors-shared-styles-polymer\"></style>\n      <slot></slot>\n    "]);_templateObject_2ab560e07a0011eab074ed896f61b288=function _templateObject_2ab560e07a0011eab074ed896f61b288(){return data};return data}/**
 * `simple-colors-polymer`
 * shared set of styles for Polymer @lrnwebcomponents
 *
 * @polymer
 * @element simple-colors-polymer
 */var SimpleColorsPolymer=/*#__PURE__*/function(_PolymerElement){babelHelpers.inherits(SimpleColorsPolymer,_PolymerElement);babelHelpers.createClass(SimpleColorsPolymer,null,[{key:"template",// render function
get:function get(){return(0,_polymerElement.html)(_templateObject_2ab560e07a0011eab074ed896f61b288())}// properties available to the custom element for data binding
},{key:"properties",get:function get(){return{/**
       * a selected accent-color: grey, red, pink, purple, etc.
       */accentColor:{name:"accentColor",type:String,value:"grey",reflectToAttribute:!0,notify:!0},/**
       * make the default theme dark?
       */dark:{name:"dark",type:Boolean,value:!1,reflectToAttribute:!0,notify:!0},/**
       * make the default theme dark?
       */colors:{name:"colors",type:Object,value:window.SimpleColorsSharedStyles.colors,notify:!0}}}},{key:"tag",get:function get(){return"simple-colors-polymer"}}]);function SimpleColorsPolymer(){var _this;babelHelpers.classCallCheck(this,SimpleColorsPolymer);_this=babelHelpers.possibleConstructorReturn(this,babelHelpers.getPrototypeOf(SimpleColorsPolymer).call(this));_this.__utils=window.SimpleColorsSharedStyles.requestAvailability();_this.colors=window.SimpleColorsSharedStyles.colors;return _this}/**
   * gets the current shade
   *
   * @param {string} the shade
   * @param {number} the inverted shade
   */babelHelpers.createClass(SimpleColorsPolymer,[{key:"invertShade",value:function invertShade(shade){return this.__utils.invertShade(shade)}/**
   * gets the color information of a given CSS variable or class
   *
   * @param {string} the CSS variable (eg. `--simple-colors-fixed-theme-red-3`) or a class (eg. `.simple-colors-fixed-theme-red-3-text`)
   * @param {object} an object that includes the theme, color, and shade information
   */},{key:"getColorInfo",value:function getColorInfo(colorName){return this.__utils.getColorInfo(colorName)}/**
   * returns a variable based on color name, shade, and fixed theme
   *
   * @param {string} the color name
   * @param {number} the color shade
   * @param {boolean} the color shade
   * @returns {string} the CSS Variable
   */},{key:"makeVariable",value:function makeVariable(){var color=0<arguments.length&&arguments[0]!==void 0?arguments[0]:"grey",shade=1<arguments.length&&arguments[1]!==void 0?arguments[1]:1,theme=2<arguments.length&&arguments[2]!==void 0?arguments[2]:"default";return this.__utils.makeVariable(color="grey",shade=1,theme="default")}/**
   * for large or small text given a color and its shade,
   * lists all the colors and shades that would be
   * WCAG 2.0 AA-compliant for contrast
   *
   * @param {boolean} large text? >= 18pt || (bold && >= 14pt)
   * @param {string} color name, e.g. "deep-purple"
   * @param {string} color shade, e.g. 3
   * @param {object} all of the WCAG 2.0 AA-compliant colors and shades
   */},{key:"getContrastingColors",value:function getContrastingColors(colorName,colorShade,isLarge){return this.__utils.getContrastingColors(colorName,colorShade,isLarge)}/**
   * for large or small text given a color and its shade,
   * lists all the shades of another color that would be
   * WCAG 2.0 AA-compliant for contrast
   *
   * @param {boolean} large text? >= 18pt || (bold && >= 14pt)
   * @param {string} color name, e.g. "deep-purple"
   * @param {string} color shade, e.g. 3
   * @param {string} contrasting color name, e.g. "grey"
   * @param {array} all of the WCAG 2.0 AA-compliant shades of the contrasting color
   */},{key:"getContrastingShades",value:function getContrastingShades(isLarge,colorName,colorShade,contrastName){return this.__utils.getContrastingShades(isLarge,colorName,colorShade,contrastName)}/**
   * determines if two shades are WCAG 2.0 AA-compliant for contrast
   *
   * @param {boolean} large text? >= 18pt || (bold && >= 14pt)
   * @param {string} color name, e.g. "deep-purple"
   * @param {string} color shade, e.g. 3
   * @param {string} contrasting color name, e.g. "grey"
   * @param {string} contrast shade, e.g. 12
   * @param {boolean} whether or not the contrasting shade is WCAG 2.0 AA-compliant
   */},{key:"isContrastCompliant",value:function isContrastCompliant(isLarge,colorName,colorShade,contrastName,contrastShade){return this.__utils.isContrastCompliant(isLarge,colorName,colorShade,contrastName,contrastShade)}}]);return SimpleColorsPolymer}(_polymerElement.PolymerElement);_exports.SimpleColorsPolymer=SimpleColorsPolymer;window.customElements.define(SimpleColorsPolymer.tag,SimpleColorsPolymer)});
