// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-submission/lrnapp-studio-submission-edit-image.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@polymer/iron-image/iron-image.js", "./lrnapp-studio-submission-media-editoverlay.js"], function (_exports, _litElement, _ironImage, _lrnappStudioSubmissionMediaEditoverlay) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioSubmissionEditImage = void 0;

  function _templateObject2_fa9533d02c1c11eaacb4bf54e86db688() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: inline-flex;\n          justify-content: space-around;\n          min-height: 100px;\n          position: relative;\n        }\n\n        iron-image {\n          --iron-image-height: 200px;\n          display: block;\n        }\n      "]);

    _templateObject2_fa9533d02c1c11eaacb4bf54e86db688 = function _templateObject2_fa9533d02c1c11eaacb4bf54e86db688() {
      return data;
    };

    return data;
  }

  function _templateObject_fa9533d02c1c11eaacb4bf54e86db688() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <lrnapp-studio-submission-media-editoverlay\n        data-index=\"", "\"\n        embedcode=\"", "\"\n      >\n        <iron-image\n          src=\"", "\"\n          style=\"width:200px; height:200px; background-color: lightgray;\"\n          sizing=\"contain\"\n          class=\"contain\"\n          preload=\"\"\n          fade=\"\"\n        ></iron-image>\n      </lrnapp-studio-submission-media-editoverlay>\n    "]);

    _templateObject_fa9533d02c1c11eaacb4bf54e86db688 = function _templateObject_fa9533d02c1c11eaacb4bf54e86db688() {
      return data;
    };

    return data;
  }

  var LrnappStudioSubmissionEditImage =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnappStudioSubmissionEditImage, _LitElement);

    function LrnappStudioSubmissionEditImage() {
      babelHelpers.classCallCheck(this, LrnappStudioSubmissionEditImage);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappStudioSubmissionEditImage).apply(this, arguments));
    }

    babelHelpers.createClass(LrnappStudioSubmissionEditImage, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_fa9533d02c1c11eaacb4bf54e86db688(), this.index, this.embedcode, this.thumbnail);
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "image") {
            _this.embedcode = _this._computeEmbedCode(_this.image);
            _this.thumbnail = _this._getThumbnail(_this.image); // notify

            _this.dispatchEvent(new CustomEvent("image-changed", {
              detail: {
                value: _this[propName]
              }
            }));
          }

          if (propName == "thumbnail") {
            // notify
            _this.dispatchEvent(new CustomEvent("thumbnail-changed", {
              detail: {
                value: _this[propName]
              }
            }));
          }
        });
      }
    }, {
      key: "_getThumbnail",
      value: function _getThumbnail(image) {
        return image.url;
      }
    }, {
      key: "_computeEmbedCode",
      value: function _computeEmbedCode(image) {
        return "![Alternative Text Here](" + image.url + ")";
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_fa9533d02c1c11eaacb4bf54e86db688())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrnapp-studio-submission-edit-image";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          index: {
            type: String
          },
          image: {
            type: Object
          },
          thumbnail: {
            type: String
          },
          embedcode: {
            type: String
          }
        };
      }
    }]);
    return LrnappStudioSubmissionEditImage;
  }(_litElement.LitElement);

  _exports.LrnappStudioSubmissionEditImage = LrnappStudioSubmissionEditImage;
  window.customElements.define(LrnappStudioSubmissionEditImage.tag, LrnappStudioSubmissionEditImage);
});
