// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-submission/lrnapp-studio-submission-critique-panel.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@polymer/marked-element/marked-element.js", "../../../lrnsys-button/lrnsys-button.js", "./lrnapp-studio-block.js", "./lrnapp-studio-submission-edit-textarea.js"], function (_exports, _litElement, _markedElement, _lrnsysButton, _lrnappStudioBlock, _lrnappStudioSubmissionEditTextarea) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioSubmissionCritiquePanel = void 0;

  function _templateObject5_e9c0cce02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        .field {\n          padding-top: 32px;\n          padding-bottom: 32px;\n        }\n\n        .actions {\n          display: flex;\n          border-top: 2px solid gainsboro;\n          margin-top: 16px;\n          font-size: 14px;\n        }\n\n        .actions .spacer {\n          flex: 1 1 auto;\n        }\n\n        .submission-critique-outline + .submission-critique-outline {\n          margin-top: 16px;\n        }\n\n        :host([edit])\n          .submission-critique-outline--feedback\n          lrnapp-studio-block {\n          padding: 0;\n        }\n      "]);

    _templateObject5_e9c0cce02da211eaa247b36e11306899 = function _templateObject5_e9c0cce02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject4_e9c0cce02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <div class=\"actions\">\n              <lrnsys-button\n                id=\"publish\"\n                icon=\"check\"\n                label=\"Publish Feedback\"\n                @click=\"", "\"\n                hover-class=\"amber lighten-5 green-text text-darken-4\"\n                icon-class=\"green-text\"\n              ></lrnsys-button>\n              <lrnsys-button\n                id=\"save-draft\"\n                icon=\"drafts\"\n                label=\"Save Draft\"\n                @click=\"", "\"\n                hover-class=\"amber lighten-5 amber-text text-darken-4\"\n                icon-class=\"amber-text text-darken-4\"\n              ></lrnsys-button>\n              <span class=\"spacer\"></span>\n              <lrnsys-button\n                id=\"delete\"\n                label=\"Delete Feedback\"\n                icon=\"delete\"\n                @click=\"", "\"\n                hover-class=\"amber lighten-5 red-text\"\n                icon-class=\"red-text text-darken-4\"\n              >\n              </lrnsys-button>\n            </div>\n          "]);

    _templateObject4_e9c0cce02da211eaa247b36e11306899 = function _templateObject4_e9c0cce02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject3_e9c0cce02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <marked-element\n                    markdown=\"", "\"\n                    @markdown-changed=\"", "\"\n                  ></marked-element>\n                "]);

    _templateObject3_e9c0cce02da211eaa247b36e11306899 = function _templateObject3_e9c0cce02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject2_e9c0cce02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <lrnapp-studio-submission-edit-textarea\n                    content=\"", "\"\n                    @content-changed=\"", "\"\n                  ></lrnapp-studio-submission-edit-textarea>\n                "]);

    _templateObject2_e9c0cce02da211eaa247b36e11306899 = function _templateObject2_e9c0cce02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject_e9c0cce02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <!-- Body -->\n      <div class=\"field\">\n        <div class=\"submission-critique-outline\">\n          <lrnapp-studio-block title=\"Directions\" icon=\"info-outline\">\n            <marked-element\n              markdown=\"", "\"\n            ></marked-element>\n          </lrnapp-studio-block>\n        </div>\n        <div\n          class=\"submission-critique-outline submission-critique-outline--feedback\"\n        >\n          <lrnapp-studio-block title=\"Feedback\" icon=\"communication:comment\">\n            ", "\n          </lrnapp-studio-block>\n        </div>\n      </div>\n      ", "\n    "]);

    _templateObject_e9c0cce02da211eaa247b36e11306899 = function _templateObject_e9c0cce02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  var LrnappStudioSubmissionCritiquePanel =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnappStudioSubmissionCritiquePanel, _LitElement);
    babelHelpers.createClass(LrnappStudioSubmissionCritiquePanel, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_e9c0cce02da211eaa247b36e11306899(), this.submission.relationships.assignment.attributes.critiqueOutline, this.edit ? (0, _litElement.html)(_templateObject2_e9c0cce02da211eaa247b36e11306899(), this.submission.attributes.body, this.submissionAttributesBody) : (0, _litElement.html)(_templateObject3_e9c0cce02da211eaa247b36e11306899(), this.submission.attributes.body, this.submissionAttributesBody), this.edit ? (0, _litElement.html)(_templateObject4_e9c0cce02da211eaa247b36e11306899(), this._publishClicked, this._saveDraftClicked, this._deleteClicked) : "");
      }
    }, {
      key: "submissionAttributesBody",
      value: function submissionAttributesBody(e) {
        var attr = this.submission.attributes;
        attr.body = e.detail.value;
        this.submission.attributes = _objectSpread({}, attr);
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject5_e9c0cce02da211eaa247b36e11306899())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrnapp-studio-submission-critique-panel";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          submission: {
            type: Object
          },
          edit: {
            type: Boolean,
            reflect: true
          }
        };
      }
    }]);

    function LrnappStudioSubmissionCritiquePanel() {
      var _this;

      babelHelpers.classCallCheck(this, LrnappStudioSubmissionCritiquePanel);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappStudioSubmissionCritiquePanel).call(this));
      _this.edit = false;
      return _this;
    }

    babelHelpers.createClass(LrnappStudioSubmissionCritiquePanel, [{
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "submission") {
            // notify
            _this2.dispatchEvent(new CustomEvent("submission-changed", {
              detail: {
                value: _this2[propName]
              }
            }));
          }
        });
      }
    }, {
      key: "_publishClicked",
      value: function _publishClicked(e) {
        var attr = this.submission.attributes;
        attr.state = "submission_ready";
        this.submission.attributes = _objectSpread({}, attr);
        this.dispatchEvent(new CustomEvent("submissionPublishClicked", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: true
        }));
      }
    }, {
      key: "_saveDraftClicked",
      value: function _saveDraftClicked(e) {
        var attr = this.submission.attributes;
        attr.state = "submission_in_progress";
        this.submission.attributes = _objectSpread({}, attr);
        this.dispatchEvent(new CustomEvent("submissionSaveDraftClicked", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: true
        }));
      }
    }, {
      key: "_deleteClicked",
      value: function _deleteClicked(e) {
        this.dispatchEvent(new CustomEvent("submissionDeleteClicked", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: true
        }));
      }
    }]);
    return LrnappStudioSubmissionCritiquePanel;
  }(_litElement.LitElement);

  _exports.LrnappStudioSubmissionCritiquePanel = LrnappStudioSubmissionCritiquePanel;
  window.customElements.define(LrnappStudioSubmissionCritiquePanel.tag, LrnappStudioSubmissionCritiquePanel);
});
