// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-submission/lrnapp-studio-submission-comment.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@polymer/paper-card/paper-card.js", "../../../../@polymer/paper-icon-button/paper-icon-button.js", "../../../../@polymer/paper-tooltip/paper-tooltip.js", "../../../word-count/word-count.js", "../../../lrndesign-avatar/lrndesign-avatar.js"], function (_exports, _litElement, _paperCard, _paperIconButton, _paperTooltip, _wordCount, _lrndesignAvatar) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioSubmissionComment = void 0;

  function _templateObject2_04c8b5e01bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: flex;\n        }\n        paper-card {\n          margin: 20px;\n          padding: 16px;\n        }\n        h1,\n        h2,\n        h3,\n        h4 {\n          text-align: left;\n        }\n        p {\n          font-size: 14px;\n          line-height: 18px;\n          text-align: left;\n        }\n        .nowrap p {\n          white-space: nowrap;\n          overflow: hidden;\n          text-overflow: ellipsis;\n        }\n        .right-actions {\n          float: right;\n        }\n        .paper-card-length-1 {\n          width: 85%;\n          margin-bottom: 8px;\n        }\n        .paper-card-length-2 {\n          width: 81%;\n          margin-top: 0;\n          margin-bottom: 8px;\n        }\n        .paper-card-length-3 {\n          width: 77%;\n          margin-top: 0;\n          margin-bottom: 8px;\n        }\n        .comment-depth-2 {\n          margin: 8px;\n        }\n        .comment-depth-3 {\n          margin: 16px;\n        }\n        .comment-depth-4 {\n          margin: 16px;\n        }\n        .comment-depth-5 {\n          margin: 16px;\n        }\n        .comment-depth-6 {\n          margin: 16px;\n        }\n        .center {\n          padding: 0;\n        }\n      "]);

    _templateObject2_04c8b5e01bb311eabd7cc30119585ddb = function _templateObject2_04c8b5e01bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject_04c8b5e01bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"center comment-depth-", "\">\n        <lrndesign-avatar\n          label=\"", "\"\n          class=\"float-left\"\n        ></lrndesign-avatar>\n      </div>\n      <paper-card\n        class=\"paper-card-length-", "\"\n      >\n        <div id=\"body\" class=\"comment-body nowrap\">\n          <h4>\n            ", "\n            <span class=\"grey-said\"> said:</span>\n          </h4>\n          <word-count>", "</word-count>\n        </div>\n        <div class=\"card-actions\">\n          <paper-icon-button\n            class=\"right-actions\"\n            id=\"reply\"\n            icon=\"icons:reply\"\n          ></paper-icon-button>\n          <paper-tooltip for=\"reply\" animation-delay=\"0\">Reply</paper-tooltip>\n          <paper-icon-button\n            class=\"right-actions\"\n            id=\"edit\"\n            icon=\"icons:create\"\n          ></paper-icon-button>\n          <paper-tooltip for=\"edit\" animation-delay=\"0\">Edit</paper-tooltip>\n          <paper-icon-button\n            class=\"right-actions\"\n            id=\"delete\"\n            icon=\"icons:delete-forever\"\n          ></paper-icon-button>\n          <paper-tooltip for=\"delete\" animation-delay=\"0\">Delete</paper-tooltip>\n        </div>\n      </paper-card>\n    "]);

    _templateObject_04c8b5e01bb311eabd7cc30119585ddb = function _templateObject_04c8b5e01bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  var LrnappStudioSubmissionComment =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnappStudioSubmissionComment, _LitElement);

    function LrnappStudioSubmissionComment() {
      babelHelpers.classCallCheck(this, LrnappStudioSubmissionComment);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappStudioSubmissionComment).apply(this, arguments));
    }

    babelHelpers.createClass(LrnappStudioSubmissionComment, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_04c8b5e01bb311eabd7cc30119585ddb(), this.comment.attributes.threadDepth, this.comment.relationships.author.data.name, this.comment.attributes.threadDepth, this.comment.relationships.author.data.name, this.comment.attributes.body);
      }
    }, {
      key: "firstUpdated",

      /**
       * attached life cycle
       */
      value: function firstUpdated() {
        var _this = this;

        setTimeout(function () {
          _this.shadowRoot.querySelector("#body").addEventListener("click", function (e) {
            _this.shadowRoot.querySelector("#body").classList.toggle("nowrap");
          });
        }, 0);
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_04c8b5e01bb311eabd7cc30119585ddb())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrnapp-studio-submission-comment";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          comment: {
            type: Object
          }
        };
      }
    }]);
    return LrnappStudioSubmissionComment;
  }(_litElement.LitElement);

  _exports.LrnappStudioSubmissionComment = LrnappStudioSubmissionComment;
  window.customElements.define(LrnappStudioSubmissionComment.tag, LrnappStudioSubmissionComment);
});
