// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-submission/lrnapp-studio-submission-edit-file.js
define(["exports", "../../../../lit-element/lit-element.js", "./lrnapp-studio-submission-media-editoverlay.js"], function (_exports, _litElement, _lrnappStudioSubmissionMediaEditoverlay) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioSubmissionEditFile = void 0;

  function _templateObject2_6d7a8df02d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: inline-flex;\n          justify-content: space-around;\n          align-items: stretch;\n          min-height: 100px;\n          width: 100px;\n          position: relative;\n          background: lightgray;\n          color: #333333;\n        }\n\n        lrnapp-studio-submission-media-editoverlay {\n          display: flex;\n        }\n\n        .file_url {\n          margin: 1em;\n        }\n      "]);

    _templateObject2_6d7a8df02d8511ea80ccf124101da415 = function _templateObject2_6d7a8df02d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject_6d7a8df02d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <lrnapp-studio-submission-media-editoverlay\n        data-index=\"", "\"\n        embedcode=\"", "\"\n      >\n        <div class=\"file_url\">", "</div>\n      </lrnapp-studio-submission-media-editoverlay>\n    "]);

    _templateObject_6d7a8df02d8511ea80ccf124101da415 = function _templateObject_6d7a8df02d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  var LrnappStudioSubmissionEditFile =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnappStudioSubmissionEditFile, _LitElement);

    function LrnappStudioSubmissionEditFile() {
      babelHelpers.classCallCheck(this, LrnappStudioSubmissionEditFile);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappStudioSubmissionEditFile).apply(this, arguments));
    }

    babelHelpers.createClass(LrnappStudioSubmissionEditFile, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_6d7a8df02d8511ea80ccf124101da415(), this.index, this.embedcode, this.file.filename);
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "file") {
            _this.embedcode = _this._computeEmbedCode(_this[propName]);
          }
        });
      }
    }, {
      key: "_computeEmbedCode",
      value: function _computeEmbedCode(file) {
        return "[Alternative Text Here](" + file.url + ")";
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_6d7a8df02d8511ea80ccf124101da415())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrnapp-studio-submission-edit-file";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          file: {
            type: Object
          },
          index: {
            type: String
          },
          embedcode: {
            type: String
          }
        };
      }
    }]);
    return LrnappStudioSubmissionEditFile;
  }(_litElement.LitElement);

  _exports.LrnappStudioSubmissionEditFile = LrnappStudioSubmissionEditFile;
  window.customElements.define(LrnappStudioSubmissionEditFile.tag, LrnappStudioSubmissionEditFile);
});
