// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-submission/lrnapp-studio-submission-edit.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@polymer/paper-input/paper-input.js", "../../../lrnsys-button/lrnsys-button.js", "./lrnapp-studio-submission-media-editoverlay.js", "./lrnapp-studio-submission-edit-images.js", "./lrnapp-studio-submission-edit-files.js", "./lrnapp-studio-submission-edit-video.js", "./lrnapp-studio-submission-edit-links.js", "./lrnapp-studio-submission-edit-textarea.js"], function (_exports, _litElement, _paperInput, _lrnsysButton, _lrnappStudioSubmissionMediaEditoverlay, _lrnappStudioSubmissionEditImages, _lrnappStudioSubmissionEditFiles, _lrnappStudioSubmissionEditVideo, _lrnappStudioSubmissionEditLinks, _lrnappStudioSubmissionEditTextarea) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioSubmissionEdit = void 0;

  function _templateObject3_05fcc9101bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          padding: 16px;\n        }\n        .field {\n          padding-bottom: 16px;\n        }\n        .actions {\n          display: flex;\n          border-top: 2px solid gainsboro;\n          margin-top: 16px;\n        }\n        .actions .spacer {\n          flex: 1 1 auto;\n        }\n        iron-icon {\n          margin-right: 6px;\n          --iron-icon-height: 32px;\n          --iron-icon-width: 32px;\n        }\n        .imagesfiled__image_delete {\n          position: absolute;\n          top: 0;\n          right: 0;\n        }\n      "]);

    _templateObject3_05fcc9101bb311eabd7cc30119585ddb = function _templateObject3_05fcc9101bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject2_05fcc9101bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <div class=\"field\">\n              <paper-input\n                label=\"Title\"\n                value=\"", "\"\n                @value-changed=\"", "\"\n              ></paper-input>\n            </div>\n            <!-- Body -->\n            <div class=\"field\">\n              <label>Submission Text</label>\n              <lrnapp-studio-submission-edit-textarea\n                content=\"", "\"\n                @content-changed=\"", "\"\n              ></lrnapp-studio-submission-edit-textarea>\n            </div>\n\n            <!-- Images -->\n            <div class=\"imagesfield field\">\n              <label for=\"image-upload\">Images</label>\n              <lrnapp-studio-submission-edit-images\n                .images=\"", "\"\n                @images-changed=\"", "\"\n                file-types=\"", "\"\n              ></lrnapp-studio-submission-edit-images>\n            </div>\n\n            <!-- Files -->\n            <div class=\"filesfield field\">\n              <label for=\"file-upload\">Files</label>\n              <lrnapp-studio-submission-edit-files\n                .files=\"", "\"\n                @files-changed=\"", "\"\n                file-types=\"", "\"\n              >\n              </lrnapp-studio-submission-edit-files>\n            </div>\n\n            <!-- Links -->\n            <div id=\"linksfield\" class=\"linksfield field\">\n              <label for=\"links-input\">Links</label>\n              <lrnapp-studio-submission-edit-links\n                .links=\"", "\"\n                @links-changed=\"", "\"\n              ></lrnapp-studio-submission-edit-links>\n            </div>\n\n            <!-- Videos -->\n            <div id=\"videosfield\" class=\"videosfield field\">\n              <label for=\"videos-input\">Videos</label>\n              <lrnapp-studio-submission-edit-video\n                .videos=\"", "\"\n                @videos-changed=\"", "\"\n                end-point=\"", "\"\n                csrf-token=\"", "\"\n              ></lrnapp-studio-submission-edit-video>\n            </div>\n\n            <div class=\"actions\">\n              <lrnsys-button\n                id=\"publish\"\n                icon=\"check\"\n                label=\"Publish to Studio\"\n                @click=\"", "\"\n                hover-class=\"amber lighten-5 green-text text-darken-4\"\n                icon-class=\"green-text\"\n              ></lrnsys-button>\n              <lrnsys-button\n                id=\"save-draft\"\n                icon=\"drafts\"\n                label=\"Save Draft\"\n                @click=\"", "\"\n                hover-class=\"amber lighten-5 amber-text text-darken-4\"\n                icon-class=\"amber-text text-darken-4\"\n              ></lrnsys-button>\n              <span class=\"spacer\"></span>\n              <lrnsys-button\n                id=\"delete\"\n                label=\"Delete\"\n                icon=\"delete\"\n                @click=\"", "\"\n                hover-class=\"amber lighten-5 red-text\"\n                icon-class=\"red-text text-darken-4\"\n              >\n              </lrnsys-button>\n            </div>\n          "]);

    _templateObject2_05fcc9101bb311eabd7cc30119585ddb = function _templateObject2_05fcc9101bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject_05fcc9101bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      ", "\n    "]);

    _templateObject_05fcc9101bb311eabd7cc30119585ddb = function _templateObject_05fcc9101bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  var LrnappStudioSubmissionEdit =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnappStudioSubmissionEdit, _LitElement);

    function LrnappStudioSubmissionEdit() {
      babelHelpers.classCallCheck(this, LrnappStudioSubmissionEdit);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappStudioSubmissionEdit).apply(this, arguments));
    }

    babelHelpers.createClass(LrnappStudioSubmissionEdit, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_05fcc9101bb311eabd7cc30119585ddb(), this.submission ? (0, _litElement.html)(_templateObject2_05fcc9101bb311eabd7cc30119585ddb(), this.submission.attributes.title, this.titleEvent, this.submission.attributes.body, this.contentEvent, this.submission.attributes.images, this.imagesEvent, this.submission.meta.imagefieldTypes, this.submission.attributes.files, this.filesEvent, this.submission.meta.filefieldTypes, this.submission.attributes.links, this.linksEvent, this.submission.attributes.video, this.submissionVideosChanged, this.endPoint, this.csrfToken, this._publishClicked, this._saveDraftClicked, this._deleteClicked) : "");
      }
    }, {
      key: "titleEvent",
      value: function titleEvent(e) {
        this.submission.attributes.title = e.detail.value;
      }
    }, {
      key: "contentEvent",
      value: function contentEvent(e) {
        this.submission.attributes.body = e.detail.value;
      }
    }, {
      key: "imagesEvent",
      value: function imagesEvent(e) {
        this.submission.attributes.images = babelHelpers.toConsumableArray(e.detail.value);
      }
    }, {
      key: "filesEvent",
      value: function filesEvent(e) {
        this.submission.attributes.files = babelHelpers.toConsumableArray(e.detail.value);
      }
    }, {
      key: "linksEvent",
      value: function linksEvent(e) {
        this.submission.attributes.links = babelHelpers.toConsumableArray(e.detail.value);
      }
    }, {
      key: "submissionVideosChanged",
      value: function submissionVideosChanged(e) {
        this.submission.attributes.video = babelHelpers.toConsumableArray(e.detail.value);
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "title") {
            _this._bodyChanged(_this[propName]);
          }

          if (propName == "submission") {
            // notify
            _this.dispatchEvent(new CustomEvent("submission-changed", {
              detail: {
                value: _this[propName]
              }
            }));
          }
        });
      }
    }, {
      key: "_publishClicked",
      value: function _publishClicked(e) {
        this.submission.attributes.state = "submission_ready";
        this.submission.attributes = babelHelpers.objectSpread({}, this.submission.attributes);
        this.dispatchEvent(new CustomEvent("submissionPublishClicked", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: true
        }));
      }
    }, {
      key: "_saveDraftClicked",
      value: function _saveDraftClicked(e) {
        this.submission.attributes.state = "submission_in_progress";
        this.submission.attributes = babelHelpers.objectSpread({}, this.submission.attributes);
        this.dispatchEvent(new CustomEvent("submissionSaveDraftClicked", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: true
        }));
      }
    }, {
      key: "_deleteClicked",
      value: function _deleteClicked(e) {
        this.dispatchEvent(new CustomEvent("submissionDeleteClicked", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: true
        }));
      }
      /**
       * Simple way to convert from object to array.
       */

    }, {
      key: "_toArray",
      value: function _toArray(obj) {
        if (obj == null) {
          return [];
        }

        return Object.keys(obj).map(function (key) {
          return obj[key];
        });
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject3_05fcc9101bb311eabd7cc30119585ddb())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrnapp-studio-submission-edit";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          submission: {
            type: Object
          },
          uploadFilesUrl: {
            type: String,
            attribute: "upload-files-url"
          },
          newlink: {
            type: String
          },
          newvideo: {
            type: String
          },
          videoGenerateSourceUrl: {
            type: String,
            attribute: "video-generate-source-url"
          }
        };
      }
    }]);
    return LrnappStudioSubmissionEdit;
  }(_litElement.LitElement);

  _exports.LrnappStudioSubmissionEdit = LrnappStudioSubmissionEdit;
  window.customElements.define(LrnappStudioSubmissionEdit.tag, LrnappStudioSubmissionEdit);
});
