// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-submission/lrnapp-studio-submission-page.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@polymer/app-route/app-route.js", "../../../../@polymer/iron-ajax/iron-ajax.js", "../../../../@polymer/app-layout/app-toolbar/app-toolbar.js", "../../../../@polymer/marked-element/marked-element.js", "../../../../@polymer/iron-icon/iron-icon.js", "../../../../@polymer/paper-dialog/paper-dialog.js", "../../../../@polymer/paper-button/paper-button.js", "../../../../@polymer/paper-badge/paper-badge.js", "../../../../@polymer/paper-toast/paper-toast.js", "../../../../@vaadin/vaadin-split-layout/vaadin-split-layout.js", "../../../lrnsys-layout/lib/lrnsys-dialog.js", "../../../lrnsys-button/lrnsys-button.js", "../../../lrnsys-comment/lib/lrnsys-comment-list.js", "../../../elmsln-loading/elmsln-loading.js", "./lrnapp-studio-submission-object.js", "./lrnapp-studio-submission-comments.js", "./lrnapp-studio-submission-comment.js", "../../../../@polymer/polymer/lib/elements/custom-style.js"], function (_exports, _litElement, _appRoute, _ironAjax, _appToolbar, _markedElement, _ironIcon, _paperDialog, _paperButton, _paperBadge, _paperToast, _vaadinSplitLayout, _lrnsysDialog, _lrnsysButton, _lrnsysCommentList, _elmslnLoading, _lrnappStudioSubmissionObject, _lrnappStudioSubmissionComments, _lrnappStudioSubmissionComment, _customStyle) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioSubmissionPage = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject5_eaf925d02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          position: relative;\n        }\n\n        p {\n          font-size: 14px;\n          line-height: 18px;\n        }\n\n        h1 {\n          margin: 0;\n          text-align: left;\n        }\n\n        iron-selector {\n          line-height: 1em;\n        }\n\n        iron-selector lrnsys-button {\n          display: inline-flex;\n        }\n\n        iron-image {\n          margin: 0 0.5em;\n        }\n\n        lrnsys-dialog {\n          display: inline;\n        }\n\n        .contain {\n          width: 10em;\n          height: 10em;\n          background: #ddd;\n        }\n\n        .center {\n          margin: auto;\n          width: 100%;\n        }\n\n        .title {\n          color: #222;\n          font-size: 2rem;\n          font-weight: 600;\n          line-height: 2.5rem;\n          padding: 0.25rem 0;\n          white-space: nowrap;\n          overflow-x: hidden;\n          text-overflow: ellipsis;\n          margin-top: 1rem;\n          text-transform: uppercase;\n          text-align: left;\n        }\n\n        .author {\n          color: #555;\n          font-size: 1rem;\n          font-weight: 500;\n          font-style: normal;\n          line-height: 1rem;\n          padding: 0.25rem 0 0.5rem 0;\n          margin: 0;\n          text-transform: capitalize;\n        }\n\n        .comments {\n          text-align: right;\n          margin: 0;\n          font-size: 1.5em;\n        }\n\n        .submission-page-wrapper {\n          padding: 0;\n        }\n\n        .submission-page {\n          width: 70vw;\n        }\n\n        .submission-page-card {\n          width: 100%;\n          margin: 0;\n          padding: 0 0 2em 1em;\n        }\n\n        .submission-comments {\n          overflow-y: hidden;\n          padding: 0;\n          margin: 0;\n        }\n\n        elmsln-loading {\n          position: absolute;\n          display: none;\n          top: 0;\n          left: 0;\n          width: 100%;\n          height: 100%;\n          background: rgba(255, 255, 255, 0.8);\n          z-index: 10;\n          align-items: flex-end;\n          justify-content: center;\n        }\n\n        :host([saving]) elmsln-loading {\n          display: block;\n        }\n\n        elmsln-loading {\n          top: calc(50% - 5vmax);\n          left: calc(50% - 5vmax);\n          position: fixed;\n          transform-origin: center center;\n          width: 10vmax !important;\n          height: 10vmax !important;\n        }\n\n        iron-pages {\n          width: 100%;\n        }\n\n        paper-dialog {\n          width: 50%;\n          width: 50vmax;\n          padding: 1em;\n          z-index: 99999;\n        }\n\n        lrnapp-studio-submission-object {\n          width: 100%;\n        }\n\n        .assignment-body {\n          padding: 2em;\n        }\n\n        iron-icon {\n          margin: 0.2em;\n        }\n      "]);

    _templateObject5_eaf925d02da211eaa247b36e11306899 = function _templateObject5_eaf925d02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject4_eaf925d02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                <lrnsys-comment-list\n                  comment-ops-base=\"", "\"\n                  csrf-token=\"", "\"\n                  source-path=\"", "\"\n                  create-stub-url=\"", "\"\n                ></lrnsys-comment-list>\n              "]);

    _templateObject4_eaf925d02da211eaa247b36e11306899 = function _templateObject4_eaf925d02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject3_eaf925d02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <lrnsys-button\n                @click=\"", "\"\n                icon=\"arrow-back\"\n                label=\"Back to project management\"\n                hover-class=\"amber darken-4 white-text\"\n              ></lrnsys-button>\n            "]);

    _templateObject3_eaf925d02da211eaa247b36e11306899 = function _templateObject3_eaf925d02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject2_eaf925d02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <lrnsys-button\n                @click=\"", "\"\n                icon=\"arrow-back\"\n                label=\"See in studio\"\n                hover-class=\"amber darken-4 white-text\"\n              ></lrnsys-button>\n            "]);

    _templateObject2_eaf925d02da211eaa247b36e11306899 = function _templateObject2_eaf925d02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject_eaf925d02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <custom-style>\n        <style>\n          .submission-page-wrapper {\n            --vaadin-split-layout-splitter: {\n              min-width: 0.4em;\n              background: var(--paper-amber-200);\n            }\n          }\n        </style>\n      </custom-style>\n      <app-route\n        .route=\"", "\"\n        @route-changed=\"", "\"\n        .tail=\"", "\"\n        @tail-changed=\"", "\"\n        pattern=\"/edit\"\n        .active=\"", "\"\n        @active-changed=\"", "\"\n      >\n      </app-route>\n\n      <iron-ajax\n        id=\"ajaxRequest\"\n        auto=\"\"\n        url=\"", "\"\n        method=\"GET\"\n        params=\"", "\"\n        content-type=\"application/json\"\n        handle-as=\"json\"\n        @response=\"", "\"\n      ></iron-ajax>\n      <!-- Update Submission Node -->\n      <iron-ajax\n        id=\"ajaxUpdateRequest\"\n        url=\"", "\"\n        method=\"PUT\"\n        .body=\"", "\"\n        params=\"", "\"\n        content-type=\"application/json\"\n        handle-as=\"json\"\n        @response=\"", "\"\n      ></iron-ajax>\n      <!-- Delete Submission Node -->\n      <iron-ajax\n        id=\"ajaxDeleteRequest\"\n        url=\"", "\"\n        method=\"DELETE\"\n        params=\"", "\"\n        content-type=\"application/json\"\n        handle-as=\"json\"\n        @response=\"", "\"\n      ></iron-ajax>\n\n      <app-toolbar class=\"amber lighten-3\" ?hidden=\"", "\">\n        ", "\n        <div spacer=\"\" main-title=\"\">", "</div>\n        <div spacer=\"\">\n          <lrnsys-dialog\n            raised\n            header=\"", "\"\n          >\n            <span slot=\"button\"\n              ><iron-icon icon=\"icons:assignment\"></iron-icon>Assignment\n              Details</span\n            >\n            <div class=\"assignment-body\">\n              <marked-element\n                id=\"assignment-body\"\n                markdown=\"", "\"\n              ></marked-element>\n            </div>\n          </lrnsys-dialog>\n        </div>\n        <div>\n          <paper-button\n            id=\"comment-count\"\n            style=\"margin:0;padding:.25em;text-transform:none;\"\n          >\n            <iron-icon icon=\"communication:forum\"></iron-icon>\n            ", " Comments\n          </paper-button>\n          <paper-badge\n            ?hidden=\"", "\"\n            for=\"comment-count\"\n            label=\"", "\"\n          ></paper-badge>\n        </div>\n        <div ?hidden=\"", "\">\n          <lrnsys-button\n            @click=\"", "\"\n            icon=\"create\"\n            label=\"Edit\"\n            hover-class=\"amber darken-4 white-text\"\n            ?hidden=\"", "\"\n          ></lrnsys-button>\n        </div>\n        <div ?hidden=\"", "\">\n          <lrnsys-button\n            @click=\"", "\"\n            icon=\"cancel\"\n            label=\"Cancel\"\n            hover-class=\"amber darken-4 white-text\"\n            ?hidden=\"", "\"\n          ></lrnsys-button>\n        </div>\n      </app-toolbar>\n\n      <vaadin-split-layout class=\"submission-page-wrapper\">\n        <div\n          id=\"commentcolumn\"\n          class=\"submission-comments\"\n          style=\"min-width: 25%; max-width: 40%; width:30%;\"\n        >\n          ", "\n        </div>\n        <div id=\"submissioncolumn\" style=\"width:70%;\">\n          <lrnapp-studio-submission-object\n            .submission=\"", "\"\n            edit=\"", "\"\n          ></lrnapp-studio-submission-object>\n        </div>\n        <iron-icon class=\"splitter-handle\" icon=\"more-vert\"></iron-icon>\n      </vaadin-split-layout>\n\n      <elmsln-loading></elmsln-loading>\n\n      <paper-dialog id=\"deletedialog\">\n        <h2>Delete submission?</h2>\n        <p>Are you sure you want to delete this submission?</p>\n        <div class=\"buttons\">\n          <paper-button dialog-dismiss=\"\">Cancel</paper-button>\n          <paper-button\n            dialog-confirm=\"\"\n            @click=\"", "\"\n            >Delete</paper-button\n          >\n        </div>\n      </paper-dialog>\n      <paper-dialog id=\"publishdialog\">\n        <h2>Ready to publish?</h2>\n        <p>\n          By publishing, the author of this submission will be able to view your\n          feedback. Are you ready to publish?\n        </p>\n        <div class=\"buttons\">\n          <paper-button dialog-dismiss=\"\">Cancel</paper-button>\n          <paper-button\n            dialog-confirm=\"\"\n            @click=\"", "\"\n            >Yes Publish</paper-button\n          >\n        </div>\n      </paper-dialog>\n\n      <paper-toast id=\"toast\"></paper-toast>\n    "]);

    _templateObject_eaf925d02da211eaa247b36e11306899 = function _templateObject_eaf925d02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  var LrnappStudioSubmissionPage =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnappStudioSubmissionPage, _LitElement);
    babelHelpers.createClass(LrnappStudioSubmissionPage, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_eaf925d02da211eaa247b36e11306899(), this.route, this.routeChangedEvent, this.tail, this.tailChangedEvent, this.editPage, this.editPageChangedEvent, this.reqUrl, this.reqParams, this._handleResponse, this.reqUrl, this.submission, this.reqParams, this._handleUpdateResponse, this.reqUrl, this.reqParams, this._handleDeleteResponse, this.hideMenuBar, this.showComments ? (0, _litElement.html)(_templateObject2_eaf925d02da211eaa247b36e11306899(), this._backToStudio) : (0, _litElement.html)(_templateObject3_eaf925d02da211eaa247b36e11306899(), this._backToKanban), this.submission.attributes.title, this.submission.relationships.assignment.attributes.title, this.submission.relationships.assignment.attributes.body, this.submission.meta.comment_count, this.displayNewBadge(this.submission.meta.comment_new), this.submission.meta.comment_new, this.editPage, this._setEditRoute, !this.submission.meta.canUpdate, !this.editPage, this._resetRoute, !this.submission.meta.canUpdate, this.showComments ? (0, _litElement.html)(_templateObject4_eaf925d02da211eaa247b36e11306899(), this.commentOpsBase, this.csrfToken, this.commentsUrl, this.createStubUrl) : "", this.submission, this.editPage, this._submissionDeleteConfirmed, this._submissionPublishConfirmed);
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject5_eaf925d02da211eaa247b36e11306899())];
      }
    }]);

    function LrnappStudioSubmissionPage() {
      var _this;

      babelHelpers.classCallCheck(this, LrnappStudioSubmissionPage);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappStudioSubmissionPage).call(this));
      _this.hideMenuBar = false;
      _this.saving = false;
      setTimeout(function () {
        _this.addEventListener("submissionDeleteClicked", _this._submissionDeleteClicked.bind(babelHelpers.assertThisInitialized(_this)));

        _this.addEventListener("submissionPublishClicked", _this._submissionPublishClicked.bind(babelHelpers.assertThisInitialized(_this)));

        _this.addEventListener("submissionSaveDraftClicked", _this._submissionSaveDraftClicked.bind(babelHelpers.assertThisInitialized(_this)));
      }, 0);
      return _this;
    }

    babelHelpers.createClass(LrnappStudioSubmissionPage, [{
      key: "editPageChangedEvent",
      value: function editPageChangedEvent(e) {
        this.editPage = e.detail.value;
      }
    }, {
      key: "routeChangedEvent",
      value: function routeChangedEvent(e) {
        this.route = _objectSpread({}, e.detail.value);
      }
    }, {
      key: "tailChangedEvent",
      value: function tailChangedEvent(e) {
        this.tail = e.detail.value;
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (["id", "endPoint"].includes(propName)) {
            _this2._urlVarsChanged(_this2.id, _this2.endPoint);
          }

          if (propName == "csrfToken") {
            _this2._paramsChanged(_this2[propName]);
          }

          if (propName == "title") {
            _this2._bodyChanged(_this2[propName]);
          }

          if (propName == "submission") {
            _this2.showComments = _this2._computeShowComments(_this2[propName]);
          }

          if (["id", "endPoint", "csrfToken"].includes(propName)) {
            _this2.commentsUrl = _this2._computeCommentsUrl(_this2.id, _this2.endPoint, _this2.csrfToken);
            _this2.createStubUrl = _this2._computeCommentsStubUrl(_this2.id, _this2.endPoint, _this2.csrfToken);
            _this2.commentOpsBase = _this2._computeCommentsOpsUrl(_this2.id, _this2.endPoint, _this2.csrfToken);
          }

          var notifiedProps = ["saving"];

          if (notifiedProps.includes(propName)) {
            // notify
            var eventName = "".concat(propName.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, "$1-$2").toLowerCase(), "-changed");

            _this2.dispatchEvent(new CustomEvent(eventName, {
              detail: {
                value: _this2[propName]
              }
            }));
          }
        });
      }
      /**
       * Go back to the studio relative to the app's path.
       */

    }, {
      key: "_backToStudio",
      value: function _backToStudio(e) {
        window.location.href = this.basePath + "lrnapp-open-studio";
      }
      /**
       * Go back to the studio relative to the app's path.
       */

    }, {
      key: "_backToKanban",
      value: function _backToKanban(e) {
        window.location.href = this.basePath + "lrnapp-studio-kanban";
      }
      /**
       * Trigger the page router to invoke edit state.
       */

    }, {
      key: "_setEditRoute",
      value: function _setEditRoute(e) {
        this.route.path = "edit";
        this.route = _objectSpread({}, this.route);
      }
      /**
       * Trigger the page router to invoke edit state.
       */

    }, {
      key: "_resetRoute",
      value: function _resetRoute(e) {
        this.route.path = "";
        this.route = _objectSpread({}, this.route);
      }
      /**
       * if we should show new badge based on new comment count.
       */

    }, {
      key: "displayNewBadge",
      value: function displayNewBadge(count) {
        if (count == 0) {
          return true;
        }

        return false;
      }
    }, {
      key: "_computeCommentsUrl",
      value: function _computeCommentsUrl(id, endPoint, csrfToken) {
        return endPoint + "/api/submissions/" + id + "/comments?token=" + csrfToken;
      }
    }, {
      key: "_computeCommentsOpsUrl",
      value: function _computeCommentsOpsUrl(id, endPoint, csrfToken) {
        return endPoint + "/api/submissions/" + id + "/comments";
      }
    }, {
      key: "_computeCommentsStubUrl",
      value: function _computeCommentsStubUrl(id, endPoint, csrfToken) {
        return endPoint + "/api/submissions/" + id + "/comments/create-stub?token=" + csrfToken;
      }
    }, {
      key: "_urlVarsChanged",
      value: function _urlVarsChanged(id, endPoint) {
        this.reqUrl = endPoint + "/api/submissions/" + id;
      }
    }, {
      key: "_paramsChanged",
      value: function _paramsChanged(csrfToken) {
        var params = this.reqParams || {};
        params.token = csrfToken;
        this.reqParams = params;
      }
    }, {
      key: "_handleResponse",
      value: function _handleResponse(data) {
        // empty response means no access or deleted item
        if (data.detail.response.data) {
          this.submission = _objectSpread({}, data.detail.response.data);
          window.dispatchEvent(new Event("resize"));
        }
      }
    }, {
      key: "_isReply",
      value: function _isReply(data) {
        this.thread = submission.relationships.comments.data.attributes.thread;
      }
    }, {
      key: "_submissionSaveDraftClicked",
      value: function _submissionSaveDraftClicked(e) {
        this.saving = true;
        this.shadowRoot.querySelector("#ajaxUpdateRequest").generateRequest();
      }
    }, {
      key: "_submissionPublishClicked",
      value: function _submissionPublishClicked(e) {
        this.shadowRoot.querySelector("#publishdialog").open();
      }
    }, {
      key: "_submissionPublishConfirmed",
      value: function _submissionPublishConfirmed(e) {
        this.saving = true;
        this.shadowRoot.querySelector("#ajaxUpdateRequest").generateRequest();
      }
    }, {
      key: "_submissionDeleteClicked",
      value: function _submissionDeleteClicked(e) {
        this.shadowRoot.querySelector("#deletedialog").open();
      }
    }, {
      key: "_submissionDeleteConfirmed",
      value: function _submissionDeleteConfirmed(e) {
        this.saving = true;
        this.shadowRoot.querySelector("#ajaxDeleteRequest").generateRequest();
      }
    }, {
      key: "_handleUpdateResponse",
      value: function _handleUpdateResponse(res) {
        var status = res.detail.response.status;
        var submission = res.detail.response.data;
        this.saving = false;

        if (status === 200) {
          this.submission = _objectSpread({}, submission);
          var attr = this.route;
          attr.path = "";
          this.route = _objectSpread({}, attr); // display a submission published notification

          if (submission.attributes.state === "submission_ready") {
            this.shadowRoot.querySelector("#toast").show("Published!");
          } // @todo replace this with the page just being there once we fix the lazy load dilema


          window.location.href = this.endPoint + "/submissions/" + submission.id;
        }
      }
    }, {
      key: "_handleDeleteResponse",
      value: function _handleDeleteResponse(res) {
        var root = this;
        var status = res.detail.response.status;
        root.saving = false;

        if (status === 200) {
          window.location.href = this.basePath + "lrnapp-studio-kanban?deletetoast";
        }
      }
    }, {
      key: "_computeShowComments",
      value: function _computeShowComments(submission) {
        try {
          var type = submission.meta.submissionType;

          if (type !== "critique" && submission.attributes.state == "submission_ready") {
            return true;
          }
        } catch (error) {}

        return false;
      }
      /**
       * Simple way to convert from object to array.
       */

    }, {
      key: "_toArray",
      value: function _toArray(obj) {
        if (babelHelpers.typeof(obj) === "object" && obj !== null) {
          return Object.keys(obj).map(function (key) {
            return obj[key];
          });
        }

        return [];
      }
    }], [{
      key: "tag",
      get: function get() {
        return "lrnapp-studio-submission-page";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          id: {
            type: String
          },
          hideMenuBar: {
            type: Boolean
          },
          elmslnCourse: {
            type: String,
            attribute: "elmsln-course"
          },
          elmslnSection: {
            type: String,
            attribute: "elmsln-section"
          },
          basePath: {
            type: String,
            attribute: "base-path"
          },
          csrfToken: {
            type: String,
            attribute: "csrf-token"
          },
          endPoint: {
            type: String,
            attribute: "end-point"
          },
          reqUrl: {
            type: String
          },
          reqParams: {
            type: Object
          },
          submission: {
            type: Object
          },
          commentsUrl: {
            type: String
          },
          createStubUrl: {
            type: String
          },
          commentOpsBase: {
            type: String
          },
          editPage: {
            type: Boolean,
            reflect: true
          },
          saving: {
            type: Boolean,
            reflect: true
          },
          showComments: {
            type: Boolean
          }
        };
      }
    }]);
    return LrnappStudioSubmissionPage;
  }(_litElement.LitElement);

  _exports.LrnappStudioSubmissionPage = LrnappStudioSubmissionPage;
  window.customElements.define(LrnappStudioSubmissionPage.tag, LrnappStudioSubmissionPage);
});
