// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-submission/lrnapp-studio-submission-critique.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@vaadin/vaadin-split-layout/vaadin-split-layout.js", "./lrnapp-studio-submission-media-editoverlay.js", "./lrnapp-studio-submission-edit-images.js", "./lrnapp-studio-submission-edit-files.js", "./lrnapp-studio-submission-edit-video.js", "./lrnapp-studio-submission-edit-links.js", "./lrnapp-studio-submission-display.js", "./lrnapp-studio-submission-edit-textarea.js", "./lrnapp-studio-block.js", "./lrnapp-studio-submission-critique-panel.js"], function (_exports, _litElement, _vaadinSplitLayout, _lrnappStudioSubmissionMediaEditoverlay, _lrnappStudioSubmissionEditImages, _lrnappStudioSubmissionEditFiles, _lrnappStudioSubmissionEditVideo, _lrnappStudioSubmissionEditLinks, _lrnappStudioSubmissionDisplay, _lrnappStudioSubmissionEditTextarea, _lrnappStudioBlock, _lrnappStudioSubmissionCritiquePanel) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioSubmissionCritique = void 0;

  function _templateObject4_f9346b502c1c11eaacb4bf54e86db688() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n\n        .submission-critique {\n          display: flex;\n          justify-content: space-around;\n        }\n\n        .submission-critique-panel {\n          flex: 1 1 auto;\n          width: 100%;\n          flex-wrap: nowrap;\n          margin: 0.3em;\n        }\n      "]);

    _templateObject4_f9346b502c1c11eaacb4bf54e86db688 = function _templateObject4_f9346b502c1c11eaacb4bf54e86db688() {
      return data;
    };

    return data;
  }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject3_f9346b502c1c11eaacb4bf54e86db688() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <vaadin-split-layout class=\"submission-critique\">\n              <!-- critique panel -->\n              <div class=\"submission-critique-panel\" id=\"crititque-panel\">\n                <lrnapp-studio-submission-display\n                  .submission=\"", "\"\n                ></lrnapp-studio-submission-display>\n              </div>\n              <!-- critique panel -->\n              <div class=\"submission-critique-panel\">\n                <lrnapp-studio-submission-critique-panel\n                  .submission=\"", "\"\n                  @submission-changed=\"", "\"\n                  ?edit=\"", "\"\n                ></lrnapp-studio-submission-critique-panel>\n              </div>\n            </vaadin-split-layout>\n          "]);

    _templateObject3_f9346b502c1c11eaacb4bf54e86db688 = function _templateObject3_f9346b502c1c11eaacb4bf54e86db688() {
      return data;
    };

    return data;
  }

  function _templateObject2_f9346b502c1c11eaacb4bf54e86db688() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <div class=\"submission-critique\">\n              <!-- critique panel -->\n              <div class=\"submission-critique-panel\" id=\"crititque-panel\">\n                <lrnapp-studio-submission-display\n                  .submission=\"", "\"\n                ></lrnapp-studio-submission-display>\n              </div>\n              <!-- critique panel -->\n              <div class=\"submission-critique-panel\">\n                <lrnapp-studio-submission-critique-panel\n                  .submission=\"", "\"\n                  ?edit=\"", "\"\n                ></lrnapp-studio-submission-critique-panel>\n              </div>\n            </div>\n          "]);

    _templateObject2_f9346b502c1c11eaacb4bf54e86db688 = function _templateObject2_f9346b502c1c11eaacb4bf54e86db688() {
      return data;
    };

    return data;
  }

  function _templateObject_f9346b502c1c11eaacb4bf54e86db688() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      ", "\n    "]);

    _templateObject_f9346b502c1c11eaacb4bf54e86db688 = function _templateObject_f9346b502c1c11eaacb4bf54e86db688() {
      return data;
    };

    return data;
  }

  var LrnappStudioSubmissionCritique =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnappStudioSubmissionCritique, _LitElement);
    babelHelpers.createClass(LrnappStudioSubmissionCritique, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_f9346b502c1c11eaacb4bf54e86db688(), this.edit ? (0, _litElement.html)(_templateObject2_f9346b502c1c11eaacb4bf54e86db688(), this.submission.relationships.relatedSubmission, this.submission, this.edit) : (0, _litElement.html)(_templateObject3_f9346b502c1c11eaacb4bf54e86db688(), this.submission.relationships.relatedSubmission, this.submission, this.submissionChanged, this.edit));
      }
    }, {
      key: "submissionChanged",
      value: function submissionChanged(e) {
        this.submission = _objectSpread({}, e.detail.value);
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "submission") {
            // notify
            _this2.dispatchEvent(new CustomEvent("submission-changed", {
              detail: {
                value: _this2[propName]
              }
            }));
          }
        });
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject4_f9346b502c1c11eaacb4bf54e86db688())];
      }
    }]);

    function LrnappStudioSubmissionCritique() {
      var _this;

      babelHelpers.classCallCheck(this, LrnappStudioSubmissionCritique);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappStudioSubmissionCritique).call(this));
      _this.edit = false;
      return _this;
    }

    babelHelpers.createClass(LrnappStudioSubmissionCritique, [{
      key: "_toArray",

      /**
       * Simple way to convert from object to array.
       */
      value: function _toArray(obj) {
        if (obj == null) {
          return [];
        }

        return Object.keys(obj).map(function (key) {
          return obj[key];
        });
      }
    }], [{
      key: "tag",
      get: function get() {
        return "lrnapp-studio-submission-critique";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          submission: {
            type: Object
          },
          edit: {
            type: Boolean,
            reflect: true
          }
        };
      }
    }]);
    return LrnappStudioSubmissionCritique;
  }(_litElement.LitElement);

  _exports.LrnappStudioSubmissionCritique = LrnappStudioSubmissionCritique;
  window.customElements.define(LrnappStudioSubmissionCritique.tag, LrnappStudioSubmissionCritique);
});
