// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-submission/lrnapp-studio-submission-display.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@polymer/paper-card/paper-card.js", "../../../../@polymer/marked-element/marked-element.js", "../../../../@polymer/iron-image/iron-image.js", "../../../image-inspector/image-inspector.js", "../../../lrndesign-avatar/lrndesign-avatar.js", "../../../word-count/word-count.js", "../../../lrnsys-button/lrnsys-button.js", "../../../lrndesign-contentblock/lrndesign-contentblock.js", "../../../lrnsys-layout/lib/lrnsys-dialog.js"], function (_exports, _litElement, _paperCard, _markedElement, _ironImage, _imageInspector, _lrndesignAvatar, _wordCount, _lrnsysButton, _lrndesignContentblock, _lrnsysDialog) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioSubmissionDisplay = void 0;

  function _templateObject12_e9ca42c02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        p {\n          font-size: 14px;\n          line-height: 18px;\n        }\n        h1,\n        h2,\n        h3 {\n          margin: 0;\n          text-align: left;\n        }\n        iron-selector {\n          line-height: 1em;\n        }\n        iron-selector lrnsys-button {\n          display: inline-flex;\n        }\n        marked-element {\n          display: block;\n          margin: 0 5em 0 5em;\n        }\n        iron-image {\n          margin: 0 0.5em;\n        }\n        lrnsys-dialog {\n          display: inline;\n        }\n        .contain {\n          width: 10em;\n          height: 10em;\n          background: #ddd;\n        }\n        .center {\n          margin: auto;\n          width: 100%;\n        }\n        .title {\n          color: #222;\n          font-size: 2rem;\n          font-weight: 600;\n          line-height: 2.5rem;\n          padding: 0.25rem 0;\n          white-space: nowrap;\n          overflow-x: hidden;\n          text-overflow: ellipsis;\n          margin-top: 1rem;\n          text-align: center;\n        }\n        .author {\n          color: #555;\n          font-size: 1rem;\n          font-weight: 500;\n          font-style: normal;\n          line-height: 1rem;\n          padding: 0.25rem 0 0.5rem 0;\n          margin: 0;\n          text-transform: capitalize;\n        }\n        .date {\n          color: #555;\n          font-size: 16px;\n          font-style: normal;\n          padding: 0.25rem 0 0.5rem 0;\n          margin: 0;\n        }\n        .flex-wrap {\n          display: flex;\n        }\n        .submission-page-card {\n          width: 100%;\n          margin: 0;\n        }\n        :host([edit-page]) .submission-page {\n          width: 100%;\n        }\n        .image-dialog-header {\n          display: flex;\n        }\n        app-toolbar {\n          width: 36em;\n          background: #ddd;\n        }\n        lrnsys-button {\n          padding: 0.75em;\n        }\n        .top {\n          top: 8em;\n        }\n      "]);

    _templateObject12_e9ca42c02da211eaa247b36e11306899 = function _templateObject12_e9ca42c02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject11_e9ca42c02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                        <p>\n                          <strong>", ":</strong>\n                          <a class=\"link\" href=\"", "\" target=\"_blank\"\n                            >", "</a\n                          >\n                        </p>\n                      "]);

    _templateObject11_e9ca42c02da211eaa247b36e11306899 = function _templateObject11_e9ca42c02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject10_e9ca42c02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <lrndesign-contentblock title=\"Files Submitted\">\n                    ", "\n                  </lrndesign-contentblock>\n                "]);

    _templateObject10_e9ca42c02da211eaa247b36e11306899 = function _templateObject10_e9ca42c02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject9_e9ca42c02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                        <p>\n                          <strong>", ":</strong>\n                          <a class=\"link\" href=\"", "\" target=\"_blank\"\n                            >", "</a\n                          >\n                        </p>\n                      "]);

    _templateObject9_e9ca42c02da211eaa247b36e11306899 = function _templateObject9_e9ca42c02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject8_e9ca42c02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <lrndesign-contentblock>\n                    ", "\n                  </lrndesign-contentblock>\n                "]);

    _templateObject8_e9ca42c02da211eaa247b36e11306899 = function _templateObject8_e9ca42c02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject7_e9ca42c02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                        <iframe\n                          width=\"460\"\n                          height=\"249\"\n                          src=\"", "\"\n                        ></iframe>\n                      "]);

    _templateObject7_e9ca42c02da211eaa247b36e11306899 = function _templateObject7_e9ca42c02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject6_e9ca42c02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <lrndesign-contentblock>\n                    ", "\n                  </lrndesign-contentblock>\n                "]);

    _templateObject6_e9ca42c02da211eaa247b36e11306899 = function _templateObject6_e9ca42c02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject5_e9ca42c02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                                  <image-inspector src=\"", "\">\n                                    <span slot=\"toolbar\">\n                                      <lrnsys-button\n                                        alt=\"Download all images\"\n                                        icon=\"icons:file-download\"\n                                        href=\"", "\"\n                                        target=\"_blank\"\n                                      ></lrnsys-button>\n                                    </span>\n                                  </image-inspector>\n                                "]);

    _templateObject5_e9ca42c02da211eaa247b36e11306899 = function _templateObject5_e9ca42c02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject4_e9ca42c02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                                  <iron-image\n                                    style=\"width:500px; height:500px; background-color: lightgray;\"\n                                    sizing=\"contain\"\n                                    class=\"contain\"\n                                    src=\"", "\"\n                                    preload=\"\"\n                                    fade=\"\"\n                                  ></iron-image>\n                                "]);

    _templateObject4_e9ca42c02da211eaa247b36e11306899 = function _templateObject4_e9ca42c02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject3_e9ca42c02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                        <lrnsys-dialog alt=\"View Image\" dynamic-images=\"\">\n                          <div class=\"image-dialog-header\"></div>\n                          <span slot=\"button\">\n                            <iron-image\n                              style=\"width:200px; height:200px; background-color: lightgray;\"\n                              sizing=\"contain\"\n                              class=\"contain\"\n                              src=\"", "\"\n                              preload=\"\"\n                              fade=\"\"\n                            ></iron-image>\n                            <span>filename: image.url</span>\n                          </span>\n                          <div style=\"text-align: center;\">\n                            ", "\n                          </div>\n                        </lrnsys-dialog>\n                      "]);

    _templateObject3_e9ca42c02da211eaa247b36e11306899 = function _templateObject3_e9ca42c02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject2_e9ca42c02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <lrndesign-contentblock class=\"center\">\n                    ", "\n                  </lrndesign-contentblock>\n                "]);

    _templateObject2_e9ca42c02da211eaa247b36e11306899 = function _templateObject2_e9ca42c02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject_e9ca42c02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"submission-page\">\n        <paper-card class=\"submission-page-card ferpa-protect\">\n          <div class=\"card-content\">\n            <lrndesign-avatar\n              class=\"center\"\n              label=\"", "\"\n              src=\"", "\"\n            ></lrndesign-avatar>\n            <div class=\"author center\">\n              ", "\n            </div>\n            <h2 class=\"title center\">", "</h2>\n            <div class=\"date center\">\n              Created: ", "\n            </div>\n            <div class=\"date center\">\n              Last updated: ", "\n            </div>\n            <!-- ----- Text Submission ----- -->\n            <word-count>\n              <marked-element id=\"markedarea\"> </marked-element>\n            </word-count>\n            <!-- ----- Images Submission ----- -->\n            ", "\n            <!-- ----- Video Submission ----- -->\n            ", "\n            <!-- ----- Links Submission ----- -->\n            ", "\n            <!-- ----- Files Submission ----- -->\n            ", "\n          </div>\n        </paper-card>\n      </div>\n    "]);

    _templateObject_e9ca42c02da211eaa247b36e11306899 = function _templateObject_e9ca42c02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  var LrnappStudioSubmissionDisplay =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnappStudioSubmissionDisplay, _LitElement);

    function LrnappStudioSubmissionDisplay() {
      babelHelpers.classCallCheck(this, LrnappStudioSubmissionDisplay);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappStudioSubmissionDisplay).apply(this, arguments));
    }

    babelHelpers.createClass(LrnappStudioSubmissionDisplay, [{
      key: "render",

      /**
       * LitElement render
       */
      value: function render() {
        var _this = this;

        return (0, _litElement.html)(_templateObject_e9ca42c02da211eaa247b36e11306899(), this.submission.relationships.author.data.name, this.submission.relationships.author.data.avatar, this.submission.relationships.author.data.display_name, this.submission.attributes.title, this.date(this.submission.meta.created), this.date(this.submission.meta.changed), this.submission.attributes.images ? (0, _litElement.html)(_templateObject2_e9ca42c02da211eaa247b36e11306899(), this._toArray(this.submission.attributes.images).map(function (image) {
          return (0, _litElement.html)(_templateObject3_e9ca42c02da211eaa247b36e11306899(), image.thumbnail, _this._isGif(image.url) ? (0, _litElement.html)(_templateObject4_e9ca42c02da211eaa247b36e11306899(), image.url) : (0, _litElement.html)(_templateObject5_e9ca42c02da211eaa247b36e11306899(), image.url, _this.submission.attributes.download_files));
        })) : "", this.submission.attributes.video ? (0, _litElement.html)(_templateObject6_e9ca42c02da211eaa247b36e11306899(), this._toArray(this.submission.attributes.video).map(function (video) {
          return (0, _litElement.html)(_templateObject7_e9ca42c02da211eaa247b36e11306899(), video.video_src);
        })) : "", this.submission.attributes.links ? (0, _litElement.html)(_templateObject8_e9ca42c02da211eaa247b36e11306899(), this._toArray(this.submission.attributes.links).map(function (link) {
          return (0, _litElement.html)(_templateObject9_e9ca42c02da211eaa247b36e11306899(), link.title, link.url, link.url);
        })) : "", this.submission.attributes.files ? (0, _litElement.html)(_templateObject10_e9ca42c02da211eaa247b36e11306899(), this._toArray(this.submission.attributes.files).map(function (file) {
          return (0, _litElement.html)(_templateObject11_e9ca42c02da211eaa247b36e11306899(), file.filename, file.url, file.url);
        })) : "");
      }
    }, {
      key: "updated",

      /**
       * LitElement ready
       */
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "submission") {
            _this2._submissionLoaded(_this2[propName]); // notify


            _this2.dispatchEvent(new CustomEvent("submission-changed", {
              detail: {
                value: _this2[propName]
              }
            }));
          }
        });
      }
    }, {
      key: "_submissionLoaded",
      value: function _submissionLoaded(newValue) {
        if (newValue) {
          if (newValue.attributes && newValue.attributes.body) {
            var mdscript = document.createElement("script");
            mdscript.type = "text/markdown";
            mdscript.innerHTML = newValue.attributes.body;
            this.shadowRoot.querySelector("#markedarea").appendChild(mdscript);
            this.shadowRoot.querySelector("#markedarea").markdown = newValue.attributes.body;
          }
        }
      }
    }, {
      key: "date",
      value: function date(time) {
        if (time) {
          var parts = time.split("-");
          var times = time.split("T");
          var month = parts[2].split("T");
          times = times[1].split("-");
          return parts[1] + "/" + month[0] + " - " + times[0];
        }
      }
      /**
       * Simple way to convert from object to array.
       */

    }, {
      key: "_toArray",
      value: function _toArray(obj) {
        if (babelHelpers.typeof(obj) === "object" && obj !== null) {
          return Object.keys(obj).map(function (key) {
            return obj[key];
          });
        }

        return [];
      }
      /**
       * See if this is a GIF, if it is then report back so we
       * know which player to display.
       */

    }, {
      key: "_isGif",
      value: function _isGif(url) {
        if (url.indexOf(".gif") != -1) {
          return true;
        }

        return false;
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject12_e9ca42c02da211eaa247b36e11306899())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrnapp-studio-submission-display";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Object that has all the details of a submission that
           * is being viewed.
           */
          submission: {
            type: Object
          },
          elmslnCourse: {
            type: String,
            attribute: "elmsln-course"
          },
          elmslnSection: {
            type: String,
            attribute: "elmsln-section"
          },
          basePath: {
            type: String,
            attribute: "base-path"
          },
          csrfToken: {
            type: String,
            attribute: "csrf-token"
          },
          endPoint: {
            type: String,
            attribute: "end-point"
          }
        };
      }
    }]);
    return LrnappStudioSubmissionDisplay;
  }(_litElement.LitElement);

  _exports.LrnappStudioSubmissionDisplay = LrnappStudioSubmissionDisplay;
  window.customElements.define(LrnappStudioSubmissionDisplay.tag, LrnappStudioSubmissionDisplay);
});
