// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-submission/lrnapp-studio-submission-media-editoverlay.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@polymer/paper-icon-button/paper-icon-button.js", "../../../../@polymer/paper-dialog/paper-dialog.js", "../../../../@polymer/paper-dialog-scrollable/paper-dialog-scrollable.js"], function (_exports, _litElement, _paperIconButton, _paperDialog, _paperDialogScrollable) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioSubmissionMediaEditoverlay = void 0;

  function _templateObject4_0627a9a01bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: flex;\n          align-items: stretch;\n          justify-content: stretch;\n        }\n\n        .wrapper {\n          position: relative;\n          z-index: 1;\n          overflow: hidden;\n          display: flex;\n          align-items: center;\n        }\n\n        .actions {\n          overflow: hidden;\n          position: absolute;\n          display: flex;\n          justify-content: center;\n          align-items: center;\n          top: 0;\n          left: 0;\n          z-index: 5;\n          color: white;\n          background: rgba(0, 0, 0, 0.7);\n          height: 100%;\n          width: 100%;\n          transition: all 0.2s ease-in-out;\n          transform: translateY(-100%);\n        }\n\n        .wrapper:hover .actions {\n          transform: translateY(0);\n        }\n\n        paper-dialog {\n          width: 50%;\n          width: 50vmax;\n          padding: 1em;\n        }\n      "]);

    _templateObject4_0627a9a01bb311eabd7cc30119585ddb = function _templateObject4_0627a9a01bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject3_0627a9a01bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <paper-dialog id=\"dialog\">\n              <h2>Embed Code</h2>\n              <p>Paste this into the text area and your image will appear.</p>\n              <paper-dialog-scrollable>\n                <pre>", "</pre>\n              </paper-dialog-scrollable>\n            </paper-dialog>\n          "]);

    _templateObject3_0627a9a01bb311eabd7cc30119585ddb = function _templateObject3_0627a9a01bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject2_0627a9a01bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                <paper-icon-button\n                  icon=\"code\"\n                  class=\"embed\"\n                  @click=\"", "\"\n                ></paper-icon-button>\n              "]);

    _templateObject2_0627a9a01bb311eabd7cc30119585ddb = function _templateObject2_0627a9a01bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject_0627a9a01bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"wrapper\">\n        <slot></slot>\n        <div class=\"actions\">\n          ", "\n          <paper-icon-button\n            icon=\"delete\"\n            class=\"delete\"\n            @click=\"", "\"\n          ></paper-icon-button>\n        </div>\n      </div>\n      ", "\n    "]);

    _templateObject_0627a9a01bb311eabd7cc30119585ddb = function _templateObject_0627a9a01bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  var LrnappStudioSubmissionMediaEditoverlay =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnappStudioSubmissionMediaEditoverlay, _LitElement);

    function LrnappStudioSubmissionMediaEditoverlay() {
      babelHelpers.classCallCheck(this, LrnappStudioSubmissionMediaEditoverlay);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappStudioSubmissionMediaEditoverlay).apply(this, arguments));
    }

    babelHelpers.createClass(LrnappStudioSubmissionMediaEditoverlay, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_0627a9a01bb311eabd7cc30119585ddb(), this.embedcode ? (0, _litElement.html)(_templateObject2_0627a9a01bb311eabd7cc30119585ddb(), this._embedClicked) : "", this._deleteClicked, this.embedcode ? (0, _litElement.html)(_templateObject3_0627a9a01bb311eabd7cc30119585ddb(), this.embedcode) : "");
      }
    }, {
      key: "_embedClicked",
      value: function _embedClicked(e) {
        this.shadowRoot.querySelector("#dialog").open();
      }
    }, {
      key: "_deleteClicked",
      value: function _deleteClicked(e) {
        this.dispatchEvent(new CustomEvent("deleted", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: true
        }));
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject4_0627a9a01bb311eabd7cc30119585ddb())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrnapp-studio-submission-media-editoverlay";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          embedcode: {
            type: String
          }
        };
      }
    }]);
    return LrnappStudioSubmissionMediaEditoverlay;
  }(_litElement.LitElement);

  _exports.LrnappStudioSubmissionMediaEditoverlay = LrnappStudioSubmissionMediaEditoverlay;
  window.customElements.define(LrnappStudioSubmissionMediaEditoverlay.tag, LrnappStudioSubmissionMediaEditoverlay);
});
