// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-submission/lrnapp-studio-submission.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../secure-request/secure-request.js", "../../../../@polymer/app-route/app-location.js", "../../../../@polymer/app-route/app-route.js", "../../../../@polymer/paper-toast/paper-toast.js", "./lrnapp-studio-submission-page.js", "./lrnapp-studio-submission-button.js"], function (_exports, _litElement, _secureRequest, _appLocation, _appRoute, _paperToast, _lrnappStudioSubmissionPage, _lrnappStudioSubmissionButton) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioSubmission = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject4_fabc91f02c1c11eaacb4bf54e86db688() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        paper-button {\n          padding: 0;\n          margin: 0;\n          min-width: 1rem;\n        }\n      "]);

    _templateObject4_fabc91f02c1c11eaacb4bf54e86db688 = function _templateObject4_fabc91f02c1c11eaacb4bf54e86db688() {
      return data;
    };

    return data;
  }

  function _templateObject3_fabc91f02c1c11eaacb4bf54e86db688() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            This is the lrnapp-studio-submission page.\n          "]);

    _templateObject3_fabc91f02c1c11eaacb4bf54e86db688 = function _templateObject3_fabc91f02c1c11eaacb4bf54e86db688() {
      return data;
    };

    return data;
  }

  function _templateObject2_fabc91f02c1c11eaacb4bf54e86db688() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <lrnapp-studio-submission-page\n              base-path=\"{", "\"\n              @base-path-changed=\"", "\"\n              .route=\"", "\"\n              @route-changed=\"", "\"\n              .data=\"", "\"\n              @data-changed=\"", "\"\n              id=\"", "\"\n              end-point=\"", "\"\n              csrf-token=\"", "\"\n            ></lrnapp-studio-submission-page>\n          "]);

    _templateObject2_fabc91f02c1c11eaacb4bf54e86db688 = function _templateObject2_fabc91f02c1c11eaacb4bf54e86db688() {
      return data;
    };

    return data;
  }

  function _templateObject_fabc91f02c1c11eaacb4bf54e86db688() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <app-location\n        .route=\"", "\"\n        @route-changed=\"", "\"\n      ></app-location>\n      <app-route\n        .route=\"", "\"\n        @route-changed=\"", "\"\n        pattern=\"", "/submissions/:submission\"\n        .data=\"", "\"\n        @data-changed=\"", "\"\n        .tail=\"", "\"\n        @tail-changed=\"", "\"\n      >\n      </app-route>\n      ", "\n      <paper-toast id=\"toast\"></paper-toast>\n    "]);

    _templateObject_fabc91f02c1c11eaacb4bf54e86db688 = function _templateObject_fabc91f02c1c11eaacb4bf54e86db688() {
      return data;
    };

    return data;
  }

  var LrnappStudioSubmission =
  /*#__PURE__*/
  function (_SecureRequestXhr) {
    babelHelpers.inherits(LrnappStudioSubmission, _SecureRequestXhr);
    babelHelpers.createClass(LrnappStudioSubmission, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_fabc91f02c1c11eaacb4bf54e86db688(), this.route, this.routeChangedEvent, this.route, this.routeChangedEvent, this.endPoint, this.data, this.dataChangedEvent, this.tail, this.tailChangedEvent, this.data.submission ? (0, _litElement.html)(_templateObject2_fabc91f02c1c11eaacb4bf54e86db688(), this.basePath, this.basePathChangedEvent, this.tail, this.tailChangedEvent, this.data, this.dataChangedEvent, this.data.submission, this.endPoint, this.csrfToken) : (0, _litElement.html)(_templateObject3_fabc91f02c1c11eaacb4bf54e86db688()));
      }
    }, {
      key: "basePathChangedEvent",
      value: function basePathChangedEvent(e) {
        this.basePath = e.detail.value;
      }
    }, {
      key: "tailChangedEvent",
      value: function tailChangedEvent(e) {
        this.tail = e.detail.value;
      }
    }, {
      key: "routeChangedEvent",
      value: function routeChangedEvent(e) {
        this.route = e.detail.value;
      }
    }, {
      key: "dataChangedEvent",
      value: function dataChangedEvent(e) {
        this.data = e.detail.value;
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (["route", "endPoint"].includes(propName)) {
            _this2._routeChanged(_this2.route, _this2.endPoint);
          }

          if (["endPoint", "csrfToken"].includes(propName)) {
            _this2._updateCookies(_this2.endPoint, _this2.csrfToken);
          }
        });
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject4_fabc91f02c1c11eaacb4bf54e86db688())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrnapp-studio-submission";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          elmslnCourse: {
            type: String,
            attribute: "elmsln-course"
          },
          elmslnSection: {
            type: String,
            attribute: "elmsln-section"
          },
          basePath: {
            type: String,
            attribute: "base-path"
          },
          csrfToken: {
            type: String,
            attribute: "csrf-token"
          },
          endPoint: {
            type: String,
            attribute: "end-point"
          },
          activePage: {
            type: String,
            attribute: "active-page"
          }
        };
      }
    }]);

    function LrnappStudioSubmission() {
      var _this;

      babelHelpers.classCallCheck(this, LrnappStudioSubmission);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappStudioSubmission).call(this));
      setTimeout(function () {
        _this.addEventListener("submissionDeleted", _this._handleSubmissionDeletion.bind(babelHelpers.assertThisInitialized(_this)));

        _this.addEventListener("displaymessage", _this._handleDisplayMessage.bind(babelHelpers.assertThisInitialized(_this)));
      }, 0);
      return _this;
    }

    babelHelpers.createClass(LrnappStudioSubmission, [{
      key: "_handleRouteChange",
      value: function _handleRouteChange(event) {
        var path = event.detail.path;

        if (path) {
          var attr = this.route;
          attr.path = path;
          this.route = _objectSpread({}, attr);
        }
      } // If the current route is outside the scope of our app
      // then allow the website to break out of the single page
      // application routing

    }, {
      key: "_routeChanged",
      value: function _routeChanged(route, endPoint) {
        if (typeof route.path === "string") {
          if (typeof endPoint === "string") {
            if (route.path.startsWith(endPoint)) {
              return;
            }
          } // reload the page which since route changed will load that page


          window.location.reload();
        }
      }
    }, {
      key: "_handleSubmissionDeletion",
      value: function _handleSubmissionDeletion(e) {
        var submission = e.detail.submission;

        if (submission) {
          var attr = this.route;
          attr.path = this.endPoint;
          this.route = _objectSpread({}, attr);
          this.shadowRoot.querySelector("#toast").show("Submission has been deleted.");
        }
      }
    }, {
      key: "_updateCookies",
      value: function _updateCookies(endPoint, csrfToken) {
        if (endPoint && csrfToken) {
          this.setCookies(endPoint, csrfToken);
        }
      }
    }, {
      key: "_handleDisplayMessage",
      value: function _handleDisplayMessage(e, detail) {
        if (typeof e.detail.messsage !== "undefined") {
          this.shadowRoot.querySelector("#toast").show(e.detail.message);
        }
      }
      /**
       * Simple way to convert from object to array.
       */

    }, {
      key: "_toArray",
      value: function _toArray(obj) {
        if (obj == null) {
          return [];
        }

        return Object.keys(obj).map(function (key) {
          return obj[key];
        });
      }
    }]);
    return LrnappStudioSubmission;
  }((0, _secureRequest.SecureRequestXhr)(_litElement.LitElement));

  _exports.LrnappStudioSubmission = LrnappStudioSubmission;
  window.customElements.define(LrnappStudioSubmission.tag, LrnappStudioSubmission);
});
