// @lrnwebcomponents/editable-table/lib/editable-table-filter.js
define(["exports","../../../lit-element/lit-element.js","../../../@polymer/paper-button/paper-button.js","../../simple-tooltip/simple-tooltip.js","../../../@polymer/iron-icons/iron-icons.js","./editable-table-iconset.js"],(function(e,t,i,l,o,n){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.EditableTableFilter=void 0;
/**
   * Copyright 2018 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class EditableTableFilter extends t.LitElement{static get styles(){return[t.css`
        paper-button {
          padding: var(--editable-table-cell-padding, 0);
          margin: 0;
          width: auto;
          min-width: unset;
          display: flex;
          justify-content: space-between;
          align-items: center;
          align-content: stretch;
          text-transform: unset;
          font-family: var(--editable-table-font-family);
        }
        paper-button > div {
          flex-grow: 1;
        }
        iron-icon {
          min-width: 24px;
        }
        .sr-only {
          position: absolute;
          left: -9999px;
          font-size: 0;
          height: 0;
          width: 0;
          overflow: hidden;
        }
        #filter-off {
          opacity: 0.25;
        }
        :host(:not([filtered])) .filtered,
        :host(:not([filtered]):not(:focus):not(:hover)) #filter,
        :host(:not([filtered]):focus) #filter-off,
        :host(:not([filtered]):hover) #filter-off,
        :host([filtered]:not(:focus):not(:hover)) #filter-off,
        :host([filtered]:focus) #filter,
        :host([filtered]:hover) #filter {
          display: none;
        }
      `]}render(){return t.html`
      <paper-button
        id="button"
        class="container"
        @click="${this._onFilterClicked}"
      >
        <span>${this.text}</span>
        <span class="sr-only" .hidden="${!this.filtered}"> (filtered)</span>
        <span class="sr-only"> Toggle filter.</span>
        <iron-icon id="filter" icon="editable-table:filter"></iron-icon>
        <iron-icon id="filter-off" icon="editable-table:filter-off"></iron-icon>
      </paper-button>
      <simple-tooltip for="button"
        >Toggle Column ${this.columnIndex} filter for
        "${this.text}"</simple-tooltip
      >
    `}static get tag(){return"editable-table-filter"}constructor(){super(),this.columnIndex=null,this.filtered=!1,this.text=""}static get properties(){return{columnIndex:{type:Number,attribute:"column-index"},filtered:{type:Boolean,reflect:!0},text:{type:String}}}_getPressed(e){return e?"true":"false"}_onFilterClicked(){this.dispatchEvent(new CustomEvent("toggle-filter",{bubbles:!0,cancelable:!0,composed:!0,detail:this}))}}e.EditableTableFilter=EditableTableFilter,window.customElements.define(EditableTableFilter.tag,EditableTableFilter)}));
