// @lrnwebcomponents/editable-table/lib/editable-table-filter.js
define(["exports","../../../@polymer/polymer/polymer-element.js","../../../@polymer/paper-button/paper-button.js","../../../@polymer/paper-tooltip/paper-tooltip.js","../../../@polymer/iron-icons/iron-icons.js","./editable-table-iconset.js"],function(_exports,_polymerElement,_paperButton,_paperTooltip,_ironIcons,_editableTableIconset){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.EditableTableFilter=void 0;class EditableTableFilter extends _polymerElement.PolymerElement{static get template(){return _polymerElement.html`
      <style is="custom-style">
        :host paper-button {
          padding: var(--editable-table-cell-padding, 0);
          margin: 0;
          width: auto;
          min-width: unset;
          display: flex;
          justify-content: space-between;
          align-items: center;
          align-content: stretch;
          text-transform: unset;
          font-family: var(--editable-table-font-family);
        }
        :host paper-button > div {
          flex-grow: 1;
        }
        :host iron-icon {
          min-width: 24px;
        }
        :host .sr-only {
          position: absolute;
          left: -9999px;
          font-size: 0;
          height: 0;
          width: 0;
          overflow: hidden;
        }
        :host #filter-off {
          opacity: 0.25;
        }
        :host(:not([filtered])) .filtered,
        :host(:not([filtered]):not(:focus):not(:hover)) #filter,
        :host(:not([filtered]):focus) #filter-off,
        :host(:not([filtered]):hover) #filter-off,
        :host([filtered]:not(:focus):not(:hover)) #filter-off,
        :host([filtered]:focus) #filter,
        :host([filtered]:hover) #filter {
          display: none;
        }
      </style>
      <paper-button id="button" class="container" on-click="_onFilterClicked">
        <span>[[text]]</span>
        <span class="sr-only" hidden\$="[[!filtered]]"> (filtered)</span>
        <span class="sr-only"> Toggle filter.</span>
        <iron-icon id="filter" icon="editable-table:filter"></iron-icon>
        <iron-icon id="filter-off" icon="editable-table:filter-off"></iron-icon>
      </paper-button>
      <paper-tooltip for="button"
        >Toggle Column [[columnIndex]] filter for "[[text]]"</paper-tooltip
      >
    `}static get tag(){return"editable-table-filter"}static get properties(){return{columnIndex:{type:Number,value:null},filtered:{type:Boolean,value:!1,reflectToAttribute:!0},text:{type:String,value:""}}}_getPressed(filtered){return filtered?"true":"false"}_onFilterClicked(){this.dispatchEvent(new CustomEvent("toggle-filter",{bubbles:!0,cancelable:!0,composed:!0,detail:this}))}}_exports.EditableTableFilter=EditableTableFilter,window.customElements.define(EditableTableFilter.tag,EditableTableFilter)});
