// @lrnwebcomponents/a11y-media-player/lib/a11y-media-controls.js
define(["exports", "../../../@polymer/polymer/polymer-element.js", "./a11y-media-behaviors.js", "../../../@polymer/paper-menu-button/paper-menu-button.js", "../../../@polymer/paper-listbox/paper-listbox.js", "../../../@polymer/paper-input/paper-input.js", "../../../@polymer/paper-item/paper-item.js", "../../../@polymer/paper-icon-button/paper-icon-button.js", "../../../@polymer/paper-toggle-button/paper-toggle-button.js", "../../dropdown-select/dropdown-select.js", "../../../@polymer/paper-slider/paper-slider.js", "../../../@polymer/paper-tooltip/paper-tooltip.js", "./a11y-media-button.js"], function (_exports, _polymerElement, _a11yMediaBehaviors, _paperMenuButton, _paperListbox, _paperInput, _paperItem, _paperIconButton, _paperToggleButton, _dropdownSelect, _paperSlider, _paperTooltip, _a11yMediaButton) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.A11yMediaControls = void 0;

  function _templateObject_e1a2b4b02da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style is=\"custom-style\" include=\"simple-colors-shared-styles-polymer\">\n        :host {\n          display: block;\n          width: 100%;\n          max-width: 100%;\n          height: 44px;\n          max-height: 44px;\n          position: relative;\n          color: var(--a11y-media-color);\n          --primary-text-color: var(--a11y-media-settings-menu-color);\n          --paper-menu-button-dropdown-background: var(--a11y-media-settings-menu-bg-color);\n          --paper-listbox-background-color: var(--a11y-media-settings-menu-bg-color);\n          --paper-listbox-color: var(--a11y-media-settings-menu-color);\n          --paper-listbox: {\n            padding: 0;\n          };\n          --paper-menu-button: {\n            background-color: var(--a11y-media-settings-menu-bg-color);\n            color: var(--a11y-media-settings-menu-color);\n          };\n          --paper-menu-button-dropdown: {\n            background-color: var(--a11y-media-settings-menu-bg-color);\n            color: var(--a11y-media-settings-menu-color);\n            margin-top: 0 !important;\n            margin-bottom: 0 !important;\n          };\n          --paper-item-selected: {\n            color: var(--a11y-media-settings-menu-hover-color);\n          };\n          --paper-item-focused: {\n            color: var(--a11y-media-settings-menu-hover-color);\n          };\n        }\n        :host > #controls-left {\n          position: absolute;\n          left: 0;\n          min-width: 200px;\n        }\n        :host > #controls-right {\n          position: absolute;\n          right: 0;\n          top: -2px;\n        }\n        :host paper-menu-button,\n        :host dropdown-select {\n          padding: 0;\n        }\n        :host paper-icon-button {\n          background-color: var(--a11y-media-settings-menu-bg-color);\n          color: var(--a11y-media-settings-menu-color);\n        }\n        :host paper-icon-button:active,\n        :host paper-icon-button:focus,\n        :host paper-icon-button:hover {\n          background-color: var(--a11y-media-settings-menu-bg-color);\n          color: var(--a11y-media-settings-menu-color);\n        }\n        :host paper-item {\n          min-height: 40;\n        }\n        :host paper-slider {\n          @apply --a11y-media-slider;\n        }\n        :host .play-status,\n        :host paper-icon-button {\n          border: none;\n          position: relative;\n        }\n        :host .play-status {\n          font-size: 85%;\n        }\n        :host .play-status.control-bar {\n          padding: 8px 13px 8px;\n        }\n        :host([hide-play-status]) .play-status {\n          display: none;\n        }\n        :host .setting {\n          display: flex;\n          justify-content: space-between;\n          align-items: center;\n          width: 100%;\n        }\n        :host .setting-text {\n          min-width: 125px;\n        }\n        :host .setting-control {\n          max-width: 100px;\n        }\n        :host .setting-slider {\n          width: 130px;\n          margin-left: -15px;\n          margin-right: -15px;\n          @apply --a11y-media-slider;\n        }\n        :host #speed {\n          --paper-slider-knob-start-color: var(--a11y-media-accent-color);\n          --paper-slider-knob-start-border-color: var(--a11y-media-accent-color);\n          --paper-slider-knob-end-color: var(--a11y-media-accent-color);\n          --paper-slider-knob-end-border-color: var(--a11y-media-accent-color);\n        }\n        :host #showvolume {\n          display: inline;\n          position: relative;\n        }\n        :host #volume {\n          z-index: 1;\n          position: absolute;\n          left: 30px;\n          top: -5px;\n          width: 0;\n          overflow: hidden;\n          transition: width 0.5s;\n          z-index: 3;\n          background-color: var(--a11y-media-bg-color);\n          --paper-slider-knob-end-color: var(--a11y-media-accent-color);\n          --paper-slider-knob-end-border-color: var(--a11y-media-accent-color);\n\n        }\n        :host #volume:active,\n        :host #volume:focus,\n        :host #volume:hover,\n        :host #volume.focus,\n        :host #showvolume:active #volume,\n        :host #showvolume:focus #volume,\n        :host #showvolume:hover #volume {\n          overflow: visible;\n          width: 100px;\n        }\n        :host([responsive-size=\"xs\"]) #volume:active,\n        :host([responsive-size=\"xs\"]) #volume:focus,\n        :host([responsive-size=\"xs\"]) #volume:hover,\n        :host([responsive-size=\"xs\"]) #volume.focus,\n        :host([responsive-size=\"xs\"]) #showvolume:active #volume,\n        :host([responsive-size=\"xs\"]) #showvolume:focus #volume,\n        :host([responsive-size=\"xs\"]) #showvolume:hover #volume {\n          top: -35px\n          border-radius: 4px;\n        }\n        :host .sr-only {\n          position: absolute;\n          left: -99999;\n          top: 0;\n          height: 0;\n          width: 0;\n          overflow: hidden;\n        }\n      </style>\n      <div id=\"controls-left\">\n        <a11y-media-button\n          action$=\"[[playPause.action]]\"\n          icon$=\"[[playPause.icon]]\"\n          label$=\"[[playPause.label]]\"\n          on-click=\"_onButtonTap\"\n        ></a11y-media-button>\n        <a11y-media-button\n          action=\"rewind\"\n          disabled$=\"[[disableSeek]]\"\n          hidden$=\"[[compactControls]]\"\n          icon$=\"[[_getLocal('rewind','icon')]]\"\n          label$=\"[[_getLocal('rewind','label')]]\"\n          on-click=\"_onButtonTap\"\n        ></a11y-media-button>\n        <a11y-media-button\n          action=\"forward\"\n          disabled$=\"[[disableSeek]]\"\n          hidden$=\"[[compactControls]]\"\n          icon$=\"[[_getLocal('forward','icon')]]\"\n          label$=\"[[_getLocal('forward','label')]]\"\n          on-click=\"_onButtonTap\"\n        ></a11y-media-button>\n        <a11y-media-button\n          action=\"restart\"\n          disabled$=\"[[disableSeek]]\"\n          hidden$=\"[[compactControls]]\"\n          icon$=\"[[_getLocal('restart','icon')]]\"\n          label$=\"[[_getLocal('restart','label')]]\"\n          on-click=\"_onButtonTap\"\n        ></a11y-media-button>\n        <div id=\"showvolume\">\n          <a11y-media-button\n            id=\"mute\"\n            action$=\"[[muteUnmute.action]]\"\n            icon$=\"[[muteUnmute.icon]]\"\n            label$=\"[[muteUnmute.label]]\"\n            on-click=\"_onButtonTap\"\n          ></a11y-media-button>\n          <paper-slider\n            id=\"volume\"\n            aria-labelledby=\"volume-slider-label\"\n            label$=\"[[volumeLabel]]\"\n            min=\"0\"\n            max=\"100\"\n            on-change=\"_onSettingsChanged\"\n            pin\n            step=\"10\"\n            value$=\"[[volume]]\"\n          ></paper-slider>\n        </div>\n        <span aria-live=\"polite\" class=\"play-status control-bar\">\n          <span id=\"statbar\"></span>\n        </span>\n      </div>\n      <div id=\"controls-right\">\n        <a11y-media-button\n          action=\"captions\"\n          disabled$=\"[[!hasCaptions]]\"\n          hidden$=\"[[!hasCaptions]]\"\n          icon$=\"[[_getLocal('captions','icon')]]\"\n          label$=\"[[_getLocal('captions','label')]]\"\n          on-click=\"_onButtonTap\"\n          toggle$=\"[[cc]]\"\n        >\n        </a11y-media-button>\n        <a11y-media-button\n          action=\"transcript\"\n          controls=\"transcript\"\n          disabled$=\"[[hideTranscriptButton]]\"\n          hidden$=\"[[hideTranscriptButton]]\"\n          icon$=\"[[_getLocal('transcript','icon')]]\"\n          label$=\"[[_getLocal('transcript','label')]]\"\n          on-click=\"_onButtonTap\"\n          toggle$=\"[[hideTranscript]]\"\n        >\n        </a11y-media-button>\n        <a11y-media-button\n          action=\"linkable\"\n          disabled$=\"[[!linkable]]\"\n          hidden$=\"[[!linkable]]\"\n          icon=\"[[_getLocal('copyLink','icon')]]\"\n          label=\"[[_getLocal('copyLink','label')]]\"\n          on-click=\"_onButtonTap\"\n        ></a11y-media-button>\n        <a11y-media-button\n          action=\"print\"\n          disabled$=\"[[noPrinting]]\"\n          hidden$=\"[[noPrinting]]\"\n          icon$=\"[[_getLocal('print','icon')]]\"\n          label=\"[[_getLocal('print','label')]]\"\n          on-click=\"_handlePrintClick\"\n        >\n        </a11y-media-button>\n        <a11y-media-button\n          action=\"download\"\n          disabled$=\"[[noPrinting]]\"\n          hidden$=\"[[noPrinting]]\"\n          icon$=\"[[_getLocal('download','icon')]]\"\n          label=\"[[_getLocal('download','label')]]\"\n          on-click=\"_handleDownloadClick\"\n        >\n        </a11y-media-button>\n        <template is=\"dom-if\" if=\"[[fullscreenButton]]\">\n          <a11y-media-button\n            action=\"fullscreen\"\n            hidden$=\"[[audioNoThumb]]\"\n            icon$=\"[[_getLocal('fullscreen','icon')]]\"\n            label$=\"[[_getLocal('fullscreen','label')]]\"\n            on-click=\"_onButtonTap\"\n            toggle$=\"[[fullscreen]]\"\n            step=\"1\"\n          >\n          </a11y-media-button>\n        </template>\n        <paper-menu-button\n          id=\"settings\"\n          allow-outside-scroll\n          horizontal-align=\"right\"\n          ignore-select\n          on-change=\"_onSettingsChanged\"\n          on-iron-activate=\"_handleSettingsActivate\"\n          on-iron-select=\"_handleSettingsSelect\"\n          vertical-align=\"bottom\"\n        >\n          <paper-icon-button\n            action=\"settings\"\n            alt$=\"[[_getLocal('settings','label')]]\"\n            icon$=\"[[_getLocal('settings','icon')]]\"\n            slot=\"dropdown-trigger\"\n          >\n          </paper-icon-button>\n          <paper-listbox id=\"settingslist\" slot=\"dropdown-content\">\n            <paper-item hidden$=\"[[!hasCaptions]]\">\n              <div class=\"setting\">\n                <div class=\"setting-text\">\n                  [[_getLocal('captions','label')]]\n                </div>\n                <div class=\"setting-control\">\n                  <dropdown-select\n                    id=\"tracks\"\n                    disabled$=\"[[!hasCaptions]]\"\n                    no-label-float\n                    on-change=\"_handleTrackChange\"\n                    value\n                  >\n                    <paper-item value=\"\"\n                      >[[_getLocal('captions','off')]]</paper-item\n                    >\n                    <template is=\"dom-repeat\" items=\"{{tracks}}\" as=\"option\">\n                      <paper-item value$=\"{{option.value}}\"\n                        >{{option.text}}</paper-item\n                      >\n                    </template>\n                  </dropdown-select>\n                </div>\n              </div>\n            </paper-item>\n            <paper-item hidden$=\"[[noTranscriptToggle]]\">\n              <div class=\"setting\">\n                <div id=\"transcript-label\" class=\"setting-text\">\n                  [[_getLocal('transcript','label')]]\n                </div>\n                <div class=\"setting-control\">\n                  <paper-toggle-button\n                    id=\"transcript-toggle\"\n                    aria-labelledby=\"transcript-label\"\n                    checked$=\"[[!hideTranscript]]\"\n                    controls=\"transcript\"\n                    disabled$=\"[[noTranscriptToggle]]\"\n                  >\n                  </paper-toggle-button>\n                </div>\n              </div>\n            </paper-item>\n            <paper-item>\n              <div class=\"setting\">\n                <div id=\"loop-label\" class=\"setting-text\">\n                  [[_getLocal('loop','label')]]\n                </div>\n                <div class=\"setting-control\">\n                  <paper-toggle-button\n                    id=\"loop\"\n                    aria-labelledby=\"loop-label\"\n                    checked$=\"[[loop]]\"\n                  ></paper-toggle-button>\n                </div>\n              </div>\n            </paper-item>\n            <paper-item>\n              <div class=\"setting\">\n                <div id=\"speed-label\" class=\"setting-text\">\n                  [[_getLocal('speed','label')]]\n                </div>\n                <div class=\"setting-control\">\n                  <paper-slider\n                    id=\"speed\"\n                    aria-labelledby=\"speed-label\"\n                    class=\"setting-slider\"\n                    min=\"0.5\"\n                    max=\"4\"\n                    pin\n                    step=\"0.5\"\n                    tab-index=\"-1\"\n                    value$=\"[[playbackRate]]\"\n                  ></paper-slider>\n                </div>\n              </div>\n            </paper-item>\n          </paper-listbox>\n        </paper-menu-button>\n        <paper-tooltip for=\"settings\">\n          [[_getLocal('settings','label')]]\n        </paper-tooltip>\n      </div>\n    "]);

    _templateObject_e1a2b4b02da211eaa247b36e11306899 = function _templateObject_e1a2b4b02da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  /**
   * `a11y-media-controls`
   * @customElement a11y-media-controls
   * `The controls bar for the a11y-media-player.`
   *
   * @microcopy - language worth noting:
   *
   * @extends A11yMediaBehaviors
  
   * @polymer
   */
  var A11yMediaControls =
  /*#__PURE__*/
  function (_A11yMediaBehaviors) {
    babelHelpers.inherits(A11yMediaControls, _A11yMediaBehaviors);
    babelHelpers.createClass(A11yMediaControls, null, [{
      key: "properties",
      // properties available to the custom element for data binding
      get: function get() {
        return {
          /**
           * Use compact controls?
           */
          compactControls: {
            name: "compactControls",
            type: Boolean,
            computed: "_getCompactControls(responsiveSize)"
          },

          /**
           * Is the player a fixed height (iframe mode) so that theure is no transcript toggle?
           */
          fixedHeight: {
            type: Boolean,
            value: false
          },

          /**
           * Is fullscreen mode?
           */
          fullscreen: {
            name: "fullscreen",
            type: Boolean,
            value: false
          },

          /**
           * show the FullscreenButton?
           */
          fullscreenButton: {
            name: "fullscreenButton",
            type: Boolean,
            value: false,
            nofity: true
          },

          /**
           * Does the player have tracks?
           */
          hasCaptions: {
            name: "hasCaptions",
            type: Boolean,
            value: false,
            notify: true
          },

          /**
           * initially hide the transcript?
           */
          hideTranscript: {
            name: "hideTranscript",
            type: Boolean,
            value: false
          },

          /**
           * hide the transcript toggle menu item?
           */
          hideTranscriptButton: {
            type: Boolean,
            computed: "_hideTranscriptButton(noTranscriptToggle,compactControls)"
          },

          /**
           * url for deeplinking
           */
          linkUrl: {
            type: String,
            value: "test"
          },

          /**
           * mute/unmute button
           */
          muteUnmute: {
            name: "muteUnmute",
            type: Object
          },

          /**
           * hide the print transcript feature available?
           */
          noPrinting: {
            type: Boolean,
            computed: "_noPrinting(standAlone,fixedHeight)"
          },

          /**
           * Is the transctipt toggle feature available?
           */
          noTranscriptToggle: {
            type: Boolean,
            computed: "_noTranscriptToggle(standAlone,fixedHeight,hasTranscript)"
          },

          /**
           * Size of the a11y media element for responsive styling
           */
          responsiveSize: {
            name: "responsiveSize",
            type: String,
            notify: true,
            value: "xs",
            reflectToAttribute: true
          }
        };
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */

    }, {
      key: "tag",
      get: function get() {
        return "a11y-media-controls";
      } //inherit styles from a11y-media-player

    }]);

    function A11yMediaControls() {
      babelHelpers.classCallCheck(this, A11yMediaControls);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(A11yMediaControls).call(this));
    } //render function


    babelHelpers.createClass(A11yMediaControls, [{
      key: "connectedCallback",

      /**
       * life cycle, element is afixed to the DOM
       */
      value: function connectedCallback() {
        babelHelpers.get(babelHelpers.getPrototypeOf(A11yMediaControls.prototype), "connectedCallback", this).call(this);
        var root = this;

        root._addResponsiveUtility({
          element: root,
          attribute: "responsive-size",
          relativeToParent: true,
          sm: 300,
          md: 600,
          lg: 900,
          xl: 1200
        });
      }
      /**
       * sets target for a11y keys
       */

    }, {
      key: "ready",
      value: function ready() {
        babelHelpers.get(babelHelpers.getPrototypeOf(A11yMediaControls.prototype), "ready", this).call(this);
        var root = this; // when the mute button is in focus,
        // add focus class to make the volume slider appear

        root.shadowRoot.querySelector("#mute").shadowRoot.querySelector("#button").onfocus = function () {
          root.shadowRoot.querySelector("#volume").classList.add("focus");
        }; // when the mute button is blurred,
        // remove focus class to make the volume slider disappear


        root.shadowRoot.querySelector("#mute").shadowRoot.querySelector("#button").onblur = function () {
          root.shadowRoot.querySelector("#volume").classList.remove("focus");
        };
      }
      /**
       * updates the play status of the media
       *
       * @param {string} the status, eg., `Loading...` or `00:02:01/00:11:43`
       */

    }, {
      key: "setStatus",
      value: function setStatus(status) {
        this.shadowRoot.querySelector("#statbar").innerText = status;
      }
      /**
       * loads tracks from array
       *
       * @param {object} the tracks of the media
       */

    }, {
      key: "setTracks",
      value: function setTracks(tracks) {
        this.set("tracks", []);
        this.set("tracks", tracks.slice(0));
      }
      /**
       * returns true if player is xs or sm and needs to use compact controls
       *
       * @param {string} the size of the player: `xs`,`sm`,`md`,`lg`, or `xl`
       * @returns {boolean} Should the player use compact controls?
       */

    }, {
      key: "_getCompactControls",
      value: function _getCompactControls(responsiveSize) {
        return this._testAttribute(responsiveSize, "xs") || this._testAttribute(responsiveSize, "sm");
      }
      /**
       * handles when the tracks dropdown selection changes
       * (when the tracks dropdown-select changes, update track and CC button)
       */

    }, {
      key: "_handleTrackChange",
      value: function _handleTrackChange(e) {
        var root = this;

        if (root.__selectedTrack !== null) {
          if (e.detail.value !== "") {
            root.dispatchEvent(new CustomEvent("select-track", {
              detail: {
                control: root,
                value: e.detail.value
              }
            }));
            root.dispatchEvent(new CustomEvent("toggle-cc", {
              detail: {
                control: root,
                value: true
              }
            }));
          } else {
            root.dispatchEvent(new CustomEvent("toggle-cc", {
              detail: {
                control: root,
                value: false
              }
            }));
          }
        }
      }
      /**
       * prevent settings menu from being closed before action is taken
       */

    }, {
      key: "_handleSettingsSelect",
      value: function _handleSettingsSelect(e) {} //e.preventDefault();

      /**
       * prevent settings menu from being closed before action is taken
       */

    }, {
      key: "_handleSettingsActivate",
      value: function _handleSettingsActivate(e) {} //if (e.target == this.shadowRoot.querySelector('#settingslist')) e.preventDefault();

      /**
       * determine which button was clicked and act accordingly
       */

    }, {
      key: "_onButtonTap",
      value: function _onButtonTap(e) {
        this.dispatchEvent(new CustomEvent("controls-change", {
          detail: e.detail
        }));
        this.shadowRoot.querySelector("#settings").close();
      }
      /**
       * determine which button was clicked and act accordingly
       */

    }, {
      key: "_onSettingsChanged",
      value: function _onSettingsChanged(e) {
        this.dispatchEvent(new CustomEvent("controls-change", {
          detail: e.target
        }));
        this.shadowRoot.querySelector("#settings").close();
      }
      /**
       * Determines if the transcript button should be hidden.
       *
       * @param {boolean} Is this feature available?
       * @param {boolean} Is the player too small to fit the extra controls?
       * @returns {boolean} Show the transcript button?
       */

    }, {
      key: "_hideTranscriptButton",
      value: function _hideTranscriptButton(noTranscriptToggle, compactControls) {
        return noTranscriptToggle || compactControls;
      }
      /**
       * Determines if the print transcript feature is available from the controls.
       *
       * @param {boolean} Is the player in stand-alone mode?
       * @param {boolean} Is the player in fixed-height/iframe mode?
       * @returns {boolean} Should print transcript be unavailable from controls?
       */

    }, {
      key: "_noPrinting",
      value: function _noPrinting(standAlone, fixedHeight) {
        return standAlone || !fixedHeight;
      }
      /**
       * Determines if the transcript toggle feature is available.
       *
       * @param {boolean} Is the player in stand-alone mode?
       * @param {boolean} Is the player in fixed-height/iframe mode?
       * @returns {boolean} Should transcript toggle be unavailable?
       */

    }, {
      key: "_noTranscriptToggle",
      value: function _noTranscriptToggle(standAlone, fixedHeight, hasTranscript) {
        return standAlone || fixedHeight || !hasTranscript;
      }
    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_e1a2b4b02da211eaa247b36e11306899());
      }
    }]);
    return A11yMediaControls;
  }(_a11yMediaBehaviors.A11yMediaBehaviors);

  _exports.A11yMediaControls = A11yMediaControls;
  window.customElements.define(A11yMediaControls.tag, A11yMediaControls);
});
