// @vaadin/vaadin-split-layout/src/vaadin-split-layout.js
define(["exports","../../../@polymer/polymer/polymer-element.js","../../../@polymer/polymer/lib/mixins/gesture-event-listeners.js","../../../@polymer/polymer/lib/utils/flattened-nodes-observer.js","../../../@polymer/iron-resizable-behavior/iron-resizable-behavior.js","../../vaadin-themable-mixin/vaadin-themable-mixin.js","../../vaadin-element-mixin/vaadin-element-mixin.js","../../../@polymer/polymer/lib/utils/html-tag.js","../../../@polymer/polymer/lib/legacy/class.js"],function(_exports,_polymerElement,_gestureEventListeners,_flattenedNodesObserver,_ironResizableBehavior,_vaadinThemableMixin,_vaadinElementMixin,_htmlTag,_class){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.SplitLayoutElement=void 0;class SplitLayoutElement extends((0,_vaadinElementMixin.ElementMixin)((0,_vaadinThemableMixin.ThemableMixin)((0,_gestureEventListeners.GestureEventListeners)((0,_class.mixinBehaviors)([_ironResizableBehavior.IronResizableBehavior],_polymerElement.PolymerElement))))){static get template(){return _htmlTag.html`
    <style>
      :host {
        display: flex;
        overflow: hidden !important;
        transform: translateZ(0);
      }

      :host([hidden]) {
        display: none !important;
      }

      :host([orientation="vertical"]) {
        flex-direction: column;
      }

      :host ::slotted(*) {
        flex: 1 1 auto;
        overflow: auto;
        -webkit-overflow-scrolling: touch;
      }

      [part="splitter"] {
        flex: none;
        position: relative;
        z-index: 1;
        overflow: visible;
        min-width: 8px;
        min-height: 8px;
      }

      :host(:not([orientation="vertical"])) > [part="splitter"] {
        cursor: ew-resize;
      }

      :host([orientation="vertical"]) > [part="splitter"] {
        cursor: ns-resize;
      }

      [part="handle"] {
        width: 40px;
        height: 40px;
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate3d(-50%, -50%, 0);
      }
    </style>
    <slot id="primary" name="primary"></slot>
    <div part="splitter" id="splitter" on-track="_onHandleTrack" on-down="_setPointerEventsNone" on-up="_restorePointerEvents">
      <div part="handle"></div>
    </div>
    <slot id="secondary" name="secondary"></slot>

    <div hidden="">
      <!-- Workaround to fix a Shady style scoping issue caused by dynamic slot naming of the child elements (primary/secondary) -->
      <slot></slot>
    </div>
`}static get is(){return"vaadin-split-layout"}static get version(){return"4.1.0"}static get properties(){return{orientation:{type:String,reflectToAttribute:!0,value:"horizontal"},_previousPrimaryPointerEvents:String,_previousSecondaryPointerEvents:String}}ready(){super.ready(),new _flattenedNodesObserver.FlattenedNodesObserver(this,this._processChildren)}_processChildren(){this.getEffectiveChildren().forEach((child,i)=>{0===i?(this._primaryChild=child,child.setAttribute("slot","primary")):1==i?(this._secondaryChild=child,child.setAttribute("slot","secondary")):child.removeAttribute("slot")})}_setFlexBasis(element,flexBasis,containerSize){0===(flexBasis=Math.max(0,Math.min(flexBasis,containerSize)))&&(flexBasis=1e-6),element.style.flex="1 1 "+flexBasis+"px"}_setPointerEventsNone(event){this._previousPrimaryPointerEvents=this._primaryChild.style.pointerEvents,this._previousSecondaryPointerEvents=this._secondaryChild.style.pointerEvents,this._primaryChild.style.pointerEvents="none",this._secondaryChild.style.pointerEvents="none",event.preventDefault()}_restorePointerEvents(){this._primaryChild.style.pointerEvents=this._previousPrimaryPointerEvents,this._secondaryChild.style.pointerEvents=this._previousSecondaryPointerEvents}_onHandleTrack(event){if(this._primaryChild&&this._secondaryChild){var size="vertical"===this.orientation?"height":"width";if("start"!==event.detail.state){var distance="vertical"===this.orientation?event.detail.dy:event.detail.dx;this._setFlexBasis(this._primaryChild,this._startSize.primary+distance,this._startSize.container),this._setFlexBasis(this._secondaryChild,this._startSize.secondary-distance,this._startSize.container),this.notifyResize(),"end"===event.detail.state&&(this.dispatchEvent(new CustomEvent("splitter-dragend")),delete this._startSize)}else this._startSize={container:this.getBoundingClientRect()[size]-this.$.splitter.getBoundingClientRect()[size],primary:this._primaryChild.getBoundingClientRect()[size],secondary:this._secondaryChild.getBoundingClientRect()[size]}}}}_exports.SplitLayoutElement=SplitLayoutElement,customElements.define(SplitLayoutElement.is,SplitLayoutElement)});
