// @vaadin/vaadin-split-layout/src/vaadin-split-layout.js
define(["exports","../../../@polymer/polymer/polymer-element.js","../../../@polymer/polymer/lib/mixins/gesture-event-listeners.js","../../../@polymer/polymer/lib/utils/flattened-nodes-observer.js","../../../@polymer/iron-resizable-behavior/iron-resizable-behavior.js","../../vaadin-themable-mixin/vaadin-themable-mixin.js","../../vaadin-element-mixin/vaadin-element-mixin.js","../../../@polymer/polymer/lib/utils/html-tag.js","../../../@polymer/polymer/lib/legacy/class.js"],(function(e,t,i,s,r,n,o,a,l){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.SplitLayoutElement=void 0;
/**
  @license
  Copyright (c) 2016 Vaadin Ltd.
  This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
  */
class SplitLayoutElement extends((0,o.ElementMixin)((0,n.ThemableMixin)((0,i.GestureEventListeners)((0,l.mixinBehaviors)([r.IronResizableBehavior],t.PolymerElement))))){static get template(){return a.html`
    <style>
      :host {
        display: flex;
        overflow: hidden !important;
        transform: translateZ(0);
      }

      :host([hidden]) {
        display: none !important;
      }

      :host([orientation="vertical"]) {
        flex-direction: column;
      }

      :host ::slotted(*) {
        flex: 1 1 auto;
        overflow: auto;
        -webkit-overflow-scrolling: touch;
      }

      [part="splitter"] {
        flex: none;
        position: relative;
        z-index: 1;
        overflow: visible;
        min-width: 8px;
        min-height: 8px;
      }

      :host(:not([orientation="vertical"])) > [part="splitter"] {
        cursor: ew-resize;
      }

      :host([orientation="vertical"]) > [part="splitter"] {
        cursor: ns-resize;
      }

      [part="handle"] {
        width: 40px;
        height: 40px;
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate3d(-50%, -50%, 0);
      }
    </style>
    <slot id="primary" name="primary"></slot>
    <div part="splitter" id="splitter" on-track="_onHandleTrack" on-down="_setPointerEventsNone" on-up="_restorePointerEvents">
      <div part="handle"></div>
    </div>
    <slot id="secondary" name="secondary"></slot>

    <div hidden="">
      <!-- Workaround to fix a Shady style scoping issue caused by dynamic slot naming of the child elements (primary/secondary) -->
      <slot></slot>
    </div>
`}static get is(){return"vaadin-split-layout"}static get version(){return"4.1.1"}static get properties(){return{orientation:{type:String,reflectToAttribute:!0,value:"horizontal"},_previousPrimaryPointerEvents:String,_previousSecondaryPointerEvents:String}}ready(){super.ready(),new s.FlattenedNodesObserver(this,this._processChildren)}_processChildren(){this.getEffectiveChildren().forEach((e,t)=>{0===t?(this._primaryChild=e,e.setAttribute("slot","primary")):1==t?(this._secondaryChild=e,e.setAttribute("slot","secondary")):e.removeAttribute("slot")})}_setFlexBasis(e,t,i){0===(t=Math.max(0,Math.min(t,i)))&&(t=1e-6),e.style.flex="1 1 "+t+"px"}_setPointerEventsNone(e){this._primaryChild&&this._secondaryChild&&(this._previousPrimaryPointerEvents=this._primaryChild.style.pointerEvents,this._previousSecondaryPointerEvents=this._secondaryChild.style.pointerEvents,this._primaryChild.style.pointerEvents="none",this._secondaryChild.style.pointerEvents="none",e.preventDefault())}_restorePointerEvents(){this._primaryChild&&this._secondaryChild&&(this._primaryChild.style.pointerEvents=this._previousPrimaryPointerEvents,this._secondaryChild.style.pointerEvents=this._previousSecondaryPointerEvents)}_onHandleTrack(e){if(this._primaryChild&&this._secondaryChild){var t="vertical"===this.orientation?"height":"width";if("start"!==e.detail.state){var i="vertical"===this.orientation?e.detail.dy:e.detail.dx;this._setFlexBasis(this._primaryChild,this._startSize.primary+i,this._startSize.container),this._setFlexBasis(this._secondaryChild,this._startSize.secondary-i,this._startSize.container),this.notifyResize(),"end"===e.detail.state&&(this.dispatchEvent(new CustomEvent("splitter-dragend")),delete this._startSize)}else this._startSize={container:this.getBoundingClientRect()[t]-this.$.splitter.getBoundingClientRect()[t],primary:this._primaryChild.getBoundingClientRect()[t],secondary:this._secondaryChild.getBoundingClientRect()[t]}}}}e.SplitLayoutElement=SplitLayoutElement,customElements.define(SplitLayoutElement.is,SplitLayoutElement)}));
