// @lrnwebcomponents/lrndesign-mapmenu/lib/lrndesign-mapmenu-item.js
define(["exports", "../../../lit-element/lit-element.js", "../../simple-icon/simple-icon.js", "../../simple-icon/lib/simple-icons.js"], function (_exports, _litElement, _simpleIcon, _simpleIcons) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrndesignMapmenuItem = void 0;

  class LrndesignMapmenuItem extends _litElement.LitElement {
    /**
     * LitElement constructable styles enhancement
     */
    static get styles() {
      return [(0, _litElement.css)`
        :host {
          display: block;
          --lrndesign-mapmenu-item-height: 16px;
        }
        simple-icon {
          --simple-icon-height: var(--lrndesign-mapmenu-item-height);
          display: inline-flex;
        }
      `];
    }

    render() {
      return (0, _litElement.html)`
      ${this.icon ? (0, _litElement.html)` <simple-icon icon="${this.icon}"></simple-icon> ` : ``}
      <span id="title">${this.title}</span>
    `;
    }

    static get tag() {
      return "lrndesign-mapmenu-item";
    }

    constructor() {
      super();
      this.icon = "";
      this.title = "";
      this.url = "";
    }

    static get properties() {
      return {
        icon: {
          type: String
        },
        title: {
          type: String
        },
        url: {
          type: String
        }
      };
    }

  }

  _exports.LrndesignMapmenuItem = LrndesignMapmenuItem;
  window.customElements.define(LrndesignMapmenuItem.tag, LrndesignMapmenuItem);
});
