// @lrnwebcomponents/hax-text-editor/lib/hax-text-editor-inline.js
define(["exports", "../../../@polymer/polymer/polymer-element.js", "../../rich-text-editor/lib/buttons/rich-text-editor-prompt-button.js"], function (_exports, _polymerElement, _richTextEditorPromptButton) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HaxTextEditorInline = void 0;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `hax-text-editor-inline`
   * @element hax-text-editor-inline
   * `a button for inline HAX widgets in hax-text-editor (custom buttons can extend this)`
   *
   * @microcopy - language worth noting:
   *  -
   *
  
   * @polymer
   */
  var HaxTextEditorInline = /*#__PURE__*/function (_RichTextEditorPrompt) {
    babelHelpers.inherits(HaxTextEditorInline, _RichTextEditorPrompt);

    var _super = _createSuper(HaxTextEditorInline);

    function HaxTextEditorInline() {
      var _this;

      babelHelpers.classCallCheck(this, HaxTextEditorInline);
      _this = _super.call(this);
      _this.inlineWidget = true;
      _this.toggles = true;
      _this.value = {
        "": null
      };
      return _this;
    } // properties available to the custom element for data binding


    babelHelpers.createClass(HaxTextEditorInline, [{
      key: "_getTag",
      value: function _getTag(element) {
        return element && element.tag ? element.tag : null;
      }
    }, {
      key: "_getIcon",
      value: function _getIcon(element) {
        return element && element.haxProperties && element.haxProperties.gizmo ? element.haxProperties.gizmo.icon : null;
      }
    }, {
      key: "_getLabel",
      value: function _getLabel(element) {
        return element && element.haxProperties && element.haxProperties.gizmo ? element.haxProperties.gizmo.title : null;
      }
    }, {
      key: "_getFields",
      value: function _getFields(element) {
        return element && element.haxProperties && element.haxProperties.settings ? element.haxProperties.settings.configure : null;
      }
    }], [{
      key: "properties",
      get: function get() {
        return {
          element: {
            type: Object,
            value: null
          },
          tag: {
            type: String,
            computed: "_getTag(element)"
          },
          icon: {
            type: String,
            computed: "_getIcon(element)"
          },
          label: {
            type: String,
            computed: "_getLabel(element)"
          },
          fields: {
            type: String,
            computed: "_getFields(element)"
          }
        };
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       */

    }, {
      key: "tag",
      get: function get() {
        return "hax-text-editor-inline";
      }
    }]);
    return HaxTextEditorInline;
  }(_richTextEditorPromptButton.RichTextEditorPromptButton);

  _exports.HaxTextEditorInline = HaxTextEditorInline;
  window.customElements.define(HaxTextEditorInline.tag, HaxTextEditorInline);
});
