// @lrnwebcomponents/lrndesign-course-banner/lrndesign-course-banner.js
define(["exports","require","../../@polymer/polymer/polymer-element.js"],(function(e,r,i){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.LrndesignCourseBanner=void 0,r=babelHelpers.interopRequireWildcard(r);class LrndesignCourseBanner extends i.PolymerElement{constructor(){super(),new Promise((e,i)=>r.default(["../../@polymer/iron-image/iron-image.js"],e,i)),new Promise((e,i)=>r.default(["../lrndesign-avatar/lrndesign-avatar.js"],e,i))}static get template(){return i.html`
      <style>
        :host {
          display: block;
        }
        /**
       * Dialog
       */
        .course-image {
        }
        .course-heading {
          position: relative;
          background-color: rgba(30, 30, 30, 0.8);
          text-align: left;
          margin: -5em 0 0 0;
          padding: 16px;
          color: #ffffff;
          height: 64px;
        }
        .course-avatar {
          float: left;
          display: inline-flex;
          padding: 0 16px 0 0;
        }
        .course-name {
          font-size: 16px;
          line-height: 16px;
          min-width: 96px;
        }
        .course-title {
          font-size: 16px;
          line-height: 16px;
          display: none;
        }
        .name-wrapper {
          display: flow-root;
          overflow: hidden;
          text-overflow: clip;
        }
        @media screen and (min-width: 420px) {
          .course-name {
            font-size: 24px;
          }
          .course-title {
            display: block;
          }
        }
      </style>
      <iron-image
        class="course-image"
        style="width:100%; height:200px; background-color: lightgray;"
        sizing="cover"
        preload=""
        fade=""
        src\$="[[image]]"
      ></iron-image>
      <div class="course-heading">
        <lrndesign-avatar
          class="course-avatar"
          label="[[name]]"
          jdenticon=""
          color="[[color]]"
        >
        </lrndesign-avatar>
        <div class="name-wrapper">
          <div class="course-name">[[name]]</div>
          <div class="course-title">[[title]]</div>
        </div>
      </div>
    `}static get tag(){return"lrndesign-course-banner"}static get properties(){return{color:{type:String},image:{type:String},name:{type:String},title:{type:String}}}}e.LrndesignCourseBanner=LrndesignCourseBanner,window.customElements.define(LrndesignCourseBanner.tag,LrndesignCourseBanner)}));
