// @lrnwebcomponents/responsive-grid/lib/responsive-grid-col.js
define(["exports","../../../@polymer/polymer/polymer-element.js"],function(_exports,_polymerElement){"use strict";Object.defineProperty(_exports,"__esModule",{value:!0});_exports.ResponsiveGridCol=void 0;/**    
`responsive-grid-col`
A column for the responsive grid

* @demo demo/index.html

@microcopy - the mental model for this element
  Example: 
  <responsive-grid-col 
    xs="1"    //the number of columns on an extra-small-width screen
    md="2"    //the number of columns on a small-width screen
    lg="3"    //the number of columns on a medium-width screen
    xl="4">   //the number of columns on a large-width screen
    CONTENT HERE
  </responsive-grid-col>

*/class ResponsiveGridCol extends _polymerElement.PolymerElement{static get template(){return _polymerElement.html`
      <style>
        :host {
          position: relative;
          min-height: 1px;
          float: left;
        }
        :host:after {
          clear: both;
        }
        #col-inner {
          padding-left: 15px;
          padding-right: 15px;
          @apply --responsive-grid-col-inner;
        }
        /* Hide Print-Only on Screen */
        @media screen {
          :host([print-only]) {
            display: none;
          }
        }
        /* Hide Screen-Only on Print */
        @media print {
          :host([screen-only]) {
            display: none;
          }
        }
      </style>
      <div id="col-inner"><slot></slot></div>
    `}static get tag(){return"responsive-grid-col"}static get properties(){return{/**
       * the width when viewed on an extra large screen
       */xl:{type:Number,value:1},/**
       * the width when viewed on a large screen
       */lg:{type:Number,value:1},/**
       * the width when viewed on a medium screen
       */md:{type:Number,value:1},/**
       * the width when viewed on a small screen
       */sm:{type:Number,value:1},/**
       * the width when viewed on an extra-small screen
       */xs:{type:Number,value:1}}}}_exports.ResponsiveGridCol=ResponsiveGridCol;window.customElements.define(ResponsiveGridCol.tag,ResponsiveGridCol)});
