// @lrnwebcomponents/hax-text-editor/hax-text-editor.js
define(["exports", "../../@polymer/polymer/polymer-element.js", "../rich-text-editor/rich-text-editor.js", "./lib/hax-text-editor-toolbar.js"], function (_exports, _polymerElement, _richTextEditor, _haxTextEditorToolbar) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HaxTextEditor = void 0;

  /**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `hax-text-editor`
   * @element hax-text-editor
   * `rich-text-editor configured for HAX`
   *
   * @microcopy - language worth noting:
   *  -
   *
  
   * @polymer
   * @demo demo/index.html
   */
  class HaxTextEditor extends _richTextEditor.RichTextEditor {
    // render function
    render() {
      return (0, _polymerElement.html)` <style></style>
      ${super.template}`;
    } // properties available to the custom element for data binding


    static get properties() {
      return { ...super.properties
      };
    }

    constructor() {
      super();
      this.type = "hax-text-editor-toolbar";
    }
    /**
     * Store the tag name to make it easier to obtain directly.
     * @notice function name must be here for tooling to operate correctly
     */


    static get tag() {
      return "hax-text-editor";
    }
    /**
     * life cycle, element is afixed to the DOM
     */


    connectedCallback() {
      super.connectedCallback();
    }
    /**
     * life cycle, element is removed from the DOM
     */
    //disconnectedCallback() {}


  }

  _exports.HaxTextEditor = HaxTextEditor;
  window.customElements.define(HaxTextEditor.tag, HaxTextEditor);
});
