diff --git a/eim.module b/eim.module
index 9f4512e..6d23f30 100644
--- a/eim.module
+++ b/eim.module
@@ -127,7 +127,7 @@ function eim_imagefield_widget_process($element, &$form_state, $form) {
 function _eim_required_fields_validate($element, &$form_state) {
   // Only do validation if the function is triggered from other places than
   // the image process form.
-  if (!in_array('file_managed_file_submit', $form_state['triggering_element']['#submit'])) {
+  if (isset($form_state['triggering_element']['#submit']) && !in_array('file_managed_file_submit', $form_state['triggering_element']['#submit'])) {
     // If the image is not there, we do not check for empty values.
     $parents = $element['#parents'];
     $field = array_pop($parents);
