// @lrnwebcomponents/license-element/license-element.js
define(["exports", "../../@polymer/polymer/polymer-element.js", "../../@polymer/polymer/lib/utils/render-status.js", "../hax-body-behaviors/lib/HAXWiring.js", "../schema-behaviors/schema-behaviors.js"], function (_exports, _polymerElement, _renderStatus, _HAXWiring, _schemaBehaviors) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LicenseElement = _exports.licenseList = void 0;

  function _templateObject_22909be0b49511e9bfd4f79a8a57e0f7() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: block;\n          font-size: 16px;\n          line-height: 32px;\n          --license-text-color: #222222;\n          --license-background-color: #f2f2f2;\n          background-color: var(--license-background-color);\n        }\n        :host:after {\n          content: \"License\";\n          position: relative;\n          float: right;\n          bottom: 36px;\n          right: 8px;\n          font-size: 28px;\n          color: #dddddd;\n          font-style: italic;\n        }\n        .license-body {\n          padding: 32px;\n          font-style: italic;\n          background-color: var(--license-background-color);\n          color: var(--license-text-color);\n          @apply --license-body;\n        }\n\n        :host([display-method=\"footnote\"]) {\n          visibility: hidden;\n          opacity: 0;\n        }\n        :host([display-method=\"popup\"]) {\n          display: block;\n        }\n        .license-link {\n          font-style: italic;\n          @apply --license-link;\n        }\n        .big-license-link img {\n          margin-right: 8px;\n          margin-bottom: 8px;\n          width: 88px;\n          height: 31px;\n          vertical-align: middle;\n        }\n        .work-title {\n          font-weight: bold;\n        }\n      </style>\n      <meta\n        rel=\"cc:license\"\n        href$=\"[[licenseLink]]\"\n        content$=\"License: [[licenseName]]\"\n      />\n      <div\n        class=\"license-body\"\n        xmlns:cc$=\"[[licenseLink]]\"\n        xmlns:dc=\"http://purl.org/dc/elements/1.1/\"\n      >\n        <a class=\"big-license-link\" target=\"_blank\" href=\"[[licenseLink]]\"\n          ><img\n            alt=\"[[licenseName]] graphic\"\n            src=\"[[licenseImage]]\"\n            hidden$=\"[[!licenseImage]]\"\n        /></a>\n        <span\n          class=\"work-title\"\n          rel=\"dc:type\"\n          href=\"http://purl.org/dc/dcmitype/Text\"\n          property=\"dc:title\"\n          >[[title]]</span\n        >\n        by\n        <a\n          rel=\"cc:attributionURL\"\n          property=\"cc:attributionName\"\n          href$=\"[[source]]\"\n          >[[creator]]</a\n        >\n        is licensed under a\n        <a class=\"license-link\" target=\"_blank\" href=\"[[licenseLink]]\"\n          >[[licenseName]]</a\n        >. <span rel=\"dc:source\" href$=\"[[source]]\"></span>\n\n        <template is=\"dom-if\" if=\"[[hasMore]]\">\n          <span\n            >Permissions beyond the scope of this license are available\n            <a rel=\"cc:morePermissions\" target=\"_blank\" href=\"[[moreLink]]\"\n              >[[moreLabel]]</a\n            >.</span\n          >\n        </template>\n      </div>\n    "], ["\n      <style>\n        :host {\n          display: block;\n          font-size: 16px;\n          line-height: 32px;\n          --license-text-color: #222222;\n          --license-background-color: #f2f2f2;\n          background-color: var(--license-background-color);\n        }\n        :host:after {\n          content: \"License\";\n          position: relative;\n          float: right;\n          bottom: 36px;\n          right: 8px;\n          font-size: 28px;\n          color: #dddddd;\n          font-style: italic;\n        }\n        .license-body {\n          padding: 32px;\n          font-style: italic;\n          background-color: var(--license-background-color);\n          color: var(--license-text-color);\n          @apply --license-body;\n        }\n\n        :host([display-method=\"footnote\"]) {\n          visibility: hidden;\n          opacity: 0;\n        }\n        :host([display-method=\"popup\"]) {\n          display: block;\n        }\n        .license-link {\n          font-style: italic;\n          @apply --license-link;\n        }\n        .big-license-link img {\n          margin-right: 8px;\n          margin-bottom: 8px;\n          width: 88px;\n          height: 31px;\n          vertical-align: middle;\n        }\n        .work-title {\n          font-weight: bold;\n        }\n      </style>\n      <meta\n        rel=\"cc:license\"\n        href\\$=\"[[licenseLink]]\"\n        content\\$=\"License: [[licenseName]]\"\n      />\n      <div\n        class=\"license-body\"\n        xmlns:cc\\$=\"[[licenseLink]]\"\n        xmlns:dc=\"http://purl.org/dc/elements/1.1/\"\n      >\n        <a class=\"big-license-link\" target=\"_blank\" href=\"[[licenseLink]]\"\n          ><img\n            alt=\"[[licenseName]] graphic\"\n            src=\"[[licenseImage]]\"\n            hidden\\$=\"[[!licenseImage]]\"\n        /></a>\n        <span\n          class=\"work-title\"\n          rel=\"dc:type\"\n          href=\"http://purl.org/dc/dcmitype/Text\"\n          property=\"dc:title\"\n          >[[title]]</span\n        >\n        by\n        <a\n          rel=\"cc:attributionURL\"\n          property=\"cc:attributionName\"\n          href\\$=\"[[source]]\"\n          >[[creator]]</a\n        >\n        is licensed under a\n        <a class=\"license-link\" target=\"_blank\" href=\"[[licenseLink]]\"\n          >[[licenseName]]</a\n        >. <span rel=\"dc:source\" href\\$=\"[[source]]\"></span>\n\n        <template is=\"dom-if\" if=\"[[hasMore]]\">\n          <span\n            >Permissions beyond the scope of this license are available\n            <a rel=\"cc:morePermissions\" target=\"_blank\" href=\"[[moreLink]]\"\n              >[[moreLabel]]</a\n            >.</span\n          >\n        </template>\n      </div>\n    "]);

    _templateObject_22909be0b49511e9bfd4f79a8a57e0f7 = function _templateObject_22909be0b49511e9bfd4f79a8a57e0f7() {
      return data;
    };

    return data;
  }

  /**
   * A list of licenses that we support the references for.
   */
  var licenseList = function licenseList() {
    var mode = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "full";
    babelHelpers.classCallCheck(this, licenseList);
    // initial list, PR to add more
    var list = {
      by: {
        name: "Attribution",
        link: "https://creativecommons.org/licenses/by/4.0/",
        image: "https://i.creativecommons.org/l/by/4.0/88x31.png"
      },
      "by-sa": {
        name: "Attribution Share a like",
        link: "https://creativecommons.org/licenses/by-sa/4.0/",
        image: "https://i.creativecommons.org/l/by-sa/4.0/88x31.png"
      },
      "by-nd": {
        name: "Attribution No derivatives",
        link: "https://creativecommons.org/licenses/by-nd/4.0/",
        image: "https://i.creativecommons.org/l/by-nd/4.0/88x31.png"
      },
      "by-nc": {
        name: "Attribution non-commercial",
        link: "https://creativecommons.org/licenses/by-nc/4.0/",
        image: "https://i.creativecommons.org/l/by-nc/4.0/88x31.png"
      },
      "by-nc-sa": {
        name: "Attribution non-commercial share a like",
        link: "https://creativecommons.org/licenses/by-nc-sa/4.0/",
        image: "https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png"
      },
      "by-nc-nd": {
        name: "Attribution Non-commercial No derivatives",
        link: "https://creativecommons.org/licenses/by-nc-nd/4.0/",
        image: "https://i.creativecommons.org/l/by-nc-nd/4.0/88x31.png"
      }
    }; // support mutating the array into a select list

    if (mode == "select") {
      var select = {};

      for (var i in list) {
        select[i] = list[i].name;
      }

      return select;
    }

    return list;
  };

  _exports.licenseList = licenseList;

  /**
   * `license-element`
   * `A simple way of applying a semantically accurate license to work.`
   * @demo demo/index.html
   * @microcopy - the mental model for this element
   */
  var LicenseElement =
  /*#__PURE__*/
  function (_SchemaBehaviors) {
    babelHelpers.inherits(LicenseElement, _SchemaBehaviors);
    babelHelpers.createClass(LicenseElement, null, [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_22909be0b49511e9bfd4f79a8a57e0f7());
      }
    }, {
      key: "tag",
      get: function get() {
        return "license-element";
      }
    }, {
      key: "properties",
      get: function get() {
        var props = {
          /**
           * Title of the work.
           */
          title: {
            type: String
          },

          /**
           * Person or group that owns / created the work.
           */
          creator: {
            type: String
          },

          /**
           * Original Source of the work in question
           */
          source: {
            type: String
          },

          /**
           * License name, calculated or supplied by the end user if we don't have them.
           */
          licenseName: {
            type: String
          },

          /**
           * License link for more details
           */
          licenseLink: {
            type: String
          },

          /**
           * License short hand. Options cc0,
           */
          license: {
            type: String,
            observer: "_licenseUpdated"
          },

          /**
           * More details label
           */
          moreLabel: {
            type: String,
            value: "on the licensing details page"
          },

          /**
           * More details link
           */
          moreLink: {
            type: String
          },

          /**
           * See if we have more things to point to
           */
          hasMore: {
            type: Boolean,
            computed: "_computeHasMore(moreLink)"
          }
        };

        if (babelHelpers.get(babelHelpers.getPrototypeOf(LicenseElement), "properties", this)) {
          props = Object.assign(props, babelHelpers.get(babelHelpers.getPrototypeOf(LicenseElement), "properties", this));
        }

        return props;
      }
    }]);

    function LicenseElement() {
      var _this;

      babelHelpers.classCallCheck(this, LicenseElement);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LicenseElement).call(this));
      _this.licenseList = new licenseList();
      (0, _renderStatus.afterNextRender)(babelHelpers.assertThisInitialized(_this), function () {
        this.HAXWiring = new _HAXWiring.HAXWiring();
        this.HAXWiring.setup(LicenseElement.haxProperties, LicenseElement.tag, this);
      });
      return _this;
    }

    babelHelpers.createClass(LicenseElement, [{
      key: "_computeHasMore",

      /**
       * Calculate if we should show the advanced details area
       */
      value: function _computeHasMore(link) {
        // only show if there's a link supplied for additional details
        if (babelHelpers.typeof(link) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined)) && link !== "") {
          return true;
        }

        return false;
      }
      /**
       * License was updated, time to update license name and link.
       */

    }, {
      key: "_licenseUpdated",
      value: function _licenseUpdated(newValue, oldValue) {
        if (babelHelpers.typeof(newValue) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined)) && babelHelpers.typeof(this.licenseList[newValue]) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          this.licenseName = this.licenseList[newValue].name;
          this.licenseLink = this.licenseList[newValue].link;
          this.licenseImage = this.licenseList[newValue].image;
        }
      }
    }], [{
      key: "haxProperties",
      get: function get() {
        return {
          canScale: false,
          canPosition: false,
          canEditSource: false,
          gizmo: {
            title: "License",
            description: "Provide a license for you rwork",
            icon: "icons:theaters",
            color: "grey",
            groups: ["Content", "Copyright"],
            handles: [{
              type: "license",
              source: "source",
              title: "title",
              author: "creator",
              license: "license"
            }],
            meta: {
              author: "LRNWebComponents"
            }
          },
          settings: {
            quick: [{
              property: "title",
              title: "Title",
              description: "The title of the work being cited.",
              inputMethod: "textfield",
              icon: "editor:title"
            }],
            configure: [{
              property: "title",
              title: "Title",
              description: "The title of the work being cited.",
              inputMethod: "textfield",
              icon: "editor:title"
            }, {
              property: "source",
              title: "Source link",
              description: "The source url for the element this is citing.",
              inputMethod: "textfield",
              icon: "link",
              validationType: "url"
            }, {
              property: "license",
              title: "License",
              description: "The source url for the element this is citing.",
              inputMethod: "select",
              options: new licenseList("select"),
              icon: "link"
            }, {
              property: "creator",
              title: "Creator",
              description: "Who made or owns this.",
              inputMethod: "textfield",
              icon: "link"
            }],
            advanced: [{
              property: "moreLink",
              title: "Source link",
              description: "Link to additional licensing details",
              inputMethod: "textfield",
              validationType: "url"
            }, {
              property: "moreLabel",
              title: "more label",
              description: "Label for more licensing details",
              inputMethod: "textfield"
            }]
          }
        };
      }
    }]);
    return LicenseElement;
  }((0, _schemaBehaviors.SchemaBehaviors)(_polymerElement.PolymerElement));

  _exports.LicenseElement = LicenseElement;
  window.customElements.define(LicenseElement.tag, LicenseElement);
});
