// @lrnwebcomponents/agency-theme/lib/agency-theme-spotlight.js
define(["exports","../../../lit-element/lit-element.js"],(function(_exports,_litElement){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.AgencyThemeSpotlight=void 0;
/**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class AgencyThemeSpotlight extends _litElement.LitElement{static get styles(){return[_litElement.css`
        :host {
          --agency-theme-spotlight-image-height: 10em;
          --agency-theme-spotlight-content-padding: 3vw;
          --agency-theme-spotlight-content-color: white;
          display: block;
          color: white;
        }

        :host([hidden]) {
          display: none;
        }

        .inner {
          display: flex;
          flex-wrap: wrap;
        }

        .image {
          display: flex;
          width: 100%;
          flex: 0 1 auto;
          justify-content: center;
        }

        .image img {
          border-radius: 100%;
          object-fit: cover;
          width: var(--agency-theme-spotlight-image-height);
          height: var(--agency-theme-spotlight-image-height);
        }

        @media screen and (min-width: 500px) {
          .inner {
            flex-wrap: nowrap;
          }
          .image {
            width: auto;
          }
          .image + .content {
            margin-left: var(--agency-theme-spotlight-content-padding);
          }
        }

        .content {
          width: 100%;
          flex: 1 1 auto;
          -ms-flex: 2;
        }
        .title {
          border-bottom: 1px solid rgba(0, 0, 0, 0.2);
          margin-bottom: 1em;
        }
      `]}static get tag(){return"agency-theme-spotlight"}static get properties(){return{image:{type:String}}}render(){return _litElement.html`
      <div class="inner">
        <div class="image">
          <img loading="lazy" src="${this.image}" alt="" />
        </div>
        <div class="content">
          <div class="title"><slot name="title"></slot></div>
          <slot></slot>
        </div>
      </div>
    `}}_exports.AgencyThemeSpotlight=AgencyThemeSpotlight,window.customElements.define(AgencyThemeSpotlight.tag,AgencyThemeSpotlight)}));
