// @lrnwebcomponents/simple-colors/lib/demo/simple-colors-swatches.js
define(["exports","require","../../../../lit-element/lit-element.js","../../simple-colors.js"],(function(t,e,s,o){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.simpleColorsSwatches=void 0,e=babelHelpers.interopRequireWildcard(e);
/**
   * Copyright 2018 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class simpleColorsSwatches extends o.SimpleColors{static get styles(){return[...super.styles,s.css`
        :host {
          display: block;
          margin: 15px 0;
        }
        :host([hidden]) {
          display: none;
        }
        :host simple-modal-template {
          --simple-modal-titlebar-color: var(--simple-colors-swatch-contrast);
          --simple-modal-titlebar-background: var(--simple-colors-swatch-color);
        }
        :host .row {
          display: flex;
          align-items: stretch;
        }
        :host button {
          width: calc(8% - 6px);
          height: 80px;
          margin: 3px;
          border: 1px solid black;
          border-radius: 3px;
          padding: 0;
          display: flex;
          flex-direction: column;
          justify-content: flex-end;
          align-items: stretch;
        }
        :host button div {
          margin: 0;
          padding: 2px;
          font-size: 10px;
          background: black;
          color: white;
          width: calc(100% - 4px);
        }
        :host button:focus,
        :host button:hover {
          opacity: 0.7;
          cursor: pointer;
        }
      `]}render(){return s.html`
      <div>
        <p>Click on the swatches above to learn more.</p>
        <div class="rows">
          ${Object.keys(this.colors).map(t=>s.html`
              <div class="row">
                ${this.colors.grey.map((e,o)=>s.html`
                    <button
                      id="${this._getId(t,o)}"
                      controls="swatchinfo"
                      style="background-color: ${this.colors[t][o]};"
                      title="${this._getTitle(t,o)}"
                      @click="${e=>this._handleClick(t,o)}"
                    >
                      <div>${this._getTitle(t,o)}</div>
                    </button>
                  `)}
              </div>
            `)}
        </div>
        <simple-modal-template id="modal" title="${this.swatchName}">
          <p slot="content">
            <simple-colors-swatch-info
              swatch-id="${this.swatchId}"
              swatch-name="${this.swatchName}"
            >
            </simple-colors-swatch-info>
          </p>
        </simple-modal-template>
      </div>
    `}static get properties(){return{swatchId:{attribute:"swatch-id",type:String},swatchName:{attribute:"swatch-name",type:String}}}static get tag(){return"simple-colors-swatches"}updated(t){t.forEach((t,e)=>{"swatchId"===e&&this.shadowRoot.querySelector("#modal")&&this.shadowRoot.querySelector("#modal").openModal&&this.shadowRoot.querySelector("#modal").openModal(this.shadowRoot.querySelector("#"+this.swatchId))})}constructor(){super(),this.swatchId="grey_0",this.swatchName="grey-1",new Promise((t,s)=>e.default(["./simple-colors-swatch-info.js"],t,s)),new Promise((t,s)=>e.default(["../../../simple-modal/lib/simple-modal-template.js"],t,s))}_getTitle(t,e){return`${t}-${this._getShade(e)}`}_getId(t,e){return`${t}_${e}`}_getShade(t){return parseInt(t)+1}_handleClick(t,e){this.swatchId=this._getId(t,e),this.swatchName=this._getTitle(t,e)}}t.simpleColorsSwatches=simpleColorsSwatches,window.customElements.define(simpleColorsSwatches.tag,simpleColorsSwatches)}));
