// @lrnwebcomponents/lorem-data/lorem-data.js
define(["exports", "require", "../../lit-element/lit-element.js"], function (_exports, _require, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LoremData = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject6_2f6c8150d1d911ea92952ff2097cf814() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          font-size: var(--lorem-data-FontSize, 14px);\n          line-height: var(--lorem-data-LineHeight, 160%);\n          color: var(--lorem-data-Color, #95989a);\n        }\n        :host([hidden]) {\n          display: none;\n        }\n        #threads {\n          background-color: var(--lorem-data-threads-BackgroundColor, #eaeaea);\n          padding: 1px;\n          width: calc(100% - 2px);\n          margin: var(--lorem-data-threads-Margin, 0);\n        }\n        #new-thread {\n          background-color: var(\n            --lorem-data-new-thread-BackgroundColor,\n            var(--lorem-data-threads-BackgroundColor, #eaeaea)\n          );\n          padding: 2px 1px;\n          width: calc(100% - 2px);\n          margin: var(\n            --lorem-data-new-thread-Margin,\n            var(--lorem-data-threads-Margin, 0)\n          );\n        }\n        .thread:focus-within,\n        #new-thread:focus-within {\n          border-left: var(\n            --lorem-data-thread-focus-BorderLeft,\n            4px solid blue\n          );\n        }\n        #new-thread:focus-within {\n          border-left: var(--lorem-data-new-thread-focus-BorderLeft);\n        }\n        .thread > *,\n        #new-thread > * {\n          background-color: var(--lorem-data-comment-BackgroundColor, #fff);\n        }\n        .comment {\n          margin: 1px 1px 0;\n          padding: var(--lorem-data-comment-Padding, 10px) 0;\n        }\n        .reply-form {\n          margin: 0;\n          width: calc(\n            100% -\n              var(\n                --lorem-data-reply-indent,\n                calc(2 * var(--lorem-data-comment-Padding, 10px))\n              )\n          );\n        }\n        .reply-form,\n        .comment-reply {\n          margin-left: var(\n            --lorem-data-reply-indent,\n            calc(2 * var(--lorem-data-comment-Padding, 10px))\n          );\n        }\n        .comment-header {\n          display: flex;\n          align-items: stretch;\n          justify-content: space-between;\n          margin-bottom: var(--lorem-data-comment-Padding, 10px);\n        }\n        .comment-header,\n        .comment-body {\n          padding: 0 var(--lorem-data-comment-Padding, 10px);\n        }\n        lrndesign-avatar {\n          margin-right: var(--lorem-data-comment-Padding, 10px);\n        }\n        .comment-header > div {\n          display: flex;\n          flex-direction: column;\n          justify-content: space-evenly;\n          align-items: flex-start;\n          flex: 1 1 auto;\n        }\n        .comment-name {\n          margin: 0;\n          font-size: var(\n            --lorem-data-name-FontSize,\n            calc(1.1 * var(--lorem-data-FontSize, 14px))\n          );\n          font-weight: var(--lorem-data-name-FontWeight, bold);\n          font-family: var(--lorem-data-name-FontFamily);\n          color: var(--lorem-data-name-Color, #4b4b4b);\n        }\n        .comment-date {\n          margin: 0;\n          font-size: var(\n            --lorem-data-date-FontSize,\n            calc(0.8 * var(--lorem-data-FontSize, 14px))\n          );\n          font-weight: var(--lorem-data-date-FontWeight, normal);\n          font-family: var(--lorem-data-date-FontFamily);\n          color: var(--lorem-data-date-Color, var(--lorem-data-Color, #95989a));\n        }\n        .comment-body {\n          font-size: var(--lorem-data-FontSize, 14px);\n        }\n      "]);

    _templateObject6_2f6c8150d1d911ea92952ff2097cf814 = function _templateObject6_2f6c8150d1d911ea92952ff2097cf814() {
      return data;
    };

    return data;
  }

  function _templateObject5_2f6c8150d1d911ea92952ff2097cf814() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <div class=\"thread\">\n              ", "\n              ", "\n              <lorem-data-form\n                button-label=\"", "\"\n                class=\"reply-form\"\n                @comment-demo=\"", "\"\n                @comment-submitted=\"", "\"\n                ?demo=\"", "\"\n                field=\"", "\"\n                ?disabled=\"", "\"\n                ?hidden=\"", "\"\n                .icon=\"", "\"\n                .submit=\"", "\"\n                textarea-label=\"", "\"\n                .thread=\"", "\"\n              >\n              </lorem-data-form>\n            </div>\n          "]);

    _templateObject5_2f6c8150d1d911ea92952ff2097cf814 = function _templateObject5_2f6c8150d1d911ea92952ff2097cf814() {
      return data;
    };

    return data;
  }

  function _templateObject4_2f6c8150d1d911ea92952ff2097cf814() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div id=\"threads\">\n        <div class=\"thread\" ?hidden=\"", "\">\n          ", "\n        </div>\n        ", "\n      </div>\n    "]);

    _templateObject4_2f6c8150d1d911ea92952ff2097cf814 = function _templateObject4_2f6c8150d1d911ea92952ff2097cf814() {
      return data;
    };

    return data;
  }

  function _templateObject3_2f6c8150d1d911ea92952ff2097cf814() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                <p>", "</p>\n              "]);

    _templateObject3_2f6c8150d1d911ea92952ff2097cf814 = function _templateObject3_2f6c8150d1d911ea92952ff2097cf814() {
      return data;
    };

    return data;
  }

  function _templateObject2_2f6c8150d1d911ea92952ff2097cf814() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div\n        aria-describedby=\"", "\"\n        class=\"comment ", "\"\n        id=\"comment-", "\"\n      >\n        <div class=\"comment-header\">\n          <lrndesign-avatar\n            .accent-color=\"", "\"\n            .label=\"", " ", "\"\n            .src=\"", "\"\n            two-chars\n          >\n          </lrndesign-avatar>\n          <div>\n            <p class=\"comment-name\">", " ", "</p>\n            <p class=\"comment-date\">", "</p>\n          </div>\n        </div>\n        <div class=\"comment-body\">\n          ", "\n        </div>\n      </div>\n    "]);

    _templateObject2_2f6c8150d1d911ea92952ff2097cf814 = function _templateObject2_2f6c8150d1d911ea92952ff2097cf814() {
      return data;
    };

    return data;
  }

  function _templateObject_2f6c8150d1d911ea92952ff2097cf814() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <slot name=\"before-discussion\"></slot>\n      ", "\n      <div id=\"new-thread\">\n        <slot name=\"before-new-thread\"></slot>\n        <lorem-data-form\n          button-label=\"", "\"\n          class=\"comment-form\"\n          @comment-demo=\"", "\"\n          @comment-submitted=\"", "\"\n          ?demo=\"", "\"\n          field=\"", "\"\n          .icon=\"", "\"\n          ?hidden=\"", "\"\n          ?disabled=\"", "\"\n          .submit=\"", "\"\n          textarea-label=\"", "\"\n        >\n        </lorem-data-form>\n        <slot name=\"after-new-thread\"></slot>\n      </div>\n      ", "\n      <slot name=\"after-discussion\"></slot>\n    "]);

    _templateObject_2f6c8150d1d911ea92952ff2097cf814 = function _templateObject_2f6c8150d1d911ea92952ff2097cf814() {
      return data;
    };

    return data;
  }

  /**
   * `lorem-data`
   * a threaded discussions component
   * 
  ### Styling
  
  `<lorem-data>` provides the following custom properties
  for styling:
  
  Custom property | Description | Default
  ----------------|-------------|----------
  `--lorem-data-FontSize` | default font-size | 14px
  `--lorem-data-LineHeight` | default line-height | 160%
  `--lorem-data-Color` | default text color | #95989a
  `--lorem-data-threads-BackgroundColor` | default thread backbround-color | #eaeaea
  `--lorem-data-threads-Margin` | default thread margin | 0
  `--lorem-data-thread-focus-BorderLeft` | default thread border left color when commenting | 4px solid blue
  `--lorem-data-new-thread-BackgroundColor` | overrides default thread backbround-color for new thread | var(--lorem-data-threads-BackgroundColor, #eaeaea)
  `--lorem-data-new-thread-Margin` | overrides default thread margin for new thread | var(--lorem-data-threads-Margin, 0))
  `--lorem-data-new-thread-focus-BorderLeft` | overrides default border left color when posting a new thread | 
  `--lorem-data-comment-BackgroundColor` | comment bacground color | #fff
  `--lorem-data-comment-Padding` | comment padding | 10px
  `--lorem-data-reply-indent` | left-indent for thread replies | calc(2 * var(--lorem-data-comment-Padding,10px))));
  `--lorem-data-name-FontSize` | comment author name font-size | calc(1.1 * var(--lorem-data-FontSize, 14px)));
  `--lorem-data-name-FontWeight` | comment author name font-weight | bold;
  `--lorem-data-name-FontFamily` | overrides default font-family for comment author name | 
  `--lorem-data-name-Color` | comment author name text color | #4b4b4b);
  `--lorem-data-date-FontSize` | comment author name font-size | calc(0.8 * var(--lorem-data-FontSize, 14px)));
  `--lorem-data-date-FontWeight` | comment author name font-weight | normal
  `--lorem-data-date-FontFamily` | overrides default font-family for comment date | 
  `--lorem-data-date-Color` | comment date text color | #95989a));
   *
   * @element lorem-data
   * @demo ./demo/index.html demo
   */
  var LoremData =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LoremData, _LitElement);
    babelHelpers.createClass(LoremData, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_2f6c8150d1d911ea92952ff2097cf814(), this.latest ? "" : this.threads, this.commentButtonLabel || "Submit", this._handleDemo, this._handleSubmit, this.demo, this.map.body || "body", this.commentIcon, this.hidden, this.disabled, this._getPath(this.submit, this.params), this.commentTextareaLabel || "Enter comment", !this.latest ? "" : this.threads);
      }
    }, {
      key: "getComment",
      value: function getComment(comment, thread) {
        return (0, _litElement.html)(_templateObject2_2f6c8150d1d911ea92952ff2097cf814(), thread || "", thread ? "comment-reply" : "", comment.id, comment.color, comment.firstName, comment.lastName, comment.avatar, comment.firstName, comment.lastName, this._getDate(comment.date), (comment.body || "").split(/[\r\n]+/).map(function (p) {
          return (0, _litElement.html)(_templateObject3_2f6c8150d1d911ea92952ff2097cf814(), p);
        }));
      }
    }, {
      key: "threads",
      get: function get() {
        var _this2 = this;

        return (0, _litElement.html)(_templateObject4_2f6c8150d1d911ea92952ff2097cf814(), !this.__loading, this.loadingText, this.sortData(this.__data).map(function (thread) {
          return (0, _litElement.html)(_templateObject5_2f6c8150d1d911ea92952ff2097cf814(), _this2.getComment(thread), (thread.replies || []).map(function (reply) {
            return _this2.getComment(reply, thread.id);
          }), _this2.replyButtonLabel || "Reply", _this2._handleDemo, _this2._handleSubmit, _this2.demo, _this2.map.replyBody || _this2.map.body || "body", _this2.disabled, _this2.hidden, _this2.replyIcon, _this2._getPath(_this2.submit, _this2._getThreadParams(thread.id)), _this2.replyTextareaLabel || "Enter reply", thread.id);
        }));
      }
    }], [{
      key: "styles",
      get: function get() {
        return [(0, _litElement.css)(_templateObject6_2f6c8150d1d911ea92952ff2097cf814())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lorem-data";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * sort comments latest-first
           */
          latest: {
            type: Boolean,
            attribute: "latest",
            reflect: true
          },

          /**
           * label for comment submit button
           */
          commentButtonLabel: {
            type: String,
            attribute: "comment-button-label"
          },

          /**
           * label for comment textarea
           */
          commentTextareaLabel: {
            type: String,
            attribute: "comment-textarea-label"
          },

          /**
           * optional icon for comment button
           */
          commentIcon: {
            type: String,
            attribute: "comment-icon"
          },

          /**
           * raw data, can be Object or Array
           */
          data: {
            type: Object
          },

          /**
           * locale for date formatting
           */
          dateLocale: {
            type: String
          },

          /**
           * JS format object for dates
           */
          dateFormat: {
            type: Object
          },

          /**
           * discussion in demo mode?
           * Instead of submitting,
           * submission is temporarily inserted into data.
           */
          demo: {
            type: Boolean
          },

          /**
           * text that displays while discussion data loads
           */
          loadingText: {
            type: String,
            attribute: "loading-text"
          },

          /**
           * object that will reformat data,
           * where key is final formatted property,
           * and value is raw data property to map.
           * Example: {avatar: "image"}
           * would use raw data's image property as a avatar
           */
          map: {
            type: Object
          },

          /**
           * query paramerters for getting and sending data
           */
          params: {
            type: Object
          },

          /**
           * label for reply submit button
           */
          replyButtonLabel: {
            type: String,
            attribute: "reply-button-label",
            reflect: true
          },

          /**
           * label for reply textarea
           */
          replyTextareaLabel: {
            type: String,
            attribute: "reply-textarea-label",
            reflect: true
          },

          /**
           * icon for reply submit button
           */
          replyIcon: {
            type: String,
            attribute: "reply-icon",
            reflect: true
          },

          /**
           * path (without parameters) for fetching data
           */
          source: {
            type: String
          },

          /**
           * path (without parameters) for sending data
           */
          submit: {
            type: String
          },

          /**
           * formatted data
           */
          __data: {
            type: Array
          },

          /**
           * whether data is still loading
           */
          __loading: {
            type: Boolean
          }
        };
      }
    }]);

    function LoremData() {
      var _this;

      babelHelpers.classCallCheck(this, LoremData);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LoremData).call(this));
      _this.latest = false;
      _this.data = [];
      _this.dateLocale = "en-US";
      _this.dateFormat = {
        year: "numeric",
        month: "long",
        day: "numeric",
        hour: "numeric",
        minute: "2-digit"
      };
      _this.demo = false;
      _this.map = {};
      _this.params = {};
      _this.submit = window.location.href;
      _this.__data = [];
      _this.__loading = false;
      new Promise(function (res, rej) {
        return _require.default(["../lrndesign-avatar/lrndesign-avatar.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/iron-icon/iron-icon.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/iron-icons/iron-icons.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["./lib/lorem-data-form.js"], res, rej);
      });
      return _this;
    }

    babelHelpers.createClass(LoremData, [{
      key: "updated",
      value: function updated(changedProperties) {
        var _this3 = this;

        if (babelHelpers.get(babelHelpers.getPrototypeOf(LoremData.prototype), "updated", this)) babelHelpers.get(babelHelpers.getPrototypeOf(LoremData.prototype), "updated", this).call(this, changedProperties);
        changedProperties.forEach(function (oldValue, propName) {
          if (["params", "source"].includes(propName) && _this3.source) {
            _this3.__loading = true;

            _this3.fetchDiscussion();
          }
        });
      }
      /**
       * gets sorted data from mapped data
       *
       * @readonly
       * @memberof LoremData
       */

    }, {
      key: "sortData",
      value: function sortData(data) {
        var _this4 = this;

        return data.sort(function (a, b) {
          return _this4.latest ? a.date - b.date : b.date - a.date;
        });
      }
      /**
       * gets data that is structured according to a map object
       *
       * @readonly
       * @memberof LoremData
       */

    }, {
      key: "_handleRawData",
      value: function _handleRawData(data) {
        var _this5 = this;

        this.__loading = false;
        return this._getArray(data || []).filter(function (comment) {
          return !_this5._getMap(comment, "thread", "replyThread");
        }).map(function (thread) {
          var id = _this5._getMap(thread, "id"),
              replies = _this5._getMap(thread, "replies") || (data || []).filter(function (comment) {
            return _this5._getMap(comment, "thread", "replyThread") === id;
          });

          return {
            //gets all threads and comments if they are not mapped as nested array of thread
            id: id,
            firstName: _this5._getMap(thread, "firstName"),
            lastName: _this5._getMap(thread, "lastName"),
            avatar: _this5._getMap(thread, "avatar"),
            body: _this5._getMap(thread, "body"),
            color: _this5._getMap(thread, "color"),
            date: _this5._getMap(thread, "date"),
            replies: _this5._getArray(replies).map(function (reply) {
              //gets all comments if they are mapped as nested array of thread
              return {
                id: _this5._getMap(reply, "id", "replyId"),
                thread: _this5._getMap(reply, "thread", "replyThread") || id,
                firstName: _this5._getMap(reply, "firstName", "replyFirstName"),
                lastName: _this5._getMap(reply, "lastName", "replyLastName"),
                avatar: _this5._getMap(reply, "avatar", "replyAvatar"),
                body: _this5._getMap(reply, "body", "replyBody"),
                color: _this5._getMap(reply, "color", "replyColor"),
                date: _this5._getMap(reply, "date", "replyDate")
              };
            })
          };
        });
      }
      /**
       * gets a property value based on a conversion map
       *
       * @param {object} obj item to check
       * @param {string} prop default property name
       * @param {string} map1 second mapped property to check
       * @param {string} map2 first mapped property to check
       * @returns {*} property value
       * @memberof LoremData
       */

    }, {
      key: "_getMap",
      value: function _getMap(obj, prop, map1, map2) {
        return obj[this._mapProp(prop, map1, map2)];
      }
      /**
       * gets a mapped property based on a conversion map
       *
       * @param {object} obj item to check
       * @param {string} prop default property name
       * @param {string} map1 second mapped property to check
       * @param {string} map2 first mapped property to check
       * @returns {*} property
       * @memberof LoremData
       */

    }, {
      key: "_mapProp",
      value: function _mapProp(prop, map1, map2) {
        var map = this.map || {};
        map1 = map1 || prop;
        map2 = map2 || map1;
        return map[map2] || map[map1] || prop;
      }
      /**
       * refreshes comments
       *
       * @param {event} e
       * @memberof LoremData
       */

    }, {
      key: "_handleSubmit",
      value: function _handleSubmit(e) {
        this.fetchDiscussion();
      }
      /**
       * demonstrates how data will appear when a comment is submitted
       *
       * @param {event} e
       * @memberof LoremData
       */

    }, {
      key: "_handleDemo",
      value: function _handleDemo(e) {
        if (e.detail && e.detail.textarea) {
          var newComment = {},
              data = this.__data;
          newComment.id = "comment-".concat(Date.now());
          newComment.firstName = "DEMO";
          newComment.lastName = "USER";
          newComment.date = this._getDate(new Date());
          newComment.body = e.detail.textarea.value;

          if (e.detail.thread) {
            var filter = data.filter(function (thread) {
              return thread.id === e.detail.thread;
            }),
                thread = filter ? filter[0] : undefined;
            newComment.thread = e.detail.thread;

            if (thread) {
              thread.replies = [].concat(babelHelpers.toConsumableArray(thread.replies), [newComment]);
            } else {
              data.push(newComment);
            }
          } else {
            newComment.replies = [];
            data.push(newComment);
          }

          this.__data = [];
          this.__data = data;
          e.detail.textarea.value = "";
        }
      }
      /**
       * fetches data from discussion
       *
       * @memberof LoremData
       */

    }, {
      key: "fetchDiscussion",
      value: function fetchDiscussion() {
        var _this6 = this;

        fetch(this._getPath(this.source, this.params)).then(function (response) {
          return response.json();
        }).then(function (data) {
          return _this6.__data = _this6._handleRawData(data);
        });
      }
      /**
       * converts object to array
       *
       * @param {object} obj
       * @returns array
       * @memberof LoremData
       */

    }, {
      key: "_getArray",
      value: function _getArray(obj) {
        return typeof obj === "array" ? obj : Object.keys(obj || {}).map(function (key) {
          var item = obj[key];
          item.id = key;
          return item;
        });
      }
      /**
       * gets formated date from date string
       *
       * @param {string} d
       * @returns string
       * @memberof LoremData
       */

    }, {
      key: "_getDate",
      value: function _getDate(d) {
        var format = typeof this.dateFormat === "string" ? JSON.parse(this.dateFormat) : this.dateFormat;
        return d && new Date(d) ? new Date(d).toLocaleString(this.dateLocale, format) : "";
      }
      /**
       * adds thread id to parameters
       *
       * @param {*} id
       * @returns
       * @memberof LoremData
       */

    }, {
      key: "_getThreadParams",
      value: function _getThreadParams(id) {
        var params = this.params;
        params[this._mapProp("id")] = id;
        return params;
      }
      /**
       * gets full path with encoded params
       *
       * @param {string} path
       * @param {object} params
       * @returns string
       * @memberof LoremData
       */

    }, {
      key: "_getPath",
      value: function _getPath(path, params) {
        var query = Object.keys(params || {}).map(function (p) {
          return "".concat(encodeURI(p), "=").concat(encodeURI(params[p]));
        }).join("&");
        return query ? "".concat(path, "?").concat(query) : path;
      }
    }]);
    return LoremData;
  }(_litElement.LitElement);

  _exports.LoremData = LoremData;
  window.customElements.define(LoremData.tag, LoremData);
});
