// @lrnwebcomponents/cms-hax/lib/cms-token.js
define(["exports","../../../@polymer/polymer/polymer-element.js","../../../@polymer/polymer/lib/utils/render-status.js","../../../@polymer/polymer/lib/utils/flattened-nodes-observer.js","../../../@polymer/polymer/lib/utils/async.js","../../../@polymer/iron-ajax/iron-ajax.js","../../../@polymer/paper-spinner/paper-spinner.js","../../utils/utils.js"],(function(e,t,n,o,i,s,a,d){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.CMSToken=void 0;class CMSToken extends t.PolymerElement{static get template(){return t.html`
      <style>
        :host {
          display: inline;
          min-width: 112px;
          min-height: 112px;
          transition: 0.6s all ease;
          background-color: transparent;
        }

        paper-spinner {
          transition: 0.6s all ease;
          position: absolute;
          visibility: hidden;
          display: none;
          opacity: 0;
          height: 0;
          width: 0;
        }

        #replacementcontent {
          transition: 0.6s all ease;
          visibility: visible;
          opacity: 1;
          height: auto;
          width: auto;
        }

        :host([loading]) {
          text-align: center;
        }

        :host([loading]) paper-spinner {
          visibility: visible;
          opacity: 1;
          position: relative;
          height: 80px;
          width: 80px;
          padding: 16px;
          display: flex;
        }

        :host([loading]) #replacementcontent {
          opacity: 0;
          visibility: hidden;
          height: 0;
          width: 0;
        }
      </style>
      <iron-ajax
        id="tokenrequest"
        method="GET"
        params="[[bodyData]]"
        url="[[tokenEndPoint]]"
        handle-as="json"
        last-response="{{tokenData}}"
      ></iron-ajax>
      <paper-spinner active="[[loading]]"></paper-spinner>
      <span id="replacementcontent">
        <slot></slot>
      </span>
    `}static get tag(){return"cms-token"}static get properties(){return{loading:{type:Boolean,reflectToAttribute:!0,value:!1},token:{type:String,reflectToAttribute:!0},tokenEndPoint:{type:String},bodyData:{type:Object,computed:"_generateBodyData(token, _clickInvoked)",observer:"_tokenChanged"},_clickInvoked:{type:String,value:!1},tokenData:{type:String,observer:"_handleTokenResponse"},tokenPrefix:{type:String,value:"["},tokenSuffix:{type:String,value:"]"},_displayMode:{type:String,value:"full",observer:"_displayModeChanged"}}}_displayModeChanged(e,t){void 0!==e&&""!=e&&void 0!==this.token&&(this.token=this.token.replace(t,e))}_generateBodyData(e,t){if(null!==e&&""!==e){return{token:`${this.tokenPrefix}${e}${this.tokenSuffix}`,cachedResponse:t}}}_handleTokenResponse(e,t){null!==e&&void 0!==e.content&&(null!=document.getElementById("cmstokenidtolockonto")&&(document.getElementById("cmstokenidtolockonto").setAttribute("href",e.editEndpoint),document.getElementById("cmstokenidtolockonto").innerHTML=e.editText,document.getElementById("cmstokenidtolockonto").addEventListener("click",this.__tokenClicked.bind(this))),(0,d.wipeSlot)(this),i.microTask.run(()=>{let t=document.createElement("template");t.innerHTML=e.content,this.appendChild(document.importNode(t.content,!0)),setTimeout(()=>{window.WCAutoload&&window.WCAutoload.process()},0),this.loading=!1}))}_tokenChanged(e,t){void 0===e||""===e||this.loading||(void 0===this.tokenEndPoint&&void 0!==window.cmstokenEndPoint&&(this.tokenEndPoint=window.cmstokenEndPoint),this.tokenEndPoint&&(this.loading=!0,i.microTask.run(()=>{this.shadowRoot.querySelector("#tokenrequest").generateRequest()})))}_windowVisibilityChanged(e){!this.loading&&this._clickInvoked&&(this.shadowRoot.querySelector("#tokenrequest").generateRequest(),this._clickInvoked=!1)}__tokenClicked(e){this._clickInvoked=!0}connectedCallback(){if(super.connectedCallback(),void 0!==this.token&&null!==this.token&&""!==this.token){0!==o.FlattenedNodesObserver.getFlattenedNodes(this).length||this.loading||(void 0===this.tokenEndPoint&&void 0!==window.cmstokenEndPoint&&(this.tokenEndPoint=window.cmstokenEndPoint),this.tokenEndPoint&&(this.loading=!0,i.microTask.run(()=>{this.shadowRoot.querySelector("#tokenrequest").generateRequest()})))}(0,n.afterNextRender)(this,(function(){document.addEventListener("visibilitychange",this._windowVisibilityChanged.bind(this))}))}disconnectedCallback(){document.removeEventListener("visibilitychange",this._windowVisibilityChanged.bind(this)),super.disconnectedCallback()}static get haxProperties(){return{canScale:!0,canPosition:!0,canEditSource:!1,gizmo:{title:"CMS Token",description:"CMS token rendered on the backend",icon:"icons:code",color:"light-blue",groups:["CMS"],handles:[{type:"cmstoken",token:"token"}],meta:{author:"ELMS:LN"}},settings:{quick:[],configure:[{property:"token",title:"Token",description:"Token from our CMS",inputMethod:"textfield",icon:"editor:title"}],advanced:[]},saveOptions:{wipeSlot:!0,unsetAttributes:["loading","token-data","body-data","token-end-point"]}}}postProcessgetHaxJSONSchema(e){let t="",n="Edit";if(void 0!==this.tokenData)for(var o in t=this.tokenData.editEndpoint,n=this.tokenData.editText,this.tokenData.schema)e.properties[o]=this.tokenData.schema[o];return e.properties.__editThis={type:"string",component:{name:"a",properties:{id:"cmstokenidtolockonto",href:t,target:"_blank"},slot:n}},e}}e.CMSToken=CMSToken,window.customElements.define(CMSToken.tag,CMSToken)}));
