// @lrnwebcomponents/pdf-element/pdf-element.js
define(["exports","require","../../@polymer/polymer/polymer-element.js","../../@polymer/paper-input/paper-input.js","../../@polymer/iron-input/iron-input.js","../../@polymer/paper-icon-button/paper-icon-button.js","../schema-behaviors/schema-behaviors.js","../../pdfjs-dist/build/pdf.js","../../pdfjs-dist/build/pdf.worker.js","./lib/main.js"],(function(e,t,i,o,n,s,a,r,l,c){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.PdfElement=void 0,t=babelHelpers.interopRequireWildcard(t);class PdfElement extends((0,a.SchemaBehaviors)(i.PolymerElement)){constructor(){super(),new Promise((e,i)=>t.default(["../../@polymer/paper-card/paper-card.js"],e,i)),new Promise((e,i)=>t.default(["../../@polymer/app-layout/app-toolbar/app-toolbar.js"],e,i)),new Promise((e,i)=>t.default(["../../@polymer/paper-spinner/paper-spinner.js"],e,i))}static get template(){return i.html`
      <style>
        :host {
          display: block;
          width: 100%;
          height: 100%;
        }

        app-toolbar.pdf-toolbar {
          --app-toolbar-background: #323639;
        }

        .pdf-viewer {
          text-align: center;
          border: 1px solid #4d4d4d;
        }

        .pdf-viewport-out {
          overflow: auto;
          background-color: #525659;
          position: relative;
          width: 100%;
          height: 100%;
        }

        .pdf-viewport {
          display: block;
          position: relative;
          border: 1px solid #eeeeee;
          transition: all 200ms ease-in-out;
          width: 100%;
          height: 100%;
        }

        .sidebar {
          background-color: gray;
          float: left;
          height: 0px;
          overflow: scroll;
          margin-left: -25%;
          visibility: hidden;
        }

        .main {
          margin-left: 0%;
        }

        .pageselector {
          width: 3ch;
          background-color: black;
          font-size: 17px;
          background-color: transparent;
          border: 0px solid;
        }

        .pageselector:focus {
          outline: none;
        }

        #input {
          -webkit-margin-start: -3px;
          color: #fff;
          line-height: 18px;
          padding: 3px;
          text-align: end;
        }

        #input:focus,
        #input:hover {
          background-color: rgba(0, 0, 0, 0.5);
          border-radius: 2px;
        }

        #slash {
          padding: 0 3px;
        }

        paper-spinner {
          position: absolute;
          left: 50%;
        }

        .textLayer {
          transition: all 200ms ease-in-out;
        }

        .positionRelative {
          position: relative;
        }
      </style>

      <paper-material elevation="{{elevation}}">
        <div class="card-content" style="width: {{width}}px">
          <paper-card class="paperCard" style="width: {{width}}px">
            <div class="pdf-viewer">
              <app-toolbar class="pdf-toolbar">
                <paper-icon-button
                  icon="menu"
                  on-click="sideBar"
                ></paper-icon-button>
                <paper-icon-button
                  icon="arrow-back"
                  on-click="showPrev"
                ></paper-icon-button>
                <input
                  class="pageselector"
                  id="input"
                  is="iron-input"
                  value="{{currentPage}}"
                  prevent-invalid-input=""
                  allowed-pattern="\\d"
                  on-change="pageNumSearch"
                />
                <span id="slash">/</span><span id="totalPages"></span>
                <paper-icon-button
                  icon="arrow-forward"
                  on-click="showNext"
                ></paper-icon-button>
                <span class="title" hidden="{{!showFileName}}">Testing</span>
                <span class="title" hidden="{{showFileName}}"></span>
                <span class="pageRendering"></span>
                <paper-icon-button
                  icon="zoom-in"
                  on-click="zoomIn"
                ></paper-icon-button>
                <paper-icon-button
                  icon="zoom-out"
                  on-click="zoomOut"
                ></paper-icon-button>
                <paper-icon-button
                  id="zoomIcon"
                  icon="fullscreen"
                  on-click="zoomFit"
                ></paper-icon-button>
                <paper-icon-button
                  icon="file-download"
                  hidden\$="{{!downloadable}}"
                  on-click="download"
                ></paper-icon-button>
              </app-toolbar>
              <div id="container" class="sidebar" style="width:25%"></div>
              <div id="main">
                <div id="test" class="pdf-viewport-out">
                  <canvas class="pdf-viewport"></canvas>
                  <div
                    id="text-layer"
                    class="textLayer"
                    hidden\$="{{!enableTextSelection}}"
                  ></div>
                </div>
                <paper-spinner
                  class="spinner"
                  hidden\$="{{!showSpinner}}"
                ></paper-spinner>
              </div>
            </div>
          </paper-card>
        </div>
      </paper-material>
    `}static get tag(){return"pdf-element"}static get properties(){return{...super.properties,src:{type:String,reflectToAttribute:!0},elevation:{type:Number,value:1},downloadable:{type:Boolean,value:!1},showFileName:{type:Boolean,value:!1},showSpinner:{type:Boolean,value:!1},enableTextSelection:{type:Boolean,value:!1},fitWidth:{type:Boolean,value:!1},width:{type:Number,value:500}}}connectedCallback(){super.connectedCallback(),this.src=this.getAttribute("src"),this._initializeReader(),this.src&&this.instance.loadPDF(),this._setFitWidth()}static get haxProperties(){return{canScale:!0,canPosition:!0,canEditSource:!1,gizmo:{title:"PDF viewer",descrption:"This can nicely present a PDF in a standard inplace, cross browser way.",icon:"image:picture-as-pdf",color:"red",groups:["Presentation","PDF","Data"],handles:[{type:"pdf",url:"src",source:"src"},{type:"document",url:"src",source:"src"}],meta:{author:"ELMS:LN"}},settings:{quick:[{property:"src",title:"File",description:"The URL for the pdf",inputMethod:"textfield",icon:"link",required:!0}],configure:[{property:"src",title:"Source",description:"The URL for this csv file",inputMethod:"textfield",icon:"link",required:!0},{property:"downloadable",title:"Downloadable",description:"User can download this",inputMethod:"boolean",icon:"file-download"},{property:"enableTextSelection",title:"Text Selection",description:"User can select text in this element.",inputMethod:"boolean",icon:"file-download"},{property:"elevation",title:"Elevation",description:"Visual elevation of the element",inputMethod:"number",icon:"flip-to-front"}],advanced:[]}}}loadPDF(){this.getAttribute("src")&&(this.instance.changePDFSource(this.getAttribute("src")),this.currentPage=1,this.totalPages=this.instance.totalPages,this.fileName=this.src.split("/").pop(),this._setFitWidth(),this.shadowRoot.querySelector("#zoomIcon").icon="fullscreen")}attributeChanged(e,t){"src"===e?void 0===this.instance?this._initializeReader():(this.loadPDF(),this.changedSideBar=!0,this.fromChange=!0,this.sideBar()):"fitWidth"===e&&this._setFitWidth()}_initializeReader(){this.instance=new Reader(this),null!=this.src&&(this.fileName=this.src.split("/").pop()),this.currentPage=1}_setFitWidth(){this.instance.setFitWidth(this.fitWidth)}zoomInOut(e){this.instance.currentZoomVal>=2?this.instance.currentZoomVal=2:this.instance.currentZoomVal<=.1?this.instance.currentZoomVal=.1:(this.shadowRoot.querySelector("#zoomIcon").icon="fullscreen",this.instance.zoomInOut(e))}zoomIn(){this.instance.pdfExists&&this.zoomInOut(.1)}zoomOut(){this.instance.pdfExists&&this.instance.zoomInOut(-.1)}zoomFit(){this.instance.pdfExists&&(this.instance.currentZoomVal==this.instance.widthZoomVal?(this.instance.zoomPageFit(),this.shadowRoot.querySelector("#zoomIcon").icon="fullscreen"):(this.instance.zoomWidthFit(),this.shadowRoot.querySelector("#zoomIcon").icon="fullscreen-exit"))}pageNumSearch(){var e=parseInt(this.shadowRoot.querySelector("#input").value);1<=e&&e<=this.instance.totalPagesNum?(this.instance.currentPage=e,this.instance.queueRenderPage(this.instance.currentPage),this.currentPage=e,this.shadowRoot.querySelector("#input").blur()):(this.shadowRoot.querySelector("#input").value=this.currentPage,this.shadowRoot.querySelector("#input").blur())}sideBarClick(e,t,i){i.src.split("/").pop();var o=t;i.sidebarOpen=!0,1<=e&&e<=t.totalPagesNum?(o.currentPage=e,o.queueRenderPage(o.currentPage),t.currentPage=e,i.currentPage=e,this.shadowRoot.querySelector("#input").blur()):(this.shadowRoot.querySelector("#input").value=o.currentPage,this.shadowRoot.querySelector("#input").blur())}showPrev(){1<this.instance.currentPage&&(this.instance.currentPage--,this.instance.queueRenderPage(this.instance.currentPage),this.currentPage--)}showNext(){this.instance.totalPagesNum>this.instance.currentPage&&(this.instance.currentPage++,this.instance.queueRenderPage(this.instance.currentPage),this.currentPage++)}sideBar(){this.instance.pdfExists&&(this.fromChange||(this.shadowRoot.querySelector("#container").style.height=this.shadowRoot.querySelector("#test").style.height,this.shadowRoot.querySelector("#container").style.width=this.shadowRoot.querySelector("#test").style.width,"25%"==this.shadowRoot.querySelector("#main").style.marginLeft?(this.sidebarOpen=!1,this.instance.setViewportPos(!1),this.shadowRoot.querySelector("#main").style.marginLeft="0%",this.shadowRoot.querySelector("#container").style.marginLeft="-25%",this.shadowRoot.querySelector("#container").style.visibility="hidden"):(this.sidebarOpen=!0,this.shadowRoot.querySelector("#main").style.marginLeft="25%",this.shadowRoot.querySelector("#container").style.marginLeft="0%",this.shadowRoot.querySelector("#container").style.visibility="visible",this.instance.setViewportPos(!0))),this.fromChange=!1,this.instance.sidebarSetup(this),this.changedSideBar=!1)}download(){this.instance.pdfExists&&this.instance.download()}}e.PdfElement=PdfElement,window.customElements.define(PdfElement.tag,PdfElement)}));
