// @lrnwebcomponents/elmsln-studio/elmsln-studio.js
define(["exports", "../../lit-element/lit-element.js", "../../@polymer/iron-flex-layout/iron-flex-layout.js", "./lib/elmsln-studio-projectboard.js"], function (_exports, _litElement, _ironFlexLayout, _elmslnStudioProjectboard) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmslnStudio = void 0;

  /**
   * Copyright 2020 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `elmsln-studio`
   * Studio App for ELMS:LN
   *
   * @microcopy - language worth noting:
   *  -
   *
   * @customElement elmsln-studio
   * @lit-html
   * @lit-element
   * @demo demo/index.html
   */
  class ElmslnStudio extends _litElement.LitElement {
    // render function
    render() {
      return (0, _litElement.html)`
      <style>
        :host {
          display: block;
        }

        :host([hidden]) {
          display: none;
        }
      </style>
      <lrnapp-studio-projectboard></lrnapp-studio-projectboard>
    `;
    } // properties available to the custom element for data binding


    static get properties() {
      return {};
    }
    /**
     * Store the tag name to make it easier to obtain directly.
     * @notice function name must be here for tooling to operate correctly
     */


    tag() {
      return "elmsln-studio";
    } // life cycle


    constructor() {
      super();
      this.tag = ElmslnStudio.tag; // map our imported properties json to real props on the element
      // @notice static getter of properties is built via tooling
      // to edit modify src/elmsln-studio-properties.json

      let obj = ElmslnStudio.properties;

      for (let p in obj) {
        if (obj.hasOwnProperty(p)) {
          if (this.hasAttribute(p)) {
            this[p] = this.getAttribute(p);
          } else {
            this.setAttribute(p, obj[p].value);
            this[p] = obj[p].value;
          }
        }
      }
    }
    /**
     * life cycle, element is afixed to the DOM
     */


    connectedCallback() {
      super.connectedCallback();
    } // static get observedAttributes() {
    //   return [];
    // }
    // disconnectedCallback() {}
    // attributeChangedCallback(attr, oldValue, newValue) {}


  }

  _exports.ElmslnStudio = ElmslnStudio;
  customElements.define("elmsln-studio", ElmslnStudio);
});
