// @lrnwebcomponents/super-daemon/lib/super-daemon-toast.js
import{css as e,html as t,unsafeCSS as i}from"../../../lit/index.js";import{SimpleToastEl as a}from"../../simple-toast/lib/simple-toast-el.js";import"../../future-terminal-text/future-terminal-text.js";import"../../simple-icon/lib/simple-icons.js";import"../../simple-icon/simple-icon.js";import"../../hax-iconset/lib/simple-hax-iconset.js";const s=new URL("./images/SpeechBubbleL.svg",import.meta.url).href,l=new URL("./images/SpeechBubbleMiddle.svg",import.meta.url).href,n=new URL("./images/SpeechBubbleR.svg",import.meta.url).href;export class SuperDaemonToast extends a{static get tag(){return"super-daemon-toast"}constructor(){super(),this.awaitingMerlinInput=!1,this.windowControllers=new AbortController,this.text="Saved",this.merlin=!1,this.classStyle="",this.future=!1,this.duration=3e3,this.accentColor="grey",this.dark=!1,this.eventCallback=null,this.fire=!1,this.hat="coffee",this.walking=!1}static get styles(){return[...super.styles,e`
        :host {
          --simple-toast-z-index: 100000000;
        }
        :host([opened]) {
          display: flex;
        }

        future-terminal-text {
          min-width: 300px;
          overflow-wrap: break-all;
          text-elipsis: ellipsis;
          line-height: 36px;
          font-size: 18px;
          text-align: left;
          padding: 36px 0px;
          max-width: 50vw;
        }

        .merlin {
          --simple-icon-height: 100px;
          --simple-icon-width: 100px;
          background-color: var(--simple-colors-default-theme-accent-1, white);
          display: block;
          height: 150px;
          width: 100px;
          margin: 6px 0 0 0;
          padding: 16px;
        }
        .awaiting-input {
          --simple-icon-height: 75px;
          --simple-icon-width: 75px;
          width: 100px;
          margin: 6px 0px 0px;
          padding: 16px;
          background-color: var(--simple-colors-default-theme-accent-2, white);
          color: var(--simple-colors-default-theme-purple-6, purple);
          vertical-align: middle;
          display: inline-flex;
          height: 100px;
        }
        :host([hidden]) {
          display: none;
        }
        :host {
          --simple-toast-bottom: 0px;
          --simple-toast-bottom-offscreen: -284px;
          height: 142px;
          display: none;
          width: var(--simple-toast-width, auto);
          color: var(
            --simple-toast-color,
            var(--simple-colors-default-theme-accent-12, black)
          );
          background-color: transparent;
          top: var(--simple-toast-top);
          margin: var(--simple-toast-margin, 4px);
          padding: var(--simple-toast-padding, 4px);
          bottom: var(--simple-toast-bottom, 36px);
          left: var(--simple-toast-left, 0px);
          border: var(--simple-toast-border);
          z-index: var(--simple-toast-z-index, 100000000);
          font-size: var(--simple-toast-font-size, 18px);
          font-family: "Press Start 2P", sans-serif;
          font-weight: bold;
          text-align: center;
          vertical-align: middle;
        }
        .bubble {
          height: 142px;
          display: inline-flex;
          margin-top: 6px;
        }
        .mid {
          line-height: 142px;
          background-color: white;
          background-repeat: repeat-x;
          background-image: url("${i(l)}");
        }
        .leftedge {
          background-image: url("${i(s)}");
          width: 24px;
          background-color: white;
          transform: scaleX(-1);
        }
        .rightedge {
          background-image: url("${i(n)}");
          width: 54px;
          background-color: white;
          transform: scaleX(-1.1);
        }
        :host([dark-mode]) .mid,
        :host([dark-mode]) .leftedge,
        :host([dark-mode]) .rightedge {
          filter: invert(1);
        }
        @media (max-width: 800px) {
          :host {
            --simple-toast-width: 80vw;
            --simple-toast-font-size: 12px;
          }
        }
        @media (max-width: 500px) {
          :host {
            height: 50px;
            line-height: 50px;
            border: none;
          }

          .rightedge {
            display: none;
          }
          .leftedge {
            display: none;
          }
          .mid {
            height: 50px;
            line-height: 50px;
            background-image: unset;
          }
          .bubble {
            height: 50px;
            margin: 0;
            border: 2px solid black;
            border-radius: 5px;
          }
        }
      `]}hideSimpleToast(e){this.alwaysvisible||(this.style.animation="forcedfadeout 0.6s .3s",this.awaitingMerlinInput=!1,setTimeout((()=>{this.hide()}),0))}static get properties(){return{...super.properties,darkMode:{type:Boolean,reflect:!0,attribute:"dark-mode"},fire:{type:Boolean},hat:{type:String},walking:{type:Boolean},opened:{type:Boolean,reflect:!0},awaitingMerlinInput:{type:Boolean,attribute:"awaiting-merlin-input"},merlin:{type:Boolean},future:{type:Boolean},text:{type:String},classStyle:{type:String,attribute:"class-style"},duration:{type:Number},eventCallback:{type:String,attribute:"event-callback"}}}render(){return t` <div class="bubble">
      ${this.awaitingMerlinInput?t`<simple-icon-lite
            class="awaiting-input"
            icon="hax:loading"
          ></simple-icon-lite>`:t`
            <simple-icon
              class="merlin"
              icon="hax:wizard-hat"
              accent-color="purple"
            ></simple-icon>
          `}
      <span class="bubble rightedge"></span>
      <span class="bubble mid">
        <slot name="pre"></slot>
        ${this.future?t` <future-terminal-text
              fadein
              glitch
              glitch-max="3"
              glitch-duration="40"
            ></future-terminal-text>`:t`${this.text}`}
        <slot></slot>
      </span>
      <span class="bubble leftedge"></span>
    </div>`}updated(e){super.updated(e),e.has("text")&&this.future&&this.shadowRoot.querySelector("future-terminal-text")&&(this.shadowRoot.querySelector("future-terminal-text").innerText=this.text,this.shadowRoot.querySelector("future-terminal-text")._doGlitch())}connectedCallback(){super.connectedCallback(),window.addEventListener("super-daemon-toast-hide",this.hideSimpleToast.bind(this),{signal:this.windowControllers.signal}),window.addEventListener("super-daemon-toast-show",this.showSimpleToast.bind(this),{signal:this.windowControllers.signal})}disconnectedCallback(){this.windowControllers.abort(),super.disconnectedCallback()}showSimpleToast(e){for(;null!==this.firstChild;)this.removeChild(this.firstChild);setTimeout((()=>{e.detail.slot&&this.appendChild(e.detail.slot)}),0),this.duration=e.detail.duration?e.detail.duration:4e3,this.fire=!!e.detail.fire&&e.detail.fire,this.hat=e.detail.hat?e.detail.hat:"coffee",this.merlin=!!e.detail.merlin&&e.detail.merlin,this.walking=!!e.detail.walking&&e.detail.walking,this.text=e.detail.text?e.detail.text:"Saved",this.future=!!e.detail.future&&e.detail.future,this.classStyle=e.detail.classStyle?e.detail.classStyle:"",this.eventCallback=e.detail.eventCallback?e.detail.eventCallback:null,this.accentColor=e.detail.accentColor?e.detail.accentColor:"grey",this.alwaysvisible=!!e.detail.alwaysvisible&&e.detail.alwaysvisible,e.detail.awaitingMerlinInput&&this.duration?setTimeout((()=>{this.style.animation="none",this.awaitingMerlinInput=e.detail.awaitingMerlinInput}),this.duration/2):this.awaitingMerlinInput=!1,this.show()}show(){this.opened||(this.style.animation="fadein 0.3s, fadeout 0.6s "+this.duration/1e3+"s",this.opened=!0)}hide(){if(this.awaitingMerlinInput)this.style.animation="fadein 0.3s";else{if(this.eventCallback){const e=new CustomEvent(this.eventCallback,{bubbles:!0,cancelable:!0,detail:!0,composed:!0});this.dispatchEvent(e)}this.alwaysvisible?this.style.animation="fadein 0.3s":this.opened=!1}}}customElements.define(SuperDaemonToast.tag,SuperDaemonToast),window.SuperDaemonToast=window.SuperDaemonToast||{},window.SuperDaemonToast.requestAvailability=()=>(window.SuperDaemonToast.instance||(window.SuperDaemonToast.instance=document.createElement(SuperDaemonToast.tag),document.body.appendChild(window.SuperDaemonToast.instance)),window.SuperDaemonToast.instance);export const SuperDaemonToastInstance=window.SuperDaemonToast.requestAvailability();
