// @lrnwebcomponents/elmsln-studio/lib/elmsln-studio-button.js
define(["exports", "../../../lit-element/lit-element.js", "../../../lit-element-router/lit-element-router.js", "../../simple-icon/simple-icon.js", "../../simple-icon/lib/simple-icons.js", "../../hax-iconset/lib/simple-hax-iconset.js"], function (_exports, _litElement, _litElementRouter, _simpleIcon, _simpleIcons, _simpleHaxIconset) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmslnStudioButton = void 0;

  /**
   * Copyright 2020 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `elmsln-studio-button`
   * Navigation Link for ELMS:LN Studio
   *
   * @customElement elmsln-studio-button
   * @lit-html
   * @lit-element
   */
  class ElmslnStudioButton extends (0, _litElementRouter.navigator)(_litElement.LitElement) {
    /**
     * Store the tag name to make it easier to obtain directly.
     * @notice function name must be here for tooling to operate correctly
     */
    static get tag() {
      return "elmsln-studio-button";
    }

    static get properties() {
      return {
        path: {
          type: String
        },
        icon: {
          type: String
        }
      };
    }

    static get styles() {
      return (0, _litElement.css)`
      :host {
        display: block;
      }
      :host([hidden]) {
        display: none;
      }
      button {
        border: none;
        background-color: transparent;
        display: flex;
        align-items: center;
        height: 26px;
      }
      button:focus,
      button:hover {
        outline: none;
      }
      iron-icon {
        color: var(--elmsln-studio-button-icon-color);
      }
    `;
    }

    constructor() {
      super();
      this.path = "";
      this.icon = "";
    }

    render() {
      return (0, _litElement.html)`
      <button @click="${this.buttonClick}">
        <slot name="before"></slot>
        <iron-icon
          aria-hidden="true"
          ?hidden="${!this.icon}"
          icon="${this.icon}"
        ></simple-icon>
        <slot></slot>
      </button>
    `;
    }

    buttonClick(event) {
      this.navigate(`${window.ElmslnStudioPath || ""}${this.path}`);
    }

  }

  _exports.ElmslnStudioButton = ElmslnStudioButton;
  customElements.define("elmsln-studio-button", ElmslnStudioButton);
});
