// @lrnwebcomponents/elmsln-studio/lib/elmsln-studio-main.js
define(["exports", "../../../lit-element/lit-element.js", "../../../lit-element-router/lit-element-router.js", "./elmsln-studio-utilities.js"], function (_exports, _litElement, _litElementRouter, _elmslnStudioUtilities) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmslnStudioMain = void 0;

  /**
   * Copyright 2020 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `elmsln-studio-main`
   * Main Content for ELMS:LN Studio
   *
   * @customElement elmsln-studio-main
   * @lit-html
   * @lit-element
   */
  class ElmslnStudioMain extends (0, _elmslnStudioUtilities.ElmslnStudioUtilities)((0, _litElementRouter.outlet)(_litElement.LitElement)) {
    /**
     * Store the tag name to make it easier to obtain directly.
     * @notice function name must be here for tooling to operate correctly
     */
    static get tag() {
      return "elmsln-studio-main";
    } //styles function


    static get styles() {
      return [(0, _litElement.css)`
        :host {
          display: block;
          font-family: var(--elmsln-studio-FontFamily, sans-serif);
        }

        :host([hidden]) {
          display: none;
        }
      `];
    }

    render() {
      return (0, _litElement.html)` <slot></slot> `;
    }

    static get properties() {
      return { ...super.properties
      };
    }

  }

  _exports.ElmslnStudioMain = ElmslnStudioMain;
  customElements.define("elmsln-studio-main", ElmslnStudioMain);
});
