// @lrnwebcomponents/lrndesign-imagemap/lib/lrndesign-imagemap-hotspot.js
define(["exports", "require", "../../../lit-element/lit-element.js"], function (_exports, _require, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrndesignImagemapHotspot = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject2_a5fd8b50e8ab11eabe6cb3de058a4220() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: none;\n        }\n        @media print {\n          :host {\n            display: block;\n          }\n        }\n      "]);

    _templateObject2_a5fd8b50e8ab11eabe6cb3de058a4220 = function _templateObject2_a5fd8b50e8ab11eabe6cb3de058a4220() {
      return data;
    };

    return data;
  }

  function _templateObject_a5fd8b50e8ab11eabe6cb3de058a4220() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <figure class=\"hotspot-print\">\n        <figcaption>\n          <relative-heading disable-link id=\"sub-heading\" parent=\"heading\">\n            <h2>", "</h2>\n          </relative-heading>\n          <div id=\"desc\"><slot></slot></div>\n        </figcaption>\n        <slot id=\"svg\" name=\"svg\"></slot>\n      </figure>\n    "]);

    _templateObject_a5fd8b50e8ab11eabe6cb3de058a4220 = function _templateObject_a5fd8b50e8ab11eabe6cb3de058a4220() {
      return data;
    };

    return data;
  }

  /**
   * `lrndesign-imagemap-hotspot`
   * creates an accessible image map
   * @demo demo/index.html
   * @element lrndesign-imagemap-hotspot
   */
  var LrndesignImagemapHotspot =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrndesignImagemapHotspot, _LitElement);
    babelHelpers.createClass(LrndesignImagemapHotspot, [{
      key: "render",

      /**
       * LitElement render
       */
      value: function render() {
        return (0, _litElement.html)(_templateObject_a5fd8b50e8ab11eabe6cb3de058a4220(), this.label);
      }
      /**
       * HTMLElement
       */

    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_a5fd8b50e8ab11eabe6cb3de058a4220())];
      }
    }]);

    function LrndesignImagemapHotspot() {
      var _this;

      babelHelpers.classCallCheck(this, LrndesignImagemapHotspot);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrndesignImagemapHotspot).call(this));
      new Promise(function (res, rej) {
        return _require.default(["../../relative-heading/relative-heading.js"], res, rej);
      });
      return _this;
    }

    babelHelpers.createClass(LrndesignImagemapHotspot, [{
      key: "loadSvg",
      value: function loadSvg(svg, hotspots) {
        var _this2 = this;

        var div = document.createElement("div");
        div.innerHTML = svg;
        var slot = div.children[0];
        slot.slot = svg;
        slot.setAttribute("aria-labelledBy", "sub-heading");
        slot.setAttribute("aria-describedBy", "sub-heading desc");
        (hotspots || []).forEach(function (hotspot) {
          var svgHotspot = slot.querySelector("#".concat(hotspot));
          svgHotspot.classList.add("hotspot");

          if (hotspot === _this2.hotspotId) {
            svgHotspot.classList.add("selected");
          }
        });
        this.appendChild(slot);
        div.remove();
      }
    }, {
      key: "setParentHeading",
      value: function setParentHeading(parent) {
        this.shadowRoot.querySelector("#heading").parent = parent;
      }
    }], [{
      key: "tag",
      get: function get() {
        return "lrndesign-imagemap-hotspot";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Label for the hotspot
           */
          label: {
            type: String
          },

          /**
           * Id of hotspot element inside the SVG
           */
          hotspotId: {
            type: String,
            attribute: "hotspot-id",
            reflect: true
          },
          position: {
            type: String
          },
          __hotspots: {
            type: Array
          }
        };
      }
    }]);
    return LrndesignImagemapHotspot;
  }(_litElement.LitElement);

  _exports.LrndesignImagemapHotspot = LrndesignImagemapHotspot;
  window.customElements.define(LrndesignImagemapHotspot.tag, LrndesignImagemapHotspot);
});
