// @lrnwebcomponents/wysiwyg-hax/wysiwyg-hax.js
define(["exports", "../../lit-element/lit-element.js", "../cms-hax/cms-hax.js"], function (_exports, _litElement, _cmsHax) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.WysiwygHax = void 0;

  function _templateObject2_b57ab7203ee611ea9b2edd59a4f93ed0() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n      "]);

    _templateObject2_b57ab7203ee611ea9b2edd59a4f93ed0 = function _templateObject2_b57ab7203ee611ea9b2edd59a4f93ed0() {
      return data;
    };

    return data;
  }

  function _templateObject_b57ab7203ee611ea9b2edd59a4f93ed0() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <textarea\n        class=\"", "\"\n        id=\"", "\"\n        name=\"", "\"\n        hidden=\"\"\n      >\n      ", "\n      </textarea\n      >\n      <cms-hax\n        hide-message=\"\"\n        redirect-location=\"", "\"\n        update-page-data=\"", "\"\n        end-point=\"", "\"\n        app-store-connection=\"", "\"\n        ?open-default=\"", "\"\n        ?sync-body=\"", "\"\n        ?hide-export-button=\"", "\"\n        ?hide-panel-ops=\"", "\"\n        ?hide-preferences-button=\"", "\"\n        align=\"", "\"\n      >\n      </cms-hax>\n    "]);

    _templateObject_b57ab7203ee611ea9b2edd59a4f93ed0 = function _templateObject_b57ab7203ee611ea9b2edd59a4f93ed0() {
      return data;
    };

    return data;
  }

  /**
   * `wysiwyg-hax`
   * `Integration of wysiwyg edit form for a page with HAX.`
   * @demo demo/index.html
   * @customElement wysiwyg-hax
   */
  var WysiwygHax =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(WysiwygHax, _LitElement);
    babelHelpers.createClass(WysiwygHax, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_b57ab7203ee611ea9b2edd59a4f93ed0(), this.fieldClass, this.fieldId, this.fieldName, this.bodyValue, this.redirectLocation, this.updatePageData, this.endPoint, this.appStoreConnection, this.openDefault, this.syncBody, this.hideExportButton, this.hidePanelOps, this.hidePreferencesButton, this.align);
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_b57ab7203ee611ea9b2edd59a4f93ed0())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "wysiwyg-hax";
      }
    }]);

    function WysiwygHax() {
      var _this;

      babelHelpers.classCallCheck(this, WysiwygHax);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(WysiwygHax).call(this)); // import child nodes before things start deleting whats in there

      var children = _this.querySelector("template");

      if (children) {
        _this.__importContent = children.innerHTML;
      }

      _this.openDefault = false;
      _this.hideExportButton = false;
      _this.align = "right";
      _this.fieldId = "textarea-input-field";
      _this.fieldName = "data[content]";
      _this.__imported = false;
      _this.redirectLocation = "";
      _this.updatePageData = "";
      window.addEventListener("hax-save", _this._bodyContentUpdated.bind(babelHelpers.assertThisInitialized(_this)));
      window.addEventListener("hax-store-property-updated", _this._haxStorePropertyUpdated.bind(babelHelpers.assertThisInitialized(_this)));
      return _this;
    }

    babelHelpers.createClass(WysiwygHax, [{
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "activeHaxBody") {
            _this2._activeHaxBodyUpdated(_this2[propName]);
          }
        });
      }
    }, {
      key: "createRenderRoot",
      value: function createRenderRoot() {
        return this;
      }
      /**
       * Ensure we've imported our content on initial setup
       */

    }, {
      key: "_activeHaxBodyUpdated",
      value: function _activeHaxBodyUpdated(newValue) {
        // ensure we import our content once we get an initial registration of active body
        if (newValue != null && !this.__imported) {
          this.__imported = true;

          if (this.__importContent) {
            newValue.importContent(this.__importContent);
          }
        }
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        window.removeEventListener("hax-save", this._bodyContentUpdated.bind(this));
        window.removeEventListener("hax-store-property-updated", this._haxStorePropertyUpdated.bind(this));
        babelHelpers.get(babelHelpers.getPrototypeOf(WysiwygHax.prototype), "disconnectedCallback", this).call(this);
      }
      /**
       * Store updated, sync.
       */

    }, {
      key: "_haxStorePropertyUpdated",
      value: function _haxStorePropertyUpdated(e) {
        if (e.detail && babelHelpers.typeof(e.detail.value) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined)) && e.detail.property) {
          if (babelHelpers.typeof(e.detail.value) === "object") {
            this[e.detail.property] = null;
          }

          this[e.detail.property] = e.detail.value;
        }
      }
      /**
       * Set the bubbled up event to the body value that just got changed
       */

    }, {
      key: "_bodyContentUpdated",
      value: function _bodyContentUpdated(e) {
        this.bodyValue = window.HaxStore.instance.activeHaxBody.haxToContent();
      }
    }], [{
      key: "properties",
      get: function get() {
        return {
          /**
           * Default the panel to open
           */
          openDefault: {
            type: Boolean,
            attribute: "open-default"
          },
          redirectLocation: {
            type: String,
            attribute: "redirect-location"
          },
          hideExportButton: {
            type: Boolean,
            attribute: "hide-export-button"
          },

          /**
           * Hide the panel operations (save and cancel),
           */
          hidePanelOps: {
            type: Boolean,
            attribute: "hide-panel-ops"
          },

          /**
           * Hide preferences button
           */
          hidePreferencesButton: {
            type: Boolean,
            attribute: "hide-preferences-button"
          },

          /**
           * Direction to align the hax edit panel
           */
          align: {
            type: String
          },

          /**
           * Data binding of a hidden text area with the value from the hax-body tag
           */
          bodyValue: {
            type: String,
            attribute: "body-value"
          },

          /**
           * Connection object for talking to an app store.
           */
          appStoreConnection: {
            type: String,
            attribute: "app-store-connection"
          },

          /**
           * class on the field
           */
          fieldClass: {
            type: String,
            attribute: "field-class"
          },

          /**
           * fieldId, id value on the input field.
           */
          fieldId: {
            type: String,
            attribute: "field-id"
          },

          /**
           * fieldName, internal to the form in whatever system it's in.
           */
          fieldName: {
            type: String,
            attribute: "field-name"
          },
          syncBody: {
            type: Boolean,
            attribute: "sync-body",
            reflect: true
          },

          /**
           * State of the panel
           */
          editMode: {
            type: Boolean,
            reflect: true,
            attribute: "edit-mode"
          },

          /**
           * Location to save content to.
           */
          endPoint: {
            type: String,
            attribute: "end-point"
          },

          /**
           * Page data, body of text as a string.
           */
          updatePageData: {
            type: String,
            attribute: "update-page-data"
          },

          /**
           * Reference to activeBody.
           */
          activeHaxBody: {
            type: Object
          },
          __imported: {
            type: Boolean
          }
        };
      }
    }]);
    return WysiwygHax;
  }(_litElement.LitElement);

  _exports.WysiwygHax = WysiwygHax;
  window.customElements.define(WysiwygHax.tag, WysiwygHax);
});
