// @lrnwebcomponents/haxcms-elements/lib/ui-components/query/site-query-menu-slice.js
define(["exports", "../../../../../lit-element/lit-element.js", "../../core/haxcms-site-store.js", "../../../../../mobx/lib/mobx.module.js"], function (_exports, _litElement, _haxcmsSiteStore, _mobxModule) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SiteQueryMenuSlice = void 0;

  /**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `site-query-menu-slice`
   * `A slice / 1 level within the hierarchy, via relative parent or deep parent`
   *
  
   * @polymer
   * @demo demo/index.html
   */
  var SiteQueryMenuSlice =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(SiteQueryMenuSlice, _LitElement);
    babelHelpers.createClass(SiteQueryMenuSlice, null, [{
      key: "tag",

      /**
       * Convention our team prefers
       */
      get: function get() {
        return "site-query-menu-slice";
      }
      /**
       * HTMLElement life cycle
       */

    }]);

    function SiteQueryMenuSlice() {
      var _this;

      babelHelpers.classCallCheck(this, SiteQueryMenuSlice);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(SiteQueryMenuSlice).call(this));
      _this.__disposer = [];
      _this.start = 1;
      _this.end = 1000;
      _this.dynamicMethodology = "active";
      _this.fixedId = false;
      _this.noDynamicLevel = false;
      return _this;
    }
    /**
     * LitElement / popular convention
     */


    babelHelpers.createClass(SiteQueryMenuSlice, [{
      key: "updated",

      /**
       * LitElement properties changed
       */
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (["start", "end", "parent", "dynamicMethodology", "_routerManifest", "noDynamicLevel"].includes(propName)) {
            _this2.result = _this2._computeItems(_this2.start, _this2.end, _this2.parent, _this2.dynamicMethodology, _this2._routerManifest, _this2.noDynamicLevel); // fire an event that this is a core piece of the system

            _this2.dispatchEvent(new CustomEvent("result-changed", {
              detail: result
            }));
          }
        });
      }
      /**
       * Compute items leveraging the site query engine
       */

    }, {
      key: "_computeItems",
      value: function _computeItems(start, end, parent, dynamicMethodology, _routerManifest, noDynamicLevel) {
        if (_routerManifest) {
          return _haxcmsSiteStore.store.computeItems(start, end, parent, dynamicMethodology, _routerManifest, noDynamicLevel);
        }
      }
      /**
       * LitElement life cycle
       */

    }, {
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        var _this3 = this;

        (0, _mobxModule.autorun)(function (reaction) {
          _this3._routerManifest = Object.assign({}, (0, _mobxModule.toJS)(_haxcmsSiteStore.store.routerManifest));

          _this3.__disposer.push(reaction);
        });

        if (!this.fixedId) {
          (0, _mobxModule.autorun)(function (reaction) {
            _this3.parent = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.activeId);

            _this3.__disposer.push(reaction);
          });
        }
      }
      /**
       * HTMLElement life cycle
       */

    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        for (var i in this.__disposer) {
          this.__disposer[i].dispose();
        }

        babelHelpers.get(babelHelpers.getPrototypeOf(SiteQueryMenuSlice.prototype), "disconnectedCallback", this).call(this);
      }
    }], [{
      key: "properties",
      get: function get() {
        return {
          /**
           * starting level for the menu items
           */
          start: {
            type: Number
          },

          /**
           * ending level for the menu items
           */
          end: {
            type: Number
          },

          /**
           * parent for the menu id
           */
          parent: {
            type: String
          },

          /**
           * How we should obtain the parent who's children we should show
           * options are active, parent, or ancestor
           */
          dynamicMethodology: {
            type: String,
            attribute: "dynamic-methodology"
          },

          /**
           * Use this boolean to force this to fix to 1 parent
           * Otherwise it will dynamically update (default behavior)
           */
          fixedId: {
            type: Boolean,
            attribute: "fixed-id"
          },

          /**
           * Allow disabling the dynamic leveling
           */
          noDynamicLevel: {
            type: Boolean,
            attribute: "no-dynamic-level"
          },

          /**
           * Results which can be binded to something else
           */
          result: {
            type: Array
          },
          _routerManifest: {
            type: Object
          }
        };
      }
    }]);
    return SiteQueryMenuSlice;
  }(_litElement.LitElement);

  _exports.SiteQueryMenuSlice = SiteQueryMenuSlice;
  window.customElements.define(SiteQueryMenuSlice.tag, SiteQueryMenuSlice);
});
