// wired-elements/lib/wired-base.js
define(["exports", "../../lit/index.js", "../../lit/decorators.js"], function (_exports, _index, _decorators) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.WiredBase = _exports.BaseCSS = void 0;
  _exports.fireEvent = fireEvent;
  _exports.randomSeed = randomSeed;

  var _templateObject_f4a163501efd11eda28c55bb65e7fc65;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var __decorate = void 0 && (void 0).__decorate || function (decorators, target, key, desc) {
    var c = arguments.length,
        r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
        d;
    if ((typeof Reflect === "undefined" ? "undefined" : babelHelpers.typeof(Reflect)) === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) {
      if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    }
    return c > 3 && r && Object.defineProperty(target, key, r), r;
  };

  var __metadata = void 0 && (void 0).__metadata || function (k, v) {
    if ((typeof Reflect === "undefined" ? "undefined" : babelHelpers.typeof(Reflect)) === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
  };

  var BaseCSS = (0, _index.css)(_templateObject_f4a163501efd11eda28c55bb65e7fc65 || (_templateObject_f4a163501efd11eda28c55bb65e7fc65 = babelHelpers.taggedTemplateLiteral(["\n:host {\n  opacity: 0;\n}\n:host(.wired-rendered) {\n  opacity: 1;\n}\n#overlay {\n  position: absolute;\n  top: 0;\n  left: 0;\n  right: 0;\n  bottom: 0;\n  pointer-events: none;\n}\nsvg {\n  display: block;\n}\npath {\n  stroke: currentColor;\n  stroke-width: 0.7;\n  fill: transparent;\n}\n.hidden {\n  display: none !important;\n}\n"])));
  _exports.BaseCSS = BaseCSS;

  var WiredBase = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(WiredBase, _LitElement);

    var _super = _createSuper(WiredBase);

    function WiredBase() {
      var _this;

      babelHelpers.classCallCheck(this, WiredBase);
      _this = _super.apply(this, arguments);
      _this.lastSize = [0, 0];
      _this.seed = Math.floor(Math.random() * Math.pow(2, 31));
      return _this;
    }

    babelHelpers.createClass(WiredBase, [{
      key: "updated",
      value: function updated(_changed) {
        this.wiredRender();
      }
    }, {
      key: "wiredRender",
      value: function wiredRender() {
        var force = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;

        if (this.svg) {
          var size = this.canvasSize();

          if (!force && size[0] === this.lastSize[0] && size[1] === this.lastSize[1]) {
            return;
          }

          while (this.svg.hasChildNodes()) {
            this.svg.removeChild(this.svg.lastChild);
          }

          this.svg.setAttribute('width', "".concat(size[0]));
          this.svg.setAttribute('height', "".concat(size[1]));
          this.draw(this.svg, size);
          this.lastSize = size;
          this.classList.add('wired-rendered');
        }
      }
    }, {
      key: "fire",
      value: function fire(name, detail) {
        fireEvent(this, name, detail);
      }
    }]);
    return WiredBase;
  }(_index.LitElement);

  _exports.WiredBase = WiredBase;

  __decorate([(0, _decorators.query)('svg'), __metadata("design:type", SVGSVGElement)], WiredBase.prototype, "svg", void 0);

  function randomSeed() {
    return Math.floor(Math.random() * Math.pow(2, 31));
  }

  function fireEvent(e, name, detail) {
    e.dispatchEvent(new CustomEvent(name, {
      composed: true,
      bubbles: true,
      detail: detail
    }));
  }
});
