// @lrnwebcomponents/app-hax/lib/v1/app-hax-search-results.js
define(["exports", "../../../simple-colors/simple-colors.js", "../../../../lit/index.js", "../../../../mobx/dist/mobx.esm.js", "../../../utils/utils.js", "./AppHaxStore.js", "./app-hax-site-bar.js", "./app-hax-site-details.js"], function (_exports, _simpleColors, _index, _mobxEsm, _utils, _AppHaxStore, _appHaxSiteBar, _appHaxSiteDetails) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.AppHaxSearchResults = void 0;

  var _templateObject_ae0dfd508e0011ed99ce491205e67b1d, _templateObject2_ae0dfd508e0011ed99ce491205e67b1d, _templateObject3_ae0dfd508e0011ed99ce491205e67b1d, _templateObject4_ae0dfd508e0011ed99ce491205e67b1d, _templateObject5_ae0dfd508e0011ed99ce491205e67b1d;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var AppHaxSearchResults = /*#__PURE__*/function (_SimpleColors) {
    babelHelpers.inherits(AppHaxSearchResults, _SimpleColors);

    var _super = _createSuper(AppHaxSearchResults);

    function AppHaxSearchResults() {
      var _this;

      babelHelpers.classCallCheck(this, AppHaxSearchResults);
      _this = _super.call(this);
      _this.searchItems = [];
      _this.displayItems = [];
      _this.searchTerm = "";
      _this.dark = false;
      (0, _mobxEsm.autorun)(function () {
        _this.searchTerm = (0, _mobxEsm.toJS)(_AppHaxStore.store.searchTerm);
      });
      (0, _mobxEsm.autorun)(function () {
        _this.dark = (0, _mobxEsm.toJS)(_AppHaxStore.store.darkMode);
      });
      (0, _mobxEsm.autorun)(function () {
        var manifest = (0, _mobxEsm.toJS)(_AppHaxStore.store.manifest);

        if (manifest && manifest.items) {
          _this.searchItems = manifest.items;
          _this.displayItems = babelHelpers.toConsumableArray(_this.searchItems);
        }
      });
      return _this;
    } // Site.json is coming from


    babelHelpers.createClass(AppHaxSearchResults, [{
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        if (babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSearchResults.prototype), "updated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSearchResults.prototype), "updated", this).call(this, changedProperties);
        }

        changedProperties.forEach(function (oldValue, propName) {
          if (propName === "searchTerm") {
            _this2.displayItems = _this2.searchItems.filter(function (word) {
              if (word.title.toLowerCase().includes(_this2.searchTerm.toLowerCase()) || word.description.toLowerCase().includes(_this2.searchTerm.toLowerCase()) || word.author.toLowerCase().includes(_this2.searchTerm.toLowerCase()) || word.slug.toLowerCase().includes(_this2.searchTerm.toLowerCase())) {
                return true;
              }

              return false;
            });
          }
        });
      }
    }, {
      key: "render",
      value: function render() {
        var _this3 = this;

        return (0, _index.html)(_templateObject_ae0dfd508e0011ed99ce491205e67b1d || (_templateObject_ae0dfd508e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral(["\n      <ul id=\"results\">\n        ", "\n      </ul>\n    "])), this.displayItems.length > 0 ? this.displayItems.map(function (item) {
          return (0, _index.html)(_templateObject2_ae0dfd508e0011ed99ce491205e67b1d || (_templateObject2_ae0dfd508e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral([" <li>\n                  <app-hax-site-bar\n                    @opened-changed=\"", "\"\n                    ?dark=\"", "\"\n                    accent-color=\"", "\"\n                    icon-link=\"", "\"\n                    icon=\"", "\"\n                  >\n                    <a href=\"", "\" slot=\"heading\">", "</a>\n                    <span slot=\"subHeading\">", "</span>\n                    <app-hax-site-details\n                      slot=\"band\"\n                      .details=\"", "\"\n                      site-id=\"", "\"\n                    >\n                      <div class=\"description\" slot=\"pre\">\n                        ", "\n                      </div>\n                    </app-hax-site-details>\n                  </app-hax-site-bar>\n                </li>"])), _this3.openedChanged, _this3.dark, (0, _utils.varGet)(item, "metadata.theme.variables.cssVariable", "orange").replace("--simple-colors-default-theme-", "").replace("-7", ""), item.slug, (0, _utils.varGet)(item, "metadata.theme.variables.icon", "link"), item.slug, item.title, item.author, _this3.getItemDetails(item), item.id, item.description);
        }) : (0, _index.html)(_templateObject3_ae0dfd508e0011ed99ce491205e67b1d || (_templateObject3_ae0dfd508e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral(["<div id=\"noResult\">\n              No results for\n              ", ".\n            </div>"])), this.searchTerm !== "" ? (0, _index.html)(_templateObject4_ae0dfd508e0011ed99ce491205e67b1d || (_templateObject4_ae0dfd508e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral(["<strong>\"", "\"</strong>"])), this.searchTerm) : "your account, try starting a new journey!"));
      }
    }, {
      key: "getItemDetails",
      value: function getItemDetails(item) {
        var details = {
          created: (0, _utils.varGet)(item, "metadata.site.created", new Date() / 1000),
          updated: (0, _utils.varGet)(item, "metadata.site.updated", new Date() / 1000),
          pages: (0, _utils.varGet)(item, "metadata.pageCount", 0),
          url: item.slug
        };
        return details;
      }
    }, {
      key: "openedChanged",
      value: function openedChanged(e) {
        _AppHaxStore.store.appEl.playSound("click");

        if (!e.detail.value) {
          this.shadowRoot.querySelector("app-hax-site-details").setAttribute("tabindex", "-1");
        } else {
          this.shadowRoot.querySelector("app-hax-site-details").removeAttribute("tabindex");
        }
      }
    }], [{
      key: "tag",
      get: // a convention I enjoy so you can change the tag name in 1 place
      function get() {
        return "app-hax-search-results";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSearchResults), "properties", this)), {}, {
          searchTerm: {
            type: String,
            reflect: true
          },
          searchItems: {
            type: Array
          },
          displayItems: {
            type: Array
          }
        });
      }
    }, {
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSearchResults), "styles", this)), [(0, _index.css)(_templateObject5_ae0dfd508e0011ed99ce491205e67b1d || (_templateObject5_ae0dfd508e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          overflow: hidden;\n        }\n        ul,\n        li {\n          margin: 0;\n          padding: 0;\n          list-style: none;\n        }\n        app-hax-site-bar {\n          margin: 8px 0;\n        }\n        .description {\n          max-height: 64px;\n          overflow: hidden;\n          text-overflow: ellipsis;\n        }\n\n        @media (max-width: 800px) {\n          app-hax-site-bar {\n            --main-banner-width: 60vw;\n          }\n        }\n        @media (max-width: 400px) {\n          app-hax-site-bar {\n            --main-banner-width: 90vw;\n          }\n        }\n        span[slot=\"band\"] {\n          height: 48px;\n          overflow: hidden;\n          text-overflow: ellipsis;\n          margin-bottom: 8px;\n        }\n        :host([dark]) #noResult {\n          color: var(--simple-colors-default-theme-grey-12, black);\n        }\n      "])))]);
      }
    }]);
    return AppHaxSearchResults;
  }(_simpleColors.SimpleColors);

  _exports.AppHaxSearchResults = AppHaxSearchResults;
  customElements.define(AppHaxSearchResults.tag, AppHaxSearchResults);
});
