// @lrnwebcomponents/app-hax/lib/v1/app-hax-site-details.js
define(["exports", "require", "../../../../lit/index.js", "../../../simple-datetime/simple-datetime.js", "../../../../mobx/dist/mobx.esm.js", "./AppHaxStore.js", "../../../simple-colors/simple-colors.js"], function (_exports, _require, _index, _simpleDatetime, _mobxEsm, _AppHaxStore, _simpleColors) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.AppHaxSiteDetails = void 0;
  _require = _interopRequireWildcard(_require);

  var _templateObject_ea948fc0ce4611ecb13bcf602138080f, _templateObject2_ea948fc0ce4611ecb13bcf602138080f, _templateObject3_ea948fc0ce4611ecb13bcf602138080f;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  // wrapper to simplify the slug if it has additional values on it
  function makeSlug(url) {
    var slug = "site";

    if (url) {
      var tmp = url.split("sites/");

      if (tmp.length > 1) {
        slug = tmp.pop().replace("/", "");
      }
    }

    return slug;
  } // EXPORT (so make available to other documents that reference this file) a class, that extends LitElement
  // which has the magic life-cycles and developer experience below added


  var AppHaxSiteDetails = /*#__PURE__*/function (_SimpleColors) {
    babelHelpers.inherits(AppHaxSiteDetails, _SimpleColors);

    var _super = _createSuper(AppHaxSiteDetails);

    // HTMLElement life-cycle, built in; use this for setting defaults
    function AppHaxSiteDetails() {
      var _this;

      babelHelpers.classCallCheck(this, AppHaxSiteDetails);
      _this = _super.call(this);
      _this.need = "all need to succeed";
      _this.details = {};
      _this.siteId = "";
      _this.detailOps = [{
        name: "Copy",
        op: "copySite",
        icon: "icons:content-copy"
      }, {
        name: "Download",
        op: "downloadSite",
        icon: "file-download"
      }, {
        name: "Archive",
        op: "archiveSite",
        icon: "icons:archive"
      }, {
        name: "Delete",
        op: "deleteSite",
        icon: "icons:delete-forever"
      }];
      return _this;
    } // properties that you wish to use as data in HTML, CSS, and the updated life-cycle


    babelHelpers.createClass(AppHaxSiteDetails, [{
      key: "siteOperation",
      value: // eslint-disable-next-line class-methods-use-this
      function siteOperation(e) {
        var _this2 = this;

        // let elements;
        _AppHaxStore.store.appEl.playSound("click");

        var target = e.target; // avoid label trigger

        if (target.tagName === "DIV") {
          target = target.parentNode;
        }

        var div = document.createElement("div");
        var op = target.getAttribute("data-site-operation");
        var opName = target.getAttribute("data-site-operation-name");
        var siteID = target.getAttribute("data-site");
        _AppHaxStore.store.activeSiteOp = op;
        _AppHaxStore.store.activeSiteId = siteID;
        new Promise(function (res, rej) {
          return _require.default(["../../../simple-modal/simple-modal.js"], res, rej);
        }).then(function () {
          setTimeout(function () {
            var site = (0, _mobxEsm.toJS)(_AppHaxStore.store.manifest.items.filter(function (item) {
              return item.id === siteID;
            }).pop());
            div.appendChild(document.createTextNode("Are you sure you want to ".concat(op.replace("Site", ""), " ").concat(site.metadata.site.name, "?")));

            if (op === "deleteSite") {
              var undone = document.createElement("div");
              undone.style.color = "red";
              undone.style.fontWeight = "bold";
              undone.appendChild(document.createTextNode("This can not be undone."));
              div.appendChild(undone);
            }

            var bcontainer = document.createElement("div");
            var b = document.createElement("button");
            b.innerText = "Confirm";
            b.classList.add("hax-modal-btn");
            b.addEventListener("click", _this2.confirmOperation.bind(_this2));
            bcontainer.appendChild(b);
            var b2 = document.createElement("button");
            b2.innerText = "Cancel";
            b2.classList.add("hax-modal-btn");
            b2.classList.add("cancel");
            b2.addEventListener("click", _this2.cancelOperation.bind(_this2));
            bcontainer.appendChild(b2);

            _this2.dispatchEvent(new CustomEvent("simple-modal-show", {
              bubbles: true,
              cancelable: true,
              composed: true,
              detail: {
                title: "".concat(opName, " ").concat(site.metadata.site.name, "?"),
                elements: {
                  content: div,
                  buttons: bcontainer
                },
                invokedBy: target,
                styles: {
                  "--simple-modal-titlebar-background": "orange",
                  "--simple-modal-titlebar-color": "black",
                  "--simple-modal-width": "30vw",
                  "--simple-modal-min-width": "300px",
                  "--simple-modal-z-index": "100000000",
                  "--simple-modal-height": "20vh",
                  "--simple-modal-min-height": "300px",
                  "--simple-modal-titlebar-height": "80px"
                }
              }
            }));
          }, 0);
        });
      }
    }, {
      key: "cancelOperation",
      value: function cancelOperation() {
        _AppHaxStore.store.activeSiteOp = "";
        _AppHaxStore.store.activeSiteId = null;
        window.dispatchEvent(new CustomEvent("simple-modal-hide"));

        _AppHaxStore.store.appEl.playSound("error");
      }
    }, {
      key: "confirmOperation",
      value: function () {
        var _confirmOperation = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
          var op, site;
          return regeneratorRuntime.wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  op = (0, _mobxEsm.toJS)(_AppHaxStore.store.activeSiteOp);
                  site = (0, _mobxEsm.toJS)(_AppHaxStore.store.activeSite); // @todo bother to implement these / translate to the path via switch

                  _context.next = 4;
                  return _AppHaxStore.store.AppHaxAPI.makeCall(op, {
                    site: {
                      name: site.metadata.site.name,
                      id: site.id
                    }
                  }, true, function () {
                    var activeOp = (0, _mobxEsm.toJS)(_AppHaxStore.store.activeSiteOp); // download is weird relative to the others

                    // download is weird relative to the others
                    if (activeOp === "downloadSite") {
                      // cheat to download a file path
                      window.open(_AppHaxStore.store.AppHaxAPI.lastResponse.downloadSite.data.link, "_blank");
                    } else {
                      _AppHaxStore.store.refreshSiteListing();
                    }
                  });

                case 4:
                  window.dispatchEvent(new CustomEvent("simple-modal-hide"));

                  _AppHaxStore.store.appEl.playSound("success");

                  _AppHaxStore.store.toast("".concat(site.metadata.site.name, " ").concat(op.replace("Site", ""), " successful!"), 3000, {
                    hat: "random"
                  });

                case 7:
                case "end":
                  return _context.stop();
              }
            }
          }, _callee);
        }));

        function confirmOperation() {
          return _confirmOperation.apply(this, arguments);
        }

        return confirmOperation;
      }() // HTML - specific to Lit

    }, {
      key: "render",
      value: function render() {
        var _this3 = this;

        return (0, _index.html)(_templateObject_ea948fc0ce4611ecb13bcf602138080f || (_templateObject_ea948fc0ce4611ecb13bcf602138080f = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"pre\"><slot name=\"pre\"></slot></div>\n      <div class=\"flex-container\" id=\"infomation\">\n          <div class=\"info-group\">\n            <div class=\"info-headings\">created</div>\n            <simple-datetime\n                format=\"m/j/y\"\n                .timestamp=\"", "\"\n                unix\n                class=\"info-date\"\n            ></simple-datetime>\n          </div>\n          <div class=\"info-group\">\n            <div class=\"info-headings\">updated</div>\n            <simple-datetime\n                format=\"m/j/y\"\n                .timestamp=\"", "\"\n                unix\n                class=\"info-date\"\n            ></simple-datetime>\n          </div>\n          <div class = info-group>\n            <div class=\"info-headings\">pages</div>\n            <div class=\"info-item\">", "</div>\n          </div>\n          <div class = info-group>\n            <div class=\"info-headings\">URL</div>\n            <a href=\"", "\" class=\"info-item\" id=\"slug\">", "</a>\n          </div>\n      </div>\n        <div class=\"flex-container\" id=\"actions\">\n          ", "\n        </div>\n      </div>\n    "])), this.details.created, this.details.updated, this.details.pages, this.details.url, makeSlug(this.details.url), this.detailOps.map(function (item) {
          return (0, _index.html)(_templateObject2_ea948fc0ce4611ecb13bcf602138080f || (_templateObject2_ea948fc0ce4611ecb13bcf602138080f = babelHelpers.taggedTemplateLiteral(["\n              <div class=\"info-group\">\n                <simple-icon-button-lite\n                  data-site=\"", "\"\n                  data-site-operation=\"", "\"\n                  data-site-operation-name=\"", "\"\n                  icon=\"", "\"\n                  id=\"op-", "\"\n                  class=\"info-icon\"\n                  @click=", "\n                >\n                  <div class=\"info-item\">", "</div>\n                </simple-icon-button-lite>\n                <simple-tooltip for=\"op-", "\" position=\"bottom\"\n                  >", " Site</simple-tooltip\n                >\n              </div>\n            "])), _this3.siteId, item.op, item.name, item.icon, item.op, _this3.siteOperation, item.name.toLowerCase(), item.op, item.name);
        }));
      }
    }], [{
      key: "tag",
      get: // a convention I enjoy so you can change the tag name in 1 place
      function get() {
        return "app-hax-site-details";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSiteDetails), "properties", this)), {}, {
          details: {
            type: Object
          },
          siteId: {
            type: String,
            attribute: "site-id"
          }
        });
      } // CSS - specific to Lit

    }, {
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSiteDetails), "styles", this)), [(0, _index.css)(_templateObject3_ea948fc0ce4611ecb13bcf602138080f || (_templateObject3_ea948fc0ce4611ecb13bcf602138080f = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: flex;\n          flex-direction: column;\n          justify-content: center;\n          font-size: 12px;\n          align-items: stretch;\n          background-color: var(--simple-colors-default-theme-grey-2);\n          height: 208px;\n        }\n\n        .flex-container {\n          flex: 1;\n          background-color: var(--simple-colors-default-theme-grey-2);\n          margin: 8px;\n          display: flex;\n          flex-direction: row;\n          justify-content: space-around;\n          align-items: center;\n        }\n        .info-group {\n          height: 100%;\n          max-width: 25%;\n          display: flex;\n          flex-direction: column;\n          justify-content: space-around;\n          align-items: center;\n          padding: 0px;\n          flex: 1;\n        }\n        simple-icon-button-lite:active,\n        simple-icon-button-lite:hover,\n        simple-icon-button-lite:focus {\n          background-color: var(--simple-colors-default-theme-grey-4, #eeeeee);\n          outline: 2px solid var(--simple-colors-default-theme-grey-12);\n          outline-offset: 1px;\n        }\n\n        .info-headings {\n          font-size: 12px;\n        }\n        .info-item {\n          font-family: \"Press Start 2P\", sans-serif;\n          display: block;\n          text-overflow: ellipsis;\n          overflow: hidden;\n          color: var(--simple-colors-default-theme-grey-12);\n          line-height: 12px;\n          max-width: 100%;\n          font-size: 12px;\n        }\n        .pre ::slotted(*) {\n          padding: 12px;\n          overflow: hidden;\n          text-overflow: ellipsis;\n          max-width: 50%;\n          display: inline-flex;\n        }\n        a {\n          text-decoration: underline;\n        }\n        .info-date {\n          color: var(--simple-colors-default-theme-grey-12);\n          line-height: 12px;\n          font-size: 12px;\n        }\n\n        .info-icon {\n          --simple-icon-width: 49px;\n          --simple-icon-height: 49px;\n          --simple-icon-button-border-radius: 0px;\n          --simple-icon-button-border: 0px;\n          outline: 0;\n          border: 2px solid var(--simple-colors-default-theme-grey-12);\n          border-radius: 4px;\n          padding: 4px;\n          width: 80%;\n        }\n        #op-deleteSite {\n          color: var(--simple-colors-default-theme-red-7);\n        }\n        .info-icon::part(button) {\n          outline: none;\n        }\n        @media (max-width: 500px) {\n          .btn-group button {\n            padding: 4px;\n            margin: 4px 0;\n          }\n          .flex-container > div {\n            margin: 0px;\n          }\n        }\n      "])))]);
      }
    }]);
    return AppHaxSiteDetails;
  }(_simpleColors.SimpleColors);

  _exports.AppHaxSiteDetails = AppHaxSiteDetails;
  customElements.define(AppHaxSiteDetails.tag, AppHaxSiteDetails);
});
