// @lrnwebcomponents/app-hax/lib/v1/app-hax-steps.js
define(["exports", "meta", "require", "../../../../lit/index.js", "../../../simple-colors/simple-colors.js", "../../../../mobx/dist/mobx.esm.js", "./AppHaxStore.js", "../../../utils/utils.js", "../../../../scrollable-component/index.js", "../../../simple-icon/lib/simple-icon-lite.js", "../../../micro-frontend-registry/micro-frontend-registry.js", "../../../micro-frontend-registry/lib/microServices.js", "./app-hax-site-button.js", "./app-hax-hat-progress.js", "./app-hax-button.js"], function (_exports, meta, _require, _index, _simpleColors, _mobxEsm, _AppHaxStore, _utils, _index2, _simpleIconLite, _microFrontendRegistry, _microServices, _appHaxSiteButton, _appHaxHatProgress, _appHaxButton) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.AppHaxSteps = void 0;
  meta = _interopRequireWildcard(meta);
  _require = _interopRequireWildcard(_require);

  var _templateObject_77261b40809211edaa2833647dc78c6c, _templateObject2_77261b40809211edaa2833647dc78c6c, _templateObject3_77261b40809211edaa2833647dc78c6c, _templateObject4_77261b40809211edaa2833647dc78c6c, _templateObject5_77261b40809211edaa2833647dc78c6c, _templateObject6_77261b40809211edaa2833647dc78c6c, _templateObject7_77261b40809211edaa2833647dc78c6c, _templateObject8_77261b40809211edaa2833647dc78c6c, _templateObject9_77261b40809211edaa2833647dc78c6c, _templateObject10_77261b40809211edaa2833647dc78c6c;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _regeneratorRuntime() { "use strict"; /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/facebook/regenerator/blob/main/LICENSE */ _regeneratorRuntime = function _regeneratorRuntime() { return exports; }; var exports = {}, Op = Object.prototype, hasOwn = Op.hasOwnProperty, $Symbol = "function" == typeof Symbol ? Symbol : {}, iteratorSymbol = $Symbol.iterator || "@@iterator", asyncIteratorSymbol = $Symbol.asyncIterator || "@@asyncIterator", toStringTagSymbol = $Symbol.toStringTag || "@@toStringTag"; function define(obj, key, value) { return Object.defineProperty(obj, key, { value: value, enumerable: !0, configurable: !0, writable: !0 }), obj[key]; } try { define({}, ""); } catch (err) { define = function define(obj, key, value) { return obj[key] = value; }; } function wrap(innerFn, outerFn, self, tryLocsList) { var protoGenerator = outerFn && babelHelpers.instanceof(outerFn.prototype, Generator) ? outerFn : Generator, generator = Object.create(protoGenerator.prototype), context = new Context(tryLocsList || []); return generator._invoke = function (innerFn, self, context) { var state = "suspendedStart"; return function (method, arg) { if ("executing" === state) throw new Error("Generator is already running"); if ("completed" === state) { if ("throw" === method) throw arg; return doneResult(); } for (context.method = method, context.arg = arg;;) { var delegate = context.delegate; if (delegate) { var delegateResult = maybeInvokeDelegate(delegate, context); if (delegateResult) { if (delegateResult === ContinueSentinel) continue; return delegateResult; } } if ("next" === context.method) context.sent = context._sent = context.arg;else if ("throw" === context.method) { if ("suspendedStart" === state) throw state = "completed", context.arg; context.dispatchException(context.arg); } else "return" === context.method && context.abrupt("return", context.arg); state = "executing"; var record = tryCatch(innerFn, self, context); if ("normal" === record.type) { if (state = context.done ? "completed" : "suspendedYield", record.arg === ContinueSentinel) continue; return { value: record.arg, done: context.done }; } "throw" === record.type && (state = "completed", context.method = "throw", context.arg = record.arg); } }; }(innerFn, self, context), generator; } function tryCatch(fn, obj, arg) { try { return { type: "normal", arg: fn.call(obj, arg) }; } catch (err) { return { type: "throw", arg: err }; } } exports.wrap = wrap; var ContinueSentinel = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} var IteratorPrototype = {}; define(IteratorPrototype, iteratorSymbol, function () { return this; }); var getProto = Object.getPrototypeOf, NativeIteratorPrototype = getProto && getProto(getProto(values([]))); NativeIteratorPrototype && NativeIteratorPrototype !== Op && hasOwn.call(NativeIteratorPrototype, iteratorSymbol) && (IteratorPrototype = NativeIteratorPrototype); var Gp = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(IteratorPrototype); function defineIteratorMethods(prototype) { ["next", "throw", "return"].forEach(function (method) { define(prototype, method, function (arg) { return this._invoke(method, arg); }); }); } function AsyncIterator(generator, PromiseImpl) { function invoke(method, arg, resolve, reject) { var record = tryCatch(generator[method], generator, arg); if ("throw" !== record.type) { var result = record.arg, value = result.value; return value && "object" == babelHelpers.typeof(value) && hasOwn.call(value, "__await") ? PromiseImpl.resolve(value.__await).then(function (value) { invoke("next", value, resolve, reject); }, function (err) { invoke("throw", err, resolve, reject); }) : PromiseImpl.resolve(value).then(function (unwrapped) { result.value = unwrapped, resolve(result); }, function (error) { return invoke("throw", error, resolve, reject); }); } reject(record.arg); } var previousPromise; this._invoke = function (method, arg) { function callInvokeWithMethodAndArg() { return new PromiseImpl(function (resolve, reject) { invoke(method, arg, resolve, reject); }); } return previousPromise = previousPromise ? previousPromise.then(callInvokeWithMethodAndArg, callInvokeWithMethodAndArg) : callInvokeWithMethodAndArg(); }; } function maybeInvokeDelegate(delegate, context) { var method = delegate.iterator[context.method]; if (undefined === method) { if (context.delegate = null, "throw" === context.method) { if (delegate.iterator.return && (context.method = "return", context.arg = undefined, maybeInvokeDelegate(delegate, context), "throw" === context.method)) return ContinueSentinel; context.method = "throw", context.arg = new TypeError("The iterator does not provide a 'throw' method"); } return ContinueSentinel; } var record = tryCatch(method, delegate.iterator, context.arg); if ("throw" === record.type) return context.method = "throw", context.arg = record.arg, context.delegate = null, ContinueSentinel; var info = record.arg; return info ? info.done ? (context[delegate.resultName] = info.value, context.next = delegate.nextLoc, "return" !== context.method && (context.method = "next", context.arg = undefined), context.delegate = null, ContinueSentinel) : info : (context.method = "throw", context.arg = new TypeError("iterator result is not an object"), context.delegate = null, ContinueSentinel); } function pushTryEntry(locs) { var entry = { tryLoc: locs[0] }; 1 in locs && (entry.catchLoc = locs[1]), 2 in locs && (entry.finallyLoc = locs[2], entry.afterLoc = locs[3]), this.tryEntries.push(entry); } function resetTryEntry(entry) { var record = entry.completion || {}; record.type = "normal", delete record.arg, entry.completion = record; } function Context(tryLocsList) { this.tryEntries = [{ tryLoc: "root" }], tryLocsList.forEach(pushTryEntry, this), this.reset(!0); } function values(iterable) { if (iterable) { var iteratorMethod = iterable[iteratorSymbol]; if (iteratorMethod) return iteratorMethod.call(iterable); if ("function" == typeof iterable.next) return iterable; if (!isNaN(iterable.length)) { var i = -1, next = function next() { for (; ++i < iterable.length;) { if (hasOwn.call(iterable, i)) return next.value = iterable[i], next.done = !1, next; } return next.value = undefined, next.done = !0, next; }; return next.next = next; } } return { next: doneResult }; } function doneResult() { return { value: undefined, done: !0 }; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, define(Gp, "constructor", GeneratorFunctionPrototype), define(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = define(GeneratorFunctionPrototype, toStringTagSymbol, "GeneratorFunction"), exports.isGeneratorFunction = function (genFun) { var ctor = "function" == typeof genFun && genFun.constructor; return !!ctor && (ctor === GeneratorFunction || "GeneratorFunction" === (ctor.displayName || ctor.name)); }, exports.mark = function (genFun) { return Object.setPrototypeOf ? Object.setPrototypeOf(genFun, GeneratorFunctionPrototype) : (genFun.__proto__ = GeneratorFunctionPrototype, define(genFun, toStringTagSymbol, "GeneratorFunction")), genFun.prototype = Object.create(Gp), genFun; }, exports.awrap = function (arg) { return { __await: arg }; }, defineIteratorMethods(AsyncIterator.prototype), define(AsyncIterator.prototype, asyncIteratorSymbol, function () { return this; }), exports.AsyncIterator = AsyncIterator, exports.async = function (innerFn, outerFn, self, tryLocsList, PromiseImpl) { void 0 === PromiseImpl && (PromiseImpl = Promise); var iter = new AsyncIterator(wrap(innerFn, outerFn, self, tryLocsList), PromiseImpl); return exports.isGeneratorFunction(outerFn) ? iter : iter.next().then(function (result) { return result.done ? result.value : iter.next(); }); }, defineIteratorMethods(Gp), define(Gp, toStringTagSymbol, "Generator"), define(Gp, iteratorSymbol, function () { return this; }), define(Gp, "toString", function () { return "[object Generator]"; }), exports.keys = function (object) { var keys = []; for (var key in object) { keys.push(key); } return keys.reverse(), function next() { for (; keys.length;) { var key = keys.pop(); if (key in object) return next.value = key, next.done = !1, next; } return next.done = !0, next; }; }, exports.values = values, Context.prototype = { constructor: Context, reset: function reset(skipTempReset) { if (this.prev = 0, this.next = 0, this.sent = this._sent = undefined, this.done = !1, this.delegate = null, this.method = "next", this.arg = undefined, this.tryEntries.forEach(resetTryEntry), !skipTempReset) for (var name in this) { "t" === name.charAt(0) && hasOwn.call(this, name) && !isNaN(+name.slice(1)) && (this[name] = undefined); } }, stop: function stop() { this.done = !0; var rootRecord = this.tryEntries[0].completion; if ("throw" === rootRecord.type) throw rootRecord.arg; return this.rval; }, dispatchException: function dispatchException(exception) { if (this.done) throw exception; var context = this; function handle(loc, caught) { return record.type = "throw", record.arg = exception, context.next = loc, caught && (context.method = "next", context.arg = undefined), !!caught; } for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i], record = entry.completion; if ("root" === entry.tryLoc) return handle("end"); if (entry.tryLoc <= this.prev) { var hasCatch = hasOwn.call(entry, "catchLoc"), hasFinally = hasOwn.call(entry, "finallyLoc"); if (hasCatch && hasFinally) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } else if (hasCatch) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); } else { if (!hasFinally) throw new Error("try statement without catch or finally"); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } } } }, abrupt: function abrupt(type, arg) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc <= this.prev && hasOwn.call(entry, "finallyLoc") && this.prev < entry.finallyLoc) { var finallyEntry = entry; break; } } finallyEntry && ("break" === type || "continue" === type) && finallyEntry.tryLoc <= arg && arg <= finallyEntry.finallyLoc && (finallyEntry = null); var record = finallyEntry ? finallyEntry.completion : {}; return record.type = type, record.arg = arg, finallyEntry ? (this.method = "next", this.next = finallyEntry.finallyLoc, ContinueSentinel) : this.complete(record); }, complete: function complete(record, afterLoc) { if ("throw" === record.type) throw record.arg; return "break" === record.type || "continue" === record.type ? this.next = record.arg : "return" === record.type ? (this.rval = this.arg = record.arg, this.method = "return", this.next = "end") : "normal" === record.type && afterLoc && (this.next = afterLoc), ContinueSentinel; }, finish: function finish(finallyLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.finallyLoc === finallyLoc) return this.complete(entry.completion, entry.afterLoc), resetTryEntry(entry), ContinueSentinel; } }, catch: function _catch(tryLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc === tryLoc) { var record = entry.completion; if ("throw" === record.type) { var thrown = record.arg; resetTryEntry(entry); } return thrown; } } throw new Error("illegal catch attempt"); }, delegateYield: function delegateYield(iterable, resultName, nextLoc) { return this.delegate = { iterator: values(iterable), resultName: resultName, nextLoc: nextLoc }, "next" === this.method && (this.arg = undefined), ContinueSentinel; } }, exports; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var homeIcon = new URL("../assets/images/Home.svg", meta.url).href;
  var disabledCircle = new URL("../assets/images/DisabledCircle.svg", meta.url).href;
  var transparentCircle = new URL("../assets/images/TransparentCircle.svg", meta.url).href;
  var enabledCircle = new URL("../assets/images/EnabledCircle.svg", meta.url).href;
  var themeContext = {
    portfolio: ["haxor-slevin", "bootstrap-theme"],
    course: ["clean-one", "clean-two", "learn-two-theme"],
    website: ["bootstrap-theme", "outline-player", "haxor-slevin"]
  };

  var AppHaxSteps = /*#__PURE__*/function (_SimpleColors) {
    babelHelpers.inherits(AppHaxSteps, _SimpleColors);

    var _super = _createSuper(AppHaxSteps);

    function AppHaxSteps() {
      var _this;

      babelHelpers.classCallCheck(this, AppHaxSteps);
      _this = _super.call(this);
      _this.nameTyped = "";
      _this.stepRoutes = [];
      _this._progressReady = false;
      _this.step = null;
      _this.loaded = false;
      _this.themeNames = [];
      _this.appSettings = {};
      (0, _mobxEsm.autorun)(function () {
        _this.appSettings = (0, _mobxEsm.toJS)(_AppHaxStore.store.appSettings);
        var contextKey = (0, _mobxEsm.toJS)(_AppHaxStore.store.site.structure);
        _this.themeNames = Object.keys(_this.appSettings.themes).filter(function (value) {
          return contextKey && themeContext[contextKey] && themeContext[contextKey].includes(value);
        });
      });
      (0, _mobxEsm.autorun)(function () {
        _this.dark = (0, _mobxEsm.toJS)(_AppHaxStore.store.darkMode);
      });
      (0, _mobxEsm.autorun)(function () {
        (0, _utils.localStorageSet)("app-hax-step", (0, _mobxEsm.toJS)(_AppHaxStore.store.step));
      });
      (0, _mobxEsm.autorun)(function () {
        (0, _utils.localStorageSet)("app-hax-site", (0, _mobxEsm.toJS)(_AppHaxStore.store.site));
        _this.step = _AppHaxStore.store.stepTest(_this.step);
      });
      (0, _mobxEsm.autorun)(function () {
        if ((0, _mobxEsm.toJS)(_AppHaxStore.store.createSiteSteps) && (0, _mobxEsm.toJS)(_AppHaxStore.store.location)) {
          _this.step = _AppHaxStore.store.stepTest(_this.step);
        }
      }); // routes, but only the ones that have a step property

      (0, _mobxEsm.autorun)(function () {
        var routes = (0, _mobxEsm.toJS)(_AppHaxStore.store.routes);
        _this.stepRoutes = routes.filter(function (item) {
          return item.step;
        });
      });
      return _this;
    }

    babelHelpers.createClass(AppHaxSteps, [{
      key: "chooseStructure",
      value: // step 1
      function chooseStructure(e) {
        if (!e.target.comingSoon) {
          var value = e.target.value;
          _AppHaxStore.store.site.structure = value;

          _AppHaxStore.store.appEl.playSound("click2");
        }
      } // step 2

    }, {
      key: "chooseType",
      value: function chooseType(e) {
        if (!e.target.comingSoon) {
          var type = e.target.type;
          _AppHaxStore.store.site.type = type;

          _AppHaxStore.store.appEl.playSound("click2");
        }
      } // step 2, doc import

    }, {
      key: "docxImport",
      value: function () {
        var _docxImport = babelHelpers.asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee2(e) {
          var _this2 = this;

          var type;
          return _regeneratorRuntime().wrap(function _callee2$(_context2) {
            while (1) {
              switch (_context2.prev = _context2.next) {
                case 0:
                  if (!e.target.comingSoon) {
                    type = e.target.type;
                    new Promise(function (res, rej) {
                      return _require.default(["../../../file-system-broker/lib/docx-file-system-broker.js"], res, rej);
                    }).then( /*#__PURE__*/function () {
                      var _ref = babelHelpers.asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee(e) {
                        var broker, file, formData, response;
                        return _regeneratorRuntime().wrap(function _callee$(_context) {
                          while (1) {
                            switch (_context.prev = _context.next) {
                              case 0:
                                // enable core services
                                (0, _microServices.enableServices)(["haxcms"]); // get the broker for docx selection

                                broker = window.FileSystemBroker.requestAvailability();
                                _context.next = 4;
                                return broker.loadFile("docx");

                              case 4:
                                file = _context.sent;
                                // tee up as a form for upload
                                formData = new FormData();
                                formData.append("method", "site"); // this is a site based importer

                                formData.append("type", (0, _mobxEsm.toJS)(_AppHaxStore.store.site.structure));
                                formData.append("upload", file);
                                _context.next = 11;
                                return _microFrontendRegistry.MicroFrontendRegistry.call("@haxcms/docxToSite", formData);

                              case 11:
                                response = _context.sent;

                                // must be a valid response and have at least SOME html to bother attempting
                                if (response.status == 200 && response.data && response.data.contents != "") {
                                  _AppHaxStore.store.items = response.data.items; // invoke a file broker for a docx file
                                  // send to the endpoint and wait
                                  // if it comes back with content, then we engineer details off of it

                                  _this2.nameTyped = response.data.filename.replace(".docx", "").replace("outline", "").replace(/\s/g, "").replace(/-/g, "").toLowerCase();
                                  setTimeout(function () {
                                    _this2.shadowRoot.querySelector("#sitename").value = _this2.nameTyped;

                                    _this2.shadowRoot.querySelector("#sitename").select();
                                  }, 800);
                                  _AppHaxStore.store.site.type = type;
                                  _AppHaxStore.store.site.theme = "clean-one";

                                  _AppHaxStore.store.appEl.playSound("click2");
                                } else {
                                  _AppHaxStore.store.appEl.playSound("error");

                                  _AppHaxStore.store.toast("File did not return valid HTML");

                                  e.preventDefault();
                                }

                              case 13:
                              case "end":
                                return _context.stop();
                            }
                          }
                        }, _callee);
                      }));

                      return function (_x2) {
                        return _ref.apply(this, arguments);
                      };
                    }());
                  }

                case 1:
                case "end":
                  return _context2.stop();
              }
            }
          }, _callee2);
        }));

        function docxImport(_x) {
          return _docxImport.apply(this, arguments);
        }

        return docxImport;
      }() // step 3

    }, {
      key: "chooseTheme",
      value: function chooseTheme(e) {
        if (!e.target.comingSoon) {
          var value = e.target.value;
          _AppHaxStore.store.site.theme = value;

          _AppHaxStore.store.appEl.playSound("click2");
        }
      } // step 4

    }, {
      key: "chooseName",
      value: function chooseName() {
        if (this.nameTyped !== "") {
          var value = this.shadowRoot.querySelector("#sitename").value;
          _AppHaxStore.store.site.name = value;

          _AppHaxStore.store.appEl.playSound("click2");
        }
      }
    }, {
      key: "progressReady",
      value: function progressReady(e) {
        var _this3 = this;

        if (e.detail) {
          this._progressReady = true;

          if (this.step === 5) {
            setTimeout(function () {
              _this3.shadowRoot.querySelector("app-hax-hat-progress").process();
            }, 300);
          }
        }
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this4 = this;

        if (babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSteps.prototype), "updated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSteps.prototype), "updated", this).call(this, changedProperties);
        }

        changedProperties.forEach(function (oldValue, propName) {
          // set input field to whats in store if we have it
          if (_this4.step === 4 && propName === "step" && _this4.shadowRoot) {
            _this4.shadowRoot.querySelector("#sitename").value = (0, _mobxEsm.toJS)(_AppHaxStore.store.site.name);
          } // progress


          if (_this4.step === 5 && propName === "step" && _this4.shadowRoot && _this4._progressReady) {
            setTimeout(function () {
              _this4.shadowRoot.querySelector("app-hax-hat-progress").process();
            }, 600);
          } // update the store for step when it changes internal to our step flow


          if (propName === "step") {
            _AppHaxStore.store.step = _this4.step;
          }
        });
      }
    }, {
      key: "connectedCallback",
      value: function connectedCallback() {
        babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSteps.prototype), "connectedCallback", this).call(this);
        window.addEventListener("resize", this.maintainScroll.bind(this));
        window.addEventListener("popstate", this.popstateListener.bind(this));
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        window.removeEventListener("resize", this.maintainScroll.bind(this));
        window.removeEventListener("popstate", this.popstateListener.bind(this));
        babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSteps.prototype), "disconnectedCallback", this).call(this);
      } // see if user navigates forward or backward while in app

    }, {
      key: "popstateListener",
      value: function popstateListener(e) {
        var _this5 = this;

        // filter out vaadin link clicks which have a state signature
        if (e.type === "popstate" && e.state === null) {
          // a lot going on here, just to be safe
          try {
            // the delay allows clicking for step to change, process, and then testing it
            setTimeout(function () {
              var link = e.target.document.location.pathname.split("/").pop(); // other links we don't care about validating state

              if (link.includes("createSite")) {
                var step = parseInt(link.replace("createSite-step-", ""));

                if (step < _AppHaxStore.store.stepTest(step)) {
                  _this5.shadowRoot.querySelector("#link-step-" + step).click();
                } else if (step > _AppHaxStore.store.stepTest(step)) {
                  _AppHaxStore.store.toast("Please select an option");

                  _this5.step = _AppHaxStore.store.stepTest(step); // forces state by maintaining where we are

                  _this5.shadowRoot.querySelector("#link-step-" + _this5.step).click();
                }
              }
            }, 0);
          } catch (e) {}
        }
      } // account for resizing

    }, {
      key: "maintainScroll",
      value: function maintainScroll() {
        var _this6 = this;

        if (this.shadowRoot && this.step) {
          this.scrollToThing("#step-".concat(this.step), {
            behavior: "instant",
            block: "start",
            inline: "nearest"
          }); // account for an animated window drag... stupid.

          setTimeout(function () {
            _this6.scrollToThing("#step-".concat(_this6.step), {
              behavior: "instant",
              block: "start",
              inline: "nearest"
            });
          }, 100);
        }
      }
    }, {
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        var _this7 = this;

        if (babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSteps.prototype), "firstUpdated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSteps.prototype), "firstUpdated", this).call(this, changedProperties);
        }

        setTimeout(function () {
          // ensure paint issues not a factor for null step
          if (_this7.step === null) {
            _this7.step = 1;
          }

          _this7.scrollToThing("#step-".concat(_this7.step), {
            behavior: "instant",
            block: "start",
            inline: "nearest"
          });
        }, 100);
        (0, _mobxEsm.autorun)(function () {
          // verify we are in the site creation process
          if ((0, _mobxEsm.toJS)(_AppHaxStore.store.createSiteSteps) && (0, _mobxEsm.toJS)(_AppHaxStore.store.appReady)) {
            var location = (0, _mobxEsm.toJS)(_AppHaxStore.store.location);

            if (location.route && location.route.step && location.route.name) {
              // account for an animated window drag... stupid.
              setTimeout(function () {
                _this7.scrollToThing("#".concat(location.route.name), {
                  behavior: "smooth",
                  block: "start",
                  inline: "nearest"
                }); /// just for step 4 since it has an input


                if (location.route.step === 4 && _AppHaxStore.store.stepTest(4) === 4) {
                  setTimeout(function () {
                    _this7.shadowRoot.querySelector("#sitename").focus();

                    _this7.scrollToThing("#step-4", {
                      behavior: "instant",
                      block: "start",
                      inline: "nearest"
                    });
                  }, 800);
                }
              }, 300); // this delay helps w/ initial paint timing but also user perception
              // there's a desire to have a delay especialy when tapping things of
              // about 300ms
            }
          }
        });
        (0, _mobxEsm.autorun)(function () {
          if (_this7.shadowRoot && (0, _mobxEsm.toJS)(_AppHaxStore.store.createSiteSteps) && (0, _mobxEsm.toJS)(_AppHaxStore.store.appReady)) {
            var activeItem = (0, _mobxEsm.toJS)(_AppHaxStore.store.activeItem);

            if (activeItem && activeItem.name && activeItem.step && !_this7.__overrideProgression) {
              _this7.shadowRoot.querySelector("#link-".concat(activeItem.name)).click();
            }
          }
        });
      }
      /**
       * Yet another reason Apple doesn't let us have nice things.
       * This detects the NONSTANDARD BS VERSION OF SCROLLINTOVIEW
       * and then ensures that it incorrectly calls to scroll into view
       * WITHOUT the wonderful params that ALL OTHER BROWSERS ACCEPT
       * AND MAKE OUR LIVES SO WONDERFUL TO SCROLL TO THINGS SMOOTHLY
       */

    }, {
      key: "scrollToThing",
      value: function scrollToThing(sel, props) {
        var isSafari = window.safari !== undefined;

        if (this.shadowRoot.querySelector(".carousel-with-snapping-item.active-step")) {
          this.shadowRoot.querySelector(".carousel-with-snapping-item.active-step").classList.remove("active-step");
        }

        if (isSafari) {
          this.shadowRoot.querySelector(sel).scrollIntoView();
        } else {
          this.shadowRoot.querySelector(sel).scrollIntoView(props);
        }

        this.shadowRoot.querySelector(sel).classList.add("active-step");
      }
    }, {
      key: "progressFinished",
      value: function progressFinished(e) {
        if (e.detail) {
          this.loaded = true;

          _AppHaxStore.store.appEl.playSound("success"); // focus the button for going to the site


          e.target.shadowRoot.querySelector(".game").focus();
          this.scrollToThing("#step-".concat(this.step), {
            behavior: "instant",
            block: "start",
            inline: "nearest"
          });
        }
      }
    }, {
      key: "typeKey",
      value: function typeKey() {
        this.nameTyped = this.shadowRoot.querySelector("#sitename").value;
      }
    }, {
      key: "keydown",
      value: function keydown(e) {
        // some trapping for common characters that make us sad
        if ([" ", "/", "\\", "&", "#", "?", "+", "=", "{", "}", "|", "^", "~", "[", "]", "`", '"', "'"].includes(e.key)) {
          _AppHaxStore.store.appEl.playSound("error");

          _AppHaxStore.store.toast("\"".concat(e.key, "\" not allowed in names"));

          e.preventDefault();
        } else if (e.key === "Enter") {
          this.chooseName();
        } else if (["ArrowUp", "ArrowRight", "ArrowDown", "ArrowLeft"].includes(e.key)) {// do nothing, directional keys for modifying word
        } else {
          _AppHaxStore.store.appEl.playSound("click");
        }
      }
    }, {
      key: "stepLinkClick",
      value: function stepLinkClick(e) {
        var clickedStep = parseInt(e.target.getAttribute("data-step"), 10);

        if (this.step < clickedStep) {
          e.preventDefault();
        } else if (e.target.getAttribute("data-step") === null) {
          _AppHaxStore.store.createSiteSteps = false;
          _AppHaxStore.store.appMode = "home";
          this.nameTyped = "";
          _AppHaxStore.store.siteReady = false;
          _AppHaxStore.store.site.structure = null;
          _AppHaxStore.store.site.type = null;
          _AppHaxStore.store.site.theme = null;
          _AppHaxStore.store.site.name = null;
        } // means user went backwards
        else if (this.step > clickedStep) {
            this.nameTyped = "";
            _AppHaxStore.store.siteReady = false;

            if (clickedStep === 1) {
              _AppHaxStore.store.site.structure = null;
              _AppHaxStore.store.site.type = null;
              _AppHaxStore.store.site.theme = null;
              _AppHaxStore.store.site.name = null;
            } else if (clickedStep === 2) {
              _AppHaxStore.store.site.type = null;
              _AppHaxStore.store.site.theme = null;
              _AppHaxStore.store.site.name = null;
            } else if (clickedStep === 3) {
              _AppHaxStore.store.site.theme = null;
              _AppHaxStore.store.site.name = null;
            } else if (clickedStep === 4) {
              _AppHaxStore.store.site.name = null;
            }

            this.step = clickedStep;
          }
      }
    }, {
      key: "renderTypes",
      value: function renderTypes(step) {
        var structure = (0, _mobxEsm.toJS)(_AppHaxStore.store.site.structure);
        var template = (0, _index.html)(_templateObject_77261b40809211edaa2833647dc78c6c || (_templateObject_77261b40809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral([""])));

        switch (structure) {
          case "portfolio":
            template = (0, _index.html)(_templateObject2_77261b40809211edaa2833647dc78c6c || (_templateObject2_77261b40809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral([" <app-hax-button\n            tabindex=\"", "\"\n            @click=", "\n          ></app-hax-button>\n          <app-hax-button\n            tabindex=\"", "\"\n            @click=", "\n            type=\"technology\"\n            coming-soon\n          ></app-hax-button>\n          <app-hax-button\n            tabindex=\"", "\"\n            @click=", "\n            type=\"business\"\n            coming-soon\n          ></app-hax-button>\n          <app-hax-button\n            tabindex=\"", "\"\n            @click=", "\n            type=\"art\"\n            coming-soon\n          ></app-hax-button>"])), step !== 2 ? "-1" : "", this.chooseType, step !== 2 ? "-1" : "", this.chooseType, step !== 2 ? "-1" : "", this.chooseType, step !== 2 ? "-1" : "", this.chooseType);
            break;

          default:
          case "course":
            template = (0, _index.html)(_templateObject3_77261b40809211edaa2833647dc78c6c || (_templateObject3_77261b40809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral([" <app-hax-button\n            tabindex=\"", "\"\n            @click=", "\n          ></app-hax-button>\n          <app-hax-button\n            tabindex=\"", "\"\n            @click=", "\n            type=\"15w\"\n          ></app-hax-button>\n          <app-hax-button\n            tabindex=\"", "\"\n            @click=", "\n            type=\"6w\"\n          ></app-hax-button>\n          <app-hax-button\n            tabindex=\"", "\"\n            @click=", "\n            type=\"training\"\n          ></app-hax-button>\n          <app-hax-button\n            tabindex=\"", "\"\n            @click=", "\n            type=\"docx import\"\n          ></app-hax-button>"])), step !== 2 ? "-1" : "", this.chooseType, step !== 2 ? "-1" : "", this.chooseType, step !== 2 ? "-1" : "", this.chooseType, step !== 2 ? "-1" : "", this.chooseType, step !== 2 ? "-1" : "", this.docxImport);
            break;

          case "website":
            template = (0, _index.html)(_templateObject4_77261b40809211edaa2833647dc78c6c || (_templateObject4_77261b40809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral([" <app-hax-button\n            tabindex=\"", "\"\n            @click=", "\n            type=\"blog\"\n          ></app-hax-button>\n          <app-hax-button\n            tabindex=\"", "\"\n            @click=", "\n            coming-soon\n          ></app-hax-button>"])), step !== 2 ? "-1" : "", this.chooseType, step !== 2 ? "-1" : "", this.chooseType);
            break;
        }

        return template;
      }
    }, {
      key: "render",
      value: function render() {
        var _this8 = this;

        return (0, _index.html)(_templateObject5_77261b40809211edaa2833647dc78c6c || (_templateObject5_77261b40809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n      <div id=\"container\">\n        <ul id=\"step-links\">\n          <li>\n            ", "\n          </li>\n          ", "\n        </ul>\n        <scrollable-component>\n          <div class=\"carousel-with-snapping-track\">\n            <div class=\"carousel-with-snapping-item\" id=\"step-1\">\n              <div class=\"step-wrapper\">\n                <app-hax-site-button\n                  tabindex=\"", "\"\n                  label=\"&gt; Course\"\n                  value=\"course\"\n                  @click=", "\n                ></app-hax-site-button>\n                <app-hax-site-button\n                  tabindex=\"", "\"\n                  label=\"&gt; Portfolio\"\n                  value=\"portfolio\"\n                  @click=", "\n                ></app-hax-site-button>\n                <app-hax-site-button\n                  tabindex=\"", "\"\n                  label=\"&gt; Website\"\n                  value=\"website\"\n                  @click=", "\n                  coming-soon\n                ></app-hax-site-button>\n              </div>\n            </div>\n            <div class=\"carousel-with-snapping-item\" id=\"step-2\">\n              <div id=\"grid-container\">", "</div>\n            </div>\n            <div class=\"carousel-with-snapping-item\" id=\"step-3\">\n              <div id=\"themeContainer\">\n                ", "\n              </div>\n            </div>\n            <div class=\"carousel-with-snapping-item\" id=\"step-4\">\n              <input\n                id=\"sitename\"\n                @input=\"", "\"\n                @keydown=\"", "\"\n                maxlength=\"30\"\n                placeholder=\"", " name..\"\n                tabindex=\"", "\"\n              />\n              <app-hax-site-button\n                class=\"sitenamebtn\"\n                tabindex=\"", "\"\n                label=\"&gt; Create journey\"\n                @click=", "\n                ?disabled=\"", "\"\n              >\n              </app-hax-site-button>\n            </div>\n            <div class=\"carousel-with-snapping-item\" id=\"step-5\">\n              <app-hax-hat-progress\n                @progress-ready=\"", "\"\n                @promise-progress-finished=\"", "\"\n                tabindex=\"", "\"\n              ></app-hax-hat-progress>\n            </div>\n          </div>\n        </scrollable-component>\n      </div>\n    "])), !(0, _mobxEsm.toJS)(_AppHaxStore.store.isNewUser) ? (0, _index.html)(_templateObject6_77261b40809211edaa2833647dc78c6c || (_templateObject6_77261b40809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n                  <a href=\"home\" class=\"homelnk\" tabindex=\"-1\">\n                    <simple-icon-lite\n                      tabindex=\"0\"\n                      src=\"", "\"\n                      id=\"homebtn\"\n                      title=\"Site list\"\n                      @click=\"", "\"\n                    ></simple-icon-lite>\n                  </a>\n                  <simple-tooltip for=\"homebtn\" position=\"bottom\"\n                    >Site list</simple-tooltip\n                  >\n                "])), homeIcon, this.stepLinkClick) : (0, _index.html)(_templateObject7_77261b40809211edaa2833647dc78c6c || (_templateObject7_77261b40809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral([""]))), this.stepRoutes.map(function (item, index) {
          return (0, _index.html)(_templateObject8_77261b40809211edaa2833647dc78c6c || (_templateObject8_77261b40809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<li\n                ?disabled=\"", "\"\n                class=\"step ", "\"\n              >\n                <a\n                  href=\"", "\"\n                  ?disabled=\"", "\"\n                  tabindex=\"", "\"\n                  @click=\"", "\"\n                  id=\"link-", "\"\n                  title=\"Step ", ": ", "\"\n                  data-step=\"", "\"\n                  >", "</a\n                >\n                <simple-tooltip for=\"link-", "\" position=\"bottom\"\n                  >Step ", ": ", "</simple-tooltip\n                >\n              </li>"])), _this8.step < item.step || _this8.step === 5 ? true : false, _this8.step === item.step ? "active-step" : "", item.path, _this8.step < item.step || _this8.step === 5 ? true : false, _this8.step <= item.step ? "-1" : "0", _this8.stepLinkClick, item.name, index + 1, item.label, item.step, index + 1, item.name, index + 1, item.label);
        }), this.step !== 1 ? "-1" : "", this.chooseStructure, this.step !== 1 ? "-1" : "", this.chooseStructure, this.step !== 1 ? "-1" : "", this.chooseStructure, this.renderTypes(this.step), this.appSettings && this.appSettings.themes ? this.themeNames.map(function (themeKey) {
          return (0, _index.html)(_templateObject9_77261b40809211edaa2833647dc78c6c || (_templateObject9_77261b40809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n                        <button\n                          aria-label=\"", " theme\"\n                          value=\"", "\"\n                          class=\"theme-button\"\n                          @click=", "\n                          tabindex=\"", "\"\n                        >\n                          <img\n                            class=\"theme-img\"\n                            src=\"", "\"\n                            alt=\"\"\n                            loading=\"lazy\"\n                            decoding=\"async\"\n                          />\n                          <div>", "</div>\n                        </button>\n                      "])), _this8.appSettings.themes[themeKey].name, themeKey, _this8.chooseTheme, _this8.step !== 3 ? "-1" : "", _this8.appSettings.themes[themeKey].thumbnail, _this8.appSettings.themes[themeKey].name);
        }) : "", this.typeKey, this.keydown, (0, _mobxEsm.toJS)(_AppHaxStore.store.site.structure), this.step !== 4 ? "-1" : "", this.step !== 4 ? "-1" : "", this.chooseName, this.nameTyped === "", this.progressReady, this.progressFinished, this.step !== 5 ? "-1" : "");
      }
    }], [{
      key: "tag",
      get: function get() {
        return "app-hax-steps";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSteps), "properties", this)), {}, {
          step: {
            type: Number,
            reflect: true
          },
          stepRoutes: {
            type: Array
          },
          themeNames: {
            type: Array
          },
          loaded: {
            type: Boolean,
            reflect: true
          },
          appSettings: {
            type: Object
          },
          nameTyped: {
            type: String
          }
        });
      }
    }, {
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(AppHaxSteps), "styles", this)), [(0, _index.css)(_templateObject10_77261b40809211edaa2833647dc78c6c || (_templateObject10_77261b40809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        scrollable-component {\n          --scrollbar-width: 0px;\n          --scrollbar-height: 0px;\n          --scrollbar-padding: 0;\n          --viewport-overflow-x: hidden;\n          overflow: hidden;\n        }\n        #grid-container {\n          display: grid;\n          grid-template-columns: 200px 200px 200px;\n          background: transparent;\n        }\n        .carousel-with-snapping-track {\n          display: grid;\n          grid-auto-flow: column;\n          grid-gap: 30px;\n        }\n        .carousel-with-snapping-item {\n          display: flex;\n          flex-direction: column;\n          align-items: center;\n          justify-content: normal;\n          scroll-snap-align: center;\n          scroll-snap-stop: always;\n          scrollbar-gutter: stable;\n          width: var(--viewport-width);\n          font-size: 1.5rem;\n          text-align: center;\n          overflow-x: hidden;\n          max-height: 60vh;\n          padding-top: 1vh;\n        }\n        #step-links {\n          padding: 0;\n          margin: 0;\n        }\n        ul,\n        li {\n          list-style: none;\n        }\n        li {\n          vertical-align: middle;\n          display: inline-flex;\n          margin: 5px;\n        }\n        li.step {\n          border-radius: 50%;\n          background-color: transparent;\n        }\n        li a {\n          font-size: 12px;\n          color: var(--simple-colors-default-theme-grey-12, white);\n          text-decoration: none;\n          padding: 5px;\n          width: 20px;\n          height: 20px;\n          line-height: 20px;\n          margin: 0;\n          display: block;\n          border: 0;\n          border-radius: 50%;\n          background-repeat: no-repeat;\n          background-size: 30px 30px;\n          background-color: var(--simple-colors-default-theme-grey-1, white);\n          background-image: url(\"", "\");\n          transition: 0.3s ease-in-out background, 0.3s ease-in-out color;\n          transition-delay: 0.6s, 0.3s;\n        }\n        li a[disabled] {\n          background-image: url(\"", "\");\n          pointer-events: none;\n          color: var(--simple-colors-default-theme-grey-7, grey);\n          user-select: none;\n        }\n        li[disabled] {\n          background-color: grey;\n        }\n        li.active-step a {\n          background-color: orange;\n          background-image: url(\"", "\");\n        }\n        app-hax-button {\n          padding: 10px 0px 10px 0px;\n          background: transparent;\n        }\n        #theme-container {\n          display: flex;\n          flex-direction: row;\n          align-items: center;\n          justify-content: center;\n        }\n        img {\n          pointer-events: none;\n        }\n        #themeContainer {\n          width: 70vw;\n          height: 55vh;\n        }\n        .theme-button {\n          background-color: transparent;\n          color: var(--simple-colors-default-theme-grey-12, white);\n          border: none;\n          margin: 8px;\n          padding: 8px;\n          width: 245px;\n        }\n\n        .theme-button div {\n          font-family: \"Press Start 2P\", sans-serif;\n          font-size: 14px;\n          margin-top: 12px;\n        }\n        .theme-button:focus,\n        .theme-button:hover {\n          outline: 4px solid var(--app-hax-accent-color, var(--accent-color));\n          outline-offset: 4px;\n          background-color: transparent;\n          border: none;\n          cursor: pointer;\n        }\n        #sitename {\n          font-family: \"Press Start 2P\", sans-serif;\n          font-size: 32px;\n          padding: 8px;\n          width: 40vw;\n        }\n        #homebtn {\n          --simple-icon-height: 30px;\n          --simple-icon-width: 30px;\n          border-radius: 50%;\n          cursor: pointer;\n          background-color: var(--simple-colors-default-theme-grey-1, white);\n        }\n        .homelnk {\n          background-image: none;\n          display: flex;\n          padding: 0;\n          margin: 0;\n          height: 30px;\n          width: 30px;\n        }\n        app-hax-site-button {\n          justify-content: center;\n          --app-hax-site-button-width: 35vw;\n          --app-hax-site-button-min-width: 240px;\n        }\n        app-hax-hat-progress {\n          height: 400px;\n          width: 400px;\n          display: block;\n        }\n\n        @media (max-width: 800px) {\n          .theme-button {\n            width: unset;\n            padding: 0;\n          }\n          .theme-button div {\n            font-size: 12px;\n            margin-top: 8px;\n          }\n          .theme-button img {\n            height: 70px;\n          }\n          app-hax-site-button {\n            width: 320px;\n            max-width: 60vw;\n            --app-hax-site-button-font-size: 2.5vw;\n          }\n          #sitename {\n            width: 70vw;\n            font-size: 20px;\n          }\n          #grid-container {\n            grid-template-columns: 150px 150px 150px;\n          }\n        }\n        @media (max-height: 600px) {\n          .carousel-with-snapping-item {\n            padding-top: 4px;\n            max-height: 57vh;\n          }\n          #sitename {\n            width: 40vw;\n            font-size: 14px;\n          }\n          app-hax-hat-progress {\n            transform: scale(0.5);\n            margin-top: -18vh;\n          }\n        }\n        @media (max-width: 500px) {\n          app-hax-hat-progress {\n            transform: scale(0.5);\n            margin-top: -15vh;\n          }\n        }\n        @media (max-height: 400px) {\n          .carousel-with-snapping-item {\n            padding-top: 4px;\n            max-height: 40vh;\n          }\n          app-hax-hat-progress {\n            transform: scale(0.3);\n          }\n          .carousel-with-snapping-item.active-step app-hax-hat-progress {\n            position: fixed;\n            top: 20%;\n            left: 20%;\n          }\n        }\n      "])), (0, _index.unsafeCSS)(enabledCircle), (0, _index.unsafeCSS)(disabledCircle), (0, _index.unsafeCSS)(transparentCircle))]);
      }
    }]);
    return AppHaxSteps;
  }(_simpleColors.SimpleColors);

  _exports.AppHaxSteps = AppHaxSteps;
  customElements.define(AppHaxSteps.tag, AppHaxSteps);
});
