// @lrnwebcomponents/haxcms-elements/lib/ui-components/layout/site-footer.js
define(["exports", "../../../../../lit/index.js", "../../core/haxcms-site-store.js", "../../../../../mobx/dist/mobx.esm.js", "../../../../license-element/license-element.js"], function (_exports, _index, _haxcmsSiteStore, _mobxEsm, _licenseElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SiteFooter = void 0;

  var _templateObject_fa7e03608e0011ed99ce491205e67b1d, _templateObject2_fa7e03608e0011ed99ce491205e67b1d, _templateObject3_fa7e03608e0011ed99ce491205e67b1d;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `site-footer`
   * `A basic site footer`
   *
   * @demo demo/index.html
   */
  var SiteFooter = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(SiteFooter, _LitElement);

    var _super = _createSuper(SiteFooter);

    function SiteFooter() {
      var _this;

      babelHelpers.classCallCheck(this, SiteFooter);
      _this = _super.call(this);
      _this.__disposer = [];
      (0, _mobxEsm.autorun)(function (reaction) {
        _this.manifest = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.manifest);

        _this.__disposer.push(reaction);
      });
      (0, _mobxEsm.autorun)(function (reaction) {
        _this.editMode = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.editMode);

        _this.__disposer.push(reaction);
      });
      (0, _mobxEsm.autorun)(function (reaction) {
        _this.siteTitle = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.siteTitle);

        _this.__disposer.push(reaction);
      });
      return _this;
    } // render function


    babelHelpers.createClass(SiteFooter, [{
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_fa7e03608e0011ed99ce491205e67b1d || (_templateObject_fa7e03608e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"wrapper\">\n        ", "\n      </div>\n    "])), this.manifest ? (0, _index.html)(_templateObject2_fa7e03608e0011ed99ce491205e67b1d || (_templateObject2_fa7e03608e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral(["\n              <license-element\n                .title=\"", "\"\n                .creator=\"", "\"\n                .source=\"", "\"\n                .license=\"", "\"\n              >\n              </license-element>\n            "])), this.siteTitle, this.manifest.author, this.manifest.domain, this.manifest.license) : "");
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        // clean up state
        for (var i in this.__disposer) {
          this.__disposer[i].dispose();
        }

        babelHelpers.get(babelHelpers.getPrototypeOf(SiteFooter.prototype), "disconnectedCallback", this).call(this);
      }
    }], [{
      key: "styles",
      get: function get() {
        return [(0, _index.css)(_templateObject3_fa7e03608e0011ed99ce491205e67b1d || (_templateObject3_fa7e03608e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          transition: 0.2s opacity linear;\n          opacity: 1;\n        }\n        :host([edit-mode]) {\n          opacity: 0.2;\n          pointer-events: none;\n        }\n      "])))];
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       */

    }, {
      key: "tag",
      get: function get() {
        return "site-footer";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          siteTitle: {
            type: String,
            attribute: "site-title"
          },
          manifest: {
            type: Object
          },
          editMode: {
            type: Boolean,
            reflect: true,
            attribute: "edit-mode"
          }
        };
      }
    }]);
    return SiteFooter;
  }(_index.LitElement);

  _exports.SiteFooter = SiteFooter;
  customElements.define(SiteFooter.tag, SiteFooter);
});
