// @lrnwebcomponents/haxcms-elements/lib/ui-components/layout/site-modal.js
define(["exports", "require", "../../../../../lit-element/lit-element.js", "../../../../simple-modal/lib/simple-modal-template.js", "../../../../simple-icon/simple-icon.js", "../../../../simple-icon/lib/simple-icons.js", "../../../../simple-icon/lib/simple-icon-button.js"], function (_exports, _require, _litElement, _simpleModalTemplate, _simpleIcon, _simpleIcons, _simpleIconButton) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SiteModal = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject2_bf28da80285c11ebbf58e33315eb0e49() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <simple-icon-button\n        ?disabled=\"", "\"\n        id=\"btn\"\n        ?dark=\"", "\"\n        accent-color=\"", "\"\n        @click=\"", "\"\n        .icon=\"", "\"\n        .title=\"", "\"\n      ></simple-icon-button>\n      <simple-tooltip for=\"btn\" .position=\"", "\" offset=\"14\">\n        ", "\n      </simple-tooltip>\n      <simple-modal-template id=\"smt\" .title=\"", "\">\n        <div id=\"content\" slot=\"content\"></div>\n      </simple-modal-template>\n    "]);

    _templateObject2_bf28da80285c11ebbf58e33315eb0e49 = function _templateObject2_bf28da80285c11ebbf58e33315eb0e49() {
      return data;
    };

    return data;
  }

  function _templateObject_bf28da80285c11ebbf58e33315eb0e49() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        simple-icon-button {\n          color: var(--site-modal-icon-color);\n        }\n      "]);

    _templateObject_bf28da80285c11ebbf58e33315eb0e49 = function _templateObject_bf28da80285c11ebbf58e33315eb0e49() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `site-modal`
   * `A basic site dialog`
   *
   * @demo demo/index.html
   */
  var SiteModal = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(SiteModal, _LitElement);

    var _super = _createSuper(SiteModal);

    babelHelpers.createClass(SiteModal, null, [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject_bf28da80285c11ebbf58e33315eb0e49())];
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       */

    }, {
      key: "tag",
      get: function get() {
        return "site-modal";
      }
    }]);

    function SiteModal() {
      var _this;

      babelHelpers.classCallCheck(this, SiteModal);
      _this = _super.call(this);
      _this.title = "Dialog";
      _this.dark = false;
      _this.accentColor = null;
      _this.icon = "icons:menu";
      _this.buttonLabel = "Open dialog";
      _this.position = "bottom";
      new Promise(function (res, rej) {
        return _require.default(["../../../../simple-tooltip/simple-tooltip.js"], res, rej);
      });
      return _this;
    } // render function


    babelHelpers.createClass(SiteModal, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject2_bf28da80285c11ebbf58e33315eb0e49(), this.editMode, this.dark, this.accentColor, this.fireEvent, this.icon, this.buttonLabel, this.position, this.buttonLabel, this.title);
      }
      /**
       * Fire an event for things to react to above us; useful for lazy loading
       */

    }, {
      key: "fireEvent",
      value: function fireEvent(e) {
        this.dispatchEvent(new CustomEvent("site-modal-click", {
          detail: {
            value: true
          }
        }));
      }
    }, {
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        var _this2 = this;

        this.shadowRoot.querySelector("#smt").associateEvents(this.shadowRoot.querySelector("#btn"));
        setTimeout(function () {
          if (_this2.children && _this2.shadowRoot.querySelector("#content")) {
            for (var i in _this2.children) {
              if (babelHelpers.typeof(_this2.children[i]) === "object") {
                _this2.shadowRoot.querySelector("#content").appendChild(_this2.children[i]);
              }
            }
          }
        }, 0);
      }
    }], [{
      key: "properties",
      get: function get() {
        return {
          disabled: {
            type: Boolean,
            reflect: true
          },
          dark: {
            type: Boolean
          },
          accentColor: {
            type: String,
            attribute: "accent-color"
          },
          title: {
            type: String
          },
          icon: {
            type: String
          },
          buttonLabel: {
            type: String,
            attribute: "button-label"
          },
          position: {
            type: String
          }
        };
      }
    }]);
    return SiteModal;
  }(_litElement.LitElement);

  _exports.SiteModal = SiteModal;
  window.customElements.define(SiteModal.tag, SiteModal);
});
