// @lrnwebcomponents/elmsln-apps/lib/lrnapp-open-studio/lrnapp-open-studio-assignments.js
define(["exports", "require", "../../../../lit-element/lit-element.js", "../../../materializecss-styles/materializecss-styles.js", "../../../../@polymer/polymer/lib/elements/custom-style.js"], function (_exports, _require, _litElement, _materializecssStyles, _customStyle) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappOpenStudioAssignments = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject6_01c11b301bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          align-content: center;\n        }\n        #loading {\n          width: 100%;\n          z-index: 1000;\n          opacity: 0.8;\n          text-align: center;\n          align-content: center;\n          justify-content: center;\n          height: 100vh;\n          position: absolute;\n          background-color: white;\n        }\n        h1.empty-title,\n        h1.assignment-title {\n          font-size: 2em;\n        }\n        .submission-title {\n          font-size: 1em;\n          font-weight: bold;\n          width: 100%;\n          margin-bottom: 2rem;\n          text-transform: none;\n        }\n        .submission-list {\n          display: flex;\n          align-items: center;\n        }\n        .submission-list-item {\n          overflow: hidden;\n          text-overflow: ellipsis;\n          word-break: break-all;\n          word-wrap: break-word;\n        }\n      "]);

    _templateObject6_01c11b301bb311eabd7cc30119585ddb = function _templateObject6_01c11b301bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject5_01c11b301bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <h1 class=\"empty-title black-text\">\n              Please select an Assignment in order to view all submissions\n            </h1>\n          "]);

    _templateObject5_01c11b301bb311eabd7cc30119585ddb = function _templateObject5_01c11b301bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject4_01c11b301bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                <lrnapp-studio-submission-display\n                  id=\"submission-", "\"\n                  .submission=\"", "\"\n                  class=\"ferpa-protect\"\n                ></lrnapp-studio-submission-display>\n                <paper-button\n                  class=\"submission-title ferpa-protect\"\n                  @click=\"", "\"\n                  data-submission-id=\"", "\"\n                  >Tap to comment on ", " by\n                  ", "\n                </paper-button>\n              "]);

    _templateObject4_01c11b301bb311eabd7cc30119585ddb = function _templateObject4_01c11b301bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject3_01c11b301bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <lrnsys-button\n                    @click=\"", "\"\n                    raised=\"\"\n                    class=\"submission-list-item\"\n                    button-class=\"submission-list-item\"\n                    hover-class=\"blue white-text\"\n                    data-submission-id=\"", "\"\n                  >\n                    <span slot=\"button\">\n                      <lrndesign-avatar\n                        src=\"", "\"\n                        label=\"", "\"\n                        style=\"display:inline-block;\"\n                        data-submission-id=\"", "\"\n                      ></lrndesign-avatar>\n                    </span>\n                    <span slot=\"label\"\n                      >", "</span\n                    >\n                  </lrnsys-button>\n                "]);

    _templateObject3_01c11b301bb311eabd7cc30119585ddb = function _templateObject3_01c11b301bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject2_01c11b301bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <h1 class=\"assignment-title black-text\">\n              ", "\n            </h1>\n            <div class=\"submission-list\">\n              ", "\n            </div>\n            ", "\n          "]);

    _templateObject2_01c11b301bb311eabd7cc30119585ddb = function _templateObject2_01c11b301bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject_01c11b301bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <custom-style>\n        <style include=\"materializecss-styles\">\n          :host {\n            display: block;\n          }\n        </style>\n      </custom-style>\n      ", "\n    "]);

    _templateObject_01c11b301bb311eabd7cc30119585ddb = function _templateObject_01c11b301bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  var LrnappOpenStudioAssignments =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnappOpenStudioAssignments, _LitElement);
    babelHelpers.createClass(LrnappOpenStudioAssignments, [{
      key: "render",
      value: function render() {
        var _this2 = this;

        return (0, _litElement.html)(_templateObject_01c11b301bb311eabd7cc30119585ddb(), this.showSubmissions ? (0, _litElement.html)(_templateObject2_01c11b301bb311eabd7cc30119585ddb(), this.activeAssignment.attributes.title, this.submissions.map(function (submission) {
          return (0, _litElement.html)(_templateObject3_01c11b301bb311eabd7cc30119585ddb(), _this2._scrollToTarget, submission.id, submission.relationships.author.data.avatar, submission.relationships.author.data.name, submission.id, submission.relationships.author.data.display_name);
        }), this.submissions.map(function (submission) {
          return (0, _litElement.html)(_templateObject4_01c11b301bb311eabd7cc30119585ddb(), submission.id, submission, _this2._loadSubmissionUrl, submission.id, submission.attributes.title, submission.relationships.author.data.display_name);
        })) : (0, _litElement.html)(_templateObject5_01c11b301bb311eabd7cc30119585ddb()));
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject6_01c11b301bb311eabd7cc30119585ddb())];
      }
    }]);

    function LrnappOpenStudioAssignments() {
      var _this;

      babelHelpers.classCallCheck(this, LrnappOpenStudioAssignments);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappOpenStudioAssignments).call(this));
      new Promise(function (res, rej) {
        return _require.default(["../../../../@polymer/paper-button/paper-button.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../../lrnsys-button/lrnsys-button.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../../lrndesign-avatar/lrndesign-avatar.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../lrnapp-studio-submission/lrnapp-studio-submission-display.js"], res, rej);
      });
      return _this;
    }

    babelHelpers.createClass(LrnappOpenStudioAssignments, [{
      key: "updated",
      value: function updated(changedProperties) {
        var _this3 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (["assignments", "activeAssignmentId"].includes(propName)) {
            _this3.activeAssignment = _this3._activeAssignmentCompute(_this3.activeAssignmentId, _this3.assignments);
          }

          if (propName == "activeAssignmentId") {
            // notify
            _this3.dispatchEvent(new CustomEvent("active-assignment-id-changed", {
              detail: {
                value: _this3[propName]
              }
            }));
          }

          if (propName == "submissions") {
            // notify
            _this3.dispatchEvent(new CustomEvent("submissions-changed", {
              detail: {
                value: _this3[propName]
              }
            }));
          }

          if (propName == "assignments") {
            // notify
            _this3.dispatchEvent(new CustomEvent("assignments-changed", {
              detail: {
                value: _this3[propName]
              }
            }));
          }

          if (propName == "activeAssignment") {
            // notify
            _this3.dispatchEvent(new CustomEvent("active-assignment-changed", {
              detail: {
                value: _this3[propName]
              }
            }));
          }

          if (propName == "activeAuthorId") {
            // notify
            _this3.dispatchEvent(new CustomEvent("active-author-id-changed", {
              detail: {
                value: _this3[propName]
              }
            }));
          }

          if (propName == "activeAssignmentId") {
            _this3.showSubmissions = _this3._showSubmissions(_this3.activeAssignmentId);
          }

          if (propName == "showSubmissions") {
            // notify
            _this3.dispatchEvent(new CustomEvent("show-submissions-changed", {
              detail: {
                value: _this3[propName]
              }
            }));
          }

          if (propName == "sourcePath") {
            // notify
            _this3.dispatchEvent(new CustomEvent("source-path-changed", {
              detail: {
                value: _this3[propName]
              }
            }));
          }

          if (propName == "basePath") {
            // notify
            _this3.dispatchEvent(new CustomEvent("base-path-changed", {
              detail: {
                value: _this3[propName]
              }
            }));
          }
        });
      }
    }, {
      key: "_loadSubmissionUrl",

      /**
       * Handle tap on paper-button above to redirect to the correct submission url.
       */
      value: function _loadSubmissionUrl(e) {
        var local = e.target; // this will have the id of the current submission

        var active = local.getAttribute("data-submission-id"); // @todo need a cleaner integration but this at least goes the right place for now

        window.location.href = this.basePath + "lrnapp-studio-submission/submissions/" + active;
      }
    }, {
      key: "_activeAssignmentCompute",
      value: function _activeAssignmentCompute(activeAssignmentId, assignments) {
        var activeAssignment = null;

        if (assignments) {
          activeAssignment = assignments.find(function (assignment) {
            return assignment.id == activeAssignmentId;
          });
        }

        return activeAssignment;
      }
    }, {
      key: "_showSubmissions",
      value: function _showSubmissions(activeAssignmentId) {
        if (activeAssignmentId) {
          return true;
        }

        return false;
      }
    }, {
      key: "_scrollToTarget",
      value: function _scrollToTarget(e) {
        var local = e.target; // this will have the id of the current submission

        var active = local.getAttribute("data-submission-id");
        this.shadowRoot.querySelector("#submission-" + active).scrollIntoView({
          block: "start",
          behavior: "smooth"
        });
      }
      /**
       * Handle response for the whole assignments object.
       */

    }, {
      key: "_handleResponse",
      value: function _handleResponse(event) {
        this.shadowRoot.querySelector("#loading").hidden = true;
      }
      /**
       * Simple way to convert from object to array.
       */

    }, {
      key: "_toArray",
      value: function _toArray(obj) {
        if (obj == null) {
          return [];
        }

        return Object.keys(obj).map(function (key) {
          return obj[key];
        });
      }
    }], [{
      key: "tag",
      get: function get() {
        return "lrnapp-open-studio-assignments";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          elmslnCourse: {
            type: String,
            attribute: "elmsln-course"
          },
          elmslnSection: {
            type: String,
            attribute: "elmsln-section"
          },
          basePath: {
            type: String,
            attribute: "base-path"
          },
          csrfToken: {
            type: String,
            attribute: "csrf-token"
          },
          endPoint: {
            type: String,
            attribute: "end-point"
          },

          /**
           * The assignments that exist so we can make other calls for data
           */
          assignments: {
            type: Array
          },

          /**
           * The submissions that exist so we can make other calls for data
           */
          submissions: {
            type: Array
          },
          activeAssignmentId: {
            type: String,
            reflect: true,
            attribute: "active-assignment-id"
          },
          activeAssignment: {
            type: Object
          },
          activeAuthorId: {
            type: String,
            reflect: true,
            attribute: "active-author-id"
          },
          showSubmissions: {
            type: Boolean,
            attribute: "show-submissions"
          },

          /**
           * Endpoint for submission data.
           */
          sourcePath: {
            type: String,
            attribute: "source-path"
          }
        };
      }
    }]);
    return LrnappOpenStudioAssignments;
  }(_litElement.LitElement);

  _exports.LrnappOpenStudioAssignments = LrnappOpenStudioAssignments;
  window.customElements.define(LrnappOpenStudioAssignments.tag, LrnappOpenStudioAssignments);
});
