// @lrnwebcomponents/haxcms-elements/lib/ui-components/layout/site-footer.js
define(["exports", "require", "../../../../../lit-element/lit-element.js", "../../core/haxcms-site-store.js", "../../../../../mobx/lib/mobx.module.js"], function (_exports, _require, _litElement, _haxcmsSiteStore, _mobxModule) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SiteFooter = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  /**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `site-footer`
   * `A basic site footer`
   *
  
   * @polymer
   * @demo demo/index.html
   */
  class SiteFooter extends _litElement.LitElement {
    static get styles() {
      return [(0, _litElement.css)`
        :host {
          display: block;
          transition: 0.2s opacity linear;
          opacity: 1;
        }
        :host([edit-mode]) {
          opacity: 0.2;
          pointer-events: none;
        }
      `];
    }
    /**
     * Store the tag name to make it easier to obtain directly.
     */


    static get tag() {
      return "site-footer";
    }

    constructor() {
      super();
      this.__disposer = [];
      new Promise((res, rej) => _require.default(["../../../../license-element/license-element.js"], res, rej));
    } // render function


    render() {
      return (0, _litElement.html)`
      <div class="wrapper">
        <license-element
          .title="${this.siteTitle}"
          .creator="${this.manifest.author}"
          .source="${this.manifest.domain}"
          .license="${this.manifest.license}"
        >
        </license-element>
      </div>
    `;
    }

    static get properties() {
      return {
        siteTitle: {
          type: String,
          attribute: "site-title"
        },
        manifest: {
          type: Object
        },
        editMode: {
          type: Boolean,
          reflect: true,
          attribute: "edit-mode"
        }
      };
    }

    connectedCallback() {
      super.connectedCallback();
      (0, _mobxModule.autorun)(reaction => {
        this.manifest = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.manifest);

        this.__disposer.push(reaction);
      });
      (0, _mobxModule.autorun)(reaction => {
        this.editMode = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.editMode);

        this.__disposer.push(reaction);
      });
      (0, _mobxModule.autorun)(reaction => {
        this.siteTitle = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.siteTitle);

        this.__disposer.push(reaction);
      });
    }

    disconnectedCallback() {
      // clean up state
      for (var i in this.__disposer) {
        this.__disposer[i].dispose();
      }

      super.disconnectedCallback();
    }

  }

  _exports.SiteFooter = SiteFooter;
  window.customElements.define(SiteFooter.tag, SiteFooter);
});
