// @lrnwebcomponents/full-screen-image/full-screen-image.js
define(["exports", "require", "../../lit-element/lit-element.js"], function (_exports, _require, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.FullScreenImage = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject2_2144eb60c5ec11eabd030b6978e72c08() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          font-variant-ligatures: common-ligatures;\n        }\n        :host([hidden]) {\n          display: none;\n        }\n        h2 {\n          font-size: 4em;\n          font-weight: 200;\n          margin: 0px !important;\n          line-height: 1em;\n          text-transform: lowercase;\n          color: white;\n          letter-spacing: -0.04em;\n          font-family: \"Source Sans Pro\", Arial, Helvetica, sans-serif;\n        }\n        p {\n          font-weight: 400;\n          color: white;\n          line-height: 1em;\n          font-family: \"Source Sans Pro\", Arial, Helvetica, sans-serif;\n          font-size: 1em;\n          letter-spacing: 0.01em;\n          margin-top: 0;\n          margin-bottom: 1em;\n        }\n        img {\n          position: absolute;\n          margin: 0px;\n          padding: 0px;\n          border: none;\n          width: 1239px;\n          height: 774.375px;\n          max-width: none;\n          z-index: -999999;\n          left: 0px;\n          top: -57.1875px;\n        }\n        paper-icon-button {\n          width: 75px;\n          height: 45px;\n          line-height: 45px;\n          bottom: 1px;\n          right: 0px;\n          display: block;\n          position: absolute;\n          text-align: center;\n          color: white;\n          background-color: rgba(0, 0, 0, 0.6);\n        }\n        .fullpage-container {\n          display: block;\n        }\n        .image-wrapper {\n          left: 0px;\n          top: 0px;\n          bottom: 0px;\n          right: 0px;\n          overflow: hidden;\n          margin: 0px;\n          padding: 0px;\n          width: 100%;\n          z-index: -999998;\n          position: absolute;\n        }\n        .overlay-container {\n          display: block;\n          position: absolute;\n          text-align: right;\n          padding: 0.1em 1.5em 0.1em 6em;\n          bottom: 60px;\n          right: 0px;\n          background-color: rgba(191, 147, 45, 0.7);\n        }\n      "]);

    _templateObject2_2144eb60c5ec11eabd030b6978e72c08 = function _templateObject2_2144eb60c5ec11eabd030b6978e72c08() {
      return data;
    };

    return data;
  }

  function _templateObject_2144eb60c5ec11eabd030b6978e72c08() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"fullpage-container\">\n        <div class=\"overlay-container\">\n          <h2>", "</h2>\n          <p>", "</p>\n        </div>\n        <paper-icon-button\n          id=\"down\"\n          icon=\"hardware:keyboard-arrow-down\"\n        ></paper-icon-button>\n        <div class=\"image-wrapper\">\n          <img loading=\"lazy\" id=\"img\" src=\"", "\" />\n        </div>\n      </div>\n    "]);

    _templateObject_2144eb60c5ec11eabd030b6978e72c08 = function _templateObject_2144eb60c5ec11eabd030b6978e72c08() {
      return data;
    };

    return data;
  }

  /**
   * `full-screen-image`
   * @element full-screen-image
   * `full screen banner image with down arrow`
   *
   * @microcopy - language worth noting:
   *  -
   *
  
   * @polymer
   * @demo demo/index.html
   */
  var FullScreenImage =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(FullScreenImage, _LitElement);
    babelHelpers.createClass(FullScreenImage, [{
      key: "render",
      // render function
      value: function render() {
        return (0, _litElement.html)(_templateObject_2144eb60c5ec11eabd030b6978e72c08(), this.title, this.subtitle, this.source);
      } // haxProperty definition

    }], [{
      key: "styles",
      //styles function
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_2144eb60c5ec11eabd030b6978e72c08())];
      }
    }, {
      key: "haxProperties",
      get: function get() {
        return {
          canScale: true,
          canPosition: true,
          canEditSource: false,
          gizmo: {
            title: "Full screen-image",
            description: "full screen banner image with down arrow",
            icon: "icons:android",
            color: "green",
            groups: ["Screen"],
            handles: [{
              type: "todo:read-the-docs-for-usage"
            }],
            meta: {
              author: "ELMS:LN",
              owner: "The Pennsylvania State University"
            }
          },
          settings: {
            quick: [{
              property: "source",
              description: "",
              inputMethod: "textfield",
              required: true,
              icon: "icons:link",
              validationType: "url"
            }],
            configure: [{
              property: "title",
              description: "",
              inputMethod: "textfield",
              required: false,
              icon: "icons:android"
            }, {
              property: "subtitle",
              description: "",
              inputMethod: "textfield",
              required: false,
              icon: "icons:android"
            }, {
              property: "source",
              description: "",
              inputMethod: "textfield",
              required: true,
              icon: "icons:link",
              validationType: "url"
            }],
            advanced: []
          }
        };
      } // properties available to the custom element for data binding

    }, {
      key: "properties",
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(FullScreenImage), "properties", this), {
          title: {
            type: String
          },
          subtitle: {
            type: String
          },
          source: {
            type: String
          }
        });
      }
      /**
       * convention
       */

    }, {
      key: "tag",
      get: function get() {
        return "full-screen-image";
      }
      /**
       * HTMLElement
       */

    }]);

    function FullScreenImage() {
      var _this;

      babelHelpers.classCallCheck(this, FullScreenImage);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(FullScreenImage).call(this));
      _this.title = "";
      _this.subtitle = "";
      _this.source = null;
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/paper-icon-button/paper-icon-button.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/iron-icons/hardware-icons.js"], res, rej);
      });
      return _this;
    }
    /**
     * LitElement ready
     */


    babelHelpers.createClass(FullScreenImage, [{
      key: "firstUpdated",
      value: function firstUpdated() {
        var _this2 = this;

        this.shadowRoot.querySelector("#down").addEventListener("click", function (e) {
          _this2.nextElementSibling.scrollIntoView({
            block: "start",
            inline: "nearest",
            behavior: "smooth"
          });
        });
      }
    }]);
    return FullScreenImage;
  }(_litElement.LitElement);

  _exports.FullScreenImage = FullScreenImage;
  window.customElements.define(FullScreenImage.tag, FullScreenImage);
});
