// @lrnwebcomponents/haxcms-elements/lib/core/utils/HAXCMSRememberRoute.js
define(["exports", "../haxcms-site-store.js", "./HAXCMSI18NMixin.js", "../../../../../mobx/dist/mobx.esm.js", "../../../../utils/utils.js"], function (_exports, _haxcmsSiteStore, _HAXCMSI18NMixin2, _mobxEsm, _utils) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HAXCMSRememberRoute = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var HAXCMSRememberRoute = function HAXCMSRememberRoute(SuperClass) {
    return /*#__PURE__*/function (_HAXCMSI18NMixin) {
      babelHelpers.inherits(_class, _HAXCMSI18NMixin);

      var _super = _createSuper(_class);

      function _class() {
        var _this;

        babelHelpers.classCallCheck(this, _class);
        _this = _super.call(this);
        var resumeMessage = "Resume where you left off last session?";
        _this.__evaluateRoute = false;
        _this.t = _this.t || {};
        _this.t = _objectSpread(_objectSpread({}, _this.t), {}, {
          resumeMessage: resumeMessage,
          resume: "Resume"
        });
        (0, _mobxEsm.autorun)(function (reaction) {
          var activePathName = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.location.pathname);

          if (activePathName && _this.__evaluateRoute) {
            (0, _utils.localStorageSet)("HAXCMSlastRoute-".concat(_haxcmsSiteStore.store.manifest.metadata.site.name), activePathName);
          }
        });
        return _this;
      }

      babelHelpers.createClass(_class, [{
        key: "firstUpdated",
        value: function firstUpdated(changedProperties) {
          var _this2 = this;

          if (babelHelpers.get(babelHelpers.getPrototypeOf(_class.prototype), "firstUpdated", this)) {
            babelHelpers.get(babelHelpers.getPrototypeOf(_class.prototype), "firstUpdated", this).call(this, changedProperties);
          }

          setTimeout(function () {
            if (_haxcmsSiteStore.store && _haxcmsSiteStore.store.location && _haxcmsSiteStore.store.location.pathname && (0, _utils.localStorageGet)("HAXCMSlastRoute-".concat(_haxcmsSiteStore.store.manifest.metadata.site.name)) && (0, _utils.localStorageGet)("HAXCMSlastRoute-".concat(_haxcmsSiteStore.store.manifest.metadata.site.name)) != (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.location.pathname)) {
              var slot = document.createElement("a");
              slot.setAttribute("href", (0, _utils.localStorageGet)("HAXCMSlastRoute-".concat(_haxcmsSiteStore.store.manifest.metadata.site.name)));
              slot.style.height = "40px";
              slot.style.display = "block";
              slot.addEventListener("click", _this2.resumeLastRoute.bind(_this2));
              slot.innerHTML = "<button style=\"padding:4px;font-weight:bold;background-color: black; color: white; border: 4px solid black; border-radius:none;margin-left:4px;cursor: pointer;\">".concat(_this2.t.resume, "</button>");
              var urlParams = new URLSearchParams(window.location.search);
              var format = urlParams.get("format"); // ensure we don't show this if we have an alternate format request

              if (!format) {
                _haxcmsSiteStore.store.toast(_this2.t.resumeMessage, 10000, {}, "capsule", null, null, slot);
              }
            }

            _this2.__evaluateRoute = true;
          }, 1000);
        }
        /**
         * Respond to confirmation of wanting to resume the previous route
         */

      }, {
        key: "resumeLastRoute",
        value: function resumeLastRoute(e) {
          window.dispatchEvent(new CustomEvent("haxcms-toast-hide", {
            bubbles: true,
            composed: true,
            cancelable: false,
            detail: false
          }));
        }
      }]);
      return _class;
    }((0, _HAXCMSI18NMixin2.HAXCMSI18NMixin)(SuperClass));
  };

  _exports.HAXCMSRememberRoute = HAXCMSRememberRoute;
});
