// @lrnwebcomponents/haxcms-elements/lib/core/utils/HAXCMSRememberRoute.js
define(["exports", "require", "../haxcms-site-store.js", "../../../../../mobx/dist/mobx.esm.js"], function (_exports, _require, _haxcmsSiteStore, _mobxEsm) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HAXCMSRememberRoute = void 0;
  _require = _interopRequireWildcard(_require);

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var HAXCMSRememberRoute = function HAXCMSRememberRoute(SuperClass) {
    return /*#__PURE__*/function (_SuperClass) {
      babelHelpers.inherits(_class, _SuperClass);

      var _super = _createSuper(_class);

      function _class() {
        var _this;

        babelHelpers.classCallCheck(this, _class);
        _this = _super.call(this);
        var resumeMessage = "Resume where you left off last session?";
        _this.menuOpen = true;
        _this.__evaluateRoute = false;

        if (_this.t) {
          _this.t.resumeMessage = resumeMessage;
          _this.t.resume = "Resume";
        } else {
          _this.t = {
            resumeMessage: resumeMessage,
            resume: "Resume"
          };
        }

        (0, _mobxEsm.autorun)(function (reaction) {
          var activePathName = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.location.pathname);

          if (activePathName && _this.__evaluateRoute) {
            window.localStorage.setItem("HAXCMSlastRoute-".concat(_haxcmsSiteStore.store.manifest.metadata.site.name), activePathName);
          }
        });
        return _this;
      }

      babelHelpers.createClass(_class, [{
        key: "firstUpdated",
        value: function firstUpdated(changedProperties) {
          var _this2 = this;

          if (babelHelpers.get(babelHelpers.getPrototypeOf(_class.prototype), "firstUpdated", this)) {
            babelHelpers.get(babelHelpers.getPrototypeOf(_class.prototype), "firstUpdated", this).call(this, changedProperties);
          }

          setTimeout(function () {
            if (_haxcmsSiteStore.store && _haxcmsSiteStore.store.location && _haxcmsSiteStore.store.location.pathname && window.localStorage.getItem("HAXCMSlastRoute-".concat(_haxcmsSiteStore.store.manifest.metadata.site.name)) && window.localStorage.getItem("HAXCMSlastRoute-".concat(_haxcmsSiteStore.store.manifest.metadata.site.name)) != (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.location.pathname)) {
              new Promise(function (res, rej) {
                return _require.default(["../../../../simple-toast/simple-toast.js"], res, rej);
              }).then(function () {
                window.SimpleToast.requestAvailability();
                var slot = document.createElement("a");
                slot.setAttribute("href", window.localStorage.getItem("HAXCMSlastRoute-".concat(_haxcmsSiteStore.store.manifest.metadata.site.name)));
                slot.addEventListener("click", _this2.resumeLastRoute.bind(_this2));
                slot.innerHTML = "<button>".concat(_this2.t.resume, "</button>");
                var urlParams = new URLSearchParams(window.location.search);
                var format = urlParams.get("format"); // ensure we don't show this if we have an alternate format request

                if (!format) {
                  _haxcmsSiteStore.store.toast(_this2.t.resumeMessage, 8000, "capsule", null, null, slot);
                }
              });
            }

            _this2.__evaluateRoute = true;
          }, 1000);
        }
        /**
         * Respond to confirmation of wanting to resume the previous route
         */

      }, {
        key: "resumeLastRoute",
        value: function resumeLastRoute(e) {
          window.dispatchEvent(new CustomEvent("simple-toast-hide", {
            bubbles: true,
            composed: true,
            cancelable: false,
            detail: false
          }));
        }
      }]);
      return _class;
    }(SuperClass);
  };

  _exports.HAXCMSRememberRoute = HAXCMSRememberRoute;
});
