// @lrnwebcomponents/haxcms-elements/lib/core/utils/PDFPageMixin.js
define(["exports", "../../../../../lit/index.js", "../haxcms-site-store.js", "../../../../hax-iconset/lib/simple-hax-iconset.js", "../../../../utils/utils.js", "../../../../micro-frontend-registry/micro-frontend-registry.js", "./HAXCMSI18NMixin.js", "../../../../../mobx/dist/mobx.esm.js", "../../../../micro-frontend-registry/lib/microServices.js"], function (_exports, _index, _haxcmsSiteStore, _simpleHaxIconset, _utils, _microFrontendRegistry, _HAXCMSI18NMixin2, _mobxEsm, _microServices) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.PDFPageMixin = void 0;

  var _templateObject_bc1c4ae01a7511eda869392c4cc68f12, _templateObject2_bc1c4ae01a7511eda869392c4cc68f12;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var PDFPageMixin = function PDFPageMixin(SuperClass) {
    return /*#__PURE__*/function (_HAXCMSI18NMixin) {
      babelHelpers.inherits(_class, _HAXCMSI18NMixin);

      var _super = _createSuper(_class);

      function _class() {
        var _this;

        babelHelpers.classCallCheck(this, _class);
        _this = _super.call(this);
        (0, _microServices.enableServices)(['core']);
        _this.t = _this.t || {};
        _this.t = _objectSpread(_objectSpread({}, _this.t), {}, {
          downloadContentPdf: "Download Content PDF"
        });
        return _this;
      }

      babelHelpers.createClass(_class, [{
        key: "PDFPageButton",
        value: function PDFPageButton() {
          return (0, _index.html)(_templateObject_bc1c4ae01a7511eda869392c4cc68f12 || (_templateObject_bc1c4ae01a7511eda869392c4cc68f12 = babelHelpers.taggedTemplateLiteral(["\n      ", "\n      "])), _microFrontendRegistry.MicroFrontendRegistry.has('@core/htmlToPdf') ? (0, _index.html)(_templateObject2_bc1c4ae01a7511eda869392c4cc68f12 || (_templateObject2_bc1c4ae01a7511eda869392c4cc68f12 = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"pdf-page-btn\" part=\"", "\">\n        <simple-icon-button-lite\n          part=\"pdf-page-btn\"\n          label=\"", "\"\n          icon=\"lrn:pdf\"\n          @click=\"", "\"\n          show-text-label\n          icon-position=\"top\"\n        >\n        </simple-icon-button-lite>\n      </div>"])), this.editMode ? "edit-mode-active" : "", this.t.downloadContentPdf, this.downloadPDFviaMicro) : "");
        }
        /**
         * Download PDF, via microservice
         */

      }, {
        key: "downloadPDFviaMicro",
        value: function () {
          var _downloadPDFviaMicro = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(e) {
            var htmlContent, base, response, link;
            return regeneratorRuntime.wrap(function _callee$(_context) {
              while (1) {
                switch (_context.prev = _context.next) {
                  case 0:
                    htmlContent = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.activeItemContent); // base helps w/ calculating URLs in content

                    base = '';

                    if (document.querySelector('base')) {
                      base = document.querySelector('base').href;
                    }

                    _context.next = 5;
                    return _microFrontendRegistry.MicroFrontendRegistry.call('@core/htmlToPdf', {
                      base: base,
                      html: htmlContent
                    });

                  case 5:
                    response = _context.sent;

                    if (response.status == 200 && response.data) {
                      link = document.createElement("a"); // click link to download file
                      // @todo this downloads but claims to be corrupt.

                      link.href = window.URL.createObjectURL((0, _utils.b64toBlob)(response.data, "application/pdf"));
                      link.download = "".concat((0, _mobxEsm.toJS)(_haxcmsSiteStore.store.activeTitle), ".pdf");
                      link.target = "_blank";
                      this.appendChild(link);
                      link.click();
                      this.removeChild(link);
                    }

                  case 7:
                  case "end":
                    return _context.stop();
                }
              }
            }, _callee, this);
          }));

          function downloadPDFviaMicro(_x) {
            return _downloadPDFviaMicro.apply(this, arguments);
          }

          return downloadPDFviaMicro;
        }()
      }]);
      return _class;
    }((0, _HAXCMSI18NMixin2.HAXCMSI18NMixin)(SuperClass));
  };

  _exports.PDFPageMixin = PDFPageMixin;
});
