// @lrnwebcomponents/haxcms-elements/lib/core/utils/PrintBranchMixin.js
define(["exports", "../../../../../lit/index.js", "../haxcms-site-store.js", "../../../../hax-iconset/lib/simple-hax-iconset.js", "../../../../utils/utils.js", "../../../../micro-frontend-registry/micro-frontend-registry.js", "./HAXCMSI18NMixin.js", "../../../../../mobx/dist/mobx.esm.js", "../../../../micro-frontend-registry/lib/microServices.js"], function (_exports, _index, _haxcmsSiteStore, _simpleHaxIconset, _utils, _microFrontendRegistry, _HAXCMSI18NMixin2, _mobxEsm, _microServices) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.PrintBranchMixin = void 0;

  var _templateObject_bc2090a01a7511eda869392c4cc68f12, _templateObject2_bc2090a01a7511eda869392c4cc68f12;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var PrintBranchMixin = function PrintBranchMixin(SuperClass) {
    return /*#__PURE__*/function (_HAXCMSI18NMixin) {
      babelHelpers.inherits(_class, _HAXCMSI18NMixin);

      var _super = _createSuper(_class);

      function _class() {
        var _this;

        babelHelpers.classCallCheck(this, _class);
        _this = _super.call(this);
        (0, _microServices.enableServices)(['haxcms']);
        _this.t = _this.t || {};
        _this.t = _objectSpread(_objectSpread({}, _this.t), {}, {
          printBranch: "Print Branch"
        });
        return _this;
      }

      babelHelpers.createClass(_class, [{
        key: "PrintBranchButton",
        value: function PrintBranchButton() {
          return (0, _index.html)(_templateObject_bc2090a01a7511eda869392c4cc68f12 || (_templateObject_bc2090a01a7511eda869392c4cc68f12 = babelHelpers.taggedTemplateLiteral(["\n      ", "\n      "])), _microFrontendRegistry.MicroFrontendRegistry.has('@haxcms/siteToHtml') ? (0, _index.html)(_templateObject2_bc2090a01a7511eda869392c4cc68f12 || (_templateObject2_bc2090a01a7511eda869392c4cc68f12 = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"print-branch-btn\" part=\"", "\">\n        <simple-icon-button-lite\n          part=\"print-branch-btn\"\n          label=\"", "\"\n          icon=\"print\"\n          @click=\"", "\"\n          show-text-label\n          icon-position=\"top\"\n        >\n        </simple-icon-button-lite>\n      </div>"])), this.editMode ? "edit-mode-active" : "", this.t.printBranch, this.printBranchOfSite) : "");
        }
        /**
         * Download PDF, via microservice
         */

      }, {
        key: "printBranchOfSite",
        value: function () {
          var _printBranchOfSite = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(e) {
            var ancestorItem, base, site, params, response, link;
            return regeneratorRuntime.wrap(function _callee$(_context) {
              while (1) {
                switch (_context.prev = _context.next) {
                  case 0:
                    ancestorItem = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.ancestorItem); // base helps w/ calculating URLs in content

                    base = '';

                    if (document.querySelector('base')) {
                      base = document.querySelector('base').href;
                    }

                    site = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.manifest);
                    params = {
                      type: 'site',
                      site: {
                        file: base,
                        id: site.id,
                        title: site.title,
                        author: site.author,
                        description: site.description,
                        license: site.license,
                        metadata: site.metadata,
                        items: site.items
                      },
                      ancestor: ancestorItem ? ancestorItem.id : (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.activeId),
                      link: base,
                      magic: window.__appCDN,
                      base: base
                    };
                    _context.next = 7;
                    return _microFrontendRegistry.MicroFrontendRegistry.call('@haxcms/siteToHtml', params);

                  case 7:
                    response = _context.sent;

                    if (response.status == 200 && response.data) {
                      link = document.createElement("a"); // click link to download file
                      // @todo this downloads but claims to be corrupt.

                      link.href = window.URL.createObjectURL((0, _utils.b64toBlob)(btoa(response.data), "text/html"));

                      if (ancestorItem) {
                        link.download = "".concat((0, _mobxEsm.toJS)(_haxcmsSiteStore.store.ancestorTitle), ".html");
                      } else {
                        link.download = "".concat((0, _mobxEsm.toJS)(_haxcmsSiteStore.store.activeTitle), ".html");
                      }

                      link.target = "_blank";
                      this.appendChild(link);
                      link.click();
                      this.removeChild(link);
                    } else {
                      // fallback in case the service fails
                      window.open(window.location.href + "?format=print-page", "", "left=0,top=0,width=800,height=800,toolbar=0,scrollbars=0,status=0,noopener=1,noreferrer=1");
                    }

                  case 9:
                  case "end":
                    return _context.stop();
                }
              }
            }, _callee, this);
          }));

          function printBranchOfSite(_x) {
            return _printBranchOfSite.apply(this, arguments);
          }

          return printBranchOfSite;
        }()
      }]);
      return _class;
    }((0, _HAXCMSI18NMixin2.HAXCMSI18NMixin)(SuperClass));
  };

  _exports.PrintBranchMixin = PrintBranchMixin;
});
