diff --git a/includes/common.inc b/includes/common.inc
index d370bb8f2..8453d0e2a 100644
--- a/includes/common.inc
+++ b/includes/common.inc
@@ -8118,7 +8118,7 @@ function entity_get_controller($entity_type) {
   $controllers = &drupal_static(__FUNCTION__, array());
   if (!isset($controllers[$entity_type])) {
     $type_info = entity_get_info($entity_type);
-    $class = $type_info['controller class'];
+    $class = isset($type_info['controller class'])?$type_info['controller class']:'DrupalDefaultEntityController';
     $controllers[$entity_type] = new $class($entity_type);
   }
   return $controllers[$entity_type];
