diff --git a/core/dslmcode/cores/drupal-7/includes/common.inc b/core/dslmcode/cores/drupal-7/includes/common.inc
index c97704c..4bafbfb 100644
--- a/core/dslmcode/cores/drupal-7/includes/common.inc
+++ b/core/dslmcode/cores/drupal-7/includes/common.inc
@@ -2657,10 +2657,10 @@ function drupal_deliver_html_page($page_callback_result) {
   // domain, but provide a variable to override this. If the code running for
   // this page request already set the X-Frame-Options header earlier, don't
   // overwrite it here.
-  $frame_options = variable_get('x_frame_options', 'SAMEORIGIN');
-  if ($frame_options && is_null(drupal_get_http_header('X-Frame-Options'))) {
-    drupal_add_http_header('X-Frame-Options', $frame_options);
-  }
+  //$frame_options = variable_get('x_frame_options', 'SAMEORIGIN');
+  //if ($frame_options && is_null(drupal_get_http_header('X-Frame-Options'))) {
+   // drupal_add_http_header('X-Frame-Options', $frame_options);
+  //}
 
   // Menu status constants are integers; page content is a string or array.
   if (is_int($page_callback_result)) {
