diff --git a/includes/module.inc b/includes/module.inc
index 494924f..4ef8f80 100644
--- a/includes/module.inc
+++ b/includes/module.inc
@@ -320,17 +320,24 @@ function module_load_install($module) {
  *   The name of the included file, if successful; FALSE otherwise.
  */
 function module_load_include($type, $module, $name = NULL) {
+  static $file_name = array();
   if (!isset($name)) {
     $name = $module;
   }
+  $key = $type . ':' . $module . ':' . $name;
+  if (!empty($file_name[$key])) {
+    return $file_name[$key];
+  }
 
   if (function_exists('drupal_get_path')) {
     $file = DRUPAL_ROOT . '/' . drupal_get_path('module', $module) . "/$name.$type";
     if (is_file($file)) {
       require_once $file;
+      $file_name[$key] = $file;
       return $file;
     }
   }
+  $file_name[$key] = FALSE;
   return FALSE;
 }
 
