diff --git a/includes/file.inc b/includes/file.inc
index 7fd6c71..0dbbd0d 100644
--- a/includes/file.inc
+++ b/includes/file.inc
@@ -277,7 +277,9 @@ function file_default_scheme() {
  *   The normalized URI.
  */
 function file_stream_wrapper_uri_normalize($uri) {
-  $scheme = file_uri_scheme($uri);
+  // Inline file_uri_scheme()
+  $position = strpos($uri, '://');
+  $scheme = $position ? substr($uri, 0, $position) : FALSE;
 
   if ($scheme && file_stream_wrapper_valid_scheme($scheme)) {
     $target = file_uri_target($uri);
