diff --git a/authcache.php b/authcache.php
index 8b83199..974fc0f 100644
--- a/authcache.php
+++ b/authcache.php
@@ -15,7 +15,8 @@
  * Root directory of Drupal installation.
  */
 define('DRUPAL_ROOT', getcwd());
-
+// support ELMSLN config loading globally
+require_once DRUPAL_ROOT . '/../../elmsln_environment/elmsln_environment.php';
 require_once DRUPAL_ROOT . '/includes/bootstrap.inc';
 drupal_bootstrap(DRUPAL_BOOTSTRAP_FULL);
 menu_execute_active_handler();
