diff --git a/modules/node/node.module b/modules/node/node.module
index 1d88834..bdf2755 100644
--- a/modules/node/node.module
+++ b/modules/node/node.module
@@ -1009,9 +1009,9 @@ function node_object_prepare($node) {
  * Performs validation checks on the given node.
  */
 function node_validate($node, $form, &$form_state) {
-  if (isset($node->nid) && (node_last_changed($node->nid) > $node->changed)) {
-    form_set_error('changed', t('The content on this page has either been modified by another user, or you have already submitted modifications using this form. As a result, your changes cannot be saved.'));
-  }
+  //if (isset($node->nid) && (node_last_changed($node->nid) > $node->changed)) {
+  //  form_set_error('changed', t('The content on this page has either been modified by another user, or you have already submitted modifications using this form. As a result, your changes cannot be saved.'));
+  //}
 
   // Validate the "authored by" field.
   if (!empty($node->name) && !($account = user_load_by_name($node->name))) {
