// @haxtheweb/haxcms-elements/lib/ui-components/site/site-random-content.js
/**
 * Copyright 2025 The Pennsylvania State University
 * @license Apache-2.0, see License.md for full text.
 */
import{html as e,css as t,nothing as n}from"../../../../../lit/index.js";import{unsafeHTML as i}from"../../../../../lit/directives/unsafe-html.js";import{store as a}from"../../core/haxcms-site-store.js";import{HAXCMSI18NMixin as o}from"../../core/utils/HAXCMSI18NMixin.js";import{autorun as d,toJS as s}from"../../../../../mobx/dist/mobx.esm.js";import{DDD as l}from"../../../../d-d-d/d-d-d.js";import"../../../../simple-icon/lib/simple-icon-button-lite.js";export class SiteRandomContent extends(o(l)){static get tag(){return"site-random-content"}constructor(){super(),this.windowControllers=new AbortController,this.HAXCMSI18NMixinBase="../../../",this.page=null,this.randomElement=null,this.allElements=[],this.currentElementIndex=0,this.loading=!1,this.editMode=!1,this.t.noContent="No content available to display",this.t.noPage="No page selected",this.t.shuffle="Shuffle content",this.t.loading="Loading content...",this.__disposer=this.__disposer?this.__disposer:[],d((e=>{a.manifest&&this.page&&this.loadPageElements(),this.__disposer.push(e)})),d((e=>{this.editMode=s(a.editMode),this.__disposer.push(e)}))}static get styles(){return[super.styles,t`
        :host {
          display: block;
          margin: var(--ddd-spacing-4) 0;
        }

        .random-content-container {
          border: var(--ddd-border-xs) var(--ddd-theme-default-limestoneGray);
          border-radius: var(--ddd-radius-sm);
          padding: var(--ddd-spacing-4);
          background: var(--ddd-theme-default-white);
          position: relative;
        }

        /* Edit mode overlay */
        :host([edit-mode]) .random-content-container::after {
          content: "";
          position: absolute;
          top: 0;
          left: 0;
          right: 0;
          bottom: 0;
          background: rgba(128, 128, 128, 0.3);
          border-radius: var(--ddd-radius-sm);
          pointer-events: none;
          z-index: 1;
        }

        :host([edit-mode]) .random-content-container {
          opacity: 0.7;
          pointer-events: none;
        }

        .header {
          display: flex;
          justify-content: space-between;
          align-items: center;
          margin-bottom: var(--ddd-spacing-3);
          padding-bottom: var(--ddd-spacing-2);
          border-bottom: var(--ddd-border-xs)
            var(--ddd-theme-default-limestoneLight);
        }

        .title {
          font-family: var(--ddd-font-navigation);
          font-size: var(--ddd-font-size-s);
          font-weight: var(--ddd-font-weight-bold);
          color: var(--ddd-theme-default-coalyGray);
          margin: 0;
        }

        .shuffle-btn {
          --simple-icon-button-border-radius: var(--ddd-radius-xs);
        }

        .random-element {
          color: var(--ddd-theme-default-coalyGray);
          line-height: var(--ddd-lh-150);
          font-size: var(--ddd-font-size-s);
        }

        /* Style common elements within the random content */
        .random-element h1,
        .random-element h2,
        .random-element h3,
        .random-element h4,
        .random-element h5,
        .random-element h6 {
          margin: 0 0 var(--ddd-spacing-3) 0;
          color: var(--ddd-theme-default-coalyGray);
        }

        .random-element p {
          margin: 0 0 var(--ddd-spacing-3) 0;
        }

        .random-element blockquote {
          margin: var(--ddd-spacing-3) 0;
          padding: var(--ddd-spacing-3);
          border-left: var(--ddd-border-sm)
            var(--ddd-theme-default-potentialMidnight);
          background-color: var(--ddd-theme-default-limestoneMaxLight);
          font-style: italic;
        }

        .random-element ul,
        .random-element ol {
          margin: 0 0 var(--ddd-spacing-3) 0;
          padding-left: var(--ddd-spacing-5);
        }

        .no-content {
          text-align: center;
          color: var(--ddd-theme-default-coalyGray);
          font-style: italic;
          padding: var(--ddd-spacing-6);
        }

        .loading {
          text-align: center;
          color: var(--ddd-theme-default-coalyGray);
          padding: var(--ddd-spacing-4);
        }

        :host([hidden]) {
          display: none;
        }
      `]}static get properties(){return{...super.properties,page:{type:String,reflect:!0},randomElement:{type:Object},allElements:{type:Array},currentElementIndex:{type:Number},loading:{type:Boolean,reflect:!0},editMode:{type:Boolean,reflect:!0,attribute:"edit-mode"}}}getPageById(e){return a.routerManifest&&a.routerManifest.items&&e?a.routerManifest.items.find((t=>t.id===e)):null}extractTopLevelElements(e){if(!e)return[];const t=globalThis.document.createElement("div");t.innerHTML=e;const n=[],i=["p","div","h1","h2","h3","h4","h5","h6","blockquote","ul","ol","pre","section","article","aside","header","main","figure"];return Array.from(t.children).forEach((e=>{0!==e.textContent.trim().length&&(["script","style","meta","link","page-break"].includes(e.tagName.toLowerCase())||(i.includes(e.tagName.toLowerCase())||e.textContent.trim().length>10)&&n.push({html:e.outerHTML,text:e.textContent.trim(),tag:e.tagName.toLowerCase()}))})),n}async loadPageElements(){if(!this.page||!a.routerManifest)return this.allElements=[],void(this.randomElement=null);this.loading=!0;if(!this.getPageById(this.page))return console.warn("Page not found:",this.page),this.allElements=[],this.randomElement=null,void(this.loading=!1);try{const e=await a.loadItemContent(this.page),t=this.extractTopLevelElements(e);this.allElements=t,t.length>0?this.selectRandomElement():this.randomElement=null}catch(e){console.warn("Failed to load content for page:",this.page,e),this.allElements=[],this.randomElement=null}this.loading=!1}selectRandomElement(){if(0===this.allElements.length)return this.randomElement=null,void(this.currentElementIndex=0);this.currentElementIndex=Math.floor(Math.random()*this.allElements.length),this.randomElement=this.allElements[this.currentElementIndex]}shuffleContent(){if(this.allElements.length<=1)return;let e;do{e=Math.floor(Math.random()*this.allElements.length)}while(e===this.currentElementIndex&&this.allElements.length>1);this.currentElementIndex=e,this.randomElement=this.allElements[this.currentElementIndex]}updated(e){super.updated(e),e.has("page")&&this.loadPageElements()}firstUpdated(e){super.firstUpdated(e),this.page&&this.loadPageElements()}disconnectedCallback(){this.__disposer.forEach((e=>{e()})),this.windowControllers.abort(),super.disconnectedCallback()}haxHooks(){return{editModeChanged:e=>{this.editMode=e},activeElementChanged:()=>{if(this.editMode)return!1},setupActiveElementForm:"haxsetupActiveElementForm"}}haxsetupActiveElementForm(e){if(globalThis.HAXCMS){const n=globalThis.HAXCMS.requestAvailability().store.getManifestItems(!0);var t=[{text:`-- ${this.t.noPage} --`,value:null}];n.forEach((e=>{if(e.id!=this.itemId){let i=e,a="- ";for(;i&&null!=i.parent;)i=n.find((e=>e.id==i.parent)),i&&(a="--"+a);t.push({text:a+e.title,value:e.id})}})),e.settings.configure.forEach(((n,i)=>{"page"===n.property&&(e.settings.configure[i].inputMethod="select",e.settings.configure[i].itemsList=t)}))}}render(){return e`
      <div class="random-content-container">
        <div class="header">
          <simple-icon-button-lite
            icon="icons:shuffle"
            @click="${this.shuffleContent}"
            title="${this.t.shuffle}"
            class="shuffle-btn"
            ?disabled="${this.editMode||this.allElements.length<=1}"
            >${this.t.shuffle}</simple-icon-button-lite
          >
        </div>

        ${this.loading?e` <div class="loading">${this.t.loading}</div> `:n}
        ${this.loading||this.page?n:e` <div class="no-content">${this.t.noPage}</div> `}
        ${this.loading||!this.page||this.randomElement&&0!==this.allElements.length?n:e` <div class="no-content">${this.t.noContent}</div> `}
        ${!this.loading&&this.randomElement?e`
              <div class="random-element">
                ${i(this.randomElement.html)}
              </div>
            `:n}
      </div>
    `}static get haxProperties(){return{canScale:!1,canPosition:!1,canEditSource:!1,gizmo:{title:"Random Page Content",description:"Display a random top-level element from a selected page",icon:"icons:shuffle",color:"purple",tags:["Other"],handles:[],meta:{author:"HAXcms"}},settings:{configure:[{property:"page",title:"Page",description:"Select the page to draw random content from. The element will randomly display one of the top-level HTML elements from this page.",inputMethod:"select",required:!0}],advanced:[]},demoSchema:[{tag:this.tag,properties:{page:null},content:""}]}}}globalThis.customElements.define(SiteRandomContent.tag,SiteRandomContent);
