// @lrnwebcomponents/editable-table/lib/editable-table-behaviors.js
/**
 * Copyright 2018 The Pennsylvania State University
 * @license Apache-2.0, see License.md for full text.
 */
import{html as t,css as e}from"../../../lit-element/lit-element.js";export const editableTableCellStyles=[e`
    .cell-button {
      padding-top: var(--editable-table-cell-vertical-padding, 10px);
      padding-bottom: var(--editable-table-cell-vertical-padding, 10px);
      padding-left: var(--editable-table-cell-horizontal-padding, 6px);
      padding-right: var(--editable-table-cell-horizontal-padding, 6px);
      margin: 0;
      width: 100%;
      display: flex;
      justify-content: var(--editable-table-cell-justify, space-between);
      align-items: center;
      align-content: stretch;
      font-family: inherit;
      font-size: inherit;
      font-weight: inherit;
      background-color: transparent;
      border: none;
      border-radius: 0;
      color: var(--editable-table-cell-color);
    }
    .sr-only {
      position: absolute;
      left: -9999px;
      font-size: 0;
      height: 0;
      width: 0;
      overflow: hidden;
      margin: 0;
      padding: 0;
    }
  `];export const editableTableStyles=[e`
    :host {
      display: block;
      width: 100%;
      max-width: 100%;
      margin: 15px 0;
      font-size: var(--editable-table-font-size, unset);
      font-family: var(--editable-table-font-family, inherit);
      font-weight: var(--editable-table-medium-weight, 300);
      color: var(--editable-table-color, #222);
      background-color: var(--editable-table-bg-color, #fff);

      --simple-picker-font-family: var(
        --editable-table-secondary-font-family,
        "Roboto",
        "Noto",
        sans-serif
      );
      --simple-picker-font-size: var(
        --editable-table-secondary-font-size,
        12px
      );
      --simple-picker-color: var(--editable-table-color, #222);
      --simple-picker-background-color: var(--editable-table-bg-color, #fff);
      --paper-font-caption: {
        font-family: var(
          --editable-table-secondary-font-family,
          "Roboto",
          "Noto",
          sans-serif
        );
      }
    }
    :host([hidden]) {
      display: none;
    }
    .sr-only {
      position: absolute;
      left: -9999px;
      font-size: 0;
      height: 0;
      width: 0;
      overflow: hidden;
      margin: 0;
      padding: 0;
    }
    table {
      width: calc(100% - 2 * var(--editable-table-border-width, 1px));
      display: table;
      border-collapse: collapse;
      border-width: var(--editable-table-border-width, 1px);
      border-style: var(--editable-table-border-style, solid);
      border-color: var(--editable-table-border-color, #999);
      font-weight: var(--editable-table-light-weight, 200);
      color: var(--editable-table-color, #222);
      background-color: var(--editable-table-bg-color, #fff);
    }
    .th,
    .td {
      font-weight: var(--editable-table-light-weight, 200);
      color: var(--editable-table-color, #222);
      background-color: var(--editable-table-bg-color, #fff);
    }
    caption {
      font-size: var(
        --editable-table-caption-font-size,
        var(--editable-table-font-size, unset)
      );
      font-weight: var(--editable-table-heavy-weight, 600);
      color: var(
        --editable-table-caption-color,
        var(--editable-table-color, #222)
      );
      background-color: var(
        --editable-table-caption-bg-color,
        var(--editable-table-bg-color, #fff)
      );
      width: 100%;
    }
    .tr {
      display: table-row;
    }
    .th-or-td {
      display: table-cell;
    }
    .thead-tr .th {
      background-color: var(--editable-table-heading-bg-color, #e0e0e0);
      font-weight: var(--editable-table-heavy-weight, 600);
      color: var(--editable-table-heading-color, #000);
    }
    .tbody-tr .th {
      font-weight: var(--editable-table-heavy-weight, 600);
      color: var(--editable-table-heading-color, #000);
      background-color: var(--editable-table-bg-color, #fff);
      text-align: left;
    }
    *[bordered] .th,
    *[bordered] .td {
      border-width: var(--editable-table-border-width, 1px);
      border-style: var(--editable-table-border-style, solid);
      border-color: var(--editable-table-border-color, #999);
    }
    *[condensed] {
      --editable-table-cell-vertical-padding: var(
        --editable-table-cell-vertical-padding-condensed,
        2px
      );
      --editable-table-cell-horizontal-padding: var(
        --editable-table-cell-horizontal-padding-condensed,
        4px
      );
    }
    *[striped] .tbody-tr:nth-child(2n + 1) .th-or-td {
      background-color: var(--editable-table-stripe-bg-color, #f0f0f0);
    }
    *[column-striped] .tbody-tr .th-or-td:nth-child(2n),
    *[column-striped] .tfoot-tr .th-or-td:nth-child(2n) {
      background-color: var(--editable-table-stripe-bg-color, #f0f0f0);
    }
    .tfoot-tr .th,
    .tfoot-tr .td {
      border-top: 2px solid var(--editable-table-color, #222);
      font-weight: var(--editable-table-heavy-weight, 600);
      color: var(--editable-table-heading-color, #000);
    }
    caption,
    .th-or-td {
      text-align: left;
    }
    *[numeric-styles] .thead-tr .th-or-td[numeric],
    *[numeric-styles] .tfoot-tr .th-or-td[numeric],
    *[numeric-styles] .th-or-td[numeric] .cell {
      text-align: right;
      --editable-table-cell-justify: flex-end;
    }
    *[numeric-styles] .tfoot-tr .th-or-td[negative],
    *[numeric-styles] .td[negative] .cell {
      color: var(--editable-table-negative-color, red);
      --editable-table-cell-color: var(--editable-table-negative-color, red);
    }
    editable-table-sort {
      width: 100%;
    }
    button {
      background-color: transparent;
      border: none;
      border-radius: 0;
    }
    ::slotted(table) {
      display: none;
    }
    @media screen {
      :host {
        overflow-x: auto;
        width: 100%;
        max-width: 100%;
      }
      :host([responsive]) {
        overflow-x: visible;
      }
    }
  `];export const editableTableDisplayStyles=[...editableTableStyles,e`
    table[sort] .thead .th,
    table[filter] .tbody .td,
    table[filter] .tfoot .td {
      padding-left: 0;
      padding-right: 0;
      padding-top: 0;
      padding-bottom: 0;
    }
    caption {
      padding-top: var(--editable-table-cell-vertical-padding, 10px);
      padding-bottom: var(--editable-table-cell-vertical-padding, 10px);
      padding: 0;
    }
    caption > div {
      display: flex;
      align-items: flex-end;
      justify-content: space-between;
    }
    caption > div > *:not(:last-child) {
      padding: 0 0 5px;
    }
    #column {
      width: calc(var(--simple-picker-option-size) + 6px);
      overflow: visible;
      display: none;
      margin-left: 10px;
      --simple-picker-border-width: 1px;
      --simple-picker-focus-border-width: 1px;
      --simple-picker-border-color: var(--editable-table-border-color, #999);
    }
    .th,
    .td {
      padding-top: var(--editable-table-cell-vertical-padding, 10px);
      padding-bottom: var(--editable-table-cell-vertical-padding, 10px);
      padding-left: var(--editable-table-cell-horizontal-padding, 6px);
      padding-right: var(--editable-table-cell-horizontal-padding, 6px);
    }
    span.cell {
      display: block;
    }
    @media screen {
      :host([responsive][responsive-size="xs"]) #column {
        display: inline-flex;
      }
      :host([responsive][responsive-size="xs"]) .th[xs-hidden],
      :host([responsive][responsive-size="xs"]) .td[xs-hidden] {
        display: none;
      }
    }
  `];export const displayProperties={bordered:{attribute:"bordered",type:Boolean},columnStriped:{attribute:"column-striped",type:Boolean},condensed:{attribute:"condensed",type:Boolean},numericStyles:{attribute:"numeric-styles",type:Boolean,reflect:!0},responsive:{attribute:"responsive",type:Boolean,reflect:!0},striped:{attribute:"striped",type:Boolean}};export const dataProperties={downloadable:{attribute:"downloadable",type:Boolean,reflect:!0},filter:{attribute:"filter",type:Boolean,reflect:!0},printable:{attribute:"printable",type:Boolean,reflect:!0},sort:{attribute:"sort",type:Boolean,reflect:!0}};export const displayBehaviors=function(t){return class extends t{static get properties(){return{...super.properties,...displayProperties,...dataProperties,caption:{type:String,value:null},columnHeader:{attribute:"column-header",type:Boolean},csvData:{type:String,attribute:"csv-data"},data:{type:Array,notify:!0,attribute:"data"},dataCsv:{type:String,attribute:"data-csv"},footer:{attribute:"footer",type:Boolean},rowHeader:{attribute:"row-header",type:Boolean}}}constructor(){super(),this.columnHeader=!1,this.downloadable=!1,this.data=[],this.filter=!1,this.footer=!1,this.rowHeader=!1,this.sort=!1,this.dataCsv=void 0,Object.keys(displayProperties).forEach((t=>{this[t]=!1})),this.fetchData()}connectedCallback(){super.connectedCallback(),this.fetchData(),setTimeout((()=>{this.dataCsv||this.loadSlottedTable(),this.__ready=!0}),0)}firstUpdated(t){super.firstUpdated&&super.firstUpdated(t),this.fetchData()}updated(t){super.updated&&super.updated(t),t.forEach(((t,e)=>{"dataCsv"===e&&this.fetchData(),"csvData"===e&&this._loadExternalData(),"striped"===e&&this.striped&&(this.columnStriped=!1),"columnStriped"===e&&this.columnStriped&&(this.striped=!1),"data"==e&&this._dataChanged(this.data,t)}))}get thead(){return this.columnHeader?(this.data||[]).slice(0,1):[]}get tbody(){return(this.data||[]).slice(this.columnHeader?1:0,this.footer?(this.data||[]).length-1:(this.data||[]).length)}get tfoot(){return this.footer?(this.data||[]).slice((this.data||[]).length-1):[]}CSVtoArray(t){let e,a="",i=[""],l=[i],o=0,r=0,d=!0;for(e in t)e=t[e],'"'===e?(d&&e===a&&(i[o]+=e),d=!d):","===e&&d?e=i[++o]="":"\n"===e&&d?("\r"===a&&(i[o]=i[o].slice(0,-1)),i=l[++r]=[e=""],o=0):i[o]+=e,a=e;return l}download(){let t=document.createElement("a"),e=(this.downloadable&&""!=this.caption.trim()?"Table as CSV":`${this.caption} CSV`).replace(/[^\w\d]/g,""),a=this.getTableCSV();t.setAttribute("href","data:text/plain;charset=UTF-8,"+encodeURIComponent(a)),t.setAttribute("download",e+".txt"),t.style.display="none",document.body.appendChild(t),t.click(),document.body.removeChild(t),this.dispatchEvent(new CustomEvent("csv-downloaded",{bubbles:!0,composed:!0,cancelable:!1,detail:{table:this,data:a,filename:e}}))}print(){let t=!!(!!this.shadowRoot&&this.shadowRoot.querySelector("table"))&&window.open("","","left=0,top=0,width=552,height=477,toolbar=0,scrollbars=0,status =0");t&&(t.document.head.innerHTML+=`<style>\n          ${editableTableDisplayStyles.map((t=>t.cssText.replace(/:host/g,"table"))).join(" ")}\n          table {\n            -webkit-print-color-adjust: exact !important;\n            color-adjust: exact !important;\n          }\n        </style>`,t.document.body.innerHTML=this.getTableHTML(!0),t.document.close(),t.focus(),t.print(),t.close()),this.dispatchEvent(new CustomEvent("table-printed",{bubbles:!0,composed:!0,cancelable:!1,detail:this}))}fetchData(){this.dataCsv&&""!==this.dataCsv&&fetch(this.dataCsv).then((t=>t.text())).then((t=>{this.csvData=t})).catch((t=>{this.loadSlottedTable()}))}getTableCSV(){return this.data.map((t=>t.map((t=>(t=this._replaceBlankCell(t),this._isNumericCell(t)?t.replace(/,/g,""):`"${t.replace(/"/g,'""')}"`))).join(","))).join("\n")}getHTML(t=" "){return this.__tempDiv=this.__tempDiv||document.createElement("template"),this.__tempDiv.innerHTML=t,this.__tempDiv.content.cloneNode(!0)}getTableHTML(t=!1){let e=[],a=[],i=[],getTR=(e,a="td",i="td",l="tbody")=>{let o=this.rowHeader?e.slice(0,1):[],r=this.rowHeader?e.slice(1):e;return`\n\t\t<tr${t?` class="${l}-tr tr"`:""}>${o.map((e=>`\n\t\t\t<th scope="row"${t?` class="${i} th-or-td"`:""}>${this._replaceBlankCell(e)}</th>`)).join("")}${r.map((e=>`\n\t\t\t<${a}${t?` class="${i} th-or-td"`:""}>${this._replaceBlankCell(e)}</${i}>`)).join("")}\n\t\t</tr>`};this.thead&&(e=this.thead.map((t=>getTR(t,'th scope="col"',"th","thead")))),this.tbody&&(a=this.tbody.map((t=>getTR(t)))),this.tfoot&&(i=this.tfoot.map((t=>getTR(t,"td","td","tfoot"))));let l=this.getTableProperties(),o="";return Object.keys(this.getTableProperties()).forEach((t=>{if(l[t]&&Object.keys(displayProperties).includes(t)){o+=`${(t=>t.split("").map(((t,e)=>t.toUpperCase()===t?`${0!==e?"-":""}${t.toLowerCase()}`:t)).join(""))(t)} `}})),[`<table ${o}>`,""!==this.caption&&null!==this.caption&&"null"!==this.caption?`\n\t<caption${t?' class="caption"':""}>\n\t\t${this.caption}\n\t</caption>`:"",e.length>0?`\n\t<thead${t?' class="thead"':""}">${e.join("")}\n\t</thead>`:"",a.length>0?`\n\t<tbody${t?' class="tbody"':""}>${a.join("")}\n\t</tbody>`:"",i.length>0?`\n\t<tfoot${t?' class="tfoot"':""}>${i.join("")}\n\t</tfoot>`:"","\n</table>"].join("")}getTableHTMLNode(){let t=document.createElement("editable-table-display"),e=this.getTableProperties();for(var a in e)t[a]=e[a];return t.innerHTML=this.getTableHTML(),t}getTableProperties(){return{bordered:!this.hideBordered&&this.bordered,caption:this.caption,columnHeader:this.columnHeader,columnStriped:this.columnStriped,condensed:!this.hideCondensed&&this.condensed,data:this.data,downloadable:this.downloadable,filter:!this.hideFilter&&this.filter,footer:this.footer,numericStyles:this.numericStyles,printable:this.printable,rowHeader:this.rowHeader,responsive:!this.hideResponsive&&this.responsive,sort:!this.hideSort&&this.sort,striped:!this.hideStriped&&this.striped,summary:this.summary}}importHTML(t){let e=[],a=[...t.querySelectorAll("tr")].map(((t,a)=>[...t.querySelectorAll("th,td")].map(((t,i)=>{let l=t.matches("th")?"th":"td",o=!!t.matches("[colspan]")&&t.getAttribute("colspan"),r=!!t.matches("[rowspan]")&&t.getAttribute("rowspan");return(o||r)&&e.push({type:l,row:a,col:i,rows:r&&""!==r.trim()?parseInt(r.trim()):void 0,cols:o&&""!==o.trim()?parseInt(o.trim()):void 0}),"string"==typeof t.innerHTML?t.innerHTML.trim():t.innerHTML}))));e.forEach((t=>{if(t.cols)for(let e=1;e<t.cols;e++)a[t.row].splice(t.col+1,0,"&nbsp;");if(t.rows)for(let e=1;e<t.rows;e++)a[t.row+1].splice(t.col,0,"&nbsp;")})),a.length>0&&a[0].length>0&&(this.data=a),this.columnHeader=this.columnHeader||t.querySelectorAll("thead").length>0,this.rowHeader=this.rowHeader||t.querySelectorAll("tbody th").length>0,this.footer=this.footer||t.querySelectorAll("tfoot").length>0,this.caption=t.querySelectorAll("caption").length>0?t.querySelector("caption").innerHTML.trim():void 0,Object.keys(displayProperties).forEach((e=>{t.matches(`.${displayProperties[e].attribute||e}`)&&(this[e]=!0)}))}loadSlottedTable(){let t=this.children.item(0);t&&"EDITABLE-TABLE-DISPLAY"===t.tagName&&(t=t.children.item(0)),t&&"TABLE"===t.tagName&&t.children&&t.children.length>0&&this.importHTML(t)}_dataChanged(t,e){}_isNegative(t){return this._isNumericCell(t)&&0===t.trim().indexOf("-")}_isNumericCell(t){return!!t&&!isNaN(t.trim().replace(/\$/g,""))}_isNumericColumn(t){let e=!0;for(let a=0;a<this.tbody.length;a++)this._isNumericCell(this.tbody[a][t])||(e=!1);return e}_isRowFiltered(t){let e,a=(this.filterText||"").trim(),i=this.filterColumn&&t?t[this.filterColumn].trim():"";return this.filterCaseSensitive||(a=a.toLowerCase(),i=i.toLowerCase()),e=this.filterContains?!i.match(a):i!==a,e}_isRowHeader(t,e){return 0===e&&t}_loadExternalData(t){let e=this.CSVtoArray(this.csvData);e.length>0&&e[0].length>0&&(this.data=e),this.columnHeader=!0}_replaceBlankCell(t){return""===String(t).trim()?"-":t}}};export const cellBehaviors=function(t){return class extends t{_getLabel(t,e){if(e)return t+1;{let e="ABCDEFGHIJKLMNOPQRSTUVWXYZ".split(""),a=this._getLetter(t).split("-").reverse(),i="";for(let t=0;t<a.length;t++)""!==a[t]&&(i+=e[a[t]]);return i}}_getLetter(t){let e=Math.floor(t/26),a="";return a+=t-26*e+"-",e>0&&e<26?a+=e-1+"-":e>=26&&(a+=this._getLetter(e-1)),a}}};
