// @lrnwebcomponents/rich-text-editor/lib/buttons/rich-text-editor-image.js
define(["exports", "../../../../lit-element/lit-element.js", "./rich-text-editor-prompt-button.js"], function (_exports, _litElement, _richTextEditorPromptButton) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.RichTextEditorImage = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `rich-text-editor-image`
   * an inline image button for rich text editor
   *
   * @element rich-text-editor-image
   * @demo ./demo/buttons.html
   */
  var RichTextEditorImage = /*#__PURE__*/function (_RichTextEditorPrompt) {
    babelHelpers.inherits(RichTextEditorImage, _RichTextEditorPrompt);

    var _super = _createSuper(RichTextEditorImage);

    babelHelpers.createClass(RichTextEditorImage, [{
      key: "render",
      // render function for template
      value: function render() {
        return babelHelpers.get(babelHelpers.getPrototypeOf(RichTextEditorImage.prototype), "render", this).call(this);
      } // properties available to the custom element for data binding

    }], [{
      key: "tag",

      /**
       * Store the tag name to make it easier to obtain directly.
       */
      get: function get() {
        return "rich-text-editor-image";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(RichTextEditorImage), "properties", this));
      }
    }]);

    function RichTextEditorImage() {
      var _this;

      babelHelpers.classCallCheck(this, RichTextEditorImage);
      _this = _super.call(this);
      _this.fields = [{
        property: "src",
        title: "Image URL",
        description: "The image URL. (Leave blank to remove.)",
        inputMethod: "textfield"
      }, {
        property: "alt",
        title: "Alt Text",
        inputMethod: "textfield"
      }, {
        property: "width",
        title: "Width",
        inputMethod: "textfield",
        inline: true
      }, {
        property: "height",
        title: "Height",
        inputMethod: "textfield",
        inline: true
      }];
      _this.command = "insertHTML";
      _this.label = "Insert Inline Image";
      _this.icon = "editor:insert-photo";
      _this.tag = "img";
      _this.value = {};
      return _this;
    }
    /**
     * overrides default block selectors
     *
     * @readonly
     * @memberof RichTextEditorLink
     */


    babelHelpers.createClass(RichTextEditorImage, [{
      key: "updatePrompt",

      /**
       * updates prompt fields with selected range data
       */
      value: function updatePrompt() {
        babelHelpers.get(babelHelpers.getPrototypeOf(RichTextEditorImage.prototype), "updatePrompt", this).call(this);
        this.__selectionContents = this.__selectionContents;
        this.value = {
          alt: this.__selectionContents.getAttribute("alt"),
          src: this.__selectionContents.getAttribute("src"),
          width: this.__selectionContents.getAttribute("width"),
          height: this.__selectionContents.getAttribute("height")
        };
      }
      /**
       * updates the insertion based on fields
       */

    }, {
      key: "updateSelection",
      value: function updateSelection() {
        var alt = this.__prompt.getPromptValue("alt"),
            src = this.__prompt.getPromptValue("src"),
            width = this.__prompt.getPromptValue("width"),
            height = this.__prompt.getPromptValue("height");

        this.__selection.range.selectNode(this.__selectionContents);

        this.toggled = !src;
        this.commandVal = !src ? "" : "<img src=\"".concat(src, "\"").concat(!alt ? "" : " alt=\"".concat(alt, "\"")).concat(!width ? "" : " width=\"".concat(width, "\"")).concat(!height ? "" : " width=\"".concat(height, "\""), ">");
        this.execCommand();
      }
    }, {
      key: "blockSelectors",
      get: function get() {
        return "img";
      }
      /**
       * whether button is toggled
       *
       * @readonly
       * @memberof RichTextEditorButton
       */

    }, {
      key: "isToggled",
      get: function get() {
        return this.toggled;
      }
    }]);
    return RichTextEditorImage;
  }((0, _richTextEditorPromptButton.RichTextEditorPromptButtonBehaviors)(_litElement.LitElement));

  _exports.RichTextEditorImage = RichTextEditorImage;
  window.customElements.define(RichTextEditorImage.tag, RichTextEditorImage);
});
