// @lrnwebcomponents/rich-text-editor/lib/buttons/rich-text-editor-more-button.js
define(["exports", "../../../../lit-element/lit-element.js", "./rich-text-editor-button.js"], function (_exports, _litElement, _richTextEditorButton) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.RichTextEditorMoreButton = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `rich-text-editor-more-button`
   * a more button to toggle collapsed buttons in the rich text editor
   *
   * @element rich-text-editor-more-button
   * @demo ./demo/buttons.html
   */
  var RichTextEditorMoreButton = /*#__PURE__*/function (_RichTextEditorButton) {
    babelHelpers.inherits(RichTextEditorMoreButton, _RichTextEditorButton);

    var _super = _createSuper(RichTextEditorMoreButton);

    babelHelpers.createClass(RichTextEditorMoreButton, [{
      key: "render",
      // render function for template
      value: function render() {
        return babelHelpers.get(babelHelpers.getPrototypeOf(RichTextEditorMoreButton.prototype), "render", this).call(this);
      }
    }], [{
      key: "tag",

      /**
       * Store the tag name to make it easier to obtain directly.
       */
      get: function get() {
        return "rich-text-editor-more-button";
      }
    }]);

    function RichTextEditorMoreButton() {
      var _this;

      babelHelpers.classCallCheck(this, RichTextEditorMoreButton);
      _this = _super.call(this);
      _this.icon = "more-vert";
      _this.collapseMax = "xs";
      _this.toggled = false;
      _this.label = "More buttons";
      _this.labelToggled = "Fewer buttons";
      return _this;
    } // properties available to the custom element for data binding


    babelHelpers.createClass(RichTextEditorMoreButton, [{
      key: "_buttonTap",

      /**
       * Fires a button tap event
       */
      value: function _buttonTap() {
        return true;
      }
    }, {
      key: "isToggled",

      /**
       * whether button is toggled
       *
       * @readonly
       * @memberof RichTextEditorButton
       */
      get: function get() {
        return this.toggled;
      }
    }], [{
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(RichTextEditorMoreButton), "properties", this)), {}, {
          /**
           * Can this button toggle?
           */
          toggled: {
            attribute: "toggled",
            type: Boolean,
            reflect: true
          },

          /**
           * The maximum size where all of the buttons display
           */
          collapseMax: {
            attribute: "collapse-max",
            type: String,
            reflect: true
          }
        });
      }
    }]);
    return RichTextEditorMoreButton;
  }((0, _richTextEditorButton.RichTextEditorButtonBehaviors)(_litElement.LitElement));

  _exports.RichTextEditorMoreButton = RichTextEditorMoreButton;
  window.customElements.define(RichTextEditorMoreButton.tag, RichTextEditorMoreButton);
});
