// @lrnwebcomponents/rich-text-editor/lib/singletons/rich-text-editor-range-behaviors.js
define(["exports", "./rich-text-editor-highlight.js", "./rich-text-editor-clipboard.js", "./rich-text-editor-source.js"], function (_exports, _richTextEditorHighlight, _richTextEditorClipboard, _richTextEditorSource) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.RichTextEditorRangeBehaviors = void 0;

  function _regeneratorRuntime() { "use strict"; /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/facebook/regenerator/blob/main/LICENSE */ _regeneratorRuntime = function _regeneratorRuntime() { return exports; }; var exports = {}, Op = Object.prototype, hasOwn = Op.hasOwnProperty, $Symbol = "function" == typeof Symbol ? Symbol : {}, iteratorSymbol = $Symbol.iterator || "@@iterator", asyncIteratorSymbol = $Symbol.asyncIterator || "@@asyncIterator", toStringTagSymbol = $Symbol.toStringTag || "@@toStringTag"; function define(obj, key, value) { return Object.defineProperty(obj, key, { value: value, enumerable: !0, configurable: !0, writable: !0 }), obj[key]; } try { define({}, ""); } catch (err) { define = function define(obj, key, value) { return obj[key] = value; }; } function wrap(innerFn, outerFn, self, tryLocsList) { var protoGenerator = outerFn && babelHelpers.instanceof(outerFn.prototype, Generator) ? outerFn : Generator, generator = Object.create(protoGenerator.prototype), context = new Context(tryLocsList || []); return generator._invoke = function (innerFn, self, context) { var state = "suspendedStart"; return function (method, arg) { if ("executing" === state) throw new Error("Generator is already running"); if ("completed" === state) { if ("throw" === method) throw arg; return doneResult(); } for (context.method = method, context.arg = arg;;) { var delegate = context.delegate; if (delegate) { var delegateResult = maybeInvokeDelegate(delegate, context); if (delegateResult) { if (delegateResult === ContinueSentinel) continue; return delegateResult; } } if ("next" === context.method) context.sent = context._sent = context.arg;else if ("throw" === context.method) { if ("suspendedStart" === state) throw state = "completed", context.arg; context.dispatchException(context.arg); } else "return" === context.method && context.abrupt("return", context.arg); state = "executing"; var record = tryCatch(innerFn, self, context); if ("normal" === record.type) { if (state = context.done ? "completed" : "suspendedYield", record.arg === ContinueSentinel) continue; return { value: record.arg, done: context.done }; } "throw" === record.type && (state = "completed", context.method = "throw", context.arg = record.arg); } }; }(innerFn, self, context), generator; } function tryCatch(fn, obj, arg) { try { return { type: "normal", arg: fn.call(obj, arg) }; } catch (err) { return { type: "throw", arg: err }; } } exports.wrap = wrap; var ContinueSentinel = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} var IteratorPrototype = {}; define(IteratorPrototype, iteratorSymbol, function () { return this; }); var getProto = Object.getPrototypeOf, NativeIteratorPrototype = getProto && getProto(getProto(values([]))); NativeIteratorPrototype && NativeIteratorPrototype !== Op && hasOwn.call(NativeIteratorPrototype, iteratorSymbol) && (IteratorPrototype = NativeIteratorPrototype); var Gp = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(IteratorPrototype); function defineIteratorMethods(prototype) { ["next", "throw", "return"].forEach(function (method) { define(prototype, method, function (arg) { return this._invoke(method, arg); }); }); } function AsyncIterator(generator, PromiseImpl) { function invoke(method, arg, resolve, reject) { var record = tryCatch(generator[method], generator, arg); if ("throw" !== record.type) { var result = record.arg, value = result.value; return value && "object" == babelHelpers.typeof(value) && hasOwn.call(value, "__await") ? PromiseImpl.resolve(value.__await).then(function (value) { invoke("next", value, resolve, reject); }, function (err) { invoke("throw", err, resolve, reject); }) : PromiseImpl.resolve(value).then(function (unwrapped) { result.value = unwrapped, resolve(result); }, function (error) { return invoke("throw", error, resolve, reject); }); } reject(record.arg); } var previousPromise; this._invoke = function (method, arg) { function callInvokeWithMethodAndArg() { return new PromiseImpl(function (resolve, reject) { invoke(method, arg, resolve, reject); }); } return previousPromise = previousPromise ? previousPromise.then(callInvokeWithMethodAndArg, callInvokeWithMethodAndArg) : callInvokeWithMethodAndArg(); }; } function maybeInvokeDelegate(delegate, context) { var method = delegate.iterator[context.method]; if (undefined === method) { if (context.delegate = null, "throw" === context.method) { if (delegate.iterator.return && (context.method = "return", context.arg = undefined, maybeInvokeDelegate(delegate, context), "throw" === context.method)) return ContinueSentinel; context.method = "throw", context.arg = new TypeError("The iterator does not provide a 'throw' method"); } return ContinueSentinel; } var record = tryCatch(method, delegate.iterator, context.arg); if ("throw" === record.type) return context.method = "throw", context.arg = record.arg, context.delegate = null, ContinueSentinel; var info = record.arg; return info ? info.done ? (context[delegate.resultName] = info.value, context.next = delegate.nextLoc, "return" !== context.method && (context.method = "next", context.arg = undefined), context.delegate = null, ContinueSentinel) : info : (context.method = "throw", context.arg = new TypeError("iterator result is not an object"), context.delegate = null, ContinueSentinel); } function pushTryEntry(locs) { var entry = { tryLoc: locs[0] }; 1 in locs && (entry.catchLoc = locs[1]), 2 in locs && (entry.finallyLoc = locs[2], entry.afterLoc = locs[3]), this.tryEntries.push(entry); } function resetTryEntry(entry) { var record = entry.completion || {}; record.type = "normal", delete record.arg, entry.completion = record; } function Context(tryLocsList) { this.tryEntries = [{ tryLoc: "root" }], tryLocsList.forEach(pushTryEntry, this), this.reset(!0); } function values(iterable) { if (iterable) { var iteratorMethod = iterable[iteratorSymbol]; if (iteratorMethod) return iteratorMethod.call(iterable); if ("function" == typeof iterable.next) return iterable; if (!isNaN(iterable.length)) { var i = -1, next = function next() { for (; ++i < iterable.length;) { if (hasOwn.call(iterable, i)) return next.value = iterable[i], next.done = !1, next; } return next.value = undefined, next.done = !0, next; }; return next.next = next; } } return { next: doneResult }; } function doneResult() { return { value: undefined, done: !0 }; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, define(Gp, "constructor", GeneratorFunctionPrototype), define(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = define(GeneratorFunctionPrototype, toStringTagSymbol, "GeneratorFunction"), exports.isGeneratorFunction = function (genFun) { var ctor = "function" == typeof genFun && genFun.constructor; return !!ctor && (ctor === GeneratorFunction || "GeneratorFunction" === (ctor.displayName || ctor.name)); }, exports.mark = function (genFun) { return Object.setPrototypeOf ? Object.setPrototypeOf(genFun, GeneratorFunctionPrototype) : (genFun.__proto__ = GeneratorFunctionPrototype, define(genFun, toStringTagSymbol, "GeneratorFunction")), genFun.prototype = Object.create(Gp), genFun; }, exports.awrap = function (arg) { return { __await: arg }; }, defineIteratorMethods(AsyncIterator.prototype), define(AsyncIterator.prototype, asyncIteratorSymbol, function () { return this; }), exports.AsyncIterator = AsyncIterator, exports.async = function (innerFn, outerFn, self, tryLocsList, PromiseImpl) { void 0 === PromiseImpl && (PromiseImpl = Promise); var iter = new AsyncIterator(wrap(innerFn, outerFn, self, tryLocsList), PromiseImpl); return exports.isGeneratorFunction(outerFn) ? iter : iter.next().then(function (result) { return result.done ? result.value : iter.next(); }); }, defineIteratorMethods(Gp), define(Gp, toStringTagSymbol, "Generator"), define(Gp, iteratorSymbol, function () { return this; }), define(Gp, "toString", function () { return "[object Generator]"; }), exports.keys = function (object) { var keys = []; for (var key in object) { keys.push(key); } return keys.reverse(), function next() { for (; keys.length;) { var key = keys.pop(); if (key in object) return next.value = key, next.done = !1, next; } return next.done = !0, next; }; }, exports.values = values, Context.prototype = { constructor: Context, reset: function reset(skipTempReset) { if (this.prev = 0, this.next = 0, this.sent = this._sent = undefined, this.done = !1, this.delegate = null, this.method = "next", this.arg = undefined, this.tryEntries.forEach(resetTryEntry), !skipTempReset) for (var name in this) { "t" === name.charAt(0) && hasOwn.call(this, name) && !isNaN(+name.slice(1)) && (this[name] = undefined); } }, stop: function stop() { this.done = !0; var rootRecord = this.tryEntries[0].completion; if ("throw" === rootRecord.type) throw rootRecord.arg; return this.rval; }, dispatchException: function dispatchException(exception) { if (this.done) throw exception; var context = this; function handle(loc, caught) { return record.type = "throw", record.arg = exception, context.next = loc, caught && (context.method = "next", context.arg = undefined), !!caught; } for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i], record = entry.completion; if ("root" === entry.tryLoc) return handle("end"); if (entry.tryLoc <= this.prev) { var hasCatch = hasOwn.call(entry, "catchLoc"), hasFinally = hasOwn.call(entry, "finallyLoc"); if (hasCatch && hasFinally) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } else if (hasCatch) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); } else { if (!hasFinally) throw new Error("try statement without catch or finally"); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } } } }, abrupt: function abrupt(type, arg) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc <= this.prev && hasOwn.call(entry, "finallyLoc") && this.prev < entry.finallyLoc) { var finallyEntry = entry; break; } } finallyEntry && ("break" === type || "continue" === type) && finallyEntry.tryLoc <= arg && arg <= finallyEntry.finallyLoc && (finallyEntry = null); var record = finallyEntry ? finallyEntry.completion : {}; return record.type = type, record.arg = arg, finallyEntry ? (this.method = "next", this.next = finallyEntry.finallyLoc, ContinueSentinel) : this.complete(record); }, complete: function complete(record, afterLoc) { if ("throw" === record.type) throw record.arg; return "break" === record.type || "continue" === record.type ? this.next = record.arg : "return" === record.type ? (this.rval = this.arg = record.arg, this.method = "return", this.next = "end") : "normal" === record.type && afterLoc && (this.next = afterLoc), ContinueSentinel; }, finish: function finish(finallyLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.finallyLoc === finallyLoc) return this.complete(entry.completion, entry.afterLoc), resetTryEntry(entry), ContinueSentinel; } }, catch: function _catch(tryLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc === tryLoc) { var record = entry.completion; if ("throw" === record.type) { var thrown = record.arg; resetTryEntry(entry); } return thrown; } } throw new Error("illegal catch attempt"); }, delegateYield: function delegateYield(iterable, resultName, nextLoc) { return this.delegate = { iterator: values(iterable), resultName: resultName, nextLoc: nextLoc }, "next" === this.method && (this.arg = undefined), ContinueSentinel; } }, exports; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var RichTextEditorRangeBehaviors = function RichTextEditorRangeBehaviors(SuperClass) {
    return /*#__PURE__*/function (_SuperClass) {
      babelHelpers.inherits(_class, _SuperClass);

      var _super = _createSuper(_class);

      function _class() {
        var _this;

        babelHelpers.classCallCheck(this, _class);
        _this = _super.call(this);
        _this.haxUIElement = true;
        _this.__highlight = window.RichTextEditorHighlight.requestAvailability();
        _this.__clipboard = window.RichTextEditorClipboard.requestAvailability();
        return _this;
      }

      babelHelpers.createClass(_class, [{
        key: "commandIsToggled",
        get: function get() {
          return this.commandToggledForRange(this.range);
        }
        /**
         * whether or not toolbar breadcrumbs
         * (override to force a toolbar to always use or not use them)
         *
         * @readonly
         */

      }, {
        key: "hasBreadcrumbs",
        get: function get() {
          return false;
        }
        /**
         * an array of all the valid commands that are toggled
         *
         * @readonly
         */

      }, {
        key: "toggledCommands",
        get: function get() {
          return this.toggledCommandsForRange();
        }
        /**
         * gets valid commands list
         *
         * @readonly
         */

      }, {
        key: "validCommands",
        get: function get() {
          return ["backColor", "bold", "copy", "createLink", "cut", "decreaseFontSize", "defaultParagraphSeparator", "delete", "fontName", "fontSize", "foreColor", "formatBlock", "forwardDelete", "hiliteColor", "increaseFontSize", "indent", "insertBrOnReturn", "insertHorizontalRule", "insertHTML", "insertImage", "insertLineBreak", "insertOrderedList", "insertUnorderedList", "insertParagraph", "insertText", "italic", "justifyCenter", "justifyFull", "justifyLeft", "justifyRight", "outdent", "paste", "redo", "removeFormat", "selectAll", "strikeThrough", "styleWithCss", "subscript", "superscript", "underline", "undo", "unlink", "useCSS", "wrapRange"];
        }
        /**
         * closes the toolbar
         *
         * @param {object} toolbar
         * @param {object} editor
         * @memberof RichTextEditorManager
         */

      }, {
        key: "closeToolbar",
        value: function closeToolbar() {
          var toolbar = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.__toolbar;
          var editor = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.__toolbar.target;
          if (editor) toolbar.disableEditing(editor);
          toolbar.target = undefined;
          document.body.append(toolbar);
        }
        /* ------ HANDLES CLIPBOARD AND HTML ------------------------- */

      }, {
        key: "sanitizeHTML",
        value: function sanitizeHTML(html) {
          return this.__toolbar.sanitizeHTML(html);
        }
        /**
         * gets trimmed version of innerHTML
         *
         * @param {obj} node
         * @returns {string}
         * @memberof RichTextEditor
         */

      }, {
        key: "trimHTML",
        value: function trimHTML(node) {
          var str = node ? node.innerHTML : undefined;
          return this.trimString(str);
        }
        /**
         * cleans and trims a string of HTML so that it has no extra spaces
         *
         * @param {string} str
         * @returns {string}
         */

      }, {
        key: "trimString",
        value: function trimString(str) {
          return (str || "").replace(/<!--[^(-->)]*-->/g, "").replace(/[\s\t\r\n]/gim, "");
        }
        /**
         * cleans up indents and extra spaces in HTML string for source code editor
         *
         * @param {string} [str=""]
         * @returns {string}
         */

      }, {
        key: "outdentHTML",
        value: function outdentHTML() {
          var str = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "";
          str = (this.sanitizeHTML(str) || "").replace(/[\s]*$/, "").replace(/^[\n\r]*/, "").replace(/[\n\r]+/gm, "\n");
          var match = str.match(/^\s+/),
              find = match ? match[0] : false,
              regex = !find ? false : new RegExp("\\n".concat(find), "gm");
          str = str.replace(/^\s+/, "");
          str = regex ? str.replace(regex, "\n ") : str;
          return str;
        }
        /**
         * pastes sanitized clipboard contents into current editor's selected range
         * @param {object} editor an HTML object that can be edited
         * @returns {void}
         */

      }, {
        key: "pasteFromClipboard",
        value: function pasteFromClipboard() {
          var _this2 = this;

          var range = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.range;

          this.__clipboard.setClipboard();

          setTimeout( /*#__PURE__*/babelHelpers.asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee() {
            return _regeneratorRuntime().wrap(function _callee$(_context) {
              while (1) {
                switch (_context.prev = _context.next) {
                  case 0:
                    _this2.paste(_this2.__clipboard.value, range, true);

                  case 1:
                  case "end":
                    return _context.stop();
                }
              }
            }, _callee);
          })), 100);
        }
        /**
         * pastes content into editor's selected range
         *
         * @param {obj} editor editor
         * @param {obj} pasteContent content to paste
         * @memberof RichTextEditorSelection
         */

      }, {
        key: "paste",
        value: function paste(pasteContent) {
          var range = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.range;
          var sanitize = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
          this.dispatchEvent(new CustomEvent("paste", {
            bubbles: true,
            cancelable: true,
            composed: true,
            detail: this.target
          }));
          var target = this.__toolbar.target;

          if (target) {
            var _range = this.range,
                div = document.createElement("div"),
                parent = _range.commonAncestorContainer.parentNode,
                closest = parent.closest("[contenteditable=true]:not([disabled]),input:not([disabled]),textarea:not([disabled])");

            if (target = closest) {
              div.innerHTML = sanitize ? this.sanitizeHTML(pasteContent || this.__clipboard.value) : pasteContent;

              if (_range && _range.extractContents) {
                _range.extractContents();
              }

              _range.insertNode(div);

              while (div.firstChild) {
                div.parentNode.insertBefore(div.firstChild, div);
              }

              div.parentNode.removeChild(div);
            }

            target.normalize();
          }
        }
        /* ------ GETS SELECTION / RANGE ------------------------- */

        /**
         * determines if range is being preserved by highlight
         *
         * @readonly
         */

      }, {
        key: "highlighted",
        get: function get() {
          return !!this.__highlight && !this.__highlight.hidden;
        }
      }, {
        key: "getRange",
        value: function getRange() {
          var sel = this.getSelection();
          return this.__toolbar ? this.__toolbar.getRange() : sel.getRangeAt(0);
        }
      }, {
        key: "getSelection",
        value: function getSelection() {
          return this.__toolbar ? this.__toolbar.getSelection() : window.getSelection();
        }
        /**
         * gets closest shadowRoot or document from node
         *
         * @param {object} node
         * @returns object
         * @memberof RichTextEditorManager
         */

      }, {
        key: "getRoot",
        value: function getRoot(node) {
          return !node || node === document ? document : node.parentNode ? this.getRoot(node.parentNode) : node;
        }
      }, {
        key: "debugRange",
        value: function debugRange() {
          var range = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.getRange();
          var contents = !range || range.collapsed || !range.cloneContents ? false : range.cloneContents();
          return {
            range: range,
            contents: contents,
            text: !contents ? false : contents.textContent
          };
        }
        /**
         * determines if command should be toggled on for a specfified range.
         * Example: if range is bold, returns 'b'
         *
         * @param {*} [range=this.range || this.getRange()]
         * @param {*} [command=this.command]
         * @returns
         */

      }, {
        key: "commandToggledForRange",
        value: function commandToggledForRange() {
          var range = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.range || this.getRange();
          var command = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.command;
          var commandVal = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : this.commandVal;
          var query = !!range && !!command ? document.queryCommandState(this.command) : false,

          /* workaround because queryCommandState("underline") returns true on links */
          block = command === "underline" ? !!this.rangeOrMatchingAncestor("u") : command === "wrapRange" ? !!this.rangeOrMatchingAncestor(commandVal) : query;
          return !!block ? true : false;
        }
        /**
         * returns an array of all toggled commands for a range.
         * Example: if range is bold and italicized, returns ['b','i']
         *
         * @param {range} [range=this.range || this.getRange()]
         * @returns {array}
         */

      }, {
        key: "toggledCommandsForRange",
        value: function toggledCommandsForRange() {
          var _this3 = this;

          var range = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.range || this.getRange();
          return this.validCommands.filter(function (command) {
            _this3.commandToggledForRange(range, command);
          });
        }
        /**
         * selects command element that contains range.
         * Example: if command is 'a' and range is text inside <a/>, returns <a>
         *
         * @param {string} [cssQuery=this.tagsList || '']
         * @param {range} [range=this.range || this.getRange()]
         * @returns {node}
         */

      }, {
        key: "rangeOrMatchingAncestor",
        value: function rangeOrMatchingAncestor() {
          var cssQuery = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.tagsList || "";
          var range = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.range || this.getRange();
          var start = this.rangeElementOrParentElement(range);
          return !start || cssQuery == "" ? undefined : start.matches(cssQuery) ? start : start.closest(cssQuery);
        }
        /**
         * returns elmement that includes range
         *
         * @param {object} [range=this.range || this.getRange()]
         * @returns {node}
         */

      }, {
        key: "rangeElementOrParentElement",
        value: function rangeElementOrParentElement() {
          var range = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.range || this.getRange();
          return this.rangeIsElement(range) ? range.startContainer : this.rangeParentElement(range);
        }
        /**
         * returns node that includes range
         *
         * @param {*} [range=this.range || this.getRange()]
         * @returns
         */

      }, {
        key: "rangeNodeOrParentNode",
        value: function rangeNodeOrParentNode() {
          var range = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.range || this.getRange();
          return this.rangeIsNode(range) ? range.startContainer : this.rangeParentNode(range);
        }
        /**
         * returns a an array ancestor nodes of range to use for breadcrumbs
         *
         * @param {range} [range=this.range || this.getRange()]
         * @returns {array}
         */

      }, {
        key: "rangeBreadcrumbs",
        value: function rangeBreadcrumbs() {
          var _this4 = this;

          var range = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.range || this.getRange();

          var nodes = [],
              getParentNode = function getParentNode(node) {
            nodes.push(node);
            if (node.parentNode && node.parentNode !== _this4.target) getParentNode(node.parentNode);
          };

          nodes.push({
            nodeName: false
          });
          return nodes.reverse();
        }
        /**
         * determines if range is a single node
         *
         * @param {range} [range=this.range || this.getRange()]
         * @returns {boolean}
         */

      }, {
        key: "rangeIsNode",
        value: function rangeIsNode() {
          var range = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.range || this.getRange();
          var startContainer = !range ? undefined : range.startContainer,
              startOffset = !range ? undefined : range.startOffset,
              endContainer = !range ? undefined : range.endContainer,
              endOffset = !range ? undefined : range.endOffset;
          return !!startContainer && startContainer === endContainer && endOffset - startOffset === 1;
        }
        /**
         * determines if range is a single element
         *
         * @param {*} [range=this.range || this.getRange()]
         * @returns {boolean}
         */

      }, {
        key: "rangeIsElement",
        value: function rangeIsElement() {
          var range = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.range || this.getRange();
          return this.rangeIsNode(range) && range.startContainer.nodeType == 1;
        }
        /**
         * gets parent element of range
         *
         * @param {range} [range=this.range || this.getRange()]
         * @returns {node}
         */

      }, {
        key: "rangeParentElement",
        value: function rangeParentElement() {
          var range = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.range || this.getRange();
          var common = !range ? undefined : range.commonAncestorContainer;
          return !common ? undefined : common.nodeType == 1 ? common : common.parentElement;
        }
        /**
         * gets parent node of range
         *
         * @param {range} [range=this.range || this.getRange()]
         * @returns {node}
         */

      }, {
        key: "rangeParentNode",
        value: function rangeParentNode() {
          var range = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.range || this.getRange();
          var common = !range ? undefined : range.commonAncestorContainer;
          return !common ? undefined : common.parentNode;
        }
        /**
         * returns an array of all node in a range
         *
         * @param {range} [range=this.range || this.getRange()]
         * @returns {array}
         */

      }, {
        key: "rangeNodes",
        value: function rangeNodes() {
          var range = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.range || this.getRange();
          var node = this.rangeFirstNode(range),
              nodes = [];

          while (!!node) {
            nodes.push(node);
            node = node.nextSibling;
          }

          return nodes;
        }
        /**
         * gets first node of range
         *
         * @param {range} [range=this.range || this.getRange()]
         * @returns {node}
         */

      }, {
        key: "rangeFirstNode",
        value: function rangeFirstNode() {
          var range = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.range || this.getRange();
          return !!range && !!range.startContainer && !!range.startContainer.childNodes ? range.startContainer.childNodes[range.startOffset - 1] : undefined;
        }
        /**
         * gets last node of range
         *
         * @param {range} [range=this.range || this.getRange()]
         * @returns {node}
         */

      }, {
        key: "rangeLastNode",
        value: function rangeLastNode() {
          var range = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.range || this.getRange();
          return !!range && !!range.endContainer && !!range.endContainer.childNodes ? range.endContainer.childNodes[range.endOffset - 1] : undefined;
        }
        /* ------ CONTROLS RANGE ------------------------- */

        /**
         * sets selection range to specified node
         * @param {object} node node to select
         * @returns {void}
         */

      }, {
        key: "selectNode",
        value: function selectNode(node) {
          var range = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.range;
          if (!node || !node.parentNode) return;

          if (!range) {
            var sel = this.getSelection();
            range = document.createRange();
            sel.removeAllRanges();
            sel.addRange(range);
          }

          if (range) {
            range.selectNode(node);

            this._rangeChanged(range);
          }
        }
        /**
         * sets selection range to specified node's contents
         * @param {object} node node to select
         * @returns {void}
         */

      }, {
        key: "selectNodeContents",
        value: function selectNodeContents(node) {
          var range = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.range;

          if (node) {
            if (!range) {
              var sel = this.getSelection();
              range = document.createRange();
              sel.removeAllRanges();
              sel.addRange(range);
            }

            if (range) {
              range.selectNodeContents(node);

              this._rangeChanged(range);
            }
          }
        }
        /**
         * selects and highlights a node
         *
         * @param {object} node
         * @param {object} range object
         */

      }, {
        key: "highlightNode",
        value: function highlightNode(node) {
          var range = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.range;
          var element = node || this.rangeOrMatchingAncestor();
          if (!!element) this.selectNode(node, range);

          this.__toolbar.updateRange();

          this.__highlight.wrap(range);
        }
        /**
         * selects or deselects(collapses) a range
         *
         * @param {object} range
         * @param {boolean} [select=true] select range?
         * @memberof RichTextEditorSelection
         */

      }, {
        key: "selectRange",
        value: function selectRange() {
          var range = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.range;
          var select = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;

          if (range) {
            if (select) {
              var sel = this.getSelection();
              sel.removeAllRanges();
              sel.addRange(range);
            } else {
              if (!range.isCollapsed) range.collapse();
            } //this._rangeChanged(range);

          }

          return range;
        }
        /**
         * sets range to content within a node
         *
         * @param {object} node
         * @param {range} range
         * @returns
         * @memberof RichTextEditorSelection
         */

      }, {
        key: "surroundRange",
        value: function surroundRange(node) {
          var range = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.range;

          if (range) {
            range.surroundContents(node); //this._rangeChanged(range);
          }
        }
      }, {
        key: "_rangeChanged",
        value: function _rangeChanged(range) {
          this.dispatchEvent(new CustomEvent("rangechange", {
            bubbles: true,
            composed: true,
            cancelable: true,
            detail: {
              element: this,
              range: range
            }
          }));
        }
        /**
         * only initialises view source if needed
         *
         */

      }, {
        key: "initViewSource",
        value: function initViewSource() {
          this.__source = this.__source || window.RichTextEditorSource.requestAvailability();
        }
        /**
         * handles commands sent from toolbar
         *
         * @param {object} toolbar toolbar element
         * @param {string} command command string
         * @param {string} commandVal string for command
         * @param {object} range
         * @memberof RichTextEditorManager
         */

      }, {
        key: "_handleCommand",
        value: function _handleCommand(command, commandVal) {
          var range = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : this.getRange();
          var toolbar = this.__toolbar,
              target = toolbar.target;

          if (this.validCommands.includes(command)) {
            commandVal = toolbar && commandVal ? toolbar.sanitizeHTML(commandVal) : commandVal;

            if (command == "wrapRange" && !!commandVal) {
              // wrap or unwrapsa range with a commandVal element
              var toggled = this.commandToggledForRange(range, "wrapRange", commandVal),
                  node = this.rangeOrMatchingAncestor(commandVal);

              if (!toggled) {
                // if button is not toggled, wrap the range
                this.__highlight.wrap(range);

                var html = this.__highlight.innerHTML;
                this.__highlight.innerHTML = "<".concat(commandVal, ">").concat(html.trim(), "</").concat(commandVal, ">");
                node = this.__highlight.querySelector(commandVal);
                range.selectNode(node);

                this.__highlight.unwrap(range);
              } else if (toggled) {
                //if button is toggled, unwrap the range
                var nodes = node ? babelHelpers.toConsumableArray(node.childNodes).reverse() : [];
                if (range) range.setStartBefore(node);
                nodes.forEach(function (node) {
                  if (range) range.insertNode(node);
                });
                if (node) node.parentElement.normalize();
                if (node) range.selectNodeContents(node);
                if (node) node.remove();
              }

              target.normalize();
            }

            if (this.__highlight.parentNode === document.body) this.__highlight.wrap(range);

            this.__highlight.unwrap(range);

            range = range || this.__highlight.range || this.range;
            this.selectRange(range);

            if (command != "paste" && command != "wrapRange") {
              document.execCommand(command, false, commandVal);
              target.normalize();
            } else if (navigator.clipboard && command == "paste") {
              this.pasteFromClipboard();
            }
          } else if (command === "cancel") {
            if (target) target.revert();
            toolbar.close(target);
          } else if (command === "close") {
            toolbar.close(target);
          } else if (command === "viewSource") {
            this.__source.toggle(toolbar);
          }
        }
      }], [{
        key: "properties",
        get:
        /**
         * Declared properties and their corresponding attributes
         */
        function get() {
          return {
            /**
             * current text selected range, which is actually a range.
             */
            range: {
              name: "range",
              type: Object
            },
            __toolbar: {
              name: toolbar,
              type: Object
            },
            __highlight: {
              type: Object
            },
            __clipboard: {
              type: Object
            },
            __source: {
              type: Object
            }
          };
        }
      }]);
      return _class;
    }(SuperClass);
  };

  _exports.RichTextEditorRangeBehaviors = RichTextEditorRangeBehaviors;
});
