// lodash-es/assignIn.js
define(["exports", "./_copyObject.js", "./_createAssigner.js", "./keysIn.js"], function (_exports, _copyObject, _createAssigner, _keysIn) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.default = void 0;
  _copyObject = babelHelpers.interopRequireDefault(_copyObject);
  _createAssigner = babelHelpers.interopRequireDefault(_createAssigner);
  _keysIn = babelHelpers.interopRequireDefault(_keysIn);

  /**
   * This method is like `_.assign` except that it iterates over own and
   * inherited source properties.
   *
   * **Note:** This method mutates `object`.
   *
   * @static
   * @memberOf _
   * @since 4.0.0
   * @alias extend
   * @category Object
   * @param {Object} object The destination object.
   * @param {...Object} [sources] The source objects.
   * @returns {Object} Returns `object`.
   * @see _.assign
   * @example
   *
   * function Foo() {
   *   this.a = 1;
   * }
   *
   * function Bar() {
   *   this.c = 3;
   * }
   *
   * Foo.prototype.b = 2;
   * Bar.prototype.d = 4;
   *
   * _.assignIn({ 'a': 0 }, new Foo, new Bar);
   * // => { 'a': 1, 'b': 2, 'c': 3, 'd': 4 }
   */
  var assignIn = (0, _createAssigner.default)(function (object, source) {
    (0, _copyObject.default)(source, (0, _keysIn.default)(source), object);
  });
  var _default = assignIn;
  _exports.default = _default;
});
