// lodash-es/forInRight.js
define(["exports", "./_baseForRight.js", "./_castFunction.js", "./keysIn.js"], function (_exports, _baseForRight, _castFunction, _keysIn) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.default = void 0;
  _baseForRight = babelHelpers.interopRequireDefault(_baseForRight);
  _castFunction = babelHelpers.interopRequireDefault(_castFunction);
  _keysIn = babelHelpers.interopRequireDefault(_keysIn);

  /**
   * This method is like `_.forIn` except that it iterates over properties of
   * `object` in the opposite order.
   *
   * @static
   * @memberOf _
   * @since 2.0.0
   * @category Object
   * @param {Object} object The object to iterate over.
   * @param {Function} [iteratee=_.identity] The function invoked per iteration.
   * @returns {Object} Returns `object`.
   * @see _.forIn
   * @example
   *
   * function Foo() {
   *   this.a = 1;
   *   this.b = 2;
   * }
   *
   * Foo.prototype.c = 3;
   *
   * _.forInRight(new Foo, function(value, key) {
   *   console.log(key);
   * });
   * // => Logs 'c', 'b', then 'a' assuming `_.forIn` logs 'a', 'b', then 'c'.
   */
  function forInRight(object, iteratee) {
    return object == null ? object : (0, _baseForRight.default)(object, (0, _castFunction.default)(iteratee), _keysIn.default);
  }

  var _default = forInRight;
  _exports.default = _default;
});
