// lodash-es/functions.js
define(["exports", "./_baseFunctions.js", "./keys.js"], function (_exports, _baseFunctions, _keys) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.default = void 0;
  _baseFunctions = babelHelpers.interopRequireDefault(_baseFunctions);
  _keys = babelHelpers.interopRequireDefault(_keys);

  /**
   * Creates an array of function property names from own enumerable properties
   * of `object`.
   *
   * @static
   * @since 0.1.0
   * @memberOf _
   * @category Object
   * @param {Object} object The object to inspect.
   * @returns {Array} Returns the function names.
   * @see _.functionsIn
   * @example
   *
   * function Foo() {
   *   this.a = _.constant('a');
   *   this.b = _.constant('b');
   * }
   *
   * Foo.prototype.c = _.constant('c');
   *
   * _.functions(new Foo);
   * // => ['a', 'b']
   */
  function functions(object) {
    return object == null ? [] : (0, _baseFunctions.default)(object, (0, _keys.default)(object));
  }

  var _default = functions;
  _exports.default = _default;
});
