// lodash-es/functionsIn.js
define(["exports", "./_baseFunctions.js", "./keysIn.js"], function (_exports, _baseFunctions, _keysIn) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.default = void 0;
  _baseFunctions = babelHelpers.interopRequireDefault(_baseFunctions);
  _keysIn = babelHelpers.interopRequireDefault(_keysIn);

  /**
   * Creates an array of function property names from own and inherited
   * enumerable properties of `object`.
   *
   * @static
   * @memberOf _
   * @since 4.0.0
   * @category Object
   * @param {Object} object The object to inspect.
   * @returns {Array} Returns the function names.
   * @see _.functions
   * @example
   *
   * function Foo() {
   *   this.a = _.constant('a');
   *   this.b = _.constant('b');
   * }
   *
   * Foo.prototype.c = _.constant('c');
   *
   * _.functionsIn(new Foo);
   * // => ['a', 'b', 'c']
   */
  function functionsIn(object) {
    return object == null ? [] : (0, _baseFunctions.default)(object, (0, _keysIn.default)(object));
  }

  var _default = functionsIn;
  _exports.default = _default;
});
