// lodash-es/math.default.js
define(["exports", "./add.js", "./ceil.js", "./divide.js", "./floor.js", "./max.js", "./maxBy.js", "./mean.js", "./meanBy.js", "./min.js", "./minBy.js", "./multiply.js", "./round.js", "./subtract.js", "./sum.js", "./sumBy.js"], function (_exports, _add, _ceil, _divide, _floor, _max, _maxBy, _mean, _meanBy, _min, _minBy, _multiply, _round, _subtract, _sum, _sumBy) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.default = void 0;
  _add = babelHelpers.interopRequireDefault(_add);
  _ceil = babelHelpers.interopRequireDefault(_ceil);
  _divide = babelHelpers.interopRequireDefault(_divide);
  _floor = babelHelpers.interopRequireDefault(_floor);
  _max = babelHelpers.interopRequireDefault(_max);
  _maxBy = babelHelpers.interopRequireDefault(_maxBy);
  _mean = babelHelpers.interopRequireDefault(_mean);
  _meanBy = babelHelpers.interopRequireDefault(_meanBy);
  _min = babelHelpers.interopRequireDefault(_min);
  _minBy = babelHelpers.interopRequireDefault(_minBy);
  _multiply = babelHelpers.interopRequireDefault(_multiply);
  _round = babelHelpers.interopRequireDefault(_round);
  _subtract = babelHelpers.interopRequireDefault(_subtract);
  _sum = babelHelpers.interopRequireDefault(_sum);
  _sumBy = babelHelpers.interopRequireDefault(_sumBy);
  var _default = {
    add: _add.default,
    ceil: _ceil.default,
    divide: _divide.default,
    floor: _floor.default,
    max: _max.default,
    maxBy: _maxBy.default,
    mean: _mean.default,
    meanBy: _meanBy.default,
    min: _min.default,
    minBy: _minBy.default,
    multiply: _multiply.default,
    round: _round.default,
    subtract: _subtract.default,
    sum: _sum.default,
    sumBy: _sumBy.default
  };
  _exports.default = _default;
});
