// @lrnwebcomponents/csv-render/csv-render.js
define(["exports","require","../../lit-element/lit-element.js","../simple-colors/simple-colors.js"],(function(_exports,_require,_litElement,_simpleColors){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.CsvRender=void 0,_require=babelHelpers.interopRequireWildcard(_require);
/**
   * Copyright 2018 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class CsvRender extends _simpleColors.SimpleColors{static get styles(){return[_litElement.css`
        :host {
          display: block;
        }
        .table {
          width: 100%;
          border: 1px solid var(--simple-colors-default-theme-accent-6);
          border-collapse: collapse;
          white-space: nowrap;
          font-size: 16px;
          background-color: var(--simple-colors-default-theme-grey-1);
        }
        .table thead {
          padding-bottom: 0.16px;
          position: sticky;
        }
        .table caption {
          background-color: var(--simple-colors-default-theme-accent-1);
          font-weight: bold;
          padding: 8px;
          border: 1px solid var(--simple-colors-default-theme-accent-6);
          border-bottom: none;
        }
        :host(:not([accent-color])) .table caption,
        :host([accent-color="grey"]) .table caption {
          background-color: var(--simple-colors-default-theme-accent-2);
        }
        .table tbody tr {
          position: relative;
          height: 48px;
          -webkit-transition-duration: 0.28s;
          transition-duration: 0.28s;
          -webkit-transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
          transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
          -webkit-transition-property: background-color;
          transition-property: background-color;
        }
        .table tbody tr:hover {
          background-color: var(--simple-colors-default-theme-accent-1);
        }
        :host(:not([accent-color])) .table tbody tr:hover,
        :host([accent-color="grey"]) .table tbody tr:hover {
          background-color: var(--simple-colors-default-theme-accent-2);
        }
        .table td,
        .table thead th,
        .table th {
          padding: 0 1.125em;
          text-align: left;
        }
        .table td {
          border-top: 1px solid var(--simple-colors-default-theme-accent-6);
          border-bottom: 1px solid var(--simple-colors-default-theme-accent-6);
        }
        .table th {
          position: relative;
          vertical-align: bottom;
          text-overflow: ellipsis;
          font-size: 16px;
          font-weight: bold;
          line-height: 24px;
          letter-spacing: 0;
          color: rgba(0, 0, 0, 0.54);
          height: 48px;
          padding-bottom: 8px;
          box-sizing: border-box;
        }
        #loading {
          position: absolute;
        }
        #download paper-button {
          color: var(--simple-colors-default-theme-accent-6);
          border-radius: 36px;
          width: 36px;
          height: 36px;
          min-width: unset;
          padding: 0;
          margin: 0;
          display: inline-flex;
        }
        iron-icon {
          display: inline-flex;
          margin: 0;
          padding: 0;
        }
        #download paper-button:hover,
        #download paper-button:focus,
        #download paper-button:active {
          color: var(--simple-colors-default-theme-accent-8);
          outline: 2px solid var(--simple-colors-default-theme-accent-6);
        }
      `]}constructor(){super(),this.table=[],this.tableHeadings=[],this.tableData="",new Promise((res,rej)=>_require.default(["../hexagon-loader/hexagon-loader.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-ajax/iron-ajax.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/iron-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icon/iron-icon.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/paper-button/paper-button.js"],res,rej)),new Promise((res,rej)=>_require.default(["../simple-tooltip/simple-tooltip.js"],res,rej))}updated(changedProperties){changedProperties.forEach((oldValue,propName)=>{"color"==propName&&this._getAccentColor(this[propName],oldValue)})}render(){return _litElement.html`
      <iron-ajax
        auto
        url="${this.dataSource}"
        handle-as="text"
        debounce-duration="500"
        @last-response-changed="${this.tableDataChanged}"
        @response="${this.handleResponse}"
      ></iron-ajax>
      <hexagon-loader
        id="loading"
        accent-color="${this.accentColor}"
        loading
        item-count="4"
        size="small"
      ></hexagon-loader>
      <a href="${this.dataSource}" id="download" tabindex="-1">
        <paper-button
          ><iron-icon icon="file-download"></iron-icon
        ></paper-button>
      </a>
      <simple-tooltip for="download" animation-delay="200" offset="14"
        >Download table data</simple-tooltip
      >
      <table class="table" summary="${this.summary}">
        ${this.caption?_litElement.html`
              <caption>
                ${this.caption}
              </caption>
            `:""}
        <thead>
          <tr>
            ${this.tableHeadings.map(heading=>_litElement.html`
                  <th scope="col">${heading}</th>
                `)}
          </tr>
        </thead>
        <tbody>
          ${this.table.map(row=>_litElement.html`
              <tr>
                ${row.map(col=>_litElement.html`
                    <td>${col}</td>
                  `)}
              </tr>
            `)}
        </tbody>
      </table>
    `}static get tag(){return"csv-render"}tableDataChanged(e){this.tableData=e.detail.value}static get properties(){return{dataSource:{type:String,attribute:"data-source"},caption:{type:String},summary:{type:String},table:{type:Array},tableHeadings:{type:Array},tableData:{type:String,attribute:"table-data"},color:{type:String}}}handleResponse(e){this.table=this.CSVtoArray(this.tableData),this.tableHeadings=this.table.shift(),this.shadowRoot.querySelector("#loading").loading=!1}CSVtoArray(text){let l,p="",row=[""],ret=[row],i=0,r=0,s=!0;for(l in text)l=text[l],'"'===l?(s&&l===p&&(row[i]+=l),s=!s):","===l&&s?l=row[++i]="":"\n"===l&&s?("\r"===p&&(row[i]=row[i].slice(0,-1)),row=ret[++r]=[l=""],i=0):row[i]+=l,p=l;return ret}_getAccentColor(color){color=color.replace("-text",""),this.accentColor&&"grey"!==this.accentColor||!this.colors[color]||(this.accentColor=color)}}_exports.CsvRender=CsvRender,window.customElements.define(CsvRender.tag,CsvRender)}));
