// @lrnwebcomponents/micro-copy-heading/micro-copy-heading.js
define(["exports","../../lit-element/lit-element.js","../hax-body-behaviors/lib/HAXWiring.js"],function(_exports,_litElement,_HAXWiring){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.MicroCopyHeading=void 0;class MicroCopyHeading extends _litElement.LitElement{render(){return _litElement.html`
      <style>
        :host {
          display: block;
          margin: 16px 0;
        }

        :host([hidden]) {
          display: none;
        }
        span {
          margin-left: 8px;
        }
        h2 {
          display: inline-flex;
          margin: unset;
          padding: unset;
          font-size: 0.75em;
          color: var(--simple-colors-default-theme-red-5, #de2654);
          border: 2px solid var(--simple-colors-default-theme-red-5, #de2654);
          line-height: 12px;
          margin-right: 10px;
          text-transform: uppercase;
          font-weight: 500;
          letter-spacing: 0.09em;
          padding: 6px 16px;
        }
      </style>
      <h2>${this.heading}<span aria-hidden="true">${this.endcap}</span></h2>
    `}static get haxProperties(){return{canScale:!0,canPosition:!0,canEditSource:!1,gizmo:{title:"Micro copy-heading",description:"small call to action / attention that acts as a heading too",icon:"icons:android",color:"green",groups:["Copy"],handles:[{type:"todo:read-the-docs-for-usage"}],meta:{author:"btopro",owner:"The Pennsylvania State University"}},settings:{quick:[{property:"heading",description:"",inputMethod:"textfield",required:!1,icon:"icons:android"},{property:"endCap",description:"",inputMethod:"textfield",required:!1,icon:"icons:android"}],configure:[{property:"heading",description:"",inputMethod:"textfield",required:!1,icon:"icons:android"},{property:"endCap",description:"",inputMethod:"textfield",required:!1,icon:"icons:android"}],advanced:[]}}}static get properties(){let props={heading:{name:"heading",type:String,value:"Telling our story"},endcap:{name:"endcap",type:String,value:"//"}};return super.properties&&(props=Object.assign(props,super.properties)),props}tag(){return"micro-copy-heading"}constructor(){super(),this.tag=MicroCopyHeading.tag;let obj=MicroCopyHeading.properties;for(let p in obj)obj.hasOwnProperty(p)&&(this.hasAttribute(p)?this[p]=this.getAttribute(p):(this.setAttribute(p,obj[p].value),this[p]=obj[p].value))}connectedCallback(){super.connectedCallback(),this.HAXWiring=new _HAXWiring.HAXWiring,this.HAXWiring.setup(MicroCopyHeading.haxProperties,MicroCopyHeading.tag,this)}}_exports.MicroCopyHeading=MicroCopyHeading,customElements.define("micro-copy-heading",MicroCopyHeading)});
