// @lrnwebcomponents/date-card/date-card.js
define(["exports", "../../lit-element/lit-element.js", "../simple-colors/simple-colors.js"], function (_exports, _litElement, _simpleColors) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.DateCard = void 0;

  function _templateObject14_15211020de5b11eaa3c04df0985e08b7() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: inline-flex;\n        }\n        .card {\n          box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);\n          width: var(--date-card-card-width, 100px);\n          border-radius: 10px;\n        }\n        .card:hover {\n          box-shadow: 0 8px 16px 0 rgba(0, 0, 0, 0.2);\n          border-radius: 10px;\n        }\n        .month,\n        .dateNumber,\n        .dayName {\n          text-align: center;\n          vertical-align: middle;\n          font-family: roboto;\n          line-height: 16px;\n          margin-bottom: 8px;\n          margin-top: 15px;\n          margin-bottom: 15px;\n        }\n        .month {\n          color: var(--simple-colors-default-theme-accent-12);\n          font-weight: 400;\n          font-size: 1em;\n          text-transform: uppercase;\n        }\n        .dateNumber {\n          font-size: 20px;\n        }\n        .dayName {\n          text-transform: uppercase;\n        }\n        .monthSection {\n          background-color: var(--simple-colors-default-theme-accent-5);\n          border: 1px solid var(--simple-colors-default-theme-accent-1);\n          border-radius: 10px 10px 0px 0px;\n        }\n        .details {\n          border: 2px solid var(--simple-colors-default-theme-accent-5);\n          padding: 4px 16px;\n          background-color: var(--simple-colors-default-theme-accent-1);\n        }\n        .title,\n        .time,\n        .location {\n          vertical-align: middle;\n          font-family: roboto;\n          line-height: 16px;\n          margin-bottom: 8px;\n        }\n        .title {\n          font-size: 20px;\n          color: var(--simple-colors-default-theme-accent-11);\n        }\n        .time {\n          font-size: 16px;\n          color: var(--simple-colors-default-theme-accent-10);\n        }\n        .location {\n          font-size: 16px;\n          color: var(--simple-colors-default-theme-accent-10);\n        }\n      "]);

    _templateObject14_15211020de5b11eaa3c04df0985e08b7 = function _templateObject14_15211020de5b11eaa3c04df0985e08b7() {
      return data;
    };

    return data;
  }

  function _templateObject13_15211020de5b11eaa3c04df0985e08b7() {
    var data = babelHelpers.taggedTemplateLiteral([""]);

    _templateObject13_15211020de5b11eaa3c04df0985e08b7 = function _templateObject13_15211020de5b11eaa3c04df0985e08b7() {
      return data;
    };

    return data;
  }

  function _templateObject12_15211020de5b11eaa3c04df0985e08b7() {
    var data = babelHelpers.taggedTemplateLiteral([""]);

    _templateObject12_15211020de5b11eaa3c04df0985e08b7 = function _templateObject12_15211020de5b11eaa3c04df0985e08b7() {
      return data;
    };

    return data;
  }

  function _templateObject11_15211020de5b11eaa3c04df0985e08b7() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                    <p class=\"location\">", "</p>\n                  "]);

    _templateObject11_15211020de5b11eaa3c04df0985e08b7 = function _templateObject11_15211020de5b11eaa3c04df0985e08b7() {
      return data;
    };

    return data;
  }

  function _templateObject10_15211020de5b11eaa3c04df0985e08b7() {
    var data = babelHelpers.taggedTemplateLiteral([""]);

    _templateObject10_15211020de5b11eaa3c04df0985e08b7 = function _templateObject10_15211020de5b11eaa3c04df0985e08b7() {
      return data;
    };

    return data;
  }

  function _templateObject9_15211020de5b11eaa3c04df0985e08b7() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                            at ", "\n                          "]);

    _templateObject9_15211020de5b11eaa3c04df0985e08b7 = function _templateObject9_15211020de5b11eaa3c04df0985e08b7() {
      return data;
    };

    return data;
  }

  function _templateObject8_15211020de5b11eaa3c04df0985e08b7() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                            from ", " - ", "\n                          "]);

    _templateObject8_15211020de5b11eaa3c04df0985e08b7 = function _templateObject8_15211020de5b11eaa3c04df0985e08b7() {
      return data;
    };

    return data;
  }

  function _templateObject7_15211020de5b11eaa3c04df0985e08b7() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                      ", "\n                    "]);

    _templateObject7_15211020de5b11eaa3c04df0985e08b7 = function _templateObject7_15211020de5b11eaa3c04df0985e08b7() {
      return data;
    };

    return data;
  }

  function _templateObject6_15211020de5b11eaa3c04df0985e08b7() {
    var data = babelHelpers.taggedTemplateLiteral([""]);

    _templateObject6_15211020de5b11eaa3c04df0985e08b7 = function _templateObject6_15211020de5b11eaa3c04df0985e08b7() {
      return data;
    };

    return data;
  }

  function _templateObject5_15211020de5b11eaa3c04df0985e08b7() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                      ", ",\n                    "]);

    _templateObject5_15211020de5b11eaa3c04df0985e08b7 = function _templateObject5_15211020de5b11eaa3c04df0985e08b7() {
      return data;
    };

    return data;
  }

  function _templateObject4_15211020de5b11eaa3c04df0985e08b7() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <span class=\"details\" .style=\"margin:", "px;\">\n              <p class=\"title\">", "</p>\n              <p class=\"time\">\n                ", "", "\n                ", "", "\n                ", "\n              </p>\n              ", "\n            </span>\n          "]);

    _templateObject4_15211020de5b11eaa3c04df0985e08b7 = function _templateObject4_15211020de5b11eaa3c04df0985e08b7() {
      return data;
    };

    return data;
  }

  function _templateObject3_15211020de5b11eaa3c04df0985e08b7() {
    var data = babelHelpers.taggedTemplateLiteral([""]);

    _templateObject3_15211020de5b11eaa3c04df0985e08b7 = function _templateObject3_15211020de5b11eaa3c04df0985e08b7() {
      return data;
    };

    return data;
  }

  function _templateObject2_15211020de5b11eaa3c04df0985e08b7() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                <p class=\"dayName\">", "</p>\n              "]);

    _templateObject2_15211020de5b11eaa3c04df0985e08b7 = function _templateObject2_15211020de5b11eaa3c04df0985e08b7() {
      return data;
    };

    return data;
  }

  function _templateObject_15211020de5b11eaa3c04df0985e08b7() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"card\" style=\"margin:", "px;\">\n        <div class=\"monthSection\">\n          <p class=\"month\"><b>", "</b></p>\n        </div>\n        <div class=\"dateSection\">\n          <p class=\"dateNumber\">\n            ", "<sup>", "</sup>\n          </p>\n          ", "\n        </div>\n      </div>\n      ", "\n    "]);

    _templateObject_15211020de5b11eaa3c04df0985e08b7 = function _templateObject_15211020de5b11eaa3c04df0985e08b7() {
      return data;
    };

    return data;
  }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  /**
   * `date-card`
   * `simple presentation of a date`
   * @litElement
   * @demo demo/index.html
   * @element date-card
   */
  var DateCard =
  /*#__PURE__*/
  function (_SimpleColors) {
    babelHelpers.inherits(DateCard, _SimpleColors);
    babelHelpers.createClass(DateCard, null, [{
      key: "tag",
      get: function get() {
        return "date-card";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(DateCard), "properties", this), {
          month: {
            type: String
          },
          date: {
            type: String
          },
          day: {
            type: String
          },
          title: {
            type: String
          },
          startTime: {
            type: String,
            attribute: "start-time"
          },
          endTime: {
            type: String,
            attribute: "end-time"
          },
          location: {
            type: String
          },
          borderSpacing: {
            type: Number,
            attribute: "border-spacing"
          }
        });
      }
    }, {
      key: "haxProperties",
      get: function get() {
        return {
          canScale: false,
          canPosition: true,
          canEditSource: false,
          gizmo: {
            title: "Date card",
            description: "Displays a calendar date with details",
            icon: "hax:calendar",
            color: "green",
            groups: ["Date"],
            handles: [],
            meta: {
              author: "McGarvelcuddy",
              owner: "elmsln"
            }
          },
          settings: {
            quick: [],
            configure: [{
              property: "month",
              title: "Month",
              description: "Month to display on card",
              inputMethod: "textfield"
            }, {
              property: "date",
              title: "Day of month",
              description: "Numeric date of the month",
              inputMethod: "number"
            }, {
              property: "day",
              title: "Day of week",
              description: "Textual day of the week",
              inputMethod: "textfield"
            }, {
              property: "title",
              title: "Title",
              description: "Name of the event",
              inputMethod: "textfield"
            }, {
              property: "location",
              title: "Location",
              description: "Where it is taking place",
              inputMethod: "textfield"
            }, {
              property: "startTime",
              title: "Start",
              description: "Start time",
              inputMethod: "textfield"
            }, {
              property: "endTime",
              title: "End",
              description: "Start time",
              inputMethod: "textfield"
            }, {
              property: "accentColor",
              title: "Accent Color",
              description: "An optional accent color.",
              inputMethod: "colorpicker"
            }, {
              property: "dark",
              title: "Dark Theme",
              description: "Enable Dark Theme",
              inputMethod: "boolean"
            }],
            advanced: []
          }
        };
      }
    }]);

    function DateCard() {
      var _this;

      babelHelpers.classCallCheck(this, DateCard);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(DateCard).call(this));
      _this.borderSpacing = 5;
      _this.accentColor = "light-blue";
      return _this;
    }

    babelHelpers.createClass(DateCard, [{
      key: "nth",
      value: function nth(d) {
        if (!d) {
          return "";
        }

        if (d > 3 && d < 21) return "th";

        switch (d % 10) {
          case 1:
            return "st";

          case 2:
            return "nd";

          case 3:
            return "rd";

          default:
            return "th";
        }
      }
    }, {
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_15211020de5b11eaa3c04df0985e08b7(), this.borderSpacing, this.month, this.date, this.nth(this.date), this.day ? (0, _litElement.html)(_templateObject2_15211020de5b11eaa3c04df0985e08b7(), this.day) : (0, _litElement.html)(_templateObject3_15211020de5b11eaa3c04df0985e08b7()), this.title ? (0, _litElement.html)(_templateObject4_15211020de5b11eaa3c04df0985e08b7(), this.borderSpacing, this.title, this.day ? (0, _litElement.html)(_templateObject5_15211020de5b11eaa3c04df0985e08b7(), this.day) : (0, _litElement.html)(_templateObject6_15211020de5b11eaa3c04df0985e08b7()), this.month, this.date, this.nth(this.date), this.startTime ? (0, _litElement.html)(_templateObject7_15211020de5b11eaa3c04df0985e08b7(), this.endTime ? (0, _litElement.html)(_templateObject8_15211020de5b11eaa3c04df0985e08b7(), this.startTime, this.endTime) : (0, _litElement.html)(_templateObject9_15211020de5b11eaa3c04df0985e08b7(), this.startTime)) : (0, _litElement.html)(_templateObject10_15211020de5b11eaa3c04df0985e08b7()), this.location ? (0, _litElement.html)(_templateObject11_15211020de5b11eaa3c04df0985e08b7(), this.location) : (0, _litElement.html)(_templateObject12_15211020de5b11eaa3c04df0985e08b7())) : (0, _litElement.html)(_templateObject13_15211020de5b11eaa3c04df0985e08b7()));
      }
    }], [{
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(DateCard), "styles", this)), [(0, _litElement.css)(_templateObject14_15211020de5b11eaa3c04df0985e08b7())]);
      }
    }]);
    return DateCard;
  }(_simpleColors.SimpleColors);

  _exports.DateCard = DateCard;
  customElements.define(DateCard.tag, DateCard);
});
