// @lrnwebcomponents/date-card/date-card.js
define(["exports", "meta", "../../lit/index.js", "../simple-colors/simple-colors.js"], function (_exports, meta, _index, _simpleColors) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.DateCard = void 0;
  meta = _interopRequireWildcard(meta);

  var _templateObject_a0d289d08e0011ed99ce491205e67b1d, _templateObject2_a0d289d08e0011ed99ce491205e67b1d, _templateObject3_a0d289d08e0011ed99ce491205e67b1d, _templateObject4_a0d289d08e0011ed99ce491205e67b1d, _templateObject5_a0d289d08e0011ed99ce491205e67b1d, _templateObject6_a0d289d08e0011ed99ce491205e67b1d, _templateObject7_a0d289d08e0011ed99ce491205e67b1d, _templateObject8_a0d289d08e0011ed99ce491205e67b1d, _templateObject9_a0d289d08e0011ed99ce491205e67b1d, _templateObject10_a0d289d08e0011ed99ce491205e67b1d, _templateObject11_a0d289d08e0011ed99ce491205e67b1d, _templateObject12_a0d289d08e0011ed99ce491205e67b1d, _templateObject13_a0d289d08e0011ed99ce491205e67b1d, _templateObject14_a0d289d08e0011ed99ce491205e67b1d;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `date-card`
   * `simple presentation of a date`
   * @litElement
   * @demo demo/index.html
   * @element date-card
   */
  var DateCard = /*#__PURE__*/function (_SimpleColors) {
    babelHelpers.inherits(DateCard, _SimpleColors);

    var _super = _createSuper(DateCard);

    function DateCard() {
      var _this;

      babelHelpers.classCallCheck(this, DateCard);
      _this = _super.call(this);
      _this.borderSpacing = 5;
      _this.accentColor = "light-blue";
      return _this;
    }

    babelHelpers.createClass(DateCard, [{
      key: "nth",
      value: function nth(d) {
        if (!d) {
          return "";
        }

        if (d > 3 && d < 21) return "th";

        switch (d % 10) {
          case 1:
            return "st";

          case 2:
            return "nd";

          case 3:
            return "rd";

          default:
            return "th";
        }
      }
    }, {
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_a0d289d08e0011ed99ce491205e67b1d || (_templateObject_a0d289d08e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"card\" style=\"margin:", "px;\">\n        <div class=\"monthSection\">\n          <p class=\"month\"><b>", "</b></p>\n        </div>\n        <div class=\"dateSection\">\n          <p class=\"dateNumber\">\n            ", "<sup>", "</sup>\n          </p>\n          ", "\n        </div>\n      </div>\n      ", "\n    "])), this.borderSpacing, this.month, this.date, this.nth(this.date), this.day ? (0, _index.html)(_templateObject2_a0d289d08e0011ed99ce491205e67b1d || (_templateObject2_a0d289d08e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral([" <p class=\"dayName\">", "</p> "])), this.day) : (0, _index.html)(_templateObject3_a0d289d08e0011ed99ce491205e67b1d || (_templateObject3_a0d289d08e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral([""]))), this.title ? (0, _index.html)(_templateObject4_a0d289d08e0011ed99ce491205e67b1d || (_templateObject4_a0d289d08e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral(["\n            <span class=\"details\" .style=\"margin:", "px;\">\n              <p class=\"title\">", "</p>\n              <p class=\"time\">\n                ", "", "\n                ", "", "\n                ", "\n              </p>\n              ", "\n            </span>\n          "])), this.borderSpacing, this.title, this.day ? (0, _index.html)(_templateObject5_a0d289d08e0011ed99ce491205e67b1d || (_templateObject5_a0d289d08e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral([" ", ", "])), this.day) : (0, _index.html)(_templateObject6_a0d289d08e0011ed99ce491205e67b1d || (_templateObject6_a0d289d08e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral([""]))), this.month, this.date, this.nth(this.date), this.startTime ? (0, _index.html)(_templateObject7_a0d289d08e0011ed99ce491205e67b1d || (_templateObject7_a0d289d08e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral(["\n                      ", "\n                    "])), this.endTime ? (0, _index.html)(_templateObject8_a0d289d08e0011ed99ce491205e67b1d || (_templateObject8_a0d289d08e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral([" from ", " - ", " "])), this.startTime, this.endTime) : (0, _index.html)(_templateObject9_a0d289d08e0011ed99ce491205e67b1d || (_templateObject9_a0d289d08e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral([" at ", " "])), this.startTime)) : (0, _index.html)(_templateObject10_a0d289d08e0011ed99ce491205e67b1d || (_templateObject10_a0d289d08e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral([""]))), this.location ? (0, _index.html)(_templateObject11_a0d289d08e0011ed99ce491205e67b1d || (_templateObject11_a0d289d08e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral([" <p class=\"location\">", "</p> "])), this.location) : (0, _index.html)(_templateObject12_a0d289d08e0011ed99ce491205e67b1d || (_templateObject12_a0d289d08e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral([""])))) : (0, _index.html)(_templateObject13_a0d289d08e0011ed99ce491205e67b1d || (_templateObject13_a0d289d08e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral([""]))));
      }
    }], [{
      key: "tag",
      get: function get() {
        return "date-card";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(DateCard), "properties", this)), {}, {
          month: {
            type: String
          },
          date: {
            type: String
          },
          day: {
            type: String
          },
          title: {
            type: String
          },
          startTime: {
            type: String,
            attribute: "start-time"
          },
          endTime: {
            type: String,
            attribute: "end-time"
          },
          location: {
            type: String
          },
          borderSpacing: {
            type: Number,
            attribute: "border-spacing"
          }
        });
      }
      /**
       * haxProperties integration via file reference
       */

    }, {
      key: "haxProperties",
      get: function get() {
        return new URL("./lib/".concat(this.tag, ".haxProperties.json"), meta.url).href;
      }
    }, {
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(DateCard), "styles", this)), [(0, _index.css)(_templateObject14_a0d289d08e0011ed99ce491205e67b1d || (_templateObject14_a0d289d08e0011ed99ce491205e67b1d = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: inline-flex;\n        }\n        .card {\n          box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);\n          width: var(--date-card-card-width, 100px);\n          border-radius: 10px;\n        }\n        .card:hover {\n          box-shadow: 0 8px 16px 0 rgba(0, 0, 0, 0.2);\n          border-radius: 10px;\n        }\n        .month,\n        .dateNumber,\n        .dayName {\n          text-align: center;\n          vertical-align: middle;\n          font-family: roboto;\n          line-height: 16px;\n          margin-bottom: 8px;\n          margin-top: 15px;\n          margin-bottom: 15px;\n        }\n        .month {\n          color: var(--simple-colors-default-theme-accent-12);\n          font-weight: 400;\n          font-size: 1em;\n          text-transform: uppercase;\n        }\n        .dateNumber {\n          font-size: 20px;\n        }\n        .dayName {\n          text-transform: uppercase;\n        }\n        .monthSection {\n          background-color: var(--simple-colors-default-theme-accent-5);\n          border: 1px solid var(--simple-colors-default-theme-accent-1);\n          border-radius: 10px 10px 0px 0px;\n        }\n        .details {\n          border: 2px solid var(--simple-colors-default-theme-accent-5);\n          padding: 4px 16px;\n          background-color: var(--simple-colors-default-theme-accent-1);\n        }\n        .title,\n        .time,\n        .location {\n          vertical-align: middle;\n          font-family: roboto;\n          line-height: 16px;\n          margin-bottom: 8px;\n        }\n        .title {\n          font-size: 20px;\n          color: var(--simple-colors-default-theme-accent-11);\n        }\n        .time {\n          font-size: 16px;\n          color: var(--simple-colors-default-theme-accent-10);\n        }\n        .location {\n          font-size: 16px;\n          color: var(--simple-colors-default-theme-accent-10);\n        }\n      "])))]);
      }
    }]);
    return DateCard;
  }(_simpleColors.SimpleColors);

  _exports.DateCard = DateCard;
  customElements.define(DateCard.tag, DateCard);
});
