// @lrnwebcomponents/lorem-data/lib/lorem-data-behaviors.js
define(["exports", "../../simple-colors/simple-colors.js", "../../simple-icon/lib/simple-iconset-demo.js", "../../simple-icon/simple-icon.js", "../../simple-icon/lib/simple-icon-lite.js", "../../simple-icon/lib/simple-iconset-manifest.js", "../../hax-iconset/lib/hax-iconset-manifest.js"], function (_exports, _simpleColors, _simpleIconsetDemo, _simpleIcon, _simpleIconLite, _simpleIconsetManifest, _haxIconsetManifest) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LoremDataBehaviors = void 0;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var LoremDataBehaviors = function LoremDataBehaviors(SuperClass) {
    return /*#__PURE__*/function (_SuperClass) {
      babelHelpers.inherits(_class, _SuperClass);

      var _super = _createSuper(_class);

      function _class() {
        babelHelpers.classCallCheck(this, _class);
        return _super.call(this);
      }

      babelHelpers.createClass(_class, [{
        key: "colors",
        get: function get() {
          var simple = window.SimpleColorsSharedStyles.requestAvailability();
          return simple && simple.colors ? Object.keys(simple.colors) : false;
        }
      }, {
        key: "icons",
        get: function get() {
          return [].concat(babelHelpers.toConsumableArray(_simpleIconsetManifest.SimpleIconIconsetsManifest || []), babelHelpers.toConsumableArray(_haxIconsetManifest.HaxIconsetManifest || [])).map(function (iconset) {
            return ((iconset || {}).icons || []).map(function (icon) {
              return iconset.name == "icons" ? icon : "".concat(iconset.name, ":").concat(icon);
            });
          }).flat();
        }
      }, {
        key: "text",
        get: function get() {
          return "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed consectetur at erat eu egestas. Cras porta volutpat vestibulum. Vestibulum eu quam fermentum, lobortis purus at, mattis massa. Sed eget velit rhoncus, ultricies metus eu, tincidunt justo. Ut sit amet congue justo, ac auctor mi. Etiam ut urna laoreet eros sodales commodo. Sed laoreet tortor ligula, vitae condimentum mi sagittis vitae. Nullam volutpat sodales ipsum vel dictum. Aenean dapibus arcu sit amet tellus placerat scelerisque sed nec dui. Maecenas laoreet maximus magna, quis consectetur massa tristique ut. Nunc rutrum elit in tincidunt commodo. Donec nec varius velit. Duis ut arcu pulvinar, venenatis felis pellentesque, egestas ante. Etiam urna dui, vestibulum at diam eget, mattis gravida ipsum. Nulla auctor, neque a ultricies laoreet, nulla diam laoreet nisl, nec rutrum ex lorem a purus. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Suspendisse gravida ante eu nisi viverra, ac porta risus laoreet. Nunc a justo vel nisl accumsan placerat eu eu libero. Donec non augue sit amet nisl pulvinar bibendum vulputate sit amet metus. Maecenas vel condimentum nulla, at elementum neque. Fusce ac dictum augue. Suspendisse potenti. Nullam eros dui, vehicula et imperdiet vel, vestibulum vel tortor. Nullam ac nunc ante. Pellentesque condimentum nec dui sed varius. Donec porttitor leo id lectus faucibus, in fermentum purus lacinia. Pellentesque vel tellus ut lectus tincidunt pharetra. Quisque elementum elit mattis mi dignissim, eleifend eleifend risus faucibus. Nullam interdum pulvinar tempor. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Etiam semper, diam sit amet condimentum interdum, dolor risus ultrices dui, et egestas ex magna at odio. Aliquam sed vulputate turpis. Integer suscipit laoreet lectus, in lobortis sapien. Ut aliquam, dui ac efficitur ullamcorper, sapien augue euismod sapien, eget consectetur odio lectus ac elit. Pellentesque varius metus vel placerat rutrum. Duis placerat diam ut tortor vestibulum, eu ultrices risus gravida. Ut vel auctor dolor. Donec sit amet ornare ipsum, nec pulvinar neque. Pellentesque sagittis orci bibendum, laoreet elit eget, vulputate ante. Donec accumsan in elit sed sollicitudin. Fusce pellentesque velit est, id vestibulum orci consectetur ac. In libero odio, vehicula non felis eu, scelerisque rhoncus arcu. Phasellus id tristique justo. Sed scelerisque quam velit, at porta leo accumsan in. Vestibulum non enim lacinia odio venenatis vehicula. Quisque ultrices augue nulla. Sed tempus porta ipsum, eu rhoncus ligula feugiat ac. Aenean et sollicitudin sem, a pharetra mi. Sed convallis velit in est molestie interdum. Fusce bibendum volutpat sapien vitae faucibus. Vivamus nec neque pellentesque, euismod elit quis, commodo est. Maecenas sed tristique augue. Phasellus facilisis fringilla sem in viverra. Sed porta in justo vitae dignissim. In cursus imperdiet elementum. Maecenas congue metus aliquam placerat semper. Mauris eu libero nunc. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Nam ultrices venenatis nibh, at lacinia nisi lacinia vitae. Quisque eu est scelerisque, mattis sem eu, rhoncus dui. Nulla facilisi. Etiam in quam a tellus mattis condimentum. Nullam facilisis, dolor ac consequat malesuada, diam felis lacinia ante, volutpat ullamcorper orci lacus non orci. Suspendisse sit amet consequat elit, ut vestibulum urna. Nam id dui sem. Suspendisse elementum ac orci vel auctor. Vivamus ornare accumsan sagittis. Nullam vitae est tempor, pretium eros nec, cursus nibh. Vivamus aliquet sodales elit. Proin non semper metus. Maecenas feugiat tincidunt iaculis. Integer lectus nisi, commodo et diam at, mattis varius nisi. Integer ultrices pharetra eleifend. Sed vulputate, ipsum congue ultricies venenatis, justo leo eleifend sem, quis lacinia urna orci in dolor. Etiam condimentum hendrerit sapien, in placerat velit porttitor lobortis. Integer non pellentesque ex. Mauris vestibulum tortor turpis, sed volutpat turpis gravida dapibus. Quisque finibus, purus sit amet aliquam consequat, sem sem ultricies risus, ut dapibus lacus elit non nulla. Mauris nisl justo, consequat et mauris tempus, placerat gravida risus. Maecenas semper vitae enim eu rutrum. Curabitur eget viverra nisl. Aliquam vitae diam risus. Sed nec nisi nec tortor convallis ultricies. Pellentesque lobortis euismod magna et malesuada. Praesent eget mi eros. Ut quis imperdiet odio. Maecenas a fermentum massa, nec semper justo. Integer a massa aliquam, vulputate est a, condimentum neque. Phasellus ligula elit, feugiat eget ante a, porttitor tempus diam. Vestibulum semper ante pretium, interdum elit ac, egestas dolor. Interdum et malesuada fames ac ante ipsum primis in faucibus. Ut consequat, nisi nec scelerisque suscipit, libero lectus dictum lectus, quis molestie sem ipsum in nisi. Curabitur finibus purus nec vehicula euismod. Duis in libero hendrerit, congue lacus vel, porta turpis. Aenean volutpat dolor faucibus varius fermentum. Proin nisi neque, condimentum id venenatis ut, fermentum id velit. Etiam pretium sem eget metus congue ultricies. Phasellus laoreet tortor eu lectus consequat viverra. Nulla eget dictum nunc, eu dictum turpis. Morbi sit amet erat et leo malesuada iaculis eget in ipsum. Vivamus eleifend orci at quam dignissim, eget congue lacus faucibus. Sed porta ultrices nibh, vitae efficitur arcu facilisis blandit. Etiam vitae ligula lobortis, tincidunt est quis, eleifend leo. Vestibulum feugiat odio non nisi vehicula dictum. Nam elementum mauris sed tortor faucibus vestibulum. Maecenas nec facilisis ligula. Quisque tellus enim, venenatis sed magna vitae, vehicula auctor orci. Suspendisse fringilla urna mauris, sit amet dictum lorem iaculis ut. Morbi tempus velit a tellus sagittis viverra. Integer ornare pharetra pharetra. Nunc sit amet ex at mi commodo mollis. Sed sed mi feugiat, semper nibh at, semper nisi. Fusce dapibus, nisi et placerat vehicula, nisi nibh maximus lorem, nec porttitor tellus eros eu metus. Proin tortor magna, cursus ultrices eleifend venenatis, tincidunt ut metus. Donec felis arcu, aliquam et odio sed, accumsan mollis nibh. Nulla nisi nunc, lobortis non convallis ac, porttitor non ligula. Suspendisse hendrerit tempor ante in congue. Nunc auctor libero sit amet lobortis efficitur. Suspendisse pretium efficitur finibus. Praesent rutrum justo in magna scelerisque mollis. In tempor sodales efficitur. Suspendisse tempus sagittis quam, fermentum vehicula odio iaculis at. Duis laoreet, neque eu aliquet venenatis, magna nisl sollicitudin risus, eu pulvinar diam ex vitae mauris. Sed eget diam justo. Quisque tristique venenatis justo nec tincidunt. Proin porta nisi nec facilisis iaculis. Ut euismod vel metus eu sodales. Pellentesque quis massa in nunc interdum ornare accumsan ac sapien. Aenean condimentum, erat nec ornare viverra, massa quam vulputate metus, at ultricies metus nunc ac dolor. In lorem nibh, luctus quis consectetur quis, ultrices quis massa. Sed ut vehicula tortor, vel blandit neque. Duis ac semper neque. Cras at faucibus tortor, sed ultricies tellus. Phasellus in lacus leo. Maecenas vehicula eu massa id ullamcorper. Duis a sem felis. Nunc fringilla felis in leo pulvinar, mattis lobortis nunc sollicitudin. Sed ut tincidunt risus. Nunc a viverra libero. Etiam sed libero cursus, porttitor purus et, aliquet nisi. Praesent ac tortor libero. Mauris vestibulum enim non turpis blandit, quis varius libero ultricies. Proin non sapien pellentesque, vulputate magna eget, dictum lacus. Curabitur tincidunt vestibulum nisl, at sodales tellus volutpat id. In nisl enim, accumsan id quam vitae, congue euismod ligula. Sed dictum luctus aliquet. Fusce consectetur porttitor imperdiet. Duis pretium sem sed nisi venenatis consequat. Aenean eget sapien id massa tincidunt dignissim vitae non tortor. Maecenas ac ligula diam. Morbi vitae bibendum turpis, sed rutrum eros. Vivamus ut mi facilisis, rutrum metus porta, condimentum urna. Vestibulum semper sed tellus vitae pellentesque. Sed porttitor nisl non felis semper, laoreet vulputate ipsum tristique. Nullam luctus sit amet orci id consectetur. Sed vel augue nisl. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque neque justo, interdum varius nulla et, aliquet pharetra nibh. Maecenas quis vehicula justo. Curabitur at elit justo. Quisque pharetra sapien vel bibendum mollis. Ut elementum ut ante a ornare. Cras suscipit, neque a suscipit laoreet, libero tellus lacinia nulla, hendrerit iaculis turpis nisi sed neque. Interdum et malesuada fames ac ante ipsum primis in faucibus. Quisque quis lacus vitae nibh consectetur lobortis. Proin varius fringilla scelerisque. Aliquam ac felis ut leo interdum fermentum. Nam accumsan risus sapien, ac maximus ex tincidunt ut. Nam ultrices et metus ut cursus. Suspendisse tempor erat pharetra nisl placerat, vitae scelerisque nisi lobortis. Suspendisse quis lorem metus. Ut nec dolor magna. Sed sem metus, hendrerit eget posuere sed, efficitur vitae urna. Sed pellentesque eu augue a tristique. Donec lacinia consequat mi non sollicitudin. In scelerisque nisi at eleifend aliquet. Donec tristique dui turpis, id congue est pharetra id. Praesent pretium libero sit amet semper aliquet. Phasellus fringilla ligula id augue accumsan, ac tempor sem consequat. Proin placerat arcu quam, eget feugiat quam dapibus a. Quisque eleifend purus ac erat fringilla laoreet. Ut sem diam, iaculis a lectus vel, volutpat eleifend lorem. Vivamus mi ligula, consectetur ut maximus in, sodales nec sem. Nam ac nulla ornare, interdum lectus sed, porta elit. Pellentesque consectetur dapibus purus. Phasellus lacinia leo at libero tincidunt maximus. Nulla facilisi. Sed feugiat sem id libero lobortis, ut scelerisque odio posuere. Pellentesque id risus nec metus bibendum consectetur. Integer ac sapien eget neque sagittis viverra. Nulla bibendum erat nisl, vitae fringilla lectus cursus ut. Ut est lorem, laoreet hendrerit nulla at, accumsan convallis quam. Praesent euismod nibh ac semper semper. Donec sit amet interdum tellus. Proin ultrices nisi augue. Sed aliquam scelerisque justo eleifend rhoncus. Mauris et leo tortor. Nunc est magna, laoreet in ex ut, sodales cursus ex. Nulla congue libero et diam euismod, ut vehicula enim egestas. Aenean et magna iaculis, placerat magna quis, hendrerit metus. Ut venenatis imperdiet risus a dignissim. Praesent lobortis consequat urna eget varius. Quisque eros nisi, dictum non luctus at, mattis et lorem. Donec elementum neque id massa ultrices, nec auctor purus sollicitudin. Etiam interdum porta massa, ut consectetur tellus lobortis malesuada. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Integer non ornare ante, eu fermentum dolor. Mauris quis venenatis arcu, eu consequat elit. Phasellus maximus placerat maximus. Ut velit turpis, consequat eu lacus at, hendrerit ultricies augue. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae; Donec non velit sapien. Vivamus sit amet sem finibus, sagittis nunc non, volutpat magna. Vivamus eget dui ut metus efficitur porttitor. Pellentesque semper vitae sem in fermentum. Maecenas sed ex massa. Fusce ac augue nec est iaculis semper. Donec aliquam, neque maximus ultricies facilisis, neque dui eleifend mi, ut tristique sem mi a risus. Quisque ut lorem quis tortor maximus condimentum suscipit nec nunc. Nulla vel enim tempus, posuere augue id, imperdiet nisl. Maecenas semper libero in felis semper, pellentesque ullamcorper nisi mollis. Vestibulum tincidunt lacinia tellus ac congue. Vivamus ultrices nisi sit amet ultricies suscipit. Nulla facilisi. Proin ultricies fermentum risus, in aliquet ligula lacinia at. Aliquam pretium faucibus erat ut varius. Morbi quam urna, mollis a velit sit amet, bibendum tempus dolor. Etiam ac tempus sapien. Etiam in rutrum purus. Nunc nec fermentum metus, sed placerat justo. Ut leo sapien, pellentesque at pulvinar vitae, volutpat ac elit. Phasellus non dolor consequat, tempor leo non, dictum nibh. Suspendisse scelerisque nisl quis sagittis facilisis. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Aenean libero eros, rutrum eu lorem eu, suscipit elementum justo. Cras viverra blandit dui in consectetur. Praesent erat est, molestie ut leo nec, efficitur placerat mauris. Nunc sollicitudin accumsan justo, sit amet interdum massa faucibus a. Pellentesque ac tempor odio. Fusce rutrum sagittis quam ut tincidunt. Aliquam consectetur luctus justo sit amet porttitor. Vivamus blandit porttitor massa, et ullamcorper lorem porttitor at. In tempus euismod est, ac ullamcorper mauris fermentum iaculis. Integer tincidunt sem id mauris tempus luctus. Phasellus sed egestas lacus. Curabitur luctus dolor arcu, aliquam convallis ligula dignissim sed. Phasellus scelerisque diam nec euismod finibus. Mauris feugiat, ipsum in condimentum blandit, justo orci ultrices neque, sed rutrum ligula magna ut nisl. Pellentesque tempus tincidunt turpis in mattis. Duis ut condimentum massa, ac cursus orci. Ut vel orci eget arcu bibendum mattis id sed felis. Morbi nisi risus, efficitur vel accumsan eget, feugiat sed ipsum. Vestibulum sit amet suscipit erat. Proin dignissim mauris eu varius suscipit. Sed consequat ex eros, eget fermentum enim interdum vel. Ut ullamcorper auctor est sed auctor. Donec lacinia turpis mauris, eget lacinia diam varius non. Vivamus vestibulum ullamcorper vestibulum. Cras quis quam at ex ultricies ornare vel eu.";
        }
      }, {
        key: "words",
        get: function get() {
          return this.text.toLowerCase().split(/\W+/);
        }
        /**
         * gets date x days from start date
         * @param {date} start starting date
         * @param {number} days number of weeks
         * @returns {date}
         */

      }, {
        key: "addDays",
        value: function addDays() {
          var start = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : new Date();
          var amt = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
          return new Date(Date.parse(start) + amt * 86400000);
        }
        /**
         * gets date x days from start date
         * @param {date} start starting date
         * @param {number} days number of weeks
         * @returns {date}
         */

      }, {
        key: "addHours",
        value: function addHours() {
          var start = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : new Date();
          var amt = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
          return new Date(Date.parse(start) + amt * 3600000);
        }
        /**
         * gets date x days from start date
         * @param {date} start starting date
         * @param {number} days number of weeks
         * @returns {date}
         */

      }, {
        key: "addMinutes",
        value: function addMinutes() {
          var start = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : new Date();
          var amt = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
          return new Date(Date.parse(start) + amt * 60000);
        }
        /**
         * gets date x days from start date
         * @param {date} start starting date
         * @param {number} days number of weeks
         * @returns {date}
         */

      }, {
        key: "addSeconds",
        value: function addSeconds() {
          var start = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : new Date();
          var amt = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
          return new Date(Date.parse(start) + amt * 1000);
        }
        /**
         * gets date x weeks from start date
         * @param {date} start starting date
         * @param {number} weeks number of weeks
         * @returns {date}
         */

      }, {
        key: "addWeeks",
        value: function addWeeks() {
          var start = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : new Date();
          var amt = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
          return new Date(Date.parse(start) + amt * 604800000);
        }
        /**
         * gets date x years from start date
         * @param {date} start starting date
         * @param {number} years number of years
         * @returns {date}
         */

      }, {
        key: "addYears",
        value: function addYears() {
          var start = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : new Date();
          var amt = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
          return new Date(Date.parse(start) + amt * 31536000000);
        }
        /**
         * convert camelcase to kebab (for converting properties in attributes)
         * @param {string} camel
         * @returns {string} kebab
         * @memberof StorybookUtilities
         */

      }, {
        key: "camelToKebab",
        value: function camelToKebab(camel) {
          return camel ? camel.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, "$1-$2").toLowerCase() : undefined;
        }
        /**
         * gets formatted date
         * @param {object} date
         * @param {format} long (Weekday, Month, Day, Year), short (Month, Day), or default (Month, Day, Year)
         * @returns {string} date as string
         */

      }, {
        key: "dateFormat",
        value: function dateFormat(d, format) {
          if (typeof d === "string") d = new Date(d);
          return !d ? "" : format === "long" ? d.toLocaleDateString(undefined, {
            weekday: "long",
            year: "numeric",
            month: "long",
            day: "numeric"
          }) : format === "short" ? d.toLocaleDateString(undefined, {
            month: "long",
            day: "numeric"
          }) : d.toLocaleDateString(undefined, {
            year: "numeric",
            month: "long",
            day: "numeric"
          });
        }
        /**
         * draws x-y items from shuffled array
         * @param {array} arr array
         * @param {number} min minimum number of items
         * @param {number} max max number of items
         * @returns {arr} shuffled array of x items
         */

      }, {
        key: "draw",
        value: function draw() {
          var arr = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
          var min = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
          var max = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : min;
          var range = max - min,
              rand = Math.random() * range,
              offset = Math.floor(rand),
              len = min + offset,
              slice = Math.max(0, len);
          return this.shuffle(arr).slice(0, slice);
        }
        /**
         * convert camelcase to kebab (for converting properties in attributes)
         * @param {string} camel
         * @returns {string} kebab
         * @memberof StorybookUtilities
         */

      }, {
        key: "kebabToCamel",
        value: function kebabToCamel(kebab) {
          return kebab ? kebab.replace(/-./g, function (x) {
            return x.toUpperCase()[1];
          }) : undefined;
        }
      }, {
        key: "randomArray",
        value: function randomArray() {
          var children = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
          var min = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 1;
          var max = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 4;
          var arr = [];

          for (var i = this.randomNumber(min, max, 1); i > 0; i--) {
            arr.push(this.randomType(children));
          }

          return arr;
        }
      }, {
        key: "randomAspect",
        value: function randomAspect() {
          var minwidth = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 200;
          var maxwidth = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 1000;
          var minheight = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 200;
          var maxheight = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 1000;
          var w = this.randomNumber(minwidth, maxwidth, 1),
              h = this.randomNumber(minheight, maxheight, 1);
          return "".concat(w, "/").concat(h);
        }
        /**
         * randomly true or false
         * @returns {boolean}
         */

      }, {
        key: "randomBool",
        value: function randomBool() {
          return Math.random() >= 0.5;
        }
        /**
         * random color from Simple Colors
         * @returns {string}
         */

      }, {
        key: "randomColor",
        value: function randomColor() {
          return this.randomOption(this.colors);
        }
        /**
         * random color from Simple Colors
         * @returns {string}
         */

      }, {
        key: "randomDate",
        value: function randomDate() {
          var start = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : Date.now();
          var units = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "years";
          var min = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : -3;
          var max = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 3;
          var amt = units === "milliseconds" ? this.randomNumber(min, max) : units === "seconds" ? this.randomNumber(min * 1000, max * 1000) : units === "minutes" ? this.randomNumber(min * 60000, max * 60000) : units === "hours" ? this.randomNumber(min * 3600000, max * 3600000) : units === "days" ? this.randomNumber(min * 86400000, max * 86400000) : units === "weeks" ? this.randomNumber(min * 604800000, max * 604800000) : this.randomNumber(min * 31536000000, max * 31536000000);
          return start + amt;
        }
        /**
         * random color as hex code
         * @returns {string}
         */

      }, {
        key: "randomHex",
        value: function randomHex() {
          var _this = this;

          return "#".concat([0, 0, 0].map(function (i) {
            return _this.randomNumber(i, 255).toString(16);
          }).map(function (i) {
            return i.length < 2 ? "0".concat(i) : "".concat(i);
          }).join(""));
        }
        /**
         * gets a random icon for an iconpicker
         *
         * @param {boolean} [includeNull=false] include a 50/50 change of no icon?
         * @returns string
         */

      }, {
        key: "randomIcon",
        value: function randomIcon() {
          var includeNull = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
          var random = this.randomOption(this.icons);
          return includeNull ? this.randomOption([].concat(babelHelpers.toConsumableArray(random), [""])) : random;
        }
        /**
         * generates random link data
         * @returns {object} link as { url, text, type }
         * */

      }, {
        key: "randomLink",
        value: function randomLink(type) {
          var min = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 1;
          var max = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 4;
          var w = this.draw(this.words, min, max),
              file = type || this.randomOption(["url", "pdf"]),
              base = this.randomOption(this.words),
              extension = file !== "url" ? ".".concat(file) : "",
              url = "http://".concat(base, ".com/").concat(w.join("/")).concat(extension),
              text = w.length > 0 ? w.join(" ") : base;
          return {
            url: url,
            text: text,
            type: file
          };
        }
      }, {
        key: "randomFlickr",
        value: function randomFlickr() {
          var aspect = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.randomAspect();
          var searchTerms = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
          var searchAll = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
          var multiple = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : -1;
          return "https://loremflickr.com/".concat(aspect).concat(searchTerms.length === 0 ? "" : "/".concat(searchTerms.join(",")).concat(searchAll ? "/all" : "")).concat(multiple > -1 ? "?random=".concat(multiple) : "");
        }
      }, {
        key: "randomImage",
        value: function randomImage() {
          var aspect = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.randomAspect();
          var greyscale = arguments.length > 1 ? arguments[1] : undefined;
          var topic = arguments.length > 2 ? arguments[2] : undefined;
          var multiple = arguments.length > 3 ? arguments[3] : undefined;
          return topic && ["man", "woman", "person"].includes(topic) ? this.randomProfileImage(aspect, topic) : babelHelpers.typeof(multiple) === (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined)) ? this.randomPlaceImg(aspect, greyscale, topic) : topic === "any" ? this.randomPicsum(aspect, greyscale, undefined, multiple) : this.randomFlickr(aspect, [topic], false, multiple);
        }
      }, {
        key: "randomImageData",
        value: function randomImageData() {
          var aspect = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.randomAspect();
          var greyscale = arguments.length > 1 ? arguments[1] : undefined;
          var topic = arguments.length > 2 ? arguments[2] : undefined;
          var multiple = arguments.length > 3 ? arguments[3] : undefined;
          return {
            src: this.randomImage(aspect, greyscale, topic, multiple),
            alt: "Random ".concat(topic ? "".concat(topic, " ") : "", "image").concat(!multiple ? "" : " #".concat(multiple)),
            longdesc: "This is a long description for image".concat(!multiple ? "" : " #".concat(multiple), ". ").concat(this.randomParagraph(1, 5))
          };
        }
      }, {
        key: "randomKitten",
        value: function randomKitten() {
          var aspect = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.randomAspect();
          var greyscale = arguments.length > 1 ? arguments[1] : undefined;
          if (babelHelpers.typeof(greyscale) === (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) greyscale = this.randomWeightedOption([{
            value: false,
            weight: 3
          }, {
            value: true,
            weight: 1
          }]);
          return "//placekitten.com".concat(greyscale ? "/g" : "", "/").concat(aspect);
        }
        /**
         * random number within a range
         * @param {number} min lowest value
         * @param {number} max highest value
         * @param {number} step
         * @returns {number}
         */

      }, {
        key: "randomNumber",
        value: function randomNumber() {
          var min = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
          var max = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 100 + min;
          var step = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 1;
          return min + Math.floor(Math.random() * Math.floor((max - min) / step)) * step;
        }
      }, {
        key: "randomObject",
        value: function randomObject() {
          var _this2 = this;

          var schema = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
          var obj = {};
          Object.keys(schema).forEach(function (key) {
            return obj[key] = _this2.randomType(schema[key]);
          });
          return obj;
        }
        /**
         * random option from and array of options
         * @param {array} options
         * @returns {*}
         */

      }, {
        key: "randomOption",
        value: function randomOption() {
          var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
          return options.length > 0 ? options[Math.floor(Math.random() * Math.floor(options.length))] : undefined;
        }
      }, {
        key: "randomPicsum",
        value: function randomPicsum() {
          var aspect = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.randomAspect();
          var greyscale = arguments.length > 1 ? arguments[1] : undefined;
          var blur = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;
          var id = arguments.length > 3 ? arguments[3] : undefined;
          var params = greyscale || this.randomWeightedOption([{
            value: false,
            weight: 3
          }, {
            value: true,
            weight: 1
          }]) ? ["greyscale"] : [];
          if (blur > 0) params.push("blur=".concat(blur));
          return "https://picsum.photos/".concat(id > -1 ? "id/".concat(id) : "", "/").concat(aspect, "/").concat(params.length == 0 ? "" : "?".concat(params.join("&")));
        }
      }, {
        key: "randomPlaceImg",
        value: function randomPlaceImg() {
          var aspect = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.randomAspect();
          var filter = arguments.length > 1 ? arguments[1] : undefined;
          var topic = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : "any";
          filter = filter || this.randomWeightedOption([{
            value: "",
            weight: 12
          }, {
            value: "greyscale",
            weight: 4
          }, {
            value: "sepia",
            weight: 1
          }]);
          return "//placeimg.com/".concat(aspect, "/").concat(topic).concat(filter === "greyscale" ? "/greyscale" : filter === "sepia" ? "/sepia" : "");
        }
      }, {
        key: "randomProfileImage",
        value: function randomProfileImage() {
          var aspect = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.randomAspect();
          var topic = arguments.length > 1 ? arguments[1] : undefined;
          var multiple = arguments.length > 2 ? arguments[2] : undefined;
          var aspects = aspect.split("/"),
              area = parseInt(aspects[0]) * parseInt(aspects[1]),
              size = area < 250 ? "thumb/" : area < 640 ? "med/" : "",
              random = Math.random() < 0.5 ? "men" : "women",
              person = topic === "man" ? "men" : topic === "woman" ? "women" : random;
          return "https://randomuser.me/api/portraits/".concat(size).concat(person, "/").concat(multiple || this.randomNumber(1, 90), ".jpg");
        }
      }, {
        key: "randomParagraph",
        value: function randomParagraph() {
          var min = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 3;
          var max = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 7;
          var wordMinPerSent = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 3;
          var wordMaxPerSent = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 10;
          var paragraph = [];

          for (var i = this.randomNumber(min, max); i > 0; i--) {
            paragraph.push(this.randomSentence(wordMinPerSent, wordMaxPerSent));
          }

          return "".concat(paragraph.join(" "));
        }
      }, {
        key: "randomPassage",
        value: function randomPassage() {
          var min = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 2;
          var max = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 5;
          var sentMinPerPara = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 2;
          var sentMaxPerPara = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 5;
          var wordMinPerSent = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : 3;
          var wordMaxPerSent = arguments.length > 5 && arguments[5] !== undefined ? arguments[5] : 10;
          var passage = [];

          for (var i = this.randomNumber(min, max); i > 0; i--) {
            passage.push("<p>".concat(this.randomParagraph(sentMinPerPara, sentMaxPerPara, wordMinPerSent, wordMaxPerSent), "</p>"));
          }

          return passage.join("");
        }
      }, {
        key: "randomPhrase",
        value: function randomPhrase() {
          var min = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;
          var max = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 5;
          var titlecase = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
          var colon = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : false;
          var phrase = [],
              wordCount = this.randomNumber(min, max),
              colonIndex = colon && wordCount > 1 ? this.randomNumber(1, wordCount - 1) : -1;

          for (var i = 0; i <= wordCount; i++) {
            var word = this.randomWord();
            phrase.push(i == colonIndex ? "".concat(word, ":") : word);
          }

          return titlecase ? this.titleCase(phrase.join(" ")) : phrase.join(" ");
        }
      }, {
        key: "randomSentence",
        value: function randomSentence() {
          var min = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 3;
          var max = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 15;
          var sentence = [],
              punctuation = this.randomWeightedOption([{
            value: ".",
            weight: 10
          }, {
            value: "?",
            weight: 3
          }, {
            value: "!",
            weight: 1
          }]);

          for (var i = this.randomNumber(min, max); i > 0; i--) {
            sentence.push(this.randomWord());
          }

          if (sentence[0]) sentence[0] = this.titleCase(sentence[0]);
          return "".concat(sentence.join(" ")).concat(punctuation);
        }
      }, {
        key: "randomType",
        value: function randomType(schema) {
          var val;
          if (schema.type) switch (schema.type) {
            case "array":
              val = this.randomArray(schema.children, schema.min, schema.max);
              break;

            case "boolean":
              val = this.randomBool();
              break;

            case "color":
              val = this.randomColor();
              break;

            case "data":
              return schema.data;
              break;

            case "date":
              val = this.randomDate(schema.start, schema.units, schema.min, schema.max);
              break;

            case "hex":
              val = this.randomHex();
              break;

            case "icon":
              val = this.randomIcon(schema.includeNull);
              break;

            case "image":
              val = this.randomImage(schema.aspect, schema.greyscale, schema.topic, schema.multiple);
              break;

            case "imageData":
              val = this.randomImageData(schema.aspect, schema.greyscale, schema.topic, schema.multiple);
              break;

            case "letter":
              val = this.randomOption("abcdefghijklmnopqrstuvwxyz");
              break;

            case "link":
              val = this.randomLink(schema.filetype, schema.minPath, schema.maxPath);
              break;

            case "number":
              val = this.randomNumber(schema.min, schema.max, schema.step);
              break;

            case "object":
              val = this.randomObject(schema.schema);
              break;

            case "option":
              return schema.weightedOptions ? this.randomWeightedOption(schema.weightedOptions) : this.randomOption(schema.options);
              break;

            case "paragraph":
              val = this.randomParagraph(schema.min, schema.max, schema.wordMinPerSent, schema.wordMaxPerSent);
              break;

            case "sentence":
              val = this.randomSentence(schema.min, schema.max);
              break;

            case "word":
              val = this.randomWord();
              break;

            default:
              val = this.randomWord();
              break;
          }
          return val;
        }
      }, {
        key: "randomWeightedOption",
        value: function randomWeightedOption() {
          var arr = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
          return this.randomOption(arr.map(function (item) {
            var arr2 = [];

            for (var i = 0; i < item.weight; i++) {
              arr2.push(item.value);
            }

            return arr2;
          }).flat());
        }
      }, {
        key: "randomWord",
        value: function randomWord() {
          return this.randomOption(this.words);
        }
        /**
         * gets shuffled array
         * @param {array} arr array
         * @returns {arr} shuffled array
         */

      }, {
        key: "shuffle",
        value: function shuffle() {
          var arr = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
          return arr.sort(function (a, b) {
            return Math.random() - Math.random();
          });
        }
        /**
         * sorts array by most recent (or by oldest)
         * @param {array} arr array
         * @param {boolean} sort by most recent? (default is true)
         * @returns {arr} sorted array
         */

      }, {
        key: "sortDates",
        value: function sortDates(arr) {
          var oldest = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
          return arr.sort(function (a, b) {
            var aa = typeof a.date === "string" ? Date.parse(a.date) : a.date,
                bb = typeof b.date === "string" ? Date.parse(b.date) : b.date;
            return !oldest ? bb - aa : aa - bb;
          });
        }
      }, {
        key: "titleCase",
        value: function titleCase(str) {
          return str.toLowerCase().split(" ").map(function (word) {
            return word.replace(word[0], word[0].toUpperCase());
          }).join(" ");
        }
      }], [{
        key: "properties",
        get: function get() {
          return {
            __demo: {
              type: Object
            }
          };
        }
      }]);
      return _class;
    }(SuperClass);
  };

  _exports.LoremDataBehaviors = LoremDataBehaviors;
});
