// @lrnwebcomponents/simple-fields/lib/simple-fields-fieldset.js
define(["exports","../../../lit-element/lit-element.js"],(function(e,i){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.SimpleFieldsFieldset=void 0;class SimpleFieldsFieldset extends i.LitElement{static get tag(){return"simple-fields-fieldset"}static get styles(){return[i.css`
        fieldset {
          padding: var(--simple-fields-margin-small, 8px)
            var(--simple-fields-margin, 16px);
          margin: var(--simple-fields-margin-small, 8px) 0
            var(--simple-fields-margin, 16px);
          border: 1px solid var(--simple-fields-border-color-light, #ccc);
          border-radius: var(--simple-fields-border-radius, 2px);
          transition: all 0.3s ease-in-out;
        }
        :host(:last-of-type) {
          margin-bottom: 0;
        }
        #label {
          font-family: var(--simple-fields-font-family, sans-serif);
          font-size: var(--simple-fields-font-size, 16px);
          line-height: var(--simple-fields-line-height, 22px);
        }
        :host([error]) #label {
          color: var(--simple-fields-error-color, #dd2c00);
          transition: all 0.3s ease-in-out;
        }
        #description {
          font-family: var(--simple-fields-detail-font-family, sans-serif);
          font-size: var(--simple-fields-detail-font-size, 12px);
          line-height: var(--simple-fields-detail-line-height, 22px);
        }
      `]}render(){return i.html`
      <fieldset>
        ${this.legend} ${this.desc} ${this.fields}
      </fieldset>
    `}get legend(){return i.html`
      <legend id="label" ?hidden="${!this.label}">
        ${this.label}${this.error?"*":""}
      </legend>
    `}get desc(){return i.html`
      <div id="description" ?hidden="${!this.description}">
        ${this.description}
      </div>
    `}get fields(){return i.html`
      <div id="item-fields">
        <slot></slot>
      </div>
    `}static get properties(){return{disabled:{type:Boolean,reflect:!0},error:{type:Boolean,reflect:!0},hidden:{type:Boolean,reflect:!0},readonly:{type:Boolean,reflect:!0},label:{type:String,reflect:!0},name:{type:String,reflect:!0,attribute:"name"},description:{type:String}}}}e.SimpleFieldsFieldset=SimpleFieldsFieldset,window.customElements.define(SimpleFieldsFieldset.tag,SimpleFieldsFieldset)}));
