// @lrnwebcomponents/sites-listing/sites-listing.umd.js
import * as _require from 'require';
!function (e, n) {
  "object" == (typeof exports === "undefined" ? "undefined" : babelHelpers.typeof(exports)) && "undefined" != typeof module ? n(exports, require("@polymer/polymer/polymer-element.js"), require("@polymer/iron-ajax/iron-ajax.js"), require("@polymer/iron-list/iron-list.js"), require("@lrnwebcomponents/sites-listing/lib/site-card.js"), require("@polymer/paper-button/paper-button.js")) : "function" == typeof define && define.amd ? define(["exports", "@polymer/polymer/polymer-element.js", "@polymer/iron-ajax/iron-ajax.js", "@polymer/iron-list/iron-list.js", "@lrnwebcomponents/sites-listing/lib/site-card.js", "@polymer/paper-button/paper-button.js"], n) : n((e = e || self).SitesListing = {}, e.polymerElement_js);
}(this, function (e, n) {
  "use strict";

  function t(e) {
    return (t = "function" == typeof Symbol && "symbol" == babelHelpers.typeof(Symbol.iterator) ? function (e) {
      return babelHelpers.typeof(e);
    } : function (e) {
      return e && "function" == typeof Symbol && e.constructor === Symbol && e !== Symbol.prototype ? "symbol" : babelHelpers.typeof(e);
    })(e);
  }

  function i(e, n) {
    for (var t = 0; t < n.length; t++) {
      var i = n[t];
      i.enumerable = i.enumerable || !1, i.configurable = !0, "value" in i && (i.writable = !0), Object.defineProperty(e, i.key, i);
    }
  }

  function o(e) {
    return (o = Object.setPrototypeOf ? Object.getPrototypeOf : function (e) {
      return e.__proto__ || Object.getPrototypeOf(e);
    })(e);
  }

  function a(e, n) {
    return (a = Object.setPrototypeOf || function (e, n) {
      return e.__proto__ = n, e;
    })(e, n);
  }

  function r(e, n) {
    return !n || "object" != babelHelpers.typeof(n) && "function" != typeof n ? function (e) {
      if (void 0 === e) throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
      return e;
    }(e) : n;
  }

  function s(e, n, t) {
    return (s = "undefined" != typeof Reflect && Reflect.get ? Reflect.get : function (e, n, t) {
      var i = function (e, n) {
        for (; !Object.prototype.hasOwnProperty.call(e, n) && null !== (e = o(e));) {
          ;
        }

        return e;
      }(e, n);

      if (i) {
        var a = Object.getOwnPropertyDescriptor(i, n);
        return a.get ? a.get.call(t) : a.value;
      }
    })(e, n, t || e);
  }

  function l() {
    var e,
        n,
        t = (e = ['\n      <style>\n        :host {\n          height: 100vh;\n          display: flex;\n          flex-direction: column;\n        }\n        iron-list {\n          flex: 1 1 auto;\n        }\n        #loading {\n          width: 100%;\n          z-index: 1000;\n          opacity: 0.8;\n          padding: 16px;\n          text-align: center;\n          align-content: center;\n          justify-content: center;\n          height: 100vh;\n          position: absolute;\n          background-color: rgba(250, 250, 250, 0.8);\n          transition: all linear 0.8s;\n          visibility: visible;\n        }\n        #loading div {\n          font-size: 32px;\n          font-weight: bold;\n          padding: 16px;\n        }\n        #loading[data-loading] {\n          background-color: rgba(0, 0, 0, 0);\n          opacity: 0;\n          visibility: hidden;\n        }\n        site-card {\n          padding: 16px;\n          font-size: 16px;\n        }\n        paper-button.site-card-wrapper {\n          margin: 0;\n          padding: 0;\n        }\n      </style>\n      <iron-ajax\n        id="loaddata"\n        auto=""\n        loading="{{__loading}}"\n        url="[[dataSource]]"\n        handle-as="json"\n        debounce-duration="250"\n        last-response="{{sitesResponse}}"\n      ></iron-ajax>\n      <div id="loading" data-loading$="[[!__loading]]">\n        <elmsln-loading size="large"></elmsln-loading>\n        <div>Loading..</div>\n      </div>\n      <iron-list id="list" items="[[sites]]" as="site" grid="">\n        <template>\n          <paper-button\n            on-focusin="_mouseEnter"\n            on-focusout="_mouseLeave"\n            data-site-id$="[[site.id]]"\n            class="site-card-wrapper"\n            on-click="_siteClicked"\n          >\n            <site-card\n              data-site-id$="[[site.id]]"\n              size="[[size]]"\n              image="[[site.metadata.image]]"\n              icon="[[site.metadata.icon]]"\n              name="[[site.title]]"\n              title="[[site.description]]"\n              elevation="2"\n            ></site-card>\n          </paper-button>\n        </template>\n      </iron-list>\n    '], (n = ['\n      <style>\n        :host {\n          height: 100vh;\n          display: flex;\n          flex-direction: column;\n        }\n        iron-list {\n          flex: 1 1 auto;\n        }\n        #loading {\n          width: 100%;\n          z-index: 1000;\n          opacity: 0.8;\n          padding: 16px;\n          text-align: center;\n          align-content: center;\n          justify-content: center;\n          height: 100vh;\n          position: absolute;\n          background-color: rgba(250, 250, 250, 0.8);\n          transition: all linear 0.8s;\n          visibility: visible;\n        }\n        #loading div {\n          font-size: 32px;\n          font-weight: bold;\n          padding: 16px;\n        }\n        #loading[data-loading] {\n          background-color: rgba(0, 0, 0, 0);\n          opacity: 0;\n          visibility: hidden;\n        }\n        site-card {\n          padding: 16px;\n          font-size: 16px;\n        }\n        paper-button.site-card-wrapper {\n          margin: 0;\n          padding: 0;\n        }\n      </style>\n      <iron-ajax\n        id="loaddata"\n        auto=""\n        loading="{{__loading}}"\n        url="[[dataSource]]"\n        handle-as="json"\n        debounce-duration="250"\n        last-response="{{sitesResponse}}"\n      ></iron-ajax>\n      <div id="loading" data-loading\\$="[[!__loading]]">\n        <elmsln-loading size="large"></elmsln-loading>\n        <div>Loading..</div>\n      </div>\n      <iron-list id="list" items="[[sites]]" as="site" grid="">\n        <template>\n          <paper-button\n            on-focusin="_mouseEnter"\n            on-focusout="_mouseLeave"\n            data-site-id\\$="[[site.id]]"\n            class="site-card-wrapper"\n            on-click="_siteClicked"\n          >\n            <site-card\n              data-site-id\\$="[[site.id]]"\n              size="[[size]]"\n              image="[[site.metadata.image]]"\n              icon="[[site.metadata.icon]]"\n              name="[[site.title]]"\n              title="[[site.description]]"\n              elevation="2"\n            ></site-card>\n          </paper-button>\n        </template>\n      </iron-list>\n    ']) || (n = e.slice(0)), Object.freeze(Object.defineProperties(e, {
      raw: {
        value: Object.freeze(n)
      }
    })));
    return l = function l() {
      return t;
    }, t;
  }

  var d = function (e) {
    function d() {
      var e;
      return function (e, n) {
        if (!babelHelpers.instanceof(e, n)) throw new TypeError("Cannot call a class as a function");
      }(this, d), e = r(this, o(d).call(this)), new Promise(function (res, rej) {
        return _require.default(["../elmsln-loading/elmsln-loading.js"], res, rej);
      }), e;
    }

    var c, u, p;
    return function (e, n) {
      if ("function" != typeof n && null !== n) throw new TypeError("Super expression must either be null or a function");
      e.prototype = Object.create(n && n.prototype, {
        constructor: {
          value: e,
          writable: !0,
          configurable: !0
        }
      }), n && a(e, n);
    }(d, n.PolymerElement), c = d, p = [{
      key: "template",
      get: function get() {
        return n.html(l());
      }
    }, {
      key: "tag",
      get: function get() {
        return "sites-listing";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          sitesResponse: {
            type: Object,
            notify: !0,
            observer: "_sitesResponseChanged"
          },
          sites: {
            type: Array,
            notify: !0
          },
          size: {
            type: String,
            value: "large"
          },
          dataSource: {
            type: String
          },
          loadLocation: {
            type: Boolean,
            value: !1
          }
        };
      }
    }], (u = [{
      key: "connectedCallback",
      value: function value() {
        s(o(d.prototype), "connectedCallback", this).call(this), window.addEventListener("sites-listing-refresh-data", this.refreshData.bind(this));
      }
    }, {
      key: "disconnectedCallback",
      value: function value() {
        window.removeEventListener("sites-listing-refresh-data", this.refreshData.bind(this)), s(o(d.prototype), "disconnectedCallback", this).call(this);
      }
    }, {
      key: "refreshData",
      value: function value(e) {
        this.shadowRoot.querySelector("#loaddata").generateRequest();
      }
    }, {
      key: "_sitesResponseChanged",
      value: function value(e, n) {
        e && "undefined" !== t(e.items) && (this.set("sites", []), this.set("sites", e.items), this.notifyPath("sites.*"));
      }
    }, {
      key: "_siteClicked",
      value: function value(e) {
        var n = e.target.getAttribute("data-site-id"),
            i = this.sites.filter(function (e) {
          return e.id === n;
        });
        i.length > 0 && (i = i.pop()), this.loadLocation && "undefined" !== t(i.location) && (window.location.href = i.location), this.dispatchEvent(new CustomEvent("sites-listing-item-selected", {
          bubbles: !0,
          cancelable: !0,
          composed: !0,
          detail: i
        }));
      }
    }, {
      key: "_mouseEnter",
      value: function value(e) {
        var n = e.target.querySelectorAll("site-card")[0];
        n.__oldElevation = n.elevation, n.elevation + 2 > 5 ? n.elevation = 5 : n.elevation += 2;
      }
    }, {
      key: "_mouseLeave",
      value: function value(e) {
        var n = e.target.querySelectorAll("site-card")[0];
        n.elevation = n.__oldElevation;
      }
    }]) && i(c.prototype, u), p && i(c, p), d;
  }();

  window.customElements.define(d.tag, d), e.SitesListing = d, Object.defineProperty(e, "__esModule", {
    value: !0
  });
});
