// @lrnwebcomponents/elmsln-apps/lib/game-show-scoreboard/game-show-scoreboard.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@polymer/polymer/lib/elements/custom-style.js", "../../../../@polymer/app-route/app-location.js", "../../../../@polymer/app-route/app-route.js", "../../../../@polymer/iron-ajax/iron-ajax.js", "../../../../@polymer/paper-toggle-button/paper-toggle-button.js", "../../../../@polymer/paper-item/paper-item.js", "../../../../@polymer/paper-button/paper-button.js", "../../../../@polymer/paper-input/paper-input.js", "../../../../@polymer/paper-icon-button/paper-icon-button.js", "../../../../@polymer/paper-tooltip/paper-tooltip.js", "../../../../@vaadin/vaadin-grid/vaadin-grid.js", "../../../../@vaadin/vaadin-grid/vaadin-grid-filter.js", "../../../../@vaadin/vaadin-grid/vaadin-grid-sorter.js", "../../../../@vaadin/vaadin-grid/vaadin-grid-sort-column.js", "../../../../@vaadin/vaadin-grid/vaadin-grid-column-group.js", "../../../../@vaadin/vaadin-grid/vaadin-grid-selection-column.js", "../../../simple-datetime/simple-datetime.js", "../../../elmsln-loading/elmsln-loading.js", "../../../simple-picker/simple-picker.js", "../../../lrndesign-avatar/lrndesign-avatar.js"], function (_exports, _litElement, _customStyle, _appLocation, _appRoute, _ironAjax, _paperToggleButton, _paperItem, _paperButton, _paperInput, _paperIconButton, _paperTooltip, _vaadinGrid, _vaadinGridFilter, _vaadinGridSorter, _vaadinGridSortColumn, _vaadinGridColumnGroup, _vaadinGridSelectionColumn, _simpleDatetime, _elmslnLoading, _simplePicker, _lrndesignAvatar) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.GameShowScoreboard = void 0;

  function _templateObject2_007eb0201bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          align-content: center;\n          padding: 0.8em;\n        }\n        vaadin-grid-table-body > vaadin-grid-cell-content {\n          height: unset !important;\n        }\n        vaadin-grid {\n          height: 75vh;\n          font-family: Roboto, sans-serif;\n          --divider-color: rgba(0, 0, 0, var(--dark-divider-opacity));\n        }\n\n        vaadin-grid .cell {\n          overflow: hidden;\n          text-overflow: ellipsis;\n          padding-right: 56px;\n        }\n\n        vaadin-grid .cell.last {\n          padding-right: 24px;\n        }\n\n        vaadin-grid .cell.numeric {\n          text-align: right;\n        }\n\n        vaadin-grid paper-checkbox {\n          --primary-color: var(--paper-indigo-500);\n          margin: 0 24px;\n        }\n\n        vaadin-grid vaadin-grid-sorter .cell {\n          flex: 1;\n          display: flex;\n          justify-content: space-between;\n          align-items: center;\n        }\n\n        vaadin-grid vaadin-grid-sorter iron-icon {\n          transform: scale(0.8);\n        }\n\n        vaadin-grid vaadin-grid-sorter:not([direction]) iron-icon {\n          color: rgba(0, 0, 0, var(--dark-disabled-opacity));\n        }\n\n        vaadin-grid vaadin-grid-sorter[direction] {\n          color: rgba(0, 0, 0, var(--dark-primary-opacity));\n        }\n\n        vaadin-grid vaadin-grid-sorter[direction=\"desc\"] iron-icon {\n          transform: scale(0.8) rotate(180deg);\n        }\n        vaadin-grid-sorter {\n          text-align: center;\n        }\n        lrndesign-avatar {\n          display: inline-block;\n        }\n        .avatar-label {\n          display: inline-block;\n          margin-left: 0.2em;\n        }\n        simple-datetime,\n        .score {\n          margin: 0;\n          padding: 8px;\n          line-height: 14px;\n          font-size: 14px;\n          color: black;\n          text-align: center;\n          border-bottom: 1px dashed black;\n        }\n        simple-datetime:hover,\n        .score:hover {\n          background-color: #dddddd;\n        }\n      "]);

    _templateObject2_007eb0201bb311eabd7cc30119585ddb = function _templateObject2_007eb0201bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject_007eb0201bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <custom-style>\n        <style>\n          vaadin-grid {\n            --vaadin-grid-cell: {\n              padding: 0;\n            }\n\n            --vaadin-grid-header-cell: {\n              height: 3.5em;\n              color: rgba(0, 0, 0, var(--dark-secondary-opacity));\n              font-size: 1em;\n            }\n\n            --vaadin-grid-body-cell: {\n              height: 3em;\n              color: rgba(0, 0, 0, var(--dark-primary-opacity));\n              font-size: 0.8em;\n            }\n\n            --vaadin-grid-body-row-hover-cell: {\n              background-color: var(--paper-grey-200);\n            }\n\n            --vaadin-grid-body-row-selected-cell: {\n              background-color: var(--paper-grey-100);\n            }\n\n            --vaadin-grid-focused-cell: {\n              box-shadow: none;\n              font-weight: bold;\n            }\n          }\n        </style>\n      </custom-style>\n      <app-location\n        .route=\"", "\"\n        @route-changed=\"", "\"\n      ></app-location>\n      <app-route\n        pattern=\"", "/submissions/:submission\"\n        .route=\"", "\"\n        @route-changed=\"", "\"\n        .data=\"", "\"\n        @data-changed=\"", "\"\n        .tail=\"", "\"\n        @tail-changed=\"", "\"\n      >\n      </app-route>\n      <iron-ajax\n        auto\n        id=\"optionsrequest\"\n        url=\"", "\"\n        handle-as=\"json\"\n        @last-response-changed=\"", "\"\n      ></iron-ajax>\n      <simple-picker\n        id=\"section\"\n        label=\"Section\"\n        value=\"", "\"\n        @value-changed=\"", "\"\n        .options=\"", "\"\n      ></simple-picker>\n      <simple-picker\n        id=\"game\"\n        label=\"Game\"\n        value=\"", "\"\n        @value-changed=\"", "\"\n        .options=\"", "\"\n      ></simple-picker>\n      <iron-ajax\n        auto\n        id=\"datarequest\"\n        url=\"", "\"\n        handle-as=\"json\"\n        @last-response-changed=\"", "\"\n      ></iron-ajax>\n      <vaadin-grid\n        ?hidden=\"", "\"\n        id=\"material\"\n        aria-label=\"Student project list\"\n        .items=\"", "\"\n        theme=\"row-dividers\"\n        column-reordering-allowed\n        multi-sort\n      >\n        <vaadin-grid-column resizable>\n          <template class=\"header\">\n            <vaadin-grid-sorter id=\"sorter\" path=\"sis.sortable_name\"\n              >Student</vaadin-grid-sorter\n            >\n          </template>\n          <template>\n            <lrndesign-avatar\n              label=\"[[item.name]]\"\n              src=\"[[item.avatar]]\"\n            ></lrndesign-avatar>\n            <span class=\"avatar-label\">[[item.sis.sortable_name]]</span>\n          </template>\n          <template class=\"footer\">\n            <vaadin-grid-filter\n              aria-label=\"Student\"\n              path=\"sis.sortable_name\"\n              value=\"[[_filterName]]\"\n            >\n              <paper-input\n                slot=\"filter\"\n                label=\"Student\"\n                value=\"{{_filterName::input}}\"\n                focus-target\n              ></paper-input>\n            </vaadin-grid-filter>\n          </template>\n        </vaadin-grid-column>\n        <vaadin-grid-column resizable>\n          <template class=\"header\">\n            <vaadin-grid-sorter id=\"sorter\" path=\"name\"\n              >Name</vaadin-grid-sorter\n            >\n          </template>\n          <template>\n            [[item.name]]\n          </template>\n          <template class=\"footer\">\n            <vaadin-grid-filter\n              aria-label=\"Student\"\n              path=\"name\"\n              value=\"[[_filterUserName]]\"\n            >\n              <paper-input\n                slot=\"filter\"\n                label=\"Username\"\n                value=\"{{_filterUserName::input}}\"\n                focus-target\n              ></paper-input>\n            </vaadin-grid-filter>\n          </template>\n        </vaadin-grid-column>\n        <vaadin-grid-sort-column\n          resizable\n          width=\"1em\"\n          header=\"Game\"\n          path=\"game\"\n        >\n          [[item.game]]\n        </vaadin-grid-sort-column>\n        <vaadin-grid-sort-column\n          resizable\n          width=\"1em\"\n          header=\"Section\"\n          path=\"section\"\n        >\n          [[item.section]]\n        </vaadin-grid-sort-column>\n        <vaadin-grid-sort-column\n          resizable\n          width=\"1em\"\n          header=\"High score\"\n          path=\"high\"\n        >\n          [[item.high]]\n        </vaadin-grid-sort-column>\n        <vaadin-grid-column\n          resizable\n          width=\"1em\"\n          header=\"All Scores\"\n          path=\"scores\"\n        >\n          <template>\n            <dom-repeat items=\"[[item.scores]]\" as=\"score\">\n              <template>\n                <div class=\"score\">[[score]]</div>\n              </template>\n            </dom-repeat>\n          </template>\n        </vaadin-grid-column>\n        <vaadin-grid-column\n          resizable\n          width=\"1em\"\n          header=\"Attempt Dates\"\n          path=\"dates\"\n        >\n          <template>\n            <dom-repeat items=\"[[item.dates]]\" as=\"date\" mutable-data>\n              <template>\n                <simple-datetime\n                  format=\"M jS, Y h:i:s\"\n                  timestamp=\"[[date]]\"\n                  unix\n                ></simple-datetime>\n              </template>\n            </dom-repeat>\n          </template>\n        </vaadin-grid-column>\n      </vaadin-grid>\n    "]);

    _templateObject_007eb0201bb311eabd7cc30119585ddb = function _templateObject_007eb0201bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  var GameShowScoreboard =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(GameShowScoreboard, _LitElement);
    babelHelpers.createClass(GameShowScoreboard, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_007eb0201bb311eabd7cc30119585ddb(), this.route, this.routeChangedEvent, this.endPoint, this.route, this.routeChangedEvent, this.data, this.dataChangedEvent, this.tail, this.tailChangedEvent, this.optionsPath, this.optionsDataChangedEvent, this.section, this.sectionEvent, this.sectionOptions, this.game, this.gameEvent, this.gameOptions, this.dataRequestUrl, this.activeDataEvent, !this.visibleData, this._toArray(this.visibleData));
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_007eb0201bb311eabd7cc30119585ddb())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "game-show-scoreboard";
      }
      /**
       * props
       */

    }, {
      key: "properties",
      get: function get() {
        return {
          elmslnCourse: {
            type: String,
            attribute: "elmsln-course"
          },
          elmslnSection: {
            type: String,
            attribute: "elmsln-section"
          },
          basePath: {
            type: String,
            attribute: "base-path"
          },
          csrfToken: {
            type: String,
            attribute: "csrf-token"
          },
          endPoint: {
            type: String,
            attribute: "end-point"
          },
          optionsPath: {
            type: String,
            attribute: "options-path"
          },
          dataPath: {
            type: String,
            attribute: "data-path"
          },
          section: {
            type: String
          },
          game: {
            type: String
          },
          dataRequestUrl: {
            type: String,
            attribute: "data-request-url"
          },

          /**
           * routing variable for url
           */
          route: {
            type: String
          },
          tail: {
            type: String
          },
          optionsData: {
            type: Object
          },
          sectionOptions: {
            type: Array
          },
          gameOptions: {
            type: Array
          },
          activeData: {
            type: Object
          },
          visibleData: {
            type: Array
          }
        };
      }
    }]);

    function GameShowScoreboard() {
      var _this;

      babelHelpers.classCallCheck(this, GameShowScoreboard);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(GameShowScoreboard).call(this));
      _this.game = "";
      _this.optionsData = {};
      _this.sectionOptions = [];
      _this.gameOptions = [];
      _this.activeData = {};
      _this.visibleData = [];
      return _this;
    }

    babelHelpers.createClass(GameShowScoreboard, [{
      key: "sectionEvent",
      value: function sectionEvent(e) {
        this.section = e.detail.value;
      }
    }, {
      key: "gameEvent",
      value: function gameEvent(e) {
        this.game = e.detail.value;
      }
    }, {
      key: "activeDataEvent",
      value: function activeDataEvent(e) {
        this.activeData = babelHelpers.objectSpread({}, e.detail.value);
      }
    }, {
      key: "optionsDataChangedEvent",
      value: function optionsDataChangedEvent(e) {
        this.optionsData = babelHelpers.objectSpread({}, e.detail.value);
      }
    }, {
      key: "dataChangedEvent",
      value: function dataChangedEvent(e) {
        this.data = e.detail.value;
      }
    }, {
      key: "routeChangedEvent",
      value: function routeChangedEvent(e) {
        this.route = e.detail.value;
      }
    }, {
      key: "tailChangedEvent",
      value: function tailChangedEvent(e) {
        this.tail = e.detail.value;
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          var notifiedProps = ["sourcePath", "response"];

          if (notifiedProps.includes(propName)) {
            // notify
            var eventName = "".concat(propName.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, "$1-$2").toLowerCase(), "-changed");

            _this2.dispatchEvent(new CustomEvent(eventName, {
              detail: {
                value: _this2[propName]
              }
            }));
          }

          if (["dataPath", "section", "game"].includes(propName)) {
            _this2.dataRequestUrl = _this2._computeDataRequestUrl(_this2.dataPath, _this2.section, _this2.game);
          }

          if (["route", "endPoint"].includes(propName)) {
            _this2._routeChanged(_this2.route, _this2.endPoint);
          }

          if (propName == "optionsData") {
            _this2._optionsDataChanged(propName);
          }

          if (propName == "activeData") {
            _this2._activeDataChanged(propName);
          }
        });
      }
    }, {
      key: "_computeDataRequestUrl",
      value: function _computeDataRequestUrl(dataPath, section, game) {
        return "".concat(dataPath, "&section=").concat(section, "&game=").concat(game);
      }
    }, {
      key: "_optionsDataChanged",
      value: function _optionsDataChanged(newValue) {
        if (newValue && newValue.status) {
          var sections = [];

          for (var i in newValue.data.sections) {
            sections.push([{
              alt: newValue.data.sections[i],
              value: i
            }]);
          }

          var games = [];

          for (var i in newValue.data.games) {
            games.push([{
              alt: newValue.data.games[i],
              value: i
            }]);
          }

          this.sectionOptions = [].concat(sections);
          this.gameOptions = [].concat(games);
        }
      }
    }, {
      key: "_activeDataChanged",
      value: function _activeDataChanged(newValue) {
        this.visibleData = babelHelpers.toConsumableArray(newValue.data);
      }
      /**
       * Route changed
       */

    }, {
      key: "_routeChanged",
      value: function _routeChanged(route, endPoint) {
        if (typeof route.path === "string") {
          if (typeof endPoint === "string") {
            // ignore "home page" as well since that's our path
            if (route.path.startsWith(endPoint) || route.path == "/") {
              return;
            }
          }

          window.location.reload();
        }
      }
    }, {
      key: "connectedCallback",
      value: function connectedCallback() {
        babelHelpers.get(babelHelpers.getPrototypeOf(GameShowScoreboard.prototype), "connectedCallback", this).call(this);

        if (document.getElementById("block-mooc-nav-block-mooc-nav-nav")) {
          document.getElementById("block-mooc-nav-block-mooc-nav-nav").remove();
        }
      }
    }, {
      key: "_toArray",
      value: function _toArray(obj) {
        if (obj == null) {
          return [];
        }

        return Object.keys(obj).map(function (key) {
          return obj[key];
        });
      }
    }]);
    return GameShowScoreboard;
  }(_litElement.LitElement);

  _exports.GameShowScoreboard = GameShowScoreboard;
  window.customElements.define(GameShowScoreboard.tag, GameShowScoreboard);
});
