// @lrnwebcomponents/image-inspector/image-inspector.js
define(["exports","require","../../@polymer/polymer/polymer-element.js"],function(_exports,_require,_polymerElement){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.ImageInspector=void 0,_require=babelHelpers.interopRequireWildcard(_require);class ImageInspector extends _polymerElement.PolymerElement{constructor(){super(),new Promise((res,rej)=>_require.default(["../../@polymer/app-layout/app-layout.js"],res,rej)),new Promise((res,rej)=>_require.default(["../img-pan-zoom/img-pan-zoom.js"],res,rej)),new Promise((res,rej)=>_require.default(["../lrnsys-button/lrnsys-button.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/iron-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/image-icons.js"],res,rej))}static get template(){return _polymerElement.html`
      <style>
        :host {
          display: block;
          --image-inspector-background: #dddddd;
        }

        app-toolbar {
          width: 90%;
          background: var(--image-inspector-background);
          margin: 32px auto;
          z-index: 1;
          display: flex;
          text-align: center;
          justify-content: space-evenly;
        }

        lrnsys-button {
          display: inline-flex;
        }

        .top {
          top: 128px;
        }
      </style>

      <app-toolbar>
        <lrnsys-button
          alt="Zoom in"
          icon="zoom-in"
          on-click="zoomIn"
          hover-class="[[hoverClass]]"
        ></lrnsys-button>
        <lrnsys-button
          alt="Zoom out"
          icon="zoom-out"
          on-click="zoomOut"
          hover-class="[[hoverClass]]"
        ></lrnsys-button>
        <lrnsys-button
          alt="Rotate right"
          icon="image:rotate-right"
          on-click="rotateRight"
          hover-class="[[hoverClass]]"
        ></lrnsys-button>
        <lrnsys-button
          alt="Rotate left"
          icon="image:rotate-left"
          on-click="rotateLeft"
          hover-class="[[hoverClass]]"
        ></lrnsys-button>
        <lrnsys-button
          alt="Mirror image"
          icon="image:flip"
          on-click="mirrorImage"
          hover-class="[[hoverClass]]"
        ></lrnsys-button>
        <lrnsys-button
          alt="Open in new window"
          icon="launch"
          href="[[src]]"
          target="_blank"
          hover-class="[[hoverClass]]"
        ></lrnsys-button>
        <slot name="toolbar"></slot>
      </app-toolbar>
      <img-pan-zoom id="img" src="[[src]]"></img-pan-zoom>
      <slot></slot>
    `}static get tag(){return"image-inspector"}static get properties(){return{degrees:{type:Number,value:0,reflectToAttribute:!0},src:{type:String,reflectToAttribute:!0},hoverClass:{type:String,value:"blue white-text"}}}connectedCallback(){super.connectedCallback(),this.__img=this.shadowRoot.querySelector("#img")}rotateRight(){this.degrees+=90,this.__img.style.transform="rotate("+this.degrees+"deg)",this.__img.toggleClass("top")}rotateLeft(){this.degrees+=-90,this.__img.style.transform="rotate("+this.degrees+"deg)",this.__img.toggleClass("top")}mirrorImage(){"scaleX(1)"===this.__img.style.transform?this.__img.style.transform="scaleX(-1)":this.__img.style.transform="scaleX(1)"}zoomIn(){this.__img.zoomIn()}zoomOut(){this.__img.zoomOut()}}_exports.ImageInspector=ImageInspector,window.customElements.define(ImageInspector.tag,ImageInspector)});
