// @lrnwebcomponents/elmsln-studio/lib/elmsln-studio-assignment.js
define(["exports", "../../../lit-element/lit-element.js", "./elmsln-studio-styles.js", "./elmsln-studio-utilities.js", "../../lrndesign-gallery/lrndesign-gallery.js", "../../hax-iconset/hax-iconset.js", "../../threaded-discussion/threaded-discussion.js", "./elmsln-studio-link.js", "./elmsln-studio-button.js"], function (_exports, _litElement, _elmslnStudioStyles, _elmslnStudioUtilities, _lrndesignGallery, _haxIconset, _threadedDiscussion, _elmslnStudioLink, _elmslnStudioButton) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmslnStudioAssignment = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject12_3de08670019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          flex-wrap: wrap;\n        }\n        h1 {\n          text-align: center;\n        }\n        #breadcrumb {\n          margin: 0 0 var(--elmsln-studio-margin, 20px);\n          flex: 1 0 100%;\n          --elmsln-studio-link-Color: var(\n            --simple-colors-fixed-theme-light-blue-8\n          );\n          --elmsln-studio-link-focus-Color: var(\n            --simple-colors-fixed-theme-light-blue-9\n          );\n        }\n        #breadcrumb > * {\n          display: inline;\n        }\n        lrndesign-avatar,\n        .student-name,\n        .project-name,\n        .assignment-name,\n        .submission-date {\n          font-weight: normal;\n          display: block;\n        }\n        .student-name {\n          font-size: calc(2 * var(--elmsln-studio-FontSize, 16px));\n          font-weight: bold;\n          color: #4b4b4b;\n        }\n        .project-name {\n          font-size: calc(1.75 * var(--elmsln-studio-FontSize, 16px));\n          color: #95989a;\n        }\n        #project-buttons,\n        .submission-header {\n          position: relative;\n          height: 40px;\n          margin-top: -40px;\n        }\n        #sort,\n        .submission-header elmsln-studio-button {\n          color: #95989a;\n          font-size: var(--elmsln-studio-FontSize, 16px);\n          text-transform: uppercase;\n          position: absolute;\n          right: 0;\n          top: 0;\n        }\n        #sort:focus,\n        #sort:hover,\n        .submission-header elmsln-studio-button:focus,\n        .submission-header elmsln-studio-button:hover {\n          color: #4b4b4b;\n        }\n        #sort.sort-latest {\n          transform: rotateX(180deg) rotateY(0deg);\n        }\n        section {\n          border-top: 2px solid #eaeaea;\n          padding: calc(0.5 * var(--elmsln-studio-margin, 20px)) 0\n            var(--elmsln-studio-margin, 20px);\n        }\n        .view-discussion {\n          border: 1px solid #eaeaea;\n          padding: calc(0.5 * var(--elmsln-studio-margin, 20px));\n        }\n        .view-discussion section {\n          opacity: 0.4;\n          display: none;\n        }\n        .view-discussion section.section-discussion {\n          display: block;\n          opacity: 1;\n        }\n        h2 {\n          margin: calc(2 * var(--elmsln-studio-margin, 20px)) auto;\n          text-align: center;\n        }\n        .assignment-name {\n          font-size: calc(1.5 * var(--elmsln-studio-FontSize, 16px));\n          color: #555555;\n        }\n        .submission-date {\n          font-size: var(--elmsln-studio-FontSize, 16px);\n          color: #95989a;\n        }\n        .submission-body {\n          color: #95989a;\n          line-height: 160%;\n          margin: calc(0.5 * var(--elmsln-studio-margin, 20px)) auto\n            var(--elmsln-studio-margin, 20px);\n        }\n        .submission-body:first-child {\n          margin-top: 0;\n        }\n        .submission-body:last-child {\n          margin-bottom: 0;\n        }\n        .submission-links {\n          list-style: none;\n          padding-inline-start: 0;\n        }\n        .submission-links > li {\n          display: flex;\n          align-items: flex-start;\n          justify-content: flex-start;\n        }\n        .submission-links a:link,\n        .submission-links a:visited {\n          font-size: calc(1.5 * var(--elmsln-studio-FontSize, 16px));\n          font-weight: bold;\n          color: #000;\n          text-decoration: none;\n        }\n        .submission-links a:focus,\n        .submission-links a:hover {\n          text-decoration: underline;\n        }\n        .submission-links iron-icon {\n          margin-right: 0.5em;\n        }\n        .submission-links a:focus iron-icon,\n        .submission-links a:hover iron-icon {\n          text-decoration: none;\n        }\n        .callout {\n          font-size: calc(0.75 * var(--elmsln-studio-FontSize, 16px));\n          border: 1px solid #eaeaea;\n          padding: 0;\n        }\n        .discussion-label {\n          text-align: left;\n          font-size: calc(1 * var(--elmsln-studio-FontSize, 16px));\n          margin: var(--elmsln-studio-margin, 20px) 0\n            calc(0.25 * var(--elmsln-studio-margin, 20px));\n        }\n        .callout > * {\n          padding: calc(0.25 * var(--elmsln-studio-margin, 20px))\n            calc(0.5 * var(--elmsln-studio-margin, 20px));\n        }\n        .callout .callout-label {\n          font-size: calc(1 * var(--elmsln-studio-FontSize, 16px));\n          font-weight: normal;\n          display: flex;\n          align-items: center;\n          justify-content: flex-start;\n          margin: 0;\n          color: #4b4b4b;\n          border-bottom: 1px solid #eaeaea;\n        }\n        .callout .callout-label iron-icon {\n          margin-right: 1em;\n        }\n        threaded-discussion {\n          background-color: #eaeaea;\n          padding: 1px;\n          --threaded-discussion-comment-button-BackgroundColor: var(\n            --simple-colors-fixed-theme-light-blue-8\n          );\n          --threaded-discussion-comment-button-focus-BackgroundColor: var(\n            --simple-colors-fixed-theme-light-blue-9\n          );\n        }\n        @media screen and (min-width: 600px) {\n          :host {\n            display: flex;\n            align-items: flex-start;\n            justify-content: space-between;\n          }\n          #primary {\n            flex: 1 0 100%;\n          }\n          #primary.view-discussion {\n            flex: 0 0 calc(50% - var(--elmsln-studio-margin, 20px));\n          }\n          #primary:not(.view-discussion) {\n            max-width: calc(1000px - 2 * var(--elmsln-studio-margin, 20px));\n            margin: 0 auto;\n            border: 1px solid #eaeaea;\n            padding: var(--elmsln-studio-margin, 20px);\n          }\n          .view-discussion section {\n            display: block;\n          }\n          .view-discussion section.section-discussion {\n            border: 4px solid #95989a;\n            padding: calc(0.5 * var(--elmsln-studio-margin, 20px))\n              calc(0.5 * var(--elmsln-studio-margin, 20px))\n              var(--elmsln-studio-margin, 20px);\n            opacity: 1;\n          }\n          #secondary {\n            flex: 0 0 calc(50% - var(--elmsln-studio-margin, 20px));\n          }\n        }\n      "]);

    _templateObject12_3de08670019d11eb9027c17497ee62cd = function _templateObject12_3de08670019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _templateObject11_3de08670019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                    <section>\n                      <h2 class=\"sr-only\">Edit Text</h2>\n                      <textarea></textarea>\n                    </section>\n                  "]);

    _templateObject11_3de08670019d11eb9027c17497ee62cd = function _templateObject11_3de08670019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _templateObject10_3de08670019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                    <section>\n                      <h2 class=\"sr-only\">Manage Links</h2>\n                      <input type=\"text\" />\n                    </section>\n                  "]);

    _templateObject10_3de08670019d11eb9027c17497ee62cd = function _templateObject10_3de08670019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _templateObject9_3de08670019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                    <section>\n                      <h2 class=\"sr-only\">Manage Uploads</h2>\n                    </section>\n                  "]);

    _templateObject9_3de08670019d11eb9027c17497ee62cd = function _templateObject9_3de08670019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _templateObject8_3de08670019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral([" <td>", "</td> "]);

    _templateObject8_3de08670019d11eb9027c17497ee62cd = function _templateObject8_3de08670019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _templateObject7_3de08670019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral([" <th scope=\"col\">", "</th> "]);

    _templateObject7_3de08670019d11eb9027c17497ee62cd = function _templateObject7_3de08670019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _templateObject6_3de08670019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                                <tr>\n                                  ", "\n                                </tr>\n                              "]);

    _templateObject6_3de08670019d11eb9027c17497ee62cd = function _templateObject6_3de08670019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _templateObject5_3de08670019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral([" <th scope=\"row\">", "</th> "]);

    _templateObject5_3de08670019d11eb9027c17497ee62cd = function _templateObject5_3de08670019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _templateObject4_3de08670019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                            <thead>\n                              ", "\n                            </thead>\n                          "]);

    _templateObject4_3de08670019d11eb9027c17497ee62cd = function _templateObject4_3de08670019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _templateObject3_3de08670019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                    <table>\n                      <caption>\n                        Rubric\n                      </caption>\n                      ", "\n                      <tbody>\n                        ", "\n                      </tbody>\n                    </table>\n                  "]);

    _templateObject3_3de08670019d11eb9027c17497ee62cd = function _templateObject3_3de08670019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _templateObject2_3de08670019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <elmsln-studio-link\n                    href=\"", "\"\n                    >", "</elmsln-studio-link\n                  >\n                  <span> > </span>\n                "]);

    _templateObject2_3de08670019d11eb9027c17497ee62cd = function _templateObject2_3de08670019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _templateObject_3de08670019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral(["\n          <div id=\"breadcrumb\">\n            <elmsln-studio-link href=\"/assignments\"\n              >Assignments</elmsln-studio-link\n            >\n            <span> > </span>\n            ", "\n            <span>", "</span>\n          </div>\n          <div\n            id=\"primary\"\n            ?hidden=\"", "\"\n            class=\"view-assignment\"\n          >\n            <article id=\"assignment\">\n              <h1>\n                <span class=\"lesson-name\">", "</span>\n                <span class=\"project-name\">", "</span>\n              </h1>\n              <p>", "</p>\n              ", "\n              ", "\n              ", "\n              ", "\n              <section>\n                ", "\n                <elmsln-studio-button class=\"delete\" icon=\"\" path=\"\"\n                  >Delete</elmsln-studio-button\n                >\n                <elmsln-studio-button class=\"save\" path=\"\"\n                  >Save Draft</elmsln-studio-button\n                >\n                <elmsln-studio-button class=\"submit\" icon=\"done\" path=\"\"\n                  >Submit Assignment</elmsln-studio-button\n                >\n              </section>\n            </article>\n          </div>\n        "]);

    _templateObject_3de08670019d11eb9027c17497ee62cd = function _templateObject_3de08670019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `elmsln-studio-assignment`
   * Studio App for ELMS:LN
   *
   * @customElement elmsln-studio-assignment
   * @lit-html
   * @lit-element
   * @demo demo/portfolio.html
   */
  var ElmslnStudioAssignment = /*#__PURE__*/function (_ElmslnStudioUtilitie) {
    babelHelpers.inherits(ElmslnStudioAssignment, _ElmslnStudioUtilitie);

    var _super = _createSuper(ElmslnStudioAssignment);

    babelHelpers.createClass(ElmslnStudioAssignment, [{
      key: "render",
      // render function
      value: function render() {
        var _this2 = this;

        return !this.assignment ? "" : (0, _litElement.html)(_templateObject_3de08670019d11eb9027c17497ee62cd(), !this.assignment.projectId || !this.assignment.project ? "" : (0, _litElement.html)(_templateObject2_3de08670019d11eb9027c17497ee62cd(), "/assignments?project=".concat(this.assignment.projectId), this.assignment.project), this.assignment.assignment, !this.assignment, this.assignment.lesson, this.assignment.assignment, this.assignment.body, !this.assignment.rubric ? "" : (0, _litElement.html)(_templateObject3_3de08670019d11eb9027c17497ee62cd(), !this.assignment.rubric || !this.assignment.rubric[0] ? "" : (0, _litElement.html)(_templateObject4_3de08670019d11eb9027c17497ee62cd(), this.assignment.rubric[0].map(function (col) {
          return (0, _litElement.html)(_templateObject5_3de08670019d11eb9027c17497ee62cd(), col);
        })), (this.assignment.rubric || []).map(function (row, i) {
          return i === 0 ? "" : (0, _litElement.html)(_templateObject6_3de08670019d11eb9027c17497ee62cd(), _this2.assignment.rubric[0].map(function (col, ii) {
            return ii === 0 ? (0, _litElement.html)(_templateObject7_3de08670019d11eb9027c17497ee62cd(), col) : (0, _litElement.html)(_templateObject8_3de08670019d11eb9027c17497ee62cd(), col);
          }));
        })), !this.assignment.uploads ? "" : (0, _litElement.html)(_templateObject9_3de08670019d11eb9027c17497ee62cd()), !this.assignment.links ? "" : (0, _litElement.html)(_templateObject10_3de08670019d11eb9027c17497ee62cd()), !this.assignment.text ? "" : (0, _litElement.html)(_templateObject11_3de08670019d11eb9027c17497ee62cd()), this.submission.draft);
      } // properties available to the custom element for data binding

    }], [{
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioAssignment), "styles", this)), [(0, _litElement.css)(_templateObject12_3de08670019d11eb9027c17497ee62cd())]);
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioAssignment), "properties", this)), {}, {
          assignment: {
            type: Object
          },
          submission: {
            type: Object
          }
        });
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */

    }, {
      key: "tag",
      get: function get() {
        return "elmsln-studio-assignment";
      } // life cycle

    }]);

    function ElmslnStudioAssignment() {
      var _this;

      babelHelpers.classCallCheck(this, ElmslnStudioAssignment);
      _this = _super.call(this);
      _this.portfolio = {};
      _this.submission = {};
      _this.submissionFilter = "";
      _this.comment = "";
      _this.tag = ElmslnStudioAssignment.tag;
      return _this;
    }
    /**
     * life cycle, element is afixed to the DOM
     */


    babelHelpers.createClass(ElmslnStudioAssignment, [{
      key: "connectedCallback",
      value: function connectedCallback() {
        babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioAssignment.prototype), "connectedCallback", this).call(this);
      }
    }, {
      key: "_getFeedbackIcon",
      value: function _getFeedbackIcon(comments) {
        if (comments === 0) {
          return "communication:comment";
        } else if (comments < 10) {
          return "hax:messages-".concat(comments);
        }

        return "hax:messages-9-plus";
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        if (babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioAssignment.prototype), "updated", this)) babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioAssignment.prototype), "updated", this).call(this, changedProperties);
        changedProperties.forEach(function (oldValue, propName) {});
      }
    }, {
      key: "sortedSubmissions",
      get: function get() {
        return !this.portfolio.submissions ? [] : this.sortDates(this.portfolio.submissions, this.sortLatest);
      }
    }, {
      key: "assignment",
      get: function get() {
        var _this3 = this;

        var filter = !this.submissionFilter || !this.portfolio || !this.portfolio.submissions ? [] : this.portfolio.submissions.filter(function (s) {
          return s.id === _this3.submissionFilter;
        });
        return !filter ? false : filter[0];
      }
    }]);
    return ElmslnStudioAssignment;
  }((0, _elmslnStudioUtilities.ElmslnStudioUtilities)((0, _elmslnStudioStyles.ElmslnStudioStyles)(_litElement.LitElement)));

  _exports.ElmslnStudioAssignment = ElmslnStudioAssignment;
  customElements.define("elmsln-studio-assignment", ElmslnStudioAssignment);
});
