// @lrnwebcomponents/elmsln-studio/lib/elmsln-studio-modal.js
define(["exports", "../../../lit-element/lit-element.js", "../../simple-modal/simple-modal.js"], function (_exports, _litElement, _simpleModal) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmslnStudioModal = void 0;

  function _templateObject2_8afa1a00e0cf11eab599756c9ebe7c2f() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <button id=\"modal\" @click=\"", "\">\n        <slot></slot>\n      </button>\n    "]);

    _templateObject2_8afa1a00e0cf11eab599756c9ebe7c2f = function _templateObject2_8afa1a00e0cf11eab599756c9ebe7c2f() {
      return data;
    };

    return data;
  }

  function _templateObject_8afa1a00e0cf11eab599756c9ebe7c2f() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      :host {\n        display: block;\n      }\n      :host([hidden]) {\n        display: none;\n      }\n      button {\n        border: none;\n        background-color: transparent;\n      }\n      button:focus,\n      button:hover {\n        outline: none;\n      }\n    "]);

    _templateObject_8afa1a00e0cf11eab599756c9ebe7c2f = function _templateObject_8afa1a00e0cf11eab599756c9ebe7c2f() {
      return data;
    };

    return data;
  }

  /**
   * `elmsln-studio-modal`
   * Navigation Link for ELMS:LN Studio
   *
   * @customElement elmsln-studio-modal
   * @lit-html
   * @lit-element
   */
  var ElmslnStudioModal =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(ElmslnStudioModal, _LitElement);
    babelHelpers.createClass(ElmslnStudioModal, null, [{
      key: "tag",

      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */
      get: function get() {
        return "elmsln-studio-modal";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          title: {
            type: Object
          },
          header: {
            type: Object
          },
          content: {
            type: Object
          },
          buttons: {
            type: Object
          },
          modal: {
            type: Object
          },
          styles: {
            type: Object
          }
        };
      }
    }, {
      key: "styles",
      get: function get() {
        return (0, _litElement.css)(_templateObject_8afa1a00e0cf11eab599756c9ebe7c2f());
      }
    }]);

    function ElmslnStudioModal() {
      var _this;

      babelHelpers.classCallCheck(this, ElmslnStudioModal);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(ElmslnStudioModal).call(this));
      _this.modal = window.SimpleModal.requestAvailability();
      return _this;
    }

    babelHelpers.createClass(ElmslnStudioModal, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject2_8afa1a00e0cf11eab599756c9ebe7c2f(), this.modalOpen);
      }
    }, {
      key: "modalOpen",
      value: function modalOpen(event) {
        var _styles,
            _this2 = this;

        var evt,
            button = this.shadowRoot.getElementById("modal"),
            styles = (_styles = {
          "--simple-modal-width": "90%",
          "--simple-modal-height": "90vh",
          "--simple-modal-titlebar-height": "40px",
          "--simple-modal-titlebar-line-height": "40px"
        }, babelHelpers.defineProperty(_styles, "--simple-modal-titlebar-height", "40px"), babelHelpers.defineProperty(_styles, "--simple-modal-titlebar-padding", "0px 5px"), babelHelpers.defineProperty(_styles, "--simple-modal-titlebar-background", "white"), babelHelpers.defineProperty(_styles, "--simple-modal-header-background", "white"), babelHelpers.defineProperty(_styles, "--simple-modal-content-padding", "0px 5px"), _styles);
        Object.keys(this.styles || {}).forEach(function (key) {
          return styles[key] = _this2.styles[key];
        });
        this.dispatchEvent(new CustomEvent("modal-button-click", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: this
        }));
        evt = new CustomEvent("simple-modal-show", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: {
            title: false,
            elements: {
              header: this.header,
              content: this.content,
              buttons: this.buttons
            },
            styles: styles,
            invokedBy: button,
            clone: false
          }
        });
        console.log(this.modal, button, evt);
        this.dispatchEvent(evt);
      }
    }]);
    return ElmslnStudioModal;
  }(_litElement.LitElement);

  _exports.ElmslnStudioModal = ElmslnStudioModal;
  customElements.define("elmsln-studio-modal", ElmslnStudioModal);
});
