// @lrnwebcomponents/elmsln-studio/lib/elmsln-studio-link.js
define(["exports", "../../../lit-element/lit-element.js", "../../../lit-element-router/lit-element-router.js"], function (_exports, _litElement, _litElementRouter) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmslnStudioLink = void 0;

  function _templateObject2_a37532c0e8ab11eabe6cb3de058a4220() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <a href=\"", "\" @click=\"", "\">\n        <slot></slot>\n      </a>\n    "]);

    _templateObject2_a37532c0e8ab11eabe6cb3de058a4220 = function _templateObject2_a37532c0e8ab11eabe6cb3de058a4220() {
      return data;
    };

    return data;
  }

  function _templateObject_a37532c0e8ab11eabe6cb3de058a4220() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      :host {\n        display: block;\n      }\n      :host([hidden]) {\n        display: none;\n      }\n      a,\n      a:link {\n        transition: var(--elmsln-studio-link-Transition);\n        background: transparent;\n        color: var(--elmsln-studio-link-Color);\n        text-decoration: var(--elmsln-studio-link-TextDecoration);\n        text-decoration-style: var(--elmsln-studio-link-TextDecorationStyle);\n        text-decoration-skip: var(--elmsln-studio-link-TextDecorationSkip);\n        text-decoration-skip-ink: var(\n          --elmsln-studio-link-TextDecorationSkipInk\n        );\n        text-decoration-line: var(--elmsln-studio-link-TextDecorationLine);\n        text-decoration-color: var(--elmsln-studio-link-TextDecorationColor);\n        text-decoration: var(--elmsln-studio-link-TextDecoration);\n        text-decoration-thickness: var(\n          --elmsln-studio-link-TextDecorationThickness\n        );\n      }\n      a:hover,\n      a:focus {\n        background: transparent;\n        color: var(--elmsln-studio-link-focus-Color, var(--elmsln-studio-link-Color));\n        text-decoration: var(--elmsln-studio-link-focus-TextDecoration, var(--elmsln-studio-link-TextDecoration));\n        text-decoration-style: var(--elmsln-studio-link-focus-TextDecorationStyle, var(--elmsln-studio-link-TextDecorationStyle));\n        text-decoration-skip: var(--elmsln-studio-link-focus-TextDecorationSkip, var(--elmsln-studio-link-TextDecorationSkip));\n        text-decoration-skip-ink: var(\n          --elmsln-studio-link-focus-TextDecorationSkipInk,\n          var(\n          --elmsln-studio-link-TextDecorationSkipInk\n        );\n        );\n        text-decoration-line: var(--elmsln-studio-link-focus-TextDecorationLine, var(--elmsln-studio-link-TextDecorationLine));\n        text-decoration-color: var(--elmsln-studio-link-focus-TextDecorationColor, ar(--elmsln-studio-link-TextDecorationColor));\n        text-decoration: var(--elmsln-studio-link-focus-TextDecoration, var(--elmsln-studio-link-TextDecoration););\n        text-decoration-thickness: var(\n          --elmsln-studio-link-focus-TextDecorationThickness,\n          var(\n            --elmsln-studio-link-TextDecorationThickness\n          )\n        );\n      }\n    "]);

    _templateObject_a37532c0e8ab11eabe6cb3de058a4220 = function _templateObject_a37532c0e8ab11eabe6cb3de058a4220() {
      return data;
    };

    return data;
  }

  /**
   * `elmsln-studio-link`
   * Navigation Link for ELMS:LN Studio
   *
   * @customElement elmsln-studio-link
   * @lit-html
   * @lit-element
   */
  var ElmslnStudioLink =
  /*#__PURE__*/
  function (_navigator) {
    babelHelpers.inherits(ElmslnStudioLink, _navigator);
    babelHelpers.createClass(ElmslnStudioLink, null, [{
      key: "tag",

      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */
      get: function get() {
        return "elmsln-studio-link";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          href: {
            type: String
          }
        };
      }
    }, {
      key: "styles",
      get: function get() {
        return (0, _litElement.css)(_templateObject_a37532c0e8ab11eabe6cb3de058a4220());
      }
    }]);

    function ElmslnStudioLink() {
      var _this;

      babelHelpers.classCallCheck(this, ElmslnStudioLink);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(ElmslnStudioLink).call(this));
      _this.href = "";

      _this.addEventListener("click", _this.linkClick);

      return _this;
    }

    babelHelpers.createClass(ElmslnStudioLink, [{
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        if (babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioLink.prototype), "disconnectedCallback", this)) babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioLink.prototype), "disconnectedCallback", this).call(this);
        this.removeEventListener("click", this.linkClick);
      }
    }, {
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject2_a37532c0e8ab11eabe6cb3de058a4220(), this.href, this.linkClick);
      }
    }, {
      key: "linkClick",
      value: function linkClick(event) {
        event.preventDefault();
        this.navigate(this.href);
      }
    }]);
    return ElmslnStudioLink;
  }((0, _litElementRouter.navigator)(_litElement.LitElement));

  _exports.ElmslnStudioLink = ElmslnStudioLink;
  customElements.define("elmsln-studio-link", ElmslnStudioLink);
});
