// @lrnwebcomponents/elmsln-studio/lib/elmsln-studio-submission-card.js
define(["exports", "../../../lit/index.js", "../../accent-card/accent-card.js"], function (_exports, _index, _accentCard) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmslnStudioSubmissionCard = void 0;

  var _templateObject_616abab0942811ecb1937969798b231a, _templateObject2_616abab0942811ecb1937969798b231a;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `elmsln-studio-submission-card`
   * Studio App for ELMS:LN
   *
   * @customElement elmsln-studio-submission-card
   * @lit-html
   * @lit-element
   * @demo demo/submission.html
   */
  var ElmslnStudioSubmissionCard = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(ElmslnStudioSubmissionCard, _LitElement);

    var _super = _createSuper(ElmslnStudioSubmissionCard);

    // life cycle
    function ElmslnStudioSubmissionCard() {
      var _this;

      babelHelpers.classCallCheck(this, ElmslnStudioSubmissionCard);
      _this = _super.call(this);
      _this.horizontal = false;
      return _this;
    }

    babelHelpers.createClass(ElmslnStudioSubmissionCard, [{
      key: "render",
      value: // render function
      function render() {
        return (0, _index.html)(_templateObject_616abab0942811ecb1937969798b231a || (_templateObject_616abab0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <accent-card\n        image-src=\"", "\"\n        .image-alt=\"", "\"\n        ?horizontal=\"", "\"\n        .image-align=\"", "\"\n        .image-valign=\"", "\"\n        .gravity=\"", "\"\n        no-border\n      >\n        <slot name=\"student\" slot=\"heading\"></slot>\n        <slot name=\"datetime\" slot=\"corner\"></slot>\n        <slot name=\"assigmment\" slot=\"subheading\"></slot>\n        <slot name=\"project\" slot=\"content\"></slot>\n        <slot name=\"feature\" slot=\"content\"></slot>\n        <slot name=\"feedback\" slot=\"footer\"></slot>\n      </accent-card>\n    "])), this.imageSrc || "", this.imageAlt || undefined, this.horizontal, this.imageAlign, this.imageValign, this.imageGravity);
      } // properties available to the custom element for data binding

    }], [{
      key: "tag",
      get:
      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */
      function get() {
        return "elmsln-studio-submission-card";
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _index.css)(_templateObject2_616abab0942811ecb1937969798b231a || (_templateObject2_616abab0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host,\n        accent-card {\n          line-height: 160%;\n          display: flex;\n          flex-direction: column;\n          align-items: stretch;\n          justify-content: stretch;\n        }\n        accent-card {\n          width: auto;\n          flex: 1 1 auto;\n          --accent-card-padding: 0;\n          --accent-card-heading-min-height: 30px;\n          --accent-card-heading-padding-top: calc(\n            0.5 * var(--elmsln-studio-margin, 20px)\n          );\n          --accent-card-heading-padding-left: calc(\n            0.5 * var(--elmsln-studio-margin, 20px)\n          );\n          --accent-card-heading-padding-right: calc(\n            0.5 * var(--elmsln-studio-margin, 20px)\n          );\n          --accent-card-subheading-padding-left: calc(\n            0.5 * var(--elmsln-studio-margin, 20px)\n          );\n          --accent-card-subheading-padding-right: calc(\n            0.5 * var(--elmsln-studio-margin, 20px)\n          );\n          --accent-card-content-padding-left: calc(\n            0.5 * var(--elmsln-studio-margin, 20px)\n          );\n          --accent-card-content-padding-right: calc(\n            0.5 * var(--elmsln-studio-margin, 20px)\n          );\n          --accent-card-content-padding-bottom: calc(\n            0.5 * var(--elmsln-studio-margin, 20px)\n          );\n          --accent-card-footer-padding-left: calc(\n            0.25 * var(--elmsln-studio-margin, 20px)\n          );\n          --accent-card-footer-padding-right: calc(\n            0.25 * var(--elmsln-studio-margin, 20px)\n          );\n          --accent-card-footer-padding-top: calc(\n            0.5 * var(--elmsln-studio-margin, 20px)\n          );\n          --accent-card-footer-padding-bottom: calc(\n            0.5 * var(--elmsln-studio-margin, 20px)\n          );\n          --accent-card-image-padding-bottom: 5px;\n          --accent-card-image-padding-right: calc(\n            0.5 * var(--elmsln-studio-margin, 20px)\n          );\n          --accent-card-image-width: 33.33333%;\n          --accent-card-image-height: 200px;\n        }\n        accent-card:focus-within,\n        accent-card:hover {\n          outline: 2px solid var(--accent-card-border-color);\n        }\n        [slot=\"feature\"] {\n          margin-top: var(--elmsln-studio-margin, 20px);\n          height: calc(\n            var(--accent-card-image-height, 200px) -\n              var(--elmsln-studio-margin, 20px)\n          );\n          overflow: auto;\n        }\n        [slot=\"heading\"] {\n          font-weight: var(--elmsln-studio-FontWeightLight, 300);\n          font-size: 22px;\n        }\n        [slot=\"corner\"] {\n          font-weight: var(--elmsln-studio-FontWeightNormal, 400);\n          font-size: 12px;\n        }\n        [slot=\"subheading\"] {\n          font-weight: var(--elmsln-studio-FontWeightBold, 500);\n          font-size: 18px;\n          font-style: normal;\n          color: #5d5e5f;\n        }\n        [slot=\"content\"] {\n          font-weight: var(--elmsln-studio-FontWeightNormal, 400);\n          font-size: 14px;\n          color: #7e7e7e;\n        }\n        [slot=\"footer\"] {\n          font-weight: var(--elmsln-studio-FontWeightNormal, 400);\n          font-size: 12px;\n          text-transform: uppercase;\n          text-align: right;\n          color: #95989a;\n        }\n        [slot=\"footer\"] elmsln-studio-link {\n          margin: 0 calc(0.5 * var(--elmsln-studio-margin, 20px));\n          text-align: right;\n          --elmsln-studio-link-Color: #7e7e7e;\n        }\n        [slot=\"footer\"] elmsln-studio-link:focus,\n        [slot=\"footer\"] elmsln-studio-link:focus-within,\n        [slot=\"footer\"] elmsln-studio-link:hover {\n          --elmsln-studio-link-Color: #95989a;\n          --elmsln-studio-link-TextDecoration: none !important;\n        }\n      "])))];
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          imageSrc: {
            type: String,
            attribute: "image-src"
          },
          imageAlt: {
            type: String,
            attribute: "image-alt"
          },
          imageAlign: {
            type: String,
            attribute: "image-align"
          },
          imageValign: {
            type: String,
            attribute: "image-valign"
          },
          horizontal: {
            type: Boolean,
            attribute: "horizontal"
          }
        };
      }
    }]);
    return ElmslnStudioSubmissionCard;
  }(_index.LitElement);

  _exports.ElmslnStudioSubmissionCard = ElmslnStudioSubmissionCard;
  customElements.define("elmsln-studio-submission-card", ElmslnStudioSubmissionCard);
});
