// @lrnwebcomponents/elmsln-studio/lib/elmsln-studio-assignments.js
define(["exports", "../../../lit-element/lit-element.js", "./elmsln-studio-utilities.js", "./elmsln-studio-styles.js"], function (_exports, _litElement, _elmslnStudioUtilities, _elmslnStudioStyles) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmslnStudioAssignments = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject6_9ef74af056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: flex;\n          color: #2c2c2c;\n          margin: 0 auto;\n          justify-content: center;\n        }\n        h2 {\n          color: #525252;\n          margin-top: calc(0.5 * var(--elmsln-studio-margin, 20px));\n        }\n        nav-card [slot=\"heading\"] {\n          color: #4d4d4d;\n          font-size: 80%;\n        }\n        nav-card-item [slot=\"label\"] {\n          font-weight: normal;\n        }\n        nav-card-item [slot=\"description\"] {\n          color: #95989a;\n        }\n        .lesson {\n          background-color: #eee;\n          padding: calc(0.5 * var(--elmsln-studio-margin, 20px));\n          --lrndesign-avatar-border-radius: 0%;\n          --nav-card-item-avatar-width: 40px;\n          border: 1px solid var(--accent-card-border-color, #ddd);\n          border-bottom: none;\n        }\n        .lesson > .assignment {\n          background-color: #fff;\n          border: 1px solid var(--accent-card-border-color, #ddd);\n          padding: calc(0.5 * var(--elmsln-studio-margin, 20px))\n            var(--elmsln-studio-margin, 20px);\n          margin-bottom: 15px;\n        }\n        nav-card-item:not([avatar]) .assignment-link:before {\n          position: absolute;\n          overflow: hidden;\n          display: inline-block;\n          -webkit-animation: ellipsis steps(4, end) 900ms infinite;\n          animation: ellipsis steps(4, end) 900ms infinite;\n          content: \"...\";\n          font-size: 300%;\n          left: 0;\n          top: 0;\n          width: 0px;\n        }\n\n        @keyframes ellipsis {\n          to {\n            width: 1.25em;\n          }\n        }\n\n        @-webkit-keyframes ellipsis {\n          to {\n            width: 1.25em;\n          }\n        }\n        @media screen and (min-width: 400px) {\n          #lessons {\n            display: flex;\n            align-items: stretch;\n            justify-content: flex-start;\n            overflow-x: auto;\n            max-width: 100%;\n            padding: 0 calc(0.5 * var(--elmsln-studio-margin, 20px));\n          }\n          .lesson {\n            width: 300px;\n            flex: 0 0 300px;\n            margin: 0 var(--elmsln-studio-margin, 20px);\n            border-bottom: 1px solid var(--accent-card-border-color, #ddd);\n          }\n        }\n      "]);

    _templateObject6_9ef74af056b811eb8081cb2f2fa80813 = function _templateObject6_9ef74af056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _templateObject5_9ef74af056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral(["\n          <nav-card-item\n            id=\"act-", "-item\"\n            accent-color=\"", "\"\n            allow-grey\n            avatar=\"", "\"\n            invert\n          >\n            <elmsln-studio-link\n              id=\"act-", "\"\n              class=\"assignment-link\"\n              aria-describedby=\"act-", "-desc act-", "-item\"\n              slot=\"label\"\n              href=\"/assignments/", "\"\n            >\n              ", "\n            </elmsln-studio-link>\n            <div id=\"act-", "-desc\" slot=\"description\">\n              Due:\n              <local-time\n                month=\"long\"\n                day=\"numeric\"\n                year=\"numeric\"\n                hour=\"2-digit\"\n                minute=\"2-digit\"\n                second=\"2-digit\"\n                time-zone-name=\"short\"\n                datetime=\"", "\"\n              >\n                ", "\n              </local-time>\n            </div>\n          </nav-card-item>\n          <simple-tooltip for=\"act-", "\" position=\"left\">\n            ", "\n          </simple-tooltip>\n        "]);

    _templateObject5_9ef74af056b811eb8081cb2f2fa80813 = function _templateObject5_9ef74af056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _templateObject4_9ef74af056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                          <nav-card flat no-border class=\"card secondary\">\n                            <span slot=\"heading\" ?hidden=\"", "\">\n                              ", "\n                            </span>\n                            <div slot=\"linklist\">\n                              ", "\n                            </div>\n                          </nav-card>\n                        "]);

    _templateObject4_9ef74af056b811eb8081cb2f2fa80813 = function _templateObject4_9ef74af056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _templateObject3_9ef74af056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                          <div class=\"assignment\">\n                            ", "\n                          </div>\n                        "]);

    _templateObject3_9ef74af056b811eb8081cb2f2fa80813 = function _templateObject3_9ef74af056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _templateObject2_9ef74af056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                <div class=\"lesson\">\n                  <h2>", "</h2>\n                  ", "\n                </div>\n              "]);

    _templateObject2_9ef74af056b811eb8081cb2f2fa80813 = function _templateObject2_9ef74af056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _templateObject_9ef74af056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <h1 class=\"sr-only\">Assignments</h1>\n      <div id=\"lessons\">\n        ", "\n      </div>\n    "]);

    _templateObject_9ef74af056b811eb8081cb2f2fa80813 = function _templateObject_9ef74af056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `elmsln-studio-assignments`
   * Studio App for ELMS:LN
   *
   * @customElement elmsln-studio-assignments
   * @lit-html
   * @lit-element
   * @demo demo/dashboard.html
   */
  var ElmslnStudioAssignments = /*#__PURE__*/function (_ElmslnStudioUtilitie) {
    babelHelpers.inherits(ElmslnStudioAssignments, _ElmslnStudioUtilitie);

    var _super = _createSuper(ElmslnStudioAssignments);

    babelHelpers.createClass(ElmslnStudioAssignments, [{
      key: "render",
      // render function

      /*
          
       */
      value: function render() {
        var _this2 = this;

        return (0, _litElement.html)(_templateObject_9ef74af056b811eb8081cb2f2fa80813(), !this.lessons ? this.loading("green", undefined, "large") : Object.keys(this.lessons || {}).map(function (l) {
          return (0, _litElement.html)(_templateObject2_9ef74af056b811eb8081cb2f2fa80813(), _this2.lessons[l].lesson, (_this2.lessons[l].assignments || []).map(function (p) {
            return !p.assignments ? (0, _litElement.html)(_templateObject3_9ef74af056b811eb8081cb2f2fa80813(), _this2.renderAssignment(p)) : (0, _litElement.html)(_templateObject4_9ef74af056b811eb8081cb2f2fa80813(), !p.project, p.project, _this2.sortDates(p.assignments || [], true).map(function (a) {
              return _this2.renderAssignment(a);
            }));
          }));
        }));
      }
    }, {
      key: "renderAssignment",
      value: function renderAssignment(assignment) {
        return !assignment ? "" : (0, _litElement.html)(_templateObject5_9ef74af056b811eb8081cb2f2fa80813(), assignment.id, !this.profile ? "grey" : this.getStatusColor(this.getSubmission(assignment.id), assignment), !this.profile ? undefined : this.getStatusIcon(this.getSubmission(assignment.id), assignment), assignment.id, assignment.id, assignment.id, assignment.id, assignment.assignment, assignment.id, assignment.date, this.dateFormat(assignment.date), assignment.id, this.getStatusMessage(this.getSubmission(assignment.id), assignment));
      } // properties available to the custom element for data binding

    }], [{
      key: "tag",

      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */
      get: function get() {
        return "elmsln-studio-assignments";
      }
    }, {
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioAssignments), "styles", this)), [(0, _litElement.css)(_templateObject6_9ef74af056b811eb8081cb2f2fa80813())]);
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioAssignments), "properties", this)), {}, {
          lessons: {
            type: Object
          },
          profile: {
            type: Object
          }
        });
      } // life cycle

    }]);

    function ElmslnStudioAssignments() {
      var _this;

      babelHelpers.classCallCheck(this, ElmslnStudioAssignments);
      _this = _super.call(this);
      _this.tag = ElmslnStudioAssignments.tag;
      return _this;
    }

    babelHelpers.createClass(ElmslnStudioAssignments, [{
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        if (babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioAssignments.prototype), "firstUpdated", this)) babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioAssignments.prototype), "firstUpdated", this).call(this, changedProperties);
        this.fetchData("profile");
        this.fetchData("lessons");
      }
    }, {
      key: "getSubmission",
      value: function getSubmission(id) {
        var _this3 = this;

        var submissions = this.profile && this.profile.id && this.profile.submissions ? this.profile.submissions.filter(function (s) {
          return s.id === "".concat(id, "-").concat(_this3.profile.id);
        }) : [];
        return submissions.length > 0 ? submissions[0] : undefined;
      } // static get observedAttributes() {
      //   return [];
      // }
      // disconnectedCallback() {}
      // attributeChangedCallback(attr, oldValue, newValue) {}

    }]);
    return ElmslnStudioAssignments;
  }((0, _elmslnStudioUtilities.ElmslnStudioUtilities)((0, _elmslnStudioStyles.ElmslnStudioStyles)(_litElement.LitElement)));

  _exports.ElmslnStudioAssignments = ElmslnStudioAssignments;
  customElements.define("elmsln-studio-assignments", ElmslnStudioAssignments);
});
