// @lrnwebcomponents/elmsln-studio/lib/elmsln-studio-modal-button.js
define(["exports", "../../../lit-element/lit-element.js", "../../../@polymer/iron-icon/iron-icon.js", "../../../@polymer/iron-icons/iron-icons.js"], function (_exports, _litElement, _ironIcon, _ironIcons) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmslnStudioModalButton = void 0;

  function _templateObject2_2f6379a0de5b11eaa3c04df0985e08b7() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <button\n        id=\"modal\"\n        @click=\"", "\"\n        alignment=\"", "\"\n        ?big=\"", "\"\n        ?border=\"", "\"\n        ?flexible=\"", "\"\n      >\n        <p>\n          <iron-icon aria-hidden=\"true\" icon=\"", "\"></iron-icon>\n          <span class=\"", "\"\n            >", "</span\n          >\n        </p>\n      </button>\n    "]);

    _templateObject2_2f6379a0de5b11eaa3c04df0985e08b7 = function _templateObject2_2f6379a0de5b11eaa3c04df0985e08b7() {
      return data;
    };

    return data;
  }

  function _templateObject_2f6379a0de5b11eaa3c04df0985e08b7() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        :host([hidden]),\n        *[hidden] {\n          display: none !important;\n        }\n        :host([flexible]) {\n          flex: 1 1 auto;\n        }\n        button {\n          border: none;\n          background-color: white;\n          width: 100%;\n        }\n        button p {\n          margin: 0;\n          display: flex;\n          align-items: center;\n          justify-content: center;\n        }\n        button[alignment=\"left\"] p {\n          justify-content: flex-start;\n        }\n        button[alignment=\"right\"] p {\n          justify-content: end;\n          flex-direction: row-reverse;\n        }\n        button[border] {\n          border: 1px solid #ddd;\n        }\n        button > * {\n          min-height: 50px;\n          line-height: 50px;\n        }\n        button[big] > * {\n          min-height: 75px;\n          line-height: 75px;\n        }\n        .sr-only {\n          position: absolute;\n          left: -9999999px;\n          width: 0;\n          overflow: hidden;\n        }\n      "]);

    _templateObject_2f6379a0de5b11eaa3c04df0985e08b7 = function _templateObject_2f6379a0de5b11eaa3c04df0985e08b7() {
      return data;
    };

    return data;
  }

  /**
   * `elmsln-studio-modal-button`
   * Navigation Link for ELMS:LN Studio
   *
   * @customElement elmsln-studio-modal-button
   * @lit-html
   * @lit-element
   */
  var ElmslnStudioModalButton =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(ElmslnStudioModalButton, _LitElement);
    babelHelpers.createClass(ElmslnStudioModalButton, null, [{
      key: "tag",

      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */
      get: function get() {
        return "elmsln-studio-modal-button";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          text: {
            type: String
          },
          icon: {
            type: String
          },
          callback: {
            type: String
          },
          hideText: {
            type: Boolean,
            attribute: "hide-text"
          },
          flexible: {
            type: Boolean,
            attribute: "flexible",
            reflect: true
          },
          border: {
            type: Boolean,
            attribute: "border"
          },
          align: {
            type: String,
            attribute: "border"
          },
          big: {
            type: Boolean,
            attribute: "big"
          }
        };
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _litElement.css)(_templateObject_2f6379a0de5b11eaa3c04df0985e08b7())];
      }
    }]);

    function ElmslnStudioModalButton() {
      var _this;

      babelHelpers.classCallCheck(this, ElmslnStudioModalButton);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(ElmslnStudioModalButton).call(this));
      _this.hideText = false;
      _this.big = false;
      _this.icon = false;
      _this.flexible = false;
      _this.border = false;
      _this.align = "center";
      return _this;
    }

    babelHelpers.createClass(ElmslnStudioModalButton, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject2_2f6379a0de5b11eaa3c04df0985e08b7(), this.callback, this.align, this.big, this.border, this.flexible, this.icon, this.hideText && this.icon ? "sr-only" : "", this.text);
      }
    }]);
    return ElmslnStudioModalButton;
  }(_litElement.LitElement);

  _exports.ElmslnStudioModalButton = ElmslnStudioModalButton;
  customElements.define("elmsln-studio-modal-button", ElmslnStudioModalButton);
});
