// @lrnwebcomponents/elmsln-studio/lib/elmsln-studio-portfolio.js
define(["exports", "../../../lit-element/lit-element.js", "./elmsln-studio-styles.js", "./elmsln-studio-utilities.js", "../../img-view-modal/img-view-modal.js", "../../hax-iconset/hax-iconset.js", "../../threaded-discussion/threaded-discussion.js"], function (_exports, _litElement, _elmslnStudioStyles, _elmslnStudioUtilities, _imgViewModal, _haxIconset, _threadedDiscussion) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmslnStudioPortfolio = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject12_9f249c8056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          flex-wrap: wrap;\n        }\n        h1 {\n          text-align: center;\n        }\n        #breadcrumb {\n          margin: 0 0 var(--elmsln-studio-margin, 20px);\n          flex: 1 0 100%;\n          --elmsln-studio-link-Color: var(\n            --simple-colors-fixed-theme-light-blue-8\n          );\n          --elmsln-studio-link-focus-Color: var(\n            --simple-colors-fixed-theme-light-blue-9\n          );\n        }\n        #breadcrumb > * {\n          display: inline;\n        }\n        lrndesign-avatar,\n        .student-name,\n        .project-name,\n        .assignment-name,\n        .submission-date {\n          font-weight: normal;\n          display: block;\n        }\n        .student-name {\n          font-size: calc(2 * var(--elmsln-studio-FontSize, 16px));\n          font-weight: bold;\n          color: #4b4b4b;\n        }\n        .project-name {\n          font-size: calc(1.75 * var(--elmsln-studio-FontSize, 16px));\n          color: #95989a;\n        }\n        #project-buttons,\n        .submission-header {\n          position: relative;\n          height: 40px;\n          margin-top: -40px;\n        }\n        #sort,\n        .submission-header elmsln-studio-button {\n          color: #95989a;\n          font-size: var(--elmsln-studio-FontSize, 16px);\n          text-transform: uppercase;\n          position: absolute;\n          right: 0;\n          top: 0;\n        }\n        #sort:focus,\n        #sort:hover,\n        .submission-header elmsln-studio-button:focus,\n        .submission-header elmsln-studio-button:hover {\n          color: #4b4b4b;\n        }\n        #sort.sort-latest {\n          transform: rotateX(180deg) rotateY(0deg);\n        }\n        section {\n          border-top: 2px solid #eaeaea;\n          padding: calc(0.5 * var(--elmsln-studio-margin, 20px)) 0\n            var(--elmsln-studio-margin, 20px);\n        }\n        .view-discussion {\n          border: 1px solid #eaeaea;\n          padding: calc(0.5 * var(--elmsln-studio-margin, 20px));\n          --elmsln-studio-discussion-multiplier: 1;\n        }\n        .view-discussion section {\n          opacity: 0.4;\n          display: none;\n        }\n        .view-discussion section.section-discussion {\n          display: block;\n          opacity: 1;\n        }\n        h2 {\n          margin: calc(2 * var(--elmsln-studio-margin, 20px)) auto;\n          text-align: center;\n        }\n        .assignment-name {\n          font-size: calc(\n            1.25 * var(--elmsln-studio-discussion-multiplier, 1) *\n              var(--elmsln-studio-FontSize, 16px)\n          );\n          color: #555555;\n        }\n        .submission-date {\n          font-size: calc(\n            0.75 * var(--elmsln-studio-discussion-multiplier, 1) *\n              var(--elmsln-studio-FontSize, 16px)\n          );\n          color: #95989a;\n        }\n        .submission-body {\n          color: #95989a;\n          line-height: 160%;\n          margin: calc(0.5 * var(--elmsln-studio-margin, 20px)) auto\n            var(--elmsln-studio-margin, 20px);\n        }\n        .submission-body:first-child {\n          margin-top: 0;\n        }\n        .submission-body:last-child {\n          margin-bottom: 0;\n        }\n        .submission-links {\n          list-style: none;\n          padding-inline-start: 0;\n        }\n        .submission-links > li {\n          display: flex;\n          align-items: flex-start;\n          justify-content: flex-start;\n        }\n        .submission-links a:link,\n        .submission-links a:visited {\n          font-size: calc(1.5 * var(--elmsln-studio-FontSize, 16px));\n          font-weight: bold;\n          color: #000;\n          text-decoration: none;\n        }\n        .submission-links a:focus,\n        .submission-links a:hover {\n          text-decoration: underline;\n        }\n        .submission-links simple-icon-lite {\n          margin-right: 0.5em;\n        }\n        .submission-links a:focus simple-icon-lite,\n        .submission-links a:hover simple-icon-lite {\n          text-decoration: none;\n        }\n        .callout {\n          font-size: calc(0.75 * var(--elmsln-studio-FontSize, 16px));\n          border: 1px solid #eaeaea;\n          padding: 0;\n        }\n        .discussion-label {\n          text-align: left;\n          font-size: calc(1 * var(--elmsln-studio-FontSize, 16px));\n          margin: var(--elmsln-studio-margin, 20px) 0\n            calc(0.25 * var(--elmsln-studio-margin, 20px));\n        }\n        .callout > * {\n          padding: calc(0.25 * var(--elmsln-studio-margin, 20px))\n            calc(0.5 * var(--elmsln-studio-margin, 20px));\n        }\n        .callout .callout-label {\n          font-size: calc(1 * var(--elmsln-studio-FontSize, 16px));\n          font-weight: normal;\n          display: flex;\n          align-items: center;\n          justify-content: flex-start;\n          margin: 0;\n          color: #4b4b4b;\n          border-bottom: 1px solid #eaeaea;\n        }\n        .callout .callout-label simple-icon-lite {\n          margin-right: 1em;\n        }\n        threaded-discussion {\n          background-color: #eaeaea;\n          padding: 1px;\n          --threaded-discussion-comment-button-BackgroundColor: var(\n            --simple-colors-fixed-theme-light-blue-8\n          );\n          --threaded-discussion-comment-button-focus-BackgroundColor: var(\n            --simple-colors-fixed-theme-light-blue-9\n          );\n        }\n        #prev-next-nav {\n          z-index: 10000;\n          display: flex;\n        }\n        @media screen and (min-width: 600px) {\n          :host {\n            display: flex;\n            align-items: flex-start;\n            justify-content: space-between;\n          }\n          #primary {\n            flex: 1 0 100%;\n          }\n          #primary.view-discussion {\n            flex: 0 0 calc(50% - var(--elmsln-studio-margin, 20px));\n          }\n          #primary:not(.view-discussion) {\n            max-width: calc(1000px - 2 * var(--elmsln-studio-margin, 20px));\n            margin: 0 auto;\n            border: 1px solid #eaeaea;\n            padding: var(--elmsln-studio-margin, 20px);\n          }\n          .view-discussion {\n            --elmsln-studio-discussion-multiplier: 0.5;\n          }\n          .assignment-name {\n            font-size: calc(\n              1.5 * var(--elmsln-studio-discussion-multiplier, 1) *\n                var(--elmsln-studio-FontSize, 16px)\n            );\n          }\n          .submission-date {\n            font-size: calc(\n              1 * var(--elmsln-studio-discussion-multiplier, 1) *\n                var(--elmsln-studio-FontSize, 16px)\n            );\n          }\n          .view-discussion section {\n            display: block;\n          }\n          .view-discussion section.section-discussion {\n            border: 4px solid #95989a;\n            padding: calc(0.5 * var(--elmsln-studio-margin, 20px))\n              calc(0.5 * var(--elmsln-studio-margin, 20px))\n              var(--elmsln-studio-margin, 20px);\n            opacity: 1;\n          }\n          #secondary {\n            flex: 0 0 calc(50% - var(--elmsln-studio-margin, 20px));\n          }\n        }\n        @media screen and (min-width: 1000px) {\n          .view-discussion {\n            --elmsln-studio-discussion-multiplier: 0.75;\n          }\n        }\n        @media screen and (min-width: 1200px) {\n          .view-discussion {\n            --elmsln-studio-discussion-multiplier: 1;\n          }\n        }\n      "]);

    _templateObject12_9f249c8056b811eb8081cb2f2fa80813 = function _templateObject12_9f249c8056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _templateObject11_9f249c8056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral([" ", " "]);

    _templateObject11_9f249c8056b811eb8081cb2f2fa80813 = function _templateObject11_9f249c8056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _templateObject10_9f249c8056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                              <li>\n                                <elmsln-studio-link\n                                  href=\"", "\"\n                                  target=\"_blank\"\n                                  ?disabled=\"", "\"\n                                >\n                                  <simple-icon-lite\n                                    aria-hidden=\"true\"\n                                    icon=\"", "\"\n                                  ></simple-icon-lite>\n                                  ", "\n                                </elmsln-studio-link>\n                              </li>\n                            "]);

    _templateObject10_9f249c8056b811eb8081cb2f2fa80813 = function _templateObject10_9f249c8056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _templateObject9_9f249c8056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                        <ul class=\"submission-links\">\n                          ", "\n                        </ul>\n                      "]);

    _templateObject9_9f249c8056b811eb8081cb2f2fa80813 = function _templateObject9_9f249c8056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _templateObject8_9f249c8056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <section\n                class=\"", "\"\n              >\n                <div class=\"submission-header\">\n                  <h2 id=\"sub-", "\">\n                    <span class=\"assignment-name\">", "</span>\n                    <span class=\"submission-date\"\n                      >Submitted:\n                      <local-time\n                        month=\"long\"\n                        day=\"numeric\"\n                        year=\"numeric\"\n                        hour=\"2-digit\"\n                        minute=\"2-digit\"\n                        second=\"2-digit\"\n                        time-zone-name=\"short\"\n                        datetime=\"", "\"\n                      >\n                        ", "\n                      </local-time>\n                    </span>\n                  </h2>\n                  <elmsln-studio-button\n                    id=\"sub-", "-toggle-button\"\n                    class=\"view-discussion-button\"\n                    aria-describedby=\"sub-", "\"\n                    icon=\"", "\"\n                    path=\"", "\"\n                  >\n                    <span class=\"sr-only\"\n                      >", "</span\n                    >\n                    <span class=\"sr-only\"\n                      >Give / View Feedback (", ")</span\n                    >\n                  </elmsln-studio-button>\n                </div>\n                <div class=\"submission-body\">\n                  ", "\n                </div>\n              </section>\n            "]);

    _templateObject8_9f249c8056b811eb8081cb2f2fa80813 = function _templateObject8_9f249c8056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _templateObject7_9f249c8056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <span class=\"project-name\">", "</span>\n                "]);

    _templateObject7_9f249c8056b811eb8081cb2f2fa80813 = function _templateObject7_9f249c8056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _templateObject6_9f249c8056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral(["<em>Student</em>"]);

    _templateObject6_9f249c8056b811eb8081cb2f2fa80813 = function _templateObject6_9f249c8056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _templateObject5_9f249c8056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <elmsln-studio-link\n                href=\"", "\"\n                >", "</elmsln-studio-link\n              >\n            "]);

    _templateObject5_9f249c8056b811eb8081cb2f2fa80813 = function _templateObject5_9f249c8056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _templateObject4_9f249c8056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral(["<em>Assignment</em>"]);

    _templateObject4_9f249c8056b811eb8081cb2f2fa80813 = function _templateObject4_9f249c8056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _templateObject3_9f249c8056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                    <elmsln-studio-link\n                      href=\"", "\"\n                      >", "</elmsln-studio-link\n                    >\n                    <span> > </span>\n                  "]);

    _templateObject3_9f249c8056b811eb8081cb2f2fa80813 = function _templateObject3_9f249c8056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _templateObject2_9f249c8056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              ", "\n            "]);

    _templateObject2_9f249c8056b811eb8081cb2f2fa80813 = function _templateObject2_9f249c8056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _templateObject_9f249c8056b811eb8081cb2f2fa80813() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div id=\"breadcrumb\">\n        <elmsln-studio-link href=\"/submissions\">Submissions</elmsln-studio-link>\n        <span> > </span>\n        ", "\n        ", "\n        <span> > </span>\n        <span\n          >", "</span\n        >\n      </div>\n      <div\n        id=\"primary\"\n        class=\"", "\"\n      >\n        <article id=\"portfolio-project\">\n          <h1>\n            <lrndesign-avatar\n              accent-color=\"", "\"\n              aria-hidden=\"true\"\n              label=\"", "\"\n              .src=\"", "\"\n              two-chars\n            >\n            </lrndesign-avatar>\n            <span class=\"student-name\"\n              >", "</span\n            >\n            ", "\n          </h1>\n          <div id=\"project-buttons\">\n            <button\n              id=\"sort\"\n              aria-pressed=\"", "\"\n              class=\"", "\"\n              @click=\"", "\"\n              controls=\"portfolio-project\"\n            >\n              <simple-icon-lite icon=\"sort\"></simple-icon-lite>\n              <simple-tooltip for=\"sort\"\n                >Sort Submissions\n                ", "</simple-tooltip\n              >\n            </button>\n          </div>\n          ", "\n        </article>\n      </div>\n      <div\n        id=\"secondary\"\n        ?hidden=", "\n      >\n        <aside>\n          <div class=\"instructions callout\">\n            <h2 class=\"callout-label\">\n              <simple-icon-lite\n                icon=\"info\"\n                aria-hidden=\"true\"\n              ></simple-icon-lite>\n              Giving Feedback\n            </h2>\n            <div>\n              <ul>\n                <li>Things to look for: blah, blah, blah</li>\n                <li>How to say it: blah, blah, blah</li>\n                <li>Other info.</li>\n              </ul>\n            </div>\n          </div>\n          <h2 class=\"discussion-label\">Feedback</h2>\n          <div class=\"discussion callout\">\n            <p class=\"callout-label\">\n              <simple-icon-lite\n                icon=\"communication:comment\"\n                aria-hidden=\"true\"\n              ></simple-icon-lite>\n              Post New Feedback\n            </p>\n            <threaded-discussion\n              comment-icon=\"send\"\n              ?demo=\"", "\"\n              .data=\"", "\"\n              latest\n            >\n            </threaded-discussion>\n          </div>\n        </aside>\n      </div>\n    "]);

    _templateObject_9f249c8056b811eb8081cb2f2fa80813 = function _templateObject_9f249c8056b811eb8081cb2f2fa80813() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `elmsln-studio-portfolio`
   * Studio App for ELMS:LN
   *
   * @customElement elmsln-studio-portfolio
   * @lit-html
   * @lit-element
   * @demo demo/portfolio.html
   */
  var ElmslnStudioPortfolio = /*#__PURE__*/function (_ElmslnStudioUtilitie) {
    babelHelpers.inherits(ElmslnStudioPortfolio, _ElmslnStudioUtilitie);

    var _super = _createSuper(ElmslnStudioPortfolio);

    babelHelpers.createClass(ElmslnStudioPortfolio, [{
      key: "render",
      // render function
      value: function render() {
        var _this2 = this;

        return (0, _litElement.html)(_templateObject_9f249c8056b811eb8081cb2f2fa80813(), !this.portfolio ? "" : (0, _litElement.html)(_templateObject2_9f249c8056b811eb8081cb2f2fa80813(), !this.portfolio.projectId || !this.portfolio.project ? "" : (0, _litElement.html)(_templateObject3_9f249c8056b811eb8081cb2f2fa80813(), "/submissions?project=".concat(this.portfolio.projectId), this.portfolio.project || "Project")), !this.portfolio || !this.assignment ? (0, _litElement.html)(_templateObject4_9f249c8056b811eb8081cb2f2fa80813()) : (0, _litElement.html)(_templateObject5_9f249c8056b811eb8081cb2f2fa80813(), "/submissions?".concat(!this.portfolio.projectId || !this.portfolio.project ? "" : "project=".concat(this.portfolio.projectId, "&"), "assignment=").concat(this.assignment.assignmentId), this.assignment.assignment || "Assignment"), !this.portfolio ? (0, _litElement.html)(_templateObject6_9f249c8056b811eb8081cb2f2fa80813()) : this.fullName(this.portfolio), this.comment && this.comment !== "" ? "view-discussion" : "", !this.portfolio ? "grey" : this.accentColor(this.fullName(this.portfolio)), !this.portfolio ? "undefined" : this.fullName(this.portfolio), this.portfolio && this.portfolio.avatar ? this.portfolio.avatar : undefined, !this.portfolio ? "Loading..." : this.fullName(this.portfolio), !this.portfolio || !this.portfolio.project ? "" : (0, _litElement.html)(_templateObject7_9f249c8056b811eb8081cb2f2fa80813(), this.portfolio.project), this.sortLatest ? "false" : "true", !this.sortLatest ? "" : "sort-latest", function (e) {
          return _this2.sortLatest = !_this2.sortLatest;
        }, !this.sortLatest ? "Oldest First" : "Newest First", this.sortedSubmissions.map(function (s) {
          return (0, _litElement.html)(_templateObject8_9f249c8056b811eb8081cb2f2fa80813(), _this2.submissionFilter === s.id ? "section-discussion" : "", s.id, s.assignment, s.date, _this2.dateFormat(s.date), s.id, s.id, _this2.submissionFilter === s.id && _this2.comment ? "close" : _this2.getFeedbackIcon(s.feedback.length), _this2.getActivityLink(s, _this2.submissionFilter === s.id && _this2.comment), _this2.submissionFilter === s.id && _this2.comment ? "Close Feedback" : "View Feedback (".concat(s.feedback.length, ")"), s.feedback.length, s.links && s.links.length > 0 ? (0, _litElement.html)(_templateObject9_9f249c8056b811eb8081cb2f2fa80813(), s.links.map(function (link) {
            return (0, _litElement.html)(_templateObject10_9f249c8056b811eb8081cb2f2fa80813(), link.url, _this2.submissionFilter !== s.id && _this2.comment, link.type === "pdf" ? "hax:file-pdf" : "link", link.text || link.url);
          })) : !s.sources || s.sources.length === 0 ? (0, _litElement.html)(_templateObject11_9f249c8056b811eb8081cb2f2fa80813(), s.body) : _this2.getThumnailGrid(s, _this2.submissionFilter !== s.id && _this2.comment));
        }), !this.portfolio || !this.comment || this.comment === "", this.demoMode, this.submission || []);
      } // properties available to the custom element for data binding

    }], [{
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioPortfolio), "styles", this)), [(0, _litElement.css)(_templateObject12_9f249c8056b811eb8081cb2f2fa80813())]);
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioPortfolio), "properties", this)), {}, {
          portfolio: {
            type: Object
          },
          feedback: {
            type: Object
          },
          comment: {
            type: String,
            attribute: "comment"
          },
          submissions: {
            type: Array
          },
          submissionFilter: {
            type: String,
            attribute: "submission-filter"
          },
          sortLatest: {
            type: Boolean,
            attribute: "sort-latest",
            reflect: true
          }
        });
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */

    }, {
      key: "tag",
      get: function get() {
        return "elmsln-studio-portfolio";
      } // life cycle

    }]);

    function ElmslnStudioPortfolio() {
      var _this;

      babelHelpers.classCallCheck(this, ElmslnStudioPortfolio);
      _this = _super.call(this);
      _this.feedback = {};
      _this.comment = "";
      _this.tag = ElmslnStudioPortfolio.tag;
      return _this;
    }

    babelHelpers.createClass(ElmslnStudioPortfolio, [{
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        if (babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioPortfolio.prototype), "firstUpdated", this)) babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioPortfolio.prototype), "firstUpdated", this).call(this, changedProperties);
        this.fetchData("submissions");
        this.fetchData("portfolios");
        this.fetchData("discussion");
        console.log("updated", this.submissions, this.portfolios, this.discussion);
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        if (babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioPortfolio.prototype), "updated", this)) babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioPortfolio.prototype), "updated", this).call(this, changedProperties);
        changedProperties.forEach(function (oldValue, propName) {});
      }
    }, {
      key: "nav",
      get: function get() {
        var _this3 = this;

        var prevHref,
            nextHref,
            portfolio = this.portfolio || {},
            getLink = function getLink(s) {
          var assignmentId = _this3.submissionFilter ? _this3.submissionFilter.replace(/-\w+$/, "") : portfolio.assignmentId,
              submissionId = "".concat(assignmentId, "-").concat(s.userId),
              submissions = !assignmentId || !s.submissions ? [] : s.submissions.filter(function (sub) {
            return sub.id === submissionId;
          }),
              submission = submissions.length > 0 ? submissions[0] : {};
          submission.activity = "submission";
          return _this3.getActivityLink(submission, !_this3.comment, _this3.sortLatest);
        },
            prevLabel = !portfolio.prev ? "" : this.fullName(portfolio.prev),
            nextLabel = !portfolio.next ? "" : this.fullName(portfolio.next);

        prevHref = portfolio.prev ? getLink(portfolio.prev) : undefined, nextHref = portfolio.next ? getLink(portfolio.next) : undefined;
        return this.prevNextNav(prevLabel, prevHref, nextLabel, nextHref);
      }
    }, {
      key: "sortedSubmissions",
      get: function get() {
        return !this.portfolio || !this.portfolio.submissions ? [] : this.sortDates(this.portfolio.submissions, this.sortLatest);
      }
    }, {
      key: "assignment",
      get: function get() {
        var _this4 = this;

        var filter = !this.portfolio || !this.portfolio.submissions ? [] : this.portfolio.submissions.filter(function (s) {
          return !_this4.submissionFilter || s.id === _this4.submissionFilter;
        });
        return !filter ? false : filter[0];
      }
    }]);
    return ElmslnStudioPortfolio;
  }((0, _elmslnStudioUtilities.ElmslnStudioUtilities)((0, _elmslnStudioStyles.ElmslnStudioStyles)(_litElement.LitElement)));

  _exports.ElmslnStudioPortfolio = ElmslnStudioPortfolio;
  customElements.define("elmsln-studio-portfolio", ElmslnStudioPortfolio);
});
