// @lrnwebcomponents/elmsln-studio/lib/elmsln-studio-button.js
define(["exports", "../../../lit-element/lit-element.js", "../../../lit-element-router/lit-element-router.js", "../../simple-icon/lib/simple-icon-lite.js"], function (_exports, _litElement, _litElementRouter, _simpleIconLite) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmslnStudioButton = void 0;

  function _templateObject3_dac84970576311eba9ce094baa5ca644() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <simple-icon-lite\n                aria-hidden=\"true\"\n                icon=\"", "\"\n              ></simple-icon-lite>\n            "]);

    _templateObject3_dac84970576311eba9ce094baa5ca644 = function _templateObject3_dac84970576311eba9ce094baa5ca644() {
      return data;
    };

    return data;
  }

  function _templateObject2_dac84970576311eba9ce094baa5ca644() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <button @click=\"", "\">\n        <slot name=\"before\"></slot>\n        ", "\n        <slot></slot>\n      </button>\n    "]);

    _templateObject2_dac84970576311eba9ce094baa5ca644 = function _templateObject2_dac84970576311eba9ce094baa5ca644() {
      return data;
    };

    return data;
  }

  function _templateObject_dac84970576311eba9ce094baa5ca644() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      :host {\n        display: block;\n      }\n      :host([hidden]) {\n        display: none;\n      }\n      button {\n        border: none;\n        background-color: transparent;\n        display: flex;\n        align-items: center;\n        height: 26px;\n        color: var(--elmsln-studio-button-icon-color);\n      }\n      button:focus,\n      button:hover {\n        outline: none;\n      }\n    "]);

    _templateObject_dac84970576311eba9ce094baa5ca644 = function _templateObject_dac84970576311eba9ce094baa5ca644() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `elmsln-studio-button`
   * Navigation Link for ELMS:LN Studio
   *
   * @customElement elmsln-studio-button
   * @lit-html
   * @lit-element
   */
  var ElmslnStudioButton = /*#__PURE__*/function (_navigator) {
    babelHelpers.inherits(ElmslnStudioButton, _navigator);

    var _super = _createSuper(ElmslnStudioButton);

    babelHelpers.createClass(ElmslnStudioButton, null, [{
      key: "tag",

      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */
      get: function get() {
        return "elmsln-studio-button";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          path: {
            type: String
          },
          icon: {
            type: String
          }
        };
      }
    }, {
      key: "styles",
      get: function get() {
        return (0, _litElement.css)(_templateObject_dac84970576311eba9ce094baa5ca644());
      }
    }]);

    function ElmslnStudioButton() {
      var _this;

      babelHelpers.classCallCheck(this, ElmslnStudioButton);
      _this = _super.call(this);
      _this.path = "";
      _this.icon = "";
      return _this;
    }

    babelHelpers.createClass(ElmslnStudioButton, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject2_dac84970576311eba9ce094baa5ca644(), this.buttonClick, !this.icon ? "" : (0, _litElement.html)(_templateObject3_dac84970576311eba9ce094baa5ca644(), this.icon));
      }
    }, {
      key: "buttonClick",
      value: function buttonClick(event) {
        this.navigate("".concat(window.ElmslnStudioPath || "").concat(this.path));
      }
    }]);
    return ElmslnStudioButton;
  }((0, _litElementRouter.navigator)(_litElement.LitElement));

  _exports.ElmslnStudioButton = ElmslnStudioButton;
  customElements.define("elmsln-studio-button", ElmslnStudioButton);
});
