// @lrnwebcomponents/elmsln-studio/lib/elmsln-studio-link.js
define(["exports", "../../../lit-element/lit-element.js", "../../../lit-element-router/lit-element-router.js"], function (_exports, _litElement, _litElementRouter) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmslnStudioLink = void 0;

  function _templateObject2_3dfdab60019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <a href=\"", "\" @click=\"", "\">\n        <slot></slot>\n      </a>\n    "]);

    _templateObject2_3dfdab60019d11eb9027c17497ee62cd = function _templateObject2_3dfdab60019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _templateObject_3dfdab60019d11eb9027c17497ee62cd() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      :host {\n        display: block;\n      }\n      :host([hidden]) {\n        display: none;\n      }\n      a,\n      a:link {\n        transition: var(--elmsln-studio-link-Transition);\n        background: transparent;\n        color: var(--elmsln-studio-link-Color);\n        text-decoration: var(--elmsln-studio-link-TextDecoration);\n        text-decoration-style: var(--elmsln-studio-link-TextDecorationStyle);\n        text-decoration-skip: var(--elmsln-studio-link-TextDecorationSkip);\n        text-decoration-skip-ink: var(\n          --elmsln-studio-link-TextDecorationSkipInk\n        );\n        text-decoration-line: var(--elmsln-studio-link-TextDecorationLine);\n        text-decoration-color: var(--elmsln-studio-link-TextDecorationColor);\n        text-decoration: var(--elmsln-studio-link-TextDecoration);\n        text-decoration-thickness: var(\n          --elmsln-studio-link-TextDecorationThickness\n        );\n      }\n      a:hover,\n      a:focus {\n        background: transparent;\n        color: var(\n          --elmsln-studio-link-focus-Color,\n          var(--elmsln-studio-link-Color)\n        );\n        text-decoration: var(\n          --elmsln-studio-link-focus-TextDecoration,\n          var(--elmsln-studio-link-TextDecoration)\n        );\n        text-decoration-style: var(\n          --elmsln-studio-link-focus-TextDecorationStyle,\n          var(--elmsln-studio-link-TextDecorationStyle)\n        );\n        text-decoration-skip: var(\n          --elmsln-studio-link-focus-TextDecorationSkip,\n          var(--elmsln-studio-link-TextDecorationSkip)\n        );\n        text-decoration-skip-ink: var(\n          --elmsln-studio-link-focus-TextDecorationSkipInk,\n          var(--elmsln-studio-link-TextDecorationSkipInk)\n        );\n        text-decoration-line: var(\n          --elmsln-studio-link-focus-TextDecorationLine,\n          var(--elmsln-studio-link-TextDecorationLine)\n        );\n        text-decoration-color: var(\n          --elmsln-studio-link-focus-TextDecorationColor,\n          ar(--elmsln-studio-link-TextDecorationColor)\n        );\n        text-decoration: var(\n          --elmsln-studio-link-focus-TextDecoration,\n          var(--elmsln-studio-link-TextDecoration)\n        );\n        text-decoration-thickness: var(\n          --elmsln-studio-link-focus-TextDecorationThickness,\n          var(--elmsln-studio-link-TextDecorationThickness)\n        );\n      }\n    "]);

    _templateObject_3dfdab60019d11eb9027c17497ee62cd = function _templateObject_3dfdab60019d11eb9027c17497ee62cd() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `elmsln-studio-link`
   * Navigation Link for ELMS:LN Studio
   *
   * @customElement elmsln-studio-link
   * @lit-html
   * @lit-element
   */
  var ElmslnStudioLink = /*#__PURE__*/function (_navigator) {
    babelHelpers.inherits(ElmslnStudioLink, _navigator);

    var _super = _createSuper(ElmslnStudioLink);

    babelHelpers.createClass(ElmslnStudioLink, null, [{
      key: "tag",

      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */
      get: function get() {
        return "elmsln-studio-link";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          href: {
            type: String
          }
        };
      }
    }, {
      key: "styles",
      get: function get() {
        return (0, _litElement.css)(_templateObject_3dfdab60019d11eb9027c17497ee62cd());
      }
    }]);

    function ElmslnStudioLink() {
      var _this;

      babelHelpers.classCallCheck(this, ElmslnStudioLink);
      _this = _super.call(this);
      _this.href = "";

      _this.addEventListener("click", _this.linkClick);

      return _this;
    }

    babelHelpers.createClass(ElmslnStudioLink, [{
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        if (babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioLink.prototype), "disconnectedCallback", this)) babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnStudioLink.prototype), "disconnectedCallback", this).call(this);
        this.removeEventListener("click", this.linkClick);
      }
    }, {
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject2_3dfdab60019d11eb9027c17497ee62cd(), this.href, this.linkClick);
      }
    }, {
      key: "linkClick",
      value: function linkClick(event) {
        event.preventDefault();
        this.navigate(this.href);
      }
    }]);
    return ElmslnStudioLink;
  }((0, _litElementRouter.navigator)(_litElement.LitElement));

  _exports.ElmslnStudioLink = ElmslnStudioLink;
  customElements.define("elmsln-studio-link", ElmslnStudioLink);
});
