// @lrnwebcomponents/simple-picker/simple-picker.js
define(["exports","require","../../lit-element/lit-element.js"],function(_exports,_require,_litElement){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.SimplePicker=void 0,_require=babelHelpers.interopRequireWildcard(_require);class SimplePicker extends _litElement.LitElement{static get styles(){return[_litElement.css`
        :host {
          display: inline-flex;
          align-items: center;
          color: var(--simple-picker-color, black);
          font-family: var(--simple-picker-font-family, inherit);
          font-size: var(--simple-picker-font-size, inherit);
          --simple-picker-height: calc(
            var(--simple-picker-option-size, 24px) +
              var(--simple-picker-sample-padding, 2px) * 2 +
              var(--simple-picker-border-width, 1px) * 2
          );
          min-height: var(--simple-picker-height);
          max-height: var(--simple-picker-height);
        }

        :host([block-label]) {
          display: block;
          margin: 0 0 15px;
          max-height: unset;
        }

        :host([disabled]) {
          --simple-picker-color: var(--simple-picker-color-disabled, #888);
          --simple-picker-background-color: var(
            --simple-picker-background-color-disabled,
            #e8e8e8
          );
          cursor: not-allowed;
        }

        :host([hidden]) {
          display: none;
        }

        div {
          margin: unset;
          padding: unset;
        }

        label:not([hidden]) {
          display: flex;
          align-items: center;
          padding-right: 5px;
          font-family: var(--simple-picker-font-family, inherit);
          color: var(
            --simple-picker-label-color,
            var(--simple-picker-color, black)
          );
        }

        :host([block-label]) label:not([hidden]) {
          display: block;
          padding-right: 0px;
          color: var(
            --simple-picker-float-label-color,
            var(--simple-picker-color-disabled, #888)
          );
          transition: all 0.5s;
          max-height: unset;
        }

        :host([block-label]:focus-within) label,
        :host([block-label]:hover) label {
          color: var(
            --simple-picker-float-label-active-color,
            var(--simple-picker-color, black)
          );
          transition: all 0.5s;
        }

        #sample,
        .rows {
          margin: 0;
          padding: 0;
        }

        #listbox {
          cursor: pointer;
          position: relative;
          flex: 1 0 auto;
          min-height: var(--simple-picker-height);
          max-height: var(--simple-picker-height);
        }

        #sample {
          display: flex;
          flex: 1 0 auto;
          justify-content: space-between;
          align-items: center;
          padding: var(--simple-picker-sample-padding, 2px);
          border-radius: var(--simple-picker-border-radius, 2px);
          color: var(--simple-picker-sample-color, black);
          background-color: var(--simple-picker-background-color, #f0f0f0);
          border-width: var(--simple-picker-border-width, 1px);
          border-style: var(--simple-picker-border-style, solid);
          border-color: var(
            --simple-picker-border-color,
            var(--simple-picker-color-disabled, #888)
          );
        }

        :host([hide-sample]) #sample {
          width: var(--simple-picker-option-size);
          overflow: visible;
        }

        :host(:focus-within) #sample {
          border-width: var(
            --simple-picker-focus-border-width,
            var(--simple-picker-border-width, 1px)
          );
          border-style: var(
            --simple-picker-focus-border-style,
            var(--simple-picker-border-style, solid)
          );
          border-color: var(
            --simple-picker-focus-border-color,
            var(
              --simple-picker-border-color,
              var(--simple-picker-color-disabled, #888)
            )
          );
          transition: all 0.5s;
        }

        #icon {
          transform: var(--simple-picker-icon-transform, rotate(0deg));
          transition: transform 0.25s;
        }

        :host([expanded]) #icon {
          transform: var(--simple-picker-expanded-icon-transform, rotate(0deg));
          transition: transform 0.25s;
        }

        #collapse {
          display: none;
          width: 100%;
          position: absolute;
          padding: var(
            --simple-picker-options-border-width,
            var(--simple-picker-border-width, 1px)
          );
          z-index: 2;
        }

        :host([expanded]:not([disabled])) #collapse {
          display: block;
        }

        .rows {
          display: block;
          position: absolute;
          z-index: 1000;
          left: calc(
            var(
              --simple-picker-options-border-width,
              var(--simple-picker-border-width, 1px)
            )
          );
          top: calc(
            0px -
              var(
                --simple-picker-options-border-width,
                var(--simple-picker-border-width, 1px)
              )
          );
          border-width: var(
            --simple-picker-options-border-width,
            var(--simple-picker-border-width, 1px)
          );
          border-style: var(
            --simple-picker-options-border-style,
            var(--simple-picker-border-style, solid)
          );
          border-color: var(
            --simple-picker-options-border-color,
            var(
              --simple-picker-border-color,
              var(--simple-picker-color-disabled, #888)
            )
          );
          background-color: var(--simple-picker-options-background-color, #fff);
          max-height: var(--simple-picker-options-max-height, 250px);
          overflow-y: auto;
        }
        :host([align-right]) #collapse .rows {
          left: unset;
          right: calc(
            var(
                --simple-picker-options-border-width,
                var(--simple-picker-border-width, 1px)
              ) * 2
          );
        }

        .row {
          display: flex;
          align-items: stretch;
          justify-content: space-between;
        }

        simple-picker-option {
          z-index: 1;
          flex: 1 1 auto;
          justify-content: flex-start;
          max-height: unset;
          min-height: var(--simple-picker-option-size, 24px);
          min-width: var(--simple-picker-option-size, 24px);
          line-height: var(--simple-picker-option-size, 24px);
          color: var(--simple-picker-color, black);
          background-color: var(--simple-picker-options-background-color, #fff);
          transition: max-height 2s;
        }

        simple-picker-option[selected] {
          z-index: 50;
          color: var(--simple-picker-color, black);
          background-color: var(
            --simple-picker-option-selected-background-color,
            var(--simple-picker-options-background-color, #fff)
          );
        }

        simple-picker-option[active] {
          z-index: 100;
          cursor: pointer;
          color: var(--simple-picker-color, black);
          background-color: var(
            --simple-picker-option-active-background-color,
            #aaddff
          );
        }

        #sample simple-picker-option {
          color: var(--simple-picker-color, black);
          background-color: var(
            --simple-picker-sample-background-color,
            transparent
          );
          --simple-picker-option-padding: var(
              --simple-picker-sample-padding,
              2px
            )
            0;
          border: none;
        }

        :host([hide-sample]) #sample simple-picker-option {
          position: absolute;
          left: -9999px;
          overflow: hidden;
          width: 0;
          height: 0;
        }

        :host(:focus-within) #sample simple-picker-option,
        :host(:hover) #sample simple-picker-option {
          --simple-picker-color: var(
            --simple-picker-color-active,
            var(--simple-picker-color, black)
          );
        }

        :host(:not([expanded])) #collapse simple-picker-option {
          max-height: 0;
          transition: max-height 1.5s;
        }

        @media screen and (max-width: 600px) {
          :host {
            position: static;
          }
          #collapse {
            top: 0;
            margin-top: 0;
            position: relative;
          }
          .rows {
            position: absolute;
          }
        }
      `]}render(){return _litElement.html`
      <label
        id="listLabel"
        for="listbox"
        .hidden="${!this.label||""===this.label.trim()}"
      >
        ${this.label&&""!==this.label.trim()?this.label.trim():""}
      </label>
      <div
        id="listbox"
        .aria-activedescendant="${this.__activeDesc}"
        .aria-labelledby="${this.ariaLabelledby}"
        .disabled="${this.disabled}"
        role="listbox"
        tabindex="0"
        @click="${this._handleListboxClick}"
        @mousedown="${this._handleListboxMousedown}"
        @keydown="${this._handleListboxKeydown}"
      >
        <div id="sample">
          <simple-picker-option
            ?hide-option-labels="${this.hideOptionLabels}"
            ?title-as-html="${this.titleAsHtml}"
            .icon="${!!this.__selectedOption&&this.__selectedOption.icon}"
            .label="${!!this.__selectedOption&&this.__selectedOption.alt}"
            .style=${!!this.__selectedOption&&this.__selectedOption.style}
            aria-hidden="true"
          >
          </simple-picker-option>
          <span id="icon"
            ><iron-icon aria-hidden="true" icon="arrow-drop-down"></iron-icon
          ></span>
        </div>
        <div id="collapse">
          <div class="rows">
            ${this.__options?this.__options.map((row,rownum)=>_litElement.html`
                    <div class="row">
                      ${row?row.map((option,colnum)=>_litElement.html`
                              <simple-picker-option
                                @option-focus="${this._handleOptionFocus}"
                                @set-selected-option="${this._handleSetSelectedOption}"
                                ?active="${`${this.__activeDesc}`===`option-${rownum}-${colnum}`}"
                                ?hide-option-labels="${this.hideOptionLabels}"
                                ?hidden="${!this.allowNull&&!option.value}"
                                ?selected="${this.value===option.value}"
                                ?title-as-html="${this.titleAsHtml}"
                                .data="${this.data}"
                                .icon="${option.icon}"
                                .id="option-${rownum}-${colnum}"
                                .label="${option.alt}"
                                .style=${option.style}
                                aria-selected="${this.value===option.value?"true":"false"}"
                                role="option"
                                tabindex="-1"
                                value="${option.value}"
                              >
                              </simple-picker-option>
                            `):""}
                    </div>
                  `):""}
          </div>
        </div>
      </div>
    `}static get properties(){return{...super.properties,allowNull:{type:Boolean,reflect:!0,attribute:"allow-null"},alignRight:{type:Boolean,reflect:!0,attribute:"align-right"},ariaLabelledby:{type:String,attribute:"aria-labelledby"},blockLabel:{type:Boolean,reflect:!0,attribute:"block-label"},disabled:{type:Boolean,reflect:!0},expanded:{type:Boolean,reflect:!0,attribute:"expanded"},hideOptionLabels:{type:Boolean,reflect:!0,attribute:"hide-option-labels"},hideSample:{type:Boolean,reflect:!0,attribute:"hide-sample"},label:{type:String},options:{type:Array},titleAsHtml:{type:Boolean,attribute:"title-as-html"},value:{type:String,reflect:!0},__activeDesc:{type:String},__options:{type:Array},__selectedOption:{type:Object}}}static get tag(){return"simple-picker"}constructor(){super(),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icon/iron-icon.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/iron-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["./lib/simple-picker-option.js"],res,rej)),this.tag=SimplePicker.tag,this.allowNull=!1,this.alignRight=!1,this.ariaLabelledby=null,this.blockLabel=!1,this.disabled=!1,this.expanded=!1,this.hideOptionLabels=!1,this.hideSample=!1,this.label=null,this.__options=[[]],this.options=[[{icon:null,style:null,alt:null,value:null}]],this.titleAsHtml=!1,this.value=null,this.__activeDesc="option-0-0",this.__hasLabel=!0,this.__selectedOption={},this.addEventListener("blur",function(e){this.expanded=!1});let obj=SimplePicker.properties;for(let p in obj)obj.hasOwnProperty(p)&&(this.hasAttribute(p)?this[p]=this.getAttribute(p):(p.reflect&&this.setAttribute(p,obj[p].value),this[p]=obj[p].value))}updated(changedProperties){changedProperties.forEach((oldValue,propName)=>{"value"===propName&&this._valueChanged(this.value,oldValue),"options"===propName&&this._optionsChanged(this.value,oldValue)}),this.dispatchEvent(new CustomEvent("changed",{detail:this}))}_getOption(options,optionId){if(void 0!==options&&null!=optionId){let coords=optionId.split("-");return options[coords[1]][coords[2]]}return null}_goToOption(rownum,colnum){let targetId=_litElement.html`
        option-${rownum}-${colnum}
      `,target=this.shadowRoot.querySelector("#"+targetId),active=this.shadowRoot.querySelector("#"+this.__activeDesc);null!==target&&(target.tabindex=0,target.focus(),active.tabindex=-1)}_handleListboxClick(e){this.dispatchEvent(new CustomEvent("click",{detail:this})),this._toggleListbox()}_handleListboxMousedown(e){this.dispatchEvent(new CustomEvent("mousedown",{detail:this}))}_handleListboxKeydown(e){this.dispatchEvent(new CustomEvent("keydown",{detail:this}));let coords=this.__activeDesc.split("-"),rownum=parseInt(coords[1]),colnum=parseInt(coords[2]);if(32===e.keyCode)e.preventDefault(),this._toggleListbox();else if(this.expanded&&[9,35,36,38,40].includes(e.keyCode))if(e.preventDefault(),35===e.keyCode){let lastrow=this.options.length-1,lastcol=this.options[lastrow].length-1;this._goToOption(lastrow,lastcol)}else 36===e.keyCode?this._goToOption(0,0):38===e.keyCode?colnum>0?this._goToOption(rownum,colnum-1):rownum>0&&this._goToOption(rownum-1,this.options[rownum-1].length-1):40===e.keyCode&&(colnum<this.options[rownum].length-1?this._goToOption(rownum,colnum+1):rownum<this.options.length-1&&this._goToOption(rownum+1,[0]))}_handleOptionFocus(e){this._setActiveOption(e.detail.id)}_setActiveOption(id){this.__activeDesc=id,this.dispatchEvent(new CustomEvent("option-focus",{detail:this}))}_valueChanged(newValue,oldValue){this._setSelectedOption(newValue,oldValue),this.dispatchEvent(new CustomEvent("value-changed",{detail:this}))}_optionsChanged(newValue,oldValue){this._setSelectedOption(newValue,oldValue)}_setSelectedOption(newVal,oldVal){let sel=!this.allowNull&&this.options.length>0&&this.options[0].length>0?this.options[0][0].value:null;if(this.options){this.__options="string"==typeof this.options?JSON.parse(this.options):this.options.slice(),this.__activeDesc=this.allowNull?"option-0-0":null;for(var i=0;i<this.__options.length;i++)for(var j=0;j<this.__options[i].length;j++)null!==this.value&&null===this.__activeDesc&&(this.__activeDesc=`option-${i}-${j}`),`${this.__options[i][j].value}`===`${this.value}`&&(this.__activeDesc=`option-${i}-${j}`,sel=this.__options[i][j])}null===sel&&(this.value=null),this.__selectedOption=sel,this.dispatchEvent(new CustomEvent("change",{bubbles:!0,detail:this}))}_toggleListbox(open=!this.expanded){if(this.disabled)return;let active=this.shadowRoot.querySelector("#"+this.__activeDesc);this.expanded=open,open?(null!==active&&active.focus(),this.dispatchEvent(new CustomEvent("expand",{detail:this}))):(null!==active&&(this.value=active.getAttribute("value")),this.dispatchEvent(new CustomEvent("collapse",{detail:this})))}setOptions(options){this.set("options",[[]]),this.set("options",options)}disconnectedCallback(){this.removeEventListener("blur",function(e){this.expanded=!1}),super.disconnectedCallback()}}_exports.SimplePicker=SimplePicker,window.customElements.define(SimplePicker.tag,SimplePicker)});
