// @lrnwebcomponents/paper-fab-speed-dial/paper-fab-speed-dial.js
define(["exports", "require", "../../@polymer/polymer/polymer-element.js"], function (_exports, _require, _polymerElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.PaperFabSpeedDial = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  /**
   * Copyright 2018 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `paper-fab-speed-dial`
   * @element paper-fab-speed-dial
   * `A speed dial setup for a floating action button`
   *
   * @demo demo/index.html
   */
  class PaperFabSpeedDial extends _polymerElement.PolymerElement {
    constructor() {
      super();
      new Promise((res, rej) => _require.default(["./lib/paper-fab-speed-dial-overlay.js"], res, rej));
    }

    static get tag() {
      return "paper-fab-speed-dial";
    }

    static get properties() {
      return {
        icon: {
          type: String,
          value: "add"
        },
        opened: {
          type: Boolean,
          notify: true
        },
        disabled: {
          type: Boolean,
          value: false
        }
      };
    } // Public methods


    open(e) {
      // Required for mobile Safari to avoid passing the tap event to an element below the FAB
      if (e) {
        e.preventDefault();
      }

      this.opened = true;
    }

    close(e) {
      // Required for mobile Safari to avoid passing the tap event to an element below the FAB
      if (e) {
        e.preventDefault();
      }

      this.opened = false;
    }

  }

  _exports.PaperFabSpeedDial = PaperFabSpeedDial;
  window.customElements.define(PaperFabSpeedDial.tag, PaperFabSpeedDial);
});
