Index: lazyloader.module
===================================================================
--- lazyloader.module	(revision xxx)
+++ lazyloader.module	(working copy)
@@ -173,12 +173,23 @@
 
     $icon = ($settings['loader_icon'] != 0) ? LAZYLOADER_PATH . '/loader/loader-' . $settings['loader_icon'] . '.gif': '';
 
-    $lazyloader_init = '
+    /**
+     * Bug with Ajax Views behaviour
+     * 
+     * @see https://drupal.org/node/1799172
+     * 
+     */
+    
+	$lazyloader_init = '
       <noscript><style type="text/css" media="all">img[data-src] { display: none; }</style></noscript>
       <script type="text/javascript">
-        jQuery(function($){
-          $("img[data-src]").lazyloader({distance: ' . trim($settings['distance']) . ', icon: "' . $icon . '" });
-        });
+                (function ($) {
+                    Drupal.behaviors.lazyloader = {
+                        attach: function (context, settings) {
+                            $("img[data-src]").lazyloader({distance: ' . trim($settings['distance']) . ', icon: "' . $icon . '" });
+                        }
+                    };
+                }(jQuery));
       </script>';
 
     $vars['page_bottom'] .= $lazyloader_init;
