// @lrnwebcomponents/lrnsys-layout/lib/lrnsys-collapselist.js
define(["exports","require","../../../lit-element/lit-element.js","../../simple-colors/simple-colors.js"],(function(s,l,e,t){"use strict";Object.defineProperty(s,"__esModule",{value:!0}),s.LrnsysCollapselist=void 0,l=babelHelpers.interopRequireWildcard(l);class LrnsysCollapselist extends t.SimpleColors{static get styles(){return[e.css`
        :host {
          display: block;
          background-color: var(--simple-colors-background1);
          --lrnsys-collapselist-text-color: var(--simple-colors-foreground1);
          --lrnsys-collapselist-item-color: var(--simple-colors-background1);
          --lrnsys-collapselist-item-active-color: var(
            --simple-colors-background2
          );
          --lrnsys-collapselist-item-border: var(--simple-colors-background5);
        }
        ul {
          list-style-type: none;
          margin: 0;
          padding: 0;
        }
        ul li {
          width: 100%;
          border: 1px solid var(--lrnsys-collapselist-item-border);
          margin-bottom: -1px;
        }
        ul li lrnsys-collapselist-item {
          height: 32px;
          padding: 8px;
          margin: 0;
          min-width: 0.16px;
          -webkit-justify-content: flex-start;
          justify-content: flex-start;
          align-items: center;
          width: 100%;
          text-transform: unset;
          border-radius: 0;
        }
        ul li lrnsys-collapselist-item iron-icon,
        ul li lrnsys-collapselist-item span {
          pointer-events: none;
        }
        iron-icon {
          display: inline-block;
        }
        .collapse-label {
          margin-left: 8px;
        }
        .collapse-content {
          padding: 16px;
        }
      `]}constructor(){super(),this.items=[],new Promise((s,e)=>l.default(["../../../@polymer/iron-icon/iron-icon.js"],s,e)),new Promise((s,e)=>l.default(["./lrnsys-collapselist-item.js"],s,e))}render(){return e.html`
      <ul>
        ${this.items.map(s=>e.html`
            <li>
              <lrnsys-collapselist-item>
                <span slot="label">
                  <iron-icon icon="${s.icon}"></iron-icon>
                  <span class="collapse-label">${s.label}</span>
                </span>
                <span slot="content"> ${s.content} </span>
              </lrnsys-collapselist-item>
            </li>
          `)}
      </ul>
    `}static get tag(){return"lrnsys-collapselist"}static get properties(){return{items:{type:Array}}}}s.LrnsysCollapselist=LrnsysCollapselist,window.customElements.define(LrnsysCollapselist.tag,LrnsysCollapselist)}));
