// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-dashboard/lrnapp-block-recent-comments-comment.js
define(["exports","../../../../@polymer/polymer/polymer-element.js","../../../../@polymer/polymer/lib/utils/render-status.js","../../../../@polymer/polymer/lib/legacy/class.js","../../../../@polymer/iron-resizable-behavior/iron-resizable-behavior.js","../../../../@polymer/paper-card/paper-card.js","../../../../@polymer/paper-button/paper-button.js","../../../lrn-icon/lrn-icon.js","../../../../@polymer/polymer/lib/elements/dom-if.js","../../../lrndesign-avatar/lrndesign-avatar.js"],(function(e,t,n,i,r,a,o,s,l,c){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.LrnappBlockRecentCommentsComment=void 0;class LrnappBlockRecentCommentsComment extends((0,i.mixinBehaviors)([r.IronResizableBehavior],t.PolymerElement)){static get template(){return t.html`
      <style>
        :host {
          display: block;
        }

        paper-card {
          padding: 2em 2em 0 2em;
          clear: right;
        }

        paper-button {
          background-color: var(--paper-grey-100);
          margin: 1em;
        }

        .card-content {
          padding-left: 2em;
          padding-right: 2em;
          width: 100%;
        }

        .card-actions {
          width: 100%;
        }

        .card-actions paper-button {
          display: flex;
        }

        lrndesign-avatar {
          float: left;
          margin-right: 1em;
        }

        .flex-wrap {
          @apply --layout-horizontal;
          @apply --layout-wrap;
        }

        .inactive {
          max-height: 4.6em;
          overflow: hidden;
        }

        paper-button {
          background: white;
          width: 100%;
          display: flex;
        }

        lrn-icon {
          color: black;
          width: 100%;
        }

        .hidden {
          display: none;
        }
      </style>
      <paper-card elevation="3" class="flex-wrap">
        <div class="card-content">
          <lrndesign-avatar
            label="[[commentUser.name]]"
            src="[[commentUser.avatar]]"
          ></lrndesign-avatar>
          <h3>[[commentUser.display_name]]</h3>
          <div id="wrapper" class="button-wrapper">
            <div id="comment" class="inactive"><slot></slot></div>
            <paper-button id="btn" class="hidden">
              <lrn-icon icon="chevron-down" id="icon"></lrn-icon>
            </paper-button>
          </div>
        </div>
        <div class="card-actions">
          <template is="dom-if" if="[[actionView]]">
            <a href$="[[actionView]]" tabindex="-1">
              <paper-button raised="" id="view">View thread</paper-button>
            </a>
          </template>
        </div>
      </paper-card>
    `}static get tag(){return"lrnapp-block-recent-comments-comment"}onHeightChange(){this.$.comment.offsetHeight>80&&this.$.btn.classList.toggle("hidden",this.hidden)}static get properties(){return{elmslnCourse:{type:String},elmslnSection:{type:String},basePath:{type:String},csrfToken:{type:String},endPoint:{type:String},commentTitle:{type:String,value:"Comment title",reflectToAttribute:!0,notify:!0},actionView:{type:String,reflectToAttribute:!0,notify:!0},dateUpdated:{type:String,reflectToAttribute:!0,notify:!0},commentUser:{type:Object,value:{},reflectToAttribute:!0,notify:!0}}}connectedCallback(){super.connectedCallback(),(0,n.afterNextRender)(this,(function(){this.$.wrapper.addEventListener("click",(function(e){this.$.comment.classList.toggle("inactive",this.inactive)})),this.addEventListener("iron-resize",this.onHeightChange.bind(this))}))}disconnectedCallback(){this.$.wrapper.removeEventListener("click",(function(e){this.$.comment.classList.toggle("inactive",this.inactive)})),this.removeEventListener("iron-resize",this.onHeightChange.bind(this)),super.disconnectedCallback()}}e.LrnappBlockRecentCommentsComment=LrnappBlockRecentCommentsComment,window.customElements.define(LrnappBlockRecentCommentsComment.tag,LrnappBlockRecentCommentsComment)}));
