// @lrnwebcomponents/simple-popover/lib/simple-tour.js
define(["exports", "../../../lit/index.js", "../../../lit/directives/unsafe-html.js", "../../simple-toolbar/simple-toolbar.js", "../../simple-icon/lib/simple-icon-button-lite.js", "../../simple-tooltip/simple-tooltip.js", "./simple-popover-manager.js"], function (_exports, _index, _unsafeHtml, _simpleToolbar, _simpleIconButtonLite, _simpleTooltip, _simplePopoverManager) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.TourStop = _exports.SimpleTourManager = _exports.SimpleTour = void 0;

  var _templateObject_6d3bc8c0942811ecb1937969798b231a, _templateObject2_6d3bc8c0942811ecb1937969798b231a;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var SimpleTour = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(SimpleTour, _LitElement);

    var _super = _createSuper(SimpleTour);

    function SimpleTour() {
      var _this;

      babelHelpers.classCallCheck(this, SimpleTour);
      _this = _super.call(this);
      _this.stacks = {};
      _this.orientation = "lr";
      _this.active = null;
      _this.tourInfo = {};
      _this.activeElementDelay = 500;
      _this.stop = -1;
      window.addEventListener("simple-tour-register", _this.registerNewTourEvent.bind(babelHelpers.assertThisInitialized(_this)));
      window.addEventListener("simple-tour-create-tour-stop", _this.createTourStopEvent.bind(babelHelpers.assertThisInitialized(_this)));
      return _this;
    }

    babelHelpers.createClass(SimpleTour, [{
      key: "registerNewTourEvent",
      value: function registerNewTourEvent(e) {
        this.registerNewTour(e.detail);
      }
    }, {
      key: "registerNewTour",
      value: function registerNewTour(newTour) {
        if (!this.stacks[newTour.key]) {
          this.stacks[newTour.key] = [];
        }

        if (!this.tourInfo[newTour.key]) {
          this.tourInfo[newTour.key] = newTour;
        }

        return this.stacks[newTour.key];
      }
      /**
       * create tour stop via events
       */

    }, {
      key: "createTourStopEvent",
      value: function createTourStopEvent(e) {
        this.createTourStop(e.detail.name, e.detail.target, e.detail.title, e.detail.description, e.detail.mode);
      }
      /**
       * Create a tour stop, add to the stack, then return the stop object
       */

    }, {
      key: "createTourStop",
      value: function createTourStop(name, target, title, description, mode) {
        var s = new TourStop();
        s.target = target;
        s.title = title;
        s.description = description;
        s.mode = mode;
        this.addStops(name, [s]);
        return s;
      }
      /**
       * It's possible we drop a target from the DOM and then
       * have to remove it from the tour
       */

    }, {
      key: "removeTarget",
      value: function removeTarget(name, target) {
        var _this2 = this;

        var dropList = [];
        this.stacks[name].forEach(function (item, index) {
          if (item.target === target) {
            dropList.push(index);
          }
        });
        dropList.forEach(function (i) {
          _this2.stacks[name].splice(i, 1);
        });
      }
      /**
       * Add stops to the tour
       */

    }, {
      key: "addStops",
      value: function addStops(name, stops) {
        if (!this.stacks[name]) {
          this.stacks[name] = [];
        }

        this.stacks[name] = this.stacks[name].concat(stops);
      }
    }, {
      key: "hasNext",
      value: function hasNext() {
        return this.stop < this.stacks[this.active].length - 1;
      }
    }, {
      key: "hasPrev",
      value: function hasPrev() {
        return this.stop > 0;
      }
      /**
       * Move ahead or back in the stack
       */

    }, {
      key: "nextStop",
      value: function nextStop(e) {
        if (this.stop < this.stacks[this.active].length - 1) {
          this.stop += 1;
        }
      }
    }, {
      key: "prevStop",
      value: function prevStop(e) {
        if (this.stop > 0) {
          this.stop -= 1;
        }
      }
    }, {
      key: "startTour",
      value: function startTour(name) {
        this.active = name;
        this.dispatchEvent(new CustomEvent("tour-changed", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: this
        }));
      }
    }, {
      key: "stopTour",
      value: function stopTour(e) {
        window.SimplePopoverManager.requestAvailability().setPopover(this, this, false, this.orientation);
        this.stop = -1;
        this.active = null;
        this.dispatchEvent(new CustomEvent("tour-changed", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: this
        }));
      }
      /**
       * Render tour buttons as block
       */

    }, {
      key: "tourButtons",
      value: function tourButtons() {
        return (0, _index.html)(_templateObject_6d3bc8c0942811ecb1937969798b231a || (_templateObject_6d3bc8c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<h1 class=\"title\" slot=\"heading\">\n        ", ",\n        <span class=\"xofy\"\n          >", "/", "</span\n        >\n      </h1>\n      <simple-icon-button-lite\n        id=\"close\"\n        slot=\"heading\"\n        @click=\"", "\"\n        label=\"Stop Tour\"\n        icon=\"close\"\n      >\n      </simple-icon-button-lite>\n      <simple-icon-button-lite\n        id=\"prev\"\n        slot=\"nav\"\n        @click=\"", "\"\n        ?disabled=\"", "\"\n        label=\"Prev\"\n        icon=\"arrow-back\"\n        show-text-label\n      >\n      </simple-icon-button-lite>\n      <simple-tooltip for=\"prev\" position=\"top\" slot=\"nav\">\n        Previous Item\n      </simple-tooltip>\n      <simple-icon-button-lite\n        id=\"next\"\n        slot=\"nav\"\n        @click=\"", "\"\n        ?disabled=\"", "\"\n        label=\"Next\"\n        icon=\"arrow-forward\"\n        show-text-label\n      >\n      </simple-icon-button-lite>\n      <simple-tooltip for=\"close\" slot=\"body\"> Stop Tour </simple-tooltip>\n      <simple-tooltip for=\"next\" position=\"top\" slot=\"nav\">\n        Next Item\n      </simple-tooltip>"])), this.tourInfo[this.active].name, this.stop + 1, this.stacks[this.active].length, this.stopTour.bind(this), this.prevStop.bind(this), !this.hasPrev(), this.nextStop.bind(this), !this.hasNext());
      }
      /**
       * Simple utility to do nice scrolling or only scroll if we can't see it
       * as that is better behavior but not in all browsers
       */

    }, {
      key: "scrollHere",
      value: function scrollHere(node) {
        // scroll to it
        if (typeof node.scrollIntoViewIfNeeded === "function") {
          node.scrollIntoViewIfNeeded(true);
        } else {
          node.scrollIntoView({
            behavior: "smooth",
            inline: "center"
          });
        }
      }
      /**
       * The manager was called and he's pissed. Do not cross Go, do not collect $200 in tips.
       * You're fired. Me, you, every body. We're fired because someone just asked for the manager
       * to do something else and we were already taking people on a tour and now the kitchen
       * is on fire and someone must be blamed.
       *
       * Also, global calls this to clean up local state when global is hijacked by another project
       * that also leverages the singleton and wants to ensure everyone cleans up after themselves
       * instead of flipping tables on their way out the door. We'll call this function pack up
       * on a Friday to avoid incidents.
       */

    }, {
      key: "managerReset",
      value: function managerReset() {
        this.stopTour();
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this3 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "active" && _this3.active) {
            _this3.stop = 0;
          }

          if (propName == "stop" && _this3.stop != -1 || propName == "active" && _this3.active) {
            (0, _index.render)(document.createElement("div"), window.SimplePopoverManager.requestAvailability());
            var title = _this3.stacks[_this3.active][_this3.stop].title;
            var description = _this3.stacks[_this3.active][_this3.stop].description; // support for live rendering of the tour info to allow the DOM
            // to translate or dynamicallly set this based on other stateful details

            if (_this3.stacks[_this3.active][_this3.stop].mode == "live") {
              var el = _this3.stacks[_this3.active][_this3.stop].target;

              if (el.getAttribute("data-stop-title") && el.getAttribute(el.getAttribute("data-stop-title"))) {
                title = el.getAttribute(el.getAttribute("data-stop-title"));
              } else if (el.querySelector("[data-stop-title]")) {
                title = el.querySelector("[data-stop-title]").innerHTML;
              }

              description = el.querySelector("[data-stop-content]").innerHTML ? el.querySelector("[data-stop-content]").innerHTML : "";
            }

            var content = (0, _index.html)(_templateObject2_6d3bc8c0942811ecb1937969798b231a || (_templateObject2_6d3bc8c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["", "\n          <h2 class=\"subheading\" slot=\"body\">\n            ", "\n          </h2>\n          ", "", ""])), _this3.tourButtons(), (0, _unsafeHtml.unsafeHTML)("<span>" + title + "</span>"), (0, _unsafeHtml.unsafeHTML)('<p slot="body">' + description + "</p>"), _this3.tourInfo[_this3.active].style ? (0, _unsafeHtml.unsafeHTML)("<style>" + _this3.tourInfo[_this3.active].style + "</style>") : "");
            (0, _index.render)(content, window.SimplePopoverManager.requestAvailability());
            window.SimplePopoverManager.requestAvailability().setPopover(_this3, _this3.stacks[_this3.active][_this3.stop].target, true, _this3.orientation, _this3.active);

            _this3.scrollHere(_this3.stacks[_this3.active][_this3.stop].target);

            var target = _this3.stacks[_this3.active][_this3.stop].target;

            var part = _this3.stacks[_this3.active][_this3.stop].target.getAttribute("part");

            target.setAttribute("part", "simple-tour-active");
            setTimeout(function () {
              if (part == null || part == "simple-tour-active") {
                target.removeAttribute("part");
              } else {
                target.setAttribute("part", part);
              }
            }, _this3.activeElementDelay);
          }
        });
      }
    }], [{
      key: "properties",
      get: function get() {
        return {
          stop: {
            type: Number
          },
          active: {
            type: String
          },
          stacks: {
            type: Object
          },
          activeElementDelay: {
            type: Number
          }
        };
      }
    }]);
    return SimpleTour;
  }(_index.LitElement);
  /**
   * Simple Tour Stop object for consistency
   */


  _exports.SimpleTour = SimpleTour;
  var TourStop = /*#__PURE__*/babelHelpers.createClass(function TourStop() {
    babelHelpers.classCallCheck(this, TourStop);
    this.target = null;
    this.title = "Title";
    this.description = "<p>Description</p>";
  });
  _exports.TourStop = TourStop;
  customElements.define("simple-tour", SimpleTour);
  // register globally so we can make sure there is only one
  window.SimpleTourManager = window.SimpleTourManager || {}; // request if this exists. This helps invoke the element existing in the dom
  // as well as that there is only one of them. That way we can ensure everything
  // is rendered through the same modal

  window.SimpleTourManager.requestAvailability = function () {
    if (!window.SimpleTourManager.instance) {
      window.SimpleTourManager.instance = document.createElement("simple-tour");
      document.body.appendChild(window.SimpleTourManager.instance);
    }

    return window.SimpleTourManager.instance;
  }; // self append


  var SimpleTourManager = window.SimpleTourManager.requestAvailability();
  _exports.SimpleTourManager = SimpleTourManager;
});
