// @lrnwebcomponents/simple-popover/lib/simple-popover-manager.js
define(["exports", "../../../lit-element/lit-element.js", "../simple-popover.js"], function (_exports, _litElement, _simplePopover) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SimplePopoverManager = void 0;

  function _templateObject2_f8254960544211ebad6c77197a564a47() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        simple-popover {\n          font-size: 14px;\n          line-height: 20px;\n          color: black;\n          --simple-popover-color: #222222;\n          --simple-popover-border-color: #222222;\n          --simple-popover-background-color: #eeeeee;\n          --simple-popover-padding: 4px;\n          max-width: var(--simple-popover-manager-max-width, 200px);\n        }\n      "]);

    _templateObject2_f8254960544211ebad6c77197a564a47 = function _templateObject2_f8254960544211ebad6c77197a564a47() {
      return data;
    };

    return data;
  }

  function _templateObject_f8254960544211ebad6c77197a564a47() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <simple-popover\n        auto\n        part=\"simple-popover\"\n        ?hidden=\"", "\"\n        position=\"", "\"\n      >\n        <slot></slot>\n      </simple-popover>\n    "]);

    _templateObject_f8254960544211ebad6c77197a564a47 = function _templateObject_f8254960544211ebad6c77197a564a47() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  var SimplePopoverManager = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(SimplePopoverManager, _LitElement);

    var _super = _createSuper(SimplePopoverManager);

    function SimplePopoverManager() {
      var _this;

      babelHelpers.classCallCheck(this, SimplePopoverManager);
      _this = _super.call(this);
      _this.popover = null;
      _this.opened = false;
      _this.context = null;
      _this.orientation = "tb";
      _this.position = "bottom";
      _this.__ignore = false;
      return _this;
    }

    babelHelpers.createClass(SimplePopoverManager, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_f8254960544211ebad6c77197a564a47(), !this.opened, this.position);
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          // ensure that changes get reflected in the direction it should point
          // or state of open. This avoids minor timing issues when element
          // being pointed to requests changes to direction / state
          if (["opened", "position", "orientation"].includes(propName)) {
            clearTimeout(_this2.timer);
            _this2.timer = setTimeout(function () {
              _this2.popover.updatePosition();
            }, 0);
          }
        });
      }
      /**
       * set target and optionally change content and open state
       */

    }, {
      key: "setPopover",
      value: function setPopover(context, el) {
        var _this3 = this;

        var opened = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
        var orientation = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : "tb";

        // this has the potential to cause 1 popover to change content and parent
        // in the same action. This would cause a open state change in 1 element
        // which would trigger a global state change to match.
        // The ignore flag implies we are actively switching an operation and thus
        // need to ignore the follow up change record, much like a debounce
        if (this.__ignore) {
          this.__ignore = false;
          setTimeout(function () {
            _this3.popover.updatePosition();
          }, 100);
        } else {
          if (el !== this.popover.target) {
            // helps manage state if multiple things leveraging this
            // yet having their own internal opened status
            if (this.context && this.context.managerReset && context !== this.context) {
              this.context.managerReset();
              this.__ignore = true;
            }

            this.context = context;
            this.popover.target = null;
            this.popover.target = el;
          }

          var position;
          var menu = el.getBoundingClientRect(); // top - bottom or left - right pointer orientation
          // Highly polarized detection of 50% in any direction
          // forces the pointer in the opposite direction

          if (orientation == "tb") {
            if (menu.y > window.innerHeight / 2) {
              position = "top";
            } else {
              position = "bottom";
            }
          } else {
            if (menu.x > window.innerWidth / 2) {
              position = "left";
            } else {
              position = "right";
            }
          } // see if we need to reposition


          this.orientation = orientation;
          this.position = position; // only open / close if told to change

          if (opened != null) {
            this.opened = opened;
          }
        }
      }
    }, {
      key: "firstUpdated",
      value: function firstUpdated() {
        this.popover = this.shadowRoot.querySelector("simple-popover");
      }
    }], [{
      key: "styles",
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_f8254960544211ebad6c77197a564a47())];
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          position: {
            type: String
          },
          opened: {
            type: Boolean
          },
          orientation: {
            type: String
          }
        };
      }
    }]);
    return SimplePopoverManager;
  }(_litElement.LitElement);

  _exports.SimplePopoverManager = SimplePopoverManager;
  customElements.define("simple-popover-manager", SimplePopoverManager);
  // register globally so we can make sure there is only one
  window.SimplePopoverManager = window.SimplePopoverManager || {}; // request if this exists. This helps invoke the element existing in the dom
  // as well as that there is only one of them. That way we can ensure everything
  // is rendered through the same modal

  window.SimplePopoverManager.requestAvailability = function () {
    if (!window.SimplePopoverManager.instance) {
      window.SimplePopoverManager.instance = document.createElement("simple-popover-manager");
      document.body.appendChild(window.SimplePopoverManager.instance);
    }

    return window.SimplePopoverManager.instance;
  }; // self append


  window.SimplePopoverManager.requestAvailability();
});
