// @lrnwebcomponents/grade-book/lib/GradeBookUIPieces.js
define(["exports", "require", "../../../lit/index.js"], function (_exports, _require, _index) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.UIRenderPieces = void 0;
  _require = _interopRequireWildcard(_require);

  var _templateObject_620dd880942811ecb1937969798b231a, _templateObject2_620dd880942811ecb1937969798b231a, _templateObject3_620dd880942811ecb1937969798b231a, _templateObject4_620dd880942811ecb1937969798b231a, _templateObject5_620dd880942811ecb1937969798b231a, _templateObject6_620dd880942811ecb1937969798b231a, _templateObject7_620dd880942811ecb1937969798b231a, _templateObject8_620dd880942811ecb1937969798b231a, _templateObject9_620dd880942811ecb1937969798b231a, _templateObject10_620dd880942811ecb1937969798b231a, _templateObject11_620dd880942811ecb1937969798b231a, _templateObject12_620dd880942811ecb1937969798b231a, _templateObject13_620dd880942811ecb1937969798b231a, _templateObject14_620dd880942811ecb1937969798b231a;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var UIRenderPieces = function UIRenderPieces(SuperClass) {
    return /*#__PURE__*/function (_SuperClass) {
      babelHelpers.inherits(_class, _SuperClass);

      var _super = _createSuper(_class);

      function _class() {
        var _this;

        babelHelpers.classCallCheck(this, _class);
        _this = _super.call(this);
        _this.hideGradeScale = true;
        _this.hideActiveStudentOverview = true;
        _this.hideActiveAssignment = true;
        _this.hideSettings = true;
        _this.hideRubricInfo = [];
        new Promise(function (res, rej) {
          return _require.default(["../../simple-popover/simple-popover.js"], res, rej);
        });
        new Promise(function (res, rej) {
          return _require.default(["./grade-book-student-block.js"], res, rej);
        });
        return _this;
      }

      babelHelpers.createClass(_class, [{
        key: "renderActiveStudentBtn",
        value: // render the active student as a button that opens a popover
        function renderActiveStudentBtn() {
          var _this2 = this;

          return (0, _index.html)(_templateObject_620dd880942811ecb1937969798b231a || (_templateObject_620dd880942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<simple-icon-button-lite\n          @click=\"", "\"\n          id=\"activestudentbtn\"\n          ?disabled=\"", "\"\n        >\n          ", "\n          <span class=\"hide-900\">", "</span>\n        </simple-icon-button-lite>\n        <simple-popover\n          ?hidden=\"", "\"\n          for=\"activestudentbtn\"\n          auto\n        >\n          <div>\n            ", "\n\n            <div class=\"active-student-grade-history\">\n              ", "\n            </div>\n          </div>\n        </simple-popover>"])), this.toggleActiveStudentOverview, !this.ready, this.database.roster[this.activeStudent] && this.database.roster[this.activeStudent].photo ? (0, _index.html)(_templateObject2_620dd880942811ecb1937969798b231a || (_templateObject2_620dd880942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<img\n                src=\"", "\"\n                loading=\"lazy\"\n              />"])), this.database.roster[this.activeStudent].photo) : (0, _index.html)(_templateObject3_620dd880942811ecb1937969798b231a || (_templateObject3_620dd880942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<simple-icon-lite icon=\"social:person\"></simple-icon-lite>"]))), this.t.activeStudent, this.hideActiveStudentOverview, this.database.roster[this.activeStudent] ? (0, _index.html)(_templateObject4_620dd880942811ecb1937969798b231a || (_templateObject4_620dd880942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral([" <grade-book-student-block\n                  .student=\"", "\"\n                ></grade-book-student-block>"])), this.database.roster[this.activeStudent]) : _index.nothing, this.activeStudentSubmissions.map(function (sRecord) {
            return (0, _index.html)(_templateObject5_620dd880942811ecb1937969798b231a || (_templateObject5_620dd880942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                  <button\n                    .value=\"", "\"\n                    @click=\"", "\"\n                    class=\"", "\"\n                  >\n                    <letter-grade\n                      mini\n                      label=\"", "\"\n                      .total=\"", "\"\n                      .score=\"", "\"\n                      .value=\"", "\"\n                    >\n                    </letter-grade>\n                  </button>\n                "])), sRecord.assignmentIndex, _this2.studentLetterGradeHistoryClick, _this2.activeAssignment === sRecord.assignmentIndex ? "activeAssignment" : "", sRecord.assignmentName, sRecord.assignmentPoints, sRecord.studentScore, sRecord.assignmentIndex);
          }));
        } // toggle logic for active student

      }, {
        key: "toggleActiveStudentOverview",
        value: function toggleActiveStudentOverview(e) {
          this.hideActiveStudentOverview = !this.hideActiveStudentOverview;

          if (!this.hideActiveStudentOverview) {
            this.hideSettings = true;
            this.hideActiveAssignment = true;
            this.hideGradeScale = true;
          }
        } // render grading scale in a popover

      }, {
        key: "renderGradeScaleBtn",
        value: function renderGradeScaleBtn() {
          return (0, _index.html)(_templateObject6_620dd880942811ecb1937969798b231a || (_templateObject6_620dd880942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral([" <simple-icon-button-lite\n          icon=\"list\"\n          @click=\"", "\"\n          id=\"gradescalebtn\"\n          ?disabled=\"", "\"\n        >\n          <span class=\"hide-900\">", "</span>\n        </simple-icon-button-lite>\n        <simple-popover\n          ?hidden=\"", "\"\n          for=\"gradescalebtn\"\n          auto\n        >\n          <editable-table-display\n            accent-color=\"", "\"\n            bordered\n            column-header\n            condensed\n            disable-responsive\n            scroll\n            striped\n          >\n            <table>\n              <tbody>\n                <tr>\n                  <td>", "</td>\n                  <td>", "</td>\n                  <td>", "</td>\n                </tr>\n                ", "\n              </tbody>\n            </table>\n          </editable-table-display>\n        </simple-popover>"])), this.toggleGradeScale, !this.ready, this.t.gradingScale, this.hideGradeScale, this.accentColor, this.t.letterGrade, this.t.highRange, this.t.lowRange, this.database.gradeScale.map(function (scale) {
            return (0, _index.html)(_templateObject7_620dd880942811ecb1937969798b231a || (_templateObject7_620dd880942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                    <tr>\n                      <td>", "</td>\n                      <td>", "</td>\n                      <td>", "</td>\n                    </tr>\n                  "])), scale.letter, scale.highRange, scale.lowRange);
          }));
        } // toggle logic for button

      }, {
        key: "toggleGradeScale",
        value: function toggleGradeScale(e) {
          this.hideGradeScale = !this.hideGradeScale;

          if (!this.hideGradeScale) {
            this.hideSettings = true;
            this.hideActiveAssignment = true;
            this.hideActiveStudentOverview = true;
          }
        } // render grading scale in a popover

      }, {
        key: "renderSettingsBtn",
        value: function renderSettingsBtn() {
          return (0, _index.html)(_templateObject8_620dd880942811ecb1937969798b231a || (_templateObject8_620dd880942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral([" <simple-icon-button-lite\n          icon=\"settings\"\n          @click=\"", "\"\n          id=\"settings\"\n          ?disabled=\"", "\"\n          class=\"divider-left\"\n        >\n          <span class=\"hide-900\">", "</span>\n        </simple-icon-button-lite>\n        <simple-popover\n          ?hidden=\"", "\"\n          for=\"settings\"\n          auto\n          @value-changed=\"", "\"\n        >\n          <simple-fields-field\n            value=\"", "\"\n            type=\"checkbox\"\n            label=\"", "\"\n            name=\"photo\"\n          ></simple-fields-field>\n          <simple-fields-field\n            value=\"", "\"\n            type=\"checkbox\"\n            label=\"", "\"\n            name=\"email\"\n          ></simple-fields-field>\n          <simple-fields-field\n            value=\"", "\"\n            type=\"checkbox\"\n            label=\"", "\"\n            name=\"fname\"\n          ></simple-fields-field>\n          <simple-fields-field\n            value=\"", "\"\n            type=\"checkbox\"\n            label=\"", "\"\n            name=\"surname\"\n          ></simple-fields-field>\n        </simple-popover>"])), this.toggleSettings, !this.ready, this.t.settings, this.hideSettings, this.settingChanged, this.settings.photo, this.t.photo, this.settings.email, this.t.email, this.settings.fname, this.t.firstName, this.settings.surname, this.t.surname);
        } // toggle logic for button

      }, {
        key: "toggleSettings",
        value: function toggleSettings(e) {
          this.hideSettings = !this.hideSettings;

          if (!this.hideSettings) {
            this.hideGradeScale = true;
            this.hideActiveAssignment = true;
            this.hideActiveStudentOverview = true;
          }
        } // render active assignment in a popover

      }, {
        key: "renderActiveAssignmentBtn",
        value: function renderActiveAssignmentBtn() {
          return (0, _index.html)(_templateObject9_620dd880942811ecb1937969798b231a || (_templateObject9_620dd880942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<simple-icon-button-lite\n          icon=\"assignment\"\n          @click=\"", "\"\n          id=\"activeassignmentbtn\"\n          ?disabled=\"", "\"\n        >\n          <span class=\"hide-900\">", "</span>\n        </simple-icon-button-lite>\n        <simple-popover\n          ?hidden=\"", "\"\n          for=\"activeassignmentbtn\"\n          auto\n        >\n          ", "\n        </simple-popover>"])), this.toggleActiveAssignment, !this.ready, this.t.activeAssignment, this.hideActiveAssignment, this.database.assignments[this.activeAssignment] ? (0, _index.html)(_templateObject10_620dd880942811ecb1937969798b231a || (_templateObject10_620dd880942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                <h3>\n                  ", "\n                </h3>\n                ", ":\n                <local-time\n                  datetime=\"", "\"\n                  month=\"short\"\n                  day=\"numeric\"\n                  year=\"numeric\"\n                  hour=\"numeric\"\n                  minute=\"numeric\"\n                  time-zone-name=\"short\"\n                >\n                </local-time>\n              "])), this.database.assignments[this.activeAssignment].name, this.t.dueDate, this.database.assignments[this.activeAssignment]._ISODueDate) : _index.nothing);
        } // toggle logic for button

      }, {
        key: "toggleActiveAssignment",
        value: function toggleActiveAssignment(e) {
          this.hideActiveAssignment = !this.hideActiveAssignment;

          if (!this.hideActiveAssignment) {
            this.hideSettings = true;
            this.hideGradeScale = true;
            this.hideActiveStudentOverview = true;
          }
        } // RUBRICS
        // render grading scale in a popover

      }, {
        key: "renderRubricInfoBtn",
        value: function renderRubricInfoBtn(index, criteria, details) {
          return (0, _index.html)(_templateObject11_620dd880942811ecb1937969798b231a || (_templateObject11_620dd880942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<simple-icon-button-lite\n          icon=\"info\"\n          @click=\"", "\"\n          id=\"rubricinfo", "\"\n          data-id=\"", "\"\n        >\n          <span class=\"hide-900\" data-id=\"", "\">", "</span>\n        </simple-icon-button-lite>\n        <simple-popover\n          ?hidden=\"", "\"\n          for=\"rubricinfo", "\"\n          auto\n        >\n          <p style=\"max-width:20vw;\">", "</p>\n        </simple-popover>"])), this.toggleRubricInfo, index, index, index, criteria, !this.hideRubricInfo[index], index, details);
        } // toggle logic for button

      }, {
        key: "toggleRubricInfo",
        value: function toggleRubricInfo(e) {
          this.hideRubricInfo[parseInt(e.target.getAttribute("data-id"))] = !this.hideRubricInfo[parseInt(e.target.getAttribute("data-id"))];
          this.requestUpdate();
        } // displayMode toggle button

      }, {
        key: "renderDisplayModeBtn",
        value: function renderDisplayModeBtn() {
          var _this3 = this;

          return (0, _index.html)(_templateObject12_620dd880942811ecb1937969798b231a || (_templateObject12_620dd880942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral([" ", ""])), this.displayModeData().map(function (value, i) {
            return (0, _index.html)(_templateObject13_620dd880942811ecb1937969798b231a || (_templateObject13_620dd880942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n          <simple-icon-button-lite\n            icon=\"", "\"\n            @click=\"", "\"\n            data-id=\"", "\"\n            ?disabled=\"", "\"\n          >\n            <span class=\"hide-900\" data-id=\"", "\">", "</span>\n          </simple-icon-button-lite>\n        "])), value.icon, _this3.toggleDisplayMode, i, _this3.displayMode === i, i, value.label);
          }));
        }
      }, {
        key: "displayModeData",
        value: function displayModeData() {
          return [{
            label: "Mode 1",
            icon: "view-array",
            layout: "1-1-2"
          }, {
            label: "Mode 2",
            icon: "view-carousel",
            layout: "1-2-1"
          }, {
            label: "Two windows",
            icon: "open-in-new",
            layout: "1-2"
          }];
        }
      }, {
        key: "toggleDisplayMode",
        value: function toggleDisplayMode(e) {
          this.displayMode = parseInt(e.target.getAttribute("data-id")); // open in new window and drop the assignment display panel

          if (this.displayMode == 2) {
            this.openWindow();
          }
        }
      }], [{
        key: "properties",
        get: function get() {
          return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(_class), "properties", this)), {}, {
            hideSettings: {
              type: Boolean
            },
            hideGradeScale: {
              type: Boolean
            },
            hideActiveStudentOverview: {
              type: Boolean
            },
            hideActiveAssignment: {
              type: Boolean
            },
            hideRubricInfo: {
              type: Array
            }
          });
        }
      }, {
        key: "styles",
        get: function get() {
          return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(_class), "styles", this)), [(0, _index.css)(_templateObject14_620dd880942811ecb1937969798b231a || (_templateObject14_620dd880942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n          .top-controls simple-icon-button-lite {\n            margin: 0;\n            border-radius: 0;\n            height: 36px;\n          }\n          .top-controls simple-icon-button-lite:not([disabled])::part(button) {\n            outline: none;\n            height: 36px;\n            border-radius: 0;\n            padding: 0 8px;\n          }\n          .top-controls\n            simple-icon-button-lite:not([disabled])::part(button):focus,\n          .top-controls\n            simple-icon-button-lite:not([disabled])::part(button):hover {\n            background-color: var(--simple-colors-default-theme-accent-11);\n            outline: 2px solid var(--simple-colors-default-theme-accent-1);\n            outline-offset: -2px;\n            color: var(--simple-colors-default-theme-accent-1);\n          }\n          #activestudentbtn img,\n          #activestudentbtn simple-icon-lite {\n            --simple-icon-width: 32px;\n            --simple-icon-height: 32px;\n            width: 32px;\n            height: 32px;\n            border-radius: 50%;\n            vertical-align: middle;\n            margin-right: 4px;\n          }\n        "])))]);
        }
      }]);
      return _class;
    }(SuperClass);
  };

  _exports.UIRenderPieces = UIRenderPieces;
});
