// @lrnwebcomponents/grade-book/lib/grade-book-table.js
define(["exports", "../../../lit/index.js", "../../editable-table/editable-table.js"], function (_exports, _index, _editableTable) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.GradeBookTable = void 0;

  var _templateObject_62194a30942811ecb1937969798b231a, _templateObject2_62194a30942811ecb1937969798b231a, _templateObject3_62194a30942811ecb1937969798b231a;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var GradeBookTable = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(GradeBookTable, _LitElement);

    var _super = _createSuper(GradeBookTable);

    function GradeBookTable() {
      var _this;

      babelHelpers.classCallCheck(this, GradeBookTable);
      _this = _super.call(this);
      _this.editMode = false;
      return _this;
    }

    babelHelpers.createClass(GradeBookTable, [{
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_62194a30942811ecb1937969798b231a || (_templateObject_62194a30942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <editable-table\n        ?edit-mode=\"", "\"\n        bordered\n        column-header\n        condensed\n        disable-responsive\n        scroll\n        striped\n      >\n        <table>\n          <tbody>\n            <tr>\n              <td>", "</td>\n              <td>", "</td>\n              <td>", "</td>\n            </tr>\n            ", "\n          </tbody>\n        </table>\n      </editable-table>\n    "])), this.editMode, this.t.letterGrade, this.t.highRange, this.t.lowRange, this.database.gradeScale.map(function (scale) {
          return (0, _index.html)(_templateObject2_62194a30942811ecb1937969798b231a || (_templateObject2_62194a30942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                <tr>\n                  <td>", "</td>\n                  <td>", "</td>\n                  <td>", "</td>\n                </tr>\n              "])), scale.letter, scale.highRange, scale.lowRange);
        }));
      }
    }], [{
      key: "tag",
      get: function get() {
        return "grade-book-table";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          editMode: {
            type: Boolean,
            attribute: "edit-mode",
            reflect: true
          }
        };
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _index.css)(_templateObject3_62194a30942811ecb1937969798b231a || (_templateObject3_62194a30942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n      "])))];
      }
    }]);
    return GradeBookTable;
  }(_index.LitElement);

  _exports.GradeBookTable = GradeBookTable;
  customElements.define(GradeBookTable.tag, GradeBookTable);
});
