// @lrnwebcomponents/grade-book/lib/grade-book-student-block.js
define(["exports", "../../../lit/index.js", "../../i18n-manager/lib/I18NMixin.js", "../../simple-icon/lib/simple-icon-lite.js", "../../simple-icon/lib/simple-icons.js"], function (_exports, _index, _I18NMixin2, _simpleIconLite, _simpleIcons) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.GradeBookStudentBlock = void 0;

  var _templateObject_62179c80942811ecb1937969798b231a, _templateObject2_62179c80942811ecb1937969798b231a, _templateObject3_62179c80942811ecb1937969798b231a, _templateObject4_62179c80942811ecb1937969798b231a, _templateObject5_62179c80942811ecb1937969798b231a, _templateObject6_62179c80942811ecb1937969798b231a, _templateObject7_62179c80942811ecb1937969798b231a, _templateObject8_62179c80942811ecb1937969798b231a, _templateObject9_62179c80942811ecb1937969798b231a;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var GradeBookStudentBlock = /*#__PURE__*/function (_I18NMixin) {
    babelHelpers.inherits(GradeBookStudentBlock, _I18NMixin);

    var _super = _createSuper(GradeBookStudentBlock);

    function GradeBookStudentBlock() {
      var _this;

      babelHelpers.classCallCheck(this, GradeBookStudentBlock);
      _this = _super.call(this);
      _this.student = {};
      _this.t = {
        profileImageFor: "Profile image for",
        userID: "User ID",
        preferredName: "Preferred name",
        emailAddress: "Email this student",
        notes: "Notes",
        interests: "Interests"
      };

      _this.registerLocalization({
        context: babelHelpers.assertThisInitialized(_this),
        namespace: "grade-book"
      });

      return _this;
    }

    babelHelpers.createClass(GradeBookStudentBlock, [{
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_62179c80942811ecb1937969798b231a || (_templateObject_62179c80942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <grid-plate layout=\"1-2\" disable-responsive>\n        <div slot=\"col-1\" style=\"text-align:center;\">\n          ", "\n        </div>\n        <div slot=\"col-2\">\n          <h2>", "</h2>\n          <ul>\n            ", "\n            ", "\n            ", "\n            ", "\n            ", "\n          </ul>\n        </div>\n      </grid-plate>\n    "])), this.student.photo ? (0, _index.html)(_templateObject2_62179c80942811ecb1937969798b231a || (_templateObject2_62179c80942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                <img\n                  src=\"", "\"\n                  alt=\"", " ", "\"\n                />\n              "])), this.student.photo, this.t.photoOf, this.student.prefName) : (0, _index.html)(_templateObject3_62179c80942811ecb1937969798b231a || (_templateObject3_62179c80942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                <simple-icon-lite icon=\"account-circle\"></simple-icon-lite>\n              "]))), this.student.name, this.student.userId ? (0, _index.html)(_templateObject4_62179c80942811ecb1937969798b231a || (_templateObject4_62179c80942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<li>", ": ", "</li>"])), this.t.userID, this.student.userId) : "", this.student.prefName ? (0, _index.html)(_templateObject5_62179c80942811ecb1937969798b231a || (_templateObject5_62179c80942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<li>", ": ", "</li>"])), this.t.preferredName, this.student.prefName) : "", this.student.email ? (0, _index.html)(_templateObject6_62179c80942811ecb1937969798b231a || (_templateObject6_62179c80942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<li>\n                  <a href=\"mailto:", "\"\n                    >", "</a\n                  >\n                </li>"])), this.student.email, this.t.emailAddress) : "", this.student.notes ? (0, _index.html)(_templateObject7_62179c80942811ecb1937969798b231a || (_templateObject7_62179c80942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<li>", ": ", "</li>"])), this.t.notes, this.student.notes) : "", this.student.interests ? (0, _index.html)(_templateObject8_62179c80942811ecb1937969798b231a || (_templateObject8_62179c80942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<li>\n                  ", ": ", "\n                </li>"])), this.t.interests, this.student.interests.join(",")) : "");
      }
    }], [{
      key: "properties",
      get: function get() {
        return {
          student: {
            type: Object
          }
        };
      }
    }, {
      key: "styles",
      get: function get() {
        return (0, _index.css)(_templateObject9_62179c80942811ecb1937969798b231a || (_templateObject9_62179c80942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      :host {\n        display: block;\n      }\n      img {\n        max-width: 150px;\n        height: 150px;\n        border-radius: 50%;\n      }\n      grid-plate {\n        --grid-plate-col-transition: none;\n        --grid-plate-item-margin: 0px;\n        --grid-plate-item-padding: 4px;\n      }\n      simple-icon-lite {\n        --simple-icon-height: 150px;\n        --simple-icon-width: 150px;\n      }\n    "])));
      }
    }, {
      key: "tag",
      get: function get() {
        return "grade-book-student-block";
      }
    }]);
    return GradeBookStudentBlock;
  }((0, _I18NMixin2.I18NMixin)(_index.LitElement));

  _exports.GradeBookStudentBlock = GradeBookStudentBlock;
  customElements.define(GradeBookStudentBlock.tag, GradeBookStudentBlock);
});
