// @lrnwebcomponents/grade-book/lib/grade-book-pop-up.js
define(["exports", "require", "../../../lit/index.js", "./grade-book-store.js", "../../i18n-manager/lib/I18NMixin.js", "../../../mobx/dist/mobx.esm.js", "../../utils/utils.js"], function (_exports, _require, _index, _gradeBookStore, _I18NMixin2, _mobxEsm, _utils) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.GradeBookPopUp = void 0;
  _require = _interopRequireWildcard(_require);

  var _templateObject_6211f730942811ecb1937969798b231a, _templateObject2_6211f730942811ecb1937969798b231a, _templateObject3_6211f730942811ecb1937969798b231a, _templateObject4_6211f730942811ecb1937969798b231a;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var GradeBookPopUp = /*#__PURE__*/function (_I18NMixin) {
    babelHelpers.inherits(GradeBookPopUp, _I18NMixin);

    var _super = _createSuper(GradeBookPopUp);

    function GradeBookPopUp() {
      var _this;

      babelHelpers.classCallCheck(this, GradeBookPopUp);
      _this = _super.call(this);
      _this.t = {
        studentSubmission: "Student submission",
        openInNewWindow: "Open in new window",
        submitted: "Submitted"
      };
      (0, _mobxEsm.autorun)(function () {
        _this.activeSubmission = (0, _mobxEsm.toJS)(_gradeBookStore.GradeBookStore.activeSubmission);
      });
      return _this;
    } // forces an open root because this will render in a pop up that cannot use css from parent window :(


    babelHelpers.createClass(GradeBookPopUp, [{
      key: "createRenderRoot",
      value: function createRenderRoot() {
        return this;
      } // weirdest way to do this award goes to...

    }, {
      key: "changeActive",
      value: function changeActive(e) {
        var tmp = e.target.value.split("-");

        switch (tmp[0]) {
          case "prev":
            _gradeBookStore.GradeBookStore["active".concat(tmp[1])]--;
            break;

          case "next":
            _gradeBookStore.GradeBookStore["active".concat(tmp[1])]++;
            break;
        }
      }
    }, {
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_6211f730942811ecb1937969798b231a || (_templateObject_6211f730942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        grade-book-pop-up .shortcut-btns {\n          position: sticky;\n          background-color: var(\n            --simple-colors-default-theme-accent-12,\n            #001333\n          );\n          color: var(--simple-colors-default-theme-accent-1, #e2ecff);\n          height: 36px;\n          vertical-align: middle;\n          width: 100%;\n          display: flex;\n        }\n        grade-book-pop-up .shortcut-btns button {\n          background: transparent;\n          margin: 0 10px;\n          padding: 4px 8px;\n          font-size: 12px;\n          color: var(--simple-colors-default-theme-accent-1, #e2ecff);\n          border: none;\n          outline: 1px solid white;\n          cursor: pointer;\n        }\n        .active-submission {\n          overflow: hidden;\n          /* 16:9 aspect ratio */\n          padding-top: 56.25%;\n          position: relative;\n        }\n\n        .active-submission iframe {\n          border: 0;\n          height: 100%;\n          left: 0;\n          position: absolute;\n          top: 0;\n          width: 100%;\n        }\n      </style>\n      <div class=\"shortcut-btns\">\n        <button @click=\"", "\" value=\"prev-Student\">\n          Previous student\n        </button>\n        <button @click=\"", "\" value=\"next-Student\">\n          Next student\n        </button>\n        <button @click=\"", "\" value=\"prev-Assignment\">\n          Previous assignment\n        </button>\n        <button @click=\"", "\" value=\"next-Assignment\">\n          Next assignment\n        </button>\n      </div>\n      <div class=\"assignment-container\">\n        ", "\n      </div>\n    "])), this.changeActive, this.changeActive, this.changeActive, this.changeActive, this.renderStudentSubmission(this.activeSubmission));
      }
    }, {
      key: "renderStudentSubmission",
      value: function renderStudentSubmission(data) {
        var pre = (0, _index.html)(_templateObject2_6211f730942811ecb1937969798b231a || (_templateObject2_6211f730942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<h3>", "</h3>\n      ", "\n      <relative-time\n        datetime=\"", "\"\n      ></relative-time> "])), this.t.studentSubmission, this.t.submitted, _gradeBookStore.GradeBookStore.database.assignments[_gradeBookStore.GradeBookStore.activeAssignment]._ISODueDate); // test if this smells like a URL

        if ((0, _utils.validURL)(data)) {
          return (0, _index.html)(_templateObject3_6211f730942811ecb1937969798b231a || (_templateObject3_6211f730942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["", "<a\n          href=\"", "\"\n          target=\"_blank\"\n          rel=\"noopener noreferrer\"\n        >\n          <button>", "</button></a\n        >\n        <div class=\"active-submission\">\n          <iframe\n            width=\"560\"\n            height=\"315\"\n            loading=\"lazy\"\n            src=\"", "\"\n            frameborder=\"0\"\n            allow=\"encrypted-media;\"\n            allowfullscreen\n          ></iframe>\n        </div>"])), pre, data, this.t.openInNewWindow, (0, _utils.cleanVideoSource)(data));
        } else {
          // see if we can just present this as data
          new Promise(function (res, rej) {
            return _require.default(["../../md-block/md-block.js"], res, rej);
          });
          return (0, _index.html)(_templateObject4_6211f730942811ecb1937969798b231a || (_templateObject4_6211f730942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["", "\n        <div class=\"active-submission\">\n          <md-block .markdown=\"", "\"></md-block>\n        </div>"])), pre, data);
        }
      }
    }], [{
      key: "tag",
      get: function get() {
        return "grade-book-pop-up";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(GradeBookPopUp), "properties", this)), {}, {
          activeSubmission: {
            type: String
          },
          data: {
            type: String
          }
        });
      }
    }]);
    return GradeBookPopUp;
  }((0, _I18NMixin2.I18NMixin)(_index.LitElement));

  _exports.GradeBookPopUp = GradeBookPopUp;
  customElements.define(GradeBookPopUp.tag, GradeBookPopUp);
});
