// @lrnwebcomponents/grade-book/lib/letter-grade.js
define(["exports", "require", "../../../lit/index.js", "./grade-book-store.js", "../../../mobx/dist/mobx.esm.js"], function (_exports, _require, _index, _gradeBookStore, _mobxEsm) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LetterGrade = void 0;
  _require = _interopRequireWildcard(_require);

  var _templateObject_63805440942811ecb1937969798b231a, _templateObject2_63805440942811ecb1937969798b231a, _templateObject3_63805440942811ecb1937969798b231a, _templateObject4_63805440942811ecb1937969798b231a, _templateObject5_63805440942811ecb1937969798b231a, _templateObject6_63805440942811ecb1937969798b231a;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var LetterGrade = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(LetterGrade, _LitElement);

    var _super = _createSuper(LetterGrade);

    function LetterGrade() {
      var _this;

      babelHelpers.classCallCheck(this, LetterGrade);
      _this = _super.call(this);
      _this.pointsSystem = "pts";
      _this.score = null;
      _this.total = null;
      _this.gradeScale = [];
      _this.displayOnly = false;
      _this._letterIndex = null;
      _this.letter = "";
      _this.showScale = false;
      _this.value = "";
      _this.mini = false;
      _this.label = "";
      _this.active = false;
      (0, _mobxEsm.autorun)(function () {
        _this.gradeScale = (0, _mobxEsm.toJS)(_gradeBookStore.GradeBookStore.gradeScale);
      });
      return _this;
    }

    babelHelpers.createClass(LetterGrade, [{
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_63805440942811ecb1937969798b231a || (_templateObject_63805440942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"wrap\">\n        <div class=\"letter\">", "</div>\n        ", "\n        ", "\n      </div>\n      ", "\n    "])), this.letter, this.showScale && this._letterIndex !== null ? (0, _index.html)(_templateObject2_63805440942811ecb1937969798b231a || (_templateObject2_63805440942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<div class=\"score\">\n              (", "/", " ", ")\n            </div>"])), this.score, this.total, this.pointsSystem) : "", this.showScale && this._letterIndex !== null ? (0, _index.html)(_templateObject3_63805440942811ecb1937969798b231a || (_templateObject3_63805440942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<div class=\"range\">\n              ", "% -\n              ", "%\n            </div>"])), this.gradeScale[this._letterIndex].lowRange, this.gradeScale[this._letterIndex].highRange) : "", this.mini ? (0, _index.html)(_templateObject4_63805440942811ecb1937969798b231a || (_templateObject4_63805440942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<simple-tooltip position=\"top\"\n            >", "<br />", "</simple-tooltip\n          >"])), this.label, this.score != null ? (0, _index.html)(_templateObject5_63805440942811ecb1937969798b231a || (_templateObject5_63805440942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["(", "/", " ", ")"])), this.score, this.total, this.pointsSystem) : "") : "");
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName === "mini" && _this2[propName] && _this2.label) {
            new Promise(function (res, rej) {
              return _require.default(["../../simple-tooltip/simple-tooltip.js"], res, rej);
            });
          }

          if (["total", "score", "gradeScale"].includes(propName) && !_this2.displayOnly) {
            if (_this2.gradeScale.length > 0) {
              _this2.letter = _this2.calculateLetterGrade();
            }
          }

          if (propName === "letter" && _this2.gradeScale[_this2._letterIndex]) {
            // progressive enhancement like PDF printing
            _this2.innerHTML = "\n        <div>---------------------------------------</div>\n        <div style=\"font-size:36px;line-height:48px;\">".concat(_this2.letter, "</div>\n        <div style=\"font-size:20px;line-height:24px;\">").concat(_this2.score, "/").concat(_this2.total, " ").concat(_this2.pointsSystem, "</div>\n        <div style=\"font-size:16px;line-height:24px;\">Range: ").concat(_this2.gradeScale[_this2._letterIndex].lowRange, "% -\n        ").concat(_this2.gradeScale[_this2._letterIndex].highRange, "%</div>\n        <div>---------------------------------------</div>\n        ");
          }
        });
      }
      /**
       * return the letter grade based on score and gradeScale existing
       */

    }, {
      key: "calculateLetterGrade",
      value: function calculateLetterGrade() {
        // support no score, meaning it is not turned in / evaluated
        if (this.score === null) {
          return "/";
        }

        var percent = this.score / this.total * 100; // account for us having MORE points than max

        if (this.gradeScale[0] && percent >= this.gradeScale[0].highRange) {
          this._letterIndex = 0;
          return this.gradeScale[0].letter;
        } else {
          for (var i in this.gradeScale) {
            var val = this.gradeScale[i];

            if (percent <= val.highRange && percent >= val.lowRange) {
              this._letterIndex = i;
              return this.gradeScale[i].letter;
            }
          }

          this._letterIndex = this.gradeScale.length - 1;
          return this.gradeScale[this.gradeScale.length - 1].letter;
        }
      }
    }], [{
      key: "styles",
      get: function get() {
        return [(0, _index.css)(_templateObject6_63805440942811ecb1937969798b231a || (_templateObject6_63805440942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        :host([letter=\"/\"]) {\n          --letter-grade-default-letter-background-color: #cccccc;\n        }\n        :host([letter=\"F\"]) {\n          --letter-grade-default-letter-background-color: #ff4444;\n        }\n        :host([letter=\"D\"]) {\n          --letter-grade-default-letter-background-color: #ff7744;\n        }\n        :host([letter=\"C\"]),\n        :host([letter=\"C+\"]) {\n          --letter-grade-default-letter-background-color: #ffff00;\n        }\n        :host([letter=\"B\"]),\n        :host([letter=\"B-\"]) {\n          --letter-grade-default-letter-background-color: #ffff66;\n        }\n        :host([letter=\"B+\"]),\n        :host([letter=\"A-\"]) {\n          --letter-grade-default-letter-background-color: #aaffaa;\n        }\n        :host([letter=\"A\"]) {\n          --letter-grade-default-letter-background-color: #55ff55;\n        }\n        :host([show-scale]) .wrap {\n          height: var(--letter-grade-wrap-height, 64px);\n          width: var(--letter-grade-wrap-width, 64px);\n        }\n        :host([mini]) .wrap {\n          height: var(--letter-grade-wrap-height, 20px);\n          width: var(--letter-grade-wrap-width, 20px);\n          border: 1px solid black;\n          border-radius: 0px;\n          padding: 8px;\n        }\n        :host([mini]) .letter {\n          font-size: var(--letter-grade-letter-font-size, 12px);\n          line-height: var(\n            --letter-grade-letter-line-height,\n            var(--letter-grade-letter-font-size, 20px)\n          );\n        }\n        .wrap {\n          height: var(--letter-grade-wrap-height, 40px);\n          width: var(--letter-grade-wrap-width, 40px);\n          padding: var(--letter-grade-wrap-padding, 16px);\n          border: var(--letter-grade-wrap-border, 4px solid black);\n          border-radius: 50%;\n          box-shadow: rgb(0 0 0 / 20%) 0px 5px 5px 0px;\n          text-align: center;\n          justify-content: center;\n          vertical-align: middle;\n          background-color: var(\n            --letter-grade-wrap-background-color,\n            var(--letter-grade-default-letter-background-color, transparent)\n          );\n        }\n        .letter {\n          font-weight: var(--letter-grade-letter-font-weight, bold);\n          font-size: var(--letter-grade-letter-font-size, 40px);\n          line-height: var(\n            --letter-grade-letter-line-height,\n            var(--letter-grade-letter-font-size, 40px)\n          );\n          letter-spacing: 2px;\n        }\n        .score,\n        .range {\n          font-weight: var(--letter-grade-range-font-weight, normal);\n          font-size: var(--letter-grade-range-font-size, 12px);\n        }\n        simple-tooltip {\n          --simple-tooltip-background: #000000;\n          --simple-tooltip-opacity: 1;\n          --simple-tooltip-text-color: #ffffff;\n          --simple-tooltip-delay-in: 0;\n          --simple-tooltip-duration-in: 200ms;\n          --simple-tooltip-duration-out: 0;\n          --simple-tooltip-border-radius: 0;\n          --simple-tooltip-font-size: 14px;\n          --simple-tooltip-width: 145px;\n        }\n      "])))];
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          score: {
            type: Number
          },
          total: {
            type: Number
          },
          value: {
            type: Number
          },
          pointsSystem: {
            type: String,
            attribute: "points-system"
          },
          gradeScale: {
            type: Array
          },
          letter: {
            type: String,
            reflect: true
          },
          showScale: {
            type: Boolean,
            reflect: true,
            attribute: "show-scale"
          },
          mini: {
            type: Boolean,
            reflect: true
          },
          label: {
            type: String
          },
          displayOnly: {
            type: Boolean,
            attribute: "display-only"
          },
          active: {
            type: Boolean,
            reflect: true
          }
        };
      }
    }, {
      key: "tag",
      get: function get() {
        return "letter-grade";
      }
    }]);
    return LetterGrade;
  }(_index.LitElement);

  _exports.LetterGrade = LetterGrade;
  customElements.define(LetterGrade.tag, LetterGrade);
});
