// @lrnwebcomponents/grade-book/lib/letter-grade-picker.js
define(["exports", "require", "../../../lit/index.js", "./grade-book-store.js", "../../../mobx/dist/mobx.esm.js", "./letter-grade.js"], function (_exports, _require, _index, _gradeBookStore, _mobxEsm, _letterGrade) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LetterGradePicker = void 0;
  _require = _interopRequireWildcard(_require);

  var _templateObject_62c96050942811ecb1937969798b231a, _templateObject2_62c96050942811ecb1937969798b231a, _templateObject3_62c96050942811ecb1937969798b231a, _templateObject4_62c96050942811ecb1937969798b231a, _templateObject5_62c96050942811ecb1937969798b231a;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `letter-grade-picker`
   * `A headless gradebook that supports multiple backends with rubrics`
   * @demo demo/letter-grade-picker.html Letter grade picker
   * @element letter-grade-picker
   */
  var LetterGradePicker = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(LetterGradePicker, _LitElement);

    var _super = _createSuper(LetterGradePicker);

    function LetterGradePicker() {
      var _this;

      babelHelpers.classCallCheck(this, LetterGradePicker);
      _this = _super.call(this);
      _this.value = null;
      _this.label = null;
      _this.reveal = false;
      _this.arrow = false;
      _this.input = false;
      _this.possible = 0;
      _this.score = 0;
      return _this;
    }

    babelHelpers.createClass(LetterGradePicker, [{
      key: "render",
      value: function render() {
        var _this2 = this;

        return (0, _index.html)(_templateObject_62c96050942811ecb1937969798b231a || (_templateObject_62c96050942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"wrapper\">\n        <slot name=\"prefix\"><span class=\"prefix\">", "</span></slot>\n        ", "\n        ", "\n        <slot name=\"suffix\"></slot>\n        ", "\n      </div>\n    "])), this.label, _gradeBookStore.GradeBookStore.gradeScale.map(function (scale, index) {
          return (0, _index.html)(_templateObject2_62c96050942811ecb1937969798b231a || (_templateObject2_62c96050942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n            <button\n              ?data-selected-value=\"", "\"\n              letter=\"", "\"\n              id=\"btn", "\"\n              @click=\"", "\"\n              @mousedown=\"", "\"\n              @mouseover=\"", "\"\n              @focusin=\"", "\"\n              @mouseout=\"", "\"\n              @focusout=\"", "\"\n            >\n              <letter-grade\n                interactive\n                display-only\n                letter=\"", "\"\n                mini\n              ></letter-grade>\n            </button>\n          "])), scale.letter == _this2.value, scale.letter, index, _this2.clickScore, _this2.activeEventOn, _this2.activeEventOn, _this2.activeEventOn, _this2.activeEventOff, _this2.activeEventOff, scale.letter);
        }), this.input ? (0, _index.html)(_templateObject3_62c96050942811ecb1937969798b231a || (_templateObject3_62c96050942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n              <span class=\"score-wrapper\">\n                <simple-fields-field\n                  type=\"number\"\n                  part=\"simple-fields-field\"\n                  min=\"0\"\n                  value=\"", "\"\n                  max=\"", "\"\n                  @value-changed=\"", "\"\n                  maxlength=\"10\"\n                ></simple-fields-field>\n                / ", "</span\n              >\n            "])), this.score, this.possible, this.scoreUpdated, this.possible) : _index.nothing, this.input ? (0, _index.html)(_templateObject4_62c96050942811ecb1937969798b231a || (_templateObject4_62c96050942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral([""]))) : _index.nothing);
      }
    }, {
      key: "activeEventOn",
      value: function activeEventOn(e) {
        e.target.setAttribute("active", "active");
      }
    }, {
      key: "activeEventOff",
      value: function activeEventOff(e) {
        e.target.removeAttribute("active");
      }
    }, {
      key: "clickScore",
      value: function clickScore(e) {
        var _this3 = this;

        this.value = e.target.getAttribute("letter"); // Value on event can influence score

        var scaleItem = (0, _mobxEsm.toJS)(_gradeBookStore.GradeBookStore.gradeScale.find(function (item) {
          return _this3.value === item.letter;
        })); // forces it to be 2 decimal max

        this.score = Math.round(scaleItem.highRange / 100 * this.possible * 100) / 100;
      }
    }, {
      key: "scoreUpdated",
      value: function scoreUpdated(e) {
        // Score on input change can influence value
        var newScore = Math.round(e.detail.value / this.possible * 100 * 100) / 100;
        var scaleItem = (0, _mobxEsm.toJS)(_gradeBookStore.GradeBookStore.gradeScale.find(function (item) {
          return item.highRange >= newScore && newScore >= item.lowRange;
        }));
        this.value = scaleItem.letter;
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this4 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName === "input" && _this4[propName]) {
            new Promise(function (res, rej) {
              return _require.default(["../../simple-fields/lib/simple-fields-field.js"], res, rej);
            });
          }

          if (propName === "reveal" && _this4[propName]) {}

          if (["value", "score"].includes(propName)) {
            _this4.dispatchEvent(new CustomEvent("".concat(propName, "-changed"), {
              composed: false,
              bubbles: false,
              cancelable: true,
              detail: _this4
            }));
          }
        });
      }
    }], [{
      key: "properties",
      get: function get() {
        return {
          value: {
            type: String
          },
          label: {
            type: String
          },
          reveal: {
            type: Boolean
          },
          arrow: {
            type: Boolean
          },
          input: {
            type: Boolean
          },
          possible: {
            type: Number
          },
          score: {
            type: Number
          }
        };
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _index.css)(_templateObject5_62c96050942811ecb1937969798b231a || (_templateObject5_62c96050942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          font-size: 18px;\n        }\n        letter-grade {\n          display: inline-flex;\n          filter: brightness(80%) saturate(50%);\n        }\n        .wrapper {\n          display: flex;\n        }\n        span.prefix,\n        .wrapper ::slotted(*) {\n          margin-right: 10px;\n          width: 90px;\n          line-height: 38px;\n          display: block;\n          padding: 0;\n          text-align: end;\n          height: 38px;\n        }\n        button {\n          background-color: transparent;\n          padding: 0;\n          margin: 0;\n          border: none;\n          border-radius: none;\n          cursor: pointer;\n        }\n        button[active=\"active\"] letter-grade,\n        letter-grade[active=\"active\"] {\n          filter: brightness(90%) saturate(4);\n          outline: 2px solid black;\n          outline-offset: -2px;\n        }\n        button[data-selected-value] letter-grade {\n          filter: brightness(100%) saturate(1);\n          outline: 4px solid black;\n          outline-offset: -4px;\n        }\n        .score-wrapper {\n          display: flex;\n          font-size: 28px;\n          line-height: 36px;\n          word-spacing: -4px;\n        }\n        simple-fields-field {\n          width: 64px;\n          margin: 0px -8px 0px 16px;\n          --simple-fields-font-size: 28px;\n        }\n      "])))];
      }
    }, {
      key: "tag",
      get: function get() {
        return "letter-grade-picker";
      }
    }]);
    return LetterGradePicker;
  }(_index.LitElement);

  _exports.LetterGradePicker = LetterGradePicker;
  customElements.define(LetterGradePicker.tag, LetterGradePicker);
});
