// @lrnwebcomponents/elmsln-loading/elmsln-loading.js
define(["exports", "require", "../../lit-element/lit-element.js", "../simple-colors/simple-colors.js"], function (_exports, _require, _litElement, _simpleColors) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ElmslnLoading = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject2_ee9c03d001c111eba2339b0c52070e79() {
    var data = babelHelpers.taggedTemplateLiteral([" <iron-icon icon=\"lrn:network\"></iron-icon> "]);

    _templateObject2_ee9c03d001c111eba2339b0c52070e79 = function _templateObject2_ee9c03d001c111eba2339b0c52070e79() {
      return data;
    };

    return data;
  }

  function _templateObject_ee9c03d001c111eba2339b0c52070e79() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        @-moz-keyframes spin {\n          100% {\n            -moz-transform: rotate(60deg);\n            filter: saturate(10) invert(0.9);\n          }\n        }\n        @-webkit-keyframes spin {\n          100% {\n            -webkit-transform: rotate(60deg);\n            filter: saturate(10) invert(0.9);\n          }\n        }\n        @keyframes spin {\n          100% {\n            -webkit-transform: rotate(60deg);\n            transform: rotate(60deg);\n          }\n        }\n        :host iron-icon {\n          color: var(--simple-colors-default-theme-accent-6);\n          display: block;\n          -webkit-animation: spin 1.25s ease-out infinite;\n          -moz-animation: spin 1.25s ease-out infinite;\n          animation: spin 1.25s ease-out infinite;\n        }\n        :host([size=\"tiny\"]) iron-icon {\n          width: 16px;\n          height: 16px;\n          -webkit-animation: spin 0.75s ease-out infinite;\n          -moz-animation: spin 0.75s ease-out infinite;\n          animation: spin 0.75s ease-out infinite;\n        }\n        :host([size=\"small\"]) iron-icon {\n          width: 32px;\n          height: 32px;\n          -webkit-animation: spin 1s ease-out infinite;\n          -moz-animation: spin 1s ease-out infinite;\n          animation: spin 1s ease-out infinite;\n        }\n        :host([size=\"medium\"]) iron-icon {\n          width: 64px;\n          height: 64px;\n          -webkit-animation: spin 1.25s ease-out infinite;\n          -moz-animation: spin 1.25s ease-out infinite;\n          animation: spin 1.25s ease-out infinite;\n        }\n        :host([size=\"large\"]) iron-icon {\n          width: 80px;\n          height: 80px;\n          -webkit-animation: spin 1.25s ease-out infinite;\n          -moz-animation: spin 1.25s ease-out infinite;\n          animation: spin 1.25s ease-out infinite;\n        }\n        :host([size=\"epic\"]) iron-icon {\n          width: 400px;\n          height: 400px;\n          -webkit-animation: spin 2s ease-out infinite;\n          -moz-animation: spin 2s ease-out infinite;\n          animation: spin 2s ease-out infinite;\n        }\n      "]);

    _templateObject_ee9c03d001c111eba2339b0c52070e79 = function _templateObject_ee9c03d001c111eba2339b0c52070e79() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `elmsln-loading`
   * A spinner to tell the user that something is loading.
   * This is just the spinner though and doesn't provide any text or anything else.
   *
   * @microcopy - language worth noting:
   *  - elmsln - an open source NGDLE to save education
   *
   * @demo demo/index.html
   * @element elmsln-loading
   */
  var ElmslnLoading = /*#__PURE__*/function (_SimpleColors) {
    babelHelpers.inherits(ElmslnLoading, _SimpleColors);

    var _super = _createSuper(ElmslnLoading);

    babelHelpers.createClass(ElmslnLoading, null, [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnLoading), "styles", this)), [(0, _litElement.css)(_templateObject_ee9c03d001c111eba2339b0c52070e79())]);
      }
    }, {
      key: "tag",
      get: function get() {
        return "elmsln-loading";
      }
      /**
       * HTMLElement
       */

    }]);

    function ElmslnLoading() {
      var _this;

      babelHelpers.classCallCheck(this, ElmslnLoading);
      _this = _super.call(this);
      _this.size = "medium";
      new Promise(function (res, rej) {
        return _require.default(["../lrn-icons/lrn-icons.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/iron-icon/iron-icon.js"], res, rej);
      });
      return _this;
    }
    /**
     * LitElement properties changed
     */


    babelHelpers.createClass(ElmslnLoading, [{
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "color") {
            _this2._getAccentColor(_this2[propName]);
          }
        });
      }
      /**
       * LitElement / popular convention
       */

    }, {
      key: "render",

      /**
       * LitElement render
       */
      value: function render() {
        return (0, _litElement.html)(_templateObject2_ee9c03d001c111eba2339b0c52070e79());
      }
    }, {
      key: "_getAccentColor",
      value: function _getAccentColor(color) {
        color = color.replace("-text", "");

        if ((!this.accentColor || this.accentColor === "grey") && this.colors[color]) {
          this.accentColor = color;
        }
      }
    }], [{
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(ElmslnLoading), "properties", this)), {}, {
          /**
           * color
           */
          color: {
            type: String
          },

          /**
           * tiny, small, medium, large, epic sizing.
           */
          size: {
            type: String,
            reflect: true
          }
        });
      }
    }]);
    return ElmslnLoading;
  }(_simpleColors.SimpleColors);

  _exports.ElmslnLoading = ElmslnLoading;
  window.customElements.define(ElmslnLoading.tag, ElmslnLoading);
});
