// @lrnwebcomponents/lrndesign-gallery/lib/lrndesign-gallery-masonry.js
define(["exports", "../../../lit/index.js", "./lrndesign-gallery-behaviors.js", "./lrndesign-gallery-zoom.js", "./lrndesign-gallery-details.js", "../../utils/utils.js"], function (_exports, _index, _lrndesignGalleryBehaviors, _lrndesignGalleryZoom, _lrndesignGalleryDetails, _utils) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrndesignGalleryMasonry = void 0;

  var _templateObject_657df900942811ecb1937969798b231a, _templateObject2_657df900942811ecb1937969798b231a, _templateObject3_657df900942811ecb1937969798b231a, _templateObject4_657df900942811ecb1937969798b231a;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `lrndesign-gallery-masonry`
   * An element that renders a collection of gallery items into a grid or a single media item into a layout.
   *
   * @element lrndesign-gallery-masonry
   * @extends LrndesignGalleryBehaviors
   * @demo ./demo/masonry.html demo
   *
   */
  var LrndesignGalleryMasonry = /*#__PURE__*/function (_LrndesignGalleryBeha) {
    babelHelpers.inherits(LrndesignGalleryMasonry, _LrndesignGalleryBeha);

    var _super = _createSuper(LrndesignGalleryMasonry);

    /**
     * life cycle, element is ready
     */
    function LrndesignGalleryMasonry() {
      var _this;

      babelHelpers.classCallCheck(this, LrndesignGalleryMasonry);
      _this = _super.call(this);
      _this.__aspects = {};
      _this.__items = [];

      var target = _this.shadowRoot.querySelector("#carouselitem");

      if (_this.selected.scroll && target) {
        _this._scrollIntoView([_this._getParentOffset(target)]);

        if (!_this.selected.zoomed) target.focus();
      }

      return _this;
    } // render function


    babelHelpers.createClass(LrndesignGalleryMasonry, [{
      key: "render",
      value: function render() {
        var _this2 = this;

        return (0, _index.html)(_templateObject_657df900942811ecb1937969798b231a || (_templateObject_657df900942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <div id=\"grid\">\n        <p class=\"sr-only\">A list of columns buttons items:</p>\n        <div id=\"galleryscreen\">\n          ", "\n        </div>\n      </div>\n    "])), (this.__items || []).map(function (row, i) {
          return (0, _index.html)(_templateObject2_657df900942811ecb1937969798b231a || (_templateObject2_657df900942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n              <div>\n                ", "\n              </div>\n            "])), (row.cols || []).map(function (col, j) {
            return (0, _index.html)(_templateObject3_657df900942811ecb1937969798b231a || (_templateObject3_657df900942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                    <lrndesign-gallery-zoom\n                      .id=\"", "\"\n                      .details=\"", "\"\n                      .heading=\"", "\"\n                      .src=\"", "\"\n                      .tooltip=\"", "\"\n                      .zoom-alt=\"", "\"\n                      style=\"", "\"\n                    >\n                      <img\n                        .alt=\"", "\"\n                        fade\n                        .src=\"", "\"\n                        @load=\"", "\"\n                      />\n                      <div class=\"zoombg\"></div>\n                      <simple-icon-lite\n                        icon=\"zoom-in\"\n                        class=\"zoomicon\"\n                      ></simple-icon-lite>\n                    </lrndesign-gallery-zoom>\n                  "])), col.id, col.details, col.heading, col.large, col.tooltip, col.zoomAlt, _this2._getStyle(col.aspect, row.aspect), col.alt, col.src, function (e) {
              return _this2._handleImgLoad(e);
            });
          }));
        }));
      }
      /**
       * handle updates
       */

    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this3 = this;

        babelHelpers.get(babelHelpers.getPrototypeOf(LrndesignGalleryMasonry.prototype), "updated", this).call(this, changedProperties);
        changedProperties.forEach(function (oldValue, propName) {
          if (propName === "sources") _this3._updateItems();
          if (propName === "responsiveSize") _this3._updateItems();
        });
      }
    }, {
      key: "_getStyle",
      value: function _getStyle(aspect, total) {
        var a = "--lrndesign-gallery-image-aspect:".concat(aspect || 100 / 75, ";"),
            h = "--lrndesign-gallery-image-height:".concat(100 / aspect || 75, "%;"),
            w = "--lrndesign-gallery-image-width:".concat(aspect * 100 / total || 0.75, "%");
        return [a, h, w].join("");
      }
    }, {
      key: "_handleImgLoad",
      value: function _handleImgLoad(e) {
        var target = (0, _utils.normalizeEventPath)(e)[0];

        if (target.naturalWidth > 0) {
          this._updateItems();
        }
      }
    }, {
      key: "_updateItems",
      value: function _updateItems() {
        var items = [],
            settings = {
          xs: 2,
          sm: 3,
          md: 4,
          lg: 5,
          xl: 6
        },
            cols = settings[this.responsiveSize || "xs"],
            rows = Math.ceil(this.sources.length / cols);

        for (var i = 0; i < rows; i++) {
          var row = {
            aspect: 0,
            cols: []
          };

          for (var j = i * cols; j < Math.min(this.sources.length, (i + 1) * cols); j++) {
            var item = JSON.parse(JSON.stringify(this.sources[j])),
                img = this.shadowRoot.querySelector("img[src=\"".concat(item.src, "\"]"));
            item.aspect = img && img.naturalWidth > 0 ? img.naturalWidth / img.naturalHeight : 100 / 75;
            row.aspect += item.aspect;
            row.cols.push(item);
          }

          items.push(row);
        }

        this.__items = items;
      }
    }], [{
      key: "tag",
      get:
      /**
       * Store the tag name to make it easier to obtain directly.
       */
      function get() {
        return "lrndesign-gallery-masonry";
      }
    }, {
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(LrndesignGalleryMasonry), "styles", this)), [(0, _index.css)(_templateObject4_657df900942811ecb1937969798b231a || (_templateObject4_657df900942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          margin: 15px 0;\n          padding: 0;\n          max-width: 100%;\n          display: block;\n        }\n        #galleryscreen > div {\n          display: flex;\n          align-items: stretch;\n          justify-content: center;\n          margin: 20px 0;\n        }\n        lrndesign-gallery-zoom {\n          position: relative;\n          margin: 0 10px;\n          flex: 0 1 var(--lrndesign-gallery-image-width, 200px);\n          max-width: 35%;\n        }\n        lrndesign-gallery-zoom,\n        lrndesign-gallery-zoom:hover,\n        lrndesign-gallery-zoom:focus,\n        lrndesign-gallery-zoom:focus-within {\n          border: 0px solid transparent;\n        }\n        lrndesign-gallery-zoom img {\n          width: 100%;\n          transition: outline 0.25s ease-in-out;\n          outline: 1px solid var(--lrndesign-gallery-color);\n        }\n        lrndesign-gallery-zoom:hover img,\n        lrndesign-gallery-zoom:focus img,\n        lrndesign-gallery-zoom:focus-within img {\n          transition: outline 0.25s ease-in-out;\n          outline: 2px solid var(--lrndesign-gallery-focus-color);\n        }\n        .zoombg,\n        .zoomicon {\n          top: unset;\n          bottom: 5px;\n          left: 5px;\n        }\n      "])))]);
      } // properties available to the custom element for data binding

    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(LrndesignGalleryMasonry), "properties", this)), {}, {
          __aspects: {
            type: Object
          },
          __columns: {
            type: Array
          },
          __items: {
            type: Array
          }
        });
      }
    }]);
    return LrndesignGalleryMasonry;
  }(_lrndesignGalleryBehaviors.LrndesignGalleryBehaviors);

  _exports.LrndesignGalleryMasonry = LrndesignGalleryMasonry;
  window.customElements.define(LrndesignGalleryMasonry.tag, LrndesignGalleryMasonry);
});
