// @lrnwebcomponents/hero-banner/hero-banner.js
define(["exports", "require", "../../@polymer/polymer/polymer-element.js", "../simple-colors/lib/simple-colors-polymer.js", "../a11y-behaviors/a11y-behaviors.js"], function (_exports, _require, _polymerElement, _simpleColorsPolymer, _a11yBehaviors) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HeroBanner = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject_ed1a333083e811eab3c8e75dc3f340b6() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style include=\"simple-colors-shared-styles-polymer\">\n        :host {\n          display: block;\n          width: 100%;\n          min-height: 600px;\n          height: 100%;\n          max-height: 600px;\n          overflow: hidden;\n          position: relative;\n          --hero-banner-font-family: \"Roboto\";\n          --hero-banner-title-weight: 500;\n          --hero-banner-text: var(--simple-colors-default-theme-grey-12);\n          --hero-banner-rgba: rgba(255, 255, 255, 0.65);\n          --hero-banner-image-bg: var(--simple-colors-default-theme-grey-3);\n          --hero-banner-button-weight: bold;\n          --hero-banner-button-color: var(\n            --simple-colors-default-theme-accent-5\n          );\n          --hero-banner-button-hover-color: var(\n            --simple-colors-default-theme-accent-4\n          );\n        }\n        :host([dark]) {\n          --hero-banner-rgba: rgba(0, 0, 0, 0.65);\n        }\n        .image {\n          position: absolute;\n          left: 0;\n          right: 0;\n          width: 100%;\n          height: 100%;\n          background-color: var(--hero-banner-image-bg);\n        }\n        .itemwrapper {\n          position: absolute;\n          bottom: 10%;\n          left: 10%;\n          width: 50%;\n        }\n        .title {\n          background-color: var(--hero-banner-rgba);\n          padding: 10px 16px;\n          font-size: 32px;\n          color: var(--hero-banner-text);\n          margin: 4px 0;\n          font-family: var(--hero-banner-font-family);\n          font-weight: var(--hero-banner-title-weight);\n        }\n        .details {\n          background-color: var(--hero-banner-rgba);\n          padding: 10px 16px;\n          font-size: 16px;\n          color: var(--hero-banner-text);\n          margin: 4px 0;\n          font-family: var(--hero-banner-font-family);\n        }\n        .linkbutton {\n          padding: 0;\n          margin: 8px 0;\n          text-decoration: none;\n          font-family: var(--hero-banner-font-family);\n        }\n        .linkbutton paper-button {\n          text-transform: none;\n          font-weight: var(--hero-banner-button-weight);\n          color: var(--hero-banner-text);\n          background-color: var(--hero-banner-button-color);\n          font-size: 16px;\n          margin: 0;\n        }\n        .linkbutton:focus paper-button,\n        .linkbutton:hover paper-button {\n          background-color: var(--hero-banner-button-hover-color);\n        }\n        @media screen and (max-width: 720px) {\n          .title {\n            font-size: 20px;\n          }\n          .details {\n            font-size: 12px;\n          }\n          .itemwrapper {\n            left: 5%;\n            width: 50%;\n          }\n        }\n        @media screen and (max-width: 500px) {\n          .title {\n            font-size: 16px;\n          }\n          .details {\n            display: none;\n          }\n          .itemwrapper {\n            left: 0;\n            width: 300px;\n          }\n        }\n      </style>\n      <iron-image\n        class=\"image\"\n        src$=\"[[image]]\"\n        fade=\"\"\n        preload=\"\"\n        sizing=\"cover\"\n      ></iron-image>\n      <div class=\"itemwrapper\">\n        <div class=\"title\">[[title]]</div>\n        <div class=\"details\">[[details]]</div>\n        <a class=\"linkbutton\" href$=\"[[buttonLink]]\"\n          ><paper-button raised=\"\">[[buttonText]]</paper-button></a\n        >\n      </div>\n    "], ["\n      <style include=\"simple-colors-shared-styles-polymer\">\n        :host {\n          display: block;\n          width: 100%;\n          min-height: 600px;\n          height: 100%;\n          max-height: 600px;\n          overflow: hidden;\n          position: relative;\n          --hero-banner-font-family: \"Roboto\";\n          --hero-banner-title-weight: 500;\n          --hero-banner-text: var(--simple-colors-default-theme-grey-12);\n          --hero-banner-rgba: rgba(255, 255, 255, 0.65);\n          --hero-banner-image-bg: var(--simple-colors-default-theme-grey-3);\n          --hero-banner-button-weight: bold;\n          --hero-banner-button-color: var(\n            --simple-colors-default-theme-accent-5\n          );\n          --hero-banner-button-hover-color: var(\n            --simple-colors-default-theme-accent-4\n          );\n        }\n        :host([dark]) {\n          --hero-banner-rgba: rgba(0, 0, 0, 0.65);\n        }\n        .image {\n          position: absolute;\n          left: 0;\n          right: 0;\n          width: 100%;\n          height: 100%;\n          background-color: var(--hero-banner-image-bg);\n        }\n        .itemwrapper {\n          position: absolute;\n          bottom: 10%;\n          left: 10%;\n          width: 50%;\n        }\n        .title {\n          background-color: var(--hero-banner-rgba);\n          padding: 10px 16px;\n          font-size: 32px;\n          color: var(--hero-banner-text);\n          margin: 4px 0;\n          font-family: var(--hero-banner-font-family);\n          font-weight: var(--hero-banner-title-weight);\n        }\n        .details {\n          background-color: var(--hero-banner-rgba);\n          padding: 10px 16px;\n          font-size: 16px;\n          color: var(--hero-banner-text);\n          margin: 4px 0;\n          font-family: var(--hero-banner-font-family);\n        }\n        .linkbutton {\n          padding: 0;\n          margin: 8px 0;\n          text-decoration: none;\n          font-family: var(--hero-banner-font-family);\n        }\n        .linkbutton paper-button {\n          text-transform: none;\n          font-weight: var(--hero-banner-button-weight);\n          color: var(--hero-banner-text);\n          background-color: var(--hero-banner-button-color);\n          font-size: 16px;\n          margin: 0;\n        }\n        .linkbutton:focus paper-button,\n        .linkbutton:hover paper-button {\n          background-color: var(--hero-banner-button-hover-color);\n        }\n        @media screen and (max-width: 720px) {\n          .title {\n            font-size: 20px;\n          }\n          .details {\n            font-size: 12px;\n          }\n          .itemwrapper {\n            left: 5%;\n            width: 50%;\n          }\n        }\n        @media screen and (max-width: 500px) {\n          .title {\n            font-size: 16px;\n          }\n          .details {\n            display: none;\n          }\n          .itemwrapper {\n            left: 0;\n            width: 300px;\n          }\n        }\n      </style>\n      <iron-image\n        class=\"image\"\n        src\\$=\"[[image]]\"\n        fade=\"\"\n        preload=\"\"\n        sizing=\"cover\"\n      ></iron-image>\n      <div class=\"itemwrapper\">\n        <div class=\"title\">[[title]]</div>\n        <div class=\"details\">[[details]]</div>\n        <a class=\"linkbutton\" href\\$=\"[[buttonLink]]\"\n          ><paper-button raised=\"\">[[buttonText]]</paper-button></a\n        >\n      </div>\n    "]);

    _templateObject_ed1a333083e811eab3c8e75dc3f340b6 = function _templateObject_ed1a333083e811eab3c8e75dc3f340b6() {
      return data;
    };

    return data;
  }

  /**
   * `hero-banner`
   * `That thing no one wants to make over and over again yet always does...`
   * @demo demo/index.html
   * @element hero-banner
   */
  var HeroBanner =
  /*#__PURE__*/
  function (_A11yBehaviors) {
    babelHelpers.inherits(HeroBanner, _A11yBehaviors);

    function HeroBanner() {
      var _this;

      babelHelpers.classCallCheck(this, HeroBanner);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(HeroBanner).call(this));
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/paper-button/paper-button.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/iron-image/iron-image.js"], res, rej);
      });
      return _this;
    }

    babelHelpers.createClass(HeroBanner, null, [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_ed1a333083e811eab3c8e75dc3f340b6());
      }
    }, {
      key: "tag",
      get: function get() {
        return "hero-banner";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(HeroBanner), "properties", this), {
          /**
           * Title
           */
          title: {
            type: String,
            value: "Title"
          },

          /**
           * Image
           */
          image: {
            type: String
          },

          /**
           * Details / teaser text
           */
          details: {
            type: String,
            value: "Details"
          },

          /**
           * button label
           */
          buttonText: {
            type: String,
            value: "Find out more"
          },

          /**
           * url for the button
           */
          buttonLink: {
            type: String
          }
        });
      }
    }, {
      key: "haxProperties",
      get: function get() {
        return {
          canScale: false,
          canPosition: false,
          canEditSource: false,
          gizmo: {
            title: "Hero image",
            description: "Typically fancy banner image calling your attention to something.",
            icon: "image:panorama",
            color: "red",
            groups: ["Image", "Media"],
            handles: [{
              type: "image",
              source: "image",
              title: "title",
              description: "details",
              link: "buttonLink"
            }],
            meta: {
              author: "ELMS:LN"
            }
          },
          settings: {
            quick: [{
              property: "title",
              title: "Title",
              description: "The title of the element",
              inputMethod: "textfield",
              icon: "editor:title"
            }, {
              property: "image",
              title: "Image",
              description: "URL of the image",
              inputMethod: "textfield",
              icon: "image:panorama"
            }, {
              property: "details",
              title: "Details",
              description: "Additional text detail / teaser data",
              inputMethod: "textfield",
              icon: "editor:text-fields"
            }, {
              property: "buttonText",
              title: "Button",
              description: "Label of the button",
              inputMethod: "textfield",
              icon: "icons:radio-button-unchecked"
            }, {
              property: "accentColor",
              title: "Accent color",
              description: "Select the accent color use",
              inputMethod: "colorpicker",
              icon: "editor:format-color-fill"
            }, {
              property: "dark",
              title: "Dark",
              description: "Use dark theme",
              inputMethod: "boolean",
              icon: "invert-colors"
            }, {
              property: "buttonLink",
              title: "Button - Link",
              description: "Label of the button",
              inputMethod: "textfield",
              validationType: "url",
              icon: "icons:link"
            }],
            configure: [{
              property: "title",
              title: "Title",
              description: "The title of the element",
              inputMethod: "textfield",
              icon: "editor:title"
            }, {
              property: "image",
              title: "Image",
              description: "URL of the image",
              inputMethod: "haxupload",
              icon: "image:panorama"
            }, {
              property: "details",
              title: "Details",
              description: "Additional text detail / teaser data",
              inputMethod: "textfield",
              icon: "editor:text-fields"
            }, {
              property: "buttonText",
              title: "Button",
              description: "Label of the button",
              inputMethod: "textfield",
              icon: "icons:radio-button-unchecked"
            }, {
              property: "accentColor",
              title: "Accent color",
              description: "Select the accent color use",
              inputMethod: "colorpicker",
              icon: "editor:format-color-fill"
            }, {
              property: "dark",
              title: "Dark",
              description: "Use dark theme",
              inputMethod: "boolean",
              icon: "invert-colors"
            }, {
              property: "buttonLink",
              title: "Button - Link",
              description: "Label of the button",
              inputMethod: "haxupload",
              validationType: "url",
              icon: "icons:link"
            }],
            advanced: []
          }
        };
      }
    }]);
    return HeroBanner;
  }((0, _a11yBehaviors.A11yBehaviors)(_simpleColorsPolymer.SimpleColorsPolymer));

  _exports.HeroBanner = HeroBanner;
  window.customElements.define(HeroBanner.tag, HeroBanner);
});
