// @lrnwebcomponents/paper-fab-speed-dial/paper-fab-speed-dial.js
define(["exports", "require", "../../@polymer/polymer/polymer-element.js"], function (_exports, _require, _polymerElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.PaperFabSpeedDial = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  /**
   * Copyright 2018 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `paper-fab-speed-dial`
   * @element paper-fab-speed-dial
   * `A speed dial setup for a floating action button`
   *
   * @demo demo/index.html
   */
  var PaperFabSpeedDial =
  /*#__PURE__*/
  function (_PolymerElement) {
    babelHelpers.inherits(PaperFabSpeedDial, _PolymerElement);

    function PaperFabSpeedDial() {
      var _this;

      babelHelpers.classCallCheck(this, PaperFabSpeedDial);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(PaperFabSpeedDial).call(this));
      new Promise(function (res, rej) {
        return _require.default(["./lib/paper-fab-speed-dial-overlay.js"], res, rej);
      });
      return _this;
    }

    babelHelpers.createClass(PaperFabSpeedDial, [{
      key: "open",
      // Public methods
      value: function open(e) {
        // Required for mobile Safari to avoid passing the tap event to an element below the FAB
        if (e) {
          e.preventDefault();
        }

        this.opened = true;
      }
    }, {
      key: "close",
      value: function close(e) {
        // Required for mobile Safari to avoid passing the tap event to an element below the FAB
        if (e) {
          e.preventDefault();
        }

        this.opened = false;
      }
    }], [{
      key: "tag",
      get: function get() {
        return "paper-fab-speed-dial";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          icon: {
            type: String,
            value: "add"
          },
          opened: {
            type: Boolean,
            notify: true
          },
          disabled: {
            type: Boolean,
            value: false
          }
        };
      }
    }]);
    return PaperFabSpeedDial;
  }(_polymerElement.PolymerElement);

  _exports.PaperFabSpeedDial = PaperFabSpeedDial;
  window.customElements.define(PaperFabSpeedDial.tag, PaperFabSpeedDial);
});
