// @lrnwebcomponents/paper-fab-speed-dial/paper-fab-speed-dial.js
define(["exports", "require", "../../@polymer/polymer/polymer-element.js"], function (_exports, _require, _polymerElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.PaperFabSpeedDial = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `paper-fab-speed-dial`
   * @element paper-fab-speed-dial
   * `A speed dial setup for a floating action button`
   *
   * @demo demo/index.html
   */
  var PaperFabSpeedDial = /*#__PURE__*/function (_PolymerElement) {
    babelHelpers.inherits(PaperFabSpeedDial, _PolymerElement);

    var _super = _createSuper(PaperFabSpeedDial);

    function PaperFabSpeedDial() {
      var _this;

      babelHelpers.classCallCheck(this, PaperFabSpeedDial);
      _this = _super.call(this);
      new Promise(function (res, rej) {
        return _require.default(["./lib/paper-fab-speed-dial-overlay.js"], res, rej);
      });
      return _this;
    }

    babelHelpers.createClass(PaperFabSpeedDial, [{
      key: "open",
      // Public methods
      value: function open(e) {
        // Required for mobile Safari to avoid passing the tap event to an element below the FAB
        if (e) {
          e.preventDefault();
        }

        this.opened = true;
      }
    }, {
      key: "close",
      value: function close(e) {
        // Required for mobile Safari to avoid passing the tap event to an element below the FAB
        if (e) {
          e.preventDefault();
        }

        this.opened = false;
      }
    }], [{
      key: "tag",
      get: function get() {
        return "paper-fab-speed-dial";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          icon: {
            type: String,
            value: "add"
          },
          opened: {
            type: Boolean,
            notify: true
          },
          disabled: {
            type: Boolean,
            value: false
          }
        };
      }
    }]);
    return PaperFabSpeedDial;
  }(_polymerElement.PolymerElement);

  _exports.PaperFabSpeedDial = PaperFabSpeedDial;
  window.customElements.define(PaperFabSpeedDial.tag, PaperFabSpeedDial);
});
