// @lrnwebcomponents/lrnsys-layout/lib/lrnsys-button-inner.js
define(["exports", "require", "../../../lit-element/lit-element.js"], function (_exports, _require, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnsysButtonInner = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject5_f49a70d02c1c11eaacb4bf54e86db688() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        :host > div > * {\n          vertical-align: middle;\n        }\n        .text-label {\n          margin-left: 8px;\n        }\n        .text-label-only {\n          margin-left: 0;\n        }\n      "]);

    _templateObject5_f49a70d02c1c11eaacb4bf54e86db688 = function _templateObject5_f49a70d02c1c11eaacb4bf54e86db688() {
      return data;
    };

    return data;
  }

  function _templateObject4_f49a70d02c1c11eaacb4bf54e86db688() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <span class=\"", "\"\n                >", "</span\n              >\n            "]);

    _templateObject4_f49a70d02c1c11eaacb4bf54e86db688 = function _templateObject4_f49a70d02c1c11eaacb4bf54e86db688() {
      return data;
    };

    return data;
  }

  function _templateObject3_f49a70d02c1c11eaacb4bf54e86db688() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <lrn-icon icon=\"", "\"></lrn-icon>\n            "]);

    _templateObject3_f49a70d02c1c11eaacb4bf54e86db688 = function _templateObject3_f49a70d02c1c11eaacb4bf54e86db688() {
      return data;
    };

    return data;
  }

  function _templateObject2_f49a70d02c1c11eaacb4bf54e86db688() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <paper-avatar src=\"", "\"></paper-avatar>\n            "]);

    _templateObject2_f49a70d02c1c11eaacb4bf54e86db688 = function _templateObject2_f49a70d02c1c11eaacb4bf54e86db688() {
      return data;
    };

    return data;
  }

  function _templateObject_f49a70d02c1c11eaacb4bf54e86db688() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div>\n        ", "\n        ", "\n        ", "\n      </div>\n      <div><slot name=\"button\"></slot><slot></slot></div>\n    "]);

    _templateObject_f49a70d02c1c11eaacb4bf54e86db688 = function _templateObject_f49a70d02c1c11eaacb4bf54e86db688() {
      return data;
    };

    return data;
  }

  /**
   * `lrnsys-button-inner`
   * @customElement lrnsys-button-inner
   */
  var LrnsysButtonInner =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnsysButtonInner, _LitElement);

    function LrnsysButtonInner() {
      babelHelpers.classCallCheck(this, LrnsysButtonInner);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnsysButtonInner).apply(this, arguments));
    }

    babelHelpers.createClass(LrnsysButtonInner, [{
      key: "render",

      /**
       * LitElement render
       */
      value: function render() {
        return (0, _litElement.html)(_templateObject_f49a70d02c1c11eaacb4bf54e86db688(), this.avatar ? (0, _litElement.html)(_templateObject2_f49a70d02c1c11eaacb4bf54e86db688(), this.avatar) : "", this.icon ? (0, _litElement.html)(_templateObject3_f49a70d02c1c11eaacb4bf54e86db688(), this.icon) : "", this.text ? (0, _litElement.html)(_templateObject4_f49a70d02c1c11eaacb4bf54e86db688(), this._getTextLabelClass(this.avatar, this.icon), this.text) : "");
      }
      /**
       * LitElement properties changed
       */

    }, {
      key: "updated",
      value: function updated(changedProperties) {
        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "avatar") {
            new Promise(function (res, rej) {
              return _require.default(["../../paper-avatar/paper-avatar.js"], res, rej);
            });
          }

          if (propName == "icon") {
            new Promise(function (res, rej) {
              return _require.default(["../../lrn-icons/lrn-icons.js"], res, rej);
            });
            new Promise(function (res, rej) {
              return _require.default(["../../../@polymer/iron-icons/iron-icons.js"], res, rej);
            });
          }
        });
      }
    }, {
      key: "_getTextLabelClass",

      /**
       * Find out if the text does not have an avatar or an icon to the left,
       * and add a class to remove the left margin.
       */
      value: function _getTextLabelClass(avatar, icon) {
        if (!avatar && !icon) {
          return "text-label-only";
        }

        return "text-label";
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject5_f49a70d02c1c11eaacb4bf54e86db688())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrnsys-button-inner";
      }
      /**
       * LitElement / popular convention
       */

    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Icon to present for clicking.
           */
          icon: {
            type: String
          },

          /**
           * Icon to present for clicking.
           */
          avatar: {
            type: String
          },

          /**
           * Text to present for clicking.
           */
          text: {
            type: String
          }
        };
      }
    }]);
    return LrnsysButtonInner;
  }(_litElement.LitElement);

  _exports.LrnsysButtonInner = LrnsysButtonInner;
  window.customElements.define(LrnsysButtonInner.tag, LrnsysButtonInner);
});
