// @haxtheweb/app-hax/app-hax.js
import{css as e,html as t,unsafeCSS as o}from"../../lit/index.js";import{toJS as a,autorun as i}from"../../mobx/dist/mobx.esm.js";import{localStorageSet as n,localStorageGet as s}from"../utils/utils.js";import"../simple-tooltip/simple-tooltip.js";import{SimpleColors as l}from"../simple-colors/simple-colors.js";import{store as r}from"./lib/v2/AppHaxStore.js";import{I18NMixin as p}from"../i18n-manager/lib/I18NMixin.js";import{AppHaxAPI as d}from"./lib/v2/AppHaxBackendAPI.js";import{SimpleTourManager as h}from"../simple-popover/lib/simple-tour.js";import{SuperDaemonInstance as m}from"../super-daemon/super-daemon.js";import"../simple-toolbar/lib/simple-toolbar-button.js";import"../simple-colors-shared-styles/simple-colors-shared-styles.js";import"./lib/v2/AppHaxRouter.js";import"./lib/v2/app-hax-label.js";import"../haxcms-elements/lib/core/ui/app-hax-top-bar.js";import{SimpleTourFinder as c}from"../simple-popover/lib/SimpleTourFinder.js";import"./lib/v2/app-hax-use-case.js";import"./lib/v2/app-hax-use-case-filter.js";import"./lib/v2/app-hax-search-results.js";import"./lib/v2/app-hax-scroll-button.js";const u=new URL("./lib/assets/images/Logout.svg",import.meta.url).href;function darkToggle(e){e.matches?r.darkMode=!0:r.darkMode=!1}function soundToggle(){r.soundStatus=!a(r.soundStatus),n("app-hax-soundStatus",a(r.soundStatus)),r.appEl.playSound("click")}export class AppHax extends(p(c(l))){static get tag(){return"app-hax"}_openExternalLink(e){globalThis.open(e,"_blank")}async _haxStoreContribute(e,t,o=null){let a="";if("merlin"==e){var i=`[${e}] New command request from HAX daemon`;a=`Location: ${globalThis.location.href}\nMerlin command: ${o}\nWhat did you want merlin to do?\n`}else{i=`[${e}] User report from HAX daemon`;if(a=`Location: ${globalThis.location.href}\nBrowser: ${navigator.userAgent}\nOS: ${navigator.userAgentData.platform} - ${navigator.deviceMemory}GB RAM - ${navigator.hardwareConcurrency} cores\nScreen: ${globalThis.screen.width}x${globalThis.screen.height}\nWindow size: ${globalThis.innerWidth}x${globalThis.innerHeight}\n`,navigator.getBattery){a+=`Battery: ${100*(await globalThis.navigator.getBattery()).level}%\n`}navigator.connection&&globalThis.navigator.connection.effectiveType&&(a+=`Connection: ${navigator.connection.effectiveType}\n`),a+=("feature"==e?"Your idea:":"Bug you experienced:")+"\n"}globalThis.open(`https://github.com/haxtheweb/issues/issues/new?assignees=&labels=${t}&template=issue-report.md&title=${i}&body=${encodeURIComponent(a)}`,"_blank")}playSound(e="coin2"){return new Promise((t=>{if(r.soundStatus&&r.appReady){let o=["click","click2","coin","coin2","hit","success"].includes(e)?e:"hit";this.audio=new Audio(new URL(`./lib/assets/sounds/${o}.mp3`,import.meta.url).href),this.audio.volume=.2,this.audio.onended=e=>{t()},this.audio.play(),setTimeout((()=>{t()}),1e3)}else t()}))}_tokenRefreshFailed(e){globalThis.dispatchEvent(new CustomEvent("jwt-login-logout",{composed:!0,bubbles:!0,cancelable:!1,detail:!0})),setTimeout((()=>{this.reset(!0)}),100)}connectedCallback(){super.connectedCallback(),globalThis.matchMedia("(prefers-color-scheme: dark)").addEventListener("change",darkToggle,{signal:this.windowControllers.signal}),globalThis.addEventListener("jwt-logged-in",this._jwtLoggedIn.bind(this),{signal:this.windowControllers.signal}),globalThis.addEventListener("jwt-login-refresh-error",this._tokenRefreshFailed.bind(this),{signal:this.windowControllers.signal})}goToLocation(e){globalThis.location=e}disconnectedCallback(){this.windowControllers.abort(),super.disconnectedCallback()}constructor(){super(),this.unlockComingSoon=!1,this.unlockTerrible=!1,this.__loginModalOpen=!1,this.t=this.t||{},this.t={...this.t,selectPage:"Select page",backToSiteList:"Back to site list",listMySites:"List my sites",cancel:"Cancel",editDetails:"Page details",add:"Add",editSettings:"Edit settings",source:"Source",viewSource:"View source",findMedia:"Find media",undo:"Undo",redo:"Redo",media:"Media",outline:"Outline",blocks:"Blocks",addBlock:"Add block",addPage:"Add page",addChildPage:"Add child page",clonePage:"Clone page",delete:"Delete page",siteSettings:"Site settings",close:"Close",settings:"Settings",edit:"Edit",configureBlock:"Configure block",configure:"Configure",save:"Save",home:"Home",startNewJourney:"Start new journey",newJourney:"New Journey",accountInfo:"Account Info",outlineDesigner:"Site Outline",pageOutline:"Structure",more:"More",siteActions:"Site actions",insights:"Insights dashboard",merlin:"Merlin",summonMerlin:"Summon Merlin",logOut:"Log out",menu:"Menu",showMore:"More",userAccess:"User Access"},void 0!==globalThis.speechSynthesis&&(globalThis.SpeechRecognition||globalThis.webkitSpeechRecognition||globalThis.mozSpeechRecognition||globalThis.msSpeechRecognition||globalThis.oSpeechRecognition)&&(m.voiceSearch=!0),m.icon="hax:wizard-hat",m.appendContext("*"),m.allowedCallback=()=>!(!a(r.appReady)||!a(r.isLoggedIn)),m.defineOption({title:"Tour of top menu buttons",icon:"help",tags:["Help","ui","tour"],priority:-1e3,value:{target:this,method:"helpClick",args:[]},eventName:"super-daemon-element-method",path:"HAX/app/tour",context:["*"]}),m.defineOption({title:"Unlock hidden features",icon:"hax:hax2022",tags:["Developer","features","hidden"],value:{target:this,method:"fireUnlocked",args:[]},eventName:"super-daemon-element-method",path:">developer/hax/unlockAll",context:[">"]}),m.defineOption({title:"Unlock terrible 2000s themes",icon:"hax:table-multiple",tags:["Developer","terrible","2000","tables"],value:{target:this,method:"fireTerrible",args:[]},eventName:"super-daemon-element-method",path:">developer/hax/terrible2000s",context:[">"]}),m.defineOption({title:"Go to site",icon:"hax:hax2022",tags:["Sites","Administration","change"],eventName:"super-daemon-run-program",path:"HAX/action/goToSite",value:{name:"Go to site",program:async(e,t)=>{let o=[];return a(r.manifest.items).forEach((async t=>{(""==e||t.metadata.site&&t.metadata.site.name&&t.metadata.site.name.includes(e))&&o.push({title:t.title,icon:t.metadata.theme&&t.metadata.theme.variables&&t.metadata.theme.variables.icon?t.metadata.theme.variables.icon:"hax:hax2022",tags:["site",t.description],value:{target:this,method:"goToLocation",args:[t.slug]},eventName:"super-daemon-element-method",context:["*","HAX/action/goToSite/"+t.metadata.site.name],path:"HAX/action/goToSite/"+t.metadata.site.name})})),o}},context:["*"]}),m.defineOption({title:"Join our Community",icon:"hax:discord",priority:-100,tags:["community","discord","chat","help"],value:{target:this,method:"_openExternalLink",args:["https://bit.ly/hax-discord"]},eventName:"super-daemon-element-method",path:"HAX/community/join",context:["logged-in","CMS","HAX","*"]}),m.defineOption({title:"User Tutorials",icon:"hax:hax2022",priority:-1e3,tags:["Documentation","community","help"],value:{target:this,method:"_openExternalLink",args:["https://haxtheweb.org/tutorials"]},eventName:"super-daemon-element-method",path:"HAX/community/tutorials",context:["logged-in","CMS","HAX","*"]}),m.defineOption({title:"User Documentation",icon:"hax:hax2022",tags:["Documentation","community","help"],value:{target:this,method:"_openExternalLink",args:["https://haxtheweb.org/documentation"]},eventName:"super-daemon-element-method",path:"HAX/community/documentation",context:["logged-in","CMS","HAX","*"]}),m.defineOption({title:"HAX Teaching Excellence",icon:"hax:hax2022",tags:["Ontology","community","pedagogy","documentation","help"],value:{target:this,method:"_openExternalLink",args:["https://haxtheweb.org/ontology"]},eventName:"super-daemon-element-method",path:"HAX/community/pedagogy",context:["logged-in","CMS","HAX","*"]}),m.defineOption({title:"Bug / issue",icon:"mdi-social:github-circle",tags:["Bug report","github","git","community","issue queue"],value:{target:this,method:"_haxStoreContribute",args:["bug","POP,bug"]},eventName:"super-daemon-element-method",path:"HAX/community/contribute",context:["logged-in","CMS","HAX","*"]}),m.defineOption({title:"Idea / Feature request",icon:"mdi-social:github-circle",tags:["Feature request","idea","github","git","community","issue queue"],value:{target:this,method:"_haxStoreContribute",args:["feature","POP,enhancement"]},context:["logged-in","CMS","HAX","*"],eventName:"super-daemon-element-method",path:"HAX/community/contribute"}),this.windowControllers=new AbortController,this.__tour=h,this.__tour.registerNewTour({key:"hax",name:"HAX top menu",style:"\n      simple-popover-manager::part(simple-popover) {\n        max-width: 250px;\n        font-family: sans-serif;\n      }\n      simple-popover-manager button {\n        font-family: sans-serif;\n        font-size: 12px;\n        margin: 0px 2px;\n        color: var(--simple-colors-default-theme-grey-12);\n      }\n      simple-popover-manager p {\n        font-family: sans-serif;\n        padding: 0;\n        margin: 0;\n        width: 250px;\n        font-size: 10px;\n        line-height: 20px;\n      }\n      simple-popover-manager h1 {\n        font-family: sans-serif;\n        margin: 0;\n        font-size: 12px;\n        width: 250px;\n        padding: 0;\n      }\n      simple-popover-manager::part(simple-popover-body),\n      simple-popover-manager::part(simple-popover-heading) {\n        color: black;\n        background-color: white;\n        font-family: sans-serif;\n      }\n      body.dark-mode simple-popover-manager::part(simple-popover-body),\n      body.dark-mode simple-popover-manager::part(simple-popover-heading) {\n        color: white;\n        background-color: black;\n        font-family: sans-serif;\n      }\n      body.dark-mode simple-popover-manager simple-icon-button-lite {\n        color: white;\n        background-color: black;\n        font-family: sans-serif;\n      }\n      "}),this.tourName="hax",i((()=>{const e=a(r.activeItem);e&&e.title&&r.setPageTitle(e.title)})),i((()=>{this.siteReady=a(r.siteReady)})),i((()=>{const e=a(r.badDevice);!1===e?import("../rpg-character/rpg-character.js"):!0===e&&globalThis.document.body.classList.add("bad-device")})),this.userMenuOpen=!1,this.courses=[],this.activeItem={},this.phrases={new:["What's ya name?","HAX to the moon","Welcome to the Jungle","We like to party","Build something awesome","Everything is awesome!","Everything is cool","When you're part of the team","When you're living our dream","Welcome to the up-side-down"],return:["Welcome back, take 2?","That wasn't very long","Stranger thiings have happened","Student driven platform","Faculty centered platform","Instructional designer influenced platform","Free, Open, Community driven","One brick at a time..","Sup?","You again? Awesome!","Let's do this","There can only be one ring...","There is another","Fancy that, I love HAX and you show up"]},this.isNewUser=null,this.basePath="/",this.searchTerm="",this.appMode="",this.soundIcon="",this.store=r,this.sound=new Audio,i((()=>{this.isNewUser=a(r.isNewUser),this.isNewUser&&"create"!==a(r.appMode)&&"404"!==a(r.appMode)&&(r.appMode="create",setTimeout((()=>{r.createSiteSteps=!0}),0))})),i((()=>{this.userName=a(r.user.name)})),i((()=>{this.appMode=a(r.appMode)})),i((()=>{this.searchTerm=a(r.searchTerm)})),i((async()=>{const e=a(r.location);e&&e.route&&(e.route.step?(r.appMode="create",r.createSiteSteps=!0):e.route.slug?(this.reset(),setTimeout((()=>{globalThis.location=e.route.slug}),0)):"404"===e.route.name?(r.createSiteSteps=!1,r.appMode="404",setTimeout((()=>{r.toast("the page miss.. it burns!!!",3e3,{fire:!0,walking:!0})}),500)):"home"!==e.route.name&&"search"!==e.route.name||(r.appMode="home",r.createSiteSteps=!1))})),i((()=>{r.routes.length>0&&null===r.location&&(r.location=a(r.routes[0]))})),null===s("app-hax-darkMode",null)&&(r.darkMode=globalThis.matchMedia("(prefers-color-scheme: dark)").matches),i((()=>{n("app-hax-darkMode",a(r.darkMode)),requestAnimationFrame((()=>{a(r.darkMode)?(globalThis.document.body.classList.add("dark-mode"),r.toast("I'm ascared of the dark",2e3,{fire:!0}),this.dark=!0,m.dark=!0,m.toastInstance.darkMode=!0):(globalThis.document.body.classList.remove("dark-mode"),r.toast("Sunny day it is",2e3,{hat:"random"}),this.dark=!1,m.dark=!1,m.toastInstance.darkMode=!1)}))})),i((()=>{const e=a(r.appMode);e&&(globalThis.document.body.classList.remove("app-hax-search"),globalThis.document.body.classList.remove("app-hax-create"),globalThis.document.body.classList.remove("app-hax-404"),globalThis.document.body.classList.remove("app-hax-home"),globalThis.document.body.classList.add(`app-hax-${e}`))}))}static get properties(){return{...super.properties,unlockComingSoon:{type:Boolean},unlockTerrible:{type:Boolean},courses:{type:Array},userName:{type:String},activeItem:{type:Object},soundIcon:{type:String},searchTerm:{type:String},appMode:{type:String},isNewUser:{type:Boolean},phrases:{type:Object},userMenuOpen:{type:Boolean},siteReady:{type:Boolean},basePath:{type:String,attribute:"base-path"},token:{type:String}}}reset(e=!1,t=!1){try{globalThis.localStorage.removeItem("app-hax-step"),globalThis.localStorage.removeItem("app-hax-site"),t&&(globalThis.localStorage.removeItem("jwt"),r.jwt=null,globalThis.dispatchEvent(new CustomEvent("jwt-login-logout",{composed:!0,bubbles:!0,cancelable:!1,detail:!0}))),e&&(document.querySelector("base")?globalThis.location=globalThis.document.querySelector("base").href:globalThis.location.reload())}catch(e){}}fireTerrible(){this.unlockTerrible=!0,r.appEl.playSound("coin").then((()=>{r.appEl.playSound("coin2").then((()=>{r.appEl.playSound("success").then((()=>{m.merlinSpeak("Enjoy these early 2000s table based layouts. May they remind you how never to web, again.")}))}))}))}fireUnlocked(){this.unlockComingSoon=!0,r.appEl.playSound("coin").then((()=>{r.appEl.playSound("coin2").then((()=>{r.appEl.playSound("success").then((()=>{m.merlinSpeak("Unbelievable! You, (Subject Name), must be the pride of (Subject Hometown). Enjoy all locked features as a boon!")}))}))}))}logout(){globalThis.dispatchEvent(new CustomEvent("jwt-login-logout",{composed:!0,bubbles:!0,cancelable:!1,detail:!0})),this.closeMenu(),this.__logoutUserAction=!0}_jwtLoggedIn(e){!1===e.detail&&this.__logoutUserAction&&(this.__logoutUserAction=!1,this.__loginModalOpen=!1,setTimeout((()=>{this.reset(!0)}),100))}login(){this.__loginModalOpen||(this.__loginModalOpen=!0,import("./lib/v2/app-hax-site-login.js").then((()=>{const e=globalThis.document.createElement("app-hax-site-login");if(this.querySelector('[slot="externalproviders"]')){const t=this.querySelector('[slot="externalproviders"]').cloneNode(!0);e.appendChild(t)}import("../simple-modal/simple-modal.js").then((()=>{setTimeout((()=>{this.dispatchEvent(new CustomEvent("simple-modal-show",{bubbles:!0,cancelable:!0,composed:!0,detail:{title:"Login to HAX",elements:{content:e},modal:!0,styles:{"--simple-modal-titlebar-background":"var(--ddd-theme-default-nittanyNavy, #001e44)","--simple-modal-titlebar-color":"var(--ddd-theme-default-white, white)","--simple-modal-width":"90vw","--simple-modal-max-width":"var(--ddd-spacing-32, 480px)","--simple-modal-min-width":"300px","--simple-modal-z-index":"1000","--simple-modal-height":"auto","--simple-modal-min-height":"300px","--simple-modal-max-height":"80vh","--simple-modal-titlebar-height":"80px","--simple-modal-border-radius":"var(--ddd-radius-md, 8px)","--simple-modal-background":"var(--ddd-theme-default-white, white)","--simple-modal-box-shadow":"var(--ddd-boxShadow-xl)"}}}))}),0)}))})))}static get styles(){return[super.styles,e`
        :host {
          display: block;
          --app-hax-background-color-active: var(--app-hax-accent-color);
        }
        #home {
          display: inline-flex;
        }
        #wt {
          border: solid 1px
            var(--simple-colors-default-theme-accent-12, var(--accent-color));
        }
        simple-toolbar-button {
          min-width: 48px;
          margin: 0;
          --simple-toolbar-border-color: #dddddddd;
          height: 48px;
          --simple-toolbar-button-disabled-opacity: 0.3;
          --simple-toolbar-button-padding: 3px 6px;
          --simple-toolbar-border-radius: 0;
        }
        simple-toolbar-button:hover,
        simple-toolbar-button:active,
        simple-toolbar-button:focus {
          background-color: var(--hax-ui-background-color-accent);
          color: var(--hax-ui-color);
        }
        simple-toolbar-button:hover,
        simple-toolbar-button:active,
        simple-toolbar-button:focus {
          --simple-toolbar-border-color: var(--hax-ui-color-accent);
        }
        .wired-button-label {
          clip: rect(0 0 0 0);
          clip-path: inset(50%);
          height: 1px;
          overflow: hidden;
          position: absolute;
          white-space: nowrap;
          width: 1px;
        }
        .topbar-character {
          cursor: pointer;
          display: inline-block;
          border: none;
          border-radius: 0px;
          padding: 0 8px;
          margin: 0 0 0 16px;
          background-color: transparent;
          height: 48px;
          max-width: 160px;
        }
        .characterbtn-name {
          color: var(--simple-colors-default-theme-grey-12);
          margin-left: 8px;
          font-size: 12px;
          vertical-align: bottom;
          line-height: 48px;
          overflow: hidden;
          text-overflow: ellipsis;
          height: 48px;
          word-break: break-all;
        }
        .topbar-character:hover,
        .topbar-character:focus {
          background-color: var(--simple-colors-default-theme-light-blue-4);
          outline: var(--haxcms-color) solid 3px;
          outline-offset: -3px;
          height: 48px;
        }
        .topbar-character rpg-character {
          margin: -4px -14px 0px -10px;
          height: 52px;
          width: 64px;
          display: inline-block;
        }
        .content {
          text-align: center;
          margin-top: 24px;
        }
        .four04-character {
          margin-top: 16px;
        }
        .start-journey {
          display: flex;
          justify-content: center;
        }
        app-hax-site-button {
          max-width: 60vw;
          justify-content: center;
        }
        app-hax-top-bar {
          top: 0;
          z-index: 1000;
          right: 0;
          left: 0;
          position: fixed;
        }
        @media (max-width: 640px) {
          simple-tooltip {
            --simple-tooltip-font-size: 10px;
          }
        }
        app-hax-label {
          animation: 0.6s ease-in-out 0s scrollin;
          -webkit-animation: 0.6s ease-in-out 0s scrollin;
          display: flex;
          align-self: flex-start;
          margin: var(--ddd-spacing-4) auto;
        }
        app-hax-label h1 {
          font-weight: normal;
          font-size: var(--ddd-font-size-m);
          margin: 0;
          padding: 0;
        }
        @keyframes scrollin {
          from {
            margin-top: -240px;
            margin-bottom: 240px;
          }
          to {
            margin-top: 0;
            margin-bottom: 0;
          }
        }
        .haxLogo {
          --simple-icon-height: 40px;
          --simple-icon-width: 40px;
          margin: 4px;
          color: var(--simple-colors-default-theme-grey-12);
          cursor: pointer;
        }
        .soundToggle {
          margin-right: 16px;
          position: relative;
          display: inline-flex;
          vertical-align: top;
        }

        .soundToggle img {
          width: 24px;
          height: 24px;
        }

        app-hax-search-bar {
          display: inline-flex;
        }
        main {
          padding-top: var(--ddd-spacing-14);
          max-width: 100%;
          margin: 0 auto;
          box-sizing: border-box;
        }
        @media (max-width: 900px) {
          .characterbtn-name {
            display: none;
          }
          main {
            padding-top: 64px;
          }
        }
        app-hax-user-menu {
          z-index: 1003;
          position: inherit;
        }
        .logout::part(menu-button) {
          background-image: url("${o(u)}");
          background-repeat: no-repeat;
          background-position: center;
          text-align: center;
          background-size: cover;
          border-top: 0px;
          border-bottom: 0px;
          padding: 10px;
        }
        app-hax-user-menu app-hax-user-menu-button::part(menu-button) {
          font-size: 12px;
        }
        #helpbtn {
          --simple-icon-height: 50px;
          --simple-icon-width: 50px;
          right: 200px;
          top: 100px;
          padding: 4px;
          background-color: var(--simple-colors-default-theme-grey-1);
          border-radius: 50%;
          position: absolute;
          color: var(--simple-colors-default-theme-grey-12);
          border: var(--simple-colors-default-theme-grey-12) 4px solid;
          cursor: pointer;
        }
        @media (max-width: 800px) {
          app-hax-site-button {
            width: 320px;
            max-width: 60vw;
            --app-hax-site-button-font-size: 16px;
          }
          #helpbtn {
            --simple-icon-height: 40px;
            --simple-icon-width: 40px;
            right: 8px;
            top: 64px;
            padding: 2px;
            border: var(--simple-colors-default-theme-grey-12) 2px solid;
          }
        }
        @media (prefers-reduced-motion: reduce) {
          app-hax-label {
            animation: none;
            -webkit-animation: none;
          }
        }
        @media (max-width: 640px) {
          .content {
            margin-top: 4px;
          }

          app-hax-site-button {
            --app-hax-site-button-font-size: 12px;
          }
        }
        @media (max-height: 500px) {
          app-hax-label h1 {
            font-family: monospace;
            font-weight: normal;
            font-size: 4vw;
            margin: 0;
            padding: 0;
          }
        }
      `]}helpClick(){r.appEl.playSound("coin2"),this.__tour.startTour("hax")}updated(e){super.updated&&super.updated(e),e.forEach(((e,t)=>{["basePath"].includes(t)&&this[t]&&(r.AppHaxAPI[t]=this[t]),["token"].includes(t)&&this[t]&&(r[t]=this[t])}))}firstUpdated(e){super.firstUpdated&&super.firstUpdated(e),r.AppHaxAPI&&this.basePath&&(r.AppHaxAPI.basePath=this.basePath),import("./lib/v2/app-hax-steps.js"),import("./lib/v2/app-hax-site-button.js"),import("../../wired-elements/lib/wired-button.js"),import("./lib/v2/app-hax-toast.js"),import("./lib/v2/app-hax-wired-toggle.js"),import("./lib/v2/app-hax-search-bar.js"),import("./lib/v2/app-hax-search-results.js"),import("./lib/v2/app-hax-user-menu.js"),import("./lib/v2/app-hax-user-menu-button.js"),this.dispatchEvent(new CustomEvent("app-hax-loaded",{composed:!0,bubbles:!0,cancelable:!1,detail:!0})),r.appReady=!0,i((()=>{a(r.appReady)?globalThis.document.body.classList.add("app-loaded"):globalThis.document.body.classList.remove("app-loaded")})),r.appEl=this,i((()=>{r.activeItem&&(this.activeItem=a(r.activeItem))})),i((()=>{this.soundIcon=a(r.soundStatus)?new URL("./lib/assets/images/FullVolume.svg",import.meta.url).href:new URL("./lib/assets/images/Silence.svg",import.meta.url).href,a(r.soundStatus)?r.toast("You can hear me now? Good.",2e3,{hat:"random"}):r.toast("Sound off.. hey! HELLO!?!",2e3,{fire:!0})})),i((async()=>{if(a(r.appReady)&&r.AppHaxAPI&&a(r.isLoggedIn)&&a(r.appSettings)&&a(r.refreshSiteList)){const e=await d.makeCall("getSitesList");r.manifest=e.data}else a(r.appReady)&&!a(r.isLoggedIn)&&this.login()}))}toggleMenu(){this.userMenuOpen=!this.userMenuOpen,r.appEl.playSound("click")}closeMenu(){this.userMenuOpen=!1}openMerlin(){r.appEl.playSound("click"),m.open()}render(){return t`<app-hax-router></app-hax-router>
      <header>
        <app-hax-top-bar>
          <slot name="app-header-pre" slot="left"></slot>
          <a
            id="home"
            title="${this.t.home}"
            href="home"
            tabindex="-1"
            slot="left"
          >
            <simple-icon-lite
              id="hlogo"
              icon="hax:hax2022"
              tabindex="0"
              class="haxLogo"
              title="${this.t.home}"
              data-simple-tour-stop
              data-stop-title="data-label"
              data-label="${this.t.home}"
            >
              <div data-stop-content slot="tour" style="display:none;">
                You can come back to this home screen whenever you click this!
              </div>
            </simple-icon-lite>
          </a>
          <simple-tooltip for="hlogo" position="right" slot="left"
            >${this.t.home}</simple-tooltip
          >
          <simple-toolbar-button
            icon="hax:wizard-hat"
            label="${this.t.merlin}"
            voice-command="${this.t.merlin}"
            icon-position="left"
            id="merlin"
            @click="${this.openMerlin}"
            slot="center"
            data-event="super-daemon"
            show-text-label
          ></simple-toolbar-button>
          <wired-button
            elevation="1"
            slot="right"
            class="soundToggle"
            id="soundtb"
            @click="${soundToggle}"
            data-simple-tour-stop
            data-stop-title="data-label"
            data-label="Sound"
            aria-label="Toggle sound effects ${this.soundIcon.includes("FullVolume")?"off":"on"}"
            aria-pressed="${this.soundIcon.includes("FullVolume")}"
          >
            <span class="wired-button-label">Toggle sound effects</span>
            <simple-icon-lite
              src="${this.soundIcon}"
              loading="lazy"
              decoding="async"
              alt="${this.soundIcon.includes("FullVolume")?"Sound enabled":"Sound disabled"}"
            >
            </simple-icon-lite>
            <div slot="tour" data-stop-content>
              Not a fan of the (awesome) sound effects? You can mute them if you
              prefer.
            </div>
          </wired-button>
          <simple-tooltip for="soundtb" position="bottom" slot="right"
            >Toggle sound</simple-tooltip
          >
          <app-hax-wired-toggle id="wt" slot="right"></app-hax-wired-toggle>
          <simple-tooltip for="wt" position="bottom" slot="right"
            >Toggle dark mode</simple-tooltip
          >
          <app-hax-user-menu
            slot="right"
            id="user-menu"
            ?is-open="${this.userMenuOpen}"
            data-simple-tour-stop
            data-stop-title="data-label"
            data-label="Menu"
          >
            <div slot="tour" data-stop-content>
              You want to log out and be someone else? Create a new site? Click
              your character. Your character is unique to you!
            </div>
            <button
              @click="${this.toggleMenu}"
              class="topbar-character"
              slot="menuButton"
              id="tbchar"
              title="System menu"
              aria-label="User menu for ${this.userName}"
              aria-expanded="${this.userMenuOpen}"
              aria-haspopup="menu"
            >
              <rpg-character
                seed="${this.userName}"
                width="68"
                height="68"
                hat="${this.userMenuOpen?"edit":"none"}"
                alt="Avatar for ${this.userName}"
                role="img"
              ></rpg-character>
              <span class="characterbtn-name" aria-hidden="true"
                >${this.userName}</span
              >
            </button>
            <a
              slot="main-menu"
              title="${this.t.listMySites}"
              href="home"
              tabindex="-1"
            >
              <app-hax-user-menu-button
                icon="hax:hax2022"
                label="${this.t.listMySites}"
                part="listMySites"
              ></app-hax-user-menu-button>
            </a>
            <!-- <app-hax-user-menu-button
              slot="main-menu"
              icon="face"
              label="Account Info"
            ></app-hax-user-menu-button> -->
            <app-hax-user-menu-button
              slot="post-menu"
              class="logout"
              label="Logout"
              @click=${this.logout}
            ></app-hax-user-menu-button>
          </app-hax-user-menu>
        </app-hax-top-bar>
      </header>
      <main @click="${this.closeMenu}">
        <confetti-container id="confetti">
          <div class="label">
            <app-hax-label heading-level="1">
              HAX Site Dashboard
              <div slot="subtitle">Let's build something awesome!</div>
            </app-hax-label>
          </div>
          <section class="content">
            <div class="start-journey">
              <app-hax-use-case-filter></app-hax-use-case-filter>
            </div>
          </section>
        </confetti-container>
      </main>`}siteReadyToGo(e){import("../multiple-choice/lib/confetti-container.js").then((e=>{setTimeout((()=>{this.shadowRoot.querySelector("#confetti").setAttribute("popped","")}),0)})),e.detail&&(r.siteReady=!0)}}customElements.define(AppHax.tag,AppHax),globalThis.AppHax=globalThis.AppHax||{},globalThis.AppHax.requestAvailability=()=>(!globalThis.AppHax.instance&&globalThis.document&&(globalThis.AppHax.instance=globalThis.document.querySelector(AppHax.tag)),globalThis.AppHax.instance);
