// @lrnwebcomponents/lrn-button/lrn-button.js
define(["exports","require","../../@polymer/polymer/polymer-element.js","../../@polymer/polymer/lib/utils/render-status.js","../../@polymer/iron-icons/iron-icons.js","../../@polymer/iron-icon/iron-icon.js","../materializecss-styles/materializecss-styles.js"],function(_exports,_require,_polymerElement,_renderStatus,_ironIcons,_ironIcon,_materializecssStyles){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.LrnButton=void 0,_require=babelHelpers.interopRequireWildcard(_require);class LrnButton extends _polymerElement.PolymerElement{static get template(){return _polymerElement.html`
      <style>
        :host {
          display: block;
          @apply --paper-font-common-base;
          @apply --paper-button;
          --lrnsys-button-height: 48px;
        }

        :host(.center) {
          text-align: center;
        }

        a {
          text-decoration: none;
          display: block;
          color: #000000;
        }

        paper-button {
          transition: 0.3s;
          margin: 0;
          max-width: 50%;
          height: inherit;
          -webkit-justify-content: flex-start;
          justify-content: flex-start;
          align-items: center;
          border-radius: unset;
        }

        paper-button iron-icon {
          height: var(--lrnsys-button-height);
          margin: 0 12px;
        }

        paper-button div.inner {
          height: var(--lrnsys-button-height);
          line-height: var(--lrnsys-button-height);
          padding: 0 12px;
        }

        paper-button span.label {
          height: var(--lrnsys-button-height);
          line-height: var(--lrnsys-button-height);
        }

        .no-margin {
          margin: 0 !important;
        }

        .no-right-padding {
          padding-right: 0 !important;
        }

        .no-left-padding {
          padding-left: 0 !important;
        }

        .center {
          text-align: center;
          margin: 0 auto;
        }
      </style>
      <style include="materializecss-styles-colors"></style>
      <a
        tabindex="-1"
        id="lrnsys-button-link"
        href$="[[showHref]]"
        data-prefetch-hover$="[[prefetch]]"
        target$="[[target]]"
      >
        <paper-button
          id="button"
          raised="[[raised]]"
          class$="[[class]] [[color]] [[textColor]]"
          disabled$="[[disabled]]"
        >
          <div class$="inner [[innerClass]]">
            <iron-icon
              icon$="[[icon]]"
              id="icon"
              class$="[[iconClass]]"
              hidden$="[[!icon]]"
            ></iron-icon>
            <span class="label" hidden$="[[!label]]">
              [[label]]
            </span>
            <slot></slot>
          </div>
        </paper-button>
      </a>
      <paper-tooltip for="lrnsys-button-link" animation-delay="0"
        >[[alt]]</paper-tooltip
      >
    `}static get properties(){let props={href:{type:String,value:"#"},showHref:{type:String,value:!1},raised:{type:Boolean},label:{type:String,value:""},target:{type:String,value:""},icon:{type:String,value:!1},hoverClass:{type:String},iconClass:{type:String},innerClass:{type:String},color:{type:String},textColor:{type:String},prefetch:{type:String},alt:{type:String},disabled:{type:Boolean,value:!1},focusState:{type:Boolean,value:!1}};return super.properties&&(props=Object.assign(props,super.properties)),props}static get tag(){return"lrn-button"}constructor(){super(),new Promise((res,rej)=>_require.default(["../../@polymer/paper-button/paper-button.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/paper-tooltip/paper-tooltip.js"],res,rej))}connectedCallback(){super.connectedCallback(),(0,_renderStatus.afterNextRender)(this,function(){this.addEventListener("mousedown",this.tapEventOn),this.addEventListener("mouseover",this.tapEventOn),this.addEventListener("mouseout",this.tapEventOff),this.$.button.addEventListener("focused-changed",this.focusToggle)})}disconnectedCallback(){this.removeEventListener("mousedown",this.tapEventOn),this.removeEventListener("mouseover",this.tapEventOn),this.removeEventListener("mouseout",this.tapEventOff),this.$.button.removeEventListener("focused-changed",this.focusToggle),super.disconnectedCallback()}ready(){super.ready(),this.disabled||(this.showHref=this.href)}tapEventOn(e){let root=this;void 0===root.hoverClass||root.disabled||root.hoverClass.split(" ").forEach(function(item,index){""!=item&&(root.$.button.classList.add(item),-1!=item.indexOf("-")&&root.$.icon.classList.add(item))})}tapEventOff(e){let root=this;void 0===root.hoverClass||root.disabled||root.hoverClass.split(" ").forEach(function(item,index){""!=item&&(root.$.button.classList.remove(item),-1!=item.indexOf("-")&&root.$.icon.classList.remove(item))})}focusToggle(e){let root=this;(this.dispatchEvent(new CustomEvent("focus-changed",{bubbles:!0,composed:!0,detail:{focus:root.focusState}})),void 0===root.hoverClass||root.disabled)||root.hoverClass.split(" ").forEach(function(item,index){""!=item&&(root.focusState?(root.$.button.classList.add(item),-1!=item.indexOf("-")&&root.$.icon.classList.add(item)):(root.$.button.classList.remove(item),-1!=item.indexOf("-")&&root.$.icon.classList.remove(item)))});root.focusState=!root.focusState}}_exports.LrnButton=LrnButton,window.customElements.define(LrnButton.tag,LrnButton)});
