// @lrnwebcomponents/hax-text-editor/lib/hax-text-editor-oer-schema.js
define(["exports", "../../../@polymer/polymer/polymer-element.js", "../../rich-text-editor/lib/buttons/rich-text-editor-prompt-button.js", "../../oer-schema/oer-schema.js"], function (_exports, _polymerElement, _richTextEditorPromptButton, _oerSchema) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HaxTextEditorOerSchema = void 0;

  /**
   * Copyright 2019 Penn State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `hax-text-editor-oer-schema`
   * @customElement hax-text-editor-oer-schema
   * `a oer-schema button for hax text editor (custom buttons can extend this)`
   *
   * @microcopy - language worth noting:
   *  -
   *
  
   * @polymer
   */
  class HaxTextEditorOerSchema extends _richTextEditorPromptButton.RichTextEditorPromptButton {
    constructor() {
      super();
      let element = _oerSchema.OerSchemaElement;
      this.fields = element.haxProperties.settings.configure;
      this.tag = element.tag;
      this.icon = element.haxProperties.gizmo.icon;
      this.label = element.haxProperties.gizmo.title;
      this.toggles = true;
      this.value = {
        "": null
      };
    } // properties available to the custom element for data binding


    static get properties() {
      return {};
    }
    /**
     * Store the tag name to make it easier to obtain directly.
     */


    static get tag() {
      return "hax-text-editor-oer-schema";
    }

  }

  _exports.HaxTextEditorOerSchema = HaxTextEditorOerSchema;
  window.customElements.define(HaxTextEditorOerSchema.tag, HaxTextEditorOerSchema);
});
