// @lrnwebcomponents/hax-text-editor/lib/hax-text-editor-vocab.js
define(["exports", "../../../@polymer/polymer/polymer-element.js", "../../rich-text-editor/lib/buttons/rich-text-editor-prompt-button.js", "../../lrn-vocab/lrn-vocab.js"], function (_exports, _polymerElement, _richTextEditorPromptButton, _lrnVocab) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HaxTextEditorVocab = void 0;

  /**
   * Copyright 2019 Penn State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `hax-text-editor-vocab`
   * @customElement hax-text-editor-vocab
   * `a lrn-vocab button for hax text editor (custom buttons can extend this)`
   *
   * @microcopy - language worth noting:
   *  -
   *
  
   * @polymer
   */
  class HaxTextEditorVocab extends _richTextEditorPromptButton.RichTextEditorPromptButton {
    constructor() {
      super();
      this.inlineWidget = true;
      let element = _lrnVocab.LrnVocab;
      this.tag = element.tag;
      this.icon = element.haxProperties.gizmo.icon;
      this.label = element.haxProperties.gizmo.title;
      this.fields = element.haxProperties.settings.configure;
      this.toggles = true;
      this.value = {
        "": null
      };
    } // properties available to the custom element for data binding


    static get properties() {
      return {};
    }
    /**
     * Store the tag name to make it easier to obtain directly.
     */


    static get tag() {
      return "hax-text-editor-vocab";
    }

  }

  _exports.HaxTextEditorVocab = HaxTextEditorVocab;
  window.customElements.define(HaxTextEditorVocab.tag, HaxTextEditorVocab);
});
