// @lrnwebcomponents/paper-input-flagged/paper-input-flagged.js
define(["exports","require","../../lit-element/lit-element.js"],(function(e,t,s){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.PaperInputFlagged=void 0,t=babelHelpers.interopRequireWildcard(t);
/**
   * Copyright 2018 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class PaperInputFlagged extends s.LitElement{static get styles(){return[s.css`
        :host {
          display: block;
        }
        iron-icon {
          transition: 0.6s all ease-in;
          width: 24px;
          height: 24px;
          margin-right: 4px;
        }
        simple-tooltip {
          --simple-tooltip-delay-in: 100;
          font-size: 11px;
        }
        #icon {
          color: var(--paper-grey-400);
          background-color: transparent;
        }
        :host([status="info"]) #icon {
          color: var(--paper-green-400);
        }
        :host([status="notice"]) #icon {
          color: var(--paper-grey-400);
        }
        :host([status="warning"]) #icon {
          color: var(--paper-yellow-700);
        }
        :host([status="error"]) #icon {
          color: var(--paper-red-900);
        }
        .element-invisible {
          position: absolute !important;
          clip: rect(1px, 1px, 1px, 1px);
          overflow: hidden;
          height: 1px;
        }
      `]}constructor(){super(),this.disabled=!1,this.label="",this.value="",this.inputSuccess={message:"You passed our simple accessibility checks.",status:"info"},this.flaggedInput=[{match:null,message:"Alt data is required for everything except decoration images.",status:"notice"},{match:"image",message:"Screenreaders will say the word image, don't put it in the descriptive text",status:"error"},{match:"photo",message:"Screenreaders will say the word image, don't put photo in the descriptive text",status:"error"},{match:"picture",message:"Screenreaders will say the word image, don't put picture in the descriptive text",status:"error"},{match:3,message:"Description not effective enough. This should be at least a sentance about what the image is.",status:"error"},{match:10,message:"Make sure your alt text is descriptive enough for those that can't see the media.",status:"warning"}],new Promise((e,s)=>t.default(["../../@polymer/paper-input/paper-input.js"],e,s)),new Promise((e,s)=>t.default(["../simple-tooltip/simple-tooltip.js"],e,s)),new Promise((e,s)=>t.default(["../../@polymer/iron-icons/iron-icons.js"],e,s)),new Promise((e,s)=>t.default(["../../@polymer/iron-icon/iron-icon.js"],e,s))}updated(e){e.forEach((e,t)=>{if(["value"].includes(t)){let e=`${t.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g,"$1-$2").toLowerCase()}-changed`;this.dispatchEvent(new CustomEvent(e,{detail:{value:this[t]}}))}"status"==t&&(this.icon=this._iconFromStatus(this.status)),"flaggedInput"!=t&&"value"!=t||(this.status=this.testStatus(this.flaggedInput,this.value))})}valueEvent(e){this.value=e.detail.value}render(){return s.html`
      <paper-input
        label="${this.label}"
        value="${this.value}"
        @value-changed="${this.valueEvent}"
        ?char-counter="${this.charCounter}"
        ?disabled="${this.disabled}"
        minlength="${this.minlength}"
        maxlength="${this.maxlength}"
      >
        <iron-icon id="icon" icon="${this.icon}" slot="prefix"></iron-icon>
      </paper-input>
      <div class="element-invisible">${this.__activeMessage}</div>
      <simple-tooltip
        for="icon"
        position="top"
        offset="20"
        fit-to-visible-bounds
      >
        ${this.__activeMessage}
      </simple-tooltip>
    `}static get tag(){return"paper-input-flagged"}static get properties(){return{label:{type:String},disabled:{type:Boolean},icon:{type:String},maxlength:{type:Number},minlength:{type:Number},status:{type:String,reflect:!0},value:{type:String},flaggedInput:{type:Array,attribute:"flagged-input"},inputSuccess:{type:Object,attribute:"input-success"},__activeMessage:{type:String}}}testStatus(e,t){for(var s in e){if(null===e[s].match&&(""===t||null===t))return this.__activeMessage=e[s].message,e[s].status;if(!isNaN(e[s].match)&&t.split(/\s+/g).length<parseInt(e[s].match))return this.__activeMessage=e[s].message,e[s].status;if(t.toLowerCase().includes(e[s].match))return this.__activeMessage=e[s].message,e[s].status}return this.__activeMessage=this.inputSuccess.message,this.inputSuccess.status}_iconFromStatus(e){switch(e){case"error":return"icons:error";case"warning":case"notice":return"icons:warning";case"info":return"icons:info-outline";default:return"icons:info"}}}e.PaperInputFlagged=PaperInputFlagged,window.customElements.define(PaperInputFlagged.tag,PaperInputFlagged)}));
