// web-social-share/dist/esm/web-social-share.entry.js
define(["exports", "./index-bef2383e.js"], function (_exports, _indexBef2383e) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.web_social_share = void 0;

  /**
   * Source: https://github.com/720kb/angular-socialshare/blob/master/dist/angular-socialshare.js
   */
  var WebSocialShareFacebook = /*#__PURE__*/function () {
    function WebSocialShareFacebook() {
      babelHelpers.classCallCheck(this, WebSocialShareFacebook);
    }

    babelHelpers.createClass(WebSocialShareFacebook, null, [{
      key: "share",
      value: function () {
        var _share = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(attrs) {
          var urlString;
          return regeneratorRuntime.wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  if (attrs.socialShareType && attrs.socialShareType === 'feed') {
                    // if user specifies that they want to use the Facebook feed dialog
                    //(https://developers.facebook.com/docs/sharing/reference/feed-dialog/v2.4)
                    urlString = 'https://www.facebook.com/dialog/feed?';

                    if (attrs.socialShareVia) {
                      urlString += '&app_id=' + encodeURIComponent(attrs.socialShareVia);
                    }

                    if (attrs.socialShareRedirectUri) {
                      urlString += '&redirect_uri=' + encodeURIComponent(attrs.socialShareRedirectUri);
                    }

                    if (attrs.socialShareUrl) {
                      urlString += '&link=' + encodeURIComponent(attrs.socialShareUrl);
                    }

                    if (attrs.socialShareTo) {
                      urlString += '&to=' + encodeURIComponent(attrs.socialShareTo);
                    }

                    if (attrs.socialShareDisplay) {
                      urlString += '&display=' + encodeURIComponent(attrs.socialShareDisplay);
                    }

                    if (attrs.socialShareRef) {
                      urlString += '&ref=' + encodeURIComponent(attrs.socialShareRef);
                    }

                    if (attrs.socialShareFrom) {
                      urlString += '&from=' + encodeURIComponent(attrs.socialShareFrom);
                    }

                    if (attrs.socialShareSource) {
                      urlString += '&source=' + encodeURIComponent(attrs.socialShareSource);
                    }

                    window.open(urlString, 'Facebook', 'toolbar=0,status=0,resizable=yes,width=' + attrs.socialSharePopupWidth + ',height=' + attrs.socialSharePopupHeight + ',top=' + (window.innerHeight - attrs.socialSharePopupHeight) / 2 + ',left=' + (window.innerWidth - attrs.socialSharePopupWidth) / 2);
                  } else if (attrs.socialShareType && attrs.socialShareType === 'share') {
                    // if user specifies that they want to use the Facebook share dialog
                    //(https://developers.facebook.com/docs/sharing/reference/share-dialog)
                    urlString = 'https://www.facebook.com/dialog/share?';

                    if (attrs.socialShareVia) {
                      urlString += '&app_id=' + encodeURIComponent(attrs.socialShareVia);
                    }

                    if (attrs.socialShareRedirectUri) {
                      urlString += '&redirect_uri=' + encodeURIComponent(attrs.socialShareRedirectUri);
                    }

                    if (attrs.socialShareUrl) {
                      urlString += '&href=' + encodeURIComponent(attrs.socialShareUrl);
                    }

                    if (attrs.socialShareQuote) {
                      urlString += '&quote=' + encodeURIComponent(attrs.socialShareQuote);
                    }

                    if (attrs.socialShareDisplay) {
                      urlString += '&display=' + encodeURIComponent(attrs.socialShareDisplay);
                    }

                    if (attrs.socialShareMobileiframe) {
                      urlString += '&mobile_iframe=' + encodeURIComponent(attrs.socialShareMobileiframe);
                    }

                    if (attrs.socialShareHashtags) {
                      urlString += '&hashtag=' + encodeURIComponent(attrs.socialShareHashtags);
                    }

                    window.open(urlString, 'Facebook', 'toolbar=0,status=0,resizable=yes,width=' + attrs.socialSharePopupWidth + ',height=' + attrs.socialSharePopupHeight + ',top=' + (window.innerHeight - attrs.socialSharePopupHeight) / 2 + ',left=' + (window.innerWidth - attrs.socialSharePopupWidth) / 2);
                  } else if (attrs.socialShareType && attrs.socialShareType === 'send') {
                    // if user specifies that they want to use the Facebook send dialog
                    //(https://developers.facebook.com/docs/sharing/reference/send-dialog)
                    urlString = 'https://www.facebook.com/dialog/send?';

                    if (attrs.socialShareVia) {
                      urlString += '&app_id=' + encodeURIComponent(attrs.socialShareVia);
                    }

                    if (attrs.socialShareRedirectUri) {
                      urlString += '&redirect_uri=' + encodeURIComponent(attrs.socialShareRedirectUri);
                    }

                    if (attrs.socialShareUrl) {
                      urlString += '&link=' + encodeURIComponent(attrs.socialShareUrl);
                    }

                    if (attrs.socialShareTo) {
                      urlString += '&to=' + encodeURIComponent(attrs.socialShareTo);
                    }

                    if (attrs.socialShareDisplay) {
                      urlString += '&display=' + encodeURIComponent(attrs.socialShareDisplay);
                    }

                    window.open(urlString, 'Facebook', 'toolbar=0,status=0,resizable=yes,width=' + attrs.socialSharePopupWidth + ',height=' + attrs.socialSharePopupHeight + ',top=' + (window.innerHeight - attrs.socialSharePopupHeight) / 2 + ',left=' + (window.innerWidth - attrs.socialSharePopupWidth) / 2);
                  } else {
                    //otherwise default to using sharer.php
                    window.open('https://www.facebook.com/sharer/sharer.php?u=' + encodeURIComponent(attrs.socialShareUrl || window.location.href), 'Facebook', 'toolbar=0,status=0,resizable=yes,width=' + attrs.socialSharePopupWidth + ',height=' + attrs.socialSharePopupHeight + ',top=' + (window.innerHeight - attrs.socialSharePopupHeight) / 2 + ',left=' + (window.innerWidth - attrs.socialSharePopupWidth) / 2);
                  }

                case 1:
                case "end":
                  return _context.stop();
              }
            }
          }, _callee);
        }));

        function share(_x) {
          return _share.apply(this, arguments);
        }

        return share;
      }()
    }]);
    return WebSocialShareFacebook;
  }();

  var WebSocialShareUtils = /*#__PURE__*/function () {
    function WebSocialShareUtils() {
      babelHelpers.classCallCheck(this, WebSocialShareUtils);
    }

    babelHelpers.createClass(WebSocialShareUtils, null, [{
      key: "isMobile",
      // Source: http://detectmobilebrowsers.com
      // Same implementation as in class DeckdeckgoUtils
      value: function isMobile() {
        if (!window) {
          return false;
        }

        var a = navigator.userAgent || navigator.vendor || window.opera;
        return /(android|bb\d+|meego).+mobile|avantgo|bada\/|blackberry|blazer|compal|elaine|fennec|hiptop|iemobile|ip(hone|od)|iris|kindle|lge |maemo|midp|mmp|mobile.+firefox|netfront|opera m(ob|in)i|palm( os)?|phone|p(ixi|re)\/|plucker|pocket|psp|series(4|6)0|symbian|treo|up\.(browser|link)|vodafone|wap|windows ce|xda|xiino/i.test(a) || /1207|6310|6590|3gso|4thp|50[1-6]i|770s|802s|a wa|abac|ac(er|oo|s\-)|ai(ko|rn)|al(av|ca|co)|amoi|an(ex|ny|yw)|aptu|ar(ch|go)|as(te|us)|attw|au(di|\-m|r |s )|avan|be(ck|ll|nq)|bi(lb|rd)|bl(ac|az)|br(e|v)w|bumb|bw\-(n|u)|c55\/|capi|ccwa|cdm\-|cell|chtm|cldc|cmd\-|co(mp|nd)|craw|da(it|ll|ng)|dbte|dc\-s|devi|dica|dmob|do(c|p)o|ds(12|\-d)|el(49|ai)|em(l2|ul)|er(ic|k0)|esl8|ez([4-7]0|os|wa|ze)|fetc|fly(\-|_)|g1 u|g560|gene|gf\-5|g\-mo|go(\.w|od)|gr(ad|un)|haie|hcit|hd\-(m|p|t)|hei\-|hi(pt|ta)|hp( i|ip)|hs\-c|ht(c(\-| |_|a|g|p|s|t)|tp)|hu(aw|tc)|i\-(20|go|ma)|i230|iac( |\-|\/)|ibro|idea|ig01|ikom|im1k|inno|ipaq|iris|ja(t|v)a|jbro|jemu|jigs|kddi|keji|kgt( |\/)|klon|kpt |kwc\-|kyo(c|k)|le(no|xi)|lg( g|\/(k|l|u)|50|54|\-[a-w])|libw|lynx|m1\-w|m3ga|m50\/|ma(te|ui|xo)|mc(01|21|ca)|m\-cr|me(rc|ri)|mi(o8|oa|ts)|mmef|mo(01|02|bi|de|do|t(\-| |o|v)|zz)|mt(50|p1|v )|mwbp|mywa|n10[0-2]|n20[2-3]|n30(0|2)|n50(0|2|5)|n7(0(0|1)|10)|ne((c|m)\-|on|tf|wf|wg|wt)|nok(6|i)|nzph|o2im|op(ti|wv)|oran|owg1|p800|pan(a|d|t)|pdxg|pg(13|\-([1-8]|c))|phil|pire|pl(ay|uc)|pn\-2|po(ck|rt|se)|prox|psio|pt\-g|qa\-a|qc(07|12|21|32|60|\-[2-7]|i\-)|qtek|r380|r600|raks|rim9|ro(ve|zo)|s55\/|sa(ge|ma|mm|ms|ny|va)|sc(01|h\-|oo|p\-)|sdk\/|se(c(\-|0|1)|47|mc|nd|ri)|sgh\-|shar|sie(\-|m)|sk\-0|sl(45|id)|sm(al|ar|b3|it|t5)|so(ft|ny)|sp(01|h\-|v\-|v )|sy(01|mb)|t2(18|50)|t6(00|10|18)|ta(gt|lk)|tcl\-|tdg\-|tel(i|m)|tim\-|t\-mo|to(pl|sh)|ts(70|m\-|m3|m5)|tx\-9|up(\.b|g1|si)|utst|v400|v750|veri|vi(rg|te)|vk(40|5[0-3]|\-v)|vm40|voda|vulc|vx(52|53|60|61|70|80|81|83|85|98)|w3c(\-| )|webc|whit|wi(g |nc|nw)|wmlb|wonu|x700|yas\-|your|zeto|zte\-/i.test(a.substr(0, 4));
      }
    }, {
      key: "staticOpenNewWindow",
      value: function staticOpenNewWindow(urlString) {
        if (window.self !== window.top) {
          window.open(urlString, '_blank');
        } else {
          window.open(urlString, '_self');
        }
      }
    }]);
    return WebSocialShareUtils;
  }();

  var WebSocialShareTwitter = /*#__PURE__*/function () {
    function WebSocialShareTwitter() {
      babelHelpers.classCallCheck(this, WebSocialShareTwitter);
    }

    babelHelpers.createClass(WebSocialShareTwitter, null, [{
      key: "share",
      value: function () {
        var _share2 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2(attrs) {
          var urlString;
          return regeneratorRuntime.wrap(function _callee2$(_context2) {
            while (1) {
              switch (_context2.prev = _context2.next) {
                case 0:
                  urlString = 'https://www.twitter.com/intent/tweet?';

                  if (attrs.socialShareText) {
                    urlString += 'text=' + encodeURIComponent(attrs.socialShareText);
                  }

                  if (attrs.socialShareVia) {
                    urlString += '&via=' + encodeURIComponent(attrs.socialShareVia);
                  }

                  if (attrs.socialShareHashtags) {
                    urlString += '&hashtags=' + encodeURIComponent(attrs.socialShareHashtags);
                  } //default to the current page if a URL isn't specified


                  urlString += '&url=' + encodeURIComponent(attrs.socialShareUrl || window.location.href);

                  if (WebSocialShareUtils.isMobile()) {
                    WebSocialShareUtils.staticOpenNewWindow(urlString);
                  } else {
                    window.open(urlString, 'Twitter', 'toolbar=0,status=0,resizable=yes,width=' + attrs.socialSharePopupWidth + ',height=' + attrs.socialSharePopupHeight + ',top=' + (window.innerHeight - attrs.socialSharePopupHeight) / 2 + ',left=' + (window.innerWidth - attrs.socialSharePopupWidth) / 2);
                  }

                case 6:
                case "end":
                  return _context2.stop();
              }
            }
          }, _callee2);
        }));

        function share(_x2) {
          return _share2.apply(this, arguments);
        }

        return share;
      }()
    }]);
    return WebSocialShareTwitter;
  }();

  var WebSocialShareEmail = /*#__PURE__*/function () {
    function WebSocialShareEmail() {
      babelHelpers.classCallCheck(this, WebSocialShareEmail);
    }

    babelHelpers.createClass(WebSocialShareEmail, null, [{
      key: "share",
      value: function () {
        var _share3 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3(attrs) {
          var urlString;
          return regeneratorRuntime.wrap(function _callee3$(_context3) {
            while (1) {
              switch (_context3.prev = _context3.next) {
                case 0:
                  urlString = 'mailto:';

                  if (attrs.socialShareTo) {
                    urlString += encodeURIComponent(attrs.socialShareTo);
                  }

                  urlString += '?';

                  if (attrs.socialShareBody) {
                    urlString += 'body=' + encodeURIComponent(attrs.socialShareBody);
                  }

                  if (attrs.socialShareSubject) {
                    urlString += '&subject=' + encodeURIComponent(attrs.socialShareSubject);
                  }

                  if (attrs.socialShareCc) {
                    urlString += '&cc=' + encodeURIComponent(attrs.socialShareCc);
                  }

                  if (attrs.socialShareBcc) {
                    urlString += '&bcc=' + encodeURIComponent(attrs.socialShareBcc);
                  }

                  WebSocialShareUtils.staticOpenNewWindow(urlString);

                case 8:
                case "end":
                  return _context3.stop();
              }
            }
          }, _callee3);
        }));

        function share(_x3) {
          return _share3.apply(this, arguments);
        }

        return share;
      }()
    }]);
    return WebSocialShareEmail;
  }();

  var WebSocialShareLinkedin = /*#__PURE__*/function () {
    function WebSocialShareLinkedin() {
      babelHelpers.classCallCheck(this, WebSocialShareLinkedin);
    }

    babelHelpers.createClass(WebSocialShareLinkedin, null, [{
      key: "share",
      value: function () {
        var _share4 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4(attrs) {
          var urlString;
          return regeneratorRuntime.wrap(function _callee4$(_context4) {
            while (1) {
              switch (_context4.prev = _context4.next) {
                case 0:
                  urlString = 'https://www.linkedin.com/shareArticle?mini=true';
                  urlString += '&url=' + encodeURIComponent(attrs.socialShareUrl || window.location.href);

                  if (attrs.socialShareText) {
                    urlString += '&title=' + encodeURIComponent(attrs.socialShareText);
                  }

                  if (attrs.socialShareDescription) {
                    urlString += '&summary=' + encodeURIComponent(attrs.socialShareDescription);
                  }

                  if (attrs.socialShareSource) {
                    urlString += '&source=' + encodeURIComponent(attrs.socialShareSource);
                  }

                  window.open(urlString, 'Linkedin', 'toolbar=0,status=0,resizable=yes,width=' + attrs.socialSharePopupWidth + ',height=' + attrs.socialSharePopupHeight + ',top=' + (window.innerHeight - attrs.socialSharePopupHeight) / 2 + ',left=' + (window.innerWidth - attrs.socialSharePopupWidth) / 2);

                case 6:
                case "end":
                  return _context4.stop();
              }
            }
          }, _callee4);
        }));

        function share(_x4) {
          return _share4.apply(this, arguments);
        }

        return share;
      }()
    }]);
    return WebSocialShareLinkedin;
  }();

  var WebSocialSharePinterest = /*#__PURE__*/function () {
    function WebSocialSharePinterest() {
      babelHelpers.classCallCheck(this, WebSocialSharePinterest);
    }

    babelHelpers.createClass(WebSocialSharePinterest, null, [{
      key: "share",
      value: function () {
        var _share5 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee5(attrs) {
          return regeneratorRuntime.wrap(function _callee5$(_context5) {
            while (1) {
              switch (_context5.prev = _context5.next) {
                case 0:
                  window.open('https://www.pinterest.com/pin/create/button/?url=' + encodeURIComponent(attrs.socialShareUrl || window.location.href) + '&media=' + encodeURIComponent(attrs.socialShareMedia) + '&description=' + encodeURIComponent(attrs.socialShareText), 'Pinterest', 'toolbar=0,status=0,resizable=yes,width=' + attrs.socialSharePopupWidth + ',height=' + attrs.socialSharePopupHeight + ',top=' + (window.innerHeight - attrs.socialSharePopupHeight) / 2 + ',left=' + (window.innerWidth - attrs.socialSharePopupWidth) / 2);

                case 1:
                case "end":
                  return _context5.stop();
              }
            }
          }, _callee5);
        }));

        function share(_x5) {
          return _share5.apply(this, arguments);
        }

        return share;
      }()
    }]);
    return WebSocialSharePinterest;
  }();

  var WebSocialShareReddit = /*#__PURE__*/function () {
    function WebSocialShareReddit() {
      babelHelpers.classCallCheck(this, WebSocialShareReddit);
    }

    babelHelpers.createClass(WebSocialShareReddit, null, [{
      key: "share",
      value: function () {
        var _share6 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee6(attrs) {
          var urlString;
          return regeneratorRuntime.wrap(function _callee6$(_context6) {
            while (1) {
              switch (_context6.prev = _context6.next) {
                case 0:
                  urlString = 'https://www.reddit.com/';

                  if (attrs.socialShareSubreddit) {
                    urlString += 'r/' + attrs.socialShareSubreddit + '/submit?url=';
                  } else {
                    urlString += 'submit?url=';
                  }
                  /*-
                  * Reddit isn't responsive and at default width for our popups (500 x 500), everything is messed up.
                  * So, overriding the width if it is less than 900 (played around to settle on this) and height if
                  * it is less than 650px.
                  */


                  if (attrs.socialSharePopupWidth < 900) {
                    attrs.socialSharePopupWidth = 900;
                  }

                  if (attrs.socialSharePopupHeight < 650) {
                    attrs.socialSharePopupHeight = 650;
                  }

                  window.open(urlString + encodeURIComponent(attrs.socialShareUrl || window.location.href) + '&title=' + encodeURIComponent(attrs.socialShareText), 'Reddit', 'toolbar=0,status=0,resizable=yes,width=' + attrs.socialSharePopupWidth + ',height=' + attrs.socialSharePopupHeight + ',top=' + (window.innerHeight - attrs.socialSharePopupHeight) / 2 + ',left=' + (window.innerWidth - attrs.socialSharePopupWidth) / 2);

                case 5:
                case "end":
                  return _context6.stop();
              }
            }
          }, _callee6);
        }));

        function share(_x6) {
          return _share6.apply(this, arguments);
        }

        return share;
      }()
    }]);
    return WebSocialShareReddit;
  }();

  var WebSocialShareWhatsapp = /*#__PURE__*/function () {
    function WebSocialShareWhatsapp() {
      babelHelpers.classCallCheck(this, WebSocialShareWhatsapp);
    }

    babelHelpers.createClass(WebSocialShareWhatsapp, null, [{
      key: "share",
      value: function () {
        var _share7 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee7(attrs) {
          var isMobile, urlString;
          return regeneratorRuntime.wrap(function _callee7$(_context7) {
            while (1) {
              switch (_context7.prev = _context7.next) {
                case 0:
                  isMobile = WebSocialShareUtils.isMobile();
                  urlString = isMobile ? 'https://api.whatsapp.com/send?text=' : 'https://web.whatsapp.com/send?text=';

                  if (attrs.socialShareText) {
                    urlString += encodeURIComponent(attrs.socialShareText) + '%0A';
                  } //default to the current page if a URL isn't specified


                  urlString += encodeURIComponent(attrs.socialShareUrl || window.location.href);

                  if (isMobile) {
                    WebSocialShareUtils.staticOpenNewWindow(urlString);
                  } else {
                    window.open(urlString, 'WhatsApp', 'toolbar=0,status=0,resizable=yes,width=' + attrs.socialSharePopupWidth + ',height=' + attrs.socialSharePopupHeight + ',top=' + (window.innerHeight - attrs.socialSharePopupHeight) / 2 + ',left=' + (window.innerWidth - attrs.socialSharePopupWidth) / 2);
                  }

                case 5:
                case "end":
                  return _context7.stop();
              }
            }
          }, _callee7);
        }));

        function share(_x7) {
          return _share7.apply(this, arguments);
        }

        return share;
      }()
    }]);
    return WebSocialShareWhatsapp;
  }();

  var WebSocialShareCopy = /*#__PURE__*/function () {
    function WebSocialShareCopy() {
      babelHelpers.classCallCheck(this, WebSocialShareCopy);
    }

    babelHelpers.createClass(WebSocialShareCopy, null, [{
      key: "share",
      value: function () {
        var _share8 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee8(attrs) {
          return regeneratorRuntime.wrap(function _callee8$(_context8) {
            while (1) {
              switch (_context8.prev = _context8.next) {
                case 0:
                  _context8.prev = 0;
                  _context8.next = 3;
                  return navigator.clipboard.writeText(attrs.socialShareUrl || window.location.href);

                case 3:
                  _context8.next = 7;
                  break;

                case 5:
                  _context8.prev = 5;
                  _context8.t0 = _context8["catch"](0);

                case 7:
                case "end":
                  return _context8.stop();
              }
            }
          }, _callee8, null, [[0, 5]]);
        }));

        function share(_x8) {
          return _share8.apply(this, arguments);
        }

        return share;
      }()
    }]);
    return WebSocialShareCopy;
  }();

  var WebSocialShareHackerNews = /*#__PURE__*/function () {
    function WebSocialShareHackerNews() {
      babelHelpers.classCallCheck(this, WebSocialShareHackerNews);
    }

    babelHelpers.createClass(WebSocialShareHackerNews, null, [{
      key: "share",
      value: function () {
        var _share9 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee9(attrs) {
          var urlString;
          return regeneratorRuntime.wrap(function _callee9$(_context9) {
            while (1) {
              switch (_context9.prev = _context9.next) {
                case 0:
                  urlString = 'https://news.ycombinator.com/submitlink?u='; //default to the current page if a URL isn't specified

                  urlString += encodeURIComponent(attrs.socialShareUrl || window.location.href);

                  if (attrs.socialShareText) {
                    urlString += '&t=' + encodeURIComponent(attrs.socialShareText);
                  }

                  WebSocialShareUtils.staticOpenNewWindow(urlString);

                case 4:
                case "end":
                  return _context9.stop();
              }
            }
          }, _callee9);
        }));

        function share(_x9) {
          return _share9.apply(this, arguments);
        }

        return share;
      }()
    }]);
    return WebSocialShareHackerNews;
  }();

  var webSocialShareCss = "div.web-social-share{visibility:hidden;opacity:0;cursor:pointer;-ms-touch-action:manipulation;touch-action:manipulation}div.web-social-share.web-social-share-open{visibility:visible;opacity:1}div.web-social-share.web-social-share-open div.web-social-share-backdrop{opacity:var(--web-social-share-backdrop-opacity, 0.25)}div.web-social-share.web-social-share-open div.web-social-share-action-sheet{opacity:1}div.web-social-share.web-social-share-open div.web-social-share-action-sheet div.web-social-share-action-sheet-container div.web-social-share-action-sheet-group{height:var(--web-social-share-height, 80px)}@media (max-width: 540px){div.web-social-share.web-social-share-open div.web-social-share-action-sheet div.web-social-share-action-sheet-container div.web-social-share-action-sheet-group{height:var(--web-social-share-height-small-device, 140px)}}div.web-social-share.web-social-share-open.web-social-share-transition-close div.web-social-share-backdrop{opacity:0}div.web-social-share.web-social-share-open.web-social-share-transition-close div.web-social-share-action-sheet div.web-social-share-action-sheet-container div.web-social-share-action-sheet-group{height:0}div.web-social-share div.web-social-share-backdrop{opacity:0;-webkit-transition:opacity 0.1s linear;transition:opacity 0.1s linear;background-color:var(--web-social-share-backdrop-background, black);z-index:var(--web-social-share-zindex, 1000);-webkit-transform:translate3d(0, 0, 2px);transform:translate3d(0, 0, 2px);left:0;top:0;position:fixed;height:100%;width:100%}div.web-social-share div.web-social-share-action-sheet{left:0;right:0;top:0;bottom:0;margin:auto;position:fixed;z-index:calc(var(--web-social-share-zindex, 1000) + 1);-webkit-transform:translate3d(0, 0, 3px);transform:translate3d(0, 0, 3px);width:100%;max-width:540px}@media (min-width: 540px){div.web-social-share div.web-social-share-action-sheet div.web-social-share-action-sheet-container div.web-social-share-action-sheet-group{border-radius:var(--web-social-share-action-sheet-group-border-radius, 8px 8px 0 0)}}div.web-social-share div.web-social-share-action-sheet div.web-social-share-action-sheet-container{display:-ms-flexbox;display:flex;-ms-flex-flow:column;flex-flow:column;-ms-flex-pack:end;justify-content:flex-end;height:100%;max-height:100%}div.web-social-share div.web-social-share-action-sheet div.web-social-share-action-sheet-container div.web-social-share-action-sheet-group{-webkit-box-shadow:var(--web-social-share-action-sheet-group-box-shadow, 0 0 8px 4px rgba(0, 0, 0, 0.1));box-shadow:var(--web-social-share-action-sheet-group-box-shadow, 0 0 8px 4px rgba(0, 0, 0, 0.1));z-index:calc(var(--web-social-share-zindex, 1000) + 10);-webkit-transform:translate3d(0, 0, 10px);transform:translate3d(0, 0, 10px);background:var(--web-social-share-action-sheet-group-background, #fafafa);display:-ms-flexbox;display:flex;-ms-flex-pack:center;justify-content:center;-ms-flex-wrap:wrap;flex-wrap:wrap;height:0;-webkit-transition-timing-function:cubic-bezier(0.36, 0.66, 0.04, 1);transition-timing-function:cubic-bezier(0.36, 0.66, 0.04, 1);-webkit-transition:height 0.2s ease-in;transition:height 0.2s ease-in}@media (max-width: 540px){div.web-social-share div.web-social-share-action-sheet div.web-social-share-action-sheet-container div.web-social-share-action-sheet-group{-ms-flex-pack:start;justify-content:flex-start}}div.web-social-share div.web-social-share-target{margin:auto;width:var(--web-social-share-target-width, 4rem);height:var(--web-social-share-target-height, 3rem);display:-ms-flexbox;display:flex;-ms-flex-direction:column;flex-direction:column;-ms-flex-align:center;align-items:center;-ms-flex-pack:center;justify-content:center}div.web-social-share div.web-social-share-target button{position:relative;cursor:pointer;border:0;background:transparent;width:var(--web-social-share-button-width, 100%);height:var(--web-social-share-button-height, 100%);font-size:var(--web-social-share-button-font-size)}div.web-social-share div.web-social-share-target p{margin:var(--web-social-share-brand-margin, 2px 0);color:var(--web-social-share-brand-color, inherit);font-size:var(--web-social-share-brand-font-size, 0.6rem)}div.web-social-share div.web-social-share-target div.web-social-share-button-icon{margin:0;display:-ms-flexbox;display:flex;-ms-flex-direction:column;flex-direction:column;-ms-flex-pack:center;justify-content:center;-ms-flex-align:center;align-items:center}::slotted([slot=facebook]),::slotted([slot=twitter]),::slotted([slot=email]),::slotted([slot=linkedin]),::slotted([slot=pinterest]),::slotted([slot=reddit]),::slotted([slot=whatsapp]),::slotted([slot=copy]),::slotted([slot=hackernews]){display:none}";

  var WebSocialShare = /*#__PURE__*/function () {
    function WebSocialShare(hostRef) {
      babelHelpers.classCallCheck(this, WebSocialShare);
      (0, _indexBef2383e.r)(this, hostRef);
      this.closed = (0, _indexBef2383e.c)(this, "closed", 7);

      this.handleFacebookShare = /*#__PURE__*/function () {
        var _ref = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee10(attributes) {
          return regeneratorRuntime.wrap(function _callee10$(_context10) {
            while (1) {
              switch (_context10.prev = _context10.next) {
                case 0:
                  _context10.next = 2;
                  return WebSocialShareFacebook.share(attributes);

                case 2:
                case "end":
                  return _context10.stop();
              }
            }
          }, _callee10);
        }));

        return function (_x10) {
          return _ref.apply(this, arguments);
        };
      }();

      this.handleTwitterShare = /*#__PURE__*/function () {
        var _ref2 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee11(attributes) {
          return regeneratorRuntime.wrap(function _callee11$(_context11) {
            while (1) {
              switch (_context11.prev = _context11.next) {
                case 0:
                  _context11.next = 2;
                  return WebSocialShareTwitter.share(attributes);

                case 2:
                case "end":
                  return _context11.stop();
              }
            }
          }, _callee11);
        }));

        return function (_x11) {
          return _ref2.apply(this, arguments);
        };
      }();

      this.handleEmailShare = /*#__PURE__*/function () {
        var _ref3 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee12(attributes) {
          return regeneratorRuntime.wrap(function _callee12$(_context12) {
            while (1) {
              switch (_context12.prev = _context12.next) {
                case 0:
                  _context12.next = 2;
                  return WebSocialShareEmail.share(attributes);

                case 2:
                case "end":
                  return _context12.stop();
              }
            }
          }, _callee12);
        }));

        return function (_x12) {
          return _ref3.apply(this, arguments);
        };
      }();

      this.handleLinkedinShare = /*#__PURE__*/function () {
        var _ref4 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee13(attributes) {
          return regeneratorRuntime.wrap(function _callee13$(_context13) {
            while (1) {
              switch (_context13.prev = _context13.next) {
                case 0:
                  _context13.next = 2;
                  return WebSocialShareLinkedin.share(attributes);

                case 2:
                case "end":
                  return _context13.stop();
              }
            }
          }, _callee13);
        }));

        return function (_x13) {
          return _ref4.apply(this, arguments);
        };
      }();

      this.handlePinterestShare = /*#__PURE__*/function () {
        var _ref5 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee14(attributes) {
          return regeneratorRuntime.wrap(function _callee14$(_context14) {
            while (1) {
              switch (_context14.prev = _context14.next) {
                case 0:
                  _context14.next = 2;
                  return WebSocialSharePinterest.share(attributes);

                case 2:
                case "end":
                  return _context14.stop();
              }
            }
          }, _callee14);
        }));

        return function (_x14) {
          return _ref5.apply(this, arguments);
        };
      }();

      this.handleRedditShare = /*#__PURE__*/function () {
        var _ref6 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee15(attributes) {
          return regeneratorRuntime.wrap(function _callee15$(_context15) {
            while (1) {
              switch (_context15.prev = _context15.next) {
                case 0:
                  _context15.next = 2;
                  return WebSocialShareReddit.share(attributes);

                case 2:
                case "end":
                  return _context15.stop();
              }
            }
          }, _callee15);
        }));

        return function (_x15) {
          return _ref6.apply(this, arguments);
        };
      }();

      this.handleWhatsappShare = /*#__PURE__*/function () {
        var _ref7 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee16(attributes) {
          return regeneratorRuntime.wrap(function _callee16$(_context16) {
            while (1) {
              switch (_context16.prev = _context16.next) {
                case 0:
                  _context16.next = 2;
                  return WebSocialShareWhatsapp.share(attributes);

                case 2:
                case "end":
                  return _context16.stop();
              }
            }
          }, _callee16);
        }));

        return function (_x16) {
          return _ref7.apply(this, arguments);
        };
      }();

      this.handleCopyShare = /*#__PURE__*/function () {
        var _ref8 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee17(attributes) {
          return regeneratorRuntime.wrap(function _callee17$(_context17) {
            while (1) {
              switch (_context17.prev = _context17.next) {
                case 0:
                  _context17.next = 2;
                  return WebSocialShareCopy.share(attributes);

                case 2:
                case "end":
                  return _context17.stop();
              }
            }
          }, _callee17);
        }));

        return function (_x17) {
          return _ref8.apply(this, arguments);
        };
      }();

      this.handleHackerNewsShare = /*#__PURE__*/function () {
        var _ref9 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee18(attributes) {
          return regeneratorRuntime.wrap(function _callee18$(_context18) {
            while (1) {
              switch (_context18.prev = _context18.next) {
                case 0:
                  _context18.next = 2;
                  return WebSocialShareHackerNews.share(attributes);

                case 2:
                case "end":
                  return _context18.stop();
              }
            }
          }, _callee18);
        }));

        return function (_x18) {
          return _ref9.apply(this, arguments);
        };
      }();
    }

    babelHelpers.createClass(WebSocialShare, [{
      key: "hide",
      value: function hide() {
        var _this = this;

        var element = this.el.shadowRoot.querySelector('div.web-social-share');

        if (element) {
          element.classList.add('web-social-share-transition-close');
          setTimeout(function () {
            // Reflect css animation speed 400ms, see css
            _this.show = false;
            element.classList.remove('web-social-share-transition-close');

            _this.closed.emit(true);
          }, 200);
        } else {
          // Well we don't find the action sheet, we could mark it as not displayed
          this.show = false;
          this.closed.emit(true);
        }
      }
    }, {
      key: "render",
      value: function render() {
        var _this2 = this;

        return (0, _indexBef2383e.h)("div", {
          class: this.show ? 'web-social-share web-social-share-open' : 'web-social-share web-social-share-close'
        }, (0, _indexBef2383e.h)("div", {
          class: 'web-social-share-backdrop',
          onClick: function onClick() {
            return _this2.hide();
          }
        }), (0, _indexBef2383e.h)("div", {
          class: 'web-social-share-action-sheet',
          onClick: function onClick() {
            return _this2.hide();
          }
        }, (0, _indexBef2383e.h)("div", {
          class: 'web-social-share-action-sheet-container'
        }, (0, _indexBef2383e.h)("div", {
          class: 'web-social-share-action-sheet-group'
        }, this.renderTargets()))));
      }
    }, {
      key: "renderTargets",
      value: function renderTargets() {
        var _this3 = this;

        if (!this.share || !this.share.config) {
          return (0, _indexBef2383e.h)("span", null);
        } else {
          return this.share.config.map(function (config) {
            return (0, _indexBef2383e.h)("div", {
              class: "web-social-share-target"
            }, _this3.renderButtons(config));
          });
        }
      }
    }, {
      key: "renderButtons",
      value: function renderButtons(share) {
        if (share.facebook) {
          return this.renderButton(share.facebook, 'facebook', this.handleFacebookShare, 'Facebook');
        } else if (share.twitter) {
          return this.renderButton(share.twitter, 'twitter', this.handleTwitterShare, 'Twitter');
        } else if (share.email) {
          return this.renderButton(share.email, 'email', this.handleEmailShare, 'Email');
        } else if (share.linkedin) {
          return this.renderButton(share.linkedin, 'linkedin', this.handleLinkedinShare, 'Linkedin');
        } else if (share.pinterest) {
          return this.renderButton(share.pinterest, 'pinterest', this.handlePinterestShare, 'Pinterest');
        } else if (share.reddit) {
          return this.renderButton(share.reddit, 'reddit', this.handleRedditShare, 'Reddit');
        } else if (share.whatsapp) {
          return this.renderButton(share.whatsapp, 'whatsapp', this.handleWhatsappShare, 'WhatsApp');
        } else if (share.copy) {
          return this.renderButton(share.copy, 'copy', this.handleCopyShare, 'Copy');
        } else if (share.hackernews) {
          return this.renderButton(share.hackernews, 'hackernews', this.handleHackerNewsShare, 'Hacker News');
        } else {
          return undefined;
        }
      }
    }, {
      key: "renderButton",
      value: function renderButton(attributes, slotName, action, defaultBrandName) {
        var _this4 = this;

        return (0, _indexBef2383e.h)("button", {
          onClick: function onClick($event) {
            return _this4.handleShare($event, attributes, action);
          },
          class: 'web-social-share-button'
        }, (0, _indexBef2383e.h)("div", {
          class: "web-social-share-button-icon"
        }, (0, _indexBef2383e.h)("slot", {
          name: slotName
        })), this.renderName(attributes, defaultBrandName));
      }
    }, {
      key: "handleShare",
      value: function () {
        var _handleShare = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee19($event, attributes, action) {
          return regeneratorRuntime.wrap(function _callee19$(_context19) {
            while (1) {
              switch (_context19.prev = _context19.next) {
                case 0:
                  $event.stopPropagation();
                  _context19.next = 3;
                  return action(attributes);

                case 3:
                  this.hide();

                case 4:
                case "end":
                  return _context19.stop();
              }
            }
          }, _callee19, this);
        }));

        function handleShare(_x19, _x20, _x21) {
          return _handleShare.apply(this, arguments);
        }

        return handleShare;
      }()
    }, {
      key: "renderName",
      value: function renderName(displayAttributes, defaultBrandName) {
        if (this.share.displayNames) {
          return (0, _indexBef2383e.h)("p", null, displayAttributes && displayAttributes.brandName && displayAttributes.brandName !== '' ? displayAttributes.brandName : defaultBrandName);
        } else {
          return (0, _indexBef2383e.h)("span", null);
        }
      }
    }, {
      key: "el",
      get: function get() {
        return (0, _indexBef2383e.g)(this);
      }
    }]);
    return WebSocialShare;
  }();

  _exports.web_social_share = WebSocialShare;
  WebSocialShare.style = webSocialShareCss;
});
