// @lrnwebcomponents/haxcms-elements/lib/core/utils/HAXCMSThemeParts.js
define(["exports", "../../../../../lit-element/lit-element.js", "../haxcms-site-store.js", "../../../../../mobx/lib/mobx.module.js"], function (_exports, _litElement, _haxcmsSiteStore, _mobxModule) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HAXCMSThemeParts = void 0;

  const HAXCMSThemeParts = function (SuperClass) {
    return class extends SuperClass {
      constructor() {
        super();
        this.__disposer = this.__disposer ? this.__disposer : [];
        (0, _mobxModule.autorun)(reaction => {
          this.editMode = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.editMode);

          this.__disposer.push(reaction);
        });
      }

      static get styles() {
        let styles = [];

        if (super.styles) {
          styles = super.styles;
        }

        return [...styles, (0, _litElement.css)`
          [part="edit-mode-active"] {
            pointer-events: var(
              --haxcms-theme-parts-edit-mode-active-pointer-events,
              none
            );
            opacity: var(--haxcms-theme-parts-edit-mode-active-opacity, 0.5);
            filter: var(
              --haxcms-theme-parts-edit-mode-active-filter,
              blur(1px)
            );
          }
        `];
      }

      static get properties() {
        let props = {};

        if (super.properties) {
          props = super.props;
        }

        return { ...props,
          editMode: {
            type: Boolean,
            attribute: "edit-mode",
            reflect: true
          }
        };
      }

    };
  };

  _exports.HAXCMSThemeParts = HAXCMSThemeParts;
});
