// @lrnwebcomponents/punnett-square/punnett-square.js
define(["exports", "../../lit-element/lit-element.js", "../../lit-html/directives/unsafe-html.js", "../../@polymer/paper-input/paper-input.js"], function (_exports, _litElement, _unsafeHtml, _paperInput) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.PunnettSquare = void 0;

  function _templateObject11_fb5b7410d0ee11ea896db74067b80894() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        ", "\n      "]);

    _templateObject11_fb5b7410d0ee11ea896db74067b80894 = function _templateObject11_fb5b7410d0ee11ea896db74067b80894() {
      return data;
    };

    return data;
  }

  function _templateObject10_fb5b7410d0ee11ea896db74067b80894() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        ", "\n      "]);

    _templateObject10_fb5b7410d0ee11ea896db74067b80894 = function _templateObject10_fb5b7410d0ee11ea896db74067b80894() {
      return data;
    };

    return data;
  }

  function _templateObject9_fb5b7410d0ee11ea896db74067b80894() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                            <td>\n                              ", "\n                            </td>\n                          "]);

    _templateObject9_fb5b7410d0ee11ea896db74067b80894 = function _templateObject9_fb5b7410d0ee11ea896db74067b80894() {
      return data;
    };

    return data;
  }

  function _templateObject8_fb5b7410d0ee11ea896db74067b80894() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <tr>\n                    <th>", "</th>\n                    ", "\n                  </tr>\n                "]);

    _templateObject8_fb5b7410d0ee11ea896db74067b80894 = function _templateObject8_fb5b7410d0ee11ea896db74067b80894() {
      return data;
    };

    return data;
  }

  function _templateObject7_fb5b7410d0ee11ea896db74067b80894() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                        <th>", "</th>\n                      "]);

    _templateObject7_fb5b7410d0ee11ea896db74067b80894 = function _templateObject7_fb5b7410d0ee11ea896db74067b80894() {
      return data;
    };

    return data;
  }

  function _templateObject6_fb5b7410d0ee11ea896db74067b80894() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  ", "\n                "]);

    _templateObject6_fb5b7410d0ee11ea896db74067b80894 = function _templateObject6_fb5b7410d0ee11ea896db74067b80894() {
      return data;
    };

    return data;
  }

  function _templateObject5_fb5b7410d0ee11ea896db74067b80894() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                ", "\n              "]);

    _templateObject5_fb5b7410d0ee11ea896db74067b80894 = function _templateObject5_fb5b7410d0ee11ea896db74067b80894() {
      return data;
    };

    return data;
  }

  function _templateObject4_fb5b7410d0ee11ea896db74067b80894() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                <paper-input\n                  value=\"", "\"\n                  @value-changed=", "\n                ></paper-input>\n              "]);

    _templateObject4_fb5b7410d0ee11ea896db74067b80894 = function _templateObject4_fb5b7410d0ee11ea896db74067b80894() {
      return data;
    };

    return data;
  }

  function _templateObject3_fb5b7410d0ee11ea896db74067b80894() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                ", "\n              "]);

    _templateObject3_fb5b7410d0ee11ea896db74067b80894 = function _templateObject3_fb5b7410d0ee11ea896db74067b80894() {
      return data;
    };

    return data;
  }

  function _templateObject2_fb5b7410d0ee11ea896db74067b80894() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                <paper-input\n                  value=\"", "\"\n                  @value-changed=", "\n                ></paper-input>\n              "]);

    _templateObject2_fb5b7410d0ee11ea896db74067b80894 = function _templateObject2_fb5b7410d0ee11ea896db74067b80894() {
      return data;
    };

    return data;
  }

  function _templateObject_fb5b7410d0ee11ea896db74067b80894() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: block;\n          --punnett-square-table-border: 1px solid black;\n          --punnett-square-table-padding: 1em;\n          --punnett-square-input-font-size: 2em;\n          --punnett-square-input-text-align: center;\n        }\n\n        table {\n          border-collapse: var(--punnett-square-table-border-collapse, collapse);\n          width: var(--punnett-square-table-width, 100%);\n          border: var(--punnett-square-table-border);\n        }\n\n        th {\n          border: var(--punnett-square-th-border, var(--punnett-square-table-border));\n          padding: var(--punnett-square-th-padding, var(--punnett-square-table-padding));\n        }\n\n        td {\n          border: var(--punnett-square-td-border, var(--punnett-square-table-border));\n          text-align: var(--punnett-square-table-text-align, center);\n          padding: var(--punnett-square-td-padding, var(--punnett-square-table-padding));\n        }\n\n        #inputs {\n          display: var(--punnett-square-inputs-display, flex);\n          align-items: var(--punnett-square-inputs-align-items, center);\n        }\n\n        .input {\n          font-size: var(--punnett-square-input-font-size);\n          text-align: var(--punnett-square-input-text-align);\n        }\n\n        paper-input {\n          --paper-input-container-input: {\n            font-size: var(--punnett-square-paper-input-font-size, --punnett-square-input-font-size);\n            text-align: var(--punnett-square-paper-input-text-align, --punnett-square-input-text-align);\n          }\n        }\n      </style>\n      <div id=\"inputs\">\n        Variant1: <div class=\"input\">", "</div>\n        Variant2: <div class=\"input\">", "</div>\n      </div>\n\n      <table>\n        <thead>\n          <tr>\n            <th></th>\n            ", "\n          </tr>\n        <tbody>\n          ", "\n        </tbody>\n        </thead>\n      </table>\n    "]);

    _templateObject_fb5b7410d0ee11ea896db74067b80894 = function _templateObject_fb5b7410d0ee11ea896db74067b80894() {
      return data;
    };

    return data;
  }

  /**
   * `punnett-square`
   * @element punnett-square
   * @lit-element
   * @demo demo/index.html
   */
  var PunnettSquare =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(PunnettSquare, _LitElement);
    babelHelpers.createClass(PunnettSquare, null, [{
      key: "tag",

      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */
      get: function get() {
        return "punnett-square";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Varient 1 - example XY
           */
          variant1: {
            type: String
          },

          /**
           * Varient 2 - example XX
           */
          variant2: {
            type: String
          },

          /**
           * Editable - allow users to change the varient values
           */
          editable: {
            type: Boolean
          },
          __results: {
            type: Array
          }
        };
      } // life cycle

    }]);

    function PunnettSquare() {
      var _this;

      babelHelpers.classCallCheck(this, PunnettSquare);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(PunnettSquare).call(this));
      _this.variant1 = "";
      _this.variant2 = "";
      _this.editable = false; // Gather Lightdom Rendered Results

      var results = _this.querySelectorAll("[data-variant]");

      var _renderedResults = [];
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = results[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var result = _step.value;
          var variant = result.getAttribute("data-variant");
          variant = _this.__sortVariant(variant);
          var item = {
            variant: variant,
            ref: result
          };
          _renderedResults = [].concat(babelHelpers.toConsumableArray(_renderedResults), [item]);
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return != null) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      _this.__results = _renderedResults;
      return _this;
    }

    babelHelpers.createClass(PunnettSquare, [{
      key: "__sortVariant",
      value: function __sortVariant(variant) {
        return variant.split("").sort().join("");
      }
      /**
       * Ensures that the value is a multidimensional array
       */

    }, {
      key: "__ensureArray",
      value: function __ensureArray(value) {
        var toplevel = typeof value === "string" ? value.split(",").map(function (i) {
          return i.trim();
        }) : [];
        return toplevel.map(function (i) {
          return typeof i === "string" ? i.split("").map(function (i) {
            return i.trim();
          }) : [];
        });
      }
    }, {
      key: "render",
      value: function render() {
        var _this2 = this;

        var value1 = this.__ensureArray(this.variant1);

        var value2 = this.__ensureArray(this.variant2);

        var asdf = value1[0].length < 1 && value2[0].length < 1;
        console.log("asdf:", asdf);
        return (0, _litElement.html)(_templateObject_fb5b7410d0ee11ea896db74067b80894(), this.editable ? (0, _litElement.html)(_templateObject2_fb5b7410d0ee11ea896db74067b80894(), this.variant1, function (e) {
          return _this2.variant1 = e.detail.value;
        }) : (0, _litElement.html)(_templateObject3_fb5b7410d0ee11ea896db74067b80894(), this.variant1), this.editable ? (0, _litElement.html)(_templateObject4_fb5b7410d0ee11ea896db74067b80894(), this.variant2, function (e) {
          return _this2.variant2 = e.detail.value;
        }) : (0, _litElement.html)(_templateObject5_fb5b7410d0ee11ea896db74067b80894(), this.variant2), value1.map(function (allele1) {
          return (0, _litElement.html)(_templateObject6_fb5b7410d0ee11ea896db74067b80894(), allele1.map(function (i) {
            return (0, _litElement.html)(_templateObject7_fb5b7410d0ee11ea896db74067b80894(), i);
          }));
        }), value2.map(function (allele2, index) {
          return allele2.map(function (allele) {
            return (0, _litElement.html)(_templateObject8_fb5b7410d0ee11ea896db74067b80894(), allele, value1.map(function (v1) {
              return v1.map(function (allele1) {
                return (0, _litElement.html)(_templateObject9_fb5b7410d0ee11ea896db74067b80894(), _this2.renderVariant("".concat(allele1).concat(allele)));
              });
            }));
          });
        }));
      }
    }, {
      key: "renderVariant",
      value: function renderVariant(variant) {
        var _variant = this.__sortVariant(variant);

        var result = this.__results.find(function (i) {
          return i.variant === _variant;
        });

        if (result) {
          return (0, _litElement.html)(_templateObject10_fb5b7410d0ee11ea896db74067b80894(), (0, _unsafeHtml.unsafeHTML)(result.ref.outerHTML));
        } else {
          return (0, _litElement.html)(_templateObject11_fb5b7410d0ee11ea896db74067b80894(), _variant);
        }
      }
    }]);
    return PunnettSquare;
  }(_litElement.LitElement);

  _exports.PunnettSquare = PunnettSquare;
  customElements.define(PunnettSquare.tag, PunnettSquare);
});
