// lit-html/async-directive.js
define(["exports", "./lit-html.js", "./directive.js", "./directive-helpers.js"], function (_exports, _litHtml, _directive, _directiveHelpers) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  Object.defineProperty(_exports, "directive", {
    enumerable: true,
    get: function get() {
      return _directive.directive;
    }
  });
  _exports.AsyncDirective = void 0;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

  function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

  function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

  /**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var r = function r(i, t) {
    var s, e;
    var o = i.N;
    if (void 0 === o) return !1;

    var _iterator = _createForOfIteratorHelper(o),
        _step;

    try {
      for (_iterator.s(); !(_step = _iterator.n()).done;) {
        var _i = _step.value;
        null === (e = (s = _i).O) || void 0 === e || e.call(s, t, !1), r(_i, t);
      }
    } catch (err) {
      _iterator.e(err);
    } finally {
      _iterator.f();
    }

    return !0;
  },
      o = function o(i) {
    var t, s;

    do {
      if (void 0 === (t = i.M)) break;
      s = t.N, s.delete(i), i = t;
    } while (0 === (null == s ? void 0 : s.size));
  },
      h = function h(i) {
    for (var _t; _t = i.M; i = _t) {
      var _s = _t.N;
      if (void 0 === _s) _t.N = _s = new Set();else if (_s.has(i)) break;
      _s.add(i), d(_t);
    }
  };

  function n(i) {
    void 0 !== this.N ? (o(this), this.M = i, h(this)) : this.M = i;
  }

  function l(i) {
    var t = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : !1;
    var s = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 0;
    var e = this.H,
        h = this.N;
    if (void 0 !== h && 0 !== h.size) if (t) {
      if (Array.isArray(e)) for (var _i2 = s; _i2 < e.length; _i2++) {
        r(e[_i2], !1), o(e[_i2]);
      } else null != e && (r(e, !1), o(e));
    } else r(this, i);
  }

  var d = function d(i) {
    var t, e, r, o;
    i.type == _directive.PartType.CHILD && (null !== (t = (r = i).P) && void 0 !== t || (r.P = l), null !== (e = (o = i).Q) && void 0 !== e || (o.Q = n));
  };

  var c = /*#__PURE__*/function (_t2) {
    babelHelpers.inherits(c, _t2);

    var _super = _createSuper(c);

    function c() {
      var _this;

      babelHelpers.classCallCheck(this, c);
      _this = _super.apply(this, arguments), _this.isConnected = !0, _this.ut = _litHtml.noChange, _this.N = void 0;
      return _this;
    }

    babelHelpers.createClass(c, [{
      key: "T",
      value: function T(i, t, s) {
        babelHelpers.get(babelHelpers.getPrototypeOf(c.prototype), "T", this).call(this, i, t, s), h(this);
      }
    }, {
      key: "O",
      value: function O(i) {
        var t = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : !0;
        this.at(i), t && (r(this, i), o(this));
      }
    }, {
      key: "at",
      value: function at(t) {
        var s, e;
        t !== this.isConnected && (t ? (this.isConnected = !0, this.ut !== _litHtml.noChange && (this.setValue(this.ut), this.ut = _litHtml.noChange), null === (s = this.reconnected) || void 0 === s || s.call(this)) : (this.isConnected = !1, null === (e = this.disconnected) || void 0 === e || e.call(this)));
      }
    }, {
      key: "S",
      value: function S(i, t) {
        if (!this.isConnected) throw Error("AsyncDirective ".concat(this.constructor.name, " was rendered while its tree was disconnected."));
        return babelHelpers.get(babelHelpers.getPrototypeOf(c.prototype), "S", this).call(this, i, t);
      }
    }, {
      key: "setValue",
      value: function setValue(i) {
        if (this.isConnected) {
          if ((0, _directiveHelpers.isSingleExpression)(this.Σdt)) this.Σdt.I(i, this);else {
            var _t3 = babelHelpers.toConsumableArray(this.Σdt.H);

            _t3[this.Σct] = i, this.Σdt.I(_t3, this, 0);
          }
        } else this.ut = i;
      }
    }, {
      key: "disconnected",
      value: function disconnected() {}
    }, {
      key: "reconnected",
      value: function reconnected() {}
    }]);
    return c;
  }(_directive.Directive);

  _exports.AsyncDirective = c;
});
