// lit-html/directive-helpers.js
define(["exports", "./lit-html.js"], function (_exports, _litHtml) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.setCommittedValue = _exports.setChildPartValue = _exports.removePart = _exports.isTemplateResult = _exports.isSingleExpression = _exports.isPrimitive = _exports.isDirectiveResult = _exports.insertPart = _exports.getDirectiveClass = _exports.getCommittedValue = _exports.clearPart = _exports.TemplateResultType = void 0;

  /**
   * @license
   * Copyright 2020 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var l = _litHtml._$LH.I,
      t = function t(o) {
    return null === o || "object" != babelHelpers.typeof(o) && "function" != typeof o;
  },
      i = {
    HTML: 1,
    SVG: 2
  },
      n = function n(o, l) {
    return void 0 === l ? void 0 !== (null == o ? void 0 : o._$litType$) : (null == o ? void 0 : o._$litType$) === l;
  },
      d = function d(o) {
    return void 0 !== (null == o ? void 0 : o._$litDirective$);
  },
      v = function v(o) {
    return null == o ? void 0 : o._$litDirective$;
  },
      e = function e(o) {
    return void 0 === o.strings;
  },
      c = function c() {
    return document.createComment("");
  },
      r = function r(o, t, i) {
    var n;
    var d = o._$AA.parentNode,
        v = void 0 === t ? o._$AB : t._$AA;

    if (void 0 === i) {
      var _t = d.insertBefore(c(), v),
          _n = d.insertBefore(c(), v);

      i = new l(_t, _n, o, o.options);
    } else {
      var _l = i._$AB.nextSibling,
          _t2 = i._$AM,
          _e = _t2 !== o;

      if (_e) {
        var _l2;

        null === (n = i._$AQ) || void 0 === n || n.call(i, o), i._$AM = o, void 0 !== i._$AP && (_l2 = o._$AU) !== _t2._$AU && i._$AP(_l2);
      }

      if (_l !== v || _e) {
        var _o = i._$AA;

        for (; _o !== _l;) {
          var _l3 = _o.nextSibling;
          d.insertBefore(_o, v), _o = _l3;
        }
      }
    }

    return i;
  },
      u = function u(o, l) {
    var t = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : o;
    return o._$AI(l, t), o;
  },
      f = {},
      s = function s(o) {
    var l = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : f;
    return o._$AH = l;
  },
      m = function m(o) {
    return o._$AH;
  },
      p = function p(o) {
    var l;
    null === (l = o._$AP) || void 0 === l || l.call(o, !1, !0);
    var t = o._$AA;
    var i = o._$AB.nextSibling;

    for (; t !== i;) {
      var _o2 = t.nextSibling;
      t.remove(), t = _o2;
    }
  },
      a = function a(o) {
    o._$AR();
  };

  _exports.clearPart = a;
  _exports.removePart = p;
  _exports.getCommittedValue = m;
  _exports.setCommittedValue = s;
  _exports.setChildPartValue = u;
  _exports.insertPart = r;
  _exports.isSingleExpression = e;
  _exports.getDirectiveClass = v;
  _exports.isDirectiveResult = d;
  _exports.isTemplateResult = n;
  _exports.TemplateResultType = i;
  _exports.isPrimitive = t;
});
