// lit-html/directive.js
define(["exports"], function (_exports) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.directive = _exports.PartType = _exports.Directive = void 0;

  /**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var t = {
    ATTRIBUTE: 1,
    CHILD: 2,
    PROPERTY: 3,
    BOOLEAN_ATTRIBUTE: 4,
    EVENT: 5,
    ELEMENT: 6
  },
      i = function i(t) {
    return function () {
      for (var _len = arguments.length, i = new Array(_len), _key = 0; _key < _len; _key++) {
        i[_key] = arguments[_key];
      }

      return {
        _$litDirective$: t,
        values: i
      };
    };
  };

  _exports.directive = i;
  _exports.PartType = t;

  var s = /*#__PURE__*/function () {
    function s(t) {
      babelHelpers.classCallCheck(this, s);
    }

    babelHelpers.createClass(s, [{
      key: "T",
      value: function T(t, i, _s) {
        this.Σdt = t, this.M = i, this.Σct = _s;
      }
    }, {
      key: "S",
      value: function S(t, i) {
        return this.update(t, i);
      }
    }, {
      key: "update",
      value: function update(t, i) {
        return this.render.apply(this, babelHelpers.toConsumableArray(i));
      }
    }]);
    return s;
  }();

  _exports.Directive = s;
});
