// lit-html/directive-helpers.js
define(["exports", "./lit-html.js"], function (_exports, _litHtml) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.setCommittedValue = _exports.setChildPartValue = _exports.removePart = _exports.isTemplateResult = _exports.isSingleExpression = _exports.isPrimitive = _exports.isDirectiveResult = _exports.insertPart = _exports.getDirectiveClass = _exports.getCommittedValue = _exports.clearPart = _exports.TemplateResultType = void 0;

  /**
   * @license
   * Copyright 2020 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var t = _litHtml._Σ.et,
      i = function i(o) {
    return null === o || "object" != babelHelpers.typeof(o) && "function" != typeof o;
  },
      n = {
    HTML: 1,
    SVG: 2
  },
      v = function v(o, t) {
    var i, n;
    return void 0 === t ? void 0 !== (null === (i = o) || void 0 === i ? void 0 : i._$litType$) : (null === (n = o) || void 0 === n ? void 0 : n._$litType$) === t;
  },
      l = function l(o) {
    var t;
    return void 0 !== (null === (t = o) || void 0 === t ? void 0 : t._$litDirective$);
  },
      r = function r(o) {
    var t;
    return null === (t = o) || void 0 === t ? void 0 : t._$litDirective$;
  },
      d = function d(o) {
    return void 0 === o.strings;
  },
      e = function e() {
    return document.createComment("");
  },
      u = function u(o, i, n) {
    var v;
    var l = o.A.parentNode,
        r = void 0 === i ? o.B : i.A;

    if (void 0 === n) {
      var _i = l.insertBefore(e(), r),
          _v = l.insertBefore(e(), r);

      n = new t(_i, _v, o, o.options);
    } else {
      var _t = n.B.nextSibling,
          _i2 = n.M !== o;

      if (_i2 && (null === (v = n.Q) || void 0 === v || v.call(n, o), n.M = o), _t !== r || _i2) {
        var _o = n.A;

        for (; _o !== _t;) {
          var _t2 = _o.nextSibling;
          l.insertBefore(_o, r), _o = _t2;
        }
      }
    }

    return n;
  },
      c = function c(o, t) {
    var i = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : o;
    return o.I(t, i), o;
  },
      s = {},
      f = function f(o) {
    var t = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : s;
    return o.H = t;
  },
      a = function a(o) {
    return o.H;
  },
      m = function m(o) {
    var t;
    null === (t = o.P) || void 0 === t || t.call(o, !1, !0);
    var i = o.A;
    var n = o.B.nextSibling;

    for (; i !== n;) {
      var _o2 = i.nextSibling;
      i.remove(), i = _o2;
    }
  },
      p = function p(o) {
    o.R();
  };

  _exports.clearPart = p;
  _exports.removePart = m;
  _exports.getCommittedValue = a;
  _exports.setCommittedValue = f;
  _exports.setChildPartValue = c;
  _exports.insertPart = u;
  _exports.isSingleExpression = d;
  _exports.getDirectiveClass = r;
  _exports.isDirectiveResult = l;
  _exports.isTemplateResult = v;
  _exports.TemplateResultType = n;
  _exports.isPrimitive = i;
});
