// lit-html/lit-html.js
define(["exports"], function (_exports) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.svg = _exports.render = _exports.nothing = _exports.noChange = _exports.html = _exports._$LH = void 0;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e3) { throw _e3; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e4) { didErr = true; err = _e4; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

  function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

  function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

  /**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var t, i;

  var s = globalThis.trustedTypes,
      e = s ? s.createPolicy("lit-html", {
    createHTML: function createHTML(t) {
      return t;
    }
  }) : void 0,
      o = "lit$".concat((Math.random() + "").slice(9), "$"),
      n = "?" + o,
      l = "<".concat(n, ">"),
      h = document,
      r = function r() {
    var t = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "";
    return h.createComment(t);
  },
      d = function d(t) {
    return null === t || "object" != babelHelpers.typeof(t) && "function" != typeof t;
  },
      u = Array.isArray,
      v = function v(t) {
    var i;
    return u(t) || "function" == typeof (null === (i = t) || void 0 === i ? void 0 : i[Symbol.iterator]);
  },
      c = /<(?:(!--|\/[^a-zA-Z])|(\/?[a-zA-Z][^>\s]*)|(\/?$))/g,
      a = /-->/g,
      f = />/g,
      _ = />|[ 	\n\r](?:([^\s"'>=/]+)([ 	\n\r]*=[ 	\n\r]*(?:[^ 	\n\r"'`<>=]|("|')|))|$)/g,
      g = /'/g,
      m = /"/g,
      $ = /^(?:script|style|textarea)$/i,
      p = function p(t) {
    return function (i) {
      for (var _len = arguments.length, s = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        s[_key - 1] = arguments[_key];
      }

      return {
        _$litType$: t,
        strings: i,
        values: s
      };
    };
  },
      y = p(1),
      b = p(2),
      T = Symbol.for("lit-noChange"),
      x = Symbol.for("lit-nothing"),
      w = new WeakMap(),
      A = function A(t, i, s) {
    var e, o;
    var n = null !== (e = null == s ? void 0 : s.renderBefore) && void 0 !== e ? e : i;
    var l = n._$litPart$;

    if (void 0 === l) {
      var _t = null !== (o = null == s ? void 0 : s.renderBefore) && void 0 !== o ? o : null;

      n._$litPart$ = l = new S(i.insertBefore(r(), _t), _t, void 0, null != s ? s : {});
    }

    return l._$AI(t), l;
  },
      C = h.createTreeWalker(h, 129, null, !1),
      P = function P(t, i) {
    var s = t.length - 1,
        n = [];
    var h,
        r = 2 === i ? "<svg>" : "",
        d = c;

    for (var _i = 0; _i < s; _i++) {
      var _s = t[_i];

      var _e = void 0,
          _u = void 0,
          _v = -1,
          _p = 0;

      for (; _p < _s.length && (d.lastIndex = _p, _u = d.exec(_s), null !== _u);) {
        _p = d.lastIndex, d === c ? "!--" === _u[1] ? d = a : void 0 !== _u[1] ? d = f : void 0 !== _u[2] ? ($.test(_u[2]) && (h = RegExp("</" + _u[2], "g")), d = _) : void 0 !== _u[3] && (d = _) : d === _ ? ">" === _u[0] ? (d = null != h ? h : c, _v = -1) : void 0 === _u[1] ? _v = -2 : (_v = d.lastIndex - _u[2].length, _e = _u[1], d = void 0 === _u[3] ? _ : '"' === _u[3] ? m : g) : d === m || d === g ? d = _ : d === a || d === f ? d = c : (d = _, h = void 0);
      }

      var _y = d === _ && t[_i + 1].startsWith("/>") ? " " : "";

      r += d === c ? _s + l : _v >= 0 ? (n.push(_e), _s.slice(0, _v) + "$lit$" + _s.slice(_v) + o + _y) : _s + o + (-2 === _v ? (n.push(void 0), _i) : _y);
    }

    var u = r + (t[s] || "<?>") + (2 === i ? "</svg>" : "");
    return [void 0 !== e ? e.createHTML(u) : u, n];
  };

  _exports.render = A;
  _exports.nothing = x;
  _exports.noChange = T;
  _exports.svg = b;
  _exports.html = y;

  var V = /*#__PURE__*/function () {
    function V(_ref, e) {
      var t = _ref.strings,
          i = _ref._$litType$;
      babelHelpers.classCallCheck(this, V);
      var l;
      this.parts = [];
      var h = 0,
          d = 0;

      var u = t.length - 1,
          v = this.parts,
          _P = P(t, i),
          _P2 = babelHelpers.slicedToArray(_P, 2),
          c = _P2[0],
          a = _P2[1];

      if (this.el = V.createElement(c, e), C.currentNode = this.el.content, 2 === i) {
        var _t2 = this.el.content,
            _i2 = _t2.firstChild;
        _i2.remove(), _t2.append.apply(_t2, babelHelpers.toConsumableArray(_i2.childNodes));
      }

      for (; null !== (l = C.nextNode()) && v.length < u;) {
        if (1 === l.nodeType) {
          if (l.hasAttributes()) {
            var _t3 = [];

            var _iterator = _createForOfIteratorHelper(l.getAttributeNames()),
                _step;

            try {
              for (_iterator.s(); !(_step = _iterator.n()).done;) {
                var _i5 = _step.value;

                if (_i5.endsWith("$lit$") || _i5.startsWith(o)) {
                  var _s2 = a[d++];

                  if (_t3.push(_i5), void 0 !== _s2) {
                    var _t5 = l.getAttribute(_s2.toLowerCase() + "$lit$").split(o),
                        _i6 = /([.?@])?(.*)/.exec(_s2);

                    v.push({
                      type: 1,
                      index: h,
                      name: _i6[2],
                      strings: _t5,
                      ctor: "." === _i6[1] ? k : "?" === _i6[1] ? H : "@" === _i6[1] ? I : M
                    });
                  } else v.push({
                    type: 6,
                    index: h
                  });
                }
              }
            } catch (err) {
              _iterator.e(err);
            } finally {
              _iterator.f();
            }

            for (var _i3 = 0, _t4 = _t3; _i3 < _t4.length; _i3++) {
              var _i4 = _t4[_i3];
              l.removeAttribute(_i4);
            }
          }

          if ($.test(l.tagName)) {
            var _t6 = l.textContent.split(o),
                _i7 = _t6.length - 1;

            if (_i7 > 0) {
              l.textContent = s ? s.emptyScript : "";

              for (var _s3 = 0; _s3 < _i7; _s3++) {
                l.append(_t6[_s3], r()), C.nextNode(), v.push({
                  type: 2,
                  index: ++h
                });
              }

              l.append(_t6[_i7], r());
            }
          }
        } else if (8 === l.nodeType) if (l.data === n) v.push({
          type: 2,
          index: h
        });else {
          var _t7 = -1;

          for (; -1 !== (_t7 = l.data.indexOf(o, _t7 + 1));) {
            v.push({
              type: 7,
              index: h
            }), _t7 += o.length - 1;
          }
        }

        h++;
      }
    }

    babelHelpers.createClass(V, null, [{
      key: "createElement",
      value: function createElement(t, i) {
        var s = h.createElement("template");
        return s.innerHTML = t, s;
      }
    }]);
    return V;
  }();

  function E(t, i) {
    var s = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : t;
    var e = arguments.length > 3 ? arguments[3] : undefined;
    var o, n, l, h;
    if (i === T) return i;
    var r = void 0 !== e ? null === (o = s._$Cl) || void 0 === o ? void 0 : o[e] : s._$Cu;
    var u = d(i) ? void 0 : i._$litDirective$;
    return (null == r ? void 0 : r.constructor) !== u && (null === (n = null == r ? void 0 : r._$AO) || void 0 === n || n.call(r, !1), void 0 === u ? r = void 0 : (r = new u(t), r._$AT(t, s, e)), void 0 !== e ? (null !== (l = (h = s)._$Cl) && void 0 !== l ? l : h._$Cl = [])[e] = r : s._$Cu = r), void 0 !== r && (i = E(t, r._$AS(t, i.values), r, e)), i;
  }

  var N = /*#__PURE__*/function () {
    function N(t, i) {
      babelHelpers.classCallCheck(this, N);
      this.v = [], this._$AN = void 0, this._$AD = t, this._$AM = i;
    }

    babelHelpers.createClass(N, [{
      key: "parentNode",
      get: function get() {
        return this._$AM.parentNode;
      }
    }, {
      key: "_$AU",
      get: function get() {
        return this._$AM._$AU;
      }
    }, {
      key: "p",
      value: function p(t) {
        var i;
        var _this$_$AD = this._$AD,
            s = _this$_$AD.el.content,
            e = _this$_$AD.parts,
            o = (null !== (i = null == t ? void 0 : t.creationScope) && void 0 !== i ? i : h).importNode(s, !0);
        C.currentNode = o;
        var n = C.nextNode(),
            l = 0,
            r = 0,
            d = e[0];

        for (; void 0 !== d;) {
          if (l === d.index) {
            var _i8 = void 0;

            2 === d.type ? _i8 = new S(n, n.nextSibling, this, t) : 1 === d.type ? _i8 = new d.ctor(n, d.name, d.strings, this, t) : 6 === d.type && (_i8 = new L(n, this, t)), this.v.push(_i8), d = e[++r];
          }

          l !== (null == d ? void 0 : d.index) && (n = C.nextNode(), l++);
        }

        return o;
      }
    }, {
      key: "m",
      value: function m(t) {
        var i = 0;

        var _iterator2 = _createForOfIteratorHelper(this.v),
            _step2;

        try {
          for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
            var _s4 = _step2.value;
            void 0 !== _s4 && (void 0 !== _s4.strings ? (_s4._$AI(t, _s4, i), i += _s4.strings.length - 2) : _s4._$AI(t[i])), i++;
          }
        } catch (err) {
          _iterator2.e(err);
        } finally {
          _iterator2.f();
        }
      }
    }]);
    return N;
  }();

  var S = /*#__PURE__*/function () {
    function S(t, i, s, e) {
      babelHelpers.classCallCheck(this, S);
      var o;
      this.type = 2, this._$AH = x, this._$AN = void 0, this._$AA = t, this._$AB = i, this._$AM = s, this.options = e, this._$Cg = null === (o = null == e ? void 0 : e.isConnected) || void 0 === o || o;
    }

    babelHelpers.createClass(S, [{
      key: "_$AU",
      get: function get() {
        var t, i;
        return null !== (i = null === (t = this._$AM) || void 0 === t ? void 0 : t._$AU) && void 0 !== i ? i : this._$Cg;
      }
    }, {
      key: "parentNode",
      get: function get() {
        var t = this._$AA.parentNode;
        var i = this._$AM;
        return void 0 !== i && 11 === t.nodeType && (t = i.parentNode), t;
      }
    }, {
      key: "startNode",
      get: function get() {
        return this._$AA;
      }
    }, {
      key: "endNode",
      get: function get() {
        return this._$AB;
      }
    }, {
      key: "_$AI",
      value: function _$AI(t) {
        var i = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this;
        t = E(this, t, i), d(t) ? t === x || null == t || "" === t ? (this._$AH !== x && this._$AR(), this._$AH = x) : t !== this._$AH && t !== T && this.$(t) : void 0 !== t._$litType$ ? this.T(t) : void 0 !== t.nodeType ? this.S(t) : v(t) ? this.M(t) : this.$(t);
      }
    }, {
      key: "A",
      value: function A(t) {
        var i = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this._$AB;
        return this._$AA.parentNode.insertBefore(t, i);
      }
    }, {
      key: "S",
      value: function S(t) {
        this._$AH !== t && (this._$AR(), this._$AH = this.A(t));
      }
    }, {
      key: "$",
      value: function $(t) {
        this._$AH !== x && d(this._$AH) ? this._$AA.nextSibling.data = t : this.S(h.createTextNode(t)), this._$AH = t;
      }
    }, {
      key: "T",
      value: function T(t) {
        var i;
        var s = t.values,
            e = t._$litType$,
            o = "number" == typeof e ? this._$AC(t) : (void 0 === e.el && (e.el = V.createElement(e.h, this.options)), e);
        if ((null === (i = this._$AH) || void 0 === i ? void 0 : i._$AD) === o) this._$AH.m(s);else {
          var _t8 = new N(o, this),
              _i9 = _t8.p(this.options);

          _t8.m(s), this.S(_i9), this._$AH = _t8;
        }
      }
    }, {
      key: "_$AC",
      value: function _$AC(t) {
        var i = w.get(t.strings);
        return void 0 === i && w.set(t.strings, i = new V(t)), i;
      }
    }, {
      key: "M",
      value: function M(t) {
        u(this._$AH) || (this._$AH = [], this._$AR());
        var i = this._$AH;
        var s,
            e = 0;

        var _iterator3 = _createForOfIteratorHelper(t),
            _step3;

        try {
          for (_iterator3.s(); !(_step3 = _iterator3.n()).done;) {
            var _o = _step3.value;
            e === i.length ? i.push(s = new S(this.A(r()), this.A(r()), this, this.options)) : s = i[e], s._$AI(_o), e++;
          }
        } catch (err) {
          _iterator3.e(err);
        } finally {
          _iterator3.f();
        }

        e < i.length && (this._$AR(s && s._$AB.nextSibling, e), i.length = e);
      }
    }, {
      key: "_$AR",
      value: function _$AR() {
        var t = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this._$AA.nextSibling;
        var i = arguments.length > 1 ? arguments[1] : undefined;
        var s;

        for (null === (s = this._$AP) || void 0 === s || s.call(this, !1, !0, i); t && t !== this._$AB;) {
          var _i10 = t.nextSibling;
          t.remove(), t = _i10;
        }
      }
    }, {
      key: "setConnected",
      value: function setConnected(t) {
        var i;
        void 0 === this._$AM && (this._$Cg = t, null === (i = this._$AP) || void 0 === i || i.call(this, t));
      }
    }]);
    return S;
  }();

  var M = /*#__PURE__*/function () {
    function M(t, i, s, e, o) {
      babelHelpers.classCallCheck(this, M);
      this.type = 1, this._$AH = x, this._$AN = void 0, this.element = t, this.name = i, this._$AM = e, this.options = o, s.length > 2 || "" !== s[0] || "" !== s[1] ? (this._$AH = Array(s.length - 1).fill(new String()), this.strings = s) : this._$AH = x;
    }

    babelHelpers.createClass(M, [{
      key: "tagName",
      get: function get() {
        return this.element.tagName;
      }
    }, {
      key: "_$AU",
      get: function get() {
        return this._$AM._$AU;
      }
    }, {
      key: "_$AI",
      value: function _$AI(t) {
        var i = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this;
        var s = arguments.length > 2 ? arguments[2] : undefined;
        var e = arguments.length > 3 ? arguments[3] : undefined;
        var o = this.strings;
        var n = !1;
        if (void 0 === o) t = E(this, t, i, 0), n = !d(t) || t !== this._$AH && t !== T, n && (this._$AH = t);else {
          var _e2 = t;

          var _l, _h;

          for (t = o[0], _l = 0; _l < o.length - 1; _l++) {
            _h = E(this, _e2[s + _l], i, _l), _h === T && (_h = this._$AH[_l]), n || (n = !d(_h) || _h !== this._$AH[_l]), _h === x ? t = x : t !== x && (t += (null != _h ? _h : "") + o[_l + 1]), this._$AH[_l] = _h;
          }
        }
        n && !e && this.k(t);
      }
    }, {
      key: "k",
      value: function k(t) {
        t === x ? this.element.removeAttribute(this.name) : this.element.setAttribute(this.name, null != t ? t : "");
      }
    }]);
    return M;
  }();

  var k = /*#__PURE__*/function (_M) {
    babelHelpers.inherits(k, _M);

    var _super = _createSuper(k);

    function k() {
      var _this;

      babelHelpers.classCallCheck(this, k);
      _this = _super.apply(this, arguments), _this.type = 3;
      return _this;
    }

    babelHelpers.createClass(k, [{
      key: "k",
      value: function k(t) {
        this.element[this.name] = t === x ? void 0 : t;
      }
    }]);
    return k;
  }(M);

  var H = /*#__PURE__*/function (_M2) {
    babelHelpers.inherits(H, _M2);

    var _super2 = _createSuper(H);

    function H() {
      var _this2;

      babelHelpers.classCallCheck(this, H);
      _this2 = _super2.apply(this, arguments), _this2.type = 4;
      return _this2;
    }

    babelHelpers.createClass(H, [{
      key: "k",
      value: function k(t) {
        t && t !== x ? this.element.setAttribute(this.name, "") : this.element.removeAttribute(this.name);
      }
    }]);
    return H;
  }(M);

  var I = /*#__PURE__*/function (_M3) {
    babelHelpers.inherits(I, _M3);

    var _super3 = _createSuper(I);

    function I(t, i, s, e, o) {
      var _this3;

      babelHelpers.classCallCheck(this, I);
      _this3 = _super3.call(this, t, i, s, e, o), _this3.type = 5;
      return _this3;
    }

    babelHelpers.createClass(I, [{
      key: "_$AI",
      value: function _$AI(t) {
        var i = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this;
        var s;
        if ((t = null !== (s = E(this, t, i, 0)) && void 0 !== s ? s : x) === T) return;
        var e = this._$AH,
            o = t === x && e !== x || t.capture !== e.capture || t.once !== e.once || t.passive !== e.passive,
            n = t !== x && (e === x || o);
        o && this.element.removeEventListener(this.name, this, e), n && this.element.addEventListener(this.name, this, t), this._$AH = t;
      }
    }, {
      key: "handleEvent",
      value: function handleEvent(t) {
        var i, s;
        "function" == typeof this._$AH ? this._$AH.call(null !== (s = null === (i = this.options) || void 0 === i ? void 0 : i.host) && void 0 !== s ? s : this.element, t) : this._$AH.handleEvent(t);
      }
    }]);
    return I;
  }(M);

  var L = /*#__PURE__*/function () {
    function L(t, i, s) {
      babelHelpers.classCallCheck(this, L);
      this.element = t, this.type = 6, this._$AN = void 0, this._$AM = i, this.options = s;
    }

    babelHelpers.createClass(L, [{
      key: "_$AU",
      get: function get() {
        return this._$AM._$AU;
      }
    }, {
      key: "_$AI",
      value: function _$AI(t) {
        E(this, t);
      }
    }]);
    return L;
  }();

  var R = {
    P: "$lit$",
    V: o,
    L: n,
    I: 1,
    N: P,
    R: N,
    D: v,
    j: E,
    H: S,
    O: M,
    F: H,
    B: I,
    W: k,
    Z: L
  };
  _exports._$LH = R;
  null === (t = globalThis.litHtmlPolyfillSupport) || void 0 === t || t.call(globalThis, V, S), (null !== (i = globalThis.litHtmlVersions) && void 0 !== i ? i : globalThis.litHtmlVersions = []).push("2.0.0");
});
