// lit-html/private-ssr-support.js
define(["exports", "./lit-html.js"], function (_exports, _litHtml) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports._$LH = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * @license
   * Copyright 2019 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var r = {
    boundAttributeSuffix: _litHtml._$LH.P,
    marker: _litHtml._$LH.L,
    markerMatch: _litHtml._$LH.V,
    HTML_RESULT: _litHtml._$LH.I,
    getTemplateHtml: _litHtml._$LH.N,
    overrideDirectiveResolve: function overrideDirectiveResolve(t, e) {
      return /*#__PURE__*/function (_t) {
        babelHelpers.inherits(_class, _t);

        var _super = _createSuper(_class);

        function _class() {
          babelHelpers.classCallCheck(this, _class);
          return _super.apply(this, arguments);
        }

        babelHelpers.createClass(_class, [{
          key: "_$AS",
          value: function _$AS(t, r) {
            return e(this, r);
          }
        }]);
        return _class;
      }(t);
    },
    setDirectiveClass: function setDirectiveClass(t, e) {
      t._$litDirective$ = e;
    },
    getAttributePartCommittedValue: function getAttributePartCommittedValue(t, r, i) {
      var a = _litHtml.noChange;
      return t.C = function (t) {
        return a = t;
      }, t._$AI(r, t, i), a;
    },
    connectedDisconnectable: function connectedDisconnectable(t) {
      return _objectSpread(_objectSpread({}, t), {}, {
        _$AU: !0
      });
    },
    resolveDirective: _litHtml._$LH.j,
    AttributePart: _litHtml._$LH.O,
    PropertyPart: _litHtml._$LH.W,
    BooleanAttributePart: _litHtml._$LH.F,
    EventPart: _litHtml._$LH.B,
    ElementPart: _litHtml._$LH.Z
  };
  _exports._$LH = r;
});
