// lit-html/private-ssr-support.js
define(["exports", "./lit-html.js"], function (_exports, _litHtml) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports._Σ = void 0;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * @license
   * Copyright 2019 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var r = {
    boundAttributeSuffix: _litHtml._Σ.Z,
    marker: _litHtml._Σ.U,
    markerMatch: _litHtml._Σ.Y,
    HTML_RESULT: _litHtml._Σ.q,
    getTemplateHtml: _litHtml._Σ.X,
    overrideDirectiveResolve: function overrideDirectiveResolve(t, e) {
      return /*#__PURE__*/function (_t) {
        babelHelpers.inherits(_class, _t);

        var _super = _createSuper(_class);

        function _class() {
          babelHelpers.classCallCheck(this, _class);
          return _super.apply(this, arguments);
        }

        babelHelpers.createClass(_class, [{
          key: "S",
          value: function S(t, r) {
            return e(this, r);
          }
        }]);
        return _class;
      }(t);
    },
    getAttributePartCommittedValue: function getAttributePartCommittedValue(t, r, a) {
      var i = _litHtml.noChange;
      return t.W = function (t) {
        return i = t;
      }, t.I(r, t, a), i;
    },
    resolveDirective: _litHtml._Σ.st,
    AttributePart: _litHtml._Σ.ot,
    PropertyPart: _litHtml._Σ.lt,
    BooleanAttributePart: _litHtml._Σ.nt,
    EventPart: _litHtml._Σ.rt,
    ElementPart: _litHtml._Σ.ht
  };
  _exports._Σ = r;
});
