// lit-html/static.js
define(["exports", "./lit-html.js"], function (_exports, _litHtml) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.withStatic = _exports.unsafeStatic = _exports.svg = _exports.literal = _exports.html = void 0;

  /**
   * @license
   * Copyright 2020 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var o = Symbol.for(""),
      e = function e(t) {
    var r, e;
    if ((null === (r = t) || void 0 === r ? void 0 : r.r) === o) return null === (e = t) || void 0 === e ? void 0 : e._$litStatic$;
  },
      i = function i(t) {
    return {
      _$litStatic$: t,
      r: o
    };
  },
      l = function l(t) {
    for (var _len = arguments.length, r = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      r[_key - 1] = arguments[_key];
    }

    return {
      _$litStatic$: r.reduce(function (r, o, e) {
        return r + function (t) {
          if (void 0 !== t._$litStatic$) return t._$litStatic$;
          throw Error("Value passed to 'literal' function must be a 'literal' result: ".concat(t, ". Use 'unsafeStatic' to pass non-literal values, but\n            take care to ensure page security."));
        }(o) + t[e + 1];
      }, t[0]),
      r: o
    };
  },
      a = new Map(),
      s = function s(t) {
    return function (r) {
      for (var _len2 = arguments.length, o = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
        o[_key2 - 1] = arguments[_key2];
      }

      var i = o.length;
      var l, s;
      var n = [],
          u = [];
      var c,
          v = 0,
          $ = !1;

      for (; v < i;) {
        for (c = r[v]; v < i && void 0 !== (s = o[v], l = e(s));) {
          c += l + r[++v], $ = !0;
        }

        u.push(s), n.push(c), v++;
      }

      if (v === i && n.push(r[i]), $) {
        var _t = n.join("$$lit$$");

        void 0 === (r = a.get(_t)) && (n.raw = n, a.set(_t, r = n)), o = u;
      }

      return t.apply(void 0, [r].concat(babelHelpers.toConsumableArray(o)));
    };
  },
      n = s(_litHtml.html),
      u = s(_litHtml.svg);

  _exports.svg = u;
  _exports.html = n;
  _exports.withStatic = s;
  _exports.literal = l;
  _exports.unsafeStatic = i;
});
