// lit-html/static.js
define(["exports", "./lit-html.js"], function (_exports, _litHtml) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.withStatic = _exports.unsafeStatic = _exports.svg = _exports.literal = _exports.html = void 0;

  /**
   * @license
   * Copyright 2020 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var o = function o(t) {
    return {
      _$litStatic$: t
    };
  },
      r = function r(t) {
    for (var _len = arguments.length, e = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      e[_key - 1] = arguments[_key];
    }

    return {
      _$litStatic$: e.reduce(function (e, o, r) {
        return e + function (t) {
          if (void 0 !== t._$litStatic$) return t._$litStatic$;
          throw Error("Value passed to 'literal' function must be a 'literal' result: ".concat(t, ". Use 'unsafeStatic' to pass non-literal values, but\n            take care to ensure page security."));
        }(o) + t[r + 1];
      }, t[0])
    };
  },
      i = new Map(),
      a = function a(t) {
    return function (e) {
      for (var _len2 = arguments.length, o = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
        o[_key2 - 1] = arguments[_key2];
      }

      var r;
      var a = o.length;
      var l, s;
      var n = [],
          u = [];
      var c,
          $ = 0,
          v = !1;

      for (; $ < a;) {
        for (c = e[$]; $ < a && void 0 !== (s = o[$], l = null === (r = s) || void 0 === r ? void 0 : r._$litStatic$);) {
          c += l + e[++$], v = !0;
        }

        u.push(s), n.push(c), $++;
      }

      if ($ === a && n.push(e[a]), v) {
        var _t = n.join("$$lit$$");

        void 0 === (e = i.get(_t)) && (n.raw = n, i.set(_t, e = n)), o = u;
      }

      return t.apply(void 0, [e].concat(babelHelpers.toConsumableArray(o)));
    };
  },
      l = a(_litHtml.html),
      s = a(_litHtml.svg);

  _exports.svg = s;
  _exports.html = l;
  _exports.withStatic = a;
  _exports.literal = r;
  _exports.unsafeStatic = o;
});
