// @lrnwebcomponents/map-menu/lib/map-menu-builder.js
define(["exports", "require", "../../../lit-element/lit-element.js"], function (_exports, _require, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.MapMenuBuilder = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject5_46b9e5c0dbf511ea82ad236b71549975() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                    <map-menu-item\n                      title=\"", "\"\n                      id=\"", "\"\n                      url=\"", "\"\n                      icon=\"", "\"\n                      track-icon=\"", "\"\n                      active-path=\"", "\"\n                      selected=\"", "\"\n                    ></map-menu-item>\n                  "]);

    _templateObject5_46b9e5c0dbf511ea82ad236b71549975 = function _templateObject5_46b9e5c0dbf511ea82ad236b71549975() {
      return data;
    };

    return data;
  }

  function _templateObject4_46b9e5c0dbf511ea82ad236b71549975() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                    <map-menu-submenu\n                      title=\"", "\"\n                      id=\"", "\"\n                      url=\"", "\"\n                      icon=\"", "\"\n                      open=\"", "\"\n                      avatar-label=\"", "\"\n                      selected=\"", "\"\n                    >\n                      <map-menu-builder\n                        .items=\"", "\"\n                        selected=\"", "\"\n                      ></map-menu-builder>\n                    </map-menu-submenu>\n                  "]);

    _templateObject4_46b9e5c0dbf511ea82ad236b71549975 = function _templateObject4_46b9e5c0dbf511ea82ad236b71549975() {
      return data;
    };

    return data;
  }

  function _templateObject3_46b9e5c0dbf511ea82ad236b71549975() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              ", "\n            "]);

    _templateObject3_46b9e5c0dbf511ea82ad236b71549975 = function _templateObject3_46b9e5c0dbf511ea82ad236b71549975() {
      return data;
    };

    return data;
  }

  function _templateObject2_46b9e5c0dbf511ea82ad236b71549975() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      ", "\n    "]);

    _templateObject2_46b9e5c0dbf511ea82ad236b71549975 = function _templateObject2_46b9e5c0dbf511ea82ad236b71549975() {
      return data;
    };

    return data;
  }

  function _templateObject_46b9e5c0dbf511ea82ad236b71549975() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n      "]);

    _templateObject_46b9e5c0dbf511ea82ad236b71549975 = function _templateObject_46b9e5c0dbf511ea82ad236b71549975() {
      return data;
    };

    return data;
  }

  var MapMenuBuilder =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(MapMenuBuilder, _LitElement);
    babelHelpers.createClass(MapMenuBuilder, null, [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject_46b9e5c0dbf511ea82ad236b71549975())];
      }
    }]);

    function MapMenuBuilder() {
      var _this;

      babelHelpers.classCallCheck(this, MapMenuBuilder);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(MapMenuBuilder).call(this));
      _this.items = [];
      new Promise(function (res, rej) {
        return _require.default(["./map-menu-submenu.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["./map-menu-item.js"], res, rej);
      });
      return _this;
    }
    /**
     * LitElement life cycle - render
     */


    babelHelpers.createClass(MapMenuBuilder, [{
      key: "render",
      value: function render() {
        var _this2 = this;

        return (0, _litElement.html)(_templateObject2_46b9e5c0dbf511ea82ad236b71549975(), this.items ? this.items.map(function (item) {
          return (0, _litElement.html)(_templateObject3_46b9e5c0dbf511ea82ad236b71549975(), item.children.length > 0 ? (0, _litElement.html)(_templateObject4_46b9e5c0dbf511ea82ad236b71549975(), item.title, item.id, item.slug, item.metadata && item.metadata.icon ? item.metadata.icon : "", item.metadata && item.metadata.active ? item.metadata.active : false, item.metadata && item.metadata.avatarLabel ? item.metadata.avatarLabel : "", _this2.selected, item.children, _this2.selected) : (0, _litElement.html)(_templateObject5_46b9e5c0dbf511ea82ad236b71549975(), item.title, item.id, item.slug, item.metadata && item.metadata.icon ? item.metadata.icon : "", item.metadata && item.metadata.accessData && item.metadata.accessData.trackIcon ? item.metadata.accessData.trackIcon : "", _this2.activePath, _this2.selected));
        }) : "");
      }
    }, {
      key: "createRenderRoot",
      value: function createRenderRoot() {
        return this;
      }
    }], [{
      key: "tag",
      get: function get() {
        return "map-menu-builder";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          items: {
            type: Array
          },
          selected: {
            type: String
          }
        };
      }
    }]);
    return MapMenuBuilder;
  }(_litElement.LitElement);

  _exports.MapMenuBuilder = MapMenuBuilder;
  window.customElements.define(MapMenuBuilder.tag, MapMenuBuilder);
});
