// @lrnwebcomponents/map-menu/lib/map-menu-header.js
define(["exports", "require", "../../../lit-element/lit-element.js", "../../simple-icon/lib/simple-icon-lite.js", "../../simple-icon/lib/simple-icons.js", "../../utils/utils.js"], function (_exports, _require, _litElement, _simpleIconLite, _simpleIcons, _utils) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.MapMenuHeader = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject4_f1e3b960544211ebad6c77197a564a47() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          transition: 0.1s all ease-in-out;\n          --map-menu-item-height: 24px;\n        }\n        :host([active]) {\n          background: var(--map-menu-active-color);\n          color: var(--map-menu-item-active-item-color, #000000);\n        }\n        #container {\n          display: flex;\n          align-items: center;\n        }\n\n        .avatarlabel {\n          margin-right: 10px;\n        }\n\n        #center {\n          flex: 1 1 auto;\n        }\n        a,\n        a:visited {\n          display: block;\n          color: var(--map-menu-item-a-color, inherit);\n          text-decoration: var(--map-menu-header-a-text-decoration, none);\n        }\n        a:hover,\n        a:active,\n        a:focus {\n          color: var(\n            --map-menu-item-a-active-color,\n            var(--map-menu-item-a-color, inherit)\n          );\n          text-decoration: var(\n            --map-menu-header-a-text-decoration-hover,\n            underline\n          );\n        }\n        lrndesign-avatar {\n          display: inline-block;\n          background: #fff;\n          border-radius: 50%;\n          box-shadow: 0 1px 1px rgba(0, 0, 0, 0.3);\n          padding: 2px;\n          position: relative;\n          margin-top: -2px;\n          transform: translateY(2px);\n        }\n\n        #link {\n          display: flex;\n          justify-content: flex-start;\n          align-items: flex-start;\n          flex-direction: column;\n        }\n\n        #right simple-icon-lite {\n          display: inline-block;\n          color: gray;\n        }\n\n        simple-icon-lite {\n          color: inherit;\n          display: inline-block;\n          --simple-icon-height: var(--map-menu-item-height);\n          --simple-icon-width: var(--map-menu-item-height);\n        }\n\n        button {\n          color: inherit;\n          background-color: transparent;\n          text-transform: none;\n          width: 100%;\n          justify-content: left;\n          margin: 0px;\n          border: 0;\n          height: 40px;\n          padding: 4px;\n          text-align: left;\n          border-radius: 0;\n        }\n      "]);

    _templateObject4_f1e3b960544211ebad6c77197a564a47 = function _templateObject4_f1e3b960544211ebad6c77197a564a47() {
      return data;
    };

    return data;
  }

  function _templateObject3_f1e3b960544211ebad6c77197a564a47() {
    var data = babelHelpers.taggedTemplateLiteral([" <simple-icon-lite icon=\"", "\"></simple-icon-lite> "]);

    _templateObject3_f1e3b960544211ebad6c77197a564a47 = function _templateObject3_f1e3b960544211ebad6c77197a564a47() {
      return data;
    };

    return data;
  }

  function _templateObject2_f1e3b960544211ebad6c77197a564a47() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <div class=\"avatarlabel\">\n                <lrndesign-avatar\n                  label=\"", "\"\n                ></lrndesign-avatar>\n              </div>\n            "]);

    _templateObject2_f1e3b960544211ebad6c77197a564a47 = function _templateObject2_f1e3b960544211ebad6c77197a564a47() {
      return data;
    };

    return data;
  }

  function _templateObject_f1e3b960544211ebad6c77197a564a47() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div id=\"container\">\n        ", "\n        ", "\n\n        <div id=\"center\">\n          <a tabindex=\"-1\" href=\"", "\">\n            <button class=\"title\" noink>\n              <div id=\"label\">", "</div>\n              <div class=\"title\">", "</div>\n            </button>\n          </a>\n        </div>\n      </div>\n    "]);

    _templateObject_f1e3b960544211ebad6c77197a564a47 = function _templateObject_f1e3b960544211ebad6c77197a564a47() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  var MapMenuHeader = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(MapMenuHeader, _LitElement);

    var _super = _createSuper(MapMenuHeader);

    babelHelpers.createClass(MapMenuHeader, [{
      key: "render",

      /**
       * LitElement life cycle - render callback
       */
      value: function render() {
        return (0, _litElement.html)(_templateObject_f1e3b960544211ebad6c77197a564a47(), this.avatarLabel ? (0, _litElement.html)(_templateObject2_f1e3b960544211ebad6c77197a564a47(), this.avatarLabel) : "", this.icon ? (0, _litElement.html)(_templateObject3_f1e3b960544211ebad6c77197a564a47(), this.icon) : "", this.url, this.label, this.title);
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject4_f1e3b960544211ebad6c77197a564a47())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "map-menu-header";
      }
      /**
       * HTMLElement
       */

    }]);

    function MapMenuHeader() {
      var _this;

      babelHelpers.classCallCheck(this, MapMenuHeader);
      _this = _super.call(this);
      _this.avatarLabel = "";
      _this.icon = null;
      _this.url = "";
      _this.opened = false;
      _this.active = false;
      setTimeout(function () {
        _this.addEventListener("click", _this.__tap.bind(babelHelpers.assertThisInitialized(_this)));

        _this.addEventListener("keypress", _this.__keypress.bind(babelHelpers.assertThisInitialized(_this)));
      }, 0);
      return _this;
    }
    /**
     * LitElement life cycle - properties changed callback
     */


    babelHelpers.createClass(MapMenuHeader, [{
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "opened") {
            _this2._openedChanged(_this2[propName], oldValue);
          }

          if (["id", "selected"].includes(propName)) {
            _this2.__selectedChanged(_this2.selected, _this2.id);
          } // only import avatar if we absolutely need it since its heavy


          if (propName == "avatarLabel" && _this2[propName] != "") {
            new Promise(function (res, rej) {
              return _require.default(["../../lrndesign-avatar/lrndesign-avatar.js"], res, rej);
            });
          }
        });
      }
      /**
       * LitElement life cycle - properties definition
       */

    }, {
      key: "_openedChanged",
      value: function _openedChanged(newValue, oldValue) {
        if (newValue) {
          this.__collapseIcon = "icons:expand-more";
          this.__collapseAria = "collapse menu";
        } else {
          this.__collapseIcon = "icons:chevron-right";
          this.__collapseAria = "expand menu";
        }
      }
    }, {
      key: "__selectedChanged",
      value: function __selectedChanged(selected, id) {
        if (selected === id) {
          this.dispatchEvent(new CustomEvent("active-item", {
            bubbles: true,
            cancelable: true,
            composed: true,
            detail: this
          }));
        }
      }
    }, {
      key: "__tap",
      value: function __tap(e) {
        // send to toggle event
        this.__toggleEventHandler(e);
      }
    }, {
      key: "__keypress",
      value: function __keypress(e) {
        // send to toggle event
        if (e.code === "Enter") {
          this.__toggleEventHandler(e);
        }
      }
    }, {
      key: "__toggleEventHandler",
      value: function __toggleEventHandler(e) {
        var target = (0, _utils.normalizeEventPath)(e)[0];

        if (e.originalTarget && e.originalTarget.id === "toggle") {
          this.dispatchEvent(new CustomEvent("toggle-header", {
            bubbles: true,
            cancelable: true,
            composed: true,
            detail: true
          }));
        } else if (target && target.id === "toggle") {
          this.dispatchEvent(new CustomEvent("toggle-header", {
            bubbles: true,
            cancelable: true,
            composed: true,
            detail: true
          }));
        }
      }
    }], [{
      key: "properties",
      get: function get() {
        return {
          title: {
            type: String
          },
          label: {
            type: String
          },
          avatarLabel: {
            type: String,
            attribute: "avatar-label"
          },
          opened: {
            type: Boolean
          },
          url: {
            type: String
          },
          id: {
            type: String,
            reflect: true
          },
          icon: {
            type: String
          },
          active: {
            type: Boolean,
            reflect: true
          },
          selected: {
            type: String,
            reflect: true
          },
          __collapseIcon: {
            type: String
          },
          __collapseAria: {
            type: String
          }
        };
      }
    }]);
    return MapMenuHeader;
  }(_litElement.LitElement);

  _exports.MapMenuHeader = MapMenuHeader;
  window.customElements.define(MapMenuHeader.tag, MapMenuHeader);
});
