// @lrnwebcomponents/map-menu/lib/map-menu-builder.js
define(["exports", "require", "../../../lit-element/lit-element.js"], function (_exports, _require, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.MapMenuBuilder = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  var _templateObject_055baf00d39311eb9590b786b63a9382, _templateObject2_055baf00d39311eb9590b786b63a9382, _templateObject3_055baf00d39311eb9590b786b63a9382, _templateObject4_055baf00d39311eb9590b786b63a9382, _templateObject5_055baf00d39311eb9590b786b63a9382;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var MapMenuBuilder = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(MapMenuBuilder, _LitElement);

    var _super = _createSuper(MapMenuBuilder);

    function MapMenuBuilder() {
      var _this;

      babelHelpers.classCallCheck(this, MapMenuBuilder);
      _this = _super.call(this);
      _this.items = [];
      new Promise(function (res, rej) {
        return _require.default(["./map-menu-submenu.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["./map-menu-item.js"], res, rej);
      });
      return _this;
    }
    /**
     * LitElement life cycle - render
     */


    babelHelpers.createClass(MapMenuBuilder, [{
      key: "render",
      value: function render() {
        var _this2 = this;

        return (0, _litElement.html)(_templateObject_055baf00d39311eb9590b786b63a9382 || (_templateObject_055baf00d39311eb9590b786b63a9382 = babelHelpers.taggedTemplateLiteral(["\n      ", "\n    "])), this.items ? this.items.map(function (item) {
          return (0, _litElement.html)(_templateObject2_055baf00d39311eb9590b786b63a9382 || (_templateObject2_055baf00d39311eb9590b786b63a9382 = babelHelpers.taggedTemplateLiteral(["\n              ", "\n            "])), item.children.length > 0 ? (0, _litElement.html)(_templateObject3_055baf00d39311eb9590b786b63a9382 || (_templateObject3_055baf00d39311eb9590b786b63a9382 = babelHelpers.taggedTemplateLiteral(["\n                    <map-menu-submenu\n                      title=\"", "\"\n                      id=\"", "\"\n                      url=\"", "\"\n                      icon=\"", "\"\n                      open=\"", "\"\n                      avatar-label=\"", "\"\n                      selected=\"", "\"\n                      .published=\"", "\"\n                    >\n                      <map-menu-builder\n                        .items=\"", "\"\n                        selected=\"", "\"\n                      ></map-menu-builder>\n                    </map-menu-submenu>\n                  "])), item.title, item.id, item.slug, item.metadata && item.metadata.icon ? item.metadata.icon : "", item.metadata && item.metadata.active ? item.metadata.active : false, item.metadata && item.metadata.avatarLabel ? item.metadata.avatarLabel : "", _this2.selected, item.metadata.published, item.children, _this2.selected) : (0, _litElement.html)(_templateObject4_055baf00d39311eb9590b786b63a9382 || (_templateObject4_055baf00d39311eb9590b786b63a9382 = babelHelpers.taggedTemplateLiteral(["\n                    <map-menu-item\n                      title=\"", "\"\n                      id=\"", "\"\n                      url=\"", "\"\n                      icon=\"", "\"\n                      track-icon=\"", "\"\n                      active-path=\"", "\"\n                      selected=\"", "\"\n                      .published=\"", "\"\n                    ></map-menu-item>\n                  "])), item.title, item.id, item.slug, item.metadata && item.metadata.icon ? item.metadata.icon : "", item.metadata && item.metadata.accessData && item.metadata.accessData.trackIcon ? item.metadata.accessData.trackIcon : "", _this2.activePath, _this2.selected, item.metadata.published));
        }) : "");
      }
    }, {
      key: "createRenderRoot",
      value: function createRenderRoot() {
        return this;
      }
    }], [{
      key: "styles",
      get:
      /**
       * LitElement constructable styles enhancement
       */
      function get() {
        return [(0, _litElement.css)(_templateObject5_055baf00d39311eb9590b786b63a9382 || (_templateObject5_055baf00d39311eb9590b786b63a9382 = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n      "])))];
      }
    }, {
      key: "tag",
      get: function get() {
        return "map-menu-builder";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          items: {
            type: Array
          },
          selected: {
            type: String
          }
        };
      }
    }]);
    return MapMenuBuilder;
  }(_litElement.LitElement);

  _exports.MapMenuBuilder = MapMenuBuilder;
  window.customElements.define(MapMenuBuilder.tag, MapMenuBuilder);
});
