// @lrnwebcomponents/elmsln-apps/lib/lrnapp-cis/lrnapp-cis.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@polymer/iron-ajax/iron-ajax.js", "../../../../@polymer/paper-item/paper-item.js", "../../../simple-modal/lib/simple-modal-template.js", "../../../../@polymer/paper-listbox/paper-listbox.js", "../../../../@polymer/paper-dropdown-menu/paper-dropdown-menu.js", "../../../../@polymer/app-layout/app-toolbar/app-toolbar.js", "../../../../@polymer/paper-dialog/paper-dialog.js", "../../../../@polymer/app-route/app-location.js", "../../../../@polymer/app-route/app-route.js", "../../../../@polymer/paper-button/paper-button.js", "../../../../@polymer/paper-toast/paper-toast.js", "../../../elmsln-loading/elmsln-loading.js", "../../../lrndesign-course-banner/lrndesign-course-banner.js", "../../../lrn-icon/lrn-icon.js", "../../../lrnsys-button/lrnsys-button.js", "../../../lrndesign-avatar/lrndesign-avatar.js", "../../../lrnsys-layout/lib/lrnsys-dialog.js", "../../../responsive-grid/lib/responsive-grid-col.js", "../../../responsive-grid/lib/responsive-grid-row.js", "../../../materializecss-styles/lib/colors.js", "./lrnapp-cis-course-card.js"], function (_exports, _litElement, _ironAjax, _paperItem, _simpleModalTemplate, _paperListbox, _paperDropdownMenu, _appToolbar, _paperDialog, _appLocation, _appRoute, _paperButton, _paperToast, _elmslnLoading, _lrndesignCourseBanner, _lrnIcon, _lrnsysButton, _lrndesignAvatar, _lrnsysDialog, _responsiveGridCol, _responsiveGridRow, _colors, _lrnappCisCourseCard) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappCis = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject17_e93fdd102da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <lrnsys-dialog id=\"confirm\">\n              <div class=\"dialog-header\" slot=\"header\">\n                Add this to the\n                <strong>", "</strong> network?\n              </div>\n              <div class=\"dialog-body\">\n                <responsive-grid-row gutter=\"5\">\n                  <responsive-grid-col\n                    xl=\"3\"\n                    lg=\"3\"\n                    md=\"3\"\n                    sm=\"3\"\n                    xs=\"3\"\n                  ></responsive-grid-col>\n                  <responsive-grid-col xl=\"1\" lg=\"1\" md=\"1\" sm=\"1\" xs=\"1\"\n                    >Add</responsive-grid-col\n                  >\n                  <responsive-grid-col xl=\"2\" lg=\"2\" md=\"2\" sm=\"2\" xs=\"2\"\n                    ><lrn-icon\n                      icon=\"", "\"\n                      class=\"", "-text elmsln-hover-icon service-confirm-icon\"\n                    ></lrn-icon\n                  ></responsive-grid-col>\n                  <responsive-grid-col xl=\"3\" lg=\"3\" md=\"3\" sm=\"3\" xs=\"3\"\n                    ><strong\n                      >", "</strong\n                    ></responsive-grid-col\n                  >\n                  <responsive-grid-col\n                    xl=\"3\"\n                    lg=\"3\"\n                    md=\"3\"\n                    sm=\"3\"\n                    xs=\"3\"\n                  ></responsive-grid-col>\n                </responsive-grid-row>\n                <responsive-grid-row gutter=\"5\">\n                  <responsive-grid-col\n                    xl=\"3\"\n                    lg=\"3\"\n                    md=\"3\"\n                    sm=\"3\"\n                    xs=\"3\"\n                  ></responsive-grid-col>\n                  <responsive-grid-col xl=\"1\" lg=\"1\" md=\"1\" sm=\"1\" xs=\"1\"\n                    >To</responsive-grid-col\n                  >\n                  <responsive-grid-col xl=\"2\" lg=\"2\" md=\"2\" sm=\"2\" xs=\"2\"\n                    ><lrndesign-avatar\n                      class=\"service-confirm-icon\"\n                      label=\"", "\"\n                      jdenticon=\"\"\n                      color=\"", " darken-4\"\n                    >\n                    </lrndesign-avatar\n                  ></responsive-grid-col>\n                  <responsive-grid-col xl=\"3\" lg=\"3\" md=\"3\" sm=\"3\" xs=\"3\"\n                    ><strong\n                      >", "</strong\n                    ></responsive-grid-col\n                  >\n                  <responsive-grid-col\n                    xl=\"3\"\n                    lg=\"3\"\n                    md=\"3\"\n                    sm=\"3\"\n                    xs=\"3\"\n                  ></responsive-grid-col>\n                </responsive-grid-row>\n                <div style=\"margin-top:1em;\">This will take a few moments.</div>\n              </div>\n              <div class=\"buttons\">\n                <paper-button\n                  raised=\"\"\n                  dialog-confirm=\"\"\n                  autofocus=\"\"\n                  @click=\"", "\"\n                  class=\"green\"\n                  >Let's do it!</paper-button\n                >\n                <paper-button dialog-dismiss=\"\" class=\"red-text\"\n                  >Oops, go back.</paper-button\n                >\n              </div>\n            </lrnsys-dialog>\n          "]);

    _templateObject17_e93fdd102da211eaa247b36e11306899 = function _templateObject17_e93fdd102da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject16_e93fdd102da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral([""]);

    _templateObject16_e93fdd102da211eaa247b36e11306899 = function _templateObject16_e93fdd102da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject15_e93fdd102da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral([""]);

    _templateObject15_e93fdd102da211eaa247b36e11306899 = function _templateObject15_e93fdd102da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject14_e93fdd102da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                          <div\n                            style=\"padding: 1em;width: 100%;margin: .5em 0;display: block;background-color:#FF2222;color:#ffffff;border: 1px solid #222222;\"\n                          >\n                            <h4>Danger zone</h4>\n                            <lrnsys-button\n                              raised=\"\"\n                              href=\"", "\"\n                              label=\"Delete\"\n                              hover-class=\"red lighten-1\"\n                              color=\"red lighten-3\"\n                              icon=\"delete\"\n                            ></lrnsys-button>\n                          </div>\n                        "]);

    _templateObject14_e93fdd102da211eaa247b36e11306899 = function _templateObject14_e93fdd102da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject13_e93fdd102da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                    <lrnsys-button\n                      raised=\"\"\n                      href=\"", "\"\n                      label=\"Edit\"\n                      hover-class=\"green lighten-4\"\n                      icon=\"create\"\n                    ></lrnsys-button>\n                    <lrnsys-button\n                      raised=\"\"\n                      href=\"", "\"\n                      label=\"Add offering\"\n                      hover-class=\"amber lighten-3\"\n                      icon=\"icons:add\"\n                    ></lrnsys-button>\n                  </template>\n                  <lrnsys-button\n                    raised=\"\"\n                    href=\"", "\"\n                    label=\"Offerings\"\n                    hover-class=\"amber lighten-4\"\n                    icon=\"social:people\"\n                  ></lrnsys-button>\n                  <lrnsys-button\n                    raised=\"\"\n                    href=\"", "\"\n                    label=\"Sync Roster\"\n                    hover-class=\"amber lighten-4\"\n                    icon=\"notification:sync\"\n                  ></lrnsys-button>\n                  <lrnsys-button\n                    raised=\"\"\n                    href=\"", "\"\n                    label=\"Course page (legacy)\"\n                    hover-class=\"brown lighten-4\"\n                    icon=\"delete\"\n                  ></lrnsys-button>\n                  ", "\n                "]);

    _templateObject13_e93fdd102da211eaa247b36e11306899 = function _templateObject13_e93fdd102da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject12_e93fdd102da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                                        <lrnsys-button\n                                          raised=\"\"\n                                          @click=\"", "\"\n                                          color=\"grey lighten-4\"\n                                          icon-class=\"grey lighten-5\"\n                                          data-machine-name=\"", "\"\n                                        >\n                                          <lrn-icon\n                                            data-machine-name=\"", "\"\n                                            icon=\"", "\"\n                                            class=\"elmsln-hover-icon\"\n                                          ></lrn-icon>\n                                          <span\n                                            data-machine-name=\"", "\"\n                                            >Make the ", "\n                                            service</span\n                                          >\n                                        </lrnsys-button>\n                                      "]);

    _templateObject12_e93fdd102da211eaa247b36e11306899 = function _templateObject12_e93fdd102da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject11_e93fdd102da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                                        <lrnsys-button\n                                          raised=\"\"\n                                          href=\"", "\"\n                                          hover-class=\"", " lighten-4\"\n                                        >\n                                          <lrn-icon\n                                            icon=\"", "\"\n                                            class=\"elmsln-hover-icon\"\n                                          ></lrn-icon>\n                                          <span>", "</span>\n                                        </lrnsys-button>\n                                      "]);

    _templateObject11_e93fdd102da211eaa247b36e11306899 = function _templateObject11_e93fdd102da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject10_e93fdd102da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                                  ", "\n                                "]);

    _templateObject10_e93fdd102da211eaa247b36e11306899 = function _templateObject10_e93fdd102da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject9_e93fdd102da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                              ", "\n                            "]);

    _templateObject9_e93fdd102da211eaa247b36e11306899 = function _templateObject9_e93fdd102da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject8_e93fdd102da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <responsive-grid-row gutter=\"5\">\n                    <responsive-grid-col xl=\"6\" lg=\"6\" md=\"6\" sm=\"12\" xs=\"12\">\n                      <div class=\"column\">\n                        <h4>Details</h4>\n                        <ul>\n                          <li\n                            ?hidden=\"", "\"\n                          >\n                            Academic unit:\n                            ", "\n                          </li>\n                          <li\n                            ?hidden=\"", "\"\n                          >\n                            Program:\n                            ", "\n                          </li>\n                        </ul>\n                        <h4>Learning Network</h4>\n                        ", "\n                      </div>\n                    </responsive-grid-col>\n                    <responsive-grid-col xl=\"6\" lg=\"6\" md=\"6\" sm=\"12\" xs=\"12\">\n                      <div class=\"column\">\n                        <h4>Operations</h4>\n                        ", "\n                      </div>\n                    </responsive-grid-col>\n                  </responsive-grid-row>\n                  <p>", "</p>\n                "]);

    _templateObject8_e93fdd102da211eaa247b36e11306899 = function _templateObject8_e93fdd102da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject7_e93fdd102da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <lrndesign-course-banner\n                    .image=\"", "\"\n                    .name=\"", "\"\n                    .title=\"", "\"\n                    .color=\"", " darken-4\"\n                  >\n                  </lrndesign-course-banner>\n                "]);

    _templateObject7_e93fdd102da211eaa247b36e11306899 = function _templateObject7_e93fdd102da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject6_e93fdd102da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <paper-button\n                data-course-id=\"", "\"\n                class=\"coursecard-wrapper\"\n                @click=\"", "\"\n              >\n                <lrnapp-cis-course-card\n                  elevation=\"2\"\n                  data-course-id=\"", "\"\n                  .name=\"", "\"\n                  .image=\"", "\"\n                  .title=\"", "\"\n                  .color=\"", "\"\n                >\n                </lrnapp-cis-course-card>\n              </paper-button>\n            "]);

    _templateObject6_e93fdd102da211eaa247b36e11306899 = function _templateObject6_e93fdd102da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject5_e93fdd102da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                <paper-item item-id=\"", "\">\n                  ", "\n                </paper-item>\n              "]);

    _templateObject5_e93fdd102da211eaa247b36e11306899 = function _templateObject5_e93fdd102da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject4_e93fdd102da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                <paper-item item-id=\"", "\">\n                  ", "\n                </paper-item>\n              "]);

    _templateObject4_e93fdd102da211eaa247b36e11306899 = function _templateObject4_e93fdd102da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject3_e93fdd102da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                <paper-item item-id=\"", "\">\n                  ", "\n                </paper-item>\n              "]);

    _templateObject3_e93fdd102da211eaa247b36e11306899 = function _templateObject3_e93fdd102da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject2_e93fdd102da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <iron-ajax\n        auto\n        .url=\"", "\"\n        handle-as=\"json\"\n        @last-response-changed=\"", "\"\n        @response=\"", "\"\n      ></iron-ajax>\n      <iron-ajax\n        .url=\"", "\"\n        handle-as=\"json\"\n        id=\"courserequest\"\n        @last-response-changed=\"", "\"\n        @response=\"", "\"\n      ></iron-ajax>\n      <iron-ajax\n        .url=\"", "\"\n        handle-as=\"json\"\n        id=\"makeservice\"\n        @last-response-changed=\"", "\"\n        @response=\"", "\"\n      ></iron-ajax>\n      <div id=\"loading\">\n        <elmsln-loading color=\"grey-text\" size=\"large\"></elmsln-loading>\n        <h3>Loading..</h3>\n      </div>\n      <app-toolbar class=\"\">\n        <span main-title=\"\"></span>\n        <span\n          top-item=\"\"\n          style=\"text-align:right;font-size:.5em;padding-right:1em;\"\n          >Displaying ", " of\n          ", "</span\n        >\n        <paper-dropdown-menu\n          label=\"Course\"\n          ?hidden=\"", "\"\n        >\n          <paper-listbox\n            slot=\"dropdown-content\"\n            class=\"dropdown-content\"\n            .selected=\"", "\"\n            @selected-changed=\"", "\"\n            attr-for-selected=\"item-id\"\n          >\n            <paper-item>-- Any --</paper-item>\n            ", "\n          </paper-listbox>\n        </paper-dropdown-menu>\n        <paper-dropdown-menu\n          label=\"Program\"\n          ?hidden=\"", "\"\n        >\n          <paper-listbox\n            slot=\"dropdown-content\"\n            class=\"dropdown-content\"\n            .selected=\"", "\"\n            @selected-changed=\"", "\"\n            attr-for-selected=\"item-id\"\n          >\n            <paper-item>-- Any --</paper-item>\n            ", "\n          </paper-listbox>\n        </paper-dropdown-menu>\n        <paper-dropdown-menu\n          label=\"Academic home\"\n          ?hidden=\"", "\"\n        >\n          <paper-listbox\n            slot=\"dropdown-content\"\n            class=\"dropdown-content\"\n            .selected=\"", "\"\n            @selected-changed=\"", "\"\n            attr-for-selected=\"item-id\"\n          >\n            <paper-item>-- Any --</paper-item>\n            ", "\n          </paper-listbox>\n        </paper-dropdown-menu>\n      </app-toolbar>\n      <div class=\"courses-grid\">\n        <div class=\"list-container\">\n          ", "\n        </div>\n      </div>\n      <app-location\n        .route=\"", "\"\n        @route-changed=\"", "\"\n        .query-params=\"", "\"\n        @query-params-changed=\"", "\"\n      ></app-location>\n      <app-route\n        .route=\"", "\"\n        @route-changed=\"", "\"\n        pattern=\"", "/:page\"\n        .tail=\"", "\"\n        @tail=\"", "\"\n        .data=\"", "\"\n        @data=\"", "\"\n        .query-params=\"", "\"\n        @query-params-changed=\"", "\"\n      >\n      </app-route>\n      <paper-dialog id=\"dialog\" with-backdrop>\n        <h2>Course details</h2>\n        <paper-dialog-scrollable>\n          <div class=\"dialog-header\">\n            ", "\n          </div>\n          <div id=\"coursedetails\">\n            ", "\n          </div>\n        </paper-dialog-scrollable>\n      </paper-dialog>\n      ", "\n      <paper-toast id=\"toast\"></paper-toast>\n    "]);

    _templateObject2_e93fdd102da211eaa247b36e11306899 = function _templateObject2_e93fdd102da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  function _templateObject_e93fdd102da211eaa247b36e11306899() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          align-content: center;\n        }\n        #loading {\n          width: 100%;\n          z-index: 1000;\n          opacity: 0.9;\n          padding: 4em 0 0 0;\n          text-align: center;\n          align-content: center;\n          justify-content: center;\n          height: 100vh;\n          position: absolute;\n          background-color: white;\n        }\n        #dialog {\n          width: 80vw;\n        }\n        iron-selector {\n          line-height: 1em;\n        }\n        iron-selector lrnsys-button {\n          display: inline-flex;\n        }\n        paper-button.coursecard-wrapper {\n          margin: 0;\n          padding: 0;\n        }\n        paper-button.coursecard-wrapper:focus {\n          outline: blue solid 1px;\n        }\n        lrnapp-cis-course-card {\n          padding: 0;\n          margin: 1em;\n          height: 15em;\n          width: 14em;\n        }\n        .courses-grid {\n          margin: 0 auto;\n          width: 95%;\n        }\n        .iron-selected .display-mode {\n          background-color: #ff6f00;\n          color: white;\n        }\n        .list-container {\n          display: block;\n          flex-direction: column;\n          min-height: 50vh;\n          flex: 1 1 auto;\n        }\n        .list-container .coursecard-wrapper {\n          display: inline-flex;\n        }\n        .dialog-header {\n          height: unset !important;\n          padding: 0 !important;\n        }\n        .course-dialog-heading.lrndesign-course-banner {\n          font-size: 1em;\n          height: 4em !important;\n        }\n        #coursedetails {\n          margin-top: 1em;\n        }\n        #confirm {\n          max-width: 400px;\n          max-height: 300px;\n        }\n        .buttons {\n          text-align: center;\n        }\n        .buttons paper-button {\n          width: 10em;\n          height: 4em;\n        }\n        .dialog-body {\n          padding: 1em;\n          font-size: 1.8em;\n          text-align: center;\n          margin: 0 auto;\n        }\n        .dialog-body lrn-icon.service-confirm-icon {\n          width: 5em;\n          height: 5em;\n        }\n        .dialog-body responsive-grid-col {\n          height: 4.5em;\n        }\n        .dialog-body lrndesign-avatar.service-confirm-icon {\n          display: inline-block;\n        }\n      "]);

    _templateObject_e93fdd102da211eaa247b36e11306899 = function _templateObject_e93fdd102da211eaa247b36e11306899() {
      return data;
    };

    return data;
  }

  /**
   `lrnapp-cis`
   A learning application for visualizing course information and listing.
  
  @demo ../../demo/index.html
  
  @microcopy - the mental model for this app
   - cis - Course Information System
   -
  
  */
  var LrnappCis =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnappCis, _LitElement);
    babelHelpers.createClass(LrnappCis, null, [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [_colors.materialCssStyles, (0, _litElement.css)(_templateObject_e93fdd102da211eaa247b36e11306899())];
      }
      /**
       * HTMLElement
       */

    }]);

    function LrnappCis() {
      var _this;

      babelHelpers.classCallCheck(this, LrnappCis);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappCis).call(this));
      _this.courses = [];
      _this.originalCourses = [];
      _this.programs = [];
      _this.academics = [];
      _this.endPoint = "/";
      _this.basePath = "/";
      _this.queryParams = {};
      _this.data = {};
      _this.activeCourse = {};
      _this._activeService = {};
      setTimeout(function () {
        _this.addEventListener("route-change", _this._routeChange.bind(babelHelpers.assertThisInitialized(_this)));
      }, 0);
      return _this;
    }
    /**
     * LitElement render
     */


    babelHelpers.createClass(LrnappCis, [{
      key: "render",
      value: function render() {
        var _this2 = this;

        return (0, _litElement.html)(_templateObject2_e93fdd102da211eaa247b36e11306899(), this.sourcePath, this.__cisResponseChanged, this._handleResponse, this.courseDataPath, this.__courseResponseChanged, this._handleCourseResponse, this.makeServicePath, this.__makeServiceResponseChanged, this._handleMakeServiceResponse, this.courses.length, this.originalCourses.length, this.courses.length == 0 ? true : false, this.queryParams.course, this._queryParamsCourseChanged, this.originalCourses.map(function (course) {
          return (0, _litElement.html)(_templateObject3_e93fdd102da211eaa247b36e11306899(), course.id, course.attributes.name);
        }), this.programs.length == 0 ? true : false, this.queryParams.program, this._queryParamsProgramChanged, this.programs.map(function (program) {
          return (0, _litElement.html)(_templateObject4_e93fdd102da211eaa247b36e11306899(), program.id, program.attributes.name);
        }), this.academics.length == 0 ? true : false, this.queryParams.academic, this._queryParamsAcademicChanged, this.academics.map(function (academic) {
          return (0, _litElement.html)(_templateObject5_e93fdd102da211eaa247b36e11306899(), academic.id, academic.attributes.name);
        }), this.courses.map(function (course) {
          return (0, _litElement.html)(_templateObject6_e93fdd102da211eaa247b36e11306899(), course.id, _this2._loadCourseUrl, course.id, course.attributes.name, course.attributes.image, course.attributes.title, course.attributes.color);
        }), this.route, this.__routeChangedEvent, this.queryParams, this.queryParamsChanged, this.route, this.__routeChangedEvent, this.endPoint, this.tail, this.__tailChanged, this.data, this.__dataChanged, this.queryParams, this.queryParamsChanged, this.activeCourse.attributes ? (0, _litElement.html)(_templateObject7_e93fdd102da211eaa247b36e11306899(), this.activeCourse.attributes.image, this.activeCourse.attributes.name, this.activeCourse.attributes.title, this.activeCourse.attributes.color) : "", this.activeCourse.attributes ? (0, _litElement.html)(_templateObject8_e93fdd102da211eaa247b36e11306899(), !this.activeCourse.relationships.academic.attributes.name, this.activeCourse.relationships.academic.attributes.name, !this.activeCourse.relationships.program.attributes.name, this.activeCourse.relationships.program.attributes.name, this.activeCourse.topology ? (0, _litElement.html)(_templateObject9_e93fdd102da211eaa247b36e11306899(), this.activeCourse.topology.Network.map(function (service) {
          return (0, _litElement.html)(_templateObject10_e93fdd102da211eaa247b36e11306899(), service._exists ? (0, _litElement.html)(_templateObject11_e93fdd102da211eaa247b36e11306899(), service.url, service.color, service.icon, service.title) : (0, _litElement.html)(_templateObject12_e93fdd102da211eaa247b36e11306899(), _this2._makeService, service.machine_name, service.machine_name, service.icon, service.machine_name, service.title));
        })) : "", this.activeCourse.meta.canUpdate ? (0, _litElement.html)(_templateObject13_e93fdd102da211eaa247b36e11306899(), this.activeCourse.uris.edit, this.activeCourse.uris.addOffering, this.activeCourse.uris.offerings, this.activeCourse.uris.sync, this.activeCourse.uris.uri, this.activeCourse.meta.canDelete ? (0, _litElement.html)(_templateObject14_e93fdd102da211eaa247b36e11306899(), this.activeCourse.uris.delete) : (0, _litElement.html)(_templateObject15_e93fdd102da211eaa247b36e11306899())) : (0, _litElement.html)(_templateObject16_e93fdd102da211eaa247b36e11306899()), this.activeCourse.attributes.body) : "", this.activeCourse.attributes ? (0, _litElement.html)(_templateObject17_e93fdd102da211eaa247b36e11306899(), this.activeCourse.attributes.title, this._activeService.icon, this._activeService.color, this._activeService.title, this.activeCourse.attributes.name, this.activeCourse.attributes.color, this.activeCourse.attributes.title, this._confirmBuild) : "");
      }
    }, {
      key: "queryParamsChanged",
      value: function queryParamsChanged(e) {
        this.queryParams = _objectSpread({}, e.detail.value);
      }
    }, {
      key: "__tailChanged",
      value: function __tailChanged(e) {
        this.tail = e.detail.value;
      }
    }, {
      key: "__dataChanged",
      value: function __dataChanged(e) {
        this.data = _objectSpread({}, e.detail.value);
      }
    }, {
      key: "_dataPageChanged",
      value: function _dataPageChanged(e) {
        var data = this.data;
        data.page = e.detail.value;
        this.data = _objectSpread({}, data);
      }
    }, {
      key: "_queryParamsAcademicChanged",
      value: function _queryParamsAcademicChanged(e) {
        var queryParams = this.queryParams;
        queryParams.academic = e.detail.value;

        if (queryParams.academic == null) {
          delete queryParams.academic;
        }

        this.queryParams = _objectSpread({}, queryParams);
      }
    }, {
      key: "_queryParamsProgramChanged",
      value: function _queryParamsProgramChanged(e) {
        var queryParams = this.queryParams;
        queryParams.program = e.detail.value;

        if (queryParams.program == null) {
          delete queryParams.program;
        }

        this.queryParams = _objectSpread({}, queryParams);
      }
    }, {
      key: "_queryParamsCourseChanged",
      value: function _queryParamsCourseChanged(e) {
        var queryParams = this.queryParams;
        queryParams.course = e.detail.value;

        if (queryParams.course == null) {
          delete queryParams.course;
        }

        this.queryParams = _objectSpread({}, queryParams);
      }
    }, {
      key: "__courseResponseChanged",
      value: function __courseResponseChanged(e) {
        this._courseResponse = e.detail.value;
      }
    }, {
      key: "__makeServiceResponseChanged",
      value: function __makeServiceResponseChanged(e) {
        this._makeServiceResponse = e.detail.value;
      }
    }, {
      key: "__cisResponseChanged",
      value: function __cisResponseChanged(e) {
        this._cisResponse = e.detail.value;
      }
    }, {
      key: "__routeChangedEvent",
      value: function __routeChangedEvent(e) {
        this.route = e.detail.value;
      }
      /**
       * LitElement properties changed
       */

    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this3 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (["route", "endPoint"].includes(propName)) {
            _routeChanged(_this3.route, _this3.endPoint);
          }

          if (["originalCourses", "queryParams"].includes(propName)) {
            _this3.courses = babelHelpers.toConsumableArray(_this3._coursesCompute(_this3.originalCourses, _this3.queryParams));
          }

          if (propName === "originalCourses") {
            // notify
            _this3.dispatchEvent(new CustomEvent("original-courses-changed", {
              detail: {
                value: _this3[propName]
              }
            }));
          }

          if (propName === "queryParams") {
            // notify
            _this3.dispatchEvent(new CustomEvent("query-params-changed", {
              detail: {
                value: _this3[propName]
              }
            }));
          }
        });
      }
      /**
       * LitElement / popular convention
       */

    }, {
      key: "_routeChanged",
      value: function _routeChanged(route, endPoint) {
        if (typeof route.path === "string") {
          if (typeof endPoint === "string") {
            // ignore "home page" as well since that's our path
            if (route.path.startsWith(endPoint) || route.path == "/") {
              return;
            }
          }

          if (this.shadowRoot) {
            this.shadowRoot.querySelector("#loading").hidden = false; // reload the page which since route changed will load that page
          }

          window.location.reload();
        }
      }
      /**
       * Change route from deeper in the app.
       */

    }, {
      key: "_routeChange",
      value: function _routeChange(e) {
        var details = e.detail;
        var queryParams = this.queryParams;
        var data = this.data;

        if (babelHelpers.typeof(details.queryParams.course) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          queryParams.course = details.queryParams.course;
        }

        if (babelHelpers.typeof(details.queryParams.academic) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          queryParams.academic = details.queryParams.academic;
        }

        if (babelHelpers.typeof(details.queryParams.program) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          queryParams.program = details.queryParams.program;
        }

        if (babelHelpers.typeof(details.data.page) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          data.page = details.data.page;
        }

        this.data = _objectSpread({}, data);
        this.queryParams = _objectSpread({}, queryParams);
      }
      /**
       * Simple way to convert from object to array.
       */

    }, {
      key: "_toArray",
      value: function _toArray(obj) {
        if (obj == null) {
          return [];
        }

        return Object.keys(obj).map(function (key) {
          return obj[key];
        });
      }
      /**
       * Handle course response for additional details about the item
       */

    }, {
      key: "_handleMakeServiceResponse",
      value: function _handleMakeServiceResponse(event) {
        // get the CIS response's data and convert to array ahead of time
        var response = this._makeServiceResponse;

        var activeCourse = this.__addServiceLinks(response.data.course);

        this.activeCourse = activeCourse;
        this.shadowRoot.querySelector("#toast").show(response.message);
      }
      /**
       * Handle course response for additional details about the item
       */

    }, {
      key: "_handleCourseResponse",
      value: function _handleCourseResponse(event) {
        // get the CIS response's data and convert to array ahead of time
        var activeCourse = this._courseResponse.data.course;

        this.__addServiceLinks(activeCourse);

        this.activeCourse = activeCourse;
      }
      /**
       * Helper to mash up services that exist with those that could.
       */

    }, {
      key: "__addServiceLinks",
      value: function __addServiceLinks(courseObject) {
        // ensure there's a part for the Network people can request
        if (babelHelpers.typeof(courseObject.topology.Network) === (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          courseObject.topology.Network = {};
        } // loop our services to see what we should add as options


        for (var key in this.services) {
          // if it's not in the topology that means it can be added
          if (babelHelpers.typeof(courseObject.topology.Network[this.services[key].attributes.machine_name]) === (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
            // if we get a miss that means we should add a "Add this" version
            courseObject.topology.Network[this.services[key].attributes.machine_name] = {
              color: this.services[key].attributes.color,
              distro: this.services[key].attributes.distro,
              icon: this.services[key].attributes.icon,
              machine_name: this.services[key].attributes.machine_name,
              title: this.services[key].attributes.title,
              url: this.services[key].attributes.url,
              weight: this.services[key].attributes.weight,
              _exists: false
            };
          }
        } // convert to array after keys in place for the object


        courseObject.topology.Network = this._toArray(courseObject.topology.Network); // sort items based on weight of the things in the network
        // so we have a consistent order to things

        courseObject.topology.Network.sort(function (a, b) {
          return a.weight - b.weight;
        });
        return courseObject;
      }
      /**
       * Handle response for the whole courses object.
       */

    }, {
      key: "_handleResponse",
      value: function _handleResponse(event) {
        var course = {};
        var program = {};
        var academic = {};
        var tmp = {
          courses: [],
          programs: [],
          academics: []
        };
        var programs = [];
        var academics = []; // get the CIS response's data and convert to array ahead of time

        var courses = this._toArray(this._cisResponse.data.courses);

        this.services = babelHelpers.toConsumableArray(this._toArray(this._cisResponse.data.services));
        this.originalCourses = babelHelpers.toConsumableArray(courses);

        for (var index = 0; index < courses.length; index++) {
          course = courses[index];
          program = courses[index].relationships.program;
          academic = courses[index].relationships.academic;
          tmp.programs[program.id] = program;
          tmp.academics[academic.id] = academic;
          tmp.courses[course.id] = course;
        } // this is stupid but we have to normalize the IDs or else dom repeats will be screwed up


        tmp.programs.forEach(function (element) {
          programs.push(element);
        }); // this is stupid but we have to normalize the IDs or else dom repeats will be screwed up

        tmp.academics.forEach(function (element) {
          academics.push(element);
        });
        this.shadowRoot.querySelector("#loading").hidden = true;
        this.academics = [].concat(academics);
        this.programs = [].concat(programs);
      }
      /**
       * Request a new service to kick off.
       */

    }, {
      key: "_makeService",
      value: function _makeService(e) {
        var active = e.target.getAttribute("data-machine-name");
        var network = this.activeCourse.topology.Network;
        var service = network.filter(function (service) {
          if (service.machine_name !== active) {
            return false;
          }

          return true;
        }); // if we found one, make it the top level item

        if (service.length > 0) {
          service = service.pop();
          this.shadowRoot.querySelector("#makeservice").params = {
            course: this.activeCourse.attributes.machine_name,
            service: service.machine_name
          };
          this._activeService = _objectSpread({}, service); // confirm via paper prompt

          this.shadowRoot.querySelector("#confirm").toggleDialog();
        } else {
          console.log("that was not a valid service..");
        }
      }
      /**
       * Confirm of build.
       */

    }, {
      key: "_confirmBuild",
      value: function _confirmBuild(e) {
        this.shadowRoot.querySelector("#makeservice").generateRequest();
      }
      /**
       * Handle tap on paper-button above to redirect to the correct course url.
       */

    }, {
      key: "_loadCourseUrl",
      value: function _loadCourseUrl(e) {
        // reset dialog to appear to be loading
        var local = e.target; // this will have the id of the current course

        var active = local.getAttribute("data-course-id"); // find the course by it's unique id and filter just to it

        var findCourse = this.originalCourses.filter(function (course) {
          if (course.id !== active) {
            return false;
          }

          return true;
        }); // if we found one, make it the top level item

        if (findCourse.length > 0) {
          findCourse = findCourse.pop();
        }

        this.activeCourse = findCourse; // formulate the post data

        this._courseDataParams = {
          id: this.activeCourse.id
        }; // @todo look at query cache mechanism to skip calls
        // if they've already happened. lrnapp-book has some stuff to do this

        this.shadowRoot.querySelector("#courserequest").params = this._courseDataParams;
        this.shadowRoot.querySelector("#courserequest").generateRequest();
        this.shadowRoot.querySelector("#dialog").toggle();
      }
      /**
       * Compute the active list of courses
       */

    }, {
      key: "_coursesCompute",
      value: function _coursesCompute(originalCourses, queryParams) {
        var _this4 = this;

        // if we don't have an original courses object to work with then we need to bail
        if (typeof originalCourses === "undefined") {
          return [];
        } // define vars


        var filteredCourses = []; // filter the courses by the query params

        filteredCourses = originalCourses.filter(function (course) {
          if (typeof _this4.queryParams.course !== "undefined") {
            if (course.id !== _this4.queryParams.course) {
              return false;
            }
          }

          if (typeof _this4.queryParams.program !== "undefined") {
            if (course.relationships.program.id !== _this4.queryParams.program) {
              return false;
            }
          }

          if (typeof _this4.queryParams.academic !== "undefined") {
            if (course.relationships.academic.id !== _this4.queryParams.academic) {
              return false;
            }
          }

          return true;
        }); // delay and repaint, can help with refresh issues

        return filteredCourses;
      }
      /**
       * Shadow Hijack
       */
      //createRenderRoot() {
      //  return this;
      //}

    }], [{
      key: "tag",
      get: function get() {
        return "lrnapp-cis";
      }
    }, {
      key: "properties",
      get: function get() {
        var _ref;

        return _ref = {
          _activeService: {
            type: Object
          },
          elmslnCourse: {
            type: String,
            attribute: "elmsln-course"
          },
          elmslnSection: {
            type: String,
            attribute: "elmsln-section"
          },
          basePath: {
            type: String,
            attribute: "base-path"
          },
          csrfToken: {
            type: String,
            attribute: "csrf-token"
          },
          endPoint: {
            type: String,
            attribute: "end-point"
          },

          /**
           * The load initial data
           */
          _cisResponse: {
            type: Object
          },

          /**
           * Load individual course data
           */
          _courseResponse: {
            type: Object
          },

          /**
           * Load service creation response
           */
          _makeServiceResponse: {
            type: Object
          },

          /**
           * The courses to render; potentially filtered
           */
          courses: {
            type: Array
          },

          /**
           * The original courses array; used to filter against
           */
          originalCourses: {
            type: Array
          },

          /**
           * The programs to render
           */
          programs: {
            type: Array
          },

          /**
           * The academics to render
           */
          academics: {
            type: Array
          },

          /**
           * sourcePath for data.
           */
          sourcePath: {
            type: String,
            attribute: "source-path"
          },

          /**
           * pathway to access info about a single course.
           */
          courseDataPath: {
            type: String,
            attribute: "course-data-path"
          },

          /**
           * pathway to creating new service instances
           */
          makeServicePath: {
            type: String,
            attribute: "make-service-path"
          }
        }, babelHelpers.defineProperty(_ref, "endPoint", {
          type: String,
          attribute: "end-point"
        }), babelHelpers.defineProperty(_ref, "activeCourse", {
          type: Array
        }), babelHelpers.defineProperty(_ref, "queryParams", {
          type: Object
        }), _ref;
      }
    }]);
    return LrnappCis;
  }(_litElement.LitElement);

  _exports.LrnappCis = LrnappCis;
  window.customElements.define(LrnappCis.tag, LrnappCis);
});
