// @lrnwebcomponents/elmsln-apps/lib/lrnapp-cis/lrnapp-cis-course-card.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@polymer/paper-card/paper-card.js", "../../../../@polymer/iron-icon/iron-icon.js", "../../../../@polymer/iron-image/iron-image.js", "../../../materializecss-styles/materializecss-styles.js", "../../../../@polymer/polymer/lib/elements/custom-style.js"], function (_exports, _litElement, _paperCard, _ironIcon, _ironImage, _materializecssStyles, _customStyle) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappCisCourseCard = void 0;

  function _templateObject2_01040cc01bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: inline-flex;\n        }\n        :host([size=\"micro\"]) {\n          transform: scale(0.5);\n        }\n        :host([size=\"small\"]) {\n          transform: scale(0.8);\n        }\n        paper-card {\n          border-radius: 4px;\n          margin: 0;\n          width: 100%;\n        }\n        .card-actions {\n          background-color: #f5f5f5;\n          border-radius: 0 0 4px 4px;\n          padding: 0 8px;\n        }\n        .card-actions .card-action-details {\n          display: inline-block;\n          vertical-align: middle;\n          vertical-align: -webkit-baseline-middle;\n          width: 80%;\n        }\n        .card-control-height {\n          height: 240px;\n        }\n        [elevation=\"0\"] {\n          border: solid 1px #eeeeee;\n        }\n        .text-right {\n          text-align: right;\n        }\n        .text-left {\n          text-align: left;\n        }\n        .name,\n        .title {\n          color: #222;\n          font-size: 14px;\n          font-weight: 600;\n          line-height: 20px;\n          padding: 0 12px;\n          overflow: hidden;\n          text-overflow: ellipsis;\n          margin-top: 8px;\n        }\n        .title {\n          font-size: 12px;\n          font-weight: 400;\n        }\n        .divider {\n          height: 1px;\n          width: 100%;\n          background: #efefef;\n        }\n        .course-icon {\n          --iron-icon-height: 100%;\n          --iron-icon-width: 100%;\n          overflow: hidden;\n          color: grey;\n        }\n        .course-icon:hover,\n        .course-icon:focus {\n          color: black;\n        }\n        .center {\n          margin: auto;\n          width: 80%;\n          padding: 16px;\n        }\n        .link {\n          font-size: 16px;\n          line-height: 16px;\n        }\n        .course-info {\n          width: 100%;\n        }\n        .course-preview {\n          height: 160px;\n        }\n        .card-content {\n          padding: 0;\n          margin: 0;\n          overflow: hidden;\n        }\n        .inline {\n          display: inline;\n        }\n      "]);

    _templateObject2_01040cc01bb311eabd7cc30119585ddb = function _templateObject2_01040cc01bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject_01040cc01bb311eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <custom-style>\n        <style include=\"materializecss-styles\">\n          :host {\n            display: block;\n          }\n        </style>\n      </custom-style>\n      <paper-card elevation=\"", "\">\n        <div class=\"card-content card-control-height card-control-center\">\n          <div class=\"course-preview\">\n            <iron-icon\n              class=\"course-icon\"\n              icon=\"", "\"\n              ?hidden=\"", "\"\n            ></iron-icon>\n            <iron-image\n              style=\"width:100%; height:100%; background-color: lightgray;\"\n              sizing=\"cover\"\n              preload=\"\"\n              fade=\"\"\n              src=\"", "\"\n              ?hidden=\"", "\"\n            ></iron-image>\n          </div>\n          <div class=\"course-info\">\n            <div class=\"divider\"></div>\n            <div class=\"name\">", "</div>\n            <div class=\"title\">", "</div>\n          </div>\n        </div>\n        <div class=\"card-actions\" hidden=\"\">\n          <div class=\"card-action-details\"></div>\n        </div>\n      </paper-card>\n    "]);

    _templateObject_01040cc01bb311eabd7cc30119585ddb = function _templateObject_01040cc01bb311eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  /*
   `lrnapp-cis-course-card`
   Present a course card
  */
  var LrnappCisCourseCard =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnappCisCourseCard, _LitElement);
    babelHelpers.createClass(LrnappCisCourseCard, [{
      key: "render",

      /**
       * LitElement render
       */
      value: function render() {
        return (0, _litElement.html)(_templateObject_01040cc01bb311eabd7cc30119585ddb(), this.elevation, this.icon, !this.icon, this.image, !this.image, this.name, this.title);
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_01040cc01bb311eabd7cc30119585ddb())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrnapp-cis-course-card";
      }
      /**
       * HTMLElement
       */

    }]);

    function LrnappCisCourseCard() {
      var _this;

      babelHelpers.classCallCheck(this, LrnappCisCourseCard);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappCisCourseCard).call(this));
      _this.icon = false;
      _this.name = "";
      _this.title = "";
      _this.color = "grey";
      _this.elevation = 1;
      setTimeout(function () {
        _this.addEventListener("mouseenter", _this._mouseEnter.bind(babelHelpers.assertThisInitialized(_this)));

        _this.addEventListener("mouseleave", _this._mouseLeave.bind(babelHelpers.assertThisInitialized(_this)));
      }, 0);
      return _this;
    }

    babelHelpers.createClass(LrnappCisCourseCard, [{
      key: "_mouseEnter",

      /**
       * Increase elevation while hovering.
       */
      value: function _mouseEnter(e) {
        this.__oldElevation = this.elevation;

        if (this.elevation + 2 > 5) {
          this.elevation = 5;
        } else {
          this.elevation += 2;
        }
      }
      /**
       * Reset the elevation.
       */

    }, {
      key: "_mouseLeave",
      value: function _mouseLeave(e) {
        this.elevation = this.__oldElevation;
      }
    }], [{
      key: "properties",
      get: function get() {
        return {
          size: {
            type: String
          },

          /**
           * Cover image src.
           */
          image: {
            type: String
          },

          /**
           * Icon to use if image isn't there.
           */
          icon: {
            type: String
          },

          /**
           * name of the course like sing100
           */
          name: {
            type: String
          },

          /**
           * title of the course like Intro to studies
           */
          title: {
            type: String
          },

          /**
           * color of the course item
           */
          color: {
            type: String
          },

          /**
           * Visual elevation of the item off the UI via paper element height
           */
          elevation: {
            type: Number,
            reflect: true
          }
        };
      }
    }]);
    return LrnappCisCourseCard;
  }(_litElement.LitElement);

  _exports.LrnappCisCourseCard = LrnappCisCourseCard;
  window.customElements.define(LrnappCisCourseCard.tag, LrnappCisCourseCard);
});
