// @lrnwebcomponents/a11y-menu-button/a11y-menu-button.js
define(["exports", "../../lit/index.js", "./lib/a11y-menu-button-item.js", "../absolute-position-behavior/absolute-position-behavior.js", "../utils/utils.js"], function (_exports, _index, _a11yMenuButtonItem, _absolutePositionBehavior, _utils) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.A11yMenuButtonBehaviors = _exports.A11yMenuButton = void 0;

  var _templateObject_1e068830942811ecb1937969798b231a, _templateObject2_1e068830942811ecb1937969798b231a, _templateObject3_1e068830942811ecb1937969798b231a, _templateObject4_1e068830942811ecb1937969798b231a, _templateObject5_1e068830942811ecb1937969798b231a, _templateObject6_1e068830942811ecb1937969798b231a;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var A11yMenuButtonBehaviors = function A11yMenuButtonBehaviors(SuperClass) {
    return /*#__PURE__*/function (_SuperClass) {
      babelHelpers.inherits(_class, _SuperClass);

      var _super = _createSuper(_class);

      function _class() {
        var _this;

        babelHelpers.classCallCheck(this, _class);
        _this = _super.call(this);
        _this.position = "bottom";
        _this.positionAlign = "start";
        _this.offset = 0;
        _this.menuItems = [];
        _this.keepOpenOnClick = false;
        _this.noOpenOnHover = false;
        babelHelpers.toConsumableArray(_this.children).filter(function (n) {
          return n.slot === "menuitem";
        }).forEach(function (item) {
          return _this.addItem(item);
        });

        _this.addEventListener("keydown", _this._handleKeydown);

        _this.addEventListener("click", _this._handleClick);

        _this.addEventListener("focus", _this._handleFocus);

        _this.addEventListener("blur", _this._handleBlur);

        _this.addEventListener("mouseover", _this._handleMouseover);

        _this.addEventListener("mouseout", _this._handleMouseout);

        _this.addEventListener("add-a11y-menu-button-item", _this._handleAddItem);

        _this.addEventListener("remove-a11y-menu-button-item", _this._handleRemoveItem);

        return _this;
      }
      /**
       * template for dropdown menu
       *
       * @readonly
       */


      babelHelpers.createClass(_class, [{
        key: "render",
        value: function render() {
          return (0, _index.html)(_templateObject_1e068830942811ecb1937969798b231a || (_templateObject_1e068830942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral([" ", " ", " "])), this.buttonTemplate, this.menuTemplate);
        }
      }, {
        key: "menuTemplate",
        get: function get() {
          var _this2 = this;

          return (0, _index.html)(_templateObject2_1e068830942811ecb1937969798b231a || (_templateObject2_1e068830942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        <absolute-position-behavior\n          ?auto=\"", "\"\n          for=\"menubutton\"\n          position=\"", "\"\n          position-align=\"", "\"\n          .offset=\"", "\"\n          fit-to-visible-bounds\n          part=\"menu-outer\"\n        >\n          <ul\n            id=\"menu\"\n            role=\"menu\"\n            aria-labelledby=\"menubutton\"\n            ?hidden=\"", "\"\n            @mousover=\"", "\"\n            @mousout=\"", "\"\n            part=\"menu\"\n          >\n            ", "\n          </ul>\n        </absolute-position-behavior>\n      "])), this.expanded, this.position, this.positionAlign, this.offset, !this.expanded, function (e) {
            return _this2.hover = true;
          }, function (e) {
            return _this2.hover = false;
          }, this.listItemTemplate);
        }
        /**
         * template for button
         *
         * @readonly
         */

      }, {
        key: "buttonTemplate",
        get: function get() {
          return (0, _index.html)(_templateObject3_1e068830942811ecb1937969798b231a || (_templateObject3_1e068830942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        <button\n          id=\"menubutton\"\n          aria-haspopup=\"true\"\n          aria-controls=\"menu\"\n          aria-expanded=\"", "\"\n          part=\"button\"\n        >\n          <slot name=\"button\"></slot>\n        </button>\n      "])), this.expanded ? "true" : "false");
        }
        /**
         * template for slotted list items
         *
         * @readonly
         */

      }, {
        key: "listItemTemplate",
        get: function get() {
          return (0, _index.html)(_templateObject4_1e068830942811ecb1937969798b231a || (_templateObject4_1e068830942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<slot name=\"menuitem\"></slot><slot></slot>"])));
        }
        /**
         * key code translations as object
         *
         * @readonly
         * @memberof A11yMenuButton
         */

      }, {
        key: "keyCode",
        get: function get() {
          return {
            TAB: 9,
            RETURN: 13,
            ESC: 27,
            SPACE: 32,
            PAGEUP: 33,
            PAGEDOWN: 34,
            END: 35,
            HOME: 36,
            LEFT: 37,
            UP: 38,
            RIGHT: 39,
            DOWN: 40
          };
        }
        /**
         * closes menu
         *
         * @param {boolean} force close even if other items have focus
         * @memberof A11yMenuButton
         */

      }, {
        key: "close",
        value: function close() {
          var force = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;

          if (force || !this.focused && !this.hovered) {
            this.expanded = false;
            /**
             * Fires when menu is closed
             * @event close
             */

            this.dispatchEvent(new CustomEvent("close", {
              bubbles: true,
              cancelable: true,
              composed: true,
              detail: this
            }));
          }
        }
        /**
         * opens menu
         *
         * @memberof A11yMenuButton
         */

      }, {
        key: "open",
        value: function open() {
          this.expanded = true;
          /**
           * Fires when menu is opened
           * @event close
           */

          this.dispatchEvent(new CustomEvent("open", {
            bubbles: true,
            cancelable: true,
            composed: true,
            detail: this
          }));
        }
        /**
         * gives focus to menu
         *
         * @memberof A11yMenuButton
         */

      }, {
        key: "focus",
        value: function focus() {
          if (this.shadowRoot && this.shadowRoot.querySelector("#menubutton")) {
            this.shadowRoot.querySelector("#menubutton").focus();
          }
        }
        /**
         * focuses on a menu item
         *
         * @param {object} item menu item
         * @memberof A11yMenuButton
         */

      }, {
        key: "focusOn",
        value: function focusOn(item) {
          item = item || this.firstItem;

          if (item) {
            this.open();
            this.focused = true;
            this.currentItem = item;
            item.focus();
          }
        }
        /**
         * focuses on item based on character pressed
         *
         * @param {string} char character pressed
         * @memberof A11yMenuButton
         */

      }, {
        key: "focusByCharacter",
        value: function focusByCharacter(char) {
          var _this3 = this;

          var start,
              index,
              char = char.toLowerCase(),
              firstChars = function firstChars(startIndex, char) {
            for (var i = startIndex; i < _this3.firstChars.length; i++) {
              if (char === _this3.firstChars[i]) {
                return i;
              }
            }

            return -1;
          }; // Get start index for search based on position of currentItem


          start = this.menuItems.indexOf(this.currentItem) + 1;

          if (start === this.menuItems.length) {
            start = 0;
          } // Check remaining slots in menu


          index = firstChars(start, char); // If not found in remaining slots, check from beginning

          if (index === -1) {
            index = firstChars(0, char);
          } // If match was found...


          if (index > -1) {
            this.menuItems[index].focus();
          }
        }
        /**
         * gets first menu item
         *
         * @returns {object}
         * @memberof A11yMenuButton
         */

      }, {
        key: "firstItem",
        get: function get() {
          return this.menuItems[0];
        }
        /**
         * gets previous menu item
         *
         * @returns {object}
         * @memberof A11yMenuButton
         */

      }, {
        key: "previousItem",
        get: function get() {
          return this.getItem(-1);
        }
        /**
         * gets next menu item
         *
         * @returns {object}
         * @memberof A11yMenuButton
         */

      }, {
        key: "nextItem",
        get: function get() {
          return this.getItem();
        }
        /**
         * gets last menu item
         *
         * @returns {object}
         * @memberof A11yMenuButton
         */

      }, {
        key: "lastItem",
        get: function get() {
          return this.menuItems[this.menuItems.length - 1];
        }
        /**
         * gets list of menu item first characters
         *
         * @readonly
         */

      }, {
        key: "firstChars",
        get: function get() {
          return this.menuItems.map(function (item) {
            var textContent = (item.textContent || "").trim();
            return (textContent || " ").substring(0, 1).toLowerCase();
          });
        }
        /**
         * gets index of a menuitem
         *
         * @returns {number}
         */

      }, {
        key: "getItemIndex",
        value: function getItemIndex() {
          var _this4 = this;

          var index = -1;
          this.menuItems.forEach(function (b, i) {
            if (b === _this4.currentItem) index = i;
          });
          return index;
        }
        /**
         * gets a menuitem relative to the current item
         *
         * @param {number} [offset=1] number of items after current menuitem in order, -1 for item before
         * @returns
         */

      }, {
        key: "getItem",
        value: function getItem() {
          var offset = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 1;
          var index = this.getItemIndex(this.currentItem) + offset;
          return !this.menuItems || index < 0 || this.menuItems.length <= index ? undefined : this.menuItems[index];
        }
        /**
         * menuitem event listeners and their handlers
         *
         * @readonly {object}
         */

      }, {
        key: "itemListeners",
        get: function get() {
          return {
            click: this._handleItemClick,
            focus: this._handleFocus,
            blur: this._handleBlur,
            mouseover: this._handleMouseover,
            mouseout: this._handleMouseout,
            keydown: this._handleItemKeydown
          };
        }
        /**
         * adds a menuitem to lists and sets up its listeners
         *
         * @param {ibject} item menu item element
         */

      }, {
        key: "addItem",
        value: function addItem(item) {
          var _this5 = this;

          var listeners = this.itemListeners;
          this.menuItems = this.menuItems || [];
          Object.keys(listeners).forEach(function (evt) {
            return item.addEventListener(evt, listeners[evt].bind(_this5));
          });
          this.menuItems.push(item);
        }
        /**
         * removes a menuitem's listners and menuitem istelf from list
         *
         * @param {ibject} item menu item element
         */

      }, {
        key: "removeItem",
        value: function removeItem(item) {
          var _this6 = this;

          var listeners = this.itemListeners;
          if (this.menuItems) this.menuItems = babelHelpers.toConsumableArray(this.menuItems.filter(function (i) {
            return item !== i;
          }));
          Object.keys(listeners).forEach(function (evt) {
            return item.removeEventListener(evt, listeners[evt].bind(_this6));
          });
        }
        /**
         * when a new menu item is added to slot,
         * updates menu items list and adds event listeners for item
         *
         * @param {event} event
         * @memberof A11yMenuButton
         */

      }, {
        key: "_handleAddItem",
        value: function _handleAddItem(event) {
          if (event.stopPropagation) event.stopPropagation();
          if (event.detail) this.addItem(event.detail);
        }
        /**
         * when a new menu item is removed from slot,
         * updates menu items list and removes event listeners for item
         *
         * @param {event} event
         * @memberof A11yMenuButton
         */

      }, {
        key: "_handleRemoveItem",
        value: function _handleRemoveItem(event) {
          event.stopPropagation();
          if (event.detail) this.addItem(event.detail);
        }
        /**
         * when menu item is clicked,
         * focus on menu button and close menu
         *
         * @param {event} event
         * @memberof A11yMenuButton
         */

      }, {
        key: "_handleItemClick",
        value: function _handleItemClick(event) {
          if (!this.keepOpenOnClick) {
            this.focus();
            this.close(true);
          }

          this.dispatchEvent(new CustomEvent("item-click", {
            bubbles: true,
            cancelable: true,
            composed: true,
            detail: event
          }));
          event.stopPropagation();
        }
        /**
         * can be overridden when extending this button,
         * so that certain elements in menuitems are excluded from keyboard handling
         *
         * @param {event} event
         * @returns
         */

      }, {
        key: "_excludeEvent",
        value: function _excludeEvent(event) {
          return false;
        }
        /**
         * handles menu item keydown, as prescribed in
         * {@link https://www.w3.org/TR/wai-aria-practices/examples/menu-button/menu-button-links.html} and
         * {@link https://www.w3.org/TR/wai-aria-practices/examples/menu-button/menu-button-actions-active-descendant.html}
         *
         * @param {event} event
         * @memberof A11yMenuButton
         */

      }, {
        key: "_handleItemKeydown",
        value: function _handleItemKeydown(event) {
          var flag = false,
              char = event.key,
              isPrintableCharacter = function isPrintableCharacter(str) {
            return str.length === 1 && str.match(/\S/);
          },
              path = (0, _utils.normalizeEventPath)(event) || [],
              target = path[0]; //don't handle form field keystrokes


          if (!isPrintableCharacter(char) && (["INPUT", "TEXTAREA", "SELECT"].includes(target.tagName) || this._excludeEvent(event))) return;

          if (event.ctrlKey || event.altKey || event.metaKey || event.keyCode === this.keyCode.SPACE || event.keyCode === this.keyCode.RETURN) {
            return;
          }

          if (event.shiftKey) {
            if (isPrintableCharacter(char)) {
              this.focusByCharacter(char);
              flag = true;
            }

            if (event.keyCode === this.keyCode.TAB) {
              this.focus();
              this.close(true);
            }
          } else {
            switch (event.keyCode) {
              case this.keyCode.ESC:
                this.focus();
                this.close(true);
                flag = true;
                break;

              case this.keyCode.UP:
                this.focusOn(this.previousItem || this.lastItem);
                flag = true;
                break;

              case this.keyCode.DOWN:
                this.focusOn(this.nextItem || this.firstItem);
                flag = true;
                break;

              case this.keyCode.HOME:
              case this.keyCode.PAGEUP:
                this.currentItem = this.firstItem;
                flag = true;
                break;

              case this.keyCode.END:
              case this.keyCode.PAGEDOWN:
                this.currentItem = this.lastItem;
                flag = true;
                break;

              case this.keyCode.TAB:
                this.focus();
                this.close(true);
                break;

              default:
                if (isPrintableCharacter(char)) {
                  this.focusByCharacter(char);
                }

                break;
            }
          }

          if (flag) {
            event.stopPropagation();
            event.preventDefault();
          }
        }
        /**
         * handles menu button keydown
         *
         * @param {event} event
         * @memberof A11yMenuButton
         */

      }, {
        key: "_handleKeydown",
        value: function _handleKeydown(event) {
          var flag = false;

          switch (event.keyCode) {
            case this.keyCode.SPACE:
            case this.keyCode.RETURN:
            case this.keyCode.DOWN:
              this.focusOn(this.firstItem);
              flag = true;
              break;

            case this.keyCode.UP:
              if (this.popupMenu) {
                this.focusOn(this.lastItem);
                flag = true;
              }

              break;

            default:
              break;
          }

          if (flag) {
            event.stopPropagation();
            event.preventDefault();
          }
        }
        /**
         * handles when menu is clicked
         *
         * @param {event} event
         * @memberof A11yMenuButton
         */

      }, {
        key: "_handleClick",
        value: function _handleClick(event) {
          if (this.expanded) {
            this.close(true);
          } else {
            this.focusOn(this.firstItem);
          }
        }
        /**
         * handles when menu has focus
         *
         * @param {event} event
         * @memberof A11yMenuButton
         */

      }, {
        key: "_handleFocus",
        value: function _handleFocus(event) {
          this.focused = true;
        }
        /**
         * handles when menu loses focus
         *
         * @param {event} event
         * @memberof A11yMenuButton
         */

      }, {
        key: "_handleBlur",
        value: function _handleBlur(event) {
          this.focused = false;
        }
        /**
         * handles menu mouseover
         *
         * @param {event} event
         * @memberof A11yMenuButton
         */

      }, {
        key: "_handleMouseover",
        value: function _handleMouseover(event) {
          this.hovered = true;
          if (!this.noOpenOnHover) this.open();
        }
        /**
         * handles menu mouseout
         *
         * @param {event} event
         * @memberof A11yMenuButton
         */

      }, {
        key: "_handleMouseout",
        value: function _handleMouseout(event) {
          this.hovered = false;
          setTimeout(this.close(), 300);
        }
      }], [{
        key: "menuButtonCoreStyles",
        get:
        /**
         * the core styles needed to make a menu button
         *
         * @readonly
         * @static
         */
        function get() {
          return [(0, _index.css)(_templateObject5_1e068830942811ecb1937969798b231a || (_templateObject5_1e068830942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n          :host {\n            padding: 0;\n            display: inline-flex;\n            position: relative;\n            z-index: 1;\n          }\n          :host([expanded]) {\n            z-index: var(--a11y-menu-button-focus-z-index, 1000);\n          }\n          button[part=\"button\"] {\n            display: block;\n            text-decoration: inherit;\n            font-family: inherit;\n            font-size: inherit;\n            margin: 0;\n            width: 100%;\n          }\n          absolute-position-behavior {\n            z-index: -1;\n            overflow: hidden;\n            transition: z-index 0s;\n          }\n          :host([expanded]) absolute-position-behavior {\n            z-index: var(--a11y-menu-button-focus-z-index, 1000);\n            transition: z-index 0s;\n          }\n          :host(:not([expanded])) absolute-position-behavior {\n            border-color: none !important;\n          }\n          ul {\n            margin: 0;\n            padding: 0;\n            list-style: none;\n          }\n        "])))];
        }
        /**
         * styles that can be customized by
         * extending and overriding this getter
         *
         * @readonly
         * @static
         */

      }, {
        key: "menuButtonThemeStyles",
        get: function get() {
          return [(0, _index.css)(_templateObject6_1e068830942811ecb1937969798b231a || (_templateObject6_1e068830942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n          button[part=\"button\"] {\n            padding: var(--a11y-menu-button-vertical-padding, 2px)\n              var(--a11y-menu-button-horizontal-padding, 5px);\n            text-align: var(--a11y-menu-button-text-align, center);\n            background-color: var(--a11y-menu-button-bg-color, white);\n            color: var(--a11y-menu-button-color, currentColor);\n            background-color: var(--a11y-menu-button-bg-color, white);\n            border-radius: var(--a11y-menu-button-border-radius, 0);\n            border-left: var(--a11y-menu-button-border-left, unset);\n            border-top: var(--a11y-menu-button-border-top, unset);\n            border-right: var(--a11y-menu-button-border-right, unset);\n            border-bottom: var(--a11y-menu-button-border-bottom, unset);\n            border: var(--a11y-menu-button-border, 1px solid #ddd);\n            box-shadow: var(--a11y-menu-button-box-shadow, unset);\n            transition: all 0.25s ease-in-out;\n          }\n          button[part=\"button\"]:focus,\n          button[part=\"button\"]:hover {\n            color: var(\n              --a11y-menu-button-focus-color,\n              var(--a11y-menu-button-color, currentColor)\n            );\n            background-color: var(\n              --a11y-menu-button-focus-bg-color,\n              var(--a11y-menu-button-bg-color, white)\n            );\n            border-left: var(\n              --a11y-menu-button-focus-border-left,\n              var(--a11y-menu-button-border-left, unset)\n            );\n            border-top: var(\n              --a11y-menu-button-focus-border-top,\n              var(--a11y-menu-button-border-top, unset)\n            );\n            border-right: var(\n              --a11y-menu-button-focus-border-right,\n              var(--a11y-menu-button-border-right, unset)\n            );\n            border-bottom: var(\n              --a11y-menu-button-focus-border-bottom,\n              var(--a11y-menu-button-border-bottom, unset)\n            );\n            border: var(\n              --a11y-menu-button-focus-border,\n              var(--a11y-menu-button-border, 1px solid #ddd)\n            );\n            box-shadow: var(\n              --a11y-menu-button-box-shadow,\n              var(--a11y-menu-button-focus-box-shadow, unset)\n            );\n          }\n          :host([expanded]) absolute-position-behavior {\n            width: var(--a11y-menu-button-list-width, unset);\n            height: var(--a11y-menu-button-list-height, unset);\n            border: var(\n              --a11y-menu-button-list-border,\n              var(--a11y-menu-button-border, 1px solid #ddd)\n            );\n            background-color: var(\n              --a11y-menu-button-bg-color,\n              var(--a11y-menu-button-list-bg-color, white)\n            );\n            box-shadow: var(--a11y-menu-button-list-box-shadow, unset);\n          }\n        "])))];
        }
      }, {
        key: "styles",
        get: function get() {
          return [].concat(babelHelpers.toConsumableArray(this.menuButtonCoreStyles), babelHelpers.toConsumableArray(this.menuButtonThemeStyles));
        }
      }, {
        key: "tag",
        get: function get() {
          return "a11y-menu-button";
        }
      }, {
        key: "properties",
        get: function get() {
          return {
            /**
             * Whether toggle is disabled
             */
            currentItem: {
              type: Object
            },

            /**
             * Whether toggle is disabled
             */
            disabled: {
              attribute: "disabled",
              type: Boolean
            },

            /**
             * Whether toggle is disabled
             */
            expanded: {
              attribute: "expanded",
              type: Boolean,
              reflect: true
            },

            /**
             * Whether button is toggled
             */
            focused: {
              attribute: "focused",
              type: Boolean
            },

            /**
             * Whether button is toggled
             */
            hovered: {
              attribute: "hovered",
              type: Boolean
            },

            /**
             * overrides default behavior of closing
             * menu after an item is clicked
             */
            keepOpenOnClick: {
              attribute: "keep-open-on-click",
              type: Boolean
            },

            /**
             * menu items in array form to move from prev to next
             */
            menuItems: {
              type: Array
            },

            /**
             * disables menu i=opening on hover
             */
            noOpenOnHover: {
              attribute: "no-open-on-hover",
              type: Boolean
            },

            /**
             * spacing between top of list and menu button
             */
            offset: {
              type: Number,
              attribute: "offset"
            },

            /**
             * Positions list to top, right, bottom, left of its content.
             */
            position: {
              type: String,
              attribute: "position",
              reflect: true
            },

            /**
             * Aligns list at start, or end fo target. Default is centered.
             */
            positionAlign: {
              type: String,
              attribute: "position-align",
              reflect: true
            }
          };
        }
      }]);
      return _class;
    }(SuperClass);
  };
  /**
   * a11y-menu-button
   * A toggle button for an property in editable-table interface (editable-table.html).
   *
  ### Styling
  
  `<a11y-menu-button>` provides custom properties for styling:
  
  Custom property | Description | Default
  ----------------|-------------|----------
  --a11y-menu-button-vertical-padding | vertical padding for menu button | 2px
  --a11y-menu-button-horizontal-padding | horizontal padding for menu button | 5px
  --a11y-menu-button-text-align | text alignment for menu button | center
  --a11y-menu-button-bg-color | default background color | white
  --a11y-menu-button-color | default text color | black
  --a11y-menu-button-box-shadow | menu button box-shadow | unset
  --a11y-menu-button-border-radius | menu button border-radius | 0
  --a11y-menu-button-border | default border | 1px solid #ddd
  --a11y-menu-button-border-left | overrides default left-border | unset
  --a11y-menu-button-border-top | overrides default top-border | unset
  --a11y-menu-button-border-right | overrides default right-border | unset
  --a11y-menu-button-border-bottom | overrides default bottom-border | unset
  --a11y-menu-button-focus-bg-color | background color for menu button when focused | --a11y-menu-button-bg-color
  --a11y-menu-button-focus-color | text color for menu button when focused | --a11y-menu-button-color
  --a11y-menu-button-focus-border | border for menu button when focused | --a11y-menu-button-border
  --a11y-menu-button-focus-border-left | menu button left-border when focused | --a11y-menu-button-border-left
  --a11y-menu-button-focus-border-top | menu button top-border when focused | --a11y-menu-button-border-top
  --a11y-menu-button-focus-border-right | menu button right-border when focused | --a11y-menu-button-border-right
  --a11y-menu-button-focus-border-bottom | menu button bottom-border when focused | --a11y-menu-button-border-bottom
  --a11y-menu-button-focus-box-shadow | menu button box-shadow when focused | --a11y-menu-button-box-shadow
  --a11y-menu-button-list-width | width of menu list | unset
  --a11y-menu-button-list-height | height of menu list | unset
  --a11y-menu-button-list-left | left position of menu list | 0
  --a11y-menu-button-list-top | top position of menu list | unset
  --a11y-menu-button-list-bottom | bottom position of menu list | unset
  --a11y-menu-button-list-right | right position of menu list | unset
  --a11y-menu-button-list-bg-color | overrides default background color for list box | --a11y-menu-button-bg-color
  --a11y-menu-button-border | overrides default border for list box | --a11y-menu-button-list-border
  --a11y-menu-button-list-box-shadow | overrides default box shadow for list box | unset
   *
   * @demo ./demo/index.html
   * @element a11y-menu-button
   */


  _exports.A11yMenuButtonBehaviors = A11yMenuButtonBehaviors;

  var A11yMenuButton = /*#__PURE__*/function (_A11yMenuButtonBehavi) {
    babelHelpers.inherits(A11yMenuButton, _A11yMenuButtonBehavi);

    var _super2 = _createSuper(A11yMenuButton);

    function A11yMenuButton() {
      babelHelpers.classCallCheck(this, A11yMenuButton);
      return _super2.apply(this, arguments);
    }

    return babelHelpers.createClass(A11yMenuButton);
  }(A11yMenuButtonBehaviors(_index.LitElement));

  _exports.A11yMenuButton = A11yMenuButton;
  window.customElements.define(A11yMenuButton.tag, A11yMenuButton);
});
