// @webcomponents/shadycss/src/common-regex.js
define(["exports"], function (_exports) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.VAR_CONSUMED = _exports.VAR_ASSIGN = _exports.MIXIN_MATCH = _exports.MEDIA_MATCH = _exports.IS_VAR = _exports.HOST_SUFFIX = _exports.HOST_PREFIX = _exports.BRACKETED = _exports.ANIMATION_MATCH = void 0;

  /**
  @license
  Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
  This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
  The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
  The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
  Code distributed by Google as part of the polymer project is also
  subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
  */
  var VAR_ASSIGN = /(?:^|[;\s{]\s*)(--[\w-]*?)\s*:\s*(?:((?:'(?:\\'|.)*?'|"(?:\\"|.)*?"|\([^)]*?\)|[^};{])+)|\{([^}]*)\}(?:(?=[;\s}])|$))/gi;
  _exports.VAR_ASSIGN = VAR_ASSIGN;
  var MIXIN_MATCH = /(?:^|\W+)@apply\s*\(?([^);\n]*)\)?/gi;
  _exports.MIXIN_MATCH = MIXIN_MATCH;
  var VAR_CONSUMED = /(--[\w-]+)\s*([:,;)]|$)/gi;
  _exports.VAR_CONSUMED = VAR_CONSUMED;
  var ANIMATION_MATCH = /(animation\s*:)|(animation-name\s*:)/;
  _exports.ANIMATION_MATCH = ANIMATION_MATCH;
  var MEDIA_MATCH = /@media\s(.*)/;
  _exports.MEDIA_MATCH = MEDIA_MATCH;
  var IS_VAR = /^--/;
  _exports.IS_VAR = IS_VAR;
  var BRACKETED = /\{[^}]*\}/g;
  _exports.BRACKETED = BRACKETED;
  var HOST_PREFIX = '(?:^|[^.#[:])';
  _exports.HOST_PREFIX = HOST_PREFIX;
  var HOST_SUFFIX = '($|[.:[\\s>+~])';
  _exports.HOST_SUFFIX = HOST_SUFFIX;
});
