// @lrnwebcomponents/map-menu/lib/map-menu-container.js
define(["exports", "../../../lit-element/lit-element.js", "./map-menu-item.js", "./map-menu-submenu.js"], function (_exports, _litElement, _mapMenuItem, _mapMenuSubmenu) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.MapMenuContainer = void 0;

  /**
  `map-menu`
  A LRN element
  
  * @demo demo/index.html
  */
  class MapMenuContainer extends _litElement.LitElement {
    /**
     * LitElement constructable styles enhancement
     */
    static get styles() {
      return [(0, _litElement.css)`
        :host {
          display: block;
        }
        #container {
          padding: 16px 32px;
        }
        :host > ::slotted(map-menu-submenu + map-menu-submenu) {
          margin-top: 16px;
        }
      `];
    }
    /**
     * LitElement life cycle - render
     */


    render() {
      return (0, _litElement.html)` <slot></slot> `;
    }

    static get tag() {
      return "map-menu-container";
    }

  }

  _exports.MapMenuContainer = MapMenuContainer;
  window.customElements.define(MapMenuContainer.tag, MapMenuContainer);
});
