// @lrnwebcomponents/map-menu/lib/map-menu-header.js
define(["exports","require","../../../lit-element/lit-element.js","../../../@polymer/iron-collapse/iron-collapse.js","../../../@polymer/iron-icon/iron-icon.js","../../../@polymer/iron-behaviors/iron-button-state.js","../../../@polymer/paper-button/paper-button.js"],(function(e,t,i,a,n,r,o){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.MapMenuHeader=void 0,t=babelHelpers.interopRequireWildcard(t);class MapMenuHeader extends i.LitElement{static get styles(){return[i.css`
        :host {
          display: block;
          transition: 0.1s all ease-in-out;
          --map-menu-item-height: 24px;
        }
        :host([active]) {
          background: var(--map-menu-active-color);
          color: var(--map-menu-item-active-item-color, #000000);
        }
        #container {
          display: flex;
          align-items: center;
        }

        .avatarlabel {
          margin-right: 10px;
        }

        #center {
          flex: 1 1 auto;
        }

        a,
        a:hover,
        a:visited,
        a:focus {
          display: block;
          color: var(--map-menu-header-a-color, inherit);
          text-decoration: var(--map-menu-header-a-text-decoration, inherit);
        }

        lrndesign-avatar {
          display: inline-block;
          background: #fff;
          border-radius: 50%;
          box-shadow: 0 1px 1px rgba(0, 0, 0, 0.3);
          padding: 2px;
          position: relative;
          margin-top: -2px;
          transform: translateY(2px);
        }

        #link {
          display: flex;
          justify-content: flex-start;
          align-items: flex-start;
          flex-direction: column;
        }

        .title {
          font-size: var(--map-menu-font-size, 16px);
        }

        #right iron-icon {
          display: inline-block;
          color: gray;
        }

        iron-icon {
          display: inline-block;
          --iron-icon-height: var(--map-menu-item-height);
          --iron-icon-width: var(--map-menu-item-height);
        }

        paper-button {
          text-transform: none;
          width: 100%;
          justify-content: left;
          margin: 0px;
        }
      `]}render(){return i.html`
      <div id="container">
        ${this.avatarLabel?i.html`
              <div class="avatarlabel">
                <lrndesign-avatar
                  label="${this.avatarLabel}"
                ></lrndesign-avatar>
              </div>
            `:""}
        ${this.icon?i.html`
              <iron-icon icon="${this.icon}"></iron-icon>
            `:""}

        <div id="center">
          <a tabindex="-1" href="${this.url}">
            <paper-button class="title" noink>
              <div id="label">${this.label}</div>
              <div class="title">${this.title}</div>
            </paper-button>
          </a>
        </div>
        <div id="right">
          <iron-icon
            id="toggle"
            icon="${this.__collapseIcon}"
            aria-role="button"
            aria-label="${this.__collapseAria}"
            tabindex="0"
          ></iron-icon>
        </div>
      </div>
    `}static get tag(){return"map-menu-header"}constructor(){super(),this.avatarLabel="",this.url="",this.opened=!1,this.active=!1,setTimeout(()=>{this.addEventListener("click",this.__tap.bind(this)),this.addEventListener("keypress",this.__keypress.bind(this))},0)}updated(e){e.forEach((e,i)=>{"opened"==i&&this._openedChanged(this[i],e),["id","selected"].includes(i)&&this.__selectedChanged(this.selected,this.id),"avatarLabel"==i&&""!=this[i]&&new Promise((e,i)=>t.default(["../../lrndesign-avatar/lrndesign-avatar.js"],e,i))})}static get properties(){return{title:{type:String},label:{type:String},avatarLabel:{type:String,attribute:"avatar-label"},opened:{type:Boolean},url:{type:String},id:{type:String,reflect:!0},icon:{type:String},active:{type:Boolean,reflect:!0},selected:{type:String,reflect:!0},__collapseIcon:{type:String},__collapseAria:{type:String}}}_openedChanged(e,t){e?(this.__collapseIcon="icons:expand-more",this.__collapseAria="collapse menu"):(this.__collapseIcon="icons:chevron-right",this.__collapseAria="expand menu")}__selectedChanged(e,t){e===t&&this.dispatchEvent(new CustomEvent("active-item",{bubbles:!0,cancelable:!0,composed:!0,detail:this}))}__tap(e){this.__toggleEventHandler(e)}__keypress(e){"Enter"===e.code&&this.__toggleEventHandler(e)}__toggleEventHandler(e){e.path&&void 0!==e.path[0]&&void 0!==e.path[0].id&&"toggle"===e.path[0].id&&this.dispatchEvent(new CustomEvent("toggle-header",{bubbles:!0,cancelable:!0,composed:!0,detail:!0}))}}e.MapMenuHeader=MapMenuHeader,window.customElements.define(MapMenuHeader.tag,MapMenuHeader)}));
