// @polymer/paper-menu-button/paper-menu-button.js
define(["exports","../polymer/polymer-legacy.js","../iron-dropdown/iron-dropdown.js","../neon-animation/animations/fade-in-animation.js","../neon-animation/animations/fade-out-animation.js","../paper-styles/default-theme.js","../paper-styles/shadow.js","./paper-menu-button-animations.js","../iron-a11y-keys-behavior/iron-a11y-keys-behavior.js","../iron-behaviors/iron-control-state.js","../polymer/lib/legacy/polymer-fn.js","../polymer/lib/legacy/polymer.dom.js","../polymer/lib/utils/html-tag.js"],(function(_exports,_polymerLegacy,_ironDropdown,_fadeInAnimation,_fadeOutAnimation,_defaultTheme,_shadow,_paperMenuButtonAnimations,_ironA11yKeysBehavior,_ironControlState,_polymerFn,_polymerDom,_htmlTag){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.PaperMenuButton=void 0;
/**
  @license
  Copyright (c) 2015 The Polymer Project Authors. All rights reserved.
  This code may only be used under the BSD style license found at
  http://polymer.github.io/LICENSE.txt The complete set of authors may be found at
  http://polymer.github.io/AUTHORS.txt The complete set of contributors may be
  found at http://polymer.github.io/CONTRIBUTORS.txt Code distributed by Google as
  part of the polymer project is also subject to an additional IP rights grant
  found at http://polymer.github.io/PATENTS.txt
  */
var config={ANIMATION_CUBIC_BEZIER:"cubic-bezier(.3,.95,.5,1)",MAX_ANIMATION_TIME_MS:400};const PaperMenuButton=(0,_polymerFn.Polymer)({_template:_htmlTag.html`
    <style>
      :host {
        display: inline-block;
        position: relative;
        padding: 8px;
        outline: none;

        @apply --paper-menu-button;
      }

      :host([disabled]) {
        cursor: auto;
        color: var(--disabled-text-color);

        @apply --paper-menu-button-disabled;
      }

      iron-dropdown {
        @apply --paper-menu-button-dropdown;
      }

      .dropdown-content {
        @apply --shadow-elevation-2dp;

        position: relative;
        border-radius: 2px;
        background-color: var(--paper-menu-button-dropdown-background, var(--primary-background-color));

        @apply --paper-menu-button-content;
      }

      :host([vertical-align="top"]) .dropdown-content {
        margin-bottom: 20px;
        margin-top: -10px;
        top: 10px;
      }

      :host([vertical-align="bottom"]) .dropdown-content {
        bottom: 10px;
        margin-bottom: -10px;
        margin-top: 20px;
      }

      #trigger {
        cursor: pointer;
      }
    </style>

    <div id="trigger" on-tap="toggle">
      <slot name="dropdown-trigger"></slot>
    </div>

    <iron-dropdown id="dropdown" opened="{{opened}}" horizontal-align="[[horizontalAlign]]" vertical-align="[[verticalAlign]]" dynamic-align="[[dynamicAlign]]" horizontal-offset="[[horizontalOffset]]" vertical-offset="[[verticalOffset]]" no-overlap="[[noOverlap]]" open-animation-config="[[openAnimationConfig]]" close-animation-config="[[closeAnimationConfig]]" no-animations="[[noAnimations]]" focus-target="[[_dropdownContent]]" allow-outside-scroll="[[allowOutsideScroll]]" restore-focus-on-close="[[restoreFocusOnClose]]" on-iron-overlay-canceled="__onIronOverlayCanceled">
      <div slot="dropdown-content" class="dropdown-content">
        <slot id="content" name="dropdown-content"></slot>
      </div>
    </iron-dropdown>
`,is:"paper-menu-button",behaviors:[_ironA11yKeysBehavior.IronA11yKeysBehavior,_ironControlState.IronControlState],properties:{opened:{type:Boolean,value:!1,notify:!0,observer:"_openedChanged"},horizontalAlign:{type:String,value:"left",reflectToAttribute:!0},verticalAlign:{type:String,value:"top",reflectToAttribute:!0},dynamicAlign:{type:Boolean},horizontalOffset:{type:Number,value:0,notify:!0},verticalOffset:{type:Number,value:0,notify:!0},noOverlap:{type:Boolean},noAnimations:{type:Boolean,value:!1},ignoreSelect:{type:Boolean,value:!1},closeOnActivate:{type:Boolean,value:!1},openAnimationConfig:{type:Object,value:function(){return[{name:"fade-in-animation",timing:{delay:100,duration:200}},{name:"paper-menu-grow-width-animation",timing:{delay:100,duration:150,easing:config.ANIMATION_CUBIC_BEZIER}},{name:"paper-menu-grow-height-animation",timing:{delay:100,duration:275,easing:config.ANIMATION_CUBIC_BEZIER}}]}},closeAnimationConfig:{type:Object,value:function(){return[{name:"fade-out-animation",timing:{duration:150}},{name:"paper-menu-shrink-width-animation",timing:{delay:100,duration:50,easing:config.ANIMATION_CUBIC_BEZIER}},{name:"paper-menu-shrink-height-animation",timing:{duration:200,easing:"ease-in"}}]}},allowOutsideScroll:{type:Boolean,value:!1},restoreFocusOnClose:{type:Boolean,value:!0},_dropdownContent:{type:Object}},hostAttributes:{role:"group","aria-haspopup":"true"},listeners:{"iron-activate":"_onIronActivate","iron-select":"_onIronSelect"},get contentElement(){for(var nodes=(0,_polymerDom.dom)(this.$.content).getDistributedNodes(),i=0,l=nodes.length;i<l;i++)if(nodes[i].nodeType===Node.ELEMENT_NODE)return nodes[i]},toggle:function(){this.opened?this.close():this.open()},open:function(){this.disabled||this.$.dropdown.open()},close:function(){this.$.dropdown.close()},_onIronSelect:function(event){this.ignoreSelect||this.close()},_onIronActivate:function(event){this.closeOnActivate&&this.close()},_openedChanged:function(opened,oldOpened){opened?(this._dropdownContent=this.contentElement,this.fire("paper-dropdown-open")):null!=oldOpened&&this.fire("paper-dropdown-close")},_disabledChanged:function(disabled){_ironControlState.IronControlState._disabledChanged.apply(this,arguments),disabled&&this.opened&&this.close()},__onIronOverlayCanceled:function(event){var uiEvent=event.detail,trigger=this.$.trigger;(0,_polymerDom.dom)(uiEvent).path.indexOf(trigger)>-1&&event.preventDefault()}});_exports.PaperMenuButton=PaperMenuButton,Object.keys(config).forEach((function(key){PaperMenuButton[key]=config[key]}))}));
