// @lit-labs/motion/animate.js
define(["exports", "../../lit/html.js", "../../lit/directive.js", "../../lit/async-directive.js", "./animate-controller.js"], function (_exports, _html, _directive, _asyncDirective, _animateController) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.Animate = void 0;
  Object.defineProperty(_exports, "AnimateController", {
    enumerable: true,
    get: function get() {
      return _animateController.AnimateController;
    }
  });
  _exports.transformProps = _exports.none = _exports.flyRight = _exports.flyLeft = _exports.flyBelow = _exports.flyAbove = _exports.fadeOut = _exports.fadeInSlow = _exports.fadeIn = _exports.fade = _exports.defaultKeyframeOptions = _exports.defaultCssProperties = _exports.animationFrame = _exports.animate = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var h = 0;

  var r = new Map(),
      n = new WeakSet(),
      a = function a() {
    return new Promise(function (t) {
      return requestAnimationFrame(t);
    });
  },
      l = [{
    transform: "translateY(100%) scale(0)",
    opacity: 0
  }],
      c = [{
    transform: "translateY(-100%) scale(0)",
    opacity: 0
  }],
      d = [{
    transform: "translateX(-100%) scale(0)",
    opacity: 0
  }],
      u = [{
    transform: "translateX(100%) scale(0)",
    opacity: 0
  }],
      v = [{}],
      f = [{
    opacity: 0
  }],
      m = f,
      p = [{
    opacity: 0
  }, {
    opacity: 1
  }],
      y = [{
    opacity: 0
  }, {
    opacity: .25,
    offset: .75
  }, {
    opacity: 1
  }],
      g = function g(t, i) {
    var s = t - i;
    return 0 === s ? void 0 : s;
  },
      w = function w(t, i) {
    var s = t / i;
    return 1 === s ? void 0 : s;
  },
      A = {
    left: function left(t, i) {
      var s = g(t, i);
      return {
        value: s,
        transform: s && "translateX(".concat(s, "px)")
      };
    },
    top: function top(t, i) {
      var s = g(t, i);
      return {
        value: s,
        transform: s && "translateY(".concat(s, "px)")
      };
    },
    width: function width(t, i) {
      var s = w(t, i);
      return {
        value: s,
        transform: s && "scaleX(".concat(s, ")")
      };
    },
    height: function height(t, i) {
      var s = w(t, i);
      return {
        value: s,
        transform: s && "scaleY(".concat(s, ")")
      };
    }
  },
      b = {
    duration: 333,
    easing: "ease-in-out"
  },
      j = ["left", "top", "width", "height", "opacity", "color", "background"],
      x = new WeakMap();

  _exports.defaultCssProperties = j;
  _exports.defaultKeyframeOptions = b;
  _exports.transformProps = A;
  _exports.fadeInSlow = y;
  _exports.fadeIn = p;
  _exports.fade = m;
  _exports.fadeOut = f;
  _exports.none = v;
  _exports.flyRight = u;
  _exports.flyLeft = d;
  _exports.flyAbove = c;
  _exports.flyBelow = l;
  _exports.animationFrame = a;

  var S = /*#__PURE__*/function (_e) {
    babelHelpers.inherits(S, _e);

    var _super = _createSuper(S);

    function S(t) {
      var _this;

      babelHelpers.classCallCheck(this, S);
      if (_this = _super.call(this, t), _this.t = null, _this.i = null, _this.o = !0, _this.shouldLog = !1, t.type === _directive.PartType.CHILD) throw Error("The `animate` directive must be used in attribute position.");

      _this.createFinished();

      return babelHelpers.possibleConstructorReturn(_this);
    }

    babelHelpers.createClass(S, [{
      key: "createFinished",
      value: function createFinished() {
        var _this2 = this;

        var t;
        null === (t = this.resolveFinished) || void 0 === t || t.call(this), this.finished = new Promise(function (t) {
          _this2.h = t;
        });
      }
    }, {
      key: "resolveFinished",
      value: function () {
        var _resolveFinished = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
          var t;
          return regeneratorRuntime.wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  null === (t = this.h) || void 0 === t || t.call(this), this.h = void 0;

                case 1:
                case "end":
                  return _context.stop();
              }
            }
          }, _callee, this);
        }));

        function resolveFinished() {
          return _resolveFinished.apply(this, arguments);
        }

        return resolveFinished;
      }()
    }, {
      key: "render",
      value: function render(i) {
        return _html.nothing;
      }
    }, {
      key: "getController",
      value: function getController() {
        return _animateController.controllerMap.get(this.l);
      }
    }, {
      key: "isDisabled",
      value: function isDisabled() {
        var t;
        return this.options.disabled || (null === (t = this.getController()) || void 0 === t ? void 0 : t.disabled);
      }
    }, {
      key: "update",
      value: function update(t, _ref) {
        var _ref2 = babelHelpers.slicedToArray(_ref, 1),
            i = _ref2[0];

        var s;
        var e = void 0 === this.l;
        return e && (this.l = null === (s = t.options) || void 0 === s ? void 0 : s.host, this.l.addController(this), this.element = t.element, x.set(this.element, this)), this.optionsOrCallback = i, (e || "function" != typeof i) && this.u(i), this.render(i);
      }
    }, {
      key: "u",
      value: function u(t) {
        var i, s;
        t = null != t ? t : {};
        var e = this.getController();
        void 0 !== e && ((t = _objectSpread(_objectSpread({}, e.defaultOptions), t)).keyframeOptions = _objectSpread(_objectSpread({}, e.defaultOptions.keyframeOptions), t.keyframeOptions)), null !== (i = (s = t).properties) && void 0 !== i || (s.properties = j), this.options = t;
      }
    }, {
      key: "v",
      value: function v() {
        var t = {},
            i = this.element.getBoundingClientRect(),
            s = getComputedStyle(this.element);
        return this.options.properties.forEach(function (e) {
          var o;
          var h = null !== (o = i[e]) && void 0 !== o ? o : A[e] ? void 0 : s[e],
              r = Number(h);
          t[e] = isNaN(r) ? h + "" : r;
        }), t;
      }
    }, {
      key: "p",
      value: function p() {
        var t,
            i = !0;
        return this.options.guard && (t = this.options.guard(), i = function (t, i) {
          if (Array.isArray(t)) {
            if (Array.isArray(i) && i.length === t.length && t.every(function (t, s) {
              return t === i[s];
            })) return !1;
          } else if (i === t) return !1;

          return !0;
        }(t, this.m)), this.o = this.l.hasUpdated && !this.isDisabled() && !this.isAnimating() && i && this.element.isConnected, this.o && (this.m = Array.isArray(t) ? Array.from(t) : t), this.o;
      }
    }, {
      key: "hostUpdate",
      value: function hostUpdate() {
        var t;
        "function" == typeof this.optionsOrCallback && this.u(this.optionsOrCallback()), this.p() && (this.g = this.v(), this.t = null !== (t = this.t) && void 0 !== t ? t : this.element.parentNode, this.i = this.element.nextSibling);
      }
    }, {
      key: "hostUpdated",
      value: function () {
        var _hostUpdated = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2() {
          var t, i, s, e, _this$j, _s, _o, _s2, _this$j2, _o2, _n;

          return regeneratorRuntime.wrap(function _callee2$(_context2) {
            while (1) {
              switch (_context2.prev = _context2.next) {
                case 0:
                  if (!(!this.o || !this.element.isConnected || this.options.skipInitial && !this.isHostRendered)) {
                    _context2.next = 2;
                    break;
                  }

                  return _context2.abrupt("return");

                case 2:
                  this.prepare();
                  _context2.next = 5;
                  return a;

                case 5:
                  i = this.A(), s = this._(this.options.keyframeOptions, i), e = this.v();

                  if (void 0 !== this.g) {
                    _this$j = this.j(this.g, e, i), _s = _this$j.from, _o = _this$j.to;
                    this.log("measured", [this.g, e, _s, _o]), t = this.calculateKeyframes(_s, _o);
                  } else {
                    _s2 = r.get(this.options.inId);

                    if (_s2) {
                      r.delete(this.options.inId);
                      _this$j2 = this.j(_s2, e, i), _o2 = _this$j2.from, _n = _this$j2.to;
                      t = this.calculateKeyframes(_o2, _n), t = this.options.in ? [_objectSpread(_objectSpread({}, this.options.in[0]), t[0])].concat(babelHelpers.toConsumableArray(this.options.in.slice(1)), [t[1]]) : t, h++, t.forEach(function (t) {
                        return t.zIndex = h;
                      });
                    } else this.options.in && (t = [].concat(babelHelpers.toConsumableArray(this.options.in), [{}]));
                  }

                  this.animate(t, s);

                case 8:
                case "end":
                  return _context2.stop();
              }
            }
          }, _callee2, this);
        }));

        function hostUpdated() {
          return _hostUpdated.apply(this, arguments);
        }

        return hostUpdated;
      }()
    }, {
      key: "resetStyles",
      value: function resetStyles() {
        var t;
        void 0 !== this.S && (this.element.setAttribute("style", null !== (t = this.S) && void 0 !== t ? t : ""), this.S = void 0);
      }
    }, {
      key: "commitStyles",
      value: function commitStyles() {
        var t, i;
        this.S = this.element.getAttribute("style"), null === (t = this.webAnimation) || void 0 === t || t.commitStyles(), null === (i = this.webAnimation) || void 0 === i || i.cancel();
      }
    }, {
      key: "reconnected",
      value: function reconnected() {}
    }, {
      key: "disconnected",
      value: function () {
        var _disconnected = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3() {
          var t, _t, _t2, _i, _s3, i;

          return regeneratorRuntime.wrap(function _callee3$(_context3) {
            while (1) {
              switch (_context3.prev = _context3.next) {
                case 0:
                  if (this.o) {
                    _context3.next = 2;
                    break;
                  }

                  return _context3.abrupt("return");

                case 2:
                  if (!(void 0 !== this.options.id && r.set(this.options.id, this.g), void 0 === this.options.out)) {
                    _context3.next = 4;
                    break;
                  }

                  return _context3.abrupt("return");

                case 4:
                  this.prepare();
                  _context3.next = 7;
                  return a();

                case 7:
                  if (!(null === (t = this.t) || void 0 === t ? void 0 : t.isConnected)) {
                    _context3.next = 10;
                    break;
                  }

                  _t = this.i && this.i.parentNode === this.t ? this.i : null;

                  if (this.t.insertBefore(this.element, _t), this.options.stabilizeOut) {
                    _t2 = this.v();
                    this.log("stabilizing out");
                    _i = this.g.left - _t2.left, _s3 = this.g.top - _t2.top;
                    !("static" === getComputedStyle(this.element).position) || 0 === _i && 0 === _s3 || (this.element.style.position = "relative"), 0 !== _i && (this.element.style.left = _i + "px"), 0 !== _s3 && (this.element.style.top = _s3 + "px");
                  }

                case 10:
                  i = this._(this.options.keyframeOptions);
                  _context3.next = 13;
                  return this.animate(this.options.out, i);

                case 13:
                  this.element.remove();

                case 14:
                case "end":
                  return _context3.stop();
              }
            }
          }, _callee3, this);
        }));

        function disconnected() {
          return _disconnected.apply(this, arguments);
        }

        return disconnected;
      }()
    }, {
      key: "prepare",
      value: function prepare() {
        this.createFinished();
      }
    }, {
      key: "start",
      value: function start() {
        var t, i;
        null === (i = (t = this.options).onStart) || void 0 === i || i.call(t, this);
      }
    }, {
      key: "didFinish",
      value: function didFinish(t) {
        var i, s;
        t && (null === (s = (i = this.options).onComplete) || void 0 === s || s.call(i, this)), this.g = void 0, this.animatingProperties = void 0, this.frames = void 0, this.resolveFinished();
      }
    }, {
      key: "A",
      value: function A() {
        var t = [];

        for (var _i2 = this.element.parentNode; _i2; _i2 = null == _i2 ? void 0 : _i2.parentNode) {
          var _s4 = x.get(_i2);

          _s4 && !_s4.isDisabled() && _s4 && t.push(_s4);
        }

        return t;
      }
    }, {
      key: "isHostRendered",
      get: function get() {
        var _this3 = this;

        var t = n.has(this.l);
        return t || this.l.updateComplete.then(function () {
          n.add(_this3.l);
        }), t;
      }
    }, {
      key: "_",
      value: function _(t) {
        var i = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.A();

        var s = _objectSpread({}, b);

        return i.forEach(function (t) {
          return Object.assign(s, t.options.keyframeOptions);
        }), Object.assign(s, t), s;
      }
    }, {
      key: "j",
      value: function j(t, i, s) {
        t = _objectSpread({}, t), i = _objectSpread({}, i);
        var e = s.map(function (t) {
          return t.animatingProperties;
        }).filter(function (t) {
          return void 0 !== t;
        });
        var o = 1,
            h = 1;
        return void 0 !== e && (e.forEach(function (t) {
          t.width && (o /= t.width), t.height && (h /= t.height);
        }), void 0 !== t.left && void 0 !== i.left && (t.left = o * t.left, i.left = o * i.left), void 0 !== t.top && void 0 !== i.top && (t.top = h * t.top, i.top = h * i.top)), {
          from: t,
          to: i
        };
      }
    }, {
      key: "calculateKeyframes",
      value: function calculateKeyframes(t, i) {
        var s = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : !1;
        var e;
        var o = {},
            h = {};
        var r = !1;
        var n = {};

        for (var _s5 in i) {
          var _a = t[_s5],
              _l = i[_s5];

          if (_s5 in A) {
            var _t3 = A[_s5];
            if (void 0 === _a || void 0 === _l) continue;

            var _i3 = _t3(_a, _l);

            void 0 !== _i3.transform && (n[_s5] = _i3.value, r = !0, o.transform = "".concat(null !== (e = o.transform) && void 0 !== e ? e : "", " ").concat(_i3.transform));
          } else _a !== _l && void 0 !== _a && void 0 !== _l && (r = !0, o[_s5] = _a, h[_s5] = _l);
        }

        return o.transformOrigin = h.transformOrigin = s ? "center center" : "top left", this.animatingProperties = n, r ? [o, h] : void 0;
      }
    }, {
      key: "animate",
      value: function () {
        var _animate = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4(t) {
          var i,
              s,
              _e2,
              _args4 = arguments;

          return regeneratorRuntime.wrap(function _callee4$(_context4) {
            while (1) {
              switch (_context4.prev = _context4.next) {
                case 0:
                  i = _args4.length > 1 && _args4[1] !== undefined ? _args4[1] : this.options.keyframeOptions;
                  this.start(), this.frames = t;
                  s = !1;

                  if (!(!this.isAnimating() && !this.isDisabled() && (this.options.onFrames && (this.frames = t = this.options.onFrames(this), this.log("modified frames", t)), void 0 !== t))) {
                    _context4.next = 15;
                    break;
                  }

                  this.log("animate", [t, i]), s = !0, this.webAnimation = this.element.animate(t, i);
                  _e2 = this.getController();
                  null == _e2 || _e2.add(this);
                  _context4.prev = 7;
                  _context4.next = 10;
                  return this.webAnimation.finished;

                case 10:
                  _context4.next = 14;
                  break;

                case 12:
                  _context4.prev = 12;
                  _context4.t0 = _context4["catch"](7);

                case 14:
                  null == _e2 || _e2.remove(this);

                case 15:
                  return _context4.abrupt("return", (this.didFinish(s), s));

                case 16:
                case "end":
                  return _context4.stop();
              }
            }
          }, _callee4, this, [[7, 12]]);
        }));

        function animate(_x) {
          return _animate.apply(this, arguments);
        }

        return animate;
      }()
    }, {
      key: "isAnimating",
      value: function isAnimating() {
        var t, i;
        return "running" === (null === (t = this.webAnimation) || void 0 === t ? void 0 : t.playState) || (null === (i = this.webAnimation) || void 0 === i ? void 0 : i.pending);
      }
    }, {
      key: "log",
      value: function log(t, i) {
        this.shouldLog && !this.isDisabled() && console.log(t, this.options.id, i);
      }
    }]);
    return S;
  }(_asyncDirective.AsyncDirective);

  _exports.Animate = S;
  var $ = (0, _directive.directive)(S);
  _exports.animate = $;
});
