// @lit-labs/motion/position.js
define(["exports", "../../lit/html.js", "../../lit/directive.js", "../../lit/async-directive.js"], function (_exports, _html, _directive, _asyncDirective) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.position = _exports.Position = void 0;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var r = ["top", "right", "bottom", "left"];

  var e = /*#__PURE__*/function (_s) {
    babelHelpers.inherits(e, _s);

    var _super = _createSuper(e);

    function e(t) {
      var _this;

      babelHelpers.classCallCheck(this, e);
      if (_this = _super.call(this, t), t.type !== _directive.PartType.ELEMENT) throw Error("The `position` directive must be used in attribute position.");
      return babelHelpers.possibleConstructorReturn(_this);
    }

    babelHelpers.createClass(e, [{
      key: "render",
      value: function render(i, o) {
        return _html.nothing;
      }
    }, {
      key: "update",
      value: function update(t, _ref) {
        var _ref2 = babelHelpers.slicedToArray(_ref, 2),
            i = _ref2[0],
            o = _ref2[1];

        var s;
        return void 0 === this.l && (this.l = null === (s = t.options) || void 0 === s ? void 0 : s.host, this.l.addController(this)), this.$ = t.element, this.k = i, this.F = null != o ? o : ["left", "top", "width", "height"], this.render(i, o);
      }
    }, {
      key: "hostUpdated",
      value: function hostUpdated() {
        this.O();
      }
    }, {
      key: "O",
      value: function O() {
        var _this2 = this;

        var t, i;
        var o = "function" == typeof this.k ? this.k() : null === (t = this.k) || void 0 === t ? void 0 : t.value,
            s = o.offsetParent;
        if (void 0 === o || !s) return;

        var _e = o.getBoundingClientRect(),
            h = s.getBoundingClientRect();

        null === (i = this.F) || void 0 === i || i.forEach(function (t) {
          var i = r.includes(t) ? _e[t] - h[t] : _e[t];
          _this2.$.style[t] = i + "px";
        });
      }
    }]);
    return e;
  }(_asyncDirective.AsyncDirective);

  _exports.Position = e;
  var h = (0, _directive.directive)(e);
  _exports.position = h;
});
