// @vowo/chart-elements/chart-property-behavior.js
define(["exports", "../../@polymer/polymer/lib/utils/mixin.js"], function (_exports, _mixin) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ChartPropertyBehavior = void 0;

  /** @polymerMixin */
  const ChartPropertyBehavior = (0, _mixin.dedupingMixin)(function (superClass) {
    return class extends superClass {
      static get properties() {
        return {
          type: {
            type: String,
            readOnly: true,
            value: 'bar'
          },
          chart: {
            notify: true
          },
          data: {
            type: Object,
            value: function () {
              return {};
            }
          },
          options: {
            type: Object,
            value: function () {
              return {};
            }
          }
        };
      }

      static get observers() {
        return ['_configurationChanged(data.*, options.*)'];
      }

      _configurationChanged(dataRecord, optionsRecord) {
        if (dataRecord.base.labels && dataRecord.base.datasets) {
          this.hasData = true;
        } else {
          this.hasData = false;
        }

        if (this.hasData && this.isAttached) {
          this._queue();
        }
      }

    };
  });
  _exports.ChartPropertyBehavior = ChartPropertyBehavior;
});
