// @lrnwebcomponents/simple-blog/lib/simple-blog-listing.js
define(["exports","../../../lit-element/lit-element.js","../../simple-colors/simple-colors.js","../../haxcms-elements/lib/ui-components/query/site-query.js","./simple-blog-overview.js"],(function(e,i,t,s,l){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.SimpleBlogListing=void 0;class SimpleBlogListing extends t.SimpleColors{static get styles(){return[...super.styles,i.css`
        :host {
          display: block;
        }
        .listing {
          width: 100%;
          max-width: 640px;
          margin: 0 auto;
          -webkit-box-sizing: border-box;
          -moz-box-sizing: border-box;
          box-sizing: border-box;
        }
        @media only screen and (max-width: 800px) {
          .listing {
            padding: 0 32px;
          }
        }
        simple-blog-overview:not(:defined) {
          display: none;
        }
        simple-blog-overview {
          width: 100%;
          border: 1px solid #f2f2f0;
        }
      `]}static get tag(){return"simple-blog-listing"}constructor(){super(),this.__items=[]}render(){return i.html`
      <site-query
        @result-changed="${this.__itemsChanged}"
        sort='{"metadata.created": "DESC"}'
      ></site-query>
      <div class="listing">
        ${this.__items.map(e=>i.html`
            <simple-blog-overview
              item-id="${e.id}"
              title="${e.title}"
              description="${e.description}"
              link="${e.location}"
              changed="${e.metadata.created}"
            ></simple-blog-overview>
          `)}
      </div>
    `}__itemsChanged(e){this.__items=[...e.detail.value]}static get properties(){return{...super.properties,__items:{type:Array}}}}e.SimpleBlogListing=SimpleBlogListing,window.customElements.define(SimpleBlogListing.tag,SimpleBlogListing)}));
