// @lrnwebcomponents/simple-blog/lib/simple-blog-footer.js
define(["exports","require","../../../lit-element/lit-element.js","../../simple-colors/simple-colors.js","../../haxcms-elements/lib/core/haxcms-site-store.js","../../utils/utils.js","../../../mobx/lib/mobx.module.js","../../../@polymer/polymer/lib/elements/custom-style.js"],(function(t,e,i,s,o,a,n,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.SimpleBlogFooter=void 0,e=babelHelpers.interopRequireWildcard(e);class SimpleBlogFooter extends s.SimpleColors{static get styles(){return[...super.styles,i.css`
        :host {
          display: block;
          width: 100%;
          position: relative;
          overflow: hidden;
          opacity: 1;
          transition: 0.2s opacity linear;
        }
        :host([edit-mode]) {
          pointer-events: none;
          opacity: 0.2;
        }
        :host *[hidden] {
          display: none;
        }
        .background-closer-image {
          position: absolute;
          top: 0;
          left: 0;
          right: 0;
          bottom: 0;
          text-indent: -9999px;
          background-size: cover;
          background-repeat: no-repeat;
          background-position: center;
          opacity: 0.4;
          transition: all 0.6s linear;
        }
        .background-closer-image-wrap {
          position: absolute;
          background-color: rgba(0, 0, 0, 0.9);
          top: 0;
          left: 0;
          right: 0;
          bottom: 0;
          text-indent: -9999px;
          transition: all 0.6s linear;
          opacity: 1;
        }
        :host([active]) .background-closer-image {
          opacity: 0.8;
        }
        :host([active]) .background-closer-image-wrap {
          background-color: rgba(0, 0, 0, 0.2);
        }
        .inner {
          width: 100%;
          position: relative;
          z-index: 99;
          max-width: 640px;
          padding: 32px 0;
          text-align: center;
          margin: 0 auto;
        }
        .blog-title {
          margin: 0;
          padding: 0 0 10px;
          font-size: 50px;
          text-align: center;
          font-weight: 700;
          letter-spacing: -2px;
          outline: 0;
          line-height: 50px;
          word-break: break-word;
          color: white;
          text-shadow: 0 1px 16px rgba(0, 0, 0, 0.5), 0 0 1px rgba(0, 0, 0, 0.5);
        }
        .blog-description {
          margin: 0 0 50px;
          padding: 0 32px;
          font-size: 18px;
          line-height: 1.5;
          color: white;
          text-align: center;
          font-weight: 400;
          text-shadow: 0 1px 16px rgba(0, 0, 0, 0.5), 0 0 1px rgba(0, 0, 0, 0.5);
        }
        paper-button,
        site-menu-button:not([disabled]) {
          display: block;
          text-align: center;
          letter-spacing: -0.32px;
          font-size: 15px;
          font-weight: 400;
          font-style: normal;
          text-decoration: none;
          cursor: pointer;
          height: 44px;
          background-color: var(--haxcms-color, red);
          color: black;
          box-sizing: border-box;
          border-radius: none;
          line-height: 44px;
          margin: 16px 0;
          padding: 0;
        }
        paper-button:focus,
        paper-button:active,
        paper-button:hover,
        site-menu-button:focus,
        site-menu-button:active,
        site-menu-button:hover {
          outline: 2px solid var(--haxcms-color);
          outline-offset: 8px;
        }
        simple-datetime {
          letter-spacing: -0.32px;
          font-weight: 300;
          font-style: normal;
          font-size: 16px;
          line-height: 1.3;
          color: black;
          display: inline-flex;
        }
      `]}static get tag(){return"simple-blog-footer"}render(){return i.html`
      <custom-style>
        <style>
          paper-button,
          site-menu-button:not([disabled]) {
            --site-menu-button-button: {
              height: 44px;
              width: 100%;
            }
          }
        </style>
      </custom-style>
      <div class="background-closer-image-wrap">
        <div
          class="background-closer-image"
          style="background-image: url(${(0,a.varExists)(this.manifest,"metadata.theme.variables.image")?this.manifest.metadata.theme.variables.image:""})"
        ></div>
      </div>
      <div class="inner">
        <site-menu-button type="prev" position="right">
          <simple-datetime
            format="M jS"
            .timestamp="${this.prevChanged}"
            unix
            slot="suffix"
          >
          </simple-datetime>
          <span slot="suffix">${this.prevTitle}</span>
        </site-menu-button>
        <site-menu-button type="next" position="right">
          <simple-datetime
            format="M jS"
            .timestamp="${this.nextChanged}"
            unix
            slot="prefix"
          ></simple-datetime>
          <span slot="prefix">${this.nextTitle}</span>
        </site-menu-button>
        <paper-button raised @click="${this._backButtonTap}"
          >Back to list</paper-button
        >
        <h2 class="blog-title">
          ${(0,a.varExists)(this.manifest,"title")?this.manifest.title:""}
        </h2>
        <h3 class="blog-description">
          ${(0,a.varExists)(this.manifest,"description")?this.manifest.description:""}
        </h3>
      </div>
    `}updated(t){t.forEach((t,e)=>{"activeManifestIndex"==e&&this._activeManifestIndexChanged(this[e],t)})}static get properties(){return{...super.properties,activeManifestIndex:{type:Number,attribute:"active-manifest-index"},manifest:{type:Object},editMode:{type:Boolean,reflect:!0,attribute:"edit-mode"},prevChanged:{type:String},nextChanged:{type:String},prevTitle:{type:String},nextTitle:{type:String}}}_activeManifestIndexChanged(t){this.manifest&&(this.manifest.items[t-1]?(this.prevTitle=" - "+this.manifest.items[t-1].title,this.prevChanged=this.manifest.items[t-1].metadata.created):(this.prevTitle="",this.prevChanged=""),this.manifest.items[t+1]?(this.nextTitle=" - "+this.manifest.items[t+1].title,this.nextChanged=this.manifest.items[t+1].metadata.created):(this.nextTitle="",this.nextChanged=""))}constructor(){super(),this.prevTitle="",this.prevChanged="",this.nextTitle="",this.nextChanged="",new Promise((t,i)=>e.default(["../../haxcms-elements/lib/ui-components/navigation/site-menu-button.js"],t,i)),new Promise((t,i)=>e.default(["../../../@polymer/paper-button/paper-button.js"],t,i)),new Promise((t,i)=>e.default(["../../simple-datetime/simple-datetime.js"],t,i))}connectedCallback(){super.connectedCallback(),this.__disposer=[],(0,n.autorun)(t=>{if(this.manifest=(0,n.toJS)(o.store.routerManifest),(0,a.varExists)(this.manifest,"title")&&(document.title=this.manifest.title),(0,a.varExists)(this.manifest,"metadata.theme.variables.cssVariable")){let t=this.manifest.metadata.theme.variables.cssVariable.replace("--simple-colors-default-theme-","").split("-");t.pop(),this.accentColor=t.join("-")}this.__disposer.push(t)}),(0,n.autorun)(t=>{this.activeManifestIndex=(0,n.toJS)(o.store.activeManifestIndex),this.__disposer.push(t)}),(0,n.autorun)(t=>{this.editMode=(0,n.toJS)(o.store.editMode),this.__disposer.push(t)})}disconnectedCallback(){for(var t in this.__disposer)this.__disposer[t].dispose();super.disconnectedCallback()}_backButtonTap(t){window.history.pushState(null,null,o.store.location.baseUrl),window.dispatchEvent(new PopStateEvent("popstate"))}}t.SimpleBlogFooter=SimpleBlogFooter,window.customElements.define(SimpleBlogFooter.tag,SimpleBlogFooter)}));
