// @lrnwebcomponents/simple-modal/lib/simple-modal-template.js
define(["exports", "../../../lit-element/lit-element.js", "../simple-modal.js"], function (_exports, _litElement, _simpleModal) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SimpleModalTemplate = void 0;

  function _templateObject2_0b92d5e0b4c911ea90ffad65c66d8e0c() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <slot name=\"header\"></slot> <slot name=\"content\"></slot>\n      <slot name=\"buttons\"></slot>\n    "]);

    _templateObject2_0b92d5e0b4c911ea90ffad65c66d8e0c = function _templateObject2_0b92d5e0b4c911ea90ffad65c66d8e0c() {
      return data;
    };

    return data;
  }

  function _templateObject_0b92d5e0b4c911ea90ffad65c66d8e0c() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: none;\n          --simple-modal-width: auto;\n          --simple-modal-height: auto;\n          --simple-modal-min-width: unset;\n          --simple-modal-min-height: unset;\n          --simple-modal-max-width: unset;\n          --simple-modal-max-height: unset;\n          --simple-modal-titlebar-color: #444;\n          --simple-modal-titlebar-background: #ddd;\n          --simple-modal-header-color: #222;\n          --simple-modal-header-background: #ccc;\n          --simple-modal-content-container-color: #222;\n          --simple-modal-content-container-background: #fff;\n          --simple-modal-buttons-color: blue;\n          --simple-modal-buttons-background: #fff;\n          --simple-modal-button-color: var(--simple-modal-buttons-color);\n          --simple-modal-button-background: var(\n            --simple-modal-buttons-background-color\n          );\n        }\n      "]);

    _templateObject_0b92d5e0b4c911ea90ffad65c66d8e0c = function _templateObject_0b92d5e0b4c911ea90ffad65c66d8e0c() {
      return data;
    };

    return data;
  }

  /**
   * `simple-modal-template`
   * `A simple modal that ensures accessibility and stack order context appropriately`
   * @demo ./demo/template.html
   * @element simple-modal-template
   */
  var SimpleModalTemplate =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(SimpleModalTemplate, _LitElement);
    babelHelpers.createClass(SimpleModalTemplate, null, [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject_0b92d5e0b4c911ea90ffad65c66d8e0c())];
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       */

    }, {
      key: "tag",
      get: function get() {
        return "simple-modal-template";
      }
    }]);

    function SimpleModalTemplate() {
      var _this;

      babelHelpers.classCallCheck(this, SimpleModalTemplate);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(SimpleModalTemplate).call(this));
      _this.title = "";
      _this.modal = window.SimpleModal.requestAvailability();
      return _this;
    } //render function


    babelHelpers.createClass(SimpleModalTemplate, [{
      key: "render",
      //render function
      value: function render() {
        return (0, _litElement.html)(_templateObject2_0b92d5e0b4c911ea90ffad65c66d8e0c());
      }
      /**
       * sets event listeners for a specified target
       *
       * @param {object} the object that will have the event listener
       * @param {string} the event name
       * @param {boolean} whether the event bubbles (default is true)
       * @param {boolean} whether the event can be canceled (default is true)
       * @returns {object} the modal object
       */

    }, {
      key: "associateEvents",
      value: function associateEvents(target) {
        var _this2 = this;

        var evt = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "click";
        var bubbles = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
        var cancelable = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : true;
        target.addEventListener(evt, function (e) {
          _this2.openModal(target, bubbles, cancelable);
        });
        return this.modal;
      }
      /**
       * dispatches event to populate and open the simple modal based template values
       *
       * @param {object} the object that will have the event listener
       * @param {boolean} whether the event bubbles (default is true)
       * @param {boolean} whether the event can be canceled (default is true)
       */

    }, {
      key: "openModal",
      value: function openModal(target) {
        var bubbles = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
        var cancelable = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
        var tplStyles = getComputedStyle(this),
            styles = {};
        ["--simple-modal-width", "--simple-modal-height", "--simple-modal-min-width", "--simple-modal-min-height", "--simple-modal-max-width", "--simple-modal-max-height", "--simple-modal-titlebar-color", "--simple-modal-titlebar-background", "--simple-modal-header-color", "--simple-modal-header-background", "--simple-modal-content-container-color", "--simple-modal-content-container-background", "--simple-modal-buttons-color", "--simple-modal-buttons-background", "--simple-modal-button-color", "--simple-modal-button-background", "--simple-modal-titlebar-button-border", "--simple-modal-titlebar-button-padding", "--simple-modal-titlebar-button-outline", "--simple-modal-titlebar-button-outline-offset", "--simple-modal-titlebar-icon-width", "--simple-modal-titlebar-icon-height"].forEach(function (prop) {
          styles[prop] = tplStyles.getPropertyValue(prop);
        });
        var evt = new CustomEvent("simple-modal-show", {
          bubbles: bubbles,
          cancelable: cancelable,
          composed: true,
          detail: {
            id: this.getAttribute("modal-id"),
            elements: {
              header: this._getSlot("header"),
              content: this._getSlot("content"),
              buttons: this._getSlot("buttons")
            },
            invokedBy: target,
            modalClass: this.getAttribute("class"),
            styles: styles,
            clone: false,
            title: this.title !== null ? this.title : false
          }
        });
        window.dispatchEvent(evt);
      }
      /**
       * clones content in a named slot
       *
       * @param {string} the name of the slot
       * @returns {object} a clone of the slotted content (or false if there is no slotted content)
       */

    }, {
      key: "_getSlot",
      value: function _getSlot(slotName) {
        var slot = this.querySelectorAll('[slot="' + slotName + '"]'); // account for slot passing down from parent element

        if (slot && slot[0] && slot[0].tagName == "SLOT") {
          slot = slot[0].assignedNodes({
            flatten: true
          });
        }

        var container = document.createElement("div");
        slot.forEach(function (el) {
          container.appendChild(el.cloneNode(true));
        });
        return slot !== null ? container.cloneNode(true) : false;
      }
    }], [{
      key: "properties",
      get: function get() {
        return {
          /**
           * the simple-modal
           */
          modal: {
            type: Object
          },

          /**
           * the modal title
           */
          title: {
            type: String
          }
        };
      }
    }]);
    return SimpleModalTemplate;
  }(_litElement.LitElement);

  _exports.SimpleModalTemplate = SimpleModalTemplate;
  window.customElements.define(SimpleModalTemplate.tag, SimpleModalTemplate);
});
