// scrollable-component/index.js
define(["exports"], function (_exports) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ScrollableComponentElement = void 0;

  function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

  function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

  function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var r = document.createElement("template");
  r.innerHTML = '<style>*{box-sizing:border-box}:host{--fade-in-transition-duration:150ms;--fade-out-transition-duration:800ms;--fade-out-transition-delay:300ms;--fill-color-transition-duration:150ms;--viewport-overflow-x:auto;--viewport-overflow-y:auto;--viewport-scroll-snap-type:none;--viewport-scroll-behavior:auto;--viewport-overscroll-behavior:auto;--viewport-z-index:0;--scrollbar-width:16px;--scrollbar-padding:2px;--scrollbar-fill-color:transparent;--scrollbar-fill-color-hover:transparent;--scrollbar-border-width:0;--scrollbar-border-style:none;--scrollbar-border-color:#999;--scrollbar-border-radius:0;--scrollbar-box-shadow:none;--scrollbar-z-index-hover:30;--vertical-scrollbar-padding:var(--scrollbar-padding);--vertical-scrollbar-background:none;--vertical-scrollbar-background-size:auto;--vertical-scrollbar-z-index:20;--horizontal-scrollbar-padding:var(--scrollbar-padding);--horizontal-scrollbar-background:none;--horizontal-scrollbar-background-size:auto;--horizontal-scrollbar-z-index:10;--scrollbar-track-fill-color:transparent;--scrollbar-track-fill-color-hover:transparent;--scrollbar-track-border-width:0;--scrollbar-track-border-style:none;--scrollbar-track-border-color:#999;--scrollbar-track-border-radius:0;--scrollbar-track-box-shadow:none;--vertical-scrollbar-track-background:none;--vertical-scrollbar-track-background-size:auto;--horizontal-scrollbar-track-background:none;--horizontal-scrollbar-track-background-size:auto;--scrollbar-thumb-fill-color:#ccc;--scrollbar-thumb-fill-color-hover:#aaa;--scrollbar-thumb-border-width:0;--scrollbar-thumb-border-style:none;--scrollbar-thumb-border-color:#999;--scrollbar-thumb-border-radius:var(--scrollbar-width);--scrollbar-thumb-box-shadow:none;--vertical-scrollbar-thumb-background:none;--vertical-scrollbar-thumb-background-size:auto;--horizontal-scrollbar-thumb-background:none;--horizontal-scrollbar-thumb-background-size:auto;--content-padding:0;position:relative;overflow:hidden;display:grid;grid-template:1fr/1fr}.viewport{z-index:var(--viewport-z-index);display:grid;overflow-x:var(--viewport-overflow-x);overflow-y:var(--viewport-overflow-y);scrollbar-width:none;outline:0;scroll-behavior:var(--viewport-scroll-behavior);overscroll-behavior:var(--viewport-overscroll-behavior);scroll-snap-type:var(--viewport-scroll-snap-type)}.viewport::-webkit-scrollbar{width:0;height:0}.content{padding:var(--content-padding)}.content.restrict-width{width:var(--viewport-width)}.content.restrict-height{height:var(--viewport-height)}.scrollbar{user-select:none;touch-action:none;position:absolute;border-width:var(--scrollbar-border-width);border-style:var(--scrollbar-border-style);border-color:var(--scrollbar-border-color);border-radius:var(--scrollbar-border-radius);box-shadow:var(--scrollbar-box-shadow);opacity:0;transition:opacity var(--fade-out-transition-duration) ease-in-out var(--fade-out-transition-delay),background-color var(--fill-color-transition-duration) ease-out}.vertical-scrollbar{z-index:var(--vertical-scrollbar-z-index);width:var(--scrollbar-width);right:0;top:0;bottom:0;padding:var(--vertical-scrollbar-padding);background:var(--vertical-scrollbar-background);background-color:var(--scrollbar-fill-color);background-size:var(--vertical-scrollbar-background-size)}.vertical-scrollbar.left-position{left:0;right:auto}.horizontal-scrollbar{z-index:var(--horizontal-scrollbar-z-index);height:var(--scrollbar-width);left:0;right:0;bottom:0;padding:var(--horizontal-scrollbar-padding);background:var(--horizontal-scrollbar-background);background-color:var(--scrollbar-fill-color);background-size:var(--horizontal-scrollbar-background-size)}.horizontal-scrollbar.top-position{top:0;bottom:auto}.scrollbar:hover,.scrollbar.scrolling-with-thumb,.viewport:hover~.scrollbar,.viewport:not(:focus):focus-within~.scrollbar,.viewport.touch~.scrollbar{opacity:1;transition:opacity var(--fade-in-transition-duration) ease-in-out 0s,background-color var(--fill-color-transition-duration) ease-out}.viewport.scrollbar-visible~.scrollbar{opacity:1;transition:none}.scrollbar:hover,.scrollbar.scrolling-with-thumb{z-index:var(--scrollbar-z-index-hover);background-color:var(--scrollbar-fill-color-hover)}.scrollbar.hidden{display:none}.scrollbar .scrollbar-track{height:100%;width:100%;border-width:var(--scrollbar-track-border-width);border-style:var(--scrollbar-track-border-style);border-color:var(--scrollbar-track-border-color);border-radius:var(--scrollbar-track-border-radius);box-shadow:var(--scrollbar-track-box-shadow);transition:background-color var(--fill-color-transition-duration) ease-out}.vertical-scrollbar .scrollbar-track{background:var(--vertical-scrollbar-track-background);background-color:var(--scrollbar-track-fill-color);background-size:var(--vertical-scrollbar-track-background-size)}.horizontal-scrollbar .scrollbar-track{background:var(--horizontal-scrollbar-track-background);background-color:var(--scrollbar-track-fill-color);background-size:var(--horizontal-scrollbar-track-background-size)}.scrollbar-track:hover,.scrollbar.scrolling-with-thumb .scrollbar-track{background-color:var(--scrollbar-track-fill-color-hover)}.scrollbar .scrollbar-thumb{height:100%;width:100%;border-width:var(--scrollbar-thumb-border-width);border-style:var(--scrollbar-thumb-border-style);border-color:var(--scrollbar-thumb-border-color);border-radius:var(--scrollbar-thumb-border-radius);transform:translate3d(0,0,0);box-shadow:var(--scrollbar-thumb-box-shadow);transition:background-color var(--fill-color-transition-duration) ease-out}.vertical-scrollbar .scrollbar-thumb{background:var(--vertical-scrollbar-thumb-background);background-color:var(--scrollbar-thumb-fill-color);background-size:var(--vertical-scrollbar-thumb-background-size)}.horizontal-scrollbar .scrollbar-thumb{background:var(--horizontal-scrollbar-thumb-background);background-color:var(--scrollbar-thumb-fill-color);background-size:var(--horizontal-scrollbar-thumb-background-size)}.scrollbar .scrollbar-thumb:hover,.scrollbar.scrolling-with-thumb .scrollbar-thumb{background-color:var(--scrollbar-thumb-fill-color-hover)}</style><div class="viewport" tabindex="-1"><div class="content"><slot></slot></div></div><div class="scrollbar vertical-scrollbar hidden"><div class="scrollbar-track"><div class="scrollbar-thumb"></div></div></div><div class="scrollbar horizontal-scrollbar hidden"><div class="scrollbar-track"><div class="scrollbar-thumb"></div></div></div>';
  var o = Symbol("vertical"),
      l = Symbol("horizontal"),
      t = [{
    key: o,
    name: "vertical",
    size: "height",
    o: "offsetHeight",
    l: "scrollHeight",
    position: "top",
    t: "scrollTop",
    i: "overflow-y",
    h: "clientY",
    v: "pageY"
  }, {
    key: l,
    name: "horizontal",
    size: "width",
    o: "offsetWidth",
    l: "scrollWidth",
    position: "left",
    t: "scrollLeft",
    i: "overflow-x",
    h: "clientX",
    v: "pageX"
  }];

  var ScrollableComponentElement = /*#__PURE__*/function (_HTMLElement) {
    babelHelpers.inherits(ScrollableComponentElement, _HTMLElement);

    var _super = _createSuper(ScrollableComponentElement);

    function ScrollableComponentElement() {
      var _this;

      babelHelpers.classCallCheck(this, ScrollableComponentElement);
      _this = _super.call(this), _this.u(), _this.p(), _this.g();
      return _this;
    }

    babelHelpers.createClass(ScrollableComponentElement, [{
      key: "connectedCallback",
      value: function connectedCallback() {
        this.k(), this.m(), this.S();
      }
    }, {
      key: "attributeChangedCallback",
      value: function attributeChangedCallback(r, t, a) {
        "scrollbar-visibility" === r ? this.viewport.classList.toggle("scrollbar-visible", "always" === a) : "vertical-scrollbar-position" === r ? this.elements[o].C.classList.toggle("left-position", "left" === a) : "horizontal-scrollbar-position" === r && this.elements[l].C.classList.toggle("top-position", "top" === a);
      }
    }, {
      key: "u",
      value: function u() {
        this.viewport, this.content, this.elements = {}, this.cache = {}, this.T = {}, this.$ = null;

        var _iterator = _createForOfIteratorHelper(t),
            _step;

        try {
          for (_iterator.s(); !(_step = _iterator.n()).done;) {
            var _r = _step.value;
            this.elements[_r.key] = {
              C: null,
              W: null,
              T: null
            }, this.cache[_r.key] = {
              A: !1,
              F: !1,
              M: 0,
              O: 0,
              P: 1
            }, this.T[_r.key] = {
              q: !0,
              L: 1,
              R: {
                v: 0,
                t: 0
              }
            };
          }
        } catch (err) {
          _iterator.e(err);
        } finally {
          _iterator.f();
        }
      }
    }, {
      key: "p",
      value: function p() {
        this.attachShadow({
          mode: "open"
        }), this.shadowRoot.appendChild(r.content.cloneNode(!0)), this.viewport = this.shadowRoot.querySelector(".viewport"), this.content = this.viewport.querySelector(".content");

        var _iterator2 = _createForOfIteratorHelper(t),
            _step2;

        try {
          for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
            var _r2 = _step2.value;
            var _o = this.elements[_r2.key];
            _o.C = this.shadowRoot.querySelector(".".concat(_r2.name, "-scrollbar")), _o.W = _o.C.querySelector(".scrollbar-track"), _o.T = _o.W.querySelector(".scrollbar-thumb");
          }
        } catch (err) {
          _iterator2.e(err);
        } finally {
          _iterator2.f();
        }
      }
    }, {
      key: "g",
      value: function g() {
        var _this2 = this;

        var r = new ResizeObserver(function () {
          _this2.H();
        });
        r.observe(this.viewport), r.observe(this.content);

        var _iterator3 = _createForOfIteratorHelper(t),
            _step3;

        try {
          var _loop = function _loop() {
            var o = _step3.value;
            var l = _this2.elements[o.key];
            r.observe(l.W), l.W.addEventListener("pointerdown", function (r) {
              r.preventDefault(), r.stopPropagation(), _this2.D(o, r);
            });

            var t = function t(r) {
              _this2.X(o, r);
            };

            l.T.addEventListener("pointerdown", function (r) {
              r.preventDefault(), r.stopPropagation(), l.T.addEventListener("pointermove", t, {
                passive: !0
              }), l.T.setPointerCapture(r.pointerId), _this2.Y(o, r);
            }), l.T.addEventListener("pointerup", function (r) {
              l.T.removeEventListener("pointermove", t, {
                passive: !0
              }), l.T.releasePointerCapture(r.pointerId), _this2.j(o);
            }, {
              passive: !0
            });
          };

          for (_iterator3.s(); !(_step3 = _iterator3.n()).done;) {
            _loop();
          }
        } catch (err) {
          _iterator3.e(err);
        } finally {
          _iterator3.f();
        }

        this.viewport.addEventListener("scroll", function () {
          _this2.B();
        }, {
          passive: !0
        }), this.viewport.addEventListener("touchstart", function () {
          _this2.G();
        }, {
          passive: !0
        }), this.viewport.addEventListener("touchend", function () {
          _this2.I();
        }, {
          passive: !0
        });
      }
    }, {
      key: "D",
      value: function D(r, o) {
        var _this3 = this;

        var l = this.elements[r.key],
            t = this.cache[r.key],
            a = l.W.getBoundingClientRect(),
            i = l.T.getBoundingClientRect(),
            s = o[r.h] - a[r.position] - i[r.size] / 2,
            c = t.O / a[r.size],
            e = s / t.P * c;
        requestAnimationFrame(function () {
          var _this3$viewport$scrol;

          _this3.viewport.scrollTo((_this3$viewport$scrol = {}, babelHelpers.defineProperty(_this3$viewport$scrol, r.position, e), babelHelpers.defineProperty(_this3$viewport$scrol, "behavior", "smooth"), _this3$viewport$scrol)), _this3.viewport.focus({
            preventScroll: !0
          });
        });
      }
    }, {
      key: "Y",
      value: function Y(r, o) {
        var _this4 = this;

        var l = this.elements[r.key],
            t = this.T[r.key],
            a = l.W.getBoundingClientRect();
        t.q = !0, t.L = this.cache[r.key].O / a[r.size], t.R.v = o.touches ? o.touches[0][r.v] : o[r.v], t.R.t = this.viewport[r.t], requestAnimationFrame(function () {
          _this4.viewport.classList.add("scrolling-with-".concat(r.name, "-thumb")), l.C.classList.add("scrolling-with-thumb"), _this4.viewport.focus({
            preventScroll: !0
          });
        });
      }
    }, {
      key: "X",
      value: function X(r, o) {
        var l = this.T[r.key];

        if (l.q) {
          var _t = ((o.touches ? o.touches[0][r.v] : o[r.v]) - l.R.v) / this.cache[r.key].P * l.L,
              a = l.R.t + _t;

          this.viewport[r.t] = a;
        }
      }
    }, {
      key: "j",
      value: function j(r) {
        var o = this.T[r.key];
        o.q && (o.q = !1, this.viewport.classList.remove("scrolling-with-".concat(r.name, "-thumb")), this.elements[r.key].C.classList.remove("scrolling-with-thumb"));
      }
    }, {
      key: "G",
      value: function G() {
        this.viewport.classList.add("touch");
      }
    }, {
      key: "I",
      value: function I() {
        this.viewport.classList.remove("touch");
      }
    }, {
      key: "B",
      value: function B() {
        var _this5 = this;

        null === this.$ && (this.$ = requestAnimationFrame(function () {
          _this5.S(), _this5.$ = null;
        }));
      }
    }, {
      key: "H",
      value: function H() {
        var _this6 = this;

        this.k(), null === this.$ && (this.$ = requestAnimationFrame(function () {
          _this6.m(), _this6.S(), _this6.$ = null;
        }));
      }
    }, {
      key: "k",
      value: function k() {
        var r = getComputedStyle(this.viewport);

        var _iterator4 = _createForOfIteratorHelper(t),
            _step4;

        try {
          for (_iterator4.s(); !(_step4 = _iterator4.n()).done;) {
            var _o2 = _step4.value;
            var _l = this.cache[_o2.key];
            _l.F = "hidden" === r.getPropertyValue(_o2.i).trim(), _l.M = Math.floor(10 * this.viewport[_o2.o]) / 10, _l.O = Math.floor(10 * this.elements[_o2.key].W[_o2.o]) / 10, _l.P = _l.O / this.viewport[_o2.l], _l.A = !_l.F && this.viewport[_o2.l] > Math.ceil(_l.M);
          }
        } catch (err) {
          _iterator4.e(err);
        } finally {
          _iterator4.f();
        }
      }
    }, {
      key: "m",
      value: function m() {
        var _iterator5 = _createForOfIteratorHelper(t),
            _step5;

        try {
          for (_iterator5.s(); !(_step5 = _iterator5.n()).done;) {
            var _r3 = _step5.value;
            var _o3 = this.elements[_r3.key],
                _l2 = this.cache[_r3.key];

            if (this.shadowRoot.host.style.setProperty("--viewport-".concat(_r3.size), "".concat(_l2.M, "px")), this.content.classList.toggle("restrict-".concat(_r3.size), _l2.F), _o3.C.classList.toggle("hidden", !_l2.A), _l2.A) {
              var _t2 = _l2.M * _l2.P;

              _o3.T.style[_r3.size] = "".concat(_t2, "px");
            }
          }
        } catch (err) {
          _iterator5.e(err);
        } finally {
          _iterator5.f();
        }
      }
    }, {
      key: "S",
      value: function S() {
        var _iterator6 = _createForOfIteratorHelper(t),
            _step6;

        try {
          for (_iterator6.s(); !(_step6 = _iterator6.n()).done;) {
            var _r4 = _step6.value;
            var _l3 = this.cache[_r4.key];

            if (_l3.A) {
              var _t3 = this.viewport[_r4.t] * _l3.P;

              this.elements[_r4.key].T.style.transform = _r4.key === o ? "translate3D(0, ".concat(_t3, "px, 0)") : "translate3D(".concat(_t3, "px, 0, 0)");
            }
          }
        } catch (err) {
          _iterator6.e(err);
        } finally {
          _iterator6.f();
        }
      }
    }], [{
      key: "observedAttributes",
      get: function get() {
        return ["scrollbar-visibility", "vertical-scrollbar-position", "horizontal-scrollbar-position"];
      }
    }]);
    return ScrollableComponentElement;
  }( /*#__PURE__*/babelHelpers.wrapNativeSuper(HTMLElement));

  _exports.ScrollableComponentElement = ScrollableComponentElement;
  window.customElements.define("scrollable-component", ScrollableComponentElement);
});
