// @lrnwebcomponents/haxcms-elements/lib/core/utils/HAXCMSOperationButtons.js
import{css as t,html as e}from"../../../../../lit/index.js";import{store as i}from"../haxcms-site-store.js";export const HAXCMSOperationButtons=function(o){return class extends o{constructor(){super(),this.t||(this.t={}),this.t&&(this.t={...this.t,edit:"Edit",save:"Save",cancel:"Cancel",editStack:"Edit stack",import:"Import",archive:"Archive"})}static get styles(){let e=[];return super.styles&&(e=super.styles),[...e,t`
          .operation-buttons-wrapper {
            display: flex;
            justify-content: flex-end;
            height: 0px;
            margin: 0;
            padding: 0;
            color: var(--simple-colors-default-theme-orange-7, orange);
            background-color: var(
              --simple-colors-default-theme-orange-1,
              white
            );
          }
          .operation-buttons-wrapper .btn {
            display: inline-flex;
            margin: 0 16px 0 0;
            height: 12px;
            max-width: 200px;
            opacity: 0.8;
            transition: all 0.3s ease-in-out;
          }
          .operation-buttons-wrapper .btn:hover,
          .operation-buttons-wrapper .btn:focus {
            opacity: 0.9;
          }
          .operation-buttons-wrapper .btn:active {
            opacity: 1;
            color: #000000;
          }
          .operation-buttons-wrapper .btn[part="edit-mode-active"] {
            opacity: 1;
          }
        `]}HAXCMSRenderOperationButtons(){return import("../../../../simple-icon/simple-icon.js"),import("../../../../simple-icon/lib/simple-icons.js"),import("../../../../simple-icon/lib/simple-icon-button-lite.js"),e`
        <div class="operation-buttons-wrapper">
          <simple-icon-button-lite
            class="btn ${this.editMode?"edit-mode-active":""}"
            icon="lrn:write"
            @click="${this._editButtonTap}"
            >${this.editMode?this.t.save:this.t.edit}</simple-icon-button-lite
          >
          <simple-icon-button-lite
            class="btn"
            icon="lrn:content"
            @click="${this._editButtonTap}"
            >${this.t.editStack}</simple-icon-button-lite
          >
          <simple-icon-button-lite class="btn" icon="icons:archive"
            >${this.t.archive}</simple-icon-button-lite
          >
          <simple-icon-button-lite class="btn" icon="icons:file-upload"
            >${this.t.import}</simple-icon-button-lite
          >
        </div>
      `}_editButtonTap(t){this.editMode=!this.editMode,this.editMode||this.dispatchEvent(new CustomEvent("haxcms-save-node",{bubbles:!0,composed:!0,cancelable:!1,detail:i.activeItem})),window.dispatchEvent(new CustomEvent("simple-modal-hide",{bubbles:!0,cancelable:!0,detail:{}}))}_editDetailsButtonTap(t){const e=new CustomEvent("haxcms-load-node-fields",{bubbles:!0,composed:!0,cancelable:!1,detail:t.target});window.dispatchEvent(e)}_cancelButtonTap(t){this.editMode=!1,this.dispatchEvent(new CustomEvent("hax-cancel",{bubbles:!0,composed:!0,cancelable:!1,detail:t.detail})),window.dispatchEvent(new CustomEvent("simple-modal-hide",{bubbles:!0,cancelable:!0,detail:{}}))}}};
