// @lrnwebcomponents/lrn-content/lrn-content.js
define(["exports", "../../lit-element/lit-element.js"], function (_exports, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnContent = void 0;

  /**
  `lrn-content`
    A LRN element for presenting content with a simple heading.
    This is to improve accessibility, consistency, and tag things
    with OER schema.
  * @demo demo/index.html
  */
  class LrnContent extends _litElement.LitElement {
    /**
     * LitElement constructable styles enhancement
     */
    static get styles() {
      return [(0, _litElement.css)`
        :host {
          display: block;
        }
      `];
    }

    render() {
      return (0, _litElement.html)`
      <div typeof="oer:SupportingMaterial">
        ${this.title ? (0, _litElement.html)`
              <h2 property="oer:name">${this.title}</h2>
            ` : ``}
        <div property="oer:description"><slot></slot></div>
      </div>
    `;
    }

    static get tag() {
      return "lrn-content";
    }

    static get properties() {
      return {
        title: {
          type: String
        }
      };
    }

  }

  _exports.LrnContent = LrnContent;
  window.customElements.define(LrnContent.tag, LrnContent);
});
