// @lrnwebcomponents/lrn-css-reset/lrn-css-reset.js
define(["exports", "../../lit/index.js"], function (_exports, _index) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnCssResetStyles = _exports.LrnCssReset = _exports.AttachStylesToHead = void 0;

  var _templateObject_1cdd2360942811ecb1937969798b231a, _templateObject2_1cdd2360942811ecb1937969798b231a;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * CSS reset and additional base styles for lrnwebcomponents and apps
   */
  var LrnCssResetStyles = [(0, _index.css)(_templateObject_1cdd2360942811ecb1937969798b231a || (_templateObject_1cdd2360942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n    @import url(\"https://unpkg.com/modern-css-reset/dist/reset.min.css\");\n    @import url(\"https://fonts.googleapis.com/css?family=Lato:400,700,900|Source+Code+Pro:400,700|Lora:400,400i,700,700i|Playfair+Display:700i,900\");\n\n    .off-screen {\n      position: absolute;\n      left: -9999999px;\n      width: 0px;\n      height: 0px;\n      overflow: hidden;\n    }\n\n    html,\n    body {\n      margin: 0;\n      width: 100%;\n      font-size: var(--haxcms-base-styles-body-font-size, 18px);\n      font-family: var(--haxcms-base-styles-body-font-family, \"Lora\");\n      line-height: var(--haxcms-base-styles-body-line-height, 1.58);\n      letter-spacing: var(--haxcms-base-styles-body-letter-spacing, -0.03px);\n    }\n\n    h1,\n    h2,\n    p,\n    i,\n    a,\n    .first-letter,\n    .authorName a {\n      color: rgba(0, 0, 0, 0.84);\n      text-rendering: optimizeLegibility;\n    }\n\n    h1 {\n      font-family: var(\n        --hax-base-styles-h1-font-family,\n        var(--hax-base-styles-accent-font-family, \"Playfair Display\", serif)\n      );\n      font-size: var(\n        --hax-base-styles-h1-font-size,\n        calc(2.66666667 * var(--haxcms-base-styles-body-font-size, 18px))\n      );\n      line-height: var(--hax-base-styles-h1-line-height, unset);\n      text-align: left;\n      margin-bottom: 8px;\n      letter-spacing: unset;\n    }\n\n    h2 {\n      font-family: var(\n        --hax-base-styles-h2-font-family,\n        var(--hax-base-styles-heading-font-family, \"Lato\", sans-serif)\n      );\n      font-size: var(\n        --hax-base-styles-h2-font-size,\n        calc(1.66666667 * var(--haxcms-base-styles-body-font-size, 18px))\n      );\n      line-height: var(--hax-base-styles-h2-line-height, 34.5px);\n      font-weight: 900;\n      padding: 0;\n      margin: 56px 0 -13px -1.883px;\n      text-align: left;\n      letter-spacing: -0.45px;\n    }\n\n    h3 {\n      font-family: var(\n        --hax-base-styles-h3-font-family,\n        var(--hax-base-styles-heading-font-family, \"Lato\", sans-serif)\n      );\n      font-size: var(\n        --hax-base-styles-h3-font-size,\n        calc(1.44444444 * var(--haxcms-base-styles-body-font-size, 18px))\n      );\n      line-height: var(--hax-base-styles-h3-line-height);\n      font-weight: 700;\n      margin: 56px 0 -13px -1.883px;\n      color: #555;\n    }\n\n    h4 {\n      font-family: var(\n        --hax-base-styles-h4-font-family,\n        var(--hax-base-styles-heading-font-family, \"Lato\", sans-serif)\n      );\n      font-size: var(\n        --hax-base-styles-h4-font-size,\n        calc(1.22222222 * var(--haxcms-base-styles-body-font-size, 18px))\n      );\n      line-height: var(--hax-base-styles-h4-line-height);\n      font-weight: 700;\n      margin: 46px 0 -13px -1.883px;\n      color: #777;\n    }\n\n    h5 {\n      font-family: var(\n        --hax-base-styles-h5-font-family,\n        var(--hax-base-styles-heading-font-family, \"Lato\", sans-serif)\n      );\n      font-size: var(\n        --hax-base-styles-h5-font-size,\n        calc(1.11111111 * var(--haxcms-base-styles-body-font-size, 18px))\n      );\n      line-height: var(--hax-base-styles-h5-line-height);\n      font-weight: 700;\n      margin: 36px 0 -13px -1.883px;\n      color: #333;\n    }\n\n    h6 {\n      font-family: var(\n        --hax-base-styles-h5-font-family,\n        var(--hax-base-styles-heading-font-family, \"Lato\", sans-serif)\n      );\n      font-size: var(\n        --hax-base-styles-h5-font-size,\n        var(--haxcms-base-styles-body-font-size, 18px)\n      );\n      line-height: var(--hax-base-styles-h5-line-height);\n      font-weight: 700;\n      margin: 26px 0 -13px -1.883px;\n      color: #777;\n    }\n\n    p {\n      min-height: var(--hax-base-styles-p-min-height);\n      font-size: var(\n        --hax-base-styles-p-font-size,\n        var(--haxcms-base-styles-body-font-size, 18px)\n      );\n      font-family: var(--haxcms-base-styles-body-font-family, \"Lora\");\n      line-height: var(\n        --hax-base-styles-p-line-height,\n        var(--haxcms-base-styles-body-line-height, 1.58)\n      );\n      letter-spacing: var(\n        --hax-base-styles-p-letter-spacing,\n        var(--haxcms-base-styles-body-letter-spacing, -0.03px)\n      );\n    }\n\n    p,\n    i,\n    a {\n      margin-top: 21px;\n    }\n\n    a {\n      text-decoration: underline;\n      font-size: var(\n        --hax-base-styles-a-font-size,\n        var(--hax-base-styles-p-font-size)\n      );\n    }\n\n    blockquote {\n      font-family: var(\n        --hax-base-styles-accent-font-family,\n        \"Playfair Display\",\n        serif\n      );\n      font-size: var(\n        --hax-base-styles-blockquote-font-size,\n        calc(1.66666667 * var(--haxcms-base-styles-body-font-size, 18px))\n      );\n      font-style: italic;\n      letter-spacing: -0.36px;\n      line-height: 44.4px;\n      overflow-wrap: break-word;\n      margin: 55px 0 33px 0;\n      color: rgba(0, 0, 0, 0.68);\n      padding: 0 0 0 50px;\n    }\n\n    figcaption {\n      font-family: var(\n        --hax-base-styles-heading-font-family,\n        \"Lato\",\n        sans-serif\n      );\n      font-size: var(\n        --hax-base-styles-small-font-size,\n        calc(0.88888889 * var(--hax-base-styles-p-font-size))\n      );\n      font-weight: 400;\n      color: #666;\n    }\n\n    ol ol {\n      list-style-type: lower-alpha;\n      font-size: 16px;\n      color: #444;\n      line-height: var(\n        --hax-base-styles-list-line-height,\n        var(--hax-base-styles-p-line-height)\n      );\n      font-size: var(\n        --hax-base-styles-smal-font-size,\n        calc(0.88888889 * var(--hax-base-styles-p-font-size))\n      );\n    }\n\n    ol ol ol {\n      list-style-type: lower-roman;\n      font-size: var(\n        --hax-base-styles-smaller-font-size,\n        calc(0.77777778 * var(--hax-base-styles-p-font-size))\n      );\n    }\n\n    code {\n      font-size: 18px;\n      background: rgba(0, 0, 0, 0.05);\n      border-radius: 2px;\n      padding: 3px 5px;\n      font-family: var(\n        --hax-base-styles-code-font-family,\n        \"Source Code Pro\",\n        monospace\n      );\n    }\n\n    pre {\n      font-family: var(\n        --hax-base-styles-code-font-family,\n        \"Source Code Pro\",\n        monospace\n      );\n    }\n\n    kbd {\n      font-family: var(\n        --hax-base-styles-code-font-family,\n        \"Source Code Pro\",\n        monospace\n      );\n      font-weight: 700;\n    }\n\n    samp {\n      font-family: var(\n        --hax-base-styles-code-font-family,\n        \"Source Code Pro\",\n        monospace\n      );\n    }\n  "])))];
  _exports.LrnCssResetStyles = LrnCssResetStyles;

  var AttachStylesToHead = function AttachStylesToHead() {
    var styles = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : LrnCssResetStyles;
    var id = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "LrnCssResetStyles";

    if (!window.LrnCssResetStyles) {
      window[id] = document.createElement("div");

      var s = document.createElement("style"),
          _css = LrnCssResetStyles.map(function (st) {
        return st.cssText;
      }).join("");

      s.setAttribute("id", id);
      s.setAttribute("type", "text/css");

      if (s.styleSheet) {
        // IE
        s.styleSheet.cssText = _css;
      } else {
        // the world
        s.appendChild(document.createTextNode(_css));
      }

      document.getElementsByTagName("head")[0].appendChild(s);
    }
  };
  /**
   * `lrn-css-reset`
   * `an element that applies CSS reset and additional base styles
   * @microcopy - language worth noting:
   *  -
   *
   * @class LrnCssReset
   * @extends {LitElement}
   * @demo demo/index.html
   * @element lrn-css-reset
   */


  _exports.AttachStylesToHead = AttachStylesToHead;

  var LrnCssReset = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(LrnCssReset, _LitElement);

    var _super = _createSuper(LrnCssReset);

    /**
     * HTMLElement
     */
    function LrnCssReset() {
      babelHelpers.classCallCheck(this, LrnCssReset);
      return _super.call(this);
    }

    babelHelpers.createClass(LrnCssReset, [{
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        if (babelHelpers.get(babelHelpers.getPrototypeOf(LrnCssReset.prototype), "firstUpdated", this)) babelHelpers.get(babelHelpers.getPrototypeOf(LrnCssReset.prototype), "firstUpdated", this).call(this, changedProperties);
        if (this.applyToHead) AttachStylesToHead(LrnCssResetStyles, "LrnCssResetStyles");
      }
    }, {
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject2_1cdd2360942811ecb1937969798b231a || (_templateObject2_1cdd2360942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<slot></slot>"])));
      }
    }], [{
      key: "tag",
      get:
      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */
      function get() {
        return "lrn-css-reset";
      }
    }, {
      key: "styles",
      get: function get() {
        return LrnCssResetStyles;
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          applyToHead: {
            type: Boolean,
            attribute: "apply-to-head"
          }
        };
      }
    }]);
    return LrnCssReset;
  }(_index.LitElement);

  _exports.LrnCssReset = LrnCssReset;
  customElements.define(LrnCssReset.tag, LrnCssReset);
});
