// @lrnwebcomponents/simple-toolbar/lib/dist/simple-toolbar-more-button.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SimpleToolbarMoreButton = void 0;

var _litElement = require("lit");

var _simpleToolbarButton = require("./simple-toolbar-button.js");

function _typeof(obj) {
  if (typeof Symbol === "function" && babelHelpers.typeof(Symbol.iterator) === "symbol") {
    _typeof = function _typeof(obj) {
      return babelHelpers.typeof(obj);
    };
  } else {
    _typeof = function _typeof(obj) {
      return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : babelHelpers.typeof(obj);
    };
  }

  return _typeof(obj);
}

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);

  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    if (enumerableOnly) symbols = symbols.filter(function (sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    });
    keys.push.apply(keys, symbols);
  }

  return keys;
}

function _objectSpread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {};

    if (i % 2) {
      ownKeys(source, true).forEach(function (key) {
        _defineProperty(target, key, source[key]);
      });
    } else if (Object.getOwnPropertyDescriptors) {
      Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
      ownKeys(source).forEach(function (key) {
        Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
      });
    }
  }

  return target;
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function _templateObject_7fceba60942811ecb1937969798b231a() {
  var data = _taggedTemplateLiteral(["\n        :host([hidden]) {\n          display: none !important;\n        }\n      "]);

  _templateObject_7fceba60942811ecb1937969798b231a = function _templateObject_7fceba60942811ecb1937969798b231a() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteral(strings, raw) {
  if (!raw) {
    raw = strings.slice(0);
  }

  return Object.freeze(Object.defineProperties(strings, {
    raw: {
      value: Object.freeze(raw)
    }
  }));
}

function _toConsumableArray(arr) {
  return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread();
}

function _nonIterableSpread() {
  throw new TypeError("Invalid attempt to spread non-iterable instance");
}

function _iterableToArray(iter) {
  if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter);
}

function _arrayWithoutHoles(arr) {
  if (Array.isArray(arr)) {
    for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) {
      arr2[i] = arr[i];
    }

    return arr2;
  }
}

function _classCallCheck(instance, Constructor) {
  if (!babelHelpers.instanceof(instance, Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _possibleConstructorReturn(self, call) {
  if (call && (_typeof(call) === "object" || typeof call === "function")) {
    return call;
  }

  return _assertThisInitialized(self);
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _inherits(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function");
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      writable: true,
      configurable: true
    }
  });
  if (superClass) _setPrototypeOf(subClass, superClass);
}

function _setPrototypeOf(o, p) {
  _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
    o.__proto__ = p;
    return o;
  };

  return _setPrototypeOf(o, p);
}

function _get(target, property, receiver) {
  if (typeof Reflect !== "undefined" && Reflect.get) {
    _get = Reflect.get;
  } else {
    _get = function _get(target, property, receiver) {
      var base = _superPropBase(target, property);

      if (!base) return;
      var desc = Object.getOwnPropertyDescriptor(base, property);

      if (desc.get) {
        return desc.get.call(receiver);
      }

      return desc.value;
    };
  }

  return _get(target, property, receiver || target);
}

function _superPropBase(object, property) {
  while (!Object.prototype.hasOwnProperty.call(object, property)) {
    object = _getPrototypeOf(object);
    if (object === null) break;
  }

  return object;
}

function _getPrototypeOf(o) {
  _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
    return o.__proto__ || Object.getPrototypeOf(o);
  };
  return _getPrototypeOf(o);
}
/**
 * `simple-toolbar-more-button`
 * a more button to toggle collapsed buttons in the rich text editor
 *
 * @customElement
 * @extends SimpleToolbarButtonBehaviors
 * @extends LitElement
 * @lit-html
 * @lit-element
 * @demo ./demo/buttons.html
 */


var SimpleToolbarMoreButton = /*#__PURE__*/function (_SimpleToolbarButtonB) {
  _inherits(SimpleToolbarMoreButton, _SimpleToolbarButtonB);

  _createClass(SimpleToolbarMoreButton, null, [{
    key: "tag",

    /**
     * Store the tag name to make it easier to obtain directly.
     */
    get: function get() {
      return "simple-toolbar-more-button";
    }
  }, {
    key: "styles",
    get: function get() {
      return [].concat(_toConsumableArray(_get(_getPrototypeOf(SimpleToolbarMoreButton), "styles", this)), [(0, _litElement.css)(_templateObject_7fceba60942811ecb1937969798b231a())]);
    }
  }]);

  function SimpleToolbarMoreButton() {
    var _this;

    _classCallCheck(this, SimpleToolbarMoreButton);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(SimpleToolbarMoreButton).call(this));
    _this.icon = "more-vert";
    _this.toggled = false;
    _this.toggles = true;
    _this.label = "More buttons";
    _this.labelToggled = "Fewer buttons";
    _this.tooltipDirection = "left";
    _this.disabled = false;
    return _this;
  }
  /**
   * toggles button if shortcutKey is pressed
   *
   * @param {string} key
   * @event toggle
   */


  _createClass(SimpleToolbarMoreButton, [{
    key: "_handleShortcutKeys",
    value: function _handleShortcutKeys(e, key) {
      this.dispatchEvent(new CustomEvent("toggle", {
        bubbles: true,
        cancelable: true,
        composed: true,
        detail: _objectSpread({}, e.detail, {
          button: this,
          shortcutKey: this
        })
      }));
    }
  }]);

  return SimpleToolbarMoreButton;
}((0, _simpleToolbarButton.SimpleToolbarButtonBehaviors)(_litElement.LitElement));

exports.SimpleToolbarMoreButton = SimpleToolbarMoreButton;
window.customElements.define(SimpleToolbarMoreButton.tag, SimpleToolbarMoreButton);
