// @lrnwebcomponents/lrndesign-stepper/lrndesign-stepper.js
define(["exports", "require", "../../@polymer/polymer/polymer-element.js"], function (_exports, _require, _polymerElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrndesignStepper = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  /**
   * Copyright 2018 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
  `lrndesign-stepper`
  visualization of steps
  
  * @demo demo/index.html
  */
  class LrndesignStepper extends _polymerElement.PolymerElement {
    constructor() {
      super();
      new Promise((res, rej) => _require.default(["./lib/lrndesign-stepper-button.js"], res, rej));
    }

    static get template() {
      return (0, _polymerElement.html)`
      <style>
        :host {
          display: block;
        }
      </style>
      <div class="buttons"><slot></slot></div>
    `;
    }

    static get tag() {
      return "lrndesign-stepper";
    }

    ready() {
      super.ready();
      var children = this.shadowRoot.querySelector("slot").assignedNodes({
        flatten: true
      }).filter(n => n.nodeType === Node.ELEMENT_NODE);

      if (children.length > 1) {
        children.forEach(function (child, index) {
          if (index === 0) {
            child.setAttribute("location", "start");
          } else if (index === children.length - 1) {
            child.setAttribute("location", "end");
          } else {
            child.setAttribute("location", "middle");
          }

          console.log(child, index);
        });
      }
    }

  }

  _exports.LrndesignStepper = LrndesignStepper;
  window.customElements.define(LrndesignStepper.tag, LrndesignStepper);
});
