// @lrnwebcomponents/haxcms-elements/lib/core/backends/haxcms-backend-nodejs.js
define(["exports", "require", "../../../../../@polymer/polymer/polymer-element.js", "../../../../jwt-login/jwt-login.js", "../haxcms-site-store.js"], function (_exports, _require, _polymerElement, _jwtLogin, _haxcmsSiteStore) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HAXCMSBackendNodeJS = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  /**
   * Copyright 2018 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `haxcms-backend-nodejs`
   * `a simple element to check for and fetch JWTs`
   *
   * @demo demo/index.html
   *
   * @microcopy - the mental model for this element
   * - jwt - a json web token which is an encrypted security token to talk
   */
  class HAXCMSBackendNodeJS extends _polymerElement.PolymerElement {
    /**
     * Store the tag name to make it easier to obtain directly.
     * @notice function name must be here for tooling to operate correctly
     */
    static get tag() {
      return "haxcms-backend-nodejs";
    } // render function


    static get template() {
      return _polymerElement.html`
      <jwt-login
        id="jwt"
        url="[[jwtLoginLocation]]"
        url-logout="[[jwtLogoutLocation]]"
        jwt="{{jwt}}"
      ></jwt-login>
    `;
    }

    static get properties() {
      return {
        /**
         * Location of what endpoint to hit for
         */
        jwtLoginLocation: {
          type: String,
          value: function () {
            if (window.appSettings) {
              return window.appSettings.login;
            }
          }
        },

        /**
         * Location of what endpoint to hit for logging out
         */
        jwtLogoutLocation: {
          type: String,
          value: function () {
            if (window.appSettings) {
              return window.appSettings.logout;
            }
          }
        },

        /**
         * JSON Web token, it'll come from a global call if it's available
         */
        jwt: {
          type: String
        }
      };
    }
    /**
     * created life cycle
     */


    constructor() {
      super();
      document.body.addEventListener("jwt-token", this._jwtTokenFired.bind(this));
    }
    /**
     * detached life cycle
     */


    disconnectedCallback() {
      document.body.removeEventListener("jwt-token", this._jwtTokenFired.bind(this));
      super.disconnectedCallback();
    }
    /**
     * JWT token fired, let's capture it
     */


    _jwtTokenFired(e) {
      this.jwt = e.detail;
      _haxcmsSiteStore.store.jwt = this.jwt;
    }
    /**
     * Attached life cycle
     */


    connectedCallback() {
      super.connectedCallback();

      if (this.jwt != null && this.jwt != "" && typeof this.jwt == "string") {
        // attempt to dynamically import the hax cms site editor
        // which will appear to be injecting into the page
        // but because of this approach it should be non-blocking
        try {
          new Promise((res, rej) => _require.default(["../haxcms-site-editor.js"], res, rej)).then(e => {
            // if we don't have appSettings by this point
            // it means we don't actually have a backend / directions
            // this would be a published state or a state where
            // there is no actual backend to bother confiring with
            // possibly a user navigated to a site that doesn't
            // have JWT credentials but isn't actually published
            if (window.appSettings) {
              _haxcmsSiteStore.store.cmsSiteEditorAvailability();

              _haxcmsSiteStore.store.cmsSiteEditor.instance.jwt = this.jwt;
              _haxcmsSiteStore.store.cmsSiteEditor.instance.saveNodePath = window.appSettings.saveNodePath;
              _haxcmsSiteStore.store.cmsSiteEditor.instance.saveManifestPath = window.appSettings.saveManifestPath;
              _haxcmsSiteStore.store.cmsSiteEditor.instance.saveOutlinePath = window.appSettings.saveOutlinePath;
              _haxcmsSiteStore.store.cmsSiteEditor.instance.getNodeFieldsPath = window.appSettings.getNodeFieldsPath;
              _haxcmsSiteStore.store.cmsSiteEditor.instance.getSiteFieldsPath = window.appSettings.getSiteFieldsPath;
              _haxcmsSiteStore.store.cmsSiteEditor.instance.getFormToken = window.appSettings.getFormToken;
              _haxcmsSiteStore.store.cmsSiteEditor.instance.publishSitePath = window.appSettings.publishSitePath;
              _haxcmsSiteStore.store.cmsSiteEditor.instance.syncSitePath = window.appSettings.syncSitePath;
              _haxcmsSiteStore.store.cmsSiteEditor.instance.revertSitePath = window.appSettings.revertSitePath;
              _haxcmsSiteStore.store.cmsSiteEditor.instance.createNodePath = window.appSettings.createNodePath;
              _haxcmsSiteStore.store.cmsSiteEditor.instance.deleteNodePath = window.appSettings.deleteNodePath;
              _haxcmsSiteStore.store.cmsSiteEditor.instance.appStore = window.appSettings.appStore;
            }
          }, e => {//import failed
          });
        } catch (err) {// error in the event this is a double registration
        }
      } else {
        // other things will have to sort out the fact that while we
        // DO have a dynamic backend, we didn't get a hit on the JWT
        // meaning that we are in a dynamic environment but logged out
        // at the moment (or viewing a site we don't have authorization to)
        window.dispatchEvent(new CustomEvent("haxcms-not-logged-in", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: this
        }));
      }
    }

  }

  _exports.HAXCMSBackendNodeJS = HAXCMSBackendNodeJS;
  window.customElements.define(HAXCMSBackendNodeJS.tag, HAXCMSBackendNodeJS);
});
