// @lrnwebcomponents/simple-blog/simple-blog.js
define(["exports", "require", "../../lit-element/lit-element.js", "../haxcms-elements/lib/core/HAXCMSLitElementTheme.js", "../haxcms-elements/lib/core/haxcms-site-store.js", "../../mobx/lib/mobx.module.js", "../simple-colors/simple-colors.js", "./lib/simple-blog-post.js", "../../@polymer/iron-pages/iron-pages.js"], function (_exports, _require, _litElement, _HAXCMSLitElementTheme, _haxcmsSiteStore, _mobxModule, _simpleColors, _simpleBlogPost, _ironPages) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SimpleBlog = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject2_db7221401bb211eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        html,\n        body {\n          background-color: #fafafa;\n        }\n        :host {\n          display: block;\n          font-family: \"Roboto\", \"Noto\", sans-serif;\n          -webkit-font-smoothing: antialiased;\n          font-size: 14px;\n          background-color: #fafafa;\n          font-family: \"Open Sans\", \"MundoSans\", helvetica neue, Arial,\n            Helvetica, sans-serif;\n          margin: 0;\n          padding: 0;\n          text-rendering: optimizeLegibility;\n          -webkit-font-smoothing: antialiased;\n          -moz-font-feature-settings: \"liga=1, dlig=1\";\n          -ms-font-feature-settings: \"liga\", \"dlig\";\n          -webkit-font-feature-settings: \"liga\", \"dlig\";\n          -o-font-feature-settings: \"liga\", \"dlig\";\n          font-feature-settings: \"liga\", \"dlig\";\n        }\n        #backbutton {\n          position: fixed;\n          top: 0px;\n          left: 0px;\n          padding: 2px;\n          width: 40px;\n          height: 40px;\n          margin: 8px;\n          z-index: 1000;\n          color: black;\n          background-color: rgba(250, 250, 250, 0.5);\n          opacity: 0.5;\n          border-radius: 50%;\n          transition: all 0.6s linear;\n        }\n        #backbutton:focus,\n        #backbutton:hover {\n          opacity: 1;\n          color: white;\n          background-color: var(--haxcms-color, black);\n        }\n        iron-pages,\n        iron-pages section {\n          width: 100vw;\n          height: 100vh;\n        }\n        simple-blog-post {\n          transition: all 0.6s ease-in-out;\n          visibility: hidden;\n        }\n        :host([selected-page=\"0\"]) simple-blog-post {\n          opacity: 0;\n          visibility: hidden;\n        }\n        :host([selected-page=\"1\"]) simple-blog-post {\n          visibility: visible;\n          opacity: 1;\n        }\n        a,\n        a:active,\n        a:hover,\n        a:focus {\n          color: inherit;\n        }\n      "]);

    _templateObject2_db7221401bb211eabd7cc30119585ddb = function _templateObject2_db7221401bb211eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject_db7221401bb211eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <iron-pages .selected=\"", "\">\n        <section>\n          <simple-blog-header></simple-blog-header>\n          <simple-blog-listing></simple-blog-listing>\n        </section>\n        <section>\n          <paper-icon-button\n            id=\"backbutton\"\n            icon=\"icons:arrow-back\"\n            @click=\"", "\"\n          ></paper-icon-button>\n          <paper-tooltip\n            for=\"backbutton\"\n            position=\"right\"\n            offset=\"14\"\n            animation-delay=\"0\"\n            >Back to listing\n          </paper-tooltip>\n          <simple-blog-post ?edit-mode=\"", "\"\n            ><slot></slot\n          ></simple-blog-post>\n          <simple-blog-footer id=\"footer\"></simple-blog-footer>\n        </section>\n      </iron-pages>\n    "]);

    _templateObject_db7221401bb211eabd7cc30119585ddb = function _templateObject_db7221401bb211eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  /**
   * `simple-blog`
   * @customElement simple-blog
   * `A simple blog and associated elements`
   * @demo demo/index.html
   */
  var SimpleBlog =
  /*#__PURE__*/
  function (_SimpleColorsSuper) {
    babelHelpers.inherits(SimpleBlog, _SimpleColorsSuper);
    babelHelpers.createClass(SimpleBlog, [{
      key: "render",
      // render function
      value: function render() {
        return (0, _litElement.html)(_templateObject_db7221401bb211eabd7cc30119585ddb(), this.selectedPage, this._goBack, this.editMode);
      }
      /**
       * Mix in an opened status
       */

    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(SimpleBlog), "styles", this)), [(0, _litElement.css)(_templateObject2_db7221401bb211eabd7cc30119585ddb())]);
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */

    }, {
      key: "tag",
      get: function get() {
        return "simple-blog";
      }
    }, {
      key: "properties",
      get: function get() {
        return babelHelpers.objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(SimpleBlog), "properties", this), {
          selectedPage: {
            type: Number,
            reflect: true,
            attribute: "selected-page"
          }
        });
      }
    }]);

    function SimpleBlog() {
      var _this;

      babelHelpers.classCallCheck(this, SimpleBlog);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(SimpleBlog).call(this));
      _this.selectedPage = 0;
      new Promise(function (res, rej) {
        return _require.default(["./lib/simple-blog-header.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/paper-icon-button/paper-icon-button.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["./lib/simple-blog-footer.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["./lib/simple-blog-listing.js"], res, rej);
      });
      return _this;
    }
    /**
     * LitElement ready
     */


    babelHelpers.createClass(SimpleBlog, [{
      key: "firstUpdated",
      value: function firstUpdated() {
        this.contentContainer = this.shadowRoot.querySelector("simple-blog-post").shadowRoot.querySelector("#contentcontainer");
      }
      /**
       * attached life cycle
       */

    }, {
      key: "connectedCallback",
      value: function connectedCallback() {
        var _this2 = this;

        babelHelpers.get(babelHelpers.getPrototypeOf(SimpleBlog.prototype), "connectedCallback", this).call(this);
        this.__disposer = [];
        (0, _mobxModule.autorun)(function (reaction) {
          _this2.activeId = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.activeId);

          _this2.__disposer.push(reaction);
        });
        (0, _mobxModule.autorun)(function (reaction) {
          _this2._locationChanged(_haxcmsSiteStore.store.location);

          _this2.__disposer.push(reaction);
        });
      }
      /**
       * detatched life cycle
       */

    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        // clean up state
        for (var i in this.__disposer) {
          this.__disposer[i].dispose();
        }

        babelHelpers.get(babelHelpers.getPrototypeOf(SimpleBlog.prototype), "disconnectedCallback", this).call(this);
      }
      /**
       * Listen for router location changes
       */

    }, {
      key: "_locationChanged",
      value: function _locationChanged(location) {
        if (!location || typeof location.route === "undefined") return;
        var name = location.route.name;

        if (name === "home" || name === "404") {
          this.selectedPage = 0;
        } else {
          window.scrollTo({
            top: 0,
            left: 0
          });
          this.selectedPage = 1;
        }
      }
      /**
       * Reset the active item to reset state
       */

    }, {
      key: "_goBack",
      value: function _goBack(e) {
        var _this3 = this;

        var prevActiveItemId = _haxcmsSiteStore.store.activeId;
        window.history.pushState(null, null, _haxcmsSiteStore.store.location.baseUrl);
        window.dispatchEvent(new PopStateEvent("popstate")); // should help account for starting on a page where popstate isn't set
        // and also generate data model mirroring

        if (prevActiveItemId) {
          setTimeout(function () {
            var active = _this3.shadowRoot.querySelector("simple-blog-listing").shadowRoot.querySelector('simple-blog-overview[item-id="' + prevActiveItemId + '"]');

            if (active) {
              active.scrollIntoView(true);
              active.focus();
            }
          }, 100);
        } else {
          window.scrollTo({
            top: 0,
            left: 0
          });
        }

        var evt = new CustomEvent("json-outline-schema-active-item-changed", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: {}
        });
        this.dispatchEvent(evt);
        this.selectedPage = 0;
      }
    }]);
    return SimpleBlog;
  }((0, _simpleColors.SimpleColorsSuper)(_HAXCMSLitElementTheme.HAXCMSLitElementTheme));

  _exports.SimpleBlog = SimpleBlog;
  window.customElements.define(SimpleBlog.tag, SimpleBlog);
});
