// lit-element/lit-element.js
define(["exports", "./node_modules/@lit/reactive-element/reactive-element.js", "./node_modules/lit-html/lit-html.js"], function (_exports, _reactiveElement, _litHtml) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  var _exportNames = {
    LitElement: true,
    UpdatingElement: true,
    _$LE: true
  };
  _exports._$LE = _exports.UpdatingElement = _exports.LitElement = void 0;
  Object.keys(_reactiveElement).forEach(function (key) {
    if (key === "default" || key === "__esModule") return;
    if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
    if (key in _exports && _exports[key] === _reactiveElement[key]) return;
    Object.defineProperty(_exports, key, {
      enumerable: true,
      get: function get() {
        return _reactiveElement[key];
      }
    });
  });
  Object.keys(_litHtml).forEach(function (key) {
    if (key === "default" || key === "__esModule") return;
    if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
    if (key in _exports && _exports[key] === _litHtml[key]) return;
    Object.defineProperty(_exports, key, {
      enumerable: true,
      get: function get() {
        return _litHtml[key];
      }
    });
  });

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var l, o, r;
  var s = _reactiveElement.ReactiveElement;
  _exports.UpdatingElement = s;

  var n = /*#__PURE__*/function (_t) {
    babelHelpers.inherits(n, _t);

    var _super = _createSuper(n);

    function n() {
      var _this;

      babelHelpers.classCallCheck(this, n);
      _this = _super.apply(this, arguments), _this.renderOptions = {
        host: babelHelpers.assertThisInitialized(_this)
      }, _this._$Dt = void 0;
      return _this;
    }

    babelHelpers.createClass(n, [{
      key: "createRenderRoot",
      value: function createRenderRoot() {
        var t, e;
        var i = babelHelpers.get(babelHelpers.getPrototypeOf(n.prototype), "createRenderRoot", this).call(this);
        return null !== (t = (e = this.renderOptions).renderBefore) && void 0 !== t || (e.renderBefore = i.firstChild), i;
      }
    }, {
      key: "update",
      value: function update(t) {
        var i = this.render();
        this.hasUpdated || (this.renderOptions.isConnected = this.isConnected), babelHelpers.get(babelHelpers.getPrototypeOf(n.prototype), "update", this).call(this, t), this._$Dt = (0, _litHtml.render)(i, this.renderRoot, this.renderOptions);
      }
    }, {
      key: "connectedCallback",
      value: function connectedCallback() {
        var t;
        babelHelpers.get(babelHelpers.getPrototypeOf(n.prototype), "connectedCallback", this).call(this), null === (t = this._$Dt) || void 0 === t || t.setConnected(!0);
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        var t;
        babelHelpers.get(babelHelpers.getPrototypeOf(n.prototype), "disconnectedCallback", this).call(this), null === (t = this._$Dt) || void 0 === t || t.setConnected(!1);
      }
    }, {
      key: "render",
      value: function render() {
        return _litHtml.noChange;
      }
    }]);
    return n;
  }(_reactiveElement.ReactiveElement);

  _exports.LitElement = n;
  n.finalized = !0, n._$litElement$ = !0, null === (l = globalThis.litElementHydrateSupport) || void 0 === l || l.call(globalThis, {
    LitElement: n
  }), null === (o = globalThis.litElementPolyfillSupport) || void 0 === o || o.call(globalThis, {
    LitElement: n
  });
  var h = {
    _$AK: function _$AK(t, e, i) {
      t._$AK(e, i);
    },
    _$AL: function _$AL(t) {
      return t._$AL;
    }
  };
  _exports._$LE = h;
  (null !== (r = globalThis.litElementVersions) && void 0 !== r ? r : globalThis.litElementVersions = []).push("3.0.0");
});
