// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-dashboard/lrnapp-block-recent-project.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@polymer/polymer/lib/elements/custom-style.js", "../../../../@polymer/iron-ajax/iron-ajax.js", "../../../../@polymer/iron-icon/iron-icon.js", "../../../../@polymer/iron-list/iron-list.js", "../../../../@polymer/paper-card/paper-card.js", "../../../../@polymer/paper-item/paper-item-shared-styles.js", "../../../../@polymer/paper-button/paper-button.js", "../../../lrnsys-button/lrnsys-button.js", "../../../elmsln-loading/elmsln-loading.js", "../../../materializecss-styles/materializecss-styles.js"], function (_exports, _litElement, _customStyle, _ironAjax, _ironIcon, _ironList, _paperCard, _paperItemSharedStyles, _paperButton, _lrnsysButton, _elmslnLoading, _materializecssStyles) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappBlockRecentProject = void 0;

  function _templateObject4_6cf7a2502d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        #loading {\n          width: 100%;\n          z-index: 1000;\n          opacity: 0.8;\n          text-align: center;\n          align-content: center;\n          justify-content: center;\n          background-color: white;\n        }\n        paper-button {\n          padding: 0;\n          margin: 0;\n          min-width: 1rem;\n          text-transform: unset;\n        }\n        .project-card {\n          width: 100%;\n          min-width: 15em;\n          max-width: 20em;\n          margin: 0 1em;\n        }\n        .button-contents {\n          display: flex;\n          align-content: center;\n        }\n        .assignment-row {\n          border: 1px solid #000000;\n          background-color: #ffffff;\n        }\n        .assignment-row .assignment-row-button.active {\n          background-color: var(--paper-amber-50);\n          font-weight: bold;\n        }\n        .assignment-row:hover .assignment-operations {\n          display: block;\n          overflow: visible;\n          margin: 0.2em;\n        }\n        .assignment-row-button {\n          width: 100%;\n          justify-content: flex-start;\n          height: 3em;\n          text-transform: none;\n        }\n        .assignment-title {\n          display: inline-flex;\n          align-items: center;\n          overflow: hidden;\n          text-overflow: ellipsis;\n          line-height: 1em;\n        }\n        .status-indicator {\n          border-right: 1px solid grey;\n          padding: 0.5em;\n          margin: 0 0.5em 0 0;\n          display: inline-flex;\n          line-height: 3em;\n          height: 3em;\n          justify-content: center;\n          align-items: center;\n        }\n        .whole-project {\n          width: 100%;\n          margin: 0;\n          padding: 0;\n        }\n      "]);

    _templateObject4_6cf7a2502d8511ea80ccf124101da415 = function _templateObject4_6cf7a2502d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject3_6cf7a2502d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <lrnsys-button\n              inner-class=\"no-left-padding\"\n              button-class=\"assignment-row-button\"\n              class=\"assignment-row-button\"\n              href=\"", "lrnapp-studio-kanban\"\n              hover-class=\"amber lighten-5\"\n            >\n              <span class=\"button-contents\">\n                <div class=\"status-indicator\">\n                  <iron-icon icon=\"assignment\"></iron-icon>\n                </div>\n                <div class=\"assignment-title\">\n                  Tap to start your first project!\n                </div>\n              </span>\n            </lrnsys-button>\n          "]);

    _templateObject3_6cf7a2502d8511ea80ccf124101da415 = function _templateObject3_6cf7a2502d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject2_6cf7a2502d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <a tabindex=\"-1\" href=\"", "lrnapp-studio-kanban\">\n              <paper-button class=\"whole-project ferpa-protect\">\n                <paper-card\n                  id=\"project-", "\"\n                  class=\"project-card grey lighten-3\"\n                  heading=\"", "\"\n                  elevation=\"2\"\n                >\n                  <div class=\"card-content\">\n                    <iron-list\n                      .items=\"", "\"\n                      as=\"assignment\"\n                      mutable-data\n                    >\n                      <template>\n                        <div class=\"assignment-row\" id=\"assignment\">\n                          <lrnsys-button\n                            inner-class=\"no-left-padding\"\n                            class=\"assignment-row-button\"\n                            button-class=\"assignment-row-button\"\n                            id=\"assignment-", "-", "\"\n                            hover-class=\"amber lighten-5\"\n                            href=\"", "lrnapp-studio-kanban\"\n                          >\n                            <span class=\"button-contents\">\n                              <div\n                                class=\"status-indicator ", "\"\n                              >\n                                <iron-icon\n                                  icon=\"", "\"\n                                ></iron-icon>\n                              </div>\n                              <div class=\"assignment-title\">\n                                ", "\n                              </div>\n                            </span>\n                          </lrnsys-button>\n                        </div>\n                      </template>\n                    </iron-list>\n                  </div>\n                </paper-card>\n              </paper-button>\n            </a>\n          "]);

    _templateObject2_6cf7a2502d8511ea80ccf124101da415 = function _templateObject2_6cf7a2502d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject_6cf7a2502d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <custom-style>\n        <style include=\"materializecss-styles paper-item-styles\">\n          :host {\n            display: block;\n          }\n        </style>\n      </custom-style>\n      <iron-ajax\n        auto=\"\"\n        url=\"", "\"\n        handle-as=\"json\"\n        @last-response-changed=\"", "\"\n        @response=\"", "\"\n      ></iron-ajax>\n      <div id=\"loading\">\n        <h3>Loading..</h3>\n        <elmsln-loading color=\"grey-text\" size=\"large\"></elmsln-loading>\n      </div>\n      ", "\n    "]);

    _templateObject_6cf7a2502d8511ea80ccf124101da415 = function _templateObject_6cf7a2502d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  var LrnappBlockRecentProject =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnappBlockRecentProject, _LitElement);

    function LrnappBlockRecentProject() {
      babelHelpers.classCallCheck(this, LrnappBlockRecentProject);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappBlockRecentProject).apply(this, arguments));
    }

    babelHelpers.createClass(LrnappBlockRecentProject, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_6cf7a2502d8511ea80ccf124101da415(), this.sourcePath, this.responseChanged, this.handleResponse, this.hasProject ? (0, _litElement.html)(_templateObject2_6cf7a2502d8511ea80ccf124101da415(), this.basePath, this.project.id, this.project.attributes.title, this._toArray(this.project.relationships.assignments), this.project.id, this.assignment.id, this.basePath, this.assignment.metadata.relatedSubmissions.complete.color, this.assignment.metadata.relatedSubmissions.complete.icon, this.assignment.title) : (0, _litElement.html)(_templateObject3_6cf7a2502d8511ea80ccf124101da415(), this.basePath));
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this = this;

        changedProperties.forEach(function (oldValue, propName) {
          // computed
          if (["project"].includes(propName)) {
            _this.hasProject = _this._getHasProject(_this[propName]);
          }

          var notifiedProps = ["sourcePath", "response", "hasProject", "project"];

          if (notifiedProps.includes(propName)) {
            // notify
            var eventName = "".concat(propName.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, "$1-$2").toLowerCase(), "-changed");

            _this.dispatchEvent(new CustomEvent(eventName, {
              detail: {
                value: _this[propName]
              }
            }));
          }
        });
      }
      /**
       * Generate the correct boolean for having a project.
       */

    }, {
      key: "_getHasProject",
      value: function _getHasProject(project) {
        if (babelHelpers.typeof(project.id) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          return true;
        }

        return false;
      }
    }, {
      key: "responseChanged",
      value: function responseChanged(e) {
        this.response = e.detail.value;
      }
      /**
       * Swap off the loading with project data.
       */

    }, {
      key: "handleResponse",
      value: function handleResponse(e) {
        this.shadowRoot.querySelector("#loading").hidden = true;
        this.project = this.response.data;
      }
      /**
       * Generate an array.
       */

    }, {
      key: "_toArray",
      value: function _toArray(obj) {
        if (obj == null) {
          return [];
        }

        return Object.keys(obj).map(function (key) {
          return obj[key];
        });
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject4_6cf7a2502d8511ea80ccf124101da415())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrnapp-block-recent-project";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          elmslnCourse: {
            type: String,
            attribute: "elmsln-course"
          },
          elmslnSection: {
            type: String,
            attribute: "elmsln-section"
          },
          basePath: {
            type: String,
            attribute: "base-path"
          },
          csrfToken: {
            type: String,
            attribute: "csrf-token"
          },
          endPoint: {
            type: String,
            attribute: "end-point"
          },
          sourcePath: {
            type: String,
            attribute: "source-path"
          },
          response: {
            type: Object
          },
          project: {
            type: Object
          },
          hasProject: {
            type: Boolean,
            reflect: true
          }
        };
      }
    }]);
    return LrnappBlockRecentProject;
  }(_litElement.LitElement);

  _exports.LrnappBlockRecentProject = LrnappBlockRecentProject;
  window.customElements.define(LrnappBlockRecentProject.tag, LrnappBlockRecentProject);
});
