// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-dashboard/lrnapp-block-recent-comments.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@polymer/iron-ajax/iron-ajax.js", "../../../../@polymer/paper-item/paper-item.js", "../../../elmsln-loading/elmsln-loading.js", "./lrnapp-block-recent-comments-comment.js", "../../../../@polymer/polymer/lib/elements/custom-style.js"], function (_exports, _litElement, _ironAjax, _paperItem, _elmslnLoading, _lrnappBlockRecentCommentsComment, _customStyle) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappBlockRecentComments = void 0;

  function _templateObject3_6cf04f502d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n      "]);

    _templateObject3_6cf04f502d8511ea80ccf124101da415 = function _templateObject3_6cf04f502d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject2_6cf04f502d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n          <lrnapp-block-recent-comments-comment\n            .comment-user=\"", "\"\n            comment-title=\"", "\"\n            action-view=\"", "\"\n            date-updated=\"", "\"\n            class=\"ferpa-protect\"\n          >\n            ", "\n          </lrnapp-block-recent-comments-comment>\n        "]);

    _templateObject2_6cf04f502d8511ea80ccf124101da415 = function _templateObject2_6cf04f502d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject_6cf04f502d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <custom-style>\n        <style include=\"paper-item-styles\">\n          :host {\n            display: block;\n          }\n        </style>\n      </custom-style>\n      <div id=\"loading\">\n        <h3>Loading..</h3>\n        <elmsln-loading color=\"grey-text\" size=\"large\"></elmsln-loading>\n      </div>\n      <iron-ajax\n        auto=\"\"\n        url=\"", "\"\n        handle-as=\"json\"\n        @last-response-changed=\"", "\"\n        @response=\"", "\"\n      ></iron-ajax>\n      ", "\n    "]);

    _templateObject_6cf04f502d8511ea80ccf124101da415 = function _templateObject_6cf04f502d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  var LrnappBlockRecentComments =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnappBlockRecentComments, _LitElement);

    function LrnappBlockRecentComments() {
      babelHelpers.classCallCheck(this, LrnappBlockRecentComments);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappBlockRecentComments).apply(this, arguments));
    }

    babelHelpers.createClass(LrnappBlockRecentComments, [{
      key: "render",
      value: function render() {
        var _this = this;

        return (0, _litElement.html)(_templateObject_6cf04f502d8511ea80ccf124101da415(), this.sourcePath, this.responseChangedEvent, this.handleResponse, this._toArray(this.response.data).map(function (comment) {
          return (0, _litElement.html)(_templateObject2_6cf04f502d8511ea80ccf124101da415(), comment.relationships.author.data, comment.attributes.subject, _this._getViewLink(comment.relationships.node.data.id), comment.attributes.changed, comment.attributes.body);
        }));
      }
    }, {
      key: "responseChangedEvent",
      value: function responseChangedEvent(e) {
        this.response = e.detail.value;
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "response") {
            // notify
            _this2.dispatchEvent(new CustomEvent("response-changed", {
              detail: {
                value: _this2[propName]
              }
            }));
          }

          if (propName == "sourcePath") {
            // notify
            _this2.dispatchEvent(new CustomEvent("source-path-changed", {
              detail: {
                value: _this2[propName]
              }
            }));
          }
        });
      }
    }, {
      key: "handleResponse",
      value: function handleResponse(e) {
        this.shadowRoot.querySelector("#loading").hidden = true;
      }
    }, {
      key: "_getViewLink",
      value: function _getViewLink(nid) {
        return this.basePath + "lrnapp-studio-submission/submissions/" + nid;
      }
    }, {
      key: "_toArray",
      value: function _toArray(obj) {
        if (obj == null) {
          return [];
        }

        return Object.keys(obj).map(function (key) {
          return obj[key];
        });
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject3_6cf04f502d8511ea80ccf124101da415())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrnapp-block-recent-comments";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          elmslnCourse: {
            type: String,
            attribute: "elmsln-course"
          },
          elmslnSection: {
            type: String,
            attribute: "elmsln-section"
          },
          basePath: {
            type: String,
            attribute: "base-path"
          },
          csrfToken: {
            type: String,
            attribute: "csrf-token"
          },
          endPoint: {
            type: String,
            attribute: "end-point"
          },
          sourcePath: {
            type: String,
            attribute: "source-path"
          },
          response: {
            type: Array
          }
        };
      }
    }]);
    return LrnappBlockRecentComments;
  }(_litElement.LitElement);

  _exports.LrnappBlockRecentComments = LrnappBlockRecentComments;
  window.customElements.define(LrnappBlockRecentComments.tag, LrnappBlockRecentComments);
});
