// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-dashboard/lrnapp-studio-dashboard.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@polymer/polymer/lib/elements/custom-style.js", "../../../../@polymer/paper-card/paper-card.js", "../../../materializecss-styles/materializecss-styles.js", "./lrnapp-block-recent-project.js", "./lrnapp-block-recent-submissions.js", "./lrnapp-block-recent-comments.js", "./lrnapp-block-need-feedback.js"], function (_exports, _litElement, _customStyle, _paperCard, _materializecssStyles, _lrnappBlockRecentProject, _lrnappBlockRecentSubmissions, _lrnappBlockRecentComments, _lrnappBlockNeedFeedback) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioDashboard = void 0;

  function _templateObject2_6cfafdb02d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        h1.title {\n          font-size: 2em;\n          color: black;\n          margin: 0;\n          padding: 0.5em 0 0 0;\n          text-transform: none;\n          text-align: left;\n        }\n        p.para {\n          margin: 0;\n          padding: 0.25em 0.5em;\n        }\n        .dashboard-row {\n          width: 100%;\n          display: inline-flex;\n        }\n        .dashboard-item {\n          width: 30%;\n        }\n      "]);

    _templateObject2_6cfafdb02d8511ea80ccf124101da415 = function _templateObject2_6cfafdb02d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject_6cfafdb02d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <custom-style>\n        <style include=\"materializecss-styles\">\n          :host {\n            display: block;\n            padding: 0 2em;\n          }\n        </style>\n      </custom-style>\n      <h1 class=\"title\">Welcome back ", "!</h1>\n      <p class=\"para\">Here's what's been going on in the studio</p>\n      <div class=\"dashboard-row\">\n        <paper-card heading=\"Your active project\" class=\"dashboard-item\">\n          <div class=\"card-content\">\n            <lrnapp-block-recent-project\n              csrf-token=\"", "\"\n              end-point=\"", "\"\n              base-path=\"", "\"\n              source-path=\"", "\"\n            >\n            </lrnapp-block-recent-project>\n          </div>\n        </paper-card>\n        <paper-card\n          heading=\"Classmates needing feedback\"\n          class=\"dashboard-item\"\n        >\n          <div class=\"card-content\">\n            <lrnapp-block-need-feedback\n              csrf-token=\"", "\"\n              end-point=\"", "\"\n              base-path=\"", "\"\n              source-path=\"", "\"\n            >\n            </lrnapp-block-need-feedback>\n          </div>\n        </paper-card>\n        <paper-card heading=\"Recent Studio submissions\" class=\"dashboard-item\">\n          <div class=\"card-content\">\n            <lrnapp-block-recent-submissions\n              csrf-token=\"", "\"\n              end-point=\"", "\"\n              base-path=\"", "\"\n              source-path=\"", "\"\n            >\n            </lrnapp-block-recent-submissions>\n          </div>\n        </paper-card>\n        <paper-card heading=\"Recent Studio comments\" class=\"dashboard-item\">\n          <div class=\"card-content\">\n            <lrnapp-block-recent-comments\n              csrf-token=\"", "\"\n              end-point=\"", "\"\n              base-path=\"", "\"\n              source-path=\"", "\"\n            >\n            </lrnapp-block-recent-comments>\n          </div>\n        </paper-card>\n      </div>\n    "]);

    _templateObject_6cfafdb02d8511ea80ccf124101da415 = function _templateObject_6cfafdb02d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  var LrnappStudioDashboard =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnappStudioDashboard, _LitElement);

    function LrnappStudioDashboard() {
      babelHelpers.classCallCheck(this, LrnappStudioDashboard);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappStudioDashboard).apply(this, arguments));
    }

    babelHelpers.createClass(LrnappStudioDashboard, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_6cfafdb02d8511ea80ccf124101da415(), this.username, this.csrfToken, this._getEndPoint(this.basePath), this.basePath, this._getDataSource(this.csrfToken, this.basePath, "recent-project"), this.csrfToken, this._getEndPoint(this.basePath), this.basePath, this._getDataSource(this.csrfToken, this.basePath, "need-feedback"), this.csrfToken, this._getEndPoint(this.basePath), this.basePath, this._getDataSource(this.csrfToken, this.basePath, "recent-submissions"), this.csrfToken, this._getEndPoint(this.basePath), this.basePath, this._getDataSource(this.csrfToken, this.basePath, "recent-comments"));
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this = this;

        changedProperties.forEach(function (oldValue, propName) {
          var notifiedProps = ["sourcePath", "basePath", "csrfToken"];

          if (notifiedProps.includes(propName)) {
            // notify
            var eventName = "".concat(propName.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, "$1-$2").toLowerCase(), "-changed");

            _this.dispatchEvent(new CustomEvent(eventName, {
              detail: {
                value: _this[propName]
              }
            }));
          }
        });
      }
    }, {
      key: "_getEndPoint",
      value: function _getEndPoint(basePath) {
        return basePath + "lrnapp-studio-dashboard/blocks";
      }
    }, {
      key: "_getDataSource",
      value: function _getDataSource(csrfToken, basePath, dataPoint) {
        return basePath + "lrnapp-studio-dashboard/blocks/" + dataPoint + "?token=" + csrfToken;
      }
      /**
       * Simple way to convert from object to array.
       */

    }, {
      key: "_toArray",
      value: function _toArray(obj) {
        if (obj == null) {
          return [];
        }

        return Object.keys(obj).map(function (key) {
          return obj[key];
        });
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_6cfafdb02d8511ea80ccf124101da415())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrnapp-studio-dashboard";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          elmslnCourse: {
            type: String,
            attribute: "elmsln-course"
          },
          elmslnSection: {
            type: String,
            attribute: "elmsln-section"
          },
          basePath: {
            type: String,
            attribute: "base-path",
            reflect: true
          },
          csrfToken: {
            type: String,
            attribute: "csrf-token",
            reflect: true
          },
          endPoint: {
            type: String,
            attribute: "end-point"
          },
          username: {
            type: String,
            reflect: true
          },
          sourcePath: {
            type: String,
            reflect: true,
            attribute: "source-path"
          }
        };
      }
    }]);
    return LrnappStudioDashboard;
  }(_litElement.LitElement);

  _exports.LrnappStudioDashboard = LrnappStudioDashboard;
  window.customElements.define(LrnappStudioDashboard.tag, LrnappStudioDashboard);
});
