// @vaadin/vaadin-progress-bar/src/vaadin-progress-bar.js
define(["exports", "../../../@polymer/polymer/polymer-element.js", "./vaadin-progress-mixin.js", "../../vaadin-themable-mixin/vaadin-themable-mixin.js", "../../vaadin-element-mixin/vaadin-element-mixin.js", "../../../@polymer/polymer/lib/utils/html-tag.js"], function (_exports, _polymerElement, _vaadinProgressMixin, _vaadinThemableMixin, _vaadinElementMixin, _htmlTag) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ProgressBarElement = void 0;

  /**
  @license
  Copyright (c) 2017 Vaadin Ltd.
  This program is available under Apache License Version 2.0, available at https://vaadin.com/license/
  */

  /**
   * `<vaadin-progress-bar>` is a Web Component for progress bars.
   *
   * ```html
   * <vaadin-progress-bar min="0" max="1" value="0.5">
   * </vaadin-progress-bar>
   * ```
   *
   * ### Styling
   *
   * The following shadow DOM parts are available for styling:
   *
   * Part name | Description
   * ----------------|----------------
   * `bar` | Progress-bar's background
   * `value` | Progress-bar's foreground
   *
   * See [ThemableMixin – how to apply styles for shadow parts](https://github.com/vaadin/vaadin-themable-mixin/wiki)
   *
   * The following custom properties are available:
   *
   * Custom property | Description | Default
   * ----------------|-------------|-------------
   * `--vaadin-progress-value` | current progress value (between 0 and 1) | 0
   *
   * The following state attributes are available for styling:
   *
   * Attribute       | Description | Part name
   * ----------------|-------------|------------
   * `indeterminate` | Set to an indeterminate progress bar | :host
   *
   * @extends PolymerElement
   * @mixes ProgressMixin
   * @mixes ThemableMixin
   * @mixes ElementMixin
   * @demo demo/index.html
   */
  class ProgressBarElement extends (0, _vaadinElementMixin.ElementMixin)((0, _vaadinThemableMixin.ThemableMixin)((0, _vaadinProgressMixin.ProgressMixin)(_polymerElement.PolymerElement))) {
    static get template() {
      return (0, _htmlTag.html)`
    <style>
      :host {
        display: block;
        width: 100%; /* prevent collapsing inside non-stretching column flex */
        height: 8px;
      }

      :host([hidden]) {
        display: none !important;
      }

      [part="bar"] {
        height: 100%;
      }

      [part="value"] {
        height: 100%;
        transform-origin: 0 50%;
        transform: scaleX(var(--vaadin-progress-value));
      }

      /* RTL specific styles */

      :host([dir="rtl"]) [part="value"] {
        transform-origin: 100% 50%;
      }
    </style>

    <div part="bar">
      <div part="value"></div>
    </div>
`;
    }

    static get is() {
      return 'vaadin-progress-bar';
    }

    static get version() {
      return '1.3.0';
    }

  }

  _exports.ProgressBarElement = ProgressBarElement;
  customElements.define(ProgressBarElement.is, ProgressBarElement);
});
