// @lrnwebcomponents/haxor-slevin/haxor-slevin.js
define(["exports","require","../../lit-element/lit-element.js","../haxcms-elements/lib/core/HAXCMSLitElementTheme.js","../haxcms-elements/lib/core/haxcms-site-store.js","../../mobx/lib/mobx.module.js","../utils/utils.js"],function(_exports,_require,_litElement,_HAXCMSLitElementTheme,_haxcmsSiteStore,_mobxModule,_utils){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.HaxorSlevin=void 0,_require=babelHelpers.interopRequireWildcard(_require);class HaxorSlevin extends _HAXCMSLitElementTheme.HAXCMSLitElementTheme{static get styles(){return[_litElement.css`
        :host {
          display: block;
          background-color: #ffffff;
          color: rgba(0, 0, 0, 0.84);
          --hax-base-styles-a-color: var(--haxcms-color, #2196f3);
          --hax-base-styles-a-color-visited: var(--haxcms-color, #2196f3);
          --hax-base-styles-a-color-active: var(--haxcms-color, #2196f3);
        }

        :host([hidden]) {
          display: none;
        }
        :host([is-logged-in]) {
          margin-left: 48px;
        }
        :host([is-logged-in][edit-mode]) {
          padding-left: 12px;
        }
        /**
      * Hide the slotted content during edit mode. This must be here to work.
      */
        :host([edit-mode]) #slot {
          display: none;
        }
        :host([edit-mode]) .contentcontainer-wrapper simple-blog-card {
          opacity: 0.2;
          pointer-events: none;
        }
        #slot {
          min-height: 50vh;
        }

        .wrapper {
          padding-bottom: 80px;
        }
        #home {
          max-width: 1032px;
          padding-left: 20px;
          padding-right: 20px;
          margin: 0 auto;
        }
        .contentcontainer-wrapper {
          max-width: 740px;
          margin: 0 auto;
          box-sizing: border-box;
          padding-left: 20px;
          padding-right: 20px;
        }
        simple-blog-card {
          padding: 8px;
          min-height: 100px;
          min-width: 100px;
        }
        .simple-blog-card-wrapper {
          margin: 0 auto;
          width: 100%;
        }
        .evenly {
          display: flex;
          justify-content: space-evenly;
        }
        simple-blog-card[size="micro"] {
          padding: 4px;
        }
        iron-pages {
          padding-top: 64px;
        }
        dom-repeat {
          padding-bottom: 16px;
          min-height: 300px;
        }
        app-toolbar {
          padding: 0 20px;
          height: 54px;
          max-width: 1032px;
          margin: 0 auto;
        }
        .backbutton {
          height: 54px;
          border-radius: 0;
          min-width: unset;
          text-transform: unset;
        }
        app-header {
          z-index: 100;
          @apply --layout-fixed-top;
          color: #ffffff;
          box-shadow: 0 4px 12px 0 rgba(0, 0, 0, 0.15);
          background-color: var(--haxcms-color, rgba(255, 0, 116, 1));
          --app-header-background-rear-layer: {
            background-color: var(--haxcms-color, rgba(255, 0, 116, 1));
          }
        }
        paper-icon-button {
          --paper-icon-button-ink-color: white;
        }
        site-active-title {
          --site-active-title-heading: {
            font-size: 42px;
            font-family: Georgia, Cambria, "Times New Roman", Times, serif;
            font-weight: 400;
            font-style: normal;
            font-weight: 400;
            line-height: 1.25;
            letter-spacing: 0;
          }
        }

        .social-float {
          top: 160px;
          position: fixed;
          z-index: 99;
          margin-left: -10vw;
          opacity: 1;
          transition: 0.2s opacity linear;
        }
        .social-float.disable-items {
          pointer-events: none;
          opacity: 0.2 !important;
        }
        .social-float ul {
          padding: 0;
          margin: 0;
          list-style: none;
        }

        site-share-widget {
          --site-share-widget-bg: var(--haxcms-color, rgba(255, 0, 116, 1));
        }
        site-share-widget:hover,
        site-share-widget:focus,
        site-share-widget:active {
          --site-share-widget-bg: var(--haxcms-system-action-color, blue);
        }

        social-share-link {
          --social-share-button-bg: var(--haxcms-color, rgba(255, 0, 116, 1));
          --social-share-button-padding: 8px;
          --social-share-button-border-radius: 50%;
        }

        .annoy-user {
          background-color: rgba(255, 255, 255, 0.9);
          display: block;
          position: fixed;
          bottom: 0;
          left: 0;
          right: 0;
          box-shadow: 0 -3px 10px 0 rgba(0, 0, 0, 0.0785);
          padding: 10px 0;
          height: 50px;
          z-index: 100;
          opacity: 1;
          transition: 0.2s opacity linear;
        }
        .annoy-user.disable-items {
          pointer-events: none;
          opacity: 0 !important;
        }
        iron-icon {
          height: 40px;
          width: 40px;
          display: flex;
          padding-right: 20px;
        }
        .annoy-user iron-icon {
          color: black;
        }
        .annoy-user span {
          flex: 1 1 auto;
          height: 40px;
          display: flex;
          vertical-align: middle;
          line-height: 40px;
        }
        .annoy-inner strong {
          padding: 0 4px;
        }
        .annoy-user .rss {
          margin-left: 50px;
        }
        .annoy-inner {
          max-width: 700px;
          margin: 0 auto;
          display: flex;
        }
        .subtitle {
          font-family: "Lucida Grande", "Lucida Sans Unicode", "Lucida Sans",
            Geneva, Arial, sans-serif;
          letter-spacing: -0.02em;
          font-weight: 300;
          font-style: normal;
          letter-spacing: 0;
          font-size: 28px;
          line-height: 1.22;
          letter-spacing: -0.012em;
        }
        site-rss-button {
          margin: 0 4px;
          padding: 0;
          --site-rss-color: #000000;
          --site-rss-bg-color: var(--haxcms-color, rgba(255, 0, 116, 1));
          --site-rss-paper-button-padding: 0 4px;
          --site-rss-paper-button-margin: 0;
        }

        @media screen and (max-width: 800px) {
          .simple-blog-card-wrapper simple-blog-card {
            margin: 0 10vw;
          }
          .simple-blog-card-wrapper {
            text-align: center;
          }
          #contentcontainer,
          #home {
            padding-left: 8px;
            padding-right: 8px;
          }
          .evenly {
            display: unset;
            justify-content: unset;
          }
          simple-blog-card {
            padding: 0;
          }
          .hide-small {
            display: none !important;
          }
          .annoy-user .rss {
            margin-left: unset;
          }
          .annoy-user {
            position: relative;
            bottom: unset;
            left: unset;
            right: unset;
            padding: 0;
            height: unset;
          }
          .annoy-user span {
            height: unset;
            line-height: unset;
          }
          .annoy-inner {
            max-width: unset;
            margin: 0;
          }
        }
      `]}render(){return _litElement.html`
      <app-header reveals>
        <app-toolbar>
          <div>
            <paper-button
              class="backbutton"
              @click="${this._goBack}"
              title="Back to listing"
            >
              <iron-icon icon="${this.icon}"></iron-icon>
              <span class="hide-small"
                >${this.title} - ${this.activeTitle}</span
              >
            </paper-button>
          </div>
          <div main-title>
            <iron-image
              src="${this.image}"
              preload
              sizing="cover"
              style="height:46px;width:100%;margin: 4px 0 2px 0;"
            ></iron-image>
          </div>
          <div>
            <site-modal
              @site-modal-click="${this.siteModalClick}"
              icon="icons:search"
              title="Search site"
              button-label="Search"
            >
              <site-search></site-search>
            </site-modal>
          </div>
        </app-toolbar>
      </app-header>
      <div class="wrapper">
        <iron-pages selected="${this.selectedPage}">
          <div id="home">
            <site-query
              @result-changed="${this.__mainPostsChanged}"
              limit="2"
              sort='{"created": "ASC"}'
            ></site-query>
            <div class="simple-blog-card-wrapper evenly">
              ${this.__mainPosts.map(post=>_litElement.html`
                  <simple-blog-card
                    alt="${post.metadata.fields&&post.metadata.fields.images&&post.metadata.fields.images[0]&&post.metadata.fields.images[0].alt?post.metadata.fields.images[0].alt:""}"
                    color="${this.color}"
                    .title="${post.title}"
                    size="large"
                    .link="${post.location}"
                    .image="${this._showImage(!!(post.metadata.fields&&post.metadata.fields.images&&post.metadata.fields.images[0]&&post.metadata.fields.images[0].alt)&&post.metadata.fields.images[0].alt)}"
                    .author="${this.author.name}"
                    .timestamp="${post.metadata.created}"
                    .readtime="${post.metadata.readtime}"
                    .authorimage="${this.author.image}"
                    .placeholder="${this.image}"
                    .authorlink="${this.author.socialLink}"
                  >
                    ${post.description}
                  </simple-blog-card>
                `)}
            </div>
            <site-query
              @result-changed="${this.__extraPostsChanged}"
              start-index="2"
              limit="6"
              sort='{"created": "ASC"}'
            ></site-query>
            <div class="simple-blog-card-wrapper">
              ${this.__extraPosts.map(post=>_litElement.html`
                  <simple-blog-card
                    alt="${post.metadata.fields&&post.metadata.fields.images&&post.metadata.fields.images[0]&&post.metadata.fields.images[0].alt?post.metadata.fields.images[0].alt:""}"
                    color="${this.color}"
                    .title="${post.title}"
                    size="medium"
                    .link="${post.location}"
                    .image="${this._showImage(!!(post.metadata.fields&&post.metadata.fields.images&&post.metadata.fields.images[0]&&post.metadata.fields.images[0].alt)&&post.metadata.fields.images[0].alt)}"
                    .author="${this.author.name}"
                    .timestamp="${post.metadata.created}"
                    .readtime="${post.metadata.readtime}"
                    .authorimage="${this.author.image}"
                    .placeholder="${this.image}"
                    .authorlink="${this.author.socialLink}"
                  >
                    ${post.description}
                  </simple-blog-card>
                `)}
            </div>
          </div>
          <div class="contentcontainer-wrapper">
            <div id="contentcontainer">
              <site-git-corner position="right"></site-git-corner>
              <site-active-title></site-active-title>
              <h3 class="subtitle" .hidden="${!this.subtitle}">
                ${this.subtitle}
              </h3>
              <div id="slot">
                <slot></slot>
              </div>
            </div>
            <site-query
              @result-changed="${this.__followUpPostsChanged}"
              limit="3"
              start-index="${this.activeManifestIndexCounter}"
              sort='{"created": "ASC"}'
            ></site-query>
            <div class="simple-blog-card-wrapper">
              ${this.__followUpPosts.map(post=>_litElement.html`
                  <simple-blog-card
                    alt="${post.metadata.fields&&post.metadata.fields.images&&post.metadata.fields.images[0]&&post.metadata.fields.images[0].alt?post.metadata.fields.images[0].alt:""}"
                    color="${this.color}"
                    .title="${post.title}"
                    size="small"
                    .link="${post.location}"
                    .image="${this._showImage(!!(post.metadata.fields&&post.metadata.fields.images&&post.metadata.fields.images[0]&&post.metadata.fields.images[0].alt)&&post.metadata.fields.images[0].alt)}"
                    .author="${this.author.name}"
                    .timestamp="${post.metadata.created}"
                    .readtime="${post.metadata.readtime}"
                    .authorimage="${this.author.image}"
                    .placeholder="${this.image}"
                    .authorlink="${this.author.socialLink}"
                  >
                    ${post.description}
                  </simple-blog-card>
                `)}
            </div>
            <div class="social-float hide-small ${this.stateClass}">
              <ul>
                <li>
                  <social-share-link
                    title="Share on twitter"
                    button-style
                    mode="icon-only"
                    message="${this.shareMsg}"
                    type="Twitter"
                  >
                  </social-share-link>
                </li>
                <li>
                  <social-share-link
                    title="Share on LinkedIn"
                    button-style
                    mode="icon-only"
                    message="${this.shareMsg}"
                    url="${this.shareUrl}"
                    type="LinkedIn"
                  >
                  </social-share-link>
                </li>
                <li>
                  <social-share-link
                    title="Share on Facebook"
                    button-style
                    mode="icon-only"
                    url="${this.shareUrl}"
                    message="${this.shareMsg}"
                    type="Facebook"
                  >
                  </social-share-link>
                </li>
                <li>
                  <social-share-link
                    title="Share on Pinterest"
                    button-style
                    mode="icon-only"
                    message="${this.shareMsg}"
                    image="${this.activeImage}"
                    url="${this.shareUrl}"
                    type="Pinterest"
                  >
                  </social-share-link>
                </li>
              </ul>
            </div>
            <div class="annoy-user ${this.stateClass}">
              <div class="annoy-inner">
                <iron-icon icon="${this.icon}" class="hide-small"></iron-icon>
                <span class="hide-small">
                  Never miss a story from <strong>${this.title}</strong> use RSS
                  today!
                </span>
                <span class="rss">
                  <site-rss-button type="atom"></site-rss-button>
                  <site-rss-button type="rss"></site-rss-button>
                </span>
                <site-share-widget
                  alt="Share on social media"
                ></site-share-widget>
              </div>
            </div>
          </div>
        </iron-pages>
      </div>
    `}static get tag(){return"haxor-slevin"}__mainPostsChanged(e){this.__mainPosts=e.detail.value}__followUpPostsChanged(e){this.__followUpPosts=e.detail.value}static get properties(){return{...super.properties,manifest:{type:Object},color:{type:String},selectedPage:{type:Number,reflect:!0,attribute:"selected-page"},stateClass:{type:String},__mainPosts:{type:Array},__extraPosts:{type:Array},__followUpPosts:{type:Array}}}__extraPostsChanged(e){this.__extraPosts=e.detail.value}_getStateClass(editMode){return editMode?"disable-items":""}_getColor(manifest){if(manifest&&(0,_utils.varExists)(manifest,"metadata.theme.variables.hexCode"))return manifest.metadata.theme.variables.hexCode}siteModalClick(e){new Promise((res,rej)=>_require.default(["../haxcms-elements/lib/ui-components/site/site-search.js"],res,rej))}constructor(){super(),this.__mainPosts=[],this.__extraPosts=[],this.__followUpPosts=[],this.selectedPage=0,new Promise((res,rej)=>_require.default(["../haxcms-elements/lib/ui-components/query/site-query.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-pages/iron-pages.js"],res,rej)),new Promise((res,rej)=>_require.default(["../haxcms-elements/lib/ui-components/active-item/site-share-widget.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/paper-button/paper-button.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-image/iron-image.js"],res,rej)),new Promise((res,rej)=>_require.default(["../simple-blog-card/simple-blog-card.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/app-layout/app-header/app-header.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/app-layout/app-toolbar/app-toolbar.js"],res,rej)),new Promise((res,rej)=>_require.default(["../haxcms-elements/lib/ui-components/active-item/site-active-title.js"],res,rej)),new Promise((res,rej)=>_require.default(["../haxcms-elements/lib/ui-components/active-item/site-git-corner.js"],res,rej)),new Promise((res,rej)=>_require.default(["../haxcms-elements/lib/ui-components/site/site-rss-button.js"],res,rej)),new Promise((res,rej)=>_require.default(["../social-share-link/social-share-link.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/iron-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icon/iron-icon.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/editor-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/device-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/hardware-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/communication-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/social-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/av-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/maps-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/places-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["../haxcms-elements/lib/ui-components/layout/site-modal.js"],res,rej)),(0,_mobxModule.autorun)(reaction=>{let location=(0,_mobxModule.toJS)(_haxcmsSiteStore.store.location);this._noticeLocationChange(location),this.__disposer.push(reaction)}),(0,_mobxModule.autorun)(reaction=>{let manifest=(0,_mobxModule.toJS)(_haxcmsSiteStore.store.manifest);this.title=(0,_utils.varGet)(manifest,"title",""),this.image=(0,_utils.varGet)(manifest,"metadata.theme.variables.image","assets/banner.jpg"),this.icon=(0,_utils.varGet)(manifest,"metadata.theme.variables.icon","icons:record-voice-over"),this.author=(0,_utils.varGet)(manifest,"metadata.author",{}),this.__disposer.push(reaction)}),(0,_mobxModule.autorun)(reaction=>{this.activeManifestIndexCounter=(0,_mobxModule.toJS)(_haxcmsSiteStore.store.activeManifestIndexCounter),this.__disposer.push(reaction)}),(0,_mobxModule.autorun)(reaction=>{this.activeTitle=(0,_mobxModule.toJS)(_haxcmsSiteStore.store.activeTitle),this.shareUrl=document.location.href,this.shareMsg=this.activeTitle+" "+this.shareUrl,(0,_utils.varGet)(_haxcmsSiteStore.store.activeItem,"metadata.fields.subtitle",!1)?this.subtitle=_haxcmsSiteStore.store.activeItem.metadata.fields.subtitle:this.subtitle=!1,(0,_utils.varGet)(_haxcmsSiteStore.store.activeItem,"metadata.fields.images.0.src",!1)&&(this.activeImage=_haxcmsSiteStore.store.activeItem.metadata.fields.images[0].src),this.__disposer.push(reaction)})}updated(changedProperties){super.updated&&super.updated(changedProperties),changedProperties.forEach((oldValue,propName)=>{"manifest"==propName&&(this.color=this._getColor(this[propName])),"editMode"==propName&&(this.stateClass=this._getStateClass(this[propName]))})}_showImage(image){return image||!!this.image&&this.image}_noticeLocationChange(location){if(!location||void 0===location.route)return;const name=location.route.name;"home"===name||"404"===name?this.selectedPage=0:(window.scrollTo({top:0,left:0}),this.selectedPage=1),setTimeout(()=>{var evt=document.createEvent("UIEvents");evt.initUIEvent("resize",!0,!1,window,0),window.dispatchEvent(evt)},50)}disconnectedCallback(){for(var i in this.__disposer)this.__disposer[i].dispose();super.disconnectedCallback()}_goBack(e){window.history.pushState(null,null,_haxcmsSiteStore.store.location.baseUrl),window.dispatchEvent(new PopStateEvent("popstate")),window.scrollTo({top:0,left:0});const evt=new CustomEvent("json-outline-schema-active-item-changed",{bubbles:!0,cancelable:!0,composed:!0,detail:{}});this.dispatchEvent(evt),this.selectedPage=0}}_exports.HaxorSlevin=HaxorSlevin,window.customElements.define(HaxorSlevin.tag,HaxorSlevin)});
