// @lrnwebcomponents/lrndesign-gallery/lrndesign-gallery.js
define(["exports", "../../lit-element/lit-element.js", "./lib/lrndesign-gallery-behaviors.js", "../responsive-utility/responsive-utility.js", "./lib/lrndesign-gallery-carousel.js", "./lib/lrndesign-gallery-grid.js"], function (_exports, _litElement, _lrndesignGalleryBehaviors, _responsiveUtility, _lrndesignGalleryCarousel, _lrndesignGalleryGrid) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrndesignGallery = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject4_60842b70939b11eaa49d972a432e908f() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n\n        :host([hidden]) {\n          display: none;\n        }\n\n        ::slotted(figure) {\n          display: none;\n        }\n      "]);

    _templateObject4_60842b70939b11eaa49d972a432e908f = function _templateObject4_60842b70939b11eaa49d972a432e908f() {
      return data;
    };

    return data;
  }

  function _templateObject3_60842b70939b11eaa49d972a432e908f() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <lrndesign-gallery-carousel\n                accent-color=\"", "\"\n                .aspect-ratio=\"", "\"\n                .extra-wide=\"", "\"\n                ?dark=\"", "\"\n                .gallery-id=\"", "\"\n                @item-changed=\"", "\"\n                responsive-size=\"", "\"\n                .selected=\"", "\"\n                sizing=\"", "\"\n                .sources=\"", "\"\n                gallery-title=\"", "\"\n              >\n                <slot></slot>\n              </lrndesign-gallery-carousel>\n            "]);

    _templateObject3_60842b70939b11eaa49d972a432e908f = function _templateObject3_60842b70939b11eaa49d972a432e908f() {
      return data;
    };

    return data;
  }

  function _templateObject2_60842b70939b11eaa49d972a432e908f() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <lrndesign-gallery-grid\n                accent-color=\"", "\"\n                .aspect-ratio=\"", "\"\n                .extra-wide=\"", "\"\n                ?dark=\"", "\"\n                .gallery-id=\"", "\"\n                @item-changed=\"", "\"\n                responsive-size=\"", "\"\n                .selected=\"", "\"\n                sizing=\"", "\"\n                .sources=\"", "\"\n                gallery-title=\"", "\"\n              >\n                <slot></slot>\n              </lrndesign-gallery-grid>\n            "]);

    _templateObject2_60842b70939b11eaa49d972a432e908f = function _templateObject2_60842b70939b11eaa49d972a432e908f() {
      return data;
    };

    return data;
  }

  function _templateObject_60842b70939b11eaa49d972a432e908f() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div id=\"gallery\">\n        ", "\n      </div>\n    "]);

    _templateObject_60842b70939b11eaa49d972a432e908f = function _templateObject_60842b70939b11eaa49d972a432e908f() {
      return data;
    };

    return data;
  }

  /**
   * `lrndesign-gallery`
   * displays images as a carousel or grid with the ability to zoom
   *
   * @element lrndesign-gallery
   * @lit-html
   * @lit-element
   * @demo demo/index.html
   */
  var LrndesignGallery =
  /*#__PURE__*/
  function (_LrndesignGalleryBeha) {
    babelHelpers.inherits(LrndesignGallery, _LrndesignGalleryBeha);
    babelHelpers.createClass(LrndesignGallery, [{
      key: "render",
      // render function
      value: function render() {
        var _this = this;

        return (0, _litElement.html)(_templateObject_60842b70939b11eaa49d972a432e908f(), this.grid ? (0, _litElement.html)(_templateObject2_60842b70939b11eaa49d972a432e908f(), this.accentColor, this.aspect, this.extra, this.dark, this.id, function (e) {
          return _this.goToItem(e.detail);
        }, this.responsiveSize, this.selected, this.sizing, this.items, this.galleryTitle) : (0, _litElement.html)(_templateObject3_60842b70939b11eaa49d972a432e908f(), this.accentColor, this.aspect, this.extra, this.dark, this.id, function (e) {
          return _this.goToItem(e.detail);
        }, this.responsiveSize, this.selected, this.sizing, this.items, this.galleryTitle));
      } // properties available to the custom element for data binding

    }, {
      key: "tag",

      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */
      value: function tag() {
        return "lrndesign-gallery";
      } // life cycle

    }], [{
      key: "styles",
      //styles function
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(LrndesignGallery), "styles", this)), [(0, _litElement.css)(_templateObject4_60842b70939b11eaa49d972a432e908f())]);
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(LrndesignGallery), "properties", this), {
          id: {
            type: String,
            reflect: true,
            attribute: "id"
          },
          responsiveSize: {
            type: String,
            reflect: true,
            attribute: "responsive-size"
          }
        });
      }
    }]);

    function LrndesignGallery() {
      babelHelpers.classCallCheck(this, LrndesignGallery);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrndesignGallery).call(this));
    }
    /**
     * life cycle, element is afixed to the DOM
     */


    babelHelpers.createClass(LrndesignGallery, [{
      key: "connectedCallback",
      value: function connectedCallback() {
        babelHelpers.get(babelHelpers.getPrototypeOf(LrndesignGallery.prototype), "connectedCallback", this).call(this);
        window.ResponsiveUtility.requestAvailability();
        this.updateGallery();
        this.observer.observe(this, {
          attributes: false,
          childList: true,
          subtree: false
        });
      }
    }, {
      key: "firstUpdated",
      value: function firstUpdated() {
        babelHelpers.get(babelHelpers.getPrototypeOf(LrndesignGallery.prototype), "firstUpdated", this).call(this);
        window.dispatchEvent(new CustomEvent("responsive-element", {
          detail: {
            element: this,
            attribute: "responsive-size",
            sm: 300,
            md: 600,
            lg: 1000,
            xl: 1500
          }
        }));
        this.anchorData = this._getAnchorData();

        if (this.anchorData.gallery === this.id) {
          this.goToItem(this.anchorData.id);
        } else {
          this.goToItem();
        }
      }
      /**
       * handle updates
       */

    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        babelHelpers.get(babelHelpers.getPrototypeOf(LrndesignGallery.prototype), "updated", this).call(this, changedProperties);
        changedProperties.forEach(function (oldValue, propName) {
          if (propName === "id" && !_this2.id) _this2.id = "gallery-".concat(_this2._generateUUID());
        });
      }
      /**
       * gets aspect ratio of images
       *
       * @readonly
       * @memberof LrndesignGallery
       */

    }, {
      key: "goToItem",

      /**
       * go to item by id, or index
       *
       * @param {string} query
       */
      value: function goToItem(query) {
        var start = this.items[0] || {};

        if (typeof query === "number" && query >= 0 && query < this.items.length) {
          this.selected = this.items[query] || start;
        } else {
          var matches = this.items.filter(function (item) {
            return item.id === query;
          });
          this.selected = matches.length > 0 ? matches[0] : start;
        }
      }
    }, {
      key: "updateGallery",
      value: function updateGallery() {
        var sources = [],
            figures = this.querySelectorAll("figure");
        figures.forEach(function (figure) {
          var id = figure.getAttribute("id"),
              img = figure.querySelector("img"),
              sizing = figure.getAttribute("sizing"),
              query = [1, 2, 3, 4, 5, 6].map(function (num) {
            return "h".concat(num, ":first-child");
          }).join(","),
              src = img && img.getAttribute("src") ? img.getAttribute("src") : undefined,
              srcset = img && img.getAttribute("srcset") ? img.getAttribute("srcset").split(",") : undefined,
              thumbset = srcset && srcset[0] ? srcset[0].split(" ") : undefined,
              largeset = srcset && srcset[srcset.length - 1] ? srcset[srcset.length - 1].split(" ") : undefined,
              thumbnail = thumbset && thumbset[0] ? thumbset[0] : undefined,
              large = largeset && largeset[0] ? largeset[0] : undefined,
              details = figure.querySelector("figcaption") ? figure.querySelector("figcaption").cloneNode(true) : undefined,
              alt = img && img.getAttribute("alt") ? img.getAttribute("alt") : undefined,
              figheading = details && details.querySelector(query) ? details.querySelector(query) : undefined,
              title = figheading.innerHTML;
          if (figheading) figheading.remove();
          sources.push({
            alt: alt,
            id: id,
            src: src,
            thumbnail: thumbnail,
            large: large,
            title: title,
            details: details.innerHTML,
            sizing: sizing
          });
        });
        if (sources.length > 0 && this.sources.length < 1) this.sources = sources;
        console.log(sources, this.items);
      }
      /**
       * gets aspect ratio of an image and
       * determine if aspect ratio is extra wide
       *
       * @param {array}
       */

    }, {
      key: "_getAnchorData",
      value: function _getAnchorData() {
        var hash = window.location.hash !== null && window.location.hash !== "" ? window.location.hash.replace("#", "") : false,
            data = hash ? hash.split("---") : [];
        return {
          id: data.length > 1 ? data[1] : -1,
          gallery: data.length > 0 ? data[0] : -1,
          zoom: scroll && data.length > 2 && data[2] === "zoom"
        };
      }
    }, {
      key: "aspect",
      get: function get() {
        var items = (this.items || []).filter(function (item) {
          return item.src && item.src != "";
        }),
            src = items && items[0] ? items[0].src : false;

        if (src) {
          var img = new Image();
          img.src = src;
          return img.naturalWidth > 0 && img.naturalHeight > 0 ? img.naturalWidth / img.naturalHeight : 1.33333333;
        } else {
          return 1.33333333;
        }
      }
      /**
       * gets aspect ratio of an image and
       * determine if aspect ratio is extra wide
       *
       * @readonly
       * @memberof LrndesignGallery
       */

    }, {
      key: "extra",
      get: function get() {
        var ew = this.aspect > 2;
        return ew;
      }
      /**
       * gets items array
       *
       * @readonly
       * @memberof LrndesignGalleryBehaviors
       */

    }, {
      key: "items",
      get: function get() {
        var _this3 = this;

        var sources = this.sources || [],
            items = typeof sources === "string" ? JSON.parse(sources) : sources,
            total = items.length,
            itemData = (items || []).map(function (item, i) {
          return {
            details: item.details,
            index: i,
            id: item.id || "gallery-".concat(_this3.id, "-item-").concat(i),
            src: item.src,
            large: item.large && item.large !== "" ? item.large : item.src,
            thumbnail: item.thumbnail && item.thumbnail != "" ? item.thumbnail : item.src,
            xofy: "".concat(i + 1, " of ").concat(total),
            next: i + 1 < total ? i + 1 : -1,
            prev: i - 1 > -1 ? i - 1 : -1,
            sizing: item.sizing && item.sizing != "" ? item.sizing : _this3.sizing,
            title: item.title,
            tooltip: "".concat(item.title || "Image ".concat(i), " (Zoom In)"),
            heading: "".concat(item.title || "Image ".concat(i), " (Full-Sized)")
          };
        });
        return itemData;
      }
      /**
       * mutation observer for tabs
       * @readonly
       * @returns {object}
       */

    }, {
      key: "observer",
      get: function get() {
        var _this4 = this;

        var callback = function callback() {
          return _this4.updateGallery();
        };

        return new MutationObserver(callback);
      }
    }]);
    return LrndesignGallery;
  }(_lrndesignGalleryBehaviors.LrndesignGalleryBehaviors);

  _exports.LrndesignGallery = LrndesignGallery;
  customElements.define("lrndesign-gallery", LrndesignGallery);
});
