// @lrnwebcomponents/wave-player/wave-player.js
define(["exports","meta","require","../../@polymer/polymer/polymer-element.js","../../@polymer/polymer/lib/utils/resolve-url.js","../../@polymer/paper-material/paper-material.js","../../@polymer/paper-fab/paper-fab.js","../../@polymer/paper-icon-button/paper-icon-button.js","../schema-behaviors/schema-behaviors.js","../es-global-bridge/es-global-bridge.js"],function(_exports,meta,_require,_polymerElement,_resolveUrl,_paperMaterial,_paperFab,_paperIconButton,_schemaBehaviors,_esGlobalBridge){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.WavePlayer=void 0,meta=babelHelpers.interopRequireWildcard(meta),_require=babelHelpers.interopRequireWildcard(_require);class WavePlayer extends((0,_schemaBehaviors.SchemaBehaviors)(_polymerElement.PolymerElement)){static get template(){return _polymerElement.html`
      <style>
        :host {
          height: 150px;
          background-color: var(--dark-primary-color);
          display: block;
        }

        paper-icon-button {
          position: absolute;
        }

        .title,
        .subtitle {
          transition: all 0.5s ease;
          padding: 10px 10px 10px 0;
          left: 160px;
          position: absolute;
        }

        .subtitle {
          bottom: 0;
        }

        .controls {
          height: 50px;
          width: 100%;
          top: 0;
          background: var(--accent-color);
          z-index: 20;
        }

        paper-fab {
          transition: all 0.5s ease;
          top: -25px;
          z-index: 25;
          border-radius: 0;
        }

        .albuminfo {
          position: relative;
          transition: all 0.5s ease;
          top: -156px;
          margin-bottom: -150px;
          z-index: 20;
          height: 150px;
          background-color: rgba(0, 0, 0, 0.4);
          color: #fff;
          font-family: Roboto, sans-serif;
        }

        .albuminfoActive {
          top: -25;
          height: 25px;
          width: 100%;
          margin-bottom: -19px;
        }

        .waveContainer {
          top: -31px;
          transition: all 0.5s ease;
          background-color: var(--dark-primary-color);
          transform: scaleY(1.5);
        }

        .circleAnimation {
          border-radius: 50%;
          overflow: auto;
          -moz-box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.4);
          box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.4);
        }

        .circleAnimation:active {
          -moz-box-shadow: 0 8px 17px 0 rgba(0, 0, 0, 0.2);
          box-shadow: 0 8px 17px 0 rgba(0, 0, 0, 0.2);
        }

        .playActive {
          top: 0;
          width: 100%;
          height: 50px;
        }

        .waveActive {
          top: 0px;
          transform: scaleY(1);
        }

        .centred,
        .titleActive {
          transform: scaleY(0);
        }

        .titleActive {
          opacity: 0;
        }

        #playbutton {
          transition: all 0.5s ease;
        }

        .coverart {
          transition: all 0.5s ease;
          width: 150px;
          height: 150px;
        }

        .title {
          font-size: 24px;
        }

        .coverartActive {
          width: 25px;
          height: 25px;
        }

        .nameActive {
          font-size: 19px;
          padding: 3px 3px 3px 0;
          left: 30px;
        }

        .centred {
          top: calc(50% - 20px);
          left: calc(50% - 20px);
          transition: all 0.3s ease;
        }

        .left,
        .middle,
        .right {
          transform: scale(1);
        }

        .left {
          left: calc(25% - 20px);
        }

        .right {
          left: calc(75% - 20px);
        }

        .hidden {
          display: none;
        }

        @media only screen and (max-width: 500px) {
          .albuminfo {
            width: 100%;
          }
        }
      </style>
      <paper-fab
        id="playbutton"
        class="circleAnimation"
        disabled=""
        icon="av:play-arrow"
        on-click="togglePlay"
      ></paper-fab>
      <paper-material id="controls" class="controls hidden" elevation="2">
        <paper-icon-button
          class="centred middle"
          style="color: white;"
          icon="av:pause"
          on-click="togglePlay"
        ></paper-icon-button>
        <paper-icon-button
          id="replay"
          class="centred"
          style="color: white;"
          icon="av:replay-30"
          on-click="throwBack"
        ></paper-icon-button>
        <paper-icon-button
          id="mute"
          class="centred"
          style="color: white;"
          icon="av:volume-up"
          on-click="toggleMute"
        ></paper-icon-button>
      </paper-material>
      <div id="container" class="waveContainer" elevation="0"></div>
      <div id="albuminfo" class="albuminfo">
        <img class="coverart" src="[[coverart]]" />
        <span class="title">[[title]]</span>
        <span class="subtitle">[[subtitle]]</span>
      </div>
    `}static get tag(){return"wave-player"}static get properties(){let props={src:{type:String,notify:!0,observer:"_srcChanged"},title:{type:String,value:"",notify:!0},subtitle:{type:String,value:"",notify:!0},coverart:{type:String,value:"",notify:!0},wavesurfer:{type:Object},lean:{type:String,value:"left",notify:!0},wavecolor:{type:String,value:"#ffffff",notify:!0},progresscolor:{type:String,value:"#CFD8DC",notify:!0}};return super.properties&&(props=Object.assign(props,super.properties)),props}_srcChanged(newValue,oldValue){void 0!==newValue&&this.__wavesurfer&&window.wavesurferobject.load(newValue)}constructor(){super(),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/iron-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/av-icons.js"],res,rej));const location=`${(0,_resolveUrl.pathFromUrl)(decodeURIComponent(meta.url))}lib/wavesurfer.js/dist/wavesurfer.js`;window.addEventListener("es-bridge-wavesurfer-loaded",this._wavesurferLoaded.bind(this)),window.ESGlobalBridge.requestAvailability(),window.ESGlobalBridge.instance.load("wavesurfer",location)}disconnectedCallback(){window.removeEventListener("es-bridge-wavesurfer-loaded",this._wavesurferLoaded.bind(this)),super.disconnectedCallback()}ready(){if(super.ready(),"right"===this.lean?(this.$.playbutton.style.right="25",this.$.controls.style.right="0"):(this.$.playbutton.style.left="25",this.$.controls.style.left="0"),""===this.name&&this.$.albuminfo.classList.add("hidden"),""===this.coverart){const basePath=(0,_resolveUrl.pathFromUrl)(decodeURIComponent(meta.url));this.coverart=`${basePath}lib/art.jpg`}}_wavesurferLoaded(){this.__wavesurfer=!0,this.initWaveSurfer()}activateAnimation(){var waveStyle=this.$.container,buttonStyle=this.$.playbutton,controlsStyle=this.$.controls,muteStyle=this.$.mute,replayStyle=this.$.replay,albumStyle=this.$.albuminfo,coverartStyle=albumStyle.querySelector(".coverart"),nameStyle=albumStyle.querySelector(".title"),titleStyle=albumStyle.querySelector(".subtitle");buttonStyle.setAttribute("icon","av:pause"),buttonStyle.classList.remove("circleAnimation"),buttonStyle.classList.add("playActive"),albumStyle.classList.add("albuminfoActive"),coverartStyle.classList.add("coverartActive"),nameStyle.classList.add("nameActive"),titleStyle.classList.add("titleActive"),"right"===this.lean?this.$.playbutton.style.right="0":this.$.playbutton.style.left="0",waveStyle.classList.add("waveActive"),setTimeout(function(){controlsStyle.classList.remove("hidden"),buttonStyle.classList.add("hidden")},500),setTimeout(function(){muteStyle.classList.add("right"),replayStyle.classList.add("left")},600)}deactivateAnimation(){var self=this,waveStyle=this.$.container,buttonStyle=this.$.playbutton,controlsStyle=this.$.controls,muteStyle=this.$.mute,replayStyle=this.$.replay,albumStyle=this.$.albuminfo,coverartStyle=albumStyle.querySelector(".coverart"),nameStyle=albumStyle.querySelector(".title"),titleStyle=albumStyle.querySelector(".subtitle");muteStyle.classList.remove("right"),replayStyle.classList.remove("left"),setTimeout(function(){controlsStyle.classList.add("hidden"),buttonStyle.classList.remove("hidden")},100),setTimeout(function(){buttonStyle.setAttribute("icon","av:play-arrow"),buttonStyle.classList.add("circleAnimation"),buttonStyle.classList.remove("playActive"),albumStyle.classList.remove("albuminfoActive"),coverartStyle.classList.remove("coverartActive"),nameStyle.classList.remove("nameActive"),titleStyle.classList.remove("titleActive"),"right"===self.lean?buttonStyle.style.right="25":buttonStyle.style.left="25",waveStyle.classList.remove("waveActive")},200)}initWaveSurfer(){window.wavesurferobject=new WaveSurfer({container:this.$.container,waveColor:this.wavecolor,progressColor:this.progresscolor,fillParent:!0,height:100}),window.wavesurferobject.init(),void 0!==this.src&&window.wavesurferobject.load(this.src),window.wavesurferobject.on("ready",()=>{this.$.playbutton.removeAttribute("disabled")}),window.wavesurferobject.on("finish",()=>{this.deactivateAnimation()})}togglePlay(e){window.wavesurferobject.playPause(),"av:play-arrow"===this.$.playbutton.getAttribute("icon")?this.activateAnimation():this.deactivateAnimation()}toggleMute(e){var muteStyle=this.$.mute,iconType=muteStyle.getAttribute("icon");window.wavesurferobject.toggleMute(),"av:volume-up"===iconType?muteStyle.setAttribute("icon","av:volume-off"):muteStyle.setAttribute("icon","av:volume-up")}throwBack(e){window.wavesurferobject.skipBackward(30)}static get haxProperties(){return{canScale:!0,canPosition:!0,canEditSource:!1,gizmo:{title:"Audio player",description:"Audio that is just like spotify.",icon:"av:play-circle-filled",color:"purple",groups:["Video","Media"],handles:[{type:"audio",source:"src",title:"title",caption:"subtitle"}],meta:{author:"LRNWebComponents"}},settings:{quick:[{property:"src",title:"Source",description:"The URL for this video.",inputMethod:"textfield",icon:"link",required:!0,validationType:"url"}],configure:[{property:"src",title:"Source",description:"The URL for this video.",inputMethod:"textfield",icon:"link",required:!0,validationType:"url"},{property:"title",title:"Title",description:"A simple title",inputMethod:"textfield",icon:"av:video-label",required:!1,validationType:"text"}],advanced:[]}}}}_exports.WavePlayer=WavePlayer,window.customElements.define(WavePlayer.tag,WavePlayer)});
