// @lit/reactive-element/decorators/query-assigned-nodes.js
define(["exports", "./base.js", "./query-assigned-elements.js"], function (_exports, _base, _queryAssignedElements) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.queryAssignedNodes = o;

  /**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  function o(o, n, r) {
    var l,
        s = o;
    return "object" == babelHelpers.typeof(o) ? (s = o.slot, l = o) : l = {
      flatten: n
    }, r ? (0, _queryAssignedElements.queryAssignedElements)({
      slot: s,
      flatten: n,
      selector: r
    }) : (0, _base.decorateProperty)({
      descriptor: function descriptor(e) {
        return {
          get: function get() {
            var e, t;
            var o = "slot" + (s ? "[name=".concat(s, "]") : ":not([name])"),
                n = null === (e = this.renderRoot) || void 0 === e ? void 0 : e.querySelector(o);
            return null !== (t = null == n ? void 0 : n.assignedNodes(l)) && void 0 !== t ? t : [];
          },
          enumerable: !0,
          configurable: !0
        };
      }
    });
  }
});
