// @lrnwebcomponents/simple-fields/lib/simple-fields-form.js
define(["exports", "../../../lit/index.js", "./simple-fields-form-lite.js", "../simple-fields.js"], function (_exports, _index, _simpleFieldsFormLite, _simpleFields) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SimpleFieldsForm = void 0;

  var _templateObject_bf6fc140012611ec9676957c409166bc;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `simple-fields-form`
   * binding and submission capabilities on top of simple-fields
   *
   * @group simple-fields
   * @element simple-fields-form
   * @demo ./demo/form.html
   */
  var SimpleFieldsForm = /*#__PURE__*/function (_SimpleFieldsFormLite) {
    babelHelpers.inherits(SimpleFieldsForm, _SimpleFieldsFormLite);

    var _super = _createSuper(SimpleFieldsForm);

    function SimpleFieldsForm() {
      babelHelpers.classCallCheck(this, SimpleFieldsForm);
      return _super.apply(this, arguments);
    }

    babelHelpers.createClass(SimpleFieldsForm, [{
      key: "render",
      value: // render function
      function render() {
        return (0, _index.html)(_templateObject_bf6fc140012611ec9676957c409166bc || (_templateObject_bf6fc140012611ec9676957c409166bc = babelHelpers.taggedTemplateLiteral(["\n      <form part=\"form\">\n        <slot name=\"before\"></slot>\n        <simple-fields\n          id=\"sf\"\n          .autofocus=\"", "\"\n          ?watchColorPrefs=\"", "\"\n          language=\"", "\"\n          .resources=\"", "\"\n          .schema=\"", "\"\n          .fields=\"", "\"\n          .schematizer=\"", "\"\n          .elementizer=\"", "\"\n          .value=\"", "\"\n          @value-changed=\"", "\"\n          part=\"fields\"\n        >\n        </simple-fields>\n        <slot></slot>\n      </form>\n    "])), !this.disableAutofocus, this.watchColorPrefs, this.language || "", this.resources, this.schema, this.fields, this.fieldsConversion, this.elementizer, this.value, this._valueChanged);
      }
      /**
       * applies loaded datda to simple-fields-lite
       *
       * @memberof SimpleFieldsFormLite
       */

    }, {
      key: "_applyLoadedData",
      value: function _applyLoadedData() {
        if (this.loadResponse.data.schema) {
          this.schema = this.loadResponse.data.schema;
        } else if (this.loadResponse.data.fields) {
          this.fields = this.loadResponse.data.fields;
        }

        if (this.loadResponse.data.value) this.value = this.loadResponse.data.value;
      }
      /**
       * properties specific to field function
       *
       * @readonly
       * @static
       * @memberof SimpleFieldsFormLite
       */

    }, {
      key: "defaultSchemaConversion",
      get:
      /**
       * gets default schemaConversion so parts of it can be overridden easily
       *
       * @readonly
       * @memberof SimpleFields
       */
      function get() {
        return _simpleFields.SimpleFields.defaultSchemaConversion;
      }
    }], [{
      key: "tag",
      get: function get() {
        return "simple-fields-form";
      }
    }, {
      key: "fieldProperties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(SimpleFieldsForm), "fieldProperties", this)), {}, {
          /**
           * Fields to convert to JSON Schema.
           */
          fields: {
            type: "Array"
          },

          /**
           * Conversion from inputMethods to JSON schema types and formats.
           * _See [Configuring fieldsConversion Property](configuring-the-fieldsconversion-property) above._
           */
          fieldsConversion: {
            type: "Object",
            attribute: "fields-conversion"
          }
        });
      }
    }]);
    return SimpleFieldsForm;
  }(_simpleFieldsFormLite.SimpleFieldsFormLite);

  _exports.SimpleFieldsForm = SimpleFieldsForm;
  window.customElements.define(SimpleFieldsForm.tag, SimpleFieldsForm);
});
