// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-submission/lrnapp-studio-submission-edit-image.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@polymer/iron-image/iron-image.js", "./lrnapp-studio-submission-media-editoverlay.js"], function (_exports, _litElement, _ironImage, _lrnappStudioSubmissionMediaEditoverlay) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioSubmissionEditImage = void 0;

  class LrnappStudioSubmissionEditImage extends _litElement.LitElement {
    /**
     * LitElement constructable styles enhancement
     */
    static get styles() {
      return [_litElement.css`
        :host {
          display: inline-flex;
          justify-content: space-around;
          min-height: 100px;
          position: relative;
        }

        iron-image {
          --iron-image-height: 200px;
          display: block;
        }
      `];
    }

    render() {
      return _litElement.html`
      <lrnapp-studio-submission-media-editoverlay
        data-index="${this.index}"
        embedcode="${this.embedcode}"
      >
        <iron-image
          src="${this.thumbnail}"
          style="width:200px; height:200px; background-color: lightgray;"
          sizing="contain"
          class="contain"
          preload=""
          fade=""
        ></iron-image>
      </lrnapp-studio-submission-media-editoverlay>
    `;
    }

    updated(changedProperties) {
      changedProperties.forEach((oldValue, propName) => {
        if (propName == "image") {
          this.embedcode = this._computeEmbedCode(this.image);
          this.thumbnail = this._getThumbnail(this.image); // notify

          this.dispatchEvent(new CustomEvent("image-changed", {
            detail: {
              value: this[propName]
            }
          }));
        }

        if (propName == "thumbnail") {
          // notify
          this.dispatchEvent(new CustomEvent("thumbnail-changed", {
            detail: {
              value: this[propName]
            }
          }));
        }
      });
    }

    static get tag() {
      return "lrnapp-studio-submission-edit-image";
    }

    static get properties() {
      return {
        index: {
          type: String
        },
        image: {
          type: Object
        },
        thumbnail: {
          type: String
        },
        embedcode: {
          type: String
        }
      };
    }

    _getThumbnail(image) {
      return image.url;
    }

    _computeEmbedCode(image) {
      return "![Alternative Text Here](" + image.url + ")";
    }

  }

  _exports.LrnappStudioSubmissionEditImage = LrnappStudioSubmissionEditImage;
  window.customElements.define(LrnappStudioSubmissionEditImage.tag, LrnappStudioSubmissionEditImage);
});
