// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-submission/lrnapp-studio-submission-edit-add-asset.js
define(["exports", "../../../../@polymer/polymer/polymer-element.js", "../../../simple-icon/simple-icon.js", "../../../simple-icon/lib/simple-icons.js", "../../../hax-iconset/lib/simple-hax-iconset.js"], function (_exports, _polymerElement, _simpleIcon, _simpleIcons, _simpleHaxIconset) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioSubmissionEditAddAsset = void 0;

  class LrnappStudioSubmissionEditAddAsset extends _polymerElement.PolymerElement {
    static get template() {
      return (0, _polymerElement.html)`
      <style>
        :host {
          cursor: pointer;
        }
        :host([display="box"]) {
          display: block;
          width: 200px;
          height: 200px;
          background: lightgray;
          display: flex;
          align-items: center;
          justify-content: center;
          position: relative;
        }
        :host([display="pill"]) {
          display: block;
          background: lightgray;
          display: inline-flex;
          align-items: center;
          justify-content: center;
          position: relative;
        }
        simple-icon {
          color: gray;
          --simple-icon-height: 50px;
          --simple-icon-width: 50px;
        }
        :host([display="pill"]) simple-icon {
          --simple-icon-height: 50px;
          --simple-icon-width: 50px;
        }
      </style>
      <button>
        <simple-icon icon="[[icon]]"></simple-icon>
      </button>
    `;
    }

    static get tag() {
      return "lrnapp-studio-submission-edit-add-asset";
    }

    static get properties() {
      return {
        icon: {
          type: String,
          value: "add"
        },

        /**
         * Change the display of the add asset element
         *
         * @type String Options: 'box', 'pill'
         */
        display: {
          type: String,
          value: "box",
          reflectToAttribute: true
        }
      };
    }

  }

  _exports.LrnappStudioSubmissionEditAddAsset = LrnappStudioSubmissionEditAddAsset;
  window.customElements.define(LrnappStudioSubmissionEditAddAsset.tag, LrnappStudioSubmissionEditAddAsset);
});
