// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-submission/lrnapp-studio-submission-critique.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@vaadin/vaadin-split-layout/vaadin-split-layout.js", "./lrnapp-studio-submission-media-editoverlay.js", "./lrnapp-studio-submission-edit-images.js", "./lrnapp-studio-submission-edit-files.js", "./lrnapp-studio-submission-edit-video.js", "./lrnapp-studio-submission-edit-links.js", "./lrnapp-studio-submission-display.js", "./lrnapp-studio-submission-edit-textarea.js", "./lrnapp-studio-block.js", "./lrnapp-studio-submission-critique-panel.js"], function (_exports, _litElement, _vaadinSplitLayout, _lrnappStudioSubmissionMediaEditoverlay, _lrnappStudioSubmissionEditImages, _lrnappStudioSubmissionEditFiles, _lrnappStudioSubmissionEditVideo, _lrnappStudioSubmissionEditLinks, _lrnappStudioSubmissionDisplay, _lrnappStudioSubmissionEditTextarea, _lrnappStudioBlock, _lrnappStudioSubmissionCritiquePanel) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioSubmissionCritique = void 0;

  class LrnappStudioSubmissionCritique extends _litElement.LitElement {
    /**
     * LitElement constructable styles enhancement
     */
    static get styles() {
      return [_litElement.css`
        :host {
          display: block;
        }

        .submission-critique {
          display: flex;
          justify-content: space-around;
        }

        .submission-critique-panel {
          flex: 1 1 auto;
          width: 100%;
          flex-wrap: nowrap;
          margin: 0.3em;
        }
      `];
    }

    render() {
      return _litElement.html`
      ${this.edit ? _litElement.html`
            <div class="submission-critique">
              <!-- critique panel -->
              <div class="submission-critique-panel" id="crititque-panel">
                <lrnapp-studio-submission-display
                  .submission="${this.submission.relationships.relatedSubmission}"
                ></lrnapp-studio-submission-display>
              </div>
              <!-- critique panel -->
              <div class="submission-critique-panel">
                <lrnapp-studio-submission-critique-panel
                  .submission="${this.submission}"
                  ?edit="${this.edit}"
                ></lrnapp-studio-submission-critique-panel>
              </div>
            </div>
          ` : _litElement.html`
            <vaadin-split-layout class="submission-critique">
              <!-- critique panel -->
              <div class="submission-critique-panel" id="crititque-panel">
                <lrnapp-studio-submission-display
                  .submission="${this.submission.relationships.relatedSubmission}"
                ></lrnapp-studio-submission-display>
              </div>
              <!-- critique panel -->
              <div class="submission-critique-panel">
                <lrnapp-studio-submission-critique-panel
                  .submission="${this.submission}"
                  @submission-changed="${this.submissionChanged}"
                  ?edit="${this.edit}"
                ></lrnapp-studio-submission-critique-panel>
              </div>
            </vaadin-split-layout>
          `}
    `;
    }

    submissionChanged(e) {
      this.submission = { ...e.detail.value
      };
    }

    updated(changedProperties) {
      changedProperties.forEach((oldValue, propName) => {
        if (propName == "submission") {
          // notify
          this.dispatchEvent(new CustomEvent("submission-changed", {
            detail: {
              value: this[propName]
            }
          }));
        }
      });
    }

    constructor() {
      super();
      this.edit = false;
    }

    static get tag() {
      return "lrnapp-studio-submission-critique";
    }

    static get properties() {
      return {
        submission: {
          type: Object
        },
        edit: {
          type: Boolean,
          reflect: true
        }
      };
    }
    /**
     * Simple way to convert from object to array.
     */


    _toArray(obj) {
      if (obj == null) {
        return [];
      }

      return Object.keys(obj).map(function (key) {
        return obj[key];
      });
    }

  }

  _exports.LrnappStudioSubmissionCritique = LrnappStudioSubmissionCritique;
  window.customElements.define(LrnappStudioSubmissionCritique.tag, LrnappStudioSubmissionCritique);
});
