// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-submission/lrnapp-studio-submission-edit-textarea.js
define(["exports", "../../../../@polymer/polymer/polymer-element.js", "../../../lrn-markdown-editor/lrn-markdown-editor.js"], function (_exports, _polymerElement, _lrnMarkdownEditor) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioSubmissionEditTextArea = void 0;

  class LrnappStudioSubmissionEditTextArea extends _polymerElement.PolymerElement {
    static get template() {
      return _polymerElement.html`
      <style>
        :host {
          display: block;
        }
      </style>
      <lrn-markdown-editor content="{{content}}"></lrn-markdown-editor>
    `;
    }

    static get tag() {
      return "lrnapp-studio-submission-edit-textarea";
    }

    static get properties() {
      return {
        content: {
          type: String,
          notify: true
        }
      };
    }

  }

  _exports.LrnappStudioSubmissionEditTextArea = LrnappStudioSubmissionEditTextArea;
  window.customElements.define(LrnappStudioSubmissionEditTextArea.tag, LrnappStudioSubmissionEditTextArea);
});
