// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-submission/lrnapp-studio-submission-edit-file.js
define(["exports", "../../../../@polymer/polymer/polymer-element.js", "./lrnapp-studio-submission-media-editoverlay.js"], function (_exports, _polymerElement, _lrnappStudioSubmissionMediaEditoverlay) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioSubmissionEditFile = void 0;

  class LrnappStudioSubmissionEditFile extends _polymerElement.PolymerElement {
    static get template() {
      return _polymerElement.html`
      <style>
        :host {
          display: inline-flex;
          justify-content: space-around;
          align-items: stretch;
          min-height: 100px;
          width: 100px;
          position: relative;
          background: lightgray;
          color: #333333;
        }

        lrnapp-studio-submission-media-editoverlay {
          display: flex;
        }

        .file_url {
          margin: 1em;
        }
      </style>

      <lrnapp-studio-submission-media-editoverlay
        data-index\$="[[index]]"
        embedcode="{{embedcode}}"
      >
        <div class="file_url">[[file.filename]]</div>
      </lrnapp-studio-submission-media-editoverlay>
    `;
    }

    static get tag() {
      return "lrnapp-studio-submission-edit-file";
    }

    static get properties() {
      return {
        file: {
          type: Object
        },
        embedcode: {
          type: String,
          computed: "_computeEmbedCode(file)"
        }
      };
    }

    _computeEmbedCode(file) {
      return "[Alternative Text Here](" + file.url + ")";
    }

  }

  _exports.LrnappStudioSubmissionEditFile = LrnappStudioSubmissionEditFile;
  window.customElements.define(LrnappStudioSubmissionEditFile.tag, LrnappStudioSubmissionEditFile);
});
