// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-submission/lrnapp-studio-submission-comments.js
define(["exports", "../../../../lit-element/lit-element.js", "./lrnapp-studio-submission-comment.js"], function (_exports, _litElement, _lrnappStudioSubmissionComment) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioSubmissionComments = void 0;

  class LrnappStudioSubmissionComments extends _litElement.LitElement {
    /**
     * LitElement constructable styles enhancement
     */
    static get styles() {
      return [_litElement.css`
        :host {
          display: block;
        }
      `];
    }

    render() {
      return _litElement.html`
      ${this._toArray(this.submission.relationships.comments.data).map(comment => _litElement.html`
          <lrnapp-studio-submission-comment
            .comment="${comment}"
          ></lrnapp-studio-submission-comment>
        `)}
    `;
    }

    static get tag() {
      return "lrnapp-studio-submission-comments";
    }

    static get properties() {
      return {
        submission: {
          type: Object
        }
      };
    }
    /**
     * Simple way to convert from object to array.
     */


    _toArray(obj) {
      if (obj == null) {
        return [];
      }

      return Object.keys(obj).map(function (key) {
        return obj[key];
      });
    }

  }

  _exports.LrnappStudioSubmissionComments = LrnappStudioSubmissionComments;
  window.customElements.define(LrnappStudioSubmissionComments.tag, LrnappStudioSubmissionComments);
});
