// @lrnwebcomponents/agency-theme/agency-theme.js
define(["exports", "require", "../../lit-element/lit-element.js"], function (_exports, _require, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.AgencyTheme = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  /**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `agency-theme`
   * @demo demo/index.html
   * @element agency-theme
   */
  class AgencyTheme extends _litElement.LitElement {
    //styles function
    static get styles() {
      return [(0, _litElement.css)`
        :host {
          display: block;
        }

        :host([hidden]) {
          display: none;
        }
      `];
    } // render function


    render() {
      return (0, _litElement.html)`
      <slot></slot>
    `;
    } // properties available to the custom element for data binding


    static get properties() {
      return { ...super.properties
      };
    }
    /**
     * Store the tag name to make it easier to obtain directly.
     * @notice function name must be here for tooling to operate correctly
     */


    static get tag() {
      return "agency-theme";
    }

    constructor() {
      super();
      new Promise((res, rej) => _require.default(["./lib/agency-theme-band.js"], res, rej));
      new Promise((res, rej) => _require.default(["./lib/agency-theme-spotlight.js"], res, rej));
    }

  }

  _exports.AgencyTheme = AgencyTheme;
  window.customElements.define(AgencyTheme.tag, AgencyTheme);
});
