// @lrnwebcomponents/tab-list/tab-list.js
define(["exports","require","../../lit-element/lit-element.js","../../@polymer/paper-tabs/paper-tabs.js"],(function(t,e,a,i){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.TabList=void 0,e=babelHelpers.interopRequireWildcard(e);class TabList extends a.LitElement{static get styles(){return[a.css`
        :host {
          display: block;
          margin: 0 auto;
          list-style: none;
          display: block;
          padding: 16px;
          border-bottom: 1px solid black;
        }
        paper-tabs {
          align-items: center;
          justify-items: center;
        }
        paper-tab a {
          text-decoration: none;
          flex: unset;
          height: unset;
          width: 100%;
          text-align: center;
        }
        paper-button {
          text-transform: unset;
          width: 100%;
          display: block;
          min-width: unset;
          margin: 0;
          text-overflow: ellipsis;
          overflow: hidden;
        }
        @media screen and (max-width: 600px) {
          paper-tab {
            display: block;
          }
        }
      `]}constructor(){super(),this.tabs=[],new Promise((t,a)=>e.default(["../../@polymer/paper-tabs/paper-tab.js"],t,a)),new Promise((t,a)=>e.default(["../../@polymer/paper-button/paper-button.js"],t,a))}render(){return a.html`
      <paper-tabs>
        ${this.tabs.map(t=>a.html`
            <paper-tab>
              <a
                target="_blank"
                href="${t.link}"
                tabindex="-1"
                rel="noopener noreferrer"
              >
                <paper-button raised>${t.label}</paper-button>
              </a>
            </paper-tab>
          `)}
      </paper-tabs>
    `}static get tag(){return"tab-list"}updated(t){t.forEach((t,e)=>{"tabs"==e&&this.dispatchEvent(new CustomEvent("tabs-changed",{bubbles:!0,cancelable:!0,composed:!0,detail:this[e]}))})}static get properties(){return{tabs:{type:Array}}}static get haxProperties(){return{canScale:!1,canPosition:!1,canEditSource:!1,gizmo:{title:"Tabs",description:"A list of links as tabs.",icon:"icons:tab",color:"grey",groups:["Presentation","Links"],handles:[],meta:{author:"ELMS:LN"}},settings:{quick:[],configure:[{property:"tabs",title:"Tabs",description:"Listing of tabs",inputMethod:"array",itemLabel:"label",properties:[{property:"link",title:"Link",description:"link to go to",inputMethod:"textfield",required:!0},{property:"label",title:"Label",description:"text to place on the tab",inputMethod:"textfield",required:!0}]}],advanced:[]}}}}t.TabList=TabList,window.customElements.define(TabList.tag,TabList)}));
