// @haxtheweb/haxcms-elements/lib/ui-components/utilities/site-available-themes.js
import{LitElement as e,html as t,css as a}from"../../../../../lit/index.js";import{DDDSuper as i}from"../../../../d-d-d/d-d-d.js";import{I18NMixin as r}from"../../../../i18n-manager/lib/I18NMixin.js";import"../../../../media-image/media-image.js";import"../../../../simple-icon/simple-icon.js";export class SiteAvailableThemes extends(i(r(e))){static get tag(){return"site-available-themes"}constructor(){super(),this.themes=[],this.loading=!0,this.error=null,this.viewMode="gallery",this.showDetails=!0,this.columns=3,this.currentTheme="",this.t=this.t||{},this.registerLocalization({context:this,localesPath:new URL("../../../../../../locales/site-available-themes/",import.meta.url).href+"/",locales:["en"]})}static get properties(){return{...super.properties,themes:{type:Array},loading:{type:Boolean},error:{type:String},viewMode:{type:String,attribute:"view-mode"},showDetails:{type:Boolean,attribute:"show-details"},columns:{type:Number},currentTheme:{type:String,attribute:"current-theme"}}}static get styles(){return[super.styles,a`
        :host {
          display: block;
          --theme-gallery-gap: var(--ddd-spacing-4);
          --theme-gallery-border-radius: var(--ddd-radius-sm);
          --theme-gallery-border: var(--ddd-border-sm);
        }

        .loading,
        .error {
          text-align: center;
          padding: var(--ddd-spacing-8);
          color: var(--ddd-theme-default-coalyGray);
        }

        .error {
          color: var(--ddd-theme-default-original87Pink);
          background: var(--ddd-theme-default-opportunityGreen);
          border-radius: var(--theme-gallery-border-radius);
          padding: var(--ddd-spacing-4);
        }

        .gallery-header {
          display: flex;
          justify-content: space-between;
          align-items: center;
          margin-bottom: var(--ddd-spacing-4);
          padding: var(--ddd-spacing-2) 0;
          border-bottom: var(--theme-gallery-border);
        }

        .gallery-title {
          margin: 0;
          color: var(--ddd-theme-default-coalyGray);
        }

        .view-toggle {
          display: flex;
          gap: var(--ddd-spacing-2);
        }

        .toggle-button {
          background: var(--ddd-theme-default-white);
          border: var(--theme-gallery-border);
          border-radius: var(--theme-gallery-border-radius);
          padding: var(--ddd-spacing-2) var(--ddd-spacing-4);
          cursor: pointer;
          transition: all 0.2s ease;
          display: flex;
          align-items: center;
          gap: var(--ddd-spacing-1);
          font-size: var(--ddd-font-size-xs);
        }

        .toggle-button:hover,
        .toggle-button.active {
          background: var(--ddd-theme-default-skyBlue);
          color: var(--ddd-theme-default-white);
        }

        .theme-count {
          color: var(--ddd-theme-default-coalyGray);
          font-size: var(--ddd-font-size-xs);
        }

        /* Gallery View */
        .gallery-grid {
          display: grid;
          grid-template-columns: repeat(var(--columns, 3), 1fr);
          gap: var(--theme-gallery-gap);
        }

        .theme-card {
          border: var(--theme-gallery-border);
          border-radius: var(--theme-gallery-border-radius);
          overflow: hidden;
          background: var(--ddd-theme-default-white);
          transition:
            transform 0.2s ease,
            box-shadow 0.2s ease;
          position: relative;
        }

        .theme-card:hover {
          transform: translateY(-2px);
          box-shadow: var(--ddd-boxShadow-md);
        }

        .theme-card.current {
          border-color: var(--ddd-theme-default-skyBlue);
          border-width: 2px;
        }

        .theme-thumbnail {
          width: 100%;
          height: 200px;
          position: relative;
        }

        .current-badge {
          position: absolute;
          top: var(--ddd-spacing-2);
          right: var(--ddd-spacing-2);
          background: var(--ddd-theme-default-skyBlue);
          color: var(--ddd-theme-default-white);
          padding: var(--ddd-spacing-1) var(--ddd-spacing-2);
          border-radius: var(--ddd-radius-xs);
          font-size: var(--ddd-font-size-3xs);
          z-index: 1;
        }

        .theme-info {
          padding: var(--ddd-spacing-4);
        }

        .theme-name {
          font-weight: var(--ddd-font-weight-bold);
          color: var(--ddd-theme-default-coalyGray);
          margin: 0 0 var(--ddd-spacing-2) 0;
          font-size: var(--ddd-font-size-sm);
        }

        .theme-description {
          font-size: var(--ddd-font-size-xs);
          color: var(--ddd-theme-default-slateGray);
          margin: 0 0 var(--ddd-spacing-2) 0;
        }

        .theme-element {
          font-size: var(--ddd-font-size-3xs);
          color: var(--ddd-theme-default-limestoneGray);
          font-family: var(--ddd-font-navigation);
          margin-bottom: var(--ddd-spacing-2);
        }

        .preview-button {
          background: var(--ddd-theme-default-skyBlue);
          color: var(--ddd-theme-default-white);
          border: none;
          border-radius: var(--theme-gallery-border-radius);
          padding: var(--ddd-spacing-2) var(--ddd-spacing-4);
          cursor: pointer;
          font-size: var(--ddd-font-size-xs);
          display: flex;
          align-items: center;
          gap: var(--ddd-spacing-1);
          transition: all 0.2s ease;
        }

        .preview-button:hover {
          background: var(--ddd-theme-default-navy);
        }

        .preview-button:disabled {
          background: var(--ddd-theme-default-limestoneGray);
          cursor: not-allowed;
        }

        /* Table View */
        .theme-table {
          width: 100%;
          border-collapse: collapse;
          border: var(--theme-gallery-border);
          border-radius: var(--theme-gallery-border-radius);
          overflow: hidden;
        }

        .theme-table th,
        .theme-table td {
          padding: var(--ddd-spacing-4);
          text-align: left;
          border-bottom: var(--theme-gallery-border);
        }

        .theme-table th {
          background: var(--ddd-theme-default-limestoneLight);
          font-weight: var(--ddd-font-weight-bold);
          color: var(--ddd-theme-default-coalyGray);
          font-size: var(--ddd-font-size-xs);
        }

        .theme-table tr:hover {
          background: var(--ddd-theme-default-limestoneLight);
        }

        .theme-table tr.current {
          background: var(--ddd-theme-default-skyMaxlight);
        }

        .table-thumbnail {
          width: 80px;
          height: 50px;
        }

        .table-name {
          font-weight: var(--ddd-font-weight-bold);
          color: var(--ddd-theme-default-coalyGray);
        }

        .table-element {
          font-family: var(--ddd-font-navigation);
          font-size: var(--ddd-font-size-xs);
          color: var(--ddd-theme-default-limestoneGray);
        }

        /* Responsive */
        @media (max-width: 768px) {
          .gallery-grid {
            grid-template-columns: repeat(2, 1fr);
          }

          .gallery-header {
            flex-direction: column;
            gap: var(--ddd-spacing-2);
            align-items: flex-start;
          }

          .theme-table {
            font-size: var(--ddd-font-size-xs);
          }

          .table-thumbnail {
            width: 60px;
            height: 40px;
          }
        }

        @media (max-width: 480px) {
          .gallery-grid {
            grid-template-columns: 1fr;
          }
        }
      `]}updated(e){super.updated(e),e.has("columns")&&this.style.setProperty("--columns",this.columns)}connectedCallback(){super.connectedCallback(),this.loadThemes(),this.detectCurrentTheme()}async loadThemes(){this.loading=!0,this.error=null;try{const e=new URL("../../themes.json",import.meta.url),t=await fetch(e);if(!t.ok)throw new Error(`HTTP error! status: ${t.status}`);const a=await t.json();this.themes=Object.entries(a).map((([e,t])=>({key:e,...t,thumbnailPath:t.thumbnail||`theme-screenshots/theme-${t.element}-thumb.jpg`,fullImagePath:t.screenshot||`theme-screenshots/theme-${t.element}.jpg`}))),this.loading=!1}catch(e){this.error=`Failed to load themes: ${e.message}`,this.loading=!1,console.error("Error loading themes:",e)}}detectCurrentTheme(){globalThis.HAXCMS&&globalThis.HAXCMS.instance&&globalThis.HAXCMS.instance.store&&globalThis.HAXCMS.instance.store.manifest&&globalThis.HAXCMS.instance.store.manifest.metadata&&globalThis.HAXCMS.instance.store.manifest.metadata.theme&&globalThis.HAXCMS.instance.store.manifest.metadata.theme.element&&(this.currentTheme=globalThis.HAXCMS.instance.store.manifest.metadata.theme.element)}async switchTheme(e){if(globalThis.HAXCMS&&globalThis.HAXCMS.setTheme)try{await globalThis.HAXCMS.setTheme(e),this.currentTheme=e,this.dispatchEvent(new CustomEvent("theme-changed",{bubbles:!0,composed:!0,detail:{themeElement:e}})),console.log(`Theme switched to: ${e}`)}catch(e){console.error("Failed to switch theme:",e)}else console.warn("HAXCMS theme switching not available")}setViewMode(e){this.viewMode=e}renderGalleryView(){return t`
      <div class="gallery-grid">
        ${this.themes.map((e=>t`
            <div
              class="theme-card ${e.element===this.currentTheme?"current":""}"
            >
              ${e.element===this.currentTheme?t` <div class="current-badge">Current</div> `:""}
              <div class="theme-thumbnail">
                <media-image
                  source="${e.fullImagePath}"
                  thumbnail="${e.thumbnailPath}"
                  alt="${e.name}"
                  accent-color="grey"
                  size="wide"
                ></media-image>
              </div>
              ${this.showDetails?t`
                    <div class="theme-info">
                      <h3 class="theme-name">${e.name}</h3>
                      ${e.description?t`
                            <p class="theme-description">
                              ${e.description}
                            </p>
                          `:""}
                      <div class="theme-element">&lt;${e.element}&gt;</div>
                      <button
                        class="preview-button"
                        ?disabled="${e.element===this.currentTheme}"
                        @click="${()=>this.switchTheme(e.element)}"
                      >
                        <simple-icon icon="visibility"></simple-icon>
                        ${e.element===this.currentTheme?"Current Theme":"Preview Theme"}
                      </button>
                    </div>
                  `:""}
            </div>
          `))}
      </div>
    `}renderTableView(){return t`
      <table class="theme-table">
        <thead>
          <tr>
            <th>Preview</th>
            <th>Name</th>
            <th>Element</th>
            ${this.showDetails?t`<th>Description</th>`:""}
            <th>Action</th>
          </tr>
        </thead>
        <tbody>
          ${this.themes.map((e=>t`
              <tr
                class="${e.element===this.currentTheme?"current":""}"
              >
                <td>
                  <div class="table-thumbnail">
                    <media-image
                      source="${e.fullImagePath}"
                      thumbnail="${e.thumbnailPath}"
                      alt="${e.name}"
                      accent-color="grey"
                      size="wide"
                    ></media-image>
                  </div>
                </td>
                <td>
                  <div class="table-name">${e.name}</div>
                  ${e.element===this.currentTheme?t`
                        <div
                          style="color: var(--ddd-theme-default-skyBlue); font-size: var(--ddd-font-size-3xs);"
                        >
                          Current
                        </div>
                      `:""}
                </td>
                <td>
                  <div class="table-element">&lt;${e.element}&gt;</div>
                </td>
                ${this.showDetails?t`<td>${e.description||""}</td>`:""}
                <td>
                  <button
                    class="preview-button"
                    ?disabled="${e.element===this.currentTheme}"
                    @click="${()=>this.switchTheme(e.element)}"
                  >
                    <simple-icon icon="visibility"></simple-icon>
                    ${e.element===this.currentTheme?"Current":"Preview"}
                  </button>
                </td>
              </tr>
            `))}
        </tbody>
      </table>
    `}render(){return this.loading?t`<div class="loading">Loading themes...</div>`:this.error?t`<div class="error">${this.error}</div>`:t`
      <div class="gallery-header">
        <div>
          <h2 class="gallery-title">Available HAXcms Themes</h2>
          <div class="theme-count">${this.themes.length} themes available</div>
        </div>
        <div class="view-toggle">
          <button
            class="toggle-button ${"gallery"===this.viewMode?"active":""}"
            @click="${()=>this.setViewMode("gallery")}"
          >
            <simple-icon icon="view-module"></simple-icon>
            Gallery
          </button>
          <button
            class="toggle-button ${"table"===this.viewMode?"active":""}"
            @click="${()=>this.setViewMode("table")}"
          >
            <simple-icon icon="view-list"></simple-icon>
            Table
          </button>
        </div>
      </div>

      ${"gallery"===this.viewMode?this.renderGalleryView():this.renderTableView()}
    `}static get haxProperties(){return{canScale:!1,canPosition:!1,canEditSource:!0,contentEditable:!1,gizmo:{title:"Available Themes",description:"Display available HAXcms themes with live preview switching",icon:"image:collections",color:"purple",tags:["Theme","Gallery","HAXcms","Developer","Documentation"],handles:[],meta:{author:"HAXTheWeb team",owner:"The Pennsylvania State University"}},settings:{configure:[{property:"viewMode",title:"View Mode",description:"Choose how to display the themes",inputMethod:"select",options:{gallery:"Gallery",table:"Table"}},{property:"showDetails",title:"Show Details",description:"Show theme descriptions and additional information",inputMethod:"boolean"},{property:"columns",title:"Columns (Gallery)",description:"Number of columns in gallery view",inputMethod:"number",min:1,max:6}]},demoSchema:[{tag:"site-available-themes",properties:{viewMode:"gallery",showDetails:!0,columns:3},content:""}]}}}globalThis.customElements.define(SiteAvailableThemes.tag,SiteAvailableThemes);
