// @lrnwebcomponents/lrn-vocab/lrn-vocab.js
define(["exports","require","../../lit-element/lit-element.js","../schema-behaviors/schema-behaviors.js","../simple-modal/simple-modal.js"],(function(e,t,i,r,o){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.LrnVocab=void 0,t=babelHelpers.interopRequireWildcard(t);
/**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class LrnVocab extends((0,r.SchemaBehaviors)(i.LitElement)){static get styles(){return[i.css`
        :host {
          display: inline-flex;
          --lrn-vocab-border: 1px dashed #ccc;
        }
        button {
          text-transform: none;
          min-width: unset;
          margin: 0;
          position: relative;
          top: 0px;
          border-radius: 0;
          border-bottom: var(--lrn-vocab-border);
          background: #f5f5f5;
          font-size: 1.1em;
          padding: 2px;
        }
        button:hover {
          background: #bbdefb;
          border-bottom: 1px dashed #2196f3;
        }
      `]}constructor(){super(),setTimeout(()=>{this.addEventListener("click",this.openDialog.bind(this))},0)}render(){return i.html`
      <button>${this.term}</button>
    `}static get tag(){return"lrn-vocab"}static get properties(){return{...super.properties,term:{type:String,reflect:!0}}}openDialog(e){let t=document.createElement("div");for(var i in this.children)this.children[i].cloneNode&&t.appendChild(this.children[i].cloneNode(!0));const r=new CustomEvent("simple-modal-show",{bubbles:!0,cancelable:!0,composed:!0,detail:{title:this.term,elements:{content:t},styles:{"--simple-modal-width":"50vw","--simple-modal-height":"50vh"},invokedBy:this}});this.dispatchEvent(r)}firstUpdated(e){super.firstUpdated&&super.firstUpdated(e),window.SimpleModal.requestAvailability()}static get haxProperties(){return new Promise((e,i)=>t.default(["../hax-iconset/hax-iconset.js"],e,i)),{canScale:!1,canPosition:!1,canEditSource:!1,gizmo:{title:"Vocab",description:"Vocabulary term",icon:"hax:vocab",color:"red",groups:["Vocab"],handles:[{type:"inline",text:"term"}],meta:{author:"ELMS:LN",inlineOnly:!0}},settings:{quick:[{property:"term",title:"Term",inputMethod:"textfield",icon:"editor:title",required:!0}],configure:[{property:"term",title:"Term",inputMethod:"textfield",icon:"editor:title",required:!0},{slot:"",title:"Contents",description:"The definitition to display when the term is clicked.",inputMethod:"code-editor",required:!0}],advanced:[]}}}}e.LrnVocab=LrnVocab,window.customElements.define(LrnVocab.tag,LrnVocab)}));
