// @lrnwebcomponents/simple-picker/lib/simple-picker-option.js
define(["exports","../../../@polymer/polymer/polymer-element.js","../../../@polymer/iron-icon/iron-icon.js","../../../@polymer/iron-icons/iron-icons.js","../../lrn-shared-styles/lrn-shared-styles.js"],function(_exports,_polymerElement,_ironIcon,_ironIcons,_lrnSharedStyles){"use strict";Object.defineProperty(_exports,"__esModule",{value:!0});_exports.SimplePickerOption=void 0;/**
 * Copyright 2018 The Pennsylvania State University
 * @license Apache-2.0, see License.md for full text.
 */ /**
 * `simple-picker-option`
 * `a simple picker for options, icons, etc.`
 *
 * @microcopy - language worth noting:
 *  -
 *
 * @customElement
 * @polymer
 * @see ../simple-picker.js
 * @see ../simple-color-picker-row.js
 */class SimplePickerOption extends _polymerElement.PolymerElement{// render function
static get template(){return _polymerElement.html`
      <style is="custom-style" include="lrn-shared-styles">
        :host {
          display: flex;
          align-items: center;
          justify-content: space-between;
          color: var(--simple-picker-color);
        }
        :host([hidden]) {
          display: none;
        }
        :host .label {
          padding: 2px 10px;
          line-height: 100%;
          @apply --simple-picker-option-label;
        }

        :host iron-icon {
          width: var(--simple-picker-option-size, 24px);
          min-height: var(--simple-picker-option-size, 24px);
          min-width: var(--simple-picker-option-size, 24px);
          line-height: var(--simple-picker-option-size, 24px);
          @apply --simple-picker-icon;
        }
      </style>
      <iron-icon
        aria-hidden="true"
        hidden$="[[_hideIcon(icon)]]"
        icon$="[[icon]]"
      ></iron-icon>
      <div id="label" class$="[[_getSrOnly(hideOptionLabels)]]">[[label]]</div>
    `}// properties available to the custom element for data binding
static get properties(){return{/**
       * Is the option active?
       */active:{name:"active",type:"Boolean",value:null,reflectToAttribute:!0},/**
       * The style of the option. (Required for accessibility.)
       */data:{name:"data",type:"Object",value:null},/**
       * If the option is hidden
       */hidden:{name:"hidden",type:"Boolean",value:!1,reflectToAttribute:!0},/**
       * Hide option labels? As color-picker or icon-picker, labels may be redundant.
       * This option would move the labels off-screen so that only screen-readers will have them.
       */hideOptionLabels:{name:"hideOptionLabels",type:"Boolean",value:!1,reflectToAttribute:!0},/**
       * Optional. If option is an iron icon, the iconset:name of the icon
       */icon:{name:"icon",type:"String",value:null,reflectToAttribute:!1},/**
       * The id of the option
       */id:{name:"order",type:"String",value:null,reflectToAttribute:!0},/**
       * The text of the option. (Required for accessibility.)
       */label:{name:"label",type:"String",value:null,reflectToAttribute:!0,observer:"_updateLabel"},/**
       * Is the option selected?
       */selected:{name:"selected",type:"Boolean",value:!1,reflectToAttribute:!0},/**
       * Renders html as title. (Good for titles with HTML in them.)
       */titleAsHtml:{name:"titleAsHtml",type:"Boolean",value:!1,reflectToAttribute:!0},/**
       * The value of the option.
       */value:{name:"label",type:"String",value:null,reflectToAttribute:!0}}}/**
   * Store the tag name to make it easier to obtain directly.
   * @notice function name must be here for tooling to operate correctly
   */static get tag(){return"simple-picker-option"}/**
   * If the option is not an iron-icon, hide the iron-icon.
   *
   * @param {string} icon the icon property
   * @returns {boolean} whether or not the iron iron should be hidden
   */_hideIcon(icon){return null===this.icon}/**
   * On keyboard focus, fires an event to the picker so that active descendant can be set.
   * @returns {void}
   */_handleFocus(){this.dispatchEvent(new CustomEvent("option-focus",{detail:this}))}/**
   * On mouse hover, fires an event to the picker so that active descendant can be set.
   * @returns {void}
   */_handleHover(){this.dispatchEvent(new CustomEvent("option-focus",{detail:this}))}/**
   * determines if a label should visible on screen
   *
   * @param {boolean} hideOptionLabels property
   * @returns {string} the sr-only (screenreader-only) class
   */_getSrOnly(hideOptionLabels){return!1===hideOptionLabels?"label":"label sr-only"}/**
   * updates the title
   * @returns {void}
   */_updateLabel(){let label=document.createElement("span");if(!1!==this.titleAsHtml){label.innerHTML=this.label;this.$.label.innerHTML="";this.$.label.appendChild(label)}}/**
   * Set event listeners
   * @returns {void}
   */ready(){super.ready();let root=this;this._updateLabel();this.addEventListener("focus",function(e){root._handleFocus()});this.addEventListener("mouseover",function(e){root._handleHover()})}/**
   * life cycle, element is afixed to the DOM
   * @returns {void}
   */connectedCallback(){super.connectedCallback()}/**
   * life cycle, element is removed from the DOM
   */ //disconnectedCallback() {}
}_exports.SimplePickerOption=SimplePickerOption;window.customElements.define(SimplePickerOption.tag,SimplePickerOption)});
