// @lrnwebcomponents/course-design/lib/worksheet-download.js
define(["exports", "../../../lit/index.js", "../../simple-icon/lib/simple-icons.js", "../../simple-icon/lib/simple-icon-lite.js"], function (_exports, _index, _simpleIcons, _simpleIconLite) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.WorksheetDownload = void 0;

  var _templateObject_60a4ab90942811ecb1937969798b231a, _templateObject2_60a4ab90942811ecb1937969798b231a;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var WorksheetDownload = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(WorksheetDownload, _LitElement);

    var _super = _createSuper(WorksheetDownload);

    function WorksheetDownload() {
      var _this;

      babelHelpers.classCallCheck(this, WorksheetDownload);
      _this = _super.call(this);
      _this.title = "";
      _this.link = "";
      return _this;
    }

    babelHelpers.createClass(WorksheetDownload, [{
      key: "haxHooks",
      value:
      /**
       * Implements haxHooks to tie into life-cycle if hax exists.
       */
      function haxHooks() {
        return {
          editModeChanged: "haxeditModeChanged",
          activeElementChanged: "haxactiveElementChanged"
        };
      }
      /**
       * double-check that we are set to inactivate click handlers
       * this is for when activated in a duplicate / adding new content state
       */

    }, {
      key: "haxactiveElementChanged",
      value: function haxactiveElementChanged(el, val) {
        if (val) {
          this._haxstate = val;
        }
      }
      /**
       * Set a flag to test if we should block link clicking on the entire card
       * otherwise when editing in hax you can't actually edit it bc its all clickable.
       * if editMode goes off this helps ensure we also become clickable again
       */

    }, {
      key: "haxeditModeChanged",
      value: function haxeditModeChanged(val) {
        this._haxstate = val;
      }
      /**
       * special support for HAX since the whole card is selectable
       */

    }, {
      key: "_clickLink",
      value: function _clickLink(e) {
        if (this._haxstate) {
          // do not do default
          e.preventDefault();
          e.stopPropagation();
          e.stopImmediatePropagation();
        }
      }
    }, {
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_60a4ab90942811ecb1937969798b231a || (_templateObject_60a4ab90942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <div id=\"button_wrap\">\n        <a\n          tabindex=\"-1\"\n          href=\"", "\"\n          target=\"_blank\"\n          download\n          rel=\"noopener noreferrer\"\n          @click=\"", "\"\n        >\n          <button>\n            <simple-icon-lite icon=\"icons:file-download\"></simple-icon-lite\n            >", "\n          </button>\n        </a>\n      </div>\n    "])), this.link, this._clickLink, this.title);
      }
    }], [{
      key: "properties",
      get: function get() {
        return {
          title: {
            type: String
          },
          link: {
            type: String
          }
        };
      }
    }, {
      key: "tag",
      get: function get() {
        return "worksheet-download";
      }
    }, {
      key: "haxProperties",
      get: function get() {
        return {
          canScale: true,
          canPosition: true,
          canEditSource: false,
          gizmo: {
            title: "Worksheet Download",
            description: "A button for displaying files available for download.",
            icon: "icons:file-download",
            color: "blue",
            groups: ["education", "link"],
            meta: {
              author: "LRNWebComponents"
            }
          },
          settings: {
            configure: [{
              property: "title",
              title: "Title",
              description: "The title of the download.",
              inputMethod: "textfield",
              icon: "editor:title"
            }, {
              property: "link",
              title: "Link",
              description: "The link for the download.",
              inputMethod: "textfield",
              icon: "editor:insert-link"
            }],
            advanced: []
          },
          demoSchema: [{
            tag: "worksheet-download",
            properties: {
              title: "Download worksheet"
            },
            content: ""
          }]
        };
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _index.css)(_templateObject2_60a4ab90942811ecb1937969798b231a || (_templateObject2_60a4ab90942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n\n        a {\n          text-decoration: none;\n          color: #0c7cd5;\n          display: block;\n        }\n\n        button {\n          text-transform: none;\n          border: solid 2px #dcdcdc;\n          width: 100%;\n          margin: 0 auto 0;\n          min-height: 48px;\n          text-align: center;\n        }\n\n        button:active,\n        button:focus,\n        button:hover {\n          cursor: pointer;\n          background-color: #0c7cd5;\n          color: #fff;\n        }\n\n        simple-icon {\n          margin-right: 5px;\n        }\n      "])))];
      }
    }]);
    return WorksheetDownload;
  }(_index.LitElement);

  _exports.WorksheetDownload = WorksheetDownload;
  customElements.define(WorksheetDownload.tag, WorksheetDownload);
});
