diff --git a/admin_menu_dropdown.module b/admin_menu_dropdown.module
index b4121f8..553beac 100644
--- a/admin_menu_dropdown.module
+++ b/admin_menu_dropdown.module
@@ -1,4 +1,5 @@
 <?php
+// $Id$
 /**
  * @file
  * Admin Menu Hider, to keep the admin menu out of the way.
@@ -7,18 +8,20 @@
  */
 
 /**
- * Implements hook_init().
+ * Implements hook_page_build().
  */
-function admin_menu_dropdown_init() {
-  $path = drupal_get_path('module', 'admin_menu_dropdown');
+function admin_menu_dropdown_page_build(&$page) {
+  if (!user_access('access administration menu') || admin_menu_suppress(FALSE)) {
+    return;
+  }
   $settings = array(
-    'key'     => variable_get('admin_menu_dropdown_key', '`'),
-    'default' => variable_get('admin_menu_dropdown_default', 1),
+    'key'    => variable_get('admin_menu_dropdown_key', '`'),
+    'onload' => variable_get('admin_menu_dropdown_default', 1),
   );
-
-  drupal_add_js($path . '/admin_menu_dropdown.js', array('defer' => TRUE));
-  drupal_add_js(array('admin_menu_dropdown' => $settings), array('type' => 'setting', 'scope' => JS_DEFAULT));
-  drupal_add_css($path . '/admin_menu_dropdown.css');
+  drupal_add_js(array('admin_menu_dropdown' => $settings), array('type' => 'setting'));
+  $path = drupal_get_path('module', 'admin_menu_dropdown');
+  drupal_add_js($path . '/admin_menu_dropdown.js', array('every_page' => TRUE, 'weight' => -999));
+  drupal_add_css($path . '/admin_menu_dropdown.css', array('every_page' => TRUE, 'weight' => -998));
 }
 
 /**
@@ -45,5 +48,4 @@ function admin_menu_dropdown_form_admin_menu_theme_settings_alter(&$form, &$form
     '#description' => t('This key will show/hide the admin menu.'),
     '#wysiwyg' => FALSE,
   );
-
 }
