diff --git a/core/dslmcode/shared/drupal-7.x/modules/ulmus/admin_menu_dropdown/admin_menu_dropdown.js b/core/dslmcode/shared/drupal-7.x/modules/ulmus/admin_menu_dropdown/admin_menu_dropdown.js
index f270214..fee3906 100644
--- a/core/dslmcode/shared/drupal-7.x/modules/ulmus/admin_menu_dropdown/admin_menu_dropdown.js
+++ b/core/dslmcode/shared/drupal-7.x/modules/ulmus/admin_menu_dropdown/admin_menu_dropdown.js
@@ -16,7 +16,8 @@
         // Attach to keypress event.
         $(document).keypress(function(e) {
           var unicode = e.keyCode? e.keyCode : e.charCode;
-          if (String.fromCharCode(unicode) == settings.admin_menu_dropdown.key) {
+          if (String.fromCharCode(unicode) == settings.admin_menu_dropdown.key && !($(document.activeElement).attr("type") == "text" || $(document.activeElement)[0].tagName == "TEXTAREA")) {
+            console.log($(document.activeElement));
             $('#admin-menu').slideToggle('fast');
             // TODO: Maybe animate the margin change so its not so jumpy?
             $('body').toggleClass('adm_menu_hidden');
