// @lrnwebcomponents/rich-text-editor/lib/buttons/rich-text-editor-button-styles.js
define(["exports", "../../../../lit-element/lit-element.js"], function (_exports, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.RichTextEditorButtonStyles = _exports.RichTextEditorButtonStylesEl = void 0;

  /**
   * Copyright 2018 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
  const RichTextEditorButtonStyles = function (SuperClass) {
    return class extends SuperClass {
      static get tag() {
        return "rich-text-editor-button-styles";
      }

      static get styles() {
        return [(0, _litElement.css)`
          :host {
            --rich-text-editor-button-min-width: 24px;
            --rich-text-editor-button-height: 24px;
            --rich-text-editor-button-margin: 3px;
            --rich-text-editor-button-padding: 0;
          }
          :host([hidden]) {
            display: none;
          }
          .offscreen {
            position: absolute;
            left: -999999px;
            top: 0;
            height: 0;
            width: 0;
            overflow: hidden;
          }
          :host(rich-text-editor-emoji-picker),
          :host(rich-text-editor-symbol-picker) {
            --simple-picker-option-label-padding: 2px;
          }
          simple-tooltip {
            z-index: 2;
          }
          .rtebutton {
            text-transform: unset;
            transition: all 0.5s;
            color: var(--rich-text-editor-button-color);
            border-color: var(--rich-text-editor-border-color);
          }
          .rtebutton[toggled] {
            color: var(--rich-text-editor-button-toggled-color);
            background-color: var(--rich-text-editor-button-toggled-bg);
          }
          .rtebutton:focus,
          .rtebutton:hover {
            color: var(--rich-text-editor-button-hover-color);
            background-color: var(--rich-text-editor-button-hover-bg);
          }
          :host([disabled]) .rtebutton {
            cursor: not-allowed;
            color: var(--rich-text-editor-button-disabled-color);
            background-color: var(--rich-text-editor-button-disabled-bg);
          }
          .rtebutton #icon:not([icon]) {
            display: none;
          }
        `];
      }

    };
  };
  /**
   * `rich-text-editor-button-styles`
   * @element rich-text-editor-button-styles
   * `a shared set of button styles for rich-text-editor`
   */


  _exports.RichTextEditorButtonStyles = RichTextEditorButtonStyles;

  class RichTextEditorButtonStylesEl extends RichTextEditorButtonStyles(_litElement.LitElement) {}

  _exports.RichTextEditorButtonStylesEl = RichTextEditorButtonStylesEl;
  window.customElements.define(RichTextEditorButtonStylesEl.tag, RichTextEditorButtonStylesEl);
});
