// @lrnwebcomponents/rich-text-editor/lib/buttons/rich-text-editor-more-button.js
define(["exports", "../../../../lit-element/lit-element.js", "./rich-text-editor-button.js"], function (_exports, _litElement, _richTextEditorButton) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.RichTextEditorMoreButton = void 0;

  /**
   * Copyright 2019 Penn State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `rich-text-editor-more-button`
   * a more button to toggle collapsed buttons in the rich text editor
   *
   * @element rich-text-editor-more-button
   * @demo ./demo/buttons.html
   */
  class RichTextEditorMoreButton extends (0, _richTextEditorButton.RichTextEditorButtonBehaviors)(_litElement.LitElement) {
    /**
     * Store the tag name to make it easier to obtain directly.
     */
    static get tag() {
      return "rich-text-editor-more-button";
    } // render function for template


    render() {
      return super.render();
    }

    constructor() {
      super();
      this.icon = "more-vert";
      this.collapseMax = "xs";
      this.toggled = false;
      this.label = "More buttons";
      this.labelToggled = "Fewer buttons";
    } // properties available to the custom element for data binding


    static get properties() {
      return { ...super.properties,

        /**
         * Can this button toggle?
         */
        toggled: {
          attribute: "toggled",
          type: Boolean,
          reflect: true
        },

        /**
         * The maximum size where all of the buttons display
         */
        collapseMax: {
          attribute: "collapse-max",
          type: String,
          reflect: true
        }
      };
    }
    /**
     * whether button is toggled
     *
     * @readonly
     * @memberof RichTextEditorButton
     */


    get isToggled() {
      return this.toggled;
    }
    /**
     * Fires a button tap event
     */


    _buttonTap() {
      return true;
    }

  }

  _exports.RichTextEditorMoreButton = RichTextEditorMoreButton;
  window.customElements.define(RichTextEditorMoreButton.tag, RichTextEditorMoreButton);
});
