// @lrnwebcomponents/rich-text-editor/lib/buttons/rich-text-editor-more-button.js
define(["exports", "../../../../@polymer/polymer/polymer-element.js", "./rich-text-editor-button.js"], function (_exports, _polymerElement, _richTextEditorButton) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.RichTextEditorMoreButton = void 0;

  /**
   * Copyright 2019 Penn State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `rich-text-editor-more-button`
   * `a more button to toggle collapsed buttons in the rich text editor`
   *
   * @microcopy - language worth noting:
   *  -
   *
  
   * @polymer
   */
  class RichTextEditorMoreButton extends _richTextEditorButton.RichTextEditorButton {
    constructor() {
      super();
      this.label = "More buttons";
      this.labelToggled = "Fewer buttons";
    } // properties available to the custom element for data binding


    static get properties() {
      return {
        /**
         * Can this button toggle?
         */
        toggled: {
          name: "toggled",
          type: Boolean,
          value: false,
          reflectToAttribute: true
        },

        /**
         * The maximum size where all of the buttons display
         */
        collapseMax: {
          name: "collapseMax",
          type: String,
          value: "xs",
          reflectToAttribute: true
        }
      };
    }
    /**
     * Store the tag name to make it easier to obtain directly.
     */


    static get tag() {
      return "rich-text-editor-more-button";
    }
    /**
     * Fires a button tap event
     */


    _buttonTap(e) {
      this.dispatchEvent(new CustomEvent("rich-text-more-button-tap", {
        detail: this
      }));
    }

  }

  _exports.RichTextEditorMoreButton = RichTextEditorMoreButton;
  window.customElements.define(RichTextEditorMoreButton.tag, RichTextEditorMoreButton);
});
