// @lrnwebcomponents/rich-text-editor/lib/buttons/rich-text-editor-underline.js
define(["exports", "../../../../@polymer/polymer/polymer-element.js", "./rich-text-editor-prompt-button.js"], function (_exports, _polymerElement, _richTextEditorPromptButton) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.RichTextEditorUnderline = void 0;

  /**
   * Copyright 2019 Penn State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `rich-text-editor-underline`
   * `a button for rich text editor (custom buttons can extend this)`
   *
   * @microcopy - language worth noting:
   *  -
   *
  
   * @polymer
   */
  class RichTextEditorUnderline extends _richTextEditorPromptButton.RichTextEditorPromptButton {
    constructor() {
      super();
      this.fields = [{
        property: "",
        title: "Text",
        description: "The link text",
        inputMethod: "textfield"
      }, {
        property: "confirm",
        title: "Underline despite usability issues",
        description: "Underlines can be confused with links on the web. To prevent usability issues, we recommend using italics instead.",
        inputMethod: "boolean"
      }];
      this.tag = "u";
      this.icon = "editor:format-underlined";
      this.label = "Underline (not recommended)";
      this.toggles = true;
      this.command = "underline";
      this.shortcutKeys = "ctrl+u";
      this.value = {
        tag: false
      };
    } // properties available to the custom element for data binding


    static get properties() {
      return {};
    }
    /**
     * Store the tag name to make it easier to obtain directly.
     */


    static get tag() {
      return "rich-text-editor-underline";
    }
    /**
     * a <u> tag is only needed if there is text content
     * and the tag usability warning is confirmed
     * @param {object} value the prompt values
     * @returns {boolean} if the tag is needed for the element
     */


    _getTagNeeded(value) {
      return value && this.getCleanValue("") && this.getCleanValue("") !== "" && this.getCleanValue("confirm") && this.getCleanValue("confirm");
    }

  }

  _exports.RichTextEditorUnderline = RichTextEditorUnderline;
  window.customElements.define(RichTextEditorUnderline.tag, RichTextEditorUnderline);
});
