// @lrnwebcomponents/haxcms-elements/lib/core/haxcms-site-insights.js
define(["exports", "./haxcms-site-store.js", "../../../../mobx/dist/mobx.esm.js", "../../../../lit/index.js", "./utils/HAXCMSI18NMixin.js", "../../../micro-frontend-registry/micro-frontend-registry.js", "../../../micro-frontend-registry/lib/microServices.js", "../../../a11y-tabs/a11y-tabs.js", "../../../accent-card/accent-card.js", "../../../retro-card/retro-card.js", "../../../simple-img/simple-img.js", "../../../simple-fields/simple-fields.js", "../../../lesson-overview/lib/lesson-highlight.js", "../../../../@github/time-elements/dist/relative-time-element.js", "../../../iframe-loader/lib/loading-indicator.js"], function (_exports, _haxcmsSiteStore, _mobxEsm, _index, _HAXCMSI18NMixin2, _microFrontendRegistry, _microServices, _a11yTabs, _accentCard, _retroCard, _simpleImg, _simpleFields, _lessonHighlight, _relativeTimeElement, _loadingIndicator) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HAXCMSShareDialog = void 0;

  var _templateObject_b07101d0809211edaa2833647dc78c6c, _templateObject2_b07101d0809211edaa2833647dc78c6c, _templateObject3_b07101d0809211edaa2833647dc78c6c, _templateObject4_b07101d0809211edaa2833647dc78c6c, _templateObject5_b07101d0809211edaa2833647dc78c6c, _templateObject6_b07101d0809211edaa2833647dc78c6c, _templateObject7_b07101d0809211edaa2833647dc78c6c, _templateObject8_b07101d0809211edaa2833647dc78c6c, _templateObject9_b07101d0809211edaa2833647dc78c6c, _templateObject10_b07101d0809211edaa2833647dc78c6c, _templateObject11_b07101d0809211edaa2833647dc78c6c, _templateObject12_b07101d0809211edaa2833647dc78c6c, _templateObject13_b07101d0809211edaa2833647dc78c6c, _templateObject14_b07101d0809211edaa2833647dc78c6c, _templateObject15_b07101d0809211edaa2833647dc78c6c, _templateObject16_b07101d0809211edaa2833647dc78c6c, _templateObject17_b07101d0809211edaa2833647dc78c6c, _templateObject18_b07101d0809211edaa2833647dc78c6c, _templateObject19_b07101d0809211edaa2833647dc78c6c, _templateObject20_b07101d0809211edaa2833647dc78c6c, _templateObject21_b07101d0809211edaa2833647dc78c6c, _templateObject22_b07101d0809211edaa2833647dc78c6c, _templateObject23_b07101d0809211edaa2833647dc78c6c, _templateObject24_b07101d0809211edaa2833647dc78c6c, _templateObject25_b07101d0809211edaa2833647dc78c6c, _templateObject26_b07101d0809211edaa2833647dc78c6c, _templateObject27_b07101d0809211edaa2833647dc78c6c, _templateObject28_b07101d0809211edaa2833647dc78c6c, _templateObject29_b07101d0809211edaa2833647dc78c6c, _templateObject30_b07101d0809211edaa2833647dc78c6c, _templateObject31_b07101d0809211edaa2833647dc78c6c, _templateObject32_b07101d0809211edaa2833647dc78c6c, _templateObject33_b07101d0809211edaa2833647dc78c6c, _templateObject34_b07101d0809211edaa2833647dc78c6c, _templateObject35_b07101d0809211edaa2833647dc78c6c, _templateObject36_b07101d0809211edaa2833647dc78c6c, _templateObject37_b07101d0809211edaa2833647dc78c6c, _templateObject38_b07101d0809211edaa2833647dc78c6c, _templateObject39_b07101d0809211edaa2833647dc78c6c, _templateObject40_b07101d0809211edaa2833647dc78c6c, _templateObject41_b07101d0809211edaa2833647dc78c6c, _templateObject42_b07101d0809211edaa2833647dc78c6c, _templateObject43_b07101d0809211edaa2833647dc78c6c, _templateObject44_b07101d0809211edaa2833647dc78c6c, _templateObject45_b07101d0809211edaa2833647dc78c6c, _templateObject46_b07101d0809211edaa2833647dc78c6c, _templateObject47_b07101d0809211edaa2833647dc78c6c, _templateObject48_b07101d0809211edaa2833647dc78c6c, _templateObject49_b07101d0809211edaa2833647dc78c6c, _templateObject50_b07101d0809211edaa2833647dc78c6c, _templateObject51_b07101d0809211edaa2833647dc78c6c, _templateObject52_b07101d0809211edaa2833647dc78c6c, _templateObject53_b07101d0809211edaa2833647dc78c6c, _templateObject54_b07101d0809211edaa2833647dc78c6c, _templateObject55_b07101d0809211edaa2833647dc78c6c, _templateObject56_b07101d0809211edaa2833647dc78c6c, _templateObject57_b07101d0809211edaa2833647dc78c6c, _templateObject58_b07101d0809211edaa2833647dc78c6c, _templateObject59_b07101d0809211edaa2833647dc78c6c;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  (0, _microServices.enableServices)(["haxcms", "core"]);
  /**
   * `haxcms-outline-editor-dialog`
   * `Dialog for presenting an editable outline`
   *
   * @demo demo/index.html
   *
   * @microcopy - the mental model for this element
   */

  var HAXCMSShareDialog = /*#__PURE__*/function (_HAXCMSI18NMixin) {
    babelHelpers.inherits(HAXCMSShareDialog, _HAXCMSI18NMixin);

    var _super = _createSuper(HAXCMSShareDialog);

    function HAXCMSShareDialog() {
      var _this;

      babelHelpers.classCallCheck(this, HAXCMSShareDialog);
      _this = _super.call(this);
      _this.base = "";
      _this.filters = {};
      _this.linkResponseData = {};
      _this.t = _this.t || {};
      _this.t = _objectSpread(_objectSpread({}, _this.t), {}, {
        pageToProvideInsightsAbout: "Page to provide insights about",
        noLinksInSelectedPages: "No links in selected pages",
        noMediaInSelectedPages: "No media in selected pages",
        recentUpdates: "Recent updates",
        created: "Created",
        lastUpdated: "Last updated",
        updateInsights: "Update insights",
        onPage: "on page",
        learningObjectives: "learning objectives",
        specialElements: "Special elements",
        headings: "Headings",
        externalLinks: "External links",
        pages: "Pages",
        videos: "videos",
        placeholders: "Placeholders",
        video: "Video",
        audio: "Audio",
        selfChecks: "Self checks",
        openInNewTab: "Open in new tab",
        links: "Links",
        images: "Images",
        dataTables: "Data tables",
        ofReading: "of reading",
        gradeReadingLevel: "grade reading level",
        words: "Words",
        longWords: "long words",
        linkReport: "Link report",
        loading: "Loading",
        fullSite: "Full site",
        insights: "Insights",
        linkChecker: "Link checker",
        mediaBrowser: "Media browser",
        contentBrowser: "Content browser",
        hour: "hour",
        hours: "hours",
        minute: "minute",
        minutes: "minutes"
      });
      _this.data = {
        readability: {},
        updatedItems: []
      };
      _this.activeTab = "insights";
      _this.loading = false;
      return _this;
    }

    babelHelpers.createClass(HAXCMSShareDialog, [{
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSShareDialog.prototype), "updated", this).call(this, changedProperties);
        changedProperties.forEach(function (oldValue, propName) {
          if (propName === "activeTab" && _this2[propName]) {
            _this2.refreshData();
          }
        });
      }
    }, {
      key: "_insightResponse",
      value: function _insightResponse(data) {
        var _this3 = this;

        this.loading = false;
        this.data = data.data; // for filtering purposes

        this._originalData = JSON.parse(JSON.stringify(data.data));
        this.filters = {};
        setTimeout(function () {
          switch (_this3.activeTab) {
            case "linkchecker":
              _this3.filters = {
                links: "all"
              };
              _this3.shadowRoot.querySelector("#schema").fields = [{
                property: "links",
                title: "Link status",
                inputMethod: "select",
                options: {
                  all: "All",
                  error: "Error only",
                  ok: "OK only"
                }
              }];
              break;

            case "contentbrowser":
              _this3.filters = {
                sort: "title",
                title: "",
                hasVideo: false,
                hasH5P: false,
                hasLinks: false,
                hasImages: false,
                hasPlaceholders: false,
                hasSiteRemoteContent: false
              };
              _this3.shadowRoot.querySelector("#schema").value = _this3.filters;
              _this3.shadowRoot.querySelector("#schema").fields = [{
                property: "sort",
                title: "Sort by",
                inputMethod: "select",
                options: {
                  title: "Title",
                  lastUpdated: "Last updated",
                  contentLength: "Content length"
                }
              }, {
                property: "title",
                title: "Title",
                inputMethod: "textfield"
              }, {
                property: "hasVideo",
                title: "Video",
                description: "Includes video",
                inputMethod: "boolean"
              }, {
                property: "hasH5P",
                title: "H5P",
                description: "Includes h5p",
                inputMethod: "boolean"
              }, {
                property: "hasPlaceholders",
                title: "Placeholders",
                description: "Includes placeholders",
                inputMethod: "boolean"
              }, {
                property: "hasSiteRemoteContent",
                title: "Remote content",
                description: "Includes remote content",
                inputMethod: "boolean"
              }, {
                property: "hasLinks",
                title: "Links",
                description: "Includes external links",
                inputMethod: "boolean"
              }, {
                property: "hasImages",
                title: "Images",
                description: "Includes images",
                inputMethod: "boolean"
              }];
              break;

            case "mediabrowser":
              _this3.filters = {
                title: "",
                type: "all",
                location: "all"
              };
              _this3.shadowRoot.querySelector("#schema").fields = [{
                property: "title",
                title: "Title",
                inputMethod: "textfield"
              }, {
                property: "status",
                title: "Status",
                inputMethod: "select",
                options: {
                  all: "All",
                  info: "No issues",
                  warning: "Warnings",
                  error: "Errors"
                }
              }, {
                property: "type",
                title: "Media type",
                inputMethod: "select",
                options: {
                  all: "All",
                  audio: _this3.t.audio,
                  video: _this3.t.video,
                  image: _this3.t.images,
                  h5p: "H5P",
                  document: "Document"
                }
              }, {
                property: "location",
                title: "Location",
                inputMethod: "select",
                options: {
                  all: "All",
                  external: "External only",
                  internal: "Internal only"
                }
              }];
              _this3.shadowRoot.querySelector("#schema").values = _this3.filters;
              break;
          }
        }, 0);
      }
    }, {
      key: "refreshData",
      value: function refreshData() {
        var base = this.base;

        if (base == "" && document.querySelector("base")) {
          base = document.querySelector("base").href;
        }

        var site = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.manifest);
        var params = {
          type: "site",
          site: {
            file: base + "site.json",
            id: site.id,
            title: site.title,
            author: site.author,
            description: site.description,
            license: site.license,
            metadata: site.metadata,
            items: site.items
          },
          activeId: this.shadowRoot.querySelector("#selector").value,
          link: base
        };
        this.loading = true;

        switch (this.activeTab) {
          case "insights":
            _microFrontendRegistry.MicroFrontendRegistry.call("@haxcms/insights", params, this._insightResponse.bind(this), this);

            break;

          case "linkchecker":
            _microFrontendRegistry.MicroFrontendRegistry.call("@haxcms/linkChecker", params, this._insightResponse.bind(this), this);

            break;

          case "contentbrowser":
            _microFrontendRegistry.MicroFrontendRegistry.call("@haxcms/contentBrowser", params, this._insightResponse.bind(this), this);

            break;

          case "mediabrowser":
            _microFrontendRegistry.MicroFrontendRegistry.call("@haxcms/mediaBrowser", params, this._insightResponse.bind(this), this);

            break;
          // bad selection

          default:
            this.loading = false;
            break;
        }
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       */

    }, {
      key: "getReadingTime",
      value: function getReadingTime(value) {
        var readVal = [];
        var hours = Math.floor(value / 60);
        var minutes = Math.floor(value - hours * 60); // handle hours of reading

        if (hours === 1) {
          readVal.push("".concat(hours, " ").concat(this.t.hour));
        } else if (hours === 0) {// do nothing for 0
        } else {
          readVal.push("".concat(hours, " ").concat(this.t.hours));
        } // minutes


        if (minutes === 1) {
          readVal.push("".concat(minutes, " ").concat(this.t.minute));
        } else if (minutes === 0) {// do nothing for 0
        } else {
          readVal.push("".concat(minutes, " ").concat(this.t.minutes));
        }

        return readVal.join(", ");
      }
    }, {
      key: "closeModal",
      value: function closeModal() {
        window.dispatchEvent(new CustomEvent("simple-modal-hide"));
      }
    }, {
      key: "activeChanged",
      value: function activeChanged(e) {
        this.activeTab = e.detail.activeTab;
      }
    }, {
      key: "linkFormChanged",
      value: function linkFormChanged(e) {
        this.filters.links = e.detail.value.links; // @todo refactor

        if (this.filters.links === "error") {
          this.hideResults = "link-status-true";
        } else if (this.filters.links === "ok") {
          this.hideResults = "link-status-false";
        }
      }
    }, {
      key: "contentBrowserFormChanged",
      value: function contentBrowserFormChanged(e) {
        var _this4 = this;

        this.data = JSON.parse(JSON.stringify(this._originalData));
        var val = e.detail.value;
        this.filters = {
          title: val.title ? val.title : "",
          sort: val.sort,
          hasVideo: val.hasVideo,
          hasH5P: val.hasH5P,
          hasPlaceholders: val.hasPlaceholders,
          hasSiteRemoteContent: val.hasSiteRemoteContent,
          hasLinks: val.hasLinks,
          hasImages: val.hasImages
        };

        if (this.data.contentData) {
          this.data.contentData = this.data.contentData.filter(function (item) {
            if (_this4.filters.hasVideo === true && item.videos === 0) {
              return false;
            }

            if (_this4.filters.hasH5P === true && item.h5p === 0) {
              return false;
            }

            if (_this4.filters.hasPlaceholders === true && item.placeholders === 0) {
              return false;
            }

            if (_this4.filters.hasSiteRemoteContent === true && item.siteremotecontent === 0) {
              return false;
            }

            if (_this4.filters.hasLinks === true && item.links === 0) {
              return false;
            }

            if (_this4.filters.hasImages === true && item.images === 0) {
              return false;
            } // no filtering, skip


            if (_this4.filters.title == "") {
              return true;
            } else if (!item.title.toLowerCase().includes(_this4.filters.title.toLowerCase())) {
              return false;
            }

            return true;
          });
          this.data.contentData.sort(function (a, b) {
            switch (_this4.filters.sort) {
              case "title":
                return a.title.localeCompare(b.title);
                break;

              case "contentLength":
                return b.readTime - a.readTime;
                break;

              case "lastUpdated":
                return new Date(b.updated) - new Date(a.updated);
                break;
            }
          });
        }
      }
    }, {
      key: "mediaBrowserFormChanged",
      value: function mediaBrowserFormChanged(e) {
        var _this5 = this;

        this.data = JSON.parse(JSON.stringify(this._originalData));
        var val = e.detail.value;
        this.filters = {
          title: val.title ? val.title : "",
          type: val.type ? val.type : "all",
          status: val.status ? val.status : "all",
          location: val.location ? val.location : "all"
        };

        if (this.data.mediaData) {
          this.data.mediaData = this.data.mediaData.filter(function (item) {
            if (_this5.filters.location != "all") {
              if (_this5.filters.location != item.locType) {
                return false;
              } else {
                return true;
              }
            }

            return true;
          });
          this.data.mediaData = this.data.mediaData.filter(function (item) {
            if (_this5.filters.status != "all") {
              if (_this5.filters.status != item.status) {
                return false;
              } else {
                return true;
              }
            }

            return true;
          });
          this.data.mediaData = this.data.mediaData.filter(function (item) {
            if (_this5.filters.type != "all") {
              if (_this5.filters.type != item.type) {
                return false;
              } else {
                return true;
              }
            }

            return true;
          });
          this.data.mediaData = this.data.mediaData.filter(function (item) {
            // no filtering, skip
            if (!_this5.filters.title || _this5.filters.title == "") {
              return true;
            } else if (_this5.filters.title.toLowerCase && !item.name.toLowerCase().includes(_this5.filters.title.toLowerCase())) {
              return false;
            }

            return true;
          });
        }
      } // render function

    }, {
      key: "render",
      value: function render() {
        var _this6 = this;

        var data = this.data;
        var base = this.base;

        if (base == "" && document.querySelector("base")) {
          base = document.querySelector("base").href;
        }

        return (0, _index.html)(_templateObject_b07101d0809211edaa2833647dc78c6c || (_templateObject_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n      ", "\n      <button\n        @click=\"", "\"\n        icon=\"refresh\"\n        ?disabled=\"", "\"\n      >\n        ", "\n      </button>\n      <a11y-tabs\n        id=\"tabs\"\n        full-width\n        @a11y-tabs-active-changed=\"", "\"\n        sticky\n      >\n        <a11y-tab\n          id=\"insights\"\n          icon=\"hax:clipboard-pulse\"\n          label=\"", "\"\n        >\n          ", "\n        </a11y-tab>\n        <a11y-tab\n          id=\"linkchecker\"\n          icon=\"icons:link\"\n          label=\"", "\"\n        >\n          ", "\n        </a11y-tab>\n        <a11y-tab\n          id=\"contentbrowser\"\n          icon=\"icons:view-module\"\n          label=\"", "\"\n        >\n          ", "\n        </a11y-tab>\n        <a11y-tab\n          id=\"mediabrowser\"\n          icon=\"icons:perm-media\"\n          label=\"", "\"\n        >\n          ", "\n        </a11y-tab>\n      </a11y-tabs>\n    "])), this.pageSelector(), this.refreshData, this.loading, this.t.updateInsights, this.activeChanged, this.t.insights, this.activeTab === "insights" ? (0, _index.html)(_templateObject2_b07101d0809211edaa2833647dc78c6c || (_templateObject2_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n                <loading-indicator\n                  full\n                  ?loading=\"", "\"\n                ></loading-indicator>\n                ", "\n              "])), this.loading, this.loading ? (0, _index.html)(_templateObject3_b07101d0809211edaa2833647dc78c6c || (_templateObject3_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<p>", " ", "..</p>"])), this.t.loading, this.t.insights) : (0, _index.html)(_templateObject4_b07101d0809211edaa2833647dc78c6c || (_templateObject4_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n        <div class=\"insights\">\n          <h2>", " ", "</h2>\n          <ul>\n            <li>\n            <lesson-highlight icon=\"editor:insert-drive-file\">\n              <p slot=\"title\">", " ", "</p>\n              <p>", " ", ",\n              ", " ", ",\n              ", " ", ",\n              ", " ", "</p>\n            </lesson-highlight>\n            </li>\n            ", "\n            ", "\n            ", "\n            <li>\n            <lesson-highlight icon=\"icons:link\">\n              <p slot=\"title\">", " ", "</p>\n            </lesson-highlight>\n            </li>\n            <li>\n            <lesson-highlight icon=\"communication:import-contacts\">\n              <p slot=\"title\">", " ", "</p>\n            </lesson-highlight>\n            </li>\n            ", "\n            <li>\n            <lesson-highlight icon=\"device:access-time\">\n              <p slot=\"title\">", ": <relative-time datetime=\"", "\"></relative-time></p>\n            </lesson-highlight>\n            </li>\n            <li>\n            <lesson-highlight icon=\"device:access-time\">\n              <p slot=\"title\">", ": <relative-time datetime=\"", "\"></relative-time></p>\n            </lesson-highlight>\n            </li>\n            <li>\n              <lesson-highlight icon=\"av:av-timer\">\n              <p slot=\"title\">", "</p>\n              <ol class=\"recently-updated-items\">\n                ", "\n              </ol>\n              </lesson-highlight>\n            </li>\n            </ul>\n            </div>\n        </div>\n      "])), data && data.title ? data.title : "", this.t.insights, data.pages ? data.pages : (0, _index.html)(_templateObject5_b07101d0809211edaa2833647dc78c6c || (_templateObject5_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["0"]))), this.t.pages, data.objectives, this.t.learningObjectives, data.specialTags, this.t.specialElements, data.dataTables, this.t.dataTables, data.headings, this.t.headings, data.video == 0 ? "" : (0, _index.html)(_templateObject6_b07101d0809211edaa2833647dc78c6c || (_templateObject6_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<li>\n                    <lesson-highlight icon=\"av:play-circle-outline\">\n                      <p slot=\"title\">", " ", "</p>\n                      ", "\n                    </lesson-highlight>\n                  </li>"])), data.video, this.t.videos, data.videoLength != 0 ? (0, _index.html)(_templateObject7_b07101d0809211edaa2833647dc78c6c || (_templateObject7_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["(\n                            <p>", "</p>\n                            )"])), toHHMMSS(data.videoLength)) : ""), data.h5p == 0 ? "" : (0, _index.html)(_templateObject8_b07101d0809211edaa2833647dc78c6c || (_templateObject8_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<li>\n                    <lesson-highlight icon=\"lrn:interact\">\n                      <p slot=\"title\">", " H5P</p>\n                    </lesson-highlight>\n                  </li>"])), data.h5p), data.audio == 0 ? "" : (0, _index.html)(_templateObject9_b07101d0809211edaa2833647dc78c6c || (_templateObject9_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<li>\n                    <lesson-highlight icon=\"av:music-video\">\n                      <p slot=\"title\">", " ", "</p>\n                    </lesson-highlight>\n                  </li>"])), data.audio, this.t.audio), data.links, this.t.externalLinks, this.getReadingTime(data.readTime), this.t.ofReading, data.readability ? (0, _index.html)(_templateObject10_b07101d0809211edaa2833647dc78c6c || (_templateObject10_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n                    <li>\n                      <lesson-highlight icon=\"editable-table:numbers\">\n                        <p slot=\"title\">\n                          ", "\n                          ", "\n                        </p>\n                      </lesson-highlight>\n                    </li>\n                    <li>\n                      <lesson-highlight icon=\"hax:format-textblock\">\n                        <p slot=\"title\">\n                          ", " ", "\n                        </p>\n                        <p>\n                          ", " ", "\n                        </p>\n                      </lesson-highlight>\n                    </li>\n                  "])), data.readability.gradeLevel, this.t.gradeReadingLevel, data.readability.lexiconCount, this.t.words, data.readability.difficultWords, this.t.longWords) : "", this.t.created, data.created, this.t.lastUpdated, data.updated, this.t.recentUpdates, data.updatedItems ? data.updatedItems.map(function (item) {
          return (0, _index.html)(_templateObject11_b07101d0809211edaa2833647dc78c6c || (_templateObject11_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral([" <li>\n                          <a @click=\"", "\" href=\"", "\"\n                            >", "\n                            <relative-time\n                              datetime=\"", "\"\n                            ></relative-time\n                          ></a>\n                        </li>"])), _this6.closeModal, item.slug, item.title, item.metadata.updated);
        }) : "")) : "", this.t.linkChecker, this.activeTab == "linkchecker" ? (0, _index.html)(_templateObject12_b07101d0809211edaa2833647dc78c6c || (_templateObject12_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n                <loading-indicator\n                  full\n                  ?loading=\"", "\"\n                ></loading-indicator>\n                ", "\n              "])), this.loading, this.loading ? (0, _index.html)(_templateObject13_b07101d0809211edaa2833647dc78c6c || (_templateObject13_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<p>", " ", "..</p>"])), this.t.loading, this.t.linkChecker) : (0, _index.html)(_templateObject14_b07101d0809211edaa2833647dc78c6c || (_templateObject14_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n                      <div>\n                        <h2>", "</h2>\n                        <form id=\"form\">\n                          <simple-fields\n                            id=\"schema\"\n                            autofocus\n                            @value-changed=\"", "\"\n                          ></simple-fields>\n                        </form>\n                        <ul class=\"link-list\">\n                          ", "\n                        </ul>\n                      </div>\n                    "])), this.t.linkReport, this.linkFormChanged, data.linkData ? Object.keys(data.linkData).map(function (key) {
          return _this6.renderLinkCheck(data.linkData, key);
        }) : (0, _index.html)(_templateObject15_b07101d0809211edaa2833647dc78c6c || (_templateObject15_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["", ""])), this.t.noLinksInSelectedPages))) : "", this.t.contentBrowser, this.activeTab == "contentbrowser" ? (0, _index.html)(_templateObject16_b07101d0809211edaa2833647dc78c6c || (_templateObject16_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n                <loading-indicator\n                  full\n                  ?loading=\"", "\"\n                ></loading-indicator>\n                ", "\n              "])), this.loading, this.loading ? (0, _index.html)(_templateObject17_b07101d0809211edaa2833647dc78c6c || (_templateObject17_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<p>", " ", "..</p>"])), this.t.loading, this.t.contentBrowser) : (0, _index.html)(_templateObject18_b07101d0809211edaa2833647dc78c6c || (_templateObject18_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n                      <div>\n                        <h2>", "</h2>\n                        <form id=\"form\">\n                          <simple-fields\n                            id=\"schema\"\n                            autofocus\n                            @value-changed=\"", "\"\n                          ></simple-fields>\n                        </form>\n                        <ul class=\"content-list\">\n                          ", "\n                        </ul>\n                      </div>\n                    "])), this.t.contentBrowser, this.contentBrowserFormChanged, data.contentData ? data.contentData.map(function (item) {
          return (0, _index.html)(_templateObject19_b07101d0809211edaa2833647dc78c6c || (_templateObject19_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n                                  <li>\n                                    <accent-card\n                                      image-src=\"https://screenshoturl.elmsln.vercel.app/api/screenshotUrl?quality=25&render=img&urlToCapture=", "", "\"\n                                      horizontal\n                                    >\n                                      <div slot=\"heading\">\n                                        <a\n                                          class=\"title-link\"\n                                          href=\"", "\"\n                                          @click=\"", "\"\n                                          >", "</a\n                                        >\n                                      </div>\n                                      <div slot=\"content\">\n                                        <div>\n                                          ", "\n                                          <relative-time\n                                            datetime=\"", "\"\n                                          ></relative-time>\n                                        </div>\n                                        <div>\n                                          ", "\n                                          <relative-time\n                                            datetime=\"", "\"\n                                          ></relative-time>\n                                        </div>\n                                      </div>\n                                      <div slot=\"footer\">\n                                        <ul class=\"stats\">\n                                          ", "\n                                          ", "\n                                          ", "\n                                          \n                                          ", "\n                                          ", "\n                                          ", "\n                                          ", "\n                                          ", "\n                                          ", "\n                                          ", "\n                                          ", "\n                                        </ul>\n                                      </div>\n                                    </accent-card>\n                                  </li>\n                                "])), base, item.location, item.slug, _this6.closeModal, item.title, _this6.t.created, item.created, _this6.t.lastUpdated, item.updated, item.objectives > 0 ? (0, _index.html)(_templateObject20_b07101d0809211edaa2833647dc78c6c || (_templateObject20_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<li>\n                                                <simple-icon\n                                                  icon=\"editor:format-list-bulleted\"\n                                                ></simple-icon\n                                                >", "\n                                                ", "\n                                              </li>"])), item.objectives, _this6.t.learningObjectives) : "", item.videos > 0 ? (0, _index.html)(_templateObject21_b07101d0809211edaa2833647dc78c6c || (_templateObject21_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<li>\n                                                <simple-icon\n                                                  icon=\"av:play-arrow\"\n                                                ></simple-icon\n                                                >", " ", "\n                                              </li>"])), item.videos, _this6.t.videos) : "", item.h5p > 0 ? (0, _index.html)(_templateObject22_b07101d0809211edaa2833647dc78c6c || (_templateObject22_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<li>\n                                                <simple-icon\n                                                  icon=\"lrn:interact\"\n                                                ></simple-icon\n                                                >", " H5P\n                                              </li>"])), item.h5p) : "", item.placeholders > 0 ? (0, _index.html)(_templateObject23_b07101d0809211edaa2833647dc78c6c || (_templateObject23_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<li>\n                                                <simple-icon\n                                                  icon=\"hax:placeholder-image\"\n                                                ></simple-icon\n                                                >", "\n                                                ", "\n                                              </li>"])), item.placeholders, _this6.t.placeholders) : "", item.audio > 0 ? (0, _index.html)(_templateObject24_b07101d0809211edaa2833647dc78c6c || (_templateObject24_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<li>\n                                                <simple-icon\n                                                  icon=\"av:music-video\"\n                                                ></simple-icon\n                                                >", " ", "\n                                              </li>"])), item.audio, _this6.t.audio) : "", item.selfChecks > 0 ? (0, _index.html)(_templateObject25_b07101d0809211edaa2833647dc78c6c || (_templateObject25_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<li>\n                                                <simple-icon\n                                                  icon=\"hardware:videogame-asset\"\n                                                ></simple-icon\n                                                >", "\n                                                ", "\n                                              </li>"])), item.selfChecks, _this6.t.selfChecks) : "", item.images > 0 ? (0, _index.html)(_templateObject26_b07101d0809211edaa2833647dc78c6c || (_templateObject26_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<li>\n                                                <simple-icon\n                                                  icon=\"image:collections\"\n                                                ></simple-icon\n                                                >", " ", "\n                                              </li>"])), item.images, _this6.t.images) : "", item.dataTables > 0 ? (0, _index.html)(_templateObject27_b07101d0809211edaa2833647dc78c6c || (_templateObject27_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<li>\n                                                <simple-icon\n                                                  icon=\"image:grid-on\"\n                                                ></simple-icon\n                                                >", "\n                                                ", "\n                                              </li>"])), item.dataTables, _this6.t.dataTables) : "", item.specialTags > 0 ? (0, _index.html)(_templateObject28_b07101d0809211edaa2833647dc78c6c || (_templateObject28_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<li>\n                                                <simple-icon\n                                                  icon=\"icons:stars\"\n                                                ></simple-icon\n                                                >", "\n                                                ", "\n                                              </li>"])), item.specialTags, _this6.t.specialElements) : "", item.links > 0 ? (0, _index.html)(_templateObject29_b07101d0809211edaa2833647dc78c6c || (_templateObject29_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<li>\n                                                <simple-icon\n                                                  icon=\"icons:link\"\n                                                ></simple-icon\n                                                >", " ", "\n                                              </li>"])), item.links, _this6.t.links) : "", item.readTime > 0 ? (0, _index.html)(_templateObject30_b07101d0809211edaa2833647dc78c6c || (_templateObject30_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<li>\n                                                <simple-icon\n                                                  icon=\"icons:chrome-reader-mode\"\n                                                ></simple-icon\n                                                >", "\n                                                ", "\n                                              </li>"])), _this6.getReadingTime(item.readTime), _this6.t.ofReading) : "");
        }) : "")) : "", this.t.mediaBrowser, this.activeTab == "mediabrowser" ? (0, _index.html)(_templateObject31_b07101d0809211edaa2833647dc78c6c || (_templateObject31_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n                <loading-indicator\n                  full\n                  ?loading=\"", "\"\n                ></loading-indicator>\n                ", "\n              "])), this.loading, this.loading ? (0, _index.html)(_templateObject32_b07101d0809211edaa2833647dc78c6c || (_templateObject32_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<p>", " ", "..</p>"])), this.t.loading, this.t.mediaBrowser) : (0, _index.html)(_templateObject33_b07101d0809211edaa2833647dc78c6c || (_templateObject33_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n                      <div class=\"mediabrowser-wrapper\">\n                        <h2>", "</h2>\n                        <form id=\"form\">\n                          <simple-fields\n                            id=\"schema\"\n                            autofocus\n                            @value-changed=\"", "\"\n                          ></simple-fields>\n                        </form>\n                        <ul class=\"media-list\">\n                          ", "\n                        </ul>\n                      </div>\n                    "])), this.t.mediaBrowser, this.mediaBrowserFormChanged, data.mediaData ? data.mediaData.map(function (item) {
          return (0, _index.html)(_templateObject34_b07101d0809211edaa2833647dc78c6c || (_templateObject34_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n                                  <li>\n                                    <retro-card\n                                      nosource\n                                      class=\"", " ", "\"\n                                      accent-color=\"", "\"\n                                      title=\"", " ", "\"\n                                      tags=\"", ", ", "\"\n                                    >\n                                      <div class=\"body\">\n                                        ", "\n                                        ", "\n                                        ", "\n                                        ", "\n                                        ", "\n                                        ", "\n                                        <div>\n                                          ", "\n                                        </div>\n                                      </div>\n                                    </retro-card>\n                                  </li>\n                                "])), item.locType, item.type, _this6.accentColorFromType(item.type), item.name, item.title, item.locType, item.type, item.type == "image" ? (0, _index.html)(_templateObject35_b07101d0809211edaa2833647dc78c6c || (_templateObject35_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<img\n                                                src=\"", "\"\n                                                alt=\"", "\"\n                                                title=\"", "\"\n                                              />\n                                              <div>\n                                                ", "\n                                              </div>"])), item.source, item.alt, item.title, _this6.analyzeAltData(item)) : "", item.type == "video" && item.source.includes("videoseries") ? (0, _index.html)(_templateObject36_b07101d0809211edaa2833647dc78c6c || (_templateObject36_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<iframe\n                                              src=\"", "\"\n                                            ></iframe>"])), item.source) : "", item.type == "video" && !item.source.includes("videoseries") ? (0, _index.html)(_templateObject37_b07101d0809211edaa2833647dc78c6c || (_templateObject37_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<video-player\n                                              source=\"", "\"\n                                            ></video-player>"])), item.source) : "", item.type == "h5p" ? (0, _index.html)(_templateObject38_b07101d0809211edaa2833647dc78c6c || (_templateObject38_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<iframe\n                                              src=\"", "\"\n                                            ></iframe>"])), item.source) : "", item.type == "audio" ? (0, _index.html)(_templateObject39_b07101d0809211edaa2833647dc78c6c || (_templateObject39_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<video-player\n                                              source=\"", "\"\n                                            ></video-player>"])), item.source) : "", item.type == "document" ? (0, _index.html)(_templateObject40_b07101d0809211edaa2833647dc78c6c || (_templateObject40_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<div>\n                                              <a\n                                                href=\"", "\"\n                                                target=\"_blank\"\n                                                >", "</a\n                                              >\n                                            </div>"])), item.source, _this6.t.openInNewTab) : "", item.itemId ? _this6.renderItemLinkById(item.itemId) : "");
        }) : "".concat(this.t.noMediaInSelectedPages))) : "");
      } // alt feedback

    }, {
      key: "analyzeAltData",
      value: function analyzeAltData(item) {
        if (item.alt == null || item.alt == "null") {
          return (0, _index.html)(_templateObject41_b07101d0809211edaa2833647dc78c6c || (_templateObject41_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<simple-icon\n        icon=\"error\"\n        accent-color=\"red\"\n        title=\"missing alt text\"\n      ></simple-icon>"])));
        } else if (item.name == item.alt || item.source == item.alt) {
          return (0, _index.html)(_templateObject42_b07101d0809211edaa2833647dc78c6c || (_templateObject42_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<simple-icon\n        icon=\"error\"\n        accent-color=\"red\"\n        title=\"alt text matches file name\"\n      ></simple-icon>"])));
        } else if (item.title == item.alt) {
          return (0, _index.html)(_templateObject43_b07101d0809211edaa2833647dc78c6c || (_templateObject43_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<simple-icon\n        icon=\"error\"\n        accent-color=\"red\"\n        title=\"alt text matches title\"\n      ></simple-icon>"])));
        } else if (item.alt == "") {
          return (0, _index.html)(_templateObject44_b07101d0809211edaa2833647dc78c6c || (_templateObject44_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<simple-icon\n        icon=\"warning\"\n        accent-color=\"yellow\"\n        title=\"alt text set to blank, ensure decorative image\"\n      ></simple-icon>"])));
        } else if (item.alt && item.alt.includes("image")) {
          return (0, _index.html)(_templateObject45_b07101d0809211edaa2833647dc78c6c || (_templateObject45_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<simple-icon\n        icon=\"warning\"\n        accent-color=\"yellow\"\n        title=\"alt text includes word 'image'\"\n      ></simple-icon>"])));
        } else if (item.alt && item.alt.includes("picture")) {
          return (0, _index.html)(_templateObject46_b07101d0809211edaa2833647dc78c6c || (_templateObject46_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<simple-icon\n        icon=\"warning\"\n        accent-color=\"yellow\"\n        title=\"alt text includes word 'picture'\"\n      ></simple-icon>"])));
        }

        return (0, _index.html)(_templateObject47_b07101d0809211edaa2833647dc78c6c || (_templateObject47_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<simple-icon\n      icon=\"info\"\n      accent-color=\"blue\"\n      title=\"alt text: ", "\"\n    ></simple-icon>"])), item.alt);
      } // response is key'ed object by link and response data

    }, {
      key: "linkValidationResponse",
      value: function linkValidationResponse(e) {
        var _this7 = this;

        var keys = Object.keys(e.data);
        this.linkResponseData[keys[0]] = e.data[keys[0]]; // basic debounce so we only update every 500ms if we had a lot of links

        clearTimeout(this.__interval);
        this.__interval = setTimeout(function () {
          _this7.requestUpdate();
        }, 500);
      }
    }, {
      key: "renderLinkCheck",
      value: function renderLinkCheck(links, key) {
        var _this8 = this;

        // kick off a call that'll satisfy later down the road, but only if its a new key
        if (!this.linkResponseData[key]) {
          // set something so we don't get looped back  over and over
          this.linkResponseData[key] = {
            ok: "loading"
          };

          _microFrontendRegistry.MicroFrontendRegistry.call("@core/linkValidator", {
            links: key
          }, this.linkValidationResponse.bind(this));
        }

        return (0, _index.html)(_templateObject48_b07101d0809211edaa2833647dc78c6c || (_templateObject48_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n      <li class=\"link-status-", "\">\n        ", "\n        ", "\n        ", "\n        (", ")\n      </li>\n    "])), this.linkResponseData[key].ok, this.linkResponseData[key].ok === "loading" ? (0, _index.html)(_templateObject49_b07101d0809211edaa2833647dc78c6c || (_templateObject49_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<simple-icon\n              icon=\"hax:loading\"\n              accent-color=\"grey\"\n            ></simple-icon>"]))) : "", this.linkResponseData[key].ok != "loading" ? this.linkResponseData[key].ok ? (0, _index.html)(_templateObject50_b07101d0809211edaa2833647dc78c6c || (_templateObject50_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<simple-icon\n                icon=\"check\"\n                accent-color=\"green\"\n              ></simple-icon>"]))) : (0, _index.html)(_templateObject51_b07101d0809211edaa2833647dc78c6c || (_templateObject51_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<simple-icon\n                icon=\"clear\"\n                accent-color=\"red\"\n                title=\"", "\"\n              ></simple-icon>"])), this.linkResponseData[key].status) : "", this.linkResponseData[key].ok != "loading" ? this.linkResponseData[key].ok ? (0, _index.html)(_templateObject52_b07101d0809211edaa2833647dc78c6c || (_templateObject52_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<a\n                href=\"", "\"\n                target=\"_blank\"\n                rel=\"noopener nofollow noreferrer\"\n                >", "</a\n              >"])), key, key) : (0, _index.html)(_templateObject53_b07101d0809211edaa2833647dc78c6c || (_templateObject53_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["", ""])), key) : "", links[key].map(function (linkUsage) {
          return (0, _index.html)(_templateObject54_b07101d0809211edaa2833647dc78c6c || (_templateObject54_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n            <strong>", "</strong> ", "\n            ", "\n          "])), linkUsage.linkTitle, _this8.t.onPage, linkUsage.itemId ? _this8.renderItemLinkById(linkUsage.itemId) : "");
        }));
      }
    }, {
      key: "accentColorFromType",
      value: function accentColorFromType(type) {
        switch (type) {
          case "video":
            return "red";

          case "h5p":
            return "purple";

          case "audio":
            return "orange";

          case "image":
            return "blue";

          case "document":
            return "green";

          default:
            return "grey";
        }
      }
    }, {
      key: "renderItemLinkById",
      value: function renderItemLinkById(itemId) {
        // trap for highest level of the site
        if (itemId != null) {
          var item = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.findItem(itemId));

          if (item) {
            return (0, _index.html)(_templateObject55_b07101d0809211edaa2833647dc78c6c || (_templateObject55_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<a href=\"", "\" @click=\"", "\"\n          >", "</a\n        >"])), item.slug, this.closeModal, item.title);
          }
        }

        return (0, _index.html)(_templateObject56_b07101d0809211edaa2833647dc78c6c || (_templateObject56_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral([""])));
      }
    }, {
      key: "pageSelector",
      value: function pageSelector() {
        var itemManifest = _haxcmsSiteStore.store.getManifestItems(true); // default to null parent as the whole site


        var items = [{
          text: "-- ".concat(this.t.fullSite, " --"),
          value: null
        }];
        itemManifest.forEach(function (el) {
          // calculate -- depth so it looks like a tree
          var itemBuilder = el; // walk back through parent tree

          var distance = "- ";

          while (itemBuilder && itemBuilder.parent != null) {
            itemBuilder = itemManifest.find(function (i) {
              return i.id == itemBuilder.parent;
            }); // double check structure is sound

            if (itemBuilder) {
              distance = "--" + distance;
            }
          }

          items.push({
            text: distance + el.title,
            value: el.id
          });
        });
        return (0, _index.html)(_templateObject57_b07101d0809211edaa2833647dc78c6c || (_templateObject57_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["<label style=\"font-weight:bold;\" for=\"selector\"\n        >", "</label\n      >:<select id=\"selector\">\n        ", "\n      </select>"])), this.t.pageToProvideInsightsAbout, items.map(function (item) {
          return (0, _index.html)(_templateObject58_b07101d0809211edaa2833647dc78c6c || (_templateObject58_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n            <option\n              .value=\"", "\"\n              ?selected=\"", "\"\n            >\n              ", "\n            </option>\n          "])), item.value, (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.activeId) == item.value, item.text);
        }));
      }
    }], [{
      key: "styles",
      get:
      /**
       * LitElement constructable styles enhancement
       */
      function get() {
        return [(0, _index.css)(_templateObject59_b07101d0809211edaa2833647dc78c6c || (_templateObject59_b07101d0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          overflow: auto;\n          height: 85vh;\n        }\n        :host([hide-results*=\"link-status-false\"]) .link-status-false,\n        :host([hide-results*=\"link-status-true\"]) .link-status-true {\n          display: none;\n        }\n        ul {\n          list-style: none;\n          margin: 0;\n          padding: 0;\n        }\n        a11y-tabs {\n          --a11y-tabs-horizontal-background: white;\n        }\n        a11y-tabs::part(tab) {\n          font-size: 24px;\n        }\n        a11y-tabs::part(tab-insights) {\n          color: var(--simple-colors-default-theme-purple-8);\n        }\n        a11y-tabs::part(tab-linkchecker) {\n          color: var(--simple-colors-default-theme-orange-8);\n        }\n        a11y-tabs::part(tab-mediabrowser) {\n          color: var(--simple-colors-default-theme-red-8);\n        }\n        a11y-tabs::part(tab-contentbrowser) {\n          color: var(--simple-colors-default-theme-blue-8);\n        }\n        a11y-tab#insights loading-indicator {\n          --loading-indicator-color: var(\n            --simple-colors-default-theme-purple-8\n          );\n        }\n        a11y-tab#linkchecker loading-indicator {\n          --loading-indicator-color: var(\n            --simple-colors-default-theme-orange-8\n          );\n        }\n        a11y-tab#mediabrowser loading-indicator {\n          --loading-indicator-color: var(--simple-colors-default-theme-red-8);\n        }\n        a11y-tab#contentbrowser loading-indicator {\n          --loading-indicator-color: var(--simple-colors-default-theme-blue-8);\n        }\n        simple-icon {\n          --simple-icon-height: 24px;\n          --simple-icon-width: 24px;\n          padding: 2px;\n        }\n        accent-card {\n          height: 225px;\n          width: 300px;\n          overflow-x: hidden;\n          overflow-y: auto;\n          --accent-card-image-width: 25%;\n          --accent-card-image-x: left;\n          --accent-card-image-y: top;\n        }\n        accent-card .title-link {\n          text-decoration: none;\n          font-family: \"Roboto Mono\", Consolas, Monospace;\n          font-size: 20px;\n          color: blue;\n        }\n        accent-card .title-link:hover,\n        accent-card .title-link:focus,\n        accent-card .title-link:active {\n          text-decoration: underline;\n          color: black;\n        }\n        accent-card .stats {\n          height: 125px;\n        }\n        .content-list li accent-card .stats li {\n          display: block;\n          padding: 0px;\n          margin: 0px;\n          list-style: none;\n        }\n        .content-list li accent-card .stats li simple-icon {\n          padding-right: 8px;\n        }\n        .content-list li {\n          display: inline-block;\n          margin: 0;\n          padding: 0;\n          font-size: 12px;\n          --accent-card-padding: 0;\n        }\n        .insights {\n          column-count: 3;\n        }\n        .insights .recently-updated-items {\n          font-size: 16px;\n          line-height: 20px;\n        }\n        .media-list li {\n          margin: 8px;\n          display: inline-flex;\n        }\n        .media-list retro-card {\n          width: 350px;\n        }\n        .media-list retro-card::part(title) {\n          padding: 0;\n          font-size: 20px;\n          overflow: hidden;\n          text-overflow: ellipsis;\n          word-break: break-word;\n          height: 75px;\n        }\n        .media-list retro-card::part(description) {\n          padding: 0;\n          font-size: 14px;\n        }\n        .media-list retro-card .body {\n          height: 275px;\n        }\n        .media-list retro-card img {\n          max-height: 225px;\n          width: 275px;\n        }\n        .mediabrowser-wrapper {\n          overflow: hidden;\n        }\n      "])))];
      }
    }, {
      key: "tag",
      get: function get() {
        return "haxcms-site-insights";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSShareDialog), "properties", this)), {}, {
          data: {
            type: Object
          },
          hideResults: {
            type: String,
            attribute: "hide-results",
            reflect: true
          },
          filters: {
            type: Object
          },
          linkResponseData: {
            type: Object
          },
          activeTab: {
            type: String
          },
          base: {
            type: String,
            reflect: true
          },
          loading: {
            type: Boolean,
            reflect: true
          }
        });
      }
    }]);
    return HAXCMSShareDialog;
  }((0, _HAXCMSI18NMixin2.HAXCMSI18NMixin)(_index.LitElement));

  _exports.HAXCMSShareDialog = HAXCMSShareDialog;
  customElements.define(HAXCMSShareDialog.tag, HAXCMSShareDialog);

  // convert seconds back into full time stamp
  function toHHMMSS(seconds) {
    var out = "";
    var snum = parseInt(seconds, 10);
    var hours = Math.floor(snum / 3600);
    var minutes = Math.floor((snum - hours * 3600) / 60);

    if (hours !== 0) {
      out += "".concat(hours, " hour");

      if (hours !== 1) {
        out += "s";
      }

      out += ", ";
    }

    if (minutes !== 0) {
      out += "".concat(minutes, " minute");

      if (hours !== 1) {
        out += "s";
      }
    }

    return out;
  }
});
