// @lrnwebcomponents/haxcms-elements/lib/core/haxcms-site-dashboard.js
define(["exports", "../../../../lit-element/lit-element.js", "./haxcms-site-store.js", "../../../utils/utils.js", "../../../../mobx/lib/mobx.module.js", "../../../../@polymer/paper-button/paper-button.js", "../../../../@polymer/paper-icon-button/paper-icon-button.js", "../../../simple-tooltip/simple-tooltip.js", "../../../../@polymer/iron-icons/iron-icons.js", "../../../../@polymer/iron-icons/notification-icons.js", "../../../simple-fields/lib/simple-fields-form.js", "../../../portal-launcher/portal-launcher.js", "../../../hax-body-behaviors/lib/HAXFields.js"], function (_exports, _litElement, _haxcmsSiteStore, _utils, _mobxModule, _paperButton, _paperIconButton, _simpleTooltip, _ironIcons, _notificationIcons, _simpleFieldsForm, _portalLauncher, _HAXFields) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HAXCMSSiteDashboard = void 0;

  function _templateObject3_9cae3a60c18011eaab345dc7f6abc3f3() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          z-index: 1;\n          display: inline-block;\n          vertical-align: top;\n          position: fixed;\n          height: 100vh;\n          width: 50vw;\n          margin-left: -50vw;\n          border-right: 2px solid #17271f;\n          overflow: scroll;\n          background-color: var(--haxcms-dashboard-bg, #37474f);\n        }\n        :host([dashboard-opened]) {\n          margin-left: 0;\n        }\n        #homebutton {\n          display: inline-block;\n          vertical-align: top;\n          margin-top: 4px;\n        }\n        #homebutton paper-icon-button {\n          color: #ffffff;\n          --paper-icon-button-ink-color: var(--haxcms-color, #ffffff);\n          background-color: #000000;\n          height: 48px;\n          width: 48px;\n          padding: 8px;\n          border-radius: 50%;\n        }\n        #homebutton paper-icon-button:hover,\n        #homebutton paper-icon-button:active,\n        #homebutton paper-icon-button:focus {\n          background-color: var(--haxcms-color, #ffffff);\n        }\n        .buttons {\n          border: 1px solid #aaaaaa;\n          background-color: var(--simple-modal-titlebar-background, #000000);\n          color: var(--simple-modal-titlebar-color, #ffffff);\n          position: absolute;\n          bottom: 0;\n          left: 0;\n          right: 0;\n          display: flex;\n        }\n        .buttons paper-button {\n          color: #ffffff;\n          background-color: #000000;\n          font-weight: bold;\n          text-transform: none;\n          border-radius: 0;\n          border: none;\n          border-right: 1px solid white;\n          margin: 0;\n        }\n        #save {\n          background-color: var(--haxcms-system-action-color);\n        }\n        #cancel {\n          background-color: var(--haxcms-system-danger-color);\n        }\n        #save:active,\n        #save:focus,\n        #save:hover,\n        #cancel:active,\n        #cancel:focus,\n        #cancel:hover {\n          background-color: var(--haxcms-color);\n        }\n        .buttons paper-button:active,\n        .buttons paper-button:focus,\n        .buttons paper-button:hover {\n          color: #ffffff;\n        }\n        .title {\n          color: white;\n          font-size: 24px;\n          margin: 0 0 0 16px;\n          padding: 0;\n          display: inline-flex;\n        }\n        @media screen and (max-width: 600px) {\n          :host {\n            width: 90vw;\n            margin-left: -90vw;\n          }\n          .title {\n            font-size: 14px;\n            margin: 0;\n          }\n          .toptext {\n            font-size: 11px;\n          }\n          #homebutton paper-icon-button {\n            height: 36px;\n            width: 36px;\n            padding: 4px;\n          }\n        }\n        paper-button {\n          background-color: white;\n          color: black;\n        }\n        .publishlink {\n          color: white;\n        }\n        .title-wrapper {\n          padding: 0 16px;\n        }\n        .toptext {\n          padding: 0;\n          margin: 0;\n          font-size: 12px;\n          font-style: italic;\n          display: inline-flex;\n        }\n        .fields-wrapper {\n          height: auto;\n          background-color: white;\n        }\n        #siteform {\n          --a11y-tabs-height: 80vh;\n          --a11y-tabs-tab-height: 75vh;\n          --a11y-tabs-tab-overflow: scroll;\n          --primary-color: var(--haxcms-color, #000000);\n          --paper-input-container-focus-color: var(--haxcms-color, #000000);\n          --lumo-primary-text-color: var(--haxcms-color, #000000);\n          --a11y-tabs-color: var(--haxcms-color, #000000);\n          --a11y-tabs-focus-color: var(--haxcms-color, #000000);\n        }\n      "]);

    _templateObject3_9cae3a60c18011eaab345dc7f6abc3f3 = function _templateObject3_9cae3a60c18011eaab345dc7f6abc3f3() {
      return data;
    };

    return data;
  }

  function _templateObject2_9cae3a60c18011eaab345dc7f6abc3f3() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <span class=\"toptext\">\n                <a\n                  class=\"publishlink\"\n                  href=\"", "\"\n                  rel=\"noopener noreferrer\"\n                  target=\"_blank\"\n                  >Published version\n                </a>\n              </span>\n            "]);

    _templateObject2_9cae3a60c18011eaab345dc7f6abc3f3 = function _templateObject2_9cae3a60c18011eaab345dc7f6abc3f3() {
      return data;
    };

    return data;
  }

  function _templateObject_9cae3a60c18011eaab345dc7f6abc3f3() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"title-wrapper\">\n        <portal-launcher>\n          <a href=\"../../\" tabindex=\"-1\" id=\"homebutton\"\n            ><paper-icon-button\n              icon=\"icons:home\"\n              title=\"Back to site list\"\n            ></paper-icon-button\n          ></a>\n        </portal-launcher>\n        <simple-tooltip for=\"homebutton\" offset=\"14\" position=\"bottom\">\n          Back to site list\n        </simple-tooltip>\n        <h2 class=\"title\">", " settings</h2>\n        ", "\n      </div>\n      <div class=\"fields-wrapper\">\n        <simple-fields-form\n          id=\"siteform\"\n          autoload\n          .headers=\"", "\"\n          .body=\"", "\"\n          .schematizer=\"", "\"\n          .elementizer=\"", "\"\n          load-endpoint=\"", "\"\n          method=\"", "\"\n        ></simple-fields-form>\n      </div>\n      <div class=\"buttons\">\n        <paper-button\n          title=\"Save site settings\"\n          id=\"save\"\n          @click=\"", "\"\n          ><iron-icon icon=\"icons:save\"></iron-icon> Save settings</paper-button\n        >\n        <paper-button\n          title=\"Cancel and close dashboard\"\n          id=\"cancel\"\n          @click=\"", "\"\n          ><iron-icon icon=\"icons:cancel\"></iron-icon> Cancel</paper-button\n        >\n      </div>\n    "]);

    _templateObject_9cae3a60c18011eaab345dc7f6abc3f3 = function _templateObject_9cae3a60c18011eaab345dc7f6abc3f3() {
      return data;
    };

    return data;
  }

  /**
   * `haxcms-site-dashboard`
   * `Off screen dashboard for modifying internal settings to the site`
   *
   * @demo demo/index.html
   */
  var HAXCMSSiteDashboard =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(HAXCMSSiteDashboard, _LitElement);
    babelHelpers.createClass(HAXCMSSiteDashboard, null, [{
      key: "tag",
      get: function get() {
        return "haxcms-site-dashboard";
      }
    }]);

    function HAXCMSSiteDashboard() {
      var _this;

      babelHelpers.classCallCheck(this, HAXCMSSiteDashboard);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(HAXCMSSiteDashboard).call(this));
      _this.manifest = {};
      _this.__disposer = []; // see up a tag to place RIGHT next to the site-builder itself

      (0, _mobxModule.autorun)(function (reaction) {
        _this.jwt = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.jwt);

        _this.__disposer.push(reaction);
      });
      (0, _mobxModule.autorun)(function (reaction) {
        _this.dashboardOpened = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.dashboardOpened);

        _this.__disposer.push(reaction);
      });
      (0, _mobxModule.autorun)(function (reaction) {
        _this.manifest = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.manifest);

        _this.__disposer.push(reaction);
      });
      return _this;
    }

    babelHelpers.createClass(HAXCMSSiteDashboard, [{
      key: "render",
      // render function
      value: function render() {
        return (0, _litElement.html)(_templateObject_9cae3a60c18011eaab345dc7f6abc3f3(), this.manifest.title, (0, _utils.varExists)(this.manifest, "metadata.site.static.publishedLocation") ? (0, _litElement.html)(_templateObject2_9cae3a60c18011eaab345dc7f6abc3f3(), (0, _utils.varGet)(this.manifest, "metadata.site.static.publishedLocation", "#")) : "", this.headers, this.body, _HAXFields.HaxSchematizer, _HAXFields.HaxElementizer, this.loadEndpoint, this.method, this._saveSiteFieldsTap, this._cancel);
      }
    }, {
      key: "disconnectedCallback",

      /**
       * Detatched life cycle
       */
      value: function disconnectedCallback() {
        for (var i in this.__disposer) {
          this.__disposer[i].dispose();
        }

        babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSSiteDashboard.prototype), "disconnectedCallback", this).call(this);
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName === "dashboardOpened" && _this2.dashboardOpened) {
            // API function so we refresh new data every time
            _this2.removeAttribute("aria-hidden");

            _this2.removeAttribute("tabindex");
          }

          if (propName === "dashboardOpened" && !_this2.dashboardOpened) {
            _this2.setAttribute("aria-hidden", "aria-hidden");

            _this2.setAttribute("tabindex", "-1");
          }
        });
      }
      /**
       * Save the fields as we get tapped
       */

    }, {
      key: "_saveSiteFieldsTap",
      value: function _saveSiteFieldsTap(e) {
        // fire event with details for saving
        window.dispatchEvent(new CustomEvent("haxcms-save-site-data", {
          bubbles: true,
          composed: true,
          cancelable: true,
          detail: this.shadowRoot.querySelector("#siteform").submit()
        }));
      }
      /**
       * Close the dashboard and reset state
       */

    }, {
      key: "_cancel",
      value: function _cancel(e) {
        window.dispatchEvent(new CustomEvent("haxcms-load-site-dashboard", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: e.target
        }));
      }
    }], [{
      key: "styles",
      get: function get() {
        return [(0, _litElement.css)(_templateObject3_9cae3a60c18011eaab345dc7f6abc3f3())];
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          dashboardOpened: {
            type: Boolean,
            reflect: true,
            attribute: "dashboard-opened"
          },

          /**
           * Allow method to be overridden, useful in local testing
           */
          method: {
            type: String
          },

          /**
           * JSON Web token, it'll come from a global call if it's available
           */
          jwt: {
            type: String
          },

          /**
           * Publishing end point, this has CDN implications so show message
           */
          publishing: {
            type: Boolean
          },

          /**
           * Outline of items in json outline schema format
           */
          manifest: {
            type: Object
          }
        };
      }
    }]);
    return HAXCMSSiteDashboard;
  }(_litElement.LitElement);

  _exports.HAXCMSSiteDashboard = HAXCMSSiteDashboard;
  window.customElements.define(HAXCMSSiteDashboard.tag, HAXCMSSiteDashboard);
});
