// @lrnwebcomponents/haxcms-elements/lib/core/haxcms-share-dialog.js
define(["exports", "../../../../lit/index.js", "./utils/HAXCMSI18NMixin.js", "../../../code-sample/code-sample.js", "../../../simple-icon/simple-icon.js", "../../../simple-icon/lib/simple-icons.js"], function (_exports, _index, _HAXCMSI18NMixin2, _codeSample, _simpleIcon, _simpleIcons) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HAXCMSShareDialog = void 0;

  var _templateObject_b011a3c0809211edaa2833647dc78c6c, _templateObject2_b011a3c0809211edaa2833647dc78c6c, _templateObject3_b011a3c0809211edaa2833647dc78c6c;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

  function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

  function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `haxcms-outline-editor-dialog`
   * `Dialog for presenting an editable outline`
   *
   * @demo demo/index.html
   *
   * @microcopy - the mental model for this element
   */
  var HAXCMSShareDialog = /*#__PURE__*/function (_HAXCMSI18NMixin) {
    babelHelpers.inherits(HAXCMSShareDialog, _HAXCMSI18NMixin);

    var _super = _createSuper(HAXCMSShareDialog);

    function HAXCMSShareDialog() {
      var _this;

      babelHelpers.classCallCheck(this, HAXCMSShareDialog);
      _this = _super.call(this);
      _this.link = window.location.href;
      return _this;
    }

    babelHelpers.createClass(HAXCMSShareDialog, [{
      key: "render",
      value: // render function
      function render() {
        var _this2 = this;

        return (0, _index.html)(_templateObject_b011a3c0809211edaa2833647dc78c6c || (_templateObject_b011a3c0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n      <form>\n        <label for=\"access\">Who are you sharing with?</label>\n        <select @change=\"", "\" name=\"access\" id=\"access\">\n          <option value=\"oer\">Open to anyone</option>\n          <option value=\"course\">Account required to view</option>\n          <option value=\"iam\">Editor rights required</option>\n        </select>\n        <details>\n          <summary>Disable features</summary>\n          <div>\n            ", "\n          </div>\n        </details>\n        <label for=\"link\">Link</label\n        ><input type=\"text\" id=\"link\" value=\"", "\" size=\"125\" /><br />\n        <label for=\"iframe\">Embed</label>\n        <code-sample\n          type=\"html\"\n          copy-clipboard-button\n          id=\"iframe\"\n        ></code-sample>\n        <label for=\"height\">Embed height</label>\n        <input\n          id=\"height\"\n          type=\"text\"\n          name=\"height\"\n          value=\"600px\"\n          @input=\"", "\"\n        />\n      </form>\n    "])), this.calculateShareCode, ["left-col", "right-col", "search", "breadcrumb", "print", "rss", "git-link", "footer", "qr-code"].map(function (item) {
          return (0, _index.html)(_templateObject2_b011a3c0809211edaa2833647dc78c6c || (_templateObject2_b011a3c0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n                <input\n                  type=\"checkbox\"\n                  name=\"disable-features\"\n                  id=\"", "\"\n                  value=\"", "\"\n                  @change=\"", "\"\n                /><label for=\"", "\">", "</label><br />\n              "])), item, item, _this2.calculateShareCode, item, item.replace("-", " "));
        }), this.link, this.calculateShareCode);
      } // generate the share

    }, {
      key: "calculateShareCode",
      value: function calculateShareCode() {
        var formData = new FormData(this.shadowRoot.querySelector("form"));
        var form = {};

        var _iterator = _createForOfIteratorHelper(formData.entries()),
            _step;

        try {
          for (_iterator.s(); !(_step = _iterator.n()).done;) {
            var pair = _step.value;

            if (form[pair[0]]) {
              form[pair[0]] += ",".concat(pair[1]);
            } else {
              form[pair[0]] = pair[1];
            }
          }
        } catch (err) {
          _iterator.e(err);
        } finally {
          _iterator.f();
        }

        this.link = "".concat(window.location.href.replace("iam.", "".concat(form.access, "."))).concat(form["disable-features"] ? "?disable-features=".concat(form["disable-features"]) : "");
        var shareCode = "<template>\n      <iframe\n        src=\"".concat(this.link, "\"\n        height=\"").concat(form.height, "\"\n        width=\"100%\"\n        frameborder=\"0\"\n        allowfullscreen=\"true\"\n        webkitallowfullscreen=\"true\"\n        mozallowfullscreen=\"true\"\n      ></iframe>\n    </template>");
        this.shadowRoot.querySelector("code-sample").innerHTML = shareCode;

        this.shadowRoot.querySelector("code-sample")._updateContent();
      }
    }, {
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSShareDialog.prototype), "firstUpdated", this).call(this, changedProperties);
        this.calculateShareCode();
      }
    }], [{
      key: "styles",
      get:
      /**
       * LitElement constructable styles enhancement
       */
      function get() {
        return [(0, _index.css)(_templateObject3_b011a3c0809211edaa2833647dc78c6c || (_templateObject3_b011a3c0809211edaa2833647dc78c6c = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          overflow: auto;\n        }\n        .buttons {\n          position: absolute;\n          bottom: 0;\n          z-index: 1000000;\n          background-color: var(--simple-modal-titlebar-background, #000000);\n          color: var(--simple-modal-titlebar-color, #ffffff);\n          left: 0;\n          right: 0;\n        }\n        .buttons button {\n          color: black;\n          background-color: white;\n        }\n        simple-icon {\n          margin-right: 4px;\n        }\n      "])))];
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       */

    }, {
      key: "tag",
      get: function get() {
        return "haxcms-share-dialog";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSShareDialog), "properties", this)), {}, {
          link: {
            type: String
          }
        });
      }
    }]);
    return HAXCMSShareDialog;
  }((0, _HAXCMSI18NMixin2.HAXCMSI18NMixin)(_index.LitElement));

  _exports.HAXCMSShareDialog = HAXCMSShareDialog;
  customElements.define(HAXCMSShareDialog.tag, HAXCMSShareDialog);
});
