// @lrnwebcomponents/code-editor/lib/code-pen-button.js
define(["exports", "../../../@polymer/polymer/polymer-element.js"], function (_exports, _polymerElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.CodePenButton = void 0;

  function _templateObject_381cf260dbdc11e9bcafdb98156fc1d6() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: block;\n        }\n      </style>\n      <form action=\"[[endPoint]]\" method=\"POST\" target=\"_blank\">\n        <input type=\"hidden\" name=\"data\" value$=\"[[dataString]]\" />\n        <input\n          type=\"image\"\n          src=\"https://s3-us-west-2.amazonaws.com/s.cdpn.io/t-1/cp-arrow-right.svg\"\n          width=\"40\"\n          height=\"40\"\n          value=\"Open code pen in a new window\"\n          class=\"codepen-mover-button\"\n        />\n      </form>\n    "], ["\n      <style>\n        :host {\n          display: block;\n        }\n      </style>\n      <form action=\"[[endPoint]]\" method=\"POST\" target=\"_blank\">\n        <input type=\"hidden\" name=\"data\" value\\$=\"[[dataString]]\" />\n        <input\n          type=\"image\"\n          src=\"https://s3-us-west-2.amazonaws.com/s.cdpn.io/t-1/cp-arrow-right.svg\"\n          width=\"40\"\n          height=\"40\"\n          value=\"Open code pen in a new window\"\n          class=\"codepen-mover-button\"\n        />\n      </form>\n    "]);

    _templateObject_381cf260dbdc11e9bcafdb98156fc1d6 = function _templateObject_381cf260dbdc11e9bcafdb98156fc1d6() {
      return data;
    };

    return data;
  }

  /**
   * `code-pen-button`
   * `Post data to codepen to form a new pen`
   * @demo demo/index.html
   */
  var CodePenButton =
  /*#__PURE__*/
  function (_PolymerElement) {
    babelHelpers.inherits(CodePenButton, _PolymerElement);

    function CodePenButton() {
      babelHelpers.classCallCheck(this, CodePenButton);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(CodePenButton).apply(this, arguments));
    }

    babelHelpers.createClass(CodePenButton, [{
      key: "connectedCallback",
      value: function connectedCallback() {
        babelHelpers.get(babelHelpers.getPrototypeOf(CodePenButton.prototype), "connectedCallback", this).call(this);
        this.setAttribute("title", this.checkItOut);
      }
    }, {
      key: "_getDataString",

      /**
       * Return string from data object so it can be posted correctly.
       */
      value: function _getDataString(data) {
        return JSON.stringify(data).replace(/"/g, "&quot;").replace(/'/g, "&apos;");
      }
    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_381cf260dbdc11e9bcafdb98156fc1d6());
      }
    }, {
      key: "tag",
      get: function get() {
        return "code-pen-button";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          checkItOut: {
            type: String,
            value: "Check it out on codepen"
          },

          /**
           * End point for posting should it change in the future.
           */
          endPoint: {
            type: String,
            value: "https://codepen.io/pen/define"
          },

          /**
           * Data object as a JSON string for the POST data in page.
           */
          dataString: {
            type: String,
            computed: "_getDataString(data)"
          },

          /**
           * Data object to post to code pen
           */
          data: {
            type: Object,
            value: {}
          }
        };
      }
    }]);
    return CodePenButton;
  }(_polymerElement.PolymerElement);

  _exports.CodePenButton = CodePenButton;
  window.customElements.define(CodePenButton.tag, CodePenButton);
});
