// @lrnwebcomponents/code-editor/lib/code-pen-button.js
define(["exports", "../../../lit/index.js"], function (_exports, _index) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.CodePenButton = void 0;

  var _templateObject_60845250942811ecb1937969798b231a, _templateObject2_60845250942811ecb1937969798b231a;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `code-pen-button`
   * `Post data to codepen to form a new pen`
   * @demo demo/index.html
   * @element code-pen-button
   */
  var CodePenButton = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(CodePenButton, _LitElement);

    var _super = _createSuper(CodePenButton);

    function CodePenButton() {
      var _this;

      babelHelpers.classCallCheck(this, CodePenButton);
      _this = _super.call(this);
      _this.checkItOut = "Check it out on codepen";
      _this.endPoint = "https://codepen.io/pen/define";
      _this.data = {};
      return _this;
    }

    babelHelpers.createClass(CodePenButton, [{
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_60845250942811ecb1937969798b231a || (_templateObject_60845250942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <form action=\"", "\" method=\"POST\" target=\"_blank\">\n        <input type=\"hidden\" name=\"data\" value=\"", "\" />\n        <input\n          type=\"image\"\n          src=\"https://s3-us-west-2.amazonaws.com/s.cdpn.io/t-1/cp-arrow-right.svg\"\n          width=\"40\"\n          height=\"40\"\n          value=\"Open code pen in a new window\"\n          class=\"codepen-mover-button\"\n          part=\"button\"\n        />\n      </form>\n    "])), this.endPoint, this.dataString);
      }
    }, {
      key: "firstUpdated",
      value: function firstUpdated() {
        this.setAttribute("title", this.checkItOut);
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "data" && _this2[propName]) {
            _this2.dataString = _this2._getDataString(_this2[propName]);
          }
        });
      }
    }, {
      key: "_getDataString",
      value:
      /**
       * Return string from data object so it can be posted correctly.
       */
      function _getDataString(data) {
        return JSON.stringify(data).replace(/"/g, "&quot;").replace(/'/g, "&apos;");
      }
    }], [{
      key: "styles",
      get:
      /**
       * LitElement constructable styles enhancement
       */
      function get() {
        return [(0, _index.css)(_templateObject2_60845250942811ecb1937969798b231a || (_templateObject2_60845250942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n      "])))];
      }
    }, {
      key: "tag",
      get: function get() {
        return "code-pen-button";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          checkItOut: {
            type: String,
            attribute: "check-it-out"
          },

          /**
           * End point for posting should it change in the future.
           */
          endPoint: {
            type: String,
            attribute: "end-point"
          },

          /**
           * Data object as a JSON string for the POST data in page.
           */
          dataString: {
            type: String,
            attribute: "data-string"
          },

          /**
           * Data object to post to code pen
           */
          data: {
            type: Object
          }
        };
      }
    }]);
    return CodePenButton;
  }(_index.LitElement);

  _exports.CodePenButton = CodePenButton;
  window.customElements.define(CodePenButton.tag, CodePenButton);
});
