// @haxtheweb/rich-text-editor/lib/open-apis/link-preview-card.js
/**
 * Copyright 2025 Alexander Manbeck
 * @license Apache-2.0, see LICENSE for full text.
 */
import{LitElement as i,html as t,css as e}from"../../../../lit/index.js";import{DDDSuper as r}from"../../../d-d-d/d-d-d.js";import"../../../micro-frontend-registry/micro-frontend-registry.js";export class LinkPreviewCard extends(r(i)){static get tag(){return"link-preview-card"}constructor(){super(),this.title="",this.description="",this.image="",this.link="",this.href="",this.loadingState=!1,this.themeColor=""}static get properties(){return{...super.properties,title:{type:String},description:{type:String},image:{type:String},link:{type:String},href:{type:String},loadingState:{type:Boolean,reflect:!0,attribute:"loading-state"},themeColor:{type:String,reflect:!0,attribute:"theme-color"}}}static get styles(){return[super.styles,e`
        :host {
          display: block;
          font-family: var(--ddd-font-navigation);
        }
        .wrapper {
          margin: var(--ddd-spacing-2);
          padding: var(--ddd-spacing-2);
        }
        .card {
          border: var(--ddd-border-md);
          padding: var(--ddd-spacing-2);
          border-radius: var(--ddd-radius-xs);
          transition: border-color 0.3s ease-in-out;
        }
        :host(:focus-within) .card,
        :host(:hover) .card {
          border-color: var(--theme-color, var(--ddd-theme-primary));
        }
        .card .title {
          margin: var(--ddd-spacing-0);
          font-size: var(--ddd-font-size-xs);
          color: var(--ddd-theme-primary);
        }
        .card a {
          font-size: var(--ddd-font-size-4xs);
          color: var(--ddd-theme-primary);
        }
        .card p {
          font-size: var(--ddd-font-size-4xs);
          margin: var(--ddd-spacing-0);
          padding: var(--ddd-spacing-2) var(--ddd-spacing-0);
        }
        .loader {
          border: var(--ddd-spacing-4) solid var(--ddd-accent-1);
          border-top: var(--ddd-spacing-4) solid var(--ddd-primary-1);
          border-radius: 50%;
          width: var(--ddd-spacing-6);
          height: var(--ddd-spacing-6);
          animation: spin 2s linear infinite;
        }
        @keyframes spin {
          0% {
            transform: rotate(0deg);
          }
          100% {
            transform: rotate(360deg);
          }
        }
        @media (max-width: 600px) {
          :host {
            max-width: 100%;
            padding: var(--ddd-spacing-3);
          }
        }

        .image {
          display: inline-block;
          background-color: var(--ddd-accent-1);
        }
        .image img {
          height: auto;
          width: var(--ddd-spacing-20);
          margin: 0 auto;
          display: block;
        }
        .details {
          display: inline-block;
          width: 78%;
          vertical-align: top;
        }
      `]}render(){return t`
      <div class="wrapper">
        ${this.loadingState?t`<div class="loader"></div>`:""}
        <div class="card" style="--theme-color:${this.themeColor}">
          <div class="image">
            ${this.image?t`<img
                  src="${this.image}"
                  alt=""
                  loading="lazy"
                  decoding="async"
                  fetchpriority="low"
                />`:""}
          </div>
          <div class="details">
            <a href="${this.link}" target="_blank" rel="noopener noreferrer">
              <div class="title">${this.title||"No preview available"}</div>
              <div class="link">${this.link}</div>
            </a>
            <p>${this.description}</p>
          </div>
        </div>
      </div>
    `}updated(i){super.updated(i),i.has("href")&&this.href&&this.fetchData(this.href)}async fetchData(i){const t=globalThis.MicroFrontendRegistry.requestAvailability();await import("../../../micro-frontend-registry/lib/microServices.js").then((i=>{t.enableServices(["core"])})),this.loadingState=!0;try{let e=await t.call("@core/websiteMetadata",{q:i});this.title=e.data["og:title"]||e.data.title||"No title available",this.description=this.truncateText(e.data["og:description"]||e.data.description||"No description available",250),this.image=e.data["og:image"]||e.data.image||"",this.image&&!this.isValidURL(this.image)&&e.data["og:url"]&&(this.image=e.data["og:url"]+this.image),this.link=e.data["og:url"]||e.data.url||link,this.themeColor=e.data?.["theme-color"]?.trim()||this.getThemeColor(link)}catch(i){console.error("Error fetching data:",i),this.title="No preview available",this.description="",this.link=link,this.image="",this.themeColor=this.getThemeColor()}finally{this.loadingState=!1}}isValidURL(i){try{return new URL(i),!0}catch(i){return!1}}truncateText(i,t){return i.length>t?i.substring(0,t)+"...":i}getThemeColor(i){try{if(new URL(i).hostname.endsWith("psu.edu"))return"var(--ddd-primary-2)"}catch(t){console.warn("Invalid URL format:",i)}return`var(--ddd-primary-${Math.floor(26*Math.random())})`}}globalThis.customElements.define(LinkPreviewCard.tag,LinkPreviewCard);
