// @lrnwebcomponents/elmsln-apps/lib/lrnapp-open-studio/lrnapp-open-studio.js
define(["exports","../../../../@polymer/polymer/polymer-element.js","../../../../@polymer/polymer/lib/utils/render-status.js","../../../../@polymer/polymer/lib/legacy/polymer.dom.js","../../../../@polymer/iron-ajax/iron-ajax.js","../../../../@polymer/iron-list/iron-list.js","../../../../@polymer/iron-pages/iron-pages.js","../../../../@polymer/iron-selector/iron-selector.js","../../../../@polymer/app-layout/app-toolbar/app-toolbar.js","../../../../@polymer/polymer/lib/elements/dom-repeat.js","../../../../@polymer/app-route/app-location.js","../../../../@polymer/app-route/app-route.js","../../../../@polymer/paper-dropdown-menu/paper-dropdown-menu.js","../../../../@polymer/paper-listbox/paper-listbox.js","../../../../@polymer/paper-item/paper-item.js","../../../../@polymer/paper-toast/paper-toast.js","../../../lrndesign-gallerycard/lrndesign-gallerycard.js","../../../elmsln-loading/elmsln-loading.js","./lrnapp-open-studio-table.js","./lrnapp-open-studio-projects.js","./lrnapp-open-studio-assignments.js"],function(_exports,_polymerElement,_renderStatus,_polymerDom,_ironAjax,_ironList,_ironPages,_ironSelector,_appToolbar,_domRepeat,_appLocation,_appRoute,_paperDropdownMenu,_paperListbox,_paperItem,_paperToast,_lrndesignGallerycard,_elmslnLoading,_lrnappOpenStudioTable,_lrnappOpenStudioProjects,_lrnappOpenStudioAssignments){"use strict";Object.defineProperty(_exports,"__esModule",{value:!0});_exports.LrnappOpenStudio=void 0;class LrnappOpenStudio extends _polymerElement.PolymerElement{static get template(){return _polymerElement.html`
      <style include="materializecss-styles">
        :host {
          display: block;
          align-content: center;
        }
        #loading {
          width: 100%;
          z-index: 1000;
          opacity: 0.8;
          text-align: center;
          align-content: center;
          justify-content: center;
          height: 100vh;
          position: absolute;
          background-color: white;
        }
        iron-selector {
          line-height: 1em;
        }
        iron-selector lrnsys-button {
          display: inline-flex;
        }
        iron-selector a {
          display: inline-block;
        }
        paper-button.gallerycard-wrapper {
          margin: 0;
          padding: 0;
        }
        lrndesign-gallerycard {
          padding: 0;
          margin: 1em;
          height: 15em;
          width: 14em;
        }
        app-toolbar {
          height: 48px;
        }
        .gallery-grid {
          margin: 0 auto;
          width: 95%;
        }
        .iron-selected .display-mode {
          background-color: var(--elmsln-system-color-dark);
          color: white;
          --lrnsys-button-link-color: white;
        }
        .iron-list-container {
          display: flex;
          flex-direction: column;
          min-height: 50vh;
        }
        iron-list {
          flex: 1 1 auto;
        }
      </style>
      <iron-ajax
        auto
        url="[[sourcePath]]"
        params=""
        handle-as="json"
        last-response="{{studioResponse}}"
        on-response="_handleResponse"
      ></iron-ajax>

      <app-location
        route="{{route}}"
        query-params="{{queryParams}}"
      ></app-location>
      <app-route
        route="{{route}}"
        pattern="[[endPoint]]/:page"
        data="{{data}}"
        tail="{{tail}}"
        query-params="{{queryParams}}"
      >
      </app-route>

      <div id="loading">
        <h3>Loading..</h3>
        <elmsln-loading color="grey-text" size="large"></elmsln-loading>
      </div>
      <app-toolbar>
        <iron-selector
          selected="{{data.page}}"
          attr-for-selected="name"
          role="navigation"
        >
          <a tabindex="-1" name="submissions" on-click="_submissionsClicked"
            ><lrnsys-button
              icon="apps"
              label="Submission display"
              hover-class="amber darken-4 white-text"
              class="display-mode"
              button-class="display-mode style-scope lrnapp-open-studio x-scope lrnsys-button-0"
            ></lrnsys-button
          ></a>
          <a tabindex="-1" name="projects" on-click="_projectsClicked"
            ><lrnsys-button
              icon="folder"
              label="Project board"
              hover-class="amber darken-4 white-text"
              class="display-mode"
              button-class="display-mode style-scope lrnapp-open-studio x-scope lrnsys-button-0"
            ></lrnsys-button
          ></a>
          <a tabindex="-1" name="assignments" on-click="_assignmentsClicked"
            ><lrnsys-button
              icon="list"
              label="Assignment centric"
              hover-class="amber darken-4 white-text"
              class="display-mode"
              button-class="display-mode style-scope lrnapp-open-studio x-scope lrnsys-button-0"
            ></lrnsys-button
          ></a>
          <a tabindex="-1" name="table" on-click="_tableClicked"
            ><lrnsys-button
              icon="view-list"
              label="Table view"
              hover-class="amber darken-4 white-text"
              class="display-mode"
              button-class="display-mode style-scope lrnapp-open-studio x-scope lrnsys-button-0"
            ></lrnsys-button
          ></a>
        </iron-selector>
        <span main-title></span>
        <span
          top-item
          style="text-align:right;font-size:.5em;padding-right:1em;"
          >Displaying [[submissions.length]] of
          [[originalSubmissions.length]]</span
        >
        <paper-dropdown-menu label="Author" hidden$="[[!authors]]">
          <paper-listbox
            slot="dropdown-content"
            class="dropdown-content"
            selected="{{queryParams.author}}"
            attr-for-selected="item-id"
          >
            <paper-item></paper-item>
            <template is="dom-repeat" items="[[_toArray(authors)]]" as="author">
              <paper-item item-id="[[author.id]]"
                >[[author.display_name]]</paper-item
              >
            </template>
          </paper-listbox>
        </paper-dropdown-menu>
        <paper-dropdown-menu label="Project" hidden$="[[!projects]]">
          <paper-listbox
            slot="dropdown-content"
            class="dropdown-content"
            selected="{{queryParams.project}}"
            attr-for-selected="item-id"
          >
            <paper-item></paper-item>
            <template
              is="dom-repeat"
              items="[[_toArray(projects)]]"
              as="project"
            >
              <paper-item item-id="[[project.id]]"
                >[[project.attributes.title]]</paper-item
              >
            </template>
          </paper-listbox>
        </paper-dropdown-menu>
        <paper-dropdown-menu label="Assignment" hidden$="[[!assignments]]">
          <paper-listbox
            slot="dropdown-content"
            class="dropdown-content"
            selected="{{queryParams.assignment}}"
            attr-for-selected="item-id"
          >
            <paper-item></paper-item>
            <template
              is="dom-repeat"
              items="[[_toArray(assignments)]]"
              as="assignment"
            >
              <paper-item item-id="[[assignment.id]]"
                >[[assignment.attributes.title]]</paper-item
              >
            </template>
          </paper-listbox>
        </paper-dropdown-menu>
      </app-toolbar>
      <div class="gallery-grid">
        <iron-pages
          selected="{{data.page}}"
          attr-for-selected="name"
          fallback-selection="submissions"
          role="main"
        >
          <div class="iron-list-container" name="submissions">
            <iron-list id="ironlist" items="[[submissions]]" as="item" grid>
              <template>
                <paper-button
                  data-submission-id$="[[item.id]]"
                  class="gallerycard-wrapper"
                  on-click="_loadSubmissionUrl"
                >
                  <lrndesign-gallerycard
                    elevation="2"
                    data-submission-id$="[[item.id]]"
                    title="[[item.attributes.title]]"
                    author="[[item.relationships.author.data]]"
                    comments="[[item.meta.comment_count]]"
                    image="[[item.display.image]]"
                    icon="[[item.display.icon]]"
                    date="[[item.meta.humandate]]"
                    class="ferpa-protect"
                  >
                  </lrndesign-gallerycard>
                </paper-button>
              </template>
            </iron-list>
          </div>
          <lrnapp-open-studio-assignments
            name="assignments"
            base-path="[[basePath]]"
            submissions="[[submissions]]"
            assignments="[[assignments]]"
            active-author-id="[[queryParams.author]]"
            active-assignment-id="[[queryParams.assignment]]"
          ></lrnapp-open-studio-assignments>
          <lrnapp-open-studio-projects
            name="projects"
            base-path="[[basePath]]"
            projects="[[projects]]"
            submissions="[[submissions]]"
            assignments="[[assignments]]"
            active-author-id="[[queryParams.author]]"
            active-project-id="[[queryParams.project]]"
          ></lrnapp-open-studio-projects>
          <lrnapp-open-studio-table
            name="table"
            base-path="[[basePath]]"
            submissions="{{submissions}}"
          ></lrnapp-open-studio-table>
        </iron-pages>
      </div>
      <paper-toast id="toast"></paper-toast>
    `}static get tag(){return"lrnapp-open-studio"}static get properties(){return{elmslnCourse:{type:String},elmslnSection:{type:String},basePath:{type:String},csrfToken:{type:String},endPoint:{type:String},/**
       * The studioResponse from server
       */studioResponse:{type:Object,notify:!0},/**
       * The submissions to render; potentially filtered
       */submissions:{type:Object,notify:!0,computed:"_submissionsCompute(originalSubmissions, queryParams)"},/**
       * The original submissions array; used to filter against
       */originalSubmissions:{type:Object,notify:!0},/**
       * The submissions to render
       */projects:{type:Array,notify:!0,value:[]},/**
       * The assignments to render
       */assignments:{type:Array,notify:!0,value:[]},/**
       * The authors to render
       */authors:{type:Array,notify:!0,value:[]},/**
       * sourcePath for submission data.
       */sourcePath:{type:String,notify:!0},/**
       * Endpoint for submission data.
       */endPoint:{type:String,notify:!0},/**
       * base path for the app
       */basePath:{type:String,notify:!0},/**
       * Active / clicked submission.
       */activeSubmission:{type:String,value:null,notify:!0},queryParams:{type:Object,notify:!0},_blockcycle:{type:Boolean,value:!1}}}connectedCallback(){super.connectedCallback();(0,_renderStatus.afterNextRender)(this,function(){this.addEventListener("route-change",this._routeChange.bind(this))})}disconnectedCallback(){this.removeEventListener("route-change",this._routeChange.bind(this));super.disconnectedCallback()}static get observers(){return["_routeChanged(route, endPoint)","_deleteToast(queryParams.deletetoast)","_assignmentFilterChanged(queryParams.assignment)","_projectFilterChanged(queryParams.project)"]}// If the current route is outside the scope of our app
// then allow the website to break out of the single page
// application routing
_routeChanged(route,endPoint){if("string"===typeof route.path){if("string"===typeof endPoint){if(route.path.startsWith(endPoint)){return}}// reload the page which since route changed will load that page
window.location.reload()}}/**
   * Change route from deeper in the app.
   */_routeChange(e){var details=e.detail;if(typeof details.queryParams.assignment!==typeof void 0){this.set("queryParams.assignment",details.queryParams.assignment)}if(typeof details.queryParams.project!==typeof void 0){this.set("queryParams.project",details.queryParams.project)}if(typeof details.queryParams.author!==typeof void 0){this.set("queryParams.author",details.queryParams.author)}if(typeof details.data.page!==typeof void 0){this.set("data.page",details.data.page)}}_submissionsCompute(originalSubmissions,queryParams){// if we don't have an original submissions object to work with then we need to bail
if("undefined"===typeof originalSubmissions){return[]}// define vars
const root=this;let filteredSubmissions=[];// filter the submissions by the query params
filteredSubmissions=originalSubmissions.filter(submission=>{if("undefined"!==typeof root.queryParams.author){if(submission.relationships.author.data.id!==root.queryParams.author){return!1}}if("undefined"!==typeof root.queryParams.project){if(submission.relationships.project.data.id!==root.queryParams.project){return!1}}if("undefined"!==typeof root.queryParams.assignment){if(submission.relationships.assignment.id!==root.queryParams.assignment){return!1}}return!0});// delay and repaint, can help with refresh issues
setTimeout(()=>{this.$.ironlist.fire("iron-resize")},200);return filteredSubmissions}/**
   * Support having a toast message because of delete or error elsewhere.
   */_deleteToast(deletetoast,old){if(typeof deletetoast!==typeof void 0){if("error"==deletetoast){this.$.toast.show("That submission on longer exists!")}else{this.$.toast.show("Submission deleted successfully!")}this.set("queryParams.deletetoast",void 0);this.notifyPath("queryParams.deletetoast")}}_assignmentFilterChanged(assignment){// if we have a assignment then we need to uncheck project
if(typeof assignment!==typeof void 0&&!this._blockcycle){this._blockcycle=!0;this.set("queryParams.project",void 0);this.notifyPath("queryParams.project");this.set("queryParams.assignment",assignment);this.notifyPath("queryParams.assignment")}else{this._blockcycle=!1}}_projectFilterChanged(project){// if we have a project then we need to uncheck assignment
if(typeof project!==typeof void 0&&!this._blockcycle){this._blockcycle=!0;this.set("queryParams.project",project);this.notifyPath("queryParams.project");this.set("queryParams.assignment",void 0);this.notifyPath("queryParams.assignment")}else{this._blockcycle=!1}}/**
   * Handle tap on paper-button above to redirect to the correct submission url.
   */_loadSubmissionUrl(e){let root=this;var normalizedEvent=(0,_polymerDom.dom)(e),local=normalizedEvent.localTarget,active=local.getAttribute("data-submission-id");// @todo need a cleaner integration but this at least goes the right place for now
window.location.href=this.basePath+"lrnapp-studio-submission/submissions/"+active}/**
   * Handle response for the whole projects object.
   */_handleResponse(event){let root=this;var author={},project={},tmp={authors:[],assignments:[]},assignment={},assignments=[],authors=[],submissions=[];if(null!=root.studioResponse.data.submissions){submissions=this._toArray(root.studioResponse.data.submissions)}var projects=[];if(null!=root.studioResponse.data.projects){projects=this._toArray(root.studioResponse.data.projects)}this.set("projects",projects);// original = active off the bat then we apply filters later to chang this
this.set("originalSubmissions",submissions);// figure out authors and assignments
for(var index=0;index<submissions.length;index++){author=submissions[index].relationships.author.data;tmp.authors[author.id]=author;project=submissions[index].relationships.project.data;assignment=submissions[index].relationships.assignment;tmp.assignments[assignment.id]=assignment;tmp.assignments[assignment.id].project=project.id}// this is stupid but we have to normalize the IDs or else dom repeats will be screwed up
tmp.authors.forEach(function(element){authors.push(element)});// this is stupid but we have to normalize the IDs or else dom repeats will be screwed up
tmp.assignments.forEach(function(element){assignments.push(element)});root.$.loading.hidden=!0;this.set("assignments",assignments);this.set("authors",authors)}_submissionsClicked(e){this.set("route.path",this.endPoint+"/submissions");this.notifyPath("route.path")}_projectsClicked(e){this.set("route.path",this.endPoint+"/projects");this.notifyPath("route.path")}_assignmentsClicked(e){this.set("route.path",this.endPoint+"/assignments");this.notifyPath("route.path")}_tableClicked(e){this.set("route.path",this.endPoint+"/table");this.notifyPath("route.path")}/**
   * Simple way to convert from object to array.
   */_toArray(obj){if(null==obj){return[]}return Object.keys(obj).map(function(key){return obj[key]})}}_exports.LrnappOpenStudio=LrnappOpenStudio;window.customElements.define(LrnappOpenStudio.tag,LrnappOpenStudio)});
