// @lrnwebcomponents/sites-listing/lib/site-card.js
define(["exports", "../../../@polymer/polymer/polymer-element.js", "../../../@polymer/paper-card/paper-card.js", "../../../@polymer/iron-icon/iron-icon.js", "../../../@polymer/iron-image/iron-image.js"], function (_exports, _polymerElement, _paperCard, _ironIcon, _ironImage) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SiteCard = void 0;

  function _templateObject_63f764b0b3b911e9b58d79562a23326c() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: inline-flex;\n          width: 250px;\n        }\n        :host(:focus) {\n          outline: none;\n        }\n\n        :host([size=\"micro\"]) {\n          transform: scale(0.5);\n        }\n\n        :host([size=\"small\"]) {\n          transform: scale(0.8);\n        }\n\n        paper-card {\n          border-radius: 4px;\n          margin: 0;\n          width: 100%;\n        }\n\n        .card-actions {\n          background-color: #f5f5f5;\n          border-radius: 0 0 4px 4px;\n          padding: 0 8px;\n        }\n\n        .card-actions .card-action-details {\n          display: inline-block;\n          vertical-align: middle;\n          vertical-align: -webkit-baseline-middle;\n          width: 80%;\n        }\n\n        .card-control-height {\n          height: 240px;\n        }\n\n        [elevation=\"0\"] {\n          border: solid 1px #eeeeee;\n        }\n\n        .text-right {\n          text-align: right;\n        }\n\n        .text-left {\n          text-align: left;\n        }\n\n        .name,\n        .title {\n          color: #222;\n          font-size: 12.8px;\n          font-weight: 600;\n          line-height: 20px;\n          padding: 0 12px;\n          overflow: hidden;\n          text-overflow: ellipsis;\n          margin-top: 8px;\n        }\n\n        .title {\n          font-size: 11.2px;\n          font-weight: 400;\n        }\n\n        .divider {\n          height: 1px;\n          width: 100%;\n          background: #efefef;\n        }\n\n        .site-icon {\n          display: inline-block;\n          vertical-align: text-top;\n          transform: scale(0.8);\n          --iron-icon-height: 100%;\n          --iron-icon-width: 100%;\n          overflow: hidden;\n          color: grey;\n          margin: -16px 8px 0 0;\n          position: absolute;\n          right: 0;\n          font-size: 16px;\n          padding: 4px;\n          height: 40px;\n          width: 40px;\n          border-radius: 50%;\n          background-color: white;\n        }\n\n        .site-icon:hover,\n        .site-icon:focus {\n          color: black;\n        }\n\n        .center {\n          margin: auto;\n          width: 80%;\n          padding: 16px;\n        }\n\n        .site-info {\n          width: 100%;\n        }\n\n        .site-preview {\n          height: 160px;\n        }\n        .card-content {\n          padding: 0;\n          margin: 0;\n          overflow: hidden;\n        }\n\n        .inline {\n          display: inline;\n        }\n      </style>\n      <paper-card elevation=\"[[elevation]]\">\n        <div class=\"card-content card-control-height card-control-center\">\n          <div class=\"site-preview\">\n            <iron-image\n              style=\"width:100%; height:100%; background-color: lightgray;\"\n              sizing=\"cover\"\n              preload=\"\"\n              fade=\"\"\n              src=\"[[image]]\"\n              hidden$=\"[[!image]]\"\n            ></iron-image>\n          </div>\n          <iron-icon\n            class=\"site-icon\"\n            icon=\"[[icon]]\"\n            hidden$=\"[[!icon]]\"\n          ></iron-icon>\n          <div class=\"site-info\">\n            <div class=\"divider\"></div>\n            <div class=\"name\">[[name]]</div>\n            <div class=\"title\">[[title]]</div>\n          </div>\n        </div>\n        <div class=\"card-actions\" hidden=\"\">\n          <div class=\"card-action-details\"></div>\n        </div>\n      </paper-card>\n    "], ["\n      <style>\n        :host {\n          display: inline-flex;\n          width: 250px;\n        }\n        :host(:focus) {\n          outline: none;\n        }\n\n        :host([size=\"micro\"]) {\n          transform: scale(0.5);\n        }\n\n        :host([size=\"small\"]) {\n          transform: scale(0.8);\n        }\n\n        paper-card {\n          border-radius: 4px;\n          margin: 0;\n          width: 100%;\n        }\n\n        .card-actions {\n          background-color: #f5f5f5;\n          border-radius: 0 0 4px 4px;\n          padding: 0 8px;\n        }\n\n        .card-actions .card-action-details {\n          display: inline-block;\n          vertical-align: middle;\n          vertical-align: -webkit-baseline-middle;\n          width: 80%;\n        }\n\n        .card-control-height {\n          height: 240px;\n        }\n\n        [elevation=\"0\"] {\n          border: solid 1px #eeeeee;\n        }\n\n        .text-right {\n          text-align: right;\n        }\n\n        .text-left {\n          text-align: left;\n        }\n\n        .name,\n        .title {\n          color: #222;\n          font-size: 12.8px;\n          font-weight: 600;\n          line-height: 20px;\n          padding: 0 12px;\n          overflow: hidden;\n          text-overflow: ellipsis;\n          margin-top: 8px;\n        }\n\n        .title {\n          font-size: 11.2px;\n          font-weight: 400;\n        }\n\n        .divider {\n          height: 1px;\n          width: 100%;\n          background: #efefef;\n        }\n\n        .site-icon {\n          display: inline-block;\n          vertical-align: text-top;\n          transform: scale(0.8);\n          --iron-icon-height: 100%;\n          --iron-icon-width: 100%;\n          overflow: hidden;\n          color: grey;\n          margin: -16px 8px 0 0;\n          position: absolute;\n          right: 0;\n          font-size: 16px;\n          padding: 4px;\n          height: 40px;\n          width: 40px;\n          border-radius: 50%;\n          background-color: white;\n        }\n\n        .site-icon:hover,\n        .site-icon:focus {\n          color: black;\n        }\n\n        .center {\n          margin: auto;\n          width: 80%;\n          padding: 16px;\n        }\n\n        .site-info {\n          width: 100%;\n        }\n\n        .site-preview {\n          height: 160px;\n        }\n        .card-content {\n          padding: 0;\n          margin: 0;\n          overflow: hidden;\n        }\n\n        .inline {\n          display: inline;\n        }\n      </style>\n      <paper-card elevation=\"[[elevation]]\">\n        <div class=\"card-content card-control-height card-control-center\">\n          <div class=\"site-preview\">\n            <iron-image\n              style=\"width:100%; height:100%; background-color: lightgray;\"\n              sizing=\"cover\"\n              preload=\"\"\n              fade=\"\"\n              src=\"[[image]]\"\n              hidden\\$=\"[[!image]]\"\n            ></iron-image>\n          </div>\n          <iron-icon\n            class=\"site-icon\"\n            icon=\"[[icon]]\"\n            hidden\\$=\"[[!icon]]\"\n          ></iron-icon>\n          <div class=\"site-info\">\n            <div class=\"divider\"></div>\n            <div class=\"name\">[[name]]</div>\n            <div class=\"title\">[[title]]</div>\n          </div>\n        </div>\n        <div class=\"card-actions\" hidden=\"\">\n          <div class=\"card-action-details\"></div>\n        </div>\n      </paper-card>\n    "]);

    _templateObject_63f764b0b3b911e9b58d79562a23326c = function _templateObject_63f764b0b3b911e9b58d79562a23326c() {
      return data;
    };

    return data;
  }

  /**
   * `site-card`
   * `A simple card for representing a site`
   * @demo demo/index.html
   */
  var SiteCard =
  /*#__PURE__*/
  function (_PolymerElement) {
    babelHelpers.inherits(SiteCard, _PolymerElement);

    function SiteCard() {
      babelHelpers.classCallCheck(this, SiteCard);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(SiteCard).apply(this, arguments));
    }

    babelHelpers.createClass(SiteCard, [{
      key: "_mouseEnter",

      /**
       * Increase elevation while hovering.
       */
      value: function _mouseEnter(e) {
        this.__oldElevation = this.elevation;

        if (this.elevation + 2 > 5) {
          this.elevation = 5;
        } else {
          this.elevation += 2;
        }
      }
      /**
       * Reset the elevation.
       */

    }, {
      key: "_mouseLeave",
      value: function _mouseLeave(e) {
        this.elevation = this.__oldElevation;
      }
    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_63f764b0b3b911e9b58d79562a23326c());
      }
    }, {
      key: "tag",
      get: function get() {
        return "site-card";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Scale of the card.
           */
          size: {
            type: String,
            reflectToAttribute: true
          },

          /**
           * Cover image src.
           */
          image: {
            type: String
          },

          /**
           * Icon to use if image isn't there.
           */
          icon: {
            type: String,
            value: false
          },

          /**
           * name of the course like sing100
           */
          name: {
            type: String,
            value: ""
          },

          /**
           * title of the course like Intro to studies
           */
          title: {
            type: String,
            value: ""
          },

          /**
           * Visual elevation of the item off the UI via paper element height
           */
          elevation: {
            type: Number,
            value: 1,
            reflectToAttribute: true
          }
        };
      }
    }]);
    return SiteCard;
  }(_polymerElement.PolymerElement);

  _exports.SiteCard = SiteCard;
  window.customElements.define(SiteCard.tag, SiteCard);
});
