// @lrnwebcomponents/data-viz/data-viz.js
define(["exports", "../../@polymer/polymer/polymer-element.js", "../chartist-render/chartist-render.js"], function (_exports, _polymerElement, _chartistRender) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.DataViz = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject_d48629102c1c11eaacb4bf54e86db688() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <chartist-render\n        id=\"barChart\"\n        type=\"bar\"\n        scale=\"ct-major-twelfth\"\n        chart-title=\"Quiz Distribution\"\n        chart-desc=\"A bar graph of quizzes completed by student\"\n      >\n      </chartist-render>\n    "]);

    _templateObject_d48629102c1c11eaacb4bf54e86db688 = function _templateObject_d48629102c1c11eaacb4bf54e86db688() {
      return data;
    };

    return data;
  }

  // register globally so we can make sure there is only one
  window.DataViz = window.DataViz || {}; // request if this exists. This helps invoke the element existing in the dom
  // as well as that there is only one of them. That way we can ensure everything
  // is rendered through the same data-viz element, making it a singleton.

  window.DataViz.requestAvailability = function () {
    // if there is no single instance, generate one and append it to end of the document
    if (!window.DataViz.instance) {
      window.DataViz.instance = document.createElement("data-viz");
      document.body.appendChild(window.DataViz.instance);
    }

    return window.DataViz.instance;
  };
  /**
   * `data-viz`
   * @customElement data-viz
   * `display pouch-db data using graphs`
   *
   * @microcopy - language worth noting:
   *  -
   *
  
   * @polymer
   * @demo demo/index.html
   */


  var DataViz =
  /*#__PURE__*/
  function (_PolymerElement) {
    babelHelpers.inherits(DataViz, _PolymerElement);

    function DataViz() {
      babelHelpers.classCallCheck(this, DataViz);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(DataViz).apply(this, arguments));
    }

    babelHelpers.createClass(DataViz, [{
      key: "connectedCallback",

      /**
       * life cycle, element is afixed to the DOM
       */
      value: function connectedCallback() {
        babelHelpers.get(babelHelpers.getPrototypeOf(DataViz.prototype), "connectedCallback", this).call(this);
        window.addEventListener("show-data", this.showDataFunction.bind(this));
      }
      /**
       * Show the data based on user selecting the view and
       * that they want to see how they did.
       */

    }, {
      key: "showDataFunction",
      value: function showDataFunction(e) {
        var queryData = e.detail;
        var whatEvent = event.target.tagName;
        var bardata = {
          labels: queryData.labels,
          series: queryData.series
        };
        this.shadowRoot.querySelector("#barChart").data = bardata;
      }
      /**
       * life cycle, element is removed from the DOM
       */

    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        babelHelpers.get(babelHelpers.getPrototypeOf(DataViz.prototype), "disconnectedCallback", this).call(this);
        window.removeEventListener("show-data", this.showDataFunction.bind(this));
      }
      /**
       * Hide callback
       */

    }, {
      key: "hideDataViz",
      value: function hideDataViz(e) {} // add your code to run when the singleton hides

      /**
       * Show / available callback
       */

    }, {
      key: "showDataViz",
      value: function showDataViz(e) {// add your code to run when the singleton is called for
      }
    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_d48629102c1c11eaacb4bf54e86db688());
      }
    }, {
      key: "properties",
      // properties available to the custom element for data binding
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(DataViz), "properties", this), {
          title: {
            name: "title",
            type: String,
            value: "data-viz-default-value",
            reflectToAttribute: false,
            observer: false
          }
        });
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */

    }, {
      key: "tag",
      get: function get() {
        return "data-viz";
      }
    }]);
    return DataViz;
  }(_polymerElement.PolymerElement);

  _exports.DataViz = DataViz;
  window.customElements.define(DataViz.tag, DataViz);
});
