// @haxtheweb/app-hax/lib/wired-darkmode-toggle/wired-darkmode-toggle.js
import{rectangle as e,hachureEllipseFill as t,ellipse as i,svgNode as o}from"../../../../wired-elements/lib/wired-lib.js";import{WiredToggle as l}from"../../../../wired-elements/lib/wired-toggle.js";import{html as s,css as r,unsafeCSS as n}from"../../../../lit/index.js";const a=new URL("./images/sunIcon.png",import.meta.url).href,d=new URL("./images/moonIcon.png",import.meta.url).href;export class WiredDarkmodeToggle extends l{constructor(){super(),this.checked=!1,this.label="Dark mode",this.knobFill=o("circle")}canvasSize(){return[100,48]}static get tag(){return"wired-darkmode-toggle"}draw(e,t){this.knob=o("g"),this.knob.classList.add("knob"),e.appendChild(this.knob),this.knobFill.setAttribute("cx",26),this.knobFill.setAttribute("cy",26),this.knobFill.setAttribute("r",20),this.knobFill.setAttribute("style","fill: var(--wired-toggle-off-color); transition: fill 0.3s ease-in-out;"),this.knobFill.classList.add("knobfill"),this.knob.appendChild(this.knobFill),i(this.knob,26,26,40,40,this.seed)}toggleMode(e){e?this.knobFill.setAttribute("style","fill: var(--wired-toggle-on-color);"):this.knobFill.setAttribute("style","fill: var(--wired-toggle-off-color);")}onChange(e){this.checked=e.target.checked,this.toggleMode(this.checked),this.dispatchEvent(new Event("change",{bubbles:!0,composed:!0}))}static get properties(){return{checked:{type:Boolean,reflect:!0},disabled:{type:Boolean,reflect:!0},label:{type:String}}}render(){return s`
      <div style="position: relative;">
        <svg class="svg"></svg>
        <label for="input">${this.label}</label>
        <input
          id="input"
          type="checkbox"
          .checked="${this.checked}"
          ?disabled="${this.disabled}"
          @change="${this.onChange}"
        />
      </div>
    `}static get styles(){return[super.styles,r`
        :host {
          opacity: 1;
          display: inline-flex;
          vertical-align: top;
        }
        :host div {
          background-image: url("${n(a)}");
          background-repeat: no-repeat;
          --wired-toggle-off-color: var(--simple-colors-fixed-theme-amber-7);
          --wired-toggle-on-color: var(
            --simple-colors-fixed-theme-light-blue-9
          );
          background-position-x: 60px;
          background-position-y: 8px;
          width: 100px;
          display: inline-flex;
        }
        :host([checked]) div {
          background-image: url("${n(d)}");
          background-position-x: 12px;
          background-position-y: 12px;
          background-size: 22px, 22px;
        }
        :host([disabled]) {
          opacity: 0.5;
          pointer-events: none;
          cursor: not-allowed;
        }
        input {
          width: 100px;
          height: 48px;
          padding: 0;
          margin: 0;
        }
        label {
          clip: rect(0 0 0 0);
          clip-path: inset(50%);
          height: 1px;
          overflow: hidden;
          position: absolute;
          white-space: nowrap;
          width: 1px;
        }
        /*img {
          top: 8px;
          max-width: 20
          height: 35;
        }*/
      `]}}globalThis.customElements.define(WiredDarkmodeToggle.tag,WiredDarkmodeToggle);
