// @haxtheweb/page-break/lib/page-template.js
/**
 * Copyright 2025 The Pennsylvania State University
 * @license Apache-2.0, see License.md for full text.
 */
import{html as e,css as t}from"../../../lit/index.js";import{SchemaBehaviors as a}from"../../schema-behaviors/schema-behaviors.js";import{I18NMixin as o}from"../../i18n-manager/lib/I18NMixin.js";import{DDD as i}from"../../d-d-d/d-d-d.js";import{generateResourceID as d}from"../../utils/utils.js";export class PageTemplate extends(o(a(i))){static get tag(){return"page-template"}static get styles(){return[super.styles,t`
        :host {
          display: block;
          margin: var(--ddd-spacing-4) 0;
          padding: var(--ddd-spacing-4);
          border: 1px dotted var(--ddd-theme-default-limestoneGray);
          border-radius: var(--ddd-radius-xs);
          background-color: transparent;
          position: relative;
          transition: border-color 0.2s ease-in-out;
          min-height: 100px;
        }

        :host(:hover) {
          border-color: var(--ddd-theme-default-coalyGray);
        }


        :host([data-hax-ray])
          .template-content
          ::slotted(*.hax-hovered)::before {
          content: " ";
          width: calc(100% + 32px);
          display: block;
          position: relative;
          margin: -28px -16px 12px;
          z-index: 2;
          height: 12px;
          border: none !important;
          transition: 0.3s all ease-in-out;
          background-color: var(--hax-body-target-background-color) !important;
        }

        .template-label {
          position: absolute;
          top: calc(-1 * var(--ddd-spacing-2));
          left: var(--ddd-spacing-2);
          background-color: var(--ddd-theme-default-skyBlue);
          color: var(--ddd-theme-default-white);
          padding: var(--ddd-spacing-1) var(--ddd-spacing-2);
          font-size: var(--ddd-font-size-xs);
          font-weight: var(--ddd-font-weight-medium);
          border-radius: var(--ddd-radius-xs);
          z-index: 10;
        }

        .template-header {
          margin-bottom: var(--ddd-spacing-4);
          padding-bottom: var(--ddd-spacing-2);
          border-bottom: 1px solid var(--ddd-theme-default-limestoneGray);
        }

        .template-name {
          font-size: var(--ddd-font-size-s);
          font-weight: var(--ddd-font-weight-medium);
          color: var(--ddd-theme-default-coalyGray);
          margin: 0;
        }

        .template-content {
          position: relative;
        }

        :host(:not([name])) .template-header {
          display: none;
        }
      `]}static get properties(){return{...super.properties,name:{type:String,reflect:!0},schema:{type:String,reflect:!0}}}constructor(){super(),this.name="",this.schema="area",this.t={pageTemplate:"Page Template",templateId:"Template ID",defaultTemplate:"Default Template",namedTemplate:"Named Template"},this.registerLocalization({context:this,localesPath:new URL("../../locales/page-break.es.json",import.meta.url).href+"/../"})}connectedCallback(){if(super.connectedCallback(),!this.getAttribute("data-haxsg-id")){const e=this.schema||"area",t=`${"page"===e?"page":"block"===e?"block":"template"}-${d("").substring(1)}`;this.setAttribute("data-haxsg-id",t)}}render(){return e`
      ${this.name?e` <div class="template-label">${this.name}</div> `:""}
      <div class="template-content">
        <slot></slot>
      </div>
    `}static get haxProperties(){return{type:"grid",canScale:!1,canPosition:!1,canEditSource:!0,contentEditable:!0,gizmo:{title:"Page Template",description:"A template component for defining reusable page layouts and styling",icon:"hax:templates",color:"blue",groups:["Other"],handles:[],meta:{author:"HAXTheWeb"}},settings:{configure:[{property:"name",title:"Template Name",description:"Name you will see in the editor",inputMethod:"textfield",required:!1},{property:"schema",title:"Template Type",description:"How this template should be categorized and used",inputMethod:"select",options:{block:"Block - Replaces element defaults",area:"Area - Shows in Templates section",page:"Page - Shows in Pages section"},required:!0}],advanced:[{slot:"",title:"Template Content",description:"The elements and content that define this template section",inputMethod:"textarea",required:!1}]},demoSchema:[{tag:"page-template",properties:{name:"Example Template",schema:"area"},content:"<h2>Template Heading</h2><p>Add your content elements inside this template. This acts as a container that can hold any HAX elements.</p>"}]}}}globalThis.customElements.define(PageTemplate.tag,PageTemplate);
