// @lrnwebcomponents/person-testimonial/person-testimonial.js
define(["exports", "require", "../../@polymer/polymer/polymer-element.js", "../simple-colors/lib/simple-colors-polymer.js"], function (_exports, _require, _polymerElement, _simpleColorsPolymer) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.PersonTestimonial = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject_33d9a110db3d11eaa14b6130b7e5d77f() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style include=\"simple-colors-shared-styles-polymer\">\n        :host {\n          display: block;\n          --person-testimonial-font-family: sans-serif;\n          --person-testimonial-bg: var(--simple-colors-default-theme-grey-1);\n          --person-testimonial-color: var(\n            --simple-colors-default-theme-accent-7\n          );\n          --person-testimonial-text: var(--simple-colors-default-theme-grey-12);\n        }\n\n        paper-card {\n          display: inline-flex;\n          background-color: var(--person-testimonial-bg);\n          color: var(--person-testimonial-text);\n          font-family: var(--person-testimonial-font-family);\n        }\n\n        iron-image {\n          display: block;\n          width: 150px;\n          height: 100%;\n        }\n        .image {\n          padding-right: 5px;\n          background-color: var(--person-testimonial-color);\n        }\n\n        iron-icon {\n          --iron-icon-height: 24px;\n          --iron-icon-width: 24px;\n          --iron-icon-fill-color: var(--person-testimonial-color);\n        }\n\n        .wrap {\n          margin: 15px;\n        }\n\n        .testimonial {\n          line-height: 24px;\n          font-size: 16px;\n          font-style: italic;\n        }\n\n        .name {\n          font-size: 21px;\n          text-transform: uppercase;\n          font-weight: bold;\n          margin-top: 20px;\n        }\n\n        .position {\n          font-size: 14px;\n          margin-top: 5px;\n        }\n\n        .arrow_right {\n          width: 0;\n          height: 0;\n          border-top: 15px solid var(--person-testimonial-bg);\n          border-bottom: 15px solid var(--person-testimonial-bg);\n          border-left: solid 15px transparent;\n          background-color: var(--person-testimonial-color);\n          position: relative;\n          top: 55px;\n        }\n\n        #quotestart {\n          display: inline-flex;\n          transform: rotateY(180deg);\n        }\n\n        #quoteend {\n          display: inline-flex;\n        }\n\n        @media screen and (max-width: 850px) {\n          paper-card {\n            display: flex;\n            flex-wrap: wrap;\n          }\n          iron-image {\n            display: block;\n            border-radius: 50%;\n            width: 200px;\n            height: 200px;\n          }\n          .image {\n            margin-top: 25px;\n            border-radius: 50%;\n            padding: 5px;\n            margin-left: auto;\n            margin-right: auto;\n          }\n          .arrow_right {\n            display: none;\n          }\n          .name,\n          .position {\n            text-align: center;\n          }\n        }\n      </style>\n      <paper-card elevation=\"[[elevation]]\">\n        <div class=\"image\">\n          <iron-image\n            src=\"[[image]]\"\n            sizing=\"cover\"\n            preload=\"\"\n            fade=\"\"\n            aria-describedby$=\"[[describedBy]]\"\n          ></iron-image>\n        </div>\n        <div class=\"arrow_right\"></div>\n        <div class=\"wrap\">\n          <div class=\"testimonial\">\n            <iron-icon\n              id=\"quotestart\"\n              icon=\"persontestimonial:format-quote\"\n            ></iron-icon>\n            <slot></slot>\n            <iron-icon\n              id=\"quoteend\"\n              icon=\"persontestimonial:format-quote\"\n            ></iron-icon>\n          </div>\n          <div class=\"name\">[[name]]</div>\n          <div class=\"position\">[[position]]</div>\n        </div>\n      </paper-card>\n    "]);

    _templateObject_33d9a110db3d11eaa14b6130b7e5d77f = function _templateObject_33d9a110db3d11eaa14b6130b7e5d77f() {
      return data;
    };

    return data;
  }

  /**
   * `person-testimonial`
   * @element person-testimonial
   * `Leaving a testimonial from a person to say your company rocks!`
   * @demo demo/index.html
   */
  var PersonTestimonial =
  /*#__PURE__*/
  function (_SimpleColorsPolymer) {
    babelHelpers.inherits(PersonTestimonial, _SimpleColorsPolymer);

    function PersonTestimonial() {
      var _this;

      babelHelpers.classCallCheck(this, PersonTestimonial);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(PersonTestimonial).call(this));
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/paper-card/paper-card.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/iron-image/iron-image.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/iron-icon/iron-icon.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["./lib/person-testimonial-icon.js"], res, rej);
      });
      return _this;
    }

    babelHelpers.createClass(PersonTestimonial, null, [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_33d9a110db3d11eaa14b6130b7e5d77f());
      }
    }, {
      key: "tag",
      get: function get() {
        return "person-testimonial";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(PersonTestimonial), "properties", this), {
          /**
           * Visual height of the card.
           */
          elevation: {
            type: Number,
            value: 1,
            reflectToAttribute: true
          },

          /**
           * Aria-describedby data passed down to appropriate tag
           */
          describedBy: {
            type: String,
            attribute: "described-by"
          },

          /**
           * The profile image to display to the left of the quote.
           */
          image: {
            type: String
          },

          /**
           * Name of the person making the quote.
           */
          name: {
            type: String
          },

          /**
           * The title / position of the person in question.
           */
          position: {
            type: String
          }
        });
      }
    }, {
      key: "haxProperties",
      get: function get() {
        return {
          canScale: true,
          canPosition: true,
          canEditSource: true,
          gizmo: {
            title: "Testimonial",
            description: "A person saying a nice thing about us",
            icon: "editor:format-quote",
            color: "orange",
            groups: ["Content", "Presentation"],
            handles: [{
              type: "image",
              source: "image",
              title: "name",
              caption: "position",
              ariaDescribedby: "describedBy"
            }],
            meta: {
              author: "EberlyODL / LRNWebComponents"
            }
          },
          settings: {
            quick: [{
              property: "image",
              title: "Image",
              description: "Adds image to this testimonial",
              inputMethod: "textfield",
              icon: "editor:insert-photo"
            }, {
              property: "name",
              title: "Full Name",
              description: "Credit the person making the testimonial",
              inputMethod: "textfield",
              icon: "account-circle"
            }, {
              property: "position",
              title: "Position or Job Title",
              description: "List the position and job title",
              inputMethod: "textfield",
              icon: "icons:work"
            }, {
              property: "accentColor",
              title: "Accent color",
              description: "Select the accent color use",
              inputMethod: "colorpicker",
              icon: "editor:format-color-fill"
            }, {
              property: "dark",
              title: "Dark",
              description: "Use dark theme",
              inputMethod: "boolean",
              icon: "invert-colors"
            }],
            configure: [{
              property: "image",
              title: "Image",
              description: "Adds image to testimonial",
              inputMethod: "haxupload",
              icon: "editor:insert-photo"
            }, {
              property: "accentColor",
              title: "Accent color",
              description: "Select the accent color use",
              inputMethod: "colorpicker",
              icon: "editor:format-color-fill"
            }, {
              property: "dark",
              title: "Dark",
              description: "Use dark theme",
              inputMethod: "boolean",
              icon: "invert-colors"
            }, {
              slot: "",
              title: "User's testimonial:",
              description: "This is where you enter your testimonial.",
              inputMethod: "code-editor",
              required: true
            }, {
              property: "name",
              title: "Full Name",
              description: "Credit the person making the testimonial",
              inputMethod: "textfield",
              icon: "account-circle"
            }, {
              property: "position",
              title: "Position or Job Title",
              description: "List the position and job title",
              inputMethod: "textfield",
              icon: "icons:work"
            }],
            advanced: [{
              property: "describedBy",
              title: "aria-describedby",
              description: "Space-separated list of IDs for elements that describe the image.",
              inputMethod: "textfield"
            }]
          }
        };
      }
    }]);
    return PersonTestimonial;
  }(_simpleColorsPolymer.SimpleColorsPolymer);

  _exports.PersonTestimonial = PersonTestimonial;
  window.customElements.define(PersonTestimonial.tag, PersonTestimonial);
});
