// @lrnwebcomponents/lrndesign-chart/lrndesign-chart.js
define(["exports", "../../lit-element/lit-element.js", "../simple-colors/simple-colors.js", "../chartist-render/chartist-render.js", "../../@polymer/iron-ajax/iron-ajax.js"], function (_exports, _litElement, _simpleColors, _chartistRender, _ironAjax) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrndesignChart = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject2_5da312d02d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          --chartist-color-1: var(--simple-colors-default-theme-red-8);\n          --chartist-color-2: var(--simple-colors-default-theme-blue-8);\n          --chartist-color-3: var(--simple-colors-default-theme-yellow-8);\n          --chartist-color-4: var(--simple-colors-default-theme-purple-8);\n          --chartist-color-5: var(--simple-colors-default-theme-green-8);\n          --chartist-color-6: var(--simple-colors-default-theme-orange-7);\n          --chartist-color-7: var(--simple-colors-default-theme-pink-8);\n          --chartist-color-8: var(--simple-colors-default-theme-deep-orange-8);\n          --chartist-color-9: var(--simple-colors-default-theme-red-9);\n          --chartist-color-10: var(--simple-colors-default-theme-blue-9);\n          --chartist-color-11: var(--simple-colors-default-theme-yellow-9);\n          --chartist-color-12: var(--simple-colors-default-theme-purple-9);\n          --chartist-color-13: var(--simple-colors-default-theme-green-9);\n          --chartist-color-14: var(--simple-colors-default-theme-orange-8);\n          --chartist-color-15: var(--simple-colors-default-theme-pink-9);\n        }\n\n        :host([dark]) {\n          --chartist-color-1: var(--simple-colors-default-theme-red-4);\n          --chartist-color-2: var(--simple-colors-default-theme-blue-4);\n          --chartist-color-3: var(--simple-colors-default-theme-yellow-4);\n          --chartist-color-4: var(--simple-colors-default-theme-purple-4);\n          --chartist-color-5: var(--simple-colors-default-theme-green-4);\n          --chartist-color-6: var(--simple-colors-default-theme-orange-5);\n          --chartist-color-7: var(--simple-colors-default-theme-pink-4);\n          --chartist-color-8: var(--simple-colors-default-theme-deep-orange-4);\n          --chartist-color-9: var(--simple-colors-default-theme-red-3);\n          --chartist-color-10: var(--simple-colors-default-theme-blue-3);\n          --chartist-color-11: var(--simple-colors-default-theme-yellow-3);\n          --chartist-color-12: var(--simple-colors-default-theme-purple-3);\n          --chartist-color-13: var(--simple-colors-default-theme-green-3);\n          --chartist-color-14: var(--simple-colors-default-theme-orange-4);\n          --chartist-color-15: var(--simple-colors-default-theme-pink-3);\n        }\n\n        :host([accent-color=\"red\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-red-8);\n          --chartist-color-2: var(--simple-colors-default-theme-pink-9);\n          --chartist-color-3: var(--simple-colors-default-theme-deep-orange-10);\n          --chartist-color-4: var(--simple-colors-default-theme-purple-8);\n          --chartist-color-5: var(--simple-colors-default-theme-orange-9);\n          --chartist-color-6: var(--simple-colors-default-theme-red-10);\n          --chartist-color-7: var(--simple-colors-default-theme-pink-8);\n          --chartist-color-8: var(--simple-colors-default-theme-deep-orange-9);\n          --chartist-color-9: var(--simple-colors-default-theme-purple-10);\n          --chartist-color-10: var(--simple-colors-default-theme-orange-8);\n          --chartist-color-11: var(--simple-colors-default-theme-red-9);\n          --chartist-color-12: var(--simple-colors-default-theme-pink-10);\n          --chartist-color-13: var(--simple-colors-default-theme-deep-orange-8);\n          --chartist-color-14: var(--simple-colors-default-theme-purple-9);\n          --chartist-color-15: var(--simple-colors-default-theme-orange-10);\n        }\n\n        :host([dark][accent-color=\"red\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-red-4);\n          --chartist-color-2: var(--simple-colors-default-theme-pink-3);\n          --chartist-color-3: var(--simple-colors-default-theme-deep-orange-2);\n          --chartist-color-4: var(--simple-colors-default-theme-purple-4);\n          --chartist-color-5: var(--simple-colors-default-theme-orange-3);\n          --chartist-color-6: var(--simple-colors-default-theme-red-2);\n          --chartist-color-7: var(--simple-colors-default-theme-pink-4);\n          --chartist-color-8: var(--simple-colors-default-theme-deep-orange-3);\n          --chartist-color-9: var(--simple-colors-default-theme-purple-2);\n          --chartist-color-10: var(--simple-colors-default-theme-orange-4);\n          --chartist-color-11: var(--simple-colors-default-theme-red-3);\n          --chartist-color-12: var(--simple-colors-default-theme-pink-2);\n          --chartist-color-13: var(--simple-colors-default-theme-deep-orange-4);\n          --chartist-color-14: var(--simple-colors-default-theme-purple-3);\n          --chartist-color-15: var(--simple-colors-default-theme-orange-2);\n        }\n\n        :host([accent-color=\"pink\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-pink-8);\n          --chartist-color-2: var(--simple-colors-default-theme-purple-9);\n          --chartist-color-3: var(--simple-colors-default-theme-red-10);\n          --chartist-color-4: var(--simple-colors-default-theme-purple-8);\n          --chartist-color-5: var(--simple-colors-default-theme-deep-orange-9);\n          --chartist-color-6: var(--simple-colors-default-theme-pink-10);\n          --chartist-color-7: var(--simple-colors-default-theme-purple-8);\n          --chartist-color-8: var(--simple-colors-default-theme-red-9);\n          --chartist-color-9: var(--simple-colors-default-theme-purple-10);\n          --chartist-color-10: var(--simple-colors-default-theme-deep-orange-8);\n          --chartist-color-11: var(--simple-colors-default-theme-pink-9);\n          --chartist-color-12: var(--simple-colors-default-theme-purple-10);\n          --chartist-color-13: var(--simple-colors-default-theme-red-8);\n          --chartist-color-14: var(--simple-colors-default-theme-purple-9);\n          --chartist-color-15: var(\n            --simple-colors-default-theme-deep-orange-10\n          );\n        }\n\n        :host([dark][accent-color=\"pink\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-pink-4);\n          --chartist-color-2: var(--simple-colors-default-theme-purple-3);\n          --chartist-color-3: var(--simple-colors-default-theme-red-2);\n          --chartist-color-4: var(--simple-colors-default-theme-purple-4);\n          --chartist-color-5: var(--simple-colors-default-theme-deep-orange-3);\n          --chartist-color-6: var(--simple-colors-default-theme-pink-2);\n          --chartist-color-7: var(--simple-colors-default-theme-purple-4);\n          --chartist-color-8: var(--simple-colors-default-theme-red-3);\n          --chartist-color-9: var(--simple-colors-default-theme-purple-2);\n          --chartist-color-10: var(--simple-colors-default-theme-deep-orange-4);\n          --chartist-color-11: var(--simple-colors-default-theme-pink-3);\n          --chartist-color-12: var(--simple-colors-default-theme-purple-2);\n          --chartist-color-13: var(--simple-colors-default-theme-red-4);\n          --chartist-color-14: var(--simple-colors-default-theme-purple-3);\n          --chartist-color-15: var(--simple-colors-default-theme-deep-orange-2);\n        }\n\n        :host([accent-color=\"purple\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-purple-8);\n          --chartist-color-2: var(--simple-colors-default-theme-deep-purple-9);\n          --chartist-color-3: var(--simple-colors-default-theme-pink-10);\n          --chartist-color-4: var(--simple-colors-default-theme-indigo-8);\n          --chartist-color-5: var(--simple-colors-default-theme-red-9);\n          --chartist-color-6: var(--simple-colors-default-theme-purple-10);\n          --chartist-color-7: var(--simple-colors-default-theme-deep-purple-8);\n          --chartist-color-8: var(--simple-colors-default-theme-pink-9);\n          --chartist-color-9: var(--simple-colors-default-theme-indigo-10);\n          --chartist-color-10: var(--simple-colors-default-theme-red-8);\n          --chartist-color-11: var(--simple-colors-default-theme-purple-9);\n          --chartist-color-12: var(\n            --simple-colors-default-theme-deep-purple-10\n          );\n          --chartist-color-13: var(--simple-colors-default-theme-pink-8);\n          --chartist-color-14: var(--simple-colors-default-theme-indigo-9);\n          --chartist-color-15: var(--simple-colors-default-theme-red-10);\n        }\n\n        :host([dark][accent-color=\"purple\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-purple-4);\n          --chartist-color-2: var(--simple-colors-default-theme-deep-purple-3);\n          --chartist-color-3: var(--simple-colors-default-theme-pink-2);\n          --chartist-color-4: var(--simple-colors-default-theme-indigo-4);\n          --chartist-color-5: var(--simple-colors-default-theme-red-3);\n          --chartist-color-6: var(--simple-colors-default-theme-purple-2);\n          --chartist-color-7: var(--simple-colors-default-theme-deep-purple-4);\n          --chartist-color-8: var(--simple-colors-default-theme-pink-3);\n          --chartist-color-9: var(--simple-colors-default-theme-indigo-2);\n          --chartist-color-10: var(--simple-colors-default-theme-red-4);\n          --chartist-color-11: var(--simple-colors-default-theme-purple-3);\n          --chartist-color-12: var(--simple-colors-default-theme-deep-purple-2);\n          --chartist-color-13: var(--simple-colors-default-theme-pink-4);\n          --chartist-color-14: var(--simple-colors-default-theme-indigo-3);\n          --chartist-color-15: var(--simple-colors-default-theme-red-2);\n        }\n\n        :host([accent-color=\"deep-purple\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-deep-purple-8);\n          --chartist-color-2: var(--simple-colors-default-theme-indigo-9);\n          --chartist-color-3: var(--simple-colors-default-theme-purple-10);\n          --chartist-color-4: var(--simple-colors-default-theme-blue-8);\n          --chartist-color-5: var(--simple-colors-default-theme-pink-9);\n          --chartist-color-6: var(--simple-colors-default-theme-deep-purple-10);\n          --chartist-color-7: var(--simple-colors-default-theme-indigo-8);\n          --chartist-color-8: var(--simple-colors-default-theme-purple-9);\n          --chartist-color-9: var(--simple-colors-default-theme-blue-10);\n          --chartist-color-10: var(--simple-colors-default-theme-pink-8);\n          --chartist-color-11: var(--simple-colors-default-theme-deep-purple-9);\n          --chartist-color-12: var(--simple-colors-default-theme-indigo-10);\n          --chartist-color-13: var(--simple-colors-default-theme-purple-8);\n          --chartist-color-14: var(--simple-colors-default-theme-blue-9);\n          --chartist-color-15: var(--simple-colors-default-theme-pink-10);\n        }\n\n        :host([dark][accent-color=\"deep-purple\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-deep-purple-4);\n          --chartist-color-2: var(--simple-colors-default-theme-indigo-3);\n          --chartist-color-3: var(--simple-colors-default-theme-purple-2);\n          --chartist-color-4: var(--simple-colors-default-theme-blue-4);\n          --chartist-color-5: var(--simple-colors-default-theme-pink-3);\n          --chartist-color-6: var(--simple-colors-default-theme-deep-purple-2);\n          --chartist-color-7: var(--simple-colors-default-theme-indigo-4);\n          --chartist-color-8: var(--simple-colors-default-theme-purple-3);\n          --chartist-color-9: var(--simple-colors-default-theme-blue-2);\n          --chartist-color-10: var(--simple-colors-default-theme-pink-4);\n          --chartist-color-11: var(--simple-colors-default-theme-deep-purple-3);\n          --chartist-color-12: var(--simple-colors-default-theme-indigo-2);\n          --chartist-color-13: var(--simple-colors-default-theme-purple-4);\n          --chartist-color-14: var(--simple-colors-default-theme-blue-3);\n          --chartist-color-15: var(--simple-colors-default-theme-pink-2);\n        }\n\n        :host([accent-color=\"indigo\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-indigo-8);\n          --chartist-color-2: var(--simple-colors-default-theme-blue-9);\n          --chartist-color-3: var(--simple-colors-default-theme-deep-purple-10);\n          --chartist-color-4: var(--simple-colors-default-theme-light-blue-8);\n          --chartist-color-5: var(--simple-colors-default-theme-purple-9);\n          --chartist-color-6: var(--simple-colors-default-theme-indigo-10);\n          --chartist-color-7: var(--simple-colors-default-theme-blue-8);\n          --chartist-color-8: var(--simple-colors-default-theme-deep-purple-9);\n          --chartist-color-9: var(--simple-colors-default-theme-light-blue-10);\n          --chartist-color-10: var(--simple-colors-default-theme-purple-8);\n          --chartist-color-11: var(--simple-colors-default-theme-indigo-9);\n          --chartist-color-12: var(--simple-colors-default-theme-blue-10);\n          --chartist-color-13: var(--simple-colors-default-theme-deep-purple-8);\n          --chartist-color-14: var(--simple-colors-default-theme-light-blue-9);\n          --chartist-color-15: var(--simple-colors-default-theme-purple-10);\n        }\n\n        :host([dark][accent-color=\"indigo\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-indigo-4);\n          --chartist-color-2: var(--simple-colors-default-theme-blue-3);\n          --chartist-color-3: var(--simple-colors-default-theme-deep-purple-2);\n          --chartist-color-4: var(--simple-colors-default-theme-light-blue-4);\n          --chartist-color-5: var(--simple-colors-default-theme-purple-3);\n          --chartist-color-6: var(--simple-colors-default-theme-indigo-2);\n          --chartist-color-7: var(--simple-colors-default-theme-blue-4);\n          --chartist-color-8: var(--simple-colors-default-theme-deep-purple-3);\n          --chartist-color-9: var(--simple-colors-default-theme-light-blue-2);\n          --chartist-color-10: var(--simple-colors-default-theme-purple-4);\n          --chartist-color-11: var(--simple-colors-default-theme-indigo-3);\n          --chartist-color-12: var(--simple-colors-default-theme-blue-2);\n          --chartist-color-13: var(--simple-colors-default-theme-deep-purple-4);\n          --chartist-color-14: var(--simple-colors-default-theme-light-blue-3);\n          --chartist-color-15: var(--simple-colors-default-theme-purple-2);\n        }\n\n        :host([accent-color=\"blue\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-blue-8);\n          --chartist-color-2: var(--simple-colors-default-theme-light-blue-9);\n          --chartist-color-3: var(--simple-colors-default-theme-indigo-10);\n          --chartist-color-4: var(--simple-colors-default-theme-cyan-8);\n          --chartist-color-5: var(--simple-colors-default-theme-deep-purple-9);\n          --chartist-color-6: var(--simple-colors-default-theme-blue-10);\n          --chartist-color-7: var(--simple-colors-default-theme-light-blue-8);\n          --chartist-color-8: var(--simple-colors-default-theme-indigo-9);\n          --chartist-color-9: var(--simple-colors-default-theme-cyan-10);\n          --chartist-color-10: var(--simple-colors-default-theme-deep-purple-8);\n          --chartist-color-11: var(--simple-colors-default-theme-blue-9);\n          --chartist-color-12: var(--simple-colors-default-theme-light-blue-10);\n          --chartist-color-13: var(--simple-colors-default-theme-indigo-8);\n          --chartist-color-14: var(--simple-colors-default-theme-cyan-9);\n          --chartist-color-15: var(\n            --simple-colors-default-theme-deep-purple-10\n          );\n        }\n\n        :host([dark][accent-color=\"blue\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-blue-4);\n          --chartist-color-2: var(--simple-colors-default-theme-light-blue-3);\n          --chartist-color-3: var(--simple-colors-default-theme-indigo-2);\n          --chartist-color-4: var(--simple-colors-default-theme-cyan-4);\n          --chartist-color-5: var(--simple-colors-default-theme-deep-purple-3);\n          --chartist-color-6: var(--simple-colors-default-theme-blue-2);\n          --chartist-color-7: var(--simple-colors-default-theme-light-blue-4);\n          --chartist-color-8: var(--simple-colors-default-theme-indigo-3);\n          --chartist-color-9: var(--simple-colors-default-theme-cyan-2);\n          --chartist-color-10: var(--simple-colors-default-theme-deep-purple-4);\n          --chartist-color-11: var(--simple-colors-default-theme-blue-3);\n          --chartist-color-12: var(--simple-colors-default-theme-light-blue-2);\n          --chartist-color-13: var(--simple-colors-default-theme-indigo-4);\n          --chartist-color-14: var(--simple-colors-default-theme-cyan-3);\n          --chartist-color-15: var(--simple-colors-default-theme-deep-purple-2);\n        }\n\n        :host([accent-color=\"light-blue\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-light-blue-8);\n          --chartist-color-2: var(--simple-colors-default-theme-cyan-9);\n          --chartist-color-3: var(--simple-colors-default-theme-blue-10);\n          --chartist-color-4: var(--simple-colors-default-theme-teal-8);\n          --chartist-color-5: var(--simple-colors-default-theme-indigo-9);\n          --chartist-color-6: var(--simple-colors-default-theme-light-blue-10);\n          --chartist-color-7: var(--simple-colors-default-theme-cyan-8);\n          --chartist-color-8: var(--simple-colors-default-theme-blue-9);\n          --chartist-color-9: var(--simple-colors-default-theme-teal-10);\n          --chartist-color-10: var(--simple-colors-default-theme-indigo-8);\n          --chartist-color-11: var(--simple-colors-default-theme-light-blue-9);\n          --chartist-color-12: var(--simple-colors-default-theme-cyan-10);\n          --chartist-color-13: var(--simple-colors-default-theme-blue-8);\n          --chartist-color-14: var(--simple-colors-default-theme-teal-9);\n          --chartist-color-15: var(--simple-colors-default-theme-indigo-10);\n        }\n\n        :host([dark][accent-color=\"light-blue\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-light-blue-4);\n          --chartist-color-2: var(--simple-colors-default-theme-cyan-3);\n          --chartist-color-3: var(--simple-colors-default-theme-blue-2);\n          --chartist-color-4: var(--simple-colors-default-theme-teal-4);\n          --chartist-color-5: var(--simple-colors-default-theme-indigo-3);\n          --chartist-color-6: var(--simple-colors-default-theme-light-blue-2);\n          --chartist-color-7: var(--simple-colors-default-theme-cyan-4);\n          --chartist-color-8: var(--simple-colors-default-theme-blue-3);\n          --chartist-color-9: var(--simple-colors-default-theme-teal-2);\n          --chartist-color-10: var(--simple-colors-default-theme-indigo-4);\n          --chartist-color-11: var(--simple-colors-default-theme-light-blue-3);\n          --chartist-color-12: var(--simple-colors-default-theme-cyan-2);\n          --chartist-color-13: var(--simple-colors-default-theme-blue-4);\n          --chartist-color-14: var(--simple-colors-default-theme-teal-3);\n          --chartist-color-15: var(--simple-colors-default-theme-indigo-2);\n        }\n\n        :host([accent-color=\"cyan\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-cyan-8);\n          --chartist-color-2: var(--simple-colors-default-theme-teal-9);\n          --chartist-color-3: var(--simple-colors-default-theme-light-blue-10);\n          --chartist-color-4: var(--simple-colors-default-theme-green-8);\n          --chartist-color-5: var(--simple-colors-default-theme-blue-9);\n          --chartist-color-6: var(--simple-colors-default-theme-cyan-10);\n          --chartist-color-7: var(--simple-colors-default-theme-teal-8);\n          --chartist-color-8: var(--simple-colors-default-theme-light-blue-9);\n          --chartist-color-9: var(--simple-colors-default-theme-green-10);\n          --chartist-color-10: var(--simple-colors-default-theme-blue-8);\n          --chartist-color-11: var(--simple-colors-default-theme-cyan-9);\n          --chartist-color-12: var(--simple-colors-default-theme-teal-10);\n          --chartist-color-13: var(--simple-colors-default-theme-light-blue-8);\n          --chartist-color-14: var(--simple-colors-default-theme-green-9);\n          --chartist-color-15: var(--simple-colors-default-theme-blue-10);\n        }\n\n        :host([dark][accent-color=\"cyan\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-cyan-4);\n          --chartist-color-2: var(--simple-colors-default-theme-teal-3);\n          --chartist-color-3: var(--simple-colors-default-theme-light-blue-2);\n          --chartist-color-4: var(--simple-colors-default-theme-green-4);\n          --chartist-color-5: var(--simple-colors-default-theme-blue-3);\n          --chartist-color-6: var(--simple-colors-default-theme-cyan-2);\n          --chartist-color-7: var(--simple-colors-default-theme-teal-4);\n          --chartist-color-8: var(--simple-colors-default-theme-light-blue-3);\n          --chartist-color-9: var(--simple-colors-default-theme-green-2);\n          --chartist-color-10: var(--simple-colors-default-theme-blue-4);\n          --chartist-color-11: var(--simple-colors-default-theme-cyan-3);\n          --chartist-color-12: var(--simple-colors-default-theme-teal-2);\n          --chartist-color-13: var(--simple-colors-default-theme-light-blue-4);\n          --chartist-color-14: var(--simple-colors-default-theme-green-3);\n          --chartist-color-15: var(--simple-colors-default-theme-blue-2);\n        }\n\n        :host([accent-color=\"teal\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-teal-8);\n          --chartist-color-2: var(--simple-colors-default-theme-green-9);\n          --chartist-color-3: var(--simple-colors-default-theme-cyan-10);\n          --chartist-color-4: var(--simple-colors-default-theme-light-green-8);\n          --chartist-color-5: var(--simple-colors-default-theme-light-blue-9);\n          --chartist-color-6: var(--simple-colors-default-theme-teal-10);\n          --chartist-color-7: var(--simple-colors-default-theme-green-8);\n          --chartist-color-8: var(--simple-colors-default-theme-cyan-9);\n          --chartist-color-9: var(--simple-colors-default-theme-light-green-10);\n          --chartist-color-10: var(--simple-colors-default-theme-light-blue-8);\n          --chartist-color-11: var(--simple-colors-default-theme-teal-9);\n          --chartist-color-12: var(--simple-colors-default-theme-green-10);\n          --chartist-color-13: var(--simple-colors-default-theme-cyan-8);\n          --chartist-color-14: var(--simple-colors-default-theme-light-green-9);\n          --chartist-color-15: var(--simple-colors-default-theme-light-blue-10);\n        }\n\n        :host([dark][accent-color=\"teal\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-teal-4);\n          --chartist-color-2: var(--simple-colors-default-theme-green-3);\n          --chartist-color-3: var(--simple-colors-default-theme-cyan-2);\n          --chartist-color-4: var(--simple-colors-default-theme-light-green-4);\n          --chartist-color-5: var(--simple-colors-default-theme-light-blue-3);\n          --chartist-color-6: var(--simple-colors-default-theme-teal-2);\n          --chartist-color-7: var(--simple-colors-default-theme-green-4);\n          --chartist-color-8: var(--simple-colors-default-theme-cyan-3);\n          --chartist-color-9: var(--simple-colors-default-theme-light-green-2);\n          --chartist-color-10: var(--simple-colors-default-theme-light-blue-4);\n          --chartist-color-11: var(--simple-colors-default-theme-teal-3);\n          --chartist-color-12: var(--simple-colors-default-theme-green-2);\n          --chartist-color-13: var(--simple-colors-default-theme-cyan-4);\n          --chartist-color-14: var(--simple-colors-default-theme-light-green-3);\n          --chartist-color-15: var(--simple-colors-default-theme-light-blue-2);\n        }\n\n        :host([accent-color=\"green\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-green-8);\n          --chartist-color-2: var(--simple-colors-default-theme-light-green-9);\n          --chartist-color-3: var(--simple-colors-default-theme-teal-10);\n          --chartist-color-4: var(--simple-colors-default-theme-lime-8);\n          --chartist-color-5: var(--simple-colors-default-theme-cyan-9);\n          --chartist-color-6: var(--simple-colors-default-theme-green-10);\n          --chartist-color-7: var(--simple-colors-default-theme-light-green-8);\n          --chartist-color-8: var(--simple-colors-default-theme-teal-9);\n          --chartist-color-9: var(--simple-colors-default-theme-lime-10);\n          --chartist-color-10: var(--simple-colors-default-theme-cyan-8);\n          --chartist-color-11: var(--simple-colors-default-theme-green-9);\n          --chartist-color-12: var(\n            --simple-colors-default-theme-light-green-10\n          );\n          --chartist-color-13: var(--simple-colors-default-theme-teal-8);\n          --chartist-color-14: var(--simple-colors-default-theme-lime-9);\n          --chartist-color-15: var(--simple-colors-default-theme-cyan-10);\n        }\n\n        :host([dark][accent-color=\"green\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-green-4);\n          --chartist-color-2: var(--simple-colors-default-theme-light-green-3);\n          --chartist-color-3: var(--simple-colors-default-theme-teal-2);\n          --chartist-color-4: var(--simple-colors-default-theme-lime-4);\n          --chartist-color-5: var(--simple-colors-default-theme-cyan-3);\n          --chartist-color-6: var(--simple-colors-default-theme-green-2);\n          --chartist-color-7: var(--simple-colors-default-theme-light-green-4);\n          --chartist-color-8: var(--simple-colors-default-theme-teal-3);\n          --chartist-color-9: var(--simple-colors-default-theme-lime-2);\n          --chartist-color-10: var(--simple-colors-default-theme-cyan-4);\n          --chartist-color-11: var(--simple-colors-default-theme-green-3);\n          --chartist-color-12: var(--simple-colors-default-theme-light-green-2);\n          --chartist-color-13: var(--simple-colors-default-theme-teal-4);\n          --chartist-color-14: var(--simple-colors-default-theme-lime-3);\n          --chartist-color-15: var(--simple-colors-default-theme-cyan-2);\n        }\n\n        :host([accent-color=\"light-green\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-light-green-8);\n          --chartist-color-2: var(--simple-colors-default-theme-lime-9);\n          --chartist-color-3: var(--simple-colors-default-theme-green-10);\n          --chartist-color-4: var(--simple-colors-default-theme-amber-8);\n          --chartist-color-5: var(--simple-colors-default-theme-teal-9);\n          --chartist-color-6: var(--simple-colors-default-theme-light-green-10);\n          --chartist-color-7: var(--simple-colors-default-theme-lime-8);\n          --chartist-color-8: var(--simple-colors-default-theme-green-9);\n          --chartist-color-9: var(--simple-colors-default-theme-amber-10);\n          --chartist-color-10: var(--simple-colors-default-theme-teal-8);\n          --chartist-color-11: var(--simple-colors-default-theme-light-green-9);\n          --chartist-color-12: var(--simple-colors-default-theme-lime-10);\n          --chartist-color-13: var(--simple-colors-default-theme-green-8);\n          --chartist-color-14: var(--simple-colors-default-theme-amber-9);\n          --chartist-color-15: var(--simple-colors-default-theme-teal-10);\n        }\n\n        :host([dark][accent-color=\"light-green\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-light-green-4);\n          --chartist-color-2: var(--simple-colors-default-theme-lime-3);\n          --chartist-color-3: var(--simple-colors-default-theme-green-2);\n          --chartist-color-4: var(--simple-colors-default-theme-amber-4);\n          --chartist-color-5: var(--simple-colors-default-theme-teal-3);\n          --chartist-color-6: var(--simple-colors-default-theme-light-green-2);\n          --chartist-color-7: var(--simple-colors-default-theme-lime-4);\n          --chartist-color-8: var(--simple-colors-default-theme-green-3);\n          --chartist-color-9: var(--simple-colors-default-theme-amber-2);\n          --chartist-color-10: var(--simple-colors-default-theme-teal-4);\n          --chartist-color-11: var(--simple-colors-default-theme-light-green-3);\n          --chartist-color-12: var(--simple-colors-default-theme-lime-2);\n          --chartist-color-13: var(--simple-colors-default-theme-green-4);\n          --chartist-color-14: var(--simple-colors-default-theme-amber-3);\n          --chartist-color-15: var(--simple-colors-default-theme-teal-2);\n        }\n\n        :host([accent-color=\"lime\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-lime-8);\n          --chartist-color-2: var(--simple-colors-default-theme-yellow-9);\n          --chartist-color-3: var(--simple-colors-default-theme-light-green-10);\n          --chartist-color-4: var(--simple-colors-default-theme-orange-8);\n          --chartist-color-5: var(--simple-colors-default-theme-green-9);\n          --chartist-color-6: var(--simple-colors-default-theme-lime-10);\n          --chartist-color-7: var(--simple-colors-default-theme-yellow-8);\n          --chartist-color-8: var(--simple-colors-default-theme-light-green-9);\n          --chartist-color-9: var(--simple-colors-default-theme-orange-10);\n          --chartist-color-10: var(--simple-colors-default-theme-green-8);\n          --chartist-color-11: var(--simple-colors-default-theme-lime-9);\n          --chartist-color-12: var(--simple-colors-default-theme-yellow-10);\n          --chartist-color-13: var(--simple-colors-default-theme-light-green-8);\n          --chartist-color-14: var(--simple-colors-default-theme-orange-9);\n          --chartist-color-15: var(--simple-colors-default-theme-green-10);\n        }\n\n        :host([dark][accent-color=\"lime\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-lime-4);\n          --chartist-color-2: var(--simple-colors-default-theme-yellow-3);\n          --chartist-color-3: var(--simple-colors-default-theme-light-green-2);\n          --chartist-color-4: var(--simple-colors-default-theme-orange-4);\n          --chartist-color-5: var(--simple-colors-default-theme-green-3);\n          --chartist-color-6: var(--simple-colors-default-theme-lime-2);\n          --chartist-color-7: var(--simple-colors-default-theme-yellow-4);\n          --chartist-color-8: var(--simple-colors-default-theme-light-green-3);\n          --chartist-color-9: var(--simple-colors-default-theme-orange-2);\n          --chartist-color-10: var(--simple-colors-default-theme-green-4);\n          --chartist-color-11: var(--simple-colors-default-theme-lime-3);\n          --chartist-color-12: var(--simple-colors-default-theme-yellow-2);\n          --chartist-color-13: var(--simple-colors-default-theme-light-green-4);\n          --chartist-color-14: var(--simple-colors-default-theme-orange-3);\n          --chartist-color-15: var(--simple-colors-default-theme-green-2);\n        }\n\n        :host([accent-color=\"yellow\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-yellow-8);\n          --chartist-color-2: var(--simple-colors-default-theme-amber-9);\n          --chartist-color-3: var(--simple-colors-default-theme-lime-10);\n          --chartist-color-4: var(--simple-colors-default-theme-deep-orange-8);\n          --chartist-color-5: var(--simple-colors-default-theme-light-green-9);\n          --chartist-color-6: var(--simple-colors-default-theme-yellow-10);\n          --chartist-color-7: var(--simple-colors-default-theme-amber-8);\n          --chartist-color-8: var(--simple-colors-default-theme-lime-9);\n          --chartist-color-9: var(--simple-colors-default-theme-deep-orange-10);\n          --chartist-color-10: var(--simple-colors-default-theme-light-green-8);\n          --chartist-color-11: var(--simple-colors-default-theme-yellow-9);\n          --chartist-color-12: var(--simple-colors-default-theme-amber-10);\n          --chartist-color-13: var(--simple-colors-default-theme-lime-8);\n          --chartist-color-14: var(--simple-colors-default-theme-deep-orange-9);\n          --chartist-color-15: var(\n            --simple-colors-default-theme-light-green-10\n          );\n        }\n\n        :host([dark][accent-color=\"yellow\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-yellow-4);\n          --chartist-color-2: var(--simple-colors-default-theme-amber-3);\n          --chartist-color-3: var(--simple-colors-default-theme-lime-2);\n          --chartist-color-4: var(--simple-colors-default-theme-deep-orange-4);\n          --chartist-color-5: var(--simple-colors-default-theme-light-green-3);\n          --chartist-color-6: var(--simple-colors-default-theme-yellow-2);\n          --chartist-color-7: var(--simple-colors-default-theme-amber-4);\n          --chartist-color-8: var(--simple-colors-default-theme-lime-3);\n          --chartist-color-9: var(--simple-colors-default-theme-deep-orange-2);\n          --chartist-color-10: var(--simple-colors-default-theme-light-green-4);\n          --chartist-color-11: var(--simple-colors-default-theme-yellow-3);\n          --chartist-color-12: var(--simple-colors-default-theme-amber-2);\n          --chartist-color-13: var(--simple-colors-default-theme-lime-4);\n          --chartist-color-14: var(--simple-colors-default-theme-deep-orange-3);\n          --chartist-color-15: var(--simple-colors-default-theme-light-green-2);\n        }\n\n        :host([accent-color=\"amber\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-amber-8);\n          --chartist-color-2: var(--simple-colors-default-theme-orange-9);\n          --chartist-color-3: var(--simple-colors-default-theme-yellow-10);\n          --chartist-color-4: var(--simple-colors-default-theme-red-8);\n          --chartist-color-5: var(--simple-colors-default-theme-lime-9);\n          --chartist-color-6: var(--simple-colors-default-theme-amber-10);\n          --chartist-color-7: var(--simple-colors-default-theme-orange-8);\n          --chartist-color-8: var(--simple-colors-default-theme-yellow-9);\n          --chartist-color-9: var(--simple-colors-default-theme-red-10);\n          --chartist-color-10: var(--simple-colors-default-theme-lime-8);\n          --chartist-color-11: var(--simple-colors-default-theme-amber-9);\n          --chartist-color-12: var(--simple-colors-default-theme-orange-10);\n          --chartist-color-13: var(--simple-colors-default-theme-yellow-8);\n          --chartist-color-14: var(--simple-colors-default-theme-red-9);\n          --chartist-color-15: var(--simple-colors-default-theme-lime-10);\n        }\n\n        :host([dark][accent-color=\"amber\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-amber-4);\n          --chartist-color-2: var(--simple-colors-default-theme-orange-3);\n          --chartist-color-3: var(--simple-colors-default-theme-yellow-2);\n          --chartist-color-4: var(--simple-colors-default-theme-red-4);\n          --chartist-color-5: var(--simple-colors-default-theme-lime-3);\n          --chartist-color-6: var(--simple-colors-default-theme-amber-2);\n          --chartist-color-7: var(--simple-colors-default-theme-orange-4);\n          --chartist-color-8: var(--simple-colors-default-theme-yellow-3);\n          --chartist-color-9: var(--simple-colors-default-theme-red-2);\n          --chartist-color-10: var(--simple-colors-default-theme-lime-4);\n          --chartist-color-11: var(--simple-colors-default-theme-amber-3);\n          --chartist-color-12: var(--simple-colors-default-theme-orange-2);\n          --chartist-color-13: var(--simple-colors-default-theme-yellow-4);\n          --chartist-color-14: var(--simple-colors-default-theme-red-3);\n          --chartist-color-15: var(--simple-colors-default-theme-lime-2);\n        }\n\n        :host([accent-color=\"orange\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-orange-8);\n          --chartist-color-2: var(--simple-colors-default-theme-deep-orange-9);\n          --chartist-color-3: var(--simple-colors-default-theme-amber-10);\n          --chartist-color-4: var(--simple-colors-default-theme-pink-8);\n          --chartist-color-5: var(--simple-colors-default-theme-yellow-9);\n          --chartist-color-6: var(--simple-colors-default-theme-orange-10);\n          --chartist-color-7: var(--simple-colors-default-theme-deep-orange-8);\n          --chartist-color-8: var(--simple-colors-default-theme-amber-9);\n          --chartist-color-9: var(--simple-colors-default-theme-pink-10);\n          --chartist-color-10: var(--simple-colors-default-theme-yellow-8);\n          --chartist-color-11: var(--simple-colors-default-theme-orange-9);\n          --chartist-color-12: var(\n            --simple-colors-default-theme-deep-orange-10\n          );\n          --chartist-color-13: var(--simple-colors-default-theme-amber-8);\n          --chartist-color-14: var(--simple-colors-default-theme-pink-9);\n          --chartist-color-15: var(--simple-colors-default-theme-yellow-10);\n        }\n\n        :host([dark][accent-color=\"orange\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-orange-4);\n          --chartist-color-2: var(--simple-colors-default-theme-deep-orange-3);\n          --chartist-color-3: var(--simple-colors-default-theme-amber-2);\n          --chartist-color-4: var(--simple-colors-default-theme-pink-4);\n          --chartist-color-5: var(--simple-colors-default-theme-yellow-3);\n          --chartist-color-6: var(--simple-colors-default-theme-orange-2);\n          --chartist-color-7: var(--simple-colors-default-theme-deep-orange-4);\n          --chartist-color-8: var(--simple-colors-default-theme-amber-3);\n          --chartist-color-9: var(--simple-colors-default-theme-pink-2);\n          --chartist-color-10: var(--simple-colors-default-theme-yellow-4);\n          --chartist-color-11: var(--simple-colors-default-theme-orange-3);\n          --chartist-color-12: var(--simple-colors-default-theme-deep-orange-2);\n          --chartist-color-13: var(--simple-colors-default-theme-amber-4);\n          --chartist-color-14: var(--simple-colors-default-theme-pink-3);\n          --chartist-color-15: var(--simple-colors-default-theme-yellow-2);\n        }\n\n        :host([accent-color=\"deep-orange\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-deep-orange-8);\n          --chartist-color-2: var(--simple-colors-default-theme-red-9);\n          --chartist-color-3: var(--simple-colors-default-theme-orange-10);\n          --chartist-color-4: var(--simple-colors-default-theme-purple-8);\n          --chartist-color-5: var(--simple-colors-default-theme-amber-9);\n          --chartist-color-6: var(--simple-colors-default-theme-deep-orange-10);\n          --chartist-color-7: var(--simple-colors-default-theme-red-8);\n          --chartist-color-8: var(--simple-colors-default-theme-orange-9);\n          --chartist-color-9: var(--simple-colors-default-theme-purple-10);\n          --chartist-color-10: var(--simple-colors-default-theme-amber-8);\n          --chartist-color-11: var(--simple-colors-default-theme-deep-orange-9);\n          --chartist-color-12: var(--simple-colors-default-theme-red-10);\n          --chartist-color-13: var(--simple-colors-default-theme-orange-8);\n          --chartist-color-14: var(--simple-colors-default-theme-purple-9);\n          --chartist-color-15: var(--simple-colors-default-theme-amber-10);\n        }\n\n        :host([dark][accent-color=\"deep-orange\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-deep-orange-4);\n          --chartist-color-2: var(--simple-colors-default-theme-red-3);\n          --chartist-color-3: var(--simple-colors-default-theme-orange-2);\n          --chartist-color-4: var(--simple-colors-default-theme-purple-4);\n          --chartist-color-5: var(--simple-colors-default-theme-amber-3);\n          --chartist-color-6: var(--simple-colors-default-theme-deep-orange-2);\n          --chartist-color-7: var(--simple-colors-default-theme-red-4);\n          --chartist-color-8: var(--simple-colors-default-theme-orange-3);\n          --chartist-color-9: var(--simple-colors-default-theme-purple-2);\n          --chartist-color-10: var(--simple-colors-default-theme-amber-4);\n          --chartist-color-11: var(--simple-colors-default-theme-deep-orange-3);\n          --chartist-color-12: var(--simple-colors-default-theme-red-2);\n          --chartist-color-13: var(--simple-colors-default-theme-orange-4);\n          --chartist-color-14: var(--simple-colors-default-theme-purple-3);\n          --chartist-color-15: var(--simple-colors-default-theme-amber-2);\n        }\n\n        :host([accent-color=\"brown\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-brown-8);\n          --chartist-color-2: var(--simple-colors-default-theme-red-9);\n          --chartist-color-3: var(--simple-colors-default-theme-deep-orange-10);\n          --chartist-color-4: var(--simple-colors-default-theme-brown-11);\n          --chartist-color-5: var(--simple-colors-default-theme-red-12);\n          --chartist-color-6: var(--simple-colors-default-theme-deep-orange-8);\n          --chartist-color-7: var(--simple-colors-default-theme-brown-9);\n          --chartist-color-8: var(--simple-colors-default-theme-red-10);\n          --chartist-color-9: var(--simple-colors-default-theme-deep-orange-11);\n          --chartist-color-10: var(--simple-colors-default-theme-brown-12);\n          --chartist-color-11: var(--simple-colors-default-theme-red-8);\n          --chartist-color-12: var(--simple-colors-default-theme-deep-orange-9);\n          --chartist-color-13: var(--simple-colors-default-theme-brown-10);\n          --chartist-color-14: var(--simple-colors-default-theme-red-11);\n          --chartist-color-15: var(\n            --simple-colors-default-theme-deep-orange-12\n          );\n        }\n\n        :host([dark][accent-color=\"brown\"]) {\n          --chartist-color-1: var(--simple-colors-default-theme-brown-4);\n          --chartist-color-2: var(--simple-colors-default-theme-red-3);\n          --chartist-color-3: var(--simple-colors-default-theme-deep-orange-2);\n          --chartist-color-4: var(--simple-colors-default-theme-brown-1);\n          --chartist-color-5: var(--simple-colors-default-theme-red-0);\n          --chartist-color-6: var(--simple-colors-default-theme-deep-orange-4);\n          --chartist-color-7: var(--simple-colors-default-theme-brown-3);\n          --chartist-color-8: var(--simple-colors-default-theme-red-2);\n          --chartist-color-9: var(--simple-colors-default-theme-deep-orange-1);\n          --chartist-color-10: var(--simple-colors-default-theme-brown-0);\n          --chartist-color-11: var(--simple-colors-default-theme-red-4);\n          --chartist-color-12: var(--simple-colors-default-theme-deep-orange-3);\n          --chartist-color-13: var(--simple-colors-default-theme-brown-2);\n          --chartist-color-14: var(--simple-colors-default-theme-red-1);\n          --chartist-color-15: var(--simple-colors-default-theme-deep-orange-0);\n        }\n\n        :host {\n          display: block;\n        }\n      "]);

    _templateObject2_5da312d02d8511ea80ccf124101da415 = function _templateObject2_5da312d02d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject_5da312d02d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <iron-ajax\n        auto\n        handle-as=\"text\"\n        url=\"", "\"\n        @response=\"", "\"\n      ></iron-ajax>\n      <chartist-render\n        id=\"chartist\"\n        .type=\"", "\"\n        .scale=\"", "\"\n        .chart-title=\"", "\"\n        .chart-desc=\"", "\"\n        .data=\"", "\"\n        .responsive-options=\"", "\"\n      ></chartist-render>\n    "]);

    _templateObject_5da312d02d8511ea80ccf124101da415 = function _templateObject_5da312d02d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  /**
   * `lrndesign-chart-behaviors`
   * @customElement lrndesign-chart-behaviors
   * a line chart
   *
  
   * @extends ChartistRender
   * @see @lrnwebcomponents/chartist-render/chartist-render.js
   * @extends SimpleColors
   * @see @lrnwebcomponents/simple-colors/simple-colors.js
   * @extends SchemaBehaviors
   * @see @lrnwebcomponents/schema-behaviors/schema-behaviors.js
   *
   * @demo ./demo/index.html
   * @demo ./demo/pie.html pie charts
   * @demo ./demo/bar.html bar charts
   * @demo ./demo/line.html line charts
   *
   */
  var LrndesignChart =
  /*#__PURE__*/
  function (_SimpleColors) {
    babelHelpers.inherits(LrndesignChart, _SimpleColors);
    babelHelpers.createClass(LrndesignChart, [{
      key: "render",
      // render function
      value: function render() {
        return (0, _litElement.html)(_templateObject_5da312d02d8511ea80ccf124101da415(), this.dataSource, this.handleResponse, this.type, this.scale, this.chartTitle, this.chartDesc, this.data, this.responsiveOptions);
      } // haxProperty definition

    }], [{
      key: "styles",
      //styles function
      get: function get() {
        var styles = [(0, _litElement.css)(_templateObject2_5da312d02d8511ea80ccf124101da415())];
        if (babelHelpers.get(babelHelpers.getPrototypeOf(LrndesignChart), "styles", this)) styles = Object.assign(babelHelpers.get(babelHelpers.getPrototypeOf(LrndesignChart), "styles", this), styles);
        return styles;
      }
    }, {
      key: "haxProperties",
      get: function get() {
        return {
          canScale: true,
          canPosition: true,
          canEditSource: false,
          gizmo: {
            description: "Creates an accessible chart based on a CSV.",
            color: "green darken-4",
            groups: ["Data", "Presentation"],
            handles: [{
              type: "data",
              url: "csvFile"
            }],
            meta: {
              author: "LRNWebComponents"
            }
          },
          settings: {
            quick: [{
              property: "chartTitle",
              title: "Chart Title",
              description: "Accessible alt text for your chart.",
              inputMethod: "textfield",
              icon: "text-field",
              required: true
            }],
            configure: [{
              property: "data-source",
              title: "CSV File",
              description: "The URL for your CSV file.",
              inputMethod: "textfield",
              icon: "link",
              validationType: "url",
              required: true
            }, {
              property: "chartTitle",
              title: "Chart Title",
              description: "Accessible alt text for your chart.",
              inputMethod: "textfield",
              icon: "text-field",
              required: true
            }, {
              property: "chartDesc",
              title: "Chart Description",
              description: "Accessible description of your chart.",
              inputMethod: "textfield",
              icon: "text-field"
            }],
            advanced: [{
              property: "scale",
              title: "Scale Name",
              description: "The ratio of width:height of the chart (See https://gionkunz.github.io/chartist-js/getting-started.html#default-sass-settings for $ct-scales and $ct-scales-names).",
              inputMethod: "textfield",
              icon: "text-field"
            }, {
              property: "reverseData",
              title: "Reverse Data",
              description: "Reverse data including labels, the series order as well as the whole series data arrays.",
              inputMethod: "boolean",
              icon: "check-box"
            }]
          }
        };
      } // properties available to the custom element for data binding

    }, {
      key: "properties",
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(LrndesignChart), "properties", this), {
          /**
           * Location of the CSV file.
           */
          dataSource: {
            type: String,
            attribute: "data-source"
          },

          /**
           * Fixed height for the chart as a string (i.e. '100px' or '50%').
           */
          height: {
            type: String
          },

          /**
           * Raw data pulled in from the csv file.
           */
          rawData: {
            type: String,
            attribute: "raw-data"
          },

          /**
           * Reverse data including labels, the series order as well as
           * the whole series data arrays.
           */
          reverseData: {
            type: Boolean,
            attribute: "reverse-data"
          },

          /**
           * Fixed width for the chart as a string (i.e. '100px' or '50%').
           */
          width: {
            type: String
          }
        });
      }
    }]);

    function LrndesignChart() {
      var _this;

      babelHelpers.classCallCheck(this, LrndesignChart);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrndesignChart).call(this));

      _this.setProperties();

      var checkReady = setInterval(function () {
        if (_this.__dataReady) {
          _this.makeChart();

          clearInterval(checkReady);
        }
      }, 1);
      return _this;
    }
    /**
     * Store the tag name to make it easier to obtain directly.
     */


    babelHelpers.createClass(LrndesignChart, [{
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          /**
           * Fired when data source changes.
           *
           * @event data-source-changed
           * @param {string} dataSource data source of the chart
           *
           */
          if (propName === "dataSource") _this2.dispatchEvent(new CustomEvent("data-source-changed", {
            detail: _this2
          }));
          /**
           * Fired when raw data changes.
           *
           * @event raw-data-changed
           * @param {string} rawData raw CSV data for the chart which will be converted into an array
           *
           */

          if (propName === "rawData") _this2.dispatchEvent(new CustomEvent("raw-data-changed", {
            detail: _this2
          }));

          if (_this2.__dataReady) {
            _this2.makeChart();
          }
        });
      }
      /**
       * refreshes the chart
       */

    }, {
      key: "makeChart",
      value: function makeChart() {
        var chart = this.shadowRoot.querySelector("#chartist");

        if (chart) {
          chart.options = this._getOptions();
          /**
           * Fired when chart options change.
           *
           * @event options-changed
           * @param {object} chart options
           *
           */

          this.dispatchEvent(new CustomEvent("options-changed", {
            detail: this
          }));
          chart.makeChart();
          /**
           * Fired when chart changes.
           *
           * @event chart-changed
           *
           */

          this.dispatchEvent(new CustomEvent("chart-changed", {
            detail: this
          }));
        }
      }
      /**
       * Sets properties for chart.
       * Specific chart types can extend this function
       * with type-specific properties.
       */

    }, {
      key: "setProperties",
      value: function setProperties() {
        this.chartTitle = null;
        this.chartDesc = null;
        this.scale = "ct-minor-seventh";
        this.reverseData = false;
        this.rawData = "";
      }
      /**
       * Sets properties specific to bar and line charts.
       * Bar and line charts can include this function
       * in their extended setProperties function.
       */

    }, {
      key: "setBarLineProperties",
      value: function setBarLineProperties() {
        this.high = undefined;
        this.low = undefined;
        this.axisXLabelOffsetX = 0;
        this.axisXLabelOffsetY = 0;
        this.axisXOffset = 30;
        this.axisXPosition = "end";
        this.axisXShowGrid = true;
        this.axisXShowLabel = true;
        this.axisXTopLeft = false;
        this.axisYLabelOffsetX = 0;
        this.axisYLabelOffsetY = 0;
        this.axisYOffset = 30;
        this.axisYOnlyInteger = false;
        this.axisYPosition = "start";
        this.axisYScaleMinSpace = 20;
        this.axisYShowGrid = true;
        this.axisYshowLabel = true;
        this.axisYTopLeft = true;
        this.showGridBackground = false;
        this.chartPaddingBottom = 5;
        this.chartPaddingLeft = 10;
        this.chartPaddingRight = 15;
        this.chartPaddingTop = 15;
      }
      /**
       * Convert from csv text to an array in the table function
       * @param {event} e event data
       */

    }, {
      key: "handleResponse",
      value: function handleResponse(e) {
        this.rawData = e.detail.response;
        var raw = this.CSVtoArray(this.rawData);
        this.data = {
          labels: raw[0],
          series: this.type !== "pie" ? raw.slice(1, raw.length) : raw[1]
        };
        this.__dataReady = true;
      }
      /**
       * override this with type-specific options
       * @returns {object} options
       */

    }, {
      key: "_getOptions",
      value: function _getOptions() {
        return {
          reverseData: this.reverseData
        };
      }
      /**
       * override this with type-specific options
       * @returns {object} options specific to both bar and line charts
       */

    }, {
      key: "_getLineBarOptions",
      value: function _getLineBarOptions() {
        return {
          high: this.high,
          low: this.low,
          axisX: {
            labelOffset: {
              x: this.axisXLabelOffsetX,
              y: this.axisXLabelOffsetY,
              offset: this.axisXOffset
            },
            position: this.axisXPosition,
            showGrid: this.axisXShowGrid,
            showLabel: this.axisXShowLabel
          },
          axisY: {
            labelOffset: {
              x: this.axisYLabelOffsetX,
              y: this.axisYLabelOffsetY,
              offset: this.axisYOffset
            },
            position: this.axisYPosition,
            showGrid: this.axisYShowGrid,
            showLabel: this.axisYShowLabel,
            onlyInteger: this.axisYOnlyInteger,
            scaleMinSpace: this.axisYScaleMinSpace
          },
          showGridBackground: this.showGridBackground,
          chartPadding: {
            bottom: this.chartPaddingBottom,
            left: this.chartPaddingLeft,
            right: this.chartPaddingRight,
            top: this.chartPaddingTop
          }
        };
      }
      /**
       * Mix of solutions from https://stackoverflow.com/questions/8493195/how-can-i-parse-a-csv-string-with-javascript-which-contains-comma-in-data
       * @param {string} text csv data
       * @returns {array} chart data
       */

    }, {
      key: "CSVtoArray",
      value: function CSVtoArray(text) {
        var p = "",
            row = [""],
            ret = [row],
            i = 0,
            r = 0,
            s = !0,
            l;

        for (l in text) {
          l = text[l];

          if ('"' === l) {
            if (s && l === p) row[i] += l;
            s = !s;
          } else if ("," === l && s) {
            if (row[i].trim().match(/^\d+$/m) !== null) row[i] = parseInt(row[i].trim());
            l = row[++i] = "";
          } else if ("\n" === l && s) {
            if ("\r" === p) row[i] = row[i].slice(0, -1);
            if (row[i].trim().match(/^\d+$/m) !== null) row[i] = parseInt(row[i].trim());
            row = ret[++r] = [l = ""];
            i = 0;
          } else row[i] += l;

          p = l;
        }

        if (row[i].trim().match(/^\d+$/m) !== null) row[i] = parseInt(row[i].trim());
        return ret;
      }
      /**
       * life cycle, element is removed from the DOM
       */
      //disconnectedCallback() {}

    }], [{
      key: "tag",
      get: function get() {
        return "lrndesign-chart";
      } //properties common to line and bar charts

    }, {
      key: "lineBarProperties",
      get: function get() {
        return {
          /**
           * Offset X of labels for X-axis
           */
          axisXLabelOffsetX: {
            attribute: "axis-x-label-offset-x",
            type: Number
          },

          /**
           * Offset Y of labels for X-axis
           */
          axisXLabelOffsetY: {
            attribute: "axis-x-label-offset-y",
            type: Number
          },

          /**
           * The offset of the chart drawing area to the border of the container.
           */
          axisXOffset: {
            attribute: "axis-x-offset",
            type: Number
          },

          /**
           * Position where labels are placed.
           * Can be set to `start` or `end`
           * where `start` is equivalent to left or top on vertical axis
           * and `end` is equivalent to right or bottom on horizontal axis.
           */
          axisXPosition: {
            attribute: "axis-x-position",
            type: String
          },

          /**
           * Show axis X grid?
           */
          axisXShowGrid: {
            attribute: "axis-x-show-grid",
            type: Boolean
          },

          /**
           * Show axis X labels?
           */
          axisXShowLabel: {
            attribute: "axis-x-show-label",
            type: Boolean
          },

          /**
           * Offset X of labels for Y-axis
           */
          axisYLabelOffsetX: {
            attribute: "axis-y-label-offset-x",
            type: Number
          },

          /**
           * Offset Y of labels for Y-axis
           */
          axisYLabelOffsetY: {
            attribute: "axis-y-label-offset-y",
            type: Number
          },

          /**
           * Position where labels are placed.
           * Can be set to `start` or `end`
           * where `start` is equivalent to left or top on vertical axis
           * and `end` is equivalent to right or bottom on horizontal axis.
           */
          axisYPosition: {
            attribute: "axis-y-position",
            type: String
          },

          /**
           * Specifies minimum height in pixel of scale steps
           */
          axisYScaleMinSpace: {
            attribute: "axis-y-scale-min-space",
            type: Number
          },

          /**
           * The offset of the chart drawing area to the border of the container.
           */
          axisYOffset: {
            attribute: "axis-y-offset",
            type: Number
          },

          /**
           * Use only integer values (whole numbers) for the scale steps
           */
          axisYOnlyInteger: {
            attribute: "axis-y-only-integer",
            type: Boolean
          },

          /**
           * Show axis Y grid?
           */
          axisYshowGrid: {
            attribute: "axis-y-show-grid",
            type: Boolean
          },

          /**
           * Show axis Y labels?
           */
          axisYshowLabel: {
            attribute: "axis-y-show-label",
            type: Boolean
          },

          /**
           * Position labels at top-left of axis?
           */
          axisYTopLeft: {
            attribute: "axis-y-top-left",
            type: Boolean
          },

          /**
           * Padding below chart drawing area
           */
          chartPaddingBottom: {
            attribute: "chart-padding-bottom",
            type: String
          },

          /**
           * Padding left of chart drawing area
           */
          chartPaddingLeft: {
            attribute: "chart-padding-left",
            type: String
          },

          /**
           * Padding right of chart drawing area
           */
          chartPaddingRight: {
            attribute: "chart-padding-right",
            type: String
          },

          /**
           * Padding above chart drawing area
           */
          chartPaddingTop: {
            attribute: "chart-padding-top",
            type: String
          },

          /**
           * Overriding the natural high of the chart allows you to zoom in
           * or limit the charts highest displayed value.
           */
          high: {
            type: Number
          },

          /**
           * Overriding the natural low of the chart allows you to zoom in
           * or limit the charts lowest displayed value.
           */
          low: {
            type: Number
          },

          /**
           * If the bar chart should add a background fill to the .ct-grids group.
           */
          showGridBackground: {
            attribute: "show-grid-background",
            type: Boolean
          }
        };
      } // extends haxProperty definition to line and bar properties

    }, {
      key: "lineBarHaxProperties",
      get: function get() {
        return {
          gridBackground: [{
            property: "showGridBackground",
            title: "Show Grid Background",
            inputMethod: "boolean"
          }],
          padding: [{
            property: "chartPaddingTop",
            title: "Chart Padding (top)",
            inputMethod: "text-field"
          }, {
            property: "chartPaddingBottom",
            title: "Chart Padding (bottom)",
            inputMethod: "text-field"
          }, {
            property: "chartPaddingLeft",
            title: "Chart Padding (left)",
            inputMethod: "text-field"
          }, {
            property: "chartPaddingRight",
            title: "Chart Padding (right)",
            inputMethod: "text-field"
          }],
          minMax: [{
            property: "low",
            title: "Chart Minimum",
            description: "\n          Overriding the natural low of the chart allows you to zoom in \n          or limit the chart's lowest displayed value",
            inputMethod: "number"
          }, {
            property: "high",
            title: "Chart Maximum",
            description: "\n          Overriding the natural high of the chart allows you to zoom in \n          or limit the chart's highest displayed value",
            inputMethod: "number"
          }],
          xAxis: [{
            property: "axisXShowGrid",
            title: "X-Axis Show Grid",
            description: "Show the X-Axis's grid.",
            inputMethod: "boolean"
          }, {
            property: "axisXOffset",
            title: "X-Axis Offset",
            inputMethod: "number"
          }, {
            property: "axisXPosition",
            title: "X-Axis Position",
            description: "\n            Position where labels are placed. Can be set to \"start\" or \"end\" \n            where \"start\" is equivalent to left or top on vertical axis\n            and \"end\" is equivalent to right or bottom on horizontal axis",
            inputMethod: "text-field"
          }, {
            property: "axisXShowLabel",
            title: "X-Axis Show Label",
            description: "Show the X-Axis's label.",
            inputMethod: "boolean"
          }, {
            property: "axisXLabelOffsetX",
            title: "X-Axis Label (horizontal offset)",
            description: "Horizontal position of the X-Axis's label.",
            inputMethod: "number"
          }, {
            property: "axisXLabelOffsetY",
            title: "X-Axis Label (vertical offset)",
            description: "Vertical position of the X-Axis's label.",
            inputMethod: "number"
          }],
          yAxis: [{
            property: "axisYShowGrid",
            title: "Y-Axis: Show Grid",
            description: "Show the Y-Axis's grid.",
            inputMethod: "boolean"
          }, {
            property: "axisYOffset",
            title: "Y-Axis Offset",
            inputMethod: "number"
          }, {
            property: "axisYPosition",
            title: "Y-Axis Position",
            description: "\n            Position where labels are placed. Can be set to \"start\" or \"end\" \n            where \"start\" is equivalent to left or top on vertical axis\n            and \"end\" is equivalent to right or bottom on horizontal axis",
            inputMethod: "text-field"
          }, {
            property: "axisYShowLabel",
            title: "Y-Axis Show Label",
            description: "Show the Y-Axis's label.",
            inputMethod: "boolean"
          }, {
            property: "axisYLabelOffsetX",
            title: "Y-Axis Label (horizontal offset)",
            description: "Horizontal position of the Y-Axis's label.",
            inputMethod: "number"
          }, {
            property: "axisYLabelOffsetY",
            title: "Y-Axis Label (vertical offset)",
            description: "Vertical position of the Y-Axis's label.",
            inputMethod: "number"
          }, {
            property: "axisYScaleMinSpace",
            title: "Y-Axis Scale Minimum Space",
            description: "Specifies minimum height in pixel of scale steps.",
            inputMethod: "number"
          }, {
            property: "axisYOnlyInteger",
            title: "Y-Axis Scale (only integers)",
            description: "Use only integer values (whole numbers) for the scale steps.",
            inputMethod: "boolean"
          }]
        };
      }
    }]);
    return LrndesignChart;
  }(_simpleColors.SimpleColors);

  _exports.LrndesignChart = LrndesignChart;
  window.customElements.define(LrndesignChart.tag, LrndesignChart);
});
