// @lrnwebcomponents/responsive-utility/responsive-utility.js
define(["exports", "../../@polymer/polymer/polymer-element.js", "../../@polymer/iron-resizable-behavior/iron-resizable-behavior.js", "../../@polymer/polymer/lib/legacy/class.js"], function (_exports, _polymerElement, _ironResizableBehavior, _class) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ResponsiveUtility = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject_eec1b4c02c1c11eaacb4bf54e86db688() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: inline;\n        }\n      </style>\n      <slot></slot>\n    "]);

    _templateObject_eec1b4c02c1c11eaacb4bf54e86db688 = function _templateObject_eec1b4c02c1c11eaacb4bf54e86db688() {
      return data;
    };

    return data;
  }

  window.ResponsiveUtility = {};
  window.ResponsiveUtility.instance = null;

  var ResponsiveUtility =
  /*#__PURE__*/
  function (_mixinBehaviors) {
    babelHelpers.inherits(ResponsiveUtility, _mixinBehaviors);
    babelHelpers.createClass(ResponsiveUtility, [{
      key: "connectedCallback",
      value: function connectedCallback() {
        babelHelpers.get(babelHelpers.getPrototypeOf(ResponsiveUtility.prototype), "connectedCallback", this).call(this);
        /* handle element registration */

        window.addEventListener("responsive-element", this.responiveElementEvent.bind(this));
        /* handle element deregistration */

        window.addEventListener("delete-responsive-element", this.deleteResponiveElementEvent.bind(this));
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        window.removeEventListener("responsive-element", this.responiveElementEvent.bind(this));
        /* handle element deregistration */

        window.removeEventListener("delete-responsive-element", this.deleteResponiveElementEvent.bind(this));
        babelHelpers.get(babelHelpers.getPrototypeOf(ResponsiveUtility.prototype), "disconnectedCallback", this).call(this);
      }
    }, {
      key: "responiveElementEvent",
      value: function responiveElementEvent(e) {
        var detail = {
          element: e.detail.element,
          attribute: e.detail.attribute !== undefined && e.detail.attribute !== null ? e.detail.attribute : "responsive-size",
          relativeToParent: e.detail.relativeToParent !== undefined && e.detail.relativeToParent !== null ? e.detail.relativeToParent : true,
          sm: e.detail.sm !== undefined && e.detail.sm !== null ? e.detail.sm : 900,
          md: e.detail.md !== undefined && e.detail.md !== null ? e.detail.md : 1200,
          lg: e.detail.lg !== undefined && e.detail.lg !== null ? e.detail.lg : 1500,
          xl: e.detail.xl !== undefined && e.detail.xl !== null ? e.detail.lg : 1800
        };

        if ("ResizeObserver" in window && detail.relativeToParent === true) {
          var resize = new ResizeObserver(function () {
            window.ResponsiveUtility.setSize(detail);
          }),
              observable = e.detail.parentNode !== undefined && e.detail.parentNode !== null ? e.detail.parentNode.nodeType === Node.DOCUMENT_FRAGMENT_NODE ? e.detail.element.parentNode.host : e.detail.element.parentNode : e.detail.element;
          resize.observe(observable);
        }

        this.details.push(detail);
        window.ResponsiveUtility.setSize(detail);
      }
    }, {
      key: "deleteResponiveElementEvent",
      value: function deleteResponiveElementEvent(e) {
        for (var i = 0; i < this.details.length; i++) {
          if (e.detail === detail[i]) this.details.splice(i, 1);
        }
      }
      /**
       * An array of objects. Each object is contains data about an element
       * that will be updated with responsive values.
       *
       * To add an element to this array, fire a 'responsive-element' event
       * with the following data:
       *
       * {
       *   "element": (the element itself),
       *   "attribute": (the attribute that will be set with the size),
       *   "relativeToParent": (true for @element query instead of @media query),
       *   "sm": (optional custom sm breakpoint, default is 600),
       *   "md": (optional custom md breakpoint, default is 900),
       *   "lg": (optional custom lg breakpoint, default is 1200),
       *   "xl": (optional custom xl breakpoint, default is 1500),
       * }
       *
       */

      /**
       * Makes sure there is a utility ready and listening for elements.
       */

    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_eec1b4c02c1c11eaacb4bf54e86db688());
      }
    }, {
      key: "tag",
      get: function get() {
        return "responsive-utility";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(ResponsiveUtility), "properties", this), {
          /**
           * Stores
           */
          details: {
            type: Array
          }
        });
      }
    }]);

    function ResponsiveUtility() {
      var _this;

      babelHelpers.classCallCheck(this, ResponsiveUtility);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(ResponsiveUtility).call(this));
      _this.details = [];

      if (window.ResponsiveUtility.instance == null) {
        window.ResponsiveUtility.instance = babelHelpers.assertThisInitialized(_this);
      }

      _this.addEventListener("iron-resize", _this._onIronResize.bind(babelHelpers.assertThisInitialized(_this)));

      return _this;
    }
    /**
     * On resize, sets sizes of any detail element that has changed.
     */


    babelHelpers.createClass(ResponsiveUtility, [{
      key: "_onIronResize",
      value: function _onIronResize() {
        for (var i = 0; i < this.details.length; i++) {
          window.ResponsiveUtility.setSize(this.details[i]);
        }
      }
    }]);
    return ResponsiveUtility;
  }((0, _class.mixinBehaviors)([_ironResizableBehavior.IronResizableBehavior], _polymerElement.PolymerElement));

  _exports.ResponsiveUtility = ResponsiveUtility;
  window.customElements.define(ResponsiveUtility.tag, ResponsiveUtility);

  /**
   * Checks to see if there is an instance available, and if not appends one
   */
  window.ResponsiveUtility.requestAvailability = function () {
    if (window.ResponsiveUtility.instance == null) {
      window.ResponsiveUtility.instance = document.createElement("responsive-utility");
    }

    document.body.appendChild(window.ResponsiveUtility.instance);
  };
  /**
   * Sets responsive size of detail.
   */


  window.ResponsiveUtility.setSize = function (detail) {
    var size,
        width = window.ResponsiveUtility._getWidth(detail);

    if (width < detail.sm) {
      size = "xs";
    } else if (width < detail.md) {
      size = "sm";
    } else if (width < detail.lg) {
      size = "md";
    } else if (width < detail.xl) {
      size = "lg";
    } else {
      size = "xl";
    }

    if (detail.element.getAttribute(detail.attribute) === undefined || size !== detail.element.getAttribute(detail.attribute)) {
      detail.element.setAttribute(detail.attribute, size);
    }
  };
  /**
   * Returns width of detail.
   *
   * @param {object} the HTML element to check
   * @returns {number} the width of the element or its parent node
   */


  window.ResponsiveUtility._getWidth = function (detail) {
    var el = detail.element;

    if (detail.relativeToParent === true) {
      if (el && el.offsetWidth !== undefined && el.offsetWidth !== null && el.offsetWidth > 0) {
        return el.offsetWidth;
      } else if (el.parentNode !== null) {
        return el.parentNode.nodeType === Node.DOCUMENT_FRAGMENT_NODE && el.parentNode.host ? el.parentNode.host.offsetWidth : el.parentNode.offsetWidth;
      }
    }

    return window.outerWidth;
  };
});
