// @lrnwebcomponents/a11y-menu-button/lib/a11y-menu-button-item.js
define(["exports", "../../../lit/index.js"], function (_exports, _index) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.A11yMenuButtonItemBehaviors = _exports.A11yMenuButtonItem = void 0;

  var _templateObject_231e8c50942811ecb1937969798b231a, _templateObject2_231e8c50942811ecb1937969798b231a, _templateObject3_231e8c50942811ecb1937969798b231a;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var A11yMenuButtonItemBehaviors = function A11yMenuButtonItemBehaviors(SuperClass) {
    return /*#__PURE__*/function (_SuperClass) {
      babelHelpers.inherits(_class, _SuperClass);

      var _super = _createSuper(_class);

      function _class() {
        var _this;

        babelHelpers.classCallCheck(this, _class);
        _this = _super.call(this);
        _this.slot = "menuitem";
        return _this;
      }
      /**
       * renders item as a link
       *
       * @readonly
       */


      babelHelpers.createClass(_class, [{
        key: "render",
        value: function render() {
          return this.href && this.href.trim() !== "" ? this.linkTemplate : this.buttonTemplate;
        }
      }, {
        key: "linkTemplate",
        get: function get() {
          return (0, _index.html)(_templateObject_231e8c50942811ecb1937969798b231a || (_templateObject_231e8c50942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral([" <li role=\"none\">\n        <a\n          role=\"menuitem\"\n          href=\"", "\"\n          ?disabled=\"", "\"\n          part=\"button\"\n        >\n          <slot></slot>\n        </a>\n      </li>"])), this.href, this.disabled);
        }
        /**
         * renders item as a button
         *
         * @readonly
         */

      }, {
        key: "buttonTemplate",
        get: function get() {
          return (0, _index.html)(_templateObject2_231e8c50942811ecb1937969798b231a || (_templateObject2_231e8c50942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        <li role=\"none\">\n          <button\n            role=\"menuitem\"\n            controls=\"", "\"\n            ?disabled=\"", "\"\n            part=\"button\"\n          >\n            <slot></slot>\n          </button>\n        </li>\n      "])), this.controls, this.disabled);
        }
        /**
         * gets item with role="menuitem"
         *
         * @readonly
         */

      }, {
        key: "menuItem",
        get: function get() {
          return this.shadowRoot && this.shadowRoot.querySelector("[role=menuitem]") ? this.shadowRoot.querySelector("[role=menuitem]") : undefined;
        }
        /**
         * allows link or button to get focus
         *
         * @memberof A11yMenuButtonItem
         */

      }, {
        key: "focus",
        value: function focus() {
          if (this.menuItem) this.menuItem.focus();
        }
      }, {
        key: "connectedCallback",
        value: function connectedCallback() {
          babelHelpers.get(babelHelpers.getPrototypeOf(_class.prototype), "connectedCallback", this).call(this);
          /**
           * Fires when menu item is added to dom
           * @event add-a11y-menu-button-item
           */

          this.dispatchEvent(new CustomEvent("add-a11y-menu-button-item", {
            bubbles: true,
            cancelable: true,
            composed: true,
            detail: this
          }));
        }
      }, {
        key: "disconnectedCallback",
        value: function disconnectedCallback() {
          babelHelpers.get(babelHelpers.getPrototypeOf(_class.prototype), "disconnectedCallback", this).call(this);
          /**
           * Fires when menu item is removed from dom
           * @event remove-a11y-menu-button-item
           */

          this.dispatchEvent(new CustomEvent("remove-a11y-menu-button-item", {
            bubbles: true,
            cancelable: true,
            composed: true,
            detail: this
          }));
        }
      }], [{
        key: "styles",
        get: function get() {
          return [(0, _index.css)(_templateObject3_231e8c50942811ecb1937969798b231a || (_templateObject3_231e8c50942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n          :host {\n            margin: 0;\n            padding: 0;\n            display: block;\n          }\n          :host([hidden]) {\n            display: none;\n          }\n          *[role=\"menuitem\"][disabled] {\n            cursor: not-allowed;\n          }\n\n          *[role=\"menuitem\"],\n          *[role=\"menuitem\"]:visited {\n            display: block;\n            margin: 0;\n            border-radius: 0;\n            font-family: inherit;\n            font-size: inherit;\n            text-decoration: var(--a11y-menu-button-item-text-decoration, none);\n            color: var(\n              --a11y-menu-button-item-color,\n              var(--a11y-menu-button-color, currentColor)\n            );\n            width: calc(\n              100% - 2 *\n                var(\n                  --a11y-menu-button-item-horizontal-padding,\n                  var(--a11y-menu-button-horizontal-padding, 5px)\n                )\n            );\n            text-align: var(--a11y-menu-button-item-text-align, left);\n            padding: var(\n                --a11y-menu-button-item-vertical-padding,\n                var(--a11y-menu-button-vertical-padding, 0)\n              )\n              var(\n                --a11y-menu-button-item-horizontal-padding,\n                var(--a11y-menu-button-horizontal-padding, 5px)\n              );\n            background-color: var(\n              --a11y-menu-button-item-bg-color,\n              var(--a11y-menu-button-bg-color, white)\n            );\n            border-left: var(--a11y-menu-button-item-border-left, none);\n            border-right: var(--a11y-menu-button-item-border-right, none);\n            border-top: var(--a11y-menu-button-item-border-top, none);\n            border-bottom: var(--a11y-menu-button-item-border-bottom, none);\n            border: var(--a11y-menu-button-item-border, none);\n            transition: all 0.25s ease-in-out;\n          }\n          button[role=\"menuitem\"],\n          button[role=\"menuitem\"]:visited {\n            width: 100%;\n          }\n\n          :host(:focus-within) *[role=\"button\"],\n          *[role=\"menuitem\"]:focus,\n          *[role=\"menuitem\"]:hover {\n            text-decoration: var(\n              --a11y-menu-button-item-focus-text-decoration,\n              none\n            );\n            color: var(--a11y-menu-button-item-focus-color, currentColor);\n            background-color: var(\n              --a11y-menu-button-item-focus-bg-color,\n              #e0e0ff\n            );\n            border-left: var(--a11y-menu-button-item-focus-border-left, unset);\n            border-right: var(\n              --a11y-menu-button-item-focus-border-right,\n              unset\n            );\n            border-top: var(--a11y-menu-button-item-focus-border-top, unset);\n            border-bottom: var(\n              --a11y-menu-button-item-focus-border-bottom,\n              unset\n            );\n            border: var(--a11y-menu-button-item-focus-border, unset);\n          }\n        "])))];
        }
      }, {
        key: "tag",
        get: function get() {
          return "a11y-menu-button-item";
        }
      }, {
        key: "properties",
        get: function get() {
          return {
            /**
             * Whether toggle is disabled
             */
            disabled: {
              attribute: "disabled",
              type: Boolean
            },

            /**
             * Whether toggle is disabled
             */
            hidden: {
              attribute: "hidden",
              type: Boolean,
              reflect: true
            },

            /**
             * Whether toggle is disabled
             */
            href: {
              attribute: "href",
              type: String
            },

            /**
             * Whether toggle is disabled
             */
            controls: {
              attribute: "controls",
              type: String
            },
            slot: {
              type: String,
              attribute: "slot",
              reflect: true
            }
          };
        }
      }]);
      return _class;
    }(SuperClass);
  };
  /**
   * a11y-menu-button-item
   * A toggle button for an property in editable-table interface (editable-table.html).
   *
  ### Styling
  
  `<a11y-menu-button-item>` provides custom properties for styling:
  
  Custom property | Description | Default
  ----------------|-------------|----------
  --a11y-menu-button-item-text-decoration | button or link text decoration | none
  --a11y-menu-button-item-color | button or link text color | --a11y-menu-button-color
  --a11y-menu-button-item-text-align | button or link text alignment | left
  --a11y-menu-button-item-vertical-padding | button or link vertical padding | --a11y-menu-button-vertical-padding
  --a11y-menu-button-item-horizontal-padding | button or link horizontal padding | --a11y-menu-button-horizontal-padding
  --a11y-menu-button-item-bg-color | button or link background color | --a11y-menu-button-bg-color
  --a11y-menu-button-item-border | default button or link border  | none
  --a11y-menu-button-item-border-left | overrides button or link left border  | none
  --a11y-menu-button-item-border-right | overrides button or link right border | none
  --a11y-menu-button-item-border-top | overrides button or link top border | none
  --a11y-menu-button-item-border-bottom | overrides button or link bottom border | none
  --a11y-menu-button-item-focus-text-decoration | button or link text decoration when focused | none
  --a11y-menu-button-item-focus-color | button or link text color when focused | black
  --a11y-menu-button-item-focus-bg-color | button or link background color when focused | #e0e0ff
  --a11y-menu-button-item-focus-border-left | overrides button or link left border when focused | unset
  --a11y-menu-button-item-focus-border-right | overrides button or link left border when focused | unset)
  --a11y-menu-button-item-focus-border-top | overrides button or link left border when focused | unset
  --a11y-menu-button-item-focus-border-bottom | overrides button or link left border when focused | unset
  --a11y-menu-button-item-focus-border | button or link border when focused | unset
   *
   * @demo ./demo/index.html
   * @element a11y-menu-button-item
   * @extends A11yMenuButtonItemBehaviors
   */


  _exports.A11yMenuButtonItemBehaviors = A11yMenuButtonItemBehaviors;

  var A11yMenuButtonItem = /*#__PURE__*/function (_A11yMenuButtonItemBe) {
    babelHelpers.inherits(A11yMenuButtonItem, _A11yMenuButtonItemBe);

    var _super2 = _createSuper(A11yMenuButtonItem);

    function A11yMenuButtonItem() {
      babelHelpers.classCallCheck(this, A11yMenuButtonItem);
      return _super2.apply(this, arguments);
    }

    return babelHelpers.createClass(A11yMenuButtonItem);
  }(A11yMenuButtonItemBehaviors(_index.LitElement));

  _exports.A11yMenuButtonItem = A11yMenuButtonItem;
  window.customElements.define(A11yMenuButtonItem.tag, A11yMenuButtonItem);
});
