// @lrnwebcomponents/elmsln-apps/lib/lrnapp-media-grid/lrnapp-media-grid.js
define(["exports", "../../../../@polymer/polymer/polymer-element.js", "../../../../@polymer/polymer/lib/utils/render-status.js", "../../../../@polymer/iron-ajax/iron-ajax.js", "../../../../@polymer/iron-scroll-threshold/iron-scroll-threshold.js", "../../../../@polymer/iron-list/iron-list.js", "../../../../@polymer/paper-button/paper-button.js", "../../../../@polymer/iron-image/iron-image.js", "../../../../@polymer/paper-dialog/paper-dialog.js", "../../../../@polymer/paper-dialog-scrollable/paper-dialog-scrollable.js", "../../../materializecss-styles/materializecss-styles.js"], function (_exports, _polymerElement, _renderStatus, _ironAjax, _ironScrollThreshold, _ironList, _paperButton, _ironImage, _paperDialog, _paperDialogScrollable, _materializecssStyles) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappMediaGrid = void 0;

  function _templateObject_f7ac9140aa5311e9b999174ecd579fc6() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style include=\"materializecss-styles\">\n        :host {\n          display: block;\n        }\n        paper-button {\n          padding: 0;\n          margin: 0;\n          min-width: 1rem;\n        }\n        #details {\n          opacity: 0.5;\n          position: absolute;\n          bottom: 0;\n          right: 0;\n          margin: 0 1rem 0 0;\n          background-color: white;\n          padding: 0.5em;\n        }\n        #details:hover {\n          opacity: 1;\n        }\n        #details span {\n          font-size: 0.6em;\n          font-weight: normal;\n        }\n        #details .comment-on-work {\n          font-size: 0.8em;\n          background-color: white;\n        }\n      </style>\n      <iron-ajax\n        id=\"ajax\"\n        url=\"[[sourcePath]]\"\n        params=\"\"\n        handle-as=\"json\"\n        last-response=\"{{images}}\"\n      ></iron-ajax>\n      <iron-scroll-threshold on-lower-threshold=\"_loadMoreData\" id=\"threshold\">\n        <iron-list grid items=\"[[_toArray(images.data)]]\" as=\"image\">\n          <template>\n            <paper-button>\n              <iron-image\n                preload=\"\"\n                title=\"{{image.filename}}\"\n                alt=\"{{image.alt}}\"\n                src=\"{{image.src}}\"\n                height=\"{{image.height}}\"\n                width=\"{{image.width}}\"\n              ></iron-image>\n            </paper-button>\n          </template>\n        </iron-list>\n      </iron-scroll-threshold>\n      <paper-dialog id=\"dialog\">\n        <paper-dialog-scrollable id=\"dialogResponse\">\n          <iron-image src$=\"[[activeImage]]\"></iron-image>\n          <div id=\"details\">\n            <div class=\"title\">\n              <span>Title:</span> <span>[[activeTitle]]</span>\n            </div>\n            <div class=\"comment-on-work\">\n              <a href$=\"[[activeUrl]]\">\n                <paper-button raised=\"\">View media</paper-button>\n              </a>\n            </div>\n          </div>\n        </paper-dialog-scrollable>\n      </paper-dialog>\n    "]);

    _templateObject_f7ac9140aa5311e9b999174ecd579fc6 = function _templateObject_f7ac9140aa5311e9b999174ecd579fc6() {
      return data;
    };

    return data;
  }

  var LrnappMediaGrid =
  /*#__PURE__*/
  function (_PolymerElement) {
    babelHelpers.inherits(LrnappMediaGrid, _PolymerElement);

    function LrnappMediaGrid() {
      babelHelpers.classCallCheck(this, LrnappMediaGrid);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappMediaGrid).apply(this, arguments));
    }

    babelHelpers.createClass(LrnappMediaGrid, [{
      key: "connectedCallback",
      value: function connectedCallback() {
        babelHelpers.get(babelHelpers.getPrototypeOf(LrnappMediaGrid.prototype), "connectedCallback", this).call(this);
        (0, _renderStatus.afterNextRender)(this, function () {
          this.addEventListener("click", this._triggerDialog.bind(this));
        });
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        this.removeEventListener("click", this._triggerDialog.bind(this));
        babelHelpers.get(babelHelpers.getPrototypeOf(LrnappMediaGrid.prototype), "disconnectedCallback", this).call(this);
      }
    }, {
      key: "_triggerDialog",

      /**
       * When someone clicks if there is a url, then we need to
       * remote load whatever is in that url.
       */
      value: function _triggerDialog(e) {
        // make sure we found an image as we're going through here
        if (e.target.nextElementSibling.nodeName == "IMG") {
          this.activeImage = e.target.nextElementSibling.src;
          this.activeTitle = e.target.parentElement.title;
          this.activeUrl = e.target.parentElement.openUrl;
          this.shadowRoot.querySelector("#dialog").toggle();
        }
      }
    }, {
      key: "_loadMoreData",
      value: function _loadMoreData(e) {
        this.shadowRoot.querySelector("#ajax").generateRequest();
        this.shadowRoot.querySelector("#threshold").clearTriggers();
      }
      /**
       * Simple way to convert from object to array.
       */

    }, {
      key: "_toArray",
      value: function _toArray(obj) {
        if (obj == null) {
          return [];
        }

        return Object.keys(obj).map(function (key) {
          return obj[key];
        });
      }
    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_f7ac9140aa5311e9b999174ecd579fc6());
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrnapp-media-grid";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          elmslnCourse: {
            type: String
          },
          elmslnSection: {
            type: String
          },
          basePath: {
            type: String
          },
          csrfToken: {
            type: String
          },
          endPoint: {
            type: String
          },
          sourcePath: {
            type: String,
            notify: true
          },
          images: {
            type: Array,
            notify: true
          },
          activeImage: {
            type: String,
            reflectToAttribute: true,
            notify: true
          },
          activeTitle: {
            type: String,
            reflectToAttribute: true,
            notify: true
          },
          activeUrl: {
            type: String,
            reflectToAttribute: true,
            notify: true
          }
        };
      }
    }]);
    return LrnappMediaGrid;
  }(_polymerElement.PolymerElement);

  _exports.LrnappMediaGrid = LrnappMediaGrid;
  window.customElements.define(LrnappMediaGrid.tag, LrnappMediaGrid);
});
