// @lit/reactive-element/css-tag.js
define(["exports"], function (_exports) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.unsafeCSS = _exports.supportsAdoptingStyleSheets = _exports.getCompatibleStyle = _exports.css = _exports.adoptStyles = _exports.CSSResult = void 0;

  function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e2) { throw _e2; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e3) { didErr = true; err = _e3; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

  function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

  function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

  /**
   * @license
   * Copyright 2019 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var t = window,
      e = t.ShadowRoot && (void 0 === t.ShadyCSS || t.ShadyCSS.nativeShadow) && "adoptedStyleSheets" in Document.prototype && "replace" in CSSStyleSheet.prototype,
      s = Symbol(),
      n = new WeakMap();
  _exports.supportsAdoptingStyleSheets = e;

  var o = /*#__PURE__*/function () {
    function o(t, e, n) {
      babelHelpers.classCallCheck(this, o);
      if (this._$cssResult$ = !0, n !== s) throw Error("CSSResult is not constructable. Use `unsafeCSS` or `css` instead.");
      this.cssText = t, this.t = e;
    }

    babelHelpers.createClass(o, [{
      key: "styleSheet",
      get: function get() {
        var t = this.o;
        var s = this.t;

        if (e && void 0 === t) {
          var _e = void 0 !== s && 1 === s.length;

          _e && (t = n.get(s)), void 0 === t && ((this.o = t = new CSSStyleSheet()).replaceSync(this.cssText), _e && n.set(s, t));
        }

        return t;
      }
    }, {
      key: "toString",
      value: function toString() {
        return this.cssText;
      }
    }]);
    return o;
  }();

  _exports.CSSResult = o;

  var r = function r(t) {
    return new o("string" == typeof t ? t : t + "", void 0, s);
  },
      i = function i(t) {
    for (var _len = arguments.length, e = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      e[_key - 1] = arguments[_key];
    }

    var n = 1 === t.length ? t[0] : e.reduce(function (e, s, n) {
      return e + function (t) {
        if (!0 === t._$cssResult$) return t.cssText;
        if ("number" == typeof t) return t;
        throw Error("Value passed to 'css' function must be a 'css' function result: " + t + ". Use 'unsafeCSS' to pass non-literal values, but take care to ensure page security.");
      }(s) + t[n + 1];
    }, t[0]);
    return new o(n, t, s);
  },
      S = function S(s, n) {
    e ? s.adoptedStyleSheets = n.map(function (t) {
      return babelHelpers.instanceof(t, CSSStyleSheet) ? t : t.styleSheet;
    }) : n.forEach(function (e) {
      var n = document.createElement("style"),
          o = t.litNonce;
      void 0 !== o && n.setAttribute("nonce", o), n.textContent = e.cssText, s.appendChild(n);
    });
  },
      c = e ? function (t) {
    return t;
  } : function (t) {
    return babelHelpers.instanceof(t, CSSStyleSheet) ? function (t) {
      var e = "";

      var _iterator = _createForOfIteratorHelper(t.cssRules),
          _step;

      try {
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          var _s = _step.value;
          e += _s.cssText;
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }

      return r(e);
    }(t) : t;
  };

  _exports.getCompatibleStyle = c;
  _exports.adoptStyles = S;
  _exports.css = i;
  _exports.unsafeCSS = r;
});
