// @haxtheweb/haxcms-elements/lib/core/micros/haxcms-page-operations.js
import{html as t,css as o}from"../../../../../lit/index.js";import{DDD as e}from"../../../../d-d-d/d-d-d.js";import{I18NMixin as i}from"../../../../i18n-manager/lib/I18NMixin.js";import{store as l}from"../haxcms-site-store.js";import{toJS as s,autorun as a}from"../../../../../mobx/dist/mobx.esm.js";import"../../../../simple-icon/lib/simple-icon-button-lite.js";import"../../../../simple-toolbar/lib/simple-toolbar-button.js";import"../../../../simple-fields/lib/simple-context-menu.js";export class HAXCMSPageOperations extends(i(e)){static get tag(){return"haxcms-page-operations"}static get styles(){return[super.styles,o`
        :host {
          display: inline-block;
          position: relative;
        }
        simple-icon-button-lite.ops {
          --simple-icon-button-border-width: 1px;
          --simple-icon-button-border-color: var(--ddd-border-1, #e0e0e0);
          --simple-icon-height: var(--haxcms-page-operations-ops-icon-size, var(--ddd-icon-4xs, 14px));
          --simple-icon-width: var(--haxcms-page-operations-ops-icon-size, var(--ddd-icon-4xs, 14px));
          padding: var(--ddd-spacing-1) var(--ddd-spacing-1);
          color: var(--ddd-theme-default-white);
          background-color: var(--ddd-theme-default-skyBlue);
          border: var(--ddd-border-xs);
          box-shadow: var(--ddd-boxShadow-sm);
          position: relative;
          height: var(--haxcms-page-operations-ops-height, 24px);
          width: var(--haxcms-page-operations-ops-width, 24px);
        }
        simple-context-menu {
          --simple-context-menu-background: var(
            --simple-toolbar-button-bg,
            white
          );
          --simple-context-menu-background-dark: var(
            --simple-toolbar-button-bg,
            #000000
          );
          --simple-context-menu-border-color: var(
            --simple-toolbar-button-border-color,
            var(--simple-toolbar-border-color, #e0e0e0)
          );
          --simple-context-menu-border-color-dark: var(
            --simple-toolbar-button-border-color,
            var(--simple-toolbar-border-color, #444)
          );
          --simple-context-menu-shadow: var(
            --simple-toolbar-menu-list-box-shadow,
            0 2px 8px rgba(0, 0, 0, 0.15)
          );
          --simple-context-menu-min-width: 180px;
        }
        simple-toolbar-button {
          --simple-toolbar-button-border-radius: 0;
          --simple-toolbar-button-hover-border-color: transparent;
          cursor: pointer;
        }
        simple-toolbar-button.delete-button {
          border-top: var(--ddd-border-sm) solid
            var(--ddd-theme-default-limestoneGray);
          margin-top: var(--ddd-spacing-2);
          padding-top: var(--ddd-spacing-2);
        }
        simple-toolbar-button.delete-button:focus,
        simple-toolbar-button.delete-button:hover {
          color: black;
          background-color: var(--ddd-theme-default-discoveryCoral);
        }
      `]}static get properties(){return{...super.properties,actionId:{type:String,attribute:"action-id"},platformAllowsOutline:{type:Boolean},platformAllowsDelete:{type:Boolean},platformAllowsAddPage:{type:Boolean},isLocked:{type:Boolean},editMode:{type:Boolean}}}constructor(){super(),this.actionId=null,this.platformAllowsOutline=!0,this.platformAllowsDelete=!0,this.platformAllowsAddPage=!0,this.isLocked=!1,this.editMode=!1,this.t={...super.t,outlineActions:"Outline actions",editPage:"Edit page",addPage:"Add child page",moveUp:"Move up",moveDown:"Move down",indent:"Indent",outdent:"Outdent",moveTo:"Move to..",siteOutline:"Site Outline",delete:"Delete",unlock:"Unlock",lock:"Lock"},this.registerLocalization({context:this,basePath:import.meta.url}),a((()=>{const t=s(l.manifest),o=t&&t.metadata&&t.metadata.platform;this.platformAllowsOutline=!o||!1!==o.outlineDesigner,this.platformAllowsDelete=!o||!1!==o.delete,this.platformAllowsAddPage=!o||!1!==o.addPage})),a((()=>{const t=s(l.activeItem);this.isLocked=!!(t&&t.metadata&&t.metadata.locked)})),a((()=>{this.editMode=s(l.editMode)}))}_toggleDialog(){const t=this.shadowRoot.querySelector("simple-context-menu"),o=this.shadowRoot.querySelector(".ops");t&&o&&t.toggle(o)}render(){return this.platformAllowsOutline?t`
      <simple-icon-button-lite
        class="ops"
        icon="icons:create"
        label="${this.t.outlineActions}"
        title="${this.t.outlineActions}"
        @click="${this._toggleDialog}"
      ></simple-icon-button-lite>
      <simple-context-menu title="${this.t.outlineActions}">
        ${this.platformAllowsAddPage&&!this.editMode?t`
              <haxcms-button-add
                id="addpagebutton"
                class="top-bar-button"
                icon="hax:add-page"
                icon-position="left"
                label="${this.t.addPage}"
                merlin
                @click="${this._closeDialog}"
                show-text-label
                autofocus
              ></haxcms-button-add>
            `:""}
        <simple-toolbar-button
          icon="icons:arrow-upward"
          icon-position="left"
          align-horizontal="left"
          show-text-label
          label="${this.t.moveUp}"
          ?disabled="${this.isLocked}"
          @click=${()=>this._op("moveUp")}
        ></simple-toolbar-button>
        <simple-toolbar-button
          icon="icons:arrow-downward"
          icon-position="left"
          align-horizontal="left"
          show-text-label
          label="${this.t.moveDown}"
          ?disabled="${this.isLocked}"
          @click=${()=>this._op("moveDown")}
        ></simple-toolbar-button>
        <simple-toolbar-button
          icon="hax:outline-designer-indent"
          icon-position="left"
          align-horizontal="left"
          show-text-label
          label="${this.t.indent}"
          ?disabled="${this.isLocked}"
          @click=${()=>this._op("indent")}
        ></simple-toolbar-button>
        <simple-toolbar-button
          icon="hax:outline-designer-outdent"
          icon-position="left"
          align-horizontal="left"
          show-text-label
          label="${this.t.outdent}"
          ?disabled="${this.isLocked}"
          @click=${()=>this._op("outdent")}
        ></simple-toolbar-button>
        <simple-toolbar-button
          icon="icons:open-with"
          icon-position="left"
          align-horizontal="left"
          show-text-label
          label="${this.t.moveTo}"
          ?disabled="${this.isLocked}"
          @click=${this._movePageProgram}
        ></simple-toolbar-button>
        <simple-toolbar-button
          icon="hax:site-map"
          icon-position="left"
          align-horizontal="left"
          show-text-label
          label="${this.t.siteOutline}"
          @click=${this._openSiteOutline}
        ></simple-toolbar-button>
        ${this.isLocked?t`
              <simple-toolbar-button
                icon="icons:lock"
                icon-position="left"
                align-horizontal="left"
                show-text-label
                label="${this.t.unlock}"
                @click=${this._toggleLock}
              ></simple-toolbar-button>
            `:t``}
        ${this.platformAllowsDelete?t`
              <simple-toolbar-button
                class="delete-button"
                icon="icons:delete"
                icon-position="left"
                align-horizontal="left"
                show-text-label
                label="${this.t.delete}"
                ?disabled="${this.isLocked}"
                @click=${this._deletePage}
              ></simple-toolbar-button>
            `:""}
      </simple-context-menu>
    `:t``}_contextItem(){return(async()=>{let t={};return t=this.actionId&&"null"!=this.actionId?await l.findItemAsObject(this.actionId):"null"==this.actionId?s(await l.getLastChildItem(null)):s(l.activeItem),t})()}async _op(t){const o=await this._contextItem();if(!o||!o.id)return void this._closeDialog();l.playSound("click");const e={id:o.id,operation:t};this.dispatchEvent(new CustomEvent("haxcms-save-node-details",{bubbles:!0,composed:!0,cancelable:!0,detail:e})),this._closeDialog()}_editPage(){this._closeDialog(),l.cmsSiteEditor.haxCmsSiteEditorUIElement._editButtonTap()}_deletePage(){this._closeDialog(),l.cmsSiteEditor.haxCmsSiteEditorUIElement._deleteButtonTap()}async _movePageProgram(){this._closeDialog();const t=await this._contextItem();if(!t||!t.id)return;const o=globalThis.SuperDaemonManager.requestAvailability();l.playSound("click"),o.waveWand(["","/",{pageId:t.id},"move-page","Move Page",""])}_closeDialog(){const t=this.shadowRoot&&this.shadowRoot.querySelector("simple-context-menu");t&&t.close()}async _toggleLock(){this._closeDialog();const t=await this._contextItem();t&&t.id&&(l.playSound("click"),globalThis.dispatchEvent(new CustomEvent("haxcms-save-node-details",{bubbles:!0,composed:!0,cancelable:!0,detail:{id:t.id,operation:"setLocked",locked:!this.isLocked}})))}_openSiteOutline(){this._closeDialog(),l.cmsSiteEditor.haxCmsSiteEditorUIElement._outlineButtonTap()}}customElements.define(HAXCMSPageOperations.tag,HAXCMSPageOperations);
