// @lrnwebcomponents/grade-book/grade-book.js
define(["exports", "meta", "require", "../../lit/index.js", "../simple-colors/simple-colors.js", "../i18n-manager/lib/I18NMixin.js", "../utils/utils.js", "../utils/lib/gSheetsInterface.js", "../es-global-bridge/es-global-bridge.js", "../simple-fields/lib/simple-fields-field.js", "../simple-fields/lib/simple-fields-tag-list.js", "../a11y-collapse/a11y-collapse.js", "../a11y-collapse/lib/a11y-collapse-group.js", "../editable-table/lib/editable-table-display.js", "../a11y-tabs/a11y-tabs.js", "../a11y-tabs/lib/a11y-tab.js", "../grid-plate/grid-plate.js", "../iframe-loader/lib/loading-indicator.js", "./lib/letter-grade.js", "./lib/letter-grade-picker.js", "../../@github/time-elements/dist/index.js", "./lib/GradeBookUIPieces.js", "./lib/grade-book-store.js", "./lib/grade-book-pop-up.js", "../../mobx/dist/mobx.esm.js", "../../idb-keyval/dist/compat.js", "../file-system-broker/lib/xlsx-file-system-broker.js"], function (_exports, meta, _require, _index, _simpleColors, _I18NMixin, _utils, _gSheetsInterface, _esGlobalBridge, _simpleFieldsField, _simpleFieldsTagList, _a11yCollapse, _a11yCollapseGroup, _editableTableDisplay, _a11yTabs, _a11yTab, _gridPlate, _loadingIndicator, _letterGrade, _letterGradePicker, _index2, _GradeBookUIPieces, _gradeBookStore, _gradeBookPopUp, _mobxEsm, _compat, _xlsxFileSystemBroker) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.GradeBook = void 0;
  meta = _interopRequireWildcard(meta);
  _require = _interopRequireWildcard(_require);

  var _templateObject_18f031c0942811ecb1937969798b231a, _templateObject2_18f031c0942811ecb1937969798b231a, _templateObject3_18f031c0942811ecb1937969798b231a, _templateObject4_18f031c0942811ecb1937969798b231a, _templateObject5_18f031c0942811ecb1937969798b231a, _templateObject6_18f031c0942811ecb1937969798b231a, _templateObject7_18f031c0942811ecb1937969798b231a, _templateObject8_18f031c0942811ecb1937969798b231a, _templateObject9_18f031c0942811ecb1937969798b231a, _templateObject10_18f031c0942811ecb1937969798b231a, _templateObject11_18f031c0942811ecb1937969798b231a, _templateObject12_18f031c0942811ecb1937969798b231a, _templateObject13_18f031c0942811ecb1937969798b231a, _templateObject14_18f031c0942811ecb1937969798b231a, _templateObject15_18f031c0942811ecb1937969798b231a, _templateObject16_18f031c0942811ecb1937969798b231a, _templateObject17_18f031c0942811ecb1937969798b231a, _templateObject18_18f031c0942811ecb1937969798b231a, _templateObject19_18f031c0942811ecb1937969798b231a, _templateObject20_18f031c0942811ecb1937969798b231a, _templateObject21_18f031c0942811ecb1937969798b231a, _templateObject22_18f031c0942811ecb1937969798b231a, _templateObject23_18f031c0942811ecb1937969798b231a, _templateObject24_18f031c0942811ecb1937969798b231a, _templateObject25_18f031c0942811ecb1937969798b231a, _templateObject26_18f031c0942811ecb1937969798b231a, _templateObject27_18f031c0942811ecb1937969798b231a, _templateObject28_18f031c0942811ecb1937969798b231a, _templateObject29_18f031c0942811ecb1937969798b231a, _templateObject30_18f031c0942811ecb1937969798b231a, _templateObject31_18f031c0942811ecb1937969798b231a, _templateObject32_18f031c0942811ecb1937969798b231a, _templateObject33_18f031c0942811ecb1937969798b231a, _templateObject34_18f031c0942811ecb1937969798b231a, _templateObject35_18f031c0942811ecb1937969798b231a, _templateObject36_18f031c0942811ecb1937969798b231a, _templateObject37_18f031c0942811ecb1937969798b231a, _templateObject38_18f031c0942811ecb1937969798b231a, _templateObject39_18f031c0942811ecb1937969798b231a, _templateObject40_18f031c0942811ecb1937969798b231a, _templateObject41_18f031c0942811ecb1937969798b231a, _templateObject42_18f031c0942811ecb1937969798b231a, _templateObject43_18f031c0942811ecb1937969798b231a, _templateObject44_18f031c0942811ecb1937969798b231a;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

  function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

  function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `grade-book`
   * `A headless gradebook that supports multiple backends with rubrics`
   * @demo demo/index.html Grade book
   * @element grade-book
   */
  var GradeBook = /*#__PURE__*/function (_UIRenderPieces) {
    babelHelpers.inherits(GradeBook, _UIRenderPieces);

    var _super = _createSuper(GradeBook);

    function GradeBook() {
      var _this;

      babelHelpers.classCallCheck(this, GradeBook);
      _this = _super.call(this);
      _this.hasFilePicker = false;
      _this.source = "googledocs";

      if (window.showOpenFilePicker) {
        _this.source = "filesystem";
        _this.hasFilePicker = true;
      }

      _this.ready = false; // 0,1 - column splits for 3 column
      // 2 - renders in a new popped up window

      _this.displayMode = 0; // storing internals of the assessmentView tab

      _this.assessmentView = _this.resetAssessmentView();
      _this.totalScore = 0; // student submission status for rendering

      _this.activeStudentSubmissions = []; // lock on score override

      _this.scoreLock = true; // active rubric data

      _this.activeRubric = []; // the active grade sheet

      _this.activeGrading = {}; // general state

      _this.settings = {
        photo: true,
        fname: true,
        surname: true,
        email: true
      };
      _this.disabled = false; // shows progress indicator as it loads

      _this.loading = false; // translatable text

      _this.t = {
        csvURL: "CSV URL",
        points: "Points",
        criteria: "Criteria",
        description: "Description",
        assessmentWeight: "Assessment weight",
        overallFeedback: "Overall feedback",
        letterGrade: "Letter grade",
        highRange: "High range",
        lowRange: "Low range",
        noSubmission: "No submission found",
        studentSubmission: "Student submission",
        openInNewWindow: "Open in new window",
        gradingScale: "Grading scale",
        activeStudent: "Active student",
        activeAssignment: "Active assignment",
        submitted: "Submitted",
        dueDate: "Due date",
        firstName: "First name",
        surname: "Surname",
        photo: "Photo",
        email: "Email",
        settings: "Settings",
        previous: "Previous",
        next: "Next",
        previousStudent: "Previous student",
        nextStudent: "Next student",
        previousAssignment: "Previous assignment",
        nextAssignment: "Next assignment",
        student: "Student",
        assessmentView: "Assessment View",
        activeAssessment: "Active assessment",
        studentReportView: "Student report view",
        loadGradebook: "Load gradebook",
        load: "Load",
        saveGradebook: "Save gradebook",
        selectGradebookSource: "Select gradebook source",
        sourceData: "Source data",
        pasteValidJSONHere: "Paste valid JSON here",
        loadingFilePleaseWait: "LOADING FILE PLEASE WAIT.."
      };

      _this.registerLocalization({
        context: babelHelpers.assertThisInitialized(_this),
        basePath: meta.url,
        locales: ["es", "fr", "de"]
      }); // notice that a category on the active grading area responded that it changed


      _this.addEventListener("simple-fields-tag-list-changed", _this.qualitativeFeedbackUpdate.bind(babelHelpers.assertThisInitialized(_this))); // value change within the rubric area


      _this.addEventListener("value-changed", _this.rubricCriteriaPointsChange); // drop event to remove the styling from droppable areas


      _this.addEventListener("drop", _this._handleDragDrop.bind(babelHelpers.assertThisInitialized(_this)));

      (0, _mobxEsm.autorun)(function () {
        _this.activeStudent = (0, _mobxEsm.toJS)(_gradeBookStore.GradeBookStore.activeStudent);
      });
      (0, _mobxEsm.autorun)(function () {
        _this.activeAssignment = (0, _mobxEsm.toJS)(_gradeBookStore.GradeBookStore.activeAssignment);
      });
      (0, _mobxEsm.autorun)(function () {
        _this.database = (0, _mobxEsm.toJS)(_gradeBookStore.GradeBookStore.database);
      });
      (0, _mobxEsm.autorun)(function () {
        _this.activeSubmission = (0, _mobxEsm.toJS)(_gradeBookStore.GradeBookStore.activeSubmission);
      });
      return _this;
    }

    babelHelpers.createClass(GradeBook, [{
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        var _this2 = this;

        if (babelHelpers.get(babelHelpers.getPrototypeOf(GradeBook.prototype), "firstUpdated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(GradeBook.prototype), "firstUpdated", this).call(this, changedProperties);
        }

        var resizeObserver = new ResizeObserver(function (entries) {
          clearTimeout(_this2.__resizer);
          _this2.__resizer = setTimeout(function () {
            var pixels = 0;

            var _iterator = _createForOfIteratorHelper(entries),
                _step;

            try {
              for (_iterator.s(); !(_step = _iterator.n()).done;) {
                var entry = _step.value;

                if (entry.contentBoxSize) {
                  pixels = Math.round(window.innerHeight - entry.contentBoxSize[0].blockSize - 122);
                } else {
                  pixels = Math.round(window.innerHeight - entry.contentRect.height - 122);
                }
              }
            } catch (err) {
              _iterator.e(err);
            } finally {
              _iterator.f();
            }

            _this2.shadowRoot.querySelector("#studentassessment").shadowRoot.querySelector("[part='content']").style.height = pixels + "px";
          }, 50);
        });
        resizeObserver.observe(this.shadowRoot.querySelector("#studentgrid")); // see if we have a previous file reference

        setTimeout( /*#__PURE__*/babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
          return regeneratorRuntime.wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  _context.next = 2;
                  return (0, _compat.get)("grade-book-prev-file");

                case 2:
                  _this2.prevLocalFileReference = _context.sent;

                  _this2.requestUpdate();

                case 4:
                case "end":
                  return _context.stop();
              }
            }
          }, _callee);
        })), 0);
      }
    }, {
      key: "resetAssessmentView",
      value: function resetAssessmentView() {
        return {
          qualitative: [],
          written: []
        };
      }
      /**
       * Return an object representing all scores on all assignments
       * which allows us to make a visual of all submissions this student
       * has had.
       */

    }, {
      key: "getStudentSubmissions",
      value: function getStudentSubmissions(activeStudent) {
        var response = [];

        for (var i in this.database.submissions) {
          var row = this.database.submissions[i]; // look for student, need a match before we render anything

          if (row.student === this.database.roster[activeStudent].student) {
            for (var j in row) {
              if (j !== "student") {
                var a = this.getAssignmentByShortName(j);

                if (a) {
                  response.push({
                    studentScore: !this.database.grades[activeStudent] || this.database.grades[activeStudent][j] == "" ? null : this.database.grades[activeStudent][j],
                    assignmentPoints: a.points,
                    assignmentName: a.name,
                    assignmentIndex: a.index
                  });
                }
              }
            }

            return response;
          }
        }

        return response;
      }
    }, {
      key: "getAssignmentByShortName",
      value: function getAssignmentByShortName(name) {
        var index;
        var item = this.database.assignments.filter(function (i, ind) {
          if (i.shortName === name) {
            index = ind;
            return true;
          }

          return false;
        });

        if (item.length === 1) {
          item[0].index = index;
          return item[0];
        }

        return null;
      } // generate current score based on student / assignment cross-section

    }, {
      key: "getCurrentScore",
      value: function getCurrentScore(activeStudent, activeAssignment) {
        // see if there's a score set in the grades setup
        if (this.database.grades[activeStudent] && this.database.grades[activeStudent][this.database.assignments[activeAssignment].shortName]) {
          return this.database.grades[activeStudent][this.database.assignments[activeAssignment].shortName];
        }

        return 0;
      } // return the active rurbic based on active assignment

    }, {
      key: "getActiveRubric",
      value: function getActiveRubric() {
        var _this3 = this;

        return this.database.rubrics.filter(function (item) {
          return item.shortName == _this3.database.assignments[_this3.activeAssignment].rubric;
        });
      }
      /**
       * LitElement life cycle for property change notification
       */

    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this4 = this;

        if (babelHelpers.get(babelHelpers.getPrototypeOf(GradeBook.prototype), "updated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(GradeBook.prototype), "updated", this).call(this, changedProperties);
        }

        changedProperties.forEach(function (oldValue, propName) {
          // set rubric based on assignment
          if (!_this4.loading && ["activeAssignment", "activeStudent", "database", "loading"].includes(propName)) {
            setTimeout(function () {
              // this will defer to whatever the "grades" db value is
              _this4.totalScore = _this4.getCurrentScore(_this4.activeStudent, _this4.activeAssignment);
              _this4.assessmentView = _this4.resetAssessmentView();
              _this4.activeRubric = babelHelpers.toConsumableArray(_this4.getActiveRubric());
              _this4.hideRubricInfo = babelHelpers.toConsumableArray(_this4.activeRubric.map(function () {
                return false;
              }));
              _this4.activeStudentSubmissions = [];
              _this4.activeStudentSubmissions = babelHelpers.toConsumableArray(_this4.getStudentSubmissions(_this4.activeStudent));

              if (_this4.displayMode == 2) {
                _this4.renderSubmissionInWindow();
              }
            }, 0);

            _this4.maintainScrollPosition();
          } // source will have to fetch ALL the pages and slowly load data as it rolls through


          if (["sourceData", "source"].includes(propName)) {
            if (_this4.sourceData) {
              switch (_this4.source) {
                case "googledocs":
                  _this4.loading = true; // gid from the google sheet. technically if you add / remove sheets this would
                  // have to be updated to match

                  _this4.gSheet = new _gSheetsInterface.gSheetInterface(_this4, _this4.sourceData, {
                    tags: 0,
                    roster: 118800528,
                    assignments: 540222065,
                    rubrics: 1744429439,
                    submissions: 2104732668,
                    gradeScale: 980501320,
                    grades: 2130903440,
                    gradesDetails: 644559151,
                    settings: 1413275461
                  });
                  setTimeout( /*#__PURE__*/babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2() {
                    var i, loadedData;
                    return regeneratorRuntime.wrap(function _callee2$(_context2) {
                      while (1) {
                        switch (_context2.prev = _context2.next) {
                          case 0:
                            _context2.t0 = regeneratorRuntime.keys(_this4.gSheet.sheetGids);

                          case 1:
                            if ((_context2.t1 = _context2.t0()).done) {
                              _context2.next = 13;
                              break;
                            }

                            i = _context2.t1.value;
                            _context2.next = 5;
                            return _this4.gSheet.loadSheetData(i);

                          case 5:
                            loadedData = _context2.sent;
                            loadedData = _this4.transformTable(loadedData);

                            if (typeof _this4["process".concat(i, "Data")] === "function") {
                              loadedData = _this4["process".concat(i, "Data")](loadedData);
                            }

                            _this4.database[i] = loadedData;

                            if (i === "gradeScale" && _this4.database.gradeScale && _this4.database.gradeScale.length > 0) {
                              _gradeBookStore.GradeBookStore.gradeScale = _this4.database.gradeScale;
                            }

                            _this4.requestUpdate();

                            _context2.next = 1;
                            break;

                          case 13:
                            _this4.importStateCleanup();

                          case 14:
                          case "end":
                            return _context2.stop();
                        }
                      }
                    }, _callee2);
                  })), 0);
                  break;

                case "url":
                  _this4.loading = true;
                  fetch(_this4.sourceData).then(function (response) {
                    if (response.ok) {
                      return response.json();
                    }
                  }).then(function (json) {
                    _gradeBookStore.GradeBookStore.database = json;

                    _this4.importStateCleanup();
                  }).catch(function (error) {
                    console.warn(error);
                  });
                  break;

                case "json":
                  _gradeBookStore.GradeBookStore.database = JSON.parse(_this4.sourceData);

                  _this4.importStateCleanup();

                  break;
              }
            } else if (_this4.source == "filesystem") {
              if (!_this4.__applied) {
                _this4.__applied = true; // this listener gets the event from the service worker

                window.addEventListener("xlsx-file-system-data", function (e) {
                  var database = e.detail.data;

                  for (var i in database) {
                    var loadedData = _this4.transformTable(database[i]);

                    if (typeof _this4["process".concat(i, "Data")] === "function") {
                      loadedData = _this4["process".concat(i, "Data")](loadedData);
                    }

                    _gradeBookStore.GradeBookStore.database[i] = loadedData;
                  }

                  _this4.importStateCleanup();
                });
              }
            }
          }
        });
      }
    }, {
      key: "importStateCleanup",
      value: function importStateCleanup() {
        this.loading = false;

        if (this.database.gradeScale && this.database.gradeScale.length > 0) {
          _gradeBookStore.GradeBookStore.gradeScale = this.database.gradeScale;
        }

        this.requestUpdate();
        this.ready = true;
      }
    }, {
      key: "transformTable",
      value: function transformTable(table) {
        var tmp = table.shift();
        var headings = {};
        var data = [];

        for (var i in tmp) {
          headings[tmp[i]] = i;
        }

        for (var i in table) {
          var item = {};

          for (var j in headings) {
            item[j] = table[i][headings[j]];
          } // push data onto the database of all data we have now as objects


          data.push(item);
        } // allow for deeper processing on the data or just return the data found


        return data;
      }
      /**
       * maintain scroll position within the table overflow scroll whenever assignment or student changes
       */

    }, {
      key: "maintainScrollPosition",
      value: function maintainScrollPosition() {
        var _this5 = this;

        // ensure we maintain visibility of the active student / assignment
        // in our studentgrid; delay to ensure paints before visibility
        if (this.shadowRoot && this.shadowRoot.querySelector("#studentgrid [data-active]")) {
          setTimeout(function () {
            var offset = 192;
            var isVisible = (0, _utils.isElementInViewport)(_this5.shadowRoot.querySelector("#studentgrid [data-active]"), {
              top: 0,
              right: window.innerWidth,
              bottom: _this5.shadowRoot.querySelector("#studentgrid").getBoundingClientRect().height + 20,
              left: offset
            });

            if (_gradeBookStore.GradeBookStore.activeStudent === 0) {
              _this5.shadowRoot.querySelector("#studentgrid").scrollTop = 0;
            } else if (!isVisible) {
              _this5.shadowRoot.querySelector("#studentgrid").scrollTop = _this5.shadowRoot.querySelector("#studentgrid [data-active]").offsetTop - _this5.shadowRoot.querySelector("#studentgrid").offsetTop;
            } // left to right


            if (_gradeBookStore.GradeBookStore.activeAssignment === 0) {
              _this5.shadowRoot.querySelector("#studentgrid").scrollLeft = 0;
            } else if (!isVisible) {
              _this5.shadowRoot.querySelector("#studentgrid").scrollLeft = _this5.shadowRoot.querySelector("#studentgrid [data-active]").offsetLeft - offset - _this5.shadowRoot.querySelector("#studentgrid").offsetLeft;
            }
          }, 0);
        }
      }
      /**
       * process assignment data to normalize date string
       */

    }, {
      key: "processassignmentsData",
      value: function processassignmentsData(data) {
        for (var i in data) {
          if (data[i].dueDate != "") {
            try {
              var event = new Date("".concat(data[i].dueDate, " ").concat(data[i].dueTime));
              data[i]._ISODueDate = event.toISOString();
            } catch (e) {}
          }
        }

        return data;
      }
      /**
       * Process our tagging structure for use in the rubric
       * Tag structure allows the instructor to drag and drop elements into
       * qualitative areas of a rubric
       */

    }, {
      key: "processtagsData",
      value: function processtagsData(data) {
        var categories = new Set([]);
        var rCategories = []; // these must all exist as keys for us to proceed

        for (var i in data) {
          data[i].category = data[i].category ? data[i].category.split(",") : [];
          data[i].associatedMaterial = data[i].associatedMaterial ? data[i].associatedMaterial.split(",") : []; // trick to dedup the categories using a Set

          data[i].category.forEach(function (item) {
            categories.add(item);
          }); // convert Set to Array for data visualization purposes

          rCategories = babelHelpers.toConsumableArray(Array.from(categories));
        }

        return {
          categories: rCategories,
          data: data
        };
      }
    }, {
      key: "processrubricsData",
      value: function processrubricsData(data) {
        for (var i in data) {
          data[i].qualitative = data[i].qualitative ? data[i].qualitative.split(",") : [];
        }

        return data;
      }
    }, {
      key: "processrosterData",
      value: function processrosterData(data) {
        for (var i in data) {
          data[i].interests = data[i].interests ? data[i].interests.split(",") : [];
        }

        return data;
      }
    }, {
      key: "processsettingsData",
      value: function processsettingsData(data) {
        var d = {};

        for (var i in data) {
          d[data[i].key] = data[i].value;
        }

        return d;
      }
    }, {
      key: "changeStudent",
      value: function () {
        var _changeStudent = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3(e) {
          return regeneratorRuntime.wrap(function _callee3$(_context3) {
            while (1) {
              switch (_context3.prev = _context3.next) {
                case 0:
                  this.shadowRoot.querySelectorAll("#studentgrid .th-or-td").forEach(function (el) {
                    el.classList.remove("col-highlight");
                  }); // have to possibly resolve UI click handler of span vs the button

                  if (e.target.getAttribute("value") == "prev" && 0 !== this.activeStudent) {
                    _gradeBookStore.GradeBookStore.activeStudent--;
                  } else if (e.target.getAttribute("value") == "next" && this.database.roster.length - 1 !== this.activeStudent) {
                    _gradeBookStore.GradeBookStore.activeStudent++;
                  }

                  _context3.next = 4;
                  return this.requestUpdate();

                case 4:
                case "end":
                  return _context3.stop();
              }
            }
          }, _callee3, this);
        }));

        function changeStudent(_x) {
          return _changeStudent.apply(this, arguments);
        }

        return changeStudent;
      }()
    }, {
      key: "changeAssignment",
      value: function () {
        var _changeAssignment = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee4(e) {
          return regeneratorRuntime.wrap(function _callee4$(_context4) {
            while (1) {
              switch (_context4.prev = _context4.next) {
                case 0:
                  this.shadowRoot.querySelectorAll("#studentgrid .th-or-td").forEach(function (el) {
                    el.classList.remove("col-highlight");
                  }); // have to possibly resolve UI click handler of span vs the button

                  if (e.target.getAttribute("value") == "prev" && 0 !== this.activeAssignment) {
                    _gradeBookStore.GradeBookStore.activeAssignment--;
                  } else if (e.target.getAttribute("value") == "next" && this.database.assignments.length - 1 !== _gradeBookStore.GradeBookStore.activeAssignment) {
                    _gradeBookStore.GradeBookStore.activeAssignment++;
                  }

                  _context4.next = 4;
                  return this.requestUpdate();

                case 4:
                case "end":
                  return _context4.stop();
              }
            }
          }, _callee4, this);
        }));

        function changeAssignment(_x2) {
          return _changeAssignment.apply(this, arguments);
        }

        return changeAssignment;
      }()
    }, {
      key: "renderSubmission",
      value: // render submission; guessing game really :)
      function renderSubmission(data) {
        var height = parseInt(this.shadowRoot.querySelector("#studentgrid").getBoundingClientRect().height);
        this.shadowRoot.querySelector("#studentassessment").shadowRoot.querySelector("[part='content']").style.height = Math.round(window.innerHeight - height - 104) + "px"; // @todo make this actually be the submitted time, right now it's due relative time'd

        var pre = (0, _index.html)(_templateObject_18f031c0942811ecb1937969798b231a || (_templateObject_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<h3>", "</h3>\n      ", "\n      <relative-time\n        datetime=\"", "\"\n      ></relative-time> "])), this.t.studentSubmission, this.t.submitted, this.database.assignments[this.activeAssignment]._ISODueDate); // test if this smells like a URL

        if ((0, _utils.validURL)(data)) {
          pre = (0, _index.html)(_templateObject2_18f031c0942811ecb1937969798b231a || (_templateObject2_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["", "<a\n          href=\"", "\"\n          target=\"_blank\"\n          rel=\"noopener noreferrer\"\n          ><simple-icon-button-lite\n            label=\"", "\"\n            icon=\"open-in-new\"\n          ></simple-icon-button-lite\n        ></a>"])), pre, data, this.t.openInNewWindow); // see if this is a video we know about

          if (data != (0, _utils.cleanVideoSource)(data)) {
            // implies it was able to clean it up in some way
            new Promise(function (res, rej) {
              return _require.default(["../video-player/video-player.js"], res, rej);
            });
            return (0, _index.html)(_templateObject3_18f031c0942811ecb1937969798b231a || (_templateObject3_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["", "<video-player\n            class=\"active-submission\"\n            source=\"", "\"\n            width=\"60%\"\n          ></video-player>"])), pre, data);
          } else {
            return (0, _index.html)(_templateObject4_18f031c0942811ecb1937969798b231a || (_templateObject4_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["", "\n          <div class=\"active-submission\">\n            <iframe\n              src=\"", "\"\n              loading=\"lazy\"\n              width=\"100%\"\n              height=\"100%\"\n            ></iframe>\n          </div>"])), pre, data);
          }
        } else {
          // see if we can just present this as data
          new Promise(function (res, rej) {
            return _require.default(["../md-block/md-block.js"], res, rej);
          });
          return (0, _index.html)(_templateObject5_18f031c0942811ecb1937969798b231a || (_templateObject5_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["", "\n        <div class=\"active-submission\">\n          <md-block .markdown=\"", "\"></md-block>\n        </div>"])), pre, data);
        }
      }
    }, {
      key: "studentLetterGradeHistoryClick",
      value: function studentLetterGradeHistoryClick(e) {
        // ensure this is numeric
        _gradeBookStore.GradeBookStore.activeAssignment = parseInt(e.target.value);
        this.requestUpdate();
      }
    }, {
      key: "activateOption",
      value: function activateOption(e) {
        var target = (0, _utils.normalizeEventPath)(e);

        if (target[0].getAttribute("data-student") && target[0].getAttribute("data-assignment")) {
          _gradeBookStore.GradeBookStore.activeAssignment = parseInt(target[0].getAttribute("data-assignment"));
          _gradeBookStore.GradeBookStore.activeStudent = parseInt(target[0].getAttribute("data-student")); // if we collapsed this but then select a specific assignment / student
          // then let's focus the user on this

          this.shadowRoot.querySelector("#studentgrid").style.height = "140px";
          this.maintainScrollPosition();
        }
      }
    }, {
      key: "mouseHighlight",
      value: function mouseHighlight(e) {
        var _this6 = this;

        var active = (0, _utils.normalizeEventPath)(e)[0];
        clearTimeout(this.__mdebounce);
        this.__mdebounce = setTimeout(function () {
          if (active && active.getAttribute("data-assignment") != _this6.__activeHoverAssignment) {
            _this6.shadowRoot.querySelectorAll('#studentgrid .th-or-td[data-assignment="' + _this6.__activeHoverAssignment + '"]').forEach(function (el) {
              el.classList.remove("col-highlight");
            }); // set active so we can clear it on previous


            _this6.__activeHoverAssignment = active.getAttribute("data-assignment");

            _this6.shadowRoot.querySelectorAll('#studentgrid .th-or-td[data-assignment="' + _this6.__activeHoverAssignment + '"]').forEach(function (el) {
              el.classList.add("col-highlight");
            });
          }
        }, 10);
      }
    }, {
      key: "mouseLeave",
      value: function mouseLeave(e) {
        var _this7 = this;

        clearTimeout(this.__mdebounce);
        this.__mdebounce = setTimeout(function () {
          _this7.shadowRoot.querySelectorAll("#studentgrid .th-or-td").forEach(function (el) {
            el.classList.remove("col-highlight");
          });
        }, 100);
      }
    }, {
      key: "settingChanged",
      value: function settingChanged(e) {
        e.stopPropagation();
        e.stopImmediatePropagation();
        e.preventDefault();
        this.settings[e.detail.name] = e.detail.value;
        this.settings = _objectSpread({}, this.settings);
      }
    }, {
      key: "handleGridScaling",
      value: function handleGridScaling(e) {
        var paths = (0, _utils.normalizeEventPath)(e);

        if (paths[0].tagName === "TABLE") {
          // toggle between height
          if (paths[0].style.height != "") {
            paths[0].style.height = null;
          } else {
            paths[0].style.height = "90vh";
          }
        }
      }
    }, {
      key: "checkTabHeight",
      value: function checkTabHeight(e) {
        // toggle between height
        var height = parseInt(this.shadowRoot.querySelector("#studentgrid").getBoundingClientRect().height); // ensure small heights are expanded because it's impossible to leverage overwise

        if (height / window.innerHeight > 0.65) {
          this.shadowRoot.querySelector("#studentgrid").style.height = null;
        }
      }
    }, {
      key: "studentreportClick",
      value: function studentreportClick(e) {
        var _this8 = this;

        _esGlobalBridge.ESGlobalBridgeStore.import("jspdf", new URL("./lib/", meta.url).href + "jspdf.min.js").then(function () {
          var pdf = new jsPDF();
          pdf.fromHTML(_this8.shadowRoot.querySelector("#studentreport").outerHTML, 15, 15);
          var cd = new Date();
          var dateTime = "".concat(cd.getFullYear(), "-").concat(cd.getMonth(), "-").concat(cd.getDate(), "__").concat(cd.getHours(), "-").concat(cd.getMinutes(), "-").concat(cd.getSeconds());
          var fname = "".concat(_this8.database.roster[_this8.activeStudent].student, "--").concat(_this8.database.assignments[_this8.activeAssignment].shortName, "--").concat(dateTime, ".pdf");
          pdf.save(fname);
        });
      } // open extra window and then render content

    }, {
      key: "openWindow",
      value: function openWindow(e) {
        var _this9 = this;

        if (this.__openWindow && !this.__openWindow.closed) {
          this.__openWindow.focus();
        } else {
          this.__openWindow = window.open("", "studentwork", "left=0,top=0,width=".concat(screen.width / 2, ",height=").concat(screen.height / 2, ",menubar=0,location=0,toolbar=0,status=0"));

          this.__openWindow.onbeforeunload = function () {
            _this9.displayMode = 0;
            _this9.__openWindow = null;
          };
        }

        this.renderSubmissionInWindow();
      } // render submission in window area

    }, {
      key: "renderSubmissionInWindow",
      value: function renderSubmissionInWindow() {
        if (this.__openWindow && this.displayMode === 2) {
          this.__openWindow.document.body.style.margin = "0";
          (0, _index.render)((0, _index.html)(_templateObject6_18f031c0942811ecb1937969798b231a || (_templateObject6_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral([" ", ""])), this.database.assignments.length && this.database.assignments[this.activeAssignment] ? (0, _index.html)(_templateObject7_18f031c0942811ecb1937969798b231a || (_templateObject7_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<grade-book-pop-up></grade-book-pop-up>"]))) : (0, _index.html)(_templateObject8_18f031c0942811ecb1937969798b231a || (_templateObject8_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["", ""])), this.t.noSubmission)), this.__openWindow.document.body);
        }
      }
      /**
       * LitElement render method
       */

    }, {
      key: "render",
      value: function render() {
        var _this10 = this;

        return (0, _index.html)(_templateObject9_18f031c0942811ecb1937969798b231a || (_templateObject9_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <loading-indicator full ?loading=\"", "\"></loading-indicator>\n      <div class=\"top-controls\">\n        <div class=\"group divider-right app-name\">\n          <slot name=\"app-name\"></slot>\n        </div>\n        <div class=\"group divider-right\">\n          ", "\n          <simple-icon-button-lite\n            @click=\"", "\"\n            value=\"prev\"\n            title=\"", "\"\n            ?disabled=\"", "\"\n            icon=\"arrow-upward\"\n          >\n            <span class=\"hide-900\" value=\"prev\">", "</span>\n          </simple-icon-button-lite>\n          <simple-icon-button-lite\n            @click=\"", "\"\n            value=\"next\"\n            title=\"", "\"\n            icon=\"arrow-downward\"\n            ?disabled=\"", "\"\n          >\n            <span class=\"hide-900\" value=\"next\">", "</span>\n          </simple-icon-button-lite>\n        </div>\n        <div class=\"group divider-right\">\n          ", "\n          <simple-icon-button-lite\n            @click=\"", "\"\n            value=\"prev\"\n            title=\"", "\"\n            icon=\"arrow-back\"\n            ?disabled=\"", "\"\n          >\n            <span class=\"hide-900\" value=\"prev\">", "</span>\n          </simple-icon-button-lite>\n          <simple-icon-button-lite\n            @click=\"", "\"\n            value=\"next\"\n            title=\"", "\"\n            ?disabled=\"", "\"\n            icon=\"arrow-forward\"\n          >\n            <span class=\"hide-900\" value=\"next\">", "</span>\n          </simple-icon-button-lite>\n        </div>\n        <div class=\"group\">", "</div>\n        <div class=\"group\">", "</div>\n        ", "\n      </div>\n      <div ?hidden=\"", "\" class=\"source-selection\">\n        <label>", "..</label>\n        <select id=\"source\" @change=\"", "\">\n          ", "\n          <option\n            value=\"googledocs\"\n            ?selected=\"", "\"\n          >\n            Google docs\n          </option>\n          <option value=\"url\">URL endpoint (JSON)</option>\n          <option value=\"json\">JSON data blob</option>\n        </select>\n        <simple-icon-button-lite\n          @click=\"", "\"\n          title=\"", "\"\n          ?disabled=\"", "\"\n          icon=\"folder-shared\"\n        >\n          <span class=\"hide-900\">", "</span>\n        </simple-icon-button-lite>\n        ", "\n        <input\n          id=\"sourcedata\"\n          placeholder=\"", "\"\n          ?hidden=\"", "\"\n        />\n        <textarea\n          id=\"sourcedatablob\"\n          rows=\"10\"\n          cols=\"40\"\n          placeholder=\"", "\"\n          ?hidden=\"", "\"\n        ></textarea>\n      </div>\n      ", "\n      <table\n        id=\"studentgrid\"\n        @mousemove=\"", "\"\n        @mouseleave=\"", "\"\n        @click=\"", "\"\n        @dblclick=\"", "\"\n        bordered\n        column-header\n        condensed\n        disable-responsive\n        scroll\n        striped\n        numeric-styles\n        sort\n        ?hidden=\"", "\"\n      >\n        ", "\n      </table>\n      <a11y-tabs\n        id=\"studentassessment\"\n        full-width\n        @click=\"", "\"\n        @a11y-tabs-active-changed=\"", "\"\n        ?hidden=\"", "\"\n      >\n        <a11y-tab\n          id=\"assessment\"\n          icon=\"assignment-ind\"\n          label=\"", "\"\n        >\n          ", "\n        </a11y-tab>\n        <a11y-tab\n          id=\"studentreporttab\"\n          icon=\"assignment\"\n          label=\"", "\"\n        >\n          <simple-icon-button-lite\n            @click=\"", "\"\n            title=\"Download PDF\"\n            icon=\"image:picture-as-pdf\"\n          >\n            <span class=\"hide-900\" value=\"prev\">Download PDF</span>\n          </simple-icon-button-lite>\n          <div id=\"studentreport\">\n            ", "\n          </div>\n        </a11y-tab>\n      </a11y-tabs>\n    "])), this.loading, this.renderActiveStudentBtn(), this.changeStudent, this.t.previousStudent, 0 === this.activeStudent || !this.ready, this.t.previous, this.changeStudent, this.t.nextStudent, this.database.roster.length - 1 === this.activeStudent || !this.ready, this.t.next, this.renderActiveAssignmentBtn(), this.changeAssignment, this.t.previousAssignment, 0 === this.activeAssignment || !this.ready, this.t.previous, this.changeAssignment, this.t.nextAssignment, this.database.assignments.length - 1 === this.activeAssignment || !this.ready, this.t.next, this.renderGradeScaleBtn(), this.renderSettingsBtn(), this.ready && this.source === "filesystem" && this.sourceData ? (0, _index.html)(_templateObject10_18f031c0942811ecb1937969798b231a || (_templateObject10_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n              <div class=\"group\">\n                <simple-icon-button-lite\n                  @click=\"", "\"\n                  title=\"", "\"\n                  ?disabled=\"", "\"\n                  icon=\"save\"\n                >\n                  <span class=\"hide-900\">", "</span>\n                </simple-icon-button-lite>\n              </div>\n            "])), this.saveToFilesystem, this.t.saveGradebook, !this.sourceData, this.t.saveGradebook) : _index.nothing, this.sourceData, this.t.selectGradebookSource, this.selectSource, this.hasFilePicker ? (0, _index.html)(_templateObject11_18f031c0942811ecb1937969798b231a || (_templateObject11_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<option\n                value=\"filesystem\"\n                ?selected=\"", "\"\n              >\n                File system\n              </option>"])), this.source === "filesystem") : _index.nothing, this.source === "googledocs", this.loadFromSource, this.t.loadGradebook, this.sourceData, this.t.loadGradebook, this.source == "filesystem" && this.prevLocalFileReference ? (0, _index.html)(_templateObject12_18f031c0942811ecb1937969798b231a || (_templateObject12_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n              <simple-icon-button-lite\n                @click=\"", "\"\n                title=\"", "\"\n                ?disabled=\"", "\"\n                icon=\"folder-shared\"\n              >\n                <span class=\"hide-900\"\n                  >", " ", "</span\n                >\n              </simple-icon-button-lite>\n            "])), this.loadFromExistingSource, this.t.loadGradebook, this.sourceData, this.t.load, this.prevLocalFileReference.name) : _index.nothing, this.t.sourceData, !["googledocs", "url"].includes(this.source), this.t.pasteValidJSONHere, this.source != "json", this.source === "filesystem" && this.loading ? (0, _index.html)(_templateObject13_18f031c0942811ecb1937969798b231a || (_templateObject13_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<p class=\"source-selection\">", "</p>"])), this.t.loadingFilePleaseWait) : _index.nothing, this.mouseHighlight, this.mouseLeave, this.activateOption, this.handleGridScaling, !(this.database.assignments && this.database.assignments[this.activeAssignment]), this.database.roster.length && this.database.assignments.length ? (0, _index.html)(_templateObject14_18f031c0942811ecb1937969798b231a || (_templateObject14_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n              <colgroup>\n                <col />\n              </colgroup>\n              ", "\n              <thead class=\"thead\">\n                <tr class=\"tr thead-tr\" part=\"tr\">\n                  <th\n                    class=\"th th-or-td student-table-label\"\n                    data-assignment=\"-1\"\n                  >\n                    <div>\n                      ", " (", " /\n                      ", ")\n                    </div>\n                  </th>\n                  ", "\n                </tr>\n              </thead>\n              <tbody class=\"tbody\">\n                ", "\n              </tbody>\n            "])), this.database.assignments.map(function (a, h) {
          return (0, _index.html)(_templateObject15_18f031c0942811ecb1937969798b231a || (_templateObject15_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<colgroup>\n                    <col />\n                  </colgroup>"])));
        }), this.t.student, parseInt(this.activeStudent + 1), this.database.roster.length, this.database.assignments.map(function (a, h) {
          return (0, _index.html)(_templateObject16_18f031c0942811ecb1937969798b231a || (_templateObject16_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<th\n                        class=\"th th-or-td assignment-name\"\n                        title=\"", "\"\n                        data-assignment=\"", "\"\n                      >\n                        <!-- @todo add once we support multi-select <simple-fields-field\n                          data-assignment=\"", "\"\n                          type=\"checkbox\"\n                          class=\"select-all\"\n                          title=\"Select all submissions ", "\"\n                          name=\"select-all-submission\"\n                          aria-label=\"", "\"\n                        ></simple-fields-field> -->\n                        ", "\n                      </th>"])), a.name, h, h, a.shortName, a.shortName, a.shortName);
        }), this.database.roster.map(function (s, i) {
          return (0, _index.html)(_templateObject17_18f031c0942811ecb1937969798b231a || (_templateObject17_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral([" <tr class=\"tr tbody-tr\">\n                    <th class=\"th th-or-td\" data-assignment=\"-1\">\n                      <div class=\"user-info\">\n                        <div class=\"user-left\">\n                          ", "\n                        </div>\n                        <div class=\"user-right\">\n                          ", "\n                          ", "\n                          ", "\n                        </div>\n                      </div>\n                    </th>\n                    ", "\n                  </tr>"])), _this10.settings.photo ? (0, _index.html)(_templateObject18_18f031c0942811ecb1937969798b231a || (_templateObject18_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["", ""])), s.photo ? (0, _index.html)(_templateObject19_18f031c0942811ecb1937969798b231a || (_templateObject19_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<img\n                                    src=\"", "\"\n                                    loading=\"lazy\"\n                                    class=\"user-photo\"\n                                  />"])), s.photo) : (0, _index.html)(_templateObject20_18f031c0942811ecb1937969798b231a || (_templateObject20_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<simple-icon-lite\n                                    icon=\"social:person\"\n                                    class=\"user-photo\"\n                                  ></simple-icon-lite>"])))) : _index.nothing, _this10.settings.fname ? (0, _index.html)(_templateObject21_18f031c0942811ecb1937969798b231a || (_templateObject21_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<div>", "</div>"])), s.student) : _index.nothing, _this10.settings.surname ? (0, _index.html)(_templateObject22_18f031c0942811ecb1937969798b231a || (_templateObject22_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<div>", "</div>"])), s.student) : _index.nothing, _this10.settings.email ? (0, _index.html)(_templateObject23_18f031c0942811ecb1937969798b231a || (_templateObject23_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<div>\n                                <a href=\"mailto:", "\" target=\"_blank\"\n                                  >", "</a\n                                >\n                              </div>"])), s.email, s.email) : _index.nothing, _this10.database.assignments.map(function (a, h) {
            return (0, _index.html)(_templateObject24_18f031c0942811ecb1937969798b231a || (_templateObject24_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral([" <td\n                        class=\"td th-or-td\"\n                        numeric\n                        data-student=\"", "\"\n                        data-assignment=\"", "\"\n                        ?data-active=\"", "\"\n                      >\n                        <button\n                          aria-label=\"", "'s assignement ", "\"\n                          data-student=\"", "\"\n                          data-assignment=\"", "\"\n                        >\n                          ", "\n                        </button>\n                      </td>"])), i, h, _this10.activeStudent === i && _this10.activeAssignment === h, s.student, a.name, i, h, _this10.database.grades[i] && _this10.database.grades[i][a.shortName] ? (0, _index.html)(_templateObject25_18f031c0942811ecb1937969798b231a || (_templateObject25_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["", ""])), _this10.database.grades[i][a.shortName]) : "-");
          }));
        })) : _index.nothing, this.checkTabHeight, this.updateStudentReport, !(this.database.assignments && this.database.assignments[this.activeAssignment]), this.t.activeAssessment, this.activeRubric[0] ? (0, _index.html)(_templateObject26_18f031c0942811ecb1937969798b231a || (_templateObject26_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                <grid-plate\n                  disable-responsive\n                  layout=\"", "\"\n                >\n                  <div slot=\"col-1\" class=\"tag-group\">\n                    ", "\n                    ", "\n                  </div>\n                  <div slot=\"col-2\">\n                    <h3>", "</h3>\n                    ", "\n                    <div class=\"student-feedback-wrap\">\n                      <div class=\"student-feedback-text\">\n                        <h3 class=\"heading\">Overall feedback</h3>\n                        <simple-fields-field\n                          type=\"textarea\"\n                          data-criteria=\"overall\"\n                          data-rubric-written\n                        ></simple-fields-field>\n                      </div>\n                      <div class=\"student-feedback-score\">\n                        <simple-icon-button-lite\n                          icon=\"", "\"\n                          @click=\"", "\"\n                        ></simple-icon-button-lite>\n                        <simple-fields-field\n                          ?disabled=\"", "\"\n                          type=\"number\"\n                          min=\"0\"\n                          id=\"totalpts\"\n                          maxlength=\"10\"\n                          @value-changed=\"", "\"\n                        ></simple-fields-field>\n                        /\n                        ", "\n                        pts\n                        <letter-grade\n                          style=\"margin:-8px 0 0 16px;\"\n                          total=\"", "\"\n                          score=\"", "\"\n                        ></letter-grade>\n                      </div>\n                    </div>\n                  </div>\n                  ", "\n                </grid-plate>\n              "])), this.displayModeData()[this.displayMode].layout, this.renderDisplayModeBtn(), this.database.tags.categories.length > 0 ? (0, _index.html)(_templateObject27_18f031c0942811ecb1937969798b231a || (_templateObject27_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                          <a11y-collapse-group heading-button>\n                            ", "\n                          </a11y-collapse-group>\n                        "])), this.database.tags.categories.map(function (category, i) {
          return (0, _index.html)(_templateObject28_18f031c0942811ecb1937969798b231a || (_templateObject28_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                                <a11y-collapse>\n                                  <div slot=\"heading\">\n                                    <simple-colors\n                                      accent-color=\"", "\"\n                                      ><span></span></simple-colors\n                                    >", "\n                                  </div>\n                                  <div slot=\"content\">\n                                    ", "\n                                  </div>\n                                </a11y-collapse>\n                              "])), _this10.pickColor(i), category, _this10.database.tags.data.filter(function (item) {
            return item.category.includes(category);
          }).map(function (term) {
            return (0, _index.html)(_templateObject29_18f031c0942811ecb1937969798b231a || (_templateObject29_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<simple-tag\n                                            draggable=\"true\"\n                                            tabindex=\"0\"\n                                            @keypress=\"", "\"\n                                            @dragstart=\"", "\"\n                                            accent-color=\"", "\"\n                                            value=\"", "\"\n                                            .data=\"", "\"\n                                          ></simple-tag>"])), _this10.keyDown, _this10.setDragTransfer, _this10.pickColor(i), term.term, term);
          }));
        })) : _index.nothing, this.activeRubric[0].name, this.activeRubric.map(function (rubric, index) {
          return (0, _index.html)(_templateObject30_18f031c0942811ecb1937969798b231a || (_templateObject30_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                        ", "\n                        <letter-grade-picker></letter-grade-picker>\n\n                        <editable-table-display\n                          accent-color=\"", "\"\n                          bordered\n                          column-header\n                          condensed\n                          disable-responsive\n                          scroll\n                          striped\n                        >\n                          <table>\n                            <tbody>\n                              <tr>\n                                ", "\n                              </tr>\n                              <tr>\n                                ", "\n                              </tr>\n                            </tbody>\n                          </table>\n                        </editable-table-display>\n                        <h4>Additional ", " feedback</h4>\n                        <simple-fields-field\n                          type=\"textarea\"\n                          data-rubric-written\n                          data-criteria=\"", "\"\n                        ></simple-fields-field>\n                      "])), _this10.renderRubricInfoBtn(index, rubric.criteria, rubric.description), _this10.accentColor, rubric.qualitative.map(function (cat) {
            return (0, _index.html)(_templateObject31_18f031c0942811ecb1937969798b231a || (_templateObject31_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral([" <td>", "</td> "])), cat);
          }), rubric.qualitative.map(function (cat) {
            return (0, _index.html)(_templateObject32_18f031c0942811ecb1937969798b231a || (_templateObject32_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                                    <td>\n                                      <simple-fields-tag-list\n                                        style=\"background-color:transparent;\"\n                                        data-criteria=\"", "\"\n                                        label=\"", "\"\n                                      ></simple-fields-tag-list>\n                                    </td>\n                                  "])), rubric.criteria, cat);
          }), rubric.criteria, rubric.criteria);
        }), this.scoreLock ? "lock" : "lock-open", this.toggleLock, this.scoreLock, this.totalScoreChangedEvent, this.database.assignments[this.activeAssignment].points, this.database.assignments[this.activeAssignment].points, this.totalScore, this.displayMode != 2 ? (0, _index.html)(_templateObject33_18f031c0942811ecb1937969798b231a || (_templateObject33_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                        <div slot=\"col-3\">\n                          ", "\n                        </div>\n                      "])), this.database.assignments.length && this.database.assignments[this.activeAssignment] ? (0, _index.html)(_templateObject34_18f031c0942811ecb1937969798b231a || (_templateObject34_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["", ""])), this.activeSubmission ? this.renderSubmission(this.activeSubmission) : (0, _index.html)(_templateObject35_18f031c0942811ecb1937969798b231a || (_templateObject35_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["", ""])), this.t.noSubmission)) : _index.nothing) : _index.nothing) : _index.nothing, this.t.studentReportView, this.studentreportClick, !this.loading && this.database.assignments && this.database.assignments[this.activeAssignment] ? (0, _index.html)(_templateObject36_18f031c0942811ecb1937969798b231a || (_templateObject36_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                  <letter-grade\n                    class=\"student-report-score\"\n                    show-scale\n                    total=\"", "\"\n                    score=\"", "\"\n                  ></letter-grade>\n                  <h2>Feedback report</h2>\n                  <div class=\"student-report-wrap\">\n                    <a11y-collapse-group\n                      heading-button\n                      expanded\n                      style=\"width:80%;\"\n                    >\n                      ", "\n                      <a11y-collapse class=\"student-feedback-text\">\n                        <div slot=\"heading\" class=\"heading\">\n                          <span style=\"font-size:20px;\"\n                            >", "</span\n                          >\n                        </div>\n                        <div slot=\"content\">\n                          <p>", "</p>\n                          <h2>Your total Score</h2>\n                          <div class=\"score-display\">\n                            ", " /\n                            ", "\n                            pts\n                          </div>\n                        </div>\n                      </a11y-collapse>\n                    </a11y-collapse-group>\n                  </div>\n                "])), this.database.assignments[this.activeAssignment].points, this.totalScore, this.database.rubrics.filter(function (item) {
          return item.shortName == _this10.database.assignments[_this10.activeAssignment].rubric;
        }).map(function (rubric) {
          return (0, _index.html)(_templateObject37_18f031c0942811ecb1937969798b231a || (_templateObject37_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                            <a11y-collapse class=\"student-feedback-text\">\n                              <div slot=\"heading\" class=\"heading\">\n                                <span style=\"font-size:20px;\"\n                                  >", "</span\n                                >\n                              </div>\n                              <div slot=\"content\">\n                                <div class=\"student-feedback-score-heading\">\n                                  <div>\n                                    ", " /\n                                    ", "\n                                  </div>\n                                  <h3>Criteria details</h3>\n                                </div>\n                                <p>", "</p>\n                                <h3>Your feedback</h3>\n                                <ul>\n                                  ", "\n                                </ul>\n                                <h3>Additional Criteria feedback</h3>\n                                <p>\n                                  ", "\n                                </p>\n                              </div>\n                            </a11y-collapse>\n                          "])), rubric.criteria, _this10.getCriteriaScore(rubric.criteria), Math.round(rubric.percentage / 100 * _this10.database.assignments[_this10.activeAssignment].points), rubric.description, rubric.qualitative.map(function (cat) {
            return (0, _index.html)(_templateObject38_18f031c0942811ecb1937969798b231a || (_templateObject38_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                                      <h4>", "</h4>\n                                      <ul>\n                                        ", "\n                                      </ul>\n                                    "])), cat, _this10.activeGrading[rubric.criteria] ? (0, _index.html)(_templateObject39_18f031c0942811ecb1937969798b231a || (_templateObject39_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["", ""])), _this10.activeGrading[rubric.criteria][cat].map(function (tag) {
              return (0, _index.html)(_templateObject40_18f031c0942811ecb1937969798b231a || (_templateObject40_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral([" <li>\n                                                <span>", "</span\n                                                >", "\n                                                ", "\n                                              </li>"])), tag.term, tag.description ? (0, _index.html)(_templateObject41_18f031c0942811ecb1937969798b231a || (_templateObject41_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral([" - ", ""])), tag.description) : "", tag.associatedMaterial ? (0, _index.html)(_templateObject42_18f031c0942811ecb1937969798b231a || (_templateObject42_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                                                      <ul>\n                                                        ", "\n                                                      </ul>\n                                                    "])), tag.associatedMaterial.map(function (material) {
                return (0, _index.html)(_templateObject43_18f031c0942811ecb1937969798b231a || (_templateObject43_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                                                            <li>\n                                                              <a\n                                                                href=\"", "\"\n                                                                target=\"_blank\"\n                                                                rel=\"noopener noreferrer\"\n                                                                >", "</a\n                                                              >\n                                                            </li>\n                                                          "])), material, material);
              })) : "");
            })) : "");
          }), _this10.getCriteriaFeedback(rubric.criteria));
        }), this.t.overallFeedback, this.getCriteriaFeedback("overall"), this.totalScore, this.database.assignments[this.activeAssignment].points) : _index.nothing);
      }
    }, {
      key: "selectSource",
      value: function selectSource(e) {
        this.source = this.shadowRoot.querySelector("#source").value;
      }
    }, {
      key: "loadFromSource",
      value: function loadFromSource(e) {
        this.source = this.shadowRoot.querySelector("#source").value;

        if (this.source === "json") {
          this.sourceData = this.shadowRoot.querySelector("#sourcedatablob").value;
        } else if (this.source == "filesystem") {
          this.loadFromFilesystem();
        } else {
          this.sourceData = this.shadowRoot.querySelector("#sourcedata").value;
        }
      }
    }, {
      key: "loadFromExistingSource",
      value: function loadFromExistingSource(e) {
        this.source = this.shadowRoot.querySelector("#source").value;
        this.loadFromFilesystem(true);
      }
    }, {
      key: "loadFromFilesystem",
      value: function loadFromFilesystem() {
        var _this11 = this;

        var existing = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;

        // implies they already selected a file and want to use that again
        if (existing) {
          this.loading = true;
          setTimeout(function () {
            _xlsxFileSystemBroker.XLSXFileSystemBrokerSingleton.processFile(_this11.prevLocalFileReference, "json");

            _this11.sourceData = _this11.prevLocalFileReference;
          }, 0);
        } else {
          _xlsxFileSystemBroker.XLSXFileSystemBrokerSingleton.loadFile("xls").then( /*#__PURE__*/function () {
            var _ref3 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee5(file) {
              return regeneratorRuntime.wrap(function _callee5$(_context5) {
                while (1) {
                  switch (_context5.prev = _context5.next) {
                    case 0:
                      _this11.loading = true; // store reference so we can add a button for recent

                      _context5.next = 3;
                      return (0, _compat.set)("grade-book-prev-file", file);

                    case 3:
                      setTimeout(function () {
                        _xlsxFileSystemBroker.XLSXFileSystemBrokerSingleton.processFile(file, "json");

                        _this11.sourceData = file;
                      }, 0);

                    case 4:
                    case "end":
                      return _context5.stop();
                  }
                }
              }, _callee5);
            }));

            return function (_x3) {
              return _ref3.apply(this, arguments);
            };
          }());
        }
      }
    }, {
      key: "saveToFilesystem",
      value: function () {
        var _saveToFilesystem = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee6(e) {
          var output, blob, file;
          return regeneratorRuntime.wrap(function _callee6$(_context6) {
            while (1) {
              switch (_context6.prev = _context6.next) {
                case 0:
                  // return as Blob based output
                  // @todo undo the table transform on import so that we can have it as an array
                  // form that they want in xlsx files
                  // ensure we save the header row into the output based on correct key names
                  // this means we'll ahve to undo the process functions as well.
                  // MAY want to consider redoing how we look info up so that we don't transform it
                  // into complex objects and instead use complex arrays (maybe)
                  output = _xlsxFileSystemBroker.XLSXFileSystemBrokerSingleton.workbookFromJSON(this.database); // treat as a Blob and then convert to a FileReader object

                  blob = new Blob([output], {
                    type: "application/octet-stream"
                  });
                  file = new FileReader();
                  file.readAsDataURL(blob); // save to file format in question!

                  _context6.next = 6;
                  return _xlsxFileSystemBroker.XLSXFileSystemBrokerSingleton.saveFile("xlsx", output);

                case 6:
                case "end":
                  return _context6.stop();
              }
            }
          }, _callee6, this);
        }));

        function saveToFilesystem(_x4) {
          return _saveToFilesystem.apply(this, arguments);
        }

        return saveToFilesystem;
      }()
      /**
       * Listen for value change coming from the fields in the active rubric
       * and update the overall point total to match
       */

    }, {
      key: "rubricCriteriaPointsChange",
      value: function rubricCriteriaPointsChange(e) {
        var _this12 = this;

        // detect score field change
        if (e.detail.getAttribute("data-rubric-score") != null) {
          clearTimeout(this.__debounce);
          this.__debounce = setTimeout( /*#__PURE__*/babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee7() {
            return regeneratorRuntime.wrap(function _callee7$(_context7) {
              while (1) {
                switch (_context7.prev = _context7.next) {
                  case 0:
                    if (!_this12.loading) {
                      // @todo we need to store and recall these values
                      _this12.updateTotalScore();

                      _this12.shadowRoot.querySelector("#totalpts").value = _this12.totalScore;
                    } // force locking the score if this changes as we're using the rubric
                    // to modify things


                    _this12.scoreLock = true;

                  case 2:
                  case "end":
                    return _context7.stop();
                }
              }
            }, _callee7);
          })), 0);
        }
      }
    }, {
      key: "updateTotalScore",
      value: function updateTotalScore() {
        var score = 0;
        var tables = this.shadowRoot.querySelectorAll("#assessment simple-fields-field[type='number']:not(#totalpts)"); // add the scores up based on values of the pieces

        for (var i in Array.from(tables)) {
          if (tables[i].value) {
            score = score + parseInt(tables[i].value);
          }
        }

        this.totalScore = score;
        this.shadowRoot.querySelector("#totalpts").value = score;
        this.requestUpdate();
      }
    }, {
      key: "totalScoreChangedEvent",
      value: function totalScoreChangedEvent(e) {
        if (this.ready) {
          e.stopPropagation();
          e.stopImmediatePropagation();
          e.preventDefault();
          this.totalScore = e.detail.value;
          this.database.grades[this.activeStudent][this.database.assignments[this.activeAssignment].shortName] = this.totalScore;
          this.activeStudentSubmissions = [];
          this.activeStudentSubmissions = babelHelpers.toConsumableArray(this.getStudentSubmissions(this.activeStudent));
          this.requestUpdate();
        }
      }
      /**
       * lock toggle
       */

    }, {
      key: "toggleLock",
      value: function toggleLock(e) {
        this.scoreLock = !this.scoreLock;
      }
      /**
       * update student report when that tab is activated
       */

    }, {
      key: "updateStudentReport",
      value: function updateStudentReport() {
        // force a repaint of the calculated values from the Assessment view
        this.requestUpdate();
      }
      /**
       * Return the criteria score, current value
       */

    }, {
      key: "getCriteriaScore",
      value: function getCriteriaScore(criteria) {
        var tables = this.shadowRoot.querySelectorAll("#assessment editable-table-display"); // add the scores up based on values of the pieces

        for (var i in Array.from(tables)) {
          if (tables[i].shadowRoot.querySelector("[data-rubric-score][data-criteria=\"".concat(criteria, "\"]"))) {
            return tables[i].shadowRoot.querySelector("[data-rubric-score][data-criteria=\"".concat(criteria, "\"]")).value;
          }
        }

        return 0;
      }
      /**
       * Return the criteria written feedback, current value
       */

    }, {
      key: "getCriteriaFeedback",
      value: function getCriteriaFeedback(criteria) {
        if (this.shadowRoot.querySelector("#assessment [data-rubric-written][data-criteria=\"".concat(criteria, "\"]"))) {
          return this.shadowRoot.querySelector("#assessment [data-rubric-written][data-criteria=\"".concat(criteria, "\"]")).value;
        }

        return "";
      }
      /**
       * A qualitative feedback field got a new value
       */

    }, {
      key: "qualitativeFeedbackUpdate",
      value: function qualitativeFeedbackUpdate(e) {
        // group grade report by criteria, then qualitative label, THEN the list of tags used
        if (!this.activeGrading[e.detail.getAttribute("data-criteria")]) {
          this.activeGrading[e.detail.getAttribute("data-criteria")] = {};
        }

        this.activeGrading[e.detail.getAttribute("data-criteria")][e.detail.label] = e.detail.tagList;
        this.requestUpdate();
      } // @todo add support for keyboard based assignment of tag to criteria

    }, {
      key: "keyDown",
      value: function keyDown(e) {
        if (e.key === "Enter") {}
      } // get color based on index in the object "colors" from SimpleColors
      // this allows us to use an index in a common way and obtain a color
      // so that our tags have a color association per category

    }, {
      key: "pickColor",
      value: function pickColor(val) {
        var colors = Object.keys(this.colors);

        while (val > colors.length) {
          val = val - colors.length;
        }

        return colors[val];
      } // ensure when we drop a tag onto the UI that it removes all the outlines
      // of fields that can have items dropped into them

    }, {
      key: "_handleDragDrop",
      value: function _handleDragDrop(e) {
        window.dispatchEvent(new CustomEvent("simple-tag-drop", {
          detail: {
            value: "drop"
          }
        }));
      } // set the drag transfer data

    }, {
      key: "setDragTransfer",
      value: function setDragTransfer(e) {
        window.dispatchEvent(new CustomEvent("simple-tag-dragstart", {
          detail: {
            value: e.target
          }
        }));
        var data = e.target.data; // have to add in color

        data.color = e.target.accentColor;
        e.dataTransfer.setData("text", JSON.stringify(data));
      }
    }], [{
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(GradeBook), "properties", this)), {}, {
          displayMode: {
            type: Number
          },
          settings: {
            type: Object
          },
          disabled: {
            type: Boolean,
            reflect: true
          },
          loading: {
            type: Boolean,
            reflect: true
          },
          ready: {
            type: Boolean,
            reflect: true
          },
          activeStudent: {
            type: Number,
            attribute: "active-student"
          },
          activeAssignment: {
            type: Number,
            attribute: "active-assignment"
          },
          totalScore: {
            type: Number
          },
          scoreLock: {
            type: Boolean
          },
          source: {
            type: String
          },
          sourceData: {
            type: String,
            attribute: "source-data"
          },
          activeSubmission: {
            type: String,
            attribute: false
          },
          database: {
            type: Object,
            attribute: false
          },
          activeRubric: {
            type: Object,
            attribute: false
          },
          assessmentView: {
            type: Object,
            attribute: false
          },
          activeGrading: {
            type: Object,
            attribute: false
          },
          activeStudentSubmissions: {
            type: Array
          }
        });
      }
    }, {
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(GradeBook), "styles", this)), [(0, _index.css)(_templateObject44_18f031c0942811ecb1937969798b231a || (_templateObject44_18f031c0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        :host [hidden] {\n          display: none !important;\n        }\n        @media (max-width: 900px) {\n          .hide-900 {\n            display: none;\n          }\n        }\n        loading-indicator {\n          --loading-indicator-background-color: var(\n            --simple-colors-default-theme-accent-2,\n            grey\n          );\n          --loading-indicator-color: var(\n            --simple-colors-default-theme-accent-10,\n            black\n          );\n        }\n        loading-indicator[full] {\n          top: 0;\n          position: absolute;\n          left: 0;\n          right: 0;\n          z-index: 1;\n        }\n        h1,\n        h2,\n        h3,\n        h4 {\n          margin: 0;\n        }\n        grid-plate {\n          --hax-layout-container-transition: none;\n          --grid-plate-col-transition: none;\n          --grid-plate-item-margin: 0px;\n          --grid-plate-item-padding: 8px;\n        }\n        #studentassessment {\n          --a11y-tabs-border-color: var(\n            --simple-colors-default-theme-accent-10,\n            black\n          );\n          --a11y-tabs-border-color: var(\n            --simple-colors-default-theme-accent-10,\n            black\n          );\n          color: var(--simple-colors-default-theme-grey-12);\n          background-color: var(--simple-colors-default-theme-grey-1);\n          --a11y-tabs-faded-color: var(--simple-colors-default-theme-grey-11);\n          --a11y-tabs-focus-color: var(--simple-colors-default-theme-grey-10);\n          --a11y-tabs-faded-background: var(\n            --simple-colors-default-theme-grey-2\n          );\n        }\n        #studentassessment::part(content) {\n          overflow: scroll;\n        }\n        a11y-collapse {\n          --a11y-collapse-border-color: var(\n            --simple-colors-default-theme-accent-10,\n            black\n          );\n          --a11y-collapse-horizontal-padding: 8px;\n          --a11y-collapse-vertical-padding: 4px;\n        }\n        a11y-collapse:not([expanded]):hover {\n          background-color: var(--simple-colors-default-theme-accent-1, grey);\n        }\n        a11y-collapse div[slot=\"heading\"] {\n          font-size: 18px;\n          font-weight: normal;\n          cursor: pointer;\n          line-height: 30px;\n          display: flex;\n        }\n        a11y-collapse[expanded] div[slot=\"heading\"] {\n          font-weight: bold;\n        }\n        .active-student-grade-history letter-grade {\n          display: inline-flex;\n          margin: 2px;\n        }\n        .active-student-grade-history button {\n          opacity: 0.4;\n          background-color: transparent;\n          border: 0;\n          padding: 0;\n          margin: 0;\n        }\n        .active-student-grade-history button.activeAssignment {\n          opacity: 0.9;\n          background-color: var(--simple-colors-default-theme-yellow-8);\n        }\n        .active-student-grade-history button:focus,\n        .active-student-grade-history button:active,\n        .active-student-grade-history button:hover {\n          opacity: 1;\n          outline: 1px solid black;\n          outline-offset: 2px;\n        }\n        simple-fields-tag-list {\n          --simple-fields-tag-list-possible: var(\n            --simple-colors-default-theme-accent-2\n          );\n          --simple-fields-tag-list-focus: var(\n            --simple-colors-default-theme-accent-10\n          );\n        }\n        simple-colors[accent-color] {\n          display: inline-flex;\n          width: 24px;\n          height: 24px;\n          margin: 4px 6px 0 0;\n        }\n        simple-colors[accent-color] span {\n          display: inline-flex;\n          width: 24px;\n          height: 24px;\n          background-color: var(\n            --simple-colors-default-theme-accent-3,\n            #eeeeee\n          );\n        }\n        simple-fields-field[type=\"textarea\"] {\n          --simple-fields-font-size: 20px;\n        }\n        .source-selection {\n          text-align: center;\n          margin-top: 30vh;\n          font-size: 32px;\n        }\n        .source-selection label {\n          display: block;\n        }\n        .source-selection select {\n          font-size: 18px;\n        }\n        #sourcedata,\n        #sourcedatablob {\n          display: block;\n          margin: 16px auto;\n          font-size: 24px;\n        }\n        simple-fields-field[type=\"number\"] {\n          --simple-fields-font-size: 40px;\n          line-height: 40px;\n        }\n\n        .student-feedback-score-heading {\n          display: flex;\n          font-size: 28px;\n          font-weight: bold;\n          line-height: 28px;\n          padding: 8px;\n        }\n        .student-feedback-score-heading h3 {\n          margin: 0 0 0 8px;\n          padding: 0;\n        }\n\n        #totalpts {\n          width: 84px;\n          margin: 0 12px;\n        }\n        #studentgrid {\n          display: block;\n          width: 100%;\n          height: 52vh;\n          max-height: 90vh;\n          min-height: 140px; /** exact height of a row to still be usable */\n          resize: vertical;\n          overflow: auto;\n          scrollbar-color: var(--simple-colors-default-theme-accent-10)\n            var(--simple-colors-default-theme-accent-1);\n          scrollbar-width: auto;\n        }\n        /** start scroll bar styling */\n        #studentgrid::-webkit-scrollbar-track,\n        #studentassessment::-webkit-scrollbar-track {\n          -webkit-box-shadow: inset 0 0 2px rgba(0, 0, 0, 0.3);\n          border-radius: 20px;\n          background-color: var(--simple-colors-default-theme-accent-1);\n        }\n        #studentgrid::-webkit-scrollbar,\n        #studentassessment::-webkit-scrollbar {\n          width: 16px;\n          height: 16px;\n          background-color: var(--simple-colors-default-theme-accent-1);\n        }\n        #studentgrid::-webkit-scrollbar-thumb {\n          border-radius: 20px;\n          -webkit-box-shadow: inset 0 0 2px rgba(0, 0, 0, 0.3);\n          background-color: var(--simple-colors-default-theme-accent-10);\n        }\n        #studentgrid::-webkit-scrollbar-thumb:vertical,\n        #studentassessment::-webkit-scrollbar-thumb:vertical {\n          height: 100px;\n          width: 16px;\n        }\n        #studentgrid::-webkit-scrollbar-corner {\n          -webkit-box-shadow: inset 0 0 2px rgba(0, 0, 0, 0.3);\n          background-color: var(--simple-colors-default-theme-accent-10);\n        }\n        /** end scroll bar styling */\n        .student-table-label {\n          text-align: center;\n          vertical-align: middle;\n        }\n        #studentgrid simple-fields-field.select-all {\n          background-color: transparent;\n          float: left;\n          margin: -2px 0 0 -2px;\n          --simple-icon-width: 16px;\n          --simple-icon-height: 16px;\n        }\n        .student-table-label div {\n          border-right: 1px solid var(--editable-table-heading-color, #000);\n          height: 32px;\n          width: 100%;\n          line-height: 32px;\n          margin-left: 3px;\n        }\n        #studentgrid [data-active] {\n          background-color: var(\n            --simple-colors-default-theme-yellow-3\n          ) !important;\n        }\n        #studentgrid button {\n          background-color: transparent;\n          border: none;\n          border-radius: 0;\n          height: 80px;\n          width: 96px;\n          display: block;\n          padding: 0;\n          margin: 0;\n          color: var(--simple-colors-default-theme-grey-12);\n        }\n        #studentgrid button:focus,\n        #studentgrid button:hover {\n          cursor: pointer;\n          outline: 2px black solid;\n          outline-offset: -1px;\n          background-color: var(\n            --simple-colors-default-theme-yellow-4\n          ) !important;\n        }\n        #studentgrid .assignment-name {\n          max-width: 96px;\n          width: 96px;\n          overflow: hidden;\n          padding: 4px;\n          font-size: 14px;\n          text-align: center;\n        }\n        .student-feedback-wrap {\n          display: flex;\n        }\n        .student-feedback-wrap .student-feedback-text {\n          width: 80%;\n          margin: 0;\n        }\n        .student-feedback-wrap .student-feedback-text div.heading {\n          padding: 20px;\n        }\n        .student-feedback-wrap .student-feedback-score {\n          font-size: 40px;\n          line-height: 68px;\n          padding: 34px 16px;\n          display: flex;\n          width: 50%;\n        }\n        simple-tag {\n          margin: 2px;\n        }\n        simple-tag:focus,\n        simple-tag:hover {\n          --simple-fields-fieldset-border-color: var(\n            --simple-colors-default-theme-accent-10,\n            #eeeeee\n          );\n        }\n        .user-info {\n          display: flex;\n          font-size: 16px;\n          min-width: 140px;\n          font-family: sans-serif;\n          font-weight: normal;\n          border-right: 1px solid black;\n        }\n        .user-left {\n          display: inline-flex;\n        }\n        .user-right {\n          display: block;\n          padding: 0 4px 0 0;\n          margin: 12px 0;\n        }\n        .user-right div a {\n          color: var(--simple-colors-default-theme-grey-12);\n          text-decoration: none;\n        }\n        .user-right div a:hover,\n        .user-right div a:focus,\n        .user-right div a:active {\n          color: var(--simple-colors-default-theme-blue-8);\n          text-decoration: underline;\n        }\n        .user-right div {\n          display: block;\n          text-align: center;\n        }\n        .user-photo {\n          --simple-icon-height: 48px;\n          --simple-icon-width: 48px;\n          width: 48px;\n          height: 48px;\n          border-radius: 50%;\n          margin: 12px 6px;\n          float: left;\n          vertical-align: middle;\n        }\n        .active-submission {\n          max-height: 400px;\n          width: 60%;\n          overflow: auto;\n          margin: 0 auto;\n        }\n        .active-submission iframe {\n          height: 400px;\n          width: 60%;\n        }\n        .tag-group {\n          position: sticky;\n          top: 0;\n        }\n        .student-report-wrap {\n          display: flex;\n          justify-content: space-evenly;\n        }\n        .student-report-score {\n          position: absolute;\n          right: 40px;\n          margin-top: -20px;\n        }\n        editable-table-display::part(td),\n        editable-table-display::part(th) {\n          text-align: center;\n          vertical-align: top;\n          max-width: 250px;\n        }\n        editable-table-display::part(simple-fields-field) {\n          --simple-fields-font-size: 32px;\n          --simple-fields-text-align: center;\n          background-color: transparent;\n          max-width: 100px;\n          padding: 0 0 28px 0;\n          margin: 0;\n        }\n\n        table thead th {\n          padding: 4px !important;\n          position: sticky;\n          top: -2px;\n          z-index: 1;\n        }\n        table thead th:first-child {\n          position: sticky;\n          left: 0;\n          z-index: 2;\n        }\n        table tbody th {\n          position: sticky;\n          left: 0;\n          z-index: 1;\n        }\n\n        .col-highlight,\n        tr:hover td,\n        tr:hover th {\n          transition: 0.3s ease-in-out all;\n          background-color: var(\n            --simple-colors-default-theme-yellow-2\n          ) !important;\n        }\n        table {\n          width: calc(100% - 2 * var(--editable-table-border-width, 1px));\n          border-collapse: collapse;\n          border-width: var(--editable-table-border-width, 1px);\n          border-style: var(--editable-table-border-style, solid);\n          border-color: var(--simple-colors-default-theme-grey-6, #999);\n          font-weight: var(--editable-table-light-weight, 200);\n          color: var(--simple-colors-default-theme-grey-10, #222);\n          background-color: var(--simple-colors-default-theme-grey-1, #fff);\n        }\n        .th,\n        .td {\n          font-weight: var(--editable-table-light-weight, 200);\n          color: var(--simple-colors-default-theme-grey-10, #222);\n          background-color: var(--simple-colors-default-theme-grey-1, #fff);\n        }\n        caption {\n          font-size: var(\n            --editable-table-caption-font-size,\n            var(--editable-table-font-size, unset)\n          );\n          font-weight: var(--editable-table-heavy-weight, 600);\n          color: var(\n            --editable-table-caption-color,\n            var(--simple-colors-default-theme-grey-10, #222)\n          );\n          background-color: var(\n            --editable-table-caption-bg-color,\n            var(--simple-colors-default-theme-grey-1, #fff)\n          );\n          width: 100%;\n        }\n        .tr {\n          display: table-row;\n        }\n        .th-or-td {\n          display: table-cell;\n        }\n        .thead .th-or-td {\n          height: 32px;\n          padding: 0;\n          margin: 0;\n        }\n        .thead-tr .th {\n          background-color: var(--simple-colors-default-theme-grey-2, #e0e0e0);\n          font-weight: var(--editable-table-heavy-weight, 600);\n          color: var(--simple-colors-default-theme-grey-12, #000);\n        }\n        .tbody-tr .th {\n          font-weight: var(--editable-table-heavy-weight, 600);\n          color: var(--simple-colors-default-theme-grey-12, #000);\n          background-color: var(--simple-colors-default-theme-grey-1, #fff);\n          text-align: left;\n        }\n        table[bordered] .th,\n        table[bordered] .td {\n          border-width: var(--editable-table-border-width, 1px);\n          border-style: var(--editable-table-border-style, solid);\n          border-color: var(--simple-colors-default-theme-grey-6, #999);\n        }\n        table[condensed] {\n          --editable-table-cell-vertical-padding: var(\n            --editable-table-cell-vertical-padding-condensed,\n            2px\n          );\n          --editable-table-cell-horizontal-padding: var(\n            --editable-table-cell-horizontal-padding-condensed,\n            4px\n          );\n        }\n        table[striped] .tbody-tr:nth-child(2n + 1) .th-or-td {\n          background-color: var(--simple-colors-default-theme-grey-2, #f0f0f0);\n        }\n        .tfoot-tr .th,\n        .tfoot-tr .td {\n          border-top: 2px solid var(--simple-colors-default-theme-grey-10, #222);\n          font-weight: var(--editable-table-heavy-weight, 600);\n          color: var(--simple-colors-default-theme-grey-12, #000);\n        }\n        caption,\n        .th-or-td {\n          text-align: left;\n        }\n        table[numeric-styles] .thead-tr .th-or-td[numeric],\n        table[numeric-styles] .tfoot-tr .th-or-td[numeric],\n        table[numeric-styles] .th-or-td[numeric] {\n          text-align: center;\n          font-size: 24px;\n          font-family: sans-serif;\n          --editable-table-cell-justify: flex-end;\n        }\n        table[numeric-styles] .tfoot-tr .th-or-td[negative],\n        table[numeric-styles] .td[negative] {\n          color: var(--simple-colors-default-theme-red-7, red);\n          --editable-table-cell-color: var(\n            --editable-table-negative-color,\n            red\n          );\n        }\n\n        caption > div {\n          display: flex;\n          align-items: flex-end;\n          justify-content: space-between;\n        }\n        caption > div > *:not(:last-child) {\n          padding: 0 0 5px;\n        }\n        #column {\n          width: calc(var(--simple-picker-option-size) + 6px);\n          overflow: visible;\n          display: none;\n          margin-left: 10px;\n          --simple-picker-border-width: 1px;\n          --simple-picker-focus-border-width: 1px;\n          --simple-picker-border-color: var(\n            --simple-colors-default-theme-grey-6,\n            #999\n          );\n        }\n        .th,\n        .td {\n          padding: 0;\n        }\n        .top-controls {\n          background-color: var(--simple-colors-default-theme-accent-12);\n          color: var(--simple-colors-default-theme-accent-1);\n          height: 36px;\n          vertical-align: middle;\n          width: 100%;\n          display: flex;\n        }\n        .top-controls .group {\n          padding: 0 8px;\n        }\n        .top-controls .app-name {\n          padding: 0 24px 0 16px;\n        }\n        .top-controls simple-icon-button-lite span {\n          padding: 0 8px 0 0;\n        }\n        .top-controls .divider-left {\n          border-left: 1px solid var(--simple-colors-default-theme-accent-1);\n        }\n        .top-controls .divider-right {\n          border-right: 1px solid var(--simple-colors-default-theme-accent-1);\n        }\n      "])))]);
      }
    }, {
      key: "tag",
      get: function get() {
        return "grade-book";
      }
    }]);
    return GradeBook;
  }((0, _GradeBookUIPieces.UIRenderPieces)((0, _I18NMixin.I18NMixin)(_simpleColors.SimpleColors)));

  _exports.GradeBook = GradeBook;
  customElements.define(GradeBook.tag, GradeBook);
  window.GradeBook = window.GradeBook || {};

  window.GradeBook.requestAvailability = function () {
    // if there is no single instance, generate one and append it to end of the document
    if (!window.GradeBook.instance) {
      if (document.querySelector("grade-book")) {
        window.GradeBook.instance = document.querySelector("grade-book");
      } else {
        window.GradeBook.instance = document.createElement("grade-book");
        document.body.appendChild(window.GradeBook.instance);
      }
    }

    return window.GradeBook.instance;
  };
});
