// @lrnwebcomponents/app-hax/lib/v1/app-hax-search-bar.js
import{LitElement as e,html as t,css as i}from"../../../../lit/index.js";import"../../../simple-tooltip/simple-tooltip.js";import{store as s}from"./AppHaxStore.js";export class AppHaxSearchBar extends e{static get tag(){return"app-hax-search-bar"}constructor(){super(),this.searchTerm="",this.disabled=!1,this.showSearch=!1}static get properties(){return{searchTerm:{type:String},showSearch:{type:Boolean,reflect:!0,attribute:"show-search"},disabled:{type:Boolean,reflect:!0}}}updated(e){e.forEach(((e,t)=>{"searchItems"===t?this.displayItems=[...this.searchItems]:"searchTerm"===t?s.searchTerm=this.searchTerm:"showSearch"===t&&void 0!==e&&!1===this[t]&&(this.searchTerm="")}))}static get styles(){return[i`
        :host {
          overflow: hidden;
        }
        input {
          visibility: none;
          opacity: 0;
          width: 0;
          transition: all ease-in-out 0.3s;
          padding: 4px;
          font-family: "Press Start 2P", sans-serif;
          font-size: 20px;
          margin: 2px 0 0 16px;
        }
        :host([show-search]) input {
          visibility: visible;
          opacity: 1;
          width: 250px;
          max-width: 25vw;
        }
        @media (max-width: 780px) {
          :host([show-search]) input {
            width: 250px;
            max-width: 20vw;
          }
        }
        @media (max-width: 600px) {
          :host([show-search]) input {
            width: 200px;
            max-width: 20vw;
          }
        }
        simple-icon-button-lite {
          color: black;
          --simple-icon-width: 40px;
          --simple-icon-height: 40px;
          padding: 2px;
          margin: 0;
          background-color: white;
        }
        simple-icon-button-lite[disabled] {
          background-color: #cccccc;
          pointer-events: none;
          cursor: help;
        }
        simple-icon-button-lite:focus,
        simple-icon-button-lite:hover {
          background-color: #eeeeee;
        }
      `]}search(){s.appEl.playSound("click"),this.searchTerm=this.shadowRoot.querySelector("#searchField").value}render(){return t`
      <simple-icon-button-lite
        id="searchico"
        ?disabled="${this.disabled}"
        label="Search"
        icon="icons:search"
        @click="${this.toggleSearch}"
      ></simple-icon-button-lite>
      <simple-tooltip for="searchico" position="bottom"
        >Toggle Search</simple-tooltip
      >
      <input
        ?disabled="${!this.showSearch}"
        id="searchField"
        @input="${this.search}"
        type="text"
        placeholder="Search.."
      />
    `}toggleSearch(){this.disabled||(this.shadowRoot.querySelector("#searchField").value="",s.appEl.playSound("click"),this.showSearch=!this.showSearch,setTimeout((()=>{this.shadowRoot.querySelector("#searchField").focus()}),300))}}customElements.define(AppHaxSearchBar.tag,AppHaxSearchBar);
