// @lrnwebcomponents/clean-one/lib/clean-one-search-box.js
define(["exports", "meta", "../../../lit/index.js", "../../i18n-manager/lib/I18NMixin.js"], function (_exports, meta, _index, _I18NMixin2) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.CleanOneSearchBox = void 0;
  meta = _interopRequireWildcard(meta);

  var _templateObject_606d47e0942811ecb1937969798b231a, _templateObject2_606d47e0942811ecb1937969798b231a;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var CleanOneSearchBox = /*#__PURE__*/function (_I18NMixin) {
    babelHelpers.inherits(CleanOneSearchBox, _I18NMixin);

    var _super = _createSuper(CleanOneSearchBox);

    function CleanOneSearchBox() {
      var _this;

      babelHelpers.classCallCheck(this, CleanOneSearchBox);
      _this = _super.call(this);
      _this.value = "";
      _this.t = {
        searchSiteContent: "Search site content",
        typeToSearch: "Type to search"
      };

      _this.registerLocalization({
        context: babelHelpers.assertThisInitialized(_this),
        namespace: "clean-one",
        localesPath: new URL("../locales", meta.url).href,
        locales: ["es", "fr", "de", "ja"]
      });

      return _this;
    }

    babelHelpers.createClass(CleanOneSearchBox, [{
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_606d47e0942811ecb1937969798b231a || (_templateObject_606d47e0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <input\n        type=\"text\"\n        aria-label=\"", "\"\n        placeholder=\"", "\"\n        .value=\"", "\"\n        @input=\"", "\"\n      />\n    "])), this.t.searchSiteContent, this.t.typeToSearch, this.value, this.searchChanged);
      }
    }, {
      key: "searchChanged",
      value: function searchChanged(e) {
        if (this.shadowRoot) {
          this.dispatchEvent(new CustomEvent("input-changed", {
            composed: false,
            bubbles: false,
            cancelable: false,
            detail: {
              value: this.shadowRoot.querySelector("input").value
            }
          }));
        }
      }
    }], [{
      key: "tag",
      get: function get() {
        return "clean-one-search-box";
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _index.css)(_templateObject2_606d47e0942811ecb1937969798b231a || (_templateObject2_606d47e0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        input {\n          line-height: normal;\n          font-family: inherit;\n          font-size: 100%;\n          margin: 0;\n        }\n        input,\n        input:focus,\n        input:hover {\n          background: 0 0;\n          border: 1px solid transparent;\n          box-shadow: none;\n          outline: 0;\n          line-height: 22px;\n          padding: 7px 7px;\n          color: inherit;\n        }\n      "])))];
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          value: {
            type: String
          }
        };
      }
    }]);
    return CleanOneSearchBox;
  }((0, _I18NMixin2.I18NMixin)(_index.LitElement));

  _exports.CleanOneSearchBox = CleanOneSearchBox;
  customElements.define(CleanOneSearchBox.tag, CleanOneSearchBox);
});
