diff --git a/uuid.entity.inc b/uuid.entity.inc
index 02ea5c2..907f05f 100644
--- a/uuid.entity.inc
+++ b/uuid.entity.inc
@@ -82,6 +82,9 @@ function uuid_get_core_entity_info() {
  */
 function uuid_entity_info_alter(&$info) {
   foreach (uuid_get_core_entity_info() as $entity_type => $core_info) {
+    if (!isset($info[$entity_type])) {
+      continue;
+    }
     $info[$entity_type]['uuid'] = TRUE;
     $info[$entity_type]['entity keys']['uuid'] = $core_info['entity keys']['uuid'];
     if (!empty($core_info['entity keys']['revision uuid'])) {
@@ -408,7 +411,7 @@ function entity_get_id_by_uuid($entity_type, $uuids, $revision = FALSE) {
   }
   $info = entity_get_info($entity_type);
   // Find out what entity keys to use.
-  if (!$revision) {
+  if (!$revision && isset($info['entity keys']['uuid'])) {
     $table = $info['base table'];
     $id_key = $info['entity keys']['id'];
     $uuid_key = $info['entity keys']['uuid'];
@@ -454,7 +457,7 @@ function entity_get_uuid_by_id($entity_type, $ids, $revision = FALSE) {
   }
   $info = entity_get_info($entity_type);
   // Find out what entity keys to use.
-  if (!$revision) {
+  if (!$revision && isset($info['entity keys']['uuid'])) {
     $table = $info['base table'];
     $id_key = $info['entity keys']['id'];
     $uuid_key = $info['entity keys']['uuid'];
