diff --git a/core/dslmcode/stacks/discuss/sites/all/modules/local_contrib/harmony_core/theme/harmony-post.tpl.php b/core/dslmcode/stacks/discuss/sites/all/modules/local_contrib/harmony_core/theme/harmony-post.tpl.php
index 0244b9dd6..5d05162cb 100644
--- a/core/dslmcode/stacks/discuss/sites/all/modules/local_contrib/harmony_core/theme/harmony-post.tpl.php
+++ b/core/dslmcode/stacks/discuss/sites/all/modules/local_contrib/harmony_core/theme/harmony-post.tpl.php
@@ -45,7 +45,7 @@
   </div>
   <div class="clearfix">
     <div class="post-left post-user-profile text-center">
-      <?php print $user_profile; ?>
+      <a href="#"><?php print user_load($post->uid)->name; ?></a>
     </div>
     <div class="post-content post-left-offset">
       <?php print render($content); ?>
