// @lrnwebcomponents/service-card/service-card.js
define(["exports", "../../lit-element/lit-element.js"], function (_exports, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ServiceCard = void 0;

  function _templateObject2_c1f4ab10e40811ea96df1524fbfe932f() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      :host {\n        display: inline-flex;\n        margin: 16px;\n      }\n      .card {\n        box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);\n        transition: 0.3s all ease-in-out;\n        width: var(--service-card-card-width, 300px);\n      }\n      .card:hover {\n        box-shadow: 0 8px 16px 0 rgba(0, 0, 0, 0.2);\n      }\n      .container {\n        padding: 2px 2px 2px 2px;\n      }\n      .avatar {\n        border-radius: 50%;\n        width: calc(var(--service-card-card-width, 300px) - 10px);\n        padding: 5px;\n      }\n      .title {\n        padding: 4px 4px 4px 4px;\n        text-align: center;\n        font-family: \"Courier New\", courier, sans-serif;\n      }\n      .info {\n        padding: 4px 4px 4px 4px;\n        height: 125px;\n        overflow-y: auto;\n      }\n    "]);

    _templateObject2_c1f4ab10e40811ea96df1524fbfe932f = function _templateObject2_c1f4ab10e40811ea96df1524fbfe932f() {
      return data;
    };

    return data;
  }

  function _templateObject_c1f4ab10e40811ea96df1524fbfe932f() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"card\">\n        <div class=\"container\">\n          <img\n            loading=\"lazy\"\n            class=\"avatar\"\n            src=\"", "\"\n            alt=\"", "\"\n          />\n          <h3 class=\"title\"><b>", "</b></h3>\n          <p class=\"info\"><slot></slot></p>\n        </div>\n      </div>\n    "]);

    _templateObject_c1f4ab10e40811ea96df1524fbfe932f = function _templateObject_c1f4ab10e40811ea96df1524fbfe932f() {
      return data;
    };

    return data;
  }

  /**
   * `service-card`
   * `simple showcase of services provided`
   * @demo demo/index.html
   * @element service-card
   */
  var ServiceCard =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(ServiceCard, _LitElement);
    babelHelpers.createClass(ServiceCard, null, [{
      key: "tag",
      get: function get() {
        return "service-card";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          source: {
            type: String
          },
          title: {
            type: String
          },
          alt: {
            type: String
          }
        };
      }
    }, {
      key: "haxProperties",
      get: function get() {
        return {
          canScale: false,
          canPosition: true,
          canEditSource: false,
          gizmo: {
            title: "Service card",
            description: "Simple presentation of services or info",
            icon: "hax:calendar",
            color: "orange",
            groups: ["Card"],
            handles: [],
            meta: {
              author: "McGarvelcuddy",
              owner: "elmsln"
            }
          },
          settings: {
            quick: [],
            configure: [{
              property: "source",
              title: "Source",
              inputMethod: "haxupload"
            }, {
              property: "alt",
              title: "Alt Text",
              description: "Alternative text for non-sighted users",
              inputMethod: "alt"
            }, {
              property: "title",
              title: "Title",
              description: "Service title / name",
              inputMethod: "textfield"
            }, {
              slot: "",
              title: "Decription",
              description: "Details of the service",
              inputMethod: "code-editor"
            }],
            advanced: []
          }
        };
      }
    }]);

    function ServiceCard() {
      var _this;

      babelHelpers.classCallCheck(this, ServiceCard);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(ServiceCard).call(this));
      _this.alt = "";
      return _this;
    }

    babelHelpers.createClass(ServiceCard, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_c1f4ab10e40811ea96df1524fbfe932f(), this.source, this.alt, this.title);
      }
    }], [{
      key: "styles",
      get: function get() {
        return (0, _litElement.css)(_templateObject2_c1f4ab10e40811ea96df1524fbfe932f());
      }
    }]);
    return ServiceCard;
  }(_litElement.LitElement);

  _exports.ServiceCard = ServiceCard;
  customElements.define(ServiceCard.tag, ServiceCard);
});
