// @lrnwebcomponents/lrn-calendar/lib/lrn-calendar-date.js
define(["exports","../../../@polymer/polymer/polymer-element.js","../../../@polymer/paper-button/paper-button.js","../../../@polymer/iron-icons/iron-icons.js","../../materializecss-styles/lib/colors.js","../../../@polymer/paper-card/paper-card.js","../../lrnsys-layout/lib/lrnsys-drawer.js","../../../@polymer/iron-ajax/iron-ajax.js"],function(_exports,_polymerElement,_paperButton,_ironIcons,_colors,_paperCard,_lrnsysDrawer,_ironAjax){"use strict";Object.defineProperty(_exports,"__esModule",{value:!0});_exports.LrnCalendarDate=void 0;/**
 * `lrn-calendar-date`
 * `A LRN element`
 *
 * @demo demo/index.html
 *
 */class LrnCalendarDate extends _polymerElement.PolymerElement{static get template(){return _polymerElement.html`
      <style include="materializecss-styles-colors">
        :host {
          display: block;
        }
        ::-webkit-scrollbar {
          width: 0px; /* remove scrollbar space */
          background: transparent; /* optional: just make scrollbar invisible */
        }

        .month {
          width: 100%;
          height: 20vh;
          vertical-align: top;
          overflow: scroll;
        }
        .week {
          width: 100%;
          height: 40vh;
          vertical-align: top;
          overflow: scroll;
        }
        .card-content {
          position: relative;
          bottom: 10px;
          white-space: nowrap;
          font-size: 12px;
          padding: 0px 5px;
          line-height: 7px;
        }
        .label {
          width: 100%;
        }
      </style>

      <div id="test">
        <div>
          <template is="dom-if" if="[[firstWeek]]">
            <paper-card class="label"> <h3>[[getWeek(date)]]</h3> </paper-card>
          </template>
        </div>
        <paper-card class$="[[view]]" id="dateHeader">
          <h5>[[getMonth(date)]]</h5>
          <template is="dom-repeat" items="{{events}}">
            <div class="card-content">
              <lrnsys-drawer
                text="[[timeString(item.event.startDate._time.hour, item.event.startDate._time.minute, item.event.endDate._time.hour,item.event.endDate._time.minute)]] {{item.event.summary}}"
                header="[[getDateString(date)]]"
                align="left"
                heading-class="orange lighten-3 blue-text text-darken-4"
                style$="[[computeStyle(item)]];overflow:auto;"
              >
                [[displayActivity(item)]]
                <br /><br />[[displayStart(item)]]<br /><br />[[displayEnd(item)]]<br /><br />[[displayDuration(item)]]<br /><br />[[displayDescription(item)]]<br /><br />[[displayLocation(item)]]
              </lrnsys-drawer>
            </div>
          </template>
        </paper-card>
      </div>
    `}static get tag(){return"lrn-calendar-date"}static get properties(){return{title:{type:String,value:"lrn-calendar-date"},name:{type:String},date:{type:Date},loadeddate:{type:Boolean},firstWeek:{type:Boolean},view:{type:String},valid:{type:Boolean,value:!1}}}/*
   * Displays the month in the heading
   */getMonth(date){var monthInt=date.getMonth(),day=date.getDate(),monthsArray=["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"];if(1==day){monthstring=monthsArray[monthInt]+" "+day}else{monthstring=day}return monthstring}/*
   * Displays the weekday for in the column for the first week
   */getWeek(date){var weekdays=date.getDay(),weekArray=["Sun","Mon","Tue","Wed","Thu","Fri","Sat"];return weekArray[weekdays]}/*
   * Creates the string that displays the time of the event
   */timeString(startHour,startMinute,endHour,endMinute){var amPM=11<startHour?"pm":"am";if(12<startHour){startHour-=12}else if(0==startHour){//If the start time and end time is 12AM
if(0==startMinute&&0==endHour&&0==endMinute){var allDayText="";return allDayText}else{startHour="12"}}if(10>startMinute){startMinute="0"+startMinute}if(0==startMinute){return startHour+amPM}return startHour+":"+startMinute+amPM}/*
   * If an event occurs the whole day the background color is yellow
   */computeStyle(item){if(0==item.event.startDate._time.hour)var s="background-color:"+" yellow";else var s="background-color:"+" clear";return s}/*
   * When click for more info, this gives the full date of the event
   */getDateString(date){var monthInt=date.getMonth(),day=date.getDate(),monthsArray=["January","February","March","April","May","June","July","August","September","October","November","December"];monthstring=monthsArray[monthInt]+" "+date.getDate()+" "+date.getFullYear();return monthstring}/*
   * Displayes the name of the activity in the more info slideout
   */displayActivity(e){var activity=e.event.summary;return"Activity: "+activity}/*
   * Displays the start time in the more info slideout
   */displayStart(e){var startTime=this.timeString(e.event.startDate._time.hour,e.event.startDate._time.minute),endTime=this.timeString(e.event.endDate._time.hour,e.event.endDate._time.minute);/*if(startTime == "12am" && endTime == "12"){
      return "";
    }*/return"Start Time: "+startTime}/*
   * Displays the end time in the more info slideout
   */displayEnd(e){var endTime=this.timeString(e.event.endDate._time.hour,e.event.endDate._time.minute);return"End Time: "+endTime}/*
   * Displays the duration in the more info slideout
   */displayDuration(e){var duration=e.event.duration,days=duration.days,hours=duration.hours,minutes=duration.minutes,weeks=duration.weeks;if(0==hours&&0==minutes&&1==days){return"Duration: All Day"}if(0==days){if(0==hours){if(0==minutes){return"Duration: "}else{return"Duration: "+minutes+"min"}}else{if(0==minutes){return"Duration: "+hours+"H "}else{return"Duration: "+hours+"H "+minutes+"min "}}}else{if(0==hours&&0==minutes){return"Duration: "+days+"Days "}else if(0<hours&&0==minutes){return"Duration: "+days+"Days "+hours+"H "}else{return"Duration: "+days+"Days "+hours+"H "+minutes+"min "}}}/*
   * Displays the description in the more info slideout
   */displayDescription(e){var description=e.event.description;if(!description){return"Description: "}return"Description: "}/*
   * Displays the location in the more info slideout
   */displayLocation(e){var location=e.event.location;location=location+"";if(null==location||"null"==location){return"Location: "}return"Location: "+location}}_exports.LrnCalendarDate=LrnCalendarDate;window.customElements.define(LrnCalendarDate.tag,LrnCalendarDate)});
