// @lrnwebcomponents/i18n-manager/i18n-manager.js
define(["exports"], function (_exports) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.I18NManagerStore = _exports.I18NManager = void 0;

  function _regeneratorRuntime() { "use strict"; /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/facebook/regenerator/blob/main/LICENSE */ _regeneratorRuntime = function _regeneratorRuntime() { return exports; }; var exports = {}, Op = Object.prototype, hasOwn = Op.hasOwnProperty, $Symbol = "function" == typeof Symbol ? Symbol : {}, iteratorSymbol = $Symbol.iterator || "@@iterator", asyncIteratorSymbol = $Symbol.asyncIterator || "@@asyncIterator", toStringTagSymbol = $Symbol.toStringTag || "@@toStringTag"; function define(obj, key, value) { return Object.defineProperty(obj, key, { value: value, enumerable: !0, configurable: !0, writable: !0 }), obj[key]; } try { define({}, ""); } catch (err) { define = function define(obj, key, value) { return obj[key] = value; }; } function wrap(innerFn, outerFn, self, tryLocsList) { var protoGenerator = outerFn && babelHelpers.instanceof(outerFn.prototype, Generator) ? outerFn : Generator, generator = Object.create(protoGenerator.prototype), context = new Context(tryLocsList || []); return generator._invoke = function (innerFn, self, context) { var state = "suspendedStart"; return function (method, arg) { if ("executing" === state) throw new Error("Generator is already running"); if ("completed" === state) { if ("throw" === method) throw arg; return doneResult(); } for (context.method = method, context.arg = arg;;) { var delegate = context.delegate; if (delegate) { var delegateResult = maybeInvokeDelegate(delegate, context); if (delegateResult) { if (delegateResult === ContinueSentinel) continue; return delegateResult; } } if ("next" === context.method) context.sent = context._sent = context.arg;else if ("throw" === context.method) { if ("suspendedStart" === state) throw state = "completed", context.arg; context.dispatchException(context.arg); } else "return" === context.method && context.abrupt("return", context.arg); state = "executing"; var record = tryCatch(innerFn, self, context); if ("normal" === record.type) { if (state = context.done ? "completed" : "suspendedYield", record.arg === ContinueSentinel) continue; return { value: record.arg, done: context.done }; } "throw" === record.type && (state = "completed", context.method = "throw", context.arg = record.arg); } }; }(innerFn, self, context), generator; } function tryCatch(fn, obj, arg) { try { return { type: "normal", arg: fn.call(obj, arg) }; } catch (err) { return { type: "throw", arg: err }; } } exports.wrap = wrap; var ContinueSentinel = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} var IteratorPrototype = {}; define(IteratorPrototype, iteratorSymbol, function () { return this; }); var getProto = Object.getPrototypeOf, NativeIteratorPrototype = getProto && getProto(getProto(values([]))); NativeIteratorPrototype && NativeIteratorPrototype !== Op && hasOwn.call(NativeIteratorPrototype, iteratorSymbol) && (IteratorPrototype = NativeIteratorPrototype); var Gp = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(IteratorPrototype); function defineIteratorMethods(prototype) { ["next", "throw", "return"].forEach(function (method) { define(prototype, method, function (arg) { return this._invoke(method, arg); }); }); } function AsyncIterator(generator, PromiseImpl) { function invoke(method, arg, resolve, reject) { var record = tryCatch(generator[method], generator, arg); if ("throw" !== record.type) { var result = record.arg, value = result.value; return value && "object" == babelHelpers.typeof(value) && hasOwn.call(value, "__await") ? PromiseImpl.resolve(value.__await).then(function (value) { invoke("next", value, resolve, reject); }, function (err) { invoke("throw", err, resolve, reject); }) : PromiseImpl.resolve(value).then(function (unwrapped) { result.value = unwrapped, resolve(result); }, function (error) { return invoke("throw", error, resolve, reject); }); } reject(record.arg); } var previousPromise; this._invoke = function (method, arg) { function callInvokeWithMethodAndArg() { return new PromiseImpl(function (resolve, reject) { invoke(method, arg, resolve, reject); }); } return previousPromise = previousPromise ? previousPromise.then(callInvokeWithMethodAndArg, callInvokeWithMethodAndArg) : callInvokeWithMethodAndArg(); }; } function maybeInvokeDelegate(delegate, context) { var method = delegate.iterator[context.method]; if (undefined === method) { if (context.delegate = null, "throw" === context.method) { if (delegate.iterator.return && (context.method = "return", context.arg = undefined, maybeInvokeDelegate(delegate, context), "throw" === context.method)) return ContinueSentinel; context.method = "throw", context.arg = new TypeError("The iterator does not provide a 'throw' method"); } return ContinueSentinel; } var record = tryCatch(method, delegate.iterator, context.arg); if ("throw" === record.type) return context.method = "throw", context.arg = record.arg, context.delegate = null, ContinueSentinel; var info = record.arg; return info ? info.done ? (context[delegate.resultName] = info.value, context.next = delegate.nextLoc, "return" !== context.method && (context.method = "next", context.arg = undefined), context.delegate = null, ContinueSentinel) : info : (context.method = "throw", context.arg = new TypeError("iterator result is not an object"), context.delegate = null, ContinueSentinel); } function pushTryEntry(locs) { var entry = { tryLoc: locs[0] }; 1 in locs && (entry.catchLoc = locs[1]), 2 in locs && (entry.finallyLoc = locs[2], entry.afterLoc = locs[3]), this.tryEntries.push(entry); } function resetTryEntry(entry) { var record = entry.completion || {}; record.type = "normal", delete record.arg, entry.completion = record; } function Context(tryLocsList) { this.tryEntries = [{ tryLoc: "root" }], tryLocsList.forEach(pushTryEntry, this), this.reset(!0); } function values(iterable) { if (iterable) { var iteratorMethod = iterable[iteratorSymbol]; if (iteratorMethod) return iteratorMethod.call(iterable); if ("function" == typeof iterable.next) return iterable; if (!isNaN(iterable.length)) { var i = -1, next = function next() { for (; ++i < iterable.length;) { if (hasOwn.call(iterable, i)) return next.value = iterable[i], next.done = !1, next; } return next.value = undefined, next.done = !0, next; }; return next.next = next; } } return { next: doneResult }; } function doneResult() { return { value: undefined, done: !0 }; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, define(Gp, "constructor", GeneratorFunctionPrototype), define(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = define(GeneratorFunctionPrototype, toStringTagSymbol, "GeneratorFunction"), exports.isGeneratorFunction = function (genFun) { var ctor = "function" == typeof genFun && genFun.constructor; return !!ctor && (ctor === GeneratorFunction || "GeneratorFunction" === (ctor.displayName || ctor.name)); }, exports.mark = function (genFun) { return Object.setPrototypeOf ? Object.setPrototypeOf(genFun, GeneratorFunctionPrototype) : (genFun.__proto__ = GeneratorFunctionPrototype, define(genFun, toStringTagSymbol, "GeneratorFunction")), genFun.prototype = Object.create(Gp), genFun; }, exports.awrap = function (arg) { return { __await: arg }; }, defineIteratorMethods(AsyncIterator.prototype), define(AsyncIterator.prototype, asyncIteratorSymbol, function () { return this; }), exports.AsyncIterator = AsyncIterator, exports.async = function (innerFn, outerFn, self, tryLocsList, PromiseImpl) { void 0 === PromiseImpl && (PromiseImpl = Promise); var iter = new AsyncIterator(wrap(innerFn, outerFn, self, tryLocsList), PromiseImpl); return exports.isGeneratorFunction(outerFn) ? iter : iter.next().then(function (result) { return result.done ? result.value : iter.next(); }); }, defineIteratorMethods(Gp), define(Gp, toStringTagSymbol, "Generator"), define(Gp, iteratorSymbol, function () { return this; }), define(Gp, "toString", function () { return "[object Generator]"; }), exports.keys = function (object) { var keys = []; for (var key in object) { keys.push(key); } return keys.reverse(), function next() { for (; keys.length;) { var key = keys.pop(); if (key in object) return next.value = key, next.done = !1, next; } return next.done = !0, next; }; }, exports.values = values, Context.prototype = { constructor: Context, reset: function reset(skipTempReset) { if (this.prev = 0, this.next = 0, this.sent = this._sent = undefined, this.done = !1, this.delegate = null, this.method = "next", this.arg = undefined, this.tryEntries.forEach(resetTryEntry), !skipTempReset) for (var name in this) { "t" === name.charAt(0) && hasOwn.call(this, name) && !isNaN(+name.slice(1)) && (this[name] = undefined); } }, stop: function stop() { this.done = !0; var rootRecord = this.tryEntries[0].completion; if ("throw" === rootRecord.type) throw rootRecord.arg; return this.rval; }, dispatchException: function dispatchException(exception) { if (this.done) throw exception; var context = this; function handle(loc, caught) { return record.type = "throw", record.arg = exception, context.next = loc, caught && (context.method = "next", context.arg = undefined), !!caught; } for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i], record = entry.completion; if ("root" === entry.tryLoc) return handle("end"); if (entry.tryLoc <= this.prev) { var hasCatch = hasOwn.call(entry, "catchLoc"), hasFinally = hasOwn.call(entry, "finallyLoc"); if (hasCatch && hasFinally) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } else if (hasCatch) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); } else { if (!hasFinally) throw new Error("try statement without catch or finally"); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } } } }, abrupt: function abrupt(type, arg) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc <= this.prev && hasOwn.call(entry, "finallyLoc") && this.prev < entry.finallyLoc) { var finallyEntry = entry; break; } } finallyEntry && ("break" === type || "continue" === type) && finallyEntry.tryLoc <= arg && arg <= finallyEntry.finallyLoc && (finallyEntry = null); var record = finallyEntry ? finallyEntry.completion : {}; return record.type = type, record.arg = arg, finallyEntry ? (this.method = "next", this.next = finallyEntry.finallyLoc, ContinueSentinel) : this.complete(record); }, complete: function complete(record, afterLoc) { if ("throw" === record.type) throw record.arg; return "break" === record.type || "continue" === record.type ? this.next = record.arg : "return" === record.type ? (this.rval = this.arg = record.arg, this.method = "return", this.next = "end") : "normal" === record.type && afterLoc && (this.next = afterLoc), ContinueSentinel; }, finish: function finish(finallyLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.finallyLoc === finallyLoc) return this.complete(entry.completion, entry.afterLoc), resetTryEntry(entry), ContinueSentinel; } }, catch: function _catch(tryLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc === tryLoc) { var record = entry.completion; if ("throw" === record.type) { var thrown = record.arg; resetTryEntry(entry); } return thrown; } } throw new Error("illegal catch attempt"); }, delegateYield: function delegateYield(iterable, resultName, nextLoc) { return this.delegate = { iterator: values(iterable), resultName: resultName, nextLoc: nextLoc }, "next" === this.method && (this.arg = undefined), ContinueSentinel; } }, exports; }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * Copyright 2021 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
  // register globally so we can make sure there is only one
  window.I18NManagerStore = window.I18NManagerStore || {};

  window.I18NManagerStore.requestAvailability = function () {
    if (!window.I18NManagerStore.instance) {
      window.I18NManagerStore.instance = document.createElement("i18n-manager");
      document.body.appendChild(window.I18NManagerStore.instance);
    }

    return window.I18NManagerStore.instance;
  };

  var I18NManagerStore = window.I18NManagerStore.requestAvailability();
  _exports.I18NManagerStore = I18NManagerStore;
  var FALLBACK_LANG = "en";
  var FALLBACK_DIR = "ltr";
  /**
   * `i18n-manager`
   * `internationalization management singleton and helper classes`
   * @demo demo/index.html
   * @element i18n-manager
   */

  var I18NManager = /*#__PURE__*/function (_HTMLElement) {
    babelHelpers.inherits(I18NManager, _HTMLElement);

    var _super = _createSuper(I18NManager);

    /**
     * HTMLElement
     */
    function I18NManager() {
      var _this;

      babelHelpers.classCallCheck(this, I18NManager);
      _this = _super.call(this); // fetch caching to reduce calls for json files

      _this.fetchTargets = {}; // reference to all elements that care about localization

      _this.elements = []; // quick set of all locales

      _this.locales = new Set([]); // set initially based on document

      _this.lang = _this.documentLang;
      _this.dir = _this.documentDir;
      return _this;
    }
    /**
     * Set document language from these common sources
     */


    babelHelpers.createClass(I18NManager, [{
      key: "documentLang",
      get: function get() {
        return document.body.getAttribute("xml:lang") || document.body.getAttribute("lang") || document.documentElement.getAttribute("xml:lang") || document.documentElement.getAttribute("lang") || navigator.language || FALLBACK_LANG;
      }
      /**
       * Set document direction from these common sources
       */

    }, {
      key: "documentDir",
      get: function get() {
        return document.body.getAttribute("xml:dir") || document.body.getAttribute("dir") || document.documentElement.getAttribute("xml:dir") || document.documentElement.getAttribute("dir") || FALLBACK_DIR;
      }
      /**
       * Life cycle
       */

    }, {
      key: "connectedCallback",
      value: function connectedCallback() {
        this.__ready = true;
        window.addEventListener("i18n-manager-register-element", this.registerLocalizationEvent.bind(this));
        window.addEventListener("languagechange", this.changeLanguageEvent.bind(this));
      }
      /**
       * Life cycle
       */

    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        window.removeEventListener("i18n-manager-register-element", this.registerLocalizationEvent.bind(this));
        window.removeEventListener("languagechange", this.changeLanguageEvent.bind(this));
      }
      /**
       * Browser level languagechange event
       */

    }, {
      key: "changeLanguageEvent",
      value: function changeLanguageEvent(e) {
        // will trigger the language to update in all related elements
        // @see attributeChangedCallback
        this.lang = e.detail;
      }
      /**
       * Register a localization via event; this allow for a 0 dependency solution!
       */

    }, {
      key: "registerLocalizationEvent",
      value: function registerLocalizationEvent(e) {
        var detail = this.detailNormalize(e.detail); // ensure we have a namespace for later use

        if (detail.namespace && detail.localesPath && detail.locales) {
          this.registerLocalization(detail);
        }
      }
      /**
       * Apply normalization to all details bubbling up to improve
       * flexibility and patching to how other people implement our
       * API. This also can improve DX downstream by making educated
       * guesses as to intent (like namespace, localesPath, updateCallback)
       */

    }, {
      key: "detailNormalize",
      value: function detailNormalize(detail) {
        if (!detail.namespace && detail.context) {
          detail.namespace = detail.context.tagName.toLowerCase();
        } // support fallback calls for requestUpdate (LitElement) and render if nothing set


        if (!detail.updateCallback && detail.context) {
          if (detail.context.requestUpdate) {
            detail.updateCallback = "requestUpdate";
          } else if (detail.context.render) {
            detail.updateCallback = "render";
          }
        }

        if (!detail.localesPath && detail.basePath) {
          // clean up path and force adding locales. part security thing as well
          detail.localesPath = "".concat(decodeURIComponent(detail.basePath), "/../locales");
        } // minimum requirement to operate but still
        // should pull from other namespace if exists


        if (detail.context && detail.namespace) {
          // establish the fallback automatically if we are supplied defaults
          if (detail.context.t) {
            detail.context._t = _objectSpread({}, detail.context.t);
          }

          var match = this.elements.filter(function (el) {
            if (el.namespace == detail.namespace && el.localesPath && el.locales) {
              return true;
            }
          });

          if (match && match.length && match[0]) {
            detail.localesPath = match[0].localesPath;
            detail.locales = match[0].locales;
          }
        }

        return detail;
      }
      /**
       * Register a localization with the manager
       */

    }, {
      key: "registerLocalization",
      value: function registerLocalization(detail) {
        var _this2 = this;

        // ensure no dual registration of context; meaning same object twice
        if (!detail.context && this.elements.filter(function (e) {
          return e.namespace === detail.namespace;
        }).length === 0 || this.elements.filter(function (e) {
          return e.context === detail.context;
        }).length === 0) {
          detail = this.detailNormalize(detail);
          this.elements.push(detail); // store in this.locales for quick "do we support this" look up

          if (detail && detail.locales) {
            detail.locales.forEach(this.locales.add, this.locales);
          } // timing issue, see if we are ready + a language and that it happened PRIOR
          // to registration just now but match against locales we support
          // and it being the set language already


          if (this.lang && this.__ready && detail.locales && detail.locales.includes(this.lang)) {
            // prevent flooding w/ lots of translatable elements
            clearTimeout(this._debounce);
            this._debounce = setTimeout(function () {
              _this2.updateLanguage(_this2.lang);
            }, 0);
          }
        }
      }
      /**
       * Convention we use
       */

    }, {
      key: "loadNamespaceFile",
      value:
      /**
       * Return language file for a specific context
       */
      function () {
        var _loadNamespaceFile = babelHelpers.asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee(ns) {
          var lang,
              langPieces,
              nsMatch,
              el,
              fetchTarget,
              _args = arguments;
          return _regeneratorRuntime().wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  lang = _args.length > 1 && _args[1] !== undefined ? _args[1] : this.lang;
                  langPieces = lang.split("-");
                  nsMatch = this.elements.filter(function (el) {
                    return el.namespace === ns && (el.locales.includes(lang) || el.locales.includes(langPieces[0]));
                  });

                  if (!(nsMatch && nsMatch.length === 1)) {
                    _context.next = 14;
                    break;
                  }

                  el = nsMatch[0];
                  fetchTarget = "";

                  if (el.locales.includes(lang)) {
                    fetchTarget = "".concat(el.localesPath, "/").concat(el.namespace, ".").concat(lang, ".json");
                  } else if (el.locales.includes(langPieces[0])) {
                    fetchTarget = "".concat(el.localesPath, "/").concat(el.namespace, ".").concat(langPieces[0], ".json");
                  }

                  if (!(fetchTarget == "")) {
                    _context.next = 9;
                    break;
                  }

                  return _context.abrupt("return", {});

                case 9:
                  if (this.fetchTargets[fetchTarget]) {
                    _context.next = 13;
                    break;
                  }

                  _context.next = 12;
                  return fetch(fetchTarget).then(function (response) {
                    if (response && response.json) return response.json();
                    return false;
                  });

                case 12:
                  this.fetchTargets[fetchTarget] = _context.sent;

                case 13:
                  return _context.abrupt("return", this.fetchTargets[fetchTarget]);

                case 14:
                case "end":
                  return _context.stop();
              }
            }
          }, _callee, this);
        }));

        function loadNamespaceFile(_x) {
          return _loadNamespaceFile.apply(this, arguments);
        }

        return loadNamespaceFile;
      }()
      /**
       * trigger an update of the language after loading everything
       */

    }, {
      key: "updateLanguage",
      value: function () {
        var _updateLanguage = babelHelpers.asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee2(lang) {
          var langPieces, processList, fallBack, i, el, _el, fetchTarget, data, id;

          return _regeneratorRuntime().wrap(function _callee2$(_context2) {
            while (1) {
              switch (_context2.prev = _context2.next) {
                case 0:
                  if (!lang) {
                    _context2.next = 21;
                    break;
                  }

                  langPieces = lang.split("-"); // get all exact matches as well as partial matches

                  processList = this.elements.filter(function (el) {
                    try {
                      return el.locales.includes(lang) || el.locales.includes(langPieces[0]);
                    } catch (e) {
                      console.error('i18n registration incorrect in:', el, e);
                    }
                  });
                  fallBack = this.elements.filter(function (el) {
                    return !el.locales.includes(lang) && !el.locales.includes(langPieces[0]);
                  }); // no matches found, now we should fallback to defaults in the elements

                  if (fallBack.length !== 0) {
                    // fallback to documentLanguage
                    for (i in fallBack) {
                      el = fallBack[i]; // verify we have a context

                      if (el.context) {
                        // reset to the fallback language t value
                        el.context.t = _objectSpread({}, el.context._t); // support a forced update / function to run when it finishes

                        if (el.updateCallback) {
                          el.context[el.updateCallback]();
                        }
                      }
                    }
                  } // run through and match exact matches


                  _context2.t0 = _regeneratorRuntime().keys(processList);

                case 6:
                  if ((_context2.t1 = _context2.t0()).done) {
                    _context2.next = 21;
                    break;
                  }

                  i = _context2.t1.value;
                  _el = processList[i];
                  fetchTarget = "";

                  if (_el.locales.includes(lang)) {
                    fetchTarget = "".concat(_el.localesPath, "/").concat(_el.namespace, ".").concat(lang, ".json");
                  } else if (_el.locales.includes(langPieces[0])) {
                    fetchTarget = "".concat(_el.localesPath, "/").concat(_el.namespace, ".").concat(langPieces[0], ".json");
                  } // see if we had this previous to avoid another request


                  if (!this.fetchTargets[fetchTarget]) {
                    _context2.next = 15;
                    break;
                  }

                  if (_el.context) {
                    data = this.fetchTargets[fetchTarget];

                    for (id in data) {
                      _el.context.t[id] = data[id];
                    }

                    _el.context.t = _objectSpread({}, _el.context.t); // support a forced update / function to run when it finishes

                    if (_el.updateCallback) {
                      _el.context[_el.updateCallback]();
                    }
                  }

                  _context2.next = 19;
                  break;

                case 15:
                  _context2.next = 17;
                  return fetch(fetchTarget).then(function (response) {
                    if (response && response.json) return response.json();
                    return false;
                  });

                case 17:
                  this.fetchTargets[fetchTarget] = _context2.sent;

                  if (_el.context) {
                    // set values
                    for (id in this.fetchTargets[fetchTarget]) {
                      _el.context.t[id] = this.fetchTargets[fetchTarget][id];
                    } // spread can generate notify statements in downstream elements
                    // this probably makes updateCallback irrelevant in reactive
                    // projects like LitElement but just to be double sure


                    _el.context.t = _objectSpread({}, _el.context.t); // support a forced update / function to run when it finishes

                    if (_el.updateCallback && _el.context) {
                      _el.context[_el.updateCallback]();
                    }
                  }

                case 19:
                  _context2.next = 6;
                  break;

                case 21:
                case "end":
                  return _context2.stop();
              }
            }
          }, _callee2, this);
        }));

        function updateLanguage(_x2) {
          return _updateLanguage.apply(this, arguments);
        }

        return updateLanguage;
      }()
      /**
       * Life cycle
       */

    }, {
      key: "attributeChangedCallback",
      value:
      /**
       * Life cycle
       */
      function attributeChangedCallback(attr, oldValue, newValue) {
        // notify of attr change
        if (attr === "lang" || attr === "dir") {
          this.dispatchEvent(new CustomEvent("".concat(attr, "-changed"), {
            detail: {
              value: newValue
            }
          }));
        } // we are NOT moving to the default from something


        if (attr === "lang" && newValue && this.__ready) {
          this.updateLanguage(newValue);
        }
      } // getters and setters to map props to attributes

    }, {
      key: "lang",
      get: function get() {
        return this.getAttribute("lang");
      },
      set: function set(val) {
        if (!val) {
          this.removeAttribute("lang");
        } else {
          this.setAttribute("lang", val);
        }
      }
    }, {
      key: "dir",
      get: function get() {
        return this.getAttribute("dir");
      },
      set: function set(val) {
        if (!val) {
          this.removeAttribute("dir");
        } else {
          this.setAttribute("dir", val);
        }
      }
    }], [{
      key: "tag",
      get: function get() {
        return "i18n-manager";
      }
    }, {
      key: "observedAttributes",
      get: function get() {
        return ["lang", "dir"];
      }
    }]);
    return I18NManager;
  }( /*#__PURE__*/babelHelpers.wrapNativeSuper(HTMLElement));

  _exports.I18NManager = I18NManager;
  customElements.define(I18NManager.tag, I18NManager);
});
