// @lrnwebcomponents/team-member/team-member.js
define(["exports", "require", "../../@polymer/polymer/polymer-element.js", "../schema-behaviors/schema-behaviors.js"], function (_exports, _require, _polymerElement, _schemaBehaviors) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.TeamMember = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject_edc58280d10411ea8ac18ff2c3761fda() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: block;\n          --team-member-circle: #1e407d;\n          --team-member-border: white;\n        }\n        .team-member {\n          text-align: center;\n          padding: 8px;\n        }\n        iron-image {\n          background-color: var(--team-member-color);\n          height: 165px;\n          width: 165px;\n          margin: 0 auto;\n          border: 7px solid var(--team-member-border);\n          border-radius: 50%;\n        }\n        .name {\n          text-transform: none;\n          font-size: 16px;\n        }\n        .line1 {\n          font-size: 12px;\n          margin: 0;\n          padding: 4px 0;\n          line-height: 22px;\n        }\n        .line2 {\n          font-size: 12px;\n          margin: 0;\n          padding: 4px 0;\n          line-height: 18px;\n        }\n        .social {\n          display: inline-block;\n          padding-left: 5px;\n          padding-right: 5px;\n        }\n        paper-icon-button {\n          color: var(--team-member-color);\n        }\n      </style>\n      <div class=\"team-member\">\n        <iron-image src=\"[[image]]\" sizing=\"cover\" alt=\"\"></iron-image>\n        <div class=\"name\">[[fullName]]</div>\n        <div hidden$=\"[[!firstLine]]\" class=\"line1\">[[firstLine]]</div>\n        <div hidden$=\"[[!secondLine]]\" class=\"line2\">[[secondLine]]</div>\n        <div class=\"social\">\n          <a tabindex=\"-1\" href$=\"[[dribble]]\" hidden$=\"[[!dribble]]\"\n            ><paper-icon-button icon=\"mdi-social:dribble\"></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[facebook]]\" hidden$=\"[[!facebook]]\"\n            ><paper-icon-button\n              icon=\"mdi-social:facebook-box\"\n            ></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[github]]\" hidden$=\"[[!github]]\"\n            ><paper-icon-button\n              icon=\"mdi-social:github-circle\"\n            ></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[google]]\" hidden$=\"[[!google]]\"\n            ><paper-icon-button\n              icon=\"mdi-social:google-plus\"\n            ></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[instagram]]\" hidden$=\"[[!instagram]]\"\n            ><paper-icon-button icon=\"mdi-social:instagram\"></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[linkedin]]\" hidden$=\"[[!linkedin]]\"\n            ><paper-icon-button icon=\"mdi-social:linkedin\"></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[pinterest]]\" hidden$=\"[[!pinterest]]\"\n            ><paper-icon-button icon=\"mdi-social:pinterest\"></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[tumblr]]\" hidden$=\"[[!tumblr]]\"\n            ><paper-icon-button icon=\"mdi-social:tumblr\"></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[twitch]]\" hidden$=\"[[!twitch]]\"\n            ><paper-icon-button icon=\"mdi-social:twitch\"></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[twitter]]\" hidden$=\"[[!twitter]]\"\n            ><paper-icon-button icon=\"mdi-social:twitter\"></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[whatsapp]]\" hidden$=\"[[!whatsapp]]\"\n            ><paper-icon-button icon=\"mdi-social:whatsapp\"></paper-icon-button\n          ></a>\n        </div>\n      </div>\n    "], ["\n      <style>\n        :host {\n          display: block;\n          --team-member-circle: #1e407d;\n          --team-member-border: white;\n        }\n        .team-member {\n          text-align: center;\n          padding: 8px;\n        }\n        iron-image {\n          background-color: var(--team-member-color);\n          height: 165px;\n          width: 165px;\n          margin: 0 auto;\n          border: 7px solid var(--team-member-border);\n          border-radius: 50%;\n        }\n        .name {\n          text-transform: none;\n          font-size: 16px;\n        }\n        .line1 {\n          font-size: 12px;\n          margin: 0;\n          padding: 4px 0;\n          line-height: 22px;\n        }\n        .line2 {\n          font-size: 12px;\n          margin: 0;\n          padding: 4px 0;\n          line-height: 18px;\n        }\n        .social {\n          display: inline-block;\n          padding-left: 5px;\n          padding-right: 5px;\n        }\n        paper-icon-button {\n          color: var(--team-member-color);\n        }\n      </style>\n      <div class=\"team-member\">\n        <iron-image src=\"[[image]]\" sizing=\"cover\" alt=\"\"></iron-image>\n        <div class=\"name\">[[fullName]]</div>\n        <div hidden\\$=\"[[!firstLine]]\" class=\"line1\">[[firstLine]]</div>\n        <div hidden\\$=\"[[!secondLine]]\" class=\"line2\">[[secondLine]]</div>\n        <div class=\"social\">\n          <a tabindex=\"-1\" href\\$=\"[[dribble]]\" hidden\\$=\"[[!dribble]]\"\n            ><paper-icon-button icon=\"mdi-social:dribble\"></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href\\$=\"[[facebook]]\" hidden\\$=\"[[!facebook]]\"\n            ><paper-icon-button\n              icon=\"mdi-social:facebook-box\"\n            ></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href\\$=\"[[github]]\" hidden\\$=\"[[!github]]\"\n            ><paper-icon-button\n              icon=\"mdi-social:github-circle\"\n            ></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href\\$=\"[[google]]\" hidden\\$=\"[[!google]]\"\n            ><paper-icon-button\n              icon=\"mdi-social:google-plus\"\n            ></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href\\$=\"[[instagram]]\" hidden\\$=\"[[!instagram]]\"\n            ><paper-icon-button icon=\"mdi-social:instagram\"></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href\\$=\"[[linkedin]]\" hidden\\$=\"[[!linkedin]]\"\n            ><paper-icon-button icon=\"mdi-social:linkedin\"></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href\\$=\"[[pinterest]]\" hidden\\$=\"[[!pinterest]]\"\n            ><paper-icon-button icon=\"mdi-social:pinterest\"></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href\\$=\"[[tumblr]]\" hidden\\$=\"[[!tumblr]]\"\n            ><paper-icon-button icon=\"mdi-social:tumblr\"></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href\\$=\"[[twitch]]\" hidden\\$=\"[[!twitch]]\"\n            ><paper-icon-button icon=\"mdi-social:twitch\"></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href\\$=\"[[twitter]]\" hidden\\$=\"[[!twitter]]\"\n            ><paper-icon-button icon=\"mdi-social:twitter\"></paper-icon-button\n          ></a>\n          <a tabindex=\"-1\" href\\$=\"[[whatsapp]]\" hidden\\$=\"[[!whatsapp]]\"\n            ><paper-icon-button icon=\"mdi-social:whatsapp\"></paper-icon-button\n          ></a>\n        </div>\n      </div>\n    "]);

    _templateObject_edc58280d10411ea8ac18ff2c3761fda = function _templateObject_edc58280d10411ea8ac18ff2c3761fda() {
      return data;
    };

    return data;
  }

  /**
   * `team-member`
   * @element team-member
   * `A simple presentation of a team member and basic info`
   * @demo demo/index.html
   */
  var TeamMember =
  /*#__PURE__*/
  function (_SchemaBehaviors) {
    babelHelpers.inherits(TeamMember, _SchemaBehaviors);

    function TeamMember() {
      var _this;

      babelHelpers.classCallCheck(this, TeamMember);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(TeamMember).call(this));
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/iron-iconset-svg/iron-iconset-svg.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/iron-image/iron-image.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/paper-icon-button/paper-icon-button.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../mdi-iconset-svg/lib/mdi-social-iconset-svg.js"], res, rej);
      });
      return _this;
    }

    babelHelpers.createClass(TeamMember, null, [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_edc58280d10411ea8ac18ff2c3761fda());
      }
    }, {
      key: "tag",
      get: function get() {
        return "team-member";
      }
    }, {
      key: "haxProperties",
      get: function get() {
        return {
          canScale: true,
          canPosition: true,
          canEditSource: false,
          gizmo: {
            title: "Team Member",
            description: "The user will be able to see this for selection in a UI.",
            icon: "av:play-circle-filled",
            color: "grey",
            groups: ["Content", "Presentation"],
            handles: [],
            meta: {
              author: "ELMS:LN"
            }
          },
          settings: {
            quick: [{
              property: "title",
              title: "Title",
              description: "The title of the element",
              inputMethod: "textfield",
              icon: "editor:title"
            }],
            configure: [{
              property: "title",
              title: "Title",
              description: "The title of the element",
              inputMethod: "textfield",
              icon: "editor:title"
            }],
            advanced: []
          }
        };
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(TeamMember), "properties", this), {
          /**
           * Image
           */
          image: {
            type: String
          },

          /**
           * Full name of the team member
           */
          fullName: {
            type: String
          },

          /**
           * first line
           */
          firstLine: {
            type: String,
            value: false
          },

          /**
           * second line
           */
          secondLine: {
            type: String,
            value: false
          },

          /**
           * dribble
           */
          dribble: {
            type: String,
            value: false
          },

          /**
           * facebook
           */
          facebook: {
            type: String,
            value: false
          },

          /**
           * github
           */
          github: {
            type: String,
            value: false
          },

          /**
           * google
           */
          google: {
            type: String,
            value: false
          },

          /**
           * instagram
           */
          instagram: {
            type: String,
            value: false
          },

          /**
           * linkedin
           */
          linkedin: {
            type: String,
            value: false
          },

          /**
           * pinterest
           */
          pinterest: {
            type: String,
            value: false
          },

          /**
           * tumblr
           */
          tumblr: {
            type: String,
            value: false
          },

          /**
           * twitch
           */
          twitch: {
            type: String,
            value: false
          },

          /**
           * twitter
           */
          twitter: {
            type: String,
            value: false
          },

          /**
           * whatsapp
           */
          whatsapp: {
            type: String,
            value: false
          }
        });
      }
    }]);
    return TeamMember;
  }((0, _schemaBehaviors.SchemaBehaviors)(_polymerElement.PolymerElement));

  _exports.TeamMember = TeamMember;
  window.customElements.define(TeamMember.tag, TeamMember);
});
