// @lrnwebcomponents/team-member/team-member.js
define(["exports", "../../@polymer/polymer/polymer-element.js", "../schema-behaviors/schema-behaviors.js", "../simple-icon/simple-icon.js", "../hax-iconset/lib/simple-hax-iconset.js"], function (_exports, _polymerElement, _schemaBehaviors, _simpleIcon, _simpleHaxIconset) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.TeamMember = void 0;

  var _templateObject_22496200942811ecb1937969798b231a;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `team-member`
   * @element team-member
   * `A simple presentation of a team member and basic info`
   * @demo demo/index.html
   */
  var TeamMember = /*#__PURE__*/function (_SchemaBehaviors) {
    babelHelpers.inherits(TeamMember, _SchemaBehaviors);

    var _super = _createSuper(TeamMember);

    function TeamMember() {
      babelHelpers.classCallCheck(this, TeamMember);
      return _super.apply(this, arguments);
    }

    babelHelpers.createClass(TeamMember, null, [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_22496200942811ecb1937969798b231a || (_templateObject_22496200942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: block;\n          --team-member-circle: #1e407d;\n          --team-member-border: white;\n        }\n        .team-member {\n          text-align: center;\n          padding: 8px;\n        }\n        img.image {\n          background-color: var(--team-member-color);\n          height: 165px;\n          width: 165px;\n          margin: 0 auto;\n          border: 7px solid var(--team-member-border);\n          border-radius: 50%;\n        }\n        .name {\n          text-transform: none;\n          font-size: 16px;\n        }\n        .line1 {\n          font-size: 12px;\n          margin: 0;\n          padding: 4px 0;\n          line-height: 22px;\n        }\n        .line2 {\n          font-size: 12px;\n          margin: 0;\n          padding: 4px 0;\n          line-height: 18px;\n        }\n        .social {\n          display: inline-block;\n          padding-left: 5px;\n          padding-right: 5px;\n        }\n        simple-icon {\n          color: var(--team-member-color);\n        }\n      </style>\n      <div class=\"team-member\">\n        <img class=\"image\" src=\"[[image]]\" loading=\"lazy\" alt=\"\" />\n        <div class=\"name\">[[fullName]]</div>\n        <div hidden$=\"[[!firstLine]]\" class=\"line1\">[[firstLine]]</div>\n        <div hidden$=\"[[!secondLine]]\" class=\"line2\">[[secondLine]]</div>\n        <div class=\"social\">\n          <a tabindex=\"-1\" href$=\"[[dribble]]\" hidden$=\"[[!dribble]]\"\n            ><simple-icon icon=\"mdi-social:dribble\"></simple-icon\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[facebook]]\" hidden$=\"[[!facebook]]\"\n            ><simple-icon icon=\"mdi-social:facebook-box\"></simple-icon\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[github]]\" hidden$=\"[[!github]]\"\n            ><simple-icon icon=\"mdi-social:github-circle\"></simple-icon\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[google]]\" hidden$=\"[[!google]]\"\n            ><simple-icon icon=\"mdi-social:google-plus\"></simple-icon\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[instagram]]\" hidden$=\"[[!instagram]]\"\n            ><simple-icon icon=\"mdi-social:instagram\"></simple-icon\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[linkedin]]\" hidden$=\"[[!linkedin]]\"\n            ><simple-icon icon=\"mdi-social:linkedin\"></simple-icon\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[pinterest]]\" hidden$=\"[[!pinterest]]\"\n            ><simple-icon icon=\"mdi-social:pinterest\"></simple-icon\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[tumblr]]\" hidden$=\"[[!tumblr]]\"\n            ><simple-icon icon=\"mdi-social:tumblr\"></simple-icon\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[twitch]]\" hidden$=\"[[!twitch]]\"\n            ><simple-icon icon=\"mdi-social:twitch\"></simple-icon\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[twitter]]\" hidden$=\"[[!twitter]]\"\n            ><simple-icon icon=\"mdi-social:twitter\"></simple-icon\n          ></a>\n          <a tabindex=\"-1\" href$=\"[[whatsapp]]\" hidden$=\"[[!whatsapp]]\"\n            ><simple-icon icon=\"mdi-social:whatsapp\"></simple-icon\n          ></a>\n        </div>\n      </div>\n    "])));
      }
    }, {
      key: "tag",
      get: function get() {
        return "team-member";
      }
    }, {
      key: "haxProperties",
      get: function get() {
        return {
          canScale: true,
          canPosition: true,
          canEditSource: true,
          gizmo: {
            title: "Team Member",
            description: "The user will be able to see this for selection in a UI.",
            icon: "av:play-circle-filled",
            color: "grey",
            groups: ["Content", "Presentation"],
            handles: [],
            meta: {
              author: "ELMS:LN"
            }
          },
          settings: {
            configure: [{
              property: "title",
              title: "Title",
              description: "The title of the element",
              inputMethod: "textfield",
              icon: "editor:title"
            }],
            advanced: []
          }
        };
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(TeamMember), "properties", this)), {}, {
          /**
           * Image
           */
          image: {
            type: String
          },

          /**
           * Full name of the team member
           */
          fullName: {
            type: String
          },

          /**
           * first line
           */
          firstLine: {
            type: String,
            value: false
          },

          /**
           * second line
           */
          secondLine: {
            type: String,
            value: false
          },

          /**
           * dribble
           */
          dribble: {
            type: String,
            value: false
          },

          /**
           * facebook
           */
          facebook: {
            type: String,
            value: false
          },

          /**
           * github
           */
          github: {
            type: String,
            value: false
          },

          /**
           * google
           */
          google: {
            type: String,
            value: false
          },

          /**
           * instagram
           */
          instagram: {
            type: String,
            value: false
          },

          /**
           * linkedin
           */
          linkedin: {
            type: String,
            value: false
          },

          /**
           * pinterest
           */
          pinterest: {
            type: String,
            value: false
          },

          /**
           * tumblr
           */
          tumblr: {
            type: String,
            value: false
          },

          /**
           * twitch
           */
          twitch: {
            type: String,
            value: false
          },

          /**
           * twitter
           */
          twitter: {
            type: String,
            value: false
          },

          /**
           * whatsapp
           */
          whatsapp: {
            type: String,
            value: false
          }
        });
      }
    }]);
    return TeamMember;
  }((0, _schemaBehaviors.SchemaBehaviors)(_polymerElement.PolymerElement));

  _exports.TeamMember = TeamMember;
  window.customElements.define(TeamMember.tag, TeamMember);
});
