// @lrnwebcomponents/a11y-media-player/lib/a11y-media-transcript-controls.js
define(["exports","../../../@polymer/polymer/polymer-element.js","./a11y-media-behaviors.js","../../simple-search/simple-search.js","./a11y-media-button.js"],function(_exports,_polymerElement,_a11yMediaBehaviors,_simpleSearch,_a11yMediaButton){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.A11yMediaTranscriptControls=void 0;class A11yMediaTranscriptControls extends _a11yMediaBehaviors.A11yMediaBehaviors{static get properties(){return{target:{type:Object,value:null}}}static get tag(){return"a11y-media-transcript-controls"}static get behaviors(){return[_a11yMediaBehaviors.A11yMediaBehaviors]}static get template(){return _polymerElement.html`
      <style>
        :host {
          display: flex;
          height: 44px;
          max-height: 44px;
          min-height: 44px;
          color: var(--a11y-media-color);
          background-color: var(--a11y-media-transcript-bg-color);
          --a11y-media-button-bg-color: var(--a11y-media-transcript-bg-color);
          --a11y-media-button-hover-bg-color: var(
            --a11y-media-transcript-bg-color
          );
          --simple-search-input-text-color: var(--a11y-media-color);
          --simple-search-input-line-color: var(--a11y-media-accent-color);
          --simple-search-input-placeholder-color: var(
            --a11y-media-transcript-color
          );
          --simple-search-button-color: var(--a11y-media-accent-color);
          --simple-search-button-hover-color: var(
            --a11y-media-faded-accent-color
          );
          --simple-search-button-bg-color: var(--a11y-media-bg-color);
          --simple-search-button-border-color: var(--a11y-media-bg-color);
          --simple-search-button-hover-border-color: var(--a11y-media-bg-color);
          --simple-search-button-disabled-color: var(
            --simple-colors-default-theme-grey-5
          );
          --simple-search-button-disabled-bg-color: var(
            --simple-colors-default-theme-grey-2
          );
          --simple-search-button-disabled-border-color: var(
            --simple-colors-default-theme-grey-3
          );
          --paper-input-container-input-color: var(--a11y-media-color);
          --simple-search-container: {
            padding: 0 15px;
          }
        }
        :host #searchbar {
          display: flex;
          align-items: stretch;
          justify-content: space-between;
          width: 100%;
        }
        :host #searching {
          flex-grow: 2;
        }
        :host #autoscroll {
          padding-right: 8px;
        }
        :host #scrolling,
        :host #printing {
          display: flex;
          align-items: center;
        }
        @media print {
          :host {
            display: none;
          }
        }
      </style>
      <div id="searchbar">
        <div id="searching">
          <simple-search
            id="simplesearch"
            controls="transcript"
            disabled$="[[disableSearch]]"
            hidden$="[[disableSearch]]"
            no-label-float
            next-button-icon$="[[_getLocal('nextResult','icon')]]"
            next-button-label$="[[_getLocal('nextResult','label')]]"
            prev-button-icon$="[[_getLocal('prevResult','icon')]]"
            prev-button-label$="[[_getLocal('prevResult','label')]]"
            search-input-icon$="[[_getLocal('search','icon')]]"
            search-input-label$="[[_getLocal('search','label')]]"
            target="[[target]]"
          >
          </simple-search>
        </div>
        <div id="scrolling">
          <a11y-media-button
            id="scroll"
            controls="transcript"
            icon="[[_getLocal('autoScroll','icon')]]"
            label="[[_getLocal('autoScroll','label')]]"
            on-tap="_handleScrollClick"
            toggle$="[[!disableScroll]]"
          >
          </a11y-media-button>
        </div>
        <div
          id="printing"
          hidden$="[[disablePrintButton]]"
          disabled$="[[disablePrintButton]]"
        >
          <a11y-media-button
            id="download"
            controls="transcript"
            icon$="[[_getLocal('download','icon')]]"
            label="[[_getLocal('download','label')]]"
            on-tap="_handleDownloadClick"
          >
          </a11y-media-button>
          <a11y-media-button
            id="print"
            controls="transcript"
            icon$="[[_getLocal('print','icon')]]"
            label="[[_getLocal('print','label')]]"
            on-tap="_handlePrintClick"
          >
          </a11y-media-button>
        </div>
      </div>
    `}connectedCallback(){super.connectedCallback();this.search=this.$.simplesearch,this.dispatchEvent(new CustomEvent("searchbar-added",{detail:this.search}))}_handleScrollClick(e){this.dispatchEvent(new CustomEvent("toggle-scroll",{detail:this}))}}_exports.A11yMediaTranscriptControls=A11yMediaTranscriptControls,window.customElements.define(A11yMediaTranscriptControls.tag,A11yMediaTranscriptControls)});
