// @material/mwc-ripple/ripple-handlers.js
define(["exports"], function (_exports) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.RippleHandlers = void 0;

  /**
  @license
  Copyright 2020 Google Inc. All Rights Reserved.
  
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at
  
      http://www.apache.org/licenses/LICENSE-2.0
  
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  */

  /**
   * Class that encapsulates the events handlers for `mwc-ripple`
   *
   *
   * Example:
   * ```
   * class XFoo extends LitElement {
   *   async getRipple() {
   *     this.renderRipple = true;
   *     await this.updateComplete;
   *     return this.renderRoot.querySelector('mwc-ripple');
   *   }
   *   rippleHandlers = new RippleHandlers(() => this.getRipple());
   *
   *   render() {
   *     return html`
   *       <div @mousedown=${this.rippleHandlers.startPress}></div>
   *       ${this.renderRipple ? html`<mwc-ripple></mwc-ripple>` : ''}
   *     `;
   *   }
   * }
   * ```
   */
  var RippleHandlers = function RippleHandlers(
  /** Function that returns a `mwc-ripple` */
  rippleFn) {
    babelHelpers.classCallCheck(this, RippleHandlers);

    this.startPress = function (ev) {
      rippleFn().then(function (r) {
        r && r.startPress(ev);
      });
    };

    this.endPress = function () {
      rippleFn().then(function (r) {
        r && r.endPress();
      });
    };

    this.startFocus = function () {
      rippleFn().then(function (r) {
        r && r.startFocus();
      });
    };

    this.endFocus = function () {
      rippleFn().then(function (r) {
        r && r.endFocus();
      });
    };

    this.startHover = function () {
      rippleFn().then(function (r) {
        r && r.startHover();
      });
    };

    this.endHover = function () {
      rippleFn().then(function (r) {
        r && r.endHover();
      });
    };
  };

  _exports.RippleHandlers = RippleHandlers;
});
