// @lrnwebcomponents/check-it-out/check-it-out.js
define(["exports", "require", "../../lit/index.js", "../simple-icon/lib/simple-icons.js", "../simple-icon/lib/simple-icon-lite.js", "../intersection-element/lib/IntersectionObserverMixin.js"], function (_exports, _require, _index, _simpleIcons, _simpleIconLite, _IntersectionObserverMixin) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.CheckItOut = void 0;
  _require = _interopRequireWildcard(_require);

  var _templateObject_1438c430942811ecb1937969798b231a, _templateObject2_1438c430942811ecb1937969798b231a, _templateObject3_1438c430942811ecb1937969798b231a, _templateObject4_1438c430942811ecb1937969798b231a, _templateObject5_1438c430942811ecb1937969798b231a, _templateObject6_1438c430942811ecb1937969798b231a, _templateObject7_1438c430942811ecb1937969798b231a, _templateObject8_1438c430942811ecb1937969798b231a, _templateObject9_1438c430942811ecb1937969798b231a, _templateObject10_1438c430942811ecb1937969798b231a;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `check-it-out`
   * `View codepen or stackblitz demos`
   * @demo demo/index.html
   * @element check-it-out
   */
  var CheckItOut = /*#__PURE__*/function (_IntersectionObserver) {
    babelHelpers.inherits(CheckItOut, _IntersectionObserver);

    var _super = _createSuper(CheckItOut);

    /**
     * HTMLElement
     */
    function CheckItOut() {
      var _this;

      babelHelpers.classCallCheck(this, CheckItOut);
      _this = _super.call(this);
      _this.typeIconObj = {
        code: "code",
        pdf: "book",
        video: "av:play-arrow"
      };
      _this.icon = "code";
      _this.modalTitle = "";
      _this.checkedOut = false;
      _this.type = "";
      _this.hideExplorer = false;
      _this.ctl = false;
      _this.view = "both";
      return _this;
    }

    babelHelpers.createClass(CheckItOut, [{
      key: "render",
      value: function render() {
        this.icon = this.icon ? this.icon : this.typeIconObj[this.type];
        return (0, _index.html)(_templateObject_1438c430942811ecb1937969798b231a || (_templateObject_1438c430942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral([" ", ""])), this.elementVisible ? (0, _index.html)(_templateObject2_1438c430942811ecb1937969798b231a || (_templateObject2_1438c430942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n          <button\n            class=\"check-it-out-btn\"\n            @click=\"", "\"\n            controls=\"m1\"\n          >\n            <simple-icon-lite icon=\"", "\"> </simple-icon-lite>\n            ", "\n            <slot></slot>\n          </button>\n          ", "\n        "])), this._handleClick, this.icon, this.label, this.modal ? (0, _index.html)(_templateObject3_1438c430942811ecb1937969798b231a || (_templateObject3_1438c430942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                <simple-modal-template\n                  title=\"", "\"\n                  @click=\"", "\"\n                  modal-id=\"m1\"\n                >\n                  <div slot=\"content\">\n                    <!-- bandaid and a terrible one -->\n                    <style>\n                      simple-modal .container {\n                        display: block;\n                        position: relative;\n                        width: var(--check-it-out-container-width, 70vw);\n                      }\n                      simple-modal .container .iframe-container {\n                        width: var(--check-it-out-content-width, 70vw);\n                        height: var(--check-it-out-content-height, 45vh);\n                      }\n\n                      simple-modal .container .video-player {\n                        width: var(--check-it-out-content-width, 70vw);\n                        height: var(--check-it-out-content-height, 45vh);\n                      }\n                      simple-modal .container iframe-loader {\n                        text-align: center;\n                        display: block;\n                        width: var(--check-it-out-content-width, 45vh);\n                      }\n                    </style>\n                    ", "\n                  </div>\n                </simple-modal-template>\n              "])), this.modalTitle, this._handleClick, this.renderLogic()) : (0, _index.html)(_templateObject4_1438c430942811ecb1937969798b231a || (_templateObject4_1438c430942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral([" ", " "])), this.renderLogic())) : "");
      }
    }, {
      key: "renderLogic",
      value: function renderLogic() {
        return (0, _index.html)(_templateObject5_1438c430942811ecb1937969798b231a || (_templateObject5_1438c430942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      ", "\n    "])), ["code", "pdf"].includes(this.type) ? (0, _index.html)(_templateObject6_1438c430942811ecb1937969798b231a || (_templateObject6_1438c430942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<div class=\"container\">\n            <iframe-loader>\n              <iframe\n                class=\"iframe-container\"\n                src=", "\n                loading=\"lazy\"\n              ></iframe>\n              ", "\n            </iframe-loader>\n          </div>"])), this.__computedSource, !this.modal ? (0, _index.html)(_templateObject7_1438c430942811ecb1937969798b231a || (_templateObject7_1438c430942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                    <button\n                      class=\"close-btn\"\n                      @click=\"", "\"\n                    ></button>\n                  "])), this._handleClick) : "") : (0, _index.html)(_templateObject8_1438c430942811ecb1937969798b231a || (_templateObject8_1438c430942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n            <div class=\"container\">\n              <video-player\n                class=\"video-player\"\n                source=", "\n              >\n              </video-player>\n              ", "\n            </div>\n          "])), this.__computedSource, !this.modal ? (0, _index.html)(_templateObject9_1438c430942811ecb1937969798b231a || (_templateObject9_1438c430942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n                    <button\n                      class=\"close-btn\"\n                      @click=\"", "\"\n                    ></button>\n                  "])), this._handleClick) : ""));
      }
    }, {
      key: "_handleClick",
      value: function _handleClick(event) {
        if (this._haxstate && this.modal) {
          event.preventDefault();
          event.stopPropagation();
          event.stopImmediatePropagation();
        } else if (!this._haxstate && !this.modal) {
          if (this.checkedOut) {
            this.checkedOut = false;
          } else {
            this.checkedOut = true;
          }
        }
      }
    }, {
      key: "haxHooks",
      value: function haxHooks() {
        return {
          editModeChanged: "haxeditModeChanged",
          activeElementChanged: "haxactiveElementChanged"
        };
      }
    }, {
      key: "haxactiveElementChanged",
      value: function haxactiveElementChanged(element, value) {
        if (value) {
          this._haxstate = value;
        }
      }
    }, {
      key: "haxeditModeChanged",
      value: function haxeditModeChanged(value) {
        this._haxstate = value;
      }
      /*
        Checks source string and determines a type
      */

    }, {
      key: "checkType",
      value: function checkType(sourceStr) {
        var videoTypes = [".mp4", ".mkv", ".flv", ".wmv", ".ovg", ".webm", ".mov"];

        if (videoTypes.filter(function (type) {
          return sourceStr.endsWith(type);
        }).length > 0 || sourceStr.includes("youtube.com")) {
          this.type = "video";
          new Promise(function (res, rej) {
            return _require.default(["../video-player/video-player.js"], res, rej);
          });
        } else if (sourceStr.endsWith(".pdf")) {
          this.type = "pdf";
        } else if (sourceStr.includes("stackblitz.com")) {
          this.type = "code";

          if (!sourceStr.includes("embed=1")) {
            if (sourceStr.includes("?")) {
              sourceStr += "&embed=1";
            } else {
              sourceStr += "?embed=1";
            }
          }

          sourceStr = this.checkStackblitzProps(sourceStr);
        } else if (sourceStr.includes("codepen.io")) {
          this.type = "code";

          if (!sourceStr.includes("embed")) {
            sourceStr = sourceStr.replace("/pen/", "/embed/");
          }
        }

        if (this.type != "video") {
          new Promise(function (res, rej) {
            return _require.default(["../iframe-loader/iframe-loader.js"], res, rej);
          });
        }

        return sourceStr;
      }
      /* 
        Apply additional stackblitz properties
      */

    }, {
      key: "checkStackblitzProps",
      value: function checkStackblitzProps(sourceStr) {
        if (this.filePath) {
          sourceStr += "&file=" + this.filePath;
        }

        if (this.hideExplorer) {
          sourceStr += "&hideExplorer=1";
        }

        if (this.ctl) {
          sourceStr += "&ctl=1";
        }

        sourceStr += "&view=".concat(this.view);
        return sourceStr;
      }
      /**
       * LitElement life cycle - property changed
       */

    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        if (babelHelpers.get(babelHelpers.getPrototypeOf(CheckItOut.prototype), "updated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(CheckItOut.prototype), "updated", this).call(this, changedProperties);
        }

        changedProperties.forEach(function (oldValue, propName) {
          // import modal if we need it only when we actually leverage it
          if (propName === "modal" && _this2[propName]) {
            // prettier-ignore
            new Promise(function (res, rej) {
              return _require.default(["../simple-modal/lib/simple-modal-template.js"], res, rej);
            }).then(function () {
              if (_this2.shadowRoot) {
                _this2.shadowRoot.querySelector('[modal-id="m1"]').associateEvents(_this2.shadowRoot.querySelector('[controls="m1"]'));
              }
            });
          }

          if (propName === "source" && _this2[propName]) {
            clearTimeout(_this2.__debounce);
            _this2.__debounce = setTimeout(function () {
              _this2.__computedSource = _this2.checkType(_this2.source);
            }, 0);
          }
        });
      }
    }], [{
      key: "haxProperties",
      get: function get() {
        return {
          canScale: false,
          canPosition: false,
          canEditSource: true,
          gizmo: {
            title: "Check it out",
            description: "Clickable button for media and code previews",
            icon: "check-circle",
            color: "grey",
            groups: ["developer", "code", "video", "pdf"],
            handles: [],
            meta: {
              author: "collinkleest",
              owner: "ELMS:LN"
            }
          },
          settings: {
            configure: [{
              property: "source",
              title: "Resource url",
              description: "Resource url",
              inputMethod: "textfield"
            }, {
              property: "icon",
              title: "Icon",
              description: "Icon for button",
              inputMethod: "textfield"
            }, {
              property: "type",
              title: "Type of resource",
              description: "Type of resource in url (video, code, pdf)",
              inputMethod: "select",
              options: {
                code: "Code Snippet",
                video: "Video",
                pdf: "PDF / Document"
              }
            }, {
              property: "label",
              title: "Button label",
              description: "Text to display in check-it-out button",
              inputMethod: "textfield"
            }, {
              property: "modal",
              title: "Enable Modal Mode",
              description: "Puts content inside a popup modal",
              inputMethod: "boolean"
            }, {
              property: "modal-title",
              title: "Popup Modal Title",
              description: "Puts a title at the top of popup modal",
              inputMethod: "textfield"
            }, {
              property: "filePath",
              title: "File Path",
              description: "File path for stackblitz embed",
              inputMethod: "textfield"
            }, {
              property: "hideExplorer",
              title: "Hide file explorer",
              description: "Hide the file explorer in a stackblitz embed",
              inputMethod: "boolean"
            }, {
              property: "ctl",
              title: "Click to load stackblitz",
              description: "Require click to load stackblitz embed",
              inputMethod: "boolean"
            }, {
              property: "view",
              title: "View",
              description: "Stackblitz default view (preview, editor, or both)",
              inputMethod: "select",
              options: {
                preview: "Preview",
                editor: "Editor",
                both: "Both"
              }
            }],
            advanced: []
          },
          demoSchema: [{
            tag: "check-it-out",
            properties: {
              org: "elmsln",
              repo: "lrnwebcomponents"
            },
            content: ""
          }]
        };
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(CheckItOut), "properties", this)), {}, {
          checkedOut: {
            type: Boolean,
            attribute: "checked-out",
            reflect: true
          },
          source: {
            type: String
          },
          // icon for button
          icon: {
            type: String
          },
          // modified source depending on media type
          __computedSource: {
            type: String
          },
          // type of media (i.e. pdf, video, codepen, stackblitz)
          type: {
            type: String,
            reflect: true
          },
          // button text
          label: {
            type: String,
            reflect: true
          },
          // stackblitz file path
          filePath: {
            type: String,
            attribute: "file-path"
          },
          // modal header
          modalTitle: {
            type: String,
            attribute: "modal-title"
          },
          // hides stackblitz file explorer pane
          hideExplorer: {
            type: Boolean,
            attribute: "hide-explorer"
          },
          // require user to click to load stackblitz embed
          ctl: {
            type: Boolean
          },
          // stackblitz "editor", "preview" or "both" open by default
          view: {
            type: String
          },
          modal: {
            type: Boolean,
            reflect: true
          }
        });
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _index.css)(_templateObject10_1438c430942811ecb1937969798b231a || (_templateObject10_1438c430942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          --check-it-out-content-width: var(\n            --check-it-out-container-width,\n            800px\n          );\n          --check-it-out-content-height: 500px;\n        }\n\n        .container {\n          display: none;\n          position: relative;\n          width: var(--check-it-out-container-width, 800px);\n        }\n\n        :host([checked-out]) .container {\n          display: flex;\n        }\n\n        .check-it-out-btn {\n          display: inline-flex;\n          background-color: transparent;\n          font-size: var(--check-it-out-modal-button-font-size, 18px);\n          border: 2px solid var(--check-it-out-button-border-color, #000);\n          font-weight: 700;\n          padding: 10px 25px;\n        }\n\n        .check-it-out-btn:active,\n        .check-it-out-btn:focus,\n        .check-it-out-btn:hover {\n          outline: 2px dashed var(--check-it-out-button-border-color, #000);\n          outline-offset: 2px;\n        }\n\n        :host([checked-out]) .check-it-out-btn {\n          display: none;\n        }\n\n        :host([checked-out]) .close-btn {\n          display: flex;\n        }\n\n        .close-btn {\n          display: none;\n          border-radius: 50%;\n          padding: 0.5em;\n          width: 30px;\n          height: 30px;\n          border: 2px solid black;\n          color: grey;\n          position: absolute;\n          top: -10px;\n          right: -10px;\n          z-index: 2;\n        }\n\n        .close-btn:hover {\n          border: 2px solid black;\n          background-color: grey;\n          color: #ffffff;\n        }\n\n        .close-btn::before {\n          content: \" \";\n          position: absolute;\n          background-color: black;\n          width: 2px;\n          left: 12px;\n          top: 5px;\n          bottom: 5px;\n          transform: rotate(45deg);\n        }\n\n        .close-btn::after {\n          content: \" \";\n          position: absolute;\n          background-color: black;\n          height: 2px;\n          top: 12px;\n          left: 5px;\n          right: 5px;\n          transform: rotate(45deg);\n        }\n        simple-modal-template[modal-id=\"m1\"] {\n          --simple-modal-width: var(--check-it-out-modal-width, 80vw);\n          --simple-modal-height: var(--check-it-out-modal-height, 55vh);\n          --simple-modal-min-width: var(--check-it-out-modal-width, 80vw);\n          --simple-modal-min-height: var(--check-it-out-modal-height, 55vh);\n        }\n\n        .iframe-container {\n          width: var(--check-it-out-content-width, 800px);\n          height: var(--check-it-out-content-height, 500px);\n        }\n\n        .video-player {\n          width: var(--check-it-out-content-width, 800px);\n          height: var(--check-it-out-content-height, 500px);\n        }\n        iframe-loader {\n          text-align: center;\n          display: block;\n          width: var(--check-it-out-content-width, 800px);\n        }\n      "])))];
      }
      /**
       * Convention we use
       */

    }, {
      key: "tag",
      get: function get() {
        return "check-it-out";
      }
    }]);
    return CheckItOut;
  }((0, _IntersectionObserverMixin.IntersectionObserverMixin)(_index.LitElement));

  _exports.CheckItOut = CheckItOut;
  customElements.define(CheckItOut.tag, CheckItOut);
});
