// @lrnwebcomponents/elmsln-apps/lib/lrnapp-studio-root/lrnapp-studio-root.js
define(["exports", "../../../../@polymer/polymer/polymer-element.js", "../../../../@polymer/polymer/lib/utils/render-status.js", "../../../../@polymer/app-route/app-location.js", "../../../../@polymer/app-route/app-route.js", "../../../../@polymer/iron-icon/iron-icon.js", "../../../../@polymer/iron-list/iron-list.js", "../../../../@polymer/iron-ajax/iron-ajax.js", "../../../../@polymer/paper-card/paper-card.js", "../../../../@polymer/iron-selector/iron-selector.js", "../../../../@polymer/iron-pages/iron-pages.js", "../../../lrnsys-button/lrnsys-button.js", "../../../elmsln-loading/elmsln-loading.js", "../lrnapp-studio-dashboard/lrnapp-studio-dashboard.js", "../lrnapp-studio-kanban/lrnapp-studio-kanban.js", "../lrnapp-open-studio/lrnapp-open-studio.js"], function (_exports, _polymerElement, _renderStatus, _appLocation, _appRoute, _ironIcon, _ironList, _ironAjax, _paperCard, _ironSelector, _ironPages, _lrnsysButton, _elmslnLoading, _lrnappStudioDashboard, _lrnappStudioKanban, _lrnappOpenStudio) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappStudioRoot = void 0;

  function _templateObject_7ccc5700add011e983121daa910d1ce6() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: block;\n        }\n      </style>\n\n      <app-location\n        route=\"{{route}}\"\n        query-params=\"{{queryParams}}\"\n      ></app-location>\n      <app-route\n        route=\"{{route}}\"\n        pattern=\"[[basePath]]/:page\"\n        data=\"{{data}}\"\n        tail=\"{{tail}}\"\n        query-params=\"{{queryParams}}\"\n      >\n      </app-route>\n\n      <iron-selector\n        selected=\"{{data.page}}\"\n        attr-for-selected=\"name\"\n        role=\"navigation\"\n      >\n        <a\n          tabindex=\"-1\"\n          name=\"lrnapp-studio-dashboard\"\n          on-click=\"_dashboardClicked\"\n          >A</a\n        >\n        <a tabindex=\"-1\" name=\"lrnapp-studio-kanban\" on-click=\"_kanbanClicked\"\n          >B</a\n        >\n        <a tabindex=\"-1\" name=\"lrnapp-open-studio\" on-click=\"_openstudioClicked\"\n          >C</a\n        >\n      </iron-selector>\n\n      <iron-pages\n        selected=\"{{data.page}}\"\n        attr-for-selected=\"name\"\n        fallback-selection=\"lrnapp-studio-dashboard\"\n        role=\"main\"\n      >\n        <lrnapp-studio-dashboard\n          name=\"lrnapp-studio-dashboard\"\n          csrf-token=\"[[csrfToken]]\"\n          end-point=\"[[_endPoint('lrnapp-studio-dashboard')]]\"\n          base-path=\"[[basePath]]\"\n          elmsln-course=\"[[elmslnCourse]]\"\n          elmsln-section=\"[[elmslnSection]]\"\n          route=\"[[tail]]\"\n        >\n        </lrnapp-studio-dashboard>\n        <lrnapp-studio-kanban\n          name=\"lrnapp-studio-kanban\"\n          csrf-token=\"[[csrfToken]]\"\n          end-point=\"[[_endPoint('lrnapp-studio-dashboard')]]\"\n          base-path=\"[[basePath]]\"\n          source-path=\"[[_sourcePath('lrnapp-studio-kanban/kanban-data')]]\"\n          elmsln-course=\"[[elmslnCourse]]\"\n          elmsln-section=\"[[elmslnSection]]\"\n          route=\"[[tail]]\"\n        ></lrnapp-studio-kanban>\n        <lrnapp-open-studio\n          name=\"lrnapp-open-studio\"\n          csrf-token=\"[[csrfToken]]\"\n          end-point=\"[[_endPoint('lrnapp-open-studio')]]\"\n          base-path=\"[[basePath]]\"\n          source-path=\"[[_sourcePath('lrnapp-open-studio/data')]]\"\n          elmsln-course=\"[[elmslnCourse]]\"\n          elmsln-section=\"[[elmslnSection]]\"\n          route=\"[[tail]]\"\n        >\n        </lrnapp-open-studio>\n      </iron-pages>\n    "]);

    _templateObject_7ccc5700add011e983121daa910d1ce6 = function _templateObject_7ccc5700add011e983121daa910d1ce6() {
      return data;
    };

    return data;
  }

  var LrnappStudioRoot =
  /*#__PURE__*/
  function (_PolymerElement) {
    babelHelpers.inherits(LrnappStudioRoot, _PolymerElement);

    function LrnappStudioRoot() {
      babelHelpers.classCallCheck(this, LrnappStudioRoot);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappStudioRoot).apply(this, arguments));
    }

    babelHelpers.createClass(LrnappStudioRoot, [{
      key: "connectedCallback",
      value: function connectedCallback() {
        babelHelpers.get(babelHelpers.getPrototypeOf(LrnappStudioRoot.prototype), "connectedCallback", this).call(this);
        (0, _renderStatus.afterNextRender)(this, function () {
          this.addEventListener("route-change", this._routeChange.bind(this));
        });
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        this.removeEventListener("route-change", this._routeChange.bind(this));
        babelHelpers.get(babelHelpers.getPrototypeOf(LrnappStudioRoot.prototype), "disconnectedCallback", this).call(this);
      }
    }, {
      key: "_endPoint",
      value: function _endPoint(path) {
        return this.basePath + path;
      }
    }, {
      key: "_sourcePath",
      value: function _sourcePath(path) {
        return this.basePath + path + "?token=" + this.csrfToken;
      }
    }, {
      key: "_routeChanged",
      // If the current route is outside the scope of our app
      // then allow the website to break out of the single page
      // application routing
      value: function _routeChanged(route, basePath) {
        if (typeof route.path === "string") {
          if (typeof basePath === "string") {
            if (route.path.startsWith(basePath)) {
              return;
            }
          } // reload the page which since route changed will load that page


          window.location.reload();
        }
      }
      /**
       * Change route from deeper in the app.
       */

    }, {
      key: "_routeChange",
      value: function _routeChange(e) {
        var details = e.detail;

        if (babelHelpers.typeof(details.queryParams.assignment) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          this.set("queryParams.assignment", details.queryParams.assignment);
          this.notifyPath("queryParams.assignment");
        }

        if (babelHelpers.typeof(details.queryParams.project) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          this.set("queryParams.project", details.queryParams.project);
          this.notifyPath("queryParams.project");
        }

        if (babelHelpers.typeof(details.queryParams.author) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          this.set("queryParams.author", details.queryParams.author);
          this.notifyPath("queryParams.author");
        }

        if (babelHelpers.typeof(details.data.page) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          this.set("data.page", details.data.page);
          this.notifyPath("data.path");
        }
      }
      /**
       * Support having a toast message because of delete or error elsewhere.
       */

    }, {
      key: "_deleteToast",
      value: function _deleteToast(deletetoast, old) {
        if (babelHelpers.typeof(deletetoast) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          if (deletetoast == "error") {
            this.$.toast.show("That submission on longer exists!");
          } else {
            this.$.toast.show("Submission deleted successfully!");
          }

          this.set("queryParams.deletetoast", undefined);
          this.notifyPath("queryParams.deletetoast");
        }
      }
    }, {
      key: "_dashboardClicked",
      value: function _dashboardClicked(e) {
        this.set("route.path", this.basePath + "/lrnapp-studio-dashboard");
        this.notifyPath("route.path");
      }
    }, {
      key: "_kanbanClicked",
      value: function _kanbanClicked(e) {
        this.set("route.path", this.basePath + "/lrnapp-studio-kanban");
        this.notifyPath("route.path");
      }
    }, {
      key: "_openstudioClicked",
      value: function _openstudioClicked(e) {
        this.set("route.path", this.basePath + "/lrnapp-open-studio");
        this.notifyPath("route.path");
      }
      /**
       * Simple way to convert from object to array.
       */

    }, {
      key: "_toArray",
      value: function _toArray(obj) {
        if (obj == null) {
          return [];
        }

        return Object.keys(obj).map(function (key) {
          return obj[key];
        });
      }
    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_7ccc5700add011e983121daa910d1ce6());
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrnapp-studio-root";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * sourcePath for submission data.
           */
          endPoint: {
            type: String
          },
          sourcePath: {
            type: String,
            notify: true,
            reflectToAttribute: true
          },
          elmslnSection: {
            type: String,
            notify: true,
            reflectToAttribute: true
          },
          elmslnCourse: {
            type: String,
            notify: true,
            reflectToAttribute: true
          },
          csrfToken: {
            type: String,
            notify: true,
            reflectToAttribute: true
          },

          /**
           * base path for the app
           */
          basePath: {
            type: String,
            notify: true
          }
        };
      }
    }, {
      key: "observers",
      get: function get() {
        return ["_routeChanged(route, basePath)", "_deleteToast(queryParams.deletetoast)"];
      }
    }]);
    return LrnappStudioRoot;
  }(_polymerElement.PolymerElement);

  _exports.LrnappStudioRoot = LrnappStudioRoot;
  window.customElements.define(LrnappStudioRoot.tag, LrnappStudioRoot);
});
