// @lrnwebcomponents/responsive-grid/lib/responsive-grid-row.js
define(["exports", "../../../lit/index.js", "../../responsive-utility/responsive-utility.js", "./responsive-grid-col.js", "./responsive-grid-clear.js"], function (_exports, _index, _responsiveUtility, _responsiveGridCol, _responsiveGridClear) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ResponsiveGridRow = void 0;

  var _templateObject_6ba32760942811ecb1937969798b231a, _templateObject2_6ba32760942811ecb1937969798b231a;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**   
  `responsive-grid-row`
  A LRN element
  
  * @demo demo/index.html
  
  @microcopy - the mental model for this element
   Example:
    <responsive-grid-row 
      sm="600"                      //custom breakpoint for small screens
      md="900"                      //custom breakpoint for medium screens
      sm="1200"                     //custom breakpoint for large screens
      md="1800"                     //custom breakpoint for extra-large screens
      responsive-to-parent="true">  //make responsive to parent container instead of window/screen
      <responsive-grid-col xl="2" lg="3" md="4" sm="6" xs="12">Content</responsive-grid-col>
      <responsive-grid-col xl="2" lg="3" md="4" sm="6" xs="12">Content</responsive-grid-col>
      <responsive-grid-clear sm/>
      <responsive-grid-col xl="2" lg="3" md="4" sm="6" xs="12">Content</responsive-grid-col>
      <responsive-grid-col xl="2" lg="3" md="4" sm="6" xs="12">Content</responsive-grid-col>
    </responsive-grid-row>
  
  */
  var ResponsiveGridRow = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(ResponsiveGridRow, _LitElement);

    var _super = _createSuper(ResponsiveGridRow);

    /**
     * HTMLElement
     */
    function ResponsiveGridRow() {
      var _this;

      babelHelpers.classCallCheck(this, ResponsiveGridRow);
      _this = _super.call(this);
      _this.xl = null;
      _this.lg = null;
      _this.md = null;
      _this.sm = null;
      _this.xs = null;
      _this.gutter = 0;
      _this.responsiveToParent = false;
      _this.screen = "xs";
      window.ResponsiveUtility.requestAvailability();
      return _this;
    }
    /**
     * LitElement ready
     */


    babelHelpers.createClass(ResponsiveGridRow, [{
      key: "render",
      value:
      /**
       * LitElement render
       */
      function render() {
        return (0, _index.html)(_templateObject_6ba32760942811ecb1937969798b231a || (_templateObject_6ba32760942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n      <responsive-utility\n        xs=\"", "\"\n        sm=\"", "\"\n        md=\"", "\"\n        lg=\"", "\"\n        xl=\"", "\"\n        responsive-to-parent=\"", "\"\n      >\n      </responsive-utility>\n      <div id=\"row-inner\" screen=\"", "\" gutter=\"", "\">\n        <slot></slot>\n      </div>\n    "])), this.xs, this.sm, this.md, this.lg, this.xl, this.responsiveToParent, this.screen, this.gutter);
      }
      /**
       * convention
       */

    }, {
      key: "firstUpdated",
      value: function firstUpdated() {
        this.dispatchEvent(new CustomEvent("responsive-element", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: {
            element: this,
            attribute: "screen",
            relativeToParent: this.responsiveToParent
          }
        }));
      }
      /**
       * LitElement properties changed
       */

    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "screen") {
            // notify
            _this2.dispatchEvent(new CustomEvent("screen-changed", {
              detail: {
                value: _this2[propName]
              }
            }));
          }
        });
      }
    }], [{
      key: "styles",
      get:
      /**
       * LitElement constructable styles enhancement
       */
      function get() {
        return [(0, _index.css)(_templateObject2_6ba32760942811ecb1937969798b231a || (_templateObject2_6ba32760942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          width: 100%;\n          margin-right: auto;\n          margin-left: auto;\n          -webkit-box-sizing: border-box;\n          -moz-box-sizing: border-box;\n          box-sizing: border-box;\n        }\n        :host:before,\n        :host:after {\n          content: \" \";\n          display: table;\n          -webkit-box-sizing: border-box;\n          -moz-box-sizing: border-box;\n          box-sizing: border-box;\n        }\n        #row-inner {\n          margin-left: var(--responsive-grid-row-inner-margin-left, -15px);\n          margin-right: var(--responsive-grid-row-inner-margin-right, -15px);\n        }\n        :host #row-inner:before,\n        :host #row-inner:after {\n          content: \" \";\n          display: table;\n        }\n        :host #row-inner:after {\n          clear: both;\n        }\n        :host #row-inner:before,\n        :host #row-inner:after {\n          -webkit-box-sizing: border-box;\n          -moz-box-sizing: border-box;\n          box-sizing: border-box;\n        }\n        #row-inner[screen=\"xs\"] ::slotted(responsive-grid-clear[xs]),\n        #row-inner[screen=\"sm\"] ::slotted(responsive-grid-clear[sm]),\n        #row-inner[screen=\"md\"] ::slotted(responsive-grid-clear[md]),\n        #row-inner[screen=\"lg\"] ::slotted(responsive-grid-clear[lg]),\n        #row-inner[screen=\"xl\"] ::slotted(responsive-grid-clear[xl]) {\n          display: block;\n        }\n        #row-inner[screen=\"sm\"] ::slotted(responsive-grid-col[xs=\"0\"]),\n        #row-inner[screen=\"md\"] ::slotted(responsive-grid-col[sm=\"0\"]),\n        #row-inner[screen=\"lg\"] ::slotted(responsive-grid-col[md=\"0\"]),\n        #row-inner[screen=\"xl\"] ::slotted(responsive-grid-col[lg=\"0\"]) {\n          display: inline-block;\n        }\n        #row-inner[screen=\"xs\"] ::slotted(responsive-grid-col[xs=\"12\"]),\n        #row-inner[screen=\"sm\"] ::slotted(responsive-grid-col[sm=\"12\"]),\n        #row-inner[screen=\"md\"] ::slotted(responsive-grid-col[md=\"12\"]),\n        #row-inner[screen=\"lg\"] ::slotted(responsive-grid-col[lg=\"12\"]),\n        #row-inner[screen=\"xl\"] ::slotted(responsive-grid-col[xl=\"12\"]) {\n          width: 100%;\n        }\n        #row-inner[screen=\"xs\"] ::slotted(responsive-grid-col[xs=\"11\"]),\n        #row-inner[screen=\"sm\"] ::slotted(responsive-grid-col[sm=\"11\"]),\n        #row-inner[screen=\"md\"] ::slotted(responsive-grid-col[md=\"11\"]),\n        #row-inner[screen=\"lg\"] ::slotted(responsive-grid-col[lg=\"11\"]),\n        #row-inner[screen=\"xl\"] ::slotted(responsive-grid-col[xl=\"11\"]) {\n          width: 91.66666667%;\n        }\n        #row-inner[screen=\"xs\"] ::slotted(responsive-grid-col[xs=\"10\"]),\n        #row-inner[screen=\"sm\"] ::slotted(responsive-grid-col[sm=\"10\"]),\n        #row-inner[screen=\"md\"] ::slotted(responsive-grid-col[md=\"10\"]),\n        #row-inner[screen=\"lg\"] ::slotted(responsive-grid-col[lg=\"10\"]),\n        #row-inner[screen=\"xl\"] ::slotted(responsive-grid-col[xl=\"10\"]) {\n          width: 83.33333333%;\n        }\n        #row-inner[screen=\"xs\"] ::slotted(responsive-grid-col[xs=\"9\"]),\n        #row-inner[screen=\"sm\"] ::slotted(responsive-grid-col[sm=\"9\"]),\n        #row-inner[screen=\"md\"] ::slotted(responsive-grid-col[md=\"9\"]),\n        #row-inner[screen=\"lg\"] ::slotted(responsive-grid-col[lg=\"9\"]),\n        #row-inner[screen=\"xl\"] ::slotted(responsive-grid-col[xl=\"9\"]) {\n          width: 75%;\n        }\n        #row-inner[screen=\"xs\"] ::slotted(responsive-grid-col[xs=\"8\"]),\n        #row-inner[screen=\"sm\"] ::slotted(responsive-grid-col[sm=\"8\"]),\n        #row-inner[screen=\"md\"] ::slotted(responsive-grid-col[md=\"8\"]),\n        #row-inner[screen=\"lg\"] ::slotted(responsive-grid-col[lg=\"8\"]),\n        #row-inner[screen=\"xl\"] ::slotted(responsive-grid-col[xl=\"8\"]) {\n          width: 66.66666667%;\n        }\n        #row-inner[screen=\"xs\"] ::slotted(responsive-grid-col[xs=\"7\"]),\n        #row-inner[screen=\"sm\"] ::slotted(responsive-grid-col[sm=\"7\"]),\n        #row-inner[screen=\"md\"] ::slotted(responsive-grid-col[md=\"7\"]),\n        #row-inner[screen=\"lg\"] ::slotted(responsive-grid-col[lg=\"7\"]),\n        #row-inner[screen=\"xl\"] ::slotted(responsive-grid-col[xl=\"7\"]) {\n          width: 58.33333333%;\n        }\n        #row-inner[screen=\"xs\"] ::slotted(responsive-grid-col[xs=\"6\"]),\n        #row-inner[screen=\"sm\"] ::slotted(responsive-grid-col[sm=\"6\"]),\n        #row-inner[screen=\"md\"] ::slotted(responsive-grid-col[md=\"6\"]),\n        #row-inner[screen=\"lg\"] ::slotted(responsive-grid-col[lg=\"6\"]),\n        #row-inner[screen=\"xl\"] ::slotted(responsive-grid-col[xl=\"6\"]) {\n          width: 50%;\n        }\n        #row-inner[screen=\"xs\"] ::slotted(responsive-grid-col[xs=\"5\"]),\n        #row-inner[screen=\"sm\"] ::slotted(responsive-grid-col[sm=\"5\"]),\n        #row-inner[screen=\"md\"] ::slotted(responsive-grid-col[md=\"5\"]),\n        #row-inner[screen=\"lg\"] ::slotted(responsive-grid-col[lg=\"5\"]),\n        #row-inner[screen=\"xl\"] ::slotted(responsive-grid-col[xl=\"5\"]) {\n          width: 41.66666667%;\n        }\n        #row-inner[screen=\"xs\"] ::slotted(responsive-grid-col[xs=\"4\"]),\n        #row-inner[screen=\"sm\"] ::slotted(responsive-grid-col[sm=\"4\"]),\n        #row-inner[screen=\"md\"] ::slotted(responsive-grid-col[md=\"4\"]),\n        #row-inner[screen=\"lg\"] ::slotted(responsive-grid-col[lg=\"4\"]),\n        #row-inner[screen=\"xl\"] ::slotted(responsive-grid-col[xl=\"4\"]) {\n          width: 33.33333333%;\n        }\n        #row-inner[screen=\"xs\"] ::slotted(responsive-grid-col[xs=\"3\"]),\n        #row-inner[screen=\"sm\"] ::slotted(responsive-grid-col[sm=\"3\"]),\n        #row-inner[screen=\"md\"] ::slotted(responsive-grid-col[md=\"3\"]),\n        #row-inner[screen=\"lg\"] ::slotted(responsive-grid-col[lg=\"3\"]),\n        #row-inner[screen=\"xl\"] ::slotted(responsive-grid-col[xl=\"3\"]) {\n          width: 25%;\n        }\n        #row-inner[screen=\"xs\"] ::slotted(responsive-grid-col[xs=\"2\"]),\n        #row-inner[screen=\"sm\"] ::slotted(responsive-grid-col[sm=\"2\"]),\n        #row-inner[screen=\"md\"] ::slotted(responsive-grid-col[md=\"2\"]),\n        #row-inner[screen=\"lg\"] ::slotted(responsive-grid-col[lg=\"2\"]),\n        #row-inner[screen=\"xl\"] ::slotted(responsive-grid-col[xl=\"2\"]) {\n          width: 16.66666667%;\n        }\n        #row-inner[screen=\"xs\"] ::slotted(responsive-grid-col[xs=\"1\"]),\n        #row-inner[screen=\"sm\"] ::slotted(responsive-grid-col[sm=\"1\"]),\n        #row-inner[screen=\"md\"] ::slotted(responsive-grid-col[md=\"1\"]),\n        #row-inner[screen=\"lg\"] ::slotted(responsive-grid-col[lg=\"1\"]),\n        #row-inner[screen=\"xl\"] ::slotted(responsive-grid-col[xl=\"1\"]) {\n          width: 8.33333333%;\n        }\n        #row-inner[screen=\"xs\"] ::slotted(responsive-grid-col[xs=\"0\"]),\n        #row-inner[screen=\"sm\"] ::slotted(responsive-grid-col[sm=\"0\"]),\n        #row-inner[screen=\"md\"] ::slotted(responsive-grid-col[md=\"0\"]),\n        #row-inner[screen=\"lg\"] ::slotted(responsive-grid-col[lg=\"0\"]),\n        #row-inner[screen=\"xl\"] ::slotted(responsive-grid-col[xl=\"0\"]) {\n          display: none;\n        }\n        :host #row-inner ::slotted(* > #col-inner) {\n          padding: 0px;\n        }\n        :host #row-inner[gutter=\"1\"] ::slotted(* > #col-inner) {\n          padding: 5px;\n        }\n        :host #row-inner[gutter=\"2\"] ::slotted(* > #col-inner) {\n          padding: 10px;\n        }\n        :host #row-inner[gutter=\"3\"] ::slotted(* > #col-inner) {\n          padding: 15px;\n        }\n        :host #row-inner[gutter=\"4\"] ::slotted(* > #col-inner) {\n          padding: 20px;\n        }\n      "])))];
      }
    }, {
      key: "tag",
      get: function get() {
        return "responsive-grid-row";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Custom x-small breakpoint
           */
          xs: {
            type: Number
          },

          /**
           * Custom small breakpoint
           */
          sm: {
            type: Number
          },

          /**
           * Custom medium breakpoint
           */
          md: {
            type: Number
          },

          /**
           * Custom large breakpoint
           */
          lg: {
            type: Number
          },

          /**
           * Custom extra-large breakpoint
           */
          xl: {
            type: Number
          },

          /**
           * the gutter-level for the columns from 0-4
           */
          gutter: {
            type: Number
          },

          /**
           * make responsive based on a container instead of the window
           */
          responsiveToParent: {
            type: Boolean,
            reflect: true,
            attribute: "responsive-to-parent"
          },

          /*
           * screen size: xs, sm, md, lg, or xl
           */
          screen: {
            type: String,
            reflect: true
          }
        };
      }
    }]);
    return ResponsiveGridRow;
  }(_index.LitElement);

  _exports.ResponsiveGridRow = ResponsiveGridRow;
  window.customElements.define(ResponsiveGridRow.tag, ResponsiveGridRow);
});
