// @lrnwebcomponents/responsive-grid/lib/responsive-grid-clear.js
define(["exports", "../../../@polymer/polymer/polymer-element.js"], function (_exports, _polymerElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ResponsiveGridClear = void 0;

  function _templateObject_4bd5a870e3b511e9b0d5a5c01151ac03() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: none;\n          clear: both;\n        }\n      </style>\n    "]);

    _templateObject_4bd5a870e3b511e9b0d5a5c01151ac03 = function _templateObject_4bd5a870e3b511e9b0d5a5c01151ac03() {
      return data;
    };

    return data;
  }

  /**  
  `responsive-grid-clear`
  A clearumn for the responsive grid
  
  * @demo demo/index.html
  
  @microcopy - the mental model for this element
    Example: 
    <responsive-grid-clear 
      xs    //clears the floated columns on an extra small-width screen
      sm    //clears the floated columns on a small-width screen
      md    //clears the floated columns on a medium-width screen
      lg/>  //clears the floated columns on a large-width screen
  */
  var ResponsiveGridClear =
  /*#__PURE__*/
  function (_PolymerElement) {
    babelHelpers.inherits(ResponsiveGridClear, _PolymerElement);

    function ResponsiveGridClear() {
      babelHelpers.classCallCheck(this, ResponsiveGridClear);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(ResponsiveGridClear).apply(this, arguments));
    }

    babelHelpers.createClass(ResponsiveGridClear, null, [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_4bd5a870e3b511e9b0d5a5c01151ac03());
      }
    }, {
      key: "tag",
      get: function get() {
        return "responsive-grid-clear";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * clear the float and force a new worw on an extra-large screen?
           */
          xl: {
            type: Boolean,
            value: false
          },

          /**
           * clear the float and force a new worw on a large screen?
           */
          lg: {
            type: Boolean,
            value: false
          },

          /**
           * clear the float and force a new worw on a medium screen?
           */
          md: {
            type: Boolean,
            value: false
          },

          /**
           * clear the float and force a new worw on a small screen?
           */
          sm: {
            type: Boolean,
            value: false
          },

          /**
           * clear the float and force a new worw on an extra-small screen?
           */
          xs: {
            type: Boolean,
            value: false
          }
        };
      }
    }]);
    return ResponsiveGridClear;
  }(_polymerElement.PolymerElement);

  _exports.ResponsiveGridClear = ResponsiveGridClear;
  window.customElements.define(ResponsiveGridClear.tag, ResponsiveGridClear);
});
