// @lrnwebcomponents/outline-player/outline-player.js
define(["exports", "../../@polymer/polymer/polymer-element.js", "../../@polymer/polymer/lib/utils/async.js", "../../@polymer/polymer/lib/mixins/element-mixin.js", "../haxcms-elements/lib/core/HAXCMSThemeWiring.js", "../haxcms-elements/lib/core/haxcms-site-store.js", "../../mobx/lib/mobx.module.js", "../../@polymer/app-layout/app-header/app-header.js", "../../@polymer/app-layout/app-toolbar/app-toolbar.js", "../../@polymer/app-layout/app-drawer/app-drawer.js", "../../@polymer/app-layout/app-drawer-layout/app-drawer-layout.js", "../../@polymer/app-layout/app-header-layout/app-header-layout.js", "../simple-colors/simple-colors.js", "../hax-body/lib/hax-shared-styles.js", "../haxcms-elements/lib/ui-components/navigation/site-menu.js", "../haxcms-elements/lib/ui-components/navigation/site-menu-button.js", "../haxcms-elements/lib/ui-components/site/site-print-button.js", "../haxcms-elements/lib/ui-components/site/site-title.js", "../haxcms-elements/lib/ui-components/active-item/site-active-title.js"], function (_exports, _polymerElement, async, _elementMixin, _HAXCMSThemeWiring, _haxcmsSiteStore, _mobxModule, _appHeader, _appToolbar, _appDrawer, _appDrawerLayout, _appHeaderLayout, _simpleColors, _haxSharedStyles, _siteMenu, _siteMenuButton, _sitePrintButton, _siteTitle, _siteActiveTitle) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.OutlinePlayer = void 0;
  async = babelHelpers.interopRequireWildcard(async);

  function _templateObject_51755e80add011e983121daa910d1ce6() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style include=\"hax-shared-styles\">\n        :host {\n          display: block;\n          font-family: libre baskerville;\n          position: relative;\n          overflow: hidden;\n          --outline-player-min-height: 100vh;\n          --app-drawer-width: 300px;\n          --outline-player-dark: #222222;\n          --outline-player-light: #f8f8f8;\n          background-color: var(--outline-player-light);\n        }\n\n        :host([closed]) {\n          --app-drawer-width: 0px;\n        }\n\n        h1 {\n          font-size: 48px;\n          line-height: 16px;\n        }\n\n        h2 {\n          font-size: 32px;\n        }\n\n        h3 {\n          font-size: 28px;\n        }\n\n        p {\n          line-height: 26px;\n          min-height: 26px;\n        }\n\n        a,\n        a:visited,\n        a:active {\n          color: #000;\n        }\n\n        a:hover {\n          color: #2196f3;\n        }\n\n        ul li {\n          padding-bottom: 24px;\n          line-height: 1.5;\n          color: #424242;\n          max-width: 448px;\n        }\n\n        ul li:last-child {\n          padding-bottom: 16px;\n        }\n\n        app-drawer-layout {\n          min-height: 100vh;\n          min-height: -moz-available; /* WebKit-based browsers will ignore this. */\n          min-height: -webkit-fill-available; /* Mozilla-based browsers will ignore this. */\n          min-height: fill-available;\n          /* if the user has set a specific value then override the defaults */\n          min-height: var(--outline-player-min-height);\n        }\n        site-title {\n          --site-title-heading: {\n            font-size: 24px;\n            font-weight: normal;\n            line-height: 32px;\n            vertical-align: middle;\n            padding: 16px;\n            height: 32px;\n            margin: 0;\n            text-align: center;\n            text-overflow: ellipsis;\n            overflow: hidden;\n            word-break: break-word;\n            border-bottom: 1px solid #eeeeee;\n            position: sticky;\n          }\n        }\n\n        site-menu {\n          padding: 8px;\n        }\n\n        outline-player-navigation {\n          --outline-player-dark: var(--outline-player-dark);\n        }\n\n        div[main-title] {\n          margin-left: 8px;\n          font-size: 16px;\n          line-height: 22px;\n          overflow-wrap: break-word;\n          text-overflow: ellipsis;\n          display: inline-block;\n          word-break: break-word;\n        }\n\n        paper-progress {\n          display: block;\n          width: 100%;\n          --paper-progress-active-color: rgba(255, 255, 255, 0.5);\n          --paper-progress-container-color: transparent;\n        }\n\n        app-header {\n          color: var(--outline-player-dark);\n          /* Enable outline to be placed anywhere in the dom */\n          /* This will override the app-header-layout forcing fixed mode */\n          /*position: absolute !important;\n        left: 0 !important;*/\n          --app-header-background-rear-layer: {\n            /* app-header-layout will force fixed */\n            background-color: var(--outline-player-light);\n          }\n        }\n\n        app-toolbar {\n          border-bottom: none;\n          background-color: #ffffff;\n          box-shadow: 0 0 6px -3px var(--outline-player-dark);\n        }\n        app-drawer {\n          box-shadow: 0 0 6px -3px var(--outline-player-dark);\n          overflow: hidden;\n          --app-drawer-scrim-background: rgba(80, 80, 80, 0.8);\n          --app-drawer-content-container: {\n            overflow: hidden;\n            background-color: var(--outline-player-light);\n          }\n        }\n        app-drawer-layout[narrow] app-toolbar {\n          position: fixed !important;\n          left: 0;\n          right: 0;\n        }\n        app-drawer-layout[narrow] #contentcontainer {\n          padding-top: 64px;\n        }\n        #content {\n          justify-content: center;\n          padding: 8px 8px 8px 8px;\n        }\n\n        #content > * {\n          display: flex;\n          flex-direction: column;\n          align-items: center;\n        }\n\n        /* Required for HAX */\n        :host([edit-mode]) #slot {\n          display: none !important;\n        }\n        :host([edit-mode]) #content {\n          padding: 32px 8px 8px 8px;\n        }\n        #contentcontainer {\n          max-width: 840px;\n          margin: 0 auto;\n          padding: 0 16px 16px 16px;\n          flex: 1 1 auto;\n          order: 1;\n          display: flex;\n        }\n        #contentcontainer > * {\n          flex: 1 1 auto;\n          display: flex;\n          flex-direction: column;\n          width: 100%;\n        }\n        #contentcontainer h-a-x {\n          margin: 0;\n        }\n        site-menu {\n          height: calc(100vh - 64px);\n          color: #000000;\n          padding: 0;\n          background-color: #ffffff;\n          --site-menu-active-color: rgba(0, 0, 0, 0.1);\n          --site-menu-scrolltrack-bg-color: rgba(0, 0, 0, 0.3);\n          --site-menu-bg-shadow: rgba(0, 0, 0, 0.3);\n          --site-menu-bg-color: #fafafa;\n          --site-menu: {\n            padding: 0;\n            background-color: #ffffff;\n            color: #000000;\n          }\n          --site-menu-container: {\n            padding: 0;\n            background-color: #ffffff;\n            color: #000000;\n          }\n          --site-menu-item-active-item: {\n            color: #000000;\n          }\n        }\n        site-menu-button {\n          --site-menu-button-button: {\n            border-radius: 50%;\n            background-color: rgba(0, 0, 0, 0.1);\n            height: 40px;\n            width: 40px;\n          }\n          --site-menu-button-button-hover: {\n            background-color: rgba(0, 0, 0, 0.2);\n          }\n        }\n      </style>\n      <!-- Control the sites query paremeters -->\n\n      <!-- Begin Layout -->\n      <app-drawer-layout narrow=\"{{narrow}}\">\n        <app-drawer id=\"drawer\" swipe-open=\"\" slot=\"drawer\" opened=\"{{opened}}\">\n          <site-title></site-title>\n          <site-menu></site-menu>\n        </app-drawer>\n        <app-header-layout>\n          <app-header slot=\"header\" reveals>\n            <app-toolbar>\n              <paper-icon-button\n                icon=\"menu\"\n                on-click=\"_toggleMenu\"\n              ></paper-icon-button>\n              <div main-title>\n                <site-active-title></site-active-title>\n                <div id=\"slotTitle\"><slot name=\"title\"></slot></div>\n              </div>\n              <site-menu-button\n                type=\"prev\"\n                position=\"bottom\"\n                label=\"Prev\"\n                raised\n              ></site-menu-button>\n              <site-menu-button\n                type=\"next\"\n                position=\"bottom\"\n                label=\"Next\"\n                raised\n              ></site-menu-button>\n              <site-print-button></site-print-button>\n            </app-toolbar>\n          </app-header>\n          <div id=\"content\">\n            <div id=\"contentcontainer\">\n              <div id=\"slot\"><slot></slot></div>\n            </div>\n          </div>\n        </app-header-layout>\n      </app-drawer-layout>\n    "]);

    _templateObject_51755e80add011e983121daa910d1ce6 = function _templateObject_51755e80add011e983121daa910d1ce6() {
      return data;
    };

    return data;
  }

  /**
   * `outline-player`
   * `A basic outline presentation`
   *
   * @demo demo/index.html
   */
  var OutlinePlayer =
  /*#__PURE__*/
  function (_HAXCMSTheme) {
    babelHelpers.inherits(OutlinePlayer, _HAXCMSTheme);

    function OutlinePlayer() {
      babelHelpers.classCallCheck(this, OutlinePlayer);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(OutlinePlayer).apply(this, arguments));
    }

    babelHelpers.createClass(OutlinePlayer, [{
      key: "_toggleMenu",

      /**
       * Link menu button to open and closing the side panel.
       */
      value: function _toggleMenu(e) {
        this.$.drawer.toggle(); // allow styling to trigger based on open status

        this.closed = !this.$.drawer.opened; // kind of silly it doesn't just work this way but
        // app-panel doesn't make any assumptions about how
        // to handle the layout when it closes

        async.microTask.run(function () {
          // trick browser into thinking we just reized
          window.dispatchEvent(new Event("resize")); // forcibly update styles via css variables

          (0, _elementMixin.updateStyles)();
        });
      }
      /**
       * active id has changed.
       */

    }, {
      key: "_activeIdChanged",
      value: function _activeIdChanged(newValue) {
        // close menu if it's narrow and something new is picked
        if (this.opened && this.narrow) {
          this.$.drawer.toggle();
        }

        window.scrollTo({
          top: 0,
          left: 0,
          behavior: "smooth"
        });
      }
      /**
       * attached life cycle
       */

    }, {
      key: "connectedCallback",
      value: function connectedCallback() {
        var _this = this;

        babelHelpers.get(babelHelpers.getPrototypeOf(OutlinePlayer.prototype), "connectedCallback", this).call(this);
        this.__disposer = [];
        (0, _mobxModule.autorun)(function (reaction) {
          _this.activeId = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.activeId);

          _this.__disposer.push(reaction);
        });
      }
      /**
       * detatched life cycle
       */

    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        for (var i in this.__disposer) {
          this.__disposer[i].dispose();
        }

        babelHelpers.get(babelHelpers.getPrototypeOf(OutlinePlayer.prototype), "disconnectedCallback", this).call(this);
      }
    }], [{
      key: "tag",

      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */
      get: function get() {
        return "outline-player";
      } // render function

    }, {
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_51755e80add011e983121daa910d1ce6());
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          opened: {
            type: Boolean,
            reflectToAttribute: true
          },
          closed: {
            type: Boolean,
            notify: true,
            reflectToAttribute: true,
            value: false
          },
          activeId: {
            type: String,
            observer: "_activeIdChanged"
          },
          narrow: {
            type: Boolean,
            reflectToAttribute: true
          }
        };
      }
    }]);
    return OutlinePlayer;
  }((0, _HAXCMSThemeWiring.HAXCMSTheme)(_polymerElement.PolymerElement));

  _exports.OutlinePlayer = OutlinePlayer;
  window.customElements.define(OutlinePlayer.tag, OutlinePlayer);
});
