// @lrnwebcomponents/cms-hax/lib/cms-views.js
define(["exports", "require", "../../../lit-element/lit-element.js", "../../utils/utils.js", "../../../@polymer/iron-ajax/iron-ajax.js"], function (_exports, _require, _litElement, _utils, _ironAjax) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.CMSViews = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject4_b68410d03ee611ea9b2edd59a4f93ed0() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          min-width: 112px;\n          min-height: 112px;\n          transition: 0.6s all ease;\n          background-color: transparent;\n        }\n        #replacementcontent {\n          visibility: visible;\n          opacity: 1;\n        }\n        :host([loading]) {\n          text-align: center;\n        }\n        :host([loading]) #replacementcontent {\n          opacity: 0;\n          visibility: hidden;\n        }\n      "]);

    _templateObject4_b68410d03ee611ea9b2edd59a4f93ed0 = function _templateObject4_b68410d03ee611ea9b2edd59a4f93ed0() {
      return data;
    };

    return data;
  }

  function _templateObject3_b68410d03ee611ea9b2edd59a4f93ed0() {
    var data = babelHelpers.taggedTemplateLiteral([""]);

    _templateObject3_b68410d03ee611ea9b2edd59a4f93ed0 = function _templateObject3_b68410d03ee611ea9b2edd59a4f93ed0() {
      return data;
    };

    return data;
  }

  function _templateObject2_b68410d03ee611ea9b2edd59a4f93ed0() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <hexagon-loader\n              item-count=\"4\"\n              loading\n              size=\"small\"\n            ></hexagon-loader>\n          "]);

    _templateObject2_b68410d03ee611ea9b2edd59a4f93ed0 = function _templateObject2_b68410d03ee611ea9b2edd59a4f93ed0() {
      return data;
    };

    return data;
  }

  function _templateObject_b68410d03ee611ea9b2edd59a4f93ed0() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <iron-ajax\n        id=\"viewsrequest\"\n        method=\"GET\"\n        url=\"", "\"\n        handle-as=\"json\"\n        @last-response-changed=\"", "\"\n      ></iron-ajax>\n      ", "\n      <span id=\"replacementcontent\"><slot></slot></span>\n    "]);

    _templateObject_b68410d03ee611ea9b2edd59a4f93ed0 = function _templateObject_b68410d03ee611ea9b2edd59a4f93ed0() {
      return data;
    };

    return data;
  }

  /**
   * `cms-views`
   * @customElement cms-views
   * `Render and process a  / views from a content management system.`
   */
  var CMSViews =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(CMSViews, _LitElement);
    babelHelpers.createClass(CMSViews, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_b68410d03ee611ea9b2edd59a4f93ed0(), this.viewsEndPoint, this.viewsDataChanged, this.loading ? (0, _litElement.html)(_templateObject2_b68410d03ee611ea9b2edd59a4f93ed0()) : (0, _litElement.html)(_templateObject3_b68410d03ee611ea9b2edd59a4f93ed0()));
      }
    }, {
      key: "viewsDataChanged",
      value: function viewsDataChanged(e) {
        this.viewsData = e.detail.value;
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject4_b68410d03ee611ea9b2edd59a4f93ed0())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "cms-views";
      }
      /**
       * HTMLElement
       */

    }]);

    function CMSViews() {
      var _this;

      babelHelpers.classCallCheck(this, CMSViews);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(CMSViews).call(this));
      _this.loading = false;
      _this.viewsPrefix = "[";
      _this.viewsSuffix = "]";
      new Promise(function (res, rej) {
        return _require.default(["../../hexagon-loader/hexagon-loader.js"], res, rej);
      });
      return _this;
    }

    babelHelpers.createClass(CMSViews, [{
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (["viewsName", "viewsDisplay"].includes(propName)) {
            _this2.bodyData = _this2._generateBodyData(_this2.viewsName, _this2.viewsDisplay);
          }

          if (propName == "bodyData") {
            _this2._viewsChanged(_this2[propName]);
          }

          if (propName == "viewsData") {
            _this2._handleviewsResponse(_this2[propName]);
          }
        });
      }
    }, {
      key: "_generateBodyData",

      /**
       * Generate body data.
       */
      value: function _generateBodyData(name, display) {
        if (name !== null && name !== "") {
          return {
            name: "".concat(name),
            display: "".concat(display)
          };
        }
      }
      /**
       * Handle the response from the views processing endpoint
       */

    }, {
      key: "_handleviewsResponse",
      value: function _handleviewsResponse(newValue, oldValue) {
        if (newValue !== null && babelHelpers.typeof(newValue.content) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          // store the text and url callbacks
          if (document.getElementById("cmstokenidtolockonto") != null) {
            document.getElementById("cmstokenidtolockonto").setAttribute("href", newValue.editEndpoint);
            document.getElementById("cmstokenidtolockonto").innerHTML = newValue.editText;
          } // wipe our own slot here


          (0, _utils.wipeSlot)(this); // now inject the content we got

          var frag = document.createElement("span");
          frag.innerHTML = newValue.content;
          var newNode = frag.cloneNode(true);
          this.appendChild(newNode);
          this.loading = false;
        }
      }
      /**
       * views end point changed
       */

    }, {
      key: "_viewsChanged",
      value: function _viewsChanged(newValue) {
        // ensure we have something and are not loading currently
        if (babelHelpers.typeof(newValue) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined)) && newValue !== "" && !this.loading) {
          // support going from a null element to a real one
          if (babelHelpers.typeof(this.viewsEndPoint) === (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined)) && babelHelpers.typeof(window.cmsviewsEndPoint) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
            this.viewsEndPoint = window.cmsviewsEndPoint;
          }

          if (this.viewsEndPoint) {
            this.loading = true;
            this.shadowRoot.querySelector("#viewsrequest").body = newValue;
            this.shadowRoot.querySelector("#viewsrequest").generateRequest();
          }
        }
      }
    }, {
      key: "postProcessgetHaxJSONSchema",

      /**
       * Implements getHaxJSONSchema post processing callback.
       */
      value: function postProcessgetHaxJSONSchema(schema) {
        schema.properties["__editThis"] = {
          type: "string",
          component: {
            name: "a",
            properties: {
              id: "cmstokenidtolockonto",
              href: "",
              target: "_blank"
            },
            slot: "Edit this view"
          }
        };
        return schema;
      }
    }], [{
      key: "properties",
      get: function get() {
        return {
          /**
           * Loading state
           */
          loading: {
            type: Boolean,
            reflect: true
          },

          /**
           * Name of the views to render
           */
          viewsName: {
            type: String,
            reflect: true,
            attribute: "views-name"
          },

          /**
           * Display from the views
           */
          viewsDisplay: {
            type: String,
            reflect: true,
            attribute: "views-display"
          },

          /**
           * views end point updated, change the way we do processing.
           */
          viewsEndPoint: {
            type: String,
            attribute: "views-end-point"
          },

          /**
           * Prefix for the views to be processed
           */
          viewsPrefix: {
            type: String,
            attribute: "views-prefix"
          },

          /**
           * Suffix for the views to be processed
           */
          viewsSuffix: {
            type: String,
            attribute: "views-suffix"
          },

          /**
           * Body data which is just views with some encapsulation.
           */
          bodyData: {
            type: Object
          },

          /**
           * views data from the end point.
           */
          viewsData: {
            type: String,
            attribute: "views-data"
          }
        };
      }
    }, {
      key: "haxProperties",
      get: function get() {
        return {
          canScale: true,
          canPosition: true,
          canEditSource: false,
          gizmo: {
            title: "CMS View",
            description: "CMS views rendered on the backend",
            icon: "icons:view-module",
            color: "light-blue",
            groups: ["CMS"],
            handles: [{
              type: "cmsviews",
              views: "views"
            }],
            meta: {
              author: "ELMS:LN"
            }
          },
          settings: {
            quick: [],
            configure: [{
              property: "viewsName",
              title: "Name",
              description: "Name of the view from our CMS",
              inputMethod: "textfield",
              icon: "editor:title"
            }, {
              property: "viewsDisplay",
              title: "Display",
              description: "Display within that view from our CMS",
              inputMethod: "textfield",
              icon: "editor:title"
            }],
            advanced: []
          },
          saveOptions: {
            wipeSlot: true,
            unsetAttributes: ["loading", "views-data", "body-data", "views-end-point"]
          }
        };
      }
    }]);
    return CMSViews;
  }(_litElement.LitElement);

  _exports.CMSViews = CMSViews;
  window.customElements.define(CMSViews.tag, CMSViews);
});
