// @lrnwebcomponents/cms-hax/lib/cms-block.js
define(["exports", "require", "../../../lit-element/lit-element.js", "../../utils/utils.js", "../../../@polymer/iron-ajax/iron-ajax.js"], function (_exports, _require, _litElement, _utils, _ironAjax) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.CMSBlock = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject4_65d0bb602d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          min-width: 112px;\n          min-height: 112px;\n          transition: 0.6s all ease;\n          background-color: transparent;\n        }\n        #replacementcontent {\n          visibility: visible;\n          opacity: 1;\n        }\n        :host([loading]) {\n          text-align: center;\n        }\n        :host([loading]) #replacementcontent {\n          opacity: 0;\n          visibility: hidden;\n        }\n      "]);

    _templateObject4_65d0bb602d8511ea80ccf124101da415 = function _templateObject4_65d0bb602d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject3_65d0bb602d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral([""]);

    _templateObject3_65d0bb602d8511ea80ccf124101da415 = function _templateObject3_65d0bb602d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject2_65d0bb602d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <hexagon-loader\n              item-count=\"4\"\n              loading\n              size=\"small\"\n            ></hexagon-loader>\n          "]);

    _templateObject2_65d0bb602d8511ea80ccf124101da415 = function _templateObject2_65d0bb602d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject_65d0bb602d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <iron-ajax\n        id=\"blockrequest\"\n        method=\"GET\"\n        url=\"", "\"\n        handle-as=\"json\"\n        @last-response-changed=\"", "\"\n      ></iron-ajax>\n      ", "\n      <span id=\"replacementcontent\"><slot></slot></span>\n    "]);

    _templateObject_65d0bb602d8511ea80ccf124101da415 = function _templateObject_65d0bb602d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  /**
   * `cms-block`
   * `Render and process a  / block from a content management system.`
   * @customElement cms-block
   */
  var CMSBlock =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(CMSBlock, _LitElement);
    babelHelpers.createClass(CMSBlock, [{
      key: "render",

      /**
       * LitElement render
       */
      value: function render() {
        return (0, _litElement.html)(_templateObject_65d0bb602d8511ea80ccf124101da415(), this.blockEndPoint, this.blockDataChanged, this.loading ? (0, _litElement.html)(_templateObject2_65d0bb602d8511ea80ccf124101da415()) : (0, _litElement.html)(_templateObject3_65d0bb602d8511ea80ccf124101da415()));
      }
    }, {
      key: "blockDataChanged",
      value: function blockDataChanged(e) {
        this._handleblockResponse(e.detail.value);
      }
    }, {
      key: "_generateBodyData",

      /**
       * Generate body data.
       */
      value: function _generateBodyData(blockModule, blockDelta) {
        if (blockModule !== null && blockModule !== "" && blockDelta !== null && blockDelta !== "") {
          return {
            module: "".concat(blockModule),
            delta: "".concat(blockDelta)
          };
        }
      }
      /**
       * Handle the response from the block processing endpoint
       */

    }, {
      key: "_handleblockResponse",
      value: function _handleblockResponse(newValue) {
        if (newValue !== null && babelHelpers.typeof(newValue.content) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          // store the text and url callbacks
          if (document.getElementById("cmstokenidtolockonto") != null) {
            document.getElementById("cmstokenidtolockonto").setAttribute("href", newValue.editEndpoint);
            document.getElementById("cmstokenidtolockonto").innerHTML = newValue.editText;
          } // wipe our own slot here


          (0, _utils.wipeSlot)(this); // now inject the content we got

          var frag = document.createElement("span");
          frag.innerHTML = newValue.content;
          var newNode = frag.cloneNode(true);
          this.appendChild(newNode);
          this.loading = false;
        }
      }
      /**
       * block end point changed
       */

    }, {
      key: "_blockChanged",
      value: function _blockChanged(newValue) {
        // ensure we have something and are not loading currently
        if (babelHelpers.typeof(newValue) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined)) && newValue !== "" && !this.loading) {
          // support going from a null element to a real one
          if (babelHelpers.typeof(this.blockEndPoint) === (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined)) && babelHelpers.typeof(window.cmsblockEndPoint) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
            this.blockEndPoint = window.cmsblockEndPoint;
          }

          if (this.blockEndPoint) {
            this.loading = true;
            this.shadowRoot.querySelector("#blockrequest").body = newValue;
            this.shadowRoot.querySelector("#blockrequest").generateRequest();
          }
        }
      }
      /**
       * HTMLElement
       */

    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject4_65d0bb602d8511ea80ccf124101da415())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "cms-block";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Loading state
           */
          loading: {
            type: Boolean,
            reflect: true
          },

          /**
           * Module supplying the block
           */
          blockModule: {
            type: String,
            reflect: true,
            attribute: "block-module"
          },

          /**
           * A delta value relative to the module
           */
          blockDelta: {
            type: String,
            reflect: true,
            attribute: "block-delta"
          },

          /**
           * block end point updated, change the way we do processing.
           */
          blockEndPoint: {
            type: String,
            attribute: "block-end-point"
          },

          /**
           * Body data which is just block with some encapsulation.
           */
          bodyData: {
            type: Object
          },

          /**
           * Prefix for the block to be processed
           */
          blockPrefix: {
            type: String,
            attribute: "block-prefix"
          },

          /**
           * Suffix for the block to be processed
           */
          blockSuffix: {
            type: String,
            attribute: "block-suffix"
          }
        };
      }
    }]);

    function CMSBlock() {
      var _this;

      babelHelpers.classCallCheck(this, CMSBlock);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(CMSBlock).call(this));
      _this.loading = false;
      _this.blockPrefix = "[";
      _this.blockSuffix = "]";
      new Promise(function (res, rej) {
        return _require.default(["../../hexagon-loader/hexagon-loader.js"], res, rej);
      });
      return _this;
    }
    /**
     * LitElement properties changed
     */


    babelHelpers.createClass(CMSBlock, [{
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (["blockModule", "blockDelta"].includes(propName)) {
            _this2.bodyData = _this2._generateBodyData(_this2.blockModule, _this2.blockDelta);
          }

          if (propName == "bodyData") {
            _this2._blockChanged(_this2[propName]);
          }
        });
      }
    }, {
      key: "postProcessgetHaxJSONSchema",

      /**
       * Implements getHaxJSONSchema post processing callback.
       */
      value: function postProcessgetHaxJSONSchema(schema) {
        schema.properties["__editThis"] = {
          type: "string",
          component: {
            name: "a",
            properties: {
              id: "cmstokenidtolockonto",
              href: "",
              target: "_blank"
            },
            slot: "Edit this block"
          }
        };
        return schema;
      }
    }], [{
      key: "haxProperties",
      get: function get() {
        return {
          canScale: true,
          canPosition: true,
          canEditSource: false,
          gizmo: {
            title: "CMS Block",
            description: "CMS block rendered on the backend",
            icon: "image:crop-square",
            color: "light-blue",
            groups: ["CMS"],
            handles: [{
              type: "cmsblock",
              block: "block"
            }],
            meta: {
              author: "ELMS:LN"
            }
          },
          settings: {
            quick: [],
            configure: [{
              property: "blockModule",
              title: "Module",
              description: "Module to load from our CMS",
              inputMethod: "textfield",
              icon: "editor:title"
            }, {
              property: "blockDelta",
              title: "Delta",
              description: "Delta of the block to load from our CMS",
              inputMethod: "textfield",
              icon: "editor:title"
            }],
            advanced: []
          },
          saveOptions: {
            wipeSlot: true,
            unsetAttributes: ["loading", "block-data", "body-data", "block-end-point"]
          }
        };
      }
    }]);
    return CMSBlock;
  }(_litElement.LitElement);

  _exports.CMSBlock = CMSBlock;
  window.customElements.define(CMSBlock.tag, CMSBlock);
});
