// @lrnwebcomponents/lrndesign-animationctrl/lib/lrndesign-animationctrl-button.js
define(["exports", "../../../@polymer/polymer/polymer-element.js", "../../../@polymer/paper-button/paper-button.js", "../../../@polymer/iron-icon/iron-icon.js"], function (_exports, _polymerElement, _paperButton, _ironIcon) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrndesignAnimationctrlButton = void 0;

  function _templateObject_fe72f970dedc11e9bf5a076df0f16721() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: block;\n        }\n        paper-button {\n          @apply --animationctrl-button;\n        }\n        iron-icon {\n          display: inline-flex;\n        }\n        :host iron-icon[hidden] {\n          display: none;\n        }\n      </style>\n      <paper-button raised=\"\" id=\"[[name]]\" on-click=\"_tap\">\n        [[name]]\n        <iron-icon icon=\"[[icon]]\" hidden$=\"[[!icon]]\"></iron-icon>\n      </paper-button>\n    "]);

    _templateObject_fe72f970dedc11e9bf5a076df0f16721 = function _templateObject_fe72f970dedc11e9bf5a076df0f16721() {
      return data;
    };

    return data;
  }

  var LrndesignAnimationctrlButton =
  /*#__PURE__*/
  function (_PolymerElement) {
    babelHelpers.inherits(LrndesignAnimationctrlButton, _PolymerElement);

    function LrndesignAnimationctrlButton() {
      babelHelpers.classCallCheck(this, LrndesignAnimationctrlButton);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrndesignAnimationctrlButton).apply(this, arguments));
    }

    babelHelpers.createClass(LrndesignAnimationctrlButton, [{
      key: "_tap",

      /**
       * Fire event
       */
      value: function _tap(e) {
        e.preventDefault();
        this.dispatchEvent(new CustomEvent("lrndesign-animationctrl-click", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: {
            button: this.name
          }
        }));
      }
    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_fe72f970dedc11e9bf5a076df0f16721());
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrndesign-animationctrl-button";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Machine name of the button. This should be unique
           * to the animationctrl set
           */
          name: {
            type: String,
            value: "buttonid"
          },

          /**
           * Name of the Icon
           */
          icon: {
            type: String,
            value: false
          }
        };
      }
    }]);
    return LrndesignAnimationctrlButton;
  }(_polymerElement.PolymerElement);

  _exports.LrndesignAnimationctrlButton = LrndesignAnimationctrlButton;
  window.customElements.define(LrndesignAnimationctrlButton.tag, LrndesignAnimationctrlButton);
});
