// @lrnwebcomponents/scroll-button/scroll-button.js
define(["exports", "require", "../../lit-element/lit-element.js", "../../@polymer/polymer/lib/elements/custom-style.js"], function (_exports, _require, _litElement, _customStyle) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ScrollButton = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject2_da1688e01bb211eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          --scroll-button-z-index: 99;\n          z-index: var(--scroll-button-z-index);\n        }\n\n        :host([hidden]) {\n          display: none;\n        }\n        paper-icon-button {\n          background-color: rgba(0, 0, 0, 0.6);\n          color: white;\n        }\n        paper-icon-button:hover,\n        paper-icon-button:active,\n        paper-icon-button:focus {\n          background-color: rgba(0, 0, 0, 1);\n        }\n        paper-tooltip {\n          --paper-tooltip-background: #000000;\n          --paper-tooltip-opacity: 1;\n          --paper-tooltip-text-color: #ffffff;\n          --paper-tooltip-delay-in: 0;\n        }\n      "]);

    _templateObject2_da1688e01bb211eabd7cc30119585ddb = function _templateObject2_da1688e01bb211eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  function _templateObject_da1688e01bb211eabd7cc30119585ddb() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <custom-style>\n        <style>\n          paper-icon-button {\n            @apply --scroll-button-button;\n          }\n          paper-icon-button:hover,\n          paper-icon-button:active,\n          paper-icon-button:focus {\n            @apply --scroll-button-button-active;\n          }\n          paper-tooltip {\n            --paper-tooltip: {\n              border-radius: 0;\n            }\n            @apply --scroll-button-tooltip;\n          }\n        </style>\n      </custom-style>\n      <paper-icon-button\n        @click=\"", "\"\n        id=\"btn\"\n        icon=\"", "\"\n        title=\"", "\"\n      ></paper-icon-button>\n      <paper-tooltip for=\"btn\" position=\"", "\" offset=\"14\">\n        ", "\n      </paper-tooltip>\n    "]);

    _templateObject_da1688e01bb211eabd7cc30119585ddb = function _templateObject_da1688e01bb211eabd7cc30119585ddb() {
      return data;
    };

    return data;
  }

  /**
   * `scroll-button`
   * `button to scroll to an area or back to top`
   * @demo demo/index.html
   * @customElement scroll-button
   */
  var ScrollButton =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(ScrollButton, _LitElement);
    babelHelpers.createClass(ScrollButton, [{
      key: "render",
      // render function
      value: function render() {
        return (0, _litElement.html)(_templateObject_da1688e01bb211eabd7cc30119585ddb(), this.scrollEvent, this.icon, this.label, this.position, this.label);
      } // haxProperty definition

    }], [{
      key: "styles",
      //styles function
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_da1688e01bb211eabd7cc30119585ddb())];
      }
    }, {
      key: "haxProperties",
      get: function get() {
        return {
          canScale: true,
          canPosition: true,
          canEditSource: false,
          gizmo: {
            title: "Scroll button",
            description: "button to scroll to an area or back to top",
            icon: "icons:android",
            color: "green",
            groups: ["Button"],
            handles: [{
              type: "todo:read-the-docs-for-usage"
            }],
            meta: {
              author: "btopro",
              owner: "The Pennsylvania State University"
            }
          },
          settings: {
            quick: [],
            configure: [{
              property: "target",
              description: "",
              inputMethod: "array",
              required: false,
              icon: "icons:android"
            }, {
              property: "icon",
              description: "",
              inputMethod: "textfield",
              required: false,
              icon: "icons:android"
            }, {
              property: "label",
              description: "",
              inputMethod: "textfield",
              required: false,
              icon: "icons:android"
            }],
            advanced: []
          }
        };
      } // properties available to the custom element for data binding

    }, {
      key: "properties",
      get: function get() {
        return babelHelpers.objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(ScrollButton), "properties", this), {
          target: {
            type: Object
          },
          icon: {
            type: String
          },
          label: {
            type: String
          },
          position: {
            type: String
          }
        });
      }
    }]);

    function ScrollButton() {
      var _this;

      babelHelpers.classCallCheck(this, ScrollButton);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(ScrollButton).call(this));
      _this.icon = "icons:expand-less";
      _this.label = "Scroll to top";
      _this.position = "top";
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/paper-icon-button/paper-icon-button.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/paper-tooltip/paper-tooltip.js"], res, rej);
      });
      new Promise(function (res, rej) {
        return _require.default(["../../@polymer/iron-icons/iron-icons.js"], res, rej);
      });
      return _this;
    }
    /**
     * Store the tag name to make it easier to obtain directly.
     * @notice function name must be here for tooling to operate correctly
     */


    babelHelpers.createClass(ScrollButton, [{
      key: "scrollEvent",

      /**
       * life cycle, element is afixed to the DOM
       */
      value: function scrollEvent(e) {
        if (this.target) {
          this.target.scrollIntoView({
            behavior: "smooth",
            block: "end",
            inline: "nearest"
          });
        } else {
          window.scrollTo({
            top: 0,
            left: 0,
            behavior: "smooth"
          });
        }
      }
    }], [{
      key: "tag",
      get: function get() {
        return "scroll-button";
      }
    }]);
    return ScrollButton;
  }(_litElement.LitElement);

  _exports.ScrollButton = ScrollButton;
  window.customElements.define(ScrollButton.tag, ScrollButton);
});
