// @lrnwebcomponents/json-outline-schema/lib/jos-render.js
define(["exports", "../../../lit-element/lit-element.js", "../json-outline-schema.js", "../../utils/utils.js", "../../dynamic-import-registry/dynamic-import-registry.js"], function (_exports, _litElement, _jsonOutlineSchema, _utils, _dynamicImportRegistry) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.JosRender = void 0;

  class JosRender extends _litElement.LitElement {
    static get tag() {
      return "jos-render";
    }

    constructor() {
      super();
      this.registry = window.DynamicImportRegistry.requestAvailability();

      if (window.WCAutoloadBasePath) {
        this.registry.basePath = window.WCAutoloadBasePath;
      }

      this.items = [];
    }

    static get styles() {
      return [(0, _litElement.css)`
        :host {
          display: block;
        }
        .children {
          display: block;
          margin: 16px auto;
        }
        .children ::slotted(*) {
          display: inline-flex;
          margin: var(--jos-render-margin, 8px);
          padding: var(--jos-render-padding, 8px);
        }
      `];
    }

    static get properties() {
      return {
        source: {
          type: String
        },
        map: {
          type: Object
        },
        items: {
          type: Array
        }
      };
    }

    render() {
      return (0, _litElement.html)`
      <div class="children">
        <slot></slot>
      </div>
    `;
    }

    updated(changedProperties) {
      changedProperties.forEach(async (oldValue, propName) => {
        if (propName == "source") {
          let site = window.JSONOutlineSchema.requestAvailability(); // load source

          if (await site.load(this[propName])) {
            this.items = [...site.items];
          }
        }

        if (propName == "map" && this.map.path && this.map.tag) {
          // register
          this.registry.register({
            tag: this.map.tag,
            path: this.map.path
          }); // load the definition

          this.registry.loadDefinition(this.map.tag);

          if (this.items.length > 0) {
            this.renderItems(this.items);
          }
        }

        if (propName == "items") {
          this.renderItems(this.items);
        }
      });
    }
    /**
     * Render any item list passed in and handle it via our map
     * Separate function to support dynamic property remapping
     */


    renderItems(items) {
      // ensure we have a map to render
      if (this.map && this.map.properties) {
        // wipe slot
        (0, _utils.wipeSlot)(this);
        items.forEach(item => {
          // create tag for the map
          let n = document.createElement(this.map.tag);

          for (var key in this.map.properties) {
            let value = this.map.properties[key]; // complex transform capability for values that need processing
            // prior to being set

            if (value === true || value === false || value === null) {
              n[key] = value;
            } else if (value.transform && value.value && (0, _utils.varExists)(item, value.value)) {
              n[key] = value.transform((0, _utils.varGet)(item, value.value), item);
            } // only set the value in the node IF we have a match in the item for data
            // odd trap but the transform case can potentially miss above and this then pass
            // which varExists requires value be a string
            else if (typeof value === "string" && (0, _utils.varExists)(item, value)) {
                n[key] = (0, _utils.varGet)(item, value);
              } else {
                n[key] = value;
              }
          }

          this.appendChild(n);
        });
      }
    }

  }

  _exports.JosRender = JosRender;
  window.customElements.define(JosRender.tag, JosRender);
});
