// @lrnwebcomponents/extensible-toolbar/extensible-toolbar.js
define(["exports","../../lit-element/lit-element.js","../responsive-utility/responsive-utility.js"],(function(_exports,_litElement,_responsiveUtility){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.ExtensibleToolbar=void 0;
/**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class ExtensibleToolbar extends _litElement.LitElement{render(){return _litElement.html`
      <style>
        :host {
          display: flex;
          opacity: 1;
          z-index: 1;
          margin: 0;
          align-items: stretch;
          flex-wrap: wrap;
          justify-content: flex-start;
          transition: all 0.5s;
          --extensible-toolbar-visible-until-sm: unset;
          --extensible-toolbar-visible-until-md: unset;
          --extensible-toolbar-visible-until-lg: unset;
          --extensible-toolbar-visible-until-xl: unset;
          --extensible-toolbar-hidden-until-sm: none;
          --extensible-toolbar-hidden-until-md: none;
          --extensible-toolbar-hidden-until-lg: none;
          --extensible-toolbar-hidden-until-xl: none;
        }
        :host([hidden]) {
          display: none;
        }
        :host([aria-hidden]) {
          visibility: hidden;
          opacity: 0;
          height: 0;
        }
        :host([sticky]) {
          position: sticky;
          top: 0;
        }
        :host([collapsed]:not([responsive-size="xs"])) {
          --extensible-toolbar-visible-until-sm: none;
          --extensible-toolbar-hidden-until-sm: unset;
        }
        :host([collapsed]:not([responsive-size*="s"])) {
          --extensible-toolbar-visible-until-md: none;
          --extensible-toolbar-hidden-until-md: none;
        }
        :host([collapsed][responsive-size*="l"]) {
          --extensible-toolbar-visible-until-lg: none;
          --extensible-toolbar-hidden-until-lg: none;
        }
        :host([collapsed][responsive-size="xl"]) {
          --extensible-toolbar-visible-until-xl: none;
          --extensible-toolbar-hidden-until-xl: none;
        }
      </style>
      <slot></slot>
    `}static get properties(){return{collapsed:{attribute:"collapsed",reflect:!0,type:"Boolean"},responsiveSize:{attribute:"responsive-size",type:"String"},sticky:{attribute:"sticky",type:"Boolean",reflectToAttribute:!0}}}tag(){return"extensible-toolbar"}constructor(){super(),this.tag=ExtensibleToolbar.tag,this.collapsed=!1,this.responsiveSize="xs",this.gte="xs sm md lg xl",this.lte="xs",this.sticky=!1}initResponsive(){window.ResponsiveUtility.requestAvailability(),window.dispatchEvent(new CustomEvent("responsive-element",{detail:{element:this,attribute:"responsive-size",relativeToParent:!0}}))}connectedCallback(){super.connectedCallback(),initResponsive()}updated(changedProperties){changedProperties.forEach((oldValue,propName)=>this.dispatchEvent(new CustomEvent(`${propName.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g,"$1-$2").toLowerCase()}-changed`,{bubbles:!0,cancelable:!0,composed:!0,detail:this})))}}_exports.ExtensibleToolbar=ExtensibleToolbar,customElements.define("extensible-toolbar",ExtensibleToolbar)}));
