// @lrnwebcomponents/elmsln-apps/lib/lrnapp-canvas-listing/lrnapp-canvas-listing.js
define(["exports", "../../../../lit-element/lit-element.js", "../../../../@vaadin/vaadin-grid/vaadin-grid.js", "../../../../@vaadin/vaadin-grid/vaadin-grid-column-group.js", "../../../../@vaadin/vaadin-grid/vaadin-grid-filter.js", "../../../../@vaadin/vaadin-grid/vaadin-grid-sorter.js", "../../../../@vaadin/vaadin-grid/vaadin-grid-selection-column.js", "../../../../@polymer/iron-ajax/iron-ajax.js", "../../../../@polymer/iron-image/iron-image.js", "../../../../@polymer/paper-tooltip/paper-tooltip.js", "../../../lrnsys-layout/lib/lrnsys-dialog.js", "../../../elmsln-loading/elmsln-loading.js", "../../../lrndesign-avatar/lrndesign-avatar.js"], function (_exports, _litElement, _vaadinGrid, _vaadinGridColumnGroup, _vaadinGridFilter, _vaadinGridSorter, _vaadinGridSelectionColumn, _ironAjax, _ironImage, _paperTooltip, _lrnsysDialog, _elmslnLoading, _lrndesignAvatar) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnappCanvasListing = void 0;

  function _templateObject8_6cb346502d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          margin: 0 2em;\n        }\n        .loading {\n          width: 100%;\n          z-index: 1000;\n          opacity: 0.8;\n          text-align: center;\n          align-content: center;\n          justify-content: center;\n          height: 100vh;\n          position: absolute;\n          background-color: white;\n        }\n        vaadin-grid {\n          height: 75vh;\n          font-family: Roboto, sans-serif;\n        }\n\n        vaadin-grid .cell {\n          overflow: hidden;\n          text-overflow: ellipsis;\n          padding-right: 56px;\n        }\n\n        vaadin-grid .cell.last {\n          padding-right: 24px;\n        }\n\n        vaadin-grid .cell.numeric {\n          text-align: right;\n        }\n\n        vaadin-grid paper-checkbox {\n          --primary-color: var(--paper-indigo-500);\n          margin: 0 24px;\n        }\n\n        vaadin-grid vaadin-grid-sorter .cell {\n          flex: 1;\n          display: flex;\n          justify-content: space-between;\n          align-items: center;\n        }\n\n        vaadin-grid vaadin-grid-sorter iron-icon {\n          transform: scale(0.8);\n        }\n\n        vaadin-grid vaadin-grid-sorter:not([direction]) iron-icon {\n          color: rgba(0, 0, 0, var(--dark-disabled-opacity));\n        }\n\n        vaadin-grid vaadin-grid-sorter[direction] {\n          color: rgba(0, 0, 0, var(--dark-primary-opacity));\n        }\n\n        vaadin-grid vaadin-grid-sorter[direction=\"desc\"] iron-icon {\n          transform: scale(0.8) rotate(180deg);\n        }\n        vaadin-grid-sorter {\n          text-align: center;\n        }\n        lrndesign-avatar {\n          display: inline-flex;\n        }\n        lrnsys-dialog {\n          display: inline-flex;\n        }\n        lrnsys-dialog #dialog-trigger span {\n          pointer-events: none;\n        }\n        .avatar-name {\n          line-height: 2em;\n          margin: 0;\n          display: inline-block;\n        }\n        .listing-select {\n          display: block;\n          height: 100%;\n          margin: 0;\n          width: 100%;\n        }\n      "]);

    _templateObject8_6cb346502d8511ea80ccf124101da415 = function _templateObject8_6cb346502d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject7_6cb346502d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <div class=\"avatar-name\" id=\"user-", "\">\n                    <lrndesign-avatar\n                      label=\"", "\"\n                      src=\"", "\"\n                    ></lrndesign-avatar>\n                  </div>\n                  <paper-tooltip for=\"user-", "\"\n                    >", "</paper-tooltip\n                  >\n                "]);

    _templateObject7_6cb346502d8511ea80ccf124101da415 = function _templateObject7_6cb346502d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject6_6cb346502d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <h2>", "s</h2>\n              ", "\n            "]);

    _templateObject6_6cb346502d8511ea80ccf124101da415 = function _templateObject6_6cb346502d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject5_6cb346502d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                      <iron-image\n                        style=\"width:100%; height:200px; background-color: lightgray;\"\n                        sizing=\"cover\"\n                        preload\n                        fade\n                        src=\"", "\"\n                      ></iron-image>\n                    "]);

    _templateObject5_6cb346502d8511ea80ccf124101da415 = function _templateObject5_6cb346502d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject4_6cb346502d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                ", "\n                <span class=\"heading\">\n                  <span>Student count: ", "</span>\n                  <span>SIS ID: ", "</span>\n                  <span>Term: ", "</span>\n                  <span>Workflow: ", "</span>\n                </span>\n              "]);

    _templateObject4_6cb346502d8511ea80ccf124101da415 = function _templateObject4_6cb346502d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject3_6cb346502d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                <div id=\"loadingRoster\" class=\"loading\">\n                  <h3>Loading..</h3>\n                  <elmsln-loading\n                    color=\"grey-text\"\n                    size=\"large\"\n                  ></elmsln-loading>\n                </div>\n              "]);

    _templateObject3_6cb346502d8511ea80ccf124101da415 = function _templateObject3_6cb346502d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject2_6cb346502d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <option value=\"", "\">\n                    ", " (", ")\n                  </option>\n                "]);

    _templateObject2_6cb346502d8511ea80ccf124101da415 = function _templateObject2_6cb346502d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  function _templateObject_6cb346502d8511ea80ccf124101da415() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <custom-style>\n        <style include=\"materializecss-styles\">\n          vaadin-grid {\n            --divider-color: rgba(0, 0, 0, var(--dark-divider-opacity));\n\n            --vaadin-grid-cell: {\n              padding: 0;\n            }\n\n            --vaadin-grid-header-cell: {\n              height: 3.5em;\n              color: rgba(0, 0, 0, var(--dark-secondary-opacity));\n              font-size: 1em;\n            }\n\n            --vaadin-grid-body-cell: {\n              height: 3em;\n              color: rgba(0, 0, 0, var(--dark-primary-opacity));\n              font-size: 0.8em;\n            }\n\n            --vaadin-grid-body-row-hover-cell: {\n              background-color: var(--paper-grey-200);\n            }\n\n            --vaadin-grid-body-row-selected-cell: {\n              background-color: var(--paper-grey-100);\n            }\n\n            --vaadin-grid-focused-cell: {\n              box-shadow: none;\n              font-weight: bold;\n            }\n          }\n          vaadin-grid vaadin-grid-sorter {\n            --vaadin-grid-sorter-arrow: {\n              display: none !important;\n            }\n          }\n        </style>\n      </custom-style>\n      <iron-ajax\n        auto\n        url=\"", "\"\n        params='{\"return\": \"courses\"}'\n        handle-as=\"json\"\n        @response=\"", "\"\n        @last-response-changed=\"", "\"\n      ></iron-ajax>\n      <div id=\"loading\" class=\"loading\">\n        <h3>Loading..</h3>\n        <elmsln-loading color=\"grey-text\" size=\"large\"></elmsln-loading>\n      </div>\n      <vaadin-grid\n        column-reordering-allowed\n        id=\"material\"\n        aria-label=\"Course list\"\n        .items=\"", "\"\n      >\n        <vaadin-grid-column width=\"50px\" flex-grow=\"0\">\n          <template class=\"header\"\n            >#</template\n          >\n          <template\n            >[[index]]</template\n          >\n          <template class=\"footer\"\n            >#</template\n          >\n        </vaadin-grid-column>\n        <vaadin-grid-column width=\"200px\" flex-grow=\"0\" resizable>\n          <template class=\"header\">\n            <vaadin-grid-sorter path=\"term\">Semester</vaadin-grid-sorter>\n          </template>\n          <template>\n            [[item.term]]\n          </template>\n          <template class=\"footer\">\n            <vaadin-grid-filter\n              aria-label=\"Semester\"\n              path=\"term\"\n              value=\"", "\"\n            >\n              <paper-input\n                slot=\"filter\"\n                label=\"Semester\"\n                @value-changed=\"", "\"\n                focus-target\n              ></paper-input>\n            </vaadin-grid-filter>\n          </template>\n        </vaadin-grid-column>\n        <vaadin-grid-column resizable>\n          <template class=\"header\">\n            <vaadin-grid-sorter path=\"name\">Name</vaadin-grid-sorter>\n          </template>\n          <template\n            >[[item.name]]</template\n          >\n          <template class=\"footer\">\n            <vaadin-grid-filter\n              aria-label=\"Course\"\n              path=\"name\"\n              value=\"", "\"\n            >\n              <paper-input\n                slot=\"filter\"\n                label=\"Course\"\n                @value-changed=\"", "\"\n                focus-target\n              ></paper-input>\n            </vaadin-grid-filter>\n          </template>\n        </vaadin-grid-column>\n        <vaadin-grid-column resizable>\n          <template class=\"header\">\n            <vaadin-grid-sorter path=\"sis_course_id\">SIS</vaadin-grid-sorter>\n          </template>\n          <template>\n            [[item.sis_course_id]]\n          </template>\n          <template class=\"footer\">\n            <vaadin-grid-filter\n              aria-label=\"Student information system ID\"\n              path=\"sis_course_id\"\n              value=\"", "\"\n            >\n              <paper-input\n                slot=\"filter\"\n                label=\"SIS\"\n                @value-changed=\"", "\"\n                focus-target\n              ></paper-input>\n            </vaadin-grid-filter>\n          </template>\n        </vaadin-grid-column>\n        <vaadin-grid-column width=\"100px\" flex-grow=\"0\" resizable>\n          <template class=\"header\">\n            <vaadin-grid-sorter path=\"student_count\"\n              >Students</vaadin-grid-sorter\n            >\n          </template>\n          <template\n            >[[item.student_count]]</template\n          >\n        </vaadin-grid-column>\n        <vaadin-grid-column width=\"100px\" flex-grow=\"0\" resizable>\n          <template class=\"header\">\n            <vaadin-grid-sorter path=\"workflow_state\">State</vaadin-grid-sorter>\n          </template>\n          <template\n            >[[item.workflow_state]]</template\n          >\n          <template class=\"footer\">\n            <vaadin-grid-filter\n              aria-label=\"Workflow state\"\n              path=\"workflow_state\"\n              value=\"", "\"\n            >\n              <paper-input\n                slot=\"filter\"\n                label=\"State\"\n                value=\"", "\"\n                focus-target\n              ></paper-input>\n            </vaadin-grid-filter>\n          </template>\n        </vaadin-grid-column>\n        <vaadin-grid-column>\n          <template class=\"header\"\n            >ELMSLN Course</template\n          >\n          <template>\n            <select\n              name=\"elmsln--map--:key:[[item.sis_course_id]]:key:[[item.term]]:key:[[item.start]]:key:[[item.end]]\"\n              class=\"listing-select\"\n              value=\"{{item.elmslnCourse::input}}\"\n            >\n              ", "\n            </select>\n          </template>\n          <template class=\"footer\">\n            <vaadin-grid-filter\n              aria-label=\"Course\"\n              path=\"elmslnCourse\"\n              value=\"", "\"\n            >\n              <paper-input\n                slot=\"filter\"\n                label=\"Course\"\n                @value-changed=\"", "\"\n                focus-target\n              ></paper-input>\n            </vaadin-grid-filter>\n          </template>\n        </vaadin-grid-column>\n        <vaadin-grid-column width=\"100px\" flex-grow=\"0\">\n          <template class=\"header\"></template>\n          <template>\n            <paper-button\n              raised\n              @click=\"", "\"\n              id=\"{{item.sis_course_id}}\"\n              >Details</paper-button\n            >\n          </template>\n          <template class=\"footer\"></template>\n        </vaadin-grid-column>\n      </vaadin-grid>\n      <iron-ajax\n        id=\"request\"\n        url=\"", "\"\n        params='{\"return\": \"users\"}'\n        handle-as=\"json\"\n        @response=\"", "\"\n        @last-response-changed=\"", "\"\n      ></iron-ajax>\n      <lrnsys-dialog\n        tabindex=\"-1\"\n        id=\"details-dialog\"\n        body-append\n        header=\"", "\"\n      >\n        <div slot=\"content\">\n          ", "\n        </div>\n        <div slot=\"header\">\n          ", "\n        </div>\n        <div id=\"loadingContent\" slot=\"content\">\n          ", "\n        </div>\n      </lrnsys-dialog>\n    "]);

    _templateObject_6cb346502d8511ea80ccf124101da415 = function _templateObject_6cb346502d8511ea80ccf124101da415() {
      return data;
    };

    return data;
  }

  var LrnappCanvasListing =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnappCanvasListing, _LitElement);
    babelHelpers.createClass(LrnappCanvasListing, [{
      key: "render",
      value: function render() {
        var _this2 = this;

        return (0, _litElement.html)(_templateObject_6cb346502d8511ea80ccf124101da415(), this.sourcePath, this.handleResponse, this.queryResponseEvent, this._toArray(this.canvasCourses), this._filterTerm, this._filterTermEvent, this._filterCourse, this._filterCourseEvent, this._filterSIS, this._filterSISChanged, this._filterWorkflow, this._filterWorkflowChanged, this.elmslnCourses.map(function (elmsCourse) {
          return (0, _litElement.html)(_templateObject2_6cb346502d8511ea80ccf124101da415(), elmsCourse.machineName, elmsCourse.name, elmsCourse.machineName);
        }), this._filterELMSLNCourse, this.filterELMSLNCourseEvent, this._triggerDialog, this.sourcePath, this.handleRosterResponse, this.queryResponseRosterEvent, this.activeCourse.name, !this.roster ? (0, _litElement.html)(_templateObject3_6cb346502d8511ea80ccf124101da415()) : "", this.roster ? (0, _litElement.html)(_templateObject4_6cb346502d8511ea80ccf124101da415(), this.activeCourse.image ? (0, _litElement.html)(_templateObject5_6cb346502d8511ea80ccf124101da415(), this.activeCourse.image) : "", this.activeCourse.student_count, this.activeCourse.sis_course_id, this.activeCourse.term, this.activeCourse.workflow_state) : "", this._toArray(this.roster).map(function (roleList) {
          return (0, _litElement.html)(_templateObject6_6cb346502d8511ea80ccf124101da415(), roleList.role, _this2._toArray(roleList.users).map(function (user) {
            return (0, _litElement.html)(_templateObject7_6cb346502d8511ea80ccf124101da415(), user.id, user.name, user.picture, user.id, user.name);
          }));
        }));
      }
    }, {
      key: "queryResponseEvent",
      value: function queryResponseEvent(e) {
        this.queryResponse = _objectSpread({}, e.detail.value);
      }
    }, {
      key: "queryResponseRosterEvent",
      value: function queryResponseRosterEvent(e) {
        this.queryResponseRoster = _objectSpread({}, e.detail.value);
      }
    }, {
      key: "_filterWorkflowChanged",
      value: function _filterWorkflowChanged(e) {
        this._filterWorkflow = e.detail.value;
      }
    }, {
      key: "_filterCourseEvent",
      value: function _filterCourseEvent(e) {
        this._filterCourse = e.detail.value;
      }
    }, {
      key: "filterELMSLNCourseEvent",
      value: function filterELMSLNCourseEvent(e) {
        this._filterELMSLNCourse = e.detail.value;
      }
    }, {
      key: "_filterTermEvent",
      value: function _filterTermEvent(e) {
        this._filterTerm = e.detail.value;
      }
    }, {
      key: "_filterSISChanged",
      value: function _filterSISChanged(e) {
        this._filterSIS = e.detail.value;
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject8_6cb346502d8511ea80ccf124101da415())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrnapp-canvas-listing";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          _filterWorkflow: {
            type: String
          },
          _filterSIS: {
            type: String
          },
          _filterCourse: {
            type: String
          },
          _filterTerm: {
            type: String
          },
          _filterELMSLNCourse: {
            type: String
          },
          elmslnCourse: {
            type: String,
            attribute: "elmsln-course"
          },
          elmslnSection: {
            type: String,
            attribute: "elmsln-section"
          },
          basePath: {
            type: String,
            attribute: "base-path"
          },
          csrfToken: {
            type: String,
            attribute: "csrf-token"
          },
          endPoint: {
            type: String,
            attribute: "end-point"
          },
          canvasCourses: {
            type: Array
          },
          roster: {
            type: Array
          },
          queryResponse: {
            type: Object
          },
          queryResponseRoster: {
            type: Object
          },
          sourcePath: {
            type: String,
            attribute: "source-path"
          },
          activeCourse: {
            type: String,
            reflect: true,
            attribute: "active-course"
          }
        };
      }
    }]);

    function LrnappCanvasListing() {
      var _this;

      babelHelpers.classCallCheck(this, LrnappCanvasListing);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnappCanvasListing).call(this));
      _this.roster = [];
      _this.elmslnCourses = [];
      _this.canvasCourses = [];
      return _this;
    }

    babelHelpers.createClass(LrnappCanvasListing, [{
      key: "updated",
      value: function updated(changedProperties) {
        var _this3 = this;

        changedProperties.forEach(function (oldValue, propName) {
          var notifiedProps = ["activeCourse", "sourcePath", "queryResponse", "roster", "canvasCourses", "elmslnCourses"];

          if (notifiedProps.includes(propName)) {
            // notify
            var eventName = "".concat(propName.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, "$1-$2").toLowerCase(), "-changed");

            _this3.dispatchEvent(new CustomEvent(eventName, {
              detail: {
                value: _this3[propName]
              }
            }));
          }
        });
      }
      /**
       * Simple way to convert from object to array.
       */

    }, {
      key: "_toArray",
      value: function _toArray(obj) {
        if (obj == null) {
          return [];
        }

        return Object.keys(obj).map(function (key) {
          return obj[key];
        });
      }
      /**
       * Toggling collapse on an iron element.
       */

    }, {
      key: "collapseToggle",
      value: function collapseToggle(e) {
        e.target.nextElementSibling.toggle();
      }
      /**
       * Trigger the dialog box to opened and kick off request for data.
       */

    }, {
      key: "_triggerDialog",
      value: function _triggerDialog(e) {
        this.querySelector("#details-dialog").toggleDialog();
        this.roster = false;
        this.activeCourse = this.canvasCourses[e.target.id];
        this.querySelector("#request").params["sis_course_id"] = this.activeCourse.sis_course_id;
        this.querySelector("#request").generateRequest();
        this.querySelector("#loadingContent").style.display = "none";
      }
    }, {
      key: "handleResponse",
      value: function handleResponse(e) {
        this.elmslnCourses = this.queryResponse.data.elmslnCourses;
        this.canvasCourses = this.queryResponse.data.canvasCourses;
        this.querySelector("#loading").hidden = true;
      }
    }, {
      key: "handleRosterResponse",
      value: function handleRosterResponse(e) {
        this.roster = this.queryResponseRoster.data;
        this.querySelector("#loadingContent").style.display = "block";
      }
    }, {
      key: "createRenderRoot",
      value: function createRenderRoot() {
        return this;
      }
    }]);
    return LrnappCanvasListing;
  }(_litElement.LitElement);

  _exports.LrnappCanvasListing = LrnappCanvasListing;
  window.customElements.define(LrnappCanvasListing.tag, LrnappCanvasListing);
});
