// @lrnwebcomponents/paper-avatar/paper-avatar.js
define(["exports","meta","../../lit-element/lit-element.js","../es-global-bridge/es-global-bridge.js","./lib/md5.min.js"],(function(e,t,r,i,a){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.PaperAvatar=void 0,t=babelHelpers.interopRequireWildcard(t),a=babelHelpers.interopRequireWildcard(a);class PaperAvatar extends r.LitElement{static get styles(){return[r.css`
        :host {
          --paper-avatar-width: 40px;
          display: inline-block;
          box-sizing: border-box;
          position: relative;
          width: var(--paper-avatar-width);
          height: var(--paper-avatar-width);
          border-radius: 50%;
          cursor: default;
          background-color: var(
            --paper-avatar-color,
            var(--paper-avatar-bgcolor)
          );
          -webkit-user-select: none;
          -moz-user-select: none;
          -ms-user-select: none;
          user-select: none;
        }

        :host > * {
          pointer-events: none;
        }
        img {
          width: var(--paper-avatar-width);
          height: var(--paper-avatar-width);
        }
        #label,
        #img,
        #jdenticon {
          position: absolute;
          top: 0;
          left: 0;
          right: 0;
          bottom: 0;
          width: 100%;
          height: 100%;
          border-radius: 50%;
        }
        #label {
          overflow: hidden;
          display: -ms-flexbox;
          display: -webkit-flex;
          display: flex;
          -webkit-flex-direction: row;
          -ms-flex-direction: row;
          flex-direction: row;
          -webkit-align-items: center;
          -ms-flex-align: center;
          align-items: center;
        }
        #label span {
          display: block;
          width: 100%;
          font-weight: 400;
          color: var(--paper-avatar-text-color, rgba(255, 255, 255, 0.8));
          text-transform: capitalize;
          font-family: "Roboto", "Noto", sans-serif;
          -webkit-font-smoothing: antialiased;
          text-align: center;
          font-size: calc(var(--paper-avatar-width) / 1.65);
        }
        #jdenticon {
          width: var(--paper-avatar-width);
          height: var(--paper-avatar-width);
        }
        #jdenticon * {
          fill: var(
            --paper-avatar-text-color,
            rgba(255, 255, 255, 0.8)
          ) !important;
        }
      `]}static get tag(){return"paper-avatar"}render(){return r.html`
      <div id="label" title="${this.label}">
        <span>${this._label(this.label)}</span>
      </div>
      <svg id="jdenticon" width="40" height="40"><slot></slot></svg>
      ${this.src?r.html`
            <img
              id="img"
              loading="lazy"
              src="${this.src}"
              title="${this.label}"
              @load="${this._onImgLoad}"
              @error="${this._onImgError}"
              alt=""
            />
          `:""}
    `}constructor(){super(),this.label=null,this.src=null,this.jdenticonExists=!1,this.twoChars=!1,this.jdenticon=!1}updated(e){e.forEach((e,t)=>{"label"==t&&this._observerLabel(this[t])})}static get properties(){return{src:{type:String},label:{type:String},jdenticonExists:{type:Boolean},twoChars:{type:Boolean,attribute:"two-chars"},colors:{type:Array},jdenticon:{type:Boolean}}}_observerLabel(e){e&&(this.jdenticonExists&&this.jdenticon&&(this.shadowRoot.querySelector("#label").hidden=!0,window.jdenticon.config={lightness:{color:[1,1],grayscale:[1,1]},saturation:1},window.jdenticon.update(this.shadowRoot.querySelector("#jdenticon"),window.md5(e))),this.style.backgroundColor=this._parseColor(e))}pathFromUrl(e){return e.substring(0,e.lastIndexOf("/")+1)}firstUpdated(e){const r=`${this.pathFromUrl(decodeURIComponent(t.url))}lib/jdenticon-1.4.0.min.js`;window.addEventListener("es-bridge-jdenticon-loaded",this._jdenticonLoaded.bind(this)),window.ESGlobalBridge.requestAvailability(),window.ESGlobalBridge.instance.load("jdenticon",r)}disconnectedCallback(){window.removeEventListener("es-bridge-jdenticon-loaded",this._jdenticonLoaded.bind(this)),super.disconnectedCallback()}_jdenticonLoaded(e){this.jdenticonExists=!0,this._observerLabel(this.label)}_label(e){if(!e)return"";if(this.twoChars){if(this.label.indexOf(" ")>-1){var t=this.label.match(/\b(\w)/g);return t[0]+t[1]}return e.substring(0,2)}return e.charAt(0)}_onImgLoad(e){e.currentTarget.hidden=!1}_onImgError(e){e.currentTarget.hidden=!0}_parseColor(e){for(var t=this.colors?this.colors:["#F44336","#E91E63","#9C27B0","#673AB7","#3F51B5","#2196F3","#03A9F4","#00BCD4","#795548","#009688","#4CAF50","#8BC34A","#CDDC39","#FFEB3B","#FFC107","#FF9800","#FF5722","#9E9E9E","#607D8B"],r=0,i=0;i<e.length;i++)r+=e.charCodeAt(i)<<5;return r>=t.length?t[r%t.length]:t[r]}}e.PaperAvatar=PaperAvatar,window.customElements.define(PaperAvatar.tag,PaperAvatar)}));
