// @lit/reactive-element/css-tag.js
define(["exports"], function (_exports) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.unsafeCSS = _exports.supportsAdoptingStyleSheets = _exports.getCompatibleStyle = _exports.css = _exports.adoptStyles = _exports.CSSResult = void 0;

  function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

  function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

  function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

  /**
   * @license
   * Copyright 2019 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var t = window.ShadowRoot && (void 0 === window.ShadyCSS || window.ShadyCSS.nativeShadow) && "adoptedStyleSheets" in Document.prototype && "replace" in CSSStyleSheet.prototype,
      e = Symbol();
  _exports.supportsAdoptingStyleSheets = t;

  var s = /*#__PURE__*/function () {
    function s(t, _s) {
      babelHelpers.classCallCheck(this, s);
      if (_s !== e) throw Error("CSSResult is not constructable. Use `unsafeCSS` or `css` instead.");
      this.cssText = t;
    }

    babelHelpers.createClass(s, [{
      key: "styleSheet",
      get: function get() {
        return t && void 0 === this.t && (this.t = new CSSStyleSheet(), this.t.replaceSync(this.cssText)), this.t;
      }
    }, {
      key: "toString",
      value: function toString() {
        return this.cssText;
      }
    }]);
    return s;
  }();

  _exports.CSSResult = s;

  var n = new Map(),
      o = function o(t) {
    var o = n.get(t);
    return void 0 === o && n.set(t, o = new s(t, e)), o;
  },
      r = function r(t) {
    return o("string" == typeof t ? t : t + "");
  },
      i = function i(t) {
    for (var _len = arguments.length, e = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      e[_key - 1] = arguments[_key];
    }

    var n = 1 === t.length ? t[0] : e.reduce(function (e, n, o) {
      return e + function (t) {
        if (babelHelpers.instanceof(t, s)) return t.cssText;
        if ("number" == typeof t) return t;
        throw Error("Value passed to 'css' function must be a 'css' function result: " + t + ". Use 'unsafeCSS' to pass non-literal values, but take care to ensure page security.");
      }(n) + t[o + 1];
    }, t[0]);
    return o(n);
  },
      S = function S(e, s) {
    t ? e.adoptedStyleSheets = s.map(function (t) {
      return babelHelpers.instanceof(t, CSSStyleSheet) ? t : t.styleSheet;
    }) : s.forEach(function (t) {
      var s = document.createElement("style");
      s.textContent = t.cssText, e.appendChild(s);
    });
  },
      u = t ? function (t) {
    return t;
  } : function (t) {
    return babelHelpers.instanceof(t, CSSStyleSheet) ? function (t) {
      var e = "";

      var _iterator = _createForOfIteratorHelper(t.cssRules),
          _step;

      try {
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          var _s2 = _step.value;
          e += _s2.cssText;
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }

      return r(e);
    }(t) : t;
  };

  _exports.getCompatibleStyle = u;
  _exports.adoptStyles = S;
  _exports.css = i;
  _exports.unsafeCSS = r;
});
