// @lit/reactive-element/reactive-element.js
define(["exports", "./css-tag.js"], function (_exports, _cssTag) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  Object.defineProperty(_exports, "CSSResult", {
    enumerable: true,
    get: function get() {
      return _cssTag.CSSResult;
    }
  });
  _exports.ReactiveElement = void 0;
  Object.defineProperty(_exports, "adoptStyles", {
    enumerable: true,
    get: function get() {
      return _cssTag.adoptStyles;
    }
  });
  Object.defineProperty(_exports, "css", {
    enumerable: true,
    get: function get() {
      return _cssTag.css;
    }
  });
  _exports.defaultConverter = void 0;
  Object.defineProperty(_exports, "getCompatibleStyle", {
    enumerable: true,
    get: function get() {
      return _cssTag.getCompatibleStyle;
    }
  });
  _exports.notEqual = void 0;
  Object.defineProperty(_exports, "supportsAdoptingStyleSheets", {
    enumerable: true,
    get: function get() {
      return _cssTag.supportsAdoptingStyleSheets;
    }
  });
  Object.defineProperty(_exports, "unsafeCSS", {
    enumerable: true,
    get: function get() {
      return _cssTag.unsafeCSS;
    }
  });

  function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e3) { throw _e3; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e4) { didErr = true; err = _e4; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

  function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

  function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var s;

  var e = window.reactiveElementPolyfillSupport,
      r = {
    toAttribute: function toAttribute(t, i) {
      switch (i) {
        case Boolean:
          t = t ? "" : null;
          break;

        case Object:
        case Array:
          t = null == t ? t : JSON.stringify(t);
      }

      return t;
    },
    fromAttribute: function fromAttribute(t, i) {
      var s = t;

      switch (i) {
        case Boolean:
          s = null !== t;
          break;

        case Number:
          s = null === t ? null : Number(t);
          break;

        case Object:
        case Array:
          try {
            s = JSON.parse(t);
          } catch (t) {
            s = null;
          }

      }

      return s;
    }
  },
      h = function h(t, i) {
    return i !== t && (i == i || t == t);
  },
      o = {
    attribute: !0,
    type: String,
    converter: r,
    reflect: !1,
    hasChanged: h
  };

  _exports.notEqual = h;
  _exports.defaultConverter = r;

  var n = /*#__PURE__*/function (_HTMLElement) {
    babelHelpers.inherits(n, _HTMLElement);

    var _super = _createSuper(n);

    function n() {
      var _this;

      babelHelpers.classCallCheck(this, n);
      _this = _super.call(this), _this._$Et = new Map(), _this.isUpdatePending = !1, _this.hasUpdated = !1, _this._$Ei = null, _this.o();
      return _this;
    }

    babelHelpers.createClass(n, [{
      key: "o",
      value: function o() {
        var _this2 = this;

        var t;
        this._$Ev = new Promise(function (t) {
          return _this2.enableUpdating = t;
        }), this._$AL = new Map(), this._$Ep(), this.requestUpdate(), null === (t = this.constructor.l) || void 0 === t || t.forEach(function (t) {
          return t(_this2);
        });
      }
    }, {
      key: "addController",
      value: function addController(t) {
        var i, s;
        (null !== (i = this._$Em) && void 0 !== i ? i : this._$Em = []).push(t), void 0 !== this.renderRoot && this.isConnected && (null === (s = t.hostConnected) || void 0 === s || s.call(t));
      }
    }, {
      key: "removeController",
      value: function removeController(t) {
        var i;
        null === (i = this._$Em) || void 0 === i || i.splice(this._$Em.indexOf(t) >>> 0, 1);
      }
    }, {
      key: "_$Ep",
      value: function _$Ep() {
        var _this3 = this;

        this.constructor.elementProperties.forEach(function (t, i) {
          _this3.hasOwnProperty(i) && (_this3._$Et.set(i, _this3[i]), delete _this3[i]);
        });
      }
    }, {
      key: "createRenderRoot",
      value: function createRenderRoot() {
        var t;
        var s = null !== (t = this.shadowRoot) && void 0 !== t ? t : this.attachShadow(this.constructor.shadowRootOptions);
        return (0, _cssTag.adoptStyles)(s, this.constructor.elementStyles), s;
      }
    }, {
      key: "connectedCallback",
      value: function connectedCallback() {
        var t;
        void 0 === this.renderRoot && (this.renderRoot = this.createRenderRoot()), this.enableUpdating(!0), null === (t = this._$Em) || void 0 === t || t.forEach(function (t) {
          var i;
          return null === (i = t.hostConnected) || void 0 === i ? void 0 : i.call(t);
        });
      }
    }, {
      key: "enableUpdating",
      value: function enableUpdating(t) {}
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        var t;
        null === (t = this._$Em) || void 0 === t || t.forEach(function (t) {
          var i;
          return null === (i = t.hostDisconnected) || void 0 === i ? void 0 : i.call(t);
        });
      }
    }, {
      key: "attributeChangedCallback",
      value: function attributeChangedCallback(t, i, s) {
        this._$AK(t, s);
      }
    }, {
      key: "_$Eg",
      value: function _$Eg(t, i) {
        var s = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : o;
        var e, h;

        var _n = this.constructor._$Eh(t, s);

        if (void 0 !== _n && !0 === s.reflect) {
          var _o = (null !== (h = null === (e = s.converter) || void 0 === e ? void 0 : e.toAttribute) && void 0 !== h ? h : r.toAttribute)(i, s.type);

          this._$Ei = t, null == _o ? this.removeAttribute(_n) : this.setAttribute(_n, _o), this._$Ei = null;
        }
      }
    }, {
      key: "_$AK",
      value: function _$AK(t, i) {
        var s, e, h;

        var o = this.constructor,
            _n2 = o._$Eu.get(t);

        if (void 0 !== _n2 && this._$Ei !== _n2) {
          var _t = o.getPropertyOptions(_n2),
              l = _t.converter,
              a = null !== (h = null !== (e = null === (s = l) || void 0 === s ? void 0 : s.fromAttribute) && void 0 !== e ? e : "function" == typeof l ? l : null) && void 0 !== h ? h : r.fromAttribute;

          this._$Ei = _n2, this[_n2] = a(i, _t.type), this._$Ei = null;
        }
      }
    }, {
      key: "requestUpdate",
      value: function requestUpdate(t, i, s) {
        var e = !0;
        void 0 !== t && (((s = s || this.constructor.getPropertyOptions(t)).hasChanged || h)(this[t], i) ? (this._$AL.has(t) || this._$AL.set(t, i), !0 === s.reflect && this._$Ei !== t && (void 0 === this._$ES && (this._$ES = new Map()), this._$ES.set(t, s))) : e = !1), !this.isUpdatePending && e && (this._$Ev = this._$EC());
      }
    }, {
      key: "_$EC",
      value: function () {
        var _$EC2 = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
          var t;
          return regeneratorRuntime.wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  this.isUpdatePending = !0;
                  _context.prev = 1;
                  _context.next = 4;
                  return this._$Ev;

                case 4:
                  _context.next = 9;
                  break;

                case 6:
                  _context.prev = 6;
                  _context.t0 = _context["catch"](1);
                  Promise.reject(_context.t0);

                case 9:
                  t = this.scheduleUpdate();
                  _context.t1 = null != t;

                  if (!_context.t1) {
                    _context.next = 14;
                    break;
                  }

                  _context.next = 14;
                  return t;

                case 14:
                  return _context.abrupt("return", !this.isUpdatePending);

                case 15:
                case "end":
                  return _context.stop();
              }
            }
          }, _callee, this, [[1, 6]]);
        }));

        function _$EC() {
          return _$EC2.apply(this, arguments);
        }

        return _$EC;
      }()
    }, {
      key: "scheduleUpdate",
      value: function scheduleUpdate() {
        return this.performUpdate();
      }
    }, {
      key: "performUpdate",
      value: function performUpdate() {
        var _this4 = this;

        var t;
        if (!this.isUpdatePending) return;
        this.hasUpdated, this._$Et && (this._$Et.forEach(function (t, i) {
          return _this4[i] = t;
        }), this._$Et = void 0);
        var i = !1;
        var s = this._$AL;

        try {
          i = this.shouldUpdate(s), i ? (this.willUpdate(s), null === (t = this._$Em) || void 0 === t || t.forEach(function (t) {
            var i;
            return null === (i = t.hostUpdate) || void 0 === i ? void 0 : i.call(t);
          }), this.update(s)) : this._$EU();
        } catch (t) {
          throw i = !1, this._$EU(), t;
        }

        i && this._$AE(s);
      }
    }, {
      key: "willUpdate",
      value: function willUpdate(t) {}
    }, {
      key: "_$AE",
      value: function _$AE(t) {
        var i;
        null === (i = this._$Em) || void 0 === i || i.forEach(function (t) {
          var i;
          return null === (i = t.hostUpdated) || void 0 === i ? void 0 : i.call(t);
        }), this.hasUpdated || (this.hasUpdated = !0, this.firstUpdated(t)), this.updated(t);
      }
    }, {
      key: "_$EU",
      value: function _$EU() {
        this._$AL = new Map(), this.isUpdatePending = !1;
      }
    }, {
      key: "updateComplete",
      get: function get() {
        return this.getUpdateComplete();
      }
    }, {
      key: "getUpdateComplete",
      value: function getUpdateComplete() {
        return this._$Ev;
      }
    }, {
      key: "shouldUpdate",
      value: function shouldUpdate(t) {
        return !0;
      }
    }, {
      key: "update",
      value: function update(t) {
        var _this5 = this;

        void 0 !== this._$ES && (this._$ES.forEach(function (t, i) {
          return _this5._$Eg(i, _this5[i], t);
        }), this._$ES = void 0), this._$EU();
      }
    }, {
      key: "updated",
      value: function updated(t) {}
    }, {
      key: "firstUpdated",
      value: function firstUpdated(t) {}
    }], [{
      key: "addInitializer",
      value: function addInitializer(t) {
        var i;
        null !== (i = this.l) && void 0 !== i || (this.l = []), this.l.push(t);
      }
    }, {
      key: "observedAttributes",
      get: function get() {
        var _this6 = this;

        this.finalize();
        var t = [];
        return this.elementProperties.forEach(function (i, s) {
          var e = _this6._$Eh(s, i);

          void 0 !== e && (_this6._$Eu.set(e, s), t.push(e));
        }), t;
      }
    }, {
      key: "createProperty",
      value: function createProperty(t) {
        var i = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : o;

        if (i.state && (i.attribute = !1), this.finalize(), this.elementProperties.set(t, i), !i.noAccessor && !this.prototype.hasOwnProperty(t)) {
          var _s = "symbol" == babelHelpers.typeof(t) ? Symbol() : "__" + t,
              _e = this.getPropertyDescriptor(t, _s, i);

          void 0 !== _e && Object.defineProperty(this.prototype, t, _e);
        }
      }
    }, {
      key: "getPropertyDescriptor",
      value: function getPropertyDescriptor(t, i, s) {
        return {
          get: function get() {
            return this[i];
          },
          set: function set(e) {
            var r = this[t];
            this[i] = e, this.requestUpdate(t, r, s);
          },
          configurable: !0,
          enumerable: !0
        };
      }
    }, {
      key: "getPropertyOptions",
      value: function getPropertyOptions(t) {
        return this.elementProperties.get(t) || o;
      }
    }, {
      key: "finalize",
      value: function finalize() {
        if (this.hasOwnProperty("finalized")) return !1;
        this.finalized = !0;
        var t = Object.getPrototypeOf(this);

        if (t.finalize(), this.elementProperties = new Map(t.elementProperties), this._$Eu = new Map(), this.hasOwnProperty("properties")) {
          var _t2 = this.properties,
              _i = [].concat(babelHelpers.toConsumableArray(Object.getOwnPropertyNames(_t2)), babelHelpers.toConsumableArray(Object.getOwnPropertySymbols(_t2)));

          var _iterator = _createForOfIteratorHelper(_i),
              _step;

          try {
            for (_iterator.s(); !(_step = _iterator.n()).done;) {
              var _s2 = _step.value;
              this.createProperty(_s2, _t2[_s2]);
            }
          } catch (err) {
            _iterator.e(err);
          } finally {
            _iterator.f();
          }
        }

        return this.elementStyles = this.finalizeStyles(this.styles), !0;
      }
    }, {
      key: "finalizeStyles",
      value: function finalizeStyles(i) {
        var s = [];

        if (Array.isArray(i)) {
          var _e2 = new Set(i.flat(1 / 0).reverse());

          var _iterator2 = _createForOfIteratorHelper(_e2),
              _step2;

          try {
            for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
              var _i2 = _step2.value;
              s.unshift((0, _cssTag.getCompatibleStyle)(_i2));
            }
          } catch (err) {
            _iterator2.e(err);
          } finally {
            _iterator2.f();
          }
        } else void 0 !== i && s.push((0, _cssTag.getCompatibleStyle)(i));

        return s;
      }
    }, {
      key: "_$Eh",
      value: function _$Eh(t, i) {
        var s = i.attribute;
        return !1 === s ? void 0 : "string" == typeof s ? s : "string" == typeof t ? t.toLowerCase() : void 0;
      }
    }]);
    return n;
  }( /*#__PURE__*/babelHelpers.wrapNativeSuper(HTMLElement));

  _exports.ReactiveElement = n;
  n.finalized = !0, n.elementProperties = new Map(), n.elementStyles = [], n.shadowRootOptions = {
    mode: "open"
  }, null == e || e({
    ReactiveElement: n
  }), (null !== (s = globalThis.reactiveElementVersions) && void 0 !== s ? s : globalThis.reactiveElementVersions = []).push("1.0.1");
});
