// @lrnwebcomponents/lrndesign-contentblock/lrndesign-contentblock.js
define(["exports", "../../@polymer/polymer/polymer-element.js"], function (_exports, _polymerElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrndesignContentblock = void 0;

  /**
   * `lrndesign-contentblock`
   * @element lrndesign-contentblock
   * `An incredibly basic content block`
   * @demo demo/index.html
   */
  class LrndesignContentblock extends _polymerElement.PolymerElement {
    static get template() {
      return (0, _polymerElement.html)`
      <style>
        :host {
          display: inline-block;
          position: relative;
          box-sizing: border-box;
        }
      </style>
      <h3>[[title]]</h3>
      <slot></slot>
    `;
    }

    static get tag() {
      return "lrndesign-contentblock";
    }

    static get properties() {
      return {
        /**
         * Heading for this block
         */
        title: {
          type: String
        }
      };
    }

  }

  _exports.LrndesignContentblock = LrndesignContentblock;
  window.customElements.define(LrndesignContentblock.tag, LrndesignContentblock);
});
