// @lrnwebcomponents/product-card/lib/course-card.js
define(["exports", "../../../lit-element/lit-element.js", "../../simple-colors/simple-colors.js", "../../../@polymer/iron-icon/iron-icon.js"], function (_exports, _litElement, _simpleColors, _ironIcon) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.CourseCard = void 0;

  function _templateObject2_73098880939b11eaa49d972a432e908f() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n\n        a {\n          text-decoration: none;\n          color: var(--simple-colors-default-theme-accent-12, #363533);\n          background-color: var(\n            --simple-colors-default-theme-accent-1,\n            #ffffff\n          );\n          display: block;\n          width: 100%;\n          min-height: 300px;\n          border: solid 1px var(--simple-colors-default-theme-accent-2, #dcdcdc);\n        }\n\n        a:active,\n        a:hover,\n        a:focus {\n          outline: 1px solid\n            var(--simple-colors-default-theme-accent-7, #e2801e);\n          box-shadow: 4px 4px 10px\n            var(--simple-colors-default-theme-accent-2, #dcdcdc);\n        }\n\n        #card_wrap {\n          display: flex;\n          flex-direction: column;\n          align-items: center;\n        }\n\n        #course_number {\n          font-size: 28px;\n          text-transform: uppercase;\n          text-align: center;\n          line-height: 1.4;\n        }\n\n        #course_name {\n          font-size: 18px;\n          text-align: center;\n          width: 90%;\n          margin: 0 0 15px 0;\n          line-height: 1.2;\n        }\n\n        #course_icon {\n          background-color: var(--simple-colors-default-theme-accent-1, #fff);\n          border-radius: 50%;\n          position: relative;\n          bottom: 50px;\n          border: solid;\n          border-color: var(--simple-colors-default-theme-accent-9, #363533);\n          border-width: 5px;\n          margin: 0 0 -40px 0;\n        }\n\n        iron-icon {\n          width: 70px;\n          height: 70px;\n          fill: var(--simple-colors-default-theme-accent-9, #363533);\n        }\n\n        #course_image {\n          background-repeat: no-repeat;\n          background-size: cover;\n          background-position: right center;\n          width: 100%;\n          height: 150px;\n        }\n        #course_author {\n          margin-top: -10px;\n        }\n      "]);

    _templateObject2_73098880939b11eaa49d972a432e908f = function _templateObject2_73098880939b11eaa49d972a432e908f() {
      return data;
    };

    return data;
  }

  function _templateObject_73098880939b11eaa49d972a432e908f() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <a href=\"", "\">\n        <div id=\"card_wrap\">\n          <div\n            id=\"course_image\"\n            style=\"background-image:url(", ")\"\n            alt=\"", "\"\n          ></div>\n          <div id=\"course_icon\">\n            <iron-icon icon=\"", "\"></iron-icon>\n          </div>\n          <div id=\"course_number\">", "</div>\n          <div id=\"course_name\">", "</div>\n          <div id=\"course_author\">By: ", "</div>\n        </div>\n      </a>\n    "]);

    _templateObject_73098880939b11eaa49d972a432e908f = function _templateObject_73098880939b11eaa49d972a432e908f() {
      return data;
    };

    return data;
  }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  var CourseCard =
  /*#__PURE__*/
  function (_SimpleColors) {
    babelHelpers.inherits(CourseCard, _SimpleColors);
    babelHelpers.createClass(CourseCard, null, [{
      key: "properties",
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(CourseCard), "properties", this), {
          url: {
            type: String
          },
          image: {
            type: String
          },
          alt: {
            type: String
          },
          number: {
            type: String
          },
          name: {
            type: String
          },
          icon: {
            type: String
          },
          author: {
            type: String
          }
        });
      }
    }]);

    function CourseCard() {
      var _this;

      babelHelpers.classCallCheck(this, CourseCard);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(CourseCard).call(this));
      _this.url = "";
      _this.image = "";
      _this.alt = "";
      _this.number = "";
      _this.name = "";
      _this.icon = "";
      _this.author = "";
      return _this;
    }

    babelHelpers.createClass(CourseCard, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_73098880939b11eaa49d972a432e908f(), this.url, this.image, this.alt, this.icon, this.number, this.name, this.author);
      }
    }], [{
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(CourseCard), "styles", this)), [(0, _litElement.css)(_templateObject2_73098880939b11eaa49d972a432e908f())]);
      }
    }, {
      key: "tag",
      get: function get() {
        return "course-card";
      }
    }]);
    return CourseCard;
  }(_simpleColors.SimpleColors);

  _exports.CourseCard = CourseCard;
  window.customElements.define(CourseCard.tag, CourseCard);
});
