// @lrnwebcomponents/product-card/lib/course-card.js
define(["exports", "../../../lit-element/lit-element.js", "../../simple-colors/simple-colors.js", "../../../@polymer/iron-icon/iron-icon.js"], function (_exports, _litElement, _simpleColors, _ironIcon) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.CourseCard = void 0;

  function _templateObject3_9006ac70e0cf11eab599756c9ebe7c2f() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n\n        a {\n          filter: var(--course-card-filter);\n          transition: 0.2s all ease-in-out;\n          text-decoration: none;\n          color: var(--simple-colors-default-theme-accent-12, #000);\n          background-color: var(--simple-colors-default-theme-grey-1, #ffffff);\n          display: block;\n          width: 100%;\n          min-height: 300px;\n          border: solid 2px var(--simple-colors-default-theme-grey-1, #ffffff);\n        }\n\n        a:active,\n        a:hover,\n        a:focus {\n          filter: var(--course-card-filter-hover);\n          outline: 2px solid\n            var(--simple-colors-default-theme-accent-7, #666666);\n          box-shadow: 2px 2px 10px\n            var(--simple-colors-default-theme-accent-2, #eeeeee);\n          border: solid 2px var(--simple-colors-default-theme-accent-7, #666666);\n        }\n\n        #card_wrap {\n          display: flex;\n          flex-direction: column;\n          align-items: center;\n        }\n\n        #course_number {\n          font-size: 28px;\n          word-break: break-all;\n          text-transform: uppercase;\n          text-align: center;\n          width: 90%;\n          line-height: 1.4;\n        }\n\n        #course_name {\n          font-size: 18px;\n          text-align: center;\n          width: 90%;\n          margin: 0 0 15px 0;\n          line-height: 1.2;\n        }\n        :host([size=\"small\"]) #course_number {\n          font-size: 18px;\n        }\n        :host([size=\"small\"]) #course_name {\n          font-size: 14px;\n        }\n        :host([size=\"small\"]) #course_author {\n          font-size: 12px;\n        }\n        #course_icon {\n          border-radius: 50%;\n          position: relative;\n          bottom: 50px;\n          border: solid;\n          background-color: var(\n            --simple-colors-default-theme-accent-7,\n            #666666\n          );\n          border-color: var(--simple-colors-default-theme-grey-1, #ffffff);\n          border-width: 5px;\n          margin: 0 0 -40px 0;\n        }\n\n        iron-icon {\n          width: 70px;\n          height: 70px;\n          fill: var(--simple-colors-default-theme-grey-1, #ffffff);\n        }\n\n        #course_image {\n          background-repeat: no-repeat;\n          background-size: cover;\n          background-position: center center;\n          background-color: var(\n            --simple-colors-default-theme-accent-11,\n            #363533\n          );\n          width: 100%;\n          height: 150px;\n          animation-name: mymove2;\n          animation-duration: 0.6s;\n          animation-iteration-count: 1;\n          animation-fill-mode: forwards;\n        }\n\n        @keyframes mymove {\n          0% {\n            background-size: 100%;\n          }\n          100% {\n            background-size: 150%;\n          }\n        }\n        @keyframes mymove2 {\n          0% {\n            background-size: 150%;\n          }\n          100% {\n            background-size: 100%;\n          }\n        }\n        @media (prefers-reduced-motion: reduce) {\n          :host([zoom]) #course_image {\n            animation: none !important;\n          }\n        }\n\n        :host([zoom]) a:active #course_image,\n        :host([zoom]) a:hover #course_image,\n        :host([zoom]) a:focus #course_image {\n          animation-name: mymove;\n          animation-duration: 0.6s;\n          animation-iteration-count: 1;\n          animation-fill-mode: forwards;\n        }\n\n        #course_author {\n          font-size: 12px;\n          margin-top: -10px;\n        }\n      "]);

    _templateObject3_9006ac70e0cf11eab599756c9ebe7c2f = function _templateObject3_9006ac70e0cf11eab599756c9ebe7c2f() {
      return data;
    };

    return data;
  }

  function _templateObject2_9006ac70e0cf11eab599756c9ebe7c2f() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                <div id=\"course_author\">By: ", "</div>\n              "]);

    _templateObject2_9006ac70e0cf11eab599756c9ebe7c2f = function _templateObject2_9006ac70e0cf11eab599756c9ebe7c2f() {
      return data;
    };

    return data;
  }

  function _templateObject_9006ac70e0cf11eab599756c9ebe7c2f() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <a href=\"", "\" title=\"", "\">\n        <div id=\"card_wrap\">\n          <div\n            id=\"course_image\"\n            style=\"background-image:url(", ")\"\n          ></div>\n          <div id=\"course_icon\">\n            <iron-icon icon=\"", "\"></iron-icon>\n          </div>\n          <div id=\"course_number\">", "</div>\n          <div id=\"course_name\">", "</div>\n          ", "\n        </div>\n      </a>\n    "]);

    _templateObject_9006ac70e0cf11eab599756c9ebe7c2f = function _templateObject_9006ac70e0cf11eab599756c9ebe7c2f() {
      return data;
    };

    return data;
  }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  var CourseCard =
  /*#__PURE__*/
  function (_SimpleColors) {
    babelHelpers.inherits(CourseCard, _SimpleColors);
    babelHelpers.createClass(CourseCard, null, [{
      key: "properties",
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(CourseCard), "properties", this), {
          size: {
            type: String,
            reflect: true
          },
          url: {
            type: String
          },
          image: {
            type: String
          },
          alt: {
            type: String
          },
          number: {
            type: String
          },
          name: {
            type: String
          },
          icon: {
            type: String
          },
          author: {
            type: String
          },
          zoom: {
            type: Boolean,
            reflect: true
          }
        });
      }
    }]);

    function CourseCard() {
      var _this;

      babelHelpers.classCallCheck(this, CourseCard);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(CourseCard).call(this));
      _this.zoom = false;
      _this.url = "";
      _this.image = "";
      _this.alt = "";
      _this.number = "";
      _this.name = "";
      _this.icon = "";
      _this.author = "";
      return _this;
    }

    babelHelpers.createClass(CourseCard, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_9006ac70e0cf11eab599756c9ebe7c2f(), this.url, this.alt, this.image, this.icon, this.number, this.name, this.author ? (0, _litElement.html)(_templateObject2_9006ac70e0cf11eab599756c9ebe7c2f(), this.author) : "");
      }
    }], [{
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(CourseCard), "styles", this)), [(0, _litElement.css)(_templateObject3_9006ac70e0cf11eab599756c9ebe7c2f())]);
      }
    }, {
      key: "tag",
      get: function get() {
        return "course-card";
      }
    }]);
    return CourseCard;
  }(_simpleColors.SimpleColors);

  _exports.CourseCard = CourseCard;
  window.customElements.define(CourseCard.tag, CourseCard);
});
