// @lrnwebcomponents/product-card/lib/hax-element-card-list.js
define(["exports", "require", "../../../lit-element/lit-element.js", "../../utils/utils.js", "../../grid-plate/grid-plate.js", "../../../@material/mwc-switch/mwc-switch.js", "../../../@material/mwc-button/mwc-button.js", "../../../@material/mwc-formfield/mwc-formfield.js", "../../simple-modal/simple-modal.js", "../product-card.js"], function (_exports, _require, _litElement, _utils, _gridPlate, _mwcSwitch, _mwcButton, _mwcFormfield, _simpleModal, _productCard) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HAXElementCardList = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject8_55a6713083fc11eab921a14edd773829() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          border: 2px solid green;\n          min-height: 100px;\n        }\n        product-card {\n          display: block;\n        }\n        product-card div[slot=\"collapse-header\"] {\n          padding: 8px 0;\n        }\n        product-card div[slot=\"details-collapse-content\"] {\n          max-height: 125px;\n          overflow-y: auto;\n        }\n      "]);

    _templateObject8_55a6713083fc11eab921a14edd773829 = function _templateObject8_55a6713083fc11eab921a14edd773829() {
      return data;
    };

    return data;
  }

  function _templateObject7_55a6713083fc11eab921a14edd773829() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                          <mwc-button\n                            data-tag=\"", "\"\n                            @click=\"", "\"\n                            >Pop up demo</mwc-button\n                          >\n                          <div class=\"demo\">\n                            ", "\n                          </div>\n                          <code-sample copy-clipboard-button>\n                            <template>\n                              ", "\n                            </template>\n                          </code-sample>\n                        "]);

    _templateObject7_55a6713083fc11eab921a14edd773829 = function _templateObject7_55a6713083fc11eab921a14edd773829() {
      return data;
    };

    return data;
  }

  function _templateObject6_55a6713083fc11eab921a14edd773829() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                      ", "\n                    "]);

    _templateObject6_55a6713083fc11eab921a14edd773829 = function _templateObject6_55a6713083fc11eab921a14edd773829() {
      return data;
    };

    return data;
  }

  function _templateObject5_55a6713083fc11eab921a14edd773829() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                            <li>\n                              <strong>", ":</strong>\n                              <span>", "</span>\n                            </li>\n                          "]);

    _templateObject5_55a6713083fc11eab921a14edd773829 = function _templateObject5_55a6713083fc11eab921a14edd773829() {
      return data;
    };

    return data;
  }

  function _templateObject4_55a6713083fc11eab921a14edd773829() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                        ", "\n                      "]);

    _templateObject4_55a6713083fc11eab921a14edd773829 = function _templateObject4_55a6713083fc11eab921a14edd773829() {
      return data;
    };

    return data;
  }

  function _templateObject3_55a6713083fc11eab921a14edd773829() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                          ", ",\n                        "]);

    _templateObject3_55a6713083fc11eab921a14edd773829 = function _templateObject3_55a6713083fc11eab921a14edd773829() {
      return data;
    };

    return data;
  }

  function _templateObject2_55a6713083fc11eab921a14edd773829() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <product-card\n              .slot=\"col-", "\"\n              ?disabled=\"", "\"\n              ?has-demo=\"", "\"\n              heading=\"", "\"\n              icon=\"", "\"\n              subheading=\"", "\"\n              accent-color=\"", "\"\n              data-index=\"", "\"\n              @product-card-demo-show=\"", "\"\n              @product-card-demo-hide=\"", "\"\n            >\n              <div class=\"switch\">\n                <mwc-formfield label=\"", "\">\n                  <mwc-switch\n                    data-index=\"", "\"\n                    data-tag=\"", "\"\n                    data-file=\"", "\"\n                    checked\n                    @change=\"", "\"\n                  ></mwc-switch>\n                </mwc-formfield>\n              </div>\n              <div slot=\"details-collapse-header\">Details</div>\n              <div slot=\"details-collapse-content\">\n                <ul>\n                  <li>\n                    <strong>Tags:</strong> ", "\n                  </li>\n                  <li><strong>Tag name:</strong> <code>", "</code></li>\n                  <li>\n                    <strong>Developer usage:</strong>\n                    <code>import \"", "\";</code>\n                  </li>\n                  ", "\n                </ul>\n              </div>\n              <div slot=\"demo-collapse-header\">Demo</div>\n              <div slot=\"demo-collapse-content\">\n                ", "\n              </div>\n            </product-card>\n          "]);

    _templateObject2_55a6713083fc11eab921a14edd773829 = function _templateObject2_55a6713083fc11eab921a14edd773829() {
      return data;
    };

    return data;
  }

  function _templateObject_55a6713083fc11eab921a14edd773829() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <grid-plate .layout=\"", "\" hide-ops>\n        ", "\n      </grid-plate>\n    "]);

    _templateObject_55a6713083fc11eab921a14edd773829 = function _templateObject_55a6713083fc11eab921a14edd773829() {
      return data;
    };

    return data;
  }

  var HAXElementCardList =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(HAXElementCardList, _LitElement);

    function HAXElementCardList() {
      var _this;

      babelHelpers.classCallCheck(this, HAXElementCardList);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(HAXElementCardList).call(this));
      _this.list = [];
      _this._layout = "1-1-1";
      window.SimpleModal.requestAvailability();
      return _this;
    }

    babelHelpers.createClass(HAXElementCardList, [{
      key: "render",
      value: function render() {
        var _this2 = this;

        return (0, _litElement.html)(_templateObject_55a6713083fc11eab921a14edd773829(), this._layout, this.list.map(function (el, i) {
          return (0, _litElement.html)(_templateObject2_55a6713083fc11eab921a14edd773829(), _this2.__getCol(i), !el.status, el.schema.demoSchema, el.schema.gizmo.title, el.schema.gizmo.icon, el.schema.gizmo.description, el.schema.gizmo.color, i, _this2.toggleShowDemo, _this2.toggleShowDemo, el.status ? "Enabled" : "Disabled", i, el.tag, el.file, _this2.elementStatusChange, el.schema.gizmo.groups.map(function (group) {
            return (0, _litElement.html)(_templateObject3_55a6713083fc11eab921a14edd773829(), group);
          }), el.tag, el.file, el.schema.gizmo.meta ? (0, _litElement.html)(_templateObject4_55a6713083fc11eab921a14edd773829(), Object.keys(el.schema.gizmo.meta).map(function (mel) {
            return (0, _litElement.html)(_templateObject5_55a6713083fc11eab921a14edd773829(), _this2.capFirst(mel), el.schema.gizmo.meta[mel]);
          })) : "", el.schema.demoSchema && el.showDemo ? (0, _litElement.html)(_templateObject6_55a6713083fc11eab921a14edd773829(), el.schema.demoSchema.map(function (demoItem) {
            return (0, _litElement.html)(_templateObject7_55a6713083fc11eab921a14edd773829(), demoItem.tag, _this2._viewDemo, _this2._haxElementToNode(demoItem), _this2._haxElementToNode(demoItem));
          })) : "");
        }));
      }
    }, {
      key: "_viewDemo",
      value: function _viewDemo(e) {
        if (e.target && e.target.nextElementSibling) {
          window.dispatchEvent(new CustomEvent("simple-modal-show", {
            bubbles: true,
            composed: true,
            cancelable: false,
            detail: {
              title: "Demo of " + e.target.getAttribute("data-tag"),
              styles: {
                "--simple-modal-width": "75vw",
                "--simple-modal-max-width": "75vw"
              },
              elements: {
                content: e.target.nextElementSibling
              },
              invokedBy: e.target,
              clone: true,
              modal: true
            }
          }));
        }
      }
      /**
       * Effectively event binding to the expanded state
       */

    }, {
      key: "toggleShowDemo",
      value: function toggleShowDemo(e) {
        this.list[e.path[0].getAttribute("data-index")].showDemo = e.detail.expanded;
        this.requestUpdate();
      }
      /**
       * Wrap our call so that we can dynamically import code sample since it has a dep tree
       */

    }, {
      key: "_haxElementToNode",
      value: function _haxElementToNode(schema) {
        if (schema && schema.tag) {
          new Promise(function (res, rej) {
            return _require.default(["../../code-sample/code-sample.js"], res, rej);
          });
          return (0, _utils.haxElementToNode)(schema);
        }

        return "";
      }
    }, {
      key: "capFirst",
      value: function capFirst(s) {
        return s.charAt(0).toUpperCase() + s.slice(1);
      }
    }, {
      key: "elementStatusChange",
      value: function elementStatusChange(e) {
        this.list[parseInt(e.path[0].getAttribute("data-index"))].status = e.path[0].checked; // send up so list can update

        this.dispatchEvent(new CustomEvent("enabled-changed", {
          detail: {
            tag: e.path[0].getAttribute("data-tag"),
            status: e.path[0].checked
          }
        }));
        this.requestUpdate();
      }
    }, {
      key: "__getCol",
      value: function __getCol(i) {
        i++;

        while (i > this.cols) {
          i = i - this.cols;
        }

        return i;
      }
      /**
       * LitElement life cycle - property changed
       */

    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this3 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "cols") {
            switch (_this3[propName]) {
              case 3:
                _this3._layout = "1-1-1";
                break;

              case 4:
                _this3._layout = "1-1-1-1";
                break;

              case 5:
                _this3._layout = "1-1-1-1-1";
                break;

              case 6:
                _this3._layout = "1-1-1-1-1-1";
                break;

              default:
                _this3._layout = "1-1";
                break;
            }
          }
        });
      }
    }], [{
      key: "tag",
      get: function get() {
        return "hax-element-card-list";
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _litElement.css)(_templateObject8_55a6713083fc11eab921a14edd773829())];
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          list: {
            type: Array
          },
          enabled: {
            type: Object
          },
          cols: {
            type: Number
          },
          _layout: {
            type: String
          }
        };
      }
    }]);
    return HAXElementCardList;
  }(_litElement.LitElement);

  _exports.HAXElementCardList = HAXElementCardList;
  customElements.define(HAXElementCardList.tag, HAXElementCardList);
});
