// @lrnwebcomponents/product-card/lib/hax-element-list-selector.js
define(["meta", "require", "../../../lit/index.js", "../../simple-icon/simple-icon.js", "../../simple-icon/lib/simple-icons.js", "../../simple-icon/lib/simple-icon-button.js", "../../hax-iconset/lib/simple-hax-iconset.js", "../../hax-body-behaviors/lib/HAXFields.js", "../../simple-fields/simple-fields.js", "../../simple-fields/lib/simple-fields-form.js", "./hax-element-card-list.js"], function (meta, _require, _index, _simpleIcon, _simpleIcons, _simpleIconButton, _simpleHaxIconset, _HAXFields, _simpleFields, _simpleFieldsForm, _haxElementCardList) {
  "use strict";

  meta = _interopRequireWildcard(meta);
  _require = _interopRequireWildcard(_require);

  var _templateObject_e3e61e401f0511ed8cc3e9a95995ad69, _templateObject2_e3e61e401f0511ed8cc3e9a95995ad69;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _regeneratorRuntime() { "use strict"; /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/facebook/regenerator/blob/main/LICENSE */ _regeneratorRuntime = function _regeneratorRuntime() { return exports; }; var exports = {}, Op = Object.prototype, hasOwn = Op.hasOwnProperty, $Symbol = "function" == typeof Symbol ? Symbol : {}, iteratorSymbol = $Symbol.iterator || "@@iterator", asyncIteratorSymbol = $Symbol.asyncIterator || "@@asyncIterator", toStringTagSymbol = $Symbol.toStringTag || "@@toStringTag"; function define(obj, key, value) { return Object.defineProperty(obj, key, { value: value, enumerable: !0, configurable: !0, writable: !0 }), obj[key]; } try { define({}, ""); } catch (err) { define = function define(obj, key, value) { return obj[key] = value; }; } function wrap(innerFn, outerFn, self, tryLocsList) { var protoGenerator = outerFn && babelHelpers.instanceof(outerFn.prototype, Generator) ? outerFn : Generator, generator = Object.create(protoGenerator.prototype), context = new Context(tryLocsList || []); return generator._invoke = function (innerFn, self, context) { var state = "suspendedStart"; return function (method, arg) { if ("executing" === state) throw new Error("Generator is already running"); if ("completed" === state) { if ("throw" === method) throw arg; return doneResult(); } for (context.method = method, context.arg = arg;;) { var delegate = context.delegate; if (delegate) { var delegateResult = maybeInvokeDelegate(delegate, context); if (delegateResult) { if (delegateResult === ContinueSentinel) continue; return delegateResult; } } if ("next" === context.method) context.sent = context._sent = context.arg;else if ("throw" === context.method) { if ("suspendedStart" === state) throw state = "completed", context.arg; context.dispatchException(context.arg); } else "return" === context.method && context.abrupt("return", context.arg); state = "executing"; var record = tryCatch(innerFn, self, context); if ("normal" === record.type) { if (state = context.done ? "completed" : "suspendedYield", record.arg === ContinueSentinel) continue; return { value: record.arg, done: context.done }; } "throw" === record.type && (state = "completed", context.method = "throw", context.arg = record.arg); } }; }(innerFn, self, context), generator; } function tryCatch(fn, obj, arg) { try { return { type: "normal", arg: fn.call(obj, arg) }; } catch (err) { return { type: "throw", arg: err }; } } exports.wrap = wrap; var ContinueSentinel = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} var IteratorPrototype = {}; define(IteratorPrototype, iteratorSymbol, function () { return this; }); var getProto = Object.getPrototypeOf, NativeIteratorPrototype = getProto && getProto(getProto(values([]))); NativeIteratorPrototype && NativeIteratorPrototype !== Op && hasOwn.call(NativeIteratorPrototype, iteratorSymbol) && (IteratorPrototype = NativeIteratorPrototype); var Gp = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(IteratorPrototype); function defineIteratorMethods(prototype) { ["next", "throw", "return"].forEach(function (method) { define(prototype, method, function (arg) { return this._invoke(method, arg); }); }); } function AsyncIterator(generator, PromiseImpl) { function invoke(method, arg, resolve, reject) { var record = tryCatch(generator[method], generator, arg); if ("throw" !== record.type) { var result = record.arg, value = result.value; return value && "object" == babelHelpers.typeof(value) && hasOwn.call(value, "__await") ? PromiseImpl.resolve(value.__await).then(function (value) { invoke("next", value, resolve, reject); }, function (err) { invoke("throw", err, resolve, reject); }) : PromiseImpl.resolve(value).then(function (unwrapped) { result.value = unwrapped, resolve(result); }, function (error) { return invoke("throw", error, resolve, reject); }); } reject(record.arg); } var previousPromise; this._invoke = function (method, arg) { function callInvokeWithMethodAndArg() { return new PromiseImpl(function (resolve, reject) { invoke(method, arg, resolve, reject); }); } return previousPromise = previousPromise ? previousPromise.then(callInvokeWithMethodAndArg, callInvokeWithMethodAndArg) : callInvokeWithMethodAndArg(); }; } function maybeInvokeDelegate(delegate, context) { var method = delegate.iterator[context.method]; if (undefined === method) { if (context.delegate = null, "throw" === context.method) { if (delegate.iterator.return && (context.method = "return", context.arg = undefined, maybeInvokeDelegate(delegate, context), "throw" === context.method)) return ContinueSentinel; context.method = "throw", context.arg = new TypeError("The iterator does not provide a 'throw' method"); } return ContinueSentinel; } var record = tryCatch(method, delegate.iterator, context.arg); if ("throw" === record.type) return context.method = "throw", context.arg = record.arg, context.delegate = null, ContinueSentinel; var info = record.arg; return info ? info.done ? (context[delegate.resultName] = info.value, context.next = delegate.nextLoc, "return" !== context.method && (context.method = "next", context.arg = undefined), context.delegate = null, ContinueSentinel) : info : (context.method = "throw", context.arg = new TypeError("iterator result is not an object"), context.delegate = null, ContinueSentinel); } function pushTryEntry(locs) { var entry = { tryLoc: locs[0] }; 1 in locs && (entry.catchLoc = locs[1]), 2 in locs && (entry.finallyLoc = locs[2], entry.afterLoc = locs[3]), this.tryEntries.push(entry); } function resetTryEntry(entry) { var record = entry.completion || {}; record.type = "normal", delete record.arg, entry.completion = record; } function Context(tryLocsList) { this.tryEntries = [{ tryLoc: "root" }], tryLocsList.forEach(pushTryEntry, this), this.reset(!0); } function values(iterable) { if (iterable) { var iteratorMethod = iterable[iteratorSymbol]; if (iteratorMethod) return iteratorMethod.call(iterable); if ("function" == typeof iterable.next) return iterable; if (!isNaN(iterable.length)) { var i = -1, next = function next() { for (; ++i < iterable.length;) { if (hasOwn.call(iterable, i)) return next.value = iterable[i], next.done = !1, next; } return next.value = undefined, next.done = !0, next; }; return next.next = next; } } return { next: doneResult }; } function doneResult() { return { value: undefined, done: !0 }; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, define(Gp, "constructor", GeneratorFunctionPrototype), define(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = define(GeneratorFunctionPrototype, toStringTagSymbol, "GeneratorFunction"), exports.isGeneratorFunction = function (genFun) { var ctor = "function" == typeof genFun && genFun.constructor; return !!ctor && (ctor === GeneratorFunction || "GeneratorFunction" === (ctor.displayName || ctor.name)); }, exports.mark = function (genFun) { return Object.setPrototypeOf ? Object.setPrototypeOf(genFun, GeneratorFunctionPrototype) : (genFun.__proto__ = GeneratorFunctionPrototype, define(genFun, toStringTagSymbol, "GeneratorFunction")), genFun.prototype = Object.create(Gp), genFun; }, exports.awrap = function (arg) { return { __await: arg }; }, defineIteratorMethods(AsyncIterator.prototype), define(AsyncIterator.prototype, asyncIteratorSymbol, function () { return this; }), exports.AsyncIterator = AsyncIterator, exports.async = function (innerFn, outerFn, self, tryLocsList, PromiseImpl) { void 0 === PromiseImpl && (PromiseImpl = Promise); var iter = new AsyncIterator(wrap(innerFn, outerFn, self, tryLocsList), PromiseImpl); return exports.isGeneratorFunction(outerFn) ? iter : iter.next().then(function (result) { return result.done ? result.value : iter.next(); }); }, defineIteratorMethods(Gp), define(Gp, toStringTagSymbol, "Generator"), define(Gp, iteratorSymbol, function () { return this; }), define(Gp, "toString", function () { return "[object Generator]"; }), exports.keys = function (object) { var keys = []; for (var key in object) { keys.push(key); } return keys.reverse(), function next() { for (; keys.length;) { var key = keys.pop(); if (key in object) return next.value = key, next.done = !1, next; } return next.done = !0, next; }; }, exports.values = values, Context.prototype = { constructor: Context, reset: function reset(skipTempReset) { if (this.prev = 0, this.next = 0, this.sent = this._sent = undefined, this.done = !1, this.delegate = null, this.method = "next", this.arg = undefined, this.tryEntries.forEach(resetTryEntry), !skipTempReset) for (var name in this) { "t" === name.charAt(0) && hasOwn.call(this, name) && !isNaN(+name.slice(1)) && (this[name] = undefined); } }, stop: function stop() { this.done = !0; var rootRecord = this.tryEntries[0].completion; if ("throw" === rootRecord.type) throw rootRecord.arg; return this.rval; }, dispatchException: function dispatchException(exception) { if (this.done) throw exception; var context = this; function handle(loc, caught) { return record.type = "throw", record.arg = exception, context.next = loc, caught && (context.method = "next", context.arg = undefined), !!caught; } for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i], record = entry.completion; if ("root" === entry.tryLoc) return handle("end"); if (entry.tryLoc <= this.prev) { var hasCatch = hasOwn.call(entry, "catchLoc"), hasFinally = hasOwn.call(entry, "finallyLoc"); if (hasCatch && hasFinally) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } else if (hasCatch) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); } else { if (!hasFinally) throw new Error("try statement without catch or finally"); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } } } }, abrupt: function abrupt(type, arg) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc <= this.prev && hasOwn.call(entry, "finallyLoc") && this.prev < entry.finallyLoc) { var finallyEntry = entry; break; } } finallyEntry && ("break" === type || "continue" === type) && finallyEntry.tryLoc <= arg && arg <= finallyEntry.finallyLoc && (finallyEntry = null); var record = finallyEntry ? finallyEntry.completion : {}; return record.type = type, record.arg = arg, finallyEntry ? (this.method = "next", this.next = finallyEntry.finallyLoc, ContinueSentinel) : this.complete(record); }, complete: function complete(record, afterLoc) { if ("throw" === record.type) throw record.arg; return "break" === record.type || "continue" === record.type ? this.next = record.arg : "return" === record.type ? (this.rval = this.arg = record.arg, this.method = "return", this.next = "end") : "normal" === record.type && afterLoc && (this.next = afterLoc), ContinueSentinel; }, finish: function finish(finallyLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.finallyLoc === finallyLoc) return this.complete(entry.completion, entry.afterLoc), resetTryEntry(entry), ContinueSentinel; } }, catch: function _catch(tryLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc === tryLoc) { var record = entry.completion; if ("throw" === record.type) { var thrown = record.arg; resetTryEntry(entry); } return thrown; } } throw new Error("illegal catch attempt"); }, delegateYield: function delegateYield(iterable, resultName, nextLoc) { return this.delegate = { iterator: values(iterable), resultName: resultName, nextLoc: nextLoc }, "next" === this.method && (this.arg = undefined), ContinueSentinel; } }, exports; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `hax-element-list-selector`
   * selects a hax element
   * @demo demo/hax.html
   * @element hax-element-list-selector
   */
  var HaxElementListSelector = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(HaxElementListSelector, _LitElement);

    var _super = _createSuper(HaxElementListSelector);

    function HaxElementListSelector() {
      var _this;

      babelHelpers.classCallCheck(this, HaxElementListSelector);
      _this = _super.call(this);
      _this.loading = false;
      _this.cols = 3;
      _this.showCardList = false;
      _this.imports = [];
      _this.haxData = [];
      _this.noSchema = {};
      _this.method = "GET";
      _this.autoload = true;
      _this.HaxSchematizer = _HAXFields.HaxSchematizer;
      _this.HaxElementizer = _HAXFields.HaxElementizer; // default fields json blob, most implementations should provide their own though obviously

      _this.fieldsEndpoint = new URL("./fields.json", meta.url).href; // allow flobal base path focibly set

      if (window.WCGlobalBasePath) {
        _this.basePath = window.WCGlobalBasePath;
      } else {
        _this.basePath = new URL("./../../../", meta.url).href;
      } // allow global definition of wc-registry for custom ones


      if (window.WCGlobalRegistryFileName) {
        _this.__regFile = window.WCGlobalRegistryFileName;
      } else {
        _this.__regFile = "wc-registry.json";
      }

      setTimeout(function () {
        window.addEventListener("a11y-tabs-active-changed", _this._activeTabChanged.bind(babelHelpers.assertThisInitialized(_this)));
      }, 0);

      _this.addEventListener("response", _this._response);

      _this.addEventListener("value-changed", _this._valueChanged);

      return _this;
    }

    babelHelpers.createClass(HaxElementListSelector, [{
      key: "render",
      value: function render() {
        var _this2 = this;

        return (0, _index.html)(_templateObject_e3e61e401f0511ed8cc3e9a95995ad69 || (_templateObject_e3e61e401f0511ed8cc3e9a95995ad69 = babelHelpers.taggedTemplateLiteral(["\n      <simple-fields-form\n        id=\"form\"\n        autoload\n        load-endpoint=\"", "\"\n        method=\"", "\"\n        .schematizer=\"", "\"\n        .elementizer=\"", "\"\n        @response=\"", "\"\n        @haxcore.search.haxcore-search-columns-value-changed=\"", "\"\n        @value-changed=\"", "\"\n      >\n      </simple-fields-form>\n    "])), this.fieldsEndpoint, this.method, this.HaxSchematizer, this.HaxElementizer, this._response, function (e) {
          return _this2.cols = e.detail.value;
        }, this._valueChanged);
      }
    }, {
      key: "_activeTabChanged",
      value: function _activeTabChanged(e) {
        if (e.detail.activeTab == "haxcore.search") {
          this.showCardList = true;
        } else {
          this.showCardList = false;
        }
      }
    }, {
      key: "_getElement",
      value: function _getElement(id) {
        return this.form && this.form.formElements && this.form.formElements[id] ? this.form.formElements[id].element : undefined;
      }
    }, {
      key: "form",
      get: function get() {
        return this.shadowRoot && this.shadowRoot.querySelector("#form") ? this.shadowRoot.querySelector("#form") : undefined;
      }
    }, {
      key: "cardList",
      get: function get() {
        return this._getElement("haxcore.search.haxcore-search-autoloader");
      }
    }, {
      key: "searchColumns",
      get: function get() {
        this._getElement("haxcore.search.haxcore-search-columns");
      }
    }, {
      key: "haxTags",
      get: function get() {
        this._getElement("haxcore.providerdetails.haxcore-providerdetails-haxtags");
      }
    }, {
      key: "otherTags",
      get: function get() {
        this._getElement("haxcore.providerdetails.haxcore-providerdetails-othertags");
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this3 = this;

        changedProperties.forEach( /*#__PURE__*/function () {
          var _ref = babelHelpers.asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee2(oldValue, propName) {
            var tag, renderHaxData, i, search;
            return _regeneratorRuntime().wrap(function _callee2$(_context3) {
              while (1) {
                switch (_context3.prev = _context3.next) {
                  case 0:
                    if (propName == "loading" && _this3.cardList) _this3.cardList.loading = _this3.loading;
                    if (propName == "showCardList" && _this3.cardList) _this3.cardList.showCardList = _this3.showCardList;
                    if (propName == "cols" && _this3.cardList) _this3.cardList.cols = _this3.cols;
                    if (propName == "cols" && _this3.searchColumns) _this3.searchColumns.value = _this3.cols;

                    if (propName == "wcRegistryEndpoint") {
                      _this3.haxData = [];
                      _this3.imports = [];
                      fetch(_this3.wcRegistryEndpoint).then(function (response) {
                        _this3.loading = true;
                        return response.json();
                      }).then(function (data) {
                        _this3.imports = data;
                      });
                    } // when imports changes make sure we import everything found


                    if (!(propName == "imports")) {
                      _context3.next = 7;
                      break;
                    }

                    return _context3.delegateYield( /*#__PURE__*/_regeneratorRuntime().mark(function _callee() {
                      var list, noSchema, _loop;

                      return _regeneratorRuntime().wrap(function _callee$(_context2) {
                        while (1) {
                          switch (_context2.prev = _context2.next) {
                            case 0:
                              list = _this3.haxData, noSchema = _this3.noSchema;
                              _loop = /*#__PURE__*/_regeneratorRuntime().mark(function _loop() {
                                var file;
                                return _regeneratorRuntime().wrap(function _loop$(_context) {
                                  while (1) {
                                    switch (_context.prev = _context.next) {
                                      case 0:
                                        file = _this3[propName][tag];
                                        _context.prev = 1;
                                        _context.next = 4;
                                        return new Promise(function (res, rej) {
                                          return _require.default(["".concat(_this3.basePath).concat(file)], res, rej);
                                        }).then(function (module) {
                                          if (module && Object.keys(module) && Object.keys(module)[0] && module[Object.keys(module)[0]] && module[Object.keys(module)[0]].haxProperties && module[Object.keys(module)[0]].haxProperties.gizmo && module[Object.keys(module)[0]].haxProperties.gizmo.title) {
                                            var detail = {
                                              tag: tag,
                                              file: file,
                                              showDemo: false,
                                              schema: module[Object.keys(module)[0]].haxProperties
                                            };
                                            list.push(detail);
                                            _this3.haxData = babelHelpers.toConsumableArray(list);
                                          } else {
                                            noSchema[tag] = file; //console.log(`${tag} doesn't have haxSchema`);
                                          }
                                        });

                                      case 4:
                                        _context.next = 9;
                                        break;

                                      case 6:
                                        _context.prev = 6;
                                        _context.t0 = _context["catch"](1);
                                        console.warn(_context.t0);

                                      case 9:
                                      case "end":
                                        return _context.stop();
                                    }
                                  }
                                }, _loop, null, [[1, 6]]);
                              });
                              _context2.t0 = _regeneratorRuntime().keys(_this3[propName]);

                            case 3:
                              if ((_context2.t1 = _context2.t0()).done) {
                                _context2.next = 8;
                                break;
                              }

                              tag = _context2.t1.value;
                              return _context2.delegateYield(_loop(), "t2", 6);

                            case 6:
                              _context2.next = 3;
                              break;

                            case 8:
                              _this3.noSchema = {};
                              _this3.noSchema = noSchema;
                              _this3.loading = false;

                            case 11:
                            case "end":
                              return _context2.stop();
                          }
                        }
                      }, _callee);
                    })(), "t0", 7);

                  case 7:
                    // this is the local data we don't let change
                    if (propName == "haxData") {
                      _this3.filteredHaxData = babelHelpers.toConsumableArray(_this3.haxData);

                      if (_this3.haxData.length > 0) {
                        renderHaxData = {};

                        for (i in _this3.haxData) {
                          renderHaxData[_this3.haxData[i].tag] = _this3.haxData[i].file;
                        }

                        if (_this3.cardList) {
                          search = _this3.form && _this3.form.value && _this3.form.value.haxcore && _this3.form.value.haxcore.search ? _this3.form.value.haxcore.search : undefined;
                          _this3.cardList.list = _this3.filteredHaxData;
                          if (search) _this3.applyFilters(search);

                          _this3.cardList.requestUpdate();
                        }

                        if (_this3.haxTags) _this3.haxTags.editorValue = JSON.stringify(renderHaxData, null, 2);
                      }
                    }

                    if (propName == "noSchema" && Object.keys(_this3.noSchema).length > 0 && _this3.otherTags) _this3.otherTags.editorValue = JSON.stringify(_this3.noSchema, null, 2);

                  case 9:
                  case "end":
                    return _context3.stop();
                }
              }
            }, _callee2);
          }));

          return function (_x, _x2) {
            return _ref.apply(this, arguments);
          };
        }());
      }
    }, {
      key: "applyFilters",
      value: function applyFilters(filters) {
        if (this.cardList) {
          var data = babelHelpers.toConsumableArray(this.haxData);
          Object.keys(filters || {}).forEach(function (key) {
            if (filters[key] != "") {
              switch (key) {
                case "haxcore-search-search":
                  data = data.filter(function (item) {
                    if (item.schema.gizmo.title.toLowerCase().includes(filters[key].toLowerCase())) {
                      return true;
                    }

                    return false;
                  });
                  break;

                case "haxcore-search-tags":
                  data = data.filter(function (item) {
                    if (item.schema.gizmo.groups.includes(filters[key])) {
                      return true;
                    }

                    return false;
                  });
                  break;

                case "haxcore-search-hasdemo":
                  // only filter if box checked otherwise show all
                  if (filters[key]) {
                    data = data.filter(function (item) {
                      if (item.schema.demoSchema) {
                        return true;
                      }

                      return false;
                    });
                  }

                  break;
              }
            }
          });
          this.cardList.filteredTags = babelHelpers.toConsumableArray(data).map(function (item) {
            return item.tag;
          });
          this.cardList.requestUpdate();
        }
      }
      /**
       * Listen for response and then apply initial settings
       */

    }, {
      key: "_response",
      value: function _response(e) {
        if (this.searchColumns) this.searchColumns.value = this.cols;
        if (this.cardList) this.cardList.cols = this.cols; //this._valueChanged(e);
      }
      /**
       * notice any value changing and then getting the form fresh
       */

    }, {
      key: "_valueChanged",
      value: function _valueChanged() {
        var _this4 = this;

        clearTimeout(this.__valueDebounce);
        this.__valueDebounce = setTimeout(function () {
          var haxcore = _this4.form && _this4.form.value && _this4.form.value.haxcore ? _this4.form.value.haxcore : undefined;

          if (haxcore) {
            var cols = haxcore.search && haxcore.search["haxcore-search-columns"] ? haxcore.search["haxcore-search-columns"] : undefined; //set columns

            if (cols) _this4.cols = cols; // look for CDN provider

            if (haxcore.providers["haxcore-providers-cdn"] == "other") {
              _this4.wcRegistryEndpoint = haxcore.providers["haxcore-providers-other"] + _this4.__regFile;
            } else {
              _this4.wcRegistryEndpoint = haxcore.providers["haxcore-providers-cdn"] + _this4.__regFile;
            } // apply filters


            _this4.applyFilters(haxcore.search);

            if (_this4.cardList) {
              _this4.dispatchEvent(new CustomEvent("appstore-changed", {
                detail: {
                  value: _this4.getAppstoreValues()
                }
              }));
            }
          }
        }, 50);
      }
      /**
       * Return the appstore values
       */

    }, {
      key: "getAppstoreValues",
      value: function getAppstoreValues() {
        // get form values
        var value = this.shadowRoot.querySelector("#form").submit();
        value.haxcore = value.haxcore || {
          templates: {},
          providers: {}
        };
        var appstore = {
          providers: {
            cdn: value.haxcore.providers["haxcore-providers-cdn"],
            other: value.haxcore.providers["haxcore-providers-other"],
            pk: value.haxcore.providers["haxcore-providers-pk"]
          },
          apps: {},
          stax: value.haxcore.templates["haxcore-templates-layouts"],
          autoloader: value.haxcore.search["haxcore-search-autoloader"]
        }; // find the API keys

        for (var key in value.haxcore.integrations) {
          appstore.apps[key.replace("haxcore-integrations-", "")] = value.haxcore.integrations[key];
        }

        return appstore;
      }
      /**
       * Autoloader is a simple keypair
       */

    }, {
      key: "getAutoloader",
      value: function getAutoloader(data) {
        var autoload = {};

        for (var i in data) {
          if (data[i].status) {
            autoload[data[i].tag] = data[i].file;
          }
        }

        return autoload;
      }
    }], [{
      key: "tag",
      get: function get() {
        return "hax-element-list-selector";
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _index.css)(_templateObject2_e3e61e401f0511ed8cc3e9a95995ad69 || (_templateObject2_e3e61e401f0511ed8cc3e9a95995ad69 = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n      "])))];
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, _simpleFieldsForm.SimpleFieldsForm.properties), {}, {
          HaxSchematizer: {
            type: Object
          },
          HaxElementizer: {
            type: Object
          },

          /**
           * Show card list so that it SEEMS like its happenign when we click HAX elements
           */
          showCardList: {
            type: Boolean
          },

          /**
           * JS imports
           */
          imports: {
            type: Object
          },

          /**
           * HAXSchema array
           */
          haxData: {
            type: Array
          },

          /**
           * Valid tags on the CDN but that don't have haxSchema.
           */
          noSchema: {
            type: Object
          },

          /**
           * Data filtered by form changes
           */
          filteredHaxData: {
            type: Array
          },

          /**
           * Columns to render
           */
          cols: {
            type: Number
          },

          /**
           * End point to load this data
           */
          fieldsEndpoint: {
            type: String,
            attribute: "fields-endpoint"
          },

          /**
           * End point to load up a list of imports
           */
          wcRegistryEndpoint: {
            type: String,
            attribute: "wc-registry-endpoint"
          },

          /**
           * Request method
           */
          method: {
            type: String
          },
          loading: {
            type: Boolean,
            reflect: true
          }
        });
      }
    }]);
    return HaxElementListSelector;
  }(_index.LitElement);

  window.customElements.define(HaxElementListSelector.tag, HaxElementListSelector);
});
