// @lrnwebcomponents/haxcms-elements/lib/ui-components/navigation/site-breadcrumb.js
define(["exports", "require", "../../../../../lit/index.js", "../../core/haxcms-site-store.js", "../../../../simple-icon/simple-icon.js", "../../../../../mobx/dist/mobx.esm.js"], function (_exports, _require, _index, _haxcmsSiteStore, _simpleIcon, _mobxEsm) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SiteBreadcrumb = void 0;
  _require = _interopRequireWildcard(_require);

  var _templateObject_4a1923c0f3af11eb8aaf6bf577118b7e, _templateObject2_4a1923c0f3af11eb8aaf6bf577118b7e;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `site-breadcrumb`
   * `A basic breadcrumb of links based on the active state in HAXcms on JSON Outline Schema`
   *
   * @demo demo/index.html
   */
  var SiteBreadcrumb = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(SiteBreadcrumb, _LitElement);

    var _super = _createSuper(SiteBreadcrumb);

    function SiteBreadcrumb() {
      var _this;

      babelHelpers.classCallCheck(this, SiteBreadcrumb);
      _this = _super.call(this);
      _this.__disposer = [];
      new Promise(function (res, rej) {
        return _require.default(["../../../../simple-icon/lib/simple-icons.js"], res, rej);
      });
      return _this;
    } // render function


    babelHelpers.createClass(SiteBreadcrumb, [{
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_4a1923c0f3af11eb8aaf6bf577118b7e || (_templateObject_4a1923c0f3af11eb8aaf6bf577118b7e = babelHelpers.taggedTemplateLiteral(["\n      <div\n        id=\"space\"\n        itemscope\n        itemtype=\"http://data-vocabulary.org/Breadcrumb\"\n      ></div>\n    "])));
      }
    }, {
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        var _this2 = this;

        if (babelHelpers.get(babelHelpers.getPrototypeOf(SiteBreadcrumb.prototype), "firstUpdated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(SiteBreadcrumb.prototype), "firstUpdated", this).call(this, changedProperties);
        } // keep editMode in sync globally


        (0, _mobxEsm.autorun)(function (reaction) {
          _this2.manifest = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.routerManifest);

          _this2.__disposer.push(reaction);
        });
        (0, _mobxEsm.autorun)(function (reaction) {
          _this2.editMode = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.editMode);

          _this2.__disposer.push(reaction);
        });
        (0, _mobxEsm.autorun)(function (reaction) {
          _this2._activeItemChanged((0, _mobxEsm.toJS)(_haxcmsSiteStore.store.activeItem));

          _this2.__disposer.push(reaction);
        });
      }
      /**
       * Notice the change and build
       */

    }, {
      key: "_activeItemChanged",
      value: function _activeItemChanged(active) {
        var _this3 = this;

        var activeItem = active;

        if (activeItem && this.shadowRoot && this.shadowRoot.querySelector("#space")) {
          var items;
          var i;

          (function () {
            // wipe out the slot and rebuild it
            while (_this3.shadowRoot.querySelector("#space").firstChild !== null) {
              _this3.shadowRoot.querySelector("#space").removeChild(_this3.shadowRoot.querySelector("#space").firstChild);
            }

            items = [{
              title: activeItem.title,
              slug: null
            }];
            var itemBuilder = activeItem; // walk back through parent tree

            while (itemBuilder && itemBuilder.parent != null) {
              itemBuilder = _this3.manifest.items.find(function (i) {
                return i.id == itemBuilder.parent;
              }); // double check structure is sound

              if (itemBuilder) {
                items.unshift({
                  title: itemBuilder.title,
                  slug: itemBuilder.slug
                });
              }
            }

            for (i in items) {
              var icon = document.createElement("simple-icon");
              icon.icon = "icons:chevron-right";

              if (items[i].slug != null) {
                var button = document.createElement("button");
                button.innerText = items[i].title;
                button.noink = true; // disable buttons if we ware editing

                if (_this3.editMode) {
                  button.setAttribute("disabled", "disabled");

                  _this3.shadowRoot.querySelector("#space").appendChild(button);
                } else {
                  var link = document.createElement("a");
                  link.setAttribute("href", items[i].slug);
                  link.setAttribute("tabindex", "-1");
                  link.setAttribute("itemprop", "url");
                  link.appendChild(button);

                  _this3.shadowRoot.querySelector("#space").appendChild(link);
                }

                _this3.shadowRoot.querySelector("#space").appendChild(icon);
              } else {
                var span = document.createElement("span");
                span.innerText = items[i].title;

                _this3.shadowRoot.querySelector("#space").appendChild(span);
              }
            }
          })();
        }
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        for (var i in this.__disposer) {
          this.__disposer[i].dispose();
        }

        babelHelpers.get(babelHelpers.getPrototypeOf(SiteBreadcrumb.prototype), "disconnectedCallback", this).call(this);
      }
    }], [{
      key: "styles",
      get: function get() {
        return [(0, _index.css)(_templateObject2_4a1923c0f3af11eb8aaf6bf577118b7e || (_templateObject2_4a1923c0f3af11eb8aaf6bf577118b7e = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          font-size: 16px;\n          color: var(--site-breadcrumb-color, #383f45);\n        }\n        #space {\n          height: 24px;\n          line-height: 24px;\n        }\n        a {\n          height: 24px;\n          font-size: 16px;\n          color: var(--site-breadcrumb-color, #383f45);\n          display: inline-flex;\n          line-height: 24px;\n          padding: 0 8px 0 0;\n          vertical-align: text-top;\n          text-decoration: var(--site-breadcrumb-text-decoration, underline);\n        }\n        button {\n          margin: 0;\n          padding: 0;\n          font-size: 16px;\n          font-family: inherit;\n          height: 24px;\n          min-width: unset;\n          display: inline-flex;\n          text-transform: unset;\n          background-color: transparent;\n          border: none;\n          letter-spacing: inherit;\n        }\n        button:hover,\n        button:focus,\n        button:active {\n          background-color: var(--site-breadcrumb-hover-bg, transparent);\n          color: var(--site-breadcrumb-hover-color, #222222);\n        }\n        span {\n          margin: 0;\n          font-size: 16px;\n          padding: 0 8px 0 0;\n          height: 24px;\n          display: inline-flex;\n        }\n        simple-icon {\n          display: inline-flex;\n          --simple-icon-height: 24px;\n          --simple-icon-width: 24px;\n          padding: 0 8px 0 0;\n          color: var(--site-breadcrumb-color, #383f45);\n        }\n      "])))];
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       */

    }, {
      key: "tag",
      get: function get() {
        return "site-breadcrumb";
      }
    }]);
    return SiteBreadcrumb;
  }(_index.LitElement);

  _exports.SiteBreadcrumb = SiteBreadcrumb;
  window.customElements.define(SiteBreadcrumb.tag, SiteBreadcrumb);
});
