// @lrnwebcomponents/simple-blog/lib/simple-blog-post.js
define(["exports", "require", "../../../@polymer/polymer/polymer-element.js", "../../../@polymer/polymer/lib/utils/render-status.js", "../../../@polymer/polymer/lib/utils/async.js", "../../haxcms-elements/lib/core/haxcms-site-store.js", "../../../mobx/lib/mobx.module.js"], function (_exports, _require, _polymerElement, _renderStatus, _async, _haxcmsSiteStore, _mobxModule) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SimpleBlogPost = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject_f42965b0e53211e9a0cfefb808989d41() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style include=\"simple-colors-shared-styles\">\n        :host {\n          display: block;\n          min-height: 80vh;\n        }\n        main {\n          transition: opacity 0.2s ease-in-out, visibility 0.2s ease-in-out;\n          width: 100%;\n          max-width: 900px;\n          margin: 0 auto;\n          -webkit-box-sizing: border-box;\n          -moz-box-sizing: border-box;\n          box-sizing: border-box;\n          opacity: 1;\n          visibility: visible;\n        }\n        @media only screen and (max-width: 900px) {\n          main {\n            padding: 0 32px;\n          }\n        }\n        section {\n          width: 100%;\n          font-family: Linux Libertine;\n          color: #333;\n        }\n        section ::slotted(*) {\n          font-weight: 400;\n          font-style: normal;\n          font-size: 22px;\n          line-height: 30px;\n          margin: 0;\n          padding-bottom: 30px;\n          color: #333;\n          -webkit-hyphens: auto;\n          -moz-hyphens: auto;\n          hyphens: auto;\n        }\n        :host([has-image]) .article-image {\n          position: absolute;\n          background-color: var(--haxcms-color, black);\n          top: 0;\n          left: 0;\n          right: 0;\n          bottom: 0;\n          overflow: hidden;\n        }\n        .post-image-image {\n          background-size: cover;\n          position: absolute;\n          top: 0;\n          left: 0;\n          right: 0;\n          bottom: 0;\n          text-indent: -9999px;\n          padding-top: 33%;\n          z-index: 1;\n        }\n        .post-meta {\n          font-family: Open Sans, MundoSans, \"Helvetica Neue\", Arial, Helvetica,\n            sans-serif;\n        }\n        :host([has-image]) .post-meta {\n          position: absolute;\n          bottom: 80px;\n          left: 30%;\n          right: 30%;\n          padding-top: 60px;\n          z-index: 9;\n          font-family: Open Sans, MundoSans, \"Helvetica Neue\", Arial, Helvetica,\n            sans-serif;\n          -webkit-box-sizing: border-box;\n          -moz-box-sizing: border-box;\n          box-sizing: border-box;\n        }\n        site-active-title {\n          --site-active-title-heading: {\n            font-weight: 700;\n            font-style: normal;\n            letter-spacing: -0.04em;\n            font-size: 50px;\n            line-height: 1.1;\n            color: black;\n          }\n        }\n        :host([has-image]) site-active-title {\n          --site-active-title-heading: {\n            font-weight: 700;\n            font-style: normal;\n            letter-spacing: -0.04em;\n            font-size: 50px;\n            line-height: 1.1;\n            margin-bottom: 16px;\n            text-shadow: 0 1px 16px rgba(0, 0, 0, 0.5),\n              0 0 1px rgba(0, 0, 0, 0.5);\n            color: white;\n          }\n        }\n        /**\n         * Hide the slotted content during edit mode\n         */\n        :host([edit-mode]) #slot {\n          display: none;\n        }\n      </style>\n      <main>\n        <article>\n          <div class=\"article-image\">\n            <template is=\"dom-if\" if=\"[[hasImage]]\">\n              <div\n                id=\"image\"\n                class=\"post-image-image\"\n                style$='background-image: url(\"[[image]]\");'\n              ></div>\n            </template>\n            <div class=\"post-meta\">\n              <site-active-title></site-active-title>\n            </div>\n          </div>\n          <section id=\"contentcontainer\">\n            <div id=\"slot\"><slot></slot></div>\n          </section>\n        </article>\n      </main>\n    "], ["\n      <style include=\"simple-colors-shared-styles\">\n        :host {\n          display: block;\n          min-height: 80vh;\n        }\n        main {\n          transition: opacity 0.2s ease-in-out, visibility 0.2s ease-in-out;\n          width: 100%;\n          max-width: 900px;\n          margin: 0 auto;\n          -webkit-box-sizing: border-box;\n          -moz-box-sizing: border-box;\n          box-sizing: border-box;\n          opacity: 1;\n          visibility: visible;\n        }\n        @media only screen and (max-width: 900px) {\n          main {\n            padding: 0 32px;\n          }\n        }\n        section {\n          width: 100%;\n          font-family: Linux Libertine;\n          color: #333;\n        }\n        section ::slotted(*) {\n          font-weight: 400;\n          font-style: normal;\n          font-size: 22px;\n          line-height: 30px;\n          margin: 0;\n          padding-bottom: 30px;\n          color: #333;\n          -webkit-hyphens: auto;\n          -moz-hyphens: auto;\n          hyphens: auto;\n        }\n        :host([has-image]) .article-image {\n          position: absolute;\n          background-color: var(--haxcms-color, black);\n          top: 0;\n          left: 0;\n          right: 0;\n          bottom: 0;\n          overflow: hidden;\n        }\n        .post-image-image {\n          background-size: cover;\n          position: absolute;\n          top: 0;\n          left: 0;\n          right: 0;\n          bottom: 0;\n          text-indent: -9999px;\n          padding-top: 33%;\n          z-index: 1;\n        }\n        .post-meta {\n          font-family: Open Sans, MundoSans, \"Helvetica Neue\", Arial, Helvetica,\n            sans-serif;\n        }\n        :host([has-image]) .post-meta {\n          position: absolute;\n          bottom: 80px;\n          left: 30%;\n          right: 30%;\n          padding-top: 60px;\n          z-index: 9;\n          font-family: Open Sans, MundoSans, \"Helvetica Neue\", Arial, Helvetica,\n            sans-serif;\n          -webkit-box-sizing: border-box;\n          -moz-box-sizing: border-box;\n          box-sizing: border-box;\n        }\n        site-active-title {\n          --site-active-title-heading: {\n            font-weight: 700;\n            font-style: normal;\n            letter-spacing: -0.04em;\n            font-size: 50px;\n            line-height: 1.1;\n            color: black;\n          }\n        }\n        :host([has-image]) site-active-title {\n          --site-active-title-heading: {\n            font-weight: 700;\n            font-style: normal;\n            letter-spacing: -0.04em;\n            font-size: 50px;\n            line-height: 1.1;\n            margin-bottom: 16px;\n            text-shadow: 0 1px 16px rgba(0, 0, 0, 0.5),\n              0 0 1px rgba(0, 0, 0, 0.5);\n            color: white;\n          }\n        }\n        /**\n         * Hide the slotted content during edit mode\n         */\n        :host([edit-mode]) #slot {\n          display: none;\n        }\n      </style>\n      <main>\n        <article>\n          <div class=\"article-image\">\n            <template is=\"dom-if\" if=\"[[hasImage]]\">\n              <div\n                id=\"image\"\n                class=\"post-image-image\"\n                style\\$='background-image: url(\"[[image]]\");'\n              ></div>\n            </template>\n            <div class=\"post-meta\">\n              <site-active-title></site-active-title>\n            </div>\n          </div>\n          <section id=\"contentcontainer\">\n            <div id=\"slot\"><slot></slot></div>\n          </section>\n        </article>\n      </main>\n    "]);

    _templateObject_f42965b0e53211e9a0cfefb808989d41 = function _templateObject_f42965b0e53211e9a0cfefb808989d41() {
      return data;
    };

    return data;
  }

  /**
   * `simple-blog-post`
   * `A simple blog and associated elements`
   */
  var SimpleBlogPost =
  /*#__PURE__*/
  function (_PolymerElement) {
    babelHelpers.inherits(SimpleBlogPost, _PolymerElement);
    babelHelpers.createClass(SimpleBlogPost, null, [{
      key: "tag",

      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */
      get: function get() {
        return "simple-blog-post";
      } // render function

    }, {
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_f42965b0e53211e9a0cfefb808989d41());
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * calculate if we have a header image.
           */
          hasImage: {
            type: Boolean,
            observer: "_hasImageChanged",
            reflectToAttribute: true
          },

          /**
           * editting state for the page
           */
          editMode: {
            type: Boolean,
            reflectToAttribute: true,
            value: false
          },
          image: {
            type: String
          }
        };
      }
      /**
       * Ready life cycle
       */

    }]);

    function SimpleBlogPost() {
      var _this;

      babelHelpers.classCallCheck(this, SimpleBlogPost);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(SimpleBlogPost).call(this));
      new Promise(function (res, rej) {
        return _require.default(["../../haxcms-elements/lib/ui-components/active-item/site-active-title.js"], res, rej);
      });
      _this.__disposer = [];
      (0, _renderStatus.afterNextRender)(babelHelpers.assertThisInitialized(_this), function () {
        var _this2 = this;

        window.addEventListener("scroll", this._scrollListener.bind(this));
        (0, _mobxModule.autorun)(function (reaction) {
          var fields = (0, _mobxModule.toJS)(_haxcmsSiteStore.store.activeItemFields);
          _this2.hasImage = _this2._computeHasImage(fields);

          if (_this2.hasImage) {
            _this2.image = fields.images[0].src;
          }

          _this2.__disposer.push(reaction);
        });
      });
      return _this;
    }
    /**
     * Detatched life cycle
     */


    babelHelpers.createClass(SimpleBlogPost, [{
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        window.removeEventListener("scroll", this._scrollListener.bind(this));

        for (var i in this.__disposer) {
          this.__disposer[i].dispose();
        }

        babelHelpers.get(babelHelpers.getPrototypeOf(SimpleBlogPost.prototype), "disconnectedCallback", this).call(this);
      }
      /**
       * Scroll event listener to do image effects.
       */

    }, {
      key: "_scrollListener",
      value: function _scrollListener(e) {
        if (this.hasImage) {
          var top = (window.pageYOffset || document.scrollTop) - (document.clientTop || 0);

          if (top < 0 || top > 1500) {
            return;
          }

          this.shadowRoot.querySelector("#image").style.transform = "translate3d(0px, " + top / 3 + "px, 0px)";
          this.shadowRoot.querySelector("#image").style.opacity = 1 - Math.max(top / 700, 0);
        }
      }
      /**
       * image has changed, ensure we set padding appropriately.
       */

    }, {
      key: "_hasImageChanged",
      value: function _hasImageChanged(newValue, oldValue) {
        var _this3 = this;

        if (newValue) {
          _async.microTask.run(function () {
            var rect = _this3.shadowRoot.querySelector("#image").getBoundingClientRect();

            _this3.$.contentcontainer.style.paddingTop = rect.height + "px";
          });
        } else {
          this.$.contentcontainer.style.paddingTop = null;
        }
      }
      /**
       * see if there's an image in the metadata blob.
       */

    }, {
      key: "_computeHasImage",
      value: function _computeHasImage(fields) {
        if (fields && babelHelpers.typeof(fields.images) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined)) && babelHelpers.typeof(fields.images[0]) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined)) && babelHelpers.typeof(fields.images[0].src) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          return true;
        }

        return false;
      }
    }]);
    return SimpleBlogPost;
  }(_polymerElement.PolymerElement);

  _exports.SimpleBlogPost = SimpleBlogPost;
  window.customElements.define(SimpleBlogPost.tag, SimpleBlogPost);
});
