// @lrnwebcomponents/simple-blog/lib/simple-blog-listing.js
define(["exports", "require", "../../../@polymer/polymer/polymer-element.js", "../../haxcms-elements/lib/ui-components/query/site-query.js", "../../../@polymer/iron-list/iron-list.js", "../../../@polymer/polymer/lib/utils/render-status.js"], function (_exports, _require, _polymerElement, _siteQuery, _ironList, _renderStatus) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SimpleBlogListing = void 0;
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject_fee34420b46e11e9978b67ee5ebb8bc8() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: block;\n        }\n        iron-list {\n          width: 100%;\n          max-width: 640px;\n          margin: 0 auto;\n          -webkit-box-sizing: border-box;\n          -moz-box-sizing: border-box;\n          box-sizing: border-box;\n        }\n        @media only screen and (max-width: 800px) {\n          iron-list {\n            padding: 0 32px;\n          }\n        }\n        simple-blog-overview:not(:defined) {\n          display: none;\n        }\n        simple-blog-overview {\n          width: 100%;\n          border: 1px solid #f2f2f0;\n        }\n      </style>\n      <site-query\n        result=\"{{__items}}\"\n        sort='{\"metadata.created\": \"DESC\"}'\n      ></site-query>\n      <iron-list items=\"[[__items]]\">\n        <template>\n          <simple-blog-overview\n            item-id=\"[[item.id]]\"\n            title=\"[[item.title]]\"\n            description=\"[[item.description]]\"\n            link=\"[[item.location]]\"\n            changed=\"[[item.metadata.created]]\"\n          ></simple-blog-overview>\n        </template>\n      </iron-list>\n    "]);

    _templateObject_fee34420b46e11e9978b67ee5ebb8bc8 = function _templateObject_fee34420b46e11e9978b67ee5ebb8bc8() {
      return data;
    };

    return data;
  }

  /**
   * `simple-blog-listing`
   * `A simple blog and associated elements`
   */
  var SimpleBlogListing =
  /*#__PURE__*/
  function (_PolymerElement) {
    babelHelpers.inherits(SimpleBlogListing, _PolymerElement);
    babelHelpers.createClass(SimpleBlogListing, null, [{
      key: "tag",

      /**
       * Store the tag name to make it easier to obtain directly.
       * @notice function name must be here for tooling to operate correctly
       */
      get: function get() {
        return "simple-blog-listing";
      }
    }]);

    function SimpleBlogListing() {
      var _this;

      babelHelpers.classCallCheck(this, SimpleBlogListing);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(SimpleBlogListing).call(this));
      new Promise(function (res, rej) {
        return _require.default(["./simple-blog-overview.js"], res, rej);
      });
      return _this;
    }

    babelHelpers.createClass(SimpleBlogListing, [{
      key: "connectedCallback",
      value: function connectedCallback() {
        babelHelpers.get(babelHelpers.getPrototypeOf(SimpleBlogListing.prototype), "connectedCallback", this).call(this);
        (0, _renderStatus.afterNextRender)(this, function () {
          this.shadowRoot.querySelector("iron-list").dispatchEvent(new CustomEvent("iron-resize", {
            bubbles: true,
            composed: true,
            cancelable: false,
            detail: true
          }));
        });
      } // render function

    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_fee34420b46e11e9978b67ee5ebb8bc8());
      }
    }]);
    return SimpleBlogListing;
  }(_polymerElement.PolymerElement);

  _exports.SimpleBlogListing = SimpleBlogListing;
  window.customElements.define(SimpleBlogListing.tag, SimpleBlogListing);
});
