// @material/mwc-button/mwc-button-base.js
define(["exports", "../../tslib/tslib.es6.js", "../mwc-icon/mwc-icon.js", "../mwc-ripple/mwc-ripple.js", "../mwc-ripple/ripple-handlers.js", "../../lit-element/lit-element.js", "../../lit-html/directives/class-map.js"], function (_exports, _tslibEs, _mwcIcon, _mwcRipple, _rippleHandlers, _litElement, _classMap) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ButtonBase = void 0;

  /**
  @license
  Copyright 2019 Google Inc. All Rights Reserved.
  
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at
  
      http://www.apache.org/licenses/LICENSE-2.0
  
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  */

  /** @soyCompatible */
  class ButtonBase extends _litElement.LitElement {
    constructor() {
      super(...arguments);
      this.raised = false;
      this.unelevated = false;
      this.outlined = false;
      this.dense = false;
      this.disabled = false;
      this.trailingIcon = false;
      this.fullwidth = false;
      this.icon = '';
      this.label = '';
      this.shouldRenderRipple = false;
      this.rippleHandlers = new _rippleHandlers.RippleHandlers(() => {
        this.shouldRenderRipple = true;
        return this.ripple;
      });
    }
    /** @soyCompatible */


    renderRipple() {
      const filled = this.raised || this.unelevated;
      return (0, _litElement.html)`${this.shouldRenderRipple ? (0, _litElement.html)`<mwc-ripple .primary="${!filled}" .disabled="${this.disabled}"></mwc-ripple>` : ''}`;
    }

    createRenderRoot() {
      return this.attachShadow({
        mode: 'open',
        delegatesFocus: true
      });
    }

    focus() {
      const buttonElement = this.buttonElement;

      if (buttonElement) {
        this.rippleHandlers.startFocus();
        buttonElement.focus();
      }
    }

    blur() {
      const buttonElement = this.buttonElement;

      if (buttonElement) {
        this.rippleHandlers.endFocus();
        buttonElement.blur();
      }
    }
    /** @soyCompatible */


    render() {
      /** @classMap */
      const classes = {
        'mdc-button--raised': this.raised,
        'mdc-button--unelevated': this.unelevated,
        'mdc-button--outlined': this.outlined,
        'mdc-button--dense': this.dense
      };
      return (0, _litElement.html)`
      <button
          id="button"
          class="mdc-button ${(0, _classMap.classMap)(classes)}"
          ?disabled="${this.disabled}"
          aria-label="${this.label || this.icon}"
          @focus="${this.handleRippleFocus}"
          @blur="${this.handleRippleBlur}"
          @mousedown="${this.handleRippleActivate}"
          @mouseenter="${this.handleRippleMouseEnter}"
          @mouseleave="${this.handleRippleMouseLeave}"
          @touchstart="${this.handleRippleActivate}"
          @touchend="${this.handleRippleDeactivate}"
          @touchcancel="${this.handleRippleDeactivate}">
        ${this.renderRipple()}
        <span class="leading-icon">
          <slot name="icon">
            ${this.icon && !this.trailingIcon ? this.renderIcon() : ''}
          </slot>
        </span>
        <span class="mdc-button__label">${this.label}</span>
        <slot></slot>
        <span class="trailing-icon">
          <slot name="trailingIcon">
            ${this.icon && this.trailingIcon ? this.renderIcon() : ''}
          </slot>
        </span>
      </button>`;
    }
    /** @soyCompatible */


    renderIcon() {
      return (0, _litElement.html)`
    <mwc-icon class="mdc-button__icon">
      ${this.icon}
    </mwc-icon>`;
    }

    handleRippleActivate(evt) {
      const onUp = () => {
        window.removeEventListener('mouseup', onUp);
        this.handleRippleDeactivate();
      };

      window.addEventListener('mouseup', onUp);
      this.rippleHandlers.startPress(evt);
    }

    handleRippleDeactivate() {
      this.rippleHandlers.endPress();
    }

    handleRippleMouseEnter() {
      this.rippleHandlers.startHover();
    }

    handleRippleMouseLeave() {
      this.rippleHandlers.endHover();
    }

    handleRippleFocus() {
      this.rippleHandlers.startFocus();
    }

    handleRippleBlur() {
      this.rippleHandlers.endFocus();
    }

  }

  _exports.ButtonBase = ButtonBase;
  (0, _tslibEs.__decorate)([(0, _litElement.property)({
    type: Boolean
  })], ButtonBase.prototype, "raised", void 0);
  (0, _tslibEs.__decorate)([(0, _litElement.property)({
    type: Boolean
  })], ButtonBase.prototype, "unelevated", void 0);
  (0, _tslibEs.__decorate)([(0, _litElement.property)({
    type: Boolean
  })], ButtonBase.prototype, "outlined", void 0);
  (0, _tslibEs.__decorate)([(0, _litElement.property)({
    type: Boolean
  })], ButtonBase.prototype, "dense", void 0);
  (0, _tslibEs.__decorate)([(0, _litElement.property)({
    type: Boolean,
    reflect: true
  })], ButtonBase.prototype, "disabled", void 0);
  (0, _tslibEs.__decorate)([(0, _litElement.property)({
    type: Boolean,
    attribute: 'trailingicon'
  })], ButtonBase.prototype, "trailingIcon", void 0);
  (0, _tslibEs.__decorate)([(0, _litElement.property)({
    type: Boolean,
    reflect: true
  })], ButtonBase.prototype, "fullwidth", void 0);
  (0, _tslibEs.__decorate)([(0, _litElement.property)({
    type: String
  })], ButtonBase.prototype, "icon", void 0);
  (0, _tslibEs.__decorate)([(0, _litElement.property)({
    type: String
  })], ButtonBase.prototype, "label", void 0);
  (0, _tslibEs.__decorate)([(0, _litElement.query)('#button')], ButtonBase.prototype, "buttonElement", void 0);
  (0, _tslibEs.__decorate)([(0, _litElement.queryAsync)('mwc-ripple')], ButtonBase.prototype, "ripple", void 0);
  (0, _tslibEs.__decorate)([(0, _litElement.internalProperty)()], ButtonBase.prototype, "shouldRenderRipple", void 0);
  (0, _tslibEs.__decorate)([(0, _litElement.eventOptions)({
    passive: true
  })], ButtonBase.prototype, "handleRippleActivate", null);
});
