// @lrnwebcomponents/paper-audio-player/paper-audio-player.js
define(["exports","../../@polymer/polymer/polymer-element.js","../../@polymer/polymer/lib/utils/render-status.js","../../@polymer/iron-a11y-keys-behavior/iron-a11y-keys-behavior.js","../../@polymer/polymer/lib/legacy/class.js","../../@polymer/paper-progress/paper-progress.js","../../@polymer/iron-icon/iron-icon.js","../../@polymer/paper-icon-button/paper-icon-button.js","../simple-colors/lib/simple-colors-polymer.js","../schema-behaviors/schema-behaviors.js","../../@polymer/iron-iconset-svg/iron-iconset-svg.js","./lib/paper-audio-icons.js"],(function(_exports,_polymerElement,_renderStatus,_ironA11yKeysBehavior,_class,_paperProgress,_ironIcon,_paperIconButton,_simpleColorsPolymer,_schemaBehaviors,_ironIconsetSvg,_paperAudioIcons){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.PaperAudioPlayer=void 0;class PaperAudioPlayer extends((0,_class.mixinBehaviors)([_ironA11yKeysBehavior.IronA11yKeysBehavior],(0,_schemaBehaviors.SchemaBehaviors)(_simpleColorsPolymer.SimpleColorsPolymer))){static get template(){return _polymerElement.html`
      <style include="simple-colors-shared-styles-polymer">
        :host {
          display: block;
          box-sizing: border-box;
          font-family: var(
            --paper-audio-player-font-family,
            "Roboto",
            "Noto",
            sans-serif
          );
          --paper-audio-player-color: var(
            --simple-colors-default-theme-accent-9
          );
          --paper-audio-player-text-color: var(
            --simple-colors-default-theme-grey-1
          );
          --paper-audio-player-shadow: var(
            --simple-colors-default-theme-grey-2
          );
          --paper-audio-player-background: var(
            --simple-colors-default-theme-grey-1
          );
          background-color: var(--paper-audio-player-background);
        }
        #wrapper {
          position: relative;
          cursor: pointer;
          height: 50px;
          box-shadow: 0 1px 2px var(--paper-audio-player-shadow);
        }
        #left,
        #center {
          border-right: 1px solid var(--paper-audio-player-shadow);
        }
        #left,
        #right {
          height: 50px;
          width: 50px;
          position: relative;
        }

        #left {
          opacity: 0.8;
          background-color: var(--paper-audio-player-color);
          transition: opacity 0.25s;
        }
        #left:focus,
        #left:hover {
          opacity: 1;
        }

        #right {
          background-color: var(--paper-audio-player-background);
        }

        paper-icon-button,
        iron-icon {
          color: var(--paper-audio-player-text-color);
        }

        #duration,
        #title,
        #progress2 {
          text-align: center;
          line-height: 50px;
        }

        #duration {
          font-size: 11px;
          color: var(--paper-audio-player-color);
        }

        paper-icon-button,
        iron-icon {
          margin: auto;
        }

        #replay {
          opacity: 0;
          color: var(--paper-audio-player-color);
        }

        #title,
        #progress2 {
          pointer-events: none;
          font-size: 15px;
        }

        #title {
          z-index: 2;
          color: var(--paper-audio-player-color);
        }

        #progress2 {
          width: 0px;
          z-index: 5;
          color: var(--paper-audio-player-text-color);
          overflow: hidden;
        }

        #center {
          position: relative;
          overflow: hidden;
          background-color: var(--paper-audio-player-background);
        }

        #progress {
          width: 100%;
          transform-origin: left;
          transform: scaleX(0);
          background-color: var(--paper-audio-player-color);
        }

        /* On hover */

        :host(:not(.cantplay)) #right:hover #replay {
          opacity: 1;
        }

        #right:hover #duration {
          opacity: 0;
        }

        #left:hover #play,
        #left:hover #pause {
          transform: scale3d(1.1, 1.1, 1.1);
          -ms-transform: scale3d(1.1, 1.1, 1.1);
          -webkit-transform: scale3d(1.1, 1.1, 1.1);
        }

        /* On Error */

        :host(.cantplay) #title {
          font-size: 12px;
        }

        :host(.cantplay) #wrapper {
          cursor: default;
        }

        :host(.cantplay) #play {
          opacity: 0;
        }

        /* Flexbox Helpers */

        .layout-horizontal {
          display: flex;
          display: -webkit-flex;
          display: -ms-flexbox;
          -ms-flex-direction: row;
          -webkit-flex-direction: row;
          flex-direction: row;
        }

        .flex {
          -ms-flex: 1;
          -webkit-flex: 1;
          flex: 1;
        }

        .fit {
          position: absolute;
          top: 0;
          right: 0;
          bottom: 0;
          left: 0;
        }

        .self-start {
          -ms-align-self: flex-start;
          -webkit-align-self: flex-start;
          align-self: flex-start;
        }

        .self-end {
          -ms-align-self: flex-end;
          -webkit-align-self: flex-end;
          align-self: flex-end;
        }
      </style>
      <div id="wrapper" class="layout-horizontal">
        <div id="left" class="self-start" on-click="playPause">
          <!-- Icon -->
          <paper-icon-button
            id="play"
            icon="paper-audio-icons:play-arrow"
            class="fit"
            hidden$="[[ _hidePlayIcon(isPlaying, canBePlayed) ]]"
            role="button"
            aria-label="Play Audio"
            tabindex="-1"
          ></paper-icon-button>
          <paper-icon-button
            id="pause"
            icon="paper-audio-icons:pause"
            class="fit"
            hidden$="[[ !isPlaying ]]"
            role="button"
            aria-label="Pause Audio"
            tabindex="-1"
          ></paper-icon-button>
          <iron-icon
            id="error"
            icon="paper-audio-icons:error-outline"
            class="fit"
            hidden$="[[ !error ]]"
          ></iron-icon>
        </div>
        <div id="center" class="flex" on-down="_onDown">
          <!-- Title -->
          <div id="title" class="fit" role="alert">[[title]]</div>
          <!-- Audio HTML5 element -->
          <audio
            id="audio"
            src$="[[src]]"
            preload$="[[ _setPreload(autoPlay, preload) ]]"
          ></audio>
          <!-- Progress bar -->
          <div id="progress" class="fit"></div>
          <!-- Secondary white title -->
          <div id="progress2" class="fit">
            <div id="title2" aria-hidden="true">[[title]]</div>
          </div>
        </div>
        <div id="right" class="self-end" on-click="restart">
          <!-- Duration -->
          <div id="duration" class="fit" hidden\$="[[ended]]">
            <span class="fit" role="timer" aria-label="Audio Track Length"
              >[[ _convertSecToMin(timeLeft) ]]</span
            >
          </div>
          <!-- Icon -->
          <paper-icon-button
            id="replay"
            class="fit"
            icon="paper-audio-icons:replay"
            tabindex="-1"
            role="button"
            aria-label="Replay Audio"
          ></paper-icon-button>
        </div>
      </div>
    `}static get tag(){return"paper-audio-player"}static get properties(){return{...super.properties,src:{type:String,observer:"_srcChanged"},title:{type:String,value:"Click to play this audio file"},autoPlay:{type:Boolean,value:!1},preload:{type:String,value:"auto"},currentTime:{type:Number,value:0,notify:!0},timeLeft:{type:Number,value:0},smallSkip:{type:Number,value:15},largeSkip:{type:Number,value:60},error:{type:Boolean},timeOffset:{type:Number,value:0}}}static get keyBindings(){return{space:"playPause",enter:"playPause",left:"_skipReverseByInterval",right:"_skipReverseByInterval",down:"_skipReverseByInterval",up:"_skipReverseByInterval"}}static get haxProperties(){return{canScale:!0,canPosition:!0,canEditSource:!1,gizmo:{title:"Mini Audio player",description:"A very small audio player good for MP3s.",icon:"image:music-note",color:"green",groups:["Audio","Media"],handles:[{type:"audio",source:"src",title:"title",color:"color"}],meta:{author:"ELMS:LN"}},settings:{quick:[{property:"src",title:"Source",description:"The URL for this audio file.",inputMethod:"textfield",icon:"link",required:!0,validationType:"url"},{property:"title",title:"Title",description:"Title of this sound track.",inputMethod:"textfield",icon:"av:video-label",required:!1,validationType:"text"},{property:"accentColor",title:"Accent color",description:"Select the accent color use",inputMethod:"colorpicker",icon:"editor:format-color-fill"},{property:"dark",title:"Dark",description:"Use dark theme",inputMethod:"boolean",icon:"invert-colors"}],configure:[{property:"src",title:"Source",description:"The URL for this audio file.",inputMethod:"haxupload",icon:"link",required:!0,validationType:"url"},{property:"title",title:"Title",description:"Title of this sound track.",inputMethod:"textfield",icon:"av:video-label",required:!1,validationType:"text"},{property:"accentColor",title:"Accent color",description:"Select the accent color use",inputMethod:"colorpicker",icon:"editor:format-color-fill"},{property:"dark",title:"Dark",description:"Use dark theme",inputMethod:"boolean",icon:"invert-colors"}],advanced:[]}}}connectedCallback(){super.connectedCallback(),(0,_renderStatus.afterNextRender)(this,(function(){this.shadowRoot.querySelector("#audio").addEventListener("loadedmetadata",this._onCanPlay.bind(this)),this.shadowRoot.querySelector("#audio").addEventListener("playing",this._onPlaying.bind(this)),this.shadowRoot.querySelector("#audio").addEventListener("pause",this._onPause.bind(this)),this.shadowRoot.querySelector("#audio").addEventListener("ended",this._onEnd.bind(this)),this.shadowRoot.querySelector("#audio").addEventListener("error",this._onError.bind(this))})),this.setAttribute("tabindex","0"),this.setAttribute("role","application"),this.setAttribute("aria-label","Audio Player"),this.setAttribute("aria-describedby","title")}disconnectedCallback(){super.disconnectedCallback(),this.shadowRoot.querySelector("#audio").removeEventListener("loadedmetadata",this._onCanPlay.bind(this)),this.shadowRoot.querySelector("#audio").removeEventListener("playing",this._onPlaying.bind(this)),this.shadowRoot.querySelector("#audio").removeEventListener("pause",this._onPause.bind(this)),this.shadowRoot.querySelector("#audio").removeEventListener("ended",this._onEnd.bind(this)),this.shadowRoot.querySelector("#audio").removeEventListener("error",this._onError.bind(this))}ready(){super.ready();this.canBePlayed=!1,this.isPlaying=!1,this.ended=!1,this.error=!1,this.shadowRoot.querySelector("#audio").currentTime=this.timeOffset}playPause(e){e&&e.preventDefault();this.canBePlayed?this.isPlaying?this._pause():this._play():"none"===this.preload&&(this.shadowRoot.querySelector("#audio").load(),this._play())}_play(){this.shadowRoot.querySelector("#audio").play()}_pause(){this.shadowRoot.querySelector("#audio").pause()}restart(e){e&&e.preventDefault();this.shadowRoot.querySelector("#audio").currentTime=0,this.isPlaying||this._play()}_onCanPlay(){if(this.canBePlayed=!0,this.timeLeft=this.shadowRoot.querySelector("#audio").duration,this.timeOffset>0){var percentagePlayed=this.timeOffset/this.shadowRoot.querySelector("#audio").duration;this._updateVisualProgress(percentagePlayed)}this.autoPlay&&this._play()}_onPlaying(){this.ended=!1,this.isPlaying=!0,this.shadowRoot.querySelector("#replay").style="",this._startProgressTimer()}_skipReverseByInterval(e){e&&e.preventDefault();var newTime=0;switch(e.detail.key){case"up":this.largeSkip<this.timeLeft&&(newTime=this.currentTime+this.largeSkip);break;case"down":this.currentTime-this.largeSkip>0&&(newTime=this.currentTime-this.largeSkip);break;case"right":this.smallSkip<this.timeLeft&&(newTime=this.currentTime+this.smallSkip);break;default:this.currentTime-this.smallSkip>0&&(newTime=this.currentTime-this.smallSkip)}this._updatePlayPosition(newTime),this.isPlaying||this._play()}_startProgressTimer(){var player=this;player.timer={},player.timer.sliderUpdateInterval&&clearInterval(player.timer.sliderUpdateInterval),player.timer.sliderUpdateInterval=setInterval((function(){if(player.isPlaying){player.currentTime=player.shadowRoot.querySelector("#audio").currentTime,player.timeLeft=player.shadowRoot.querySelector("#audio").duration-player.currentTime;var percentagePlayed=player.currentTime/player.shadowRoot.querySelector("#audio").duration;player._updateVisualProgress(percentagePlayed)}else clearInterval(player.timer.sliderUpdateInterval)}),60)}_onPause(){this.isPlaying=!1}_onEnd(){this.ended=!0,this.isPlaying=!1,this.shadowRoot.querySelector("#replay").style.opacity=1}_onError(){this.classList.add("cantplay"),this.title="Sorry, can't play track: "+this.title,this.error=!0,this.setAttribute("aria-invalid","true")}_convertSecToMin(seconds){if(0===seconds)return"";var minutes=Math.floor(seconds/60),secondsToCalc=Math.floor(seconds%60)+"";return minutes+":"+(secondsToCalc.length<2?"0"+secondsToCalc:secondsToCalc)}_onDown(e){e.preventDefault();var player=this;player.canBePlayed?(player._updateProgressBar(e),player.isPlaying||player._play()):"none"===player.preload&&(player.shadowRoot.querySelector("#audio").load(),player.shadowRoot.querySelector("#audio").addEventListener("loadedmetadata",(function(){player._updateProgressBar(e),player.isPlaying||player._play()}),!1))}_updateProgressBar(e){var r=(e.detail.x-this.shadowRoot.querySelector("#center").getBoundingClientRect().left)/this.shadowRoot.querySelector("#center").getBoundingClientRect().width*this.shadowRoot.querySelector("#audio").duration;this._updatePlayPosition(r)}_updatePlayPosition(newTime){this.currentTime=this.shadowRoot.querySelector("#audio").currentTime=newTime;var percentagePlayed=this.currentTime/this.shadowRoot.querySelector("#audio").duration;this._updateVisualProgress(percentagePlayed)}_updateVisualProgress(percentagePlayed){this.shadowRoot.querySelector("#progress").style.transform="scaleX("+percentagePlayed+")",this.shadowRoot.querySelector("#progress2").style.width=100*percentagePlayed+"%",this.shadowRoot.querySelector("#title2").style.width=1/percentagePlayed*100+"%"}_srcChanged(newValue,oldValue){this.isPlaying&&(this._pause(),this._play())}_changeColor(newValue){this.shadowRoot.querySelector("#left").style.backgroundColor=newValue,this.shadowRoot.querySelector("#title").style.color=newValue,this.shadowRoot.querySelector("#duration").style.color=newValue,this.shadowRoot.querySelector("#progress").style.backgroundColor=newValue,this.shadowRoot.querySelector("#replay").style.color=newValue}_hidePlayIcon(isPlaying,canBePlayed){return!!isPlaying||!(canBePlayed||"none"===this.preload)}_setPreload(autoplay,preload){return autoplay?"auto":preload}}_exports.PaperAudioPlayer=PaperAudioPlayer,window.customElements.define(PaperAudioPlayer.tag,PaperAudioPlayer)}));
