// @lrnwebcomponents/a11y-media-player/a11y-media-player.js
define(["exports","meta","require","../../@polymer/polymer/polymer-element.js","../../@polymer/polymer/lib/utils/render-status.js","./lib/a11y-media-behaviors.js","../simple-colors/lib/simple-colors-polymer.js","../es-global-bridge/es-global-bridge.js","../anchor-behaviors/anchor-behaviors.js","../../@polymer/paper-toast/paper-toast.js","./lib/a11y-media-state-manager.js","./lib/a11y-media-controls.js","./lib/a11y-media-button.js","./lib/a11y-media-html5.js","./lib/a11y-media-transcript.js","./lib/a11y-media-transcript-controls.js","./lib/a11y-media-youtube.js"],(function(_exports,meta,_require,_polymerElement,_renderStatus,_a11yMediaBehaviors,_simpleColorsPolymer,_esGlobalBridge,_anchorBehaviors,_paperToast,_a11yMediaStateManager,_a11yMediaControls,_a11yMediaButton,_a11yMediaHtml,_a11yMediaTranscript,_a11yMediaTranscriptControls,_a11yMediaYoutube){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.A11yMediaPlayer=void 0,meta=babelHelpers.interopRequireWildcard(meta),_require=babelHelpers.interopRequireWildcard(_require);
/**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class A11yMediaPlayer extends _a11yMediaBehaviors.A11yMediaBehaviors{static get template(){return _polymerElement.html`
      <style>
        :host {
          display: block;
          width: calc(100% - 2px);
          border: 1px solid var(--simple-colors-default-theme-grey-3);
          --a11y-media-color: var(--simple-colors-default-theme-grey-11);
          --a11y-media-bg-color: var(--simple-colors-default-theme-grey-2);
          --a11y-media-hover-color: var(--simple-colors-default-theme-grey-12);
          --a11y-media-hover-bg-color: var(
            --simple-colors-default-theme-grey-2
          );
          --a11y-media-accent-color: var(
            --simple-colors-default-theme-accent-9
          );
          --a11y-media-faded-accent-color: var(
            --simple-colors-default-theme-accent-8
          );
          --paper-toast-color: var(--simple-colors-default-theme-grey-11);
          --paper-toast-background-color: var(
            --simple-colors-default-theme-grey-2
          );

          --a11y-media-settings-menu-color: var(--a11y-media-color);
          --a11y-media-settings-menu-bg-color: var(--a11y-media-bg-color);
          --a11y-media-settings-menu-hover-color: var(--a11y-media-hover-color);
          --a11y-media-settings-menu-hover-bg-color: var(
            --a11y-media-hover-bg-color
          );

          --a11y-media-button-color: var(--a11y-media-color);
          --a11y-media-button-bg-color: var(--a11y-media-bg-color);
          --a11y-media-button-hover-color: var(--a11y-media-accent-color);
          --a11y-media-button-hover-bg-color: var(--a11y-media-hover-bg-color);
          --a11y-media-button-toggle-color: var(
            --a11y-media-faded-accent-color
          );

          --paper-toggle-button-unchecked-bar-color: var(--a11y-media-color);
          --paper-toggle-button-unchecked-button-color: var(--a11y-media-color);
          --paper-toggle-button-checked-bar-color: var(
            --a11y-media-accent-color
          );
          --paper-toggle-button-checked-button-color: var(
            --a11y-media-accent-color
          );

          --paper-slider-active-color: var(--a11y-media-accent-color);
          --paper-slider-secondary-color: var(--a11y-media-faded-accent-color);
          --paper-slider-pin-color: var(--a11y-media-faded-bg-color);
          --paper-slider-pin-start-color: var(--a11y-media-faded-bg-color);
          --paper-slider-pin-end-color: var(--a11y-media-faded-bg-color);
          --paper-slider-knob-color: var(--a11y-media-accent-color);
          --paper-slider-knob-start-color: var(--a11y-media-bg-color);
          --paper-slider-knob-end-color: var(--a11y-media-bg-color);
          --paper-slider-knob-border-color: var(--a11y-media-accent-color);
          --paper-slider-knob-start-border-color: var(--a11y-media-bg-color);
          --paper-slider-knob-end-border-color: var(--a11y-media-bg-color);

          --a11y-media-transcript-color: var(
            --simple-colors-default-theme-grey-7
          );
          --a11y-media-transcript-bg-color: var(
            --simple-colors-default-theme-grey-1
          );
          --a11y-media-transcript-accent-color: var(
            --simple-colors-default-theme-accent-8
          );
          --a11y-media-transcript-faded-accent-color: var(
            --simple-colors-default-theme-accent-10
          );
          --a11y-media-transcript-cue-color: var(
            --simple-colors-fixed-theme-grey-12
          );
          --a11y-media-transcript-cue-bg-color: var(
            --simple-colors-fixed-theme-grey-1
          );
          --a11y-media-transcript-active-cue-color: var(
            --simple-colors-fixed-theme-grey-12
          );
          --a11y-media-transcript-active-cue-bg-color: var(
            --simple-colors-fixed-theme-accent-1
          );
          --a11y-media-transcript-focused-cue-color: var(
            --simple-colors-fixed-theme-grey-12
          );
          --a11y-media-transcript-focused-cue-bg-color: var(
            --simple-colors-fixed-theme-grey-2
          );
          --a11y-media-transcript-match-color: var(
            --simple-colors-fixed-theme-grey-1
          );
          --a11y-media-transcript-match-bg-color: var(
            --simple-colors-fixed-theme-accent-10
          );
          --a11y-media-transcript-match-border-color: var(
            --simple-colors-fixed-theme-accent-12
          );
        }
        :host([dark]) {
          border: 1px solid var(--simple-colors-default-theme-grey-1);
        }
        :host([dark-transcript]) {
          --a11y-media-transcript-bg-color: var(
            --simple-colors-dark-theme-grey-1
          );
          --a11y-media-transcript-cue-color: var(
            --simple-colors-dark-theme-grey-12
          );
          --a11y-media-transcript-cue-bg-color: var(
            --simple-colors-dark-theme-grey-1
          );
          --a11y-media-transcript-active-cue-color: var(
            --simple-colors-dark-theme-accent-10
          );
          --a11y-media-transcript-active-cue-bg-color: var(
            --simple-colors-dark-theme-grey-1
          );
          --a11y-media-transcript-match-color: var(
            --simple-colors-dark-theme-grey-1
          );
          --a11y-media-transcript-match-bg-color: var(
            --simple-colors-dark-theme-accent-10
          );
          --a11y-media-transcript-match-border-color: var(
            --simple-colors-dark-theme-accent-12
          );
          --a11y-media-transcript-focused-cue-color: var(
            --simple-colors-dark-theme-grey-12
          );
          --a11y-media-transcript-focused-cue-bg-color: var(
            --simple-colors-dark-theme-grey-2
          );
        }
        :host,
        :host #outerplayer {
          color: var(--simple-colors-default-theme-grey-12);
          background-color: var(--simple-colors-default-theme-grey-2);
        }
        :host > * {
          transition: all 0.5s;
        }
        :host,
        :host #outerplayer,
        :host #player,
        :host #outertranscript,
        :host #innertranscript {
          display: flex;
          flex-flow: column;
          align-items: stretch;
          align-content: stretch;
        }
        :host #captionlink:link {
          text-decoration: none;
        }
        :host #innerplayer {
          display: flex;
        }
        :host([hidden]),
        :host *[hidden] {
          display: none !important;
        }
        :host #innerplayer,
        :host #player,
        :host #player > *,
        :host #customcc,
        :host #customcctxt,
        :host #slider,
        :host #controls,
        :host #outertranscript,
        :host #innertranscript {
          width: 100%;
        }
        :host #innertranscript > * {
          width: calc(100% - 1px);
        }
        :host > *,
        :host #innerplayer,
        :host #player,
        :host #player > *,
        :host #customcctxt {
          flex: 1 1 auto;
        }
        :host #controls,
        :host #tcontrols {
          flex: 0 0 44px;
        }
        :host #innerplayer {
          margin: 0 auto;
        }
        :host #player {
          height: 400px;
          position: relative;
        }
        :host #player > * {
          position: absolute;
          top: 0;
          left: 0;
          height: 100%;
        }
        :host #playbutton,
        :host #slider,
        :host #controls {
          z-index: 2;
        }
        :host([audio-only]) #playbutton {
          opacity: 0;
        }
        :host #slider {
          flex: 0 0 32px;
          height: 32px;
        }
        :host([thumbnail-src]) #youtube {
          opacity: 0;
        }
        :host #youtube[elapsed] {
          opacity: 1;
          transition: opacity 0.5s;
        }
        :host #customcc:not([hidden]) {
          font-size: 20px;
          transition: font-size 0.25s;
          display: flex;
        }
        :host #customcctxt:not(:empty) {
          align-self: flex-end;
          font-family: sans-serif;
          color: white;
          margin: 4px 10px;
          padding: 0.15em 4px;
          background-color: black;
          background-color: rgba(0, 0, 0, 0.8);
          transition: all 0.5s;
        }
        :host([audio-only]:not([thumbnail-src])) #customcctxt {
          align-self: center;
          color: var(--a11y-media-color);
          background-color: transparent;
        }
        :host #printthumb {
          width: 100%;
          margin: 0;
          display: block;
          border-top: 1px solid #aaaaaa;
        }
        :host .media-caption:not(:empty) {
          width: calc(100% - 30px);
          padding: 5px 15px;
        }
        :host .media-type {
          font-style: italic;
        }
        :host #outertranscript {
          padding: 0 1px 0 0;
        }
        :host #innertranscript {
          flex: 1 0 194px;
        }
        :host #transcript {
          flex: 1 0 150px;
          overflow-y: scroll;
        }
        :host .sr-only {
          position: absolute;
          left: -9999px;
          font-size: 0;
          height: 0;
          width: 0;
          overflow: hidden;
        }
        @media screen {
          :host([flex-layout]:not([responsive-size="xs"])) {
            flex-flow: row;
            padding: 0;
          }
          :host([flex-layout]:not([responsive-size="xs"])) #outerplayer {
            flex: 1 0 auto;
          }
          :host #printthumb,
          :host([height]) #outertranscript,
          :host([stand-alone]) #outertranscript,
          :host([hide-transcript]) #outertranscript {
            display: none;
          }
          :host([sticky]:not([sticky-corner="none"])) #outerplayer {
            position: fixed;
            top: 5px;
            right: 5px;
            width: 200px;
            max-width: 200px;
            z-index: 999999;
            border: 1px solid var(--a11y-media-bg-color);
            box-shadow: 1px 1px 20px 1px rgba(125, 125, 125);
            border-radius: 3.2px;
          }
          :host([dark][sticky]:not([sticky-corner="none"])) #outerplayer {
            border: 1px solid var(--a11y-media-bg-color);
          }
          :host([sticky][sticky-corner="top-left"]) #outerplayer {
            right: unset;
            left: 5px;
          }
          :host([flex-layout]:not([responsive-size="xs"])) > div {
            width: 50%;
            flex: 1 1 auto;
          }
          :host #innertranscript {
            position: relative;
          }
          :host([hide-transcript]) #outerplayer {
            min-width: 50%;
            max-width: 100%;
          }
          :host([hide-transcript]) #outertranscript {
            display: none;
          }
          :host(:not([no-height]):not([stacked-layout]):not([responsive-size="xs"]))
            #transcript {
            position: absolute;
            top: 44px;
            left: 0;
            right: 0;
            bottom: 0;
            overflow-y: scroll;
          }
          :host(:not([no-height]):not([stacked-layout]):not([responsive-size="xs"]))
            #innerplayer.totop {
            position: absolute;
            top: 0;
            left: 0;
            width: 200px !important;
            z-index: 9999;
          }
          :host([sticky][sticky-corner="bottom-left"]) #innerplayer {
            top: unset;
            right: unset;
            bottom: 5px;
          }
          :host([sticky][sticky-corner="bottom-right"]) #innerplayer {
            top: unset;
            bottom: 5px;
          }
          :host([sticky]:not([sticky-corner="none"]):not([no-height]):not([stacked-layout]):not([responsive-size="xs"]))
            #controls {
            display: none;
          }
          :host([responsive-size="lg"]) #customcc {
            font-size: 16px;
          }
          :host([responsive-size="md"]) #customcc,
          :host([flex-layout][responsive-size="xl"]) #customcc {
            font-size: 14px;
          }
          :host([responsive-size="sm"]) #customcc,
          :host([flex-layout][responsive-size="lg"]) #customcc {
            font-size: 12px;
          }
          :host([responsive-size="xs"]) #customcc,
          :host([flex-layout][responsive-size="md"]) #customcc,
          :host([flex-layout][responsive-size="sm"]) #customcc {
            font-size: 10px;
          }
          :host([sticky]:not([sticky-corner="none"])) #customcc {
            display: none;
          }
          :host .media-caption {
            color: var(--a11y-media-bg-color);
            background-color: var(--a11y-media-accent-color);
          }
          :host #audio-only {
            text-align: center;
            font-style: italic;
            width: 100%;
            line-height: 160%;
          }
          :host .print-only {
            display: none;
          }
        }

        @media print {
          :host,
          :host([dark]) {
            outline: 1px solid #aaaaaa;
            background-color: #ffffff;
          }
          :host .screen-only,
          :host #printthumb:not([src]),
          :host(:not([thumbnail-src])) #player {
            display: none;
          }
          :host #searchbar {
            display: none;
          }
          :host .media-caption {
            background-color: #cccccc;
            color: #000000;
            font-size: 120%;
          }
        }
      </style>
      <style include="simple-colors-shared-styles-polymer"></style>
      <div class="sr-only">
        <a href$="[[__captionHref]]">[[mediaCaption]]</a>
      </div>
      <div id="outerplayer">
        <div id="innerplayer">
          <div
            id="player"
            style$="[[_getThumbnailCSS(thumbnailSrc,isYoutube,audioOnly)]]"
          >
            <a11y-media-play-button
              id="playbutton"
              action$="[[playPause.action]]"
              audio-only$="[[audioOnly]]"
              disabled="true"
              elapsed$="[[_hidePlayButton(thumbnailSrc, isYoutube, __elapsed)]]"
              hidden$="[[audioNoThumb]]"
              disabled$="[[audioNoThumb]]"
              on-controls-change="_onControlsChanged"
              localization$="[[localization]]"
            >
            </a11y-media-play-button>
            <a11y-media-html5
              id="html5"
              audio-only$="[[audioOnly]]"
              autoplay$="[[autoplay]]"
              cc$="[[cc]]"
              crossorigin$="[[crossorigin]]"
              hidden$="[[isYoutube]]"
              media-lang$="[[mediaLang]]"
              loop$="[[loop]]"
              muted$="[[muted]]"
              manifest$="[[manifest]]"
              on-media-loaded="_handleMediaLoaded"
              ontimeupdate="_handleTimeUpdate"
              playing$="[[__playing]]"
              playback-rate$="[[playbackRate]]"
              thumbnail-src$="[[thumbnailSrc]]"
              preload$="[[preload]]"
              volume$="[[volume]]"
            >
              <slot></slot>
            </a11y-media-html5>
            <div
              id="youtube"
              elapsed$="[[__elapsed]]"
              lang$="[[mediaLang]]"
              video-id$="[[videoId]]"
            ></div>
            <div
              id="customcc"
              aria-live="polite"
              class="screen-only"
              hidden$="[[!showCustomCaptions]]"
            >
              <div id="customcctxt"></div>
            </div>
          </div>
        </div>
        <paper-slider
          id="slider"
          class="screen-only"
          disabled$="[[disableSeek]]"
          label$="[[seekSlider.label]]"
          min="0"
          max="[[__duration]]"
          on-mousedown="_handleSliderStart"
          on-mouseup="_handleSliderStop"
          on-keyup="_handleSliderStop"
          on-keydown="_handleSliderStart"
          on-blur="_handleSliderStop"
          secondary-progress="[[__buffered]]"
          value="[[__elapsed]]"
        >
        </paper-slider>
        <a11y-media-controls
          id="controls"
          cc$="[[cc]]"
          disable-seek$="[[disableSeek]]"
          fixed-height$="[[height]]"
          fullscreen$="[[fullscreen]]"
          fullscreen-button$="[[fullscreenButton]]"
          has-captions$="[[hasCaptions]]"
          has-transcript$="[[hasTranscript]]"
          hide-transcript$="[[hideTranscript]]"
          linkable$="[[linkable]]"
          mute-unmute="[[muteUnmute]]"
          on-controls-change="_onControlsChanged"
          on-print-transcript="_handlePrinting"
          on-download-transcript="_handleDownload"
          responsive-size$="[[responsiveSize]]"
          play-pause="[[playPause]]"
          stand-alone$="[[standAlone]]"
          volume="[[__volume]]"
        >
        </a11y-media-controls>
        <a id="captionlink" href$="[[__captionHref]]">
          <div
            aria-hidden="true"
            class="screen-only media-caption"
            hidden$="[[!_hasAttribute(mediaCaption)]]"
          >
            [[mediaCaption]]
          </div>
        </a>
        <div class="print-only media-caption">[[printCaption]]</div>
      </div>
      <img
        loading="lazy"
        id="printthumb"
        aria-hidden="true"
        src$="[[thumbnailSrc]]"
      />
      <div id="outertranscript" hidden$="[[standAlone]]">
        <div id="innertranscript" hidden$="[[hideTranscript]]">
          <a11y-media-transcript-controls
            id="tcontrols"
            accent-color$="[[accentColor]]"
            localization$="[[localization]]"
            dark$="[[darkTranscript]]"
            disable-print-button$="[[disablePrintButton]]"
            disable-scroll$="[[disableScroll]]"
            disable-search$="[[disableSearch]]"
            has-transcript$="[[hasTranscript]]"
            localization$="[[localization]]"
            on-searchbar-added="_handleSearchAdded"
            on-toggle-scroll="_handleTranscriptScrollToggle"
            on-print-transcript="_handlePrinting"
            on-download-transcript="_handleDownload"
            stand-alone$="[[standAlone]]"
          >
          </a11y-media-transcript-controls>
          <a11y-media-transcript
            id="transcript"
            accent-color$="[[accentColor]]"
            dark$="[[darkTranscript]]"
            disable-scroll$="[[disableScroll]]"
            disable-search$="[[disableSearch]]"
            disable-seek$="[[disableSeek]]"
            disable-interactive$="[[disableInteractive]]"
            hide-timestamps$="[[hideTimestamps]]"
            media-id$="[[id]]"
            on-transcript-seek="_handleTranscriptSeek"
            localization$="[[localization]]"
            search="[[search]]"
            selected-transcript$="[[__selectedTrack]]"
          >
          </a11y-media-transcript>
        </div>
      </div>
      <paper-toast
        id="link"
        disabled$="[[!linkable]]"
        hidden$="[[!linkable]]"
        duration="5000"
        text="Copied to clipboard: [[shareLink]]"
      >
        <a11y-media-button
          action="linkable"
          icon="[[_getLocal('closeLink','icon')]]"
          label="[[_getLocal('closeLink','label')]]"
          on-click="_handleCloseLink"
          tooltip-position="top"
        ></a11y-media-button>
      </paper-toast>
    `}static get properties(){return{...super.properties,allowConcurrent:{name:"allowConcurrent",type:Boolean,value:!1},audioNoThumb:{name:"audioNoThumb",type:Boolean,computed:"_getAudioNoThumb(audioOnly,thumbnailSrc)"},darkTranscript:{name:"darkTranscript",type:Boolean,value:!1},crossorigin:{type:String,value:null},disableFullscreen:{name:"disableFullscreen",type:Boolean,value:!1},disableInteractive:{name:"disableInteractive",type:Boolean,value:!1},flexLayout:{name:"flexLayout",type:Boolean,computed:"_isFlexLayout(standAlone,hideTranscript,audioNoThumb,stackedLayout)",reflectToAttribute:!0},fullscreen:{name:"fullscreen",type:Boolean,value:!1},fullscreenButton:{name:"fullscreenButton",type:Boolean,computed:"_getFullscreenButton(disableFullscreen,audioNoThumb,screenfullLoaded)",notify:!0},hasCaptions:{name:"hasCaptions",type:Boolean,value:!1},hideElapsedTime:{name:"hideElapsedTime",type:Boolean,value:!1},hideTimestamps:{name:"hideTimestamps",type:Boolean,value:!1},hideTranscript:{name:"hideTranscript",type:Boolean,value:!1,reflectToAttribute:!0},id:{name:"id",type:String,value:null,reflectToAttribute:!0},mediaCaption:{name:"mediaCaption",type:String,computed:"_getMediaCaption(audioOnly,localization,mediaTitle)"},mediaLang:{name:"mediaLang",type:String,value:"en"},muteUnmute:{name:"muteUnmute",type:Object,computed:"_getMuteUnmute(muted)"},printCaption:{name:"printCaption",type:String,computed:"_getPrintCaption(audioOnly,audioLabel,videoLabel,mediaTitle)"},responsiveSize:{name:"responsiveSize",type:String,notify:!0,value:"xs",reflectToAttribute:!0},screenfullLoaded:{name:"screenfullLoaded",type:Boolean,value:!1,notify:!0},shareLink:{name:"shareLink",type:String,computed:"_getShareLink(__elapsed)"},showCustomCaptions:{name:"showCustomCaptions",type:Boolean,computed:"_showCustomCaptions(isYoutube, audioOnly, hasCaptions, cc)"},sources:{name:"sources",type:Array,value:[]},stackedLayout:{name:"stackedLayout",type:Boolean,value:!1},sticky:{name:"sticky",type:Boolean,value:!1,reflectToAttribute:!0},stickyCorner:{name:"stickyCorner",type:String,value:"top-right",reflectToAttribute:!0},thumbnailSrc:{name:"thumbnailSrc",type:String,value:null,reflectToAttribute:!0},tracks:{name:"tracks",type:Array,value:[]},playPause:{name:"playPause",type:Object,computed:"_getPlayPause(__playing)"},__elapsed:{name:"__elapsed",type:Number,value:null,notify:!0},__playing:{name:"__playing",type:Boolean,value:!1,notify:!0,reflectToAttribute:!0},__captionHref:{name:"__captionHref",type:String,value:null,notify:!0}}}constructor(){super(),new Promise((res,rej)=>_require.default(["../../@polymer/paper-slider/paper-slider.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/iron-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../@polymer/iron-icons/av-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["./lib/a11y-media-play-button.js"],res,rej));const location=`${this.pathFromUrl(decodeURIComponent(meta.url))}lib/screenfull/dist/screenfull.js`;window.ESGlobalBridge.requestAvailability(),window.ESGlobalBridge.instance.load("screenfullLib",location),window.addEventListener("es-bridge-screenfullLib-loaded",this._onScreenfullLoaded.bind(this)),this.__playerAttached=!0,window.A11yMediaStateManager.requestAvailability()}pathFromUrl(url){return url.substring(0,url.lastIndexOf("/")+1)}static get tag(){return"a11y-media-player"}connectedCallback(){super.connectedCallback(),this._addResponsiveUtility(),null===this.id&&(this.id="a11y-media-player"+Date.now()),window.dispatchEvent(new CustomEvent("a11y-player",{detail:this})),this.isYoutube&&this._youTubeRequest()}ready(){super.ready();let root=this;new Array,new Array;"object"==typeof screenfull&&root._onScreenfullLoaded.bind(root),root.__playerReady=!0,root.target=root.shadowRoot.querySelector("#transcript"),root.__status=root._getLocal("loading","label"),root.__slider=root.shadowRoot.querySelector("#slider"),root.__slider.min=0,root.__volume=root.muted?0:Math.max(this.volume,10),root.__resumePlaying=!1,root.__duration=0,root.shadowRoot.querySelector("#controls").setStatus(root.__status),root.width=null!==root.width?root.width:"100%",root.style.maxWidth=null!==root.width?root.width:"100%",root._setPlayerHeight(16/9),root.isYoutube?(root._youTubeRequest(),document.addEventListener("timeupdate",e=>{e.detail===root.media&&root._handleTimeUpdate(e)})):(root.media=root.shadowRoot.querySelector("#html5"),root.media.media.addEventListener("timeupdate",e=>{root._handleTimeUpdate(e)}),root._addSourcesAndTracks()),root.shadowRoot.querySelector("#transcript").setMedia(root.shadowRoot.querySelector("#innerplayer"))}play(){let root=this;root.__playing;root.isYoutube&&!root.__ytAppended?ytInit():(root.__playing=!0,root.media.play(),window.dispatchEvent(new CustomEvent("a11y-player-playing",{detail:root})))}pause(){this.__playing=!1,this.media.pause()}stop(){this.pause(),this.seek(0)}restart(){this.seek(0),this.play()}rewind(amt){amt=void 0!==amt?amt:this.media.duration/20,this.__resumePlaying=this.__playing,this.seek(this.media.getCurrentTime()-amt,0),this.__resumePlaying&&this.play(),this.__resumePlaying=!1}forward(amt){amt=void 0!==amt?amt:this.media.duration/20,this.__resumePlaying=this.__playing,this.seek(this.media.getCurrentTime()+amt),this.__resumePlaying&&this.play(),this.__resumePlaying=!1}seek(time){let seekable=void 0!==this.media&&null!==this.media?this.media.seekable:[];seekable.length>0&&time>=seekable.start(0)&&time<=seekable.end(0)&&this.media.seek(time)}selectTrack(index){this.__selectedTrack=index,this.shadowRoot.querySelector("#html5").selectTrack(index)}setVolume(value){this.volume=null!==value?value:70,this.media.setVolume(this.volume),this.muted=0===this.volume}setPlaybackRate(value){value=null!==value?value:1,this.media.setPlaybackRate(value)}toggleCC(mode){this.cc=void 0===mode?!this.cc:mode,this.shadowRoot.querySelector("#html5").setCC(this.cc)}toggleLoop(mode){this.isYoutube||(this.loop=void 0===mode?!this.loop:mode,this.media.setLoop(this.loop))}toggleMute(mode){this.muted=void 0===mode?!this.muted:mode,this.__volume=this.muted?0:Math.max(this.volume,10),this.media.setMute(this.muted)}toggleSticky(mode){mode=void 0===mode?!this.sticky:mode,this.sticky=mode,this.dispatchEvent(new CustomEvent("player-sticky",{detail:this}))}toggleTranscript(mode){mode=void 0===mode?this.hideTranscript:mode,this.hideTranscript=!mode,void 0!==this.shadowRoot.querySelector("#transcript")&&null!==this.shadowRoot.querySelector("#transcript")&&this.dispatchEvent(new CustomEvent("transcript-toggle",{detail:this}))}_appendToPlayer(data,type){if(null!=data&&data!==[]){let root=this,arr=Array.isArray(data)?data:JSON.parse(data);for(let i=0;i<arr.length;i++){let el=document.createElement(type);this.__captionHref||"source"!==type||(this.__captionHref=arr[i].src);for(let key in arr[i])el.setAttribute(key,arr[i][key]);root.shadowRoot.querySelector("#html5").media.appendChild(el)}}}_setPlayerHeight(aspect){let root=this;root.shadowRoot.querySelector("#player").style.height="unset",root.audioOnly&&null===root.thumbnailSrc&&null===root.height?root.shadowRoot.querySelector("#player").style.height="60px":null===root.height?(root.shadowRoot.querySelector("#player").style.paddingTop=100/aspect+"%",root.shadowRoot.querySelector("#innerplayer").style.maxWidth="calc("+100*aspect+"vh - "+80*aspect+"px)"):root.shadowRoot.querySelector("#outerplayer").style.height=root.height}_getMediaCaption(audioOnly,localization,mediaTitle){let audioLabel=this._getLocal("audio","label"),hasMediaTitle=null!=mediaTitle&&""!==mediaTitle;return audioOnly&&hasMediaTitle?mediaTitle+" ("+audioLabel+")":audioOnly?audioLabel:hasMediaTitle?mediaTitle:null}_getMuteUnmute(muted){return muted?{label:this._getLocal("unmute","label"),icon:this._getLocal("unmute","icon"),action:"unmute"}:{label:this._getLocal("mute","label"),icon:this._getLocal("mute","icon"),action:"mute"}}_getPrintCaption(audioOnly,localization,mediaTitle){let audioLabel=this._getLocal("audio","label"),videoLabel=this._getLocal("video","label"),hasMediaTitle=null!=mediaTitle&&""!==mediaTitle;return audioOnly&&hasMediaTitle?mediaTitle+" ("+audioLabel+")":audioOnly?audioLabel:hasMediaTitle?mediaTitle+" ("+videoLabel+")":videoLabel}_getThumbnailCSS(thumbnailSrc,isYoutube,audioOnly){return null!=thumbnailSrc&&(isYoutube||audioOnly)?"background-image: url("+thumbnailSrc+"); background-size: cover;":null}_addSourcesAndTracks(){let root=this,counter=0;root.audioOnly=root.audioOnly||null!==root.querySelector("audio"),root.querySelectorAll("source,track").forEach(node=>{root.__captionHref||"SOURCE"!==node.tagName||(root.__captionHref=node.getAttribute("src")),root.shadowRoot.querySelector("#html5").media.appendChild(node)}),root._appendToPlayer(root.tracks,"track"),root._appendToPlayer(root.sources,"source"),root.shadowRoot.querySelector("#html5").media.textTracks.onaddtrack=e=>{root.hasCaptions=!0,root.hasTranscript=!root.standAlone,root._getTrackData(e.track,counter++)}}_getAudioNoThumb(audioOnly,thumbnailSrc){return audioOnly&&null==thumbnailSrc}_getFullscreenButton(disableFullscreen,audioNoThumb,screenfullLoaded){return"object"==typeof screenfull&&this._onScreenfullLoaded.bind(this),!(/Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(navigator.userAgent)||disableFullscreen||audioNoThumb||"object"!=typeof screenfull)}_getPlayPause(__playing){return!1!==__playing?{label:this._getLocal("pause","label"),icon:this._getLocal("pause","icon"),action:"pause"}:{label:this._getLocal("play","label"),icon:this._getLocal("play","icon"),action:"play"}}_getShareLink(__elapsed){let url=window.location.href.split(/[#?]/)[0],id=this.id?`?id=${this.id}`:"";return`${url}${id}${""!==id&&this.__elapsed&&0!==this.__elapsed?`&t=${this.__elapsed}`:""}`}_getTrackData(track,id){let loadCueData,root=this,selected=!0===track.default||void 0===root.__selectedTrack;selected&&root.selectTrack(id),track.mode=selected&&!0===this.cc?"showing":"hidden",loadCueData=setInterval(()=>{if(void 0!==track.cues&&null!==track.cues&&track.cues.length>0){clearInterval(loadCueData);let cues=Object.keys(track.cues).map(key=>({order:""!==track.cues[key].id?track.cues[key].id:key,seek:track.cues[key].startTime,seekEnd:track.cues[key].endTime,start:root._getHHMMSS(track.cues[key].startTime,root.media.duration),end:root._getHHMMSS(track.cues[key].endTime,root.media.duration),text:track.cues[key].text}));void 0===root.__tracks&&(root.__tracks=[]),root.push("__tracks",{value:id,language:track.language,text:void 0!==track.label?track.label:void 0!==track.language?track.language:"Track "+id,cues:cues}),root.shadowRoot.querySelector("#controls").setTracks(root.__tracks),root.shadowRoot.querySelector("#transcript").setTracks(root.__tracks),root.push("__tracks"),track.oncuechange=e=>{root.shadowRoot.querySelector("#transcript").setActiveCues(Object.keys(e.currentTarget.activeCues).map(key=>e.currentTarget.activeCues[key].id))}}},1)}_handleCloseLink(){this.shadowRoot.querySelector("#link").close(),this.__resumePlaying&&this.play(),this.__resumePlaying=!1}_handleCopyLink(){let el=document.createElement("textarea");this.__resumePlaying=this.__playing,this.pause,el.value=this.shareLink,document.body.appendChild(el),el.select(),document.execCommand("copy"),document.body.removeChild(el),this.shadowRoot.querySelector("#link").open()}_handleTranscriptSeek(e){let root=this;root.standAlone||void 0===root.shadowRoot.querySelector("#transcript")||null===root.shadowRoot.querySelector("#transcript")||(root.__resumePlaying=root.__playing,root.seek(e.detail))}_handleMediaLoaded(e){let anchor=window.AnchorBehaviors,target=anchor.getTarget(this),params=anchor.params,aspect=this.media.aspectRatio;this._setPlayerHeight(aspect),this.shadowRoot.querySelector("#playbutton").removeAttribute("disabled"),this._setElapsedTime(),this._getTrackData(this.shadowRoot.querySelector("#html5").media),target===this&&this.seek(this._getSeconds(params.t))}_hidePlayButton(thumbnailSrc,isYoutube,__elapsed){return isYoutube&&null===thumbnailSrc||!(void 0===__elapsed||0===__elapsed)}_handlePrinting(e){this.dispatchEvent(new CustomEvent("printing-transcript",{detail:this})),this.shadowRoot.querySelector("#transcript").print(this.mediaTitle)}_handleSearchAdded(e){this.search=e.detail}_handleSliderStart(e){this.__resumePlaying=!this.paused,this.pause(),this.__seeking=!0}_handleSliderStop(e){this.seek(this.shadowRoot.querySelector("#slider").immediateValue),this.__seeking=!1,this.__resumePlaying&&(this.play(),this.__resumePlaying=null)}_handleTimeUpdate(e){let root=this;root.isYoutube&&root.media.duration!==root.media.getDuration()&&(root.__duration=root.media.duration=root.media.getDuration(),root.disableSeek=!1,root._addSourcesAndTracks(),root.media.seekable&&root.media.seekable.length>0&&0!==root.media.seekable.start(0)&&(root.shadowRoot.querySelector("#slider").min=root.media.seekable.start(0))),void 0!==root.media.seekable&&root.media.seekable.length>0&&root.media.seekable.end(0)<=root.media.getCurrentTime()&&(root.stop(),root.__playing=!1),root._updateCustomTracks(),root._setElapsedTime()}_handleTranscriptScrollToggle(e){this.disableScroll=!this.disableScroll}_isFlexLayout(standAlone,hideTranscript,audioNoThumb,stackedLayout){return!(standAlone||hideTranscript||audioNoThumb||stackedLayout)}_onControlsChanged(e){let root=this,action=void 0!==e.detail.action?e.detail.action:e.detail.id;"backward"===action||"rewind"===action?root.rewind():"captions"===action?root.toggleCC():"transcript"===action||"transcript-toggle"===action?root.toggleTranscript():"tracks"===e.detail.id?""===e.detail.value?root.toggleCC(!1):(root.toggleCC(!0),root.selectTrack(e.detail.value)):"forward"===action?root.forward():"fullscreen"===action&&root.fullscreenButton?(root.toggleTranscript(root.fullscreen),screenfull.toggle(root.shadowRoot.querySelector("#outerplayer"))):"loop"===action?root.toggleLoop():"mute"===action||"unmute"===action?root.toggleMute():"pause"===action?root.pause():"play"===action?root.play():"restart"===action?(root.seek(0),root.play()):"speed"===action?root.setPlaybackRate(e.detail.value):"volume"===action?root.setVolume(e.detail.value):"linkable"===action&&root._handleCopyLink()}_onScreenfullLoaded(){let root=this;root.screenfullLoaded=!0,screenfull&&screenfull.on("change",()=>{screenfull.enabled&&(root.fullscreen=screenfull.isFullscreen)})}_setElapsedTime(){let elapsed=!0===this.__seeking?this.shadowRoot.querySelector("#slider").immediateValue:this.media.getCurrentTime()>0?this.media.getCurrentTime():0,duration=this.media.duration>0?this.media.duration:0;this.__elapsed=elapsed,this.__duration=duration,void 0!==this.media.seekable&&this.media.seekable.length>0&&(void 0!==this.media.seekable.start(0)&&(elapsed-=this.media.seekable.start(0)),void 0!==this.media.seekable.end(0)&&(duration=this.media.seekable.end(0)-(void 0!==this.media.seekable.start(0)?this.media.seekable.start(0):0))),this.__status=this._getHHMMSS(elapsed,duration)+"/"+this._getHHMMSS(duration),this.shadowRoot.querySelector("#controls").setStatus(this.__status)}_showCustomCaptions(isYoutube,audioOnly,hasCaptions,cc){return(isYoutube||audioOnly)&&hasCaptions&&cc}_useYoutubeIframe(thumbnailSrc,isYoutube,__elapsed){return isYoutube&&(null===thumbnailSrc||void 0===__elapsed||0===__elapsed)}_youTubeRequest(){window.A11yMediaYoutube.requestAvailability();let root=this,ytUtil=window.A11yMediaYoutube.instance;if(root.disableSeek=!0,root.__playerAttached&&root.__playerReady){let ytInit=()=>{root.media=ytUtil.initYoutubePlayer({width:"100%",height:"100%",videoId:root.youtubeId}),root.__status=root._getLocal("youTubeLoading","label"),root.shadowRoot.querySelector("#controls").setStatus(root.__status),root.shadowRoot.querySelector("#youtube").appendChild(root.media.a),root.__ytAppended=!0,root._updateCustomTracks()},checkApi=e=>{ytUtil.apiReady&&(document.removeEventListener("youtube-api-ready",checkApi),root.__ytAppended||ytInit())};ytUtil.apiReady?root.__ytAppended||ytInit():document.addEventListener("youtube-api-ready",checkApi)}}_updateCustomTracks(){if((this.isYoutube||this.audioOnly)&&this.__tracks){let root=this,track=root.__tracks[this.shadowRoot.querySelector("#transcript").selectedTranscript],active=[],caption="";if(null!=track&&void 0!==track.cues&&null!==track.cues){for(let i=0;i<track.cues.length;i++)track.cues[i].seek<root.__elapsed&&track.cues[i].seekEnd>root.__elapsed&&(active.push(track.cues[i].order),caption=""===caption?track.cues[i].text:caption);root.shadowRoot.querySelector("#customcctxt").innerText=caption,root.shadowRoot.querySelector("#transcript").setActiveCues(active)}}}disconnectedCallback(){window.removeEventListener("es-bridge-screenfullLib-loaded",this._onScreenfullLoaded.bind(this)),super.disconnectedCallback()}}_exports.A11yMediaPlayer=A11yMediaPlayer,window.customElements.define(A11yMediaPlayer.tag,A11yMediaPlayer)}));
