// @juggle/resize-observer/lib/ResizeObservation.js
define(["exports", "./ResizeObserverBoxOptions.js", "./algorithms/calculateBoxSize.js", "./utils/element.js"], function (_exports, _ResizeObserverBoxOptions, _calculateBoxSize, _element) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ResizeObservation = void 0;

  var skipNotifyOnElement = function skipNotifyOnElement(target) {
    return !(0, _element.isSVG)(target) && !(0, _element.isReplacedElement)(target) && getComputedStyle(target).display === 'inline';
  };

  var ResizeObservation = function () {
    function ResizeObservation(target, observedBox) {
      this.target = target;
      this.observedBox = observedBox || _ResizeObserverBoxOptions.ResizeObserverBoxOptions.CONTENT_BOX;
      this.lastReportedSize = {
        inlineSize: 0,
        blockSize: 0
      };
    }

    ResizeObservation.prototype.isActive = function () {
      var size = (0, _calculateBoxSize.calculateBoxSize)(this.target, this.observedBox, true);

      if (skipNotifyOnElement(this.target)) {
        this.lastReportedSize = size;
      }

      if (this.lastReportedSize.inlineSize !== size.inlineSize || this.lastReportedSize.blockSize !== size.blockSize) {
        return true;
      }

      return false;
    };

    return ResizeObservation;
  }();

  _exports.ResizeObservation = ResizeObservation;
});
