// @lrnwebcomponents/mtz-marked-editor/lib/mtz-marked-control-behavior.js
define(["exports"], function (_exports) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.mtzMarkedControlBehavior = void 0;

  var mtzMarkedControlBehavior = function mtzMarkedControlBehavior(SuperClass) {
    return (
      /*#__PURE__*/
      function (_SuperClass) {
        babelHelpers.inherits(_class, _SuperClass);

        function _class() {
          babelHelpers.classCallCheck(this, _class);
          return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(_class).apply(this, arguments));
        }

        babelHelpers.createClass(_class, [{
          key: "connectedCallback",

          /**
           * Registers the control with the editor to obtain the editor instance
           * @fires register-control
           */
          value: function connectedCallback() {
            babelHelpers.get(babelHelpers.getPrototypeOf(_class.prototype), "connectedCallback", this).call(this);
            this.dispatchEvent(new CustomEvent("register-control", {
              bubbles: true,
              cancelable: true,
              composed: true,
              detail: true
            }));
          }
        }], [{
          key: "properties",
          get: function get() {
            return babelHelpers.objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(_class), "properties", this), {
              /* Reference to the editor. Set programatically */
              __editor: Object
            });
          }
        }]);
        return _class;
      }(SuperClass)
    );
  };

  _exports.mtzMarkedControlBehavior = mtzMarkedControlBehavior;
});
