// @lrnwebcomponents/lrndesign-abbreviation/lrndesign-abbreviation.js
define(["exports","require","../../@polymer/polymer/polymer-element.js","../schema-behaviors/schema-behaviors.js"],(function(e,t,i,r){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.LrndesignAbbreviation=void 0,t=babelHelpers.interopRequireWildcard(t);
/**
   * Copyright 2018 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class LrndesignAbbreviation extends((0,r.SchemaBehaviors)(i.PolymerElement)){constructor(){super(),new Promise((e,i)=>t.default(["../simple-tooltip/simple-tooltip.js"],e,i))}static get template(){return i.html`
      <style>
        :host {
          display: inline-block;
        }
        abbr {
          transition: 0.6s all ease;
          padding: 2px 4px;
          font-style: italic;
          background-color: var(--abbreviation-bg, #f9f9f9);
          text-underline-position: under;
          text-decoration: underline double;
          cursor: help;
          outline: var(--abbreviation-selection, #ffff33);
          @apply --abbreviation-main;
        }
        abbr:focus,
        abbr:active,
        abbr:hover {
          text-decoration: none;
          background-color: var(--abbreviation-selection, #ffff33);
          @apply --abbreviation-hover;
        }
        abbr::-moz-selection,
        abbr::selection {
          text-decoration: none;
          background-color: var(--abbreviation-selection, #ffff33);
          @apply --abbreviation-selection;
        }
      </style>
      <abbr tabindex="0" title$="[[phrase]]" id="abbr">[[abbr]]</abbr>
      <simple-tooltip for="abbr" position="top" offset="2" animation-delay="300"
        >[[phrase]]</simple-tooltip
      >
    `}static get tag(){return"lrndesign-abbreviation"}static get properties(){return{...super.properties,abbr:{type:String,reflectToAttribute:!0,notify:!0},phrase:{type:String,reflectToAttribute:!0,notify:!0}}}static get haxProperties(){return{canScale:!1,canPosition:!1,canEditSource:!1,gizmo:{title:"Abbreviation",description:"Simple abbreviation with tooltip of full word",icon:"editor:title",color:"grey",groups:["Instructional","Term"],handles:[{type:"inline",text:"text"}],meta:{author:"ELMS:LN"}},settings:{quick:[{property:"abbr",title:"Abbreviation",description:"Abbreviation word",inputMethod:"textfield",icon:"editor:title"},{property:"phrase",title:"Phrase",description:"The phrase / original words",inputMethod:"textfield",icon:"editor:title"}],configure:[{property:"abbr",title:"Abbreviation",description:"Abbreviation word",inputMethod:"textfield",icon:"editor:title"},{property:"phrase",title:"Phrase",description:"The phrase / original words",inputMethod:"textfield",icon:"editor:title"}],advanced:[]}}}}e.LrndesignAbbreviation=LrndesignAbbreviation,window.customElements.define(LrndesignAbbreviation.tag,LrndesignAbbreviation)}));
