// @lrnwebcomponents/fullscreen-behaviors/fullscreen-behaviors.js
define(["exports", "meta", "../es-global-bridge/es-global-bridge.js", "../../lit-element/lit-element.js"], function (_exports, meta, _esGlobalBridge, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.FullscreenBehaviors = void 0;
  meta = babelHelpers.interopRequireWildcard(meta);

  /**
   * Copyright 2018 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */

  /**
   * `fullscreen-behaviors`
   *
   * @demo demo/viewer.html
   * @element fullscreen-behaviors
   *
   */
  class FullscreenBehaviors extends _litElement.LitElement {
    static get tag() {
      return "fullscreen-behaviors";
    } // properties available to the custom element for data binding


    static get properties() {
      return {
        __fullscreenBehaviorsLoaded: {
          type: Boolean
        }
      };
    }

    constructor() {
      super();
      this.__fullscreenBehaviorsLoaded = false;

      if (typeof screenfull === "object") {
        this._onFullscreenBehaviorsLoaded();
      } else {
        const basePath = this.pathFromUrl(decodeURIComponent(meta.url));
        const location = `${basePath}screenfull/dist/screenfull.js`;
        window.ESGlobalBridge.requestAvailability();
        window.ESGlobalBridge.instance.load("screenfullLib", location);
        window.addEventListener("es-bridge-screenfullLib-loaded", this._onFullscreenBehaviorsLoaded);
      }
    }
    /**
     * life cycle, element is removed from the DOM
     */


    disconnectedCallback() {
      window.removeEventListener("es-bridge-screenfullLib-loaded", this._onFullscreenBehaviorsLoaded);
      super.disconnectedCallback();
    }
    /**
     * whether or not the fullscreen mode is be disabled
     * @returns {boolean}
     */


    get fullscreenEnabled() {
      let mobile = /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(navigator.userAgent);
      return typeof screenfull === "object" && !mobile;
    }
    /**
     * sets the element's __screenfullLoaded variable to true once screenfull is loaded
     * and adds an event listener for screenfull
     * @param {event} e screenfull load
     */


    _onFullscreenBehaviorsLoaded() {
      this.__fullscreenBehaviorsLoaded = true;
    }

  }

  _exports.FullscreenBehaviors = FullscreenBehaviors;
  window.customElements.define(FullscreenBehaviors.tag, FullscreenBehaviors);
  // register globally so we can make sure there is only one
  window.FullscreenBehaviors = window.FullscreenBehaviors || {}; // request if this exists. This helps invoke the element existing in the dom
  // as well as that there is only one of them. That way we can ensure everything
  // is rendered through the same modal

  window.FullscreenBehaviors.requestAvailability = () => {
    if (!window.FullscreenBehaviors.instance) {
      window.FullscreenBehaviors.instance = document.createElement("fullscreen-behaviors");
      document.body.appendChild(window.FullscreenBehaviors.instance);
    }

    return window.FullscreenBehaviors.instance;
  };
});
