// @lrnwebcomponents/hax-body/lib/hax-context-behaviors.js
import{LitElement as t,html as o,css as e}from"../../../lit-element/lit-element.js";import{SimpleTourFinder as n}from"../../simple-popover/lib/SimpleTourFinder.js";export const HaxContextBehaviors=function(t){return class extends(n(t)){static get styles(){return[e`
          :host {
            display: block;
            pointer-events: none;
            --hax-ui-spacing-sm: 1px;
          }
          :host [hidden] {
            display: none;
          }
          :host > * {
            margin: 0 auto;
          }
          .selected-buttons {
            transition: 0.1s all ease-in-out;
            width: 0;
          }
          :host([has-selected-text]) .selected-buttons {
            width: 100%;
          }
          :host(.hax-context-pin-top) #toolbar {
            position: fixed;
            top: 0px;
          }
          :host(:hover),
          :host(:focus-within) {
            z-index: var(--hax-ui-focus-z-index) !important;
          }
          .group {
            padding: 0;
          }
          hax-toolbar {
            flex: 0 1 auto;
            border: none !important;
          }
          hax-toolbar[collapse-disabled]::part(morebutton) {
            display: none !important;
          }
          hax-toolbar *[hidden] {
            display: none !important;
          }
          hax-toolbar[collapse-disabled]::part(morebutton) {
            display: none !important;
          }
        `]}constructor(){super()}render(){return o`<slot></slot> `}static get tag(){return"hax-context-behaviors"}}};
