// @lrnwebcomponents/hax-body/lib/hax-context-behaviors.js
import{LitElement as t,html as e,css as o}from"../../../lit/index.js";import{SimpleTourFinder as a}from"../../simple-popover/lib/SimpleTourFinder.js";import{autorun as s,toJS as r}from"../../../mobx/dist/mobx.esm.js";import{HAXStore as i}from"./hax-store.js";export const HaxContextBehaviors=function(t){return class extends(a(t)){static get styles(){return[o`
          :host {
            display: block;
            --hax-ui-spacing-sm: 1px;
            max-width: 100%;
          }
          :host [hidden] {
            display: none;
          }
          .selected-buttons {
            transition: 0.1s all ease-in-out;
            width: 0;
          }
          :host([has-selected-text]) .selected-buttons {
            width: 100%;
          }
          :host(.hax-context-pin-top) #toolbar {
            position: fixed;
            top: 0px;
          }
          :host(:hover),
          :host(:focus-within) {
            z-index: var(--hax-ui-focus-z-index) !important;
          }
          .group {
            padding: 0;
            background-color: var(--hax-ui-background-color);
          }
          hax-toolbar {
            flex: 0 1 auto;
          }
          hax-toolbar::part(morebutton) {
            border: 1px solid
              var(
                --simple-toolbar-group-border-width,
                var(--simple-toolbar-border-width, 1px)
              ) !important;
          }
          hax-toolbar[collapse-disabled]::part(morebutton) {
            display: none !important;
          }
          hax-toolbar *[hidden] {
            display: none !important;
          }
          hax-toolbar[collapse-disabled]::part(morebutton) {
            display: none !important;
          }
        `]}constructor(){super(),this.viewSource=!1,s((()=>{this.hasSelectedText=r(i.haxSelectedText).length>0})),s((()=>{r(i.editMode);const t=r(i.activeNode);if(this.sourceView=!1,t&&t.tagName){let e=i.haxSchemaFromTag(t.tagName);this.parentSchema=t&&t.parentNode?i.haxSchemaFromTag(t.parentNode.tagName):void 0,this.sourceView=e.canEditSource}}))}render(){return e`<slot></slot> `}static get tag(){return"hax-context-behaviors"}static get properties(){return{...super.properties,activeNode:{type:Object},parentSchema:{type:Object},realSelectedValue:{type:String},sourceView:{type:Boolean},viewSource:{type:Boolean}}}get slotSchema(){let t;if(this.activeNode&&this.parentSchema){let e=this.activeNode.slot||"";Object.keys(this.parentSchema.settings||{}).forEach((o=>{(this.parentSchema.settings[o]||[]).forEach((o=>{o.slot&&o.slot===e&&(t=o)}))}))}return t}getFilteredBlocks(t=[]){return t.filter((t=>{if(!t.tag)return;let e=t.tag||"",o=this.slotSchema&&this.slotSchema.slotWrapper&&this.slotSchema.slotWrapper?this.slotSchema.slotWrapper:void 0,a=this.slotSchema&&this.slotSchema.slotWrapper&&this.slotSchema.allowedSlotWrappers?this.slotSchema.allowedSlotWrappers:void 0,s=this.slotSchema&&this.slotSchema.slotWrapper&&this.slotSchema.excludedSlotWrappers?this.slotSchema.excludedSlotWrappers:void 0,r=!this.slotSchema||!a,i=!!o&&o===e||!!a&&a.includes(e);return!(!!s&&s.includes(e))&&(r||i)}))}updated(t){super.updated&&super.updated(t),t.forEach(((t,e)=>{"activeNode"===e&&this.activeNode!==t&&this.setTarget(this.activeNode)}))}setTarget(t=this.activeNode){super.setTarget&&super.setTarget(t),this.parentSchema=t&&t.parentNode?i.haxSchemaFromTag(t.parentNode.tagName):void 0}}};
