// @lrnwebcomponents/editable-table/lib/editable-table-sort.js
define(["exports", "../../../lit/index.js", "./editable-table-behaviors.js", "../../simple-icon/lib/simple-icon-lite.js", "../../simple-icon/lib/simple-icons.js", "../../hax-iconset/lib/simple-hax-iconset.js"], function (_exports, _index, _editableTableBehaviors, _simpleIconLite, _simpleIcons, _simpleHaxIconset) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.EditableTableSort = void 0;

  var _templateObject_ee0b60a0feb911eb8439e3eec9305af7, _templateObject2_ee0b60a0feb911eb8439e3eec9305af7;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `editable-table-editor-sort`
   * A column header that functions as a three-state sort button (no sort, sort ascending, sort descending) for table-editor-display mode (table-editor-display.html).
   *
   * @demo ./demo/display.html
   * @customElement
   * @extends LitElement
   * @extends editableTableCellStyles
   */
  var EditableTableSort = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(EditableTableSort, _LitElement);

    var _super = _createSuper(EditableTableSort);

    function EditableTableSort() {
      var _this;

      babelHelpers.classCallCheck(this, EditableTableSort);
      _this = _super.call(this);
      _this.sortMode = "none";
      _this.sortColumn = -1;
      return _this;
    }
    /**
     *
     * Whether column is being sorted
     * @readonly
     * @memberof EditableTableSort
     */


    babelHelpers.createClass(EditableTableSort, [{
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_ee0b60a0feb911eb8439e3eec9305af7 || (_templateObject_ee0b60a0feb911eb8439e3eec9305af7 = babelHelpers.taggedTemplateLiteral(["\n      <button id=\"button\" class=\"cell-button\" @click=\"", "\">\n        <slot></slot>\n        <span class=\"sr-only asc\">(ascending)</span>\n        <span class=\"sr-only desc\">(descending)</span>\n        <span class=\"sr-only\"> Toggle sort mode.</span>\n        <simple-icon-lite\n          icon=\"", "\"\n        ></simple-icon-lite>\n      </button>\n    "])), this._onSortClicked, this.sortMode == "asc" ? "arrow-drop-up" : this.sortMode == "desc" ? "arrow-drop-down" : "editable-table:sortable");
      }
    }, {
      key: "sorting",
      get: function get() {
        return this.columnIndex === this.sortColumn;
      }
      /**
       * Fires when sort button is clicked
       * @event change-sort-mode
       */

    }, {
      key: "_onSortClicked",
      value: function _onSortClicked() {
        this.dispatchEvent(new CustomEvent("change-sort-mode", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: this
        }));
      }
      /**
       * Changes sort mode
       * @param {string} mode sort mode: `asc` for ascending or `desc` for descending;
       */

    }, {
      key: "setSortMode",
      value: function setSortMode(mode) {
        this.sortMode = mode;
        this.__checked = mode === "asc" ? true : mode === "desc" ? mode : false;
      }
    }], [{
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(EditableTableSort), "styles", this) || []), babelHelpers.toConsumableArray(_editableTableBehaviors.editableTableCellStyles), [(0, _index.css)(_templateObject2_ee0b60a0feb911eb8439e3eec9305af7 || (_templateObject2_ee0b60a0feb911eb8439e3eec9305af7 = babelHelpers.taggedTemplateLiteral([""])))]);
      }
    }, {
      key: "tag",
      get: function get() {
        return "editable-table-sort";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Sort ascending, descending or none
           */
          columnIndex: {
            attribute: "column-index",
            type: Number,
            reflect: true
          },

          /**
           * Sort mode: ascending, descending or none
           */
          sortMode: {
            attribute: "sort-mode",
            type: String,
            reflect: true
          },

          /**
           * Index of current sort column
           */
          sortColumn: {
            attribute: "sort-column",
            type: Number,
            reflect: true
          }
        };
      }
    }]);
    return EditableTableSort;
  }(_index.LitElement);

  _exports.EditableTableSort = EditableTableSort;
  window.customElements.define(EditableTableSort.tag, EditableTableSort);
});
