// lit-html/directives/async-append.js
define(["exports", "../lit-html.js", "../directive.js", "../directive-helpers.js", "../async-directive.js"], function (_exports, _litHtml, _directive, _directiveHelpers, _asyncDirective) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.asyncAppend = void 0;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var c = (0, _directive.directive)( /*#__PURE__*/function (_n) {
    babelHelpers.inherits(_class, _n);

    var _super = _createSuper(_class);

    function _class(t) {
      var _this;

      babelHelpers.classCallCheck(this, _class);
      if (_this = _super.call(this, t), t.type !== _directive.PartType.CHILD) throw Error("asyncAppend can only be used in child expressions");
      return babelHelpers.possibleConstructorReturn(_this);
    }

    babelHelpers.createClass(_class, [{
      key: "render",
      value: function render(e, i) {
        return _litHtml.noChange;
      }
    }, {
      key: "update",
      value: function update(e, _ref) {
        var _ref2 = babelHelpers.slicedToArray(_ref, 2),
            i = _ref2[0],
            s = _ref2[1];

        if (i !== this.vt) return this.vt = i, this.Σft(e, s), _litHtml.noChange;
      }
    }, {
      key: "\u03A3ft",
      value: function () {
        var _Σft = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(t, e) {
          var i, n, _iteratorAbruptCompletion, _didIteratorError, _iteratorError, _iterator, _step, _c, h;

          return regeneratorRuntime.wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  i = 0;
                  n = this.vt;
                  _iteratorAbruptCompletion = false;
                  _didIteratorError = false;
                  _context.prev = 4;
                  _iterator = babelHelpers.asyncIterator(n);

                case 6:
                  _context.next = 8;
                  return _iterator.next();

                case 8:
                  if (!(_iteratorAbruptCompletion = !(_step = _context.sent).done)) {
                    _context.next = 23;
                    break;
                  }

                  _c = _step.value;

                  if (!(this.vt !== n)) {
                    _context.next = 12;
                    break;
                  }

                  return _context.abrupt("break", 23);

                case 12:
                  _context.t0 = this.wt;

                  if (!_context.t0) {
                    _context.next = 16;
                    break;
                  }

                  _context.next = 16;
                  return this.wt;

                case 16:
                  0 === i && (0, _directiveHelpers.clearPart)(t);
                  void 0 !== e && (_c = e(_c, i));
                  h = (0, _directiveHelpers.insertPart)(t);
                  (0, _directiveHelpers.setChildPartValue)(h, _c), i++;

                case 20:
                  _iteratorAbruptCompletion = false;
                  _context.next = 6;
                  break;

                case 23:
                  _context.next = 29;
                  break;

                case 25:
                  _context.prev = 25;
                  _context.t1 = _context["catch"](4);
                  _didIteratorError = true;
                  _iteratorError = _context.t1;

                case 29:
                  _context.prev = 29;
                  _context.prev = 30;

                  if (!(_iteratorAbruptCompletion && _iterator.return != null)) {
                    _context.next = 34;
                    break;
                  }

                  _context.next = 34;
                  return _iterator.return();

                case 34:
                  _context.prev = 34;

                  if (!_didIteratorError) {
                    _context.next = 37;
                    break;
                  }

                  throw _iteratorError;

                case 37:
                  return _context.finish(34);

                case 38:
                  return _context.finish(29);

                case 39:
                case "end":
                  return _context.stop();
              }
            }
          }, _callee, this, [[4, 25, 29, 39], [30,, 34, 38]]);
        }));

        function Σft(_x, _x2) {
          return _Σft.apply(this, arguments);
        }

        return Σft;
      }()
    }, {
      key: "disconnected",
      value: function disconnected() {
        var _this2 = this;

        this.wt = new Promise(function (t) {
          return _this2.yt = t;
        });
      }
    }, {
      key: "reconnected",
      value: function reconnected() {
        this.wt = void 0, this.yt();
      }
    }]);
    return _class;
  }(_asyncDirective.AsyncDirective));
  _exports.asyncAppend = c;
});
