// lit-html/directives/async-replace.js
define(["exports", "../lit-html.js", "../directive.js", "../async-directive.js"], function (_exports, _litHtml, _directive, _asyncDirective) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.asyncReplace = void 0;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var s = (0, _directive.directive)( /*#__PURE__*/function (_e) {
    babelHelpers.inherits(_class, _e);

    var _super = _createSuper(_class);

    function _class() {
      babelHelpers.classCallCheck(this, _class);
      return _super.apply(this, arguments);
    }

    babelHelpers.createClass(_class, [{
      key: "render",
      value: function render(i, e) {
        return _litHtml.noChange;
      }
    }, {
      key: "update",
      value: function update(i, _ref) {
        var _ref2 = babelHelpers.slicedToArray(_ref, 2),
            e = _ref2[0],
            s = _ref2[1];

        if (e !== this.vt) return this.vt = e, this.Σft(s), _litHtml.noChange;
      }
    }, {
      key: "\u03A3ft",
      value: function () {
        var _Σft = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(t) {
          var i, e, _iteratorAbruptCompletion, _didIteratorError, _iteratorError, _iterator, _step, _s;

          return regeneratorRuntime.wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  i = 0;
                  e = this.vt;
                  _iteratorAbruptCompletion = false;
                  _didIteratorError = false;
                  _context.prev = 4;
                  _iterator = babelHelpers.asyncIterator(e);

                case 6:
                  _context.next = 8;
                  return _iterator.next();

                case 8:
                  if (!(_iteratorAbruptCompletion = !(_step = _context.sent).done)) {
                    _context.next = 22;
                    break;
                  }

                  _s = _step.value;

                  if (!(this.vt !== e)) {
                    _context.next = 12;
                    break;
                  }

                  return _context.abrupt("break", 22);

                case 12:
                  _context.t0 = this.wt;

                  if (!_context.t0) {
                    _context.next = 16;
                    break;
                  }

                  _context.next = 16;
                  return this.wt;

                case 16:
                  void 0 !== t && (_s = t(_s, i));
                  this.setValue(_s);
                  i++;

                case 19:
                  _iteratorAbruptCompletion = false;
                  _context.next = 6;
                  break;

                case 22:
                  _context.next = 28;
                  break;

                case 24:
                  _context.prev = 24;
                  _context.t1 = _context["catch"](4);
                  _didIteratorError = true;
                  _iteratorError = _context.t1;

                case 28:
                  _context.prev = 28;
                  _context.prev = 29;

                  if (!(_iteratorAbruptCompletion && _iterator.return != null)) {
                    _context.next = 33;
                    break;
                  }

                  _context.next = 33;
                  return _iterator.return();

                case 33:
                  _context.prev = 33;

                  if (!_didIteratorError) {
                    _context.next = 36;
                    break;
                  }

                  throw _iteratorError;

                case 36:
                  return _context.finish(33);

                case 37:
                  return _context.finish(28);

                case 38:
                case "end":
                  return _context.stop();
              }
            }
          }, _callee, this, [[4, 24, 28, 38], [29,, 33, 37]]);
        }));

        function Σft(_x) {
          return _Σft.apply(this, arguments);
        }

        return Σft;
      }()
    }, {
      key: "disconnected",
      value: function disconnected() {
        var _this = this;

        this.wt = new Promise(function (t) {
          return _this.yt = t;
        });
      }
    }, {
      key: "reconnected",
      value: function reconnected() {
        this.wt = void 0, this.yt();
      }
    }]);
    return _class;
  }(_asyncDirective.AsyncDirective));
  _exports.asyncReplace = s;
});
