// lit-html/directives/private-async-helpers.js
define(["exports"], function (_exports) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.forAwaitOf = _exports.PseudoWeakRef = _exports.Pauser = void 0;

  function _asyncIterator(iterable) { var method, async, sync, retry = 2; for ("undefined" != typeof Symbol && (async = Symbol.asyncIterator, sync = Symbol.iterator); retry--;) { if (async && null != (method = iterable[async])) return method.call(iterable); if (sync && null != (method = iterable[sync])) return new AsyncFromSyncIterator(method.call(iterable)); async = "@@asyncIterator", sync = "@@iterator"; } throw new TypeError("Object is not async iterable"); }

  function AsyncFromSyncIterator(s) { function AsyncFromSyncIteratorContinuation(r) { if (Object(r) !== r) return Promise.reject(new TypeError(r + " is not an object.")); var done = r.done; return Promise.resolve(r.value).then(function (value) { return { value: value, done: done }; }); } return AsyncFromSyncIterator = function AsyncFromSyncIterator(s) { this.s = s, this.n = s.next; }, AsyncFromSyncIterator.prototype = { s: null, n: null, next: function next() { return AsyncFromSyncIteratorContinuation(this.n.apply(this.s, arguments)); }, return: function _return(value) { var ret = this.s.return; return void 0 === ret ? Promise.resolve({ value: value, done: !0 }) : AsyncFromSyncIteratorContinuation(ret.apply(this.s, arguments)); }, throw: function _throw(value) { var thr = this.s.return; return void 0 === thr ? Promise.reject(value) : AsyncFromSyncIteratorContinuation(thr.apply(this.s, arguments)); } }, new AsyncFromSyncIterator(s); }

  /**
   * @license
   * Copyright 2021 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var t = /*#__PURE__*/function () {
    var _ref = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_t, s) {
      var _iteratorAbruptCompletion, _didIteratorError, _iteratorError, _iterator, _step, _i;

      return regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _iteratorAbruptCompletion = false;
              _didIteratorError = false;
              _context.prev = 2;
              _iterator = _asyncIterator(_t);

            case 4:
              _context.next = 6;
              return _iterator.next();

            case 6:
              if (!(_iteratorAbruptCompletion = !(_step = _context.sent).done)) {
                _context.next = 17;
                break;
              }

              _i = _step.value;
              _context.t0 = !1;
              _context.next = 11;
              return s(_i);

            case 11:
              _context.t1 = _context.sent;

              if (!(_context.t0 === _context.t1)) {
                _context.next = 14;
                break;
              }

              return _context.abrupt("return");

            case 14:
              _iteratorAbruptCompletion = false;
              _context.next = 4;
              break;

            case 17:
              _context.next = 23;
              break;

            case 19:
              _context.prev = 19;
              _context.t2 = _context["catch"](2);
              _didIteratorError = true;
              _iteratorError = _context.t2;

            case 23:
              _context.prev = 23;
              _context.prev = 24;

              if (!(_iteratorAbruptCompletion && _iterator.return != null)) {
                _context.next = 28;
                break;
              }

              _context.next = 28;
              return _iterator.return();

            case 28:
              _context.prev = 28;

              if (!_didIteratorError) {
                _context.next = 31;
                break;
              }

              throw _iteratorError;

            case 31:
              return _context.finish(28);

            case 32:
              return _context.finish(23);

            case 33:
            case "end":
              return _context.stop();
          }
        }
      }, _callee, null, [[2, 19, 23, 33], [24,, 28, 32]]);
    }));

    return function t(_x, _x2) {
      return _ref.apply(this, arguments);
    };
  }();

  _exports.forAwaitOf = t;

  var s = /*#__PURE__*/function () {
    function s(t) {
      babelHelpers.classCallCheck(this, s);
      this.U = t;
    }

    babelHelpers.createClass(s, [{
      key: "disconnect",
      value: function disconnect() {
        this.U = void 0;
      }
    }, {
      key: "reconnect",
      value: function reconnect(t) {
        this.U = t;
      }
    }, {
      key: "deref",
      value: function deref() {
        return this.U;
      }
    }]);
    return s;
  }();

  _exports.PseudoWeakRef = s;

  var i = /*#__PURE__*/function () {
    function i() {
      babelHelpers.classCallCheck(this, i);
      this.Y = void 0, this.q = void 0;
    }

    babelHelpers.createClass(i, [{
      key: "get",
      value: function get() {
        return this.Y;
      }
    }, {
      key: "pause",
      value: function pause() {
        var _this = this;

        var t;
        null !== (t = this.Y) && void 0 !== t || (this.Y = new Promise(function (t) {
          return _this.q = t;
        }));
      }
    }, {
      key: "resume",
      value: function resume() {
        var t;
        null === (t = this.q) || void 0 === t || t.call(this), this.Y = this.q = void 0;
      }
    }]);
    return i;
  }();

  _exports.Pauser = i;
});
