// lit-html/directives/ref.js
define(["exports", "../lit-html.js", "../directive.js", "../async-directive.js"], function (_exports, _litHtml, _directive, _asyncDirective) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ref = _exports.createRef = void 0;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * @license
   * Copyright 2020 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var e = function e() {
    return new o();
  };

  _exports.createRef = e;

  var o = function o() {
    babelHelpers.classCallCheck(this, o);
  };

  var h = new WeakMap(),
      n = (0, _directive.directive)( /*#__PURE__*/function (_s) {
    babelHelpers.inherits(_class, _s);

    var _super = _createSuper(_class);

    function _class() {
      babelHelpers.classCallCheck(this, _class);
      return _super.apply(this, arguments);
    }

    babelHelpers.createClass(_class, [{
      key: "render",
      value: function render(i) {
        return _litHtml.nothing;
      }
    }, {
      key: "update",
      value: function update(i, _ref) {
        var _ref2 = babelHelpers.slicedToArray(_ref, 1),
            s = _ref2[0];

        var e;
        var o = s !== this.gt;
        return o && void 0 !== this.gt && this.xt(void 0), (o || this.Tt !== this.Et) && (this.gt = s, this.At = null === (e = i.options) || void 0 === e ? void 0 : e.host, this.xt(this.Et = i.element)), _litHtml.nothing;
      }
    }, {
      key: "xt",
      value: function xt(t) {
        "function" == typeof this.gt ? (void 0 !== h.get(this.gt) && this.gt.call(this.At, void 0), h.set(this.gt, t), void 0 !== t && this.gt.call(this.At, t)) : this.gt.value = t;
      }
    }, {
      key: "Tt",
      get: function get() {
        var t;
        return "function" == typeof this.gt ? h.get(this.gt) : null === (t = this.gt) || void 0 === t ? void 0 : t.value;
      }
    }, {
      key: "disconnected",
      value: function disconnected() {
        this.Tt === this.Et && this.xt(void 0);
      }
    }, {
      key: "reconnected",
      value: function reconnected() {
        this.xt(this.Et);
      }
    }]);
    return _class;
  }(_asyncDirective.AsyncDirective));
  _exports.ref = n;
});
