// lit-html/directives/ref.js
define(["exports", "../lit-html.js", "../async-directive.js", "../directive.js"], function (_exports, _litHtml, _asyncDirective, _directive) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ref = _exports.createRef = void 0;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * @license
   * Copyright 2020 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var e = function e() {
    return new o();
  };

  _exports.createRef = e;
  var o = /*#__PURE__*/babelHelpers.createClass(function o() {
    babelHelpers.classCallCheck(this, o);
  });
  var h = new WeakMap(),
      n = (0, _directive.directive)( /*#__PURE__*/function (_t) {
    babelHelpers.inherits(_class, _t);

    var _super = _createSuper(_class);

    function _class() {
      babelHelpers.classCallCheck(this, _class);
      return _super.apply(this, arguments);
    }

    babelHelpers.createClass(_class, [{
      key: "render",
      value: function render(t) {
        return _litHtml.nothing;
      }
    }, {
      key: "update",
      value: function update(t, _ref) {
        var _ref2 = babelHelpers.slicedToArray(_ref, 1),
            s = _ref2[0];

        var e;
        var o = s !== this.U;
        return o && void 0 !== this.U && this.ot(void 0), (o || this.rt !== this.lt) && (this.U = s, this.ht = null === (e = t.options) || void 0 === e ? void 0 : e.host, this.ot(this.lt = t.element)), _litHtml.nothing;
      }
    }, {
      key: "ot",
      value: function ot(i) {
        var t;

        if ("function" == typeof this.U) {
          var _s = null !== (t = this.ht) && void 0 !== t ? t : globalThis;

          var _e = h.get(_s);

          void 0 === _e && (_e = new WeakMap(), h.set(_s, _e)), void 0 !== _e.get(this.U) && this.U.call(this.ht, void 0), _e.set(this.U, i), void 0 !== i && this.U.call(this.ht, i);
        } else this.U.value = i;
      }
    }, {
      key: "rt",
      get: function get() {
        var i, t, s;
        return "function" == typeof this.U ? null === (t = h.get(null !== (i = this.ht) && void 0 !== i ? i : globalThis)) || void 0 === t ? void 0 : t.get(this.U) : null === (s = this.U) || void 0 === s ? void 0 : s.value;
      }
    }, {
      key: "disconnected",
      value: function disconnected() {
        this.rt === this.lt && this.ot(void 0);
      }
    }, {
      key: "reconnected",
      value: function reconnected() {
        this.ot(this.lt);
      }
    }]);
    return _class;
  }(_asyncDirective.AsyncDirective));
  _exports.ref = n;
});
