// lit-html/directives/repeat.js
define(["exports", "../lit-html.js", "../directive.js", "../directive-helpers.js"], function (_exports, _litHtml, _directive, _directiveHelpers) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.repeat = void 0;

  function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e5) { throw _e5; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e6) { didErr = true; err = _e6; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

  function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

  function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var u = function u(e, s, t) {
    var r = new Map();

    for (var _l = s; _l <= t; _l++) {
      r.set(e[_l], _l);
    }

    return r;
  },
      c = (0, _directive.directive)( /*#__PURE__*/function (_t) {
    babelHelpers.inherits(_class, _t);

    var _super = _createSuper(_class);

    function _class(e) {
      var _this;

      babelHelpers.classCallCheck(this, _class);
      if (_this = _super.call(this, e), e.type !== _directive.PartType.CHILD) throw Error("repeat() can only be used in text expressions");
      return babelHelpers.possibleConstructorReturn(_this);
    }

    babelHelpers.createClass(_class, [{
      key: "Mt",
      value: function Mt(e, s, t) {
        var r;
        void 0 === t ? t = s : void 0 !== s && (r = s);
        var l = [],
            o = [];
        var i = 0;

        var _iterator = _createForOfIteratorHelper(e),
            _step;

        try {
          for (_iterator.s(); !(_step = _iterator.n()).done;) {
            var _s = _step.value;
            l[i] = r ? r(_s, i) : i, o[i] = t(_s, i), i++;
          }
        } catch (err) {
          _iterator.e(err);
        } finally {
          _iterator.f();
        }

        return {
          values: o,
          keys: l
        };
      }
    }, {
      key: "render",
      value: function render(e, s, t) {
        return this.Mt(e, s, t).values;
      }
    }, {
      key: "update",
      value: function update(s, _ref) {
        var _ref2 = babelHelpers.slicedToArray(_ref, 3),
            t = _ref2[0],
            r = _ref2[1],
            c = _ref2[2];

        var d;

        var p = (0, _directiveHelpers.getCommittedValue)(s),
            _this$Mt = this.Mt(t, r, c),
            v = _this$Mt.values,
            a = _this$Mt.keys;

        if (!p) return this.Pt = a, v;
        var h = null !== (d = this.Pt) && void 0 !== d ? d : this.Pt = [],
            m = [];
        var x,
            y,
            j = 0,
            k = p.length - 1,
            w = 0,
            b = v.length - 1;

        for (; j <= k && w <= b;) {
          if (null === p[j]) j++;else if (null === p[k]) k--;else if (h[j] === a[w]) m[w] = (0, _directiveHelpers.setChildPartValue)(p[j], v[w]), j++, w++;else if (h[k] === a[b]) m[b] = (0, _directiveHelpers.setChildPartValue)(p[k], v[b]), k--, b--;else if (h[j] === a[b]) m[b] = (0, _directiveHelpers.setChildPartValue)(p[j], v[b]), (0, _directiveHelpers.insertPart)(s, m[b + 1], p[j]), j++, b--;else if (h[k] === a[w]) m[w] = (0, _directiveHelpers.setChildPartValue)(p[k], v[w]), (0, _directiveHelpers.insertPart)(s, p[j], p[k]), k--, w++;else if (void 0 === x && (x = u(a, w, b), y = u(h, j, k)), x.has(h[j])) {
            if (x.has(h[k])) {
              var _e = y.get(a[w]),
                  _t2 = void 0 !== _e ? p[_e] : null;

              if (null === _t2) {
                var _e2 = (0, _directiveHelpers.insertPart)(s, p[j]);

                (0, _directiveHelpers.setChildPartValue)(_e2, v[w]), m[w] = _e2;
              } else m[w] = (0, _directiveHelpers.setChildPartValue)(_t2, v[w]), (0, _directiveHelpers.insertPart)(s, p[j], _t2), p[_e] = null;

              w++;
            } else (0, _directiveHelpers.removePart)(p[k]), k--;
          } else (0, _directiveHelpers.removePart)(p[j]), j++;
        }

        for (; w <= b;) {
          var _e3 = (0, _directiveHelpers.insertPart)(s, m[b + 1]);

          (0, _directiveHelpers.setChildPartValue)(_e3, v[w]), m[w++] = _e3;
        }

        for (; j <= k;) {
          var _e4 = p[j++];
          null !== _e4 && (0, _directiveHelpers.removePart)(_e4);
        }

        return this.Pt = a, (0, _directiveHelpers.setCommittedValue)(s, m), _litHtml.noChange;
      }
    }]);
    return _class;
  }(_directive.Directive));

  _exports.repeat = c;
});
