// lit-html/directives/until.js
define(["exports", "../lit-html.js", "../directive.js", "../directive-helpers.js", "../async-directive.js"], function (_exports, _litHtml, _directive, _directiveHelpers, _asyncDirective) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.until = void 0;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var e = function e(t) {
    return !(0, _directiveHelpers.isPrimitive)(t) && "function" == typeof t.then;
  },
      o = (0, _directive.directive)( /*#__PURE__*/function (_i) {
    babelHelpers.inherits(_class, _i);

    var _super = _createSuper(_class);

    function _class() {
      var _this;

      babelHelpers.classCallCheck(this, _class);
      _this = _super.apply(this, arguments), _this.Ct = 2147483647, _this.Rt = [];
      return _this;
    }

    babelHelpers.createClass(_class, [{
      key: "render",
      value: function render() {
        var s;

        for (var _len = arguments.length, r = new Array(_len), _key = 0; _key < _len; _key++) {
          r[_key] = arguments[_key];
        }

        return null !== (s = r.find(function (t) {
          return !e(t);
        })) && void 0 !== s ? s : _litHtml.noChange;
      }
    }, {
      key: "update",
      value: function update(r, s) {
        var _this2 = this;

        var i = this.Rt;
        var o = i.length;
        this.Rt = s;

        var _loop = function _loop(_t) {
          var r = s[_t];
          if (!e(r)) return {
            v: (_this2.Ct = _t, r)
          };
          _t < o && r === i[_t] || (_this2.Ct = 2147483647, o = 0, Promise.resolve(r).then(function (t) {
            var s = _this2.Rt.indexOf(r);

            s > -1 && s < _this2.Ct && (_this2.Ct = s, _this2.setValue(t));
          }));
        };

        for (var _t = 0; _t < s.length && !(_t > this.Ct); _t++) {
          var _ret = _loop(_t);

          if (babelHelpers.typeof(_ret) === "object") return _ret.v;
        }

        return _litHtml.noChange;
      }
    }]);
    return _class;
  }(_asyncDirective.AsyncDirective));

  _exports.until = o;
});
