// lit-html/directives/until.js
define(["exports", "../lit-html.js", "../directive.js", "../directive-helpers.js", "../async-directive.js", "./private-async-helpers.js"], function (_exports, _litHtml, _directive, _directiveHelpers, _asyncDirective, _privateAsyncHelpers) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.until = _exports.UntilDirective = void 0;

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var n = function n(t) {
    return !(0, _directiveHelpers.isPrimitive)(t) && "function" == typeof t.then;
  };

  var h = /*#__PURE__*/function (_r) {
    babelHelpers.inherits(h, _r);

    var _super = _createSuper(h);

    function h() {
      var _this;

      babelHelpers.classCallCheck(this, h);
      _this = _super.apply(this, arguments), _this._$Cwt = 1073741823, _this._$Cyt = [], _this._$CG = new _privateAsyncHelpers.PseudoWeakRef(babelHelpers.assertThisInitialized(_this)), _this._$CK = new _privateAsyncHelpers.Pauser();
      return _this;
    }

    babelHelpers.createClass(h, [{
      key: "render",
      value: function render() {
        var i;

        for (var _len = arguments.length, s = new Array(_len), _key = 0; _key < _len; _key++) {
          s[_key] = arguments[_key];
        }

        return null !== (i = s.find(function (t) {
          return !n(t);
        })) && void 0 !== i ? i : _litHtml.noChange;
      }
    }, {
      key: "update",
      value: function update(s, i) {
        var _this2 = this;

        var r = this._$Cyt;
        var e = r.length;
        this._$Cyt = i;
        var o = this._$CG,
            _h = this._$CK;
        this.isConnected || this.disconnected();

        var _loop = function _loop(_t) {
          var s = i[_t];
          if (!n(s)) return {
            v: (_this2._$Cwt = _t, s)
          };
          _t < e && s === r[_t] || (_this2._$Cwt = 1073741823, e = 0, Promise.resolve(s).then( /*#__PURE__*/function () {
            var _ref = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(t) {
              var i, _r2;

              return regeneratorRuntime.wrap(function _callee$(_context) {
                while (1) {
                  switch (_context.prev = _context.next) {
                    case 0:
                      if (!_h.get()) {
                        _context.next = 5;
                        break;
                      }

                      _context.next = 3;
                      return _h.get();

                    case 3:
                      _context.next = 0;
                      break;

                    case 5:
                      i = o.deref();

                      if (void 0 !== i) {
                        _r2 = i._$Cyt.indexOf(s);
                        _r2 > -1 && _r2 < i._$Cwt && (i._$Cwt = _r2, i.setValue(t));
                      }

                    case 7:
                    case "end":
                      return _context.stop();
                  }
                }
              }, _callee);
            }));

            return function (_x) {
              return _ref.apply(this, arguments);
            };
          }()));
        };

        for (var _t = 0; _t < i.length && !(_t > this._$Cwt); _t++) {
          var _ret = _loop(_t);

          if (babelHelpers.typeof(_ret) === "object") return _ret.v;
        }

        return _litHtml.noChange;
      }
    }, {
      key: "disconnected",
      value: function disconnected() {
        this._$CG.disconnect(), this._$CK.pause();
      }
    }, {
      key: "reconnected",
      value: function reconnected() {
        this._$CG.reconnect(this), this._$CK.resume();
      }
    }]);
    return h;
  }(_asyncDirective.AsyncDirective);

  _exports.UntilDirective = h;
  var c = (0, _directive.directive)(h);
  _exports.until = c;
});
