// @haxtheweb/flash-card/lib/flash-card-prompt-img.js
import{html as t,css as e}from"../../../lit/index.js";import"../../simple-icon/lib/simple-icons.js";import"../../simple-icon/lib/simple-icon-lite.js";import{DDD as r}from"../../d-d-d/d-d-d.js";export class FlashCardPromptImg extends r{static get tag(){return"flash-card-image-prompt"}static get styles(){return e`
      :host {
        display: block;
        width: 320px;
        height: 265px;
      }

      img {
        display: flex;
        margin: 25px auto auto;
        height: 200px;
        width: 275px;
        border: 5px solid
          light-dark(
            var(--ddd-theme-default-white),
            var(--ddd-theme-default-coalyGray)
          );
        border-radius: 19px;
        box-shadow: 0 0 10px
          light-dark(rgba(0, 0, 0, 0.5), rgba(255, 255, 255, 0.3));
      }

      .backgroundbox {
        display: flex;
        background-color: light-dark(
          var(--ddd-theme-default-accent),
          var(--ddd-theme-default-accent8)
        );
        border-radius: 19px 19px 0 0;
        height: 265px;
        width: 320px;
      }

      .overlay {
        position: relative;
      }

      .overlay::before {
        content: "";
        width: 100%;
        height: 100%;
        position: absolute;
        border: 1px;
        border-radius: 19px 19px 0px 0px;
      }

      simple-icon-lite {
        --simple-icon-height: 100px;
        --simple-icon-width: 100px;
        color: var(--ddd-theme-default-white);
        transform: translate(-50%, -190%);
        top: 50%;
        left: 50%;
        z-index: 100;
        filter: drop-shadow(2px 2px 4px rgba(0, 0, 0, 0.5));
        animation: iconPop 0.3s ease-out;
      }

      @keyframes iconPop {
        0% {
          transform: translate(-50%, -190%) scale(0);
        }
        80% {
          transform: translate(-50%, -190%) scale(1.1);
        }
        100% {
          transform: translate(-50%, -190%) scale(1);
        }
      }

      :host([status="pending"]) .overlay::before {
        display: flex;
        background: transparent;
      }

      :host([status="correct"]) .overlay::before {
        display: flex;
        background: var(--ddd-theme-default-opportunityGreen);
        filter: opacity(0.75);
        animation: correctOverlay 0.6s ease-in-out;
      }

      :host([status="incorrect"]) .overlay::before {
        display: flex;
        background: var(--ddd-theme-default-wonderPurple);
        filter: opacity(0.75);
        animation: incorrectOverlay 0.4s ease-in-out;
      }

      /* Feedback animations */
      @keyframes correctOverlay {
        0% {
          opacity: 0;
          transform: scale(0.8);
        }
        50% {
          opacity: 1;
          transform: scale(1.05);
        }
        100% {
          opacity: 0.75;
          transform: scale(1);
        }
      }

      @keyframes incorrectOverlay {
        0% {
          opacity: 0;
        }
        25% {
          opacity: 1;
        }
        75% {
          opacity: 1;
        }
        100% {
          opacity: 0.75;
        }
      }
    `}constructor(){super(),this.imgSrc="",this.imgKeyword="grey box",this.status="pending",this.answerIcon=!1,this.icon=""}static get properties(){return{...super.properties,imgSrc:{type:String,reflect:!0,attribute:"img-src"},imgKeyword:{type:String,attribute:"img-keyword"},status:{type:String,reflect:!0},answerIcon:{type:Boolean,reflect:!0},icon:{type:String}}}updated(t){t.forEach(((t,e)=>{"status"===e&&"correct"===this[e]&&(this.answerIcon=!0,this.icon="check"),"status"===e&&"incorrect"===this[e]&&(this.answerIcon=!0,this.icon="cancel"),"status"===e&&"pending"===this[e]&&(this.answerIcon=!1)}))}render(){return t`
      <div>
        <div class="overlay">
          <div class="backgroundbox">
            ${""!==this.imgSrc?t`<img src="${this.imgSrc}" alt="" />`:t`<img
                  src="https://loremflickr.com/320/240/${this.imgKeyword}?lock=1"
                  alt=""
                />`}
          </div>
        </div>
        ${this.answerIcon?t`<simple-icon-lite icon="${this.icon}"></simple-icon-lite>`:""}
      </div>
    `}}globalThis.customElements.define(FlashCardPromptImg.tag,FlashCardPromptImg);
