// @haxtheweb/flash-card/flash-card.js
import{html as t,css as e}from"../../lit/index.js";import{DDD as r}from"../d-d-d/d-d-d.js";import"../simple-toolbar/lib/simple-toolbar-button.js";import"../simple-toast/simple-toast.js";import"./lib/flash-card-answer-box.js";import"./lib/flash-card-prompt-img.js";export class FlashCard extends r{static get tag(){return"flash-card"}constructor(){super(),this.imgKeyword="",this.imgSource=""}static get properties(){return{...super.properties,inverted:{type:Boolean},imgSource:{type:String,attribute:"img-source",reflect:!0},imgKeyword:{type:String,attribute:"img-keyword"},status:{type:String,reflect:!0}}}static get styles(){return[super.styles,e`
        :host {
          display: block;
          box-sizing: content-box !important;
          border: var(--ddd-border-md);
          border-radius: var(--ddd-radius-sm);
          box-shadow: var(--ddd-boxShadow-sm);
          min-width: 320px;
          min-height: 155px;
          padding: var(--ddd-spacing-4);
          margin: var(--ddd-spacing-4);
          transition: all 0.3s ease-in-out;
          background-color: light-dark(
            var(--ddd-theme-default-white),
            var(--ddd-theme-default-coalyGray)
          );
          color: light-dark(
            var(--ddd-theme-default-coalyGray),
            var(--ddd-theme-default-white)
          );
        }
        :host(:focus),
        :host(:focus-within),
        :host(:hover) {
          border-color: var(--ddd-theme-primary);
        }

        /* Status-based styling for correct/incorrect feedback */
        :host([status="correct"]) {
          outline: 3px solid var(--ddd-theme-default-opportunityGreen);
          outline-offset: -3px;
          animation: correctPulse 0.6s ease-in-out;
        }

        :host([status="incorrect"]) {
          outline: 3px dotted var(--ddd-theme-default-wonderPurple);
          outline-offset: -3px;
          animation: incorrectShake 0.4s ease-in-out;
        }

        /* Animations for feedback */
        @keyframes correctPulse {
          0%,
          100% {
            transform: scale(1);
          }
          50% {
            transform: scale(1.02);
          }
        }

        @keyframes incorrectShake {
          0%,
          100% {
            transform: translateX(0);
          }
          25% {
            transform: translateX(-5px);
          }
          75% {
            transform: translateX(5px);
          }
        }

        p {
          color: light-dark(
            var(--ddd-theme-primary),
            var(--ddd-theme-default-linkLight)
          );
          font-family: var(--ddd-font-navigation);
        }

        confetti-container {
          display: flex;
          flex-direction: column;
          align-items: center;
          transition: all 0.3s ease-in-out;
        }

        flash-card-image-prompt {
          align-self: center;
          margin-bottom: var(--ddd-spacing-4);
        }

        flash-card-answer-box {
          width: 100%;
          display: flex;
          flex-direction: column;
          align-items: center;
        }
      `]}statusChanged(t){this.status=t.detail,this.provideFeedback()}provideFeedback(){let t,e,r;if("correct"===this.status?(t="green",e="icons:thumb-up",r="Correct!",this.makeItRain(),this.playSound("success")):"incorrect"===this.status&&(t="red",e="icons:thumb-down",r="Try again!",this.playSound("error")),"correct"===this.status||"incorrect"===this.status){let a=globalThis.document.createElement("simple-icon-lite");a.icon=e,a.style.marginLeft="16px",a.style.setProperty("--simple-icon-height","24px"),a.style.setProperty("--simple-icon-width","24px"),globalThis.dispatchEvent(new CustomEvent("simple-toast-show",{bubbles:!0,composed:!0,cancelable:!0,detail:{text:r,accentColor:t,duration:3e3,slot:a}}))}}makeItRain(){import("../multiple-choice/lib/confetti-container.js").then((t=>{setTimeout((()=>{this.shadowRoot.querySelector("#confetti").setAttribute("popped","")}),0)}))}playSound(t){globalThis.dispatchEvent(new CustomEvent("playaudio",{detail:{sound:t}}))}render(){return t`
      <confetti-container id="confetti">
        ${this.imgSource||this.imgKeyword?t`
              <flash-card-image-prompt
                img-src="${this.imgSource}"
                img-keyword="${this.imgKeyword}"
                status="${this.status}"
              ></flash-card-image-prompt>
            `:""}
        <flash-card-answer-box
          @flash-card-status-change="${this.statusChanged}"
        >
          <div slot="front">
            <slot slot="front" name="front"></slot>
          </div>
          <div slot="back">
            <slot slot="back" name="back"></slot>
          </div>
        </flash-card-answer-box>
      </confetti-container>
    `}static get haxProperties(){return new URL(`./lib/${this.tag}.haxProperties.json`,import.meta.url).href}}globalThis.customElements.define(FlashCard.tag,FlashCard);
