// @lrnwebcomponents/simple-fields/simple-fields.js
define(["exports","require","../../@polymer/polymer/polymer-element.js","../../@polymer/polymer/lib/mixins/mutable-data.js","../utils/utils.js","../hax-body-behaviors/lib/HAXWiring.js","../eco-json-schema-form/eco-json-schema-form.js","../eco-json-schema-form/lib/eco-json-schema-object.js","../simple-colors/lib/simple-colors-polymer.js"],(function(e,s,t,l,o,i,a,r,c){"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.SimpleFields=void 0,s=babelHelpers.interopRequireWildcard(s);
/**
   * Copyright 2019 The Pennsylvania State University
   * @license Apache-2.0, see License.md for full text.
   */
class SimpleFields extends((0,l.MutableData)(t.PolymerElement)){static get template(){return t.html`
      <style>
        :host {
          display: block;
          background-color: #ffffff;
          overflow: visible;
        }

        :host([hidden]) {
          display: none;
        }

        eco-json-schema-object {
          width: 50%;
        }
        eco-json-schema-object {
          color: var(--hax-text-color);
          --eco-json-form-color: var(--hax-text-color);
          --eco-json-schema-object-form : {
            -ms-flex: unset;
            -webkit-flex: unset;
            flex: unset;
            -webkit-flex-basis: unset;
            flex-basis: unset;
          }
        }
        eco-json-schema-object .hax-code-editor {
          padding: 0;
        }
      </style>
      <style include="simple-colors-shared-styles-polymer"></style>
      <eco-json-schema-object
        id="schemaobject"
        autofocus$="[[autofocus]]"
        hide-line-numbers$="[[hideLineNumbers]]"
        on-form-changed="_formFieldsChanged"
        schema="[[__validatedSchema]]"
        value="{{value}}"
      ></eco-json-schema-object>
    `}static get haxProperties(){return{}}static get properties(){return{...super.properties,autofocus:{type:Boolean,value:!1},hideLineNumbers:{type:Boolean,value:!1},fields:{type:Array,value:[],observer:"_formFieldsChanged"},value:{type:Object,notify:!0,value:{},observer:"_valueChanged"},__validatedSchema:{type:Array,notify:!0,value:{properties:{}}}}}static get tag(){return"simple-fields"}connectedCallback(){super.connectedCallback(),new Promise((e,t)=>s.default(["./lib/simple-fields-imports.js"],e,t))}_formFieldsChanged(e){this.dispatchEvent(new CustomEvent("fields-changed",{bubbles:!0,cancelable:!0,composed:!0,detail:e?e.detail:this}))}_valueChanged(e,s){JSON.stringify(s)!==JSON.stringify(e)&&(this._setValues(),this.dispatchEvent(new CustomEvent("value-changed",{bubbles:!0,cancelable:!0,composed:!0,detail:this})))}_fieldsChanged(e,s){JSON.stringify(s)!==JSON.stringify(e)&&this._setValues()}_setValues(){let e=new i.HAXWiring,s=e._getHaxJSONSchemaProperty(this.fields,e);for(let t in this.value)s[t]&&(s[t].value=this.value[t]);this.set("__validatedSchema",{properties:s}),this.notifyPath("__validatedSchema.properties.*")}}e.SimpleFields=SimpleFields,window.customElements.define(SimpleFields.tag,SimpleFields)}));
