// @lrnwebcomponents/eco-json-schema-form/lib/eco-json-schema-fieldset.js
define(["exports", "../../../@polymer/polymer/polymer-element.js", "../../../@polymer/app-localize-behavior/app-localize-behavior.js", "../../../@polymer/polymer/lib/legacy/class.js", "../../../@polymer/polymer/lib/utils/render-status.js", "../../../@polymer/iron-flex-layout/iron-flex-layout-classes.js"], function (_exports, _polymerElement, _appLocalizeBehavior, _class, _renderStatus, _ironFlexLayoutClasses) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.EcoJsonSchemaFieldset = void 0;

  function _templateObject_ef010fc0d10411ea8ac18ff2c3761fda() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style include=\"iron-flex iron-flex-alignment\">\n        :host {\n          color: var(--eco-json-form-color);\n          background-color: var(--eco-json-form-bg);\n          font-family: var(--eco-json-form-font-family);\n          margin-bottom: 15px;\n        }\n        :host([hidden]) {\n          display: none;\n        }\n        :host fieldset {\n          border-radius: var(--eco-json-form-border-radius);\n          border-style: solid;\n          border-width: 1px;\n          border-color: var(--eco-json-form-faded-color);\n          transition: all 0.5s;\n        }\n        :host legend {\n          transition: all 0.5s;\n          color: var(--eco-json-form-faded-color);\n        }\n        :host fieldset:focus #legend,\n        :host fieldset:focus-within #legend {\n          color: var(--eco-json-form-active-color);\n        }\n        :host eco-json-schema-object {\n          flex: 1 0 auto;\n        }\n      </style>\n      <fieldset>\n        <legend id=\"legend\" class=\"flex\" hidden$=\"[[!schema.title]]\">\n          [[schema.title]]\n        </legend>\n        <div hidden$=\"[[!schema.description]]\">[[item.description]]</div>\n        <div class=\"item-fields\"></div>\n      </fieldset>\n    "], ["\n      <style include=\"iron-flex iron-flex-alignment\">\n        :host {\n          color: var(--eco-json-form-color);\n          background-color: var(--eco-json-form-bg);\n          font-family: var(--eco-json-form-font-family);\n          margin-bottom: 15px;\n        }\n        :host([hidden]) {\n          display: none;\n        }\n        :host fieldset {\n          border-radius: var(--eco-json-form-border-radius);\n          border-style: solid;\n          border-width: 1px;\n          border-color: var(--eco-json-form-faded-color);\n          transition: all 0.5s;\n        }\n        :host legend {\n          transition: all 0.5s;\n          color: var(--eco-json-form-faded-color);\n        }\n        :host fieldset:focus #legend,\n        :host fieldset:focus-within #legend {\n          color: var(--eco-json-form-active-color);\n        }\n        :host eco-json-schema-object {\n          flex: 1 0 auto;\n        }\n      </style>\n      <fieldset>\n        <legend id=\"legend\" class=\"flex\" hidden\\$=\"[[!schema.title]]\">\n          [[schema.title]]\n        </legend>\n        <div hidden$=\"[[!schema.description]]\">[[item.description]]</div>\n        <div class=\"item-fields\"></div>\n      </fieldset>\n    "]);

    _templateObject_ef010fc0d10411ea8ac18ff2c3761fda = function _templateObject_ef010fc0d10411ea8ac18ff2c3761fda() {
      return data;
    };

    return data;
  }

  /**
  `eco-json-schema-fieldset` takes in a JSON schema of type fieldset and builds a form,
  exposing a `value` property that represents an array described by the schema.
  
  Please see the `eco-json-schema-object` documentation for further information.
  
  @group eco Elements
  @element eco-json-schema-fieldset
  * @demo demo/index.html
  */
  var EcoJsonSchemaFieldset =
  /*#__PURE__*/
  function (_mixinBehaviors) {
    babelHelpers.inherits(EcoJsonSchemaFieldset, _mixinBehaviors);
    babelHelpers.createClass(EcoJsonSchemaFieldset, null, [{
      key: "tag",
      get: function get() {
        return "eco-json-schema-fieldset";
      }
    }]);

    function EcoJsonSchemaFieldset() {
      babelHelpers.classCallCheck(this, EcoJsonSchemaFieldset);
      return babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(EcoJsonSchemaFieldset).call(this));
    }

    babelHelpers.createClass(EcoJsonSchemaFieldset, [{
      key: "ready",
      value: function ready() {
        babelHelpers.get(babelHelpers.getPrototypeOf(EcoJsonSchemaFieldset.prototype), "ready", this).call(this);

        this._schemaChanged();
      }
      /**
       * updates the array fields if the schema (which includes values) changes
       */

    }, {
      key: "_schemaChanged",
      value: function _schemaChanged() {
        var _this = this;

        //make sure the content is there first
        (0, _renderStatus.afterNextRender)(this, function () {
          _this.shadowRoot.querySelectorAll(".item-fields").forEach(function (item, index) {
            var prefix = "".concat(_this.propertyName);

            _this.dispatchEvent(new CustomEvent("build-fieldset", {
              bubbles: false,
              cancelable: true,
              composed: true,
              detail: {
                container: item,
                path: prefix,
                prefix: prefix,
                properties: _this.schema.properties,
                type: EcoJsonSchemaFieldset.tag,
                value: _this.schema.properties[index].value || {}
              }
            }));
          });
        });
      }
    }], [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_ef010fc0d10411ea8ac18ff2c3761fda());
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          propertyName: {
            type: String,
            value: null
          },
          schema: {
            type: Object,
            value: {}
          }
        };
      }
    }]);
    return EcoJsonSchemaFieldset;
  }((0, _class.mixinBehaviors)([_appLocalizeBehavior.AppLocalizeBehavior], _polymerElement.PolymerElement));

  _exports.EcoJsonSchemaFieldset = EcoJsonSchemaFieldset;
  window.customElements.define(EcoJsonSchemaFieldset.tag, EcoJsonSchemaFieldset);
});
