// @lrnwebcomponents/chartist-render/chartist-render.js
define(["exports", "meta", "../../lit-element/lit-element.js", "../schema-behaviors/schema-behaviors.js", "../utils/utils.js", "../es-global-bridge/es-global-bridge.js"], function (_exports, meta, _litElement, _schemaBehaviors, _utils, _esGlobalBridge) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ChartistRender = void 0;
  meta = babelHelpers.interopRequireWildcard(meta);

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject15_5b205ae0425211ea8ef51d3abfe05f78() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        .ct-label {\n          fill: rgba(0, 0, 0, 0.4);\n          color: rgba(0, 0, 0, 0.4);\n          font-size: 0.75rem;\n          line-height: 1;\n        }\n\n        .ct-chart-line .ct-label,\n        .ct-chart-bar .ct-label {\n          display: block;\n          display: -webkit-box;\n          display: -moz-box;\n          display: -ms-flexbox;\n          display: -webkit-flex;\n          display: flex;\n        }\n\n        .ct-chart-pie .ct-label,\n        .ct-chart-donut .ct-label {\n          dominant-baseline: central;\n        }\n\n        .ct-label.ct-horizontal.ct-start {\n          -webkit-box-align: flex-end;\n          -webkit-align-items: flex-end;\n          -ms-flex-align: flex-end;\n          align-items: flex-end;\n          -webkit-box-pack: flex-start;\n          -webkit-justify-content: flex-start;\n          -ms-flex-pack: flex-start;\n          justify-content: flex-start;\n          text-align: left;\n          text-anchor: start;\n        }\n\n        .ct-label.ct-horizontal.ct-end {\n          -webkit-box-align: flex-start;\n          -webkit-align-items: flex-start;\n          -ms-flex-align: flex-start;\n          align-items: flex-start;\n          -webkit-box-pack: flex-start;\n          -webkit-justify-content: flex-start;\n          -ms-flex-pack: flex-start;\n          justify-content: flex-start;\n          text-align: left;\n          text-anchor: start;\n        }\n\n        .ct-label.ct-vertical.ct-start {\n          -webkit-box-align: flex-end;\n          -webkit-align-items: flex-end;\n          -ms-flex-align: flex-end;\n          align-items: flex-end;\n          -webkit-box-pack: flex-end;\n          -webkit-justify-content: flex-end;\n          -ms-flex-pack: flex-end;\n          justify-content: flex-end;\n          text-align: right;\n          text-anchor: end;\n        }\n\n        .ct-label.ct-vertical.ct-end {\n          -webkit-box-align: flex-end;\n          -webkit-align-items: flex-end;\n          -ms-flex-align: flex-end;\n          align-items: flex-end;\n          -webkit-box-pack: flex-start;\n          -webkit-justify-content: flex-start;\n          -ms-flex-pack: flex-start;\n          justify-content: flex-start;\n          text-align: left;\n          text-anchor: start;\n        }\n\n        .ct-chart-bar .ct-label.ct-horizontal.ct-start {\n          -webkit-box-align: flex-end;\n          -webkit-align-items: flex-end;\n          -ms-flex-align: flex-end;\n          align-items: flex-end;\n          -webkit-box-pack: center;\n          -webkit-justify-content: center;\n          -ms-flex-pack: center;\n          justify-content: center;\n          text-align: center;\n          text-anchor: start;\n        }\n\n        .ct-chart-bar .ct-label.ct-horizontal.ct-end {\n          -webkit-box-align: flex-start;\n          -webkit-align-items: flex-start;\n          -ms-flex-align: flex-start;\n          align-items: flex-start;\n          -webkit-box-pack: center;\n          -webkit-justify-content: center;\n          -ms-flex-pack: center;\n          justify-content: center;\n          text-align: center;\n          text-anchor: start;\n        }\n\n        .ct-chart-bar.ct-horizontal-bars .ct-label.ct-horizontal.ct-start {\n          -webkit-box-align: flex-end;\n          -webkit-align-items: flex-end;\n          -ms-flex-align: flex-end;\n          align-items: flex-end;\n          -webkit-box-pack: flex-start;\n          -webkit-justify-content: flex-start;\n          -ms-flex-pack: flex-start;\n          justify-content: flex-start;\n          text-align: left;\n          text-anchor: start;\n        }\n\n        .ct-chart-bar.ct-horizontal-bars .ct-label.ct-horizontal.ct-end {\n          -webkit-box-align: flex-start;\n          -webkit-align-items: flex-start;\n          -ms-flex-align: flex-start;\n          align-items: flex-start;\n          -webkit-box-pack: flex-start;\n          -webkit-justify-content: flex-start;\n          -ms-flex-pack: flex-start;\n          justify-content: flex-start;\n          text-align: left;\n          text-anchor: start;\n        }\n\n        .ct-chart-bar.ct-horizontal-bars .ct-label.ct-vertical.ct-start {\n          -webkit-box-align: center;\n          -webkit-align-items: center;\n          -ms-flex-align: center;\n          align-items: center;\n          -webkit-box-pack: flex-end;\n          -webkit-justify-content: flex-end;\n          -ms-flex-pack: flex-end;\n          justify-content: flex-end;\n          text-align: right;\n          text-anchor: end;\n        }\n\n        .ct-chart-bar.ct-horizontal-bars .ct-label.ct-vertical.ct-end {\n          -webkit-box-align: center;\n          -webkit-align-items: center;\n          -ms-flex-align: center;\n          align-items: center;\n          -webkit-box-pack: flex-start;\n          -webkit-justify-content: flex-start;\n          -ms-flex-pack: flex-start;\n          justify-content: flex-start;\n          text-align: left;\n          text-anchor: end;\n        }\n\n        .ct-grid {\n          stroke: rgba(0, 0, 0, 0.2);\n          stroke-width: 1px;\n          stroke-dasharray: 2px;\n        }\n\n        .ct-grid-background {\n          fill: none;\n        }\n\n        .ct-point {\n          stroke-width: 10px;\n          stroke-linecap: round;\n        }\n\n        .ct-line {\n          fill: none;\n          stroke-width: 4px;\n        }\n\n        .ct-area {\n          stroke: none;\n          fill-opacity: 0.1;\n        }\n\n        .ct-bar {\n          fill: none;\n          stroke-width: 10px;\n        }\n\n        .ct-slice-donut {\n          fill: none;\n          stroke-width: 60px;\n        }\n\n        .ct-series-a .ct-point,\n        .ct-series-a .ct-line,\n        .ct-series-a .ct-bar,\n        .ct-series-a .ct-slice-donut {\n          stroke: #d70206;\n        }\n\n        .ct-series-a .ct-slice-pie,\n        .ct-series-a .ct-slice-donut-solid,\n        .ct-series-a .ct-area {\n          fill: #d70206;\n        }\n\n        .ct-series-b .ct-point,\n        .ct-series-b .ct-line,\n        .ct-series-b .ct-bar,\n        .ct-series-b .ct-slice-donut {\n          stroke: #f05b4f;\n        }\n\n        .ct-series-b .ct-slice-pie,\n        .ct-series-b .ct-slice-donut-solid,\n        .ct-series-b .ct-area {\n          fill: #f05b4f;\n        }\n\n        .ct-series-c .ct-point,\n        .ct-series-c .ct-line,\n        .ct-series-c .ct-bar,\n        .ct-series-c .ct-slice-donut {\n          stroke: #f4c63d;\n        }\n\n        .ct-series-c .ct-slice-pie,\n        .ct-series-c .ct-slice-donut-solid,\n        .ct-series-c .ct-area {\n          fill: #f4c63d;\n        }\n\n        .ct-series-d .ct-point,\n        .ct-series-d .ct-line,\n        .ct-series-d .ct-bar,\n        .ct-series-d .ct-slice-donut {\n          stroke: #d17905;\n        }\n\n        .ct-series-d .ct-slice-pie,\n        .ct-series-d .ct-slice-donut-solid,\n        .ct-series-d .ct-area {\n          fill: #d17905;\n        }\n\n        .ct-series-e .ct-point,\n        .ct-series-e .ct-line,\n        .ct-series-e .ct-bar,\n        .ct-series-e .ct-slice-donut {\n          stroke: #453d3f;\n        }\n\n        .ct-series-e .ct-slice-pie,\n        .ct-series-e .ct-slice-donut-solid,\n        .ct-series-e .ct-area {\n          fill: #453d3f;\n        }\n\n        .ct-series-f .ct-point,\n        .ct-series-f .ct-line,\n        .ct-series-f .ct-bar,\n        .ct-series-f .ct-slice-donut {\n          stroke: #59922b;\n        }\n\n        .ct-series-f .ct-slice-pie,\n        .ct-series-f .ct-slice-donut-solid,\n        .ct-series-f .ct-area {\n          fill: #59922b;\n        }\n\n        .ct-series-g .ct-point,\n        .ct-series-g .ct-line,\n        .ct-series-g .ct-bar,\n        .ct-series-g .ct-slice-donut {\n          stroke: #0544d3;\n        }\n\n        .ct-series-g .ct-slice-pie,\n        .ct-series-g .ct-slice-donut-solid,\n        .ct-series-g .ct-area {\n          fill: #0544d3;\n        }\n\n        .ct-series-h .ct-point,\n        .ct-series-h .ct-line,\n        .ct-series-h .ct-bar,\n        .ct-series-h .ct-slice-donut {\n          stroke: #6b0392;\n        }\n\n        .ct-series-h .ct-slice-pie,\n        .ct-series-h .ct-slice-donut-solid,\n        .ct-series-h .ct-area {\n          fill: #6b0392;\n        }\n\n        .ct-series-i .ct-point,\n        .ct-series-i .ct-line,\n        .ct-series-i .ct-bar,\n        .ct-series-i .ct-slice-donut {\n          stroke: #f05b4f;\n        }\n\n        .ct-series-i .ct-slice-pie,\n        .ct-series-i .ct-slice-donut-solid,\n        .ct-series-i .ct-area {\n          fill: #f05b4f;\n        }\n\n        .ct-series-j .ct-point,\n        .ct-series-j .ct-line,\n        .ct-series-j .ct-bar,\n        .ct-series-j .ct-slice-donut {\n          stroke: #dda458;\n        }\n\n        .ct-series-j .ct-slice-pie,\n        .ct-series-j .ct-slice-donut-solid,\n        .ct-series-j .ct-area {\n          fill: #dda458;\n        }\n\n        .ct-series-k .ct-point,\n        .ct-series-k .ct-line,\n        .ct-series-k .ct-bar,\n        .ct-series-k .ct-slice-donut {\n          stroke: #eacf7d;\n        }\n\n        .ct-series-k .ct-slice-pie,\n        .ct-series-k .ct-slice-donut-solid,\n        .ct-series-k .ct-area {\n          fill: #eacf7d;\n        }\n\n        .ct-series-l .ct-point,\n        .ct-series-l .ct-line,\n        .ct-series-l .ct-bar,\n        .ct-series-l .ct-slice-donut {\n          stroke: #86797d;\n        }\n\n        .ct-series-l .ct-slice-pie,\n        .ct-series-l .ct-slice-donut-solid,\n        .ct-series-l .ct-area {\n          fill: #86797d;\n        }\n\n        .ct-series-m .ct-point,\n        .ct-series-m .ct-line,\n        .ct-series-m .ct-bar,\n        .ct-series-m .ct-slice-donut {\n          stroke: #b2c326;\n        }\n\n        .ct-series-m .ct-slice-pie,\n        .ct-series-m .ct-slice-donut-solid,\n        .ct-series-m .ct-area {\n          fill: #b2c326;\n        }\n\n        .ct-series-n .ct-point,\n        .ct-series-n .ct-line,\n        .ct-series-n .ct-bar,\n        .ct-series-n .ct-slice-donut {\n          stroke: #6188e2;\n        }\n\n        .ct-series-n .ct-slice-pie,\n        .ct-series-n .ct-slice-donut-solid,\n        .ct-series-n .ct-area {\n          fill: #6188e2;\n        }\n\n        .ct-series-o .ct-point,\n        .ct-series-o .ct-line,\n        .ct-series-o .ct-bar,\n        .ct-series-o .ct-slice-donut {\n          stroke: #a748ca;\n        }\n\n        .ct-series-o .ct-slice-pie,\n        .ct-series-o .ct-slice-donut-solid,\n        .ct-series-o .ct-area {\n          fill: #a748ca;\n        }\n\n        .ct-square {\n          display: block;\n          position: relative;\n          width: 100%;\n        }\n        .ct-square:before {\n          display: block;\n          float: left;\n          content: \"\";\n          width: 0;\n          height: 0;\n          padding-bottom: 100%;\n        }\n        .ct-square:after {\n          content: \"\";\n          display: table;\n          clear: both;\n        }\n        .ct-square > svg {\n          display: block;\n          position: absolute;\n          top: 0;\n          left: 0;\n        }\n\n        .ct-minor-second {\n          display: block;\n          position: relative;\n          width: 100%;\n        }\n        .ct-minor-second:before {\n          display: block;\n          float: left;\n          content: \"\";\n          width: 0;\n          height: 0;\n          padding-bottom: 93.75%;\n        }\n        .ct-minor-second:after {\n          content: \"\";\n          display: table;\n          clear: both;\n        }\n        .ct-minor-second > svg {\n          display: block;\n          position: absolute;\n          top: 0;\n          left: 0;\n        }\n\n        .ct-major-second {\n          display: block;\n          position: relative;\n          width: 100%;\n        }\n        .ct-major-second:before {\n          display: block;\n          float: left;\n          content: \"\";\n          width: 0;\n          height: 0;\n          padding-bottom: 88.88889%;\n        }\n        .ct-major-second:after {\n          content: \"\";\n          display: table;\n          clear: both;\n        }\n        .ct-major-second > svg {\n          display: block;\n          position: absolute;\n          top: 0;\n          left: 0;\n        }\n\n        .ct-minor-third {\n          display: block;\n          position: relative;\n          width: 100%;\n        }\n        .ct-minor-third:before {\n          display: block;\n          float: left;\n          content: \"\";\n          width: 0;\n          height: 0;\n          padding-bottom: 83.33333%;\n        }\n        .ct-minor-third:after {\n          content: \"\";\n          display: table;\n          clear: both;\n        }\n        .ct-minor-third > svg {\n          display: block;\n          position: absolute;\n          top: 0;\n          left: 0;\n        }\n\n        .ct-major-third {\n          display: block;\n          position: relative;\n          width: 100%;\n        }\n        .ct-major-third:before {\n          display: block;\n          float: left;\n          content: \"\";\n          width: 0;\n          height: 0;\n          padding-bottom: 80%;\n        }\n        .ct-major-third:after {\n          content: \"\";\n          display: table;\n          clear: both;\n        }\n        .ct-major-third > svg {\n          display: block;\n          position: absolute;\n          top: 0;\n          left: 0;\n        }\n\n        .ct-perfect-fourth {\n          display: block;\n          position: relative;\n          width: 100%;\n        }\n        .ct-perfect-fourth:before {\n          display: block;\n          float: left;\n          content: \"\";\n          width: 0;\n          height: 0;\n          padding-bottom: 75%;\n        }\n        .ct-perfect-fourth:after {\n          content: \"\";\n          display: table;\n          clear: both;\n        }\n        .ct-perfect-fourth > svg {\n          display: block;\n          position: absolute;\n          top: 0;\n          left: 0;\n        }\n\n        .ct-perfect-fifth {\n          display: block;\n          position: relative;\n          width: 100%;\n        }\n        .ct-perfect-fifth:before {\n          display: block;\n          float: left;\n          content: \"\";\n          width: 0;\n          height: 0;\n          padding-bottom: 66.66667%;\n        }\n        .ct-perfect-fifth:after {\n          content: \"\";\n          display: table;\n          clear: both;\n        }\n        .ct-perfect-fifth > svg {\n          display: block;\n          position: absolute;\n          top: 0;\n          left: 0;\n        }\n\n        .ct-minor-sixth {\n          display: block;\n          position: relative;\n          width: 100%;\n        }\n        .ct-minor-sixth:before {\n          display: block;\n          float: left;\n          content: \"\";\n          width: 0;\n          height: 0;\n          padding-bottom: 62.5%;\n        }\n        .ct-minor-sixth:after {\n          content: \"\";\n          display: table;\n          clear: both;\n        }\n        .ct-minor-sixth > svg {\n          display: block;\n          position: absolute;\n          top: 0;\n          left: 0;\n        }\n\n        .ct-golden-section {\n          display: block;\n          position: relative;\n          width: 100%;\n        }\n        .ct-golden-section:before {\n          display: block;\n          float: left;\n          content: \"\";\n          width: 0;\n          height: 0;\n          padding-bottom: 61.8047%;\n        }\n        .ct-golden-section:after {\n          content: \"\";\n          display: table;\n          clear: both;\n        }\n        .ct-golden-section > svg {\n          display: block;\n          position: absolute;\n          top: 0;\n          left: 0;\n        }\n\n        .ct-major-sixth {\n          display: block;\n          position: relative;\n          width: 100%;\n        }\n        .ct-major-sixth:before {\n          display: block;\n          float: left;\n          content: \"\";\n          width: 0;\n          height: 0;\n          padding-bottom: 60%;\n        }\n        .ct-major-sixth:after {\n          content: \"\";\n          display: table;\n          clear: both;\n        }\n        .ct-major-sixth > svg {\n          display: block;\n          position: absolute;\n          top: 0;\n          left: 0;\n        }\n\n        .ct-minor-seventh {\n          display: block;\n          position: relative;\n          width: 100%;\n        }\n        .ct-minor-seventh:before {\n          display: block;\n          float: left;\n          content: \"\";\n          width: 0;\n          height: 0;\n          padding-bottom: 56.25%;\n        }\n        .ct-minor-seventh:after {\n          content: \"\";\n          display: table;\n          clear: both;\n        }\n        .ct-minor-seventh > svg {\n          display: block;\n          position: absolute;\n          top: 0;\n          left: 0;\n        }\n\n        .ct-major-seventh {\n          display: block;\n          position: relative;\n          width: 100%;\n        }\n        .ct-major-seventh:before {\n          display: block;\n          float: left;\n          content: \"\";\n          width: 0;\n          height: 0;\n          padding-bottom: 53.33333%;\n        }\n        .ct-major-seventh:after {\n          content: \"\";\n          display: table;\n          clear: both;\n        }\n        .ct-major-seventh > svg {\n          display: block;\n          position: absolute;\n          top: 0;\n          left: 0;\n        }\n\n        .ct-octave {\n          display: block;\n          position: relative;\n          width: 100%;\n        }\n        .ct-octave:before {\n          display: block;\n          float: left;\n          content: \"\";\n          width: 0;\n          height: 0;\n          padding-bottom: 50%;\n        }\n        .ct-octave:after {\n          content: \"\";\n          display: table;\n          clear: both;\n        }\n        .ct-octave > svg {\n          display: block;\n          position: absolute;\n          top: 0;\n          left: 0;\n        }\n\n        .ct-major-tenth {\n          display: block;\n          position: relative;\n          width: 100%;\n        }\n        .ct-major-tenth:before {\n          display: block;\n          float: left;\n          content: \"\";\n          width: 0;\n          height: 0;\n          padding-bottom: 40%;\n        }\n        .ct-major-tenth:after {\n          content: \"\";\n          display: table;\n          clear: both;\n        }\n        .ct-major-tenth > svg {\n          display: block;\n          position: absolute;\n          top: 0;\n          left: 0;\n        }\n\n        .ct-major-eleventh {\n          display: block;\n          position: relative;\n          width: 100%;\n        }\n        .ct-major-eleventh:before {\n          display: block;\n          float: left;\n          content: \"\";\n          width: 0;\n          height: 0;\n          padding-bottom: 37.5%;\n        }\n        .ct-major-eleventh:after {\n          content: \"\";\n          display: table;\n          clear: both;\n        }\n        .ct-major-eleventh > svg {\n          display: block;\n          position: absolute;\n          top: 0;\n          left: 0;\n        }\n\n        .ct-major-twelfth {\n          display: block;\n          position: relative;\n          width: 100%;\n        }\n        .ct-major-twelfth:before {\n          display: block;\n          float: left;\n          content: \"\";\n          width: 0;\n          height: 0;\n          padding-bottom: 33.33333%;\n        }\n        .ct-major-twelfth:after {\n          content: \"\";\n          display: table;\n          clear: both;\n        }\n        .ct-major-twelfth > svg {\n          display: block;\n          position: absolute;\n          top: 0;\n          left: 0;\n        }\n\n        .ct-double-octave {\n          display: block;\n          position: relative;\n          width: 100%;\n        }\n        .ct-double-octave:before {\n          display: block;\n          float: left;\n          content: \"\";\n          width: 0;\n          height: 0;\n          padding-bottom: 25%;\n        }\n        .ct-double-octave:after {\n          content: \"\";\n          display: table;\n          clear: both;\n        }\n        .ct-double-octave > svg {\n          display: block;\n          position: absolute;\n          top: 0;\n          left: 0;\n        }\n\n        :host {\n          display: block;\n        }\n\n        .a11y {\n          position: absolute;\n          left: -999999px;\n          height: 0;\n          overflow: hidden;\n        }\n\n        .ct-label {\n          fill: var(--chartist-text-color, rgba(0, 0, 0, 0.4));\n          color: var(--chartist-text-color, rgba(0, 0, 0, 0.4));\n          font-size: var(--chartist-text-size, 0.75rem);\n          line-height: var(--chartist-line-height, 1);\n        }\n\n        .ct-grid {\n          stroke: var(--chartist-grid-color, rgba(0, 0, 0, 0.2));\n          stroke-width: 1px;\n          stroke-dasharray: 2px;\n        }\n\n        .ct-series-a .ct-point,\n        .ct-series-a .ct-line,\n        .ct-series-a .ct-bar,\n        .ct-series-a .ct-slice-donut {\n          stroke: var(--chartist-color-1, #d70206);\n        }\n\n        .ct-series-a .ct-slice-pie,\n        .ct-series-a .ct-slice-donut-solid,\n        .ct-series-a .ct-area {\n          fill: var(--chartist-color-1, #d70206);\n        }\n\n        .ct-series-b .ct-point,\n        .ct-series-b .ct-line,\n        .ct-series-b .ct-bar,\n        .ct-series-b .ct-slice-donut {\n          stroke: var(--chartist-color-2, #f05b4f);\n        }\n\n        .ct-series-b .ct-slice-pie,\n        .ct-series-b .ct-slice-donut-solid,\n        .ct-series-b .ct-area {\n          fill: var(--chartist-color-2, #f05b4f);\n        }\n\n        .ct-series-c .ct-point,\n        .ct-series-c .ct-line,\n        .ct-series-c .ct-bar,\n        .ct-series-c .ct-slice-donut {\n          stroke: var(--chartist-color-3, #f4c63d);\n        }\n\n        .ct-series-c .ct-slice-pie,\n        .ct-series-c .ct-slice-donut-solid,\n        .ct-series-c .ct-area {\n          fill: var(--chartist-color-3, #f4c63d);\n        }\n\n        .ct-series-d .ct-point,\n        .ct-series-d .ct-line,\n        .ct-series-d .ct-bar,\n        .ct-series-d .ct-slice-donut {\n          stroke: var(--chartist-color-4, #d17905);\n        }\n\n        .ct-series-d .ct-slice-pie,\n        .ct-series-d .ct-slice-donut-solid,\n        .ct-series-d .ct-area {\n          fill: var(--chartist-color-4, #d17905);\n        }\n\n        .ct-series-e .ct-point,\n        .ct-series-e .ct-line,\n        .ct-series-e .ct-bar,\n        .ct-series-e .ct-slice-donut {\n          stroke: var(--chartist-color-5, #453d3f);\n        }\n\n        .ct-series-e .ct-slice-pie,\n        .ct-series-e .ct-slice-donut-solid,\n        .ct-series-e .ct-area {\n          fill: var(--chartist-color-5, #453d3f);\n        }\n\n        .ct-series-f .ct-point,\n        .ct-series-f .ct-line,\n        .ct-series-f .ct-bar,\n        .ct-series-f .ct-slice-donut {\n          stroke: var(--chartist-color-6, #59922b);\n        }\n\n        .ct-series-f .ct-slice-pie,\n        .ct-series-f .ct-slice-donut-solid,\n        .ct-series-f .ct-area {\n          fill: var(--chartist-color-6, #59922b);\n        }\n\n        .ct-series-g .ct-point,\n        .ct-series-g .ct-line,\n        .ct-series-g .ct-bar,\n        .ct-series-g .ct-slice-donut {\n          stroke: var(--chartist-color-7, #0544d3);\n        }\n\n        .ct-series-g .ct-slice-pie,\n        .ct-series-g .ct-slice-donut-solid,\n        .ct-series-g .ct-area {\n          fill: var(--chartist-color-7, #0544d3);\n        }\n\n        .ct-series-h .ct-point,\n        .ct-series-h .ct-line,\n        .ct-series-h .ct-bar,\n        .ct-series-h .ct-slice-donut {\n          stroke: var(--chartist-color-8, #6b0392);\n        }\n\n        .ct-series-h .ct-slice-pie,\n        .ct-series-h .ct-slice-donut-solid,\n        .ct-series-h .ct-area {\n          fill: var(--chartist-color-8, #6b0392);\n        }\n\n        .ct-series-i .ct-point,\n        .ct-series-i .ct-line,\n        .ct-series-i .ct-bar,\n        .ct-series-i .ct-slice-donut {\n          stroke: var(--chartist-color-9, #f05b4f);\n        }\n\n        .ct-series-i .ct-slice-pie,\n        .ct-series-i .ct-slice-donut-solid,\n        .ct-series-i .ct-area {\n          fill: var(--chartist-color-9, #f05b4f);\n        }\n\n        .ct-series-j .ct-point,\n        .ct-series-j .ct-line,\n        .ct-series-j .ct-bar,\n        .ct-series-j .ct-slice-donut {\n          stroke: var(--chartist-color-10, #dda458);\n        }\n\n        .ct-series-j .ct-slice-pie,\n        .ct-series-j .ct-slice-donut-solid,\n        .ct-series-j .ct-area {\n          fill: var(--chartist-color-10, #dda458);\n        }\n\n        .ct-series-k .ct-point,\n        .ct-series-k .ct-line,\n        .ct-series-k .ct-bar,\n        .ct-series-k .ct-slice-donut {\n          stroke: var(--chartist-color-11, #eacf7d);\n        }\n\n        .ct-series-k .ct-slice-pie,\n        .ct-series-k .ct-slice-donut-solid,\n        .ct-series-k .ct-area {\n          fill: var(--chartist-color-11, #eacf7d);\n        }\n\n        .ct-series-l .ct-point,\n        .ct-series-l .ct-line,\n        .ct-series-l .ct-bar,\n        .ct-series-l .ct-slice-donut {\n          stroke: var(--chartist-color-12, #86797d);\n        }\n\n        .ct-series-l .ct-slice-pie,\n        .ct-series-l .ct-slice-donut-solid,\n        .ct-series-l .ct-area {\n          fill: var(--chartist-color-12, #86797d);\n        }\n\n        .ct-series-m .ct-point,\n        .ct-series-m .ct-line,\n        .ct-series-m .ct-bar,\n        .ct-series-m .ct-slice-donut {\n          stroke: var(--chartist-color-13, #b2c326);\n        }\n\n        .ct-series-m .ct-slice-pie,\n        .ct-series-m .ct-slice-donut-solid,\n        .ct-series-m .ct-area {\n          fill: var(--chartist-color-13, #b2c326);\n        }\n\n        .ct-series-n .ct-point,\n        .ct-series-n .ct-line,\n        .ct-series-n .ct-bar,\n        .ct-series-n .ct-slice-donut {\n          stroke: var(--chartist-color-14, #6188e2);\n        }\n\n        .ct-series-n .ct-slice-pie,\n        .ct-series-n .ct-slice-donut-solid,\n        .ct-series-n .ct-area {\n          fill: var(--chartist-color-14, #6188e2);\n        }\n\n        .ct-series-o .ct-point,\n        .ct-series-o .ct-line,\n        .ct-series-o .ct-bar,\n        .ct-series-o .ct-slice-donut {\n          stroke: var(--chartist-color-15, #a748ca);\n        }\n\n        .ct-series-o .ct-slice-pie,\n        .ct-series-o .ct-slice-donut-solid,\n        .ct-series-o .ct-area {\n          fill: var(--chartist-color-15, #a748ca);\n        }\n      "]);

    _templateObject15_5b205ae0425211ea8ef51d3abfe05f78 = function _templateObject15_5b205ae0425211ea8ef51d3abfe05f78() {
      return data;
    };

    return data;
  }

  function _templateObject14_5b205ae0425211ea8ef51d3abfe05f78() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  A ", " chart.\n                "]);

    _templateObject14_5b205ae0425211ea8ef51d3abfe05f78 = function _templateObject14_5b205ae0425211ea8ef51d3abfe05f78() {
      return data;
    };

    return data;
  }

  function _templateObject13_5b205ae0425211ea8ef51d3abfe05f78() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                  <div id=\"", "-title\" class=\"a11y\">\n                    <div>", "</div>\n                    <div>", "</div>\n                  </div>\n                "]);

    _templateObject13_5b205ae0425211ea8ef51d3abfe05f78 = function _templateObject13_5b205ae0425211ea8ef51d3abfe05f78() {
      return data;
    };

    return data;
  }

  function _templateObject12_5b205ae0425211ea8ef51d3abfe05f78() {
    var data = babelHelpers.taggedTemplateLiteral(["", "\n</div>"]);

    _templateObject12_5b205ae0425211ea8ef51d3abfe05f78 = function _templateObject12_5b205ae0425211ea8ef51d3abfe05f78() {
      return data;
    };

    return data;
  }

  function _templateObject11_5b205ae0425211ea8ef51d3abfe05f78() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                              <td>", "</td>\n                            "]);

    _templateObject11_5b205ae0425211ea8ef51d3abfe05f78 = function _templateObject11_5b205ae0425211ea8ef51d3abfe05f78() {
      return data;
    };

    return data;
  }

  function _templateObject10_5b205ae0425211ea8ef51d3abfe05f78() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                        <tr>\n                          ", "\n                        </tr>\n                      "]);

    _templateObject10_5b205ae0425211ea8ef51d3abfe05f78 = function _templateObject10_5b205ae0425211ea8ef51d3abfe05f78() {
      return data;
    };

    return data;
  }

  function _templateObject9_5b205ae0425211ea8ef51d3abfe05f78() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                            <td>", "</td>\n                          "]);

    _templateObject9_5b205ae0425211ea8ef51d3abfe05f78 = function _templateObject9_5b205ae0425211ea8ef51d3abfe05f78() {
      return data;
    };

    return data;
  }

  function _templateObject8_5b205ae0425211ea8ef51d3abfe05f78() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                      <tr>\n                        ", "\n                      </tr>\n                    "]);

    _templateObject8_5b205ae0425211ea8ef51d3abfe05f78 = function _templateObject8_5b205ae0425211ea8ef51d3abfe05f78() {
      return data;
    };

    return data;
  }

  function _templateObject7_5b205ae0425211ea8ef51d3abfe05f78() {
    var data = babelHelpers.taggedTemplateLiteral([""]);

    _templateObject7_5b205ae0425211ea8ef51d3abfe05f78 = function _templateObject7_5b205ae0425211ea8ef51d3abfe05f78() {
      return data;
    };

    return data;
  }

  function _templateObject6_5b205ae0425211ea8ef51d3abfe05f78() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                            <td>", "</td>\n                          "]);

    _templateObject6_5b205ae0425211ea8ef51d3abfe05f78 = function _templateObject6_5b205ae0425211ea8ef51d3abfe05f78() {
      return data;
    };

    return data;
  }

  function _templateObject5_5b205ae0425211ea8ef51d3abfe05f78() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                    <thead>\n                      <tr>\n                        ", "\n                      </tr>\n                    </thead>\n                  "]);

    _templateObject5_5b205ae0425211ea8ef51d3abfe05f78 = function _templateObject5_5b205ae0425211ea8ef51d3abfe05f78() {
      return data;
    };

    return data;
  }

  function _templateObject4_5b205ae0425211ea8ef51d3abfe05f78() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                      A ", " chart.\n                    "]);

    _templateObject4_5b205ae0425211ea8ef51d3abfe05f78 = function _templateObject4_5b205ae0425211ea8ef51d3abfe05f78() {
      return data;
    };

    return data;
  }

  function _templateObject3_5b205ae0425211ea8ef51d3abfe05f78() {
    var data = babelHelpers.taggedTemplateLiteral(["\n                      ", "", "\n                    "]);

    _templateObject3_5b205ae0425211ea8ef51d3abfe05f78 = function _templateObject3_5b205ae0425211ea8ef51d3abfe05f78() {
      return data;
    };

    return data;
  }

  function _templateObject2_5b205ae0425211ea8ef51d3abfe05f78() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <table id=\"", "-desc\" class=\"a11y\">\n              <caption>\n                ", "\n              </caption>\n              ", "\n              <tbody>\n                ", "\n              </tbody>\n            </table>\n          "]);

    _templateObject2_5b205ae0425211ea8ef51d3abfe05f78 = function _templateObject2_5b205ae0425211ea8ef51d3abfe05f78() {
      return data;
    };

    return data;
  }

  function _templateObject_5b205ae0425211ea8ef51d3abfe05f78() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div\n        id=\"chart\"\n        chart=\"", "\"\n        class=\"ct-chart ", "\"\n      ></div>\n      ", "\n    "]);

    _templateObject_5b205ae0425211ea8ef51d3abfe05f78 = function _templateObject_5b205ae0425211ea8ef51d3abfe05f78() {
      return data;
    };

    return data;
  }

  /**
   * `chartist-render`
   * uses chartist library to render a chart
   *
  ### Styling
  
  `<chartist-render>` provides the following custom properties
  for styling:
  
  Custom property | Description | Default
  ----------------|-------------|----------
  `--chartist-label-color` | default label color for charts | #000
  `--chartist-pie-label-color` | label color for pie charts | `--chartist-label-color`
  `--chartist-color-a` | background color for 1st series |  #d70206
  `--chartist-color-label-a` | color for 1st series label |  `--chartist-label-color`
  `--chartist-color-b` | background color for 2nd series |  #f05b4f
  `--chartist-color-label-b` | color for 2nd series label |  `--chartist-label-color`
  `--chartist-color-c` | background color for 3rd series |  #f4c63d
  `--chartist-color-label-c` | color for 3rd series label |  `--chartist-label-color`
  `--chartist-color-d` | background color for 4th series |  #d17905
  `--chartist-color-label-d` | color for 4th series label |  `--chartist-label-color`
  `--chartist-color-e` | background color for 5th series |  #453d3f
  `--chartist-color-label-e` | color for 5th series label |  `--chartist-label-color`
  `--chartist-color-f` | background color for 6th series |  #59922b
  `--chartist-color-label-f` | color for 6th series label |  `--chartist-label-color`
  `--chartist-color-g` | background color for 7th series |  #0544d3
  `--chartist-color-label-g` | color for 7th series label |  `--chartist-label-color`
  `--chartist-color-h` | background color for 8th series |  #6b0392
  `--chartist-color-label-h` | color for 8th series label |  `--chartist-label-color`
  `--chartist-color-i` | background color for 9th series |  #f05b4f
  `--chartist-color-label-i` | color for 9th series label |  `--chartist-label-color`
  `--chartist-color-j` | background color for 10th series |  #dda458
  `--chartist-color-label-j` | color for 10th series label |  `--chartist-label-color`
  `--chartist-color-k` | background color for 11th series |  #eacf7d
  `--chartist-color-label-k` | color for 11th series label |  `--chartist-label-color`
  `--chartist-color-l` | background color for 12th series |  #86797d
  `--chartist-color-label-l` | color for 12th series label |  `--chartist-label-color`
  `--chartist-color-m` | background color for 13th series |  #b2c326
  `--chartist-color-label-m` | color for 13th series label |  `--chartist-label-color`
  `--chartist-color-n` | background color for 14th series |  #6188e2
  `--chartist-color-label-n` | color for 15th series label |  `--chartist-label-color`
  `--chartist-color-0` | background color for 15th series |  #a748ca
  `--chartist-color-label-o` | color for 15th series label |  `--chartist-label-color`
   * @extends SchemaBehaviors
   * @demo ./demo/index.html 
   * @customElement chartist-render
   */
  var ChartistRender =
  /*#__PURE__*/
  function (_SchemaBehaviors) {
    babelHelpers.inherits(ChartistRender, _SchemaBehaviors);
    babelHelpers.createClass(ChartistRender, [{
      key: "render",
      // render function
      value: function render() {
        return (0, _litElement.html)(_templateObject_5b205ae0425211ea8ef51d3abfe05f78(), this.__chartId, this.scale, this.data && this.data.series ? (0, _litElement.html)(_templateObject2_5b205ae0425211ea8ef51d3abfe05f78(), this.__chartId, this.chartTitle || this.chartDesc ? (0, _litElement.html)(_templateObject3_5b205ae0425211ea8ef51d3abfe05f78(), this.chartTitle, this.chartDesc) : (0, _litElement.html)(_templateObject4_5b205ae0425211ea8ef51d3abfe05f78(), this.type), this.data.labels ? (0, _litElement.html)(_templateObject5_5b205ae0425211ea8ef51d3abfe05f78(), this.data.labels.map(function (cell) {
          return (0, _litElement.html)(_templateObject6_5b205ae0425211ea8ef51d3abfe05f78(), cell || "-");
        })) : (0, _litElement.html)(_templateObject7_5b205ae0425211ea8ef51d3abfe05f78()), this.type === "pie" ? (0, _litElement.html)(_templateObject8_5b205ae0425211ea8ef51d3abfe05f78(), this.data.series.map(function (cell) {
          return (0, _litElement.html)(_templateObject9_5b205ae0425211ea8ef51d3abfe05f78(), cell || "-");
        })) : this.data.series.map(function (row) {
          return (0, _litElement.html)(_templateObject10_5b205ae0425211ea8ef51d3abfe05f78(), row.map(function (cell) {
            return (0, _litElement.html)(_templateObject11_5b205ae0425211ea8ef51d3abfe05f78(), cell || "-");
          }));
        })) : (0, _litElement.html)(_templateObject12_5b205ae0425211ea8ef51d3abfe05f78(), this.chartTitle || this.chartDesc ? (0, _litElement.html)(_templateObject13_5b205ae0425211ea8ef51d3abfe05f78(), this.__chartId, this.chartTitle, this.chartDesc) : (0, _litElement.html)(_templateObject14_5b205ae0425211ea8ef51d3abfe05f78(), this.type)));
      } // properties available to the custom element for data binding

    }], [{
      key: "styles",
      //styles function
      get: function get() {
        return [(0, _litElement.css)(_templateObject15_5b205ae0425211ea8ef51d3abfe05f78())];
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(ChartistRender), "properties", this), {
          /**
           * The unique identifier of the chart.
           */
          id: {
            type: String
          },

          /**
           * The type of chart:bar, line, or pie
           */
          type: {
            type: String
          },

          /**
          * The scale of the chart. (See https://gionkunz.github.io/chartist-js/api-documentation.html)```
          Container class	Ratio
          .ct-square          1
          .ct-minor-second	  15:16
          .ct-major-second	  8:9
          .ct-minor-third	    5:6
          .ct-major-third	    4:5
          .ct-perfect-fourth	3:4
          .ct-perfect-fifth	  2:3
          .ct-minor-sixth	    5:8
          .ct-golden-section	1:1.618
          .ct-major-sixth	    3:5
          .ct-minor-seventh	  9:16
          .ct-major-seventh	  8:15
          .ct-octave	        1:2
          .ct-major-tenth	    2:5
          .ct-major-eleventh	3:8
          .ct-major-twelfth	  1:3
          .ct-double-octave	  1:4```
          */
          scale: {
            type: String
          },

          /**
           * The chart title used for accessibility.
           */
          chartTitle: {
            type: String,
            attribute: "chart-title"
          },

          /**
           * The chart description used for accessibility.
           */
          chartDesc: {
            type: String,
            attribute: "chart-desc"
          },

          /**
           * The chart data.
           */
          data: {
            type: Object
          },

          /**
           * The options available at  https://gionkunz.github.io/chartist-js/api-documentation.html.
           */
          options: {
            type: Object
          },

          /**
          * The responsive options.
          From https://gionkunz.github.io/chartist-js/api-documentation.html:
          In addition to the regular options we specify responsive option 
          overrides that will override the default configutation based 
          on the matching media queries.
          `var responsiveOptions = [
          ['screen and (min-width: 641px) and (max-width: 1024px)', {
            showPoint: false,
            axisX: {
              labelInterpolationFnc: function(value) {
                // Will return Mon, Tue, Wed etc. on medium screens
                return value.slice(0, 3);
              }
            }
          }],
          ['screen and (max-width: 640px)', {
            showLine: false,
            axisX: {
              labelInterpolationFnc: function(value) {
                // Will return M, T, W etc. on small screens
                return value[0];
              }
            }
          }]
          ];`
          */
          responsiveOptions: {
            type: Array,
            attribute: "responsive-options"
          },

          /**
           * The show data in table form as well? Default is false.
           */
          showTable: {
            type: Boolean,
            attribute: "show-table"
          }
        });
      }
    }]);

    function ChartistRender() {
      var _this;

      babelHelpers.classCallCheck(this, ChartistRender);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(ChartistRender).call(this));
      _this.id = "chart";
      _this.type = "bar";
      _this.scale = "ct-minor-seventh";
      _this.chartTitle = null;
      _this.chartDesc = null;
      _this.data = null;
      _this.options = null;
      _this.responsiveOptions = [];
      _this.showTable = false;
      _this.__chartId = (0, _utils.generateResourceID)("chart-");

      var basePath = _this.pathFromUrl(decodeURIComponent(meta.url));

      var location = "".concat(basePath, "lib/chartist/dist/chartist.min.js");
      window.addEventListener("es-bridge-chartistLib-loaded", _this._chartistLoaded.bind(babelHelpers.assertThisInitialized(_this)));
      window.ESGlobalBridge.requestAvailability();
      window.ESGlobalBridge.instance.load("chartistLib", location);
      /**
       * Fired once once chart is ready.
       *
       * @event chartist-render-ready
       *
       */

      _this.dispatchEvent(new CustomEvent("chartist-render-ready", {
        bubbles: true,
        cancelable: true,
        composed: true,
        detail: babelHelpers.assertThisInitialized(_this)
      }));

      if ((typeof Chartist === "undefined" ? "undefined" : babelHelpers.typeof(Chartist)) === "object") _this._chartistLoaded.bind(babelHelpers.assertThisInitialized(_this));
      return _this;
    }
    /**
     * Store the tag name to make it easier to obtain directly.
     * @notice function name must be here for tooling to operate correctly
     */


    babelHelpers.createClass(ChartistRender, [{
      key: "updated",
      value: function updated(changedProperties) {
        this._renderChart();
      } // simple path from a url modifier

    }, {
      key: "pathFromUrl",
      value: function pathFromUrl(url) {
        return url.substring(0, url.lastIndexOf("/") + 1);
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        window.removeEventListener("es-bridge-chartistLib-loaded", this._chartistLoaded.bind(this));
        babelHelpers.get(babelHelpers.getPrototypeOf(ChartistRender.prototype), "disconnectedCallback", this).call(this);
      }
      /**
       * determines if char is ready
       */

    }, {
      key: "_chartistLoaded",
      value: function _chartistLoaded() {
        this.__chartistLoaded = true;

        this._renderChart();
      }
      /**
       * Makes chart and returns the chart object.
       */

    }, {
      key: "makeChart",
      value: function makeChart() {
        var _this2 = this;

        setTimeout(function () {
          _this2.chart = _this2._renderChart();
        }, 100);
      }
      /**
       * Renders chart and returns the chart object.
       */

    }, {
      key: "_renderChart",
      value: function _renderChart() {
        var _this3 = this;

        var chart = null,
            target = this.shadowRoot.querySelector("#chart");

        if (target !== null && (typeof Chartist === "undefined" ? "undefined" : babelHelpers.typeof(Chartist)) === "object" && this.data !== null) {
          if (this.type == "bar") {
            if (this.responsiveOptions !== undefined && this.responsiveOptions.length > 0) {
              this.responsiveOptions.forEach(function (option) {
                if (option[1] !== undefined) {
                  if (option[1].axisX && option[1].axisX.labelInterpolationFnc == "noop") option[1].axisX.labelInterpolationFnc = Chartist.noop;
                  if (option[1].axisY && option[1].axisY.labelInterpolationFnc == "noop") option[1].axisY.labelInterpolationFnc = Chartist.noop;
                }
              });
            }

            chart = Chartist.Bar(target, this.data, this.options, this.responsiveOptions);
          } else if (this.type === "line") {
            chart = Chartist.Line(target, this.data, this.options, this.responsiveOptions);
          } else if (this.type === "pie") {
            chart = Chartist.Pie(target, this.data, this.options, this.responsiveOptions);
          }
          /**
           * Fired when chart is being drawn.
           *
           * @event chartist-render-draw
           *
           */


          this.dispatchEvent(new CustomEvent("chartist-render-draw", {
            bubbles: true,
            cancelable: true,
            composed: true,
            detail: chart
          }));
          chart.on("created", function () {
            _this3.addA11yFeatures(chart.container.children[0]);
            /**
             * Fired once chart is created and accessibility features are added.
             *
             * @event chartist-render-created
             *
             */


            _this3.dispatchEvent(new CustomEvent("chartist-render-created", {
              bubbles: true,
              cancelable: true,
              composed: true,
              detail: chart
            }));
          });
        }

        return chart;
      }
      /**
       * Add accessibility features.
       * @param {object} svg chart SVG
       */

    }, {
      key: "addA11yFeatures",
      value: function addA11yFeatures(svg) {
        if (this.data && this.data.series) {
          svg.setAttribute("aria-labelledby", "".concat(this.__chartId, "-desc"));
        } else {
          svg.setAttribute("aria-labelledby", "".concat(this.__chartId, "-title"));
        }
      }
      /**
       * Get unique ID from the chart
       * @param {string} prefix for unique ID
       * @returns {string} unique ID
       */

    }, {
      key: "_getUniqueId",
      value: function _getUniqueId(prefix) {
        var id = prefix + Date.now();
        return id;
      }
    }], [{
      key: "tag",
      get: function get() {
        return "chartist-render";
      }
    }]);
    return ChartistRender;
  }((0, _schemaBehaviors.SchemaBehaviors)(_litElement.LitElement));

  _exports.ChartistRender = ChartistRender;
  window.customElements.define(ChartistRender.tag, ChartistRender);
});
