// @lrnwebcomponents/page-contents-menu/page-contents-menu.js
define(["exports", "../../lit-element/lit-element.js"], function (_exports, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.PageContentsMenu = void 0;

  function _templateObject2_0266f730b4c911ea90ffad65c66d8e0c() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n\n        :host([hidden]) {\n          display: none;\n        }\n      "]);

    _templateObject2_0266f730b4c911ea90ffad65c66d8e0c = function _templateObject2_0266f730b4c911ea90ffad65c66d8e0c() {
      return data;
    };

    return data;
  }

  function _templateObject_0266f730b4c911ea90ffad65c66d8e0c() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"reset-3c756112--pageSideSection-542f1fd5\">\n        <div\n          class=\"reset-3c756112--menu-5b8a7448--pageToc-16c35922--pageTocLive-542f1fd5\"\n        >\n          <div\n            class=\"reset-3c756112--withScrollbar-39338630--scrollAxisY-7680295e--menuItems-2e6f017b--pageTocLiveItems-67c61496\"\n          >\n            <div\n              class=\"reset-3c756112--menuHeader-342906de--menuHeaderLight-2c5d8183\"\n            >\n              <div class=\"reset-3c756112--menuHeaderIcon-0c3ed569\">\n                <svg\n                  preserveAspectRatio=\"xMidYMid meet\"\n                  height=\"1em\"\n                  width=\"1em\"\n                  fill=\"none\"\n                  xmlns=\"http://www.w3.org/2000/svg\"\n                  viewBox=\"0 0 24 24\"\n                  stroke-width=\"2\"\n                  stroke-linecap=\"round\"\n                  stroke-linejoin=\"round\"\n                  stroke=\"currentColor\"\n                  class=\"icon-7f6730be--text-3f89f380\"\n                >\n                  <g>\n                    <line x1=\"21\" y1=\"10\" x2=\"7\" y2=\"10\"></line>\n                    <line x1=\"21\" y1=\"6\" x2=\"3\" y2=\"6\"></line>\n                    <line x1=\"21\" y1=\"14\" x2=\"3\" y2=\"14\"></line>\n                    <line x1=\"21\" y1=\"18\" x2=\"7\" y2=\"18\"></line>\n                  </g>\n                </svg>\n              </div>\n              <div class=\"reset-3c756112\">\n                <span\n                  class=\"text-4505230f--InfoH100-1e92e1d1--textContentFamily-49a318e1\"\n                  >Contents</span\n                >\n              </div>\n            </div>\n            <a\n              class=\"reset-3c756112--menuItem-aa02f6ec--menuItemLight-757d5235--menuItemInline-173bdf97--pageTocItem-f4427024\"\n              href=\"#i-have-a-github-sync-error\"\n              style=\"\"\n              ><div class=\"reset-3c756112--menuItemContent-c44ec79e\">\n                <span\n                  class=\"text-4505230f--UIH300-2063425d--textContentFamily-49a318e1\"\n                  ><span\n                    class=\"text-4505230f--UIH200-50ead35f--textContentFamily-49a318e1\"\n                  >\n                    I have a GitHub sync error</span\n                  ></span\n                >\n              </div></a\n            ><a\n              class=\"reset-3c756112--menuItem-aa02f6ec--menuItemLight-757d5235--menuItemInline-173bdf97--pageTocItem-f4427024\"\n              href=\"#gitbook-is-not-using-my-docs-folder\"\n              style=\"\"\n              ><div class=\"reset-3c756112--menuItemContent-c44ec79e\">\n                <span\n                  class=\"text-4505230f--UIH300-2063425d--textContentFamily-49a318e1\"\n                  ><span\n                    class=\"text-4505230f--UIH200-50ead35f--textContentFamily-49a318e1\"\n                    >&#8203; GitBook is not using my docs folder?</span\n                  ></span\n                >\n              </div></a\n            ><a\n              class=\"reset-3c756112--menuItem-aa02f6ec--menuItemLight-757d5235--menuItemInline-173bdf97--pageTocItem-f4427024\"\n              href=\"#gitbook-is-creating-new-markdown-files\"\n              style=\"\"\n              ><div class=\"reset-3c756112--menuItemContent-c44ec79e\">\n                <span\n                  class=\"text-4505230f--UIH300-2063425d--textContentFamily-49a318e1\"\n                  ><span\n                    class=\"text-4505230f--UIH200-50ead35f--textContentFamily-49a318e1\"\n                  >\n                    GitBook is creating new markdown files?</span\n                  ></span\n                >\n              </div></a\n            ><a\n              class=\"reset-3c756112--menuItem-aa02f6ec--menuItemLight-757d5235--menuItemInline-173bdf97--pageTocItem-f4427024\"\n              href=\"#my-repository-is-not-listed\"\n              style=\"\"\n              ><div class=\"reset-3c756112--menuItemContent-c44ec79e\">\n                <span\n                  class=\"text-4505230f--UIH300-2063425d--textContentFamily-49a318e1\"\n                  ><span\n                    class=\"text-4505230f--UIH200-50ead35f--textContentFamily-49a318e1\"\n                    >&#8203; My repository is not listed?</span\n                  ></span\n                >\n              </div></a\n            ><a\n              class=\"reset-3c756112--menuItem-aa02f6ec--menuItemLight-757d5235--menuItemInline-173bdf97--pageTocItem-f4427024\"\n              href=\"#nothing-happens-on-gitbook-after-adding-a-new-file-to-my-repository\"\n              style=\"\"\n              ><div class=\"reset-3c756112--menuItemContent-c44ec79e\">\n                <span\n                  class=\"text-4505230f--UIH300-2063425d--textContentFamily-49a318e1\"\n                  ><span\n                    class=\"text-4505230f--UIH200-50ead35f--textContentFamily-49a318e1\"\n                    >&#8203; Nothing happens on GitBook after adding a new file\n                    to my repository</span\n                  ></span\n                >\n              </div></a\n            >\n          </div>\n        </div>\n      </div>\n    "]);

    _templateObject_0266f730b4c911ea90ffad65c66d8e0c = function _templateObject_0266f730b4c911ea90ffad65c66d8e0c() {
      return data;
    };

    return data;
  }

  /**
   * `page-contents-menu`
   * `Links that jump you to the right place in the page&#39;s content`
   * @demo demo/index.html
   * @element page-contents-menu
   */
  var PageContentsMenu =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(PageContentsMenu, _LitElement);
    babelHelpers.createClass(PageContentsMenu, [{
      key: "render",
      // render function
      value: function render() {
        return (0, _litElement.html)(_templateObject_0266f730b4c911ea90ffad65c66d8e0c());
      } // haxProperty definition

    }], [{
      key: "styles",
      //styles function
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_0266f730b4c911ea90ffad65c66d8e0c())];
      }
    }, {
      key: "haxProperties",
      get: function get() {
        return {
          canScale: true,
          canPosition: true,
          canEditSource: false,
          gizmo: {
            title: "Page contents-menu",
            description: "Links that jump you to the right place in the page's content",
            icon: "icons:android",
            color: "green",
            groups: ["Contents"],
            handles: [{
              type: "todo:read-the-docs-for-usage"
            }],
            meta: {
              author: "btopro",
              owner: "The Pennsylvania State University"
            }
          },
          settings: {
            quick: [],
            configure: [],
            advanced: []
          }
        };
      } // properties available to the custom element for data binding

    }, {
      key: "properties",
      get: function get() {
        return {
          contentContainer: {
            type: Object
          },
          relationship: {
            type: String
          },
          _items: {
            type: Array
          }
        };
      }
      /**
       * Convention we use
       */

    }, {
      key: "tag",
      get: function get() {
        return "page-contents-menu";
      }
      /**
       * HTMLElement
       */

    }]);

    function PageContentsMenu() {
      var _this;

      babelHelpers.classCallCheck(this, PageContentsMenu);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(PageContentsMenu).call(this)); // default is to use the parent container unless a content container
      // is defined otherwise. This would imply usage of placing this at the TOP of
      // content area though next, pervious and none are valid

      _this.relationship = "parent";
      _this._items = [];
      _this.contentContainer = null;
      return _this;
    }

    babelHelpers.createClass(PageContentsMenu, [{
      key: "firstUpdated",
      value: function firstUpdated() {
        // if we are told to use the parent and we're connected...
        if (this.relationship == "parent") {
          // which will kick that off
          this.contentContainer = this.parentElement;
        } else if (this.relationship == "next") {
          // which will kick that off
          this.contentContainer = this.nextElementSibling;
        } else if (this.relationship == "previous") {
          // which will kick that off
          this.contentContainer = this.previousElementSibling;
        } // anything else we ignore sot hat contentContainer can be set manually
        else {}
      }
      /**
       * LitElement life cycle - property changed
       */

    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          /* notify example
          // notify
          if (propName == 'format') {
            this.dispatchEvent(
              new CustomEvent(`${propName}-changed`, {
                detail: {
                  value: this[propName],
                }
              })
            );
          }
          */
          if (propName == "contentContainer") {
            _this2._contentContainerChanged(_this2[propName]);
          }
          /* computed example
          if (['id', 'selected'].includes(propName)) {
            this.__selectedChanged(this.selected, this.id);
          }
          */

        });
      }
      /**
       * When our content container changes, process the hierarchy in question
       */

    }, {
      key: "_contentContainerChanged",
      value: function _contentContainerChanged(newValue) {
        if (newValue && newValue.innerHTML) {
          // @todo use HAX element scrappign thing to take a blob of content
          // and convert it into a hierarchy, much like the map does
          this._items = [];
        }
      }
    }]);
    return PageContentsMenu;
  }(_litElement.LitElement);

  _exports.PageContentsMenu = PageContentsMenu;
  customElements.define(PageContentsMenu.tag, PageContentsMenu);
});
