// @lrnwebcomponents/future-terminal-text/lib/future-terminal-text-lite.js
define(["exports", "../../../lit/index.js", "./FutureTerminalTextSuper.js"], function (_exports, _index, _FutureTerminalTextSuper) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.FutureTerminalTextLite = void 0;

  var _templateObject_61bb24a0942811ecb1937969798b231a, _templateObject2_61bb24a0942811ecb1937969798b231a, _templateObject3_61bb24a0942811ecb1937969798b231a;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `future-terminal-text`
   * `this is to provide a futuristic terminal text type of environment`
   * @demo demo/index.html
   * @element future-terminal-text
   */
  var FutureTerminalTextLite = /*#__PURE__*/function (_FutureTerminalTextLi) {
    babelHelpers.inherits(FutureTerminalTextLite, _FutureTerminalTextLi);

    var _super = _createSuper(FutureTerminalTextLite);

    function FutureTerminalTextLite() {
      babelHelpers.classCallCheck(this, FutureTerminalTextLite);
      return _super.apply(this, arguments);
    }

    babelHelpers.createClass(FutureTerminalTextLite, [{
      key: "updated",
      value:
      /**
       * LitElement life cycle - property changed
       */
      function updated(changedProperties) {
        var _this = this;

        if (babelHelpers.get(babelHelpers.getPrototypeOf(FutureTerminalTextLite.prototype), "updated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(FutureTerminalTextLite.prototype), "updated", this).call(this, changedProperties);
        }

        changedProperties.forEach(function (oldValue, propName) {
          if (propName === "glitch" && _this[propName]) {
            _this._doGlitch();
          }
        });
      } // Template return function

    }, {
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_61bb24a0942811ecb1937969798b231a || (_templateObject_61bb24a0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<span><slot></slot></span>"])));
      } // properties available to the custom element for data binding

    }], [{
      key: "styles",
      get:
      /**
       * LitElement shadow styling convention
       */
      function get() {
        var styles = (0, _index.css)(_templateObject2_61bb24a0942811ecb1937969798b231a || (_templateObject2_61bb24a0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral([""])));

        if (babelHelpers.get(babelHelpers.getPrototypeOf(FutureTerminalTextLite), "styles", this)) {
          styles = babelHelpers.get(babelHelpers.getPrototypeOf(FutureTerminalTextLite), "styles", this);
        }

        return [styles, (0, _index.css)(_templateObject3_61bb24a0942811ecb1937969798b231a || (_templateObject3_61bb24a0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          font-weight: bold;\n          display: inline-flex;\n          --flicker-easing: cubic-bezier(0.32, 0.32, 0, 0.92);\n          --flicker-duration: 300ms;\n          --fade-in-duration: 500ms;\n        }\n        span {\n          color: #5fa4a5;\n          text-shadow: 0 0 4px #5fa4a5;\n          animation: flicker var(--flicker-duration) var(--flicker-easing);\n        }\n        :host([red]) span {\n          color: #b35b5a;\n          text-shadow: 0 0 4px #b35b5a;\n        }\n        :host([fadein]) span {\n          animation: fade-in var(--fade-in-duration),\n            flicker 300ms var(--flicker-easing)\n              calc(var(--fade-in-duration) * 0.8);\n          transform: translate(0, 0);\n          opacity: 1;\n        }\n        @keyframes flicker {\n          0% {\n            opacity: 0.75;\n          }\n          50% {\n            opacity: 0.45;\n          }\n          100% {\n            opacity: 1;\n          }\n        }\n        @keyframes fade-in {\n          from {\n            transform: translate(-30px, 0px);\n            opacity: 0;\n          }\n        }\n      "])))];
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(FutureTerminalTextLite), "properties", this)), {}, {
          glitch: {
            type: Boolean
          },
          red: {
            type: Boolean,
            reflect: true
          },
          fadein: {
            type: Boolean,
            reflect: true
          },
          glitchMax: {
            type: Number,
            attribute: "glitch-max"
          },
          glitchDuration: {
            type: Number,
            attribute: "glitch-duration"
          }
        });
      }
      /**
       * Convention we use
       */

    }, {
      key: "tag",
      get: function get() {
        return "future-terminal-text-lite";
      }
    }]);
    return FutureTerminalTextLite;
  }((0, _FutureTerminalTextSuper.FutureTerminalTextLiteSuper)(_index.LitElement));

  _exports.FutureTerminalTextLite = FutureTerminalTextLite;
  customElements.define(FutureTerminalTextLite.tag, FutureTerminalTextLite);
});
