diff --git a/question_types/multichoice/multichoice.classes.inc b/question_types/multichoice/multichoice.classes.inc
index b1b246c..2262b48 100644
--- a/question_types/multichoice/multichoice.classes.inc
+++ b/question_types/multichoice/multichoice.classes.inc
@@ -579,6 +579,7 @@ class MultichoiceQuestion extends QuizQuestion {
       }
       $form['alternatives'][$i]['correct'] = array(
         '#type' => 'checkbox',
+        '#title' => t('Correct'),
         '#default_value' => $correct_default,
         '#attributes' => array('onchange' => 'Multichoice.refreshScores(this, ' . variable_get('multichoice_def_scoring', 0) . ')'),
       );
