// @lrnwebcomponents/lrn-aside/lrn-aside.js
define(["exports","require","../../lit-element/lit-element.js"],(function(t,e,i){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.LrnAside=void 0,e=babelHelpers.interopRequireWildcard(e);class LrnAside extends i.LitElement{constructor(){super(),this.title="Related content",this.sticky=!1,this.direction="",setTimeout(()=>{new Promise((t,i)=>e.default(["../lrndesign-panelcard/lrndesign-panelcard.js"],t,i))},0)}static get styles(){return[i.css`
        :host {
          display: flex;
          padding: 8px;
        }
        :host([sticky]) {
          top: 0;
          position: sticky;
        }
        :host([direction="left"]) {
          float: left;
          max-width: 480px;
        }
        :host([direction="right"]) {
          float: right;
          max-width: 480px;
        }
      `]}render(){return i.html`
      <aside>
        <lrndesign-panelcard title="${this.title}">
          <slot></slot>
        </lrndesign-panelcard>
      </aside>
    `}static get tag(){return"lrn-aside"}static get properties(){return{title:{type:String},sticky:{type:Boolean,reflect:!0},direction:{type:String,reflect:!0}}}static get haxProperties(){return{canScale:!0,canPosition:!0,canEditSource:!1,gizmo:{title:"Sticky note",description:"A sticky note to present some basic info offset on the page.",icon:"av:note",color:"yellow",groups:["Content"],handles:[{type:"text",title:"title"}],meta:{author:"ELMS:LN"}},settings:{quick:[{property:"title",title:"Title",description:"Enter title for sticky note",inputMethod:"textfield",required:!0}],configure:[{property:"title",title:"Title",description:"Enter title for sticky note.",inputMethod:"textfield",required:!0},{slot:"",title:"Content",description:"Content of the sticky note",inputMethod:"code-editor",required:!0},{property:"sticky",title:"Stick to page on scroll",description:"Appear sticky when the user scrolls past it",inputMethod:"boolean"},{property:"direction",title:"Direction to hang",description:"Location of the sticky note to hang",inputMethod:"select",options:{"":"none",right:"Right",left:"Left"}}],advanced:[]}}}}t.LrnAside=LrnAside,window.customElements.define(LrnAside.tag,LrnAside)}));
