// @lrnwebcomponents/hax-body/lib/hax-manager.js
define(["exports","require","../../../@polymer/polymer/polymer-element.js","../../../@polymer/polymer/lib/utils/render-status.js","../../../@polymer/polymer/lib/legacy/polymer.dom.js","../../../@polymer/paper-styles/paper-styles.js","../../simple-colors/simple-colors.js","./hax-shared-styles.js","../../../@polymer/iron-pages/iron-pages.js"],function(_exports,_require,_polymerElement,_renderStatus,_polymerDom,_paperStyles,_simpleColors,_haxSharedStyles,_ironPages){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.HaxManager=void 0,_require=babelHelpers.interopRequireWildcard(_require);class HaxManager extends _polymerElement.PolymerElement{constructor(){super(),new Promise((res,rej)=>_require.default(["../../../@polymer/iron-icons/iron-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../../@polymer/iron-icon/iron-icon.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../../@polymer/paper-input/paper-input.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../../@polymer/paper-button/paper-button.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../../@polymer/paper-icon-button/paper-icon-button.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../../@vaadin/vaadin-upload/vaadin-upload.js"],res,rej)),new Promise((res,rej)=>_require.default(["./hax-preview.js"],res,rej)),new Promise((res,rej)=>_require.default(["./hax-app-browser.js"],res,rej)),new Promise((res,rej)=>_require.default(["./hax-gizmo-browser.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../../@polymer/app-layout/app-drawer/app-drawer.js"],res,rej))}static get template(){return _polymerElement.html`
      <style include="hax-shared-styles">
        @import url("https://fonts.googleapis.com/css?family=Noto+Serif");
        :host {
          display: block;
          color: var(--hax-color-text);
        }
        app-drawer:not(:defined),
        hax-preview:not(:defined),
        hax-app-browser:not(:defined),
        hax-gizmo-browser:not(:defined),
        iron-icon:not(:defined) {
          display: none;
        }
        #dialog {
          color: var(--hax-color-text);
          z-index: 10000;
          padding: 56px 0;
          margin-top: 56px;
          --app-drawer-width: 400px;
          --app-drawer-content-container: {
            background-color: #ffffff;
          }
        }
        #closedialog {
          float: right;
          top: 124px;
          right: 0;
          position: absolute;
          padding: 8px;
          margin: 0;
          color: var(--hax-color-text);
          background-color: transparent;
          width: 40px;
          height: 40px;
          min-width: unset;
        }
        :host([active-page="0"]) #dialog {
          --app-drawer-width: 300px;
        }
        :host([active-page="1"]) #dialog {
          --app-drawer-width: 60%;
        }
        :host([active-page="2"]) #dialog {
          --app-drawer-width: 60%;
        }
        :host([active-step]) #dialog {
          --app-drawer-width: 60%;
        }
        :host([searching]) #dialog {
          --app-drawer-width: 60%;
        }
        @media screen and (max-width: 800px) {
          :host([active-page="0"]) #dialog {
            --app-drawer-width: 50%;
          }
          :host([active-page="1"]) #dialog,
          :host([active-page="2"]) #dialog,
          :host([active-step]) #dialog,
          :host([searching]) #dialog {
            --app-drawer-width: 80%;
          }
        }
        .title {
          position: relative;
          padding: 16px;
          outline: 0;
          font-weight: 600;
          text-align: left;
          margin: 0;
          background-color: var(--hax-color-menu-heading-bg);
          font-size: 18px;
          line-height: 18px;
          font-family: "Noto Serif", serif;
          color: var(--hax-color-text);
        }

        #preview {
          height: 100%;
        }
        vaadin-upload {
          --primary-color: var(--hax-color-accent1);
          --primary-font-color: #ffffff;
          --dark-primary-color: #ffffff;
          --light-primary-color: var(--hax-color-accent1);
          --error-color: darkred;
          color: #ffffff;
          display: block;
          padding: 32px !important;
          --vaadin-upload-button-add-wrapper: {
            border: 2px solid #ffffff;
            background-color: var(--hax-color-accent1);
            color: #ffffff;
            display: block;
          }
          --vaadin-upload-buttons-primary: {
            display: block;
            width: 100%;
            flex: unset;
            -webkit-flex: unset;
          }
          --vaadin-upload-button-add: {
            color: #000000;
            display: block;
            flex: unset;
            -webkit-flex: unset;
            text-align: center;
          }
          --vaadin-upload-drop-label: {
            color: #ffffff;
            display: block;
            padding: 16px;
          }
          --vaadin-upload-drop-label-dragover: {
            color: #ffffff;
          }
          --vaadin-upload-file-list: {
            padding: 16px;
            margin: 0;
            color: #ffffff;
          }
          --vaadin-upload-file: {
            color: #ffffff;
          }
        }
        vaadin-upload[dragover] {
          border-color: #396;
        }
        vaadin-upload-file {
          --disabled-text-color: #222222;
        }
        .add-area-content-wrapper {
          padding: 0 16px;
        }
        .add-url-area,
        .add-upload-area {
          margin: 16px 0;
        }
        .url-description {
          font-size: 18px;
          color: #ffffff;
          line-height: 22px;
          font-family: sans-serif;
          letter-spacing: 1px;
        }
        #steppages {
          height: 100%;
        }
        #newassetconfigure {
          width: 100%;
          margin: 0;
          padding: 16px;
          background-color: var(--hax-color-accent1);
          color: var(--hax-color-accent1-color);
        }
        paper-input {
          color: var(--hax-color-text);
        }
        @media screen and (max-width: 550px) {
          .hide-on-mobile {
            opacity: 0;
            visibility: hidden;
            position: absolute;
            left: -9999px;
          }
          .page-area.hax-manager {
            padding: 6px;
          }
        }
      </style>
      <app-drawer id="dialog" opened="{{opened}}" disable-swipe="">
        <div
          class="dialog-contents"
          id="dialogcontent"
          style="height: 100%; overflow: auto;"
        >
          <iron-pages
            id="steppages"
            selected="{{activeStep}}"
            fallback-selection="select"
            role="main"
          >
            <div data-value="select">
              <iron-pages
                id="activepage"
                selected="{{activePage}}"
                fallback-selection="link"
              >
                <div class="page-area add-area">
                  <h3 class="title">[[addTitle]]</h3>
                  <div class="add-area-content-wrapper">
                    <div class="add-url-area">
                      <paper-input
                        id="url"
                        label="URL"
                        type="url"
                        auto-validate=""
                      ></paper-input>
                      <div class="url-description">
                        A full URL with https:// referencing a link that already
                        exists on the web.
                      </div>
                    </div>
                    <div class="add-upload-area">
                      <vaadin-upload
                        form-data-name="file-upload"
                        id="fileupload"
                        hidden$="[[!canSupportUploads]]"
                      ></vaadin-upload>
                    </div>
                    <paper-button id="newassetconfigure" raised=""
                      >Configure item</paper-button
                    >
                  </div>
                </div>
                <div class="page-area">
                  <hax-app-browser id="appbrowser">
                    <slot></slot>
                  </hax-app-browser>
                </div>
                <div class="page-area">
                  <hax-gizmo-browser id="gizmobrowser"></hax-gizmo-browser>
                </div>
              </iron-pages>
            </div>
            <div style="height:100%;">
              <hax-preview
                id="preview"
                element="{{activeHaxElement}}"
              ></hax-preview>
            </div>
          </iron-pages>
          <paper-button
            id="closedialog"
            on-click="cancel"
            hidden$="[[activeStep]]"
          >
            <iron-icon icon="icons:cancel" title="Close dialog"></iron-icon>
          </paper-button>
        </div>
      </app-drawer>
    `}static get tag(){return"hax-manager"}static get properties(){return{opened:{type:Boolean,value:!1,reflectToAttribute:!0,observer:"_openedChanged"},editExistingNode:{type:Boolean,value:!1},editTitle:{type:String,computed:"_computeEditTitle(editExistingNode)"},addTitle:{type:String,value:"Add content"},activeStep:{type:Boolean,reflectToAttribute:!0,value:!1,observer:"_activeStepChanged"},searching:{type:Boolean,reflectToAttribute:!0,value:!1},activePage:{type:String,reflectToAttribute:!0,value:0,observer:"_activePageChanged"},canSupportUploads:{type:Boolean,value:!1},activeHaxElement:{type:Object,observer:"_activeHaxElementChanged"},appList:{type:Array,value:[]},__allowUpload:{type:Boolean,value:!1}}}ready(){super.ready(),this.dispatchEvent(new CustomEvent("hax-register-manager",{bubbles:!0,cancelable:!0,composed:!0,detail:this})),(0,_renderStatus.afterNextRender)(this,function(){document.body.addEventListener("hax-store-property-updated",this._haxStorePropertyUpdated.bind(this)),document.body.addEventListener("hax-app-picker-selection",this._haxAppPickerSelection.bind(this)),document.body.addEventListener("place-holder-file-drop",this._placeHolderFileDrop.bind(this)),this.shadowRoot.querySelector("#dialog").addEventListener("iron-overlay-canceled",this.close.bind(this)),this.shadowRoot.querySelector("#dialog").addEventListener("iron-overlay-closed",this.close.bind(this)),this.shadowRoot.querySelector("#closedialog").addEventListener("click",this.close.bind(this)),this.shadowRoot.querySelector("#newassetconfigure").addEventListener("click",this.newAssetConfigure.bind(this)),this.shadowRoot.querySelector("#fileupload").addEventListener("upload-before",this._fileAboutToUpload.bind(this)),this.shadowRoot.querySelector("#fileupload").addEventListener("upload-response",this._fileUploadResponse.bind(this))})}togglePanelSize(e){this.shadowRoot.querySelector("#dialog").classList.toggle("grow"),this.updateStyles(),window.dispatchEvent(new Event("resize"))}_computeEditTitle(updateExisting){return void 0!==this.activeHaxElement&&updateExisting?"Update":"Insert"}_placeHolderFileDrop(e){this.resetManager(),this.open(),window.HaxStore.instance.activePlaceHolder=e.detail.placeHolderElement,this.editExistingNode=!0,this.shadowRoot.querySelector("#fileupload")._onDrop(e.detail)}_fileAboutToUpload(e){if(this.__allowUpload)this.__allowUpload=!1;else{e.preventDefault(),e.stopPropagation();let values={source:e.detail.file.name,type:e.detail.file.type};var type=window.HaxStore.guessGizmoType(values);let targets=window.HaxStore.getHaxAppStoreTargets(type);1===targets.length?this._haxAppPickerSelection({detail:targets[0]}):0!==targets.length?window.HaxStore.instance.haxAppPicker.presentOptions(targets,type,"Where would you like to upload this "+type+"?","app"):window.HaxStore.toast("Sorry, you don't have a storage location that can handle "+type+" uploads!",5e3)}}_haxAppPickerSelection(e){let connection=e.detail.connection;this.__appUsed=e.detail,this.shadowRoot.querySelector("#fileupload").method=connection.operations.add.method;let requestEndPoint=connection.protocol+"://"+connection.url;"/"!=requestEndPoint.substr(requestEndPoint.length-1)&&(requestEndPoint+="/"),void 0!==connection.operations.add.endPoint&&(requestEndPoint+=connection.operations.add.endPoint),null!=window.HaxStore.instance.connectionRewrites.appendUploadEndPoint&&(requestEndPoint+="?"+window.HaxStore.instance.connectionRewrites.appendUploadEndPoint),null!=window.HaxStore.instance.connectionRewrites.appendJwt&&(requestEndPoint+="&"+window.HaxStore.instance.connectionRewrites.appendJwt+"="+localStorage.getItem(window.HaxStore.instance.connectionRewrites.appendJwt)),this.shadowRoot.querySelector("#fileupload").headers=connection.headers,this.shadowRoot.querySelector("#fileupload").target=requestEndPoint,this.__allowUpload=!0,this.shadowRoot.querySelector("#fileupload").uploadFiles()}_fileUploadResponse(e){this.editExistingNode=!0;let response=JSON.parse(e.detail.xhr.response),map=this.__appUsed.connection.operations.add.resultMap,data={},item={};for(var prop in void 0!==this._resolveObjectPath(map.item,response)&&(data=this._resolveObjectPath(map.item,response)),item.type=map.defaultGizmoType,map.gizmo)item[prop]=this._resolveObjectPath(map.gizmo[prop],data);void 0===item.url&&void 0!==item.source&&(item.url=item.source),void 0!==map.gizmo.type&&(item.type=this._resolveObjectPath(map.gizmo.type,data)),this.shadowRoot.querySelector("#url").value=item.url,this.newAssetConfigure()}_activePageChanged(newValue,oldValue){void 0!==newValue&&(this.searching=!1,this.updateStyles(),1===newValue?this.shadowRoot.querySelector("#appbrowser").resetBrowser():2===newValue&&this.shadowRoot.querySelector("#gizmobrowser").resetBrowser())}_haxStorePropertyUpdated(e){e.detail&&void 0!==e.detail.value&&e.detail.property&&this.set(e.detail.property,e.detail.value)}_activeHaxElementChanged(newValue,oldValue){void 0!==oldValue&&(this.shadowRoot.querySelector("#preview").advancedForm=!1,void 0===newValue.tag?this.resetManager():(this.shadowRoot.querySelector("#fileupload").set("files",[]),this.shadowRoot.querySelector("#dialog").scrollTop=0,this.selectStep("configure")))}insertHaxElement(e){let previewNode=this.shadowRoot.querySelector("#preview").previewNode;previewNode.removeAttribute("hax-preview-mode"),null!=previewNode.getAttribute("data-hax-slot")&&(previewNode.setAttribute("slot",previewNode.getAttribute("data-hax-slot")),previewNode.removeAttribute("data-hax-slot"));let element=window.HaxStore.nodeToHaxElement(previewNode);element.replace=this.editExistingNode,void 0!==this.activeHaxElement.__type&&(element.__type=this.activeHaxElement.__type),element.replacement=previewNode,this.dispatchEvent(new CustomEvent("hax-insert-content",{bubbles:!0,cancelable:!0,composed:!0,detail:element}));let toast="New element added!";this.editExistingNode&&(toast="Element updated!"),window.HaxStore.toast(toast,2e3),this.close()}resetManager(activePage=0){this.selectStep("select"),this.activePage=activePage,document.body.style.overflow=null,this.appList=window.HaxStore.instance.appList,this.searching=!1,window.HaxStore.write("activeApp",null,this),window.dispatchEvent(new Event("resize")),this.editExistingNode=!1,this.shadowRoot.querySelector("#url").value="",this.shadowRoot.querySelector("#fileupload").headers="",this.shadowRoot.querySelector("#fileupload").method="",this.shadowRoot.querySelector("#fileupload").target="",this.shadowRoot.querySelector("#preview").shadowRoot.querySelector("#ppanel1").style.flex="",this.shadowRoot.querySelector("#preview").shadowRoot.querySelector("#ppanel2").style.flex="",this.__allowUpload=!1}cancel(e){this.close()}_openedChanged(newValue,oldValue){oldValue&&!newValue?document.body.style.overflow=null:newValue&&!oldValue&&(document.body.style.overflow="hidden")}close(e){var local=(0,_polymerDom.dom)(e).localTarget;void 0!==e&&local!==this.shadowRoot.querySelector("#dialog")&&local!==this.shadowRoot.querySelector("#closedialog")||(window.HaxStore.write("activeHaxElement",{},this),this.opened=!1,this.resetManager())}open(e){this.opened=!0}newAssetConfigure(){let values={source:this.shadowRoot.querySelector("#url").value};var type=window.HaxStore.guessGizmoType(values);let haxElements=window.HaxStore.guessGizmo(type,values);haxElements.length>0?1===haxElements.length?void 0!==haxElements[0].tag&&window.HaxStore.write("activeHaxElement",haxElements[0],this):window.HaxStore.instance.haxAppPicker.presentOptions(haxElements,type,"Pick how to present the "+type,"gizmo"):window.HaxStore.toast("Sorry, HAX doesn't know how to handle that type of link yet.")}toggleDialog(toggle=!0){this.opened&&toggle?this.close():window.HaxStore.instance.closeAllDrawers(this)}selectStep(step){this.activeStep="configure"==step}_activeStepChanged(newValue,oldValue){!newValue&&newValue||(this.updateStyles(),window.dispatchEvent(new Event("resize")))}_resolveObjectPath(path,obj){return path.split(".").reduce(function(prev,curr){return prev?prev[curr]:null},obj||self)}}_exports.HaxManager=HaxManager,window.customElements.define(HaxManager.tag,HaxManager)});
