// @lrnwebcomponents/hax-body/lib/hax-app-browser-item.js
define(["exports","require","../../../lit-element/lit-element.js"],function(_exports,_require,_litElement){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.HAXAppBrowserItem=void 0,_require=babelHelpers.interopRequireWildcard(_require);class HAXAppBrowserItem extends _litElement.LitElement{static get tag(){return"hax-app-browser-item"}static get properties(){return{title:{type:String},index:{type:Number},icon:{type:String},image:{type:String},color:{type:String},author:{type:String},description:{type:String},examples:{type:Array},status:{type:Array}}}constructor(){super(),this.image=!1,new Promise((res,rej)=>_require.default(["../../../@polymer/paper-button/paper-button.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../../@polymer/iron-icon/iron-icon.js"],res,rej))}static get styles(){return[_litElement.css`
        :host {
          display: block;
          max-width: 90px;
        }
        paper-button {
          color: var(--hax-color-text);
          text-transform: none;
          background-color: var(--hax-color-bg-accent);
          min-width: unset;
          cursor: pointer;
          display: flex;
          width: 80px;
          padding: 5px;
          margin: 5px;
          color: #ffffff;
          border-radius: 0;
          transition: box-shadow 0.3s;
        }
        paper-button:hover,
        paper-button:focus {
          box-shadow: 0 4px 4px 0 rgba(0, 0, 0, 0.14),
            0 2px 10px 0 rgba(0, 0, 0, 0.12), 0 6px 2px -4px rgba(0, 0, 0, 0.2);
        }
        paper-button:active {
          box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14),
            0 1px 5px 0 rgba(0, 0, 0, 0.12), 0 3px 1px -2px rgba(0, 0, 0, 0.2);
        }
        paper-button iron-icon {
          height: 32px;
          width: 32px;
          color: var(--simple-colors-default-theme-grey-1);
          display: inline-block;
        }
        .item-title {
          margin-top: 8px;
          color: var(--hax-color-text);
          width: 100%;
          white-space: nowrap;
          overflow: hidden;
          text-overflow: ellipsis;
          font-size: 12px;
          line-height: 12px;
          height: 12px;
          text-align: center;
        }
        .button-inner {
          display: flex;
        }
        .flip-icon {
          transform: rotateY(180deg);
        }
      `]}render(){return _litElement.html`
      <paper-button
        @click="${this._fireEvent}"
        .data-voicecommand="select ${this.title}"
        title="${this.title}"
      >
        <hax-item-button-inner
          color="${this.color}"
          icon="${this.icon}"
          label="${this.title}"
        >
        </hax-item-button-inner>
      </paper-button>
    `}_fireEvent(e){this.dispatchEvent(new CustomEvent("hax-app-selected",{bubbles:!0,cancelable:!0,composed:!0,detail:this.index}))}}_exports.HAXAppBrowserItem=HAXAppBrowserItem,customElements.define(HAXAppBrowserItem.tag,HAXAppBrowserItem)});
