// @lrnwebcomponents/hax-body/lib/hax-toolbar-menu.js
define(["exports","../../../@polymer/polymer/polymer-element.js","../../../@polymer/polymer/lib/utils/render-status.js","../../../@polymer/iron-icon/iron-icon.js","../../../@polymer/iron-icons/iron-icons.js","../../../@polymer/iron-icons/editor-icons.js","../../../@polymer/iron-icons/device-icons.js","../../../@polymer/iron-icons/hardware-icons.js","../../../@polymer/iron-icons/social-icons.js","../../../@polymer/iron-icons/av-icons.js","../../../@polymer/iron-icons/maps-icons.js","../../../@polymer/paper-item/paper-item.js","../../../@polymer/paper-listbox/paper-listbox.js","../../../@polymer/paper-menu-button/paper-menu-button.js","./hax-toolbar-item.js"],function(_exports,_polymerElement,_renderStatus,_ironIcon,_ironIcons,_editorIcons,_deviceIcons,_hardwareIcons,_socialIcons,_avIcons,_mapsIcons,_paperItem,_paperListbox,_paperMenuButton,_haxToolbarItem){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.HaxToolbarMenu=void 0;class HaxToolbarMenu extends _polymerElement.PolymerElement{constructor(){super(),this.addEventListener("click",this._menubuttonTap.bind(this))}static get template(){return _polymerElement.html`
      <style>
        :host {
          display: block;
          box-sizing: border-box;
        }
        paper-menu-button {
          color: rgba(0, 0, 0, 0.66);
          margin: 0;
          padding: 0;
          text-transform: none;
          background-color: #ffffff;
          display: flex;
          min-width: 24px;
        }

        paper-menu-button .label {
          font-size: 12px;
          margin-top: 4px;
        }

        paper-menu-button .button-inner {
          padding-top: 8px;
          text-align: center;
        }

        .flip-icon {
          transform: rotateY(180deg);
        }

        paper-tooltip {
          pointer-events: none;
        }
        paper-listbox {
          padding: 0;
        }
      </style>
      <paper-menu-button>
        <hax-toolbar-item
          id="button"
          slot="dropdown-trigger"
          icon="[[icon]]"
          hidden\$="[[!icon]]"
          class\$="[[iconClass]]"
          tooltip="[[tooltip]]"
        ></hax-toolbar-item>
        <paper-listbox
          id="listbox"
          slot="dropdown-content"
          selected="{{selected}}"
        >
          <slot></slot>
        </paper-listbox>
      </paper-menu-button>
    `}static get tag(){return"hax-toolbar-menu"}static get properties(){return{corner:{type:String,reflectToAttribute:!0,value:""},resetOnSelect:{type:Boolean,value:!1},tooltip:{type:String,value:""},tooltipDirection:{type:String,value:""},selected:{type:String,value:"",notify:!0,observer:"_selectChanged"}}}_selectChanged(e){this.shadowRoot.querySelector("#button").focus()}_menubuttonTap(e){this.shadowRoot.querySelector("#listbox").style.display="inherit",this.resetOnSelect&&(this.selected="")}hideMenu(){this.shadowRoot.querySelector("#listbox").style.display="none"}}_exports.HaxToolbarMenu=HaxToolbarMenu,window.customElements.define(HaxToolbarMenu.tag,HaxToolbarMenu)});
