// @lrnwebcomponents/hax-body/lib/hax-panel-item.js
define(["exports","../../../lit-element/lit-element.js","../../../@polymer/paper-button/paper-button.js","../../simple-tooltip/simple-tooltip.js","../../../@polymer/iron-icon/iron-icon.js"],(function(_exports,_litElement,_paperButton,_simpleTooltip,_ironIcon){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.HAXPanelItem=void 0;class HAXPanelItem extends _litElement.LitElement{constructor(){super(),this.disabled=!1,this.edged="",this.icon="editor:text-fields",this.label="editor:text-fields",this.eventName="button",this.value="",setTimeout(()=>{this.addEventListener("click",this._fireEvent)},0)}static get properties(){return{light:{type:Boolean,reflect:!0},voiceCommand:{type:String,attribute:"voice-command"},disabled:{type:Boolean,reflect:!0},edged:{type:String,reflect:!0},icon:{type:String,reflect:!0},label:{type:String,reflect:!0},eventName:{type:String,reflect:!0,attribute:"event-name"},value:{type:String,reflect:!0}}}static get tag(){return"hax-panel-item"}static get styles(){return[_litElement.css`
        :host {
          display: inline-flex;
        }
        paper-button {
          height: 48px;
          line-height: 48px;
          width: 48px;
          overflow: hidden;
          margin: 0;
          text-transform: none;
          background-color: var(--hax-panel-item-bg);
          color: var(--hax-panel-item-text);
          display: flex;
          padding: 8px;
          border-radius: 50%;
          border: 1px solid var(--hax-panel-item-border-color);
          min-width: unset;
          --paper-button-ink-color: var(--hax-panel-item-ink, black);
        }
        paper-button .button-inner {
          text-align: center;
          margin: 0 auto;
        }
        paper-button iron-icon {
          height: 24px;
          width: 24px;
          display: inline-flex;
        }
        paper-button:hover,
        paper-button:focus {
          color: var(--hax-panel-item-active);
          border: 1px solid var(--hax-panel-item-active-border-color);
        }

        paper-button[disabled] {
          opacity: 0.5;
        }
        .flip-icon {
          transform: rotateY(180deg);
        }
        :host([dark]) paper-button {
          border: solid 2px #000000;
          background-color: #000000 !important;
          color: var(--hax-color-bg-accent);
        }
        :host([large]) paper-button {
          height: 40px;
          width: 40px;
        }
        :host([dark]) paper-button:hover iron-icon,
        :host([dark]) paper-button:focus iron-icon {
          color: #ffffff !important;
        }
        :host([dark]) paper-button:hover {
          border: solid #0085ba 1px;
        }
        simple-tooltip {
          font-size: 16px;
          --simple-tooltip-background: #000000;
          --simple-tooltip-opacity: 1;
          --simple-tooltip-text-color: #ffffff;
          --simple-tooltip-delay-in: 0;
          --simple-tooltip-duration-in: 100ms;
          --simple-tooltip-duration-out: 0;
          --simple-tooltip-border-radius: 0;
          --simple-tooltip-font-size: 14px;
        }
      `]}render(){return _litElement.html`
      <paper-button raised id="button" .disabled="${this.disabled}">
        <div class="button-inner">
          <iron-icon icon="${this.icon}"></iron-icon>
        </div>
      </paper-button>
      <simple-tooltip for="button" position="bottom" offset="10">
        ${this.label}
      </simple-tooltip>
    `}updated(changedProperties){changedProperties.forEach((oldValue,propName)=>{"voiceCommand"==propName&&this.dispatchEvent(new CustomEvent("hax-add-voice-command",{bubbles:!0,composed:!0,cancelable:!1,detail:{command:":name: "+this[propName],context:this,callback:"_fireEvent"}}))})}_fireEvent(e){this.disabled||this.dispatchEvent(new CustomEvent("hax-item-selected",{bubbles:!0,cancelable:!1,composed:!0,detail:{target:this,value:this.value,eventName:this.eventName}}))}}_exports.HAXPanelItem=HAXPanelItem,customElements.define(HAXPanelItem.tag,HAXPanelItem)}));
