// @lrnwebcomponents/hax-body/lib/hax-preferences-dialog.js
define(["exports", "../../../lit-element/lit-element.js", "../../hax-body-behaviors/lib/HAXFields.js", "../../simple-icon/lib/simple-icon-lite.js", "../../simple-fields/simple-fields.js", "./hax-store.js", "../../../mobx/dist/mobx.esm.js"], function (_exports, _litElement, _HAXFields, _simpleIconLite, _simpleFields, _haxStore, _mobxEsm) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HaxPreferencesDialog = void 0;

  /**
   * `hax-preferences-dialog`
   * @element hax-preferences-dialog
   * `Export dialog with all export options and settings provided.`
   */
  class HaxPreferencesDialog extends _litElement.LitElement {
    /**
     * LitElement constructable styles enhancement
     */
    static get styles() {
      return [(0, _litElement.css)`
        :host {
          display: block;
        }
        .title {
          position: relative;
          padding: 16px;
          outline: 0;
          font-weight: 600;
          text-align: left;
          margin: 0;
          background-color: var(--hax-color-menu-heading-bg, #eeeeee);
          color: var(--hax-color-menu-heading-color, black);
          font-size: 18px;
          line-height: 18px;
          font-family: "Noto Serif", serif;
        }
        .pref-container {
          text-align: left;
          padding: 16px;
        }
        #link {
          color: #81a3a9;
          font-size: 18px;
          padding: 16px;
          font-style: italic;
        }
      `];
    }

    constructor() {
      super();
      this.hideLink = false;
      this.title = "Advanced settings"; // JSON schema object needs delayed to ensure page repaints the form

      this.schema = [{
        property: "haxVoiceCommands",
        title: "Voice commands",
        description: "Experimental: Voice based control system",
        inputMethod: "boolean",
        value: false
      }];
      this.schemaValues = {
        haxVoiceCommands: false
      };
      (0, _mobxEsm.autorun)(() => {
        this.globalPreferences = (0, _mobxEsm.toJS)(_haxStore.HAXStore.globalPreferences);
        this.schemaValues = (0, _mobxEsm.toJS)(_haxStore.HAXStore.globalPreferences);
      });
    }

    render() {
      return (0, _litElement.html)`
      <h3 class="title">
        <simple-icon-lite icon="hax:settings"></simple-icon-lite>
        ${this.title}
      </h3>
      <div style="height: 100%; overflow: auto;" class="pref-container">
        <simple-fields
          id="settingsform"
          .schematizer="${_HAXFields.HaxSchematizer}"
          .elementizer="${_HAXFields.HaxElementizer}"
        >
        </simple-fields>
      </div>
      ${!this.hideLink ? (0, _litElement.html)`<a
            href="https://haxtheweb.org/"
            rel="noopener"
            id="link"
            target="_blank"
            >Learn more about HAX</a
          >` : ``}
    `;
    }

    static get tag() {
      return "hax-preferences-dialog";
    }

    static get properties() {
      return {
        /**
         * Developer prop to hide the link
         */
        hideLink: {
          type: Boolean
        },

        /**
         * Title.
         */
        title: {
          type: String
        },

        /**
         * Schema that has all of inputs / manages state
         */
        schema: {
          type: Object
        },

        /**
         * Preferences managed for everything global about HAX.
         */
        globalPreferences: {
          type: Object
        }
      };
    }

    firstUpdated(changedProperties) {
      this.shadowRoot.querySelector("#settingsform").fields = [...this.schema];
      this.shadowRoot.querySelector("#settingsform").value = { ...this.schemaValues
      };
      this.shadowRoot.querySelector("#settingsform").addEventListener("value-changed", this.__valueChangedEvent.bind(this));
    }

    __valueChangedEvent(e) {
      if (e.detail.value) {
        _haxStore.HAXStore.globalPreferences = { ...e.detail.value
        };
      }
    }
    /**
     * force an update of settings
     */


    reloadPreferencesForm() {
      this.shadowRoot.querySelector("#settingsform").fields = [...this.schema];
      this.shadowRoot.querySelector("#settingsform").value = { ...this.schemaValues
      };
    }

  }

  _exports.HaxPreferencesDialog = HaxPreferencesDialog;
  window.customElements.define(HaxPreferencesDialog.tag, HaxPreferencesDialog);
});
