// @lrnwebcomponents/hax-body/lib/hax-blox-picker.js
define(["exports","require","../../../lit-element/lit-element.js"],function(_exports,_require,_litElement){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.HaxBloxPicker=void 0,_require=babelHelpers.interopRequireWildcard(_require);class HaxBloxPicker extends _litElement.LitElement{constructor(){super(),this.title="Layouts",new Promise((res,rej)=>_require.default(["../../../@polymer/paper-button/paper-button.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../../@polymer/app-layout/app-drawer/app-drawer.js"],res,rej)),new Promise((res,rej)=>_require.default(["./hax-blox-browser.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../../@polymer/iron-icons/iron-icons.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../../@polymer/iron-icon/iron-icon.js"],res,rej)),this.dispatchEvent(new CustomEvent("hax-register-blox-picker",{bubbles:!0,cancelable:!0,composed:!0,detail:this}))}static get styles(){return[_litElement.css`
        :host {
          display: block;
        }
        iron-icon:not(:defined),
        paper-button:not(:defined),
        app-drawer:not(:defined) {
          display: none;
        }
        #dialog {
          --app-drawer-width: 320px;
          z-index: 1000;
          margin-top: 56px;
        }
        #closedialog {
          float: right;
          top: 124px;
          right: 0;
          position: absolute;
          padding: 8px;
          margin: 0;
          color: var(--hax-color-text);
          background-color: transparent;
          width: 40px;
          height: 40px;
          min-width: unset;
        }
        .title {
          position: relative;
          padding: 16px;
          outline: 0;
          font-weight: 600;
          text-align: left;
          margin: 0;
          background-color: var(--hax-color-menu-heading-bg);
          font-size: 18px;
          line-height: 18px;
          font-family: "Noto Serif", serif;
          color: var(--hax-color-text);
        }
        app-drawer {
          --app-drawer-width: 320px;
        }
        .pref-container {
          text-align: left;
          padding: 16px;
        }
      `]}render(){return _litElement.html`
      <app-drawer id="dialog" align="left" transition-duration="300">
        <h3 class="title">${this.title}</h3>
        <div style="height: 100%; overflow: auto;" class="pref-container">
          <hax-blox-browser id="bloxbrowser"></hax-blox-browser>
        </div>
        <paper-button id="closedialog" @click="${this.close}">
          <iron-icon icon="icons:cancel" title="Close dialog"></iron-icon>
        </paper-button>
      </app-drawer>
    `}static get tag(){return"hax-blox-picker"}static get properties(){return{title:{type:String}}}open(){this.shadowRoot.querySelector("#dialog").open()}close(){this.shadowRoot.querySelector("#dialog").close()}toggleDialog(){this.shadowRoot.querySelector("#dialog").opened?this.close():window.HaxStore.instance.closeAllDrawers(this)}}_exports.HaxBloxPicker=HaxBloxPicker,window.customElements.define(HaxBloxPicker.tag,HaxBloxPicker)});
