// @lrnwebcomponents/hax-body/lib/hax-app-picker-item.js
define(["exports","../../../lit-element/lit-element.js","../../../@polymer/paper-button/paper-button.js","./hax-item-button-inner.js"],function(_exports,_litElement,_paperButton,_haxItemButtonInner){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.HaxAppPickerItem=void 0;class HaxAppPickerItem extends _litElement.LitElement{constructor(){super()}static get styles(){return[_litElement.css`
        :host {
          display: block;
          max-width: 90px;
        }
        paper-button {
          color: black;
          text-transform: none;
          min-width: unset;
          cursor: pointer;
          width: 80px;
          padding: 5px;
          margin: 5px;
          --paper-button-ink-color: var(--hax-ink-color, black);
          -webkit-transition: box-shadow 0.3s;
          -moz-transition: box-shadow 0.3s;
          -ms-transition: box-shadow 0.3s;
          -o-transition: box-shadow 0.3s;
          transition: box-shadow 0.3s;
        }
        paper-button:hover,
        paper-button:focus {
          box-shadow: 0 4px 4px 0 rgba(0, 0, 0, 0.14),
            0 2px 10px 0 rgba(0, 0, 0, 0.12), 0 6px 2px -4px rgba(0, 0, 0, 0.2);
        }
        paper-button:active {
          box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14),
            0 1px 5px 0 rgba(0, 0, 0, 0.12), 0 3px 1px -2px rgba(0, 0, 0, 0.2);
        }
      `]}render(){return _litElement.html`
      <paper-button class="icon">
        <hax-item-button-inner
          color="${this.color}"
          icon="${this.icon}"
          label="${this.label}"
        >
        </hax-item-button-inner>
      </paper-button>
    `}static get tag(){return"hax-app-picker-item"}static get properties(){return{color:{type:String},icon:{type:String},label:{type:String}}}}_exports.HaxAppPickerItem=HaxAppPickerItem,window.customElements.define(HaxAppPickerItem.tag,HaxAppPickerItem)});
