// @lrnwebcomponents/hax-body/lib/hax-upload-field.js
define(["exports","require","../../../lit-element/lit-element.js","../../simple-colors/simple-colors.js","../../utils/utils.js","../../simple-picker/simple-picker.js"],(function(_exports,_require,_litElement,_simpleColors,_utils,_simplePicker){Object.defineProperty(_exports,"__esModule",{value:!0}),_exports.HaxUploadField=void 0,_require=babelHelpers.interopRequireWildcard(_require);class HaxUploadField extends((0,_utils.winEventsElement)(_simpleColors.SimpleColors)){static get styles(){return[...super.styles,_litElement.css`
        :host {
          display: block;
          visibility: visible;
          transition: 0.3s all ease;
          box-sizing: border-box;
          pointer-events: all;
          overflow: visible;
          --simple-camera-snap-width: 300px;
          --simple-camera-snap-height: calc(300px * 9 / 16);
          --simple-camera-snap-color: var(--eco-json-form-color, #222);
          --simple-camera-snap-background: var(--eco-json-form-bg, white);
          --simple-camera-snap-border-radius: 2px;
          --lumo-font-family: var(
            --eco-json-form-font-family,
            var(--paper-font-caption_-_font-family, unset)
          );
          --lumo-base-color: var(
            --eco-json-form-bg,
            var(--primary-background-color, #fff)
          );
          --lumo-primary-contrast-color: var(
            --eco-json-form-bg,
            var(--primary-background-color, #fff)
          );
          --lumo-primary-color: var(
            --eco-json-form-active-color,
            var(--primary-color, #000)
          );
          --lumo-primary-text-color: var(
            --eco-json-form-color,
            var(--primary-text-color, #222)
          );
          --lumo-body-text-color: var(
            --eco-json-form-color,
            var(--primary-text-color, #222)
          );
          --lumo-header-text-color: var(
            --eco-json-form-color,
            var(--primary-text-color, #222)
          );
          --lumo-secondary-text-color: var(
            --eco-json-form-faded-color,
            var(--secondary-text-color, #888)
          );
          --lumo-disabled-text-color: var(
            --eco-json-form-faded-color,
            var(--secondary-text-color, #888)
          );
          background-color: var(
            --eco-json-form-bg,
            var(--primary-background-color, #fff)
          );
        }
        :host #legend {
          transition: all 0.5s;
          color: var(
            --eco-json-form-faded-color,
            var(--secondary-text-color, #888)
          );
        }
        :host(:focus-within) #legend {
          color: var(--eco-json-form-active-color, var(--primary-color, #000));
        }
        :host #fieldset {
          border-radius: 2px;
          transition: all 0.5s;
          border: 1px solid
            var(--eco-json-form-faded-color, var(--secondary-text-color, #888));
        }
        :host #fieldset > div {
          display: flex;
          align-items: center;
          justify-content: space-between;
        }
        :host #fieldset > div > *:not(#picker) {
          flex: 1 1 auto;
        }
        #picker {
          margin-bottom: 0;
          margin-right: 5px;
        }
        vaadin-upload {
          padding: 0;
          margin: 0;
        }
        simple-camera-snap {
          position: relative;
          --simple-camera-snap-button-container-position: absolute;
          --simple-camera-snap-button-container-bottom: 2px;
          --simple-camera-snap-button-container-z-index: 5;
          --simple-camera-snap-button-border-radius: 100%;
          --simple-camera-snap-button-opacity: 0.7;
        }
      `]}constructor(){super(),this.__winEvents={"hax-app-picker-selection":"_haxAppPickerSelection"},this.label=null,this.noCamera=!1,this.noVoiceRecord=!0,new Promise((res,rej)=>_require.default(["../../../@polymer/paper-input/paper-input.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../../@polymer/paper-icon-button/paper-icon-button.js"],res,rej)),new Promise((res,rej)=>_require.default(["../../../@vaadin/vaadin-upload/vaadin-upload.js"],res,rej))}render(){return _litElement.html`
      <fieldset id="fieldset">
        <legend id="legend" ?hidden="${!this.label}">${this.label}</legend>
        <div>
          <simple-picker
            id="picker"
            aria-label="Source..."
            required
            value="${this.option}"
            @value-changed="${this.optionChanged}"
            .options="${this.options}"
          >
          </simple-picker>
          <paper-input
            id="url"
            ?hidden="${"url"!==this.option}"
            value="${this.value}"
            @value-changed="${this.valueChanged}"
            label="URL"
            type="url"
            auto-validate=""
          ></paper-input>
          <vaadin-upload
            capture
            form-data-name="file-upload"
            ?hidden="${"fileupload"!==this.option}"
            id="fileupload"
            @upload-before="${this._fileAboutToUpload}"
            @upload-response="${this._fileUploadResponse}"
          ></vaadin-upload>
          <div id="camerahole" ?hidden="${"selfie"!==this.option}"></div>
          <div id="voicerecorder" ?hidden="${"audio"!==this.option}"></div>
        </div>
      </fieldset>
    `}optionChanged(e){this.option=e.detail.value}valueChanged(e){this.value=e.detail.value}updated(changedProperties){super.updated&&super.updated(changedProperties),changedProperties.forEach((oldValue,propName)=>{"value"==propName&&this.dispatchEvent(new CustomEvent("value-changed",{detail:{value:this[propName]}}))})}static get properties(){return{label:{type:String},value:{type:String},option:{type:String},options:{type:Array},noCamera:{type:Boolean,attribute:"no-camera"},noVoiceRecord:{type:Boolean,attribute:"no-voice-record"}}}_fileAboutToUpload(e){if(this.__allowUpload)this.__allowUpload=!1;else{e.preventDefault(),e.stopPropagation();let values={source:e.detail.file.name.toLowerCase(),type:e.detail.file.type};var type=window.HaxStore.guessGizmoType(values);let targets=window.HaxStore.getHaxAppStoreTargets(type);1===targets.length?this._haxAppPickerSelection({detail:targets[0]}):0!==targets.length?window.HaxStore.instance.haxAppPicker.presentOptions(targets,type,"Where would you like to upload this "+type+"?","app"):window.HaxStore.toast("Sorry, you don't have a storage location that can handle "+type+" uploads!",5e3)}}_fileUploadResponse(e){let response=JSON.parse(e.detail.xhr.response),map=this.__appUsed.connection.operations.add.resultMap,data={},item={};for(var prop in void 0!==this._resolveObjectPath(map.item,response)&&(data=this._resolveObjectPath(map.item,response)),item.type=map.defaultGizmoType,map.gizmo)item[prop]=this._resolveObjectPath(map.gizmo[prop],data);void 0===item.url&&void 0!==item.source&&(item.url=item.source),void 0!==map.gizmo.type&&(item.type=this._resolveObjectPath(map.gizmo.type,data)),this.shadowRoot.querySelector("#url").value=item.url}_haxAppPickerSelection(e){let connection=e.detail.connection;this.__appUsed=e.detail,this.shadowRoot.querySelector("#fileupload").method=connection.operations.add.method;let requestEndPoint=connection.protocol+"://"+connection.url;"/"!=requestEndPoint.substr(requestEndPoint.length-1)&&(requestEndPoint+="/"),void 0!==connection.operations.add.endPoint&&(requestEndPoint+=connection.operations.add.endPoint),null!=window.HaxStore.instance.connectionRewrites.appendUploadEndPoint&&(requestEndPoint+="?"+window.HaxStore.instance.connectionRewrites.appendUploadEndPoint),null!=window.HaxStore.instance.connectionRewrites.appendJwt&&(requestEndPoint+="&"+window.HaxStore.instance.connectionRewrites.appendJwt+"="+localStorage.getItem(window.HaxStore.instance.connectionRewrites.appendJwt)),this.shadowRoot.querySelector("#fileupload").headers=connection.headers,this.shadowRoot.querySelector("#fileupload").target=requestEndPoint,this.__allowUpload=!0,this.shadowRoot.querySelector("#fileupload").uploadFiles()}_setInputOptions(){let options=[[{alt:"URL",icon:"icons:link",value:"url"}],[{alt:"Upload",icon:"icons:file-upload",value:"fileupload"}]];return!navigator.mediaDevices||this.noCamera?this.shadowRoot.querySelector("#camerahole").style.display="none":options.push([{alt:"Camera",icon:"image:photo-camera",value:"selfie"}]),navigator.mediaDevices&&!this.noVoiceRecord||(this.shadowRoot.querySelector("#voicerecorder").style.display="none"),options}firstUpdated(changedProperties){super.firstUpdated&&super.firstUpdated(changedProperties),this.options=[...this._setInputOptions()],this.value?this.option="url":this.option="fileupload",this.shadowRoot.querySelector("#picker").addEventListener("change",e=>{e&&e.detail&&"selfie"===e.detail.value&&this._takeSelfie(e),e&&e.detail&&"audio"===e.detail.value&&this._voiceRecorder(e)})}__newPhotoShowedUp(e){let file=new File([e.detail.raw],"headshot"+e.timeStamp+".jpg");this.shadowRoot.querySelector("#fileupload")._addFile(file)}__newAudioShowedUp(e){let file=new File([e.detail.value],"voice-memo"+e.timeStamp+".mp3");this.shadowRoot.querySelector("#fileupload")._addFile(file)}_takeSelfie(e){this.camera||(new Promise((res,rej)=>_require.default(["../../simple-login/lib/simple-camera-snap.js"],res,rej)),this.camera=document.createElement("simple-camera-snap"),this.camera.autoplay=!0,this.camera.addEventListener("simple-camera-snap-image",this.__newPhotoShowedUp.bind(this)),this.shadowRoot.querySelector("#camerahole").appendChild(this.camera))}_voiceRecorder(e){this.voice||(new Promise((res,rej)=>_require.default(["../../voice-recorder/voice-recorder.js"],res,rej)),this.voice=document.createElement("voice-recorder"),this.voice.addEventListener("voice-recorder-recording",this.__newAudioShowedUp.bind(this)),this.shadowRoot.querySelector("#voicerecorder").appendChild(this.voice))}_resolveObjectPath(path,obj){return path.split(".").reduce((function(prev,curr){return prev?prev[curr]:null}),obj||self)}}_exports.HaxUploadField=HaxUploadField,window.customElements.define("hax-upload-field",HaxUploadField)}));
