// @lrnwebcomponents/map-element/map-element.js
define(["exports", "../../@polymer/polymer/polymer-element.js", "../../@polymer/polymer/lib/utils/render-status.js", "../hax-body-behaviors/lib/HAXWiring.js", "../schema-behaviors/schema-behaviors.js", "../web-map/web-map.js"], function (_exports, _polymerElement, _renderStatus, _HAXWiring, _schemaBehaviors, _webMap) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.MapElement = void 0;

  function _templateObject_fde64380b48f11e9ae0d876050e9d419() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: block;\n        }\n      </style>\n      <map\n        is=\"web-map\"\n        zoom=\"17\"\n        lat=\"45.398043\"\n        lon=\"-75.70683\"\n        width=\"700\"\n        height=\"400\"\n        controls=\"\"\n      >\n        <layer-\n          id=\"osm\"\n          src=\"https://geogratis.gc.ca/mapml/en/osmtile/osm/\"\n          label=\"Open Street Map\"\n          checked=\"\"\n          hidden=\"\"\n        ></layer->\n        <layer-\n          id=\"cbmt\"\n          src=\"https://geogratis.gc.ca/mapml/en/osmtile/cbmt/\"\n          label=\"Canada Base Map\"\n        ></layer->\n        <layer-\n          id=\"canvec\"\n          src=\"https://geogratis.gc.ca/api/beta/vectors/canvec/50k/features/\"\n          label=\"CanVec+ 031G\"\n          class=\"transparency\"\n        ></layer->\n        <layer- id=\"marker\" label=\"Marker layer\" src=\"marker.mapml\"></layer->\n        <area\n          is=\"map-area\"\n          id=\"marker2\"\n          href=\"https://example.com/marker/\"\n          alt=\"Marker\"\n          coords=\"265,185\"\n          shape=\"marker\"\n        />\n        <area\n          is=\"map-area\"\n          id=\"line\"\n          href=\"https://example.com/line/\"\n          alt=\"Line\"\n          coords=\"275,275,540,107\"\n          shape=\"line\"\n        />\n        <area\n          is=\"map-area\"\n          id=\"doughnut\"\n          alt=\"Circle\"\n          href=\"https://example.com/circle/\"\n          coords=\"250,250,25\"\n          shape=\"circle\"\n          style=\"fill: white; stroke: aqua; stroke-width: 5px;fill-opacity: 0.0\"\n        />\n        <area\n          is=\"map-area\"\n          id=\"hole\"\n          coords=\"250,250,7\"\n          shape=\"circle\"\n          style=\"fill: blue; stroke: none;fill-opacity: 0.3;\"\n        />\n        <area\n          is=\"map-area\"\n          id=\"rect\"\n          href=\"https://example.com/rectangle/\"\n          alt=\"Rectangle\"\n          coords=\"345,290,415,320\"\n          shape=\"rect\"\n          style=\"fill: greenyellow; stroke: blue; stroke-width: 3px;fill-opacity: 0.4\"\n        />\n        <area\n          is=\"map-area\"\n          id=\"poly\"\n          href=\"https://example.com/polygon/\"\n          alt=\"Polygon\"\n          coords=\"392,116,430,100,441,128,405,145\"\n          shape=\"poly\"\n          style=\"fill: pink; stroke: blue; stroke-width: 3px;fill-opacity: 0.4\"\n        />\n      </map>\n    "]);

    _templateObject_fde64380b48f11e9ae0d876050e9d419 = function _templateObject_fde64380b48f11e9ae0d876050e9d419() {
      return data;
    };

    return data;
  }

  /**
   * `map-element`
   * `A mapping data bound element`
   *
   * @demo demo/index.html
   */
  var MapElement =
  /*#__PURE__*/
  function (_SchemaBehaviors) {
    babelHelpers.inherits(MapElement, _SchemaBehaviors);

    function MapElement() {
      var _this;

      babelHelpers.classCallCheck(this, MapElement);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(MapElement).call(this));
      (0, _renderStatus.afterNextRender)(babelHelpers.assertThisInitialized(_this), function () {
        this.HAXWiring = new _HAXWiring.HAXWiring();
        this.HAXWiring.setup(MapElement.haxProperties, MapElement.tag, this);
      });
      return _this;
    }

    babelHelpers.createClass(MapElement, null, [{
      key: "template",
      get: function get() {
        return (0, _polymerElement.html)(_templateObject_fde64380b48f11e9ae0d876050e9d419());
      }
    }, {
      key: "tag",
      get: function get() {
        return "map-element";
      }
    }, {
      key: "properties",
      get: function get() {
        var props = {
          /**
           * Title
           */
          title: {
            type: String
          }
        };

        if (babelHelpers.get(babelHelpers.getPrototypeOf(MapElement), "properties", this)) {
          props = Object.assign(props, babelHelpers.get(babelHelpers.getPrototypeOf(MapElement), "properties", this));
        }

        return props;
      }
    }, {
      key: "haxProperties",
      get: function get() {
        return {
          canScale: true,
          canPosition: true,
          canEditSource: false,
          gizmo: {
            title: "Sample gizmo",
            description: "The user will be able to see this for selection in a UI.",
            icon: "av:play-circle-filled",
            color: "grey",
            groups: ["Video", "Media"],
            handles: [{
              type: "video",
              url: "source"
            }],
            meta: {
              author: "Your organization on github"
            }
          },
          settings: {
            quick: [{
              property: "title",
              title: "Title",
              description: "The title of the element",
              inputMethod: "textfield",
              icon: "editor:title"
            }],
            configure: [{
              property: "title",
              title: "Title",
              description: "The title of the element",
              inputMethod: "textfield",
              icon: "editor:title"
            }],
            advanced: []
          }
        };
      }
    }]);
    return MapElement;
  }((0, _schemaBehaviors.SchemaBehaviors)(_polymerElement.PolymerElement));

  _exports.MapElement = MapElement;
  window.customElements.define(MapElement.tag, MapElement);
});
