// @lit/reactive-element/decorators/base.js
define(["exports"], function (_exports) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.standardPrototypeMethod = _exports.legacyPrototypeMethod = _exports.decorateProperty = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  /**
   * @license
   * Copyright 2017 Google LLC
   * SPDX-License-Identifier: BSD-3-Clause
   */
  var e = function e(_e, t, o) {
    Object.defineProperty(t, o, _e);
  },
      t = function t(e, _t) {
    return {
      kind: "method",
      placement: "prototype",
      key: _t.key,
      descriptor: e
    };
  },
      o = function o(_ref) {
    var e = _ref.finisher,
        t = _ref.descriptor;
    return function (o, n) {
      var r;

      if (void 0 === n) {
        var _n = null !== (r = o.originalKey) && void 0 !== r ? r : o.key,
            i = null != t ? {
          kind: "method",
          placement: "prototype",
          key: _n,
          descriptor: t(o.key)
        } : _objectSpread(_objectSpread({}, o), {}, {
          key: _n
        });

        return null != e && (i.finisher = function (t) {
          e(t, _n);
        }), i;
      }

      {
        var _r = o.constructor;
        void 0 !== t && Object.defineProperty(o, n, t(n)), null == e || e(_r, n);
      }
    };
  };

  _exports.decorateProperty = o;
  _exports.standardPrototypeMethod = t;
  _exports.legacyPrototypeMethod = e;
});
