// @lrnwebcomponents/lrs-elements/lib/lrs-service/src/commands/deploy.js
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var _require = require("@oclif/command"),
    Command = _require.Command,
    flags = _require.flags;

var _require2 = require("child_process"),
    spawn = _require2.spawn;

var path = require("path");

var DeployCommand = /*#__PURE__*/function (_Command) {
  "use strict";

  babelHelpers.inherits(DeployCommand, _Command);

  var _super = _createSuper(DeployCommand);

  function DeployCommand() {
    babelHelpers.classCallCheck(this, DeployCommand);
    return _super.apply(this, arguments);
  }

  babelHelpers.createClass(DeployCommand, [{
    key: "run",
    value: function () {
      var _run = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
        var _this$parse, flags;

        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _this$parse = this.parse(DeployCommand), flags = _this$parse.flags;
                spawn("prisma", ["deploy"], {
                  cwd: path.join(__dirname, "../../prisma"),
                  stdio: "inherit",
                  shell: true
                });

              case 2:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function run() {
        return _run.apply(this, arguments);
      }

      return run;
    }()
  }]);
  return DeployCommand;
}(Command);

DeployCommand.description = "Deploy Prisma configuration to active Prisma database.";
DeployCommand.flags = {
  name: flags.string({
    char: "n",
    description: "name to print"
  })
};
module.exports = DeployCommand;
