// @lrnwebcomponents/lrs-elements/lib/lrs-service/src/commands/start.js
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var path = require("path");

var _require = require("@oclif/command"),
    Command = _require.Command,
    flags = _require.flags;

var _require2 = require("child_process"),
    spawn = _require2.spawn;

var startServicePath = require.resolve("../../gateway");

var StartCommand = /*#__PURE__*/function (_Command) {
  "use strict";

  babelHelpers.inherits(StartCommand, _Command);

  var _super = _createSuper(StartCommand);

  function StartCommand() {
    babelHelpers.classCallCheck(this, StartCommand);
    return _super.apply(this, arguments);
  }

  babelHelpers.createClass(StartCommand, [{
    key: "run",
    value: function () {
      var _run = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
        var _this$parse, flags, prismaDockerStart;

        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _this$parse = this.parse(StartCommand), flags = _this$parse.flags;
                prismaDockerStart = spawn("docker-compose", ["up", "-d"], {
                  cwd: path.join(__dirname, "../../prisma"),
                  stdio: "inherit",
                  shell: true
                });
                prismaDockerStart.on("exit", function (code) {
                  spawn("node", [startServicePath], {
                    stdio: "inherit",
                    shell: true
                  });
                });

              case 3:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function run() {
        return _run.apply(this, arguments);
      }

      return run;
    }()
  }]);
  return StartCommand;
}(Command);

StartCommand.description = "Start Prisma Server and Graphql Gateway";
StartCommand.flags = {
  name: flags.string({
    char: "n",
    description: "name to print"
  })
};
module.exports = StartCommand;
