// @lrnwebcomponents/simple-toast/lib/simple-toast-el.js
define(["exports", "../../../lit/index.js", "../../simple-colors/simple-colors.js"], function (_exports, _index, _simpleColors) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SimpleToastEl = void 0;

  var _templateObject_6d4369e0942811ecb1937969798b231a, _templateObject2_6d4369e0942811ecb1937969798b231a;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var SimpleToastEl = /*#__PURE__*/function (_SimpleColors) {
    babelHelpers.inherits(SimpleToastEl, _SimpleColors);

    var _super = _createSuper(SimpleToastEl);

    function SimpleToastEl() {
      var _this;

      babelHelpers.classCallCheck(this, SimpleToastEl);
      _this = _super.call(this);
      _this.text = "";
      _this.duration = 3000;
      _this.opened = false;

      _this.addEventListener("animationend", function (e) {
        if (e.animationName == "fadeout") {
          _this.opened = false;
        }
      });

      return _this;
    }

    babelHelpers.createClass(SimpleToastEl, [{
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        babelHelpers.get(babelHelpers.getPrototypeOf(SimpleToastEl.prototype), "updated", this).call(this, changedProperties);
        changedProperties.forEach(function (oldValue, propName) {
          if (propName === "opened") {
            // notify for others listening
            _this2.dispatchEvent(new CustomEvent("opened-changed", {
              detail: {
                value: _this2[propName]
              }
            }));

            if (_this2[propName]) {
              _this2.show(_this2.text);
            } else {
              _this2.hide();
            }
          }

          if (propName === "duration" && _this2[propName]) {
            _this2.style.animation = "fadein 0.2s, fadeout 0.2s " + _this2[propName] / 1000 + "s";
          }
        });
      }
    }, {
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_6d4369e0942811ecb1937969798b231a || (_templateObject_6d4369e0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["<span><span>", "</span><slot></slot></span>"])), this.text);
      } // To read out loud the toast

    }, {
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        babelHelpers.get(babelHelpers.getPrototypeOf(SimpleToastEl.prototype), "updated", this).call(this, changedProperties);
        this.style.setProperty("aria-live", "assertive");
        this.style.setProperty("role", "alert");
        this.style.setProperty("aria-atomic", "true");
        this.style.setProperty("aria-relevant", "all");
      }
    }, {
      key: "hide",
      value: function hide() {
        this.classList.remove("show");
      }
    }, {
      key: "show",
      value: function show() {
        var text = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "";
        this.text = text;
        this.classList.add("show");
      }
    }], [{
      key: "tag",
      get: function get() {
        return "simple-toast-el";
      }
    }, {
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(SimpleToastEl), "styles", this)), [(0, _index.css)(_templateObject2_6d4369e0942811ecb1937969798b231a || (_templateObject2_6d4369e0942811ecb1937969798b231a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: none;\n          justify-content: center;\n          position: fixed;\n          min-width: 100px;\n          text-align: center;\n          color: var(--simple-colors-default-theme-accent-1, white);\n          background-color: var(--simple-colors-default-theme-accent-12, black);\n          border-radius: var(--simple-toast-border-radius, 2px);\n          border: var(--simple-toast-border, none);\n          font-size: var(--simple-toast-font-size, 1em);\n          font-family: var(--simple-toast-font-family, sans-serif);\n        }\n        :host(.show) {\n          display: flex;\n        }\n        @-webkit-keyframes fadein {\n          from {\n            bottom: 0;\n            opacity: 0;\n          }\n          to {\n            bottom: var(--simple-toast-bottom, 40px);\n            opacity: 1;\n          }\n        }\n        @keyframes fadein {\n          from {\n            bottom: 0;\n            opacity: 0;\n          }\n          to {\n            bottom: var(--simple-toast-bottom, 40px);\n            opacity: 1;\n          }\n        }\n        @-webkit-keyframes fadeout {\n          from {\n            bottom: var(--simple-toast-bottom, 40px);\n            opacity: 1;\n          }\n          to {\n            bottom: 0;\n            opacity: 0;\n          }\n        }\n        @keyframes fadeout {\n          from {\n            bottom: var(--simple-toast-bottom, 40px);\n            opacity: 1;\n          }\n          to {\n            bottom: 0;\n            opacity: 0;\n          }\n        }\n      "])))]);
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(SimpleToastEl), "properties", this)), {}, {
          text: {
            type: String
          },
          duration: {
            type: Number
          },
          opened: {
            type: Boolean
          }
        });
      }
    }]);
    return SimpleToastEl;
  }(_simpleColors.SimpleColors);

  _exports.SimpleToastEl = SimpleToastEl;
  customElements.define(SimpleToastEl.tag, SimpleToastEl);
});
