// @lrnwebcomponents/code-sample/code-sample.js
define(["exports","../../@polymer/polymer/polymer-element.js","../hax-body-behaviors/lib/HAXWiring.js","../../@polymer/polymer/lib/utils/flattened-nodes-observer.js","./lib/themes/one-dark.js","./lib/highlightjs/highlight.js","./lib/highlightjs/languages/javascript.js","./lib/highlightjs/languages/xml.js"],function(_exports,_polymerElement,_HAXWiring,_flattenedNodesObserver,_oneDark,_highlight,_javascript,_xml){"use strict";Object.defineProperty(_exports,"__esModule",{value:!0});_exports.CodeSample=void 0;/**
 * @license MIT, see License.md for full text.
 */_highlight.hljs.registerLanguage("javascript",_javascript.javascript);_highlight.hljs.registerLanguage("xml",_xml.xml);window.hljs=_highlight.hljs;/**
 * `code-sample`
 * `<code-sample>` uses [highlight.js](https://highlightjs.org/) for syntax highlighting.
 * @polymer
 * @customElement
 * @extends {PolymerElement}
 * @demo demo/index.html
 */class CodeSample extends _polymerElement.PolymerElement{// render function
static get template(){return _polymerElement.html`
      <style>
        :host {
          display: block;
        }

        :host([hidden]),
        [hidden] {
          display: none;
        }

        pre {
          margin: 0;
          @apply --code-sample-pre;
        }

        pre,
        code {
          font-family: var(
            --code-sample-font-family,
            Operator Mono,
            Inconsolata,
            Roboto Mono,
            monaco,
            consolas,
            monospace
          );
          font-size: var(--code-sample-font-size, 0.875rem);
        }

        .hljs {
          padding: 0 1.25rem;
          line-height: var(--code-sample-line-height, 1.3);
          @apply --code-sample-hljs;
        }

        .demo:not(:empty) {
          padding: var(--code-sample-demo-padding, 0 0 1.25rem);
          @apply --code-sample-demo-not-empty;
        }

        .demo {
          @apply --code-sample-demo;
        }

        #code-container {
          position: relative;
          @apply --code-sample-code-container;
        }

        #code-container:hover {
          @apply --code-sample-code-container-hover;
        }

        #code-container:hover > button {
          @apply --code-sample-code-container-hover-button;
        }

        button {
          background: var(--code-sample-copy-button-bg-color, #e0e0e0);
          border: none;
          cursor: pointer;
          display: block;
          position: absolute;
          right: 0;
          top: 0;
          text-transform: uppercase;
          @apply --code-sample-copy-clipboard-button;
        }
      </style>
      <div id="theme">${this.constructor.theme||_oneDark.oneDark}</div>
      <div id="demo" class="demo"></div>
      <slot id="content"></slot>

      <div id="code-container">
        <button
          type="button"
          hidden="[[!copyClipboardButton]]"
          id="copyButton"
          title="Copy to clipboard"
          on-click="_copyToClipboard"
        >
          Copy
        </button>
        <pre id="code"></pre>
      </div>
    `}// haxProperty definition
static get haxProperties(){return{canScale:!0,canPosition:!0,canEditSource:!1,gizmo:{title:"Code sample",description:"A sample of code highlighted in the page",icon:"icons:code",color:"blue",groups:["Code","Development"],meta:{author:"kuscamara"}},settings:{quick:[],configure:[{slot:"",slotWrapper:"template",slotAttributes:{"preserve-content":"preserve-content"},title:"Source",description:"The URL for this video.",inputMethod:"code-editor"},{attribute:"copy-clipboard-button",title:"Copy to clipboard button",description:"button in top right that says copy to clipboard",inputMethod:"boolean"}],advanced:[]}}}// properties available to the custom element for data binding
static get properties(){return{// Set to true to show a copy to clipboard button.
copyClipboardButton:{type:"Boolean",value:!1},// Tagged template literal with custom styles.
// Only supported in Shadow DOM.
theme:{type:"String",observer:"_themeChanged"},// Set to true to render the code inside the template.
render:{type:"Boolean",value:!1},// Code type (optional). (eg.: html, js, css)
// Options are the same as the available classes for `<code>` tag using highlight.js
type:{type:"String"}}}/**
   * Store the tag name to make it easier to obtain directly.
   * @notice function name must be here for tooling to operate correctly
   */static get tag(){return"code-sample"}/**
   * life cycle, element is afixed to the DOM
   */connectedCallback(){super.connectedCallback();this.HAXWiring=new _HAXWiring.HAXWiring;this.HAXWiring.setup(CodeSample.haxProperties,CodeSample.tag,this);setTimeout(()=>{if(this.querySelector("template")){this._observer=new _flattenedNodesObserver.FlattenedNodesObserver(this.$.content,()=>this._updateContent())}else if(this.childNodes.length){console.error("<code-sample>:","content must be provided inside a <template> tag")}})}/**
   * life cycle, element is removed from the DOM
   */disconnectedCallback(){super.disconnectedCallback();if(this._observer){this._observer.disconnect();this._observer=null}}_themeChanged(theme){if(theme&&this._themeCanBeChanged()){const previousTheme=this.$.theme.querySelector("style");this.$.theme.replaceChild(document.importNode(theme.content,!0),previousTheme)}}_themeCanBeChanged(){if(window.ShadyCSS){console.error("<code-sample>:","Theme changing is not supported in Shady DOM.");return}if("TEMPLATE"!==this.theme.tagName){console.error("<code-sample>:","theme must be a template");return}return!0}_updateContent(){if(this._code)this._code.parentNode.removeChild(this._code);if(this._demo)this.$.demo.innerHTML="";const template=this._getCodeTemplate();if(this.render){this._demo=this.$.demo.appendChild(document.importNode(template.content,!0))}this._highlight(template.innerHTML)}_getCodeTemplate(){const nodes=_flattenedNodesObserver.FlattenedNodesObserver.getFlattenedNodes(this.$.content);return[].filter.call(nodes,node=>node.nodeType===Node.ELEMENT_NODE)[0]}_highlight(str){this._code=document.createElement("code");if(this.type)this._code.classList.add(this.type);this._code.innerHTML=this._entitize(this._cleanIndentation(str));this.$.code.appendChild(this._code);_highlight.hljs.highlightBlock(this._code)}_cleanIndentation(str){const pattern=str.match(/\s*\n[\t\s]*/);return str.replace(new RegExp(pattern,"g"),"\n")}_entitize(str){return(str+"").replace(/&lt;/g,"<").replace(/&gt;/g,">").replace(/=""/g,"").replace(/=&gt;/g,"=>").replace(/</g,"&lt;").replace(/>/g,"&gt;").replace(/"/g,"&quot;")}_copyToClipboard(event){const copyButton=event.target,textarea=this._textAreaWithClonedContent();textarea.select();try{document.execCommand("copy",!1);copyButton.textContent="Done"}catch(err){console.error(err);copyButton.textContent="Error"}textarea.remove();setTimeout(()=>{copyButton.textContent="Copy"},1e3)}_textAreaWithClonedContent(){const textarea=document.createElement("textarea");document.body.appendChild(textarea);textarea.value=this._cleanIndentation(this._getCodeTemplate().innerHTML);return textarea}}_exports.CodeSample=CodeSample;window.customElements.define(CodeSample.tag,CodeSample)});
